

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Phylo.Applications package &mdash; Biopython 1.74 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Phylo.PAML package" href="Bio.Phylo.PAML.html" />
    <link rel="prev" title="Bio.Phylo package" href="Bio.Phylo.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Phylo.html#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html#module-Bio.Phylo">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.Phylo.html">Bio.Phylo package</a> &raquo;</li>
        
      <li>Bio.Phylo.Applications package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Phylo/Applications/__init__.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Phylo.PAML.html" class="btn btn-neutral float-right" title="Bio.Phylo.PAML package" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Phylo.html" class="btn btn-neutral float-left" title="Bio.Phylo package" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="bio-phylo-applications-package">
<h1>Bio.Phylo.Applications package<a class="headerlink" href="#bio-phylo-applications-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-Bio.Phylo.Applications">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.Phylo.Applications" title="Permalink to this headline">¶</a></h2>
<p>Phylogenetics command line tool wrappers.</p>
<dl class="py class">
<dt id="Bio.Phylo.Applications.PhymlCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.Applications.</code><code class="sig-name descname">PhymlCommandline</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cmd</span><span class="o">=</span><span class="default_value">'phyml'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command-line wrapper for the tree inference program PhyML.</p>
<p>Homepage: <a class="reference external" href="http://www.atgc-montpellier.fr/phyml">http://www.atgc-montpellier.fr/phyml</a></p>
<p class="rubric">References</p>
<p>Guindon S, Gascuel O.
A simple, fast, and accurate algorithm to estimate large phylogenies by maximum
likelihood.
Systematic Biology, 2003 Oct;52(5):696-704.
PubMed PMID: 14530136.</p>
<p>Guindon S, Dufayard JF, Lefort V, Anisimova M, Hordijk W, Gascuel O.
New Algorithms and Methods to Estimate Maximum-Likelihood Phylogenies: Assessing
the Performance of PhyML 3.0.
Systematic Biology, 2010 59(3):307-21.</p>
<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">cmd</span><span class="o">=</span><span class="default_value">'phyml'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.alpha">
<em class="property">property </em><code class="sig-name descname">alpha</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.alpha" title="Permalink to this definition">¶</a></dt>
<dd><p>Distribution of the gamma distribution shape parameter.</p>
<blockquote>
<div><p>Can be a fixed positive value, or ‘e’ to get the
maximum-likelihood estimate.</p>
</div></blockquote>
<p>This controls the addition of the -a parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.bootstrap">
<em class="property">property </em><code class="sig-name descname">bootstrap</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.bootstrap" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of bootstrap replicates, if value is &gt; 0.</p>
<blockquote>
<div><p>Otherwise:</p>
<dl class="simple">
<dt>0: neither approximate likelihood ratio test nor bootstrap</dt><dd><p>values are computed.</p>
</dd>
</dl>
<p>-1: approximate likelihood ratio test returning aLRT statistics.
-2: approximate likelihood ratio test returning Chi2-based</p>
<blockquote>
<div><p>parametric branch supports.</p>
</div></blockquote>
<p>-4: SH-like branch supports alone.</p>
</div></blockquote>
<p>This controls the addition of the -b parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.datatype">
<em class="property">property </em><code class="sig-name descname">datatype</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.datatype" title="Permalink to this definition">¶</a></dt>
<dd><p>Datatype ‘nt’ for nucleotide (default) or ‘aa’ for amino-acids.</p>
<p>This controls the addition of the -d parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.frequencies">
<em class="property">property </em><code class="sig-name descname">frequencies</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.frequencies" title="Permalink to this definition">¶</a></dt>
<dd><p>Character frequencies.</p>
<blockquote>
<div><p>-f e, m, or “fA fC fG fT”</p>
<p>e : Empirical frequencies, determined as follows :</p>
<blockquote>
<div><ul class="simple">
<li><p>Nucleotide sequences: (Empirical) the equilibrium base
frequencies are estimated by counting the occurrence
of the different bases in the alignment.</p></li>
<li><p>Amino-acid sequences: (Empirical) the equilibrium
amino-acid frequencies are estimated by counting the
occurrence of the different amino-acids in the alignment.</p></li>
</ul>
</div></blockquote>
<p>m : ML/model-based frequencies, determined as follows :</p>
<blockquote>
<div><ul class="simple">
<li><p>Nucleotide sequences: (ML) the equilibrium base
frequencies are estimated using maximum likelihood</p></li>
<li><p>Amino-acid sequences: (Model) the equilibrium amino-acid
frequencies are estimated using the frequencies defined by
the substitution model.</p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>“fA fC fG fT”<span class="classifier">only valid for nucleotide-based models.</span></dt><dd><p>fA, fC, fG and fT are floating-point numbers that correspond
to the frequencies of A, C, G and T, respectively.</p>
</dd>
</dl>
</div></blockquote>
<p>This controls the addition of the -f parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.input">
<em class="property">property </em><code class="sig-name descname">input</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.input" title="Permalink to this definition">¶</a></dt>
<dd><p>PHYLIP format input nucleotide or amino-acid sequence filenam.</p>
<p>This controls the addition of the -i parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.input_tree">
<em class="property">property </em><code class="sig-name descname">input_tree</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.input_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Starting tree filename. The tree must be in Newick format.</p>
<p>This controls the addition of the -u parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.model">
<em class="property">property </em><code class="sig-name descname">model</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.model" title="Permalink to this definition">¶</a></dt>
<dd><p>Substitution model name.</p>
<blockquote>
<div><p>Nucleotide-based models:</p>
<p>HKY85 (default) | JC69 | K80 | F81 | F84 | TN93 | GTR | custom</p>
<p>For the custom option, a string of six digits identifies the
model. For instance, 000000 corresponds to F81 (or JC69,
provided the distribution of nucleotide frequencies is uniform).
012345 corresponds to GTR. This option can be used for encoding
any model that is a nested within GTR.</p>
<p>Amino-acid based models:</p>
<p>LG (default) | WAG | JTT | MtREV | Dayhoff | DCMut | RtREV |
CpREV | VT | Blosum62 | MtMam | MtArt | HIVw | HIVb | custom</p>
</div></blockquote>
<p>This controls the addition of the -m parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.multiple">
<em class="property">property </em><code class="sig-name descname">multiple</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.multiple" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of data sets to analyse (integer).</p>
<p>This controls the addition of the -n parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.n_rand_starts">
<em class="property">property </em><code class="sig-name descname">n_rand_starts</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.n_rand_starts" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of initial random trees to be used.</p>
<blockquote>
<div><p>Only valid if SPR searches are to be performed.</p>
</div></blockquote>
<p>This controls the addition of the –n_rand_starts parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.nclasses">
<em class="property">property </em><code class="sig-name descname">nclasses</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.nclasses" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of relative substitution rate categories.</p>
<blockquote>
<div><p>Default 1. Must be a positive integer.</p>
</div></blockquote>
<p>This controls the addition of the -c parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.optimize">
<em class="property">property </em><code class="sig-name descname">optimize</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.optimize" title="Permalink to this definition">¶</a></dt>
<dd><p>Specific parameter optimisation.</p>
<blockquote>
<div><dl class="simple">
<dt>tlr<span class="classifier">tree topology (t), branch length (l) and</span></dt><dd><p>rate parameters (r) are optimised.</p>
</dd>
</dl>
<p>tl  : tree topology and branch length are optimised.
lr  : branch length and rate parameters are optimised.
l   : branch length are optimised.
r   : rate parameters are optimised.
n   : no parameter is optimised.</p>
</div></blockquote>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.pars">
<em class="property">property </em><code class="sig-name descname">pars</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.pars" title="Permalink to this definition">¶</a></dt>
<dd><p>Use a minimum parsimony starting tree.</p>
<blockquote>
<div><p>This option is taken into account when the ‘-u’ option is absent
and when tree topology modifications are to be done.</p>
</div></blockquote>
<p>This property controls the addition of the -p switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.print_site_lnl">
<em class="property">property </em><code class="sig-name descname">print_site_lnl</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.print_site_lnl" title="Permalink to this definition">¶</a></dt>
<dd><p>Print the likelihood for each site in file <a href="#id1"><span class="problematic" id="id2">*</span></a>_phyml_lk.txt.</p>
<p>This property controls the addition of the –print_site_lnl switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.print_trace">
<em class="property">property </em><code class="sig-name descname">print_trace</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.print_trace" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Print each phylogeny explored during the tree search process</dt><dd><p>in file <a href="#id3"><span class="problematic" id="id4">*</span></a>_phyml_trace.txt.</p>
</dd>
</dl>
<p>This property controls the addition of the –print_trace switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.prop_invar">
<em class="property">property </em><code class="sig-name descname">prop_invar</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.prop_invar" title="Permalink to this definition">¶</a></dt>
<dd><p>Proportion of invariable sites.</p>
<blockquote>
<div><p>Can be a fixed value in the range [0,1], or ‘e’ to get the
maximum-likelihood estimate.</p>
</div></blockquote>
<p>This controls the addition of the -v parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.quiet">
<em class="property">property </em><code class="sig-name descname">quiet</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.quiet" title="Permalink to this definition">¶</a></dt>
<dd><p>No interactive questions (for running in batch mode).</p>
<p>This property controls the addition of the –quiet switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.r_seed">
<em class="property">property </em><code class="sig-name descname">r_seed</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.r_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Seed used to initiate the random number generator.</p>
<blockquote>
<div><p>Must be an integer.</p>
</div></blockquote>
<p>This controls the addition of the –r_seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.rand_start">
<em class="property">property </em><code class="sig-name descname">rand_start</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.rand_start" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the initial tree to random.</p>
<blockquote>
<div><p>Only valid if SPR searches are to be performed.</p>
</div></blockquote>
<p>This property controls the addition of the –rand_start switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.run_id">
<em class="property">property </em><code class="sig-name descname">run_id</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.run_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Append the given string at the end of each PhyML output file.</p>
<blockquote>
<div><p>This option may be useful when running simulations involving
PhyML.</p>
</div></blockquote>
<p>This controls the addition of the –run_id parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.search">
<em class="property">property </em><code class="sig-name descname">search</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.search" title="Permalink to this definition">¶</a></dt>
<dd><p>Tree topology search operation option.</p>
<blockquote>
<div><p>Can be one of:</p>
<blockquote>
<div><p>NNI : default, fast
SPR : a bit slower than NNI
BEST : best of NNI and SPR search</p>
</div></blockquote>
</div></blockquote>
<p>This controls the addition of the -s parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.sequential">
<em class="property">property </em><code class="sig-name descname">sequential</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.sequential" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes interleaved format (default) to sequential format.</p>
<p>This property controls the addition of the -q switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.PhymlCommandline.ts_tv_ratio">
<em class="property">property </em><code class="sig-name descname">ts_tv_ratio</code><a class="headerlink" href="#Bio.Phylo.Applications.PhymlCommandline.ts_tv_ratio" title="Permalink to this definition">¶</a></dt>
<dd><p>Transition/transversion ratio. (DNA sequences only.)</p>
<blockquote>
<div><p>Can be a fixed positive value (ex:4.0) or e to get the
maximum-likelihood estimate.</p>
</div></blockquote>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.Applications.RaxmlCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.Applications.</code><code class="sig-name descname">RaxmlCommandline</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cmd</span><span class="o">=</span><span class="default_value">'raxmlHPC'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command-line wrapper for the tree inference program RAxML.</p>
<p>The required parameters are ‘sequences’ (-s), ‘model’ (-m) and ‘name’ (-n).
The parameter ‘parsimony_seed’ (-p) must also be set for RAxML, but if you
do not specify it, this wrapper will set the seed to 10000 for you.</p>
<p class="rubric">References</p>
<p>Stamatakis A.
RAxML-VI-HPC: Maximum Likelihood-based Phylogenetic Analyses with
Thousands of Taxa and Mixed Models.
Bioinformatics 2006, 22(21):2688-2690.</p>
<p>Homepage: <a class="reference external" href="http://sco.h-its.org/exelixis/software.html">http://sco.h-its.org/exelixis/software.html</a></p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Phylo.Applications</span> <span class="kn">import</span> <span class="n">RaxmlCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">raxml_cline</span> <span class="o">=</span> <span class="n">RaxmlCommandline</span><span class="p">(</span><span class="n">sequences</span><span class="o">=</span><span class="s2">&quot;Tests/Phylip/interlaced2.phy&quot;</span><span class="p">,</span>
<span class="gp">... </span>                               <span class="n">model</span><span class="o">=</span><span class="s2">&quot;PROTCATWAG&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;interlaced2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">raxml_cline</span><span class="p">)</span>
<span class="go">raxmlHPC -m PROTCATWAG -n interlaced2 -p 10000 -s Tests/Phylip/interlaced2.phy</span>
</pre></div>
</div>
<p>You would typically run the command line with raxml_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">cmd</span><span class="o">=</span><span class="default_value">'raxmlHPC'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.parsimony_seed">
<em class="property">property </em><code class="sig-name descname">parsimony_seed</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.parsimony_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Random number seed for the parsimony inferences. This allows you to reproduce your results and will help developers debug the program. This option HAS NO EFFECT in the parallel MPI version.</p>
<p>This controls the addition of the -p parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.algorithm">
<em class="property">property </em><code class="sig-name descname">algorithm</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.algorithm" title="Permalink to this definition">¶</a></dt>
<dd><p>Select algorithm:</p>
<blockquote>
<div><dl class="simple">
<dt>a: Rapid Bootstrap analysis and search for best-scoring ML</dt><dd><p>tree in one program run.</p>
</dd>
<dt>b: Draw bipartition information on a tree provided with ‘-t’</dt><dd><p>based on multiple trees (e.g. form a bootstrap) in a file
specifed by ‘-z’.</p>
</dd>
</dl>
<p>c: Check if the alignment can be properly read by RAxML.
d: New rapid hill-climbing (DEFAULT).
e: Optimize model+branch lengths for given input tree under</p>
<blockquote>
<div><p>GAMMA/GAMMAI only.</p>
</div></blockquote>
<dl class="simple">
<dt>g: Compute per site log Likelihoods for one ore more trees</dt><dd><p>passed via ‘-z’ and write them to a file that can be read
by CONSEL.</p>
</dd>
<dt>h: Compute log likelihood test (SH-test) between best tree</dt><dd><p>passed via ‘-t’ and a bunch of other trees passed via ‘-z’.</p>
</dd>
<dt>i: Perform a really thorough bootstrap, refinement of final</dt><dd><p>bootstrap tree under GAMMA and a more exhaustive algorithm.</p>
</dd>
<dt>j: Generate a bunch of bootstrapped alignment files from an</dt><dd><p>original alignemnt file.</p>
</dd>
<dt>m: Compare bipartitions between two bunches of trees passed</dt><dd><p>via ‘-t’ and ‘-z’ respectively. This will return the
Pearson correlation between all bipartitions found in the
two tree files. A file called
RAxML_bipartitionFrequencies.outputFileName will be
printed that contains the pair-wise bipartition
frequencies of the two sets.</p>
</dd>
<dt>n: Compute the log likelihood score of all trees contained</dt><dd><p>in a tree file provided by ‘-z’ under GAMMA or
GAMMA+P-Invar.</p>
</dd>
</dl>
<p>o: Old and slower rapid hill-climbing.
p: Perform pure stepwise MP addition of new sequences to an</p>
<blockquote>
<div><p>incomplete starting tree.</p>
</div></blockquote>
<dl class="simple">
<dt>s: Split up a multi-gene partitioned alignment into the</dt><dd><p>respective subalignments.</p>
</dd>
</dl>
<p>t: Do randomized tree searches on one fixed starting tree.
w: Compute ELW test on a bunch of trees passed via ‘-z’.
x: Compute pair-wise ML distances, ML model parameters will</p>
<blockquote>
<div><p>be estimated on an MP starting tree or a user-defined
tree passed via ‘-t’, only allowed for GAMMA-based models
of rate heterogeneity.</p>
</div></blockquote>
</div></blockquote>
<p>This controls the addition of the -f parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.binary_constraint">
<em class="property">property </em><code class="sig-name descname">binary_constraint</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.binary_constraint" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of a binary constraint tree. This tree does not need to be comprehensive, i.e. contain all taxa.</p>
<p>This controls the addition of the -r parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.bipartition_filename">
<em class="property">property </em><code class="sig-name descname">bipartition_filename</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.bipartition_filename" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of a file containing multiple trees, e.g. from a bootstrap run, that shall be used to draw bipartition values onto a tree provided with ‘-t’. It can also be used to compute per-site log likelihoods in combination with ‘-f g’, and to read a bunch of trees for a couple of other options (‘-f h’, ‘-f m’, ‘-f n’).</p>
<p>This controls the addition of the -z parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.bootstrap_branch_lengths">
<em class="property">property </em><code class="sig-name descname">bootstrap_branch_lengths</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.bootstrap_branch_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Print bootstrapped trees with branch lengths. The bootstraps will run a bit longer, because model parameters will be optimized at the end of each run. Use with CATMIX/PROTMIX or GAMMA/GAMMAI.</p>
<p>This property controls the addition of the -k switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.bootstrap_seed">
<em class="property">property </em><code class="sig-name descname">bootstrap_seed</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.bootstrap_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Random seed for bootstrapping.</p>
<p>This controls the addition of the -b parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.checkpoints">
<em class="property">property </em><code class="sig-name descname">checkpoints</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.checkpoints" title="Permalink to this definition">¶</a></dt>
<dd><p>Write checkpoints (intermediate tree topologies).</p>
<p>This property controls the addition of the -j switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.cluster_threshold">
<em class="property">property </em><code class="sig-name descname">cluster_threshold</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.cluster_threshold" title="Permalink to this definition">¶</a></dt>
<dd><p>Threshold for sequence similarity clustering. RAxML will then print out an alignment to a file called sequenceFileName.reducedBy.threshold that only contains sequences &lt;= the specified threshold that must be between 0.0 and 1.0. RAxML uses the QT-clustering algorithm to perform this task. In addition, a file called RAxML_reducedList.outputFileName will be written that contains clustering information.</p>
<p>This controls the addition of the -l parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.cluster_threshold_fast">
<em class="property">property </em><code class="sig-name descname">cluster_threshold_fast</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.cluster_threshold_fast" title="Permalink to this definition">¶</a></dt>
<dd><p>Same functionality as ‘-l’, but uses a less exhaustive and thus faster clustering algorithm. This is intended for very large datasets with more than 20,000-30,000 sequences.</p>
<p>This controls the addition of the -L parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.epsilon">
<em class="property">property </em><code class="sig-name descname">epsilon</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.epsilon" title="Permalink to this definition">¶</a></dt>
<dd><p>Set model optimization precision in log likelihood units for final optimization of tree topology under MIX/MIXI or GAMMA/GAMMAI.Default: 0.1 for models not using proportion of invariant sites estimate; 0.001 for models using proportion of invariant sites estimate.</p>
<p>This controls the addition of the -e parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.exclude_filename">
<em class="property">property </em><code class="sig-name descname">exclude_filename</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.exclude_filename" title="Permalink to this definition">¶</a></dt>
<dd><p>An exclude file name, containing a specification of alignment positions you wish to exclude.  Format is similar to Nexus, the file shall contain entries like ‘100-200 300-400’; to exclude a single column write, e.g., ‘100-100’. If you use a mixed model, an appropriately adapted model file will be written.</p>
<p>This controls the addition of the -E parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.grouping_constraint">
<em class="property">property </em><code class="sig-name descname">grouping_constraint</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.grouping_constraint" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of a multifurcating constraint tree. this tree does not need to be comprehensive, i.e. contain all taxa.</p>
<p>This controls the addition of the -g parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.model">
<em class="property">property </em><code class="sig-name descname">model</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.model" title="Permalink to this definition">¶</a></dt>
<dd><p>Model of Nucleotide or Amino Acid Substitution:</p>
<blockquote>
<div><p>NUCLEOTIDES:</p>
<dl class="simple">
<dt>GTRCAT<span class="classifier">GTR + Optimization of substitution rates + Optimization of site-specific</span></dt><dd><p>evolutionary rates which are categorized into numberOfCategories distinct
rate categories for greater computational efficiency
if you do a multiple analysis with  ‘-#’ or ‘-N’ but without bootstrapping the program
will use GTRMIX instead</p>
</dd>
<dt>GTRGAMMA<span class="classifier">GTR + Optimization of substitution rates + GAMMA model of rate</span></dt><dd><p>heterogeneity (alpha parameter will be estimated)</p>
</dd>
<dt>GTRMIX<span class="classifier">Inference of the tree under GTRCAT</span></dt><dd><p>and thereafter evaluation of the final tree topology under GTRGAMMA</p>
</dd>
<dt>GTRCAT_GAMMA<span class="classifier">Inference of the tree with site-specific evolutionary rates.</span></dt><dd><p>However, here rates are categorized using the 4 discrete GAMMA rates.
Evaluation of the final tree topology under GTRGAMMA</p>
</dd>
</dl>
<p>GTRGAMMAI      : Same as GTRGAMMA, but with estimate of proportion of invariable sites
GTRMIXI        : Same as GTRMIX, but with estimate of proportion of invariable sites
GTRCAT_GAMMAI  : Same as GTRCAT_GAMMA, but with estimate of proportion of invariable sites</p>
<p>AMINO ACIDS:</p>
<dl class="simple">
<dt>PROTCATmatrixName[F]<span class="classifier">specified AA matrix + Optimization of substitution rates + Optimization of site-specific</span></dt><dd><p>evolutionary rates which are categorized into numberOfCategories distinct
rate categories for greater computational efficiency
if you do a multiple analysis with  ‘-#’ or ‘-N’ but without bootstrapping the program
will use PROTMIX… instead</p>
</dd>
<dt>PROTGAMMAmatrixName[F]<span class="classifier">specified AA matrix + Optimization of substitution rates + GAMMA model of rate</span></dt><dd><p>heterogeneity (alpha parameter will be estimated)</p>
</dd>
<dt>PROTMIXmatrixName[F]<span class="classifier">Inference of the tree under specified AA matrix + CAT</span></dt><dd><p>and thereafter evaluation of the final tree topology under specified AA matrix + GAMMA</p>
</dd>
<dt>PROTCAT_GAMMAmatrixName[F]<span class="classifier">Inference of the tree under specified AA matrix and site-specific evolutionary rates.</span></dt><dd><p>However, here rates are categorized using the 4 discrete GAMMA rates.
Evaluation of the final tree topology under specified AA matrix + GAMMA</p>
</dd>
</dl>
<p>PROTGAMMAImatrixName[F] : Same as PROTGAMMAmatrixName[F], but with estimate of proportion of invariable sites
PROTMIXImatrixName[F]   : Same as PROTMIXmatrixName[F], but with estimate of proportion of invariable sites
PROTCAT_GAMMAImatrixName[F] : Same as PROTCAT_GAMMAmatrixName[F], but with estimate of proportion of invariable sites</p>
<p>Available AA substitution models: DAYHOFF, DCMUT, JTT, MTREV, WAG, RTREV, CPREV, VT, BLOSUM62, MTMAM, GTR
With the optional ‘F’ appendix you can specify if you want to use empirical base frequencies
Please not that for mixed models you can in addition specify the per-gene AA model in
the mixed model file (see manual for details)</p>
</div></blockquote>
<p>This controls the addition of the -m parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.name">
<em class="property">property </em><code class="sig-name descname">name</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name used in the output files.</p>
<p>This controls the addition of the -n parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.num_bootstrap_searches">
<em class="property">property </em><code class="sig-name descname">num_bootstrap_searches</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.num_bootstrap_searches" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of multiple bootstrap searches per replicate. Use this to obtain better ML trees for each replicate. Default: 1 ML search per bootstrap replicate.</p>
<p>This controls the addition of the -u parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.num_categories">
<em class="property">property </em><code class="sig-name descname">num_categories</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.num_categories" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of distinct rate categories for RAxML when evolution model is set to GTRCAT or GTRMIX.Individual per-site rates are categorized into this many rate categories to accelerate computations. Default: 25.</p>
<p>This controls the addition of the -c parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.num_replicates">
<em class="property">property </em><code class="sig-name descname">num_replicates</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.num_replicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of alternative runs on distinct starting trees. In combination with the ‘-b’ option, this will invoke a multiple bootstrap analysis. DEFAULT: 1 single analysis.Note that ‘-N’ has been added as an alternative since ‘-#’ sometimes caused problems with certain MPI job submission systems, since ‘-#’ is often used to start comments.</p>
<p>This controls the addition of the -N parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.outgroup">
<em class="property">property </em><code class="sig-name descname">outgroup</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of a single outgroup or a comma-separated list of outgroups, eg ‘-o Rat’ or ‘-o Rat,Mouse’. In case that multiple outgroups are not monophyletic the first name in the list will be selected as outgroup. Don’t leave spaces between taxon names!</p>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.parsimony">
<em class="property">property </em><code class="sig-name descname">parsimony</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.parsimony" title="Permalink to this definition">¶</a></dt>
<dd><p>Only compute a parsimony starting tree, then exit.</p>
<p>This property controls the addition of the -y switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.partition_branch_lengths">
<em class="property">property </em><code class="sig-name descname">partition_branch_lengths</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.partition_branch_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Switch on estimation of individual per-partition branch lengths. Only has effect when used in combination with ‘partition_filename’ (‘-q’). Branch lengths for individual partitions will be printed to separate files.  A weighted average of the branch lengths is computed by using the respective partition lengths.</p>
<p>This property controls the addition of the -M switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.partition_filename">
<em class="property">property </em><code class="sig-name descname">partition_filename</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.partition_filename" title="Permalink to this definition">¶</a></dt>
<dd><p>File name containing the assignment of models to alignment partitions for multiple models of substitution. For the syntax of this file please consult the RAxML manual.</p>
<p>This controls the addition of the -q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.protein_model">
<em class="property">property </em><code class="sig-name descname">protein_model</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.protein_model" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of a user-defined AA (Protein) substitution model. This file must contain 420 entries, the first 400 being the AA substitution rates (this must be a symmetric matrix) and the last 20 are the empirical base frequencies.</p>
<p>This controls the addition of the -P parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.random_starting_tree">
<em class="property">property </em><code class="sig-name descname">random_starting_tree</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.random_starting_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Start ML optimization from random starting tree.</p>
<p>This property controls the addition of the -d switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.rapid_bootstrap_seed">
<em class="property">property </em><code class="sig-name descname">rapid_bootstrap_seed</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.rapid_bootstrap_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Random seed for rapid bootstrapping.</p>
<p>This controls the addition of the -x parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.rearrangements">
<em class="property">property </em><code class="sig-name descname">rearrangements</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.rearrangements" title="Permalink to this definition">¶</a></dt>
<dd><p>Initial rearrangement setting for the subsequent application of topological changes phase.</p>
<p>This controls the addition of the -i parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.sequences">
<em class="property">property </em><code class="sig-name descname">sequences</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of the alignment data file, in PHYLIP format.</p>
<p>This controls the addition of the -s parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.starting_tree">
<em class="property">property </em><code class="sig-name descname">starting_tree</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.starting_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of a user starting tree, in Newick format.</p>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.threads">
<em class="property">property </em><code class="sig-name descname">threads</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.threads" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of threads to run. PTHREADS VERSION ONLY! Make sure to set this at most the number of CPUs you have on your machine, otherwise, there will be a huge performance decrease!</p>
<p>This controls the addition of the -T parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.version">
<em class="property">property </em><code class="sig-name descname">version</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.version" title="Permalink to this definition">¶</a></dt>
<dd><p>Display version information.</p>
<p>This property controls the addition of the -v switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.weight_filename">
<em class="property">property </em><code class="sig-name descname">weight_filename</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.weight_filename" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of a column weight file to assign individual weights to each column of the alignment. Those weights must be integers separated by any type and number of whitespaces within a separate file.</p>
<p>This controls the addition of the -a parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.RaxmlCommandline.working_dir">
<em class="property">property </em><code class="sig-name descname">working_dir</code><a class="headerlink" href="#Bio.Phylo.Applications.RaxmlCommandline.working_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of the working directory where RAxML will write its output files. Default: current directory.</p>
<p>This controls the addition of the -w parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.Applications.FastTreeCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.Applications.</code><code class="sig-name descname">FastTreeCommandline</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cmd</span><span class="o">=</span><span class="default_value">'fasttree'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command-line wrapper for FastTree.</p>
<p>Only the <code class="docutils literal notranslate"><span class="pre">input</span></code> and <code class="docutils literal notranslate"><span class="pre">out</span></code> parameters are mandatory.</p>
<p>From the terminal command line use <code class="docutils literal notranslate"><span class="pre">fasttree.exe</span> <span class="pre">-help</span></code> or <code class="docutils literal notranslate"><span class="pre">fasttree.exe</span> <span class="pre">-expert</span></code>
for more explanation of usage options.</p>
<p>Homepage: <a class="reference external" href="http://www.microbesonline.org/fasttree/">http://www.microbesonline.org/fasttree/</a></p>
<p class="rubric">References</p>
<p>Price, M.N., Dehal, P.S., and Arkin, A.P. (2010) FastTree 2 – Approximately
Maximum-Likelihood Trees for Large Alignments. PLoS ONE, 5(3):e9490.
<a class="reference external" href="https://doi.org/10.1371/journal.pone.0009490">https://doi.org/10.1371/journal.pone.0009490</a>.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">_Fasttree</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fasttree_exe</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;C:\FasttreeWin32\fasttree.exe&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cmd</span> <span class="o">=</span> <span class="n">_Fasttree</span><span class="o">.</span><span class="n">FastTreeCommandline</span><span class="p">(</span><span class="n">fasttree_exe</span><span class="p">,</span>
<span class="gp">... </span>                                    <span class="nb">input</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;C:\Input\ExampleAlignment.fsa&#39;</span><span class="p">,</span>
<span class="gp">... </span>                                    <span class="n">out</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;C:\Output\ExampleTree.tree&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out</span><span class="p">,</span> <span class="n">err</span> <span class="o">=</span> <span class="n">cmd</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">cmd</span><span class="o">=</span><span class="default_value">'fasttree'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.bionj">
<em class="property">property </em><code class="sig-name descname">bionj</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.bionj" title="Permalink to this definition">¶</a></dt>
<dd><p>Join options: weighted joins as in BIONJ.</p>
<blockquote>
<div><p>FastTree will also weight joins during NNIs.</p>
</div></blockquote>
<p>This property controls the addition of the -bionj switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.boot">
<em class="property">property </em><code class="sig-name descname">boot</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.boot" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify the number of resamples for support values.</p>
<blockquote>
<div><p>Support value options:
By default, FastTree computes local support values by resampling the site
likelihoods 1,000 times and the Shimodaira Hasegawa test. If you specify -nome,
it will compute minimum-evolution bootstrap supports instead
In either case, the support values are proportions ranging from 0 to 1</p>
<p>Use -nosupport to turn off support values or -boot 100 to use just 100 resamples.</p>
</div></blockquote>
<p>This controls the addition of the -boot parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.cat">
<em class="property">property </em><code class="sig-name descname">cat</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.cat" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum likelihood model options.</p>
<blockquote>
<div><p>Specify the number of rate categories of sites (default 20).</p>
</div></blockquote>
<p>This controls the addition of the -cat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.close">
<em class="property">property </em><code class="sig-name descname">close</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Modify the close heuristic for the top-hit list</p>
<blockquote>
<div><p>Top-hit heuristics:
By default, FastTree uses a top-hit list to speed up search
-close 0.75 – modify the close heuristic, lower is more conservative.</p>
</div></blockquote>
<p>This controls the addition of the -close parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.constraintWeight">
<em class="property">property </em><code class="sig-name descname">constraintWeight</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.constraintWeight" title="Permalink to this definition">¶</a></dt>
<dd><p>Weight strength of contraints in topology searching.</p>
<blockquote>
<div><p>Constrained topology search options:
-constraintWeight – how strongly to weight the constraints. A value of 1
means a penalty of 1 in tree length for violating a constraint
Default: 100.0</p>
</div></blockquote>
<p>This controls the addition of the -constraintWeight parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.constraints">
<em class="property">property </em><code class="sig-name descname">constraints</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.constraints" title="Permalink to this definition">¶</a></dt>
<dd><p>Specifies an alignment file for use with constrained topology searching</p>
<blockquote>
<div><p>Constrained topology search options:
-constraints alignmentfile – an alignment with values of 0, 1, and -
Not all sequences need be present. A column of 0s and 1s defines a
constrained split. Some constraints may be violated
(see ‘violating constraints:’ in standard error).</p>
</div></blockquote>
<p>This controls the addition of the -constraints parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.expert">
<em class="property">property </em><code class="sig-name descname">expert</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.expert" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the expert level help.</p>
<p>This property controls the addition of the -expert switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.fastest">
<em class="property">property </em><code class="sig-name descname">fastest</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.fastest" title="Permalink to this definition">¶</a></dt>
<dd><p>Search the visible set (the top hit for each node) only.</p>
<blockquote>
<div><p>Searching for the best join:
By default, FastTree combines the ‘visible set’ of fast neighbor-joining with
local hill-climbing as in relaxed neighbor-joining
-fastest – search the visible set (the top hit for each node) only
Unlike the original fast neighbor-joining, -fastest updates visible(C)
after joining A and B if join(AB,C) is better than join(C,visible(C))
-fastest also updates out-distances in a very lazy way,
-fastest sets -2nd on as well, use -fastest -no2nd to avoid this</p>
</div></blockquote>
<p>This property controls the addition of the -fastest switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.gamma">
<em class="property">property </em><code class="sig-name descname">gamma</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.gamma" title="Permalink to this definition">¶</a></dt>
<dd><p>Report the likelihood under the discrete gamma model.</p>
<blockquote>
<div><p>Maximum likelihood model options:
-gamma – after the final round of optimizing branch lengths with the CAT model,
report the likelihood under the discrete gamma model with the same
number of categories. FastTree uses the same branch lengths but
optimizes the gamma shape parameter and the scale of the lengths.
The final tree will have rescaled lengths. Used with -log, this
also generates per-site likelihoods for use with CONSEL, see
GammaLogToPaup.pl and documentation on the FastTree web site.</p>
</div></blockquote>
<p>This property controls the addition of the -gamma switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.gtr">
<em class="property">property </em><code class="sig-name descname">gtr</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.gtr" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum likelihood model options.</p>
<blockquote>
<div><p>Use generalized time-reversible instead of (default)
Jukes-Cantor (nt only)</p>
</div></blockquote>
<p>This property controls the addition of the -gtr switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.gtrfreq">
<em class="property">property </em><code class="sig-name descname">gtrfreq</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.gtrfreq" title="Permalink to this definition">¶</a></dt>
<dd><p>-gtrfreq A C G T</p>
<p>This controls the addition of the -gtrfreq parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.gtrrates">
<em class="property">property </em><code class="sig-name descname">gtrrates</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.gtrrates" title="Permalink to this definition">¶</a></dt>
<dd><p>-gtrrates ac ag at cg ct gt</p>
<p>This controls the addition of the -gtrrates parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.help">
<em class="property">property </em><code class="sig-name descname">help</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.help" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the help.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.input">
<em class="property">property </em><code class="sig-name descname">input</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.input" title="Permalink to this definition">¶</a></dt>
<dd><p>Enter &lt;input file&gt;</p>
<blockquote>
<div><p>An input file of sequence alignments in fasta or phylip format
is needed.  By default FastTree expects protein
alignments, use -nt for nucleotides.</p>
</div></blockquote>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.intree">
<em class="property">property </em><code class="sig-name descname">intree</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.intree" title="Permalink to this definition">¶</a></dt>
<dd><p>-intree newickfile – read the starting tree in from newickfile.</p>
<blockquote>
<div><p>Any branch lengths in the starting trees are ignored.
-intree with -n will read a separate starting tree for each alignment.</p>
</div></blockquote>
<p>This controls the addition of the -intree parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.intree1">
<em class="property">property </em><code class="sig-name descname">intree1</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.intree1" title="Permalink to this definition">¶</a></dt>
<dd><p>intree1 newickfile – read the same starting tree for each alignment.</p>
<p>This controls the addition of the -intree1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.log">
<em class="property">property </em><code class="sig-name descname">log</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.log" title="Permalink to this definition">¶</a></dt>
<dd><p>Create log files of data such as intermediate trees and per-site rates</p>
<blockquote>
<div><p>-log logfile – save intermediate trees so you can extract
the trees and restart long-running jobs if they crash
-log also reports the per-site rates (1 means slowest category).</p>
</div></blockquote>
<p>This controls the addition of the -log parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.makematrix">
<em class="property">property </em><code class="sig-name descname">makematrix</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.makematrix" title="Permalink to this definition">¶</a></dt>
<dd><p>-makematrix [alignment]</p>
<p>This controls the addition of the -makematrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.matrix">
<em class="property">property </em><code class="sig-name descname">matrix</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify a matrix for nucleotide or amino acid distances</p>
<blockquote>
<div><p>Distances:
Default: For protein sequences, log-corrected distances and an
amino acid dissimilarity matrix derived from BLOSUM45
or for nucleotide sequences, Jukes-Cantor distances
To specify a different matrix, use -matrix FilePrefix or -nomatrix</p>
</div></blockquote>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.mlacc">
<em class="property">property </em><code class="sig-name descname">mlacc</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.mlacc" title="Permalink to this definition">¶</a></dt>
<dd><p>Option for optimization of branches at each NNI.</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs,
and -spr to set the rounds of SPRs.
Use -mlacc 2 or -mlacc 3 to always optimize all 5 branches at each NNI,
and to optimize all 5 branches in 2 or 3 rounds.</p>
</div></blockquote>
<p>This controls the addition of the -mlacc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.mllen">
<em class="property">property </em><code class="sig-name descname">mllen</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.mllen" title="Permalink to this definition">¶</a></dt>
<dd><p>Optimize branch lengths on a fixed topology.</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs,
and -spr to set the rounds of SPRs.
Use -mllen to optimize branch lengths without ML NNIs
Use -mllen -nome with -intree to optimize branch lengths on a fixed topology.</p>
</div></blockquote>
<p>This property controls the addition of the -mllen switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.mlnni">
<em class="property">property </em><code class="sig-name descname">mlnni</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.mlnni" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the number of rounds of maximum-likelihood NNIs.</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs,
and -spr to set the rounds of SPRs.
Use -mlnni to set the number of rounds of maximum-likelihood NNIs.</p>
</div></blockquote>
<p>This controls the addition of the -mlnni parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.n">
<em class="property">property </em><code class="sig-name descname">n</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.n" title="Permalink to this definition">¶</a></dt>
<dd><p>-n – read N multiple alignments in.</p>
<blockquote>
<div><p>This only works with phylip interleaved format. For example, you can
use it with the output from phylip’s seqboot. If you use -n, FastTree
will write 1 tree per line to standard output.</p>
</div></blockquote>
<p>This controls the addition of the -n parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.nj">
<em class="property">property </em><code class="sig-name descname">nj</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.nj" title="Permalink to this definition">¶</a></dt>
<dd><p>Join options: regular (unweighted) neighbor-joining (default)</p>
<p>This property controls the addition of the -nj switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.nni">
<em class="property">property </em><code class="sig-name descname">nni</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.nni" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the rounds of minimum-evolution nearest-neighbor interchanges</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs.</p>
</div></blockquote>
<p>This controls the addition of the -nni parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.no2nd">
<em class="property">property </em><code class="sig-name descname">no2nd</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.no2nd" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn 2nd-level top hits heuristic off.</p>
<blockquote>
<div><p>Top-hit heuristics:
By default, FastTree uses a top-hit list to speed up search
Use -notop (or -slow) to turn this feature off
and compare all leaves to each other,
and all new joined nodes to each other</p>
<p>-2nd or -no2nd to turn 2nd-level top hits heuristic on or off
This reduces memory usage and running time but may lead to
marginal reductions in tree quality.
(By default, -fastest turns on -2nd.)</p>
</div></blockquote>
<p>This property controls the addition of the -no2nd switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.nocat">
<em class="property">property </em><code class="sig-name descname">nocat</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.nocat" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum likelihood model options: No CAT model (just 1 category)</p>
<p>This property controls the addition of the -nocat switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.nomatrix">
<em class="property">property </em><code class="sig-name descname">nomatrix</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.nomatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify that no matrix should be used for nucleotide or amino acid distances</p>
<blockquote>
<div><p>Distances:
Default: For protein sequences, log-corrected distances and an
amino acid dissimilarity matrix derived from BLOSUM45
or for nucleotide sequences, Jukes-Cantor distances
To specify a different matrix, use -matrix FilePrefix or -nomatrix</p>
</div></blockquote>
<p>This property controls the addition of the -nomatrix switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.nome">
<em class="property">property </em><code class="sig-name descname">nome</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.nome" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes support values calculation to a minimum-evolution bootstrap method.</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs,
and -spr to set the rounds of SPRs.
Use -mllen to optimize branch lengths without ML NNIs
Use -mllen -nome with -intree to optimize branch lengths on a fixed topology</p>
<p>Support value options:
By default, FastTree computes local support values by resampling the site
likelihoods 1,000 times and the Shimodaira Hasegawa test. If you specify -nome,
it will compute minimum-evolution bootstrap supports instead
In either case, the support values are proportions ranging from 0 to 1.</p>
</div></blockquote>
<p>This property controls the addition of the -nome switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.noml">
<em class="property">property </em><code class="sig-name descname">noml</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.noml" title="Permalink to this definition">¶</a></dt>
<dd><p>Deactivate min-evo NNIs and SPRs.</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs,
and -spr to set the rounds of SPRs.
Use -noml to turn off both min-evo NNIs and SPRs (useful if refining
an approximately maximum-likelihood tree with further NNIs).</p>
</div></blockquote>
<p>This property controls the addition of the -noml switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.nopr">
<em class="property">property </em><code class="sig-name descname">nopr</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.nopr" title="Permalink to this definition">¶</a></dt>
<dd><p>-nopr – do not write the progress indicator to stderr.</p>
<p>This property controls the addition of the -nopr switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.nosupport">
<em class="property">property </em><code class="sig-name descname">nosupport</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.nosupport" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn off support values.</p>
<blockquote>
<div><p>Support value options:
By default, FastTree computes local support values by resampling the site
likelihoods 1,000 times and the Shimodaira Hasegawa test. If you specify -nome,
it will compute minimum-evolution bootstrap supports instead
In either case, the support values are proportions ranging from 0 to 1</p>
<p>Use -nosupport to turn off support values or -boot 100 to use just 100 resamples.</p>
</div></blockquote>
<p>This property controls the addition of the -nosupport switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.notop">
<em class="property">property </em><code class="sig-name descname">notop</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.notop" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn off top-hit list to speed up search</p>
<blockquote>
<div><p>Top-hit heuristics:
By default, FastTree uses a top-hit list to speed up search
Use -notop (or -slow) to turn this feature off
and compare all leaves to each other,
and all new joined nodes to each other.</p>
</div></blockquote>
<p>This property controls the addition of the -notop switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.nt">
<em class="property">property </em><code class="sig-name descname">nt</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.nt" title="Permalink to this definition">¶</a></dt>
<dd><p>By default FastTree expects protein alignments, use -nt for nucleotides</p>
<p>This property controls the addition of the -nt switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.out">
<em class="property">property </em><code class="sig-name descname">out</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.out" title="Permalink to this definition">¶</a></dt>
<dd><p>Enter &lt;output file&gt;</p>
<blockquote>
<div><p>The path to a Newick Tree output file needs to be specified.</p>
</div></blockquote>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.pseudo">
<em class="property">property </em><code class="sig-name descname">pseudo</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.pseudo" title="Permalink to this definition">¶</a></dt>
<dd><p>-pseudo [weight] – Pseudocounts are used with sequence distance estimation.</p>
<blockquote>
<div><p>Use pseudocounts to estimate distances between sequences with little or no
overlap. (Off by default.) Recommended if analyzing the alignment has
sequences with little or no overlap.
If the weight is not specified, it is 1.0</p>
</div></blockquote>
<p>This controls the addition of the -pseudo parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.quiet">
<em class="property">property </em><code class="sig-name descname">quiet</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.quiet" title="Permalink to this definition">¶</a></dt>
<dd><p>-quiet – do not write to standard error during normal operation</p>
<blockquote>
<div><p>(no progress indicator, no options summary, no likelihood values, etc.)</p>
</div></blockquote>
<p>This property controls the addition of the -quiet switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.quote">
<em class="property">property </em><code class="sig-name descname">quote</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.quote" title="Permalink to this definition">¶</a></dt>
<dd><p>-quote – add quotes to sequence names in output.</p>
<blockquote>
<div><p>Quote sequence names in the output and allow spaces, commas,
parentheses, and colons in them but not ‘ characters (fasta files only).</p>
</div></blockquote>
<p>This property controls the addition of the -quote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.rawdist">
<em class="property">property </em><code class="sig-name descname">rawdist</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.rawdist" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn off or adjust log-correction in AA or NT distances.</p>
<blockquote>
<div><p>Use -rawdist to turn the log-correction off or to use
%different instead of Jukes-Cantor in AA or NT distances</p>
<p>Distances:
Default: For protein sequences, log-corrected distances and an
amino acid dissimilarity matrix derived from BLOSUM45
or for nucleotide sequences, Jukes-Cantor distances
To specify a different matrix, use -matrix FilePrefix or -nomatrix</p>
</div></blockquote>
<p>This property controls the addition of the -rawdist switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.refresh">
<em class="property">property </em><code class="sig-name descname">refresh</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.refresh" title="Permalink to this definition">¶</a></dt>
<dd><p>Parameter for conditions that joined nodes are compared to other nodes</p>
<blockquote>
<div><p>Top-hit heuristics:
By default, FastTree uses a top-hit list to speed up search
-refresh 0.8 – compare a joined node to all other nodes if its
top-hit list is less than 80% of the desired length,
or if the age of the top-hit list is log2(m) or greater.</p>
</div></blockquote>
<p>This controls the addition of the -refresh parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.second">
<em class="property">property </em><code class="sig-name descname">second</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.second" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn 2nd-level top hits heuristic on.</p>
<blockquote>
<div><p>Top-hit heuristics:
By default, FastTree uses a top-hit list to speed up search
Use -notop (or -slow) to turn this feature off
and compare all leaves to each other,
and all new joined nodes to each other</p>
<p>-2nd or -no2nd to turn 2nd-level top hits heuristic on or off
This reduces memory usage and running time but may lead to
marginal reductions in tree quality.
(By default, -fastest turns on -2nd.)</p>
</div></blockquote>
<p>This property controls the addition of the -2nd switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.seed">
<em class="property">property </em><code class="sig-name descname">seed</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Use -seed to initialize the random number generator.</p>
<blockquote>
<div><p>Support value options:
By default, FastTree computes local support values by resampling the site
likelihoods 1,000 times and the Shimodaira Hasegawa test. If you specify -nome,
it will compute minimum-evolution bootstrap supports instead
In either case, the support values are proportions ranging from 0 to 1.</p>
</div></blockquote>
<p>This controls the addition of the -seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.slow">
<em class="property">property </em><code class="sig-name descname">slow</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.slow" title="Permalink to this definition">¶</a></dt>
<dd><p>Use an exhaustive search.</p>
<blockquote>
<div><p>Searching for the best join:
By default, FastTree combines the ‘visible set’ of fast neighbor-joining with
local hill-climbing as in relaxed neighbor-joining
-slow – exhaustive search (like NJ or BIONJ, but different gap handling)
-slow takes half an hour instead of 8 seconds for 1,250 proteins</p>
</div></blockquote>
<p>This property controls the addition of the -slow switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.slownni">
<em class="property">property </em><code class="sig-name descname">slownni</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.slownni" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn off heuristics to avoid constant subtrees with NNIs.</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs,
and -spr to set the rounds of SPRs.
Use -slownni to turn off heuristics to avoid constant subtrees
(affects both ML and ME NNIs).</p>
</div></blockquote>
<p>This property controls the addition of the -slownni switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.spr">
<em class="property">property </em><code class="sig-name descname">spr</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.spr" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the rounds of subtree-prune-regraft moves</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs,
and -spr to set the rounds of SPRs.</p>
</div></blockquote>
<p>This controls the addition of the -spr parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.sprlength">
<em class="property">property </em><code class="sig-name descname">sprlength</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.sprlength" title="Permalink to this definition">¶</a></dt>
<dd><p>Set maximum SPR move length in topology refinement (default 10).</p>
<blockquote>
<div><p>Topology refinement:
By default, FastTree tries to improve the tree with up to 4*log2(N)
rounds of minimum-evolution nearest-neighbor interchanges (NNI),
where N is the number of unique sequences, 2 rounds of
subtree-prune-regraft (SPR) moves (also min. evo.), and
up to 2*log(N) rounds of maximum-likelihood NNIs.
Use -nni to set the number of rounds of min. evo. NNIs,
and -spr to set the rounds of SPRs.</p>
</div></blockquote>
<p>This controls the addition of the -sprlength parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.top">
<em class="property">property </em><code class="sig-name descname">top</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.top" title="Permalink to this definition">¶</a></dt>
<dd><p>Top-hit list to speed up search</p>
<blockquote>
<div><p>Top-hit heuristics:
By default, FastTree uses a top-hit list to speed up search
Use -notop (or -slow) to turn this feature off
and compare all leaves to each other,
and all new joined nodes to each other.</p>
</div></blockquote>
<p>This property controls the addition of the -top switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.topm">
<em class="property">property </em><code class="sig-name descname">topm</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.topm" title="Permalink to this definition">¶</a></dt>
<dd><p>Change the top hits calculation method</p>
<blockquote>
<div><p>Top-hit heuristics:
By default, FastTree uses a top-hit list to speed up search
-topm 1.0 – set the top-hit list size to parameter*sqrt(N)
FastTree estimates the top m hits of a leaf from the
top 2*m hits of a ‘close’ neighbor, where close is
defined as d(seed,close) &lt; 0.75 * d(seed, hit of rank 2*m),
and updates the top-hits as joins proceed.</p>
</div></blockquote>
<p>This controls the addition of the -topm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.Applications.FastTreeCommandline.wag">
<em class="property">property </em><code class="sig-name descname">wag</code><a class="headerlink" href="#Bio.Phylo.Applications.FastTreeCommandline.wag" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum likelihood model options.</p>
<blockquote>
<div><p>Whelan-And-Goldman 2001 model instead of (default)
Jones-Taylor-Thorton 1992 model (a.a. only)</p>
</div></blockquote>
<p>This property controls the addition of the -wag switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.Phylo.PAML.html" class="btn btn-neutral float-right" title="Bio.Phylo.PAML package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Phylo.html" class="btn btn-neutral float-left" title="Bio.Phylo package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.74
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>

      <dl>

        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>