

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.SCOP.Raf module &mdash; Biopython 1.74 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SCOP.Residues module" href="Bio.SCOP.Residues.html" />
    <link rel="prev" title="Bio.SCOP.Hie module" href="Bio.SCOP.Hie.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SCOP.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html#module-Bio.SCOP">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.SCOP.html">Bio.SCOP package</a> &raquo;</li>
        
      <li>Bio.SCOP.Raf module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SCOP/Raf.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SCOP.Residues.html" class="btn btn-neutral float-right" title="Bio.SCOP.Residues module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SCOP.Hie.html" class="btn btn-neutral float-left" title="Bio.SCOP.Hie module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.SCOP.Raf">
<span id="bio-scop-raf-module"></span><h1>Bio.SCOP.Raf module<a class="headerlink" href="#module-Bio.SCOP.Raf" title="Permalink to this headline">¶</a></h1>
<p>ASTRAL RAF (Rapid Access Format) Sequence Maps.</p>
<p>The ASTRAL RAF Sequence Maps record the relationship between the PDB SEQRES
records (representing the sequence of the molecule used in an experiment) to
the ATOM records (representing the atoms experimentally observed).</p>
<p>This data is derived from the Protein Data Bank CIF files. Known errors in the
CIF files are corrected manually, with the original PDB file serving as the
final arbiter in case of discrepancies.</p>
<p>Residues are referenced by residue ID. This consists of a the PDB residue
sequence number (up to 4 digits) and an optional PDB insertion code (an
ascii alphabetic character, a-z, A-Z). e.g. “1”, “10A”, “1010b”, “-1”</p>
<p>See “ASTRAL RAF Sequence Maps”:<a class="reference external" href="http://astral.stanford.edu/raf.html">http://astral.stanford.edu/raf.html</a></p>
<p>Dictionary <cite>protein_letters_3to1</cite> provides a mapping from the 3-letter amino
acid codes found in PDB files to 1-letter codes.  The 3-letter codes include
chemically modified residues.</p>
<dl class="py function">
<dt id="Bio.SCOP.Raf.normalize_letters">
<code class="sig-prename descclassname">Bio.SCOP.Raf.</code><code class="sig-name descname">normalize_letters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">one_letter_code</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.normalize_letters" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert RAF one-letter amino acid codes into IUPAC standard codes.</p>
<p>Letters are uppercased, and “.” (“Unknown”) is converted to “X”.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.SCOP.Raf.SeqMapIndex">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SCOP.Raf.</code><code class="sig-name descname">SeqMapIndex</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMapIndex" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
<p>An RAF file index.</p>
<p>The RAF file itself is about 50 MB. This index provides rapid, random
access of RAF records without having to load the entire file into memory.</p>
<p>The index key is a concatenation of the  PDB ID and chain ID. e.g
“2drcA”, <code class="docutils literal notranslate"><span class="pre">&quot;155c_&quot;</span></code>. RAF uses an underscore to indicate blank
chain IDs.</p>
<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMapIndex.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">filename</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMapIndex.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the RAF file index.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>filename  – The file to index</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMapIndex.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">key</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMapIndex.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an item from the indexed file.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMapIndex.getSeqMap">
<code class="sig-name descname">getSeqMap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">residues</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMapIndex.getSeqMap" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the sequence map for a collection of residues.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>residues – A Residues instance, or a string that can be
converted into a Residues instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SCOP.Raf.SeqMap">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SCOP.Raf.</code><code class="sig-name descname">SeqMap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">line</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An ASTRAL RAF (Rapid Access Format) Sequence Map.</p>
<p>This is a list like object; You can find the location of particular residues
with index(), slice this SeqMap into fragments, and glue fragments back
together with extend().</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>pdbid – The PDB 4 character ID</p></li>
<li><p>pdb_datestamp – From the PDB file</p></li>
<li><p>version – The RAF format version. e.g. 0.01</p></li>
<li><p>flags – RAF flags. (See release notes for more information.)</p></li>
<li><p>res – A list of Res objects, one for each residue in this sequence map</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMap.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">line</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMap.index">
<code class="sig-name descname">index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">resid</span></em>, <em class="sig-param"><span class="n">chainid</span><span class="o">=</span><span class="default_value">'_'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the index of the SeqMap for the given resid and chainid.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMap.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a single Res object from the SeqMap.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMap.append">
<code class="sig-name descname">append</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">res</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append another Res object onto the list of residue mappings.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMap.extend">
<code class="sig-name descname">extend</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Append another SeqMap onto the end of self.</p>
<p>Both SeqMaps must have the same PDB ID, PDB datestamp and
RAF version.  The RAF flags are erased if they are inconsistent. This
may happen when fragments are taken from different chains.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMap.__iadd__">
<code class="sig-name descname">__iadd__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap.__iadd__" title="Permalink to this definition">¶</a></dt>
<dd><p>In place addition of SeqMap objects.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMap.__add__">
<code class="sig-name descname">__add__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Addition of SeqMap objects.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SCOP.Raf.SeqMap.getAtoms">
<code class="sig-name descname">getAtoms</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">pdb_handle</span></em>, <em class="sig-param"><span class="n">out_handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.SeqMap.getAtoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract all relevant ATOM and HETATOM records from a PDB file.</p>
<p>The PDB file is scanned for ATOM and HETATOM records. If the
chain ID, residue ID (seqNum and iCode), and residue type match
a residue in this sequence map, then the record is echoed to the
output handle.</p>
<p>This is typically used to find the coordinates of a domain, or other
residue subset.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>pdb_handle – A handle to the relevant PDB file.</p></li>
<li><p>out_handle – All output is written to this file like object.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SCOP.Raf.Res">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SCOP.Raf.</code><code class="sig-name descname">Res</code><a class="headerlink" href="#Bio.SCOP.Raf.Res" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A single residue mapping from a RAF record.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>chainid – A single character chain ID.</p></li>
<li><p>resid   – The residue ID.</p></li>
<li><p>atom    – amino acid one-letter code from ATOM records.</p></li>
<li><p>seqres  – amino acid one-letter code from SEQRES records.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.SCOP.Raf.Res.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.Res.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.SCOP.Raf.parse">
<code class="sig-prename descclassname">Bio.SCOP.Raf.</code><code class="sig-name descname">parse</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Raf.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over RAF file, giving a SeqMap object for each line.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>handle – file-like object.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.SCOP.Residues.html" class="btn btn-neutral float-right" title="Bio.SCOP.Residues module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SCOP.Hie.html" class="btn btn-neutral float-left" title="Bio.SCOP.Hie module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.74
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>

      <dl>

        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>