

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.PopGen.GenePop.Controller module &mdash; Biopython 1.74 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.PopGen.GenePop.EasyController module" href="Bio.PopGen.GenePop.EasyController.html" />
    <link rel="prev" title="Bio.PopGen.GenePop package" href="Bio.PopGen.GenePop.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html" class="icon icon-home"> Biopython
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.74
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.PopGen.html#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html#module-Bio.PopGen">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.PopGen.html">Bio.PopGen package</a> &raquo;</li>
        
          <li><a href="Bio.PopGen.GenePop.html">Bio.PopGen.GenePop package</a> &raquo;</li>
        
      <li>Bio.PopGen.GenePop.Controller module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.PopGen.GenePop.EasyController.html" class="btn btn-neutral float-right" title="Bio.PopGen.GenePop.EasyController module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.PopGen.GenePop.html" class="btn btn-neutral float-left" title="Bio.PopGen.GenePop package" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.PopGen.GenePop.Controller">
<span id="bio-popgen-genepop-controller-module"></span><h1>Bio.PopGen.GenePop.Controller module<a class="headerlink" href="#module-Bio.PopGen.GenePop.Controller" title="Permalink to this headline">¶</a></h1>
<p>Module to control GenePop.</p>
<dl class="class">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController">
<em class="property">class </em><code class="sig-prename descclassname">Bio.PopGen.GenePop.Controller.</code><code class="sig-name descname">GenePopController</code><span class="sig-paren">(</span><em class="sig-param">genepop_dir=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Define a class to interface with the GenePop program.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_allele_genotype_freqs" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_allele_genotype_freqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_allele_genotype_freqs</span></code></a>(self, fname)</p></td>
<td><p>Calculate allele and genotype frequencies per locus and per sample.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_identity" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_identity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_diversities_fis_with_identity</span></code></a>(self, fname)</p></td>
<td><p>Compute identity-base Gene diversities and Fis.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_size" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_diversities_fis_with_size</span></code></a>(self, fname)</p></td>
<td><p>Provision to Computer Allele size-based Gene diversities and Fis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_all" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_all"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_fst_all</span></code></a>(self, fname)</p></td>
<td><p>Execute GenePop and gets Fst/Fis/Fit (all populations).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_pair" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_fst_pair</span></code></a>(self, fname)</p></td>
<td><p>Estimate spatial structure from Allele identity for all population pairs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_diplo" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_diplo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_ibd_diplo</span></code></a>(self, fname[, stat, scale, …])</p></td>
<td><p>Calculate isolation by distance statistics for diploid data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_haplo" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_haplo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_ibd_haplo</span></code></a>(self, fname[, stat, scale, …])</p></td>
<td><p>Calculate isolation by distance statistics for haploid data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_all" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_all"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_rho_all</span></code></a>(self, fname)</p></td>
<td><p>Provision for estimating spatial structure from Allele size for all populations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_pair" title="Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calc_rho_pair</span></code></a>(self, fname)</p></td>
<td><p>Provision for estimating spatial structure from Allele size for all population pairs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.create_contingency_tables" title="Bio.PopGen.GenePop.Controller.GenePopController.create_contingency_tables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">create_contingency_tables</span></code></a>(self, fname)</p></td>
<td><p>Provision for creating Genotypic contingency tables.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.estimate_nm" title="Bio.PopGen.GenePop.Controller.GenePopController.estimate_nm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">estimate_nm</span></code></a>(self, fname)</p></td>
<td><p>Estimate the Number of Migrants.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_all" title="Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_all"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_genic_diff_all</span></code></a>(self, fname[, …])</p></td>
<td><p>Provision for Genic differentiation for all populations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_pair" title="Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_genic_diff_pair</span></code></a>(self, fname[, …])</p></td>
<td><p>Provision for Genic differentiation for all population pairs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_all" title="Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_all"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_genotypic_diff_all</span></code></a>(self, fname[, …])</p></td>
<td><p>Provision for Genotypic differentiation for all populations.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_pair" title="Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_genotypic_diff_pair</span></code></a>(self, fname[, …])</p></td>
<td><p>Provision for Genotypic differentiation for all population pairs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_deficiency" title="Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_deficiency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_global_hz_deficiency</span></code></a>(self, fname[, …])</p></td>
<td><p>Use Global Hardy-Weinberg test for heterozygote deficiency.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_excess" title="Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_excess"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_global_hz_excess</span></code></a>(self, fname[, …])</p></td>
<td><p>Use Global Hardy-Weinberg test for heterozygote excess.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_ld" title="Bio.PopGen.GenePop.Controller.GenePopController.test_ld"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_ld</span></code></a>(self, fname[, dememorization, …])</p></td>
<td><p>Test for linkage disequilibrium on each pair of loci in each population.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_deficiency" title="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_deficiency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_pop_hz_deficiency</span></code></a>(self, fname[, …])</p></td>
<td><p>Use Hardy-Weinberg test for heterozygote deficiency.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_excess" title="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_excess"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_pop_hz_excess</span></code></a>(self, fname[, enum_test, …])</p></td>
<td><p>Use Hardy-Weinberg test for heterozygote deficiency.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_prob" title="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_prob"><code class="xref py py-obj docutils literal notranslate"><span class="pre">test_pop_hz_prob</span></code></a>(self, fname, ext[, …])</p></td>
<td><p>Use Hardy-Weinberg test based on probability.</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_allele_genotype_freqs">
<code class="sig-name descname">calc_allele_genotype_freqs</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_allele_genotype_freqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate allele and genotype frequencies per locus and per sample.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>fname - file name</p></li>
</ul>
</dd>
<dt>Returns tuple with 2 elements:</dt><dd><ul class="simple">
<li><p>Population iterator with</p>
<ul>
<li><p>population name</p></li>
<li><p>Locus dictionary with key = locus name and content tuple as
Genotype List with
(Allele1, Allele2, observed, expected)
(expected homozygotes, observed hm,
expected heterozygotes, observed ht)
Allele frequency/Fis dictionary with allele as key and
(count, frequency, Fis Weir &amp; Cockerham)</p></li>
<li><p>Totals as a pair</p></li>
<li><p>count</p></li>
<li><p>Fis Weir &amp; Cockerham,</p></li>
<li><p>Fis Robertson &amp; Hill</p></li>
</ul>
</li>
<li><p>Locus iterator with</p>
<ul>
<li><p>Locus name</p></li>
<li><p>allele list</p></li>
<li><p>Population list with a triple</p>
<ul>
<li><p>population name</p></li>
<li><p>list of allele frequencies in the same order as allele list above</p></li>
<li><p>number of genes</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</dd>
</dl>
<p>Will create a file called fname.INF</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_identity">
<code class="sig-name descname">calc_diversities_fis_with_identity</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_identity" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute identity-base Gene diversities and Fis.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_size">
<code class="sig-name descname">calc_diversities_fis_with_size</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Provision to Computer Allele size-based Gene diversities and Fis.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_all">
<code class="sig-name descname">calc_fst_all</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute GenePop and gets Fst/Fis/Fit (all populations).</p>
<dl class="simple">
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>fname - file name</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><ul class="simple">
<li><p>(multiLocusFis, multiLocusFst, multiLocus Fit),</p></li>
<li><p>Iterator of tuples
(Locus name, Fis, Fst, Fit, Qintra, Qinter)</p></li>
</ul>
</dd>
</dl>
<p>Will create a file called <cite>fname.FST</cite>.</p>
<p>This does not return the genotype frequencies.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_pair">
<code class="sig-name descname">calc_fst_pair</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimate spatial structure from Allele identity for all population pairs.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_diplo">
<code class="sig-name descname">calc_ibd_diplo</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">stat='a'</em>, <em class="sig-param">scale='Log'</em>, <em class="sig-param">min_dist=1e-05</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_diplo" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate isolation by distance statistics for diploid data.</p>
<p>See _calc_ibd for parameter details.</p>
<p>Note that each pop can only have a single individual and
the individual name has to be the sample coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_haplo">
<code class="sig-name descname">calc_ibd_haplo</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">stat='a'</em>, <em class="sig-param">scale='Log'</em>, <em class="sig-param">min_dist=1e-05</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_haplo" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate isolation by distance statistics for haploid data.</p>
<p>See _calc_ibd for parameter details.</p>
<p>Note that each pop can only have a single individual and
the individual name has to be the sample coordinates.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_all">
<code class="sig-name descname">calc_rho_all</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Provision for estimating spatial structure from Allele size for all populations.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_pair">
<code class="sig-name descname">calc_rho_pair</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Provision for estimating spatial structure from Allele size for all population pairs.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.create_contingency_tables">
<code class="sig-name descname">create_contingency_tables</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.create_contingency_tables" title="Permalink to this definition">¶</a></dt>
<dd><p>Provision for creating Genotypic contingency tables.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.estimate_nm">
<code class="sig-name descname">estimate_nm</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.estimate_nm" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimate the Number of Migrants.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>fname - file name</p></li>
</ul>
</dd>
<dt>Returns</dt><dd><ul class="simple">
<li><p>Mean sample size</p></li>
<li><p>Mean frequency of private alleles</p></li>
<li><p>Number of migrants for Ne=10</p></li>
<li><p>Number of migrants for Ne=25</p></li>
<li><p>Number of migrants for Ne=50</p></li>
<li><p>Number of migrants after correcting for expected size</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_all">
<code class="sig-name descname">test_genic_diff_all</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Provision for Genic differentiation for all populations.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_pair">
<code class="sig-name descname">test_genic_diff_pair</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Provision for Genic differentiation for all population pairs.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_all">
<code class="sig-name descname">test_genotypic_diff_all</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Provision for Genotypic differentiation for all populations.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_pair">
<code class="sig-name descname">test_genotypic_diff_pair</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Provision for Genotypic differentiation for all population pairs.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_deficiency">
<code class="sig-name descname">test_global_hz_deficiency</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">enum_test=True</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_deficiency" title="Permalink to this definition">¶</a></dt>
<dd><p>Use Global Hardy-Weinberg test for heterozygote deficiency.</p>
<dl class="simple">
<dt>Returns a triple with:</dt><dd><ul class="simple">
<li><p>An list per population containing (pop_name, P-val, SE, switches).
Some pops have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>An list per loci containing (locus_name, P-val, SE, switches).
Some loci have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>Overall results (P-val, SE, switches).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_excess">
<code class="sig-name descname">test_global_hz_excess</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">enum_test=True</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_excess" title="Permalink to this definition">¶</a></dt>
<dd><p>Use Global Hardy-Weinberg test for heterozygote excess.</p>
<dl class="simple">
<dt>Returns a triple with:</dt><dd><ul class="simple">
<li><p>A list per population containing (pop_name, P-val, SE, switches).
Some pops have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>A list per loci containing (locus_name, P-val, SE, switches).
Some loci have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>Overall results (P-val, SE, switches)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_ld">
<code class="sig-name descname">test_ld</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_ld" title="Permalink to this definition">¶</a></dt>
<dd><p>Test for linkage disequilibrium on each pair of loci in each population.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_deficiency">
<code class="sig-name descname">test_pop_hz_deficiency</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">enum_test=True</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_deficiency" title="Permalink to this definition">¶</a></dt>
<dd><p>Use Hardy-Weinberg test for heterozygote deficiency.</p>
<p>Returns a population iterator containing a dictionary wehre
dictionary[locus]=(P-val, SE, Fis-WC, Fis-RH, steps).</p>
<p>Some loci have a None if the info is not available.
SE might be none (for enumerations).</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_excess">
<code class="sig-name descname">test_pop_hz_excess</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">enum_test=True</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_excess" title="Permalink to this definition">¶</a></dt>
<dd><p>Use Hardy-Weinberg test for heterozygote deficiency.</p>
<p>Returns a population iterator containing a dictionary where
dictionary[locus]=(P-val, SE, Fis-WC, Fis-RH, steps).</p>
<p>Some loci have a None if the info is not available.
SE might be none (for enumerations).</p>
</dd></dl>

<dl class="method">
<dt id="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_prob">
<code class="sig-name descname">test_pop_hz_prob</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fname</em>, <em class="sig-param">ext</em>, <em class="sig-param">enum_test=False</em>, <em class="sig-param">dememorization=10000</em>, <em class="sig-param">batches=20</em>, <em class="sig-param">iterations=5000</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_prob" title="Permalink to this definition">¶</a></dt>
<dd><p>Use Hardy-Weinberg test based on probability.</p>
<p>Returns 2 iterators and a final tuple:</p>
<blockquote>
<div><ol class="arabic simple">
<li><dl class="simple">
<dt>Returns a loci iterator containing:</dt><dd><ul class="simple">
<li><p>A dictionary[pop_pos]=(P-val, SE, Fis-WC, Fis-RH, steps).
Some pops have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>Result of Fisher’s test (Chi2, deg freedom, prob).</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Returns a population iterator containing:</dt><dd><ul class="simple">
<li><p>A dictionary[locus]=(P-val, SE, Fis-WC, Fis-RH, steps).
Some loci have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>Result of Fisher’s test (Chi2, deg freedom, prob).</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Final tuple (Chi2, deg freedom, prob).</p></li>
</ol>
</div></blockquote>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.PopGen.GenePop.EasyController.html" class="btn btn-neutral float-right" title="Bio.PopGen.GenePop.EasyController module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.PopGen.GenePop.html" class="btn btn-neutral float-left" title="Bio.PopGen.GenePop package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2017, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>