

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.AlignIO.MauveIO module &mdash; Biopython 1.74 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.AlignIO.NexusIO module" href="Bio.AlignIO.NexusIO.html" />
    <link rel="prev" title="Bio.AlignIO.MafIO module" href="Bio.AlignIO.MafIO.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html" class="icon icon-home"> Biopython
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.74
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.AlignIO.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html#module-Bio.AlignIO">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.AlignIO.html">Bio.AlignIO package</a> &raquo;</li>
        
      <li>Bio.AlignIO.MauveIO module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.AlignIO.NexusIO.html" class="btn btn-neutral float-right" title="Bio.AlignIO.NexusIO module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.AlignIO.MafIO.html" class="btn btn-neutral float-left" title="Bio.AlignIO.MafIO module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.AlignIO.MauveIO">
<span id="bio-alignio-mauveio-module"></span><h1>Bio.AlignIO.MauveIO module<a class="headerlink" href="#module-Bio.AlignIO.MauveIO" title="Permalink to this headline">¶</a></h1>
<p>Bio.AlignIO support for “xmfa” output from Mauve/ProgressiveMauve.</p>
<p>You are expected to use this module via the Bio.AlignIO functions (or the
Bio.SeqIO functions if you want to work directly with the gapped sequences).</p>
<p>For example, consider a progressiveMauve alignment file containing the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#FormatVersion Mauve1</span>
<span class="c1">#Sequence1File      a.fa</span>
<span class="c1">#Sequence1Entry     1</span>
<span class="c1">#Sequence1Format    FastA</span>
<span class="c1">#Sequence2File      b.fa</span>
<span class="c1">#Sequence2Entry     2</span>
<span class="c1">#Sequence2Format    FastA</span>
<span class="c1">#Sequence3File      c.fa</span>
<span class="c1">#Sequence3Entry     3</span>
<span class="c1">#Sequence3Format    FastA</span>
<span class="c1">#BackboneFile       three.xmfa.bbcols</span>
<span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span><span class="mi">0</span><span class="o">-</span><span class="mi">0</span> <span class="o">+</span> <span class="n">a</span><span class="o">.</span><span class="n">fa</span>
<span class="o">--------------------------------------------------------------------------------</span>
<span class="o">--------------------------------------------------------------------------------</span>
<span class="o">--------------------------------------------------------------------------------</span>
<span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span><span class="mi">5417</span><span class="o">-</span><span class="mi">5968</span> <span class="o">+</span> <span class="n">b</span><span class="o">.</span><span class="n">fa</span>
<span class="n">TTTAAACATCCCTCGGCCCGTCGCCCTTTTATAATAGCAGTACGTGAGAGGAGCGCCCTAAGCTTTGGGAAATTCAAGC</span><span class="o">-</span>
<span class="o">--------------------------------------------------------------------------------</span>
<span class="n">CTGGAACGTACTTGCTGGTTTCGCTACTATTTCAAACAAGTTAGAGGCCGTTACCTCGGGCGAACGTATAAACCATTCTG</span>
<span class="o">&gt;</span> <span class="mi">3</span><span class="p">:</span><span class="mi">9476</span><span class="o">-</span><span class="mi">10076</span> <span class="o">-</span> <span class="n">c</span><span class="o">.</span><span class="n">fa</span>
<span class="n">TTTAAACACCTTTTTGGATG</span><span class="o">--</span><span class="n">GCCCAGTTCGTTCAGTTGTG</span><span class="o">-</span><span class="n">GGGAGGAGATCGCCCCAAACGTATGGTGAGTCGGGCG</span>
<span class="n">TTTCCTATAGCTATAGGACCAATCCACTTACCATACGCCCGGCGTCGCCCAGTCCGGTTCGGTACCCTCCATGACCCACG</span>
<span class="o">---------------------------------------------------------</span><span class="n">AAATGAGGGCCCAGGGTATGCTT</span>
<span class="o">=</span>
<span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span><span class="mi">5969</span><span class="o">-</span><span class="mi">6015</span> <span class="o">+</span> <span class="n">b</span><span class="o">.</span><span class="n">fa</span>
<span class="o">-----------------------</span>
<span class="n">GGGCGAACGTATAAACCATTCTG</span>
<span class="o">&gt;</span> <span class="mi">3</span><span class="p">:</span><span class="mi">9429</span><span class="o">-</span><span class="mi">9476</span> <span class="o">-</span> <span class="n">c</span><span class="o">.</span><span class="n">fa</span>
<span class="n">TTCGGTACCCTCCATGACCCACG</span>
<span class="n">AAATGAGGGCCCAGGGTATGCTT</span>
</pre></div>
</div>
<p>This is a multiple sequence alignment with multiple aligned sections, so you
would probably load this using the Bio.AlignIO.parse() function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">AlignIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Mauve/simple_short.xmfa&quot;</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">aln</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">SingleLetterAlphabet() alignment with 3 rows and 240 columns</span>
<span class="go">--------------------------------------------...--- a.fa</span>
<span class="go">TTTAAACATCCCTCGGCCCGTCGCCCTTTTATAATAGCAGTACG...CTG b.fa/5416-5968</span>
<span class="go">TTTAAACACCTTTTTGGATG--GCCCAGTTCGTTCAGTTGTG-G...CTT c.fa/9475-10076</span>
<span class="go">SingleLetterAlphabet() alignment with 2 rows and 46 columns</span>
<span class="go">-----------------------GGGCGAACGTATAAACCATTCTG b.fa/5968-6015</span>
<span class="go">TTCGGTACCCTCCATGACCCACGAAATGAGGGCCCAGGGTATGCTT c.fa/9428-9476</span>
</pre></div>
</div>
<p>Additional information is extracted from the XMFA file and available through
the annotation attribute of each record:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">))</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  start: </span><span class="si">%d</span><span class="s2">, end: </span><span class="si">%d</span><span class="s2">, strand: </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s1">&#39;strand&#39;</span><span class="p">]))</span>
<span class="gp">...</span>
<span class="go">a.fa 240</span>
<span class="go">  start: 0, end: 0, strand: 1</span>
<span class="go">b.fa/5416-5968 240</span>
<span class="go">  start: 5416, end: 5968, strand: 1</span>
<span class="go">c.fa/9475-10076 240</span>
<span class="go">  start: 9475, end: 10076, strand: -1</span>
</pre></div>
</div>
<dl class="class">
<dt id="Bio.AlignIO.MauveIO.MauveIterator">
<em class="property">class </em><code class="sig-prename descclassname">Bio.AlignIO.MauveIO.</code><code class="sig-name descname">MauveIterator</code><span class="sig-paren">(</span><em class="sig-param">handle</em>, <em class="sig-param">seq_count=None</em>, <em class="sig-param">alphabet=SingleLetterAlphabet()</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.MauveIO.MauveIterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.AlignIO.Interfaces.html#Bio.AlignIO.Interfaces.AlignmentIterator" title="Bio.AlignIO.Interfaces.AlignmentIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.AlignIO.Interfaces.AlignmentIterator</span></code></a></p>
<p>Mauve xmfa alignment iterator.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.AlignIO.MauveIO.MauveWriter">
<em class="property">class </em><code class="sig-prename descclassname">Bio.AlignIO.MauveIO.</code><code class="sig-name descname">MauveWriter</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.MauveIO.MauveWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.AlignIO.Interfaces.html#Bio.AlignIO.Interfaces.SequentialAlignmentWriter" title="Bio.AlignIO.Interfaces.SequentialAlignmentWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.AlignIO.Interfaces.SequentialAlignmentWriter</span></code></a></p>
<p>Mauve/XMFA alignment writer.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">clean</span></code>(self, text)</p></td>
<td><p>Use this to avoid getting newlines in the output.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.AlignIO.MauveIO.MauveWriter.write_alignment" title="Bio.AlignIO.MauveIO.MauveWriter.write_alignment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_alignment</span></code></a>(self, alignment)</p></td>
<td><p>Use this to write (another) single alignment to an open file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_file</span></code>(self, alignments)</p></td>
<td><p>Use this to write an entire file containing the given alignments.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_footer</span></code>(self)</p></td>
<td><p>Use this to write any footer.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_header</span></code>(self)</p></td>
<td><p>Use this to write any header.</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.AlignIO.MauveIO.MauveWriter.write_alignment">
<code class="sig-name descname">write_alignment</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">alignment</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.MauveIO.MauveWriter.write_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Use this to write (another) single alignment to an open file.</p>
<p>Note that sequences and their annotation are recorded
together (rather than having a block of annotation followed
by a block of aligned sequences).</p>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.AlignIO.NexusIO.html" class="btn btn-neutral float-right" title="Bio.AlignIO.NexusIO module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.AlignIO.MafIO.html" class="btn btn-neutral float-left" title="Bio.AlignIO.MafIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2017, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>