

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Nexus.Trees module &mdash; Biopython 1.77 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.PDB package" href="Bio.PDB.html" />
    <link rel="prev" title="Bio.Nexus.StandardData module" href="Bio.Nexus.StandardData.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Nexus.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html#module-Bio.Nexus">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.Nexus.html">Bio.Nexus package</a> &raquo;</li>
        
      <li>Bio.Nexus.Trees module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Nexus/Trees.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.PDB.html" class="btn btn-neutral float-right" title="Bio.PDB package" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Nexus.StandardData.html" class="btn btn-neutral float-left" title="Bio.Nexus.StandardData module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.Nexus.Trees">
<span id="bio-nexus-trees-module"></span><h1>Bio.Nexus.Trees module<a class="headerlink" href="#module-Bio.Nexus.Trees" title="Permalink to this headline">Â¶</a></h1>
<p>Tree class to handle phylogenetic trees.</p>
<p>Provides a set of methods to read and write newick-format tree descriptions,
get information about trees (monphyly of taxon sets, congruence between trees,
common ancestors,â¦) and to manipulate trees (re-root trees, split terminal
nodes).</p>
<dl class="py exception">
<dt id="Bio.Nexus.Trees.TreeError">
<em class="property">exception </em><code class="sig-prename descclassname">Bio.Nexus.Trees.</code><code class="sig-name descname">TreeError</code><a class="headerlink" href="#Bio.Nexus.Trees.TreeError" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Provision for the management of Tree exceptions.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.Nexus.Trees.NodeData">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Nexus.Trees.</code><code class="sig-name descname">NodeData</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxon</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">branchlength</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">support</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">comment</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.NodeData" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Store tree-relevant data associated with nodes (e.g. branches or otus).</p>
<dl class="py method">
<dt id="Bio.Nexus.Trees.NodeData.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">taxon</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">branchlength</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">support</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">comment</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.NodeData.__init__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Nexus.Trees.Tree">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Nexus.Trees.</code><code class="sig-name descname">Tree</code><span class="sig-paren">(</span><em class="sig-param">tree=None</em>, <em class="sig-param">weight=1.0</em>, <em class="sig-param">rooted=False</em>, <em class="sig-param">name=''</em>, <em class="sig-param">data=&lt;class 'Bio.Nexus.Trees.NodeData'&gt;</em>, <em class="sig-param">values_are_support=False</em>, <em class="sig-param">max_support=1.0</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Nexus.Nodes.html#Bio.Nexus.Nodes.Chain" title="Bio.Nexus.Nodes.Chain"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Nexus.Nodes.Chain</span></code></a></p>
<p>Represent a tree using a chain of nodes with on predecessor (=ancestor) and multiple successors (=subclades).</p>
<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">tree=None</em>, <em class="sig-param">weight=1.0</em>, <em class="sig-param">rooted=False</em>, <em class="sig-param">name=''</em>, <em class="sig-param">data=&lt;class 'Bio.Nexus.Trees.NodeData'&gt;</em>, <em class="sig-param">values_are_support=False</em>, <em class="sig-param">max_support=1.0</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.__init__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Ntree(self,tree).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.node">
<code class="sig-name descname">node</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node_id</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.node" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the instance of node_id.</p>
<p>node = node(self,node_id)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.split">
<code class="sig-name descname">split</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">parent_id</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">n</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">branchlength</span><span class="o">=</span><span class="default_value">1.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.split" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Speciation: generates n (default two) descendants of a node.</p>
<p>[new ids] = split(self,parent_id=None,n=2,branchlength=1.0):</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.search_taxon">
<code class="sig-name descname">search_taxon</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">taxon</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.search_taxon" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the first matching taxon in self.data.taxon. Not restricted to terminal nodes.</p>
<p>node_id = search_taxon(self,taxon)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.prune">
<code class="sig-name descname">prune</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">taxon</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.prune" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Prune a terminal taxon from the tree.</p>
<p>id_of_previous_node = prune(self,taxon)
If taxon is from a bifurcation, the connectiong node will be collapsed
and its branchlength added to remaining terminal node. This might be no
longer a meaningful valueâ</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.get_taxa">
<code class="sig-name descname">get_taxa</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node_id</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.get_taxa" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a list of all otus downwards from a node.</p>
<p>nodes = get_taxa(self,node_id=None)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.get_terminals">
<code class="sig-name descname">get_terminals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.get_terminals" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a list of all terminal nodes.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.is_terminal">
<code class="sig-name descname">is_terminal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.is_terminal" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True if node is a terminal node.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.is_internal">
<code class="sig-name descname">is_internal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.is_internal" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True if node is an internal node.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.is_preterminal">
<code class="sig-name descname">is_preterminal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.is_preterminal" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True if all successors of a node are terminal ones.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.count_terminals">
<code class="sig-name descname">count_terminals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.count_terminals" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Count the number of terminal nodes that are attached to a node.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.collapse_genera">
<code class="sig-name descname">collapse_genera</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">space_equals_underscore</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.collapse_genera" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Collapse all subtrees which belong to the same genus.</p>
<p>(i.e share the same first word in their taxon name.)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.sum_branchlength">
<code class="sig-name descname">sum_branchlength</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">root</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">node</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.sum_branchlength" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Add up the branchlengths from root (default self.root) to node.</p>
<p>sum = sum_branchlength(self,root=None,node=None)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.set_subtree">
<code class="sig-name descname">set_subtree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.set_subtree" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return subtree as a set of nested sets.</p>
<p>sets = set_subtree(self,node)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.is_identical">
<code class="sig-name descname">is_identical</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">tree2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.is_identical" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Compare tree and tree2 for identity.</p>
<p>result = is_identical(self,tree2)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.is_compatible">
<code class="sig-name descname">is_compatible</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">tree2</span></em>, <em class="sig-param"><span class="n">threshold</span></em>, <em class="sig-param"><span class="n">strict</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.is_compatible" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Compare branches with support&gt;threshold for compatibility.</p>
<p>result = is_compatible(self,tree2,threshold)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.common_ancestor">
<code class="sig-name descname">common_ancestor</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node1</span></em>, <em class="sig-param"><span class="n">node2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.common_ancestor" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the common ancestor that connects two nodes.</p>
<p>node_id = common_ancestor(self,node1,node2)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.distance">
<code class="sig-name descname">distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node1</span></em>, <em class="sig-param"><span class="n">node2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.distance" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Add and return the sum of the branchlengths between two nodes.</p>
<p>dist = distance(self,node1,node2)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.is_monophyletic">
<code class="sig-name descname">is_monophyletic</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">taxon_list</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.is_monophyletic" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return node_id of common ancestor if taxon_list is monophyletic, -1 otherwise.</p>
<p>result = is_monophyletic(self,taxon_list)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.is_bifurcating">
<code class="sig-name descname">is_bifurcating</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.is_bifurcating" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True if tree downstream of node is strictly bifurcating.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.branchlength2support">
<code class="sig-name descname">branchlength2support</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.branchlength2support" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Move values stored in data.branchlength to data.support, and set branchlength to 0.0.</p>
<p>This is necessary when support has been stored as branchlength (e.g. paup), and has thus
been read in as branchlength.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.convert_absolute_support">
<code class="sig-name descname">convert_absolute_support</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">nrep</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.convert_absolute_support" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Convert absolute support (clade-count) to rel. frequencies.</p>
<p>Some software (e.g. PHYLIP consense) just calculate how often clades appear, instead of
calculating relative frequencies.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.has_support">
<code class="sig-name descname">has_support</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">node</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.has_support" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True if any of the nodes has data.support != None.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.randomize">
<code class="sig-name descname">randomize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">ntax</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">taxon_list</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">branchlength</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">branchlength_sd</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">bifurcate</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.randomize" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Generate a random tree with ntax taxa and/or taxa from taxlabels.</p>
<p>new_tree = randomize(self,ntax=None,taxon_list=None,branchlength=1.0,branchlength_sd=None,bifurcate=True)
Trees are bifurcating by default. (Polytomies not yet supported).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.display">
<code class="sig-name descname">display</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.display" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Quick and dirty lists of all nodes.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.to_string">
<code class="sig-name descname">to_string</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">support_as_branchlengths</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">branchlengths_only</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">plain</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">plain_newick</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ladderize</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ignore_comments</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.to_string" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a paup compatible tree line.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.__str__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Short version of to_string(), gives plain tree.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.unroot">
<code class="sig-name descname">unroot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.unroot" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Define a unrooted Tree structure, using data of a rooted Tree.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.root_with_outgroup">
<code class="sig-name descname">root_with_outgroup</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">outgroup</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.root_with_outgroup" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Define a treeâs root with a reference group outgroup.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Trees.Tree.merge_with_support">
<code class="sig-name descname">merge_with_support</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">bstrees</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">constree</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">threshold</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">outgroup</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.Tree.merge_with_support" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Merge clade support (from consensus or list of bootstrap-trees) with phylogeny.</p>
<p>tree=merge_bootstrap(phylo,bs_tree=&lt;list_of_trees&gt;)
or
tree=merge_bootstrap(phylo,consree=consensus_tree with clade support)</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.Nexus.Trees.consensus">
<code class="sig-prename descclassname">Bio.Nexus.Trees.</code><code class="sig-name descname">consensus</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">trees</span></em>, <em class="sig-param"><span class="n">threshold</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">outgroup</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Trees.consensus" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Compute a majority rule consensus tree of all clades with relative frequency&gt;=threshold from a list of trees.</p>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.PDB.html" class="btn btn-neutral float-right" title="Bio.PDB package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Nexus.StandardData.html" class="btn btn-neutral float-left" title="Bio.Nexus.StandardData module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.77
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.76/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>

      <dl>

        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>