<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.SeqIO.SffIO module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.SeqIO.SnapGeneIO module" href="Bio.SeqIO.SnapGeneIO.html" />
    <link rel="prev" title="Bio.SeqIO.SeqXmlIO module" href="Bio.SeqIO.SeqXmlIO.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
      <li class="breadcrumb-item active">Bio.SeqIO.SffIO module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SeqIO/SffIO.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.SeqIO.SeqXmlIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.SeqXmlIO module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqIO.SnapGeneIO.html" class="btn btn-neutral float-right" title="Bio.SeqIO.SnapGeneIO module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.SeqIO.SffIO">
<span id="bio-seqio-sffio-module"></span><h1>Bio.SeqIO.SffIO module<a class="headerlink" href="#module-Bio.SeqIO.SffIO" title="Permalink to this headline"></a></h1>
<p>Bio.SeqIO support for the binary Standard Flowgram Format (SFF) file format.</p>
<p>SFF was designed by 454 Life Sciences (Roche), the Whitehead Institute for
Biomedical Research and the Wellcome Trust Sanger Institute. SFF was also used
as the native output format from early versions of Ion Torrent’s PGM platform
as well. You are expected to use this module via the Bio.SeqIO functions under
the format name “sff” (or “sff-trim” as described below).</p>
<p>For example, to iterate over the records in an SFF file,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2"> </span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">20</span><span class="p">]))</span>
<span class="gp">...</span>
<span class="go">E3MFGYR02JWQ7T 265 tcagGGTCTACATGTTGGTT...</span>
<span class="go">E3MFGYR02JA6IL 271 tcagTTTTTTTTGGAAAGGA...</span>
<span class="go">E3MFGYR02JHD4H 310 tcagAAAGACAAGTGGTATC...</span>
<span class="go">E3MFGYR02GFKUC 299 tcagCGGCCGGGCCTCTCAT...</span>
<span class="go">E3MFGYR02FTGED 281 tcagTGGTAATGGGGGGAAA...</span>
<span class="go">E3MFGYR02FR9G7 261 tcagCTCCGTAAGAAGGTGC...</span>
<span class="go">E3MFGYR02GAZMS 278 tcagAAAGAAGTAAGGTAAA...</span>
<span class="go">E3MFGYR02HHZ8O 221 tcagACTTTCTTCTTTACCG...</span>
<span class="go">E3MFGYR02GPGB1 269 tcagAAGCAGTGGTATCAAC...</span>
<span class="go">E3MFGYR02F7Z7G 219 tcagAATCATCCACTTTTTA...</span>
</pre></div>
</div>
<p>Each SeqRecord object will contain all the annotation from the SFF file,
including the PHRED quality scores.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)))</span>
<span class="go">E3MFGYR02F7Z7G 219</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">tcagAATCAT...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%r</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s2">&quot;phred_quality&quot;</span><span class="p">][:</span><span class="mi">10</span><span class="p">]))</span>
<span class="go">[22, 21, 23, 28, 26, 15, 12, 21, 28, 21]...</span>
</pre></div>
</div>
<p>Notice that the sequence is given in mixed case, the central upper case region
corresponds to the trimmed sequence. This matches the output of the Roche
tools (and the 3rd party tool sff_extract) for SFF to FASTA.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;clip_qual_left&quot;</span><span class="p">])</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;clip_qual_right&quot;</span><span class="p">])</span>
<span class="go">134</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">4</span><span class="p">])</span>
<span class="go">tcag</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">...</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">20</span><span class="p">],</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">120</span><span class="p">:</span><span class="mi">134</span><span class="p">]))</span>
<span class="go">AATCATCCACTTTTTA...CAAAACACAAACAG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">134</span><span class="p">:])</span>
<span class="go">atcttatcaacaaaactcaaagttcctaactgagacacgcaacaggggataagacaaggcacacaggggataggnnnnnnnnnnn</span>
</pre></div>
</div>
<p>The annotations dictionary also contains any adapter clip positions
(usually zero), and information about the flows. e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
<span class="go">12</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;flow_key&quot;</span><span class="p">])</span>
<span class="go">TCAG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;flow_values&quot;</span><span class="p">][:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">(83, 1, 128, 7, 4, 84, 6, 106, 3, 172)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;flow_values&quot;</span><span class="p">]))</span>
<span class="go">400</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;flow_index&quot;</span><span class="p">][:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">(1, 2, 3, 2, 2, 0, 3, 2, 3, 3)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;flow_index&quot;</span><span class="p">]))</span>
<span class="go">219</span>
</pre></div>
</div>
<p>Note that to convert from a raw reading in flow_values to the corresponding
homopolymer stretch estimate, the value should be rounded to the nearest 100:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%r</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="p">))</span> <span class="o">//</span> <span class="mi">100</span>
<span class="gp">... </span>                 <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;flow_values&quot;</span><span class="p">][:</span><span class="mi">10</span><span class="p">]])</span>
<span class="gp">...</span>
<span class="go">[1, 0, 1, 0, 0, 1, 0, 1, 0, 2]...</span>
</pre></div>
</div>
<p>If a read name is exactly 14 alphanumeric characters, the annotations
dictionary will also contain meta-data about the read extracted by
interpreting the name as a 454 Sequencing System “Universal” Accession
Number. Note that if a read name happens to be exactly 14 alphanumeric
characters but was not generated automatically, these annotation records
will contain nonsense information.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;region&quot;</span><span class="p">])</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;time&quot;</span><span class="p">])</span>
<span class="go">[2008, 1, 9, 16, 16, 0]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;coords&quot;</span><span class="p">])</span>
<span class="go">(2434, 1658)</span>
</pre></div>
</div>
<p>As a convenience method, you can read the file with SeqIO format name “sff-trim”
instead of “sff” to get just the trimmed sequences (without any annotation
except for the PHRED quality scores and anything encoded in the read names):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff-trim&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2"> </span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">20</span><span class="p">]))</span>
<span class="gp">...</span>
<span class="go">E3MFGYR02JWQ7T 260 GGTCTACATGTTGGTTAACC...</span>
<span class="go">E3MFGYR02JA6IL 265 TTTTTTTTGGAAAGGAAAAC...</span>
<span class="go">E3MFGYR02JHD4H 292 AAAGACAAGTGGTATCAACG...</span>
<span class="go">E3MFGYR02GFKUC 295 CGGCCGGGCCTCTCATCGGT...</span>
<span class="go">E3MFGYR02FTGED 277 TGGTAATGGGGGGAAATTTA...</span>
<span class="go">E3MFGYR02FR9G7 256 CTCCGTAAGAAGGTGCTGCC...</span>
<span class="go">E3MFGYR02GAZMS 271 AAAGAAGTAAGGTAAATAAC...</span>
<span class="go">E3MFGYR02HHZ8O 150 ACTTTCTTCTTTACCGTAAC...</span>
<span class="go">E3MFGYR02GPGB1 221 AAGCAGTGGTATCAACGCAG...</span>
<span class="go">E3MFGYR02F7Z7G 130 AATCATCCACTTTTTAACGT...</span>
</pre></div>
</div>
<p>Looking at the final record in more detail, note how this differs to the
example above:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)))</span>
<span class="go">E3MFGYR02F7Z7G 130</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">AATCATCCAC...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%r</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="n">record</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s2">&quot;phred_quality&quot;</span><span class="p">][:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">[26, 15, 12, 21, 28, 21, 36, 28, 27, 27]...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;region&quot;</span><span class="p">])</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;coords&quot;</span><span class="p">])</span>
<span class="go">(2434, 1658)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;time&quot;</span><span class="p">])</span>
<span class="go">[2008, 1, 9, 16, 16, 0]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">])</span>
<span class="go">DNA</span>
</pre></div>
</div>
<p>You might use the Bio.SeqIO.convert() function to convert the (trimmed) SFF
reads into a FASTQ file (or a FASTA file and a QUAL file), e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out_handle</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">count</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">,</span>
<span class="gp">... </span>                      <span class="n">out_handle</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Converted </span><span class="si">%i</span><span class="s2"> records&quot;</span> <span class="o">%</span> <span class="n">count</span><span class="p">)</span>
<span class="go">Converted 10 records</span>
</pre></div>
</div>
<p>The output FASTQ file would start like this:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="n">out_handle</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()[:</span><span class="mi">50</span><span class="p">])</span>
<span class="go">@E3MFGYR02JWQ7T</span>
<span class="go">tcagGGTCTACATGTTGGTTAACCCGTACTGATT...</span>
</pre></div>
</div>
<p>Bio.SeqIO.index() provides memory efficient random access to the reads in an
SFF file by name. SFF files can include an index within the file, which can
be read in making this very fast. If the index is missing (or in a format not
yet supported in Biopython) the file is indexed by scanning all the reads -
which is a little slower. For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="s2">&quot;E3MFGYR02JHD4H&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2"> </span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">20</span><span class="p">]))</span>
<span class="go">E3MFGYR02JHD4H 310 tcagAAAGACAAGTGGTATC...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Or, using the trimmed reads:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff-trim&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">reads</span><span class="p">[</span><span class="s2">&quot;E3MFGYR02JHD4H&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2"> </span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">20</span><span class="p">]))</span>
<span class="go">E3MFGYR02JHD4H 292 AAAGACAAGTGGTATCAACG...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>You can also use the Bio.SeqIO.write() function with the “sff” format. Note
that this requires all the flow information etc, and thus is probably only
useful for SeqRecord objects originally from reading another SFF file (and
not the trimmed SeqRecord objects from parsing an SFF file as “sff-trim”).</p>
<p>As an example, let’s pretend this example SFF file represents some DNA which
was pre-amplified with a PCR primers AAAGANNNNN. The following script would
produce a sub-file containing all those reads whose post-quality clipping
region (i.e. the sequence after trimming) starts with AAAGA exactly (the non-
degenerate bit of this pretend primer):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="p">(</span><span class="n">record</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span>
<span class="gp">... </span>           <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">)</span>
<span class="gp">... </span>           <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;clip_qual_left&quot;</span><span class="p">]:]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;AAAGA&quot;</span><span class="p">))</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">count</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="s2">&quot;temp_filtered.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Selected </span><span class="si">%i</span><span class="s2"> records&quot;</span> <span class="o">%</span> <span class="n">count</span><span class="p">)</span>
<span class="go">Selected 2 records</span>
</pre></div>
</div>
<p>Of course, for an assembly you would probably want to remove these primers.
If you want FASTA or FASTQ output, you could just slice the SeqRecord. However,
if you want SFF output we have to preserve all the flow information - the trick
is just to adjust the left clip position!</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">filter_and_trim</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">primer</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;clip_qual_left&quot;</span><span class="p">]:]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">primer</span><span class="p">):</span>
<span class="gp">... </span>            <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;clip_qual_left&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">primer</span><span class="p">)</span>
<span class="gp">... </span>            <span class="k">yield</span> <span class="n">record</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">count</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">filter_and_trim</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="s2">&quot;AAAGA&quot;</span><span class="p">),</span>
<span class="gp">... </span>                    <span class="s2">&quot;temp_filtered.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Selected </span><span class="si">%i</span><span class="s2"> records&quot;</span> <span class="o">%</span> <span class="n">count</span><span class="p">)</span>
<span class="go">Selected 2 records</span>
</pre></div>
</div>
<p>We can check the results, note the lower case clipped region now includes the “AAAGA”
sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;temp_filtered.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2"> </span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">20</span><span class="p">]))</span>
<span class="gp">...</span>
<span class="go">E3MFGYR02JHD4H 310 tcagaaagaCAAGTGGTATC...</span>
<span class="go">E3MFGYR02GAZMS 278 tcagaaagaAGTAAGGTAAA...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;temp_filtered.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff-trim&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2"> </span><span class="si">%s</span><span class="s2">...&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">20</span><span class="p">]))</span>
<span class="gp">...</span>
<span class="go">E3MFGYR02JHD4H 287 CAAGTGGTATCAACGCAGAG...</span>
<span class="go">E3MFGYR02GAZMS 266 AGTAAGGTAAATAACAAACG...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s2">&quot;temp_filtered.sff&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>For a description of the file format, please see the Roche manuals and:
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/Traces/trace.cgi?cmd=show&amp;f=formats&amp;m=doc&amp;s=formats">http://www.ncbi.nlm.nih.gov/Traces/trace.cgi?cmd=show&amp;f=formats&amp;m=doc&amp;s=formats</a></p>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.ReadRocheXmlManifest">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.SffIO.</span></span><span class="sig-name descname"><span class="pre">ReadRocheXmlManifest</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.ReadRocheXmlManifest" title="Permalink to this definition"></a></dt>
<dd><p>Read any Roche style XML manifest data in the SFF “index”.</p>
<p>The SFF file format allows for multiple different index blocks, and Roche
took advantage of this to define their own index block which also embeds
an XML manifest string. This is not a publicly documented extension to
the SFF file format, this was reverse engineered.</p>
<p>The handle should be to an SFF file opened in binary mode. This function
will use the handle seek/tell functions and leave the handle in an
arbitrary location.</p>
<p>Any XML manifest found is returned as a Python string, which you can then
parse as appropriate, or reuse when writing out SFF files with the
SffWriter class.</p>
<p>Returns a string, or raises a ValueError if an Roche manifest could not be
found.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.SffIO.</span></span><span class="sig-name descname"><span class="pre">SffIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceIterator" title="Bio.SeqIO.Interfaces.SequenceIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceIterator</span></code></a></p>
<p>Parser for Standard Flowgram Format (SFF) files.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffIterator.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over Standard Flowgram Format (SFF) reads (as SeqRecord objects).</p>
<blockquote>
<div><ul class="simple">
<li><p>source - path to an SFF file, e.g. from Roche 454 sequencing,
or a file-like object opened in binary mode.</p></li>
<li><p>alphabet - optional alphabet, unused. Leave as None.</p></li>
<li><p>trim - should the sequences be trimmed?</p></li>
</ul>
</div></blockquote>
<p>The resulting SeqRecord objects should match those from a paired FASTA
and QUAL file converted from the SFF file using the Roche 454 tool
ssfinfo. i.e. The sequence will be mixed case, with the trim regions
shown in lower case.</p>
<p>This function is used internally via the Bio.SeqIO functions:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;sff&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)))</span>
<span class="gp">...</span>
<span class="go">E3MFGYR02JWQ7T 265</span>
<span class="go">E3MFGYR02JA6IL 271</span>
<span class="go">E3MFGYR02JHD4H 310</span>
<span class="go">E3MFGYR02GFKUC 299</span>
<span class="go">E3MFGYR02FTGED 281</span>
<span class="go">E3MFGYR02FR9G7 261</span>
<span class="go">E3MFGYR02GAZMS 278</span>
<span class="go">E3MFGYR02HHZ8O 221</span>
<span class="go">E3MFGYR02GPGB1 269</span>
<span class="go">E3MFGYR02F7Z7G 219</span>
</pre></div>
</div>
<p>You can also call it directly:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SffIterator</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)))</span>
<span class="gp">...</span>
<span class="go">E3MFGYR02JWQ7T 265</span>
<span class="go">E3MFGYR02JA6IL 271</span>
<span class="go">E3MFGYR02JHD4H 310</span>
<span class="go">E3MFGYR02GFKUC 299</span>
<span class="go">E3MFGYR02FTGED 281</span>
<span class="go">E3MFGYR02FR9G7 261</span>
<span class="go">E3MFGYR02GAZMS 278</span>
<span class="go">E3MFGYR02HHZ8O 221</span>
<span class="go">E3MFGYR02GPGB1 269</span>
<span class="go">E3MFGYR02F7Z7G 219</span>
</pre></div>
</div>
<p>Or, with the trim option:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Roche/E3MFGYR02_random_10_reads.sff&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SffIterator</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)))</span>
<span class="gp">...</span>
<span class="go">E3MFGYR02JWQ7T 260</span>
<span class="go">E3MFGYR02JA6IL 265</span>
<span class="go">E3MFGYR02JHD4H 292</span>
<span class="go">E3MFGYR02GFKUC 295</span>
<span class="go">E3MFGYR02FTGED 277</span>
<span class="go">E3MFGYR02FR9G7 256</span>
<span class="go">E3MFGYR02GAZMS 271</span>
<span class="go">E3MFGYR02HHZ8O 150</span>
<span class="go">E3MFGYR02GPGB1 221</span>
<span class="go">E3MFGYR02F7Z7G 130</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffIterator.parse">
<span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffIterator.parse" title="Permalink to this definition"></a></dt>
<dd><p>Start parsing the file, and return a SeqRecord generator.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffIterator.iterate">
<span class="sig-name descname"><span class="pre">iterate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffIterator.iterate" title="Permalink to this definition"></a></dt>
<dd><p>Parse the file and generate SeqRecord objects.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.SeqIO.SffIO.SffIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffIterator.__parameters__">
<span class="sig-name descname"><span class="pre">__parameters__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.SeqIO.SffIO.SffIterator.__parameters__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.SffIO.</span></span><span class="sig-name descname"><span class="pre">SffWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xml</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffWriter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceWriter" title="Bio.SeqIO.Interfaces.SequenceWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceWriter</span></code></a></p>
<p>SFF file writer.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffWriter.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xml</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffWriter.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize an SFF writer object.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>target - Output stream opened in binary mode, or a path to a file.</p></li>
<li><p>index - Boolean argument, should we try and write an index?</p></li>
<li><p>xml - Optional string argument, xml manifest to be recorded
in the index block (see function ReadRocheXmlManifest for
reading this data).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffWriter.write_file">
<span class="sig-name descname"><span class="pre">write_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffWriter.write_file" title="Permalink to this definition"></a></dt>
<dd><p>Use this to write an entire file containing the given records.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffWriter.write_header">
<span class="sig-name descname"><span class="pre">write_header</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffWriter.write_header" title="Permalink to this definition"></a></dt>
<dd><p>Write the SFF file header.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SffIO.SffWriter.write_record">
<span class="sig-name descname"><span class="pre">write_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SffIO.SffWriter.write_record" title="Permalink to this definition"></a></dt>
<dd><p>Write a single additional record to the output file.</p>
<p>This assumes the header has been done.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.SeqIO.SeqXmlIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.SeqXmlIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqIO.SnapGeneIO.html" class="btn btn-neutral float-right" title="Bio.SeqIO.SnapGeneIO module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>