

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.SearchIO.BlatIO module &mdash; Biopython 1.77.dev0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SearchIO.FastaIO module" href="Bio.SearchIO.FastaIO.html" />
    <link rel="prev" title="Bio.SearchIO.InterproscanIO.interproscan_xml module" href="Bio.SearchIO.InterproscanIO.interproscan_xml.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.77.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html#subpackages">Subpackages</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SearchIO.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html#module-Bio.SearchIO">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.SearchIO.html">Bio.SearchIO package</a> &raquo;</li>
        
      <li>Bio.SearchIO.BlatIO module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SearchIO/BlatIO.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SearchIO.FastaIO.html" class="btn btn-neutral float-right" title="Bio.SearchIO.FastaIO module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SearchIO.InterproscanIO.interproscan_xml.html" class="btn btn-neutral float-left" title="Bio.SearchIO.InterproscanIO.interproscan_xml module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.SearchIO.BlatIO">
<span id="bio-searchio-blatio-module"></span><h1>Bio.SearchIO.BlatIO module<a class="headerlink" href="#module-Bio.SearchIO.BlatIO" title="Permalink to this headline">¶</a></h1>
<p>Bio.SearchIO parser for BLAT output formats.</p>
<p>This module adds support for parsing BLAT outputs. BLAT (BLAST-Like Alignment
Tool) is a sequence similarity search program initially built for annotating
the human genome.</p>
<p>Bio.SearchIO.BlastIO was tested using standalone BLAT version 34, psLayout
version 3. It should be able to parse psLayout version 4 without problems.</p>
<p>More information on BLAT is available from these sites:</p>
<blockquote>
<div><ul class="simple">
<li><p>Publication: <a class="reference external" href="http://genome.cshlp.org/content/12/4/656">http://genome.cshlp.org/content/12/4/656</a></p></li>
<li><p>User guide: <a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/blatSpec.html">http://genome.ucsc.edu/goldenPath/help/blatSpec.html</a></p></li>
<li><p>Source download: <a class="reference external" href="http://www.soe.ucsc.edu/~kent/src">http://www.soe.ucsc.edu/~kent/src</a></p></li>
<li><p>Executable download: <a class="reference external" href="http://hgdownload.cse.ucsc.edu/admin/exe/">http://hgdownload.cse.ucsc.edu/admin/exe/</a></p></li>
<li><p>Blat score calculation: <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQblat.html#blat4">http://genome.ucsc.edu/FAQ/FAQblat.html#blat4</a></p></li>
</ul>
</div></blockquote>
<div class="section" id="supported-formats">
<h2>Supported Formats<a class="headerlink" href="#supported-formats" title="Permalink to this headline">¶</a></h2>
<p>BlatIO supports parsing, indexing, and writing for both PSL and PSLX output
formats, with or without header. To parse, index, or write PSLX files, use the
‘pslx’ keyword argument and set it to True.</p>
<blockquote>
<div><p># blat-psl defaults to PSL files
&gt;&gt;&gt; from Bio import SearchIO
&gt;&gt;&gt; psl = ‘Blat/psl_34_004.psl’
&gt;&gt;&gt; qresult = SearchIO.read(psl, ‘blat-psl’)
&gt;&gt;&gt; qresult
QueryResult(id=’hg19_dna’, 10 hits)</p>
<p># set the pslx flag to parse PSLX files
&gt;&gt;&gt; pslx = ‘Blat/pslx_34_004.pslx’
&gt;&gt;&gt; qresult = SearchIO.read(pslx, ‘blat-psl’, pslx=True)
&gt;&gt;&gt; qresult
QueryResult(id=’hg19_dna’, 10 hits)</p>
</div></blockquote>
<p>For parsing and indexing, you do not need to specify whether the file has a
header or not. For writing, if you want to write a header, you can set the
‘header’ keyword argument to True. This will write a ‘psLayout version 3’ header
to your output file.</p>
<blockquote>
<div><p>from Bio import SearchIO
qresult = SearchIO.read(psl, ‘blat-psl’)
SearchIO.write(qresult, ‘header.psl’, header=True)
&lt;stdout&gt; (1, 10, 19, 23)</p>
</div></blockquote>
<p>Note that the number of HSPFragments written may exceed the number of HSP
objects. This is because in PSL files, it is possible to have single matches
consisting of noncontiguous sequence fragments. This is where the HSPFragment
object comes into play. These fragments are grouped into a single HSP because
they share the same statistics (e.g. match numbers, BLAT score, etc.). However,
they do not share the same sequence attributes, such as the start and end
coordinates, making them distinct objects.</p>
<p>In addition to parsing PSL(X) files, BlatIO also computes the percent identities
and scores of your search results. This is done using the calculation formula
posted here: <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQblat.html#blat4">http://genome.ucsc.edu/FAQ/FAQblat.html#blat4</a>. It mimics the score
and percent identity calculation done by UCSC’s web BLAT service.</p>
<p>Since BlatIO parses the file in a single pass, it expects all results from
the same query to be in consecutive rows. If the results from one query are
spread in nonconsecutive rows, BlatIO will consider them to be separate
QueryResult objects.</p>
<p>In most cases, the PSL(X) format uses the same coordinate system as Python
(zero-based, half open). These coordinates are anchored on the plus strand.
However, if the query aligns on the minus strand, BLAT will anchor the qStarts
coordinates on the minus strand instead. BlatIO is aware of this, and will
re-anchor the qStarts coordinates to the plus strand whenever it sees a minus
strand query match. Conversely, when you write out to a PSL(X) file, BlatIO will
reanchor qStarts to the minus strand again.</p>
<p>BlatIO provides the following attribute-column mapping:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 21%" />
<col style="width: 33%" />
<col style="width: 46%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Object</p></th>
<th class="head"><p>Attribute</p></th>
<th class="head"><p>Column Name, Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td rowspan="2"><p>QueryResutl</p></td>
<td><p>id</p></td>
<td><p>Q name, query sequence ID</p></td>
</tr>
<tr class="row-odd"><td><p>seq_len</p></td>
<td><p>Q size, query sequence full
length</p></td>
</tr>
<tr class="row-even"><td rowspan="2"><p>Hit</p></td>
<td><p>id</p></td>
<td><p>T name, hit sequence ID</p></td>
</tr>
<tr class="row-odd"><td><p>seq_len</p></td>
<td><p>T size, hit sequence full length</p></td>
</tr>
<tr class="row-even"><td rowspan="17"><p>HSP</p></td>
<td><p>hit_end</p></td>
<td><p>T end, end coordinate of the last
hit fragment</p></td>
</tr>
<tr class="row-odd"><td><p>hit_gap_num</p></td>
<td><p>T gap bases, number of bases
inserted in hit</p></td>
</tr>
<tr class="row-even"><td><p>hit_gapopen_num</p></td>
<td><p>T gap count, number of hit gap
inserts</p></td>
</tr>
<tr class="row-odd"><td><p>hit_span_all</p></td>
<td><p>blockSizes, sizes of each
fragment</p></td>
</tr>
<tr class="row-even"><td><p>hit_start</p></td>
<td><p>T start, start coordinate of the
first hit fragment</p></td>
</tr>
<tr class="row-odd"><td><p>hit_start_all</p></td>
<td><p>tStarts, start coordinate of each
hit fragment</p></td>
</tr>
<tr class="row-even"><td><p>match_num</p></td>
<td><p>match, number of non-repeat
matches</p></td>
</tr>
<tr class="row-odd"><td><p>mismatch_num</p></td>
<td><p>mismatch, number of mismatches</p></td>
</tr>
<tr class="row-even"><td><p>match_rep_num</p></td>
<td><p>rep. match, number of matches
that are part of repeats</p></td>
</tr>
<tr class="row-odd"><td><p>n_num</p></td>
<td><p>N’s, number of N bases</p></td>
</tr>
<tr class="row-even"><td><p>query_end</p></td>
<td><p>Q end, end coordinate of the last</p></td>
</tr>
<tr class="row-odd"><td><p>query_gap_num</p></td>
<td><p>query fragment
Q gap bases, number of bases
inserted in query</p></td>
</tr>
<tr class="row-even"><td><p>query_gapopen_num</p></td>
<td><p>Q gap count, number of query gap
inserts</p></td>
</tr>
<tr class="row-odd"><td><p>query_span_all</p></td>
<td><p>blockSizes, sizes of each
fragment</p></td>
</tr>
<tr class="row-even"><td><p>query_start</p></td>
<td><p>Q start, start coordinate of the
first query block</p></td>
</tr>
<tr class="row-odd"><td><p>query_start_all</p></td>
<td><p>qStarts, start coordinate of each
query fragment</p></td>
</tr>
<tr class="row-even"><td><p>len <a class="footnote-reference brackets" href="#id2" id="id1">*</a></p></td>
<td><p>block count, the number of blocks
in the alignment</p></td>
</tr>
<tr class="row-odd"><td rowspan="4"><p>HSPFragment</p></td>
<td><p>hit</p></td>
<td><p>hit sequence, if present</p></td>
</tr>
<tr class="row-even"><td><p>hit_strand</p></td>
<td><p>strand, hit sequence strand</p></td>
</tr>
<tr class="row-odd"><td><p>query</p></td>
<td><p>query sequence, if present</p></td>
</tr>
<tr class="row-even"><td><p>query_strand</p></td>
<td><p>strand, query sequence strand</p></td>
</tr>
</tbody>
</table>
<p>In addition to the column mappings above, BlatIO also provides the following
object attributes:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 21%" />
<col style="width: 33%" />
<col style="width: 46%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Object</p></th>
<th class="head"><p>Attribute</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td rowspan="5"><p>HSP</p></td>
<td><p>gapopen_num</p></td>
<td><dl class="simple">
<dt>Q gap count + T gap count, total</dt><dd><p>number of gap openings</p>
</dd>
</dl>
</td>
</tr>
<tr class="row-odd"><td><p>ident_num</p></td>
<td><p>matches + repmatches, total
number of identical residues</p></td>
</tr>
<tr class="row-even"><td><p>ident_pct</p></td>
<td><p>percent identity, calculated
using UCSC’s formula</p></td>
</tr>
<tr class="row-odd"><td><p>query_is_protein</p></td>
<td><p>boolean, whether the query
sequence is a protein</p></td>
</tr>
<tr class="row-even"><td><p>score</p></td>
<td><p>HSP score, calculated using
UCSC’s formula</p></td>
</tr>
</tbody>
</table>
<p>Finally, the default HSP and HSPFragment properties are also provided. See the
HSP and HSPFragment documentation for more details on these properties.</p>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets"><a class="fn-backref" href="#id1">*</a></span></dt>
<dd><p>You can obtain the number of blocks / fragments in the HSP by invoking
<code class="docutils literal notranslate"><span class="pre">len</span></code> on the HSP</p>
</dd>
</dl>
<dl class="class">
<dt id="Bio.SearchIO.BlatIO.BlatPslParser">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SearchIO.BlatIO.</code><code class="sig-name descname">BlatPslParser</code><span class="sig-paren">(</span><em class="sig-param">handle</em>, <em class="sig-param">pslx=False</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SearchIO.BlatIO.BlatPslParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Parser for the BLAT PSL format.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.SearchIO.BlatIO.BlatPslIndexer">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SearchIO.BlatIO.</code><code class="sig-name descname">BlatPslIndexer</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">pslx=False</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SearchIO.BlatIO.BlatPslIndexer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SearchIO._index.SearchIndexer</span></code></p>
<p>Indexer class for BLAT PSL output.</p>
<dl class="method">
<dt id="Bio.SearchIO.BlatIO.BlatPslIndexer.get_raw">
<code class="sig-name descname">get_raw</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">offset</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SearchIO.BlatIO.BlatPslIndexer.get_raw" title="Permalink to this definition">¶</a></dt>
<dd><p>Return raw bytes string of a QueryResult object from the given offset.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.SearchIO.BlatIO.BlatPslWriter">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SearchIO.BlatIO.</code><code class="sig-name descname">BlatPslWriter</code><span class="sig-paren">(</span><em class="sig-param">handle</em>, <em class="sig-param">header=False</em>, <em class="sig-param">pslx=False</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SearchIO.BlatIO.BlatPslWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Writer for the blat-psl format.</p>
<dl class="method">
<dt id="Bio.SearchIO.BlatIO.BlatPslWriter.write_file">
<code class="sig-name descname">write_file</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">qresults</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SearchIO.BlatIO.BlatPslWriter.write_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Write query results to file.</p>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.SearchIO.FastaIO.html" class="btn btn-neutral float-right" title="Bio.SearchIO.FastaIO module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SearchIO.InterproscanIO.interproscan_xml.html" class="btn btn-neutral float-left" title="Bio.SearchIO.InterproscanIO.interproscan_xml module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2017, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>