<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.PopGen.GenePop.FileParser module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.PopGen.GenePop.LargeFileParser module" href="Bio.PopGen.GenePop.LargeFileParser.html" />
    <link rel="prev" title="Bio.PopGen.GenePop.EasyController module" href="Bio.PopGen.GenePop.EasyController.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PopGen.html">Bio.PopGen package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PopGen.GenePop.html">Bio.PopGen.GenePop package</a></li>
      <li class="breadcrumb-item active">Bio.PopGen.GenePop.FileParser module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/PopGen/GenePop/FileParser.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.PopGen.GenePop.EasyController.html" class="btn btn-neutral float-left" title="Bio.PopGen.GenePop.EasyController module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PopGen.GenePop.LargeFileParser.html" class="btn btn-neutral float-right" title="Bio.PopGen.GenePop.LargeFileParser module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.PopGen.GenePop.FileParser">
<span id="bio-popgen-genepop-fileparser-module"></span><h1>Bio.PopGen.GenePop.FileParser module<a class="headerlink" href="#module-Bio.PopGen.GenePop.FileParser" title="Permalink to this headline"></a></h1>
<p>Code to parse BIG GenePop files.</p>
<p>The difference between this class and the standard Bio.PopGen.GenePop.Record
class is that this one does not read the whole file to memory.
It provides an iterator interface, slower but consuming much mess memory.
Should be used with big files (Thousands of markers and individuals).</p>
<p>See <a class="reference external" href="http://wbiomed.curtin.edu.au/genepop/">http://wbiomed.curtin.edu.au/genepop/</a> , the format is documented
here: <a class="reference external" href="http://wbiomed.curtin.edu.au/genepop/help_input.html">http://wbiomed.curtin.edu.au/genepop/help_input.html</a> .</p>
<dl class="simple">
<dt>Classes:</dt><dd><ul class="simple">
<li><p>FileRecord           Holds GenePop data.</p></li>
</ul>
</dd>
</dl>
<p>Functions:</p>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.read">
<span class="sig-prename descclassname"><span class="pre">Bio.PopGen.GenePop.FileParser.</span></span><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.read" title="Permalink to this definition"></a></dt>
<dd><p>Parse a file containing a GenePop file.</p>
<p>fname is a file name that contains a GenePop record.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PopGen.GenePop.FileParser.</span></span><span class="sig-name descname"><span class="pre">FileRecord</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Hold information from a GenePop record.</p>
<p>Attributes:
- marker_len         The marker length (2 or 3 digit code per allele).
- comment_line       Comment line.
- loci_list          List of loci names.</p>
<p>Methods:
- get_individual     Returns the next individual of the current population.
- skip_population    Skips the current population.</p>
<p>skip_population skips the individuals of the current population, returns
True if there are more populations.</p>
<p>get_individual returns an individual of the current population (or None
if the list ended).</p>
<p>Each individual is a pair composed by individual name and a list of alleles
(2 per marker or 1 for haploid data). Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="s1">&#39;Ind1&#39;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span> <span class="p">(</span><span class="mi">200</span><span class="p">,</span><span class="mi">201</span><span class="p">)]</span>
<span class="p">(</span><span class="s1">&#39;Ind2&#39;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">2</span><span class="p">,</span><span class="kc">None</span><span class="p">),</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span><span class="kc">None</span><span class="p">)]</span>
<span class="p">(</span><span class="s1">&#39;Other1&#39;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span>  <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span> <span class="p">(</span><span class="mi">200</span><span class="p">,</span><span class="mi">200</span><span class="p">)]</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Return (reconstructs) a GenePop textual representation.</p>
<p>This might take a lot of memory.
Marker length will be 3.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.start_read">
<span class="sig-name descname"><span class="pre">start_read</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.start_read" title="Permalink to this definition"></a></dt>
<dd><p>Start parsing a file containing a GenePop file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.skip_header">
<span class="sig-name descname"><span class="pre">skip_header</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.skip_header" title="Permalink to this definition"></a></dt>
<dd><p>Skip the Header. To be done after a re-open.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.seek_position">
<span class="sig-name descname"><span class="pre">seek_position</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indiv</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.seek_position" title="Permalink to this definition"></a></dt>
<dd><p>Seek a certain position in the file.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>pop - pop position (0 is first)</p></li>
<li><p>indiv - individual in pop</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.skip_population">
<span class="sig-name descname"><span class="pre">skip_population</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.skip_population" title="Permalink to this definition"></a></dt>
<dd><p>Skip the current population. Returns true if there is another pop.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.get_individual">
<span class="sig-name descname"><span class="pre">get_individual</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.get_individual" title="Permalink to this definition"></a></dt>
<dd><p>Get the next individual.</p>
<p>Returns individual information if there are more individuals
in the current population.
Returns True if there are no more individuals in the current
population, but there are more populations. Next read will
be of the following pop.
Returns False if at end of file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.remove_population">
<span class="sig-name descname"><span class="pre">remove_population</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.remove_population" title="Permalink to this definition"></a></dt>
<dd><p>Remove a population (by position).</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>pos - position</p></li>
<li><p>fname - file to be created with population removed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.remove_locus_by_position">
<span class="sig-name descname"><span class="pre">remove_locus_by_position</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.remove_locus_by_position" title="Permalink to this definition"></a></dt>
<dd><p>Remove a locus by position.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>pos - position</p></li>
<li><p>fname - file to be created with locus removed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.remove_loci_by_position">
<span class="sig-name descname"><span class="pre">remove_loci_by_position</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">positions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.remove_loci_by_position" title="Permalink to this definition"></a></dt>
<dd><p>Remove a set of loci by position.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>positions - positions</p></li>
<li><p>fname - file to be created with locus removed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.remove_locus_by_name">
<span class="sig-name descname"><span class="pre">remove_locus_by_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.remove_locus_by_name" title="Permalink to this definition"></a></dt>
<dd><p>Remove a locus by name.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>name - name</p></li>
<li><p>fname - file to be created with locus removed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.FileParser.FileRecord.remove_loci_by_name">
<span class="sig-name descname"><span class="pre">remove_loci_by_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.FileParser.FileRecord.remove_loci_by_name" title="Permalink to this definition"></a></dt>
<dd><p>Remove a loci list (by name).</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>names - names</p></li>
<li><p>fname - file to be created with loci removed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.PopGen.GenePop.EasyController.html" class="btn btn-neutral float-left" title="Bio.PopGen.GenePop.EasyController module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PopGen.GenePop.LargeFileParser.html" class="btn btn-neutral float-right" title="Bio.PopGen.GenePop.LargeFileParser module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>