<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.motifs.jaspar.db module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.motifs.alignace module" href="Bio.motifs.alignace.html" />
    <link rel="prev" title="Bio.motifs.jaspar package" href="Bio.motifs.jaspar.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.motifs.html">Bio.motifs package</a></li>
          <li class="breadcrumb-item"><a href="Bio.motifs.jaspar.html">Bio.motifs.jaspar package</a></li>
      <li class="breadcrumb-item active">Bio.motifs.jaspar.db module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/motifs/jaspar/db.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.motifs.jaspar.html" class="btn btn-neutral float-left" title="Bio.motifs.jaspar package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.motifs.alignace.html" class="btn btn-neutral float-right" title="Bio.motifs.alignace module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.motifs.jaspar.db">
<span id="bio-motifs-jaspar-db-module"></span><h1>Bio.motifs.jaspar.db module<a class="headerlink" href="#module-Bio.motifs.jaspar.db" title="Permalink to this headline">ÔÉÅ</a></h1>
<p>Provides read access to a JASPAR5 formatted database.</p>
<p>This modules requires MySQLdb to be installed.</p>
<p>Example, substitute the your database credentials as
appropriate:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    from Bio.motifs.jaspar.db import JASPAR5
    JASPAR_DB_HOST = &quot;hostname.example.org&quot;
    JASPAR_DB_NAME = &quot;JASPAR2018&quot;
    JASPAR_DB_USER = &quot;guest&quot;
    JASPAR_DB_PASS = &quot;guest&quot;

    jdb = JASPAR5(
        host=JASPAR_DB_HOST,
        name=JASPAR_DB_NAME,
        user=JASPAR_DB_USER,
        password=JASPAR_DB_PASS
    )
    ets1 = jdb.fetch_motif_by_id(&#39;MA0098&#39;)
    print(ets1)
TF name ETS1
Matrix ID   MA0098.3
Collection  CORE
TF class    [&#39;Tryptophan cluster factors&#39;]
TF family   [&#39;Ets-related factors&#39;]
Species 9606
Taxonomic group vertebrates
Accession   [&#39;P14921&#39;]
Data type used  HT-SELEX
Medline 20517297
PAZAR ID    TF0000070
Comments    Data is from Taipale HTSELEX DBD (2013)
Matrix:
        0      1      2      3      4      5      6      7      8      9
A: 2683.00 180.00 425.00   0.00   0.00 2683.00 2683.00 1102.00  89.00 803.00
C: 210.00 2683.00 2683.00  21.00   0.00   0.00   9.00  21.00 712.00 401.00
G: 640.00 297.00   7.00 2683.00 2683.00   0.00  31.00 1580.00 124.00 1083.00
T: 241.00  22.00   0.00   0.00  12.00   0.00 909.00  12.00 1970.00 396.00

    motifs = jdb.fetch_motifs(
        collection = &#39;CORE&#39;,
        tax_group = [&#39;vertebrates&#39;, &#39;insects&#39;],
        tf_class = &#39;Homeo domain factors&#39;,
        tf_family = [&#39;TALE-type homeo domain factors&#39;, &#39;POU domain factors&#39;],
        min_ic = 12
    )
    for motif in motifs:
        pass # do something with the motif
</pre></div>
</div>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.motifs.jaspar.db.JASPAR5">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.motifs.jaspar.db.</span></span><span class="sig-name descname"><span class="pre">JASPAR5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">password</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.jaspar.db.JASPAR5" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class representing a JASPAR5 database.</p>
<p>Class representing a JASPAR5 DB. The methods within are loosely based
on the perl TFBS::DB::JASPAR5 module.</p>
<p>Note: We will only implement reading of JASPAR motifs from the DB.
Unlike the perl module, we will not attempt to implement any methods to
store JASPAR motifs or create a new DB at this time.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.jaspar.db.JASPAR5.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">host</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">password</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.jaspar.db.JASPAR5.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Construct a JASPAR5 instance and connect to specified DB.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>host - host name of the the JASPAR DB server</p></li>
<li><p>name - name of the JASPAR database</p></li>
<li><p>user - user name to connect to the JASPAR DB</p></li>
<li><p>password - JASPAR DB password</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.jaspar.db.JASPAR5.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.jaspar.db.JASPAR5.__str__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return a string represention of the JASPAR5 DB connection.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.jaspar.db.JASPAR5.fetch_motif_by_id">
<span class="sig-name descname"><span class="pre">fetch_motif_by_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.jaspar.db.JASPAR5.fetch_motif_by_id" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Fetch a single JASPAR motif from the DB by its JASPAR matrix ID.</p>
<p>Example id ‚ÄòMA0001.1‚Äô.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>id - JASPAR matrix ID. This may be a fully specified ID including</dt><dd><p>the version number (e.g. MA0049.2) or just the base ID (e.g.
MA0049). If only a base ID is provided, the latest version is
returned.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><ul class="simple">
<li><p>A Bio.motifs.jaspar.Motif object</p></li>
</ul>
</dd>
</dl>
<p><strong>NOTE:</strong> The perl TFBS module allows you to specify the type of matrix
to return (PFM, PWM, ICM) but matrices are always stored in JASPAR as
PFMs so this does not really belong here. Once a PFM is fetched the
pwm() and pssm() methods can be called to return the normalized and
log-odds matrices.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.jaspar.db.JASPAR5.fetch_motifs_by_name">
<span class="sig-name descname"><span class="pre">fetch_motifs_by_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.jaspar.db.JASPAR5.fetch_motifs_by_name" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Fetch a list of JASPAR motifs from a JASPAR DB by the given TF name(s).</p>
<p>Arguments:
name - a single name or list of names
Returns:
A list of Bio.motifs.jaspar.Motif objects</p>
<p>Notes:
Names are not guaranteed to be unique. There may be more than one
motif with the same name. Therefore even if name specifies a single
name, a list of motifs is returned. This just calls
self.fetch_motifs(collection = None, tf_name = name).</p>
<p>This behaviour is different from the TFBS perl module‚Äôs
get_Matrix_by_name() method which always returns a single matrix,
issuing a warning message and returning the first matrix retrieved
in the case where multiple matrices have the same name.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.jaspar.db.JASPAR5.fetch_motifs">
<span class="sig-name descname"><span class="pre">fetch_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">collection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">JASPAR_DFLT_COLLECTION</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf_family</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tax_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pazar_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">medline</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_ic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_sites</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_versions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.jaspar.db.JASPAR5.fetch_motifs" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Fetch jaspar.Record (list) of motifs using selection criteria.</p>
<p>Arguments:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Except</span> <span class="n">where</span> <span class="n">obvious</span><span class="p">,</span> <span class="nb">all</span> <span class="n">selection</span> <span class="n">criteria</span> <span class="n">arguments</span> <span class="n">may</span> <span class="n">be</span>
<span class="n">specified</span> <span class="k">as</span> <span class="n">a</span> <span class="n">single</span> <span class="n">value</span> <span class="ow">or</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">values</span><span class="o">.</span> <span class="n">Motifs</span> <span class="n">must</span>
<span class="n">meet</span> <span class="n">ALL</span> <span class="n">the</span> <span class="n">specified</span> <span class="n">selection</span> <span class="n">criteria</span> <span class="n">to</span> <span class="n">be</span> <span class="n">returned</span> <span class="k">with</span>
<span class="n">the</span> <span class="n">precedent</span> <span class="n">exceptions</span> <span class="n">noted</span> <span class="n">below</span><span class="o">.</span>

<span class="nb">all</span>         <span class="o">-</span> <span class="n">Takes</span> <span class="n">precedent</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">other</span> <span class="n">selection</span> <span class="n">criteria</span><span class="o">.</span>
              <span class="n">Every</span> <span class="n">motif</span> <span class="ow">is</span> <span class="n">returned</span><span class="o">.</span> <span class="n">If</span> <span class="s1">&#39;all_versions&#39;</span> <span class="ow">is</span> <span class="n">also</span>
              <span class="n">specified</span><span class="p">,</span> <span class="nb">all</span> <span class="n">versions</span> <span class="n">of</span> <span class="n">every</span> <span class="n">motif</span> <span class="n">are</span> <span class="n">returned</span><span class="p">,</span>
              <span class="n">otherwise</span> <span class="n">just</span> <span class="n">the</span> <span class="n">latest</span> <span class="n">version</span> <span class="n">of</span> <span class="n">every</span> <span class="n">motif</span> <span class="ow">is</span>
              <span class="n">returned</span><span class="o">.</span>
<span class="n">matrix_id</span>   <span class="o">-</span> <span class="n">Takes</span> <span class="n">precedence</span> <span class="n">over</span> <span class="nb">all</span> <span class="n">other</span> <span class="n">selection</span> <span class="n">criteria</span>
              <span class="k">except</span> <span class="s1">&#39;all&#39;</span><span class="o">.</span>  <span class="n">Only</span> <span class="n">motifs</span> <span class="k">with</span> <span class="n">the</span> <span class="n">given</span> <span class="n">JASPAR</span>
              <span class="n">matrix</span> <span class="n">ID</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span> <span class="n">A</span> <span class="n">matrix</span> <span class="n">ID</span> <span class="n">may</span> <span class="n">be</span>
              <span class="n">specified</span> <span class="k">as</span> <span class="n">just</span> <span class="n">a</span> <span class="n">base</span> <span class="n">ID</span> <span class="ow">or</span> <span class="n">full</span> <span class="n">JASPAR</span> <span class="n">IDs</span>
              <span class="n">including</span> <span class="n">version</span> <span class="n">number</span><span class="o">.</span> <span class="n">If</span> <span class="n">only</span> <span class="n">a</span> <span class="n">base</span> <span class="n">ID</span> <span class="ow">is</span>
              <span class="n">provided</span> <span class="k">for</span> <span class="n">specific</span> <span class="n">motif</span><span class="p">(</span><span class="n">s</span><span class="p">),</span> <span class="n">then</span> <span class="n">just</span> <span class="n">the</span> <span class="n">latest</span>
              <span class="n">version</span> <span class="n">of</span> <span class="n">those</span> <span class="n">motif</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="n">are</span> <span class="n">returned</span> <span class="n">unless</span>
              <span class="s1">&#39;all_versions&#39;</span> <span class="ow">is</span> <span class="n">also</span> <span class="n">specified</span><span class="o">.</span>
<span class="n">collection</span>  <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="kn">from</span> <span class="nn">the</span> <span class="n">specified</span> <span class="n">JASPAR</span> <span class="n">collection</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
              <span class="n">are</span> <span class="n">returned</span><span class="o">.</span> <span class="n">NOTE</span> <span class="o">-</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">specified</span><span class="p">,</span> <span class="n">the</span> <span class="n">collection</span>
              <span class="n">defaults</span> <span class="n">to</span> <span class="n">CORE</span> <span class="k">for</span> <span class="nb">all</span> <span class="n">other</span> <span class="n">selection</span> <span class="n">criteria</span>
              <span class="k">except</span> <span class="s1">&#39;all&#39;</span> <span class="ow">and</span> <span class="s1">&#39;matrix_id&#39;</span><span class="o">.</span> <span class="n">To</span> <span class="n">apply</span> <span class="n">the</span> <span class="n">other</span>
              <span class="n">selection</span> <span class="n">criteria</span> <span class="n">across</span> <span class="nb">all</span> <span class="n">JASPAR</span> <span class="n">collections</span><span class="p">,</span>
              <span class="n">explicitly</span> <span class="nb">set</span> <span class="n">collection</span><span class="o">=</span><span class="kc">None</span><span class="o">.</span>
<span class="n">tf_name</span>     <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="k">with</span> <span class="n">the</span> <span class="n">given</span> <span class="n">name</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
<span class="n">tf_class</span>    <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="n">of</span> <span class="n">the</span> <span class="n">given</span> <span class="n">TF</span> <span class="n">class</span><span class="p">(</span><span class="n">es</span><span class="p">)</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
<span class="n">tf_family</span>   <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="kn">from</span> <span class="nn">the</span> <span class="n">given</span> <span class="n">TF</span> <span class="n">families</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
<span class="n">tax_group</span>   <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="n">belonging</span> <span class="n">to</span> <span class="n">the</span> <span class="n">given</span> <span class="n">taxonomic</span>
              <span class="n">supergroups</span> <span class="n">are</span> <span class="n">returned</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="s1">&#39;vertebrates&#39;</span><span class="p">,</span>
              <span class="s1">&#39;insects&#39;</span><span class="p">,</span> <span class="s1">&#39;nematodes&#39;</span> <span class="n">etc</span><span class="o">.</span><span class="p">)</span>
<span class="n">species</span>     <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="n">derived</span> <span class="kn">from</span> <span class="nn">the</span> <span class="n">given</span> <span class="n">species</span> <span class="n">are</span>
              <span class="n">returned</span><span class="o">.</span>  <span class="n">Species</span> <span class="n">are</span> <span class="n">specified</span> <span class="k">as</span> <span class="n">taxonomy</span> <span class="n">IDs</span><span class="o">.</span>
<span class="n">data_type</span>   <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="n">generated</span> <span class="k">with</span> <span class="n">the</span> <span class="n">given</span> <span class="n">data</span> <span class="nb">type</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span>
              <span class="p">(</span><span class="s1">&#39;ChIP-seq&#39;</span><span class="p">,</span> <span class="s1">&#39;PBM&#39;</span><span class="p">,</span> <span class="s1">&#39;SELEX&#39;</span> <span class="n">etc</span><span class="o">.</span><span class="p">)</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
              <span class="n">NOTE</span> <span class="o">-</span> <span class="n">must</span> <span class="n">match</span> <span class="n">exactly</span> <span class="k">as</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">database</span><span class="o">.</span>
<span class="n">pazar_id</span>    <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="k">with</span> <span class="n">the</span> <span class="n">given</span> <span class="n">PAZAR</span> <span class="n">TF</span> <span class="n">ID</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
<span class="n">medline</span>     <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="k">with</span> <span class="n">the</span> <span class="n">given</span> <span class="n">medline</span> <span class="p">(</span><span class="n">PubmMed</span> <span class="n">IDs</span><span class="p">)</span> <span class="n">are</span>
              <span class="n">returned</span><span class="o">.</span>
<span class="n">min_ic</span>      <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="n">whose</span> <span class="n">profile</span> <span class="n">matrices</span> <span class="n">have</span> <span class="n">at</span> <span class="n">least</span> <span class="n">this</span>
              <span class="n">information</span> <span class="n">content</span> <span class="p">(</span><span class="n">specificty</span><span class="p">)</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
<span class="n">min_length</span>  <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="n">whose</span> <span class="n">profiles</span> <span class="n">are</span> <span class="n">of</span> <span class="n">at</span> <span class="n">least</span> <span class="n">this</span>
              <span class="n">length</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
<span class="n">min_sites</span>   <span class="o">-</span> <span class="n">Only</span> <span class="n">motifs</span> <span class="n">compiled</span> <span class="kn">from</span> <span class="nn">at</span> <span class="n">least</span> <span class="n">these</span> <span class="n">many</span> <span class="n">binding</span>
              <span class="n">sites</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
<span class="n">all_versions</span><span class="o">-</span> <span class="n">Unless</span> <span class="n">specified</span><span class="p">,</span> <span class="n">just</span> <span class="n">the</span> <span class="n">latest</span> <span class="n">version</span> <span class="n">of</span> <span class="n">motifs</span>
              <span class="n">determined</span> <span class="n">by</span> <span class="n">the</span> <span class="n">other</span> <span class="n">selection</span> <span class="n">criteria</span> <span class="n">are</span>
              <span class="n">returned</span><span class="o">.</span> <span class="n">Otherwise</span> <span class="nb">all</span> <span class="n">versions</span> <span class="n">of</span> <span class="n">the</span> <span class="n">selected</span>
              <span class="n">motifs</span> <span class="n">are</span> <span class="n">returned</span><span class="o">.</span>
</pre></div>
</div>
<dl class="simple">
<dt>Returns:</dt><dd><ul class="simple">
<li><p>A Bio.motifs.jaspar.Record (list) of motifs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.motifs.jaspar.html" class="btn btn-neutral float-left" title="Bio.motifs.jaspar package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.motifs.alignace.html" class="btn btn-neutral float-right" title="Bio.motifs.alignace module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>