<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.KEGG.REST module &mdash; Biopython 1.80.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Medline package" href="Bio.Medline.html" />
    <link rel="prev" title="Bio.KEGG.Map package" href="Bio.KEGG.Map.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
            <a href="index.html">
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html#subpackages">Subpackages</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.KEGG.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html#module-Bio.KEGG">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
          <li><a href="Bio.KEGG.html">Bio.KEGG package</a> &raquo;</li>
      <li>Bio.KEGG.REST module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/KEGG/REST.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.KEGG.Map.html" class="btn btn-neutral float-left" title="Bio.KEGG.Map package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Medline.html" class="btn btn-neutral float-right" title="Bio.Medline package" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.KEGG.REST">
<span id="bio-kegg-rest-module"></span><h1>Bio.KEGG.REST module<a class="headerlink" href="#module-Bio.KEGG.REST" title="Permalink to this headline"></a></h1>
<p>Provides code to access the REST-style KEGG online API.</p>
<p>This module aims to make the KEGG online REST-style API easier to use. See:
<a class="reference external" href="http://www.kegg.jp/kegg/rest/keggapi.html">http://www.kegg.jp/kegg/rest/keggapi.html</a></p>
<p>The KEGG REST-style API provides simple access to a range of KEGG databases.
This works using simple URLs (which this module will construct for you),
with any errors indicated via HTTP error levels.</p>
<p>The functionality is somewhat similar to Biopython’s Bio.TogoWS and Bio.Entrez
modules.</p>
<p>Currently KEGG does not provide any usage guidelines (unlike the NCBI whose
requirements are reasonably clear). To avoid risking overloading the service,
Biopython will only allow three calls per second.</p>
<p>References:
Kanehisa, M. and Goto, S.; KEGG: Kyoto Encyclopedia of Genes and Genomes.
Nucleic Acids Res. 28, 29-34 (2000).</p>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.KEGG.REST.kegg_info">
<span class="sig-prename descclassname"><span class="pre">Bio.KEGG.REST.</span></span><span class="sig-name descname"><span class="pre">kegg_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.REST.kegg_info" title="Permalink to this definition"></a></dt>
<dd><p>KEGG info - Displays the current statistics of a given database.</p>
<p>db - database or organism (string)</p>
<p>The argument db can be a KEGG database name (e.g. ‘pathway’ or its
official abbreviation, ‘path’), or a KEGG organism code or T number
(e.g. ‘hsa’ or ‘T01001’ for human).</p>
<p>A valid list of organism codes and their T numbers can be obtained
via kegg_info(‘organism’) or <a class="reference external" href="http://rest.kegg.jp/list/organism">http://rest.kegg.jp/list/organism</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.KEGG.REST.kegg_list">
<span class="sig-prename descclassname"><span class="pre">Bio.KEGG.REST.</span></span><span class="sig-name descname"><span class="pre">kegg_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">org</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.REST.kegg_list" title="Permalink to this definition"></a></dt>
<dd><p>KEGG list - Entry list for database, or specified database entries.</p>
<p>db - database or organism (string)
org - optional organism (string), see below.</p>
<p>For the pathway and module databases the optional organism can be
used to restrict the results.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.KEGG.REST.kegg_find">
<span class="sig-prename descclassname"><span class="pre">Bio.KEGG.REST.</span></span><span class="sig-name descname"><span class="pre">kegg_find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">option</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.REST.kegg_find" title="Permalink to this definition"></a></dt>
<dd><p>KEGG find - Data search.</p>
<p>Finds entries with matching query keywords or other query data in
a given database.</p>
<p>db - database or organism (string)
query - search terms (string)
option - search option (string), see below.</p>
<p>For the compound and drug database, set option to the string ‘formula’,
‘exact_mass’ or ‘mol_weight’ to search on that field only. The
chemical formula search is a partial match irrespective of the order
of atoms given. The exact mass (or molecular weight) is checked by
rounding off to the same decimal place as the query data. A range of
values may also be specified with the minus(-) sign.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.KEGG.REST.kegg_get">
<span class="sig-prename descclassname"><span class="pre">Bio.KEGG.REST.</span></span><span class="sig-name descname"><span class="pre">kegg_get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbentries</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">option</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.REST.kegg_get" title="Permalink to this definition"></a></dt>
<dd><p>KEGG get - Data retrieval.</p>
<p>dbentries - Identifiers (single string, or list of strings), see below.
option - One of “aaseq”, “ntseq”, “mol”, “kcf”, “image”, “kgml” (string)</p>
<p>The input is limited up to 10 entries.
The input is limited to one pathway entry with the image or kgml option.
The input is limited to one compound/glycan/drug entry with the image option.</p>
<p>Returns a handle.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.KEGG.REST.kegg_conv">
<span class="sig-prename descclassname"><span class="pre">Bio.KEGG.REST.</span></span><span class="sig-name descname"><span class="pre">kegg_conv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">option</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.REST.kegg_conv" title="Permalink to this definition"></a></dt>
<dd><p>KEGG conv - convert KEGG identifiers to/from outside identifiers.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>target_db - Target database</p></li>
<li><p>source_db_or_dbentries - source database or database entries</p></li>
<li><p>option - Can be “turtle” or “n-triple” (string).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.KEGG.REST.kegg_link">
<span class="sig-prename descclassname"><span class="pre">Bio.KEGG.REST.</span></span><span class="sig-name descname"><span class="pre">kegg_link</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">option</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.REST.kegg_link" title="Permalink to this definition"></a></dt>
<dd><p>KEGG link - find related entries by using database cross-references.</p>
<p>target_db - Target database
source_db_or_dbentries - source database
option - Can be “turtle” or “n-triple” (string).</p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.KEGG.Map.html" class="btn btn-neutral float-left" title="Bio.KEGG.Map package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Medline.html" class="btn btn-neutral float-right" title="Bio.Medline package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2020, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.80.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.78/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>