<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.Sequencing.Applications package &mdash; Biopython 1.80.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Sequencing.Ace module" href="Bio.Sequencing.Ace.html" />
    <link rel="prev" title="Bio.Sequencing package" href="Bio.Sequencing.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
            <a href="index.html">
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Sequencing.html#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html#module-Bio.Sequencing">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
          <li><a href="Bio.Sequencing.html">Bio.Sequencing package</a> &raquo;</li>
      <li>Bio.Sequencing.Applications package</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Sequencing/Applications/__init__.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Sequencing.html" class="btn btn-neutral float-left" title="Bio.Sequencing package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Sequencing.Ace.html" class="btn btn-neutral float-right" title="Bio.Sequencing.Ace module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bio-sequencing-applications-package">
<h1>Bio.Sequencing.Applications package<a class="headerlink" href="#bio-sequencing-applications-package" title="Permalink to this headline"></a></h1>
<section id="module-Bio.Sequencing.Applications">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.Sequencing.Applications" title="Permalink to this headline"></a></h2>
<p>Sequencing related command line application wrappers (OBSOLETE).</p>
<p>We have decided to remove this module in future, and instead recommend
building your command and invoking it via the subprocess module directly.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaIndexCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">BwaIndexCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaIndexCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for Burrows Wheeler Aligner (BWA) index.</p>
<p>Index database sequences in the FASTA format, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ bwa index [-p prefix] [-a algoType] [-c] &lt;in.db.fasta&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a> for details.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">BwaIndexCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_genome</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference_genome.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index_cmd</span> <span class="o">=</span> <span class="n">BwaIndexCommandline</span><span class="p">(</span><span class="n">infile</span><span class="o">=</span><span class="n">reference_genome</span><span class="p">,</span> <span class="n">algorithm</span><span class="o">=</span><span class="s2">&quot;bwtsw&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">index_cmd</span><span class="p">)</span>
<span class="go">bwa index -a bwtsw /path/to/reference_genome.fasta</span>
</pre></div>
</div>
<p>You would typically run the command using index_cmd() or via the
Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaIndexCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaIndexCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaIndexCommandline.algorithm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">algorithm</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaIndexCommandline.algorithm" title="Permalink to this definition"></a></dt>
<dd><p>Algorithm for constructing BWT index.</p>
<blockquote>
<div><dl class="simple">
<dt>Available options are:</dt><dd><ul class="simple">
<li><p>is:    IS linear-time algorithm for constructing suffix array.
It requires 5.37N memory where N is the size of the database.
IS is moderately fast, but does not work with database larger
than 2GB. IS is the default algorithm due to its simplicity.</p></li>
<li><p>bwtsw: Algorithm implemented in BWT-SW. This method works with the
whole human genome, but it does not work with database
smaller than 10MB and it is usually slower than IS.</p></li>
</ul>
</dd>
</dl>
</div></blockquote>
<p>This controls the addition of the -a parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaIndexCommandline.c">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">c</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaIndexCommandline.c" title="Permalink to this definition"></a></dt>
<dd><p>Build color-space index. The input fasta should be in nucleotide space.</p>
<p>This property controls the addition of the -c switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaIndexCommandline.infile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">infile</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaIndexCommandline.infile" title="Permalink to this definition"></a></dt>
<dd><p>Input file name</p>
<p>This controls the addition of the infile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaIndexCommandline.prefix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prefix</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaIndexCommandline.prefix" title="Permalink to this definition"></a></dt>
<dd><p>Prefix of the output database [same as db filename]</p>
<p>This controls the addition of the -p parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">BwaAlignCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for Burrows Wheeler Aligner (BWA) aln.</p>
<p>Run a BWA alignment, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ bwa aln [...] &lt;in.db.fasta&gt; &lt;in.query.fq&gt; &gt; &lt;out.sai&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a> for details.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">BwaAlignCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_genome</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference_genome.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.fq&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_sai_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.sai&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align_cmd</span> <span class="o">=</span> <span class="n">BwaAlignCommandline</span><span class="p">(</span><span class="n">reference</span><span class="o">=</span><span class="n">reference_genome</span><span class="p">,</span> <span class="n">read_file</span><span class="o">=</span><span class="n">read_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align_cmd</span><span class="p">)</span>
<span class="go">bwa aln /path/to/reference_genome.fasta /path/to/read_1.fq</span>
</pre></div>
</div>
<p>You would typically run the command line using align_cmd(stdout=output_sai_file)
or via the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.B">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">B</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.B" title="Permalink to this definition"></a></dt>
<dd><p>Length of barcode starting from the 5-end. When INT is positive, the barcode of each read will be trimmed before mapping and will be written at the BC SAM tag. For paired-end reads, the barcode from both ends are concatenated. [0]</p>
<p>This controls the addition of the -B parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.E">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">E</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.E" title="Permalink to this definition"></a></dt>
<dd><p>Gap extension penalty [4]</p>
<p>This controls the addition of the -E parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.I">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">I</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.I" title="Permalink to this definition"></a></dt>
<dd><p>The input is in the Illumina 1.3+ read format (quality equals ASCII-64).</p>
<p>This property controls the addition of the -I switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.M">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">M</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.M" title="Permalink to this definition"></a></dt>
<dd><p>Mismatch penalty. BWA will not search for suboptimal hits with a score lower than (bestScore-misMsc). [3]</p>
<p>This controls the addition of the -M parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.N">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">N</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.N" title="Permalink to this definition"></a></dt>
<dd><p>Disable iterative search. All hits with no more than maxDiff differences will be found. This mode is much slower than the default.</p>
<p>This property controls the addition of the -N switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.O">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">O</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.O" title="Permalink to this definition"></a></dt>
<dd><p>Gap open penalty [11]</p>
<p>This controls the addition of the -O parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.R">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">R</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.R" title="Permalink to this definition"></a></dt>
<dd><p>Proceed with suboptimal alignments if there are no more than INT equally best hits.</p>
<blockquote>
<div><p>This option only affects paired-end mapping. Increasing this threshold helps
to improve the pairing accuracy at the cost of speed, especially for short
reads (~32bp).</p>
</div></blockquote>
<p>This controls the addition of the -R parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.b">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.b" title="Permalink to this definition"></a></dt>
<dd><p>Specify the input read sequence file is the BAM format</p>
<p>This property controls the addition of the -b switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.b1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b1</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.b1" title="Permalink to this definition"></a></dt>
<dd><p>When -b is specified, only use the first read in a read pair in mapping (skip single-end reads and the second reads).</p>
<p>This property controls the addition of the -b1 switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.b2">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b2</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.b2" title="Permalink to this definition"></a></dt>
<dd><p>When -b is specified, only use the second read in a read pair in mapping.</p>
<p>This property controls the addition of the -b2 switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.c">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">c</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.c" title="Permalink to this definition"></a></dt>
<dd><p>Reverse query but not complement it, which is required for alignment in the color space.</p>
<p>This property controls the addition of the -c switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.d">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">d</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.d" title="Permalink to this definition"></a></dt>
<dd><p>Disallow a long deletion within INT bp towards the 3-end [16]</p>
<p>This controls the addition of the -d parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.e">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">e</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.e" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of gap extensions, -1 for k-difference mode (disallowing long gaps) [-1]</p>
<p>This controls the addition of the -e parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.i">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">i</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.i" title="Permalink to this definition"></a></dt>
<dd><p>Disallow an indel within INT bp towards the ends [5]</p>
<p>This controls the addition of the -i parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.k">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">k</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.k" title="Permalink to this definition"></a></dt>
<dd><p>Maximum edit distance in the seed [2]</p>
<p>This controls the addition of the -k parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.l">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">l</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.l" title="Permalink to this definition"></a></dt>
<dd><p>Take the first INT subsequence as seed.</p>
<blockquote>
<div><p>If INT is larger than the query sequence, seeding will be disabled.
For long reads, this option is typically ranged from 25 to 35 for
-k 2. [inf]</p>
</div></blockquote>
<p>This controls the addition of the -l parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.n">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.n" title="Permalink to this definition"></a></dt>
<dd><p>Maximum edit distance if the value is INT, or the fraction of missing alignments given 2% uniform base error rate if FLOAT. In the latter case, the maximum edit distance is automatically chosen for different read lengths. [0.04]</p>
<p>This controls the addition of the -n parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.o">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">o</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.o" title="Permalink to this definition"></a></dt>
<dd><p>Maximum edit distance if the value is INT, or the fraction of missing alignments given 2% uniform base error rate if FLOAT. In the latter case, the maximum edit distance is automatically chosen for different read lengths. [0.04]</p>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">q</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.q" title="Permalink to this definition"></a></dt>
<dd><p>Parameter for read trimming [0].</p>
<blockquote>
<div><p>BWA trims a read down to argmax_x{sum_{i=x+1}^l(INT-q_i)} if q_l&lt;INT
where l is the original read length.</p>
</div></blockquote>
<p>This controls the addition of the -q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.read_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.read_file" title="Permalink to this definition"></a></dt>
<dd><p>Read file name</p>
<p>This controls the addition of the read_file parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.reference">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.reference" title="Permalink to this definition"></a></dt>
<dd><p>Reference file name</p>
<p>This controls the addition of the reference parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaAlignCommandline.t">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">t</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaAlignCommandline.t" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads (multi-threading mode) [1]</p>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSamseCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">BwaSamseCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSamseCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for Burrows Wheeler Aligner (BWA) samse.</p>
<p>Generate alignments in the SAM format given single-end reads.
Equvialent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ bwa samse [-n maxOcc] &lt;in.db.fasta&gt; &lt;in.sai&gt; &lt;in.fq&gt; &gt; &lt;out.sam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a> for details.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">BwaSamseCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_genome</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference_genome.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.fq&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sai_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.sai&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_sam_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.sam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samse_cmd</span> <span class="o">=</span> <span class="n">BwaSamseCommandline</span><span class="p">(</span><span class="n">reference</span><span class="o">=</span><span class="n">reference_genome</span><span class="p">,</span>
<span class="gp">... </span>                                <span class="n">read_file</span><span class="o">=</span><span class="n">read_file</span><span class="p">,</span> <span class="n">sai_file</span><span class="o">=</span><span class="n">sai_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samse_cmd</span><span class="p">)</span>
<span class="go">bwa samse /path/to/reference_genome.fasta /path/to/read_1.sai /path/to/read_1.fq</span>
</pre></div>
</div>
<p>You would typically run the command line using samse_cmd(stdout=output_sam_file)
or via the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSamseCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSamseCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSamseCommandline.n">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSamseCommandline.n" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of alignments to output in the XA tag for reads paired properly.</p>
<blockquote>
<div><p>If a read has more than INT hits, the XA tag will not be written. [3]</p>
</div></blockquote>
<p>This controls the addition of the -n parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSamseCommandline.r">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSamseCommandline.r" title="Permalink to this definition"></a></dt>
<dd><p>Specify the read group in a format like <a class="reference external" href="mailto:'&#37;&#52;&#48;RG">‘<span>&#64;</span>RG</a>    ID:foo  SM:bar’. [null]</p>
<p>This controls the addition of the -r parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSamseCommandline.read_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSamseCommandline.read_file" title="Permalink to this definition"></a></dt>
<dd><p>Read  file name</p>
<p>This controls the addition of the read_file parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSamseCommandline.reference">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSamseCommandline.reference" title="Permalink to this definition"></a></dt>
<dd><p>Reference file name</p>
<p>This controls the addition of the reference parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSamseCommandline.sai_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sai_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSamseCommandline.sai_file" title="Permalink to this definition"></a></dt>
<dd><p>Sai file name</p>
<p>This controls the addition of the sai_file parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">BwaSampeCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for Burrows Wheeler Aligner (BWA) sampe.</p>
<p>Generate alignments in the SAM format given paired-end reads.
Equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ bwa sampe [...] &lt;in.db.fasta&gt; &lt;in1.sai&gt; &lt;in2.sai&gt; &lt;in1.fq&gt; &lt;in2.fq&gt; &gt; &lt;out.sam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a> for details.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">BwaSampeCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_genome</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference_genome.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read_file1</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.fq&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read_file2</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_2.fq&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sai_file1</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.sai&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sai_file2</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_2.sai&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_sam_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/output.sam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read_group</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;@RG\tID:foo\tSM:bar&quot;</span>  <span class="c1"># BWA will turn backslash-t into tab</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sampe_cmd</span> <span class="o">=</span> <span class="n">BwaSampeCommandline</span><span class="p">(</span><span class="n">reference</span><span class="o">=</span><span class="n">reference_genome</span><span class="p">,</span>
<span class="gp">... </span>                                <span class="n">sai_file1</span><span class="o">=</span><span class="n">sai_file1</span><span class="p">,</span> <span class="n">sai_file2</span><span class="o">=</span><span class="n">sai_file2</span><span class="p">,</span>
<span class="gp">... </span>                                <span class="n">read_file1</span><span class="o">=</span><span class="n">read_file1</span><span class="p">,</span> <span class="n">read_file2</span><span class="o">=</span><span class="n">read_file2</span><span class="p">,</span>
<span class="gp">... </span>                                <span class="n">r</span><span class="o">=</span><span class="n">read_group</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sampe_cmd</span><span class="p">)</span>
<span class="go">bwa sampe /path/to/reference_genome.fasta /path/to/read_1.sai /path/to/read_2.sai /path/to/read_1.fq /path/to/read_2.fq -r @RG\tID:foo\tSM:bar</span>
</pre></div>
</div>
<p>You would typically run the command line using sampe_cmd(stdout=output_sam_file)
or via the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.N">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">N</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.N" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of alignments to output in the XA tag for disconcordant read pairs (excluding singletons) [10].</p>
<blockquote>
<div><p>If a read has more than INT hits, the XA tag will not be written.</p>
</div></blockquote>
<p>This controls the addition of the -N parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.a">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">a</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.a" title="Permalink to this definition"></a></dt>
<dd><p>Maximum insert size for a read pair to be considered being mapped properly [500].</p>
<blockquote>
<div><p>Since 0.4.5, this option is only used when there are not enough
good alignments to infer the distribution of insert sizes.</p>
</div></blockquote>
<p>This controls the addition of the -a parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.n">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.n" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of alignments to output in the XA tag for reads paired properly [3].</p>
<blockquote>
<div><p>If a read has more than INT hits, the XA tag will not be written.</p>
</div></blockquote>
<p>This controls the addition of the -n parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.o">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">o</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.o" title="Permalink to this definition"></a></dt>
<dd><p>Maximum occurrences of a read for pairing [100000].</p>
<blockquote>
<div><p>A read with more occurrences will be treated as a single-end read.
Reducing this parameter helps faster pairing.</p>
</div></blockquote>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.r">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.r" title="Permalink to this definition"></a></dt>
<dd><p>Specify the read group in a format like <a class="reference external" href="mailto:'&#37;&#52;&#48;RG">‘<span>&#64;</span>RG</a>    ID:foo  SM:bar’. [null]</p>
<p>This controls the addition of the -r parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.read_file1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_file1</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.read_file1" title="Permalink to this definition"></a></dt>
<dd><p>Read  file 1</p>
<p>This controls the addition of the read_file1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.read_file2">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_file2</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.read_file2" title="Permalink to this definition"></a></dt>
<dd><p>Read  file 2</p>
<p>This controls the addition of the read_file2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.reference">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.reference" title="Permalink to this definition"></a></dt>
<dd><p>Reference file name</p>
<p>This controls the addition of the reference parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.sai_file1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sai_file1</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.sai_file1" title="Permalink to this definition"></a></dt>
<dd><p>Sai file 1</p>
<p>This controls the addition of the sai_file1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaSampeCommandline.sai_file2">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sai_file2</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaSampeCommandline.sai_file2" title="Permalink to this definition"></a></dt>
<dd><p>Sai file 2</p>
<p>This controls the addition of the sai_file2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">BwaBwaswCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for Burrows Wheeler Aligner (BWA) bwasw.</p>
<p>Align query sequences from FASTQ files. Equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ bwa bwasw [...] &lt;in.db.fasta&gt; &lt;in.fq&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a> for details.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">BwaBwaswCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_genome</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference_genome.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.fq&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bwasw_cmd</span> <span class="o">=</span> <span class="n">BwaBwaswCommandline</span><span class="p">(</span><span class="n">reference</span><span class="o">=</span><span class="n">reference_genome</span><span class="p">,</span> <span class="n">read_file</span><span class="o">=</span><span class="n">read_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bwasw_cmd</span><span class="p">)</span>
<span class="go">bwa bwasw /path/to/reference_genome.fasta /path/to/read_1.fq</span>
</pre></div>
</div>
<p>You would typically run the command line using bwasw_cmd() or via the
Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.N">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">N</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.N" title="Permalink to this definition"></a></dt>
<dd><p>Minimum number of seeds supporting the resultant alignment to skip reverse alignment. [5]</p>
<p>This controls the addition of the -N parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.T">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">T</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.T" title="Permalink to this definition"></a></dt>
<dd><p>Minimum score threshold divided by a [37]</p>
<p>This controls the addition of the -T parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.a">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">a</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.a" title="Permalink to this definition"></a></dt>
<dd><p>Score of a match [1]</p>
<p>This controls the addition of the -a parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.b">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.b" title="Permalink to this definition"></a></dt>
<dd><p>Mismatch penalty [3]</p>
<p>This controls the addition of the -b parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.c">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">c</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.c" title="Permalink to this definition"></a></dt>
<dd><p>Coefficient for threshold adjustment according to query length [5.5].</p>
<blockquote>
<div><p>Given an l-long query, the threshold for a hit to be retained is
a*max{T,c*log(l)}.</p>
</div></blockquote>
<p>This controls the addition of the -c parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.mate_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mate_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.mate_file" title="Permalink to this definition"></a></dt>
<dd><p>Mate file</p>
<p>This controls the addition of the mate_file parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">q</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.q" title="Permalink to this definition"></a></dt>
<dd><p>Gap open penalty [5]</p>
<p>This controls the addition of the -q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.r">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.r" title="Permalink to this definition"></a></dt>
<dd><p>Gap extension penalty. The penalty for a contiguous gap of size k is q+k*r. [2]</p>
<p>This controls the addition of the -r parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.read_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.read_file" title="Permalink to this definition"></a></dt>
<dd><p>Read file</p>
<p>This controls the addition of the read_file parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.reference">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.reference" title="Permalink to this definition"></a></dt>
<dd><p>Reference file name</p>
<p>This controls the addition of the reference parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">s</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.s" title="Permalink to this definition"></a></dt>
<dd><p>Maximum SA interval size for initiating a seed [3].</p>
<blockquote>
<div><p>Higher -s increases accuracy at the cost of speed.</p>
</div></blockquote>
<p>This controls the addition of the -s parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.t">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">t</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.t" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads in the multi-threading mode [1]</p>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.w">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">w</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.w" title="Permalink to this definition"></a></dt>
<dd><p>Band width in the banded alignment [33]</p>
<p>This controls the addition of the -w parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaBwaswCommandline.z">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">z</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaBwaswCommandline.z" title="Permalink to this definition"></a></dt>
<dd><p>Z-best heuristics. Higher -z increases accuracy at the cost of speed. [1]</p>
<p>This controls the addition of the -z parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">BwaMemCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for Burrows Wheeler Aligner (BWA) mem.</p>
<p>Run a BWA-MEM alignment, with single- or paired-end reads, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ bwa mem [...] &lt;in.db.fasta&gt; &lt;in1.fq&gt; &lt;in2.fq&gt; &gt; &lt;out.sam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a> for details.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">BwaMemCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_genome</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference_genome.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/read_1.fq&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_sam_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/output.sam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align_cmd</span> <span class="o">=</span> <span class="n">BwaMemCommandline</span><span class="p">(</span><span class="n">reference</span><span class="o">=</span><span class="n">reference_genome</span><span class="p">,</span> <span class="n">read_file1</span><span class="o">=</span><span class="n">read_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align_cmd</span><span class="p">)</span>
<span class="go">bwa mem /path/to/reference_genome.fasta /path/to/read_1.fq</span>
</pre></div>
</div>
<p>You would typically run the command line using align_cmd(stdout=output_sam_file)
or via the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'bwa'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.A">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">A</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.A" title="Permalink to this definition"></a></dt>
<dd><p>Matching score. [1]</p>
<p>This controls the addition of the -A parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.B">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">B</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.B" title="Permalink to this definition"></a></dt>
<dd><p>Mismatch penalty. The sequence error rate is approximately: {.75 * exp[-log(4) * B/A]}. [4]</p>
<p>This controls the addition of the -B parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.C">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">C</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.C" title="Permalink to this definition"></a></dt>
<dd><p>Append FASTA/Q comment to SAM output. This option can be used to transfer read meta information (e.g. barcode) to the SAM output. Note that the FASTA/Q comment (the string after a space in the header line) must conform the SAM spec (e.g. BC:Z:CGTAC). Malformated comments lead to incorrect SAM output.</p>
<p>This property controls the addition of the -C switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.E">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">E</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.E" title="Permalink to this definition"></a></dt>
<dd><p>Gap extension penalty. A gap of length k costs O + k*E (i.e. -O is for opening a zero-length gap). [1]</p>
<p>This controls the addition of the -E parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.H">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">H</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.H" title="Permalink to this definition"></a></dt>
<dd><p>Use hard clipping ‘H’ in the SAM output. This option may dramatically reduce the redundancy of output when mapping long contig or BAC sequences.</p>
<p>This property controls the addition of the -H switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.L">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">L</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.L" title="Permalink to this definition"></a></dt>
<dd><p>Clipping penalty. When performing SW extension, BWA-MEM keeps track of the best score reaching the end of query. If this score is larger than the best SW score minus the clipping penalty, clipping will not be applied. Note that in this case, the SAM AS tag reports the best SW score; clipping penalty is not deducted. [5]</p>
<p>This controls the addition of the -L parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.M">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">M</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.M" title="Permalink to this definition"></a></dt>
<dd><p>Mark shorter split hits as secondary (for Picard compatibility).</p>
<p>This property controls the addition of the -M switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.O">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">O</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.O" title="Permalink to this definition"></a></dt>
<dd><p>Gap open penalty. [6]</p>
<p>This controls the addition of the -O parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.P">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">P</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.P" title="Permalink to this definition"></a></dt>
<dd><p>In the paired-end mode, perform SW to rescue missing hits only but do not try to find hits that fit a proper pair.</p>
<p>This property controls the addition of the -P switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.R">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">R</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.R" title="Permalink to this definition"></a></dt>
<dd><p>Complete read group header line. ‘t’ can be used in STR and will be converted to a TAB in the output SAM. The read group ID will be attached to every read in the output. An example is <a class="reference external" href="mailto:'&#37;&#52;&#48;RG">‘<span>&#64;</span>RG</a>    ID:foo  SM:bar’. [null]</p>
<p>This controls the addition of the -R parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.T">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">T</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.T" title="Permalink to this definition"></a></dt>
<dd><p>Don’t output alignment with score lower than INT. This option only affects output. [30]</p>
<p>This controls the addition of the -T parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.U">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">U</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.U" title="Permalink to this definition"></a></dt>
<dd><p>Penalty for an unpaired read pair. BWA-MEM scores an unpaired read pair as scoreRead1+scoreRead2-INT and scores a paired as scoreRead1+scoreRead2-insertPenalty. It compares these two scores to determine whether we should force pairing. [9]</p>
<p>This controls the addition of the -U parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.a">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">a</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.a" title="Permalink to this definition"></a></dt>
<dd><p>Output all found alignments for single-end or unpaired paired-end reads. These alignments will be flagged as secondary alignments.</p>
<p>This property controls the addition of the -a switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.c">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">c</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.c" title="Permalink to this definition"></a></dt>
<dd><p>Discard a MEM if it has more than INT occurrence in the genome. This is an insensitive parameter. [10000]</p>
<p>This controls the addition of the -c parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.d">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">d</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.d" title="Permalink to this definition"></a></dt>
<dd><p>Off-diagonal X-dropoff (Z-dropoff). Stop extension when the difference between the best and the current extension score is above |i-j|*A+INT, where i and j are the current positions of the query and reference, respectively, and A is the matching score. Z-dropoff is similar to BLAST’s X-dropoff except that it doesn’t penalize gaps in one of the sequences in the alignment. Z-dropoff not only avoids unnecessary extension, but also reduces poor alignments inside a long good alignment. [100]</p>
<p>This controls the addition of the -d parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.k">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">k</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.k" title="Permalink to this definition"></a></dt>
<dd><p>Minimum seed length. Matches shorter than INT will be missed. The alignment speed is usually insensitive to this value unless it significantly deviates 20. [19]</p>
<p>This controls the addition of the -k parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.p">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">p</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.p" title="Permalink to this definition"></a></dt>
<dd><p>Assume the first input query file is interleaved paired-end FASTA/Q. See the command description for details.</p>
<p>This property controls the addition of the -p switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.r">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.r" title="Permalink to this definition"></a></dt>
<dd><p>Trigger re-seeding for a MEM longer than minSeedLen*FLOAT. This is a key heuristic parameter for tuning the performance. Larger value yields fewer seeds, which leads to faster alignment speed but lower accuracy. [1.5]</p>
<p>This controls the addition of the -r parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.read_file1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_file1</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.read_file1" title="Permalink to this definition"></a></dt>
<dd><p>Read 1 file name</p>
<p>This controls the addition of the read_file1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.read_file2">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_file2</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.read_file2" title="Permalink to this definition"></a></dt>
<dd><p>Read 2 file name</p>
<p>This controls the addition of the read_file2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.reference">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.reference" title="Permalink to this definition"></a></dt>
<dd><p>Reference file name</p>
<p>This controls the addition of the reference parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.t">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">t</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.t" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads [1]</p>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.v">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">v</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.v" title="Permalink to this definition"></a></dt>
<dd><p>Control the verbose level of the output. This option has not been fully supported throughout BWA. Ideally, a value 0 for disabling all the output to stderr; 1 for outputting errors only; 2 for warnings and errors; 3 for all normal messages; 4 or higher for debugging. When this option takes value 4, the output is not SAM. [3]</p>
<p>This controls the addition of the -v parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.BwaMemCommandline.w">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">w</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.BwaMemCommandline.w" title="Permalink to this definition"></a></dt>
<dd><p>Band width. Essentially, gaps longer than INT will not be found. Note that the maximum gap length is also affected by the scoring matrix and the hit length, not solely determined by this option. [100]</p>
<p>This controls the addition of the -w parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">NovoalignCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'novoalign'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for novoalign by Novocraft.</p>
<p>See www.novocraft.com - novoalign is a short read alignment program.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">NovoalignCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">novoalign_cline</span> <span class="o">=</span> <span class="n">NovoalignCommandline</span><span class="p">(</span><span class="n">database</span><span class="o">=</span><span class="s1">&#39;some_db&#39;</span><span class="p">,</span>
<span class="gp">... </span>                                       <span class="n">readfile</span><span class="o">=</span><span class="s1">&#39;some_seq.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">novoalign_cline</span><span class="p">)</span>
<span class="go">novoalign -d some_db -f some_seq.txt</span>
</pre></div>
</div>
<p>As with all the Biopython application wrappers, you can also add or
change options after creating the object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">novoalign_cline</span><span class="o">.</span><span class="n">format</span> <span class="o">=</span> <span class="s1">&#39;PRBnSEQ&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">novoalign_cline</span><span class="o">.</span><span class="n">r_method</span><span class="o">=</span><span class="s1">&#39;0.99&#39;</span> <span class="c1"># limited valid values</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">novoalign_cline</span><span class="o">.</span><span class="n">fragment</span> <span class="o">=</span> <span class="s1">&#39;250 20&#39;</span> <span class="c1"># must be given as a string</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">novoalign_cline</span><span class="o">.</span><span class="n">miRNA</span> <span class="o">=</span> <span class="mi">100</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">novoalign_cline</span><span class="p">)</span>
<span class="go">novoalign -d some_db -f some_seq.txt -F PRBnSEQ -r 0.99 -i 250 20 -m 100</span>
</pre></div>
</div>
<p>You would typically run the command line with novoalign_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<p>Last checked against version: 2.05.04</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'novoalign'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.adapter3">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">adapter3</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.adapter3" title="Permalink to this definition"></a></dt>
<dd><p>Strips a 3’ adapter sequence prior to alignment.</p>
<p>With paired ends two adapters can be specified</p>
<p>This controls the addition of the -a parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.adapter5">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">adapter5</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.adapter5" title="Permalink to this definition"></a></dt>
<dd><p>Strips a 5’ adapter sequence.</p>
<p>Similar to -a (adaptor3), but on the 5’ end.</p>
<p>This controls the addition of the -5 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.cores">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cores</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.cores" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads, disabled on free versions [default: number of cores]</p>
<p>This controls the addition of the -c parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.database">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">database</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.database" title="Permalink to this definition"></a></dt>
<dd><p>database filename</p>
<p>This controls the addition of the -d parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.format">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">format</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.format" title="Permalink to this definition"></a></dt>
<dd><p>Format of read files.</p>
<p>Allowed values: FA, SLXFQ, STDFQ, ILMFQ, PRB, PRBnSEQ</p>
<p>This controls the addition of the -F parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.fragment">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fragment</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.fragment" title="Permalink to this definition"></a></dt>
<dd><p>Fragment length (2 reads + insert) and standard deviation [default: 250 30]</p>
<p>This controls the addition of the -i parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.gap_extend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gap_extend</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.gap_extend" title="Permalink to this definition"></a></dt>
<dd><p>Gap extend penalty [default: 15]</p>
<p>This controls the addition of the -x parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.gap_open">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gap_open</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.gap_open" title="Permalink to this definition"></a></dt>
<dd><p>Gap opening penalty [default: 40]</p>
<p>This controls the addition of the -g parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.good_bases">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">good_bases</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.good_bases" title="Permalink to this definition"></a></dt>
<dd><p>Minimum number of good quality bases [default: log(N_g, 4) + 5]</p>
<p>This controls the addition of the -l parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.homopolymer">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">homopolymer</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.homopolymer" title="Permalink to this definition"></a></dt>
<dd><p>Homopolymer read filter [default: 20; disable: negative value]</p>
<p>This controls the addition of the -h parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.miRNA">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">miRNA</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.miRNA" title="Permalink to this definition"></a></dt>
<dd><p>Sets miRNA mode and optionally sets a value for the region scanned [default: off]</p>
<p>This controls the addition of the -m parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.qual_digits">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qual_digits</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.qual_digits" title="Permalink to this definition"></a></dt>
<dd><p>Decimal digits for quality scores [default: 0]</p>
<p>This controls the addition of the -q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.quality">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">quality</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.quality" title="Permalink to this definition"></a></dt>
<dd><p>Lower threshold for an alignment to be reported [default: 0]</p>
<p>This controls the addition of the -Q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.r_method">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r_method</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.r_method" title="Permalink to this definition"></a></dt>
<dd><p>Methods to report reads with multiple matches.</p>
<p>Allowed values: None, Random, All, Exhaustive, 0.99
‘All’ and ‘Exhaustive’ accept limits.</p>
<p>This controls the addition of the -r parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.read_cal">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_cal</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.read_cal" title="Permalink to this definition"></a></dt>
<dd><p>Read quality calibration from file (mismatch counts)</p>
<p>This controls the addition of the -k parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.readfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">readfile</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.readfile" title="Permalink to this definition"></a></dt>
<dd><p>read file</p>
<p>This controls the addition of the -f parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.recorded">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">recorded</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.recorded" title="Permalink to this definition"></a></dt>
<dd><p>Alignments recorded with score equal to the best.</p>
<p>Default: 1000 in default read method, otherwise no limit.</p>
<p>This controls the addition of the -e parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.repeats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">repeats</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.repeats" title="Permalink to this definition"></a></dt>
<dd><p>If score difference is higher, report repeats.</p>
<p>Otherwise -r read method applies [default: 5]</p>
<p>This controls the addition of the -R parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.report">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">report</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.report" title="Permalink to this definition"></a></dt>
<dd><p>Specifies the report format.</p>
<p>Allowed values: Native, Pairwise, SAM
Default: Native</p>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.threshold" title="Permalink to this definition"></a></dt>
<dd><p>Threshold for alignment score</p>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.trimming">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">trimming</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.trimming" title="Permalink to this definition"></a></dt>
<dd><p>If fail to align, trim by s bases until they map or become shorter than l.</p>
<p>Ddefault: 2</p>
<p>This controls the addition of the -s parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.truncate">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">truncate</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.truncate" title="Permalink to this definition"></a></dt>
<dd><p>Truncate to specific length before alignment</p>
<p>This controls the addition of the -n parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.unconverted">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">unconverted</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.unconverted" title="Permalink to this definition"></a></dt>
<dd><p>Experimental: unconverted cytosines penalty in bisulfite mode</p>
<p>Default: no penalty</p>
<p>This controls the addition of the -u parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.variation">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">variation</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.variation" title="Permalink to this definition"></a></dt>
<dd><p>Structural variation penalty [default: 70]</p>
<p>This controls the addition of the -v parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.NovoalignCommandline.write_cal">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">write_cal</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.NovoalignCommandline.write_cal" title="Permalink to this definition"></a></dt>
<dd><p>Accumulate mismatch counts and write to file</p>
<p>This controls the addition of the -K parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsViewCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools view.</p>
<p>Extract/print all or sub alignments in SAM or BAM format, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools view [-bchuHS] [-t in.refList] [-o output] [-f reqFlag]
                [-F skipFlag] [-q minMapQ] [-l library] [-r readGroup]
                [-R rgFile] &lt;in.bam&gt;|&lt;in.sam&gt; [region1 [...]]
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsViewCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/sam_or_bam_file&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_view_cmd</span> <span class="o">=</span> <span class="n">SamtoolsViewCommandline</span><span class="p">(</span><span class="n">input_file</span><span class="o">=</span><span class="n">input_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_view_cmd</span><span class="p">)</span>
<span class="go">samtools view /path/to/sam_or_bam_file</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.F">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">F</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.F" title="Permalink to this definition"></a></dt>
<dd><p>Skip alignments with bits present in INT</p>
<p>This controls the addition of the -F parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.H">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">H</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.H" title="Permalink to this definition"></a></dt>
<dd><p>Output the header only</p>
<p>This property controls the addition of the -H switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.R">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">R</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.R" title="Permalink to this definition"></a></dt>
<dd><p>Output reads in read groups listed in FILE</p>
<p>This controls the addition of the -R parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.S">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">S</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.S" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Input is in SAM.</dt><dd><p>If &#64;SQ header lines are absent,
the ‘-t’ option is required.</p>
</dd>
</dl>
<p>This property controls the addition of the -S switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.b">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.b" title="Permalink to this definition"></a></dt>
<dd><p>Output in the BAM format</p>
<p>This property controls the addition of the -b switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.c">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">c</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.c" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>Instead of printing the alignments, only count them and</dt><dd><p>print the total number.</p>
<p>All filter options, such as ‘-f’, ‘-F’ and ‘-q’,
are taken into account</p>
</dd>
</dl>
<p>This property controls the addition of the -c switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.f">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">f</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.f" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Only output alignments with all bits in</dt><dd><p>INT present in the FLAG field</p>
</dd>
</dl>
<p>This controls the addition of the -f parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.fast_bam">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fast_bam</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.fast_bam" title="Permalink to this definition"></a></dt>
<dd><p>Use zlib compression level 1 to compress the output</p>
<p>This property controls the addition of the -1 switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Include the header in the output</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.input_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.input_file" title="Permalink to this definition"></a></dt>
<dd><p>Input File Name</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.l">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">l</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.l" title="Permalink to this definition"></a></dt>
<dd><p>Only output reads in library STR</p>
<p>This controls the addition of the -l parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.o">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">o</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.o" title="Permalink to this definition"></a></dt>
<dd><p>Output file</p>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">q</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.q" title="Permalink to this definition"></a></dt>
<dd><p>Skip alignments with MAPQ smaller than INT</p>
<p>This controls the addition of the -q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.r">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.r" title="Permalink to this definition"></a></dt>
<dd><p>Only output reads in read group STR</p>
<p>This controls the addition of the -r parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.region">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">region</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.region" title="Permalink to this definition"></a></dt>
<dd><p>Region</p>
<p>This controls the addition of the region parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.t">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">t</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.t" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>This file is TAB-delimited.</dt><dd><p>Each line must contain the reference name and the
length of the reference, one line for each
distinct reference; additional fields are ignored.</p>
<p>This file also defines the order of the reference
sequences in sorting.
If you run   ‘samtools faidx &lt;ref.fa&gt;’,
the resultant index file &lt;ref.fa&gt;.fai can be used
as this &lt;in.ref_list&gt; file.</p>
</dd>
</dl>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsViewCommandline.u">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">u</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsViewCommandline.u" title="Permalink to this definition"></a></dt>
<dd><p>Output uncompressed BAM.</p>
<blockquote>
<div><p>This option saves time spent on compression/decompression
and is thus preferred when the output is piped to
another samtools command</p>
</div></blockquote>
<p>This property controls the addition of the -u switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsCalmdCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools calmd.</p>
<p>Generate the MD tag, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools calmd [-EeubSr] [-C capQcoef] &lt;aln.bam&gt; &lt;ref.fasta&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsCalmdCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/aln.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_fasta</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">calmd_cmd</span> <span class="o">=</span> <span class="n">SamtoolsCalmdCommandline</span><span class="p">(</span><span class="n">input_bam</span><span class="o">=</span><span class="n">input_bam</span><span class="p">,</span>
<span class="gp">... </span>                                     <span class="n">reference</span><span class="o">=</span><span class="n">reference_fasta</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">calmd_cmd</span><span class="p">)</span>
<span class="go">samtools calmd /path/to/aln.bam /path/to/reference.fasta</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.A">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">A</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.A" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>When used jointly with -r this option overwrites</dt><dd><p>the original base quality</p>
</dd>
</dl>
<p>This property controls the addition of the -A switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.C">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">C</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.C" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>Coefficient to cap mapping quality</dt><dd><p>of poorly mapped reads.</p>
<p>See the pileup command for details.</p>
</dd>
</dl>
<p>This controls the addition of the -C parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.E">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">E</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.E" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Extended BAQ calculation.</dt><dd><p>This option trades specificity for sensitivity,
though the effect is minor.</p>
</dd>
</dl>
<p>This property controls the addition of the -E switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.S">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">S</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.S" title="Permalink to this definition"></a></dt>
<dd><p>The input is SAM with header lines</p>
<p>This property controls the addition of the -S switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.b">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.b" title="Permalink to this definition"></a></dt>
<dd><p>Output compressed BAM</p>
<p>This property controls the addition of the -b switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.e">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">e</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.e" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>Convert the read base to = if it is</dt><dd><p>identical to the aligned reference base.</p>
<p>Indel caller does not support the = bases
at the moment.</p>
</dd>
</dl>
<p>This property controls the addition of the -e switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.input_bam">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_bam</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.input_bam" title="Permalink to this definition"></a></dt>
<dd><p>Input BAM</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.r">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.r" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Compute the BQ tag (without -A)</dt><dd><p>or cap base quality by BAQ (with -A).</p>
</dd>
</dl>
<p>This property controls the addition of the -r switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.ref">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ref</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.ref" title="Permalink to this definition"></a></dt>
<dd><p>Reference FASTA to be indexed</p>
<p>This controls the addition of the reference parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCalmdCommandline.u">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">u</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCalmdCommandline.u" title="Permalink to this definition"></a></dt>
<dd><p>Output uncompressed BAM</p>
<p>This property controls the addition of the -u switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCatCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsCatCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCatCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools cat.</p>
<p>Concatenate BAMs, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools cat [-h header.sam] [-o out.bam] &lt;in1.bam&gt; &lt;in2.bam&gt; [ ... ]
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsCatCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bam1</span> <span class="o">=</span> <span class="s2">&quot;/path/to/input_bam1&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bam2</span> <span class="o">=</span> <span class="s2">&quot;/path/to/input_bam2&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bams</span> <span class="o">=</span> <span class="p">[</span><span class="n">input_bam1</span><span class="p">,</span> <span class="n">input_bam2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_cat_cmd</span> <span class="o">=</span> <span class="n">SamtoolsCatCommandline</span><span class="p">(</span><span class="n">input_bam</span><span class="o">=</span><span class="n">input_bams</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_cat_cmd</span><span class="p">)</span>
<span class="go">samtools cat /path/to/input_bam1 /path/to/input_bam2</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCatCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCatCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCatCommandline.bams">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bams</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCatCommandline.bams" title="Permalink to this definition"></a></dt>
<dd><p>Input BAM files</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCatCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCatCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Header SAM file</p>
<p>This controls the addition of the -h parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsCatCommandline.o">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">o</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsCatCommandline.o" title="Permalink to this definition"></a></dt>
<dd><p>Output SAM file</p>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsFaidxCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsFaidxCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsFaidxCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools faidx.</p>
<p>Retrieve and print stats in the index file, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools faidx &lt;ref.fasta&gt; [region1 [...]]
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsFaidxCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_faidx_cmd</span> <span class="o">=</span> <span class="n">SamtoolsFaidxCommandline</span><span class="p">(</span><span class="n">reference</span><span class="o">=</span><span class="n">reference</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_faidx_cmd</span><span class="p">)</span>
<span class="go">samtools faidx /path/to/reference.fasta</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsFaidxCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsFaidxCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsFaidxCommandline.ref">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ref</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsFaidxCommandline.ref" title="Permalink to this definition"></a></dt>
<dd><p>Reference FASTA to be indexed</p>
<p>This controls the addition of the reference parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsFixmateCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsFixmateCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsFixmateCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools fixmate.</p>
<p>Fill in mate coordinates, ISIZE and mate related
flags from a name-sorted alignment, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools fixmate &lt;in.nameSrt.bam&gt; &lt;out.bam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsFixmateCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">in_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/in.nameSrt.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/out.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fixmate_cmd</span> <span class="o">=</span> <span class="n">SamtoolsFixmateCommandline</span><span class="p">(</span><span class="n">input_bam</span><span class="o">=</span><span class="n">in_bam</span><span class="p">,</span>
<span class="gp">... </span>                                         <span class="n">out_bam</span><span class="o">=</span><span class="n">out_bam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fixmate_cmd</span><span class="p">)</span>
<span class="go">samtools fixmate /path/to/in.nameSrt.bam /path/to/out.bam</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsFixmateCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsFixmateCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsFixmateCommandline.input_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsFixmateCommandline.input_file" title="Permalink to this definition"></a></dt>
<dd><p>Name Sorted Alignment File</p>
<p>This controls the addition of the in_bam parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsFixmateCommandline.output_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">output_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsFixmateCommandline.output_file" title="Permalink to this definition"></a></dt>
<dd><p>Output file</p>
<p>This controls the addition of the out_bam parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsIdxstatsCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsIdxstatsCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsIdxstatsCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools idxstats.</p>
<p>Retrieve and print stats in the index file, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools idxstats &lt;aln.bam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsIdxstatsCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;/path/to/aln_bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_idxstats_cmd</span> <span class="o">=</span> <span class="n">SamtoolsIdxstatsCommandline</span><span class="p">(</span><span class="n">input_bam</span><span class="o">=</span><span class="nb">input</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_idxstats_cmd</span><span class="p">)</span>
<span class="go">samtools idxstats /path/to/aln_bam</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsIdxstatsCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsIdxstatsCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsIdxstatsCommandline.input_bam">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_bam</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsIdxstatsCommandline.input_bam" title="Permalink to this definition"></a></dt>
<dd><p>BAM file to be indexed</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsIndexCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsIndexCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsIndexCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools index.</p>
<p>Index sorted alignment for fast random access, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools index &lt;aln.bam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsIndexCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="s2">&quot;/path/to/aln_bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_index_cmd</span> <span class="o">=</span> <span class="n">SamtoolsIndexCommandline</span><span class="p">(</span><span class="n">input_bam</span><span class="o">=</span><span class="nb">input</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_index_cmd</span><span class="p">)</span>
<span class="go">samtools index /path/to/aln_bam</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsIndexCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsIndexCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsIndexCommandline.input_bam">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_bam</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsIndexCommandline.input_bam" title="Permalink to this definition"></a></dt>
<dd><p>BAM file to be indexed</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsMergeCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools merge.</p>
<p>Merge multiple sorted alignments, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools merge [-nur1f] [-h inh.sam] [-R reg]
                 &lt;out.bam&gt; &lt;in1.bam&gt; &lt;in2.bam&gt; [...]
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsMergeCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/out_bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">in_bam</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;/path/to/input_bam1&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/input_bam2&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">merge_cmd</span> <span class="o">=</span> <span class="n">SamtoolsMergeCommandline</span><span class="p">(</span><span class="n">out_bam</span><span class="o">=</span><span class="n">out_bam</span><span class="p">,</span>
<span class="gp">... </span>                                     <span class="n">input_bam</span><span class="o">=</span><span class="n">in_bam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">merge_cmd</span><span class="p">)</span>
<span class="go">samtools merge /path/to/out_bam /path/to/input_bam1 /path/to/input_bam2</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.R">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">R</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.R" title="Permalink to this definition"></a></dt>
<dd><p>Merge files in the specified region indicated by STR</p>
<p>This controls the addition of the -R parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.bam">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bam</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.bam" title="Permalink to this definition"></a></dt>
<dd><p>Input BAM</p>
<p>This controls the addition of the input_bam parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.f">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">f</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.f" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Force to overwrite the</dt><dd><p>output file if present</p>
</dd>
</dl>
<p>This property controls the addition of the -f switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.fast_bam">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fast_bam</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.fast_bam" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Use zlib compression level 1</dt><dd><p>to compress the output</p>
</dd>
</dl>
<p>This property controls the addition of the -1 switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.h" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Use the lines of FILE as ‘&#64;’</dt><dd><p>headers to be copied to out.bam</p>
</dd>
</dl>
<p>This controls the addition of the -h parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.n">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.n" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>The input alignments are sorted by read names</dt><dd><p>rather than by chromosomal coordinates</p>
</dd>
</dl>
<p>This property controls the addition of the -n switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.output">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">output</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.output" title="Permalink to this definition"></a></dt>
<dd><p>Output BAM file</p>
<p>This controls the addition of the output_bam parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.r">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.r" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Attach an RG tag to each alignment.</dt><dd><p>The tag value is inferred from file names</p>
</dd>
</dl>
<p>This property controls the addition of the -r switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMergeCommandline.u">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">u</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMergeCommandline.u" title="Permalink to this definition"></a></dt>
<dd><p>Uncompressed BAM output</p>
<p>This property controls the addition of the -u switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsMpileupCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools mpileup.</p>
<p>Generate BCF or pileup for one or multiple BAM files, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools mpileup [-EBug] [-C capQcoef] [-r reg] [-f in.fa]
                   [-l list] [-M capMapQ] [-Q minBaseQ]
                   [-q minMapQ] in.bam [in2.bam [...]]
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsMpileupCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;/path/to/sam_or_bam_file&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_mpileup_cmd</span> <span class="o">=</span> <span class="n">SamtoolsMpileupCommandline</span><span class="p">(</span><span class="n">input_file</span><span class="o">=</span><span class="nb">input</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_mpileup_cmd</span><span class="p">)</span>
<span class="go">samtools mpileup /path/to/sam_or_bam_file</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.A">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">A</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.A" title="Permalink to this definition"></a></dt>
<dd><p>Do not skip anomalous read pairs in variant calling.</p>
<p>This property controls the addition of the -A switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.B">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">B</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.B" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>Disable probabilistic realignment for the</dt><dd><p>computation of base alignment quality (BAQ).</p>
<p>BAQ is the Phred-scaled probability of a read base being
misaligned.
Applying this option greatly helps to reduce false SNPs
caused by misalignments</p>
</dd>
</dl>
<p>This property controls the addition of the -B switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.C">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">C</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.C" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>Coefficient for downgrading mapping quality for</dt><dd><p>reads containing excessive mismatches.</p>
<p>Given a read with a phred-scaled probability q of
being generated from the mapped position,
the new mapping quality is about sqrt((INT-q)/INT)*INT.
A zero value disables this functionality;
if enabled, the recommended value for BWA is 50</p>
</dd>
</dl>
<p>This controls the addition of the -C parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.D">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">D</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.D" title="Permalink to this definition"></a></dt>
<dd><p>Output per-sample read depth</p>
<p>This property controls the addition of the -D switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.E">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">E</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.E" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Extended BAQ computation.</dt><dd><p>This option helps sensitivity especially
for MNPs, but may hurt specificity a little bit</p>
</dd>
</dl>
<p>This property controls the addition of the -E switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.I">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">I</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.I" title="Permalink to this definition"></a></dt>
<dd><p>Do not perform INDEL calling</p>
<p>This property controls the addition of the -I switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.L">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">L</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.L" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Skip INDEL calling if the average per-sample</dt><dd><p>depth is above INT</p>
</dd>
</dl>
<p>This controls the addition of the -L parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.M">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">M</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.M" title="Permalink to this definition"></a></dt>
<dd><p>Cap Mapping Quality at M</p>
<p>This controls the addition of the -M parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.Q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Q</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.Q" title="Permalink to this definition"></a></dt>
<dd><p>Minimum base quality for a base to be considered</p>
<p>This controls the addition of the -Q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.S">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">S</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.S" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Output per-sample Phred-scaled</dt><dd><p>strand bias P-value</p>
</dd>
</dl>
<p>This property controls the addition of the -S switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.b">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.b" title="Permalink to this definition"></a></dt>
<dd><p>List of input BAM files, one file per line</p>
<p>This controls the addition of the -b parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.d">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">d</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.d" title="Permalink to this definition"></a></dt>
<dd><p>At a position, read maximally INT reads per input BAM</p>
<p>This controls the addition of the -d parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.e">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">e</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.e" title="Permalink to this definition"></a></dt>
<dd><p>Phred-scaled gap extension sequencing error probability.</p>
<blockquote>
<div><p>Reducing INT leads to longer indels</p>
</div></blockquote>
<p>This controls the addition of the -e parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.f">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">f</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.f" title="Permalink to this definition"></a></dt>
<dd><p>The faidx-indexed reference file in the FASTA format.</p>
<blockquote>
<div><p>The file can be optionally compressed by razip</p>
</div></blockquote>
<p>This controls the addition of the -f parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.g">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">g</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.g" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Compute genotype likelihoods and output them in the</dt><dd><p>binary call format (BCF)</p>
</dd>
</dl>
<p>This property controls the addition of the -g switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Coefficient for modeling homopolymer errors.</p>
<blockquote>
<div><p>Given an l-long homopolymer run, the sequencing error
of an indel of size s is modeled as INT*s/l</p>
</div></blockquote>
<p>This controls the addition of the -h parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.illumina_13">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">illumina_13</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.illumina_13" title="Permalink to this definition"></a></dt>
<dd><p>Assume the quality is in the Illumina 1.3+ encoding</p>
<p>This property controls the addition of the -6 switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.input_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.input_file" title="Permalink to this definition"></a></dt>
<dd><p>Input File for generating mpileup</p>
<p>This controls the addition of the input_file parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.l">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">l</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.l" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>BED or position list file containing a list of regions</dt><dd><p>or sites where pileup or BCF should be generated</p>
</dd>
</dl>
<p>This controls the addition of the -l parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.o">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">o</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.o" title="Permalink to this definition"></a></dt>
<dd><p>Phred-scaled gap open sequencing error probability.</p>
<blockquote>
<div><p>Reducing INT leads to more indel calls.</p>
</div></blockquote>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.p">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">p</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.p" title="Permalink to this definition"></a></dt>
<dd><dl>
<dt>Comma delimited list of platforms (determined by &#64;RG-PL)</dt><dd><p>from which indel candidates are obtained.</p>
<p>It is recommended to collect indel candidates from
sequencing technologies that have low indel error rate
such as ILLUMINA</p>
</dd>
</dl>
<p>This controls the addition of the -p parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">q</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.q" title="Permalink to this definition"></a></dt>
<dd><p>Minimum mapping quality for an alignment to be used</p>
<p>This controls the addition of the -q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.r">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">r</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.r" title="Permalink to this definition"></a></dt>
<dd><p>Only generate pileup in region STR</p>
<p>This controls the addition of the -r parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsMpileupCommandline.u">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">u</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsMpileupCommandline.u" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Similar to -g except that the output is</dt><dd><p>uncompressed BCF, which is preferred for piping</p>
</dd>
</dl>
<p>This property controls the addition of the -u switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsPhaseCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools phase.</p>
<p>Call and phase heterozygous SNPs, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools phase [-AF] [-k len] [-b prefix]
                 [-q minLOD] [-Q minBaseQ] &lt;in.bam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsPhaseCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/in.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_phase_cmd</span> <span class="o">=</span> <span class="n">SamtoolsPhaseCommandline</span><span class="p">(</span><span class="n">input_bam</span><span class="o">=</span><span class="n">input_bam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_phase_cmd</span><span class="p">)</span>
<span class="go">samtools phase /path/to/in.bam</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline.A">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">A</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline.A" title="Permalink to this definition"></a></dt>
<dd><p>Drop reads with ambiguous phase</p>
<p>This property controls the addition of the -A switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline.F">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">F</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline.F" title="Permalink to this definition"></a></dt>
<dd><p>Do not attempt to fix chimeric reads</p>
<p>This property controls the addition of the -F switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline.Q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Q</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline.Q" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Minimum base quality to be</dt><dd><p>used in het calling</p>
</dd>
</dl>
<p>This controls the addition of the -Q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline.b">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">b</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline.b" title="Permalink to this definition"></a></dt>
<dd><p>Prefix of BAM output</p>
<p>This controls the addition of the -b parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline.in_bam">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">in_bam</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline.in_bam" title="Permalink to this definition"></a></dt>
<dd><p>Input file</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline.k">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">k</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline.k" title="Permalink to this definition"></a></dt>
<dd><p>Maximum length for local phasing</p>
<p>This controls the addition of the -k parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsPhaseCommandline.q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">q</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsPhaseCommandline.q" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Minimum Phred-scaled LOD to</dt><dd><p>call a heterozygote</p>
</dd>
</dl>
<p>This controls the addition of the -q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsReheaderCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsReheaderCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsReheaderCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools reheader.</p>
<p>Replace the header in in.bam with the header
in in.header.sam, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools reheader &lt;in.header.sam&gt; &lt;in.bam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsReheaderCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_header</span> <span class="o">=</span> <span class="s2">&quot;/path/to/header_sam_file&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/input_bam_file&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reheader_cmd</span> <span class="o">=</span> <span class="n">SamtoolsReheaderCommandline</span><span class="p">(</span><span class="n">input_header</span><span class="o">=</span><span class="n">input_header</span><span class="p">,</span>
<span class="gp">... </span>                                           <span class="n">input_bam</span><span class="o">=</span><span class="n">input_bam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">reheader_cmd</span><span class="p">)</span>
<span class="go">samtools reheader /path/to/header_sam_file /path/to/input_bam_file</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsReheaderCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsReheaderCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsReheaderCommandline.bam_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bam_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsReheaderCommandline.bam_file" title="Permalink to this definition"></a></dt>
<dd><p>BAM file for writing header to</p>
<p>This controls the addition of the input_bam parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsReheaderCommandline.sam_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sam_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsReheaderCommandline.sam_file" title="Permalink to this definition"></a></dt>
<dd><p>Sam file with header</p>
<p>This controls the addition of the input_header parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsRmdupCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsRmdupCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsRmdupCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools rmdup.</p>
<p>Remove potential PCR duplicates, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools rmdup [-sS] &lt;input.srt.bam&gt; &lt;out.bam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsRmdupCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_sorted_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/input.srt.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/out.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rmdup_cmd</span> <span class="o">=</span> <span class="n">SamtoolsRmdupCommandline</span><span class="p">(</span><span class="n">input_bam</span><span class="o">=</span><span class="n">input_sorted_bam</span><span class="p">,</span>
<span class="gp">... </span>                                     <span class="n">out_bam</span><span class="o">=</span><span class="n">out_bam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rmdup_cmd</span><span class="p">)</span>
<span class="go">samtools rmdup /path/to/input.srt.bam /path/to/out.bam</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsRmdupCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsRmdupCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsRmdupCommandline.S">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">S</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsRmdupCommandline.S" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Treat paired-end reads</dt><dd><p>as single-end reads</p>
</dd>
</dl>
<p>This property controls the addition of the -S switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsRmdupCommandline.input_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsRmdupCommandline.input_file" title="Permalink to this definition"></a></dt>
<dd><p>Name Sorted Alignment File</p>
<p>This controls the addition of the in_bam parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsRmdupCommandline.output_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">output_file</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsRmdupCommandline.output_file" title="Permalink to this definition"></a></dt>
<dd><p>Output file</p>
<p>This controls the addition of the out_bam parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsRmdupCommandline.s">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">s</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsRmdupCommandline.s" title="Permalink to this definition"></a></dt>
<dd><p>Remove duplicates for single-end reads.</p>
<blockquote>
<div><p>By default, the command works for paired-end
reads only</p>
</div></blockquote>
<p>This property controls the addition of the -s switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsSortCommandline">
<span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsSortCommandline</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsSortCommandline" title="Permalink to this definition"></a></dt>
<dd><p>alias of <a class="reference internal" href="#Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline" title="Bio.Sequencing.Applications._samtools.SamtoolsVersion0xSortCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Sequencing.Applications._samtools.SamtoolsVersion0xSortCommandline</span></code></a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsVersion0xSortCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools version 0.1.x sort.</p>
<p>Concatenate BAMs, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools sort [-no] [-m maxMem] &lt;in.bam&gt; &lt;out.prefix&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsVersion0xSortCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/input_bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out_prefix</span> <span class="o">=</span> <span class="s2">&quot;/path/to/out_prefix&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_sort_cmd</span> <span class="o">=</span> <span class="n">SamtoolsVersion0xSortCommandline</span><span class="p">(</span><span class="nb">input</span><span class="o">=</span><span class="n">input_bam</span><span class="p">,</span> <span class="n">out_prefix</span><span class="o">=</span><span class="n">out_prefix</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_sort_cmd</span><span class="p">)</span>
<span class="go">samtools sort /path/to/input_bam /path/to/out_prefix</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.input">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.input" title="Permalink to this definition"></a></dt>
<dd><p>Input BAM file</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.m">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">m</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.m" title="Permalink to this definition"></a></dt>
<dd><p>Approximately the maximum required memory</p>
<p>This controls the addition of the -m parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.n">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.n" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Sort by read names rather</dt><dd><p>than by chromosomal coordinates</p>
</dd>
</dl>
<p>This property controls the addition of the -n switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.o">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">o</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.o" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Output the final alignment</dt><dd><p>to the standard output</p>
</dd>
</dl>
<p>This property controls the addition of the -o switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.out_prefix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out_prefix</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion0xSortCommandline.out_prefix" title="Permalink to this definition"></a></dt>
<dd><p>Output prefix</p>
<p>This controls the addition of the out_prefix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsVersion1xSortCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools version 1.3.x sort.</p>
<p>Concatenate BAMs, equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools sort [-n] [-T FREFIX] [-o file] [-I INT] [-m maxMem] &lt;in.bam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsVersion1xSortCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/input_bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">FREFIX</span> <span class="o">=</span> <span class="s2">&quot;/path/to/out_prefix&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="s2">&quot;/path/to/out_file&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_sort_cmd</span> <span class="o">=</span> <span class="n">SamtoolsVersion1xSortCommandline</span><span class="p">(</span><span class="nb">input</span><span class="o">=</span><span class="n">input_bam</span><span class="p">,</span> <span class="n">T</span><span class="o">=</span><span class="n">FREFIX</span><span class="p">,</span> <span class="n">o</span><span class="o">=</span><span class="n">file_name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_sort_cmd</span><span class="p">)</span>
<span class="go">samtools sort -o /path/to/out_file -T /path/to/out_prefix /path/to/input_bam</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.I">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">I</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.I" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>(INT) Set the desired compression level for the final output file,</dt><dd><p>ranging from 0 (uncompressed) or 1 (fastest but minimal compression)
to 9 (best compression but slowest to write), similarly to gzip(1)’s compression level setting.</p>
</dd>
</dl>
<p>This controls the addition of the -I parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.O">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">O</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.O" title="Permalink to this definition"></a></dt>
<dd><p>(FORMAT) Write the final output as sam, bam, or cram</p>
<p>This controls the addition of the -O parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.T">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">T</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.T" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>(PREFIX) Write temporary files to PREFIX.nnnn.bam, or if the specified PREFIX</dt><dd><p>is an existing directory, to PREFIX/samtools.mmm.mmm.tmp.nnnn.bam,
where mmm is unique to this invocation of the sort command</p>
</dd>
</dl>
<p>This controls the addition of the -T parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.input">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.input" title="Permalink to this definition"></a></dt>
<dd><p>Input SAM/BAM/CRAM file</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.m">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">m</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.m" title="Permalink to this definition"></a></dt>
<dd><p>Approximately the maximum required memory</p>
<p>This controls the addition of the -m parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.n">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">n</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.n" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>Sort by read names rather</dt><dd><p>than by chromosomal coordinates</p>
</dd>
</dl>
<p>This property controls the addition of the -n switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.o">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">o</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsVersion1xSortCommandline.o" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>(file) Write the final sorted output to FILE,</dt><dd><p>rather than to standard output</p>
</dd>
</dl>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Sequencing.Applications.</span></span><span class="sig-name descname"><span class="pre">SamtoolsTargetcutCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for samtools targetcut.</p>
<p>This command identifies target regions by examining the continuity
of read depth, computes haploid consensus sequences of targets
and outputs a SAM with each sequence corresponding to a target,
equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ samtools targetcut [-Q minBaseQ] [-i inPenalty] [-0 em0]
                     [-1 em1] [-2 em2] [-f ref] &lt;in.bam&gt;
</pre></div>
</div>
<p>See <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a> for more details</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Sequencing.Applications</span> <span class="kn">import</span> <span class="n">SamtoolsTargetcutCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">input_bam</span> <span class="o">=</span> <span class="s2">&quot;/path/to/aln.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">samtools_targetcut_cmd</span> <span class="o">=</span> <span class="n">SamtoolsTargetcutCommandline</span><span class="p">(</span><span class="n">input_bam</span><span class="o">=</span><span class="n">input_bam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">samtools_targetcut_cmd</span><span class="p">)</span>
<span class="go">samtools targetcut /path/to/aln.bam</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'samtools'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.Q">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Q</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.Q" title="Permalink to this definition"></a></dt>
<dd><p>Minimum Base Quality</p>
<p>This controls the addition of the -Q parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.em0">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">em0</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.em0" title="Permalink to this definition"></a></dt>
<dd><p>This controls the addition of the -0 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.em1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">em1</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.em1" title="Permalink to this definition"></a></dt>
<dd><p>This controls the addition of the -1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.em2">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">em2</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.em2" title="Permalink to this definition"></a></dt>
<dd><p>This controls the addition of the -2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.f">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">f</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.f" title="Permalink to this definition"></a></dt>
<dd><p>Reference Filename</p>
<p>This controls the addition of the -f parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.i">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">i</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.i" title="Permalink to this definition"></a></dt>
<dd><p>Insertion Penalty</p>
<p>This controls the addition of the -i parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.in_bam">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">in_bam</span></span><a class="headerlink" href="#Bio.Sequencing.Applications.SamtoolsTargetcutCommandline.in_bam" title="Permalink to this definition"></a></dt>
<dd><p>Input file</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Sequencing.html" class="btn btn-neutral float-left" title="Bio.Sequencing package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Sequencing.Ace.html" class="btn btn-neutral float-right" title="Bio.Sequencing.Ace module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2020, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.80.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.78/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>