

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Bio.SeqFeature module &mdash; Biopython 1.79.dev0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SeqRecord module" href="Bio.SeqRecord.html" />
    <link rel="prev" title="Bio.Seq module" href="Bio.Seq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.File.html">Bio.File module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.LogisticRegression.html">Bio.LogisticRegression module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.MarkovModel.html">Bio.MarkovModel module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.MaxEntropy.html">Bio.MaxEntropy module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NaiveBayes.html">Bio.NaiveBayes module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Seq.html">Bio.Seq module</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.SeqFeature module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#base-class-to-hold-a-feature">Base class to hold a Feature</a></li>
<li class="toctree-l4"><a class="reference internal" href="#hold-information-about-a-reference">Hold information about a Reference</a></li>
<li class="toctree-l4"><a class="reference internal" href="#specify-locations-of-a-feature-on-a-sequence">Specify locations of a feature on a Sequence</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqRecord.html">Bio.SeqRecord module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.bgzf.html">Bio.bgzf module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.kNN.html">Bio.kNN module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.pairwise2.html">Bio.pairwise2 module</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          



















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
      <li>Bio.SeqFeature module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SeqFeature.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SeqRecord.html" class="btn btn-neutral float-right" title="Bio.SeqRecord module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
      
      
        <a href="Bio.Seq.html" class="btn btn-neutral float-left" title="Bio.Seq module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="module-Bio.SeqFeature">
<span id="bio-seqfeature-module"></span><h1>Bio.SeqFeature module<a class="headerlink" href="#module-Bio.SeqFeature" title="Permalink to this headline">¶</a></h1>
<p>Represent a Sequence Feature holding info about a part of a sequence.</p>
<p>This is heavily modeled after the Biocorba SeqFeature objects, and
may be pretty biased towards GenBank stuff since I’m writing it
for the GenBank parser output…</p>
<p>What’s here:</p>
<section id="base-class-to-hold-a-feature">
<h2>Base class to hold a Feature<a class="headerlink" href="#base-class-to-hold-a-feature" title="Permalink to this headline">¶</a></h2>
<dl class="simple">
<dt>Classes:</dt><dd><ul class="simple">
<li><p>SeqFeature</p></li>
</ul>
</dd>
</dl>
</section>
<section id="hold-information-about-a-reference">
<h2>Hold information about a Reference<a class="headerlink" href="#hold-information-about-a-reference" title="Permalink to this headline">¶</a></h2>
<p>This is an attempt to create a General class to hold Reference type
information.</p>
<dl class="simple">
<dt>Classes:</dt><dd><ul class="simple">
<li><p>Reference</p></li>
</ul>
</dd>
</dl>
</section>
<section id="specify-locations-of-a-feature-on-a-sequence">
<h2>Specify locations of a feature on a Sequence<a class="headerlink" href="#specify-locations-of-a-feature-on-a-sequence" title="Permalink to this headline">¶</a></h2>
<p>This aims to handle, in Ewan Birney’s words, ‘the dreaded fuzziness issue’.
This has the advantages of allowing us to handle fuzzy stuff in case anyone
needs it, and also be compatible with BioPerl etc and BioSQL.</p>
<dl class="simple">
<dt>Classes:</dt><dd><ul class="simple">
<li><p>FeatureLocation - Specify the start and end location of a feature.</p></li>
<li><p>CompoundLocation - Collection of FeatureLocation objects (for joins etc).</p></li>
<li><p>ExactPosition - Specify the position as being exact.</p></li>
<li><p>WithinPosition - Specify a position occurring within some range.</p></li>
<li><p>BetweenPosition - Specify a position occurring between a range (OBSOLETE?).</p></li>
<li><p>BeforePosition - Specify the position as being found before some base.</p></li>
<li><p>AfterPosition - Specify the position as being found after some base.</p></li>
<li><p>OneOfPosition - Specify a position where the location can be multiple positions.</p></li>
<li><p>UncertainPosition - Specify a specific position which is uncertain.</p></li>
<li><p>UnknownPosition - Represents missing information like ‘?’ in UniProt.</p></li>
</ul>
</dd>
</dl>
<dl class="py class">
<dt id="Bio.SeqFeature.SeqFeature">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">SeqFeature</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location_operator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'&lt;unknown</span> <span class="pre">id&gt;'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qualifiers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sub_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represent a Sequence Feature on an object.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>location - the location of the feature on the sequence (FeatureLocation)</p></li>
<li><p>type - the specified type of the feature (ie. CDS, exon, repeat…)</p></li>
<li><p>location_operator - a string specifying how this SeqFeature may
be related to others. For example, in the example GenBank feature
shown below, the location_operator would be “join”. This is a proxy
for feature.location.operator and only applies to compound locations.</p></li>
<li><p>strand - A value specifying on which strand (of a DNA sequence, for
instance) the feature deals with. 1 indicates the plus strand, -1
indicates the minus strand, 0 indicates stranded but unknown (? in GFF3),
while the default of None indicates that strand doesn’t apply (dot in GFF3,
e.g. features on proteins). Note this is a shortcut for accessing the
strand property of the feature’s location.</p></li>
<li><p>id - A string identifier for the feature.</p></li>
<li><p>ref - A reference to another sequence. This could be an accession
number for some different sequence. Note this is a shortcut for the
reference property of the feature’s location.</p></li>
<li><p>ref_db - A different database for the reference accession number.
Note this is a shortcut for the reference property of the location</p></li>
<li><p>qualifiers - A dictionary of qualifiers on the feature. These are
analogous to the qualifiers from a GenBank feature table. The keys of
the dictionary are qualifier names, the values are the qualifier
values. As of Biopython 1.69 this is an ordered dictionary.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location_operator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'&lt;unknown</span> <span class="pre">id&gt;'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qualifiers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sub_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a SeqFeature on a Sequence.</p>
<p>location can either be a FeatureLocation (with strand argument also
given if required), or None.</p>
<p>e.g. With no strand, on the forward strand, and on the reverse strand:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="n">f1</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="kc">None</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">110</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="n">f2</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="o">+</span><span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f3</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">108</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f3</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="n">f3</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span>
<span class="go">True</span>
</pre></div>
</div>
<p>An invalid strand will trigger an exception:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f4</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">60</span><span class="p">),</span> <span class="n">strand</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Strand should be +1, -1, 0 or None, not 2</span>
</pre></div>
</div>
<p>Similarly if set via the FeatureLocation directly:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loc4</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">60</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Strand should be +1, -1, 0 or None, not 2</span>
</pre></div>
</div>
<p>For exact start/end positions, an integer can be used (as shown above)
as shorthand for the ExactPosition object. For non-exact locations, the
FeatureLocation must be specified via the appropriate position objects.</p>
<p>Note that the strand, ref and ref_db arguments to the SeqFeature are
now obsolete and will be deprecated in a future release (which will
give warning messages) and later removed. Set them via the location
object instead.</p>
<p>Note that location_operator and sub_features arguments can no longer
be used, instead do this via the CompoundLocation object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.strand">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">strand</span></code><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Feature’s strand</p>
<p>This is a shortcut for feature.location.strand</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.ref">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">ref</span></code><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.ref" title="Permalink to this definition">¶</a></dt>
<dd><p>Feature location reference (e.g. accession).</p>
<p>This is a shortcut for feature.location.ref</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.ref_db">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">ref_db</span></code><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.ref_db" title="Permalink to this definition">¶</a></dt>
<dd><p>Feature location reference’s database.</p>
<p>This is a shortcut for feature.location.ref_db</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.location_operator">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">location_operator</span></code><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.location_operator" title="Permalink to this definition">¶</a></dt>
<dd><p>Location operator for compound locations (e.g. join).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the feature as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the full feature as a python string.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.extract">
<code class="sig-name descname"><span class="pre">extract</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">references</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.extract" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the feature’s sequence from supplied parent sequence.</p>
<p>The parent_sequence can be a Seq like object or a string, and will
generally return an object of the same type. The exception to this is
a MutableSeq as the parent sequence will return a Seq object.</p>
<p>This should cope with complex locations including complements, joins
and fuzzy positions. Even mixed strand features should work! This
also covers features on protein sequences (e.g. domains), although
here reverse strand features are not permitted. If the
location refers to other records, they must be supplied in the
optional dictionary references.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MKQHKAMIVALIVICITAVVAAL&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">Seq(&#39;VALIVIC&#39;)</span>
</pre></div>
</div>
<p>If the FeatureLocation is None, e.g. when parsing invalid locus
locations in the GenBank parser, extract() will raise a ValueError.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MKQHKAMIVALIVICITAVVAAL&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">The feature&#39;s .location is None. Check the sequence file for a valid location.</span>
</pre></div>
</div>
<p>Note - currently only compound features of type “join” are supported.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.translate">
<code class="sig-name descname"><span class="pre">translate</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Standard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'*'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a translation of the feature’s sequence.</p>
<p>This method is intended for CDS or other features that code proteins
and is a shortcut that will both extract the feature and
translate it, taking into account the codon_start and transl_table
qualifiers, if they are present. If they are not present the
value of the arguments “table” and “start_offset” are used.</p>
<p>The “cds” parameter is set to “True” if the feature is of type
“CDS” but can be overridden by giving an explicit argument.</p>
<p>The arguments stop_symbol, to_stop and gap have the same meaning
as Seq.translate, refer to that documentation for further information.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>parent_sequence - A DNA or RNA sequence.</p></li>
<li><p>table - Which codon table to use if there is no transl_table
qualifier for this feature. This can be either a name
(string), an NCBI identifier (integer), or a CodonTable
object (useful for non-standard genetic codes).  This
defaults to the “Standard” table.</p></li>
<li><p>start_offset - offset at which the first complete codon of a
coding feature can be found, relative to the first base of
that feature. Has a valid value of 0, 1 or 2. NOTE: this
uses python’s 0-based numbering whereas the codon_start
qualifier in files from NCBI use 1-based numbering.
Will override a codon_start qualifier</p></li>
</ul>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GGTTACACTTACCGATAATGTCTCTGATGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">30</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;transl_table&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">11</span><span class="p">]</span>
</pre></div>
</div>
<p>Note that features of type CDS are subject to the usual
checks at translation. But you can override this behaviour
by giving explicit arguments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Seq(&#39;GYTYR*CL**&#39;)</span>
</pre></div>
</div>
<p>Now use the start_offset argument to change the frame. Note
this uses python 0-based numbering.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">start_offset</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Seq(&#39;VTLTDNVSD&#39;)</span>
</pre></div>
</div>
<p>Alternatively use the codon_start qualifier to do the same
thing. Note: this uses 1-based numbering, which is found
in files from NCBI.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;codon_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Seq(&#39;VTLTDNVSD&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.__bool__">
<code class="sig-name descname"><span class="pre">__bool__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.__bool__" title="Permalink to this definition">¶</a></dt>
<dd><p>Boolean value of an instance of this class (True).</p>
<p>This behaviour is for backwards compatibility, since until the
__len__ method was added, a SeqFeature always evaluated as True.</p>
<p>Note that in comparison, Seq objects, strings, lists, etc, will all
evaluate to False if they have length zero.</p>
<p>WARNING: The SeqFeature may in future evaluate to False when its
length is zero (in order to better match normal python behaviour)!</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.__len__">
<code class="sig-name descname"><span class="pre">__len__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.__len__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the region where the feature is located.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MKQHKAMIVALIVICITAVVAAL&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">Seq(&#39;VALIVIC&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
<span class="go">7</span>
</pre></div>
</div>
<p>This is a proxy for taking the length of the feature’s location:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">location</span><span class="p">)</span>
<span class="go">7</span>
</pre></div>
</div>
<p>For simple features this is the same as the region spanned (end
position minus start position using Pythonic counting). However, for
a compound location (e.g. a CDS as the join of several exons) the
gaps are not counted (e.g. introns). This ensures that len(f) matches
len(f.extract(parent_seq)), and also makes sure things work properly
with features wrapping the origin etc.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.__iter__">
<code class="sig-name descname"><span class="pre">__iter__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.__iter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over the parent positions within the feature.</p>
<p>The iteration order is strand aware, and can be thought of as moving
along the feature using the parent sequence coordinates:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
<span class="go">9</span>
<span class="go">8</span>
<span class="go">7</span>
<span class="go">6</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">[9, 8, 7, 6, 5]</span>
</pre></div>
</div>
<p>This is a proxy for iterating over the location,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">location</span><span class="p">)</span>
<span class="go">[9, 8, 7, 6, 5]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.SeqFeature.__contains__">
<code class="sig-name descname"><span class="pre">__contains__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.SeqFeature.__contains__" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if an integer position is within the feature.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f</span><span class="p">]</span>
<span class="go">[5, 6, 7, 8, 9]</span>
</pre></div>
</div>
<p>For example, to see which features include a SNP position, you could
use this:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;GenBank/NC_000932.gb&quot;</span><span class="p">,</span> <span class="s2">&quot;gb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="mi">1750</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">location</span><span class="p">))</span>
<span class="go">source [0:154478](+)</span>
<span class="go">gene [1716:4347](-)</span>
<span class="go">tRNA join{[4310:4347](-), [1716:1751](-)}</span>
</pre></div>
</div>
<p>Note that for a feature defined as a join of several subfeatures (e.g.
the union of several exons) the gaps are not checked (e.g. introns).
In this example, the tRNA location is defined in the GenBank file as
complement(join(1717..1751,4311..4347)), so that position 1760 falls
in the gap:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="mi">1760</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">location</span><span class="p">))</span>
<span class="go">source [0:154478](+)</span>
<span class="go">gene [1716:4347](-)</span>
</pre></div>
</div>
<p>Note that additional care may be required with fuzzy locations, for
example just before a BeforePosition:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">BeforePosition</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">BeforePosition</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span> <span class="mi">8</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;domain&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f</span><span class="p">]</span>
<span class="go">[3, 4, 5, 6, 7]</span>
</pre></div>
</div>
<p>Note that is is a proxy for testing membership on the location.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">location</span><span class="p">]</span>
<span class="go">[3, 4, 5, 6, 7]</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.Reference">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">Reference</span></code><a class="headerlink" href="#Bio.SeqFeature.Reference" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represent a Generic Reference object.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>location - A list of Location objects specifying regions of
the sequence that the references correspond to. If no locations are
specified, the entire sequence is assumed.</p></li>
<li><p>authors - A big old string, or a list split by author, of authors
for the reference.</p></li>
<li><p>title - The title of the reference.</p></li>
<li><p>journal - Journal the reference was published in.</p></li>
<li><p>medline_id - A medline reference for the article.</p></li>
<li><p>pubmed_id - A pubmed reference for the article.</p></li>
<li><p>comment - A place to stick any comments about the reference.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.SeqFeature.Reference.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.Reference.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.Reference.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.Reference.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the full Reference object as a python string.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.Reference.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.Reference.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the Reference object as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.Reference.__eq__">
<code class="sig-name descname"><span class="pre">__eq__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.Reference.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if two Reference objects should be considered equal.</p>
<p>Note prior to Biopython 1.70 the location was not compared, as
until then __eq__ for the FeatureLocation class was not defined.</p>
</dd></dl>

<dl class="py attribute">
<dt id="Bio.SeqFeature.Reference.__hash__">
<code class="sig-name descname"><span class="pre">__hash__</span></code><em class="property"> <span class="pre">=</span> <span class="pre">None</span></em><a class="headerlink" href="#Bio.SeqFeature.Reference.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.FeatureLocation">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">FeatureLocation</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Specify the location of a feature along a sequence.</p>
<p>The FeatureLocation is used for simple continuous features, which can
be described as running from a start position to and end position
(optionally with a strand and reference information).  More complex
locations made up from several non-continuous parts (e.g. a coding
sequence made up of several exons) are described using a SeqFeature
with a CompoundLocation.</p>
<p>Note that the start and end location numbering follow Python’s scheme,
thus a GenBank entry of 123..150 (one based counting) becomes a location
of [122:150] (zero based counting).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">122</span><span class="p">,</span> <span class="mi">150</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">[122:150]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
<span class="go">122</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
<span class="go">150</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
<span class="go">None</span>
</pre></div>
</div>
<p>Note the strand defaults to None. If you are working with nucleotide
sequences you’d want to be explicit if it is the forward strand:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">122</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="n">strand</span><span class="o">=+</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">[122:150](+)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
<span class="go">1</span>
</pre></div>
</div>
<p>Note that for a parent sequence of length n, the FeatureLocation
start and end must satisfy the inequality 0 &lt;= start &lt;= end &lt;= n.
This means even for features on the reverse strand of a nucleotide
sequence, we expect the ‘start’ coordinate to be less than the
‘end’.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">122</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
<span class="go">[122:150](-)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
<span class="go">122</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
<span class="go">150</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
<span class="go">-1</span>
</pre></div>
</div>
<p>i.e. Rather than thinking of the ‘start’ and ‘end’ biologically in a
strand aware manner, think of them as the ‘left most’ or ‘minimum’
boundary, and the ‘right most’ or ‘maximum’ boundary of the region
being described. This is particularly important with compound
locations describing non-continuous regions.</p>
<p>In the example above we have used standard exact positions, but there
are also specialised position objects used to represent fuzzy positions
as well, for example a GenBank location like complement(&lt;123..150)
would use a BeforePosition object for the start.</p>
<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
<p>start and end arguments specify the values where the feature begins
and ends. These can either by any of the <code class="docutils literal notranslate"><span class="pre">*Position</span></code> objects that
inherit from AbstractPosition, or can just be integers specifying the
position. In the case of integers, the values are assumed to be
exact and are converted in ExactPosition arguments. This is meant
to make it easy to deal with non-fuzzy ends.</p>
<p>i.e. Short form:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">[5:10](-)</span>
</pre></div>
</div>
<p>Explicit form:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span><span class="p">,</span> <span class="n">ExactPosition</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">ExactPosition</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">[5:10](-)</span>
</pre></div>
</div>
<p>Other fuzzy positions are used similarly,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">BeforePosition</span><span class="p">,</span> <span class="n">AfterPosition</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loc2</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">BeforePosition</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">loc2</span><span class="p">)</span>
<span class="go">[&lt;5:&gt;10](-)</span>
</pre></div>
</div>
<p>For nucleotide features you will also want to specify the strand,
use 1 for the forward (plus) strand, -1 for the reverse (negative)
strand, 0 for stranded but strand unknown (? in GFF3), or None for
when the strand does not apply (dot in GFF3), e.g. features on
proteins.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">strand</span><span class="o">=+</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">[5:10](+)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">loc</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
<span class="go">1</span>
</pre></div>
</div>
<p>Normally feature locations are given relative to the parent
sequence you are working with, but an explicit accession can
be given with the optional ref and db_ref strings:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">105172</span><span class="p">,</span> <span class="mi">108462</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s2">&quot;AL391218.9&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">AL391218.9[105172:108462](+)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">loc</span><span class="o">.</span><span class="n">ref</span><span class="p">)</span>
<span class="go">AL391218.9</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.strand">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">strand</span></code><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Strand of the location (+1, -1, 0 or None).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the FeatureLocation object (with python counting).</p>
<p>For the simple case this uses the python splicing syntax, [122:150]
(zero based counting) which GenBank would call 123..150 (one based
counting).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the FeatureLocation object as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__add__">
<code class="sig-name descname"><span class="pre">__add__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Combine location with another FeatureLocation object, or shift it.</p>
<p>You can add two feature locations to make a join CompoundLocation:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">combined</span> <span class="o">=</span> <span class="n">f1</span> <span class="o">+</span> <span class="n">f2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">combined</span><span class="p">)</span>
<span class="go">join{[5:10], [20:30]}</span>
</pre></div>
</div>
<p>This is thus equivalent to:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">CompoundLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">join</span> <span class="o">=</span> <span class="n">CompoundLocation</span><span class="p">([</span><span class="n">f1</span><span class="p">,</span> <span class="n">f2</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">join</span><span class="p">)</span>
<span class="go">join{[5:10], [20:30]}</span>
</pre></div>
</div>
<p>You can also use sum(…) in this way:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">join</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">f1</span><span class="p">,</span> <span class="n">f2</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">join</span><span class="p">)</span>
<span class="go">join{[5:10], [20:30]}</span>
</pre></div>
</div>
<p>Furthermore, you can combine a FeatureLocation with a CompoundLocation
in this way.</p>
<p>Separately, adding an integer will give a new FeatureLocation with
its start and end offset by that amount. For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f1</span><span class="p">)</span>
<span class="go">[5:10]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f1</span> <span class="o">+</span> <span class="mi">100</span><span class="p">)</span>
<span class="go">[105:110]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="mi">200</span> <span class="o">+</span> <span class="n">f1</span><span class="p">)</span>
<span class="go">[205:210]</span>
</pre></div>
</div>
<p>This can be useful when editing annotation.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__radd__">
<code class="sig-name descname"><span class="pre">__radd__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__radd__" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a feature locationanother FeatureLocation object to the left.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__nonzero__">
<code class="sig-name descname"><span class="pre">__nonzero__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__nonzero__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True regardless of the length of the feature.</p>
<p>This behaviour is for backwards compatibility, since until the
__len__ method was added, a FeatureLocation always evaluated as True.</p>
<p>Note that in comparison, Seq objects, strings, lists, etc, will all
evaluate to False if they have length zero.</p>
<p>WARNING: The FeatureLocation may in future evaluate to False when its
length is zero (in order to better match normal python behaviour)!</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__len__">
<code class="sig-name descname"><span class="pre">__len__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__len__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the region described by the FeatureLocation object.</p>
<p>Note that extra care may be needed for fuzzy locations, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">BeforePosition</span><span class="p">,</span> <span class="n">AfterPosition</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">BeforePosition</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">5</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__contains__">
<code class="sig-name descname"><span class="pre">__contains__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__contains__" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if an integer position is within the FeatureLocation object.</p>
<p>Note that extra care may be needed for fuzzy locations, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">BeforePosition</span><span class="p">,</span> <span class="n">AfterPosition</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">BeforePosition</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">loc</span><span class="p">]</span>
<span class="go">[5, 6, 7, 8, 9]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__iter__">
<code class="sig-name descname"><span class="pre">__iter__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__iter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over the parent positions within the FeatureLocation object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">BeforePosition</span><span class="p">,</span> <span class="n">AfterPosition</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">BeforePosition</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">loc</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
<span class="go">5</span>
<span class="go">6</span>
<span class="go">7</span>
<span class="go">8</span>
<span class="go">9</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">[5, 6, 7, 8, 9]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">loc</span><span class="p">]</span>
<span class="go">[5, 6, 7, 8, 9]</span>
</pre></div>
</div>
<p>Note this is strand aware:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">BeforePosition</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span> <span class="n">strand</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
<span class="go">[9, 8, 7, 6, 5]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.__eq__">
<code class="sig-name descname"><span class="pre">__eq__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement equality by comparing all the location attributes.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.parts">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">parts</span></code><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.parts" title="Permalink to this definition">¶</a></dt>
<dd><p>Read only list of sections (always one, the FeatureLocation object).</p>
<p>This is a convenience property allowing you to write code handling
both simple FeatureLocation objects (with one part) and more complex
CompoundLocation objects (with multiple parts) interchangeably.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.start">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">start</span></code><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.start" title="Permalink to this definition">¶</a></dt>
<dd><p>Start location - left most (minimum) value, regardless of strand.</p>
<p>Read only, returns an integer like position object, possibly a fuzzy
position.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.end">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">end</span></code><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.end" title="Permalink to this definition">¶</a></dt>
<dd><p>End location - right most (maximum) value, regardless of strand.</p>
<p>Read only, returns an integer like position object, possibly a fuzzy
position.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.nofuzzy_start">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">nofuzzy_start</span></code><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.nofuzzy_start" title="Permalink to this definition">¶</a></dt>
<dd><p>Start position (integer, approximated if fuzzy, read only) (OBSOLETE).</p>
<p>This is now an alias for int(feature.start), which should be
used in preference – unless you are trying to support old
versions of Biopython.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.nofuzzy_end">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">nofuzzy_end</span></code><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.nofuzzy_end" title="Permalink to this definition">¶</a></dt>
<dd><p>End position (integer, approximated if fuzzy, read only) (OBSOLETE).</p>
<p>This is now an alias for int(feature.end), which should be
used in preference – unless you are trying to support old
versions of Biopython.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.FeatureLocation.extract">
<code class="sig-name descname"><span class="pre">extract</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">references</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.extract" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the sequence from supplied parent sequence using the FeatureLocation object.</p>
<p>The parent_sequence can be a Seq like object or a string, and will
generally return an object of the same type. The exception to this is
a MutableSeq as the parent sequence will return a Seq object.
If the location refers to other records, they must be supplied
in the optional dictionary references.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MKQHKAMIVALIVICITAVVAAL&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">feature_loc</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">feature_loc</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">Seq(&#39;VALIVIC&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt id="Bio.SeqFeature.FeatureLocation.__hash__">
<code class="sig-name descname"><span class="pre">__hash__</span></code><em class="property"> <span class="pre">=</span> <span class="pre">None</span></em><a class="headerlink" href="#Bio.SeqFeature.FeatureLocation.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.CompoundLocation">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">CompoundLocation</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">operator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'join'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>For handling joins etc where a feature location has several parts.</p>
<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">operator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'join'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span><span class="p">,</span> <span class="n">CompoundLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="n">strand</span><span class="o">=+</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">59</span><span class="p">,</span> <span class="n">strand</span><span class="o">=+</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">CompoundLocation</span><span class="p">([</span><span class="n">f1</span><span class="p">,</span> <span class="n">f2</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">f1</span><span class="p">)</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">f2</span><span class="p">)</span> <span class="o">==</span> <span class="mi">39</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">f</span><span class="p">))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">operator</span><span class="p">)</span>
<span class="go">join</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">5</span> <span class="ow">in</span> <span class="n">f</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">15</span> <span class="ow">in</span> <span class="n">f</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">strand</span>
<span class="go">1</span>
</pre></div>
</div>
<p>Notice that the strand of the compound location is computed
automatically - in the case of mixed strands on the sub-locations
the overall strand is set to None.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">CompoundLocation</span><span class="p">([</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="n">strand</span><span class="o">=+</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>                      <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
<span class="go">None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">[3, 4, 5, 12, 11, 10]</span>
</pre></div>
</div>
<p>The example above doing list(f) iterates over the coordinates within the
feature. This allows you to use max and min on the location, to find the
range covered:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">min</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">max</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">12</span>
</pre></div>
</div>
<p>More generally, you can use the compound location’s start and end which
give the full span covered, 0 &lt;= start &lt;= end &lt;= full sequence length.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">start</span> <span class="o">==</span> <span class="nb">min</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">end</span> <span class="o">==</span> <span class="nb">max</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
<span class="go">True</span>
</pre></div>
</div>
<p>This is consistent with the behaviour of the simple FeatureLocation for
a single region, where again the ‘start’ and ‘end’ do not necessarily
give the biological start and end, but rather the ‘minimal’ and ‘maximal’
coordinate boundaries.</p>
<p>Note that adding locations provides a more intuitive method of
construction:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="n">strand</span><span class="o">=+</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">13</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">[3, 4, 5, 12, 11, 10]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the CompoundLocation object (with python counting).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the CompoundLocation object as string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.strand">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">strand</span></code><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Overall strand of the compound location.</p>
<p>If all the parts have the same strand, that is returned. Otherwise
for mixed strands, this returns None.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span><span class="p">,</span> <span class="n">CompoundLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">17</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">f1</span> <span class="o">+</span> <span class="n">f2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">strand</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">strand</span>
<span class="go">-1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">strand</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">strand</span> <span class="ow">is</span> <span class="kc">None</span>
<span class="go">True</span>
</pre></div>
</div>
<p>If you set the strand of a CompoundLocation, this is applied to
all the parts - use with caution:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span><span class="o">.</span><span class="n">strand</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span><span class="o">.</span><span class="n">strand</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">strand</span>
<span class="go">1</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__add__">
<code class="sig-name descname"><span class="pre">__add__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Combine locations, or shift the location by an integer offset.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f1</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">17</span><span class="p">)</span> <span class="o">+</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">30</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f1</span><span class="p">)</span>
<span class="go">join{[15:17], [20:30]}</span>
</pre></div>
</div>
<p>You can add another FeatureLocation:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f1</span> <span class="o">+</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">40</span><span class="p">,</span> <span class="mi">50</span><span class="p">))</span>
<span class="go">join{[15:17], [20:30], [40:50]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span> <span class="o">+</span> <span class="n">f1</span><span class="p">)</span>
<span class="go">join{[5:10], [15:17], [20:30]}</span>
</pre></div>
</div>
<p>You can also add another CompoundLocation:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f2</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">40</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span> <span class="o">+</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">60</span><span class="p">,</span> <span class="mi">70</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f2</span><span class="p">)</span>
<span class="go">join{[40:50], [60:70]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f1</span> <span class="o">+</span> <span class="n">f2</span><span class="p">)</span>
<span class="go">join{[15:17], [20:30], [40:50], [60:70]}</span>
</pre></div>
</div>
<p>Also, as with the FeatureLocation, adding an integer shifts the
location’s co-ordinates by that offset:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f1</span> <span class="o">+</span> <span class="mi">100</span><span class="p">)</span>
<span class="go">join{[115:117], [120:130]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="mi">200</span> <span class="o">+</span> <span class="n">f1</span><span class="p">)</span>
<span class="go">join{[215:217], [220:230]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">f1</span> <span class="o">+</span> <span class="p">(</span><span class="o">-</span><span class="mi">5</span><span class="p">))</span>
<span class="go">join{[10:12], [15:25]}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__radd__">
<code class="sig-name descname"><span class="pre">__radd__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__radd__" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a feature to the left.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__contains__">
<code class="sig-name descname"><span class="pre">__contains__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__contains__" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if an integer position is within the CompoundLocation object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__nonzero__">
<code class="sig-name descname"><span class="pre">__nonzero__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__nonzero__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True regardless of the length of the feature.</p>
<p>This behaviour is for backwards compatibility, since until the
__len__ method was added, a FeatureLocation always evaluated as True.</p>
<p>Note that in comparison, Seq objects, strings, lists, etc, will all
evaluate to False if they have length zero.</p>
<p>WARNING: The FeatureLocation may in future evaluate to False when its
length is zero (in order to better match normal python behaviour)!</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__len__">
<code class="sig-name descname"><span class="pre">__len__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__len__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the CompoundLocation object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__iter__">
<code class="sig-name descname"><span class="pre">__iter__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__iter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over the parent positions within the CompoundLocation object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.__eq__">
<code class="sig-name descname"><span class="pre">__eq__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if all parts of CompoundLocation are equal to all parts of other CompoundLocation.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.start">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">start</span></code><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.start" title="Permalink to this definition">¶</a></dt>
<dd><p>Start location - left most (minimum) value, regardless of strand.</p>
<p>Read only, returns an integer like position object, possibly a fuzzy
position.</p>
<p>For the special case of a CompoundLocation wrapping the origin of a
circular genome, this will return zero.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.end">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">end</span></code><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.end" title="Permalink to this definition">¶</a></dt>
<dd><p>End location - right most (maximum) value, regardless of strand.</p>
<p>Read only, returns an integer like position object, possibly a fuzzy
position.</p>
<p>For the special case of a CompoundLocation wrapping the origin of
a circular genome this will match the genome length (minus one
given how Python counts from zero).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.nofuzzy_start">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">nofuzzy_start</span></code><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.nofuzzy_start" title="Permalink to this definition">¶</a></dt>
<dd><p>Start position (integer, approximated if fuzzy, read only) (OBSOLETE).</p>
<p>This is an alias for int(feature.start), which should be used in
preference – unless you are trying to support old versions of
Biopython.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.nofuzzy_end">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">nofuzzy_end</span></code><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.nofuzzy_end" title="Permalink to this definition">¶</a></dt>
<dd><p>End position (integer, approximated if fuzzy, read only) (OBSOLETE).</p>
<p>This is an alias for int(feature.end), which should be used in
preference – unless you are trying to support old versions of
Biopython.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.ref">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">ref</span></code><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.ref" title="Permalink to this definition">¶</a></dt>
<dd><p>Not present in CompoundLocation, dummy method for API compatibility.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.ref_db">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">ref_db</span></code><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.ref_db" title="Permalink to this definition">¶</a></dt>
<dd><p>Not present in CompoundLocation, dummy method for API compatibility.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.CompoundLocation.extract">
<code class="sig-name descname"><span class="pre">extract</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">references</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.extract" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the sequence from supplied parent sequence using the CompoundLocation object.</p>
<p>The parent_sequence can be a Seq like object or a string, and will
generally return an object of the same type. The exception to this is
a MutableSeq as the parent sequence will return a Seq object.
If the location refers to other records, they must be supplied
in the optional dictionary references.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">FeatureLocation</span><span class="p">,</span> <span class="n">CompoundLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MKQHKAMIVALIVICITAVVAAL&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fl1</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fl2</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fl3</span> <span class="o">=</span> <span class="n">CompoundLocation</span><span class="p">([</span><span class="n">fl1</span><span class="p">,</span><span class="n">fl2</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fl3</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">Seq(&#39;QHKAMILIVIC&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt id="Bio.SeqFeature.CompoundLocation.__hash__">
<code class="sig-name descname"><span class="pre">__hash__</span></code><em class="property"> <span class="pre">=</span> <span class="pre">None</span></em><a class="headerlink" href="#Bio.SeqFeature.CompoundLocation.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.AbstractPosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">AbstractPosition</span></code><a class="headerlink" href="#Bio.SeqFeature.AbstractPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Abstract base class representing a position.</p>
<dl class="py method">
<dt id="Bio.SeqFeature.AbstractPosition.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.AbstractPosition.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the AbstractPosition object as a string for debugging.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.ExactPosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">ExactPosition</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.ExactPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <a class="reference internal" href="#Bio.SeqFeature.AbstractPosition" title="Bio.SeqFeature.AbstractPosition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.AbstractPosition</span></code></a></p>
<p>Specify the specific position of a boundary.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>position - The position of the boundary.</p></li>
<li><p>extension - An optional argument which must be zero since we don’t
have an extension. The argument is provided so that the same number
of arguments can be passed to all position types.</p></li>
</ul>
</dd>
</dl>
<p>In this case, there is no fuzziness associated with the position.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">ExactPosition(5)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">5</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">AbstractPosition</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Integer comparisons and operations should work as expected:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="mi">5</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">&lt;</span> <span class="mi">6</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">&lt;=</span> <span class="mi">5</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">+</span> <span class="mi">10</span>
<span class="go">15</span>
</pre></div>
</div>
<dl class="py method">
<dt id="Bio.SeqFeature.ExactPosition.__new__">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">__new__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.ExactPosition.__new__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an ExactPosition object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.ExactPosition.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.ExactPosition.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the ExactPosition object (with python counting).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.ExactPosition.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.ExactPosition.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the ExactPosition object as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.ExactPosition.position">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">position</span></code><a class="headerlink" href="#Bio.SeqFeature.ExactPosition.position" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get position as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.ExactPosition.extension">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">extension</span></code><a class="headerlink" href="#Bio.SeqFeature.ExactPosition.extension" title="Permalink to this definition">¶</a></dt>
<dd><p>Not present in this object, return zero (OBSOLETE).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.UncertainPosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">UncertainPosition</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.UncertainPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.SeqFeature.ExactPosition" title="Bio.SeqFeature.ExactPosition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.ExactPosition</span></code></a></p>
<p>Specify a specific position which is uncertain.</p>
<p>This is used in UniProt, e.g. ?222 for uncertain position 222, or in the
XML format explicitly marked as uncertain. Does not apply to GenBank/EMBL.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.UnknownPosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">UnknownPosition</span></code><a class="headerlink" href="#Bio.SeqFeature.UnknownPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.SeqFeature.AbstractPosition" title="Bio.SeqFeature.AbstractPosition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.AbstractPosition</span></code></a></p>
<p>Specify a specific position which is unknown (has no position).</p>
<p>This is used in UniProt, e.g. ? or in the XML as unknown.</p>
<dl class="py method">
<dt id="Bio.SeqFeature.UnknownPosition.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.UnknownPosition.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the UnknownPosition object as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.UnknownPosition.__hash__">
<code class="sig-name descname"><span class="pre">__hash__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.UnknownPosition.__hash__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the hash value of the UnknownPosition object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.UnknownPosition.position">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">position</span></code><a class="headerlink" href="#Bio.SeqFeature.UnknownPosition.position" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get location (None) (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.UnknownPosition.extension">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">extension</span></code><a class="headerlink" href="#Bio.SeqFeature.UnknownPosition.extension" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get extension (zero) as integer (OBSOLETE).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.WithinPosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">WithinPosition</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.WithinPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <a class="reference internal" href="#Bio.SeqFeature.AbstractPosition" title="Bio.SeqFeature.AbstractPosition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.AbstractPosition</span></code></a></p>
<p>Specify the position of a boundary within some coordinates.</p>
<p>Arguments:
- position - The default integer position
- left - The start (left) position of the boundary
- right - The end (right) position of the boundary</p>
<p>This allows dealing with a location like ((1.4)..100). This
indicates that the start of the sequence is somewhere between 1
and 4. Since this is a start coordinate, it should acts like
it is at position 1 (or in Python counting, 0).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">WithinPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">13</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">WithinPosition(10, left=10, right=13)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">(10.13)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">10</span>
</pre></div>
</div>
<p>Basic integer comparisons and operations should work as though
this were a plain integer:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="mi">10</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">11</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">&lt;</span> <span class="mi">11</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">+</span> <span class="mi">10</span>
<span class="go">20</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">WithinPosition</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">AbstractPosition</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Note this also applies for comparison to other position objects,
where again the integer behaviour is used:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="mi">10</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="n">BeforePosition</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>If this were an end point, you would want the position to be 13:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">WithinPosition</span><span class="p">(</span><span class="mi">13</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">13</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span>
<span class="go">WithinPosition(13, left=10, right=13)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span>
<span class="go">(10.13)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span>
<span class="go">13</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">==</span> <span class="mi">13</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">==</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">13</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The old legacy properties of position and extension give the
starting/lower/left position as an integer, and the distance
to the ending/higher/right position as an integer. Note that
the position object will act like either the left or the right
end-point depending on how it was created:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">position</span> <span class="o">==</span> <span class="n">p2</span><span class="o">.</span><span class="n">position</span> <span class="o">==</span> <span class="mi">10</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">extension</span> <span class="o">==</span> <span class="n">p2</span><span class="o">.</span><span class="n">extension</span> <span class="o">==</span> <span class="mi">3</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="mi">10</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">==</span> <span class="mi">13</span>
<span class="go">True</span>
</pre></div>
</div>
<dl class="py method">
<dt id="Bio.SeqFeature.WithinPosition.__new__">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">__new__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.WithinPosition.__new__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a WithinPosition object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.WithinPosition.__getnewargs__">
<code class="sig-name descname"><span class="pre">__getnewargs__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.WithinPosition.__getnewargs__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the arguments accepted by __new__.</p>
<p>Necessary to allow pickling and unpickling of class instances.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.WithinPosition.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.WithinPosition.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the WithinPosition object as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.WithinPosition.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.WithinPosition.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the WithinPosition object (with python counting).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.WithinPosition.position">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">position</span></code><a class="headerlink" href="#Bio.SeqFeature.WithinPosition.position" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get (left) position as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.WithinPosition.extension">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">extension</span></code><a class="headerlink" href="#Bio.SeqFeature.WithinPosition.extension" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get extension (from left to right) as an integer (OBSOLETE).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.BetweenPosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">BetweenPosition</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BetweenPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <a class="reference internal" href="#Bio.SeqFeature.AbstractPosition" title="Bio.SeqFeature.AbstractPosition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.AbstractPosition</span></code></a></p>
<p>Specify the position of a boundary between two coordinates (OBSOLETE?).</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>position - The default integer position</p></li>
<li><p>left - The start (left) position of the boundary</p></li>
<li><p>right - The end (right) position of the boundary</p></li>
</ul>
</dd>
</dl>
<p>This allows dealing with a position like 123^456. This
indicates that the start of the sequence is somewhere between
123 and 456. It is up to the parser to set the position argument
to either boundary point (depending on if this is being used as
a start or end of the feature). For example as a feature end:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">BetweenPosition</span><span class="p">(</span><span class="mi">456</span><span class="p">,</span> <span class="mi">123</span><span class="p">,</span> <span class="mi">456</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">BetweenPosition(456, left=123, right=456)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">(123^456)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">456</span>
</pre></div>
</div>
<p>Integer equality and comparison use the given position,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="mi">456</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">455</span><span class="p">,</span> <span class="mi">456</span><span class="p">,</span> <span class="mi">457</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">&gt;</span> <span class="mi">300</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The old legacy properties of position and extension give the
starting/lower/left position as an integer, and the distance
to the ending/higher/right position as an integer. Note that
the position object will act like either the left or the right
end-point depending on how it was created:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">BetweenPosition</span><span class="p">(</span><span class="mi">123</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">456</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">position</span> <span class="o">==</span> <span class="n">p2</span><span class="o">.</span><span class="n">position</span> <span class="o">==</span> <span class="mi">123</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">extension</span>
<span class="go">333</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span><span class="o">.</span><span class="n">extension</span>
<span class="go">333</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">extension</span> <span class="o">==</span> <span class="n">p2</span><span class="o">.</span><span class="n">extension</span> <span class="o">==</span> <span class="mi">333</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="mi">456</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">==</span> <span class="mi">123</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Note this potentially surprising behaviour:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">BetweenPosition</span><span class="p">(</span><span class="mi">123</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">456</span><span class="p">)</span> <span class="o">==</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">BetweenPosition</span><span class="p">(</span><span class="mi">123</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">456</span><span class="p">)</span> <span class="o">==</span> <span class="n">BeforePosition</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">BetweenPosition</span><span class="p">(</span><span class="mi">123</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="mi">123</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">456</span><span class="p">)</span> <span class="o">==</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>i.e. For equality (and sorting) the position objects behave like
integers.</p>
<dl class="py method">
<dt id="Bio.SeqFeature.BetweenPosition.__new__">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">__new__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BetweenPosition.__new__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance in BetweenPosition object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BetweenPosition.__getnewargs__">
<code class="sig-name descname"><span class="pre">__getnewargs__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BetweenPosition.__getnewargs__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the arguments accepted by __new__.</p>
<p>Necessary to allow pickling and unpickling of class instances.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BetweenPosition.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BetweenPosition.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the BetweenPosition object as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BetweenPosition.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BetweenPosition.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the BetweenPosition object (with python counting).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BetweenPosition.position">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">position</span></code><a class="headerlink" href="#Bio.SeqFeature.BetweenPosition.position" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get (left) position as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BetweenPosition.extension">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">extension</span></code><a class="headerlink" href="#Bio.SeqFeature.BetweenPosition.extension" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get extension (from left to right) as an integer (OBSOLETE).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.BeforePosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">BeforePosition</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BeforePosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <a class="reference internal" href="#Bio.SeqFeature.AbstractPosition" title="Bio.SeqFeature.AbstractPosition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.AbstractPosition</span></code></a></p>
<p>Specify a position where the actual location occurs before it.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>position - The upper boundary of where the location can occur.</p></li>
<li><p>extension - An optional argument which must be zero since we don’t
have an extension. The argument is provided so that the same number
of arguments can be passed to all position types.</p></li>
</ul>
</dd>
</dl>
<p>This is used to specify positions like (&lt;10..100) where the location
occurs somewhere before position 10.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">BeforePosition</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">BeforePosition(5)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">&lt;5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">+</span> <span class="mi">10</span>
<span class="go">15</span>
</pre></div>
</div>
<p>Note this potentially surprising behaviour:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Just remember that for equality and sorting the position objects act
like integers.</p>
<dl class="py method">
<dt id="Bio.SeqFeature.BeforePosition.__new__">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">__new__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BeforePosition.__new__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance in BeforePosition object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BeforePosition.position">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">position</span></code><a class="headerlink" href="#Bio.SeqFeature.BeforePosition.position" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get position as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BeforePosition.extension">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">extension</span></code><a class="headerlink" href="#Bio.SeqFeature.BeforePosition.extension" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get extension (zero) as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BeforePosition.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BeforePosition.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the location as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.BeforePosition.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.BeforePosition.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the BeforePosition object (with python counting).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.AfterPosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">AfterPosition</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.AfterPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <a class="reference internal" href="#Bio.SeqFeature.AbstractPosition" title="Bio.SeqFeature.AbstractPosition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.AbstractPosition</span></code></a></p>
<p>Specify a position where the actual location is found after it.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>position - The lower boundary of where the location can occur.</p></li>
<li><p>extension - An optional argument which must be zero since we don’t
have an extension. The argument is provided so that the same number
of arguments can be passed to all position types.</p></li>
</ul>
</dd>
</dl>
<p>This is used to specify positions like (&gt;10..100) where the location
occurs somewhere after position 10.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">AfterPosition</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">AfterPosition(7)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">&gt;7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">+</span> <span class="mi">10</span>
<span class="go">17</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">AfterPosition</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">AbstractPosition</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Note this potentially surprising behaviour:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="n">BeforePosition</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Just remember that for equality and sorting the position objects act
like integers.</p>
<dl class="py method">
<dt id="Bio.SeqFeature.AfterPosition.__new__">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">__new__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.AfterPosition.__new__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance of the AfterPosition object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.AfterPosition.position">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">position</span></code><a class="headerlink" href="#Bio.SeqFeature.AfterPosition.position" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get position as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.AfterPosition.extension">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">extension</span></code><a class="headerlink" href="#Bio.SeqFeature.AfterPosition.extension" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get extension (zero) as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.AfterPosition.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.AfterPosition.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the location as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.AfterPosition.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.AfterPosition.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the AfterPosition object (with python counting).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.OneOfPosition">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">OneOfPosition</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choices</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.OneOfPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <a class="reference internal" href="#Bio.SeqFeature.AbstractPosition" title="Bio.SeqFeature.AbstractPosition"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.AbstractPosition</span></code></a></p>
<p>Specify a position where the location can be multiple positions.</p>
<p>This models the GenBank ‘one-of(1888,1901)’ function, and tries
to make this fit within the Biopython Position models. If this was
a start position it should act like 1888, but as an end position 1901.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">OneOfPosition</span><span class="p">(</span><span class="mi">1888</span><span class="p">,</span> <span class="p">[</span><span class="n">ExactPosition</span><span class="p">(</span><span class="mi">1888</span><span class="p">),</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">1901</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span>
<span class="go">OneOfPosition(1888, choices=[ExactPosition(1888), ExactPosition(1901)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="go">1888</span>
</pre></div>
</div>
<p>Integer comparisons and operators act like using int(p),</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="mi">1888</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">&lt;=</span> <span class="mi">1888</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">&gt;</span> <span class="mi">1888</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">+</span> <span class="mi">100</span>
<span class="go">1988</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">OneOfPosition</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">AbstractPosition</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The old legacy properties of position and extension give the
starting/lowest/left-most position as an integer, and the
distance to the ending/highest/right-most position as an integer.
Note that the position object will act like one of the list of
possible locations depending on how it was created:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">=</span> <span class="n">OneOfPosition</span><span class="p">(</span><span class="mi">1901</span><span class="p">,</span> <span class="p">[</span><span class="n">ExactPosition</span><span class="p">(</span><span class="mi">1888</span><span class="p">),</span> <span class="n">ExactPosition</span><span class="p">(</span><span class="mi">1901</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">position</span> <span class="o">==</span> <span class="n">p2</span><span class="o">.</span><span class="n">position</span> <span class="o">==</span> <span class="mi">1888</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span><span class="o">.</span><span class="n">extension</span> <span class="o">==</span> <span class="n">p2</span><span class="o">.</span><span class="n">extension</span> <span class="o">==</span> <span class="mi">13</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">==</span> <span class="mi">1888</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p2</span> <span class="o">==</span> <span class="mi">1901</span>
<span class="go">True</span>
</pre></div>
</div>
<dl class="py method">
<dt id="Bio.SeqFeature.OneOfPosition.__new__">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">__new__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">choices</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.OneOfPosition.__new__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize with a set of possible positions.</p>
<p>position_list is a list of AbstractPosition derived objects,
specifying possible locations.</p>
<p>position is an integer specifying the default behaviour.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.OneOfPosition.__getnewargs__">
<code class="sig-name descname"><span class="pre">__getnewargs__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.OneOfPosition.__getnewargs__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the arguments accepted by __new__.</p>
<p>Necessary to allow pickling and unpickling of class instances.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.OneOfPosition.position">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">position</span></code><a class="headerlink" href="#Bio.SeqFeature.OneOfPosition.position" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get (left) position as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.OneOfPosition.extension">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">extension</span></code><a class="headerlink" href="#Bio.SeqFeature.OneOfPosition.extension" title="Permalink to this definition">¶</a></dt>
<dd><p>Legacy attribute to get extension as integer (OBSOLETE).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.OneOfPosition.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.OneOfPosition.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the OneOfPosition object as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.OneOfPosition.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.OneOfPosition.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the OneOfPosition object (with python counting).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SeqFeature.PositionGap">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.SeqFeature.</span></code><code class="sig-name descname"><span class="pre">PositionGap</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gap_size</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.PositionGap" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Simple class to hold information about a gap between positions.</p>
<dl class="py method">
<dt id="Bio.SeqFeature.PositionGap.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_size</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.PositionGap.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Intialize with a position object containing the gap information.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.PositionGap.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.PositionGap.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the position gap as a string for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqFeature.PositionGap.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqFeature.PositionGap.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the PositionGap object (with python counting).</p>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="Bio.SeqRecord.html" class="btn btn-neutral float-right" title="Bio.SeqRecord module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="Bio.Seq.html" class="btn btn-neutral float-left" title="Bio.Seq module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 1999-2020, The Biopython Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.79.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.77/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>