<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.PopGen.GenePop.Controller module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.PopGen.GenePop.EasyController module" href="Bio.PopGen.GenePop.EasyController.html" />
    <link rel="prev" title="Bio.PopGen.GenePop package" href="Bio.PopGen.GenePop.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PopGen.html">Bio.PopGen package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PopGen.GenePop.html">Bio.PopGen.GenePop package</a></li>
      <li class="breadcrumb-item active">Bio.PopGen.GenePop.Controller module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/PopGen/GenePop/Controller.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.PopGen.GenePop.html" class="btn btn-neutral float-left" title="Bio.PopGen.GenePop package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PopGen.GenePop.EasyController.html" class="btn btn-neutral float-right" title="Bio.PopGen.GenePop.EasyController module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.PopGen.GenePop.Controller">
<span id="bio-popgen-genepop-controller-module"></span><h1>Bio.PopGen.GenePop.Controller module<a class="headerlink" href="#module-Bio.PopGen.GenePop.Controller" title="Permalink to this headline"></a></h1>
<p>Module to control GenePop.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PopGen.GenePop.Controller.</span></span><span class="sig-name descname"><span class="pre">GenePopController</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genepop_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Define a class to interface with the GenePop program.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genepop_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the controller.</p>
<p>genepop_dir is the directory where GenePop is.</p>
<p>The binary should be called Genepop (capital G)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_deficiency">
<span class="sig-name descname"><span class="pre">test_pop_hz_deficiency</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enum_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_deficiency" title="Permalink to this definition"></a></dt>
<dd><p>Use Hardy-Weinberg test for heterozygote deficiency.</p>
<p>Returns a population iterator containing a dictionary wehre
dictionary[locus]=(P-val, SE, Fis-WC, Fis-RH, steps).</p>
<p>Some loci have a None if the info is not available.
SE might be none (for enumerations).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_excess">
<span class="sig-name descname"><span class="pre">test_pop_hz_excess</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enum_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_excess" title="Permalink to this definition"></a></dt>
<dd><p>Use Hardy-Weinberg test for heterozygote deficiency.</p>
<p>Returns a population iterator containing a dictionary where
dictionary[locus]=(P-val, SE, Fis-WC, Fis-RH, steps).</p>
<p>Some loci have a None if the info is not available.
SE might be none (for enumerations).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_prob">
<span class="sig-name descname"><span class="pre">test_pop_hz_prob</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ext</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enum_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_pop_hz_prob" title="Permalink to this definition"></a></dt>
<dd><p>Use Hardy-Weinberg test based on probability.</p>
<p>Returns 2 iterators and a final tuple:</p>
<blockquote>
<div><ol class="arabic simple">
<li><dl class="simple">
<dt>Returns a loci iterator containing:</dt><dd><ul class="simple">
<li><p>A dictionary[pop_pos]=(P-val, SE, Fis-WC, Fis-RH, steps).
Some pops have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>Result of Fisher’s test (Chi2, deg freedom, prob).</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Returns a population iterator containing:</dt><dd><ul class="simple">
<li><p>A dictionary[locus]=(P-val, SE, Fis-WC, Fis-RH, steps).
Some loci have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>Result of Fisher’s test (Chi2, deg freedom, prob).</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Final tuple (Chi2, deg freedom, prob).</p></li>
</ol>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_deficiency">
<span class="sig-name descname"><span class="pre">test_global_hz_deficiency</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enum_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_deficiency" title="Permalink to this definition"></a></dt>
<dd><p>Use Global Hardy-Weinberg test for heterozygote deficiency.</p>
<dl class="simple">
<dt>Returns a triple with:</dt><dd><ul class="simple">
<li><p>An list per population containing (pop_name, P-val, SE, switches).
Some pops have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>An list per loci containing (locus_name, P-val, SE, switches).
Some loci have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>Overall results (P-val, SE, switches).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_excess">
<span class="sig-name descname"><span class="pre">test_global_hz_excess</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enum_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_global_hz_excess" title="Permalink to this definition"></a></dt>
<dd><p>Use Global Hardy-Weinberg test for heterozygote excess.</p>
<dl class="simple">
<dt>Returns a triple with:</dt><dd><ul class="simple">
<li><p>A list per population containing (pop_name, P-val, SE, switches).
Some pops have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>A list per loci containing (locus_name, P-val, SE, switches).
Some loci have a None if the info is not available.
SE might be none (for enumerations).</p></li>
<li><p>Overall results (P-val, SE, switches)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_ld">
<span class="sig-name descname"><span class="pre">test_ld</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_ld" title="Permalink to this definition"></a></dt>
<dd><p>Test for linkage disequilibrium on each pair of loci in each population.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.create_contingency_tables">
<span class="sig-name descname"><span class="pre">create_contingency_tables</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.create_contingency_tables" title="Permalink to this definition"></a></dt>
<dd><p>Provision for creating Genotypic contingency tables.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_all">
<span class="sig-name descname"><span class="pre">test_genic_diff_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_all" title="Permalink to this definition"></a></dt>
<dd><p>Provision for Genic differentiation for all populations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_pair">
<span class="sig-name descname"><span class="pre">test_genic_diff_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genic_diff_pair" title="Permalink to this definition"></a></dt>
<dd><p>Provision for Genic differentiation for all population pairs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_all">
<span class="sig-name descname"><span class="pre">test_genotypic_diff_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_all" title="Permalink to this definition"></a></dt>
<dd><p>Provision for Genotypic differentiation for all populations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_pair">
<span class="sig-name descname"><span class="pre">test_genotypic_diff_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.test_genotypic_diff_pair" title="Permalink to this definition"></a></dt>
<dd><p>Provision for Genotypic differentiation for all population pairs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.estimate_nm">
<span class="sig-name descname"><span class="pre">estimate_nm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.estimate_nm" title="Permalink to this definition"></a></dt>
<dd><p>Estimate the Number of Migrants.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>fname - file name</p></li>
</ul>
</dd>
<dt>Returns</dt><dd><ul class="simple">
<li><p>Mean sample size</p></li>
<li><p>Mean frequency of private alleles</p></li>
<li><p>Number of migrants for Ne=10</p></li>
<li><p>Number of migrants for Ne=25</p></li>
<li><p>Number of migrants for Ne=50</p></li>
<li><p>Number of migrants after correcting for expected size</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_allele_genotype_freqs">
<span class="sig-name descname"><span class="pre">calc_allele_genotype_freqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_allele_genotype_freqs" title="Permalink to this definition"></a></dt>
<dd><p>Calculate allele and genotype frequencies per locus and per sample.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>fname - file name</p></li>
</ul>
</dd>
<dt>Returns tuple with 2 elements:</dt><dd><ul class="simple">
<li><p>Population iterator with</p>
<ul>
<li><p>population name</p></li>
<li><p>Locus dictionary with key = locus name and content tuple as
Genotype List with
(Allele1, Allele2, observed, expected)
(expected homozygotes, observed hm,
expected heterozygotes, observed ht)
Allele frequency/Fis dictionary with allele as key and
(count, frequency, Fis Weir &amp; Cockerham)</p></li>
<li><p>Totals as a pair</p></li>
<li><p>count</p></li>
<li><p>Fis Weir &amp; Cockerham,</p></li>
<li><p>Fis Robertson &amp; Hill</p></li>
</ul>
</li>
<li><p>Locus iterator with</p>
<ul>
<li><p>Locus name</p></li>
<li><p>allele list</p></li>
<li><p>Population list with a triple</p>
<ul>
<li><p>population name</p></li>
<li><p>list of allele frequencies in the same order as allele list above</p></li>
<li><p>number of genes</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</dd>
</dl>
<p>Will create a file called fname.INF</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_identity">
<span class="sig-name descname"><span class="pre">calc_diversities_fis_with_identity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_identity" title="Permalink to this definition"></a></dt>
<dd><p>Compute identity-base Gene diversities and Fis.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_size">
<span class="sig-name descname"><span class="pre">calc_diversities_fis_with_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_diversities_fis_with_size" title="Permalink to this definition"></a></dt>
<dd><p>Provision to Computer Allele size-based Gene diversities and Fis.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_all">
<span class="sig-name descname"><span class="pre">calc_fst_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_all" title="Permalink to this definition"></a></dt>
<dd><p>Execute GenePop and gets Fst/Fis/Fit (all populations).</p>
<dl class="simple">
<dt>Parameters:</dt><dd><ul class="simple">
<li><p>fname - file name</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><ul class="simple">
<li><p>(multiLocusFis, multiLocusFst, multiLocus Fit),</p></li>
<li><p>Iterator of tuples
(Locus name, Fis, Fst, Fit, Qintra, Qinter)</p></li>
</ul>
</dd>
</dl>
<p>Will create a file called <code class="docutils literal notranslate"><span class="pre">fname.FST</span></code>.</p>
<p>This does not return the genotype frequencies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_pair">
<span class="sig-name descname"><span class="pre">calc_fst_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_fst_pair" title="Permalink to this definition"></a></dt>
<dd><p>Estimate spatial structure from Allele identity for all population pairs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_all">
<span class="sig-name descname"><span class="pre">calc_rho_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_all" title="Permalink to this definition"></a></dt>
<dd><p>Provision for estimating spatial structure from Allele size for all populations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_pair">
<span class="sig-name descname"><span class="pre">calc_rho_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_rho_pair" title="Permalink to this definition"></a></dt>
<dd><p>Provision for estimating spatial structure from Allele size for all population pairs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_diplo">
<span class="sig-name descname"><span class="pre">calc_ibd_diplo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Log'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.00001</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_diplo" title="Permalink to this definition"></a></dt>
<dd><p>Calculate isolation by distance statistics for diploid data.</p>
<p>See _calc_ibd for parameter details.</p>
<p>Note that each pop can only have a single individual and
the individual name has to be the sample coordinates.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_haplo">
<span class="sig-name descname"><span class="pre">calc_ibd_haplo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Log'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.00001</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.Controller.GenePopController.calc_ibd_haplo" title="Permalink to this definition"></a></dt>
<dd><p>Calculate isolation by distance statistics for haploid data.</p>
<p>See _calc_ibd for parameter details.</p>
<p>Note that each pop can only have a single individual and
the individual name has to be the sample coordinates.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.PopGen.GenePop.html" class="btn btn-neutral float-left" title="Bio.PopGen.GenePop package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PopGen.GenePop.EasyController.html" class="btn btn-neutral float-right" title="Bio.PopGen.GenePop.EasyController module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>