<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.PDB.internal_coords module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.PDB.mmcifio module" href="Bio.PDB.mmcifio.html" />
    <link rel="prev" title="Bio.PDB.ic_rebuild module" href="Bio.PDB.ic_rebuild.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Welcome to Biopython’s API documentation!</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Welcome to Biopython’s API documentation!</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PDB.html">Bio.PDB package</a></li>
      <li class="breadcrumb-item active">Bio.PDB.internal_coords module</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/api/Bio.PDB.internal_coords.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.PDB.ic_rebuild.html" class="btn btn-neutral float-left" title="Bio.PDB.ic_rebuild module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PDB.mmcifio.html" class="btn btn-neutral float-right" title="Bio.PDB.mmcifio module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.PDB.internal_coords">
<span id="bio-pdb-internal-coords-module"></span><h1>Bio.PDB.internal_coords module<a class="headerlink" href="#module-Bio.PDB.internal_coords" title="Permalink to this headline"></a></h1>
<p>Classes to support internal coordinates for protein structures.</p>
<p>Internal coordinates comprise Psi, Omega and Phi dihedral angles along the
protein backbone, Chi angles along the sidechains, and all 3-atom angles and
bond lengths defining a protein chain.  These routines can compute internal
coordinates from atom XYZ coordinates, and compute atom XYZ coordinates from
internal coordinates.</p>
<p>Secondary benefits include the ability to align and compare residue
environments in 3D structures, support for 2D atom distance plots, converting a
distance plot plus chirality information to a structure, generating an OpenSCAD
description of a structure for 3D printing, and reading/writing structures as
internal coordinate data files.</p>
<p><strong>Usage:</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.PDB.PDBParser</span> <span class="kn">import</span> <span class="n">PDBParser</span>
<span class="kn">from</span> <span class="nn">Bio.PDB.Chain</span> <span class="kn">import</span> <span class="n">Chain</span>
<span class="kn">from</span> <span class="nn">Bio.PDB.internal_coords</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">Bio.PDB.PICIO</span> <span class="kn">import</span> <span class="n">write_PIC</span><span class="p">,</span> <span class="n">read_PIC</span><span class="p">,</span> <span class="n">read_PIC_seq</span>
<span class="kn">from</span> <span class="nn">Bio.PDB.ic_rebuild</span> <span class="kn">import</span> <span class="n">write_PDB</span><span class="p">,</span> <span class="n">IC_duplicate</span><span class="p">,</span> <span class="n">structure_rebuild_test</span>
<span class="kn">from</span> <span class="nn">Bio.PDB.SCADIO</span> <span class="kn">import</span> <span class="n">write_SCAD</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio.PDB.PDBIO</span> <span class="kn">import</span> <span class="n">PDBIO</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># load a structure as normal, get first chain</span>
<span class="n">parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span>
<span class="n">myProtein</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;7rsa&quot;</span><span class="p">,</span> <span class="s2">&quot;pdb7rsa.ent&quot;</span><span class="p">)</span>
<span class="n">myChain</span> <span class="o">=</span> <span class="n">myProtein</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;A&quot;</span><span class="p">]</span>

<span class="c1"># compute bond lengths, angles, dihedral angles</span>
<span class="n">myChain</span><span class="o">.</span><span class="n">atom_to_internal_coordinates</span><span class="p">(</span><span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># check myChain makes sense (can get angles and rebuild same structure)</span>
<span class="n">resultDict</span> <span class="o">=</span> <span class="n">structure_rebuild_test</span><span class="p">(</span><span class="n">myChain</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">resultDict</span><span class="p">[</span><span class="s1">&#39;pass&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="kc">True</span>

<span class="c1"># get residue 1 chi2 angle</span>
<span class="n">r1</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">myChain</span><span class="o">.</span><span class="n">get_residues</span><span class="p">())</span>
<span class="n">r1chi2</span> <span class="o">=</span> <span class="n">r1</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;chi2&quot;</span><span class="p">)</span>

<span class="c1"># rotate residue 1 chi2 angle by 120 degrees (loops w/in +/-180)</span>
<span class="n">r1</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">set_angle</span><span class="p">(</span><span class="s2">&quot;chi2&quot;</span><span class="p">,</span> <span class="n">r1chi2</span> <span class="o">+</span> <span class="mf">120.0</span><span class="p">)</span>
<span class="c1"># or</span>
<span class="n">r1</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">bond_rotate</span><span class="p">(</span><span class="s2">&quot;chi2&quot;</span><span class="p">,</span> <span class="mf">120.0</span><span class="p">)</span>
<span class="c1"># update myChain XYZ coordinates with chi2 changed</span>
<span class="n">myChain</span><span class="o">.</span><span class="n">internal_to_atom_coordinates</span><span class="p">()</span>
<span class="c1"># write new conformation with PDBIO</span>
<span class="n">write_PDB</span><span class="p">(</span><span class="n">myProtein</span><span class="p">,</span> <span class="s2">&quot;myChain.pdb&quot;</span><span class="p">)</span>
<span class="c1"># or just the ATOM records without headers:</span>
<span class="n">io</span> <span class="o">=</span> <span class="n">PDBIO</span><span class="p">()</span>
<span class="n">io</span><span class="o">.</span><span class="n">set_structure</span><span class="p">(</span><span class="n">myProtein</span><span class="p">)</span>
<span class="n">io</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s2">&quot;myChain2.pdb&quot;</span><span class="p">)</span>

<span class="c1"># write chain as &#39;protein internal coordinates&#39; (.pic) file</span>
<span class="n">write_PIC</span><span class="p">(</span><span class="n">myProtein</span><span class="p">,</span> <span class="s2">&quot;myChain.pic&quot;</span><span class="p">)</span>
<span class="c1"># read .pic file</span>
<span class="n">myProtein2</span> <span class="o">=</span> <span class="n">read_PIC</span><span class="p">(</span><span class="s2">&quot;myChain.pic&quot;</span><span class="p">)</span>

<span class="c1"># create default structure for random sequence by reading as .pic file</span>
<span class="n">myProtein3</span> <span class="o">=</span> <span class="n">read_PIC_seq</span><span class="p">(</span>
    <span class="n">SeqRecord</span><span class="p">(</span>
        <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GAVLIMFPSTCNQYWDEHKR&quot;</span><span class="p">),</span>
        <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;1RND&quot;</span><span class="p">,</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;my random sequence&quot;</span><span class="p">,</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">myProtein3</span><span class="o">.</span><span class="n">internal_to_atom_coordinates</span><span class="p">()</span>
<span class="n">write_PDB</span><span class="p">(</span><span class="n">myProtein3</span><span class="p">,</span> <span class="s2">&quot;myRandom.pdb&quot;</span><span class="p">)</span>

<span class="c1"># access the all-dihedrals array for the chain, e.g. residue 1 chi2 angle:</span>
<span class="n">r1chi2_obj</span> <span class="o">=</span> <span class="n">r1</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">pick_angle</span><span class="p">(</span><span class="s2">&quot;chi2&quot;</span><span class="p">)</span>
<span class="c1"># or same thing: r1chi2_obj = r1.internal_coord.pick_angle(&quot;CA:CB:CG:CD&quot;)</span>
<span class="n">r1chi2_key</span> <span class="o">=</span> <span class="n">r1chi2_obj</span><span class="o">.</span><span class="n">atomkeys</span>
<span class="c1"># r1chi2_key is tuple of AtomKeys (1_K_CA, 1_K_CB, 1_K_CG, 1_K_CD)</span>
<span class="n">r1chi2_index</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">dihedraNdx</span><span class="p">[</span><span class="n">r1chi2_key</span><span class="p">]</span>
<span class="c1"># or same thing: r1chi2_index = r1chi2_obj.ndx</span>
<span class="n">r1chi2_value</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">dihedraAngle</span><span class="p">[</span><span class="n">r1chi2_index</span><span class="p">]</span>
<span class="c1"># also true: r1chi2_obj == myChain.internal_coord.dihedra[r1chi2_index]</span>

<span class="c1"># access the array of all atoms for the chain, e.g. residue 1 C-beta</span>
<span class="n">r1_cBeta_index</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">atomArrayIndex</span><span class="p">[</span><span class="n">AtomKey</span><span class="p">(</span><span class="s2">&quot;1_K_CB&quot;</span><span class="p">)]</span>
<span class="n">r1_cBeta_coords</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">atomArray</span><span class="p">[</span><span class="n">r1_cBeta_index</span><span class="p">]</span>
<span class="c1"># r1_cBeta_coords = [ x, y, z, 1.0 ]</span>

<span class="c1"># the Biopython Atom coord array is now a view into atomArray, so</span>
<span class="k">assert</span> <span class="n">r1_cBeta_coords</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">r1</span><span class="p">[</span><span class="s2">&quot;CB&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">r1_cBeta_coords</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="mf">1.0</span>  <span class="c1"># change the Y coord 1 angstrom</span>
<span class="k">assert</span> <span class="n">r1_cBeta_coords</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">r1</span><span class="p">[</span><span class="s2">&quot;CB&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c1"># they are always the same (they share the same memory)</span>
<span class="n">r1_cBeta_coords</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-=</span> <span class="mf">1.0</span>  <span class="c1"># restore</span>

<span class="c1"># create a selector to filter just the C-alpha atoms from the all atom array</span>
<span class="n">atmNameNdx</span> <span class="o">=</span> <span class="n">AtomKey</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">atm</span>
<span class="n">atomArrayIndex</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">atomArrayIndex</span>
<span class="n">CaSelect</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">akl</span><span class="p">[</span><span class="n">atmNameNdx</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span>
<span class="p">]</span>
<span class="c1"># now the ordered array of C-alpha atom coordinates is:</span>
<span class="n">CA_coords</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">atomArray</span><span class="p">[</span><span class="n">CaSelect</span><span class="p">]</span>
<span class="c1"># note this uses Numpy fancy indexing, so CA_coords is a new copy</span>

<span class="c1"># create a C-alpha distance plot</span>
<span class="n">caDistances</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">distance_plot</span><span class="p">(</span><span class="n">CaSelect</span><span class="p">)</span>
<span class="c1"># display with e.g. MatPlotLib:</span>
<span class="c1"># import matplotlib.pyplot as plt</span>
<span class="c1"># plt.imshow(caDistances, cmap=&quot;hot&quot;, interpolation=&quot;nearest&quot;)</span>
<span class="c1"># plt.show()</span>

<span class="c1"># build structure from distance plot:</span>
<span class="c1">## create the all-atom distance plot</span>
<span class="n">distances</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">distance_plot</span><span class="p">()</span>
<span class="c1">## get the sign of the dihedral angles</span>
<span class="n">chirality</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">dihedral_signs</span><span class="p">()</span>
<span class="c1">## get new, empty data structure : copy data structure from myChain</span>
<span class="n">myChain2</span> <span class="o">=</span> <span class="n">IC_duplicate</span><span class="p">(</span><span class="n">myChain</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;A&quot;</span><span class="p">]</span>
<span class="n">cic2</span> <span class="o">=</span> <span class="n">myChain2</span><span class="o">.</span><span class="n">internal_coord</span>
<span class="c1">## clear the new atomArray and di/hedra value arrays, just for proof</span>
<span class="n">cic2</span><span class="o">.</span><span class="n">atomArray</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">cic2</span><span class="o">.</span><span class="n">AAsiz</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="n">cic2</span><span class="o">.</span><span class="n">dihedraAngle</span><span class="p">[:]</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="n">cic2</span><span class="o">.</span><span class="n">hedraAngle</span><span class="p">[:]</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="n">cic2</span><span class="o">.</span><span class="n">hedraL12</span><span class="p">[:]</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="n">cic2</span><span class="o">.</span><span class="n">hedraL23</span><span class="p">[:]</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1">## copy just the first N-Ca-C coords so structures will superimpose:</span>
<span class="n">cic2</span><span class="o">.</span><span class="n">copy_initNCaCs</span><span class="p">(</span><span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="p">)</span>
<span class="c1">## copy distances to chain arrays:</span>
<span class="n">cic2</span><span class="o">.</span><span class="n">distplot_to_dh_arrays</span><span class="p">(</span><span class="n">distances</span><span class="p">,</span> <span class="n">chirality</span><span class="p">)</span>
<span class="c1">## compute angles and dihedral angles from distances:</span>
<span class="n">cic2</span><span class="o">.</span><span class="n">distance_to_internal_coordinates</span><span class="p">()</span>
<span class="c1">## generate XYZ coordinates from internal coordinates:</span>
<span class="n">myChain2</span><span class="o">.</span><span class="n">internal_to_atom_coordinates</span><span class="p">()</span>
<span class="c1">## confirm result atomArray matches original structure:</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">cic2</span><span class="o">.</span><span class="n">atomArray</span><span class="p">,</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">atomArray</span><span class="p">)</span>

<span class="c1"># superimpose all phe-phe pairs - quick hack just to demonstrate concept</span>
<span class="c1"># for analyzing pairwise residue interactions.  Generates PDB ATOM records</span>
<span class="c1"># placing each PHE at origin and showing all other PHEs in environment</span>
<span class="c1">## shorthand for key variables:</span>
<span class="n">cic</span> <span class="o">=</span> <span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span>
<span class="n">resNameNdx</span> <span class="o">=</span> <span class="n">AtomKey</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">resname</span>
<span class="n">aaNdx</span> <span class="o">=</span> <span class="n">cic</span><span class="o">.</span><span class="n">atomArrayIndex</span>
<span class="c1">## select just PHE atoms:</span>
<span class="n">pheAtomSelect</span> <span class="o">=</span> <span class="p">[</span><span class="n">aaNdx</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">aaNdx</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">akl</span><span class="p">[</span><span class="n">resNameNdx</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;F&quot;</span><span class="p">]</span>
<span class="n">aaF</span> <span class="o">=</span> <span class="n">cic</span><span class="o">.</span><span class="n">atomArray</span><span class="p">[</span> <span class="n">pheAtomSelect</span> <span class="p">]</span>  <span class="c1"># numpy fancy indexing makes COPY not view</span>

<span class="k">for</span> <span class="n">ric</span> <span class="ow">in</span> <span class="n">cic</span><span class="o">.</span><span class="n">ordered_aa_ic_list</span><span class="p">:</span>  <span class="c1"># internal_coords version of get_residues()</span>
    <span class="k">if</span> <span class="n">ric</span><span class="o">.</span><span class="n">rbase</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;F&quot;</span><span class="p">:</span>  <span class="c1"># if PHE, get transform matrices for chi1 dihedral</span>
        <span class="n">chi1</span> <span class="o">=</span> <span class="n">ric</span><span class="o">.</span><span class="n">pick_angle</span><span class="p">(</span><span class="s2">&quot;N:CA:CB:CG&quot;</span><span class="p">)</span>  <span class="c1"># chi1 space has C-alpha at origin</span>
        <span class="n">cst</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">chi1</span><span class="o">.</span><span class="n">cst</span><span class="p">)</span>  <span class="c1"># transform TO chi1 space</span>
        <span class="c1"># rcst = np.transpose(chi1.rcst)  # transform FROM chi1 space</span>
        <span class="n">cic</span><span class="o">.</span><span class="n">atomArray</span><span class="p">[</span><span class="n">pheAtomSelect</span><span class="p">]</span> <span class="o">=</span> <span class="n">aaF</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">cst</span><span class="p">)</span>  <span class="c1"># transform just the PHEs</span>
        <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">myChain</span><span class="o">.</span><span class="n">get_residues</span><span class="p">():</span>  <span class="c1"># print PHEs in new coordinate space</span>
            <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">resname</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PHE&quot;</span><span class="p">]:</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">internal_coord</span><span class="o">.</span><span class="n">pdb_residue_string</span><span class="p">())</span>
        <span class="n">cic</span><span class="o">.</span><span class="n">atomArray</span><span class="p">[</span><span class="n">pheAtomSelect</span><span class="p">]</span> <span class="o">=</span> <span class="n">aaF</span>  <span class="c1"># restore coordinate space from copy</span>

<span class="c1"># write OpenSCAD program of spheres and cylinders to 3d print myChain backbone</span>
<span class="c1">## set atom load filter to accept backbone only:</span>
<span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_atoms</span> <span class="o">=</span> <span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_backbone</span>
<span class="c1">## delete existing data to force re-read of all atoms:</span>
<span class="n">myChain</span><span class="o">.</span><span class="n">internal_coord</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">write_SCAD</span><span class="p">(</span><span class="n">myChain</span><span class="p">,</span> <span class="s2">&quot;myChain.scad&quot;</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mf">10.0</span><span class="p">)</span>
</pre></div>
</div>
<p>See the <cite>‘’Internal coordinates module’’</cite> section of the <cite>Biopython Tutorial
and Cookbook</cite> for further discussion.</p>
<p><strong>Terms and key data structures:</strong>
Internal coordinates are defined on sequences of atoms which span
residues or follow accepted nomenclature along sidechains.  To manage these
sequences and support Biopython’s disorder mechanisms, <a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomKey</span></code></a>
specifiers are implemented to capture residue, atom and variant identification
in a single object.  A <a class="reference internal" href="#Bio.PDB.internal_coords.Hedron" title="Bio.PDB.internal_coords.Hedron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Hedron</span></code></a> object is specified as three sequential
AtomKeys, comprising two bond lengths and the bond angle between them.  A
<a class="reference internal" href="#Bio.PDB.internal_coords.Dihedron" title="Bio.PDB.internal_coords.Dihedron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dihedron</span></code></a> consists of four sequential AtomKeys, linking two Hedra with
a dihedral angle between them.</p>
<p><strong>Algorithmic overview:</strong>
The Internal Coordinates module combines a specification of connected atoms as
hedra and dihedra in the <a class="reference internal" href="Bio.PDB.ic_data.html#module-Bio.PDB.ic_data" title="Bio.PDB.ic_data"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ic_data</span></code></a> file with routines here to transform
XYZ coordinates of these atom sets between a local coordinate system and the
world coordinates supplied in e.g. a PDB or mmCif data file.  The local
coordinate system places the center atom of a hedron at the origin (0,0,0), one
leg on the +Z axis, and the other leg on the XZ plane (see <a class="reference internal" href="#Bio.PDB.internal_coords.Hedron" title="Bio.PDB.internal_coords.Hedron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Hedron</span></code></a>).
Measurement and creation or manipulation of hedra and dihedra in the local
coordinate space is straightforward, and the calculated transformation matrices
enable assembling these subunits into a protein chain starting from supplied
(PDB) coordinates for the initial N-Ca-C atoms.</p>
<p>Psi and Phi angles are defined on atoms from adjacent residues in a protein
chain, see e.g. <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pick_angle" title="Bio.PDB.internal_coords.IC_Residue.pick_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pick_angle()</span></code></a> and <a class="reference internal" href="Bio.PDB.ic_data.html#module-Bio.PDB.ic_data" title="Bio.PDB.ic_data"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ic_data</span></code></a> for the relevant
mapping between residues and backbone dihedral angles.</p>
<p>Transforms to and from the dihedron local coordinate space described above are
accessible via <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.dCoordSpace"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Chain.dCoordSpace</span></code></a> and <a class="reference internal" href="#Bio.PDB.internal_coords.Dihedron" title="Bio.PDB.internal_coords.Dihedron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dihedron</span></code></a> attributes
.cst and .rcst, and may be applied in the alignment and comparison of residues
and their environments with code along the lines of:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chi1</span> <span class="o">=</span> <span class="n">ric0</span><span class="o">.</span><span class="n">pick_angle</span><span class="p">(</span><span class="s2">&quot;chi1&quot;</span><span class="p">)</span> <span class="c1"># chi1 space defined with CA at origin</span>
<span class="n">cst</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">chi1</span><span class="o">.</span><span class="n">cst</span><span class="p">)</span> <span class="c1"># transform TO chi1 local space</span>
<span class="n">newAtomCoords</span> <span class="o">=</span> <span class="n">oldAtomCoords</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">cst</span><span class="p">)</span>
</pre></div>
</div>
<p>The core algorithms were developed independently during 1993-4 for
<a class="reference external" href="https://www.google.com/books/edition/Advances_in_Molecular_Bioinformatics/VmFSNNm7k6cC?gbpv=1">‘’Development and Application of a Three-dimensional Description of Amino Acid
Environments in Protein,’’ Miller, Douthart, and Dunker, Advances in Molecular
Bioinformatics, IOS Press, 1994, ISBN 90 5199 172 x, pp. 9-30.</a></p>
<p>A Protein Internal Coordinate (.pic) file format is defined to capture
sufficient detail to reproduce a PDB file from chain starting coordinates
(first residue N, Ca, C XYZ coordinates) and remaining internal coordinates.
These files are used internally to verify that a given structure can be
regenerated from its internal coordinates.  See <a class="reference internal" href="Bio.PDB.PICIO.html#module-Bio.PDB.PICIO" title="Bio.PDB.PICIO"><code class="xref py py-mod docutils literal notranslate"><span class="pre">PICIO</span></code></a> for reading and
writing .pic files and <a class="reference internal" href="Bio.PDB.ic_rebuild.html#Bio.PDB.ic_rebuild.structure_rebuild_test" title="Bio.PDB.ic_rebuild.structure_rebuild_test"><code class="xref py py-func docutils literal notranslate"><span class="pre">structure_rebuild_test()</span></code></a> to determine if a
specific PDB or mmCif datafile has sufficient information to interconvert
between cartesian and internal coordinates.</p>
<p>Internal coordinates may also be exported as <a class="reference external" href="https://www.openscad.org">OpenSCAD</a>
data arrays for generating 3D printed protein models.  OpenSCAD software is
provided as a starting point and proof-of-concept for generating such models.
See <a class="reference internal" href="Bio.PDB.SCADIO.html#module-Bio.PDB.SCADIO" title="Bio.PDB.SCADIO"><code class="xref py py-mod docutils literal notranslate"><span class="pre">SCADIO</span></code></a> and this <a class="reference external" href="https://www.thingiverse.com/thing:3957471">Thingiverse project</a>
for a more advanced example.</p>
<p>Refer to <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.distance_plot" title="Bio.PDB.internal_coords.IC_Chain.distance_plot"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distance_plot()</span></code></a> and <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distance_to_internal_coordinates()</span></code></a>
for converting structure data to/from 2D distance plots.</p>
<p>The following classes comprise the core functionality for processing internal
coordinates and are sufficiently related and coupled to place them together in
this module:</p>
<dl class="simple">
<dt><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Chain</span></code></a>: Extends Biopython Chain on .internal_coord attribute.</dt><dd><p>Manages connected sequence of residues and chain breaks; holds numpy arrays
for all atom coordinates and bond geometries. For ‘parallel’ processing
IC_Chain methods operate on these arrays with single numpy commands.</p>
</dd>
<dt><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue" title="Bio.PDB.internal_coords.IC_Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Residue</span></code></a>: Extends Biopython Residue on .internal_coord attribute.</dt><dd><p>Access for per residue views on internal coordinates and methods for serial
(residue by residue) assembly.</p>
</dd>
<dt><a class="reference internal" href="#Bio.PDB.internal_coords.Dihedron" title="Bio.PDB.internal_coords.Dihedron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dihedron</span></code></a>: four joined atoms forming a dihedral angle.</dt><dd><p>Dihedral angle, homogeneous atom coordinates in local coordinate space,
references to relevant Hedra and IC_Residue.  Getter methods for
residue dihedral angles, bond angles and bond lengths.</p>
</dd>
<dt><a class="reference internal" href="#Bio.PDB.internal_coords.Hedron" title="Bio.PDB.internal_coords.Hedron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Hedron</span></code></a>: three joined atoms forming a plane.</dt><dd><p>Contains homogeneous atom coordinates in local coordinate space as well as
bond lengths and angle between them.</p>
</dd>
<dt><a class="reference internal" href="#Bio.PDB.internal_coords.Edron" title="Bio.PDB.internal_coords.Edron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edron</span></code></a>: base class for Hedron and Dihedron classes.</dt><dd><p>Tuple of AtomKeys comprising child, string ID, mainchain membership boolean
and other routines common for both Hedra and Dihedra.  Implements rich
comparison.</p>
</dd>
<dt><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomKey</span></code></a>: keys (dictionary and string) for referencing atom sequences.</dt><dd><p>Capture residue and disorder/occupancy information, provides a
no-whitespace key for .pic files, and implements rich comparison.</p>
</dd>
</dl>
<p>Custom exception classes: <a class="reference internal" href="#Bio.PDB.internal_coords.HedronMatchError" title="Bio.PDB.internal_coords.HedronMatchError"><code class="xref py py-class docutils literal notranslate"><span class="pre">HedronMatchError</span></code></a> and
<a class="reference internal" href="#Bio.PDB.internal_coords.MissingAtomError" title="Bio.PDB.internal_coords.MissingAtomError"><code class="xref py py-class docutils literal notranslate"><span class="pre">MissingAtomError</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">IC_Chain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Chain</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to extend Biopython Chain with internal coordinate data.</p>
<dl class="field-list">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl>
<dt><strong>chain: object reference</strong></dt><dd><p>The Biopython <a class="reference internal" href="Bio.PDB.Chain.html#Bio.PDB.Chain.Chain" title="Bio.PDB.Chain.Chain"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.PDB.Chain.Chain</span></code></a> object this extends</p>
</dd>
<dt><strong>MaxPeptideBond: float</strong></dt><dd><p><strong>Class</strong> attribute to detect chain breaks.
Override for fully contiguous chains with some very long bonds - e.g.
for 3D printing (OpenSCAD output) a structure with missing residues.
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.MaxPeptideBond" title="Bio.PDB.internal_coords.IC_Chain.MaxPeptideBond"><code class="xref py py-data docutils literal notranslate"><span class="pre">MaxPeptideBond</span></code></a></p>
</dd>
<dt><strong>ParallelAssembleResidues: bool</strong></dt><dd><p><strong>Class</strong> attribute affecting internal_to_atom_coords.
Short (50 residue and less) chains are faster to assemble without the
overhead of creating numpy arrays, and the algorithm is easier to
understand and trace processing a single residue at a time.  Clearing
(set to False) this flag will switch to the serial algorithm</p>
</dd>
<dt><strong>ordered_aa_ic_list: list</strong></dt><dd><p>IC_Residue objects internal_coords algorithms can process (e.g. no
waters)</p>
</dd>
<dt><strong>initNCaC: List of N, Ca, C AtomKey tuples (NCaCKeys).</strong></dt><dd><p>NCaCKeys start chain segments (first residue or after chain break).
These 3 atoms define the coordinate space for a contiguous chain
segment, as initially specified by PDB or mmCIF file.</p>
</dd>
<dt><strong>AAsiz = int</strong></dt><dd><p>AtomArray size, number of atoms in this chain</p>
</dd>
<dt><strong>atomArray: numpy array</strong></dt><dd><p>homogeneous atom coords ([x,, y, z, 1.0]) for every atom in chain</p>
</dd>
<dt><strong>atomArrayIndex: dict</strong></dt><dd><p>maps AtomKeys to atomArray indexes</p>
</dd>
<dt><strong>hedra: dict</strong></dt><dd><p>Hedra forming residues in this chain; indexed by 3-tuples of AtomKeys.</p>
</dd>
<dt><strong>hedraLen: int</strong></dt><dd><p>length of hedra dict</p>
</dd>
<dt><strong>hedraNdx: dict</strong></dt><dd><p>maps hedra AtomKeys to numeric index into hedra data arrays e.g.
hedraL12 below</p>
</dd>
<dt><strong>a2ha_map: [hedraLen x 3]</strong></dt><dd><p>atom indexes in hedraNdx order</p>
</dd>
<dt><strong>dihedra: dict</strong></dt><dd><p>Dihedra forming residues in this chain; indexed by 4-tuples of AtomKeys.</p>
</dd>
<dt><strong>dihedraLen: int</strong></dt><dd><p>length of dihedra dict</p>
</dd>
<dt><strong>dihedraNdx: dict</strong></dt><dd><p>maps dihedra AtomKeys to dihedra data arrays e.g. dihedraAngle</p>
</dd>
<dt><strong>a2da_map</strong><span class="classifier">[dihedraLen x 4]</span></dt><dd><p>AtomNdx’s in dihedraNdx order</p>
</dd>
<dt><strong>d2a_map</strong><span class="classifier">[dihedraLen x [4]]</span></dt><dd><p>AtomNdx’s for each dihedron (reshaped a2da_map)</p>
</dd>
<dt><strong>Numpy arrays for vector processing of chain di/hedra:</strong></dt><dd></dd>
<dt><strong>hedraL12: numpy array</strong></dt><dd><p>bond length between hedron 1st and 2nd atom</p>
</dd>
<dt><strong>hedraAngle: numpy array</strong></dt><dd><p>bond angle for each hedron, in degrees</p>
</dd>
<dt><strong>hedraL23: numpy array</strong></dt><dd><p>bond length between hedron 2nd and 3rd atom</p>
</dd>
<dt><strong>id3_dh_index: dict</strong></dt><dd><p>maps hedron key to list of dihedra starting with hedron, used by
assemble and bond_rotate to find dihedra with h1 key</p>
</dd>
<dt><strong>id32_dh_index: dict</strong></dt><dd><p>like id3_dh_index, find dihedra from h2 key</p>
</dd>
<dt><strong>hAtoms: numpy array</strong></dt><dd><p>homogeneous atom coordinates (3x4) of hedra, central atom at origin</p>
</dd>
<dt><strong>hAtomsR: numpy array</strong></dt><dd><p>hAtoms in reverse orientation</p>
</dd>
<dt><strong>hAtoms_needs_update: numpy array of bool</strong></dt><dd><p>indicates whether hAtoms represent hedraL12/A/L23</p>
</dd>
<dt><strong>dihedraAngle: numpy array</strong></dt><dd><p>dihedral angles (degrees) for each dihedron</p>
</dd>
<dt><strong>dAtoms: numpy array</strong></dt><dd><p>homogeneous atom coordinates (4x4) of dihedra, second atom at origin</p>
</dd>
<dt><strong>dAtoms_needs_update: numpy array of bool</strong></dt><dd><p>indicates whether dAtoms represent dihedraAngle</p>
</dd>
<dt><strong>dCoordSpace: numpy array</strong></dt><dd><p>forward and reverse transform matrices standardising positions of first
hedron.  See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.dCoordSpace"><code class="xref py py-data docutils literal notranslate"><span class="pre">dCoordSpace</span></code></a>.</p>
</dd>
<dt><strong>dcsValid: bool</strong></dt><dd><p>indicates dCoordSpace up to date</p>
</dd>
<dt><strong>See also attributes generated by</strong><span class="classifier">meth:<cite>build_edraArrays</cite> for indexing</span></dt><dd></dd>
<dt><strong>di/hedra data elements.</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 17%" />
<col style="width: 83%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>internal_to_atom_coordinates:</strong></p></td>
<td><p>Process ic data to Residue/Atom coordinates; calls assemble_residues()</p></td>
</tr>
<tr class="row-even"><td><p><strong>assemble_residues:</strong></p></td>
<td><p>Generate IC_Chain atom coords from internal coordinates (parallel)</p></td>
</tr>
<tr class="row-odd"><td><p><strong>assemble_residues_ser:</strong></p></td>
<td><p>Generate IC_Residue atom coords from internal coordinates (serial)</p></td>
</tr>
<tr class="row-even"><td><p><strong>atom_to_internal_coordinates:</strong></p></td>
<td><p>Calculate dihedrals, angles, bond lengths (internal coordinates) for Atom data</p></td>
</tr>
<tr class="row-odd"><td><p><strong>write_SCAD:</strong></p></td>
<td><p>Write OpenSCAD matrices for internal coordinate data comprising chain; this is a support routine, see <a class="reference internal" href="Bio.PDB.SCADIO.html#Bio.PDB.SCADIO.write_SCAD" title="Bio.PDB.SCADIO.write_SCAD"><code class="xref py py-func docutils literal notranslate"><span class="pre">SCADIO.write_SCAD()</span></code></a> to generate OpenSCAD description of a protein chain.</p></td>
</tr>
<tr class="row-even"><td><p><strong>distance_plot:</strong></p></td>
<td><p>Generate 2D plot of interatomic distances with optional filter</p></td>
</tr>
<tr class="row-odd"><td><p><strong>distance_to_internal_coordinates:</strong></p></td>
<td><p>Compute internal coordinates from distance plot and array of dihedral angle signs.</p></td>
</tr>
<tr class="row-even"><td><p><strong>make_extended:</strong></p></td>
<td><p>Arbitrarily sets all psi and phi backbone angles to 123 and -104 degrees.</p></td>
</tr>
</tbody>
</table>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.MaxPeptideBond">
<span class="sig-name descname"><span class="pre">MaxPeptideBond</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1.4</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.MaxPeptideBond" title="Permalink to this definition"></a></dt>
<dd><p>Larger C-N distance than this will be chain break</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.ParallelAssembleResidues">
<span class="sig-name descname"><span class="pre">ParallelAssembleResidues</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">True</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.ParallelAssembleResidues" title="Permalink to this definition"></a></dt>
<dd><p>Enable parallel internal_to_atom algorithm, is slower for short chains</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.AAsiz">
<span class="sig-name descname"><span class="pre">AAsiz</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.AAsiz" title="Permalink to this definition"></a></dt>
<dd><p>Number of atoms in this chain (size of atomArray)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.atomArray">
<span class="sig-name descname"><span class="pre">atomArray</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">numpy.array</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.atomArray" title="Permalink to this definition"></a></dt>
<dd><p>AAsiz x [4] of float np.float64 homogeneous atom coordinates, all atoms
in chain.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.dCoordSpace">
<span class="sig-name descname"><span class="pre">dCoordSpace</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.dCoordSpace" title="Permalink to this definition"></a></dt>
<dd><p>[2][dihedraLen][4][4] : 2 arrays of 4x4 coordinate space transforms for
each dihedron.  The first [0] converts TO standard space with first atom on
the XZ plane, the second atom at the origin, the third on the +Z axis, and
the fourth placed according to the dihedral angle.  The second [1] transform
returns FROM the standard space to world coordinates (PDB file input or
whatever is current).  Also accessible as .cst (forward
transform) and .rcst (reverse transform) in <a class="reference internal" href="#Bio.PDB.internal_coords.Dihedron" title="Bio.PDB.internal_coords.Dihedron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dihedron</span></code></a>.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.dcsValid">
<span class="sig-name descname"><span class="pre">dcsValid</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.dcsValid" title="Permalink to this definition"></a></dt>
<dd><p>True if dCoordSpace is up to date.  Use <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.update_dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.update_dCoordSpace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">update_dCoordSpace()</span></code></a>
if needed.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Chain</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize IC_Chain object, with or without residue/Atom data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>parent</strong> (<em>Bio.PDB.Chain</em>) – Biopython Chain object
Chain object this extends</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.__deepcopy__">
<span class="sig-name descname"><span class="pre">__deepcopy__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><span class="pre">Bio.PDB.internal_coords.IC_Chain</span></a></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.__deepcopy__" title="Permalink to this definition"></a></dt>
<dd><p>Implement deepcopy for IC_Chain.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.clear_ic">
<span class="sig-name descname"><span class="pre">clear_ic</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.clear_ic" title="Permalink to this definition"></a></dt>
<dd><p>Clear residue internal_coord settings for this chain.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.build_atomArray">
<span class="sig-name descname"><span class="pre">build_atomArray</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.build_atomArray" title="Permalink to this definition"></a></dt>
<dd><p>Build <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Chain</span></code></a> numpy coordinate array from biopython atoms.</p>
<p>See also <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.init_edra" title="Bio.PDB.internal_coords.IC_Chain.init_edra"><code class="xref py py-meth docutils literal notranslate"><span class="pre">init_edra()</span></code></a> for more complete initialization of IC_Chain.</p>
<dl class="simple">
<dt>Inputs:</dt><dd><dl class="simple">
<dt>self.akset<span class="classifier">set</span></dt><dd><p><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomKey</span></code></a> s in this chain</p>
</dd>
</dl>
</dd>
<dt>Generates:</dt><dd><dl class="simple">
<dt>self.AAsiz<span class="classifier">int</span></dt><dd><p>number of atoms in chain (len(akset))</p>
</dd>
<dt>self.aktuple<span class="classifier">AAsiz x AtomKeys</span></dt><dd><p>sorted akset AtomKeys</p>
</dd>
<dt>self.atomArrayIndex<span class="classifier">[AAsiz] of int</span></dt><dd><p>numerical index for each AtomKey in aktuple</p>
</dd>
<dt>self.atomArrayValid<span class="classifier">AAsiz x bool</span></dt><dd><p>atomArray coordinates current with internal coordinates if True</p>
</dd>
<dt>self.atomArray<span class="classifier">AAsiz x np.float64[4]</span></dt><dd><p>homogeneous atom coordinates; Biopython <a class="reference internal" href="Bio.PDB.Atom.html#Bio.PDB.Atom.Atom" title="Bio.PDB.Atom.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Atom</span></code></a>
coordinates are view into this array after execution</p>
</dd>
<dt>rak_cache<span class="classifier">dict</span></dt><dd><p>lookup cache for AtomKeys for each residue</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.build_edraArrays">
<span class="sig-name descname"><span class="pre">build_edraArrays</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.build_edraArrays" title="Permalink to this definition"></a></dt>
<dd><p>Build chain level hedra and dihedra arrays.</p>
<p>Used by <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.init_edra" title="Bio.PDB.internal_coords.IC_Chain.init_edra"><code class="xref py py-meth docutils literal notranslate"><span class="pre">init_edra()</span></code></a> and <code class="xref py py-meth docutils literal notranslate"><span class="pre">_hedraDict2chain()</span></code>.  Should be
private method but exposed for documentation.</p>
<dl class="simple">
<dt>Inputs:</dt><dd><dl class="simple">
<dt>self.dihedraLen<span class="classifier">int</span></dt><dd><p>number of dihedra needed</p>
</dd>
<dt>self.hedraLen<span class="classifier">int</span></dt><dd><p>number of hedra needed</p>
</dd>
<dt>self.AAsiz<span class="classifier">int</span></dt><dd><p>length of atomArray</p>
</dd>
<dt>self.hedraNdx<span class="classifier">dict</span></dt><dd><p>maps hedron keys to range(hedraLen)</p>
</dd>
<dt>self.dihedraNdx<span class="classifier">dict</span></dt><dd><p>maps dihedron keys to range(dihedraLen)</p>
</dd>
<dt>self.hedra<span class="classifier">dict</span></dt><dd><p>maps Hedra keys to Hedra for chain</p>
</dd>
<dt>self.atomArray<span class="classifier">AAsiz x np.float64[4]</span></dt><dd><p>homogeneous atom coordinates for chain</p>
</dd>
<dt>self.atomArrayIndex<span class="classifier">dict</span></dt><dd><p>maps AtomKeys to atomArray</p>
</dd>
<dt>self.atomArrayValid<span class="classifier">AAsiz x bool</span></dt><dd><p>indicates coord is up-to-date</p>
</dd>
</dl>
</dd>
<dt>Generates:</dt><dd><dl class="simple">
<dt>self.dCoordSpace<span class="classifier">[2][dihedraLen][4][4]</span></dt><dd><p>transforms to/from dihedron coordinate space</p>
</dd>
<dt>self.dcsValid<span class="classifier">dihedraLen x bool</span></dt><dd><p>indicates dCoordSpace is current</p>
</dd>
<dt>self.hAtoms<span class="classifier">hedraLen x 3 x np.float64[4]</span></dt><dd><p>atom coordinates in hCoordSpace</p>
</dd>
<dt>self.hAtomsR<span class="classifier">hedraLen x 3 x np.float64[4]</span></dt><dd><p>hAtoms in reverse order (trading space for time)</p>
</dd>
<dt>self.hAtoms_needs_update<span class="classifier">hedraLen x bool</span></dt><dd><p>indicates hAtoms, hAtoms current</p>
</dd>
<dt>self.a2h_map<span class="classifier">AAsiz x [int …]</span></dt><dd><p>maps atomArrayIndex to hedraNdx’s with that atom</p>
</dd>
<dt>self.a2ha_map<span class="classifier">[hedraLen x 3]</span></dt><dd><p>AtomNdx’s in hedraNdx order</p>
</dd>
<dt>self.h2aa<span class="classifier">hedraLen x [int …]</span></dt><dd><p>maps hedraNdx to atomNdx’s in hedron (reshaped later)</p>
</dd>
<dt>Hedron.ndx<span class="classifier">int</span></dt><dd><p>self.hedraNdx value stored inside Hedron object</p>
</dd>
<dt>self.dRev<span class="classifier">dihedraLen x bool</span></dt><dd><p>dihedron reversed if true</p>
</dd>
<dt>self.dH1ndx, dH2ndx<span class="classifier">[dihedraLen]</span></dt><dd><p>hedraNdx’s for 1st and 2nd hedra</p>
</dd>
<dt>self.h1d_map<span class="classifier">hedraLen x []</span></dt><dd><p>hedraNdx -&gt; [dihedra using hedron]</p>
</dd>
<dt>Dihedron.h1key, h2key<span class="classifier">[AtomKey …]</span></dt><dd><p>hedron keys for dihedron, reversed as needed</p>
</dd>
<dt>Dihedron.hedron1, hedron2<span class="classifier">Hedron</span></dt><dd><p>references inside dihedron to hedra</p>
</dd>
<dt>Dihedron.ndx<span class="classifier">int</span></dt><dd><p>self.dihedraNdx info inside Dihedron object</p>
</dd>
<dt>Dihedron.cst, rcst<span class="classifier">np.float64p4][4]</span></dt><dd><p>dCoordSpace references inside Dihedron</p>
</dd>
<dt>self.a2da_map<span class="classifier">[dihedraLen x 4]</span></dt><dd><p>AtomNdx’s in dihedraNdx order</p>
</dd>
<dt>self.d2a_map<span class="classifier">[dihedraLen x [4]]</span></dt><dd><p>AtomNdx’s for each dihedron (reshaped a2da_map)</p>
</dd>
<dt>self.dFwd<span class="classifier">bool</span></dt><dd><p>dihedron is not Reversed if True</p>
</dd>
<dt>self.a2d_map<span class="classifier">AAsiz x [[dihedraNdx]</span></dt><dd><p>[atom ndx 0-3 of atom in dihedron]], maps atom indexes to
dihedra and atoms in them</p>
</dd>
<dt>self.dAtoms_needs_update<span class="classifier">dihedraLen x bool</span></dt><dd><p>atoms in h1, h2 are current if False</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.assemble_residues">
<span class="sig-name descname"><span class="pre">assemble_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues" title="Permalink to this definition"></a></dt>
<dd><p>Generate atom coords from internal coords (vectorised).</p>
<p>This is the ‘Numpy parallel’ version of <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues_ser()</span></code></a>.</p>
<p>Starting with dihedra already formed by <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.init_atom_coords" title="Bio.PDB.internal_coords.IC_Chain.init_atom_coords"><code class="xref py py-meth docutils literal notranslate"><span class="pre">init_atom_coords()</span></code></a>, transform
each from dihedron local coordinate space into protein chain coordinate
space.  Iterate until all dependencies satisfied.</p>
<p>Does not update <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.dCoordSpace"><code class="xref py py-data docutils literal notranslate"><span class="pre">dCoordSpace</span></code></a> as <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues_ser()</span></code></a>
does.  Call <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.update_dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.update_dCoordSpace"><code class="xref py py-meth docutils literal notranslate"><span class="pre">update_dCoordSpace()</span></code></a> if needed.  Faster to do in
single operation once all atom coordinates finished.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – default False.
Report number of iterations to compute changed dihedra</p>
</dd>
</dl>
<dl class="simple">
<dt>generates:</dt><dd><dl class="simple">
<dt>self.dSet: AAsiz x dihedraLen x 4</dt><dd><p>maps atoms in dihedra to atomArray</p>
</dd>
<dt>self.dSetValid<span class="classifier">[dihedraLen][4] of bool</span></dt><dd><p>map of valid atoms into dihedra to detect 3 or 4 atoms valid</p>
</dd>
</dl>
</dd>
</dl>
<p>Output coordinates written to <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.atomArray" title="Bio.PDB.internal_coords.IC_Chain.atomArray"><code class="xref py py-data docutils literal notranslate"><span class="pre">atomArray</span></code></a>.  Biopython
<a class="reference internal" href="Bio.PDB.Atom.html#module-Bio.PDB.Atom" title="Bio.PDB.Atom"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.PDB.Atom</span></code></a> coordinates are a view on this data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser">
<span class="sig-name descname"><span class="pre">assemble_residues_ser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser" title="Permalink to this definition"></a></dt>
<dd><p>Generate IC_Residue atom coords from internal coordinates (serial).</p>
<p>See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues()</span></code></a> for ‘numpy parallel’ version.</p>
<p>Filter positions between start and fin if set, find appropriate start
coordinates for each residue and pass to <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.assemble" title="Bio.PDB.internal_coords.IC_Residue.assemble"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>verbose</strong> (<em>bool</em>) – default False.
Describe runtime problems</p></li>
<li><p><strong>start</strong><strong>,</strong><strong>fin</strong> (<em>int</em>) – default None.
Sequence position for begin, end of subregion to generate coords
for.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.init_edra">
<span class="sig-name descname"><span class="pre">init_edra</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.init_edra" title="Permalink to this definition"></a></dt>
<dd><p>Create chain and residue di/hedra structures, arrays, atomArray.</p>
<dl class="simple">
<dt>Inputs:</dt><dd><p>self.ordered_aa_ic_list : list of IC_Residue</p>
</dd>
<dt>Generates:</dt><dd><ul class="simple">
<li><p>edra objects, self.di/hedra (executes <code class="xref py py-meth docutils literal notranslate"><span class="pre">_create_edra()</span></code>)</p></li>
<li><p>atomArray and support (executes <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.build_atomArray" title="Bio.PDB.internal_coords.IC_Chain.build_atomArray"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_atomArray()</span></code></a>)</p></li>
<li><p>self.hedraLen : number of hedra in structure</p></li>
<li><p>hedraL12 : numpy arrays for lengths, angles (empty)</p></li>
<li><p>hedraAngle ..</p></li>
<li><p>hedraL23 ..</p></li>
<li><p>self.hedraNdx : dict mapping hedrakeys to hedraL12 etc</p></li>
<li><p>self.dihedraLen : number of dihedra in structure</p></li>
<li><p>dihedraAngle ..</p></li>
<li><p>dihedraAngleRads : np arrays for angles (empty)</p></li>
<li><p>self.dihedraNdx : dict mapping dihedrakeys to dihedraAngle</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.init_atom_coords">
<span class="sig-name descname"><span class="pre">init_atom_coords</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.init_atom_coords" title="Permalink to this definition"></a></dt>
<dd><p>Set chain level di/hedra initial coords from angles and distances.</p>
<p>Initializes atom coordinates in local coordinate space for hedra and
dihedra, will be transformed appropriately later by <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.dCoordSpace"><code class="xref py py-data docutils literal notranslate"><span class="pre">dCoordSpace</span></code></a>
matrices for assembly.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.update_dCoordSpace">
<span class="sig-name descname"><span class="pre">update_dCoordSpace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">workSelector</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.update_dCoordSpace" title="Permalink to this definition"></a></dt>
<dd><p>Compute/update coordinate space transforms for chain dihedra.</p>
<p>Requires all atoms updated so calls <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues()</span></code></a>
(returns immediately if all atoms already assembled).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>workSelector</strong> (<em>[</em><em>bool</em><em>]</em>) – Optional mask to select dihedra for update</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.propagate_changes">
<span class="sig-name descname"><span class="pre">propagate_changes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.propagate_changes" title="Permalink to this definition"></a></dt>
<dd><p>Track through di/hedra to invalidate dependent atoms.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.internal_to_atom_coordinates">
<span class="sig-name descname"><span class="pre">internal_to_atom_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.internal_to_atom_coordinates" title="Permalink to this definition"></a></dt>
<dd><p>Process IC data to Residue/Atom coords.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>verbose</strong> (<em>bool</em>) – default False.
Describe runtime problems</p></li>
<li><p><strong>start</strong><strong>,</strong><strong>fin</strong> (<em>int</em>) – Optional sequence positions for begin, end of subregion
to process.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Setting start or fin activates serial <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues_ser()</span></code></a>
instead of (Numpy parallel) <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues()</span></code></a>.
Start C-alpha will be at origin.</p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.ParallelAssembleResidues" title="Bio.PDB.internal_coords.IC_Chain.ParallelAssembleResidues"><code class="xref py py-data docutils literal notranslate"><span class="pre">ParallelAssembleResidues</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates">
<span class="sig-name descname"><span class="pre">atom_to_internal_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates" title="Permalink to this definition"></a></dt>
<dd><p>Calculate dihedrals, angles, bond lengths for Atom data.</p>
<p>Generates atomArray (through init_edra), value arrays for hedra and
dihedra, and coordinate space transforms for dihedra.</p>
<p>Generates Gly C-beta if specified, see <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.gly_Cbeta" title="Bio.PDB.internal_coords.IC_Residue.gly_Cbeta"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Residue.gly_Cbeta</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – default False.
describe runtime problems</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.distance_plot">
<span class="sig-name descname"><span class="pre">distance_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">numpy.ndarray</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.distance_plot" title="Permalink to this definition"></a></dt>
<dd><p>Generate 2D distance plot from atomArray.</p>
<p>Default is to calculate distances for all atoms.  To generate the
classic C-alpha distance plot, pass a boolean mask array like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">atmNameNdx</span> <span class="o">=</span> <span class="n">internal_coords</span><span class="o">.</span><span class="n">AtomKey</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">atm</span>
<span class="n">CaSelect</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">akl</span><span class="p">[</span><span class="n">atmNameNdx</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span>
<span class="p">]</span>
<span class="n">plot</span> <span class="o">=</span> <span class="n">cic</span><span class="o">.</span><span class="n">distance_plot</span><span class="p">(</span><span class="n">CaSelect</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternatively, this will select all backbone atoms:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">backboneSelect</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">is_backbone</span><span class="p">()</span>
<span class="p">]</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filter</strong> (<em>[</em><em>bool</em><em>]</em>) – restrict atoms for calculation</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distance_to_internal_coordinates()</span></code></a>, which requires the
default all atom distance plot.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.dihedral_signs">
<span class="sig-name descname"><span class="pre">dihedral_signs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">numpy.ndarray</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.dihedral_signs" title="Permalink to this definition"></a></dt>
<dd><p>Get sign array (+1/-1) for each element of chain dihedraAngle array.</p>
<p>Required for <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distance_to_internal_coordinates()</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.distplot_to_dh_arrays">
<span class="sig-name descname"><span class="pre">distplot_to_dh_arrays</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distplot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">numpy.ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dihedra_signs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">numpy.ndarray</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.distplot_to_dh_arrays" title="Permalink to this definition"></a></dt>
<dd><p>Load di/hedra distance arays from distplot.</p>
<p>Fill <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Chain</span></code></a> arrays hedraL12, L23, L13 and dihedraL14
distance value arrays from input distplot, dihedra_signs array from
input dihedra_signs.  Distplot and di/hedra distance arrays must index
according to AtomKey mappings in <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Chain</span></code></a> .hedraNdx and .dihedraNdx
(created in <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.init_edra" title="Bio.PDB.internal_coords.IC_Chain.init_edra"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Chain.init_edra()</span></code></a>)</p>
<p>Call <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">atom_to_internal_coordinates()</span></code></a> (or at least <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.init_edra" title="Bio.PDB.internal_coords.IC_Chain.init_edra"><code class="xref py py-meth docutils literal notranslate"><span class="pre">init_edra()</span></code></a>)
to generate a2ha_map and d2a_map before running this.</p>
<p>Explcitly removed from <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distance_to_internal_coordinates()</span></code></a> so
user may populate these chain di/hedra arrays by other
methods.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates">
<span class="sig-name descname"><span class="pre">distance_to_internal_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resetAtoms</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates" title="Permalink to this definition"></a></dt>
<dd><p>Compute chain di/hedra from from distance and chirality data.</p>
<p>Distance properties on hedra L12, L23, L13 and dihedra L14 configured
by <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.distplot_to_dh_arrays" title="Bio.PDB.internal_coords.IC_Chain.distplot_to_dh_arrays"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distplot_to_dh_arrays()</span></code></a> or alternative loader.</p>
<p>dihedraAngles result is multiplied by dihedra_signs at final step
recover chirality information lost in distance plot (mirror image of
structure has same distances but opposite sign dihedral angles).</p>
<p>Note that chain breaks will cause errors in rebuilt structure, use
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.copy_initNCaCs" title="Bio.PDB.internal_coords.IC_Chain.copy_initNCaCs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">copy_initNCaCs()</span></code></a> to avoid this</p>
<p>Based on Blue, the Hedronometer’s answer to <a class="reference external" href="https://math.stackexchange.com/a/49340/972353">The dihedral angles of a tetrahedron
in terms of its edge lengths</a>
on <a class="reference external" href="https://math.stackexchange.com/">math.stackexchange.com</a>.  See also:
<a class="reference external" href="http://daylateanddollarshort.com/mathdocs/Heron-like-Results-for-Tetrahedral-Volume.pdf">“Heron-like Hedronometric Results for Tetrahedral Volume”</a>.</p>
<p>Other values from that analysis included here as comments for
completeness:</p>
<ul class="simple">
<li><p>oa = hedron1 L12 if reverse else hedron1 L23</p></li>
<li><p>ob = hedron1 L23 if reverse else hedron1 L12</p></li>
<li><p>ac = hedron2 L12 if reverse else hedron2 L23</p></li>
<li><p>ab = hedron1 L13 = law of cosines on OA, OB (hedron1 L12, L23)</p></li>
<li><p>oc = hedron2 L13 = law of cosines on OA, AC (hedron2 L12, L23)</p></li>
<li><p>bc = dihedron L14</p></li>
</ul>
<p>target is OA, the dihedral angle along edge oa.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>resetAtoms</strong> (<em>bool</em>) – default True.
Mark all atoms in di/hedra and atomArray for updating by
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.internal_to_atom_coordinates" title="Bio.PDB.internal_coords.IC_Chain.internal_to_atom_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">internal_to_atom_coordinates()</span></code></a>.  Alternatvely set this to
False and manipulate <cite>atomArrayValid</cite>, <cite>dAtoms_needs_update</cite> and
<cite>hAtoms_needs_update</cite> directly to reduce computation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.copy_initNCaCs">
<span class="sig-name descname"><span class="pre">copy_initNCaCs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><span class="pre">Bio.PDB.internal_coords.IC_Chain</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.copy_initNCaCs" title="Permalink to this definition"></a></dt>
<dd><p>Copy atom coordinates for initNCaC atoms from other IC_Chain.</p>
<p>Copies the coordinates and sets atomArrayValid flags True for initial
NCaC and after any chain breaks.</p>
<p>Needed for <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.distance_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">distance_to_internal_coordinates()</span></code></a> if target has
chain breaks (otherwise each fragment will start at origin).</p>
<p>Also useful if copying internal coordinates from another chain.</p>
<p>N.B. <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.set_angle" title="Bio.PDB.internal_coords.IC_Residue.set_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Residue.set_angle()</span></code></a> and <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.set_length" title="Bio.PDB.internal_coords.IC_Residue.set_length"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Residue.set_length()</span></code></a>
invalidate their relevant atoms, so apply them before calling this
function.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.make_extended">
<span class="sig-name descname"><span class="pre">make_extended</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.make_extended" title="Permalink to this definition"></a></dt>
<dd><p>Set all psi and phi angles to extended conformation (123, -104).</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Chain.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'atomArray':</span> <span class="pre">&lt;built-in</span> <span class="pre">function</span> <span class="pre">array&gt;,</span> <span class="pre">'atomArrayIndex':</span> <span class="pre">&quot;Dict['AtomKey',</span> <span class="pre">int]&quot;,</span> <span class="pre">'bpAtomArray':</span> <span class="pre">&quot;List['Atom']&quot;,</span> <span class="pre">'ordered_aa_ic_list':</span> <span class="pre">'List[IC_Residue]'}</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Chain.__annotations__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">IC_Residue</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Residue</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to extend Biopython Residue with internal coordinate data.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>parent: biopython Residue object this class extends</strong></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>no_altloc: bool default False</strong></dt><dd><p><strong>Class</strong> variable, disable processing of ALTLOC atoms if True, use
only selected atoms.</p>
</dd>
<dt><strong>accept_atoms: tuple</strong></dt><dd><p><strong>Class</strong> variable <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.accept_atoms" title="Bio.PDB.internal_coords.IC_Residue.accept_atoms"><code class="xref py py-data docutils literal notranslate"><span class="pre">accept_atoms</span></code></a>, list of PDB atom names to use
when generating internal coordinates.
Default is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">accept_atoms</span> <span class="o">=</span> <span class="n">accept_mainchain</span> <span class="o">+</span> <span class="n">accept_hydrogens</span>
</pre></div>
</div>
<p>to exclude hydrogens in internal coordinates and generated PDB files,
override as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_atoms</span> <span class="o">=</span> <span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_mainchain</span>
</pre></div>
</div>
<p>to get only mainchain atoms plus amide proton, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_atoms</span> <span class="o">=</span> <span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_mainchain</span> <span class="o">+</span> <span class="p">(</span><span class="s1">&#39;H&#39;</span><span class="p">,)</span>
</pre></div>
</div>
<p>to convert D atoms to H, set <a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey.d2h" title="Bio.PDB.internal_coords.AtomKey.d2h"><code class="xref py py-data docutils literal notranslate"><span class="pre">AtomKey.d2h</span></code></a> = True and use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_atoms</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">accept_mainchain</span> <span class="o">+</span> <span class="n">accept_hydrogens</span> <span class="o">+</span> <span class="n">accept_deuteriums</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Note that <cite>accept_mainchain = accept_backbone + accept_sidechain</cite>.
Thus to generate sequence-agnostic conformational data for e.g.
structure alignment in dihedral angle space, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_atoms</span> <span class="o">=</span> <span class="n">accept_backbone</span>
</pre></div>
</div>
<p>or set gly_Cbeta = True and use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">IC_Residue</span><span class="o">.</span><span class="n">accept_atoms</span> <span class="o">=</span> <span class="n">accept_backbone</span> <span class="o">+</span> <span class="p">(</span><span class="s1">&#39;CB&#39;</span><span class="p">,)</span>
</pre></div>
</div>
<p>Changing accept_atoms will cause the default <cite>structure_rebuild_test</cite> in
<a class="reference internal" href="Bio.PDB.ic_rebuild.html#module-Bio.PDB.ic_rebuild" title="Bio.PDB.ic_rebuild"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ic_rebuild</span></code></a> to fail if some atoms are filtered (obviously).  Use
the <cite>quick=True</cite> option to test only the coordinates of filtered atoms
to avoid this.</p>
<p>There is currently no option to output internal coordinates with D
instead of H.</p>
</dd>
<dt><strong>accept_resnames: tuple</strong></dt><dd><p><strong>Class</strong> variable <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.accept_resnames" title="Bio.PDB.internal_coords.IC_Residue.accept_resnames"><code class="xref py py-data docutils literal notranslate"><span class="pre">accept_resnames</span></code></a>, list of 3-letter residue
names for HETATMs to accept when generating internal coordinates from
atoms.  HETATM sidechain will be ignored, but normal backbone atoms (N,
CA, C, O, CB) will be included.  Currently only CYG, YCM and UNK;
override at your own risk.  To generate sidechain, add appropriate
entries to <cite>ic_data_sidechains</cite> in <a class="reference internal" href="Bio.PDB.ic_data.html#module-Bio.PDB.ic_data" title="Bio.PDB.ic_data"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ic_data</span></code></a> and support in
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Chain.atom_to_internal_coordinates()</span></code></a>.</p>
</dd>
<dt><strong>gly_Cbeta: bool default False</strong></dt><dd><p><strong>Class</strong> variable <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.gly_Cbeta" title="Bio.PDB.internal_coords.IC_Residue.gly_Cbeta"><code class="xref py py-data docutils literal notranslate"><span class="pre">gly_Cbeta</span></code></a>, override to True to generate
internal coordinates for glycine CB atoms in
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Chain.atom_to_internal_coordinates()</span></code></a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">IC_Residue</span><span class="o">.</span><span class="n">gly_Cbeta</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
</dd>
<dt><strong>pic_accuracy: str default “17.13f”</strong></dt><dd><p><strong>Class</strong> variable <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pic_accuracy" title="Bio.PDB.internal_coords.IC_Residue.pic_accuracy"><code class="xref py py-data docutils literal notranslate"><span class="pre">pic_accuracy</span></code></a> sets accuracy for numeric values
(angles, lengths) in .pic files.  Default set high to support mmCIF file
accuracy in rebuild tests.  If you find rebuild tests fail with
‘ERROR -COORDINATES-’ and verbose=True shows only small discrepancies,
try raising this value (or lower it to 9.5 if only working with PDB
format files).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">IC_Residue</span><span class="o">.</span><span class="n">pic_accuracy</span> <span class="o">=</span> <span class="s2">&quot;9.5f&quot;</span>
</pre></div>
</div>
</dd>
<dt><strong>residue: Biopython Residue object reference</strong></dt><dd><p>The <a class="reference internal" href="Bio.PDB.Residue.html#Bio.PDB.Residue.Residue" title="Bio.PDB.Residue.Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">Residue</span></code></a> object this extends</p>
</dd>
<dt><strong>hedra: dict indexed by 3-tuples of AtomKeys</strong></dt><dd><p>Hedra forming this residue</p>
</dd>
<dt><strong>dihedra: dict indexed by 4-tuples of AtomKeys</strong></dt><dd><p>Dihedra forming (overlapping) this residue</p>
</dd>
<dt><strong>rprev, rnext: lists of IC_Residue objects</strong></dt><dd><p>References to adjacent (bonded, not missing, possibly disordered)
residues in chain</p>
</dd>
<dt><strong>atom_coords: AtomKey indexed dict of numpy [4] arrays</strong></dt><dd><p><strong>removed</strong>
Use AtomKeys and atomArrayIndex to build if needed</p>
</dd>
<dt><strong>ak_set: set of AtomKeys in dihedra</strong></dt><dd><p>AtomKeys in all dihedra overlapping this residue (see __contains__())</p>
</dd>
<dt><strong>alt_ids: list of char</strong></dt><dd><p>AltLoc IDs from PDB file</p>
</dd>
<dt><strong>bfactors: dict</strong></dt><dd><p>AtomKey indexed B-factors as read from PDB file</p>
</dd>
<dt><strong>NCaCKey: List of tuples of AtomKeys</strong></dt><dd><p>List of tuples of N, Ca, C backbone atom AtomKeys; usually only 1
but more if backbone altlocs.</p>
</dd>
<dt><strong>is20AA: bool</strong></dt><dd><p>True if residue is one of 20 standard amino acids, based on
Residue resname</p>
</dd>
<dt><strong>isAccept: bool</strong></dt><dd><p>True if is20AA or in accept_resnames below</p>
</dd>
<dt><strong>rbase: tuple</strong></dt><dd><p>residue position, insert code or none, resname (1 letter if standard
amino acid)</p>
</dd>
<dt><strong>cic: IC_Chain default None</strong></dt><dd><p>parent chain <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Chain</span></code></a> object</p>
</dd>
<dt><strong>scale: optional float</strong></dt><dd><p>used for OpenSCAD output to generate gly_Cbeta bond length</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 26%" />
<col style="width: 74%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>assemble(atomCoordsIn, resetLocation, verbose)</strong></p></td>
<td><p>Compute atom coordinates for this residue from internal coordinates</p></td>
</tr>
<tr class="row-even"><td><p><strong>get_angle()</strong></p></td>
<td><p>Return angle for passed key</p></td>
</tr>
<tr class="row-odd"><td><p><strong>get_length()</strong></p></td>
<td><p>Return bond length for specified pair</p></td>
</tr>
<tr class="row-even"><td><p><strong>pick_angle()</strong></p></td>
<td><p>Find Hedron or Dihedron for passed key</p></td>
</tr>
<tr class="row-odd"><td><p><strong>pick_length()</strong></p></td>
<td><p>Find hedra for passed AtomKey pair</p></td>
</tr>
<tr class="row-even"><td><p><strong>set_angle()</strong></p></td>
<td><p>Set angle for passed key (no position updates)</p></td>
</tr>
<tr class="row-odd"><td><p><strong>set_length()</strong></p></td>
<td><p>Set bond length in all relevant hedra for specified pair</p></td>
</tr>
<tr class="row-even"><td><p><strong>bond_rotate(delta)</strong></p></td>
<td><p>adjusts related dihedra angles by delta, e.g. rotating psi (N-Ca-C-N) will adjust the adjacent N-Ca-C-O by the same amount to avoid clashes</p></td>
</tr>
<tr class="row-odd"><td><p><strong>bond_set(angle)</strong></p></td>
<td><p>uses bond_rotate to set specified dihedral to angle and adjust related dihedra accordingly</p></td>
</tr>
<tr class="row-even"><td><p><strong>rak(atom info)</strong></p></td>
<td><p>cached AtomKeys for this residue</p></td>
</tr>
</tbody>
</table>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.accept_resnames">
<span class="sig-name descname"><span class="pre">accept_resnames</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('CYG',</span> <span class="pre">'YCM',</span> <span class="pre">'UNK')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.accept_resnames" title="Permalink to this definition"></a></dt>
<dd><p>Add 3-letter residue name here for non-standard residues with
normal backbone.  CYG included for test case 4LGY (1305 residue
contiguous chain).  Safe to add more names for N-CA-C-O backbones, any
more complexity will need additions to <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.accept_atoms" title="Bio.PDB.internal_coords.IC_Residue.accept_atoms"><code class="xref py py-data docutils literal notranslate"><span class="pre">accept_atoms</span></code></a>,
<cite>ic_data_sidechains</cite> in <a class="reference internal" href="Bio.PDB.ic_data.html#module-Bio.PDB.ic_data" title="Bio.PDB.ic_data"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ic_data</span></code></a> and support in
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates" title="Bio.PDB.internal_coords.IC_Chain.atom_to_internal_coordinates"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Chain.atom_to_internal_coordinates()</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.no_altloc">
<span class="sig-name descname"><span class="pre">no_altloc</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">False</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.no_altloc" title="Permalink to this definition"></a></dt>
<dd><p>Set True to filter altloc atoms on input and only work with Biopython
default Atoms</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.gly_Cbeta">
<span class="sig-name descname"><span class="pre">gly_Cbeta</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">False</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.gly_Cbeta" title="Permalink to this definition"></a></dt>
<dd><p>Create beta carbons on all Gly residues.</p>
<p>Setting this to True will generate internal coordinates for Gly C-beta
carbons in <code class="xref py py-meth docutils literal notranslate"><span class="pre">atom_to_internal_coordinates()</span></code>.</p>
<p>Data averaged from Sep 2019 Dunbrack cullpdb_pc20_res2.2_R1.0
restricted to structures with amide protons.
Please see</p>
<p><a class="reference external" href="https://dunbrack.fccc.edu/pisces/">PISCES: A Protein Sequence Culling Server</a></p>
<p>‘G. Wang and R. L. Dunbrack, Jr. PISCES: a protein sequence culling
server. Bioinformatics, 19:1589-1591, 2003.’</p>
<p>Ala avg rotation of OCCACB from NCACO query:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">select</span> <span class="n">avg</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">rslt</span><span class="p">)</span> <span class="k">as</span> <span class="n">avg_rslt</span><span class="p">,</span> <span class="n">stddev</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">rslt</span><span class="p">)</span> <span class="k">as</span> <span class="n">sd_rslt</span><span class="p">,</span> <span class="n">count</span><span class="p">(</span><span class="o">*</span><span class="p">)</span>
<span class="kn">from</span>
<span class="p">(</span><span class="n">select</span> <span class="n">f</span><span class="o">.</span><span class="n">d1d</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">d2d</span><span class="p">,</span>
<span class="p">(</span><span class="n">case</span> <span class="n">when</span> <span class="n">f</span><span class="o">.</span><span class="n">rslt</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="n">then</span> <span class="n">f</span><span class="o">.</span><span class="n">rslt</span><span class="o">-</span><span class="mf">360.0</span> <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">rslt</span> <span class="n">end</span><span class="p">)</span> <span class="k">as</span> <span class="n">rslt</span>
<span class="kn">from</span> <span class="p">(</span><span class="n">select</span> <span class="n">d1</span><span class="o">.</span><span class="n">angle</span> <span class="k">as</span> <span class="n">d1d</span><span class="p">,</span> <span class="n">d2</span><span class="o">.</span><span class="n">angle</span> <span class="k">as</span> <span class="n">d2d</span><span class="p">,</span>
<span class="p">(</span><span class="n">d2</span><span class="o">.</span><span class="n">angle</span> <span class="o">-</span> <span class="n">d1</span><span class="o">.</span><span class="n">angle</span><span class="p">)</span> <span class="k">as</span> <span class="n">rslt</span> <span class="kn">from</span> <span class="nn">dihedron</span> <span class="n">d1</span><span class="p">,</span>
<span class="n">dihedron</span> <span class="n">d2</span> <span class="n">where</span> <span class="n">d1</span><span class="o">.</span><span class="n">re_class</span><span class="o">=</span><span class="s1">&#39;AOACACAACB&#39;</span> <span class="ow">and</span>
<span class="n">d2</span><span class="o">.</span><span class="n">re_class</span><span class="o">=</span><span class="s1">&#39;ANACAACAO&#39;</span> <span class="ow">and</span> <span class="n">d1</span><span class="o">.</span><span class="n">pdb</span><span class="o">=</span><span class="n">d2</span><span class="o">.</span><span class="n">pdb</span> <span class="ow">and</span> <span class="n">d1</span><span class="o">.</span><span class="n">chn</span><span class="o">=</span><span class="n">d2</span><span class="o">.</span><span class="n">chn</span>
<span class="ow">and</span> <span class="n">d1</span><span class="o">.</span><span class="n">res</span><span class="o">=</span><span class="n">d2</span><span class="o">.</span><span class="n">res</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">)</span> <span class="k">as</span> <span class="n">g</span>
</pre></div>
</div>
<p>results:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">|</span> <span class="n">avg_rslt</span>          <span class="o">|</span> <span class="n">sd_rslt</span>          <span class="o">|</span> <span class="n">count</span>   <span class="o">|</span>
<span class="o">|</span> <span class="o">-</span><span class="mf">122.682194862932</span> <span class="o">|</span> <span class="mf">5.04403040513919</span> <span class="o">|</span> <span class="mi">14098</span>   <span class="o">|</span>
</pre></div>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.pic_accuracy">
<span class="sig-name descname"><span class="pre">pic_accuracy</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'17.13f'</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.pic_accuracy" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.accept_backbone">
<span class="sig-name descname"><span class="pre">accept_backbone</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('N',</span> <span class="pre">'CA',</span> <span class="pre">'C',</span> <span class="pre">'O',</span> <span class="pre">'OXT')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.accept_backbone" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.accept_sidechain">
<span class="sig-name descname"><span class="pre">accept_sidechain</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('CB',</span> <span class="pre">'CG',</span> <span class="pre">'CG1',</span> <span class="pre">'OG1',</span> <span class="pre">'OG',</span> <span class="pre">'SG',</span> <span class="pre">'CG2',</span> <span class="pre">'CD',</span> <span class="pre">'CD1',</span> <span class="pre">'SD',</span> <span class="pre">'OD1',</span> <span class="pre">'ND1',</span> <span class="pre">'CD2',</span> <span class="pre">'ND2',</span> <span class="pre">'CE',</span> <span class="pre">'CE1',</span> <span class="pre">'NE',</span> <span class="pre">'OE1',</span> <span class="pre">'NE1',</span> <span class="pre">'CE2',</span> <span class="pre">'OE2',</span> <span class="pre">'NE2',</span> <span class="pre">'CE3',</span> <span class="pre">'CZ',</span> <span class="pre">'NZ',</span> <span class="pre">'CZ2',</span> <span class="pre">'CZ3',</span> <span class="pre">'OD2',</span> <span class="pre">'OH',</span> <span class="pre">'CH2',</span> <span class="pre">'NH1',</span> <span class="pre">'NH2')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.accept_sidechain" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.accept_mainchain">
<span class="sig-name descname"><span class="pre">accept_mainchain</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('N',</span> <span class="pre">'CA',</span> <span class="pre">'C',</span> <span class="pre">'O',</span> <span class="pre">'OXT',</span> <span class="pre">'CB',</span> <span class="pre">'CG',</span> <span class="pre">'CG1',</span> <span class="pre">'OG1',</span> <span class="pre">'OG',</span> <span class="pre">'SG',</span> <span class="pre">'CG2',</span> <span class="pre">'CD',</span> <span class="pre">'CD1',</span> <span class="pre">'SD',</span> <span class="pre">'OD1',</span> <span class="pre">'ND1',</span> <span class="pre">'CD2',</span> <span class="pre">'ND2',</span> <span class="pre">'CE',</span> <span class="pre">'CE1',</span> <span class="pre">'NE',</span> <span class="pre">'OE1',</span> <span class="pre">'NE1',</span> <span class="pre">'CE2',</span> <span class="pre">'OE2',</span> <span class="pre">'NE2',</span> <span class="pre">'CE3',</span> <span class="pre">'CZ',</span> <span class="pre">'NZ',</span> <span class="pre">'CZ2',</span> <span class="pre">'CZ3',</span> <span class="pre">'OD2',</span> <span class="pre">'OH',</span> <span class="pre">'CH2',</span> <span class="pre">'NH1',</span> <span class="pre">'NH2')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.accept_mainchain" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.accept_hydrogens">
<span class="sig-name descname"><span class="pre">accept_hydrogens</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('H',</span> <span class="pre">'H1',</span> <span class="pre">'H2',</span> <span class="pre">'H3',</span> <span class="pre">'HA',</span> <span class="pre">'HA2',</span> <span class="pre">'HA3',</span> <span class="pre">'HB',</span> <span class="pre">'HB1',</span> <span class="pre">'HB2',</span> <span class="pre">'HB3',</span> <span class="pre">'HG2',</span> <span class="pre">'HG3',</span> <span class="pre">'HD2',</span> <span class="pre">'HD3',</span> <span class="pre">'HE2',</span> <span class="pre">'HE3',</span> <span class="pre">'HZ1',</span> <span class="pre">'HZ2',</span> <span class="pre">'HZ3',</span> <span class="pre">'HG11',</span> <span class="pre">'HG12',</span> <span class="pre">'HG13',</span> <span class="pre">'HG21',</span> <span class="pre">'HG22',</span> <span class="pre">'HG23',</span> <span class="pre">'HZ',</span> <span class="pre">'HD1',</span> <span class="pre">'HE1',</span> <span class="pre">'HD11',</span> <span class="pre">'HD12',</span> <span class="pre">'HD13',</span> <span class="pre">'HG',</span> <span class="pre">'HG1',</span> <span class="pre">'HD21',</span> <span class="pre">'HD22',</span> <span class="pre">'HD23',</span> <span class="pre">'NH1',</span> <span class="pre">'NH2',</span> <span class="pre">'HE',</span> <span class="pre">'HH11',</span> <span class="pre">'HH12',</span> <span class="pre">'HH21',</span> <span class="pre">'HH22',</span> <span class="pre">'HE21',</span> <span class="pre">'HE22',</span> <span class="pre">'HE2',</span> <span class="pre">'HH',</span> <span class="pre">'HH2')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.accept_hydrogens" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.accept_deuteriums">
<span class="sig-name descname"><span class="pre">accept_deuteriums</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('D',</span> <span class="pre">'D1',</span> <span class="pre">'D2',</span> <span class="pre">'D3',</span> <span class="pre">'DA',</span> <span class="pre">'DA2',</span> <span class="pre">'DA3',</span> <span class="pre">'DB',</span> <span class="pre">'DB1',</span> <span class="pre">'DB2',</span> <span class="pre">'DB3',</span> <span class="pre">'DG2',</span> <span class="pre">'DG3',</span> <span class="pre">'DD2',</span> <span class="pre">'DD3',</span> <span class="pre">'DE2',</span> <span class="pre">'DE3',</span> <span class="pre">'DZ1',</span> <span class="pre">'DZ2',</span> <span class="pre">'DZ3',</span> <span class="pre">'DG11',</span> <span class="pre">'DG12',</span> <span class="pre">'DG13',</span> <span class="pre">'DG21',</span> <span class="pre">'DG22',</span> <span class="pre">'DG23',</span> <span class="pre">'DZ',</span> <span class="pre">'DD1',</span> <span class="pre">'DE1',</span> <span class="pre">'DD11',</span> <span class="pre">'DD12',</span> <span class="pre">'DD13',</span> <span class="pre">'DG',</span> <span class="pre">'DG1',</span> <span class="pre">'DD21',</span> <span class="pre">'DD22',</span> <span class="pre">'DD23',</span> <span class="pre">'ND1',</span> <span class="pre">'ND2',</span> <span class="pre">'DE',</span> <span class="pre">'DH11',</span> <span class="pre">'DH12',</span> <span class="pre">'DH21',</span> <span class="pre">'DH22',</span> <span class="pre">'DE21',</span> <span class="pre">'DE22',</span> <span class="pre">'DE2',</span> <span class="pre">'DH',</span> <span class="pre">'DH2')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.accept_deuteriums" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.accept_atoms">
<span class="sig-name descname"><span class="pre">accept_atoms</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('N',</span> <span class="pre">'CA',</span> <span class="pre">'C',</span> <span class="pre">'O',</span> <span class="pre">'OXT',</span> <span class="pre">'CB',</span> <span class="pre">'CG',</span> <span class="pre">'CG1',</span> <span class="pre">'OG1',</span> <span class="pre">'OG',</span> <span class="pre">'SG',</span> <span class="pre">'CG2',</span> <span class="pre">'CD',</span> <span class="pre">'CD1',</span> <span class="pre">'SD',</span> <span class="pre">'OD1',</span> <span class="pre">'ND1',</span> <span class="pre">'CD2',</span> <span class="pre">'ND2',</span> <span class="pre">'CE',</span> <span class="pre">'CE1',</span> <span class="pre">'NE',</span> <span class="pre">'OE1',</span> <span class="pre">'NE1',</span> <span class="pre">'CE2',</span> <span class="pre">'OE2',</span> <span class="pre">'NE2',</span> <span class="pre">'CE3',</span> <span class="pre">'CZ',</span> <span class="pre">'NZ',</span> <span class="pre">'CZ2',</span> <span class="pre">'CZ3',</span> <span class="pre">'OD2',</span> <span class="pre">'OH',</span> <span class="pre">'CH2',</span> <span class="pre">'NH1',</span> <span class="pre">'NH2',</span> <span class="pre">'H',</span> <span class="pre">'H1',</span> <span class="pre">'H2',</span> <span class="pre">'H3',</span> <span class="pre">'HA',</span> <span class="pre">'HA2',</span> <span class="pre">'HA3',</span> <span class="pre">'HB',</span> <span class="pre">'HB1',</span> <span class="pre">'HB2',</span> <span class="pre">'HB3',</span> <span class="pre">'HG2',</span> <span class="pre">'HG3',</span> <span class="pre">'HD2',</span> <span class="pre">'HD3',</span> <span class="pre">'HE2',</span> <span class="pre">'HE3',</span> <span class="pre">'HZ1',</span> <span class="pre">'HZ2',</span> <span class="pre">'HZ3',</span> <span class="pre">'HG11',</span> <span class="pre">'HG12',</span> <span class="pre">'HG13',</span> <span class="pre">'HG21',</span> <span class="pre">'HG22',</span> <span class="pre">'HG23',</span> <span class="pre">'HZ',</span> <span class="pre">'HD1',</span> <span class="pre">'HE1',</span> <span class="pre">'HD11',</span> <span class="pre">'HD12',</span> <span class="pre">'HD13',</span> <span class="pre">'HG',</span> <span class="pre">'HG1',</span> <span class="pre">'HD21',</span> <span class="pre">'HD22',</span> <span class="pre">'HD23',</span> <span class="pre">'NH1',</span> <span class="pre">'NH2',</span> <span class="pre">'HE',</span> <span class="pre">'HH11',</span> <span class="pre">'HH12',</span> <span class="pre">'HH21',</span> <span class="pre">'HH22',</span> <span class="pre">'HE21',</span> <span class="pre">'HE22',</span> <span class="pre">'HE2',</span> <span class="pre">'HH',</span> <span class="pre">'HH2')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.accept_atoms" title="Permalink to this definition"></a></dt>
<dd><p>Change accept_atoms to restrict atoms processed. See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue" title="Bio.PDB.internal_coords.IC_Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Residue</span></code></a>
for usage.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Residue</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize IC_Residue with parent Biopython Residue.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>parent</strong> (<a class="reference internal" href="Bio.PDB.Residue.html#Bio.PDB.Residue.Residue" title="Bio.PDB.Residue.Residue"><em>Residue</em></a>) – Biopython Residue object.
The Biopython Residue this object extends</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.__deepcopy__">
<span class="sig-name descname"><span class="pre">__deepcopy__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.__deepcopy__" title="Permalink to this definition"></a></dt>
<dd><p>Deep copy implementation for IC_Residue.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.__contains__">
<span class="sig-name descname"><span class="pre">__contains__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ak</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.__contains__" title="Permalink to this definition"></a></dt>
<dd><p>Return True if atomkey is in this residue.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.rak">
<span class="sig-name descname"><span class="pre">rak</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="Bio.PDB.Atom.html#Bio.PDB.Atom.Atom" title="Bio.PDB.Atom.Atom"><span class="pre">Bio.PDB.Atom.Atom</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.rak" title="Permalink to this definition"></a></dt>
<dd><p>Cache calls to AtomKey for this residue.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Print string is parent Residue ID.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.pretty_str">
<span class="sig-name descname"><span class="pre">pretty_str</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.pretty_str" title="Permalink to this definition"></a></dt>
<dd><p>Nice string for residue ID.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.set_flexible">
<span class="sig-name descname"><span class="pre">set_flexible</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.set_flexible" title="Permalink to this definition"></a></dt>
<dd><p>For OpenSCAD, mark N-CA and CA-C bonds to be flexible joints.</p>
<p>See <a class="reference internal" href="Bio.PDB.SCADIO.html#Bio.PDB.SCADIO.write_SCAD" title="Bio.PDB.SCADIO.write_SCAD"><code class="xref py py-func docutils literal notranslate"><span class="pre">SCADIO.write_SCAD()</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.set_hbond">
<span class="sig-name descname"><span class="pre">set_hbond</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.set_hbond" title="Permalink to this definition"></a></dt>
<dd><p>For OpenSCAD, mark H-N and C-O bonds to be hbonds (magnets).</p>
<p>See <a class="reference internal" href="Bio.PDB.SCADIO.html#Bio.PDB.SCADIO.write_SCAD" title="Bio.PDB.SCADIO.write_SCAD"><code class="xref py py-func docutils literal notranslate"><span class="pre">SCADIO.write_SCAD()</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.clear_transforms">
<span class="sig-name descname"><span class="pre">clear_transforms</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.clear_transforms" title="Permalink to this definition"></a></dt>
<dd><p>Invalidate dihedra coordinate space attributes before assemble().</p>
<p>Coordinate space attributes are Dihedron.cst and .rcst, and
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.dCoordSpace"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Chain.dCoordSpace</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.assemble">
<span class="sig-name descname"><span class="pre">assemble</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resetLocation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">numpy.array</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">numpy.array</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.assemble" title="Permalink to this definition"></a></dt>
<dd><p>Compute atom coordinates for this residue from internal coordinates.</p>
<p>This is the IC_Residue part of the <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues_ser"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues_ser()</span></code></a> serial
version, see <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues()</span></code></a> for numpy vectorized approach
which works at the <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Chain</span></code></a> level.</p>
<p>Join prepared dihedra starting from N-CA-C and N-CA-CB hedrons,
computing protein space coordinates for backbone and sidechain atoms</p>
<p>Sets forward and reverse transforms on each Dihedron to convert from
protein coordinates to dihedron space coordinates for first three
atoms (see <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.dCoordSpace"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Chain.dCoordSpace</span></code></a>)</p>
<p>Call <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.init_atom_coords" title="Bio.PDB.internal_coords.IC_Chain.init_atom_coords"><code class="xref py py-meth docutils literal notranslate"><span class="pre">init_atom_coords()</span></code></a> to update any modified di/hedra before
coming here, this only assembles dihedra into protein coordinate space.</p>
<p><strong>Algorithm</strong></p>
<p>Form double-ended queue, start with c-ca-n, o-c-ca, n-ca-cb, n-ca-c.</p>
<p>if resetLocation=True, use initial coords from generating dihedron
for n-ca-c initial positions (result in dihedron coordinate space)</p>
<dl>
<dt>while queue not empty</dt><dd><p>get 3-atom hedron key</p>
<p>for each dihedron starting with hedron key (1st hedron of dihedron)</p>
<blockquote>
<div><dl>
<dt>if have coordinates for all 4 atoms already</dt><dd><p>add 2nd hedron key to back of queue</p>
</dd>
<dt>else if have coordinates for 1st 3 atoms</dt><dd><p>compute forward and reverse transforms to take 1st 3 atoms
to/from dihedron initial coordinate space</p>
<p>use reverse transform to get position of 4th atom in
current coordinates from dihedron initial coordinates</p>
<p>add 2nd hedron key to back of queue</p>
</dd>
<dt>else</dt><dd><p>ordering failed, put hedron key at back of queue and hope
next time we have 1st 3 atom positions (should not happen)</p>
</dd>
</dl>
</div></blockquote>
</dd>
</dl>
<p>loop terminates (queue drains) as hedron keys which do not start any
dihedra are removed without action</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>resetLocation</strong> (<em>bool</em>) – default False.
- Option to ignore start location and orient so initial N-Ca-C
hedron at origin.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>Dict of AtomKey -&gt; homogeneous atom coords for residue in protein
space relative to previous residue</p>
<p><strong>Also</strong> directly updates <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.atomArray" title="Bio.PDB.internal_coords.IC_Chain.atomArray"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Chain.atomArray</span></code></a> as
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.assemble_residues" title="Bio.PDB.internal_coords.IC_Chain.assemble_residues"><code class="xref py py-meth docutils literal notranslate"><span class="pre">assemble_residues()</span></code></a> does.</p>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.split_akl">
<span class="sig-name descname"><span class="pre">split_akl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missingOK</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="p"><span class="pre">...</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.split_akl" title="Permalink to this definition"></a></dt>
<dd><p>Get AtomKeys for this residue (ak_set) for generic list of AtomKeys.</p>
<p>Changes and/or expands a list of ‘generic’ AtomKeys (e.g. ‘N, C, C’) to
be specific to this Residue’s altlocs etc., e.g.
‘(N-Ca_A_0.3-C, N-Ca_B_0.7-C)’</p>
<dl>
<dt>Given a list of AtomKeys for a Hedron or Dihedron,</dt><dd><dl>
<dt>return:</dt><dd><blockquote>
<div><p>list of matching atomkeys that have id3_dh in this residue
(ak may change if occupancy != 1.00)</p>
</div></blockquote>
<dl class="simple">
<dt>or</dt><dd><p>multiple lists of matching atomkeys expanded for all atom altlocs</p>
</dd>
<dt>or</dt><dd><p>empty list if any of atom_coord(ak) missing and not missingOK</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lst</strong> (<em>list</em>) – list[3] or [4] of AtomKeys.
Non-altloc AtomKeys to match to specific AtomKeys for this residue</p></li>
<li><p><strong>missingOK</strong> (<em>bool</em>) – default False, see above.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.atom_sernum">
<span class="sig-name descname"><span class="pre">atom_sernum</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.atom_sernum" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.atom_chain">
<span class="sig-name descname"><span class="pre">atom_chain</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.atom_chain" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.pdb_residue_string">
<span class="sig-name descname"><span class="pre">pdb_residue_string</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.pdb_residue_string" title="Permalink to this definition"></a></dt>
<dd><p>Generate PDB ATOM records for this residue as string.</p>
<p>Convenience method for functionality not exposed in PDBIO.py.
Increments <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.atom_sernum" title="Bio.PDB.internal_coords.IC_Residue.atom_sernum"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Residue.atom_sernum</span></code></a> if not None</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>IC_Residue.atom_sernum</strong> – Class variable default None.
Override and increment atom serial number if not None</p></li>
<li><p><strong>IC_Residue.atom_chain</strong> – Class variable.
Override atom chain id if not None</p></li>
</ul>
</dd>
</dl>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>move to PDBIO</p>
</div>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.pic_flags">
<span class="sig-name descname"><span class="pre">pic_flags</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_pfDef(psi=1,</span> <span class="pre">omg=2,</span> <span class="pre">phi=4,</span> <span class="pre">tau=8,</span> <span class="pre">chi1=16,</span> <span class="pre">chi2=32,</span> <span class="pre">chi3=64,</span> <span class="pre">chi4=128,</span> <span class="pre">chi5=256,</span> <span class="pre">pomg=512,</span> <span class="pre">chi=496,</span> <span class="pre">classic_b=525,</span> <span class="pre">classic=1021,</span> <span class="pre">hedra=1024,</span> <span class="pre">primary=2048,</span> <span class="pre">secondary=4096,</span> <span class="pre">all=7168,</span> <span class="pre">initAtoms=8192,</span> <span class="pre">bFactors=16384)</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.pic_flags" title="Permalink to this definition"></a></dt>
<dd><p>Used by <a class="reference internal" href="Bio.PDB.PICIO.html#Bio.PDB.PICIO.write_PIC" title="Bio.PDB.PICIO.write_PIC"><code class="xref py py-func docutils literal notranslate"><span class="pre">PICIO.write_PIC()</span></code></a> to control classes of values to be defaulted.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.picFlagsDefault">
<span class="sig-name descname"><span class="pre">picFlagsDefault</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">31744</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.picFlagsDefault" title="Permalink to this definition"></a></dt>
<dd><p>Default is all dihedra + initial tau atoms + bFactors.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.picFlagsDict">
<span class="sig-name descname"><span class="pre">picFlagsDict</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'all':</span> <span class="pre">7168,</span> <span class="pre">'bFactors':</span> <span class="pre">16384,</span> <span class="pre">'chi':</span> <span class="pre">496,</span> <span class="pre">'chi1':</span> <span class="pre">16,</span> <span class="pre">'chi2':</span> <span class="pre">32,</span> <span class="pre">'chi3':</span> <span class="pre">64,</span> <span class="pre">'chi4':</span> <span class="pre">128,</span> <span class="pre">'chi5':</span> <span class="pre">256,</span> <span class="pre">'classic':</span> <span class="pre">1021,</span> <span class="pre">'classic_b':</span> <span class="pre">525,</span> <span class="pre">'hedra':</span> <span class="pre">1024,</span> <span class="pre">'initAtoms':</span> <span class="pre">8192,</span> <span class="pre">'omg':</span> <span class="pre">2,</span> <span class="pre">'phi':</span> <span class="pre">4,</span> <span class="pre">'pomg':</span> <span class="pre">512,</span> <span class="pre">'primary':</span> <span class="pre">2048,</span> <span class="pre">'psi':</span> <span class="pre">1,</span> <span class="pre">'secondary':</span> <span class="pre">4096,</span> <span class="pre">'tau':</span> <span class="pre">8}</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.picFlagsDict" title="Permalink to this definition"></a></dt>
<dd><p>Dictionary of pic_flags values to use as needed.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.pick_angle">
<span class="sig-name descname"><span class="pre">pick_angle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angle_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.Hedron" title="Bio.PDB.internal_coords.Hedron"><span class="pre">Bio.PDB.internal_coords.Hedron</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.Dihedron" title="Bio.PDB.internal_coords.Dihedron"><span class="pre">Bio.PDB.internal_coords.Dihedron</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.pick_angle" title="Permalink to this definition"></a></dt>
<dd><p>Get Hedron or Dihedron for angle_key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>angle_key</strong> – <ul class="simple">
<li><p>tuple of 3 or 4 AtomKeys</p></li>
<li><p>string of atom names (‘CA’) separated by :’s</p></li>
<li><p>string of [-1, 0, 1]&lt;atom name&gt; separated by ‘:’s. -1 is
previous residue, 0 is this residue, 1 is next residue</p></li>
<li><p>psi, phi, omg, omega, chi1, chi2, chi3, chi4, chi5</p></li>
<li><p>tau (N-CA-C angle) see Richardson1981</p></li>
<li><p>tuples of AtomKeys is only access for alternate disordered atoms</p></li>
</ul>
</p>
</dd>
</dl>
<p>Observe that a residue’s phi and omega dihedrals, as well as the hedra
comprising them (including the N:Ca:C <cite>tau</cite> hedron), are stored in the
n-1 di/hedra sets; this overlap is handled here, but may be an issue if
accessing directly.</p>
<p>The following print commands are equivalent (except for sidechains with
non-carbon atoms for chi2):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ric</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">internal_coord</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="n">r</span><span class="p">,</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;psi&quot;</span><span class="p">),</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;phi&quot;</span><span class="p">),</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;omg&quot;</span><span class="p">),</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;tau&quot;</span><span class="p">),</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;chi2&quot;</span><span class="p">),</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="n">r</span><span class="p">,</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;N:CA:C:1N&quot;</span><span class="p">),</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;-1C:N:CA:C&quot;</span><span class="p">),</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;-1CA:-1C:N:CA&quot;</span><span class="p">),</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;N:CA:C&quot;</span><span class="p">),</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_angle</span><span class="p">(</span><span class="s2">&quot;CA:CB:CG:CD&quot;</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
<p>See ic_data.py for detail of atoms in the enumerated sidechain angles
and the backbone angles which do not span the peptide bond. Using ‘s’
for current residue (‘self’) and ‘n’ for next residue, the spanning
(overlapping) angles are:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">sN</span><span class="p">,</span> <span class="n">sCA</span><span class="p">,</span> <span class="n">sC</span><span class="p">,</span> <span class="n">nN</span><span class="p">)</span>   <span class="c1"># psi</span>
<span class="p">(</span><span class="n">sCA</span><span class="p">,</span> <span class="n">sC</span><span class="p">,</span> <span class="n">nN</span><span class="p">,</span> <span class="n">nCA</span><span class="p">)</span>  <span class="c1"># omega i+1</span>
<span class="p">(</span><span class="n">sC</span><span class="p">,</span> <span class="n">nN</span><span class="p">,</span> <span class="n">nCA</span><span class="p">,</span> <span class="n">nC</span><span class="p">)</span>   <span class="c1"># phi i+1</span>
<span class="p">(</span><span class="n">sCA</span><span class="p">,</span> <span class="n">sC</span><span class="p">,</span> <span class="n">nN</span><span class="p">)</span>
<span class="p">(</span><span class="n">sC</span><span class="p">,</span> <span class="n">nN</span><span class="p">,</span> <span class="n">nCA</span><span class="p">)</span>
<span class="p">(</span><span class="n">nN</span><span class="p">,</span> <span class="n">nCA</span><span class="p">,</span> <span class="n">nC</span><span class="p">)</span>       <span class="c1"># tau i+1</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Matching Hedron, Dihedron, or None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.get_angle">
<span class="sig-name descname"><span class="pre">get_angle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angle_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.get_angle" title="Permalink to this definition"></a></dt>
<dd><p>Get dihedron or hedron angle for specified key.</p>
<p>See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pick_angle" title="Bio.PDB.internal_coords.IC_Residue.pick_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pick_angle()</span></code></a> for key specifications.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.set_angle">
<span class="sig-name descname"><span class="pre">set_angle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angle_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.set_angle" title="Permalink to this definition"></a></dt>
<dd><p>Set dihedron or hedron angle for specified key.</p>
<p>If angle is a <cite>Dihedron</cite> and <cite>overlap</cite> is True (default), overlapping
dihedra are also changed as appropriate.  The overlap is a result of
protein chain definitions in <a class="reference internal" href="Bio.PDB.ic_data.html#module-Bio.PDB.ic_data" title="Bio.PDB.ic_data"><code class="xref py py-mod docutils literal notranslate"><span class="pre">ic_data</span></code></a> and <code class="xref py py-meth docutils literal notranslate"><span class="pre">_create_edra()</span></code>
(e.g. psi overlaps N-CA-C-O).</p>
<p>Te default overlap=True is probably what you want for:
<cite>set_angle(“chi1”, val)</cite></p>
<p>The default is probably NOT what you want when processing all dihedrals
in a chain or residue (such as copying from another structure), as the
overlaping dihedra will likely be in the set as well.</p>
<p>N.B. setting e.g. PRO chi2 is permitted without error or warning!</p>
<p>See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pick_angle" title="Bio.PDB.internal_coords.IC_Residue.pick_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pick_angle()</span></code></a> for angle_key specifications.
See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.bond_rotate" title="Bio.PDB.internal_coords.IC_Residue.bond_rotate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">bond_rotate()</span></code></a> to change a dihedral by a number of degrees</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>angle_key</strong> – angle identifier.</p></li>
<li><p><strong>v</strong> (<em>float</em>) – new angle in degrees (result adjusted to +/-180).</p></li>
<li><p><strong>overlap</strong> (<em>bool</em>) – default True.
Modify overlapping dihedra as needed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.bond_rotate">
<span class="sig-name descname"><span class="pre">bond_rotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angle_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.bond_rotate" title="Permalink to this definition"></a></dt>
<dd><p>Rotate set of overlapping dihedrals by delta degrees.</p>
<p>Changes a dihedral angle by a given delta, i.e.
new_angle = current_angle + delta
Values are adjusted so new_angle iwll be within +/-180.</p>
<p>Changes overlapping dihedra as in <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.set_angle" title="Bio.PDB.internal_coords.IC_Residue.set_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_angle()</span></code></a></p>
<p>See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pick_angle" title="Bio.PDB.internal_coords.IC_Residue.pick_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pick_angle()</span></code></a> for key specifications.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.bond_set">
<span class="sig-name descname"><span class="pre">bond_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angle_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.bond_set" title="Permalink to this definition"></a></dt>
<dd><p>Set dihedron to val, update overlapping dihedra by same amount.</p>
<p>Redundant to <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.set_angle" title="Bio.PDB.internal_coords.IC_Residue.set_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_angle()</span></code></a>, retained for compatibility.  Unlike
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.set_angle" title="Bio.PDB.internal_coords.IC_Residue.set_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">set_angle()</span></code></a> this is for dihedra only and no option to not update
overlapping dihedra.</p>
<p>See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pick_angle" title="Bio.PDB.internal_coords.IC_Residue.pick_angle"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pick_angle()</span></code></a> for key specifications.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.pick_length">
<span class="sig-name descname"><span class="pre">pick_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ak_spec</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.Hedron" title="Bio.PDB.internal_coords.Hedron"><span class="pre">Bio.PDB.internal_coords.Hedron</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.pick_length" title="Permalink to this definition"></a></dt>
<dd><p>Get list of hedra containing specified atom pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ak_spec</strong> – <ul class="simple">
<li><p>tuple of two AtomKeys</p></li>
<li><p>string: two atom names separated by ‘:’, e.g. ‘N:CA’ with
optional position specifier relative to self, e.g. ‘-1C:N’ for
preceding peptide bond.  Position specifiers are -1, 0, 1.</p></li>
</ul>
</p>
</dd>
</dl>
<p>The following are equivalent:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ric</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">internal_coord</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="n">r</span><span class="p">,</span>
    <span class="n">ric</span><span class="o">.</span><span class="n">get_length</span><span class="p">(</span><span class="s2">&quot;0C:1N&quot;</span><span class="p">),</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="n">r</span><span class="p">,</span>
    <span class="kc">None</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">ric</span><span class="o">.</span><span class="n">rnext</span>
    <span class="k">else</span> <span class="n">ric</span><span class="o">.</span><span class="n">get_length</span><span class="p">((</span><span class="n">ric</span><span class="o">.</span><span class="n">rak</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">),</span> <span class="n">ric</span><span class="o">.</span><span class="n">rnext</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">rak</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">))),</span>
<span class="p">)</span>
</pre></div>
</div>
<p>If atom not found on current residue then will look on rprev[0] to
handle cases like Gly N:CA.  For finer control please access
<cite>IC_Chain.hedra</cite> directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of hedra containing specified atom pair as tuples of
AtomKeys</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.get_length">
<span class="sig-name descname"><span class="pre">get_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ak_spec</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.get_length" title="Permalink to this definition"></a></dt>
<dd><p>Get bond length for specified atom pair.</p>
<p>See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pick_length" title="Bio.PDB.internal_coords.IC_Residue.pick_length"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pick_length()</span></code></a> for ak_spec and details.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.set_length">
<span class="sig-name descname"><span class="pre">set_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ak_spec</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.set_length" title="Permalink to this definition"></a></dt>
<dd><p>Set bond length for specified atom pair.</p>
<p>See <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pick_length" title="Bio.PDB.internal_coords.IC_Residue.pick_length"><code class="xref py py-meth docutils literal notranslate"><span class="pre">pick_length()</span></code></a> for ak_spec.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.applyMtx">
<span class="sig-name descname"><span class="pre">applyMtx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mtx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">numpy.array</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.applyMtx" title="Permalink to this definition"></a></dt>
<dd><p>Apply matrix to atom_coords for this IC_Residue.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.IC_Residue.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'_AllBonds':</span> <span class="pre">&lt;class</span> <span class="pre">'bool'&gt;,</span> <span class="pre">'ak_set':</span> <span class="pre">'Set[AtomKey]',</span> <span class="pre">'akc':</span> <span class="pre">'Dict[Union[str,</span> <span class="pre">Atom],</span> <span class="pre">AtomKey]',</span> <span class="pre">'alt_ids':</span> <span class="pre">'Union[List[str],</span> <span class="pre">None]',</span> <span class="pre">'bfactors':</span> <span class="pre">'Dict[str,</span> <span class="pre">float]',</span> <span class="pre">'cic':</span> <span class="pre">'IC_Chain',</span> <span class="pre">'dihedra':</span> <span class="pre">'Dict[DKT,</span> <span class="pre">Dihedron]',</span> <span class="pre">'gly_Cbeta':</span> <span class="pre">&lt;class</span> <span class="pre">'bool'&gt;,</span> <span class="pre">'hedra':</span> <span class="pre">'Dict[HKT,</span> <span class="pre">Hedron]',</span> <span class="pre">'no_altloc':</span> <span class="pre">&lt;class</span> <span class="pre">'bool'&gt;,</span> <span class="pre">'pic_accuracy':</span> <span class="pre">&lt;class</span> <span class="pre">'str'&gt;,</span> <span class="pre">'rnext':</span> <span class="pre">'List[IC_Residue]',</span> <span class="pre">'rprev':</span> <span class="pre">'List[IC_Residue]'}</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.IC_Residue.__annotations__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">Edron</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for Hedron and Dihedron classes.</p>
<p>Supports rich comparison based on lists of AtomKeys.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>atomkeys: tuple</strong></dt><dd><p>3 (hedron) or 4 (dihedron) <a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomKey</span></code></a> s defining this di/hedron</p>
</dd>
<dt><strong>id: str</strong></dt><dd><p>‘:’-joined string of AtomKeys for this di/hedron</p>
</dd>
<dt><strong>needs_update: bool</strong></dt><dd><p>indicates di/hedron local atom_coords do NOT reflect current di/hedron
angle and length values in hedron local coordinate space</p>
</dd>
<dt><strong>e_class: str</strong></dt><dd><p>sequence of atoms (no position or residue) comprising di/hedron
for statistics</p>
</dd>
<dt><strong>re_class: str</strong></dt><dd><p>sequence of residue, atoms comprising di/hedron for statistics</p>
</dd>
<dt><strong>cre_class: str</strong></dt><dd><p>sequence of covalent radii classses comprising di/hedron for statistics</p>
</dd>
<dt><strong>edron_re: compiled regex (Class Attribute)</strong></dt><dd><p>A compiled regular expression matching string IDs for Hedron
and Dihedron objects</p>
</dd>
<dt><strong>cic: IC_Chain reference</strong></dt><dd><p>Chain internal coords object containing this hedron</p>
</dd>
<dt><strong>ndx: int</strong></dt><dd><p>index into IC_Chain level numpy data arrays for di/hedra.
Set in <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.init_edra" title="Bio.PDB.internal_coords.IC_Chain.init_edra"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Chain.init_edra()</span></code></a></p>
</dd>
<dt><strong>rc: int</strong></dt><dd><p>number of residues involved in this edron</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 54%" />
<col style="width: 46%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>gen_key([AtomKey, …] or AtomKey, …) (Static Method)</strong></p></td>
<td><p>generate a ‘:’-joined string of AtomKey Ids</p></td>
</tr>
<tr class="row-even"><td><p><strong>is_backbone()</strong></p></td>
<td><p>Return True if all atomkeys atoms are N, Ca, C or O</p></td>
</tr>
</tbody>
</table>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.edron_re">
<span class="sig-name descname"><span class="pre">edron_re</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">re.compile('^(?P&lt;pdbid&gt;\\w+)?\\s(?P&lt;chn&gt;[\\w|\\s])?\\s(?P&lt;a1&gt;[\\w\\-\\.]+):(?P&lt;a2&gt;[\\w\\-\\.]+):(?P&lt;a3&gt;[\\w\\-\\.]+)(:(?P&lt;a4&gt;[\\w\\-\\.]+))?\\s+(((?P&lt;len12&gt;\\S+)\\s+(?P&lt;angle&gt;\\S+)\\s+(?P&lt;len23&gt;\\S+)\\s*$)|((?P&lt;)</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.edron_re" title="Permalink to this definition"></a></dt>
<dd><p>A compiled regular expression matching string IDs for Hedron and
Dihedron objects</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.gen_key">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gen_key</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.gen_key" title="Permalink to this definition"></a></dt>
<dd><p>Generate string of ‘:’-joined AtomKey strings from input.</p>
<p>Generate ‘2_A_C:3_P_N:3_P_CA’ from (2_A_C, 3_P_N, 3_P_CA)
:param list lst: list of AtomKey objects</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.gen_tuple">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gen_tuple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">akstr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Tuple</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.gen_tuple" title="Permalink to this definition"></a></dt>
<dd><p>Generate AtomKey tuple for ‘:’-joined AtomKey string.</p>
<p>Generate (2_A_C, 3_P_N, 3_P_CA) from ‘2_A_C:3_P_N:3_P_CA’
:param str akstr: string of ‘:’-separated AtomKey strings</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize Edron with sequence of AtomKeys.</p>
<p>Acceptable input:</p>
<blockquote>
<div><p>[ AtomKey, … ]  : list of AtomKeys
AtomKey, …      : sequence of AtomKeys as args
{‘a1’: str, ‘a2’: str, … }  : dict of AtomKeys as ‘a1’, ‘a2’ …</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__deepcopy__">
<span class="sig-name descname"><span class="pre">__deepcopy__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__deepcopy__" title="Permalink to this definition"></a></dt>
<dd><p>Deep copy implementation for Edron.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__contains__">
<span class="sig-name descname"><span class="pre">__contains__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ak</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__contains__" title="Permalink to this definition"></a></dt>
<dd><p>Return True if atomkey is in this edron.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.is_backbone">
<span class="sig-name descname"><span class="pre">is_backbone</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.is_backbone" title="Permalink to this definition"></a></dt>
<dd><p>Report True for contains only N, C, CA, O, H atoms.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Tuple of AtomKeys is default repr string.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__hash__" title="Permalink to this definition"></a></dt>
<dd><p>Hash calculated at init from atomkeys tuple.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__eq__" title="Permalink to this definition"></a></dt>
<dd><p>Test for equality.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__ne__">
<span class="sig-name descname"><span class="pre">__ne__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__ne__" title="Permalink to this definition"></a></dt>
<dd><p>Test for inequality.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__gt__">
<span class="sig-name descname"><span class="pre">__gt__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__gt__" title="Permalink to this definition"></a></dt>
<dd><p>Test greater than.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__ge__">
<span class="sig-name descname"><span class="pre">__ge__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__ge__" title="Permalink to this definition"></a></dt>
<dd><p>Test greater or equal.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__lt__">
<span class="sig-name descname"><span class="pre">__lt__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__lt__" title="Permalink to this definition"></a></dt>
<dd><p>Test less than.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Edron.__le__">
<span class="sig-name descname"><span class="pre">__le__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Edron.__le__" title="Permalink to this definition"></a></dt>
<dd><p>Test less or equal.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">Hedron</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.PDB.internal_coords.Edron" title="Bio.PDB.internal_coords.Edron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.PDB.internal_coords.Edron</span></code></a></p>
<p>Class to represent three joined atoms forming a plane.</p>
<p>Contains atom coordinates in local coordinate space: central atom
at origin, one terminal atom on XZ plane, and the other on the +Z axis.
Stored in two orientations, with the 3rd (forward) or first (reversed)
atom on the +Z axis.  See <a class="reference internal" href="#Bio.PDB.internal_coords.Dihedron" title="Bio.PDB.internal_coords.Dihedron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dihedron</span></code></a> for use of forward and
reverse orientations.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>len12: float</strong></dt><dd><p>distance between first and second atoms</p>
</dd>
<dt><strong>len23: float</strong></dt><dd><p>distance between second and third atoms</p>
</dd>
<dt><strong>angle: float</strong></dt><dd><p>angle (degrees) formed by three atoms in hedron</p>
</dd>
<dt><strong>xrh_class: string</strong></dt><dd><p>only for hedron spanning 2 residues, will have ‘X’ for residue
contributing only one atom</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 37%" />
<col style="width: 63%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>get_length()</strong></p></td>
<td><p>get bond length for specified atom pair</p></td>
</tr>
<tr class="row-even"><td><p><strong>set_length()</strong></p></td>
<td><p>set bond length for specified atom pair</p></td>
</tr>
<tr class="row-odd"><td><p><strong>angle(), len12(), len23()</strong></p></td>
<td><p>setters for relevant attributes (angle in degrees)</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize Hedron with sequence of AtomKeys, kwargs.</p>
<dl class="simple">
<dt>Acceptable input:</dt><dd><p>As for Edron, plus optional ‘len12’, ‘angle’, ‘len23’
keyworded values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Print string for Hedron object.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron.angle">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">angle</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">float</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron.angle" title="Permalink to this definition"></a></dt>
<dd><p>Get this hedron angle.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron.len12">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">len12</span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron.len12" title="Permalink to this definition"></a></dt>
<dd><p>Get first length for Hedron.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron.len23">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">len23</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">float</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron.len23" title="Permalink to this definition"></a></dt>
<dd><p>Get second length for Hedron.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron.get_length">
<span class="sig-name descname"><span class="pre">get_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ak_tpl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron.get_length" title="Permalink to this definition"></a></dt>
<dd><p>Get bond length for specified atom pair.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ak_tpl</strong> (<em>tuple</em>) – tuple of AtomKeys.
Pair of atoms in this Hedron</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron.set_length">
<span class="sig-name descname"><span class="pre">set_length</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ak_tpl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newLength</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron.set_length" title="Permalink to this definition"></a></dt>
<dd><p>Set bond length for specified atom pair; sets needs_update.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ak_tpl</strong> (<em>tuple</em>) – tuple of AtomKeys
Pair of atoms in this Hedron</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Hedron.cic">
<span class="sig-name descname"><span class="pre">cic</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><span class="pre">Bio.PDB.internal_coords.IC_Chain</span></a></em><a class="headerlink" href="#Bio.PDB.internal_coords.Hedron.cic" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">Dihedron</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.PDB.internal_coords.Edron" title="Bio.PDB.internal_coords.Edron"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.PDB.internal_coords.Edron</span></code></a></p>
<p>Class to represent four joined atoms forming a dihedral angle.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>angle: float</strong></dt><dd><p>Measurement or specification of dihedral angle in degrees; prefer
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.bond_set" title="Bio.PDB.internal_coords.IC_Residue.bond_set"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Residue.bond_set()</span></code></a> to set</p>
</dd>
<dt><strong>hedron1, hedron2: Hedron object references</strong></dt><dd><p>The two hedra which form the dihedral angle</p>
</dd>
<dt><strong>h1key, h2key: tuples of AtomKeys</strong></dt><dd><p>Hash keys for hedron1 and hedron2</p>
</dd>
<dt><strong>id3,id32: tuples of AtomKeys</strong></dt><dd><p>First 3 and second 3 atoms comprising dihedron; hxkey orders may differ</p>
</dd>
<dt><strong>ric: IC_Residue object reference</strong></dt><dd><p><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue" title="Bio.PDB.internal_coords.IC_Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Residue</span></code></a> object containing this dihedral</p>
</dd>
<dt><strong>reverse: bool</strong></dt><dd><p>Indicates order of atoms in dihedron is reversed from order of atoms
in hedra</p>
</dd>
<dt><strong>primary: bool</strong></dt><dd><p>True if this is psi, phi, omega or a sidechain chi angle</p>
</dd>
<dt><strong>pclass: string (primary angle class)</strong></dt><dd><p>re_class with X for adjacent residue according to nomenclature
(psi, omega, phi)</p>
</dd>
<dt><strong>cst, rcst: numpy [4][4] arrays</strong></dt><dd><p>transformations to (cst) and from (rcst) Dihedron coordinate space
defined with atom 2 (Hedron 1 center atom) at the origin.  Views on
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain.dCoordSpace" title="Bio.PDB.internal_coords.IC_Chain.dCoordSpace"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Chain.dCoordSpace</span></code></a>.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 12%" />
<col style="width: 88%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>angle()</strong></p></td>
<td><p>getter/setter for dihdral angle in degrees; prefer <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.bond_set" title="Bio.PDB.internal_coords.IC_Residue.bond_set"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Residue.bond_set()</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p><strong>bits()</strong></p></td>
<td><p>return <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pic_flags" title="Bio.PDB.internal_coords.IC_Residue.pic_flags"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Residue.pic_flags</span></code></a> bitmask for dihedron psi, omega, etc</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Init Dihedron with sequence of AtomKeys and optional dihedral angle.</p>
<dl class="simple">
<dt>Acceptable input:</dt><dd><p>As for Edron, plus optional ‘dihedral’ keyworded angle value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Print string for Dihedron object.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.angle">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">angle</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">float</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.angle" title="Permalink to this definition"></a></dt>
<dd><p>Get dihedral angle.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.angle_dif">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">angle_dif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">numpy.ndarray</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.angle_dif" title="Permalink to this definition"></a></dt>
<dd><p>Get angle difference between two +/- 180 angles.</p>
<p><a class="reference external" href="https://stackoverflow.com/a/36001014/2783487">https://stackoverflow.com/a/36001014/2783487</a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.angle_avg">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">angle_avg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_rads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_rads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.angle_avg" title="Permalink to this definition"></a></dt>
<dd><p>Get average of list of +/-180 angles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alst</strong> (<em>List</em>) – list of angles to average</p></li>
<li><p><strong>in_rads</strong> (<em>bool</em>) – input values are in radians</p></li>
<li><p><strong>out_rads</strong> (<em>bool</em>) – report result in radians</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.angle_pop_sd">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">angle_pop_sd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alst</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.angle_pop_sd" title="Permalink to this definition"></a></dt>
<dd><p>Get population standard deviation for list of +/-180 angles.</p>
<p>should be sample std dev but avoid len(alst)=1 -&gt; div by 0</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.difference">
<span class="sig-name descname"><span class="pre">difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#Bio.PDB.internal_coords.Dihedron" title="Bio.PDB.internal_coords.Dihedron"><span class="pre">Bio.PDB.internal_coords.Dihedron</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.difference" title="Permalink to this definition"></a></dt>
<dd><p>Get angle difference between this and other +/- 180 angles.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.bits">
<span class="sig-name descname"><span class="pre">bits</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.bits" title="Permalink to this definition"></a></dt>
<dd><p>Get <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.pic_flags" title="Bio.PDB.internal_coords.IC_Residue.pic_flags"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Residue.pic_flags</span></code></a> bitmasks for self is psi, omg, phi, pomg, chiX.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.Dihedron.cic">
<span class="sig-name descname"><span class="pre">cic</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Chain" title="Bio.PDB.internal_coords.IC_Chain"><span class="pre">Bio.PDB.internal_coords.IC_Chain</span></a></em><a class="headerlink" href="#Bio.PDB.internal_coords.Dihedron.cic" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">AtomKey</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue" title="Bio.PDB.internal_coords.IC_Residue"><span class="pre">Bio.PDB.internal_coords.IC_Residue</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="Bio.PDB.Atom.html#Bio.PDB.Atom.Atom" title="Bio.PDB.Atom.Atom"><span class="pre">Bio.PDB.Atom.Atom</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class for dict keys to reference atom coordinates.</p>
<p>AtomKeys capture residue and disorder information together, and
provide a no-whitespace string key for .pic files.</p>
<p>Supports rich comparison and multiple ways to instantiate.</p>
<dl class="simple">
<dt>AtomKeys contain:</dt><dd><p>residue position (respos), insertion code (icode), 1 or 3 char residue
name (resname), atom name (atm), altloc (altloc), and occupancy (occ)</p>
</dd>
</dl>
<p>Use <a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey.fields" title="Bio.PDB.internal_coords.AtomKey.fields"><code class="xref py py-data docutils literal notranslate"><span class="pre">AtomKey.fields</span></code></a> to get the index to the component of interest by
name:</p>
<p>Get C-alpha atoms from IC_Chain atomArray and atomArrayIndex with
AtomKeys:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">atmNameNdx</span> <span class="o">=</span> <span class="n">internal_coords</span><span class="o">.</span><span class="n">AtomKey</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">atm</span>
<span class="n">CaSelection</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">akl</span><span class="p">[</span><span class="n">atmNameNdx</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span>
<span class="p">]</span>
<span class="n">AtomArrayCa</span> <span class="o">=</span> <span class="n">atomArray</span><span class="p">[</span><span class="n">CaSelection</span><span class="p">]</span>
</pre></div>
</div>
<p>Get all phenylalanine atoms in a chain:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">resNameNdx</span> <span class="o">=</span> <span class="n">internal_coords</span><span class="o">.</span><span class="n">AtomKey</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">resname</span>
<span class="n">PheSelection</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">atomArrayIndex</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">k</span><span class="o">.</span><span class="n">akl</span><span class="p">[</span><span class="n">resNameNdx</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;F&quot;</span>
<span class="p">]</span>
<span class="n">AtomArrayPhe</span> <span class="o">=</span> <span class="n">atomArray</span><span class="p">[</span><span class="n">PheSelection</span><span class="p">]</span>
</pre></div>
</div>
<p>‘resname’ will be the uppercase 1-letter amino acid code if one of the 20
standard residues, otherwise the supplied 3-letter code.  Supplied as input
or read from .rbase attribute of <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue" title="Bio.PDB.internal_coords.IC_Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Residue</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>akl: tuple</strong></dt><dd><p>All six fields of AtomKey</p>
</dd>
<dt><strong>fieldNames: tuple (Class Attribute)</strong></dt><dd><p>Mapping of key index positions to names</p>
</dd>
<dt><strong>fields: namedtuple (Class Attribute)</strong></dt><dd><p>Mapping of field names to index positions.</p>
</dd>
<dt><strong>id: str</strong></dt><dd><p>‘_’-joined AtomKey fields, excluding ‘None’ fields</p>
</dd>
<dt><strong>atom_re: compiled regex (Class Attribute)</strong></dt><dd><p>A compiled regular expression matching the string form of the key</p>
</dd>
<dt><strong>d2h: bool (Class Attribute) default False</strong></dt><dd><p>Convert D atoms to H on input if True; must also modify
<a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.accept_atoms" title="Bio.PDB.internal_coords.IC_Residue.accept_atoms"><code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Residue.accept_atoms</span></code></a></p>
</dd>
<dt><strong>missing: bool default False</strong></dt><dd><p>AtomKey __init__’d from string is probably missing, set this flag to
note the issue.  Set by <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue.rak" title="Bio.PDB.internal_coords.IC_Residue.rak"><code class="xref py py-meth docutils literal notranslate"><span class="pre">IC_Residue.rak()</span></code></a></p>
</dd>
<dt><strong>ric: IC_Residue default None</strong></dt><dd><p><em>If</em> initialised with IC_Residue, this references the IC_residue</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 21%" />
<col style="width: 79%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>altloc_match(other)</strong></p></td>
<td><p>Returns True if this AtomKey matches other AtomKey excluding altloc and occupancy fields</p></td>
</tr>
<tr class="row-even"><td><p><strong>is_backbone()</strong></p></td>
<td><p>Returns True if atom is N, CA, C, O or H</p></td>
</tr>
<tr class="row-odd"><td><p><strong>atm()</strong></p></td>
<td><p>Returns atom name, e.g. N, CA, CB, etc.</p></td>
</tr>
<tr class="row-even"><td><p><strong>cr_class()</strong></p></td>
<td><p>Returns covalent radii class e.g. Csb</p></td>
</tr>
</tbody>
</table>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.atom_re">
<span class="sig-name descname"><span class="pre">atom_re</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">re.compile('^(?P&lt;respos&gt;-?\\d+)(?P&lt;icode&gt;[A-Za-z])?_(?P&lt;resname&gt;[a-zA-Z]+)_(?P&lt;atm&gt;[A-Za-z0-9]+)(?:_(?P&lt;altloc&gt;\\w))?(?:_(?P&lt;occ&gt;-?\\d\\.\\d+?))?$')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.atom_re" title="Permalink to this definition"></a></dt>
<dd><p>Pre-compiled regular expression to match an AtomKey string.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.fieldNames">
<span class="sig-name descname"><span class="pre">fieldNames</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('respos',</span> <span class="pre">'icode',</span> <span class="pre">'resname',</span> <span class="pre">'atm',</span> <span class="pre">'altloc',</span> <span class="pre">'occ')</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.fieldNames" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.fields">
<span class="sig-name descname"><span class="pre">fields</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">_fieldsDef(respos=0,</span> <span class="pre">icode=1,</span> <span class="pre">resname=2,</span> <span class="pre">atm=3,</span> <span class="pre">altloc=4,</span> <span class="pre">occ=5)</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.fields" title="Permalink to this definition"></a></dt>
<dd><p>Use this namedtuple to access AtomKey fields.  See <a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><code class="xref py py-class docutils literal notranslate"><span class="pre">AtomKey</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.d2h">
<span class="sig-name descname"><span class="pre">d2h</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">False</span></em><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.d2h" title="Permalink to this definition"></a></dt>
<dd><p>Set True to convert D Deuterium to H Hydrogen on input.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue" title="Bio.PDB.internal_coords.IC_Residue"><span class="pre">Bio.PDB.internal_coords.IC_Residue</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="Bio.PDB.Atom.html#Bio.PDB.Atom.Atom" title="Bio.PDB.Atom.Atom"><span class="pre">Bio.PDB.Atom.Atom</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize AtomKey with residue and atom data.</p>
<p>Examples of acceptable input:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="o">&lt;</span><span class="n">IC_Residue</span><span class="o">&gt;</span><span class="p">,</span> <span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>    <span class="p">:</span> <span class="n">IC_Residue</span> <span class="k">with</span> <span class="n">atom</span> <span class="n">info</span>
<span class="p">(</span><span class="o">&lt;</span><span class="n">IC_Residue</span><span class="o">&gt;</span><span class="p">,</span> <span class="o">&lt;</span><span class="n">Atom</span><span class="o">&gt;</span><span class="p">)</span>       <span class="p">:</span> <span class="n">IC_Residue</span> <span class="k">with</span> <span class="n">Biopython</span> <span class="n">Atom</span>
<span class="p">([</span><span class="mi">52</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="o">...</span><span class="p">])</span>  <span class="p">:</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">ordered</span> <span class="n">data</span> <span class="n">fields</span>
<span class="p">(</span><span class="mi">52</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>    <span class="p">:</span> <span class="n">multiple</span> <span class="n">ordered</span> <span class="n">arguments</span>
<span class="p">({</span><span class="n">respos</span><span class="p">:</span> <span class="mi">52</span><span class="p">,</span> <span class="n">icode</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="n">atm</span><span class="p">:</span> <span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="o">...</span><span class="p">})</span> <span class="p">:</span> <span class="nb">dict</span> <span class="k">with</span> <span class="n">fieldNames</span>
<span class="p">(</span><span class="n">respos</span><span class="p">:</span> <span class="mi">52</span><span class="p">,</span> <span class="n">icode</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="n">atm</span><span class="p">:</span> <span class="s1">&#39;CA&#39;</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span> <span class="p">:</span> <span class="n">kwargs</span> <span class="k">with</span> <span class="n">fieldNames</span>
<span class="mi">52</span><span class="n">_G_CA</span><span class="p">,</span> <span class="mi">52</span><span class="n">B_G_CA</span><span class="p">,</span> <span class="mi">52</span><span class="n">_G_CA_0</span><span class="mf">.33</span><span class="p">,</span> <span class="mi">52</span><span class="n">_G_CA_B_0</span><span class="mf">.33</span>  <span class="p">:</span> <span class="nb">id</span> <span class="n">strings</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__deepcopy__">
<span class="sig-name descname"><span class="pre">__deepcopy__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">memo</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__deepcopy__" title="Permalink to this definition"></a></dt>
<dd><p>Deep copy implementation for AtomKey.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Repr string from id.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">int</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__hash__" title="Permalink to this definition"></a></dt>
<dd><p>Hash calculated at init from akl tuple.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.altloc_match">
<span class="sig-name descname"><span class="pre">altloc_match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#Bio.PDB.internal_coords.AtomKey" title="Bio.PDB.internal_coords.AtomKey"><span class="pre">Bio.PDB.internal_coords.AtomKey</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.altloc_match" title="Permalink to this definition"></a></dt>
<dd><p>Test AtomKey match to other discounting occupancy and altloc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.is_backbone">
<span class="sig-name descname"><span class="pre">is_backbone</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.is_backbone" title="Permalink to this definition"></a></dt>
<dd><p>Return True if is N, C, CA, O, or H.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.atm">
<span class="sig-name descname"><span class="pre">atm</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.atm" title="Permalink to this definition"></a></dt>
<dd><p>Return atom name : N, CA, CB, O etc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.cr_class">
<span class="sig-name descname"><span class="pre">cr_class</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.cr_class" title="Permalink to this definition"></a></dt>
<dd><p>Return covalent radii class for atom or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__ne__">
<span class="sig-name descname"><span class="pre">__ne__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__ne__" title="Permalink to this definition"></a></dt>
<dd><p>Test for inequality.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__eq__" title="Permalink to this definition"></a></dt>
<dd><p>Test for equality.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__gt__">
<span class="sig-name descname"><span class="pre">__gt__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__gt__" title="Permalink to this definition"></a></dt>
<dd><p>Test greater than.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__ge__">
<span class="sig-name descname"><span class="pre">__ge__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__ge__" title="Permalink to this definition"></a></dt>
<dd><p>Test greater or equal.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__lt__">
<span class="sig-name descname"><span class="pre">__lt__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__lt__" title="Permalink to this definition"></a></dt>
<dd><p>Test less than.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.AtomKey.__le__">
<span class="sig-name descname"><span class="pre">__le__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.AtomKey.__le__" title="Permalink to this definition"></a></dt>
<dd><p>Test less or equal.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.set_accuracy_95">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">set_accuracy_95</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#Bio.PDB.internal_coords.set_accuracy_95" title="Permalink to this definition"></a></dt>
<dd><p>Reduce floating point accuracy to 9.5 (xxxx.xxxxx).</p>
<p>Used by <a class="reference internal" href="#Bio.PDB.internal_coords.IC_Residue" title="Bio.PDB.internal_coords.IC_Residue"><code class="xref py py-class docutils literal notranslate"><span class="pre">IC_Residue</span></code></a> class writing PIC and SCAD
files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>num</strong> (<em>float</em>) – input number</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float with specified accuracy</p>
</dd>
</dl>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.HedronMatchError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">HedronMatchError</span></span><a class="headerlink" href="#Bio.PDB.internal_coords.HedronMatchError" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Cannot find hedron in residue for given key.</p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="Bio.PDB.internal_coords.MissingAtomError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.internal_coords.</span></span><span class="sig-name descname"><span class="pre">MissingAtomError</span></span><a class="headerlink" href="#Bio.PDB.internal_coords.MissingAtomError" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Missing atom coordinates for hedron or dihedron.</p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.PDB.ic_rebuild.html" class="btn btn-neutral float-left" title="Bio.PDB.ic_rebuild module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PDB.mmcifio.html" class="btn btn-neutral float-right" title="Bio.PDB.mmcifio module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>