

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.codonalign.codonseq module &mdash; Biopython 1.75.dev0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.motifs package" href="Bio.motifs.html" />
    <link rel="prev" title="Bio.codonalign.codonalphabet module" href="Bio.codonalign.codonalphabet.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html" class="icon icon-home"> Biopython
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.75.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.codonalign.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html#module-Bio.codonalign">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.codonalign.html">Bio.codonalign package</a> &raquo;</li>
        
      <li>Bio.codonalign.codonseq module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.motifs.html" class="btn btn-neutral float-right" title="Bio.motifs package" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.codonalign.codonalphabet.html" class="btn btn-neutral float-left" title="Bio.codonalign.codonalphabet module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.codonalign.codonseq">
<span id="bio-codonalign-codonseq-module"></span><h1>Bio.codonalign.codonseq module<a class="headerlink" href="#module-Bio.codonalign.codonseq" title="Permalink to this headline">¶</a></h1>
<p>Code for dealing with coding sequence.</p>
<p>CodonSeq class is inherited from Seq class. This is the core class to
deal with sequences in CodonAlignment in biopython.</p>
<dl class="class">
<dt id="Bio.codonalign.codonseq.CodonSeq">
<em class="property">class </em><code class="sig-prename descclassname">Bio.codonalign.codonseq.</code><code class="sig-name descname">CodonSeq</code><span class="sig-paren">(</span><em class="sig-param">data=''</em>, <em class="sig-param">alphabet=CodonAlphabet(Standard)</em>, <em class="sig-param">gap_char='-'</em>, <em class="sig-param">rf_table=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Seq.html#Bio.Seq.Seq" title="Bio.Seq.Seq"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code></a></p>
<p>CodonSeq is designed to be within the SeqRecords of a CodonAlignment class.</p>
<p>CodonSeq is useful as it allows the user to specify
reading frame when translate CodonSeq</p>
<p>CodonSeq also accepts codon style slice by calling
get_codon() method.</p>
<p><strong>Important:</strong> Ungapped CodonSeq can be any length if you
specify the rf_table. Gapped CodonSeq should be a
multiple of three.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">codonseq</span> <span class="o">=</span> <span class="n">CodonSeq</span><span class="p">(</span><span class="s2">&quot;AAATTTGGGCCAAATTT&quot;</span><span class="p">,</span> <span class="n">rf_table</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">14</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">codonseq</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
<span class="go">KFGAKF</span>
</pre></div>
</div>
<p>test get_full_rf_table method</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">CodonSeq</span><span class="p">(</span><span class="s1">&#39;AAATTTCCCGG-TGGGTTTAA&#39;</span><span class="p">,</span> <span class="n">rf_table</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">17</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">full_rf_table</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">get_full_rf_table</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">full_rf_table</span><span class="p">)</span>
<span class="go">[0, 3, 6, 9, 12, 15, 18]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">rf_table</span><span class="o">=</span><span class="n">full_rf_table</span><span class="p">,</span> <span class="n">ungap_seq</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="go">KFPPWV*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">CodonSeq</span><span class="p">(</span><span class="s1">&#39;AAATTTCCCGGGAA-TTTTAA&#39;</span><span class="p">,</span> <span class="n">rf_table</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">17</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">get_full_rf_table</span><span class="p">())</span>
<span class="go">[0, 3, 6, 9, 12.0, 15, 18]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">CodonSeq</span><span class="p">(</span><span class="s1">&#39;AAA------------TAA&#39;</span><span class="p">,</span> <span class="n">rf_table</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">get_full_rf_table</span><span class="p">())</span>
<span class="go">[0, 3.0, 6.0, 9.0, 12.0, 15]</span>
</pre></div>
</div>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">back_transcribe</span></code>(self)</p></td>
<td><p>Return the DNA sequence from an RNA sequence by creating a new Seq object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complement</span></code>(self)</p></td>
<td><p>Return the complement sequence by creating a new Seq object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code>(self, sub[, start, end])</p></td>
<td><p>Return a non-overlapping count, like that of a python string.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_overlap</span></code>(self, sub[, start, end])</p></td>
<td><p>Return an overlapping count.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">endswith</span></code>(self, suffix[, start, end])</p></td>
<td><p>Return True if the Seq ends with the given suffix, False otherwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">find</span></code>(self, sub[, start, end])</p></td>
<td><p>Find method, like that of a python string.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.codonalign.codonseq.CodonSeq.from_seq" title="Bio.codonalign.codonseq.CodonSeq.from_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_seq</span></code></a>(seq[, alphabet, rf_table])</p></td>
<td><p>Get codon sequence from sequence data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.codonalign.codonseq.CodonSeq.full_translate" title="Bio.codonalign.codonseq.CodonSeq.full_translate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">full_translate</span></code></a>(self[, codon_table, stop_symbol])</p></td>
<td><p>Apply full translation with gaps considered.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.codonalign.codonseq.CodonSeq.get_codon" title="Bio.codonalign.codonseq.CodonSeq.get_codon"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_codon</span></code></a>(self, index)</p></td>
<td><p>Get the index codon from the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.codonalign.codonseq.CodonSeq.get_codon_num" title="Bio.codonalign.codonseq.CodonSeq.get_codon_num"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_codon_num</span></code></a>(self)</p></td>
<td><p>Return the number of codons in the CodonSeq.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.codonalign.codonseq.CodonSeq.get_full_rf_table" title="Bio.codonalign.codonseq.CodonSeq.get_full_rf_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_full_rf_table</span></code></a>(self)</p></td>
<td><p>Return full rf_table of the CodonSeq records.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join</span></code>(self, other)</p></td>
<td><p>Return a merge of the sequences in other, spaced by the sequence from self.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">lower</span></code>(self)</p></td>
<td><p>Return a lower case copy of the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">lstrip</span></code>(self[, chars])</p></td>
<td><p>Return a new Seq object with leading (left) end stripped.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse_complement</span></code>(self)</p></td>
<td><p>Return the reverse complement sequence by creating a new Seq object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rfind</span></code>(self, sub[, start, end])</p></td>
<td><p>Find from right method, like that of a python string.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rsplit</span></code>(self[, sep, maxsplit])</p></td>
<td><p>Do a right split method, like that of a python string.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rstrip</span></code>(self[, chars])</p></td>
<td><p>Return a new Seq object with trailing (right) end stripped.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">split</span></code>(self[, sep, maxsplit])</p></td>
<td><p>Split method, like that of a python string.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">startswith</span></code>(self, prefix[, start, end])</p></td>
<td><p>Return True if the Seq starts with the given prefix, False otherwise.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>(self[, chars])</p></td>
<td><p>Return a new Seq object with leading and trailing ends stripped.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.codonalign.codonseq.CodonSeq.toSeq" title="Bio.codonalign.codonseq.CodonSeq.toSeq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">toSeq</span></code></a>(self[, alphabet])</p></td>
<td><p>Convert DNA to seq object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">tomutable</span></code>(self)</p></td>
<td><p>Return the full sequence as a MutableSeq object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">transcribe</span></code>(self)</p></td>
<td><p>Return the RNA sequence from a DNA sequence by creating a new Seq object.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.codonalign.codonseq.CodonSeq.translate" title="Bio.codonalign.codonseq.CodonSeq.translate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate</span></code></a>(self[, codon_table, stop_symbol, …])</p></td>
<td><p>Translate the CodonSeq based on the reading frame in rf_table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.codonalign.codonseq.CodonSeq.ungap" title="Bio.codonalign.codonseq.CodonSeq.ungap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ungap</span></code></a>(self[, gap])</p></td>
<td><p>Return a copy of the sequence without the gap character(s).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">upper</span></code>(self)</p></td>
<td><p>Return an upper case copy of the sequence.</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.codonalign.codonseq.CodonSeq.from_seq">
<em class="property">classmethod </em><code class="sig-name descname">from_seq</code><span class="sig-paren">(</span><em class="sig-param">seq</em>, <em class="sig-param">alphabet=CodonAlphabet(Standard)</em>, <em class="sig-param">rf_table=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.from_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Get codon sequence from sequence data.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.codonalign.codonseq.CodonSeq.full_translate">
<code class="sig-name descname">full_translate</code><span class="sig-paren">(</span><em class="sig-param">self, codon_table=NCBICodonTable(id=1, names=['Standard', 'SGC0'], ...), stop_symbol='*'</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.full_translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply full translation with gaps considered.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.codonalign.codonseq.CodonSeq.get_codon">
<code class="sig-name descname">get_codon</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">index</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.get_codon" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the index codon from the sequence.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.codonalign.codonseq.CodonSeq.get_codon_num">
<code class="sig-name descname">get_codon_num</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.get_codon_num" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the number of codons in the CodonSeq.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.codonalign.codonseq.CodonSeq.get_full_rf_table">
<code class="sig-name descname">get_full_rf_table</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.get_full_rf_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Return full rf_table of the CodonSeq records.</p>
<p>A full rf_table is different from a normal rf_table in that
it translate gaps in CodonSeq. It is helpful to construct
alignment containing frameshift.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.codonalign.codonseq.CodonSeq.toSeq">
<code class="sig-name descname">toSeq</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">alphabet=DNAAlphabet()</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.toSeq" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert DNA to seq object.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.codonalign.codonseq.CodonSeq.translate">
<code class="sig-name descname">translate</code><span class="sig-paren">(</span><em class="sig-param">self, codon_table=NCBICodonTable(id=1, names=['Standard', 'SGC0'], ...), stop_symbol='*', rf_table=None, ungap_seq=True</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate the CodonSeq based on the reading frame in rf_table.</p>
<p>It is possible for the user to specify
a rf_table at this point. If you want to include
gaps in the translated sequence, this is the only
way. ungap_seq should be set to true for this
purpose.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.codonalign.codonseq.CodonSeq.ungap">
<code class="sig-name descname">ungap</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">gap=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.ungap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy of the sequence without the gap character(s).</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="Bio.codonalign.codonseq.cal_dn_ds">
<code class="sig-prename descclassname">Bio.codonalign.codonseq.</code><code class="sig-name descname">cal_dn_ds</code><span class="sig-paren">(</span><em class="sig-param">codon_seq1, codon_seq2, method='NG86', codon_table=NCBICodonTable(id=1, names=['Standard', 'SGC0'], ...), k=1, cfreq=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.cal_dn_ds" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate dN and dS of the given two sequences.</p>
<dl class="simple">
<dt>Available methods:</dt><dd><ul class="simple">
<li><p>NG86  - <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/3444411">Nei and Gojobori (1986)</a> (PMID 3444411).</p></li>
<li><p>LWL85 - <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/3916709">Li et al. (1985)</a> (PMID 3916709).</p></li>
<li><p>ML    - <a class="reference external" href="http://mbe.oxfordjournals.org/content/11/5/725">Goldman and Yang (1994)</a> (PMID 7968486).</p></li>
<li><p>YN00  - <a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a026236">Yang and Nielsen (2000)</a> (PMID 10666704).</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>codon_seq1 - CodonSeq or or SeqRecord that contains a CodonSeq</p></li>
<li><p>codon_seq2 - CodonSeq or or SeqRecord that contains a CodonSeq</p></li>
<li><p>w  - transition/transversion ratio</p></li>
<li><p>cfreq - Current codon frequency vector can only be specified
when you are using ML method. Possible ways of
getting cfreq are: F1x4, F3x4 and F61.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.motifs.html" class="btn btn-neutral float-right" title="Bio.motifs package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.codonalign.codonalphabet.html" class="btn btn-neutral float-left" title="Bio.codonalign.codonalphabet module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2017, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>