

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Sequencing.Ace module &mdash; Biopython 1.76.dev0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Sequencing.Phd module" href="Bio.Sequencing.Phd.html" />
    <link rel="prev" title="Bio.Sequencing.Applications package" href="Bio.Sequencing.Applications.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.76.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html#subpackages">Subpackages</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Sequencing.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html#module-Bio.Sequencing">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.Sequencing.html">Bio.Sequencing package</a> &raquo;</li>
        
      <li>Bio.Sequencing.Ace module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Sequencing/Ace.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Sequencing.Phd.html" class="btn btn-neutral float-right" title="Bio.Sequencing.Phd module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Sequencing.Applications.html" class="btn btn-neutral float-left" title="Bio.Sequencing.Applications package" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.Sequencing.Ace">
<span id="bio-sequencing-ace-module"></span><h1>Bio.Sequencing.Ace module<a class="headerlink" href="#module-Bio.Sequencing.Ace" title="Permalink to this headline">¶</a></h1>
<p>Parser for ACE files output by PHRAP.</p>
<p>Written by Frank Kauff (<a class="reference external" href="mailto:fkauff&#37;&#52;&#48;duke&#46;edu">fkauff<span>&#64;</span>duke<span>&#46;</span>edu</a>) and
Cymon J. Cox (<a class="reference external" href="mailto:cymon&#37;&#52;&#48;duke&#46;edu">cymon<span>&#64;</span>duke<span>&#46;</span>edu</a>)</p>
<p>Usage:</p>
<p>There are two ways of reading an ace file:</p>
<ol class="arabic simple">
<li><p>The function ‘read’ reads the whole file at once;</p></li>
<li><p>The function ‘parse’ reads the file contig after contig.</p></li>
</ol>
<p>First option, parse whole ace file at once:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Sequencing</span> <span class="k">import</span> <span class="n">Ace</span>
<span class="n">acefilerecord</span> <span class="o">=</span> <span class="n">Ace</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;my_ace_file.ace&#39;</span><span class="p">))</span>
</pre></div>
</div>
<dl class="simple">
<dt>This gives you:</dt><dd><ul class="simple">
<li><p>acefilerecord.ncontigs (the number of contigs in the ace file)</p></li>
<li><p>acefilerecord.nreads (the number of reads in the ace file)</p></li>
<li><p>acefilerecord.contigs[] (one instance of the Contig class for each contig)</p></li>
</ul>
</dd>
</dl>
<p>The Contig class holds the info of the CO tag, CT and WA tags, and all the reads used
for this contig in a list of instances of the Read class, e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">contig3</span> <span class="o">=</span> <span class="n">acefilerecord</span><span class="o">.</span><span class="n">contigs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="n">read4</span> <span class="o">=</span> <span class="n">contig3</span><span class="o">.</span><span class="n">reads</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">RD_of_read4</span> <span class="o">=</span> <span class="n">read4</span><span class="o">.</span><span class="n">rd</span>
<span class="n">DS_of_read4</span> <span class="o">=</span> <span class="n">read4</span><span class="o">.</span><span class="n">ds</span>
</pre></div>
</div>
<p>CT, WA, RT tags from the end of the file can appear anywhere are automatically
sorted into the right place.</p>
<p>see _RecordConsumer for details.</p>
<p>The second option is to  iterate over the contigs of an ace file one by one
in the ususal way:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Sequencing</span> <span class="k">import</span> <span class="n">Ace</span>
<span class="n">contigs</span> <span class="o">=</span> <span class="n">Ace</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;my_ace_file.ace&#39;</span><span class="p">))</span>
<span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="n">contigs</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">contig</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="o">...</span>
</pre></div>
</div>
<p>Please note that for memory efficiency, when using the iterator approach, only one
contig is kept in memory at once.  However, there can be a footer to the ACE file
containing WA, CT, RT or WR tags which contain additional meta-data on the contigs.
Because the parser doesn’t see this data until the final record, it cannot be added to
the appropriate records.  Instead these tags will be returned with the last contig record.
Thus an ace file does not entirerly suit the concept of iterating. If WA, CT, RT, WR tags
are needed, the ‘read’ function rather than the ‘parse’ function might be more appropriate.</p>
<dl class="class">
<dt id="Bio.Sequencing.Ace.ACEFileRecord">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">ACEFileRecord</code><a class="headerlink" href="#Bio.Sequencing.Ace.ACEFileRecord" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Holds data of an ACE file.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.Sequencing.Ace.ACEFileRecord.sort" title="Bio.Sequencing.Ace.ACEFileRecord.sort"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code></a>(self)</p></td>
<td><p>Sorts wr, rt and ct tags into the appropriate contig / read instance, if possible.</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.Sequencing.Ace.ACEFileRecord.sort">
<code class="sig-name descname">sort</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.ACEFileRecord.sort" title="Permalink to this definition">¶</a></dt>
<dd><p>Sorts wr, rt and ct tags into the appropriate contig / read instance, if possible.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.Contig">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">Contig</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.Contig" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Holds information about a contig from an ACE record.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.Reads">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">Reads</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.Reads" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Holds information about a read supporting an ACE contig.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.af">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">af</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.af" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>AF lines, define the location of the read within the contig.</p>
<p>Note attribute coru is short for complemented (C) or uncomplemented (U),
since the strand information is stored in an ACE file using either the
C or U character.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.bs">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">bs</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.bs" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>BS (base segment), which read was chosen as the consensus at each position.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.ct">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">ct</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.ct" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>CT (consensus tags).</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.ds">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">ds</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.ds" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>DS lines, include file name of a read’s chromatogram file.</p>
</dd></dl>

<dl class="function">
<dt id="Bio.Sequencing.Ace.parse">
<code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">parse</code><span class="sig-paren">(</span><em class="sig-param">handle</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate of ACE file contig by contig.</p>
<p>Argument handle is a file-like object.</p>
<p>This function returns an iterator that allows you to iterate
over the ACE file record by record:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">records</span> <span class="o">=</span> <span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
    <span class="c1"># do something with the record</span>
</pre></div>
</div>
<p>where each record is a Contig object.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.qa">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">qa</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.qa" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>QA (read quality), including which part if any was used as the consensus.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.rd">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">rd</code><a class="headerlink" href="#Bio.Sequencing.Ace.rd" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>RD (reads), store a read with its name, sequence etc.</p>
<p>The location and strand each read is mapped to is held in the AF lines.</p>
</dd></dl>

<dl class="function">
<dt id="Bio.Sequencing.Ace.read">
<code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">read</code><span class="sig-paren">(</span><em class="sig-param">handle</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.read" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a full ACE file into a list of contigs.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.rt">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">rt</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.rt" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>RT (transient read tags), generated by crossmatch and phrap.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.wa">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">wa</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.wa" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>WA (whole assembly tag), holds the assembly program name, version, etc.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.Sequencing.Ace.wr">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Sequencing.Ace.</code><code class="sig-name descname">wr</code><span class="sig-paren">(</span><em class="sig-param">line=None</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Sequencing.Ace.wr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>WR lines.</p>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.Sequencing.Phd.html" class="btn btn-neutral float-right" title="Bio.Sequencing.Phd module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Sequencing.Applications.html" class="btn btn-neutral float-left" title="Bio.Sequencing.Applications package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2017, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>