<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.SeqUtils package &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.SeqUtils.CheckSum module" href="Bio.SeqUtils.CheckSum.html" />
    <link rel="prev" title="Bio.SeqIO.XdnaIO module" href="Bio.SeqIO.XdnaIO.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
      <li class="breadcrumb-item active">Bio.SeqUtils package</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SeqUtils/__init__.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.SeqIO.XdnaIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.XdnaIO module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqUtils.CheckSum.html" class="btn btn-neutral float-right" title="Bio.SeqUtils.CheckSum module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bio-sequtils-package">
<h1>Bio.SeqUtils package<a class="headerlink" href="#bio-sequtils-package" title="Permalink to this headline"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.SeqUtils.CheckSum.html">Bio.SeqUtils.CheckSum module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SeqUtils.IsoelectricPoint.html">Bio.SeqUtils.IsoelectricPoint module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SeqUtils.MeltingTemp.html">Bio.SeqUtils.MeltingTemp module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SeqUtils.ProtParam.html">Bio.SeqUtils.ProtParam module</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Bio.SeqUtils.ProtParam.html#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SeqUtils.ProtParamData.html">Bio.SeqUtils.ProtParamData module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SeqUtils.lcc.html">Bio.SeqUtils.lcc module</a></li>
</ul>
</div>
</section>
<section id="module-Bio.SeqUtils">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.SeqUtils" title="Permalink to this headline"></a></h2>
<p>Miscellaneous functions for dealing with sequences.</p>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.gc_fraction">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">gc_fraction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'remove'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.gc_fraction" title="Permalink to this definition"></a></dt>
<dd><p>Calculate G+C percentage in seq (float between 0 and 1).</p>
<p>Copes with mixed case sequences. Ambiguous Nucleotides in this context are
those different from ATCGSW (S is G or C, and W is A or T).</p>
<p>If ambiguous equals “remove” (default), will only count GCS and will only
include ACTGSW when calculating the sequence length. Equivalent to removing
all characters in the set BDHKMNRVXY before calculating the GC content, as
each of these ambiguous nucleotides can either be in (A,T) or in (C,G).</p>
<p>If ambiguous equals “ignore”, it will treat only unambiguous nucleotides (GCS)
as counting towards the GC percentage, but will include all ambiguous and
unambiguous nucleotides when calculating the sequence length.</p>
<p>If ambiguous equals “weighted”, will use a “mean” value when counting the
ambiguous characters, for example, G and C will be counted as 1, N and X will
be counted as 0.5, D will be counted as 0.33 etc. See Bio.SeqUtils._gc_values
for a full list.</p>
<p>Will raise a ValueError for any other value of the ambiguous parameter.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">gc_fraction</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;ACTG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content of </span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="s2"> : </span><span class="si">{</span><span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content of ACTG : 0.50</span>
</pre></div>
</div>
<p>S and W are ambiguous for the purposes of calculating the GC content.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;ACTGSSSS&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;remove&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content of </span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="s2"> : </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content of ACTGSSSS : 0.75</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content of </span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="s2"> : </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content of ACTGSSSS : 0.75</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;weighted&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content with ambiguous counting: </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content with ambiguous counting: 0.75</span>
</pre></div>
</div>
<p>Some examples with ambiguous nucleotides.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;ACTGN&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content of </span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="s2"> : </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content of ACTGN : 0.40</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;weighted&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content with ambiguous counting: </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content with ambiguous counting: 0.50</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;remove&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content with ambiguous removing: </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content with ambiguous removing: 0.50</span>
</pre></div>
</div>
<p>Ambiguous nucleotides are also removed from the length of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;GDVV&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content of </span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="s2"> : </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content of GDVV : 0.25</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;weighted&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content with ambiguous counting: </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content with ambiguous counting: 0.6667</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc</span> <span class="o">=</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;remove&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GC content with ambiguous removing: </span><span class="si">{</span><span class="n">gc</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">GC content with ambiguous removing: 1.00</span>
</pre></div>
</div>
<p>Note that this will return zero for an empty sequence.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.GC123">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">GC123</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.GC123" title="Permalink to this definition"></a></dt>
<dd><p>Calculate G+C content: total, for first, second and third positions.</p>
<p>Returns a tuple of four floats (percentages between 0 and 100) for the
entire sequence, and the three codon positions.  e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">GC123</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">GC123</span><span class="p">(</span><span class="s2">&quot;ACTGTN&quot;</span><span class="p">)</span>
<span class="go">(40.0, 50.0, 50.0, 0.0)</span>
</pre></div>
</div>
<p>Copes with mixed case sequences, but does NOT deal with ambiguous
nucleotides.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.GC_skew">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">GC_skew</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.GC_skew" title="Permalink to this definition"></a></dt>
<dd><p>Calculate GC skew (G-C)/(G+C) for multiple windows along the sequence.</p>
<p>Returns a list of ratios (floats), controlled by the length of the sequence
and the size of the window.</p>
<p>Returns 0 for windows without any G/C by handling zero division errors.</p>
<p>Does NOT look at any ambiguous nucleotides.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.xGC_skew">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">xGC_skew</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zoom</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">px</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">py</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.xGC_skew" title="Permalink to this definition"></a></dt>
<dd><p>Calculate and plot normal and accumulated GC skew (GRAPHICS !!!).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.nt_search">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">nt_search</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subseq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.nt_search" title="Permalink to this definition"></a></dt>
<dd><p>Search for a DNA subseq in seq, return list of [subseq, positions].</p>
<p>Use ambiguous values (like N = A or T or C or G, R = A or G etc.),
searches only on forward strand.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.seq3">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">seq3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">custom_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">undef_code</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Xaa'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.seq3" title="Permalink to this definition"></a></dt>
<dd><p>Convert protein sequence from one-letter to three-letter code.</p>
<p>The single required input argument ‘seq’ should be a protein sequence using
single letter codes, either as a Python string or as a Seq or MutableSeq
object.</p>
<p>This function returns the amino acid sequence as a string using the three
letter amino acid codes. Output follows the IUPAC standard (including
ambiguous characters B for “Asx”, J for “Xle” and X for “Xaa”, and also U
for “Sel” and O for “Pyl”) plus “Ter” for a terminator given as an
asterisk. Any unknown character (including possible gap characters),
is changed into ‘Xaa’ by default.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">seq3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq3</span><span class="p">(</span><span class="s2">&quot;MAIVMGRWKGAR*&quot;</span><span class="p">)</span>
<span class="go">&#39;MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer&#39;</span>
</pre></div>
</div>
<p>You can set a custom translation of the codon termination code using the
dictionary “custom_map” argument (which defaults to {‘*’: ‘Ter’}), e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq3</span><span class="p">(</span><span class="s2">&quot;MAIVMGRWKGAR*&quot;</span><span class="p">,</span> <span class="n">custom_map</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;*&quot;</span><span class="p">:</span> <span class="s2">&quot;***&quot;</span><span class="p">})</span>
<span class="go">&#39;MetAlaIleValMetGlyArgTrpLysGlyAlaArg***&#39;</span>
</pre></div>
</div>
<p>You can also set a custom translation for non-amino acid characters, such
as ‘-’, using the “undef_code” argument, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq3</span><span class="p">(</span><span class="s2">&quot;MAIVMGRWKGA--R*&quot;</span><span class="p">,</span> <span class="n">undef_code</span><span class="o">=</span><span class="s1">&#39;---&#39;</span><span class="p">)</span>
<span class="go">&#39;MetAlaIleValMetGlyArgTrpLysGlyAla------ArgTer&#39;</span>
</pre></div>
</div>
<p>If not given, “undef_code” defaults to “Xaa”, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq3</span><span class="p">(</span><span class="s2">&quot;MAIVMGRWKGA--R*&quot;</span><span class="p">)</span>
<span class="go">&#39;MetAlaIleValMetGlyArgTrpLysGlyAlaXaaXaaArgTer&#39;</span>
</pre></div>
</div>
<p>This function was inspired by BioPerl’s seq3.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.seq1">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">seq1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">custom_map</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">undef_code</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.seq1" title="Permalink to this definition"></a></dt>
<dd><p>Convert protein sequence from three-letter to one-letter code.</p>
<p>The single required input argument ‘seq’ should be a protein sequence
using three-letter codes, either as a Python string or as a Seq or
MutableSeq object.</p>
<p>This function returns the amino acid sequence as a string using the one
letter amino acid codes. Output follows the IUPAC standard (including
ambiguous characters “B” for “Asx”, “J” for “Xle”, “X” for “Xaa”, “U” for
“Sel”, and “O” for “Pyl”) plus “*” for a terminator given the “Ter” code.
Any unknown character (including possible gap characters), is changed
into ‘-’ by default.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">seq1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="p">(</span><span class="s2">&quot;MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer&quot;</span><span class="p">)</span>
<span class="go">&#39;MAIVMGRWKGAR*&#39;</span>
</pre></div>
</div>
<p>The input is case insensitive, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">seq1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="p">(</span><span class="s2">&quot;METalaIlEValMetGLYArgtRplysGlyAlaARGTer&quot;</span><span class="p">)</span>
<span class="go">&#39;MAIVMGRWKGAR*&#39;</span>
</pre></div>
</div>
<p>You can set a custom translation of the codon termination code using the
dictionary “custom_map” argument (defaulting to {‘Ter’: ‘*’}), e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="p">(</span><span class="s2">&quot;MetAlaIleValMetGlyArgTrpLysGlyAla***&quot;</span><span class="p">,</span> <span class="n">custom_map</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;***&quot;</span><span class="p">:</span> <span class="s2">&quot;*&quot;</span><span class="p">})</span>
<span class="go">&#39;MAIVMGRWKGA*&#39;</span>
</pre></div>
</div>
<p>You can also set a custom translation for non-amino acid characters, such
as ‘-’, using the “undef_code” argument, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="p">(</span><span class="s2">&quot;MetAlaIleValMetGlyArgTrpLysGlyAla------ArgTer&quot;</span><span class="p">,</span> <span class="n">undef_code</span><span class="o">=</span><span class="s1">&#39;?&#39;</span><span class="p">)</span>
<span class="go">&#39;MAIVMGRWKGA??R*&#39;</span>
</pre></div>
</div>
<p>If not given, “undef_code” defaults to “X”, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span><span class="p">(</span><span class="s2">&quot;MetAlaIleValMetGlyArgTrpLysGlyAla------ArgTer&quot;</span><span class="p">)</span>
<span class="go">&#39;MAIVMGRWKGAXXR*&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.molecular_weight">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">molecular_weight</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DNA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">double_stranded</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">circular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">monoisotopic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.molecular_weight" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the molecular mass of DNA, RNA or protein sequences as float.</p>
<p>Only unambiguous letters are allowed. Nucleotide sequences are assumed to
have a 5’ phosphate.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>seq: string, Seq, or SeqRecord object.</p></li>
<li><p>seq_type: The default is to assume DNA; override this with a string
“DNA”, “RNA”, or “protein”.</p></li>
<li><p>double_stranded: Calculate the mass for the double stranded molecule?</p></li>
<li><p>circular: Is the molecule circular (has no ends)?</p></li>
<li><p>monoisotopic: Use the monoisotopic mass tables?</p></li>
</ul>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">molecular_weight</span><span class="p">(</span><span class="s2">&quot;AGC&quot;</span><span class="p">))</span>
<span class="go">949.61</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">molecular_weight</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AGC&quot;</span><span class="p">)))</span>
<span class="go">949.61</span>
</pre></div>
</div>
<p>However, it is better to be explicit - for example with strings:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">molecular_weight</span><span class="p">(</span><span class="s2">&quot;AGC&quot;</span><span class="p">,</span> <span class="s2">&quot;DNA&quot;</span><span class="p">))</span>
<span class="go">949.61</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">molecular_weight</span><span class="p">(</span><span class="s2">&quot;AGC&quot;</span><span class="p">,</span> <span class="s2">&quot;RNA&quot;</span><span class="p">))</span>
<span class="go">997.61</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">molecular_weight</span><span class="p">(</span><span class="s2">&quot;AGC&quot;</span><span class="p">,</span> <span class="s2">&quot;protein&quot;</span><span class="p">))</span>
<span class="go">249.29</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SeqUtils.six_frame_translations">
<span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">six_frame_translations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetic_code</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.six_frame_translations" title="Permalink to this definition"></a></dt>
<dd><p>Return pretty string showing the 6 frame translations and GC content.</p>
<p>Nice looking 6 frame translation with GC content - code from xbbtools
similar to DNA Striders six-frame translation</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">six_frame_translations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">six_frame_translations</span><span class="p">(</span><span class="s2">&quot;AUGGCCAUUGUAAUGGGCCGCUGA&quot;</span><span class="p">))</span>
<span class="go">GC_Frame: a:5 t:0 g:8 c:5</span>
<span class="go">Sequence: auggccauug ... gggccgcuga, 24 nt, 54.17 %GC</span>


<span class="go">1/1</span>
<span class="go">  G  H  C  N  G  P  L</span>
<span class="go"> W  P  L  *  W  A  A</span>
<span class="go">M  A  I  V  M  G  R  *</span>
<span class="go">auggccauuguaaugggccgcuga   54 %</span>
<span class="go">uaccgguaacauuacccggcgacu</span>
<span class="go">A  M  T  I  P  R  Q</span>
<span class="go"> H  G  N  Y  H  A  A  S</span>
<span class="go">  P  W  Q  L  P  G  S</span>

</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqUtils.CodonAdaptationIndex">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.</span></span><span class="sig-name descname"><span class="pre">CodonAdaptationIndex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">standard_dna_table</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.CodonAdaptationIndex" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
<p>A codon adaptation index (CAI) implementation.</p>
<p>Implements the codon adaptation index (CAI) described by Sharp and
Li (Nucleic Acids Res. 1987 Feb 11;15(3):1281-95).</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.CodonAdaptationIndex.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">standard_dna_table</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.CodonAdaptationIndex.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Generate a codon adaptiveness table from the coding DNA sequences.</p>
<p>This calculates the relative adaptiveness of each codon (w_ij) as
defined by Sharp &amp; Li (Nucleic Acids Research 15(3): 1281-1295 (1987))
from the provided codon DNA sequences.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>sequences: An iterable over DNA sequences, which may be plain</dt><dd><p>strings, Seq objects, MutableSeq objects, or SeqRecord
objects.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>table:     A Bio.Data.CodonTable.CodonTable object defining the</dt><dd><p>genetic code. By default, the standard genetic code is
used.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.CodonAdaptationIndex.calculate">
<span class="sig-name descname"><span class="pre">calculate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.CodonAdaptationIndex.calculate" title="Permalink to this definition"></a></dt>
<dd><p>Calculate and return the CAI (float) for the provided DNA sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.CodonAdaptationIndex.optimize">
<span class="sig-name descname"><span class="pre">optimize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DNA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.CodonAdaptationIndex.optimize" title="Permalink to this definition"></a></dt>
<dd><p>Return a new DNA sequence with preferred codons only.</p>
<p>Uses the codon adaptiveness table defined by the CodonAdaptationIndex
object to generate DNA sequences with only preferred codons.
May be useful when designing DNA sequences for transgenic protein
expression or codon-optimized proteins like fluorophores.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>sequence: DNA, RNA, or protein sequence to codon-optimize.</dt><dd><p>Supplied as a str, Seq, or SeqRecord object.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>seq_type: String specifying type of sequence provided.</dt><dd><p>Options are “DNA”, “RNA”, and “protein”. Default is “DNA”.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>strict:   Determines whether an exception should be raised when</dt><dd><p>two codons are equally prefered for a given amino acid.</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>Returns:</dt><dd><p>Seq object with DNA encoding the same protein as the sequence argument,
but using only preferred codons as defined by the codon adaptation index.
If multiple codons are equally preferred, a warning is issued
and one codon is chosen for use in the optimzed sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.CodonAdaptationIndex.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.CodonAdaptationIndex.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Return str(self).</p>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.SeqIO.XdnaIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.XdnaIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqUtils.CheckSum.html" class="btn btn-neutral float-right" title="Bio.SeqUtils.CheckSum module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>