<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.Align package &mdash; Biopython 1.82.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Align.Applications package" href="Bio.Align.Applications.html" />
    <link rel="prev" title="Bio.Affy.CelFile module" href="Bio.Affy.CelFile.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="index.html">
            
              <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.Align package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-Bio.Align">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
      <li class="breadcrumb-item active">Bio.Align package</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Align/__init__.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Affy.CelFile.html" class="btn btn-neutral float-left" title="Bio.Affy.CelFile module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Align.Applications.html" class="btn btn-neutral float-right" title="Bio.Align.Applications package" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bio-align-package">
<h1>Bio.Align package<a class="headerlink" href="#bio-align-package" title="Permalink to this headline"></a></h1>
<section id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.Applications.html">Bio.Align.Applications package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Bio.Align.Applications.html#module-Bio.Align.Applications">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.substitution_matrices.html">Bio.Align.substitution_matrices package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Bio.Align.substitution_matrices.html#module-Bio.Align.substitution_matrices">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</section>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.AlignInfo.html">Bio.Align.AlignInfo module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.a2m.html">Bio.Align.a2m module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.bed.html">Bio.Align.bed module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.bigbed.html">Bio.Align.bigbed module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.bigmaf.html">Bio.Align.bigmaf module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.bigpsl.html">Bio.Align.bigpsl module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.clustal.html">Bio.Align.clustal module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.emboss.html">Bio.Align.emboss module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.exonerate.html">Bio.Align.exonerate module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.fasta.html">Bio.Align.fasta module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.hhr.html">Bio.Align.hhr module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.interfaces.html">Bio.Align.interfaces module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.maf.html">Bio.Align.maf module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.mauve.html">Bio.Align.mauve module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.msf.html">Bio.Align.msf module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.nexus.html">Bio.Align.nexus module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.phylip.html">Bio.Align.phylip module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.psl.html">Bio.Align.psl module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.sam.html">Bio.Align.sam module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.stockholm.html">Bio.Align.stockholm module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Align.tabular.html">Bio.Align.tabular module</a></li>
</ul>
</div>
</section>
<section id="module-Bio.Align">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.Align" title="Permalink to this headline"></a></h2>
<p>Code for dealing with sequence alignments.</p>
<p>One of the most important things in this module is the MultipleSeqAlignment
class, used in the Bio.AlignIO module.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.AlignmentCounts">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">AlignmentCounts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gaps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identities</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatches</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.AlignmentCounts" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.AlignmentCounts.__getnewargs__">
<span class="sig-name descname"><span class="pre">__getnewargs__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.AlignmentCounts.__getnewargs__" title="Permalink to this definition"></a></dt>
<dd><p>Return self as a plain tuple.  Used by copy and pickle.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.AlignmentCounts.__new__">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">__new__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">_cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gaps</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identities</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mismatches</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.AlignmentCounts.__new__" title="Permalink to this definition"></a></dt>
<dd><p>Create new instance of AlignmentCounts(gaps, identities, mismatches)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.AlignmentCounts.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.AlignmentCounts.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Return a nicely formatted representation string</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.AlignmentCounts.__slots__">
<span class="sig-name descname"><span class="pre">__slots__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.Align.AlignmentCounts.__slots__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.AlignmentCounts.gaps">
<span class="sig-name descname"><span class="pre">gaps</span></span><a class="headerlink" href="#Bio.Align.AlignmentCounts.gaps" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.AlignmentCounts.identities">
<span class="sig-name descname"><span class="pre">identities</span></span><a class="headerlink" href="#Bio.Align.AlignmentCounts.identities" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.AlignmentCounts.mismatches">
<span class="sig-name descname"><span class="pre">mismatches</span></span><a class="headerlink" href="#Bio.Align.AlignmentCounts.mismatches" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">MultipleSeqAlignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents a classical multiple sequence alignment (MSA).</p>
<p>By this we mean a collection of sequences (usually shown as rows) which
are all the same length (usually with gap characters for insertions or
padding). The data can then be regarded as a matrix of letters, with well
defined columns.</p>
<p>You would typically create an MSA by loading an alignment file with the
AlignIO module:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;Clustalw/opuntia.aln&quot;</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">Alignment with 7 rows and 156 columns</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191</span>
<span class="go">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191</span>
</pre></div>
</div>
<p>In some respects you can treat these objects as lists of SeqRecord objects,
each representing a row of the alignment. Iterating over an alignment gives
the SeqRecord object for each row:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">align</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)))</span>
<span class="gp">...</span>
<span class="go">gi|6273285|gb|AF191659.1|AF191 156</span>
<span class="go">gi|6273284|gb|AF191658.1|AF191 156</span>
<span class="go">gi|6273287|gb|AF191661.1|AF191 156</span>
<span class="go">gi|6273286|gb|AF191660.1|AF191 156</span>
<span class="go">gi|6273290|gb|AF191664.1|AF191 156</span>
<span class="go">gi|6273289|gb|AF191663.1|AF191 156</span>
<span class="go">gi|6273291|gb|AF191665.1|AF191 156</span>
</pre></div>
</div>
<p>You can also access individual rows as SeqRecord objects via their index:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="go">gi|6273285|gb|AF191659.1|AF191</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="go">gi|6273291|gb|AF191665.1|AF191</span>
</pre></div>
</div>
<p>And extract columns as strings:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">AAAAAAA</span>
</pre></div>
</div>
<p>Or, take just the first ten columns as a sub-alignment:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">Alignment with 7 rows and 10 columns</span>
<span class="go">TATACATTAA gi|6273285|gb|AF191659.1|AF191</span>
<span class="go">TATACATTAA gi|6273284|gb|AF191658.1|AF191</span>
<span class="go">TATACATTAA gi|6273287|gb|AF191661.1|AF191</span>
<span class="go">TATACATAAA gi|6273286|gb|AF191660.1|AF191</span>
<span class="go">TATACATTAA gi|6273290|gb|AF191664.1|AF191</span>
<span class="go">TATACATTAA gi|6273289|gb|AF191663.1|AF191</span>
<span class="go">TATACATTAA gi|6273291|gb|AF191665.1|AF191</span>
</pre></div>
</div>
<p>Combining this alignment slicing with alignment addition allows you to
remove a section of the alignment. For example, taking just the first
and last ten columns:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">10</span><span class="p">]</span> <span class="o">+</span> <span class="n">align</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">10</span><span class="p">:])</span>
<span class="go">Alignment with 7 rows and 20 columns</span>
<span class="go">TATACATTAAGTGTACCAGA gi|6273285|gb|AF191659.1|AF191</span>
<span class="go">TATACATTAAGTGTACCAGA gi|6273284|gb|AF191658.1|AF191</span>
<span class="go">TATACATTAAGTGTACCAGA gi|6273287|gb|AF191661.1|AF191</span>
<span class="go">TATACATAAAGTGTACCAGA gi|6273286|gb|AF191660.1|AF191</span>
<span class="go">TATACATTAAGTGTACCAGA gi|6273290|gb|AF191664.1|AF191</span>
<span class="go">TATACATTAAGTATACCAGA gi|6273289|gb|AF191663.1|AF191</span>
<span class="go">TATACATTAAGTGTACCAGA gi|6273291|gb|AF191665.1|AF191</span>
</pre></div>
</div>
<p>Note - This object does NOT attempt to model the kind of alignments used
in next generation sequencing with multiple sequencing reads which are
much shorter than the alignment, and where there is usually a consensus or
reference sequence with special status.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize a new MultipleSeqAlignment object.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>records - A list (or iterator) of SeqRecord objects, whose</dt><dd><p>sequences are all the same length.  This may be an be an
empty list.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>alphabet - For backward compatibility only; its value should always</dt><dd><p>be None.</p>
</dd>
</dl>
</li>
<li><p>annotations - Information about the whole alignment (dictionary).</p></li>
<li><dl class="simple">
<dt>column_annotations - Per column annotation (restricted dictionary).</dt><dd><p>This holds Python sequences (lists, strings, tuples)
whose length matches the number of columns. A typical
use would be a secondary structure consensus string.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<p>You would normally load a MSA from a file using Bio.AlignIO, but you
can do this from a list of SeqRecord objects too:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAACGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA-CGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAAGGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">],</span>
<span class="gp">... </span>                             <span class="n">annotations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;demo&quot;</span><span class="p">},</span>
<span class="gp">... </span>                             <span class="n">column_annotations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;stats&quot;</span><span class="p">:</span> <span class="s2">&quot;CCCXCCC&quot;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 7 columns</span>
<span class="go">AAAACGT Alpha</span>
<span class="go">AAA-CGT Beta</span>
<span class="go">AAAAGGT Gamma</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;tool&#39;: &#39;demo&#39;}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="o">.</span><span class="n">column_annotations</span>
<span class="go">{&#39;stats&#39;: &#39;CCCXCCC&#39;}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.column_annotations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">column_annotations</span></span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.column_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Dictionary of per-letter-annotation for the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Return a multi-line string summary of the alignment.</p>
<p>This output is intended to be readable, but large alignments are
shown truncated.  A maximum of 20 rows (sequences) and 50 columns
are shown, with the record identifiers.  This should fit nicely on a
single screen. e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACT-CTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGATAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 12 columns</span>
<span class="go">ACTGCTAGCTAG Alpha</span>
<span class="go">ACT-CTAGCTAG Beta</span>
<span class="go">ACTGCTAGATAG Gamma</span>
</pre></div>
</div>
<p>See also the alignment’s format method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Return a representation of the object for debugging.</p>
<p>The representation cannot be used with eval() to recreate the object,
which is usually possible with simple python objects.  For example:</p>
<p>&lt;Bio.Align.MultipleSeqAlignment instance (2 records of length 14)
at a3c184c&gt;</p>
<p>The hex string is the memory address of the object, see help(id).
This provides a simple way to visually distinguish alignments of
the same size.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.__format__">
<span class="sig-name descname"><span class="pre">__format__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">format_spec</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.__format__" title="Permalink to this definition"></a></dt>
<dd><p>Return the alignment as a string in the specified file format.</p>
<p>The format should be a lower case string supported as an output
format by Bio.AlignIO (such as “fasta”, “clustal”, “phylip”,
“stockholm”, etc), which is used to turn the alignment into a
string.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACT-CTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGATAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">align</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>
<span class="go">&gt;Alpha</span>
<span class="go">ACTGCTAGCTAG</span>
<span class="go">&gt;Beta</span>
<span class="go">ACT-CTAGCTAG</span>
<span class="go">&gt;Gamma</span>
<span class="go">ACTGCTAGATAG</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">align</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">))</span>
<span class="go"> 3 12</span>
<span class="go">Alpha      ACTGCTAGCT AG</span>
<span class="go">Beta       ACT-CTAGCT AG</span>
<span class="go">Gamma      ACTGCTAGAT AG</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.__iter__" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over alignment rows as SeqRecord objects.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACT-CTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGATAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">align</span><span class="p">:</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">Alpha</span>
<span class="go">ACTGCTAGCTAG</span>
<span class="go">Beta</span>
<span class="go">ACT-CTAGCTAG</span>
<span class="go">Gamma</span>
<span class="go">ACTGCTAGATAG</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.__len__" title="Permalink to this definition"></a></dt>
<dd><p>Return the number of sequences in the alignment.</p>
<p>Use len(alignment) to get the number of sequences (i.e. the number of
rows), and alignment.get_alignment_length() to get the length of the
longest sequence (i.e. the number of columns).</p>
<p>This is easy to remember if you think of the alignment as being like a
list of SeqRecord objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.get_alignment_length">
<span class="sig-name descname"><span class="pre">get_alignment_length</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.get_alignment_length" title="Permalink to this definition"></a></dt>
<dd><p>Return the maximum length of the alignment.</p>
<p>All objects in the alignment should (hopefully) have the same
length. This function will go through and find this length
by finding the maximum length of sequences in the alignment.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACT-CTAGCTAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGATAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="o">.</span><span class="n">get_alignment_length</span><span class="p">()</span>
<span class="go">12</span>
</pre></div>
</div>
<p>If you want to know the number of sequences in the alignment,
use len(align) instead:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.extend">
<span class="sig-name descname"><span class="pre">extend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.extend" title="Permalink to this definition"></a></dt>
<dd><p>Add more SeqRecord objects to the alignment as rows.</p>
<p>They must all have the same length as the original alignment. For
example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAACGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA-CGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAAGGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAACGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Delta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">e</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA-GGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Epsilon&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>First we create a small alignment (three rows):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 7 columns</span>
<span class="go">AAAACGT Alpha</span>
<span class="go">AAA-CGT Beta</span>
<span class="go">AAAAGGT Gamma</span>
</pre></div>
</div>
<p>Now we can extend this alignment with another two rows:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">d</span><span class="p">,</span> <span class="n">e</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">Alignment with 5 rows and 7 columns</span>
<span class="go">AAAACGT Alpha</span>
<span class="go">AAA-CGT Beta</span>
<span class="go">AAAAGGT Gamma</span>
<span class="go">AAAACGT Delta</span>
<span class="go">AAA-GGT Epsilon</span>
</pre></div>
</div>
<p>Because the alignment object allows iteration over the rows as
SeqRecords, you can use the extend method with a second alignment
(provided its sequences have the same length as the original alignment).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.append">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.append" title="Permalink to this definition"></a></dt>
<dd><p>Add one more SeqRecord object to the alignment as a new row.</p>
<p>This must have the same length as the original alignment (unless this is
the first record).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;Clustalw/opuntia.aln&quot;</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">Alignment with 7 rows and 156 columns</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191</span>
<span class="go">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">7</span>
</pre></div>
</div>
<p>We’ll now construct a dummy record to append as an example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dummy</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="o">*</span><span class="mi">156</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;dummy&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Now append this to the alignment,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dummy</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">Alignment with 8 rows and 156 columns</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191</span>
<span class="go">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191</span>
<span class="go">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191</span>
<span class="go">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191</span>
<span class="go">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN dummy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>
<span class="go">8</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.__add__">
<span class="sig-name descname"><span class="pre">__add__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.__add__" title="Permalink to this definition"></a></dt>
<dd><p>Combine two alignments with the same number of rows by adding them.</p>
<p>If you have two multiple sequence alignments (MSAs), there are two ways to think
about adding them - by row or by column. Using the extend method adds by row.
Using the addition operator adds by column. For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAAC&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA-C&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">left</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a1</span><span class="p">,</span> <span class="n">b1</span><span class="p">,</span> <span class="n">c1</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="n">annotations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;demo&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;start&quot;</span><span class="p">},</span>
<span class="gp">... </span>                            <span class="n">column_annotations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;stats&quot;</span><span class="p">:</span> <span class="s2">&quot;CCCXC&quot;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">right</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a2</span><span class="p">,</span> <span class="n">b2</span><span class="p">,</span> <span class="n">c2</span><span class="p">],</span>
<span class="gp">... </span>                            <span class="n">annotations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;demo&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;end&quot;</span><span class="p">},</span>
<span class="gp">... </span>                            <span class="n">column_annotations</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;stats&quot;</span><span class="p">:</span> <span class="s2">&quot;CC&quot;</span><span class="p">})</span>
</pre></div>
</div>
<p>Now, let’s look at these two alignments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">left</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 5 columns</span>
<span class="go">AAAAC Alpha</span>
<span class="go">AAA-C Beta</span>
<span class="go">AAAAG Gamma</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">right</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 2 columns</span>
<span class="go">GT Alpha</span>
<span class="go">GT Beta</span>
<span class="go">GT Gamma</span>
</pre></div>
</div>
<p>And add them:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">combined</span> <span class="o">=</span> <span class="n">left</span> <span class="o">+</span> <span class="n">right</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">combined</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 7 columns</span>
<span class="go">AAAACGT Alpha</span>
<span class="go">AAA-CGT Beta</span>
<span class="go">AAAAGGT Gamma</span>
</pre></div>
</div>
<p>For this to work, both alignments must have the same number of records (here
they both have 3 rows):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">left</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">right</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">combined</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
<p>The individual rows are SeqRecord objects, and these can be added together. Refer
to the SeqRecord documentation for details of how the annotation is handled. This
example is a special case in that both original alignments shared the same names,
meaning when the rows are added they also get the same name.</p>
<p>Any common annotations are preserved, but differing annotation is lost. This is
the same behaviour used in the SeqRecord annotations and is designed to prevent
accidental propagation of inappropriate values:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">combined</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;tool&#39;: &#39;demo&#39;}</span>
</pre></div>
</div>
<p>Similarly any common per-column-annotations are combined:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">combined</span><span class="o">.</span><span class="n">column_annotations</span>
<span class="go">{&#39;stats&#39;: &#39;CCCXCCC&#39;}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.__getitem__" title="Permalink to this definition"></a></dt>
<dd><p>Access part of the alignment.</p>
<p>Depending on the indices, you can get a SeqRecord object
(representing a single row), a Seq object (for a single column),
a string (for a single character) or another alignment
(representing some part or all of the alignment).</p>
<p>align[r,c] gives a single character as a string
align[r] gives a row as a SeqRecord
align[r,:] gives a row as a SeqRecord
align[:,c] gives a column as a Seq</p>
<p>align[:] and align[:,:] give a copy of the alignment</p>
<p>Anything else gives a sub alignment, e.g.
align[0:2] or align[0:2,:] uses only row 0 and 1
align[:,1:3] uses only columns 1 and 2
align[0:2,1:3] uses only rows 0 &amp; 1 and only cols 1 &amp; 2</p>
<p>We’ll use the following example alignment here for illustration:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAACGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA-CGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAAGGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAACGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Delta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">e</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA-GGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Epsilon&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">,</span> <span class="n">e</span><span class="p">])</span>
</pre></div>
</div>
<p>You can access a row of the alignment as a SeqRecord using an integer
index (think of the alignment as a list of SeqRecord objects here):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">first_record</span> <span class="o">=</span> <span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">first_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">first_record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="go">Alpha AAAACGT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">last_record</span> <span class="o">=</span> <span class="n">align</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">last_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">last_record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="go">Epsilon AAA-GGT</span>
</pre></div>
</div>
<p>You can also access use python’s slice notation to create a sub-alignment
containing only some of the SeqRecord objects:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sub_alignment</span> <span class="o">=</span> <span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">:</span><span class="mi">5</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sub_alignment</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 7 columns</span>
<span class="go">AAAAGGT Gamma</span>
<span class="go">AAAACGT Delta</span>
<span class="go">AAA-GGT Epsilon</span>
</pre></div>
</div>
<p>This includes support for a step, i.e. align[start:end:step], which
can be used to select every second sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sub_alignment</span> <span class="o">=</span> <span class="n">align</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sub_alignment</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 7 columns</span>
<span class="go">AAAACGT Alpha</span>
<span class="go">AAAAGGT Gamma</span>
<span class="go">AAA-GGT Epsilon</span>
</pre></div>
</div>
<p>Or to get a copy of the alignment with the rows in reverse order:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rev_alignment</span> <span class="o">=</span> <span class="n">align</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rev_alignment</span><span class="p">)</span>
<span class="go">Alignment with 5 rows and 7 columns</span>
<span class="go">AAA-GGT Epsilon</span>
<span class="go">AAAACGT Delta</span>
<span class="go">AAAAGGT Gamma</span>
<span class="go">AAA-CGT Beta</span>
<span class="go">AAAACGT Alpha</span>
</pre></div>
</div>
<p>You can also use two indices to specify both rows and columns. Using simple
integers gives you the entry as a single character string. e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;C&#39;</span>
</pre></div>
</div>
<p>This is equivalent to:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">][</span><span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;C&#39;</span>
</pre></div>
</div>
<p>or:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;C&#39;</span>
</pre></div>
</div>
<p>To get a single column (as a string) use this syntax:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="p">[:,</span> <span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;CCGCG&#39;</span>
</pre></div>
</div>
<p>Or, to get part of a column,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]</span>
<span class="go">&#39;CG&#39;</span>
</pre></div>
</div>
<p>However, in general you get a sub-alignment,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="mi">6</span><span class="p">])</span>
<span class="go">Alignment with 4 rows and 3 columns</span>
<span class="go">-CG Beta</span>
<span class="go">AGG Gamma</span>
<span class="go">ACG Delta</span>
<span class="go">-GG Epsilon</span>
</pre></div>
</div>
<p>This should all seem familiar to anyone who has used the NumPy
array or matrix objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.sort">
<span class="sig-name descname"><span class="pre">sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.sort" title="Permalink to this definition"></a></dt>
<dd><p>Sort the rows (SeqRecord objects) of the alignment in place.</p>
<p>This sorts the rows alphabetically using the SeqRecord object id by
default. The sorting can be controlled by supplying a key function
which must map each SeqRecord to a sort value.</p>
<p>This is useful if you want to add two alignments which use the same
record identifiers, but in a different order. For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align1</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span>
<span class="gp">... </span>             <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Human&quot;</span><span class="p">),</span>
<span class="gp">... </span>             <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">),</span>
<span class="gp">... </span>             <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGC&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Chicken&quot;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align2</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span>
<span class="gp">... </span>             <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">),</span>
<span class="gp">... </span>             <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGTT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Human&quot;</span><span class="p">),</span>
<span class="gp">... </span>             <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGCT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Chicken&quot;</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">])</span>
</pre></div>
</div>
<p>If you simple try and add these without sorting, you get this:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align1</span> <span class="o">+</span> <span class="n">align2</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 8 columns</span>
<span class="go">ACGTCGGT &lt;unknown id&gt;</span>
<span class="go">ACGGCGTT &lt;unknown id&gt;</span>
<span class="go">ACGCCGCT Chicken</span>
</pre></div>
</div>
<p>Consult the SeqRecord documentation which explains why you get a
default value when annotation like the identifier doesn’t match up.
However, if we sort the alignments first, then add them we get the
desired result:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align1</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align2</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align1</span> <span class="o">+</span> <span class="n">align2</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 8 columns</span>
<span class="go">ACGCCGCT Chicken</span>
<span class="go">ACGTCGTT Human</span>
<span class="go">ACGGCGGT Mouse</span>
</pre></div>
</div>
<p>As an example using a different sort order, you could sort on the
GC content of each sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">gc_fraction</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align1</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 4 columns</span>
<span class="go">ACGC Chicken</span>
<span class="go">ACGT Human</span>
<span class="go">ACGG Mouse</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align1</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">record</span><span class="p">:</span> <span class="n">gc_fraction</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align1</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 4 columns</span>
<span class="go">ACGT Human</span>
<span class="go">ACGC Chicken</span>
<span class="go">ACGG Mouse</span>
</pre></div>
</div>
<p>There is also a reverse argument, so if you wanted to sort by ID
but backwards:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align1</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">align1</span><span class="p">)</span>
<span class="go">Alignment with 3 rows and 4 columns</span>
<span class="go">ACGG Mouse</span>
<span class="go">ACGT Human</span>
<span class="go">ACGC Chicken</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.MultipleSeqAlignment.substitutions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">substitutions</span></span><a class="headerlink" href="#Bio.Align.MultipleSeqAlignment.substitutions" title="Permalink to this definition"></a></dt>
<dd><p>Return an Array with the number of substitutions of letters in the alignment.</p>
<p>As an example, consider a multiple sequence alignment of three DNA sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;A--A&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;seq2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;seq3&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq4</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TTTC&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;seq4&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">seq3</span><span class="p">,</span> <span class="n">seq4</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">Alignment with 4 rows and 4 columns</span>
<span class="go">ACGT seq1</span>
<span class="go">A--A seq2</span>
<span class="go">ACGT seq3</span>
<span class="go">TTTC seq4</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">substitutions</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="go">    A   C   G   T</span>
<span class="go">A 3.0 0.5 0.0 2.5</span>
<span class="go">C 0.5 1.0 0.0 2.0</span>
<span class="go">G 0.0 0.0 1.0 1.0</span>
<span class="go">T 2.5 2.0 1.0 1.0</span>
</pre></div>
</div>
<p>Note that the matrix is symmetric, with counts divided equally on both
sides of the diagonal. For example, the total number of substitutions
between A and T in the alignment is 3.5 + 3.5 = 7.</p>
<p>Any weights associated with the sequences are taken into account when
calculating the substitution matrix.  For example, given the following
multiple sequence alignment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GTATC</span>  <span class="mf">0.5</span>
<span class="n">AT</span><span class="o">--</span><span class="n">C</span>  <span class="mf">0.8</span>
<span class="n">CTGTC</span>  <span class="mf">1.0</span>
</pre></div>
</div>
<p>For the first column we have:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">)</span> <span class="p">:</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">0.8</span> <span class="o">=</span> <span class="mf">0.4</span>
<span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">)</span> <span class="p">:</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="mf">1.0</span> <span class="o">=</span> <span class="mf">0.5</span>
<span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">)</span> <span class="p">:</span> <span class="mf">0.8</span> <span class="o">*</span> <span class="mf">1.0</span> <span class="o">=</span> <span class="mf">0.8</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.Alignment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">Alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents a sequence alignment.</p>
<p>An Alignment object has a <cite>.sequences</cite> attribute storing the sequences
(Seq, MutableSeq, SeqRecord, or string objects) that were aligned, as well
as a <cite>.coordinates</cite> attribute storing the sequence coordinates defining the
alignment as a numpy array.</p>
<dl class="simple">
<dt>Other commonly used attributes (which may or may not be present) are:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>annotations        - A dictionary with annotations describing the</dt><dd><p>alignment;</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>column_annotations - A dictionary with annotations describing each</dt><dd><p>column in the alignment;</p>
</dd>
</dl>
</li>
<li><p>score              - The alignment score.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.infer_coordinates">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">infer_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lines</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skipped_columns</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.infer_coordinates" title="Permalink to this definition"></a></dt>
<dd><p>Infer the coordinates from a printed alignment.</p>
<p>This method is primarily employed in Biopython’s alignment parsers,
though it may be useful for other purposes.</p>
<p>For an alignment consisting of N sequences, printed as N lines with
the same number of columns, where gaps are represented by dashes,
this method will calculate the sequence coordinates that define the
alignment. The coordinates are returned as a numpy array of integers,
and can be used to create an Alignment object.</p>
<p>The argument skipped columns should be None (the default) or an empty
list. If skipped_columns is a list, then the indices of any columns in
the alignment with a gap in all lines are appended to skipped_columns.</p>
<p>This is an example for the alignment of three sequences TAGGCATACGTG,
AACGTACGT, and ACGCATACTTG, with gaps in the second and third sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">Alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lines</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;TAGGCATACGTG&quot;</span><span class="p">,</span>
<span class="gp">... </span>         <span class="s2">&quot;AACG--TACGT-&quot;</span><span class="p">,</span>
<span class="gp">... </span>         <span class="s2">&quot;-ACGCATACTTG&quot;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span>
<span class="go">[&#39;TAGGCATACGTG&#39;, &#39;AACGTACGT&#39;, &#39;ACGCATACTTG&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span> <span class="o">=</span> <span class="n">Alignment</span><span class="o">.</span><span class="n">infer_coordinates</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span>
<span class="go">array([[ 0,  1,  4,  6, 11, 12],</span>
<span class="go">       [ 0,  1,  4,  4,  9,  9],</span>
<span class="go">       [ 0,  0,  3,  5, 10, 11]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">coordinates</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize a new Alignment object.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>sequences   - A list of the sequences (Seq, MutableSeq, SeqRecord,</dt><dd><p>or string objects) that were aligned.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>coordinates - The sequence coordinates that define the alignment.</dt><dd><p>If None (the default value), assume that the sequences
align to each other without any gaps.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__array__">
<span class="sig-name descname"><span class="pre">__array__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__array__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.Alignment.target">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">target</span></span><a class="headerlink" href="#Bio.Align.Alignment.target" title="Permalink to this definition"></a></dt>
<dd><p>Return self.sequences[0] for a pairwise alignment.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.Alignment.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Align.Alignment.query" title="Permalink to this definition"></a></dt>
<dd><p>Return self.sequences[1] for a pairwise alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__eq__">
<span class="sig-name descname"><span class="pre">__eq__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__eq__" title="Permalink to this definition"></a></dt>
<dd><p>Check if two Alignment objects specify the same alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__ne__">
<span class="sig-name descname"><span class="pre">__ne__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__ne__" title="Permalink to this definition"></a></dt>
<dd><p>Check if two Alignment objects have different alignments.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__lt__">
<span class="sig-name descname"><span class="pre">__lt__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__lt__" title="Permalink to this definition"></a></dt>
<dd><p>Check if self should come before other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__le__">
<span class="sig-name descname"><span class="pre">__le__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__le__" title="Permalink to this definition"></a></dt>
<dd><p>Check if self should come before or is equal to other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__gt__">
<span class="sig-name descname"><span class="pre">__gt__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__gt__" title="Permalink to this definition"></a></dt>
<dd><p>Check if self should come after other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__ge__">
<span class="sig-name descname"><span class="pre">__ge__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__ge__" title="Permalink to this definition"></a></dt>
<dd><p>Check if self should come after or is equal to other.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.Alignment.path">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">path</span></span><a class="headerlink" href="#Bio.Align.Alignment.path" title="Permalink to this definition"></a></dt>
<dd><p>Return the path through the trace matrix.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__getitem__" title="Permalink to this definition"></a></dt>
<dd><p>Return self[key].</p>
<p>Indices of the form</p>
<p>self[:, :]</p>
<p>return a copy of the Alignment object;</p>
<p>self[:, i:]
self[:, :j]
self[:, i:j]
self[:, iterable] (where iterable returns integers)</p>
<p>return a new Alignment object spanning the selected columns;</p>
<p>self[k, i]
self[k, i:]
self[k, :j]
self[k, i:j]
self[k, iterable] (where iterable returns integers)
self[k] (equivalent to self[k, :])</p>
<p>return a string with the aligned sequence (including gaps) for the
selected columns, where k = 0 represents the target and k = 1
represents the query sequence; and</p>
<p>self[:, i]</p>
<p>returns a string with the selected column in the alignment.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;ACCGGTTT&quot;</span><span class="p">,</span> <span class="s2">&quot;ACGGGTT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 ACCGG-TTT 8</span>
<span class="go">                  0 ||-||-||- 9</span>
<span class="go">query             0 AC-GGGTT- 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span>
<span class="go">&#39;ACCGG-TTT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
<span class="go">&#39;AC-GGGTT-&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;ACCGG-TTT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;AC-GGGTT-&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;CCGG-T&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;C-GGGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">)]</span>
<span class="go">&#39;C-C&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">::</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;A-GT-&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">2</span><span class="p">)]</span>
<span class="go">&#39;A-GT-&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;AA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">5</span><span class="p">]</span>
<span class="go">&#39;-G&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:]</span>  
<span class="go">&lt;Alignment object (2 rows x 8 columns) at 0x...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:])</span>
<span class="go">target            1 CCGG-TTT 8</span>
<span class="go">                  0 |-||-||- 8</span>
<span class="go">query             1 C-GGGTT- 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">:])</span>
<span class="go">target            2 CGG-TTT 8</span>
<span class="go">                  0 -||-||- 7</span>
<span class="go">query             2 -GGGTT- 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">3</span><span class="p">:])</span>
<span class="go">target            3 GG-TTT 8</span>
<span class="go">                  0 ||-||- 6</span>
<span class="go">query             2 GGGTT- 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">3</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="go">target            3 GG-TT 7</span>
<span class="go">                  0 ||-|| 5</span>
<span class="go">query             2 GGGTT 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="p">::</span><span class="mi">2</span><span class="p">])</span>
<span class="go">target            0 ACGTT 5</span>
<span class="go">                  0 |-||- 5</span>
<span class="go">query             0 A-GT- 3</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">2</span><span class="p">)])</span>
<span class="go">target            0 CG-T 3</span>
<span class="go">                  0 ||-| 4</span>
<span class="go">query             0 CGGT 4</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">3</span><span class="p">)])</span>
<span class="go">target            0 CTG 3</span>
<span class="go">                  0 -|| 3</span>
<span class="go">query             0 -TG 2</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__format__">
<span class="sig-name descname"><span class="pre">__format__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">format_spec</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__format__" title="Permalink to this definition"></a></dt>
<dd><p>Return the alignment as a string in the specified file format.</p>
<p>Wrapper for self.format().</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.format">
<span class="sig-name descname"><span class="pre">format</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fmt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.format" title="Permalink to this definition"></a></dt>
<dd><p>Return the alignment as a string in the specified file format.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>fmt       - File format. Acceptable values are an empty string to</dt><dd><p>create a human-readable representation of the alignment,
or any of the alignment file formats supported by
<cite>Bio.Align</cite> (some have not yet been implemented).</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>All other arguments are passed to the format-specific writer functions:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>mask      - PSL format only. Specify if repeat regions in the target</dt><dd><p>sequence are masked and should be reported in the
<cite>repMatches</cite> field of the PSL file instead of in the
<cite>matches</cite> field. Acceptable values are
None   : no masking (default);
“lower”: masking by lower-case characters;
“upper”: masking by upper-case characters.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>wildcard  - PSL format only. Report alignments to the wildcard</dt><dd><p>character in the target or query sequence in the
<cite>nCount</cite> field of the PSL file instead of in the
<cite>matches</cite>, <cite>misMatches</cite>, or <cite>repMatches</cite> fields.
Default value is ‘N’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>md        - SAM format only. If True, calculate the MD tag from</dt><dd><p>the alignment and include it in the output. If False
(default), do not include the MD tag in the output.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Return a human-readable string representation of the alignment.</p>
<p>For sequence alignments, each line has at most 80 columns.
The first 10 columns show the (possibly truncated) sequence name,
which may be the id attribute of a SeqRecord, or otherwise ‘target’
or ‘query’ for pairwise alignments.
The next 10 columns show the sequence coordinate, using zero-based
counting as usual in Python.
The remaining 60 columns shown the sequence, using dashes to represent
gaps.
At the end of the alignment, the end coordinates are shown on the right
of the sequence, again in zero-based coordinates.</p>
<p>Pairwise alignments have an additional line between the two sequences
showing whether the sequences match (‘|’) or mismatch (‘.’), or if
there is a gap (‘-‘).
The coordinates shown for this line are the column indices, which can
be useful when extracting a subalignment.</p>
<p>For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">()</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqA</span> <span class="o">=</span> <span class="s2">&quot;TTAACCCCATTTG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqB</span> <span class="o">=</span> <span class="s2">&quot;AAGCCCCTTT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqC</span> <span class="o">=</span> <span class="s2">&quot;AAAGGGGCTT&quot;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">seqA</span><span class="p">,</span> <span class="n">seqB</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 TTAA-CCCCATTTG 13</span>
<span class="go">                  0 --||-||||-|||- 14</span>
<span class="go">query             0 --AAGCCCC-TTT- 10</span>
</pre></div>
</div>
<p>Note that seqC is the reverse complement of seqB. Aligning it to the
reverse strand gives the same alignment, but the query coordinates are
switched:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">seqA</span><span class="p">,</span> <span class="n">seqC</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 TTAA-CCCCATTTG 13</span>
<span class="go">                  0 --||-||||-|||- 14</span>
<span class="go">query            10 --AAGCCCC-TTT-  0</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Return a representation of the alignment, including its shape.</p>
<p>The representation cannot be used with eval() to recreate the object,
which is usually possible with simple python objects.  For example:</p>
<p>&lt;Alignment object (2 rows x 14 columns) at 0x10403d850&gt;</p>
<p>The hex string is the memory address of the object and can be used to
distinguish different Alignment objects.  See help(id) for more
information.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">Alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">((</span><span class="s2">&quot;ACCGT&quot;</span><span class="p">,</span> <span class="s2">&quot;ACGT&quot;</span><span class="p">),</span>
<span class="gp">... </span>                      <span class="n">coordinates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
<span class="gp">... </span>                                              <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>                                             <span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 ACCGT 5</span>
<span class="go">                  0 ||-|| 5</span>
<span class="go">query             0 AC-GT 4</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (2 rows x 5 columns) at 0x...&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.__len__" title="Permalink to this definition"></a></dt>
<dd><p>Return the number of sequences in the alignment.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.Alignment.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#Bio.Align.Alignment.shape" title="Permalink to this definition"></a></dt>
<dd><p>Return the shape of the alignment as a tuple of two integer values.</p>
<p>The first integer value is the number of sequences in the alignment as
returned by len(alignment), which is always 2 for pairwise alignments.</p>
<p>The second integer value is the number of columns in the alignment when
it is printed, and is equal to the sum of the number of matches, number
of mismatches, and the total length of gaps in the target and query.
Sequence sections beyond the aligned segment are not included in the
number of columns.</p>
<p>For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;global&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;GACCTG&quot;</span><span class="p">,</span> <span class="s2">&quot;CGATCG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 -GACCT-G 6</span>
<span class="go">                  0 -||--|-| 8</span>
<span class="go">query             0 CGA--TCG 6</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(2, 8)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;local&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;GACCTG&quot;</span><span class="p">,</span> <span class="s2">&quot;CGATCG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 GACCT-G 6</span>
<span class="go">                  0 ||--|-| 7</span>
<span class="go">query             1 GA--TCG 6</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(2, 7)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.Alignment.aligned">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aligned</span></span><a class="headerlink" href="#Bio.Align.Alignment.aligned" title="Permalink to this definition"></a></dt>
<dd><p>Return the indices of subsequences aligned to each other.</p>
<p>This property returns the start and end indices of subsequences
in the target and query sequence that were aligned to each other.
If the alignment between target (t) and query (q) consists of N
chunks, you get two tuples of length N:</p>
<blockquote>
<div><dl class="simple">
<dt>(((t_start1, t_end1), (t_start2, t_end2), …, (t_startN, t_endN)),</dt><dd><p>((q_start1, q_end1), (q_start2, q_end2), …, (q_startN, q_endN)))</p>
</dd>
</dl>
</div></blockquote>
<p>For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;GAACT&quot;</span><span class="p">,</span> <span class="s2">&quot;GAT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 GAACT 5</span>
<span class="go">                  0 ||--| 5</span>
<span class="go">query             0 GA--T 3</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 2],</span>
<span class="go">        [4, 5]],</span>

<span class="go">       [[0, 2],</span>
<span class="go">        [2, 3]]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 GAACT 5</span>
<span class="go">                  0 |-|-| 5</span>
<span class="go">query             0 G-A-T 3</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 1],</span>
<span class="go">        [2, 3],</span>
<span class="go">        [4, 5]],</span>

<span class="go">       [[0, 1],</span>
<span class="go">        [1, 2],</span>
<span class="go">        [2, 3]]])</span>
</pre></div>
</div>
<p>Note that different alignments may have the same subsequences
aligned to each other. In particular, this may occur if alignments
differ from each other in terms of their gap placement only:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mismatch_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;AAACAAA&quot;</span><span class="p">,</span> <span class="s2">&quot;AAAGAAA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 AAAC-AAA 7</span>
<span class="go">                  0 |||--||| 8</span>
<span class="go">query             0 AAA-GAAA 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 3],</span>
<span class="go">        [4, 7]],</span>

<span class="go">       [[0, 3],</span>
<span class="go">        [4, 7]]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">target            0 AAA-CAAA 7</span>
<span class="go">                  0 |||--||| 8</span>
<span class="go">query             0 AAAG-AAA 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 3],</span>
<span class="go">        [4, 7]],</span>

<span class="go">       [[0, 3],</span>
<span class="go">        [4, 7]]])</span>
</pre></div>
</div>
<p>The property can be used to identify alignments that are identical
to each other in terms of their aligned sequences.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.Alignment.indices">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">indices</span></span><a class="headerlink" href="#Bio.Align.Alignment.indices" title="Permalink to this definition"></a></dt>
<dd><p>Return the sequence index of each lettter in the alignment.</p>
<p>This property returns a 2D numpy array with the sequence index of each
letter in the alignment. Gaps are indicated by -1.  The array has the
same number of rows and columns as the alignment, as given by
<cite>self.shape</cite>.</p>
<p>For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;local&quot;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;GAACTGG&quot;</span><span class="p">,</span> <span class="s2">&quot;AATG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            1 AACTG 6</span>
<span class="go">                  0 ||-|| 5</span>
<span class="go">query             0 AA-TG 4</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">indices</span>
<span class="go">array([[ 1,  2,  3,  4,  5],</span>
<span class="go">       [ 0,  1, -1,  2,  3]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            1 AACTGG 7</span>
<span class="go">                  0 ||-|-| 6</span>
<span class="go">query             0 AA-T-G 4</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">indices</span>
<span class="go">array([[ 1,  2,  3,  4,  5,  6],</span>
<span class="go">       [ 0,  1, -1,  2, -1,  3]])</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;GAACTGG&quot;</span><span class="p">,</span> <span class="s2">&quot;CATT&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            1 AACTG 6</span>
<span class="go">                  0 ||-|| 5</span>
<span class="go">query             4 AA-TG 0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">indices</span>
<span class="go">array([[ 1,  2,  3,  4,  5],</span>
<span class="go">       [ 3,  2, -1,  1,  0]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            1 AACTGG 7</span>
<span class="go">                  0 ||-|-| 6</span>
<span class="go">query             4 AA-T-G 0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">indices</span>
<span class="go">array([[ 1,  2,  3,  4,  5,  6],</span>
<span class="go">       [ 3,  2, -1,  1, -1,  0]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.Alignment.inverse_indices">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">inverse_indices</span></span><a class="headerlink" href="#Bio.Align.Alignment.inverse_indices" title="Permalink to this definition"></a></dt>
<dd><p>Return the alignment column index for each letter in each sequence.</p>
<p>This property returns a list of 1D numpy arrays; the number of arrays
is equal to the number of aligned sequences, and the length of each
array is equal to the length of the corresponding sequence. For each
letter in each sequence, the array contains the corresponding column
index in the alignment. Letters not included in the alignment are
indicated by -1.</p>
<p>For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;local&quot;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;GAACTGG&quot;</span><span class="p">,</span> <span class="s2">&quot;AATG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            1 AACTG 6</span>
<span class="go">                  0 ||-|| 5</span>
<span class="go">query             0 AA-TG 4</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">inverse_indices</span>
<span class="go">[array([-1,  0,  1,  2,  3,  4, -1]), array([0, 1, 3, 4])]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            1 AACTGG 7</span>
<span class="go">                  0 ||-|-| 6</span>
<span class="go">query             0 AA-T-G 4</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">inverse_indices</span>
<span class="go">[array([-1,  0,  1,  2,  3,  4,  5]), array([0, 1, 3, 5])]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;GAACTGG&quot;</span><span class="p">,</span> <span class="s2">&quot;CATT&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            1 AACTG 6</span>
<span class="go">                  0 ||-|| 5</span>
<span class="go">query             4 AA-TG 0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">inverse_indices</span>
<span class="go">[array([-1,  0,  1,  2,  3,  4, -1]), array([4, 3, 1, 0])]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            1 AACTGG 7</span>
<span class="go">                  0 ||-|-| 6</span>
<span class="go">query             4 AA-T-G 0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">inverse_indices</span>
<span class="go">[array([-1,  0,  1,  2,  3,  4,  5]), array([5, 3, 1, 0])]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.sort">
<span class="sig-name descname"><span class="pre">sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.sort" title="Permalink to this definition"></a></dt>
<dd><p>Sort the sequences of the alignment in place.</p>
<p>By default, this sorts the sequences alphabetically using their id
attribute if available, or by their sequence contents otherwise.
For example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;AATAA&quot;</span><span class="p">,</span> <span class="s2">&quot;AAGAA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 AATAA 5</span>
<span class="go">                  0 ||.|| 5</span>
<span class="go">query             0 AAGAA 5</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 AAGAA 5</span>
<span class="go">                  0 ||.|| 5</span>
<span class="go">query             0 AATAA 5</span>
</pre></div>
</div>
<p>Alternatively, a key function can be supplied that maps each sequence
to a sort value.  For example, you could sort on the GC content of each
sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">gc_fraction</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">gc_fraction</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 AATAA 5</span>
<span class="go">                  0 ||.|| 5</span>
<span class="go">query             0 AAGAA 5</span>
</pre></div>
</div>
<p>You can reverse the sort order by passing <cite>reverse=True</cite>:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">gc_fraction</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 AAGAA 5</span>
<span class="go">                  0 ||.|| 5</span>
<span class="go">query             0 AATAA 5</span>
</pre></div>
</div>
<p>The sequences are now sorted by decreasing GC content value.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.map">
<span class="sig-name descname"><span class="pre">map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.map" title="Permalink to this definition"></a></dt>
<dd><p>Map the alignment to self.target and return the resulting alignment.</p>
<p>Here, self.query and alignment.target are the same sequence.</p>
<p>A typical example is where self is the pairwise alignment between a
chromosome and a transcript, the argument is the pairwise alignment
between the transcript and a sequence (e.g., as obtained by RNA-seq),
and we want to find the alignment of the sequence to the chromosome:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;local&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">open_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">extend_gap_score</span> <span class="o">=</span> <span class="mi">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromosome</span> <span class="o">=</span> <span class="s2">&quot;AAAAAAAACCCCCCCAAAAAAAAAAAGGGGGGAAAAAAAA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">transcript</span> <span class="o">=</span> <span class="s2">&quot;CCCCCCCGGGGGG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments1</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">transcript</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments1</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span> <span class="o">=</span> <span class="n">alignments1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment1</span><span class="p">)</span>
<span class="go">target            8 CCCCCCCAAAAAAAAAAAGGGGGG 32</span>
<span class="go">                  0 |||||||-----------|||||| 24</span>
<span class="go">query             0 CCCCCCC-----------GGGGGG 13</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;CCCCGGGG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments2</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">transcript</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments2</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span> <span class="o">=</span> <span class="n">alignments2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="go">target            3 CCCCGGGG 11</span>
<span class="go">                  0 ||||||||  8</span>
<span class="go">query             0 CCCCGGGG  8</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignment1</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target           11 CCCCAAAAAAAAAAAGGGG 30</span>
<span class="go">                  0 ||||-----------|||| 19</span>
<span class="go">query             0 CCCC-----------GGGG  8</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="go">&#39;8\t0\t0\t0\t0\t0\t1\t11\t+\tquery\t8\t0\t8\ttarget\t40\t11\t30\t2\t4,4,\t0,4,\t11,26,\n&#39;</span>
</pre></div>
</div>
<p>Mapping the alignment does not depend on the sequence contents. If we
delete the sequence contents, the same alignment is found in PSL format
(though we obviously lose the ability to print the sequence alignment):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment1</span><span class="o">.</span><span class="n">target</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span><span class="o">.</span><span class="n">query</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment1</span><span class="o">.</span><span class="n">query</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">target</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span><span class="o">.</span><span class="n">query</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">query</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignment1</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="go">&#39;8\t0\t0\t0\t0\t0\t1\t11\t+\tquery\t8\t0\t8\ttarget\t40\t11\t30\t2\t4,4,\t0,4,\t11,26,\n&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Align.Alignment.substitutions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">substitutions</span></span><a class="headerlink" href="#Bio.Align.Alignment.substitutions" title="Permalink to this definition"></a></dt>
<dd><p>Return an Array with the number of substitutions of letters in the alignment.</p>
<p>As an example, consider a sequence alignment of two RNA sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;ATACTTACCTGGCAGGGGAGATACCATGATCACGAAGGTGGTTTTCCCAGGGCGAGGCTTATCCATTGCACTCCGGATGTGCTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACTGCATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTTTCCCCTG&quot;</span>  <span class="c1"># human spliceosomal small nuclear RNA U1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;ATACTTACCTGACAGGGGAGGCACCATGATCACACAGGTGGTCCTCCCAGGGCGAGGCTCTTCCATTGCACTGCGGGAGGGTTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACTGTATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTATCCCCCG&quot;</span>  <span class="c1"># sea lamprey spliceosomal small RNA U1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 ATACTTACCTGGCAGGGGAGATACCATGATCACGAAGGTGGTTTTCCCAGGGCGAGGCTT</span>
<span class="go">                  0 |||||||||||.||||||||..|||||||||||..|||||||..|||||||||||||||.</span>
<span class="go">query             0 ATACTTACCTGACAGGGGAGGCACCATGATCACACAGGTGGTCCTCCCAGGGCGAGGCTC</span>

<span class="go">target           60 ATCCATTGCACTCCGGATGTGCTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACT</span>
<span class="go">                 60 .|||||||||||.|||..|.|.||||||||||||||||||||||||||||||||||||||</span>
<span class="go">query            60 TTCCATTGCACTGCGGGAGGGTTGACCCCTGCGATTTCCCCAAATGTGGGAAACTCGACT</span>

<span class="go">target          120 GCATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTTTCCCCTG 164</span>
<span class="go">                120 |.||||||||||||||||||||||||||||||||||.|||||.| 164</span>
<span class="go">query           120 GTATAATTTGTGGTAGTGGGGGACTGCGTTCGCGCTATCCCCCG 164</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">substitutions</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="go">     A    C    G    T</span>
<span class="go">A 28.0  1.0  2.0  1.0</span>
<span class="go">C  0.0 39.0  1.0  2.0</span>
<span class="go">G  2.0  0.0 45.0  0.0</span>
<span class="go">T  2.0  5.0  1.0 35.0</span>
</pre></div>
</div>
<p>Note that the matrix is not symmetric: rows correspond to the target
sequence, and columns to the query sequence.  For example, the number
of T’s in the target sequence that are aligned to a C in the query
sequence is</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">]</span>
<span class="go">5.0</span>
</pre></div>
</div>
<p>and the number of C’s in the query sequence tat are aligned to a T in
the query sequence is</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
<span class="go">2.0</span>
</pre></div>
</div>
<p>For some applications (for example, to define a scoring matrix from
the substitution matrix), a symmetric matrix may be preferred, which
can be calculated as follows:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">+=</span> <span class="n">m</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">/=</span> <span class="mf">2.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="go">     A    C    G    T</span>
<span class="go">A 28.0  0.5  2.0  1.5</span>
<span class="go">C  0.5 39.0  0.5  3.5</span>
<span class="go">G  2.0  0.5 45.0  0.5</span>
<span class="go">T  1.5  3.5  0.5 35.0</span>
</pre></div>
</div>
<p>The matrix is now symmetric, with counts divided equally on both sides
of the diagonal:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
<span class="go">3.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">]</span>
<span class="go">3.5</span>
</pre></div>
</div>
<p>The total number of substitutions between T’s and C’s in the alignment
is 3.5 + 3.5 = 7.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.Alignment.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Alignment.counts" title="Permalink to this definition"></a></dt>
<dd><p>Return number of identities, mismatches, and gaps, of a pairwise alignment.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;global&#39;</span><span class="p">,</span> <span class="n">match_score</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">mismatch_score</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;TACCG&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score = </span><span class="si">%.1f</span><span class="s2">:&quot;</span> <span class="o">%</span> <span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">c</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">counts</span><span class="p">()</span>  <span class="c1"># namedtuple</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">c</span><span class="o">.</span><span class="n">gaps</span><span class="si">}</span><span class="s2"> gaps, </span><span class="si">{</span><span class="n">c</span><span class="o">.</span><span class="n">identities</span><span class="si">}</span><span class="s2"> identities, </span><span class="si">{</span><span class="n">c</span><span class="o">.</span><span class="n">mismatches</span><span class="si">}</span><span class="s2"> mismatches&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">Score = 6.0:</span>
<span class="go">2 gaps, 3 identities, 0 mismatches</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -||-| 5</span>
<span class="go">query             0 -AC-G 3</span>

<span class="go">Score = 6.0:</span>
<span class="go">2 gaps, 3 identities, 0 mismatches</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -|-|| 5</span>
<span class="go">query             0 -A-CG 3</span>
</pre></div>
</div>
<p>This classifies each pair of letters in a pairwise alignment into gaps,
perfect matches, or mismatches. It has been defined as a method (not a
property) so that it may in future take optional argument(s) allowing
the behaviour to be customised. These three values are returned as a
namedtuple. This is calculated for all the pairs of sequences in the
alignment.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.Alignment.__hash__">
<span class="sig-name descname"><span class="pre">__hash__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#Bio.Align.Alignment.__hash__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignments">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">PairwiseAlignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">paths</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignments" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Implements an iterator over pairwise alignments returned by the aligner.</p>
<p>This class also supports indexing, which is fast for increasing indices,
but may be slow for random access of a large number of alignments.</p>
<p>Note that pairwise aligners can return an astronomical number of alignments,
even for relatively short sequences, if they align poorly to each other. We
therefore recommend to first check the number of alignments, accessible as
len(alignments), which can be calculated quickly even if the number of
alignments is very large.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignments.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">paths</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignments.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize a new PairwiseAlignments object.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>seqA  - The first sequence, as a plain string, without gaps.</p></li>
<li><p>seqB  - The second sequence, as a plain string, without gaps.</p></li>
<li><p>score - The alignment score.</p></li>
<li><dl class="simple">
<dt>paths - An iterator over the paths in the traceback matrix;</dt><dd><p>each path defines one alignment.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<p>You would normally obtain a PairwiseAlignments object by calling
aligner.align(seqA, seqB), where aligner is a PairwiseAligner object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignments.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignments.__len__" title="Permalink to this definition"></a></dt>
<dd><p>Return the number of alignments.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignments.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignments.__getitem__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignments.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignments.__iter__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignments.__next__">
<span class="sig-name descname"><span class="pre">__next__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignments.__next__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignments.rewind">
<span class="sig-name descname"><span class="pre">rewind</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignments.rewind" title="Permalink to this definition"></a></dt>
<dd><p>Rewind the iterator to let it loop over the alignments from the beginning.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAligner">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">PairwiseAligner</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scoring</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAligner" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">_algorithms.PairwiseAligner</span></code></p>
<p>Performs pairwise sequence alignment using dynamic programming.</p>
<p>This provides functions to get global and local alignments between two
sequences.  A global alignment finds the best concordance between all
characters in two sequences.  A local alignment finds just the
subsequences that align the best.</p>
<p>To perform a pairwise sequence alignment, first create a PairwiseAligner
object.  This object stores the match and mismatch scores, as well as the
gap scores.  Typically, match scores are positive, while mismatch scores
and gap scores are negative or zero.  By default, the match score is 1,
and the mismatch and gap scores are zero.  Based on the values of the gap
scores, a PairwiseAligner object automatically chooses the appropriate
alignment algorithm (the Needleman-Wunsch, Smith-Waterman, Gotoh, or
Waterman-Smith-Beyer global or local alignment algorithm).</p>
<p>Calling the “score” method on the aligner with two sequences as arguments
will calculate the alignment score between the two sequences.
Calling the “align” method on the aligner with two sequences as arguments
will return a generator yielding the alignments between the two
sequences.</p>
<p>Some examples:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;TACCG&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">alignments</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score = </span><span class="si">%.1f</span><span class="s2">:&quot;</span> <span class="o">%</span> <span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">Score = 3.0:</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -|-|| 5</span>
<span class="go">query             0 -A-CG 3</span>

<span class="go">Score = 3.0:</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -||-| 5</span>
<span class="go">query             0 -AC-G 3</span>
</pre></div>
</div>
<p>Specify the aligner mode as local to generate local alignments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;local&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;TACCG&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">alignments</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score = </span><span class="si">%.1f</span><span class="s2">:&quot;</span> <span class="o">%</span> <span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">Score = 3.0:</span>
<span class="go">target            1 ACCG 5</span>
<span class="go">                  0 |-|| 4</span>
<span class="go">query             0 A-CG 3</span>

<span class="go">Score = 3.0:</span>
<span class="go">target            1 ACCG 5</span>
<span class="go">                  0 ||-| 4</span>
<span class="go">query             0 AC-G 3</span>
</pre></div>
</div>
<p>Do a global alignment.  Identical characters are given 2 points,
1 point is deducted for each non-identical character.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;global&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">match_score</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mismatch_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;TACCG&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score = </span><span class="si">%.1f</span><span class="s2">:&quot;</span> <span class="o">%</span> <span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">Score = 6.0:</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -||-| 5</span>
<span class="go">query             0 -AC-G 3</span>

<span class="go">Score = 6.0:</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -|-|| 5</span>
<span class="go">query             0 -A-CG 3</span>
</pre></div>
</div>
<p>Same as above, except now 0.5 points are deducted when opening a
gap, and 0.1 points are deducted when extending it.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">open_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">extend_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">target_end_gap_score</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">query_end_gap_score</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;TACCG&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score = </span><span class="si">%.1f</span><span class="s2">:&quot;</span> <span class="o">%</span> <span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">Score = 5.5:</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -|-|| 5</span>
<span class="go">query             0 -A-CG 3</span>

<span class="go">Score = 5.5:</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -||-| 5</span>
<span class="go">query             0 -AC-G 3</span>
</pre></div>
</div>
<p>The alignment function can also use known matrices already included in
Biopython:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">substitution_matrices</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;BLOSUM62&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;KEVLA&quot;</span><span class="p">,</span> <span class="s2">&quot;EVL&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of alignments: </span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">))</span>
<span class="go">Number of alignments: 1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score = </span><span class="si">%.1f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">Score = 13.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 KEVLA 5</span>
<span class="go">                  0 -|||- 5</span>
<span class="go">query             0 -EVL- 3</span>
</pre></div>
</div>
<p>You can also set the value of attributes directly during construction
of the PairwiseAligner object by providing them as keyword arguments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;global&#39;</span><span class="p">,</span> <span class="n">match_score</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">mismatch_score</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;TACCG&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score = </span><span class="si">%.1f</span><span class="s2">:&quot;</span> <span class="o">%</span> <span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">Score = 6.0:</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -||-| 5</span>
<span class="go">query             0 -AC-G 3</span>

<span class="go">Score = 6.0:</span>
<span class="go">target            0 TACCG 5</span>
<span class="go">                  0 -|-|| 5</span>
<span class="go">query             0 -A-CG 3</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAligner.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scoring</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAligner.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize a new PairwiseAligner as specified by the keyword arguments.</p>
<p>If scoring is None, use the default scoring scheme match = 1.0,
mismatch = 0.0, gap score = 0.0
If scoring is “blastn”, “megablast”, or “blastp”, use the default
substitution matrix and gap scores for BLASTN, MEGABLAST, or BLASTP,
respectively.</p>
<p>Loops over the remaining keyword arguments and sets them as attributes
on the object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAligner.__setattr__">
<span class="sig-name descname"><span class="pre">__setattr__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAligner.__setattr__" title="Permalink to this definition"></a></dt>
<dd><p>Implement setattr(self, name, value).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAligner.align">
<span class="sig-name descname"><span class="pre">align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAligner.align" title="Permalink to this definition"></a></dt>
<dd><p>Return the alignments of two sequences using PairwiseAligner.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAligner.score">
<span class="sig-name descname"><span class="pre">score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'+'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAligner.score" title="Permalink to this definition"></a></dt>
<dd><p>Return the alignments score of two sequences using PairwiseAligner.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAligner.__getstate__">
<span class="sig-name descname"><span class="pre">__getstate__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAligner.__getstate__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAligner.__setstate__">
<span class="sig-name descname"><span class="pre">__setstate__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">state</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAligner.__setstate__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">PairwiseAlignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignment" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Align.Alignment" title="Bio.Align.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Align.Alignment</span></code></a></p>
<p>Represents a pairwise sequence alignment.</p>
<p>Internally, the pairwise alignment is stored as the path through
the traceback matrix, i.e. a tuple of pairs of indices corresponding
to the vertices of the path in the traceback matrix.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.PairwiseAlignment.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.PairwiseAlignment.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize a new PairwiseAlignment object.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>target  - The first sequence, as a plain string, without gaps.</p></li>
<li><p>query   - The second sequence, as a plain string, without gaps.</p></li>
<li><dl class="simple">
<dt>path    - The path through the traceback matrix, defining an</dt><dd><p>alignment.</p>
</dd>
</dl>
</li>
<li><p>score   - The alignment score.</p></li>
</ul>
</dd>
</dl>
<p>You would normally obtain a PairwiseAlignment object by iterating
over a PairwiseAlignments object.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.Align.write">
<span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fmt</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.write" title="Permalink to this definition"></a></dt>
<dd><p>Write alignments to a file.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>alignments - List (or iterator) of Alignment objects, or a single
Alignment.</p></li>
<li><p>target     - File or file-like object to write to, or filename as string.</p></li>
<li><p>fmt        - String describing the file format (case-insensitive).</p></li>
</ul>
</dd>
</dl>
<p>Note if providing a file or file-like object, your code should close the
target after calling this function, or call .flush(), to ensure the data
gets flushed to disk.</p>
<p>Returns the number of alignments written (as an integer).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.Align.parse">
<span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fmt</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.parse" title="Permalink to this definition"></a></dt>
<dd><p>Parse an alignment file and return an iterator over alignments.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>source - File or file-like object to read from, or filename as string.</p></li>
<li><p>fmt    - String describing the file format (case-insensitive).</p></li>
</ul>
</dd>
</dl>
<p>Typical usage, opening a file to read in, and looping over the aligments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;Exonerate/exn_22_m_ner_cigar.exn&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;exonerate&quot;</span><span class="p">):</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of sequences in alignment&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="p">))</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Alignment score:&quot;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">Number of sequences in alignment 2</span>
<span class="go">Alignment score: 6150.0</span>
<span class="go">Number of sequences in alignment 2</span>
<span class="go">Alignment score: 502.0</span>
<span class="go">Number of sequences in alignment 2</span>
<span class="go">Alignment score: 440.0</span>
</pre></div>
</div>
<p>For lazy-loading file formats such as bigMaf, for which the file contents
is read on demand only, ensure that the file remains open while extracting
alignment data.</p>
<p>You can use the Bio.Align.read(…) function when the file contains only
one alignment.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.Align.read">
<span class="sig-prename descclassname"><span class="pre">Bio.Align.</span></span><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fmt</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.read" title="Permalink to this definition"></a></dt>
<dd><p>Parse a file containing one alignment, and return it.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>source - File or file-like object to read from, or filename as string.</p></li>
<li><p>fmt    - String describing the file format (case-insensitive).</p></li>
</ul>
</dd>
</dl>
<p>This function is for use parsing alignment files containing exactly one
alignment.  For example, reading a Clustal file:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;Clustalw/opuntia.aln&quot;</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Alignment shape:&quot;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="go">Alignment shape: (7, 156)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">sequence</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">))</span>
<span class="go">gi|6273285|gb|AF191659.1|AF191 146</span>
<span class="go">gi|6273284|gb|AF191658.1|AF191 148</span>
<span class="go">gi|6273287|gb|AF191661.1|AF191 146</span>
<span class="go">gi|6273286|gb|AF191660.1|AF191 146</span>
<span class="go">gi|6273290|gb|AF191664.1|AF191 150</span>
<span class="go">gi|6273289|gb|AF191663.1|AF191 150</span>
<span class="go">gi|6273291|gb|AF191665.1|AF191 156</span>
</pre></div>
</div>
<p>If the file contains no records, or more than one record, an exception is
raised.  For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;Exonerate/exn_22_m_ner_cigar.exn&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;exonerate&quot;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">More than one alignment found in file</span>
</pre></div>
</div>
<p>Use the Bio.Align.parse function if you want to read a file containing
more than one alignment.</p>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Affy.CelFile.html" class="btn btn-neutral float-left" title="Bio.Affy.CelFile module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Align.Applications.html" class="btn btn-neutral float-right" title="Bio.Align.Applications package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2020, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.82.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.80/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>