

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Bio.KEGG.KGML.KGML_pathway module &mdash; Biopython 1.79.dev0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.KEGG.Map package" href="Bio.KEGG.Map.html" />
    <link rel="prev" title="Bio.KEGG.KGML.KGML_parser module" href="Bio.KEGG.KGML.KGML_parser.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.KEGG.html#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html#module-Bio.KEGG">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          



















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.KEGG.html">Bio.KEGG package</a> &raquo;</li>
        
          <li><a href="Bio.KEGG.KGML.html">Bio.KEGG.KGML package</a> &raquo;</li>
        
      <li>Bio.KEGG.KGML.KGML_pathway module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/KEGG/KGML/KGML_pathway.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.KEGG.Map.html" class="btn btn-neutral float-right" title="Bio.KEGG.Map package" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
      
      
        <a href="Bio.KEGG.KGML.KGML_parser.html" class="btn btn-neutral float-left" title="Bio.KEGG.KGML.KGML_parser module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.KEGG.KGML.KGML_pathway">
<span id="bio-kegg-kgml-kgml-pathway-module"></span><h1>Bio.KEGG.KGML.KGML_pathway module<a class="headerlink" href="#module-Bio.KEGG.KGML.KGML_pathway" title="Permalink to this headline">¶</a></h1>
<p>Classes to represent a KGML Pathway Map.</p>
<p>The KGML definition is as of release KGML v0.7.2
(<a class="reference external" href="http://www.kegg.jp/kegg/xml/docs/">http://www.kegg.jp/kegg/xml/docs/</a>)</p>
<dl class="simple">
<dt>Classes:</dt><dd><ul class="simple">
<li><p>Pathway - Specifies graph information for the pathway map</p></li>
<li><p>Relation - Specifies a relationship between two proteins or KOs,
or protein and compound. There is an implied direction to the
relationship in some cases.</p></li>
<li><p>Reaction - A specific chemical reaction between a substrate and
a product.</p></li>
<li><p>Entry - A node in the pathway graph</p></li>
<li><p>Graphics - Entry subelement describing its visual representation</p></li>
</ul>
</dd>
</dl>
<dl class="py class">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.KEGG.KGML.KGML_pathway.</span></code><code class="sig-name descname"><span class="pre">Pathway</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents a KGML pathway from KEGG.</p>
<p>Specifies graph information for the pathway map, as described in
release KGML v0.7.2 (<a class="reference external" href="http://www.kegg.jp/kegg/xml/docs/">http://www.kegg.jp/kegg/xml/docs/</a>)</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>name - KEGGID of the pathway map</p></li>
<li><p>org - ko/ec/[org prefix]</p></li>
<li><p>number - map number (integer)</p></li>
<li><p>title - the map title</p></li>
<li><p>image - URL of the image map for the pathway</p></li>
<li><p>link - URL of information about the pathway</p></li>
<li><p>entries - Dictionary of entries in the pathway, keyed by node ID</p></li>
<li><p>reactions - Set of reactions in the pathway</p></li>
</ul>
</dd>
</dl>
<p>The name attribute has a restricted format, so we make it a property and
enforce the formatting.</p>
<p>The Pathway object is the only allowed route for adding/removing
Entry, Reaction, or Relation elements.</p>
<p>Entries are held in a dictionary and keyed by the node ID for the
pathway graph - this allows for ready access via the Reaction/Relation
etc. elements.  Entries must be added before reference by any other
element.</p>
<p>Reactions are held in a dictionary, keyed by node ID for the path.
The elements referred to in the reaction must be added before the
reaction itself.</p>
<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.get_KGML">
<code class="sig-name descname"><span class="pre">get_KGML</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.get_KGML" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the pathway as a string in prettified KGML format.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.add_entry">
<code class="sig-name descname"><span class="pre">add_entry</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entry</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.add_entry" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an Entry element to the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.remove_entry">
<code class="sig-name descname"><span class="pre">remove_entry</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entry</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.remove_entry" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove an Entry element from the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.add_reaction">
<code class="sig-name descname"><span class="pre">add_reaction</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reaction</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.add_reaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a Reaction element to the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.remove_reaction">
<code class="sig-name descname"><span class="pre">remove_reaction</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reaction</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.remove_reaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove a Reaction element from the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.add_relation">
<code class="sig-name descname"><span class="pre">add_relation</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">relation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.add_relation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a Relation element to the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.remove_relation">
<code class="sig-name descname"><span class="pre">remove_relation</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">relation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.remove_relation" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove a Relation element from the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a readable summary description string.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.name">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">name</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.name" title="Permalink to this definition">¶</a></dt>
<dd><p>The KEGGID for the pathway map.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.number">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">number</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.number" title="Permalink to this definition">¶</a></dt>
<dd><p>The KEGG map number.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.compounds">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">compounds</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.compounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of entries of type compound.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.maps">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">maps</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.maps" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of entries of type map.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.orthologs">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">orthologs</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.orthologs" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of entries of type ortholog.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.genes">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">genes</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of entries of type gene.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.reactions">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">reactions</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.reactions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of reactions in the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.reaction_entries">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">reaction_entries</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.reaction_entries" title="Permalink to this definition">¶</a></dt>
<dd><p>List of entries corresponding to each reaction in the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.relations">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">relations</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.relations" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of relations in the pathway.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.element">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">element</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.element" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Pathway as a valid KGML element.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Pathway.bounds">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">bounds</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Pathway.bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Coordinate bounds for all Graphics elements in the Pathway.</p>
<p>Returns the [(xmin, ymin), (xmax, ymax)] coordinates for all
Graphics elements in the Pathway</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.KEGG.KGML.KGML_pathway.</span></code><code class="sig-name descname"><span class="pre">Entry</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represent an Entry from KGML.</p>
<p>Each Entry element is a node in the pathway graph, as described in
release KGML v0.7.2 (<a class="reference external" href="http://www.kegg.jp/kegg/xml/docs/">http://www.kegg.jp/kegg/xml/docs/</a>)</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>id - The ID of the entry in the pathway map (integer)</p></li>
<li><p>names - List of KEGG IDs for the entry</p></li>
<li><p>type - The type of the entry</p></li>
<li><p>link - URL of information about the entry</p></li>
<li><p>reaction - List of KEGG IDs of the corresponding reactions
(integer)</p></li>
<li><p>graphics -    List of Graphics objects describing the Entry’s visual
representation</p></li>
<li><p>components - List of component node ID for this Entry (‘group’)</p></li>
<li><p>alt - List of alternate names for the Entry</p></li>
</ul>
</dd>
</dl>
<p>NOTE: The alt attribute represents a subelement of the substrate and
product elements in the KGML file</p>
<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return readable descriptive string.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.add_component">
<code class="sig-name descname"><span class="pre">add_component</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">element</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.add_component" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an element to the entry.</p>
<p>If the Entry is already part of a pathway, make sure
the component already exists.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.remove_component">
<code class="sig-name descname"><span class="pre">remove_component</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.remove_component" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove the entry with the passed ID from the group.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.add_graphics">
<code class="sig-name descname"><span class="pre">add_graphics</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entry</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.add_graphics" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the Graphics entry.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.remove_graphics">
<code class="sig-name descname"><span class="pre">remove_graphics</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entry</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.remove_graphics" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove the Graphics entry with the passed ID from the group.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.name">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">name</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.name" title="Permalink to this definition">¶</a></dt>
<dd><p>List of KEGG identifiers for the Entry.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.reaction">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">reaction</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.reaction" title="Permalink to this definition">¶</a></dt>
<dd><p>List of reaction KEGG IDs for this Entry.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.id">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">id</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.id" title="Permalink to this definition">¶</a></dt>
<dd><p>The pathway graph node ID for the Entry.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.element">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">element</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.element" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Entry as a valid KGML element.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.bounds">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">bounds</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Coordinate bounds for all Graphics elements in the Entry.</p>
<p>Return the [(xmin, ymin), (xmax, ymax)] co-ordinates for the Entry
Graphics elements.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Entry.is_reactant">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">is_reactant</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Entry.is_reactant" title="Permalink to this definition">¶</a></dt>
<dd><p>Return true if this Entry participates in any reaction in its parent pathway.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.KEGG.KGML.KGML_pathway.Component">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.KEGG.KGML.KGML_pathway.</span></code><code class="sig-name descname"><span class="pre">Component</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Component" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An Entry subelement used to represents a complex node.</p>
<p>A subelement of the Entry element, used when the Entry is a complex
node, as described in release KGML v0.7.2
(<a class="reference external" href="http://www.kegg.jp/kegg/xml/docs/">http://www.kegg.jp/kegg/xml/docs/</a>)</p>
<p>The Component acts as a collection (with type ‘group’, and typically
its own Graphics subelement), having only an ID.</p>
<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Component.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Component.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Component.id">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">id</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Component.id" title="Permalink to this definition">¶</a></dt>
<dd><p>The pathway graph node ID for the Entry</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Component.element">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">element</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Component.element" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Component as a valid KGML element.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.KEGG.KGML.KGML_pathway.</span></code><code class="sig-name descname"><span class="pre">Graphics</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An Entry subelement used to represents the visual representation.</p>
<p>A subelement of Entry, specifying its visual representation, as
described in release KGML v0.7.2 (<a class="reference external" href="http://www.kegg.jp/kegg/xml/docs/">http://www.kegg.jp/kegg/xml/docs/</a>)</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>name         Label for the graphics object</p></li>
<li><p>x            X-axis position of the object (int)</p></li>
<li><p>y            Y-axis position of the object (int)</p></li>
<li><p>coords       polyline co-ordinates, list of (int, int) tuples</p></li>
<li><p>type         object shape</p></li>
<li><p>width        object width (int)</p></li>
<li><p>height       object height (int)</p></li>
<li><p>fgcolor      object foreground color (hex RGB)</p></li>
<li><p>bgcolor      object background color (hex RGB)</p></li>
</ul>
</dd>
</dl>
<p>Some attributes are present only for specific graphics types.  For
example, line types do not (typically) have a width.
We permit non-DTD attributes and attribute settings, such as</p>
<p>dash         List of ints, describing an on/off pattern for dashes</p>
<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.x">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">x</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.x" title="Permalink to this definition">¶</a></dt>
<dd><p>The X coordinate for the graphics element.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.y">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">y</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.y" title="Permalink to this definition">¶</a></dt>
<dd><p>The Y coordinate for the graphics element.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.width">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">width</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.width" title="Permalink to this definition">¶</a></dt>
<dd><p>The width of the graphics element.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.height">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">height</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.height" title="Permalink to this definition">¶</a></dt>
<dd><p>The height of the graphics element.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.coords">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">coords</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Polyline coordinates for the graphics element.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.fgcolor">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">fgcolor</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.fgcolor" title="Permalink to this definition">¶</a></dt>
<dd><p>Foreground color.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.bgcolor">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">bgcolor</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.bgcolor" title="Permalink to this definition">¶</a></dt>
<dd><p>Background color.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.element">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">element</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.element" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Graphics as a valid KGML element.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.bounds">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">bounds</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Coordinate bounds for the Graphics element.</p>
<p>Return the bounds of the Graphics object as an [(xmin, ymin),
(xmax, ymax)] tuple.  Co-ordinates give the centre of the
circle, rectangle, roundrectangle elements, so we have to
adjust for the relevant width/height.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Graphics.centre">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">centre</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Graphics.centre" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the centre of the Graphics object as an (x, y) tuple.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.KEGG.KGML.KGML_pathway.</span></code><code class="sig-name descname"><span class="pre">Reaction</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A specific chemical reaction with substrates and products.</p>
<p>This describes a specific chemical reaction between one or more
substrates and one or more products.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>id             Pathway graph node ID of the entry</p></li>
<li><p>names          List of KEGG identifier(s) from the REACTION database</p></li>
<li><p>type           String: reversible or irreversible</p></li>
<li><p>substrate      Entry object of the substrate</p></li>
<li><p>product        Entry object of the product</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an informative human-readable string.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.add_substrate">
<code class="sig-name descname"><span class="pre">add_substrate</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substrate_id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.add_substrate" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a substrate, identified by its node ID, to the reaction.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.add_product">
<code class="sig-name descname"><span class="pre">add_product</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">product_id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.add_product" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a product, identified by its node ID, to the reaction.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.id">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">id</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.id" title="Permalink to this definition">¶</a></dt>
<dd><p>Node ID for the reaction.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.name">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">name</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.name" title="Permalink to this definition">¶</a></dt>
<dd><p>List of KEGG identifiers for the reaction.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.substrates">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">substrates</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.substrates" title="Permalink to this definition">¶</a></dt>
<dd><p>Return list of substrate Entry elements.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.products">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">products</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.products" title="Permalink to this definition">¶</a></dt>
<dd><p>Return list of product Entry elements.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.entry">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">entry</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.entry" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Entry corresponding to this reaction.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.reactant_ids">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">reactant_ids</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.reactant_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of substrate and product reactant IDs.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Reaction.element">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">element</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Reaction.element" title="Permalink to this definition">¶</a></dt>
<dd><p>Return KGML element describing the Reaction.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.KEGG.KGML.KGML_pathway.Relation">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.KEGG.KGML.KGML_pathway.</span></code><code class="sig-name descname"><span class="pre">Relation</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Relation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A relationship between to products, KOs, or protein and compound.</p>
<p>This describes a relationship between two products, KOs, or protein
and compound, as described in release KGML v0.7.2
(<a class="reference external" href="http://www.kegg.jp/kegg/xml/docs/">http://www.kegg.jp/kegg/xml/docs/</a>)</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>entry1 - The first Entry object node ID defining the
relation (int)</p></li>
<li><p>entry2 - The second Entry object node ID defining the
relation (int)</p></li>
<li><p>type - The relation type</p></li>
<li><p>subtypes - List of subtypes for the relation, as a list of
(name, value) tuples</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Relation.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Relation.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Relation.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Relation.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a useful human-readable string.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Relation.entry1">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">entry1</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Relation.entry1" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry1 of the relation.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Relation.entry2">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">entry2</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Relation.entry2" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry2 of the relation.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.KEGG.KGML.KGML_pathway.Relation.element">
<em class="property"><span class="pre">property</span> </em><code class="sig-name descname"><span class="pre">element</span></code><a class="headerlink" href="#Bio.KEGG.KGML.KGML_pathway.Relation.element" title="Permalink to this definition">¶</a></dt>
<dd><p>Return KGML element describing the Relation.</p>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="Bio.KEGG.Map.html" class="btn btn-neutral float-right" title="Bio.KEGG.Map package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="Bio.KEGG.KGML.KGML_parser.html" class="btn btn-neutral float-left" title="Bio.KEGG.KGML.KGML_parser module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 1999-2020, The Biopython Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.79.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.77/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>