

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Align.Applications package &mdash; Biopython 1.76.dev0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Align.substitution_matrices package" href="Bio.Align.substitution_matrices.html" />
    <link rel="prev" title="Bio.Align package" href="Bio.Align.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.76.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Align.html#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html#module-Bio.Align">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.Align.html">Bio.Align package</a> &raquo;</li>
        
      <li>Bio.Align.Applications package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Align/Applications/__init__.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Align.substitution_matrices.html" class="btn btn-neutral float-right" title="Bio.Align.substitution_matrices package" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Align.html" class="btn btn-neutral float-left" title="Bio.Align package" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="bio-align-applications-package">
<h1>Bio.Align.Applications package<a class="headerlink" href="#bio-align-applications-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-Bio.Align.Applications">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.Align.Applications" title="Permalink to this headline">¶</a></h2>
<p>Alignment command line tool wrappers.</p>
<dl class="class">
<dt id="Bio.Align.Applications.MuscleCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">MuscleCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='muscle'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for the multiple alignment program MUSCLE.</p>
<p><a class="reference external" href="http://www.drive5.com/muscle/">http://www.drive5.com/muscle/</a></p>
<p class="rubric">Notes</p>
<p>Last checked against version: 3.7, briefly against 3.8</p>
<p class="rubric">References</p>
<p>Edgar, Robert C. (2004), MUSCLE: multiple sequence alignment with high
accuracy and high throughput, Nucleic Acids Research 32(5), 1792-97.</p>
<p>Edgar, R.C. (2004) MUSCLE: a multiple sequence alignment method with
reduced time and space complexity. BMC Bioinformatics 5(1): 113.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">MuscleCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">muscle_exe</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;C:\Program Files\Aligments\muscle3.8.31_i86win32.exe&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">in_file</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;C:\My Documents\unaligned.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out_file</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;C:\My Documents\aligned.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">muscle_cline</span> <span class="o">=</span> <span class="n">MuscleCommandline</span><span class="p">(</span><span class="n">muscle_exe</span><span class="p">,</span> <span class="nb">input</span><span class="o">=</span><span class="n">in_file</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="n">out_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">muscle_cline</span><span class="p">)</span>
<span class="go">&quot;C:\Program Files\Aligments\muscle3.8.31_i86win32.exe&quot; -in &quot;C:\My Documents\unaligned.fasta&quot; -out &quot;C:\My Documents\aligned.fasta&quot;</span>
</pre></div>
</div>
<p>You would typically run the command line with muscle_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.anchors">
<em class="property">property </em><code class="sig-name descname">anchors</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.anchors" title="Permalink to this definition">¶</a></dt>
<dd><p>Use anchor optimisation in tree dependent refinement iterations</p>
<p>This property controls the addition of the -anchors switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.anchorspacing">
<em class="property">property </em><code class="sig-name descname">anchorspacing</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.anchorspacing" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum spacing between anchor columns</p>
<p>This controls the addition of the -anchorspacing parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.brenner">
<em class="property">property </em><code class="sig-name descname">brenner</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.brenner" title="Permalink to this definition">¶</a></dt>
<dd><p>Use Steve Brenner’s root alignment method</p>
<p>This property controls the addition of the -brenner switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.center">
<em class="property">property </em><code class="sig-name descname">center</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.center" title="Permalink to this definition">¶</a></dt>
<dd><p>Center parameter - should be negative</p>
<p>This controls the addition of the -center parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.cluster">
<em class="property">property </em><code class="sig-name descname">cluster</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform fast clustering of input sequences, use -tree1 to save tree</p>
<p>This property controls the addition of the -cluster switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.cluster1">
<em class="property">property </em><code class="sig-name descname">cluster1</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.cluster1" title="Permalink to this definition">¶</a></dt>
<dd><p>Clustering method used in iteration 1</p>
<p>This controls the addition of the -cluster1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.cluster2">
<em class="property">property </em><code class="sig-name descname">cluster2</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.cluster2" title="Permalink to this definition">¶</a></dt>
<dd><p>Clustering method used in iteration 2</p>
<p>This controls the addition of the -cluster2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.clw">
<em class="property">property </em><code class="sig-name descname">clw</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.clw" title="Permalink to this definition">¶</a></dt>
<dd><p>Write output in CLUSTALW format (with a MUSCLE header)</p>
<p>This property controls the addition of the -clw switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.clwout">
<em class="property">property </em><code class="sig-name descname">clwout</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.clwout" title="Permalink to this definition">¶</a></dt>
<dd><p>Write CLUSTALW output (with MUSCLE header) to specified filename</p>
<p>This controls the addition of the -clwout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.clwstrict">
<em class="property">property </em><code class="sig-name descname">clwstrict</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.clwstrict" title="Permalink to this definition">¶</a></dt>
<dd><p>Write output in CLUSTALW format with version1.81 header</p>
<p>This property controls the addition of the -clwstrict switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.clwstrictout">
<em class="property">property </em><code class="sig-name descname">clwstrictout</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.clwstrictout" title="Permalink to this definition">¶</a></dt>
<dd><p>Write CLUSTALW output (with version 1.81 header) to specified filename</p>
<p>This controls the addition of the -clwstrictout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.core">
<em class="property">property </em><code class="sig-name descname">core</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.core" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not catch exceptions</p>
<p>This property controls the addition of the -core switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.diagbreak">
<em class="property">property </em><code class="sig-name descname">diagbreak</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.diagbreak" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum distance between two diagonals that allows them to merge into one diagonal</p>
<p>This controls the addition of the -diagbreak parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.diaglength">
<em class="property">property </em><code class="sig-name descname">diaglength</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.diaglength" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum length of diagonal</p>
<p>This controls the addition of the -diaglength parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.diagmargin">
<em class="property">property </em><code class="sig-name descname">diagmargin</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.diagmargin" title="Permalink to this definition">¶</a></dt>
<dd><p>Discard this many positions at ends of diagonal</p>
<p>This controls the addition of the -diagmargin parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.diags">
<em class="property">property </em><code class="sig-name descname">diags</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.diags" title="Permalink to this definition">¶</a></dt>
<dd><p>Find diagonals (faster for similar sequences)</p>
<p>This property controls the addition of the -diags switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.dimer">
<em class="property">property </em><code class="sig-name descname">dimer</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.dimer" title="Permalink to this definition">¶</a></dt>
<dd><p>Use faster (slightly less accurate) dimer approximationfor the SP score</p>
<p>This property controls the addition of the -dimer switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.distance1">
<em class="property">property </em><code class="sig-name descname">distance1</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.distance1" title="Permalink to this definition">¶</a></dt>
<dd><p>Distance measure for iteration 1</p>
<p>This controls the addition of the -distance1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.distance2">
<em class="property">property </em><code class="sig-name descname">distance2</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.distance2" title="Permalink to this definition">¶</a></dt>
<dd><p>Distance measure for iteration 2</p>
<p>This controls the addition of the -distance2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.fasta">
<em class="property">property </em><code class="sig-name descname">fasta</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Write output in FASTA format</p>
<p>This property controls the addition of the -fasta switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.fastaout">
<em class="property">property </em><code class="sig-name descname">fastaout</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.fastaout" title="Permalink to this definition">¶</a></dt>
<dd><p>Write FASTA format output to specified filename</p>
<p>This controls the addition of the -fastaout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.gapextend">
<em class="property">property </em><code class="sig-name descname">gapextend</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.gapextend" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap extension penalty</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.gapopen">
<em class="property">property </em><code class="sig-name descname">gapopen</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.gapopen" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap open score - negative number</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.group">
<em class="property">property </em><code class="sig-name descname">group</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.group" title="Permalink to this definition">¶</a></dt>
<dd><p>Group similar sequences in output</p>
<p>This property controls the addition of the -group switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.html">
<em class="property">property </em><code class="sig-name descname">html</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.html" title="Permalink to this definition">¶</a></dt>
<dd><p>Write output in HTML format</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.htmlout">
<em class="property">property </em><code class="sig-name descname">htmlout</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.htmlout" title="Permalink to this definition">¶</a></dt>
<dd><p>Write HTML output to specified filename</p>
<p>This controls the addition of the -htmlout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.hydro">
<em class="property">property </em><code class="sig-name descname">hydro</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.hydro" title="Permalink to this definition">¶</a></dt>
<dd><p>Window size for hydrophobic region</p>
<p>This controls the addition of the -hydro parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.hydrofactor">
<em class="property">property </em><code class="sig-name descname">hydrofactor</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.hydrofactor" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiplier for gap penalties in hydrophobic regions</p>
<p>This controls the addition of the -hydrofactor parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.in1">
<em class="property">property </em><code class="sig-name descname">in1</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.in1" title="Permalink to this definition">¶</a></dt>
<dd><p>First input filename for profile alignment</p>
<p>This controls the addition of the -in1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.in2">
<em class="property">property </em><code class="sig-name descname">in2</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.in2" title="Permalink to this definition">¶</a></dt>
<dd><p>Second input filename for a profile alignment</p>
<p>This controls the addition of the -in2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.input">
<em class="property">property </em><code class="sig-name descname">input</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.input" title="Permalink to this definition">¶</a></dt>
<dd><p>Input filename</p>
<p>This controls the addition of the -in parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.le">
<em class="property">property </em><code class="sig-name descname">le</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.le" title="Permalink to this definition">¶</a></dt>
<dd><p>Use log-expectation profile score (VTML240)</p>
<p>This property controls the addition of the -le switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.log">
<em class="property">property </em><code class="sig-name descname">log</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.log" title="Permalink to this definition">¶</a></dt>
<dd><p>Log file name</p>
<p>This controls the addition of the -log parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.loga">
<em class="property">property </em><code class="sig-name descname">loga</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.loga" title="Permalink to this definition">¶</a></dt>
<dd><p>Log file name (append to existing file)</p>
<p>This controls the addition of the -loga parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.matrix">
<em class="property">property </em><code class="sig-name descname">matrix</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>path to NCBI or WU-BLAST format protein substitution matrix - also set -gapopen, -gapextend and -center</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.maxdiagbreak">
<em class="property">property </em><code class="sig-name descname">maxdiagbreak</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.maxdiagbreak" title="Permalink to this definition">¶</a></dt>
<dd><p>Deprecated in v3.8, use -diagbreak instead.</p>
<p>This controls the addition of the -maxdiagbreak parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.maxhours">
<em class="property">property </em><code class="sig-name descname">maxhours</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.maxhours" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum time to run in hours</p>
<p>This controls the addition of the -maxhours parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.maxiters">
<em class="property">property </em><code class="sig-name descname">maxiters</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.maxiters" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum number of iterations</p>
<p>This controls the addition of the -maxiters parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.maxtrees">
<em class="property">property </em><code class="sig-name descname">maxtrees</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.maxtrees" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum number of trees to build in iteration 2</p>
<p>This controls the addition of the -maxtrees parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.minbestcolscore">
<em class="property">property </em><code class="sig-name descname">minbestcolscore</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.minbestcolscore" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum score a column must have to be an anchor</p>
<p>This controls the addition of the -minbestcolscore parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.minsmoothscore">
<em class="property">property </em><code class="sig-name descname">minsmoothscore</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.minsmoothscore" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum smoothed score a column must have to be an anchor</p>
<p>This controls the addition of the -minsmoothscore parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.msf">
<em class="property">property </em><code class="sig-name descname">msf</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.msf" title="Permalink to this definition">¶</a></dt>
<dd><p>Write output in MSF format</p>
<p>This property controls the addition of the -msf switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.msfout">
<em class="property">property </em><code class="sig-name descname">msfout</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.msfout" title="Permalink to this definition">¶</a></dt>
<dd><p>Write MSF format output to specified filename</p>
<p>This controls the addition of the -msfout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.noanchors">
<em class="property">property </em><code class="sig-name descname">noanchors</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.noanchors" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not use anchor optimisation in tree dependent refinement iterations</p>
<p>This property controls the addition of the -noanchors switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.nocore">
<em class="property">property </em><code class="sig-name descname">nocore</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.nocore" title="Permalink to this definition">¶</a></dt>
<dd><p>Catch exceptions</p>
<p>This property controls the addition of the -nocore switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.objscore">
<em class="property">property </em><code class="sig-name descname">objscore</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.objscore" title="Permalink to this definition">¶</a></dt>
<dd><p>Objective score used by tree dependent refinement</p>
<p>This controls the addition of the -objscore parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.out">
<em class="property">property </em><code class="sig-name descname">out</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.out" title="Permalink to this definition">¶</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.phyi">
<em class="property">property </em><code class="sig-name descname">phyi</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.phyi" title="Permalink to this definition">¶</a></dt>
<dd><p>Write output in PHYLIP interleaved format</p>
<p>This property controls the addition of the -phyi switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.phyiout">
<em class="property">property </em><code class="sig-name descname">phyiout</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.phyiout" title="Permalink to this definition">¶</a></dt>
<dd><p>Write PHYLIP interleaved output to specified filename</p>
<p>This controls the addition of the -phyiout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.phys">
<em class="property">property </em><code class="sig-name descname">phys</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.phys" title="Permalink to this definition">¶</a></dt>
<dd><p>Write output in PHYLIP sequential format</p>
<p>This property controls the addition of the -phys switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.physout">
<em class="property">property </em><code class="sig-name descname">physout</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.physout" title="Permalink to this definition">¶</a></dt>
<dd><p>Write PHYLIP sequential format to specified filename</p>
<p>This controls the addition of the -physout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.profile">
<em class="property">property </em><code class="sig-name descname">profile</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.profile" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform a profile alignment</p>
<p>This property controls the addition of the -profile switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.quiet">
<em class="property">property </em><code class="sig-name descname">quiet</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.quiet" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not display progress messages</p>
<p>This property controls the addition of the -quiet switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.refine">
<em class="property">property </em><code class="sig-name descname">refine</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.refine" title="Permalink to this definition">¶</a></dt>
<dd><p>Only do tree dependent refinement</p>
<p>This property controls the addition of the -refine switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.refinew">
<em class="property">property </em><code class="sig-name descname">refinew</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.refinew" title="Permalink to this definition">¶</a></dt>
<dd><p>Only do tree dependent refinement using sliding window approach</p>
<p>This property controls the addition of the -refinew switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.refinewindow">
<em class="property">property </em><code class="sig-name descname">refinewindow</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.refinewindow" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of window for -refinew</p>
<p>This controls the addition of the -refinewindow parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.root1">
<em class="property">property </em><code class="sig-name descname">root1</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.root1" title="Permalink to this definition">¶</a></dt>
<dd><p>Method used to root tree in iteration 1</p>
<p>This controls the addition of the -root1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.root2">
<em class="property">property </em><code class="sig-name descname">root2</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.root2" title="Permalink to this definition">¶</a></dt>
<dd><p>Method used to root tree in iteration 2</p>
<p>This controls the addition of the -root2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.scorefile">
<em class="property">property </em><code class="sig-name descname">scorefile</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.scorefile" title="Permalink to this definition">¶</a></dt>
<dd><p>Score file name, contains one line for each column in the alignment with average BLOSUM62 score</p>
<p>This controls the addition of the -scorefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.seqtype">
<em class="property">property </em><code class="sig-name descname">seqtype</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.seqtype" title="Permalink to this definition">¶</a></dt>
<dd><p>Sequence type</p>
<p>This controls the addition of the -seqtype parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.smoothscoreceil">
<em class="property">property </em><code class="sig-name descname">smoothscoreceil</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.smoothscoreceil" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum value of column score for smoothing</p>
<p>This controls the addition of the -smoothscoreceil parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.smoothwindow">
<em class="property">property </em><code class="sig-name descname">smoothwindow</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.smoothwindow" title="Permalink to this definition">¶</a></dt>
<dd><p>Window used for anchor column smoothing</p>
<p>This controls the addition of the -smoothwindow parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.sp">
<em class="property">property </em><code class="sig-name descname">sp</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.sp" title="Permalink to this definition">¶</a></dt>
<dd><p>Use sum-of-pairs protein profile score (PAM200)</p>
<p>This property controls the addition of the -sp switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.spn">
<em class="property">property </em><code class="sig-name descname">spn</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.spn" title="Permalink to this definition">¶</a></dt>
<dd><p>Use sum-of-pairs protein nucleotide profile score</p>
<p>This property controls the addition of the -spn switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.spscore">
<em class="property">property </em><code class="sig-name descname">spscore</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.spscore" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute SP objective score of multiple alignment</p>
<p>This controls the addition of the -spscore parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.stable">
<em class="property">property </em><code class="sig-name descname">stable</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.stable" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not group similar sequences in output (not supported in v3.8)</p>
<p>This property controls the addition of the -stable switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.sueff">
<em class="property">property </em><code class="sig-name descname">sueff</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.sueff" title="Permalink to this definition">¶</a></dt>
<dd><p>Constant used in UPGMB clustering</p>
<p>This controls the addition of the -sueff parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.sv">
<em class="property">property </em><code class="sig-name descname">sv</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.sv" title="Permalink to this definition">¶</a></dt>
<dd><p>Use sum-of-pairs profile score (VTML240)</p>
<p>This property controls the addition of the -sv switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.tree1">
<em class="property">property </em><code class="sig-name descname">tree1</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.tree1" title="Permalink to this definition">¶</a></dt>
<dd><p>Save Newick tree from iteration 1</p>
<p>This controls the addition of the -tree1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.tree2">
<em class="property">property </em><code class="sig-name descname">tree2</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.tree2" title="Permalink to this definition">¶</a></dt>
<dd><p>Save Newick tree from iteration 2</p>
<p>This controls the addition of the -tree2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.usetree">
<em class="property">property </em><code class="sig-name descname">usetree</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.usetree" title="Permalink to this definition">¶</a></dt>
<dd><p>Use given Newick tree as guide tree</p>
<p>This controls the addition of the -usetree parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.verbose">
<em class="property">property </em><code class="sig-name descname">verbose</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.verbose" title="Permalink to this definition">¶</a></dt>
<dd><p>Write parameter settings and progress</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.version">
<em class="property">property </em><code class="sig-name descname">version</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.version" title="Permalink to this definition">¶</a></dt>
<dd><p>Write version string to stdout and exit</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.weight1">
<em class="property">property </em><code class="sig-name descname">weight1</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.weight1" title="Permalink to this definition">¶</a></dt>
<dd><p>Weighting scheme used in iteration 1</p>
<p>This controls the addition of the -weight1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MuscleCommandline.weight2">
<em class="property">property </em><code class="sig-name descname">weight2</code><a class="headerlink" href="#Bio.Align.Applications.MuscleCommandline.weight2" title="Permalink to this definition">¶</a></dt>
<dd><p>Weighting scheme used in iteration 2</p>
<p>This controls the addition of the -weight2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Align.Applications.ClustalwCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">ClustalwCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='clustalw'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for clustalw (version one or two).</p>
<p><a class="reference external" href="http://www.clustal.org/">http://www.clustal.org/</a></p>
<p class="rubric">Notes</p>
<p>Last checked against versions: 1.83 and 2.1</p>
<p class="rubric">References</p>
<p>Larkin MA, Blackshields G, Brown NP, Chenna R, McGettigan PA,
McWilliam H, Valentin F, Wallace IM, Wilm A, Lopez R, Thompson JD,
Gibson TJ, Higgins DG. (2007). Clustal W and Clustal X version 2.0.
Bioinformatics, 23, 2947-2948.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">ClustalwCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">in_file</span> <span class="o">=</span> <span class="s2">&quot;unaligned.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clustalw_cline</span> <span class="o">=</span> <span class="n">ClustalwCommandline</span><span class="p">(</span><span class="s2">&quot;clustalw2&quot;</span><span class="p">,</span> <span class="n">infile</span><span class="o">=</span><span class="n">in_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">clustalw_cline</span><span class="p">)</span>
<span class="go">clustalw2 -infile=unaligned.fasta</span>
</pre></div>
</div>
<p>You would typically run the command line with clustalw_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.align">
<em class="property">property </em><code class="sig-name descname">align</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.align" title="Permalink to this definition">¶</a></dt>
<dd><p>Do full multiple alignment.</p>
<p>This property controls the addition of the -align switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.bootlabels">
<em class="property">property </em><code class="sig-name descname">bootlabels</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.bootlabels" title="Permalink to this definition">¶</a></dt>
<dd><p>Node OR branch position of bootstrap values in tree display</p>
<p>This controls the addition of the -bootlabels parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.bootstrap">
<em class="property">property </em><code class="sig-name descname">bootstrap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.bootstrap" title="Permalink to this definition">¶</a></dt>
<dd><p>Bootstrap a NJ tree (n= number of bootstraps; def. = 1000).</p>
<p>This controls the addition of the -bootstrap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.case">
<em class="property">property </em><code class="sig-name descname">case</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.case" title="Permalink to this definition">¶</a></dt>
<dd><p>LOWER or UPPER (for GDE output only)</p>
<p>This controls the addition of the -case parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.check">
<em class="property">property </em><code class="sig-name descname">check</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.check" title="Permalink to this definition">¶</a></dt>
<dd><p>Outline the command line params.</p>
<p>This property controls the addition of the -check switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.clustering">
<em class="property">property </em><code class="sig-name descname">clustering</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.clustering" title="Permalink to this definition">¶</a></dt>
<dd><p>NJ or UPGMA</p>
<p>This controls the addition of the -clustering parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.convert">
<em class="property">property </em><code class="sig-name descname">convert</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.convert" title="Permalink to this definition">¶</a></dt>
<dd><p>Output the input sequences in a different file format.</p>
<p>This property controls the addition of the -convert switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.dnamatrix">
<em class="property">property </em><code class="sig-name descname">dnamatrix</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.dnamatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>DNA weight matrix=IUB, CLUSTALW or filename</p>
<p>This controls the addition of the -dnamatrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.endgaps">
<em class="property">property </em><code class="sig-name descname">endgaps</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.endgaps" title="Permalink to this definition">¶</a></dt>
<dd><p>No end gap separation pen.</p>
<p>This property controls the addition of the -endgaps switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.fullhelp">
<em class="property">property </em><code class="sig-name descname">fullhelp</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.fullhelp" title="Permalink to this definition">¶</a></dt>
<dd><p>Output full help content.</p>
<p>This property controls the addition of the -fullhelp switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.gapdist">
<em class="property">property </em><code class="sig-name descname">gapdist</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.gapdist" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap separation pen. range</p>
<p>This controls the addition of the -gapdist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.gapext">
<em class="property">property </em><code class="sig-name descname">gapext</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.gapext" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap extension penalty</p>
<p>This controls the addition of the -gapext parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.gapopen">
<em class="property">property </em><code class="sig-name descname">gapopen</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.gapopen" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap opening penalty</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.helixendin">
<em class="property">property </em><code class="sig-name descname">helixendin</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.helixendin" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of residues inside helix to be treated as terminal</p>
<p>This controls the addition of the -helixendin parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.helixendout">
<em class="property">property </em><code class="sig-name descname">helixendout</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.helixendout" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of residues outside helix to be treated as terminal</p>
<p>This controls the addition of the -helixendout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.helixgap">
<em class="property">property </em><code class="sig-name descname">helixgap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.helixgap" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap penalty for helix core residues</p>
<p>This controls the addition of the -helixgap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.help">
<em class="property">property </em><code class="sig-name descname">help</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.help" title="Permalink to this definition">¶</a></dt>
<dd><p>Outline the command line params.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.hgapresidues">
<em class="property">property </em><code class="sig-name descname">hgapresidues</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.hgapresidues" title="Permalink to this definition">¶</a></dt>
<dd><p>List hydrophilic res.</p>
<p>This property controls the addition of the -hgapresidues switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.infile">
<em class="property">property </em><code class="sig-name descname">infile</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.infile" title="Permalink to this definition">¶</a></dt>
<dd><p>Input sequences.</p>
<p>This controls the addition of the -infile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.iteration">
<em class="property">property </em><code class="sig-name descname">iteration</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.iteration" title="Permalink to this definition">¶</a></dt>
<dd><p>NONE or TREE or ALIGNMENT</p>
<p>This controls the addition of the -iteration parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.kimura">
<em class="property">property </em><code class="sig-name descname">kimura</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.kimura" title="Permalink to this definition">¶</a></dt>
<dd><p>Use Kimura’s correction.</p>
<p>This property controls the addition of the -kimura switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.ktuple">
<em class="property">property </em><code class="sig-name descname">ktuple</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.ktuple" title="Permalink to this definition">¶</a></dt>
<dd><p>Word size</p>
<p>This controls the addition of the -ktuple parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.loopgap">
<em class="property">property </em><code class="sig-name descname">loopgap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.loopgap" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap penalty for loop regions</p>
<p>This controls the addition of the -loopgap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.matrix">
<em class="property">property </em><code class="sig-name descname">matrix</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Protein weight matrix=BLOSUM, PAM, GONNET, ID or filename</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.maxdiv">
<em class="property">property </em><code class="sig-name descname">maxdiv</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.maxdiv" title="Permalink to this definition">¶</a></dt>
<dd><p>% ident. for delay</p>
<p>This controls the addition of the -maxdiv parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.maxseqlen">
<em class="property">property </em><code class="sig-name descname">maxseqlen</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.maxseqlen" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum allowed input sequence length</p>
<p>This controls the addition of the -maxseqlen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.negative">
<em class="property">property </em><code class="sig-name descname">negative</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.negative" title="Permalink to this definition">¶</a></dt>
<dd><p>Protein alignment with negative values in matrix</p>
<p>This property controls the addition of the -negative switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.newtree">
<em class="property">property </em><code class="sig-name descname">newtree</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.newtree" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file name for newly created guide tree</p>
<p>This controls the addition of the -newtree parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.newtree1">
<em class="property">property </em><code class="sig-name descname">newtree1</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.newtree1" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file name for new guide tree of profile1</p>
<p>This controls the addition of the -newtree1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.newtree2">
<em class="property">property </em><code class="sig-name descname">newtree2</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.newtree2" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file for new guide tree of profile2</p>
<p>This controls the addition of the -newtree2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.nohgap">
<em class="property">property </em><code class="sig-name descname">nohgap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.nohgap" title="Permalink to this definition">¶</a></dt>
<dd><p>Hydrophilic gaps off</p>
<p>This property controls the addition of the -nohgap switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.nopgap">
<em class="property">property </em><code class="sig-name descname">nopgap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.nopgap" title="Permalink to this definition">¶</a></dt>
<dd><p>Residue-specific gaps off</p>
<p>This property controls the addition of the -nopgap switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.nosecstr1">
<em class="property">property </em><code class="sig-name descname">nosecstr1</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.nosecstr1" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not use secondary structure-gap penalty mask for profile 1</p>
<p>This property controls the addition of the -nosecstr1 switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.nosecstr2">
<em class="property">property </em><code class="sig-name descname">nosecstr2</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.nosecstr2" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not use secondary structure-gap penalty mask for profile 2</p>
<p>This property controls the addition of the -nosecstr2 switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.noweights">
<em class="property">property </em><code class="sig-name descname">noweights</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.noweights" title="Permalink to this definition">¶</a></dt>
<dd><p>Disable sequence weighting</p>
<p>This property controls the addition of the -noweights switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.numiter">
<em class="property">property </em><code class="sig-name descname">numiter</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.numiter" title="Permalink to this definition">¶</a></dt>
<dd><p>maximum number of iterations to perform</p>
<p>This controls the addition of the -numiter parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.options">
<em class="property">property </em><code class="sig-name descname">options</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.options" title="Permalink to this definition">¶</a></dt>
<dd><p>List the command line parameters</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.outfile">
<em class="property">property </em><code class="sig-name descname">outfile</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.outfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Output sequence alignment file name</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.outorder">
<em class="property">property </em><code class="sig-name descname">outorder</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.outorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Output taxon order: INPUT or ALIGNED</p>
<p>This controls the addition of the -outorder parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.output">
<em class="property">property </em><code class="sig-name descname">output</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.output" title="Permalink to this definition">¶</a></dt>
<dd><p>Output format: CLUSTAL(default), GCG, GDE, PHYLIP, PIR, NEXUS and FASTA</p>
<p>This controls the addition of the -output parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.outputtree">
<em class="property">property </em><code class="sig-name descname">outputtree</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.outputtree" title="Permalink to this definition">¶</a></dt>
<dd><p>nj OR phylip OR dist OR nexus</p>
<p>This controls the addition of the -outputtree parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.pairgap">
<em class="property">property </em><code class="sig-name descname">pairgap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.pairgap" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap penalty</p>
<p>This controls the addition of the -pairgap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.pim">
<em class="property">property </em><code class="sig-name descname">pim</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.pim" title="Permalink to this definition">¶</a></dt>
<dd><p>Output percent identity matrix (while calculating the tree).</p>
<p>This property controls the addition of the -pim switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.profile">
<em class="property">property </em><code class="sig-name descname">profile</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.profile" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge two alignments by profile alignment</p>
<p>This property controls the addition of the -profile switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.profile1">
<em class="property">property </em><code class="sig-name descname">profile1</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.profile1" title="Permalink to this definition">¶</a></dt>
<dd><p>Profiles (old alignment).</p>
<p>This controls the addition of the -profile1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.profile2">
<em class="property">property </em><code class="sig-name descname">profile2</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.profile2" title="Permalink to this definition">¶</a></dt>
<dd><p>Profiles (old alignment).</p>
<p>This controls the addition of the -profile2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.pwdnamatrix">
<em class="property">property </em><code class="sig-name descname">pwdnamatrix</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.pwdnamatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>DNA weight matrix=IUB, CLUSTALW or filename</p>
<p>This controls the addition of the -pwdnamatrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.pwgapext">
<em class="property">property </em><code class="sig-name descname">pwgapext</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.pwgapext" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap extension penalty</p>
<p>This controls the addition of the -pwgapext parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.pwgapopen">
<em class="property">property </em><code class="sig-name descname">pwgapopen</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.pwgapopen" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap opening penalty</p>
<p>This controls the addition of the -pwgapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.pwmatrix">
<em class="property">property </em><code class="sig-name descname">pwmatrix</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.pwmatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Protein weight matrix=BLOSUM, PAM, GONNET, ID or filename</p>
<p>This controls the addition of the -pwmatrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.quicktree">
<em class="property">property </em><code class="sig-name descname">quicktree</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.quicktree" title="Permalink to this definition">¶</a></dt>
<dd><p>Use FAST algorithm for the alignment guide tree</p>
<p>This property controls the addition of the -quicktree switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.quiet">
<em class="property">property </em><code class="sig-name descname">quiet</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.quiet" title="Permalink to this definition">¶</a></dt>
<dd><p>Reduce console output to minimum</p>
<p>This property controls the addition of the -quiet switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.range">
<em class="property">property </em><code class="sig-name descname">range</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.range" title="Permalink to this definition">¶</a></dt>
<dd><p>Sequence range to write starting m to m+n. Input as string eg. ‘24,200’</p>
<p>This controls the addition of the -range parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.score">
<em class="property">property </em><code class="sig-name descname">score</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.score" title="Permalink to this definition">¶</a></dt>
<dd><p>Either: PERCENT or ABSOLUTE</p>
<p>This controls the addition of the -score parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.secstrout">
<em class="property">property </em><code class="sig-name descname">secstrout</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.secstrout" title="Permalink to this definition">¶</a></dt>
<dd><p>STRUCTURE or MASK or BOTH or NONE output in alignment file</p>
<p>This controls the addition of the -secstrout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.seed">
<em class="property">property </em><code class="sig-name descname">seed</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Seed number for bootstraps.</p>
<p>This controls the addition of the -seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.seqno_range">
<em class="property">property </em><code class="sig-name descname">seqno_range</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.seqno_range" title="Permalink to this definition">¶</a></dt>
<dd><p>OFF or ON (NEW- for all output formats)</p>
<p>This controls the addition of the -seqno_range parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.seqnos">
<em class="property">property </em><code class="sig-name descname">seqnos</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.seqnos" title="Permalink to this definition">¶</a></dt>
<dd><p>OFF or ON (for Clustal output only)</p>
<p>This controls the addition of the -seqnos parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.sequences">
<em class="property">property </em><code class="sig-name descname">sequences</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Sequentially add profile2 sequences to profile1 alignment</p>
<p>This property controls the addition of the -sequences switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.stats">
<em class="property">property </em><code class="sig-name descname">stats</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Log some alignment statistics to file</p>
<p>This controls the addition of the -stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.strandendin">
<em class="property">property </em><code class="sig-name descname">strandendin</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.strandendin" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of residues inside strand to be treated as terminal</p>
<p>This controls the addition of the -strandendin parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.strandendout">
<em class="property">property </em><code class="sig-name descname">strandendout</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.strandendout" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of residues outside strand to be treated as terminal</p>
<p>This controls the addition of the -strandendout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.strandgap">
<em class="property">property </em><code class="sig-name descname">strandgap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.strandgap" title="Permalink to this definition">¶</a></dt>
<dd><p>gap penalty for strand core residues</p>
<p>This controls the addition of the -strandgap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.terminalgap">
<em class="property">property </em><code class="sig-name descname">terminalgap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.terminalgap" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap penalty for structure termini</p>
<p>This controls the addition of the -terminalgap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.topdiags">
<em class="property">property </em><code class="sig-name descname">topdiags</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.topdiags" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of best diags.</p>
<p>This controls the addition of the -topdiags parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.tossgaps">
<em class="property">property </em><code class="sig-name descname">tossgaps</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.tossgaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Ignore positions with gaps.</p>
<p>This property controls the addition of the -tossgaps switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.transweight">
<em class="property">property </em><code class="sig-name descname">transweight</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.transweight" title="Permalink to this definition">¶</a></dt>
<dd><p>Transitions weighting</p>
<p>This controls the addition of the -transweight parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.tree">
<em class="property">property </em><code class="sig-name descname">tree</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate NJ tree.</p>
<p>This property controls the addition of the -tree switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.type">
<em class="property">property </em><code class="sig-name descname">type</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.type" title="Permalink to this definition">¶</a></dt>
<dd><p>PROTEIN or DNA sequences</p>
<p>This controls the addition of the -type parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.usetree">
<em class="property">property </em><code class="sig-name descname">usetree</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.usetree" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of guide tree</p>
<p>This controls the addition of the -usetree parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.usetree1">
<em class="property">property </em><code class="sig-name descname">usetree1</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.usetree1" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of guide tree for profile1</p>
<p>This controls the addition of the -usetree1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.usetree2">
<em class="property">property </em><code class="sig-name descname">usetree2</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.usetree2" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of guide tree for profile2</p>
<p>This controls the addition of the -usetree2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalwCommandline.window">
<em class="property">property </em><code class="sig-name descname">window</code><a class="headerlink" href="#Bio.Align.Applications.ClustalwCommandline.window" title="Permalink to this definition">¶</a></dt>
<dd><p>Window around best diags.</p>
<p>This controls the addition of the -window parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">ClustalOmegaCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='clustalo'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for clustal omega.</p>
<p><a class="reference external" href="http://www.clustal.org/omega">http://www.clustal.org/omega</a></p>
<p class="rubric">Notes</p>
<p>Last checked against version: 1.2.0</p>
<p class="rubric">References</p>
<p>Sievers F, Wilm A, Dineen DG, Gibson TJ, Karplus K, Li W, Lopez R,
McWilliam H, Remmert M, Söding J, Thompson JD, Higgins DG (2011).
Fast, scalable generation of high-quality protein multiple
sequence alignments using Clustal Omega.
Molecular Systems Biology 7:539 <a class="reference external" href="https://doi.org/10.1038/msb.2011.75">https://doi.org/10.1038/msb.2011.75</a></p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">ClustalOmegaCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">in_file</span> <span class="o">=</span> <span class="s2">&quot;unaligned.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out_file</span> <span class="o">=</span> <span class="s2">&quot;aligned.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clustalomega_cline</span> <span class="o">=</span> <span class="n">ClustalOmegaCommandline</span><span class="p">(</span><span class="n">infile</span><span class="o">=</span><span class="n">in_file</span><span class="p">,</span> <span class="n">outfile</span><span class="o">=</span><span class="n">out_file</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auto</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">clustalomega_cline</span><span class="p">)</span>
<span class="go">clustalo -i unaligned.fasta -o aligned.fasta --auto -v</span>
</pre></div>
</div>
<p>You would typically run the command line with clustalomega_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.auto">
<em class="property">property </em><code class="sig-name descname">auto</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.auto" title="Permalink to this definition">¶</a></dt>
<dd><p>Set options automatically (might overwrite some of your options)</p>
<p>This property controls the addition of the –auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.clusteringout">
<em class="property">property </em><code class="sig-name descname">clusteringout</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.clusteringout" title="Permalink to this definition">¶</a></dt>
<dd><p>Clustering output file</p>
<p>This controls the addition of the –clustering-out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.clustersize">
<em class="property">property </em><code class="sig-name descname">clustersize</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.clustersize" title="Permalink to this definition">¶</a></dt>
<dd><p>soft maximum of sequences in sub-clusters</p>
<p>This controls the addition of the –cluster-size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.dealign">
<em class="property">property </em><code class="sig-name descname">dealign</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.dealign" title="Permalink to this definition">¶</a></dt>
<dd><p>Dealign input sequences</p>
<p>This property controls the addition of the –dealign switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.distmat_full">
<em class="property">property </em><code class="sig-name descname">distmat_full</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.distmat_full" title="Permalink to this definition">¶</a></dt>
<dd><p>Use full distance matrix for guide-tree calculation (slow; mBed is default)</p>
<p>This property controls the addition of the –full switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.distmat_full_iter">
<em class="property">property </em><code class="sig-name descname">distmat_full_iter</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.distmat_full_iter" title="Permalink to this definition">¶</a></dt>
<dd><p>Use full distance matrix for guide-tree calculation during iteration (mBed is default)</p>
<p>This property controls the addition of the –full-iter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.distmat_in">
<em class="property">property </em><code class="sig-name descname">distmat_in</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.distmat_in" title="Permalink to this definition">¶</a></dt>
<dd><p>Pairwise distance matrix input file (skips distance computation).</p>
<p>This controls the addition of the –distmat-in parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.distmat_out">
<em class="property">property </em><code class="sig-name descname">distmat_out</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.distmat_out" title="Permalink to this definition">¶</a></dt>
<dd><p>Pairwise distance matrix output file.</p>
<p>This controls the addition of the –distmat-out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.force">
<em class="property">property </em><code class="sig-name descname">force</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.force" title="Permalink to this definition">¶</a></dt>
<dd><p>Force file overwriting.</p>
<p>This property controls the addition of the –force switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.guidetree_in">
<em class="property">property </em><code class="sig-name descname">guidetree_in</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.guidetree_in" title="Permalink to this definition">¶</a></dt>
<dd><p>Guide tree input file (skips distance computation and guide-tree clustering step).</p>
<p>This controls the addition of the –guidetree-in parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.guidetree_out">
<em class="property">property </em><code class="sig-name descname">guidetree_out</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.guidetree_out" title="Permalink to this definition">¶</a></dt>
<dd><p>Guide tree output file.</p>
<p>This controls the addition of the –guidetree-out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.help">
<em class="property">property </em><code class="sig-name descname">help</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.help" title="Permalink to this definition">¶</a></dt>
<dd><p>Print help and exit.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.hmm_input">
<em class="property">property </em><code class="sig-name descname">hmm_input</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.hmm_input" title="Permalink to this definition">¶</a></dt>
<dd><p>HMM input files</p>
<p>This controls the addition of the –hmm-in parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.infile">
<em class="property">property </em><code class="sig-name descname">infile</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.infile" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiple sequence input file</p>
<p>This controls the addition of the -i parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.infmt">
<em class="property">property </em><code class="sig-name descname">infmt</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.infmt" title="Permalink to this definition">¶</a></dt>
<dd><p>Forced sequence input file format (default: auto)</p>
<blockquote>
<div><p>Allowed values: a2m, fa[sta], clu[stal], msf, phy[lip], selex, st[ockholm], vie[nna]</p>
</div></blockquote>
<p>This controls the addition of the –infmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.isprofile">
<em class="property">property </em><code class="sig-name descname">isprofile</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.isprofile" title="Permalink to this definition">¶</a></dt>
<dd><p>disable check if profile, force profile (default no)</p>
<p>This property controls the addition of the –is-profile switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.iterations">
<em class="property">property </em><code class="sig-name descname">iterations</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.iterations" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of (combined guide-tree/HMM) iterations</p>
<p>This controls the addition of the –iterations parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.log">
<em class="property">property </em><code class="sig-name descname">log</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.log" title="Permalink to this definition">¶</a></dt>
<dd><p>Log all non-essential output to this file.</p>
<p>This controls the addition of the -l parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.long_version">
<em class="property">property </em><code class="sig-name descname">long_version</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.long_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Print long version information and exit</p>
<p>This property controls the addition of the –long-version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.max_guidetree_iterations">
<em class="property">property </em><code class="sig-name descname">max_guidetree_iterations</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.max_guidetree_iterations" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum number of guidetree iterations</p>
<p>This controls the addition of the –max-guidetree-iterations parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.max_hmm_iterations">
<em class="property">property </em><code class="sig-name descname">max_hmm_iterations</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.max_hmm_iterations" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum number of HMM iterations</p>
<p>This controls the addition of the –max-hmm-iterations parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.maxnumseq">
<em class="property">property </em><code class="sig-name descname">maxnumseq</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.maxnumseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum allowed number of sequences</p>
<p>This controls the addition of the –maxnumseq parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.maxseqlen">
<em class="property">property </em><code class="sig-name descname">maxseqlen</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.maxseqlen" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum allowed sequence length</p>
<p>This controls the addition of the –maxseqlen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.outfile">
<em class="property">property </em><code class="sig-name descname">outfile</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.outfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiple sequence alignment output file (default: stdout).</p>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.outfmt">
<em class="property">property </em><code class="sig-name descname">outfmt</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.outfmt" title="Permalink to this definition">¶</a></dt>
<dd><p>MSA output file format: a2m=fa[sta],clu[stal],msf,phy[lip],selex,st[ockholm],vie[nna] (default: fasta).</p>
<p>This controls the addition of the –outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.outputorder">
<em class="property">property </em><code class="sig-name descname">outputorder</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.outputorder" title="Permalink to this definition">¶</a></dt>
<dd><p>MSA output order like in input/guide-tree</p>
<p>This controls the addition of the –output-order parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.percentid">
<em class="property">property </em><code class="sig-name descname">percentid</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.percentid" title="Permalink to this definition">¶</a></dt>
<dd><p>convert distances into percent identities (default no)</p>
<p>This property controls the addition of the –percent-id switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.profile1">
<em class="property">property </em><code class="sig-name descname">profile1</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.profile1" title="Permalink to this definition">¶</a></dt>
<dd><p>Pre-aligned multiple sequence file (aligned columns will be kept fix).</p>
<p>This controls the addition of the –profile1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.profile2">
<em class="property">property </em><code class="sig-name descname">profile2</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.profile2" title="Permalink to this definition">¶</a></dt>
<dd><p>Pre-aligned multiple sequence file (aligned columns will be kept fix).</p>
<p>This controls the addition of the –profile2 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.residuenumber">
<em class="property">property </em><code class="sig-name descname">residuenumber</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.residuenumber" title="Permalink to this definition">¶</a></dt>
<dd><p>in Clustal format print residue numbers (default no)</p>
<p>This property controls the addition of the –residuenumber switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.seqtype">
<em class="property">property </em><code class="sig-name descname">seqtype</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.seqtype" title="Permalink to this definition">¶</a></dt>
<dd><p>{Protein, RNA, DNA} Force a sequence type (default: auto).</p>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.threads">
<em class="property">property </em><code class="sig-name descname">threads</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.threads" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of processors to use</p>
<p>This controls the addition of the –threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.usekimura">
<em class="property">property </em><code class="sig-name descname">usekimura</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.usekimura" title="Permalink to this definition">¶</a></dt>
<dd><p>use Kimura distance correction for aligned sequences (default no)</p>
<p>This property controls the addition of the –use-kimura switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.verbose">
<em class="property">property </em><code class="sig-name descname">verbose</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.verbose" title="Permalink to this definition">¶</a></dt>
<dd><p>Verbose output</p>
<p>This property controls the addition of the -v switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.version">
<em class="property">property </em><code class="sig-name descname">version</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.version" title="Permalink to this definition">¶</a></dt>
<dd><p>Print version information and exit</p>
<p>This property controls the addition of the –version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ClustalOmegaCommandline.wrap">
<em class="property">property </em><code class="sig-name descname">wrap</code><a class="headerlink" href="#Bio.Align.Applications.ClustalOmegaCommandline.wrap" title="Permalink to this definition">¶</a></dt>
<dd><p>number of residues before line-wrap in output</p>
<p>This controls the addition of the –wrap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Align.Applications.PrankCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">PrankCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='prank'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for the multiple alignment program PRANK.</p>
<p><a class="reference external" href="http://www.ebi.ac.uk/goldman-srv/prank/prank/">http://www.ebi.ac.uk/goldman-srv/prank/prank/</a></p>
<p class="rubric">Notes</p>
<p>Last checked against version: 081202</p>
<p class="rubric">References</p>
<p>Loytynoja, A. and Goldman, N. 2005. An algorithm for progressive
multiple alignment of sequences with insertions. Proceedings of
the National Academy of Sciences, 102: 10557–10562.</p>
<p>Loytynoja, A. and Goldman, N. 2008. Phylogeny-aware gap placement
prevents errors in sequence alignment and evolutionary analysis.
Science, 320: 1632.</p>
<p class="rubric">Examples</p>
<p>To align a FASTA file (unaligned.fasta) with the output in aligned
FASTA format with the output filename starting with “aligned” (you
can’t pick the filename explicitly), no tree output and no XML output,
use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">PrankCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prank_cline</span> <span class="o">=</span> <span class="n">PrankCommandline</span><span class="p">(</span><span class="n">d</span><span class="o">=</span><span class="s2">&quot;unaligned.fasta&quot;</span><span class="p">,</span>
<span class="gp">... </span>                               <span class="n">o</span><span class="o">=</span><span class="s2">&quot;aligned&quot;</span><span class="p">,</span> <span class="c1"># prefix only!</span>
<span class="gp">... </span>                               <span class="n">f</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="c1"># FASTA output</span>
<span class="gp">... </span>                               <span class="n">notree</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">noxml</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">prank_cline</span><span class="p">)</span>
<span class="go">prank -d=unaligned.fasta -o=aligned -f=8 -noxml -notree</span>
</pre></div>
</div>
<p>You would typically run the command line with prank_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.F">
<em class="property">property </em><code class="sig-name descname">F</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.F" title="Permalink to this definition">¶</a></dt>
<dd><p>Force insertions to be always skipped: same as +F</p>
<p>This property controls the addition of the -F switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.codon">
<em class="property">property </em><code class="sig-name descname">codon</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.codon" title="Permalink to this definition">¶</a></dt>
<dd><p>Codon aware alignment or not</p>
<p>This property controls the addition of the -codon switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.convert">
<em class="property">property </em><code class="sig-name descname">convert</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.convert" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert input alignment to new format. Do not perform alignment</p>
<p>This property controls the addition of the -convert switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.d">
<em class="property">property </em><code class="sig-name descname">d</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.d" title="Permalink to this definition">¶</a></dt>
<dd><p>Input filename</p>
<p>This controls the addition of the -d parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.dnafreqs">
<em class="property">property </em><code class="sig-name descname">dnafreqs</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.dnafreqs" title="Permalink to this definition">¶</a></dt>
<dd><p>DNA frequencies - ‘A,C,G,T’. eg ‘25,25,25,25’ as a quote surrounded string value. Default: empirical</p>
<p>This controls the addition of the -dnafreqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.dots">
<em class="property">property </em><code class="sig-name descname">dots</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.dots" title="Permalink to this definition">¶</a></dt>
<dd><p>Show insertion gaps as dots</p>
<p>This property controls the addition of the -dots switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.f">
<em class="property">property </em><code class="sig-name descname">f</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.f" title="Permalink to this definition">¶</a></dt>
<dd><p>Output alignment format. Default: 8 FASTA
Option are:
1. IG/Stanford  8. Pearson/Fasta
2. GenBank/GB   11. Phylip3.2
3. NBRF         12. Phylip
4. EMBL         14. PIR/CODATA
6. DNAStrider   15. MSF
7. Fitch        17. PAUP/NEXUS</p>
<p>This controls the addition of the -f parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.fixedbranches">
<em class="property">property </em><code class="sig-name descname">fixedbranches</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.fixedbranches" title="Permalink to this definition">¶</a></dt>
<dd><p>Use fixed branch lengths of input value</p>
<p>This controls the addition of the -fixedbranches parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.gapext">
<em class="property">property </em><code class="sig-name descname">gapext</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.gapext" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap extension probability. Default: dna 0.5 / prot 0.5</p>
<p>This controls the addition of the -gapext parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.gaprate">
<em class="property">property </em><code class="sig-name descname">gaprate</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.gaprate" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap opening rate. Default: dna 0.025 prot 0.0025</p>
<p>This controls the addition of the -gaprate parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.kappa">
<em class="property">property </em><code class="sig-name descname">kappa</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.kappa" title="Permalink to this definition">¶</a></dt>
<dd><p>Transition/transversion ratio. Default: 2</p>
<p>This controls the addition of the -kappa parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.longseq">
<em class="property">property </em><code class="sig-name descname">longseq</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.longseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Save space in pairwise alignments</p>
<p>This property controls the addition of the -longseq switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.m">
<em class="property">property </em><code class="sig-name descname">m</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.m" title="Permalink to this definition">¶</a></dt>
<dd><p>User-defined alignment model filename. Default: HKY2/WAG</p>
<p>This controls the addition of the -m parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.matinitsize">
<em class="property">property </em><code class="sig-name descname">matinitsize</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.matinitsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix initial size multiplier</p>
<p>This controls the addition of the -matinitsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.matresize">
<em class="property">property </em><code class="sig-name descname">matresize</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.matresize" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix resizing multiplier</p>
<p>This controls the addition of the -matresize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.maxbranches">
<em class="property">property </em><code class="sig-name descname">maxbranches</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.maxbranches" title="Permalink to this definition">¶</a></dt>
<dd><p>Use maximum branch lengths of input value</p>
<p>This controls the addition of the -maxbranches parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.mttranslate">
<em class="property">property </em><code class="sig-name descname">mttranslate</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.mttranslate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate to protein using mt table</p>
<p>This property controls the addition of the -mttranslate switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.nopost">
<em class="property">property </em><code class="sig-name descname">nopost</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.nopost" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not compute posterior support. Default: compute</p>
<p>This property controls the addition of the -nopost switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.notree">
<em class="property">property </em><code class="sig-name descname">notree</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.notree" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not output dnd tree files (PRANK versions earlier than v.120626)</p>
<p>This property controls the addition of the -notree switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.noxml">
<em class="property">property </em><code class="sig-name descname">noxml</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.noxml" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not output XML files (PRANK versions earlier than v.120626)</p>
<p>This property controls the addition of the -noxml switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.o">
<em class="property">property </em><code class="sig-name descname">o</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.o" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Output filenames prefix. Default: ‘output’</dt><dd><p>Will write: output.?.fas (depending on requested format), output.?.xml and output.?.dnd</p>
</dd>
</dl>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.once">
<em class="property">property </em><code class="sig-name descname">once</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.once" title="Permalink to this definition">¶</a></dt>
<dd><p>Run only once. Default: twice if no guidetree given</p>
<p>This property controls the addition of the -once switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.printnodes">
<em class="property">property </em><code class="sig-name descname">printnodes</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.printnodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Output each node; mostly for debugging</p>
<p>This property controls the addition of the -printnodes switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.pwdist">
<em class="property">property </em><code class="sig-name descname">pwdist</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.pwdist" title="Permalink to this definition">¶</a></dt>
<dd><p>Expected pairwise distance for computing guidetree. Default: dna 0.25 / prot 0.5</p>
<p>This controls the addition of the -pwdist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.pwgenomic">
<em class="property">property </em><code class="sig-name descname">pwgenomic</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.pwgenomic" title="Permalink to this definition">¶</a></dt>
<dd><p>Do pairwise alignment, no guidetree</p>
<p>This property controls the addition of the -pwgenomic switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.pwgenomicdist">
<em class="property">property </em><code class="sig-name descname">pwgenomicdist</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.pwgenomicdist" title="Permalink to this definition">¶</a></dt>
<dd><p>Distance for pairwise alignment. Default: 0.3</p>
<p>This controls the addition of the -pwgenomicdist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.quiet">
<em class="property">property </em><code class="sig-name descname">quiet</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.quiet" title="Permalink to this definition">¶</a></dt>
<dd><p>Reduce verbosity</p>
<p>This property controls the addition of the -quiet switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.realbranches">
<em class="property">property </em><code class="sig-name descname">realbranches</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.realbranches" title="Permalink to this definition">¶</a></dt>
<dd><p>Disable branch length truncation</p>
<p>This property controls the addition of the -realbranches switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.rho">
<em class="property">property </em><code class="sig-name descname">rho</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.rho" title="Permalink to this definition">¶</a></dt>
<dd><p>Purine/pyrimidine ratio. Default: 1</p>
<p>This controls the addition of the -rho parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.scalebranches">
<em class="property">property </em><code class="sig-name descname">scalebranches</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.scalebranches" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale branch lengths. Default: dna 1 / prot 2</p>
<p>This controls the addition of the -scalebranches parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.shortnames">
<em class="property">property </em><code class="sig-name descname">shortnames</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.shortnames" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncate names at first space</p>
<p>This property controls the addition of the -shortnames switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.showtree">
<em class="property">property </em><code class="sig-name descname">showtree</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.showtree" title="Permalink to this definition">¶</a></dt>
<dd><p>Output dnd tree files (PRANK v.120626 and later)</p>
<p>This property controls the addition of the -showtree switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.showxml">
<em class="property">property </em><code class="sig-name descname">showxml</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.showxml" title="Permalink to this definition">¶</a></dt>
<dd><p>Output XML files (PRANK v.120626 and later)</p>
<p>This property controls the addition of the -showxml switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.skipins">
<em class="property">property </em><code class="sig-name descname">skipins</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.skipins" title="Permalink to this definition">¶</a></dt>
<dd><p>Skip insertions in posterior support</p>
<p>This property controls the addition of the -skipins switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.t">
<em class="property">property </em><code class="sig-name descname">t</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.t" title="Permalink to this definition">¶</a></dt>
<dd><p>Input guide tree filename</p>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.termgap">
<em class="property">property </em><code class="sig-name descname">termgap</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.termgap" title="Permalink to this definition">¶</a></dt>
<dd><p>Penalise terminal gaps normally</p>
<p>This property controls the addition of the -termgap switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.translate">
<em class="property">property </em><code class="sig-name descname">translate</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate to protein</p>
<p>This property controls the addition of the -translate switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.tree">
<em class="property">property </em><code class="sig-name descname">tree</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Input guide tree as Newick string</p>
<p>This controls the addition of the -tree parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.twice">
<em class="property">property </em><code class="sig-name descname">twice</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.twice" title="Permalink to this definition">¶</a></dt>
<dd><p>Always run twice</p>
<p>This property controls the addition of the -twice switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.uselogs">
<em class="property">property </em><code class="sig-name descname">uselogs</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.uselogs" title="Permalink to this definition">¶</a></dt>
<dd><p>Slower but should work for a greater number of sequences</p>
<p>This property controls the addition of the -uselogs switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.PrankCommandline.writeanc">
<em class="property">property </em><code class="sig-name descname">writeanc</code><a class="headerlink" href="#Bio.Align.Applications.PrankCommandline.writeanc" title="Permalink to this definition">¶</a></dt>
<dd><p>Output ancestral sequences</p>
<p>This property controls the addition of the -writeanc switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Align.Applications.MafftCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">MafftCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='mafft'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for the multiple alignment program MAFFT.</p>
<p><a class="reference external" href="http://align.bmr.kyushu-u.ac.jp/mafft/software/">http://align.bmr.kyushu-u.ac.jp/mafft/software/</a></p>
<p class="rubric">Notes</p>
<p>Last checked against version: MAFFT v6.717b (2009/12/03)</p>
<p class="rubric">References</p>
<p>Katoh, Toh (BMC Bioinformatics 9:212, 2008) Improved accuracy of
multiple ncRNA alignment by incorporating structural information into
a MAFFT-based framework (describes RNA structural alignment methods)</p>
<p>Katoh, Toh (Briefings in Bioinformatics 9:286-298, 2008) Recent
developments in the MAFFT multiple sequence alignment program
(outlines version 6)</p>
<p>Katoh, Toh (Bioinformatics 23:372-374, 2007)  Errata PartTree: an
algorithm to build an approximate tree from a large number of
unaligned sequences (describes the PartTree algorithm)</p>
<p>Katoh, Kuma, Toh, Miyata (Nucleic Acids Res. 33:511-518, 2005) MAFFT
version 5: improvement in accuracy of multiple sequence alignment
(describes [ancestral versions of] the G-INS-i, L-INS-i and E-INS-i
strategies)</p>
<p>Katoh, Misawa, Kuma, Miyata (Nucleic Acids Res. 30:3059-3066, 2002)</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">MafftCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mafft_exe</span> <span class="o">=</span> <span class="s2">&quot;/opt/local/mafft&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">in_file</span> <span class="o">=</span> <span class="s2">&quot;../Doc/examples/opuntia.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mafft_cline</span> <span class="o">=</span> <span class="n">MafftCommandline</span><span class="p">(</span><span class="n">mafft_exe</span><span class="p">,</span> <span class="nb">input</span><span class="o">=</span><span class="n">in_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mafft_cline</span><span class="p">)</span>
<span class="go">/opt/local/mafft ../Doc/examples/opuntia.fasta</span>
</pre></div>
</div>
<p>If the mafft binary is on the path (typically the case on a Unix style
operating system) then you don’t need to supply the executable location:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">MafftCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">in_file</span> <span class="o">=</span> <span class="s2">&quot;../Doc/examples/opuntia.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mafft_cline</span> <span class="o">=</span> <span class="n">MafftCommandline</span><span class="p">(</span><span class="nb">input</span><span class="o">=</span><span class="n">in_file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mafft_cline</span><span class="p">)</span>
<span class="go">mafft ../Doc/examples/opuntia.fasta</span>
</pre></div>
</div>
<p>You would typically run the command line with mafft_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<p>Note that MAFFT will write the alignment to stdout, which you may
want to save to a file and then parse, e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdout</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">mafft_cline</span><span class="p">()</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;aligned.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
    <span class="n">handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">stdout</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">AlignIO</span>
<span class="n">align</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;aligned.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternatively, to parse the output with AlignIO directly you can
use StringIO to turn the string into a handle:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdout</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">mafft_cline</span><span class="p">()</span>
<span class="kn">from</span> <span class="nn">StringIO</span> <span class="k">import</span> <span class="n">StringIO</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">AlignIO</span>
<span class="n">align</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">stdout</span><span class="p">),</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.LEXP">
<em class="property">property </em><code class="sig-name descname">LEXP</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.LEXP" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap extension penalty to skip the alignment. Default: 0.00</p>
<p>This controls the addition of the –LEXP parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.LOP">
<em class="property">property </em><code class="sig-name descname">LOP</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.LOP" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap opening penalty to skip the alignment. Default: -6.00</p>
<p>This controls the addition of the –LOP parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.aamatrix">
<em class="property">property </em><code class="sig-name descname">aamatrix</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.aamatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Use a user-defined AA scoring matrix. Default: BLOSUM62</p>
<p>This controls the addition of the –aamatrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.adjustdirection">
<em class="property">property </em><code class="sig-name descname">adjustdirection</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.adjustdirection" title="Permalink to this definition">¶</a></dt>
<dd><p>Adjust direction according to the first sequence. Default off.</p>
<p>This property controls the addition of the –adjustdirection switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.adjustdirectionaccurately">
<em class="property">property </em><code class="sig-name descname">adjustdirectionaccurately</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.adjustdirectionaccurately" title="Permalink to this definition">¶</a></dt>
<dd><p>Adjust direction according to the first sequence,for highly diverged data; very slowDefault off.</p>
<p>This property controls the addition of the –adjustdirectionaccurately switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.amino">
<em class="property">property </em><code class="sig-name descname">amino</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.amino" title="Permalink to this definition">¶</a></dt>
<dd><p>Assume the sequences are amino acid (True/False). Default: auto</p>
<p>This property controls the addition of the –amino switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.auto">
<em class="property">property </em><code class="sig-name descname">auto</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.auto" title="Permalink to this definition">¶</a></dt>
<dd><p>Automatically select strategy. Default off.</p>
<p>This property controls the addition of the –auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.bl">
<em class="property">property </em><code class="sig-name descname">bl</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.bl" title="Permalink to this definition">¶</a></dt>
<dd><p>BLOSUM number matrix is used. Default: 62</p>
<p>This controls the addition of the –bl parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.clustalout">
<em class="property">property </em><code class="sig-name descname">clustalout</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.clustalout" title="Permalink to this definition">¶</a></dt>
<dd><p>Output format: clustal (True) or fasta (False, default)</p>
<p>This property controls the addition of the –clustalout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.dpparttree">
<em class="property">property </em><code class="sig-name descname">dpparttree</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.dpparttree" title="Permalink to this definition">¶</a></dt>
<dd><p>The PartTree algorithm is used with distances based on DP. Default: off</p>
<p>This property controls the addition of the –dpparttree switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.ep">
<em class="property">property </em><code class="sig-name descname">ep</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.ep" title="Permalink to this definition">¶</a></dt>
<dd><p>Offset value, which works like gap extension penalty, for group-to- group alignment. Default: 0.123</p>
<p>This controls the addition of the –ep parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.fastapair">
<em class="property">property </em><code class="sig-name descname">fastapair</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.fastapair" title="Permalink to this definition">¶</a></dt>
<dd><p>All pairwise alignments are computed with FASTA (Pearson and Lipman 1988). Default: off</p>
<p>This property controls the addition of the –fastapair switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.fastaparttree">
<em class="property">property </em><code class="sig-name descname">fastaparttree</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.fastaparttree" title="Permalink to this definition">¶</a></dt>
<dd><p>The PartTree algorithm is used with distances based on FASTA. Default: off</p>
<p>This property controls the addition of the –fastaparttree switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.fft">
<em class="property">property </em><code class="sig-name descname">fft</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.fft" title="Permalink to this definition">¶</a></dt>
<dd><p>Use FFT approximation in group-to-group alignment. Default: on</p>
<p>This property controls the addition of the –fft switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.fmodel">
<em class="property">property </em><code class="sig-name descname">fmodel</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.fmodel" title="Permalink to this definition">¶</a></dt>
<dd><p>Incorporate the AA/nuc composition information into the scoring matrix (True) or not (False, default)</p>
<p>This property controls the addition of the –fmodel switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.genafpair">
<em class="property">property </em><code class="sig-name descname">genafpair</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.genafpair" title="Permalink to this definition">¶</a></dt>
<dd><p>All pairwise alignments are computed with a local algorithm with the generalized affine gap cost (Altschul 1998). Default: off</p>
<p>This property controls the addition of the –genafpair switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.globalpair">
<em class="property">property </em><code class="sig-name descname">globalpair</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.globalpair" title="Permalink to this definition">¶</a></dt>
<dd><p>All pairwise alignments are computed with the Needleman-Wunsch algorithm. Default: off</p>
<p>This property controls the addition of the –globalpair switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.groupsize">
<em class="property">property </em><code class="sig-name descname">groupsize</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.groupsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not make alignment larger than number sequences. Default: the number of input sequences</p>
<p>This property controls the addition of the –groupsize switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.input">
<em class="property">property </em><code class="sig-name descname">input</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.input" title="Permalink to this definition">¶</a></dt>
<dd><p>Input file name</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.input1">
<em class="property">property </em><code class="sig-name descname">input1</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.input1" title="Permalink to this definition">¶</a></dt>
<dd><p>Second input file name for the mafft-profile command</p>
<p>This controls the addition of the input1 parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.inputorder">
<em class="property">property </em><code class="sig-name descname">inputorder</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.inputorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Output order: same as input (True, default) or alignment based (False)</p>
<p>This property controls the addition of the –inputorder switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.jtt">
<em class="property">property </em><code class="sig-name descname">jtt</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.jtt" title="Permalink to this definition">¶</a></dt>
<dd><p>JTT PAM number (Jones et al. 1992) matrix is used. number&gt;0. Default: BLOSUM62</p>
<p>This controls the addition of the –jtt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.lep">
<em class="property">property </em><code class="sig-name descname">lep</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.lep" title="Permalink to this definition">¶</a></dt>
<dd><p>Offset value at local pairwise alignment. Default: 0.1</p>
<p>This controls the addition of the –lep parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.lexp">
<em class="property">property </em><code class="sig-name descname">lexp</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.lexp" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap extension penalty at local pairwise alignment. Default: -0.1</p>
<p>This controls the addition of the –lexp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.localpair">
<em class="property">property </em><code class="sig-name descname">localpair</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.localpair" title="Permalink to this definition">¶</a></dt>
<dd><p>All pairwise alignments are computed with the Smith-Waterman algorithm. Default: off</p>
<p>This property controls the addition of the –localpair switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.lop">
<em class="property">property </em><code class="sig-name descname">lop</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.lop" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap opening penalty at local pairwise alignment. Default: 0.123</p>
<p>This controls the addition of the –lop parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.maxiterate">
<em class="property">property </em><code class="sig-name descname">maxiterate</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.maxiterate" title="Permalink to this definition">¶</a></dt>
<dd><p>Number cycles of iterative refinement are performed. Default: 0</p>
<p>This controls the addition of the –maxiterate parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.memsave">
<em class="property">property </em><code class="sig-name descname">memsave</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.memsave" title="Permalink to this definition">¶</a></dt>
<dd><p>Use the Myers-Miller (1988) algorithm. Default: automatically turned on when the alignment length exceeds 10,000 (aa/nt).</p>
<p>This property controls the addition of the –memsave switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.namelength">
<em class="property">property </em><code class="sig-name descname">namelength</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.namelength" title="Permalink to this definition">¶</a></dt>
<dd><p>Name length in CLUSTAL and PHYLIP output.</p>
<blockquote>
<div><p>MAFFT v6.847 (2011) added –namelength for use with
the –clustalout option for CLUSTAL output.</p>
<p>MAFFT v7.024 (2013) added support for this with the
–phylipout option for PHYLIP output (default 10).</p>
</div></blockquote>
<p>This controls the addition of the –namelength parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.nofft">
<em class="property">property </em><code class="sig-name descname">nofft</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.nofft" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not use FFT approximation in group-to-group alignment. Default: off</p>
<p>This property controls the addition of the –nofft switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.noscore">
<em class="property">property </em><code class="sig-name descname">noscore</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.noscore" title="Permalink to this definition">¶</a></dt>
<dd><p>Alignment score is not checked in the iterative refinement stage. Default: off (score is checked)</p>
<p>This property controls the addition of the –noscore switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.nuc">
<em class="property">property </em><code class="sig-name descname">nuc</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.nuc" title="Permalink to this definition">¶</a></dt>
<dd><p>Assume the sequences are nucleotide (True/False). Default: auto</p>
<p>This property controls the addition of the –nuc switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.op">
<em class="property">property </em><code class="sig-name descname">op</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.op" title="Permalink to this definition">¶</a></dt>
<dd><p>Gap opening penalty at group-to-group alignment. Default: 1.53</p>
<p>This controls the addition of the –op parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.partsize">
<em class="property">property </em><code class="sig-name descname">partsize</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.partsize" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of partitions in the PartTree algorithm. Default: 50</p>
<p>This controls the addition of the –partsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.parttree">
<em class="property">property </em><code class="sig-name descname">parttree</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.parttree" title="Permalink to this definition">¶</a></dt>
<dd><p>Use a fast tree-building method with the 6mer distance. Default: off</p>
<p>This property controls the addition of the –parttree switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.phylipout">
<em class="property">property </em><code class="sig-name descname">phylipout</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.phylipout" title="Permalink to this definition">¶</a></dt>
<dd><p>Output format: phylip (True), or fasta (False, default)</p>
<p>This property controls the addition of the –phylipout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.quiet">
<em class="property">property </em><code class="sig-name descname">quiet</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.quiet" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not report progress (True) or not (False, default).</p>
<p>This property controls the addition of the –quiet switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.reorder">
<em class="property">property </em><code class="sig-name descname">reorder</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.reorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Output order: aligned (True) or in input order (False, default)</p>
<p>This property controls the addition of the –reorder switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.retree">
<em class="property">property </em><code class="sig-name descname">retree</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.retree" title="Permalink to this definition">¶</a></dt>
<dd><p>Guide tree is built number times in the progressive stage. Valid with 6mer distance. Default: 2</p>
<p>This controls the addition of the –retree parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.seed">
<em class="property">property </em><code class="sig-name descname">seed</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Seed alignments given in alignment_n (fasta format) are aligned with sequences in input.</p>
<p>This controls the addition of the –seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.sixmerpair">
<em class="property">property </em><code class="sig-name descname">sixmerpair</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.sixmerpair" title="Permalink to this definition">¶</a></dt>
<dd><p>Distance is calculated based on the number of shared 6mers. Default: on</p>
<p>This property controls the addition of the –6merpair switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.thread">
<em class="property">property </em><code class="sig-name descname">thread</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.thread" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of threads to use. Default: 1</p>
<p>This controls the addition of the –thread parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.tm">
<em class="property">property </em><code class="sig-name descname">tm</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.tm" title="Permalink to this definition">¶</a></dt>
<dd><p>Transmembrane PAM number (Jones et al. 1994) matrix is used. number&gt;0. Default: BLOSUM62</p>
<p>This controls the addition of the –tm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.treeout">
<em class="property">property </em><code class="sig-name descname">treeout</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.treeout" title="Permalink to this definition">¶</a></dt>
<dd><p>Guide tree is output to the input.tree file (True) or not (False, default)</p>
<p>This property controls the addition of the –treeout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MafftCommandline.weighti">
<em class="property">property </em><code class="sig-name descname">weighti</code><a class="headerlink" href="#Bio.Align.Applications.MafftCommandline.weighti" title="Permalink to this definition">¶</a></dt>
<dd><p>Weighting factor for the consistency term calculated from pairwise alignments. Default: 2.7</p>
<p>This controls the addition of the –weighti parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Align.Applications.DialignCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">DialignCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='dialign2-2'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for the multiple alignment program DIALIGN2-2.</p>
<p><a class="reference external" href="http://bibiserv.techfak.uni-bielefeld.de/dialign/welcome.html">http://bibiserv.techfak.uni-bielefeld.de/dialign/welcome.html</a></p>
<p class="rubric">Notes</p>
<p>Last checked against version: 2.2</p>
<p class="rubric">References</p>
<p>B. Morgenstern (2004). DIALIGN: Multiple DNA and Protein Sequence
Alignment at BiBiServ. Nucleic Acids Research 32, W33-W36.</p>
<p class="rubric">Examples</p>
<p>To align a FASTA file (unaligned.fasta) with the output files names
aligned.* including a FASTA output file (aligned.fa), use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">DialignCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dialign_cline</span> <span class="o">=</span> <span class="n">DialignCommandline</span><span class="p">(</span><span class="nb">input</span><span class="o">=</span><span class="s2">&quot;unaligned.fasta&quot;</span><span class="p">,</span>
<span class="gp">... </span>                                   <span class="n">fn</span><span class="o">=</span><span class="s2">&quot;aligned&quot;</span><span class="p">,</span> <span class="n">fa</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dialign_cline</span><span class="p">)</span>
<span class="go">dialign2-2 -fa -fn aligned unaligned.fasta</span>
</pre></div>
</div>
<p>You would typically run the command line with dialign_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.afc">
<em class="property">property </em><code class="sig-name descname">afc</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.afc" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates additional output file ‘*.afc’ containing data of all fragments considered for alignment WARNING: this file can be HUGE !</p>
<p>This property controls the addition of the -afc switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.afc_v">
<em class="property">property </em><code class="sig-name descname">afc_v</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.afc_v" title="Permalink to this definition">¶</a></dt>
<dd><p>Like ‘-afc’ but verbose: fragments are explicitly printed. WARNING: this file can be EVEN BIGGER !</p>
<p>This property controls the addition of the -afc_v switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.anc">
<em class="property">property </em><code class="sig-name descname">anc</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.anc" title="Permalink to this definition">¶</a></dt>
<dd><p>Anchored alignment. Requires a file &lt;seq_file&gt;.anc containing anchor points.</p>
<p>This property controls the addition of the -anc switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.cs">
<em class="property">property </em><code class="sig-name descname">cs</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.cs" title="Permalink to this definition">¶</a></dt>
<dd><p>If segments are translated, not only the ‘Watson strand’ but also the ‘Crick strand’ is looked at.</p>
<p>This property controls the addition of the -cs switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.cw">
<em class="property">property </em><code class="sig-name descname">cw</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.cw" title="Permalink to this definition">¶</a></dt>
<dd><p>Additional output file in CLUSTAL W format.</p>
<p>This property controls the addition of the -cw switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.ds">
<em class="property">property </em><code class="sig-name descname">ds</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.ds" title="Permalink to this definition">¶</a></dt>
<dd><p>‘dna alignment speed up’ - non-translated nucleic acid fragments are taken into account only if they start with at least two matches. Speeds up DNA alignment at the expense of sensitivity.</p>
<p>This property controls the addition of the -ds switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.fa">
<em class="property">property </em><code class="sig-name descname">fa</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.fa" title="Permalink to this definition">¶</a></dt>
<dd><p>Additional output file in FASTA format.</p>
<p>This property controls the addition of the -fa switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.ff">
<em class="property">property </em><code class="sig-name descname">ff</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.ff" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates file *.frg containing information about all fragments that are part of the respective optimal pairwise alignmnets plus information about consistency in the multiple alignment</p>
<p>This property controls the addition of the -ff switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.fn">
<em class="property">property </em><code class="sig-name descname">fn</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.fn" title="Permalink to this definition">¶</a></dt>
<dd><p>Output files are named &lt;out_file&gt;.&lt;extension&gt;.</p>
<p>This controls the addition of the -fn parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.fop">
<em class="property">property </em><code class="sig-name descname">fop</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.fop" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates file *.fop containing coordinates of all fragments that are part of the respective pairwise alignments.</p>
<p>This property controls the addition of the -fop switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.fsm">
<em class="property">property </em><code class="sig-name descname">fsm</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.fsm" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates file *.fsm containing coordinates of all fragments that are part of the final alignment</p>
<p>This property controls the addition of the -fsm switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.input">
<em class="property">property </em><code class="sig-name descname">input</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.input" title="Permalink to this definition">¶</a></dt>
<dd><p>Input file name. Must be FASTA format</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.iw">
<em class="property">property </em><code class="sig-name descname">iw</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.iw" title="Permalink to this definition">¶</a></dt>
<dd><p>Overlap weights switched off (by default, overlap weights are used if up to 35 sequences are aligned). This option speeds up the alignment but may lead to reduced alignment quality.</p>
<p>This property controls the addition of the -iw switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.lgs">
<em class="property">property </em><code class="sig-name descname">lgs</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.lgs" title="Permalink to this definition">¶</a></dt>
<dd><p>‘long genomic sequences’ - combines the following options: -ma, -thr 2, -lmax 30, -smin 8, -nta, -ff, -fop, -ff, -cs, -ds, -pst</p>
<p>This property controls the addition of the -lgs switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.lgs_t">
<em class="property">property </em><code class="sig-name descname">lgs_t</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.lgs_t" title="Permalink to this definition">¶</a></dt>
<dd><p>Like ‘-lgs’ but with all segment pairs assessed at the peptide level (rather than ‘mixed alignments’ as with the ‘-lgs’ option). Therefore faster than -lgs but not very sensitive for non-coding regions.</p>
<p>This property controls the addition of the -lgs_t switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.lmax">
<em class="property">property </em><code class="sig-name descname">lmax</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.lmax" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum fragment length = x  (default: x = 40 or x = 120 for ‘translated’ fragments). Shorter x speeds up the program but may affect alignment quality.</p>
<p>This controls the addition of the -lmax parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.lo">
<em class="property">property </em><code class="sig-name descname">lo</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.lo" title="Permalink to this definition">¶</a></dt>
<dd><p>(Long Output) Additional file *.log with information about fragments selected for pairwise alignment and about consistency in multi-alignment procedure.</p>
<p>This property controls the addition of the -lo switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.ma">
<em class="property">property </em><code class="sig-name descname">ma</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.ma" title="Permalink to this definition">¶</a></dt>
<dd><p>‘mixed alignments’ consisting of P-fragments and N-fragments if nucleic acid sequences are aligned.</p>
<p>This property controls the addition of the -ma switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.mask">
<em class="property">property </em><code class="sig-name descname">mask</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Residues not belonging to selected fragments are replaced by ‘*’ characters in output alignment (rather than being printed in lower-case characters)</p>
<p>This property controls the addition of the -mask switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.mat">
<em class="property">property </em><code class="sig-name descname">mat</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.mat" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates file *mat with substitution counts derived from the fragments that have been selected for alignment.</p>
<p>This property controls the addition of the -mat switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.mat_thr">
<em class="property">property </em><code class="sig-name descname">mat_thr</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.mat_thr" title="Permalink to this definition">¶</a></dt>
<dd><p>Like ‘-mat’ but only fragments with weight score &gt; t are considered</p>
<p>This property controls the addition of the -mat_thr switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.max_link">
<em class="property">property </em><code class="sig-name descname">max_link</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.max_link" title="Permalink to this definition">¶</a></dt>
<dd><p>‘maximum linkage’ clustering used to construct sequence tree (instead of UPGMA).</p>
<p>This property controls the addition of the -max_link switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.min_link">
<em class="property">property </em><code class="sig-name descname">min_link</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.min_link" title="Permalink to this definition">¶</a></dt>
<dd><p>‘minimum linkage’ clustering used.</p>
<p>This property controls the addition of the -min_link switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.mot">
<em class="property">property </em><code class="sig-name descname">mot</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.mot" title="Permalink to this definition">¶</a></dt>
<dd><p>‘motif’ option.</p>
<p>This controls the addition of the -mot parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.msf">
<em class="property">property </em><code class="sig-name descname">msf</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.msf" title="Permalink to this definition">¶</a></dt>
<dd><p>Separate output file in MSF format.</p>
<p>This property controls the addition of the -msf switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.n">
<em class="property">property </em><code class="sig-name descname">n</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.n" title="Permalink to this definition">¶</a></dt>
<dd><p>Input sequences are nucleic acid sequences. No translation of fragments.</p>
<p>This property controls the addition of the -n switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.nt">
<em class="property">property </em><code class="sig-name descname">nt</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.nt" title="Permalink to this definition">¶</a></dt>
<dd><p>Input sequences are nucleic acid sequences and ‘nucleic acid segments’ are translated to ‘peptide segments’.</p>
<p>This property controls the addition of the -nt switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.nta">
<em class="property">property </em><code class="sig-name descname">nta</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.nta" title="Permalink to this definition">¶</a></dt>
<dd><p>‘no textual alignment’ - textual alignment suppressed. This option makes sense if other output files are of interest – e.g. the fragment files created with -ff, -fop, -fsm or -lo.</p>
<p>This property controls the addition of the -nta switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.o">
<em class="property">property </em><code class="sig-name descname">o</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.o" title="Permalink to this definition">¶</a></dt>
<dd><p>Fast version, resulting alignments may be slightly different.</p>
<p>This property controls the addition of the -o switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.ow">
<em class="property">property </em><code class="sig-name descname">ow</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.ow" title="Permalink to this definition">¶</a></dt>
<dd><p>Overlap weights enforced (By default, overlap weights are used only if up to 35 sequences are aligned since calculating overlap weights is time consuming).</p>
<p>This property controls the addition of the -ow switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.pst">
<em class="property">property </em><code class="sig-name descname">pst</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.pst" title="Permalink to this definition">¶</a></dt>
<dd><p>‘print status’. Creates and updates a file *.sta with information about the current status of the program run.  This option is recommended if large data sets are aligned since it allows the user to estimate the remaining running time.</p>
<p>This property controls the addition of the -pst switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.smin">
<em class="property">property </em><code class="sig-name descname">smin</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.smin" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum similarity value for first residue pair (or codon pair) in fragments. Speeds up protein alignment or alignment of translated DNA fragments at the expense of sensitivity.</p>
<p>This property controls the addition of the -smin switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.stars">
<em class="property">property </em><code class="sig-name descname">stars</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.stars" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum number of ‘*’ characters indicating degree of local similarity among sequences. By default, no stars are used but numbers between 0 and 9, instead.</p>
<p>This controls the addition of the -stars parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.stdo">
<em class="property">property </em><code class="sig-name descname">stdo</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.stdo" title="Permalink to this definition">¶</a></dt>
<dd><p>Results written to standard output.</p>
<p>This property controls the addition of the -stdo switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.ta">
<em class="property">property </em><code class="sig-name descname">ta</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.ta" title="Permalink to this definition">¶</a></dt>
<dd><p>Standard textual alignment printed (overrides suppression of textual alignments in special options, e.g. -lgs)</p>
<p>This property controls the addition of the -ta switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.thr">
<em class="property">property </em><code class="sig-name descname">thr</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.thr" title="Permalink to this definition">¶</a></dt>
<dd><p>Threshold T = x.</p>
<p>This controls the addition of the -thr parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.DialignCommandline.xfr">
<em class="property">property </em><code class="sig-name descname">xfr</code><a class="headerlink" href="#Bio.Align.Applications.DialignCommandline.xfr" title="Permalink to this definition">¶</a></dt>
<dd><p>‘exclude fragments’ - list of fragments can be specified that are NOT considered for pairwise alignment</p>
<p>This property controls the addition of the -xfr switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Align.Applications.ProbconsCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">ProbconsCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='probcons'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for the multiple alignment program PROBCONS.</p>
<p><a class="reference external" href="http://probcons.stanford.edu/">http://probcons.stanford.edu/</a></p>
<p class="rubric">Notes</p>
<p>Last checked against version: 1.12</p>
<p class="rubric">References</p>
<p>Do, C.B., Mahabhashyam, M.S.P., Brudno, M., and Batzoglou, S. 2005.
PROBCONS: Probabilistic Consistency-based Multiple Sequence Alignment.
Genome Research 15: 330-340.</p>
<p class="rubric">Examples</p>
<p>To align a FASTA file (unaligned.fasta) with the output in ClustalW
format, and otherwise default settings, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">ProbconsCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">probcons_cline</span> <span class="o">=</span> <span class="n">ProbconsCommandline</span><span class="p">(</span><span class="nb">input</span><span class="o">=</span><span class="s2">&quot;unaligned.fasta&quot;</span><span class="p">,</span>
<span class="gp">... </span>                                     <span class="n">clustalw</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">probcons_cline</span><span class="p">)</span>
<span class="go">probcons -clustalw unaligned.fasta</span>
</pre></div>
</div>
<p>You would typically run the command line with probcons_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<p>Note that PROBCONS will write the alignment to stdout, which you may
want to save to a file and then parse, e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdout</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">probcons_cline</span><span class="p">()</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;aligned.aln&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
    <span class="n">handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">stdout</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">AlignIO</span>
<span class="n">align</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;aligned.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;clustalw&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternatively, to parse the output with AlignIO directly you can
use StringIO to turn the string into a handle:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">stdout</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">probcons_cline</span><span class="p">()</span>
<span class="kn">from</span> <span class="nn">StringIO</span> <span class="k">import</span> <span class="n">StringIO</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">AlignIO</span>
<span class="n">align</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">stdout</span><span class="p">),</span> <span class="s2">&quot;clustalw&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.a">
<em class="property">property </em><code class="sig-name descname">a</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.a" title="Permalink to this definition">¶</a></dt>
<dd><p>Print sequences in alignment order rather than input order (default: off)</p>
<p>This property controls the addition of the -a switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.annot">
<em class="property">property </em><code class="sig-name descname">annot</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.annot" title="Permalink to this definition">¶</a></dt>
<dd><p>Write annotation for multiple alignment to FILENAME</p>
<p>This controls the addition of the -annot parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.clustalw">
<em class="property">property </em><code class="sig-name descname">clustalw</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.clustalw" title="Permalink to this definition">¶</a></dt>
<dd><p>Use CLUSTALW output format instead of MFA</p>
<p>This property controls the addition of the -clustalw switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.consistency">
<em class="property">property </em><code class="sig-name descname">consistency</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.consistency" title="Permalink to this definition">¶</a></dt>
<dd><p>Use 0 &lt;= REPS &lt;= 5 (default: 2) passes of consistency transformation</p>
<p>This controls the addition of the -c parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.emissions">
<em class="property">property </em><code class="sig-name descname">emissions</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.emissions" title="Permalink to this definition">¶</a></dt>
<dd><p>Also reestimate emission probabilities (default: off)</p>
<p>This property controls the addition of the -e switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.input">
<em class="property">property </em><code class="sig-name descname">input</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.input" title="Permalink to this definition">¶</a></dt>
<dd><p>Input file name. Must be multiple FASTA alignment (MFA) format</p>
<p>This controls the addition of the input parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.ir">
<em class="property">property </em><code class="sig-name descname">ir</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.ir" title="Permalink to this definition">¶</a></dt>
<dd><p>Use 0 &lt;= REPS &lt;= 1000 (default: 100) passes of iterative-refinement</p>
<p>This controls the addition of the -ir parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.pairs">
<em class="property">property </em><code class="sig-name descname">pairs</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate all-pairs pairwise alignments</p>
<p>This property controls the addition of the -pairs switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.paramfile">
<em class="property">property </em><code class="sig-name descname">paramfile</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.paramfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Read parameters from FILENAME</p>
<p>This controls the addition of the -p parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.pre">
<em class="property">property </em><code class="sig-name descname">pre</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.pre" title="Permalink to this definition">¶</a></dt>
<dd><p>Use 0 &lt;= REPS &lt;= 20 (default: 0) rounds of pretraining</p>
<p>This controls the addition of the -pre parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.train">
<em class="property">property </em><code class="sig-name descname">train</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.train" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute EM transition probabilities, store in FILENAME (default: no training)</p>
<p>This controls the addition of the -t parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.verbose">
<em class="property">property </em><code class="sig-name descname">verbose</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.verbose" title="Permalink to this definition">¶</a></dt>
<dd><p>Report progress while aligning (default: off)</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.ProbconsCommandline.viterbi">
<em class="property">property </em><code class="sig-name descname">viterbi</code><a class="headerlink" href="#Bio.Align.Applications.ProbconsCommandline.viterbi" title="Permalink to this definition">¶</a></dt>
<dd><p>Use Viterbi algorithm to generate all pairs (automatically enables -pairs)</p>
<p>This property controls the addition of the -viterbi switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Align.Applications.TCoffeeCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">TCoffeeCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='t_coffee'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Commandline object for the TCoffee alignment program.</p>
<p><a class="reference external" href="http://www.tcoffee.org/Projects_home_page/t_coffee_home_page.html">http://www.tcoffee.org/Projects_home_page/t_coffee_home_page.html</a></p>
<p>The T-Coffee command line tool has a lot of switches and options.
This wrapper implements a VERY limited number of options - if you
would like to help improve it please get in touch.</p>
<p class="rubric">Notes</p>
<p>Last checked against: Version_6.92</p>
<p class="rubric">References</p>
<p>T-Coffee: A novel method for multiple sequence alignments.
Notredame, Higgins, Heringa, JMB,302(205-217) 2000</p>
<p class="rubric">Examples</p>
<p>To align a FASTA file (unaligned.fasta) with the output in ClustalW
format (file aligned.aln), and otherwise default settings, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">TCoffeeCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tcoffee_cline</span> <span class="o">=</span> <span class="n">TCoffeeCommandline</span><span class="p">(</span><span class="n">infile</span><span class="o">=</span><span class="s2">&quot;unaligned.fasta&quot;</span><span class="p">,</span>
<span class="gp">... </span>                                   <span class="n">output</span><span class="o">=</span><span class="s2">&quot;clustalw&quot;</span><span class="p">,</span>
<span class="gp">... </span>                                   <span class="n">outfile</span><span class="o">=</span><span class="s2">&quot;aligned.aln&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">tcoffee_cline</span><span class="p">)</span>
<span class="go">t_coffee -output clustalw -infile unaligned.fasta -outfile aligned.aln</span>
</pre></div>
</div>
<p>You would typically run the command line with tcoffee_cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="attribute">
<dt id="Bio.Align.Applications.TCoffeeCommandline.SEQ_TYPES">
<code class="sig-name descname">SEQ_TYPES</code><em class="property"> = ['dna', 'protein', 'dna_protein']</em><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.SEQ_TYPES" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.convert">
<em class="property">property </em><code class="sig-name descname">convert</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.convert" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify you want to perform a file conversion</p>
<p>This property controls the addition of the -convert switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.gapext">
<em class="property">property </em><code class="sig-name descname">gapext</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.gapext" title="Permalink to this definition">¶</a></dt>
<dd><p>Indicates the penalty applied for extending a gap (negative integer)</p>
<p>This controls the addition of the -gapext parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.gapopen">
<em class="property">property </em><code class="sig-name descname">gapopen</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.gapopen" title="Permalink to this definition">¶</a></dt>
<dd><p>Indicates the penalty applied for opening a gap (negative integer)</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.infile">
<em class="property">property </em><code class="sig-name descname">infile</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.infile" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify the input file.</p>
<p>This controls the addition of the -infile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.matrix">
<em class="property">property </em><code class="sig-name descname">matrix</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify the filename of the substitution matrix to use.Default: blosum62mt</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.mode">
<em class="property">property </em><code class="sig-name descname">mode</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.mode" title="Permalink to this definition">¶</a></dt>
<dd><p>Specifies a special mode: genome, quickaln, dali, 3dcoffee</p>
<p>This controls the addition of the -mode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.outfile">
<em class="property">property </em><code class="sig-name descname">outfile</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.outfile" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify the output file. Default: &lt;your sequences&gt;.aln</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.outorder">
<em class="property">property </em><code class="sig-name descname">outorder</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.outorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify the order of sequence to outputEither ‘input’, ‘aligned’ or &lt;filename&gt; of Fasta file with sequence order</p>
<p>This controls the addition of the -outorder parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.output">
<em class="property">property </em><code class="sig-name descname">output</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.output" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify the output type.</p>
<blockquote>
<div><p>One (or more separated by a comma) of:
‘clustalw_aln’, ‘clustalw’, ‘gcg’, ‘msf_aln’,
‘pir_aln’, ‘fasta_aln’, ‘phylip’, ‘pir_seq’, ‘fasta_seq’</p>
</div></blockquote>
<p>This controls the addition of the -output parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.quiet">
<em class="property">property </em><code class="sig-name descname">quiet</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.quiet" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn off log output</p>
<p>This property controls the addition of the -quiet switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.TCoffeeCommandline.type">
<em class="property">property </em><code class="sig-name descname">type</code><a class="headerlink" href="#Bio.Align.Applications.TCoffeeCommandline.type" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify the type of sequence being aligned</p>
<p>This controls the addition of the -type parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Align.Applications.MSAProbsCommandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Align.Applications.</code><code class="sig-name descname">MSAProbsCommandline</code><span class="sig-paren">(</span><em class="sig-param">cmd='msaprobs'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Command line wrapper for MSAProbs.</p>
<p><a class="reference external" href="http://msaprobs.sourceforge.net">http://msaprobs.sourceforge.net</a></p>
<p class="rubric">Notes</p>
<p>Last checked against version: 0.9.7</p>
<p class="rubric">References</p>
<p>Yongchao Liu, Bertil Schmidt, Douglas L. Maskell: “MSAProbs: multiple
sequence alignment based on pair hidden Markov models and partition
function posterior probabilities”. Bioinformatics, 2010, 26(16): 1958 -1964</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.Applications</span> <span class="k">import</span> <span class="n">MSAProbsCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">in_file</span> <span class="o">=</span> <span class="s2">&quot;unaligned.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">out_file</span> <span class="o">=</span> <span class="s2">&quot;aligned.cla&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">MSAProbsCommandline</span><span class="p">(</span><span class="n">infile</span><span class="o">=</span><span class="n">in_file</span><span class="p">,</span> <span class="n">outfile</span><span class="o">=</span><span class="n">out_file</span><span class="p">,</span> <span class="n">clustalw</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">msaprobs -o aligned.cla -clustalw unaligned.fasta</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via
the Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.alignment_order">
<em class="property">property </em><code class="sig-name descname">alignment_order</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.alignment_order" title="Permalink to this definition">¶</a></dt>
<dd><p>print sequences in alignment order rather than input order (default: off)</p>
<p>This property controls the addition of the -a switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.annot">
<em class="property">property </em><code class="sig-name descname">annot</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.annot" title="Permalink to this definition">¶</a></dt>
<dd><p>write annotation for multiple alignment to FILENAME</p>
<p>This controls the addition of the -annot parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.clustalw">
<em class="property">property </em><code class="sig-name descname">clustalw</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.clustalw" title="Permalink to this definition">¶</a></dt>
<dd><p>use CLUSTALW output format instead of FASTA format</p>
<p>This property controls the addition of the -clustalw switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.consistency">
<em class="property">property </em><code class="sig-name descname">consistency</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.consistency" title="Permalink to this definition">¶</a></dt>
<dd><p>use 0 &lt;= REPS &lt;= 5 (default: 2) passes of consistency transformation</p>
<p>This controls the addition of the -c parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.infile">
<em class="property">property </em><code class="sig-name descname">infile</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.infile" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiple sequence input file</p>
<p>This controls the addition of the infile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.iterative_refinement">
<em class="property">property </em><code class="sig-name descname">iterative_refinement</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.iterative_refinement" title="Permalink to this definition">¶</a></dt>
<dd><p>use 0 &lt;= REPS &lt;= 1000 (default: 10) passes of iterative-refinement</p>
<p>This controls the addition of the -ir parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.numthreads">
<em class="property">property </em><code class="sig-name descname">numthreads</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.numthreads" title="Permalink to this definition">¶</a></dt>
<dd><p>specify the number of threads used, and otherwise detect automatically</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.outfile">
<em class="property">property </em><code class="sig-name descname">outfile</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.outfile" title="Permalink to this definition">¶</a></dt>
<dd><p>specify the output file name (STDOUT by default)</p>
<p>This controls the addition of the -o parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.verbose">
<em class="property">property </em><code class="sig-name descname">verbose</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.verbose" title="Permalink to this definition">¶</a></dt>
<dd><p>report progress while aligning (default: off)</p>
<p>This property controls the addition of the -v switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.Align.Applications.MSAProbsCommandline.version">
<em class="property">property </em><code class="sig-name descname">version</code><a class="headerlink" href="#Bio.Align.Applications.MSAProbsCommandline.version" title="Permalink to this definition">¶</a></dt>
<dd><p>print out version of MSAPROBS</p>
<p>This controls the addition of the -version parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.Align.substitution_matrices.html" class="btn btn-neutral float-right" title="Bio.Align.substitution_matrices package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Align.html" class="btn btn-neutral float-left" title="Bio.Align package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2017, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>