<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.PopGen.GenePop.EasyController module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.PopGen.GenePop.FileParser module" href="Bio.PopGen.GenePop.FileParser.html" />
    <link rel="prev" title="Bio.PopGen.GenePop.Controller module" href="Bio.PopGen.GenePop.Controller.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PopGen.html">Bio.PopGen package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PopGen.GenePop.html">Bio.PopGen.GenePop package</a></li>
      <li class="breadcrumb-item active">Bio.PopGen.GenePop.EasyController module</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/api/Bio.PopGen.GenePop.EasyController.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.PopGen.GenePop.Controller.html" class="btn btn-neutral float-left" title="Bio.PopGen.GenePop.Controller module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PopGen.GenePop.FileParser.html" class="btn btn-neutral float-right" title="Bio.PopGen.GenePop.FileParser module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.PopGen.GenePop.EasyController">
<span id="bio-popgen-genepop-easycontroller-module"></span><h1>Bio.PopGen.GenePop.EasyController module<a class="headerlink" href="#module-Bio.PopGen.GenePop.EasyController" title="Permalink to this headline">ÔÉÅ</a></h1>
<p>Control GenePop through an easier interface.</p>
<p>This interface is less efficient than the standard GenePopControler</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PopGen.GenePop.EasyController.</span></span><span class="sig-name descname"><span class="pre">EasyController</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genepop_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Define a class for an easier interface with the GenePop program.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genepop_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the controller.</p>
<p>genepop_dir is the directory where GenePop is.</p>
<p>The binary should be called Genepop (capital G)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_basic_info">
<span class="sig-name descname"><span class="pre">get_basic_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_basic_info" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Obtain the population list and loci list from the file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.test_hw_pop">
<span class="sig-name descname"><span class="pre">test_hw_pop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'probability'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.test_hw_pop" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Perform Hardy-Weinberg test on the given position.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.test_hw_global">
<span class="sig-name descname"><span class="pre">test_hw_global</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'deficiency'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enum_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.test_hw_global" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Perform Hardy-Weinberg global Heterozygote test.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.test_ld_all_pair">
<span class="sig-name descname"><span class="pre">test_ld_all_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dememorization</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batches</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.test_ld_all_pair" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Test for linkage disequilibrium for each pair of loci in each population.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.estimate_nm">
<span class="sig-name descname"><span class="pre">estimate_nm</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.estimate_nm" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Estimate Nm. Just a simple bridge.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_heterozygosity_info">
<span class="sig-name descname"><span class="pre">get_heterozygosity_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_heterozygosity_info" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the heterozygosity info for a certain locus on a population.</p>
<p>Returns (Expected homozygotes, observed homozygotes,
Expected heterozygotes, observed heterozygotes)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_genotype_count">
<span class="sig-name descname"><span class="pre">get_genotype_count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_genotype_count" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the genotype counts for a certain population and locus.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_fis">
<span class="sig-name descname"><span class="pre">get_fis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_fis" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the Fis for a certain population and locus.</p>
<p>Below CW means Cockerham and Weir and RH means Robertson and Hill.</p>
<p>Returns a pair:</p>
<ul class="simple">
<li><p>dictionary [allele] = (repetition count, frequency, Fis CW )
with information for each allele</p></li>
<li><p>a triple with total number of alleles, Fis CW, Fis RH</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_alleles">
<span class="sig-name descname"><span class="pre">get_alleles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_alleles" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the alleles for a certain population and locus.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_alleles_all_pops">
<span class="sig-name descname"><span class="pre">get_alleles_all_pops</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_alleles_all_pops" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the alleles for a certain population and locus.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_allele_frequency">
<span class="sig-name descname"><span class="pre">get_allele_frequency</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pop_pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">locus_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_allele_frequency" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate the allele frequency for a certain locus on a population.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_multilocus_f_stats">
<span class="sig-name descname"><span class="pre">get_multilocus_f_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_multilocus_f_stats" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the multilocus F stats.</p>
<p>Explain averaging.
Returns Fis(CW), Fst, Fit</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_f_stats">
<span class="sig-name descname"><span class="pre">get_f_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_f_stats" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return F stats for a locus.</p>
<p>Returns Fis(CW), Fst, Fit, Qintra, Qinter</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_avg_fis">
<span class="sig-name descname"><span class="pre">get_avg_fis</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_avg_fis" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate identity-base average Fis.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_avg_fst_pair">
<span class="sig-name descname"><span class="pre">get_avg_fst_pair</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_avg_fst_pair" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate Allele size-base average Fis for all population pairs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.get_avg_fst_pair_locus">
<span class="sig-name descname"><span class="pre">get_avg_fst_pair_locus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locus</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.get_avg_fst_pair_locus" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate Allele size-base average Fis for all population pairs of the given locus.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PopGen.GenePop.EasyController.EasyController.calc_ibd">
<span class="sig-name descname"><span class="pre">calc_ibd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">is_diplo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'a'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Log'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.00001</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PopGen.GenePop.EasyController.EasyController.calc_ibd" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate isolation by distance statistics for Diploid or Haploid.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.PopGen.GenePop.Controller.html" class="btn btn-neutral float-left" title="Bio.PopGen.GenePop.Controller module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PopGen.GenePop.FileParser.html" class="btn btn-neutral float-right" title="Bio.PopGen.GenePop.FileParser module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>