

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Alphabet package &mdash; Biopython 1.76.dev0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Alphabet.IUPAC module" href="Bio.Alphabet.IUPAC.html" />
    <link rel="prev" title="Bio.AlignIO.StockholmIO module" href="Bio.AlignIO.StockholmIO.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.76.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.Alphabet package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-Bio.Alphabet">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
      <li>Bio.Alphabet package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Alphabet/__init__.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Alphabet.IUPAC.html" class="btn btn-neutral float-right" title="Bio.Alphabet.IUPAC module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.AlignIO.StockholmIO.html" class="btn btn-neutral float-left" title="Bio.AlignIO.StockholmIO module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="bio-alphabet-package">
<h1>Bio.Alphabet package<a class="headerlink" href="#bio-alphabet-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.Alphabet.IUPAC.html">Bio.Alphabet.IUPAC module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.Alphabet.Reduced.html">Bio.Alphabet.Reduced module</a></li>
</ul>
</div>
</div>
<div class="section" id="module-Bio.Alphabet">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.Alphabet" title="Permalink to this headline">¶</a></h2>
<p>Alphabets used in Seq objects etc to declare sequence type and letters (OBSOLETE).</p>
<p>This is used by sequences which contain a finite number of similar words.</p>
<p>The design of Bio.Aphabet included a number of historic design choices
which, with the benefit of hindsight, were regretable. While the details
remain to be agreed, we intend to remove or replace Bio.Alphabet in 2020.
Please avoid using this module explicitly in your code. See also:
<a class="reference external" href="https://github.com/biopython/biopython/issues/2046">https://github.com/biopython/biopython/issues/2046</a></p>
<dl class="class">
<dt id="Bio.Alphabet.Alphabet">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">Alphabet</code><a class="headerlink" href="#Bio.Alphabet.Alphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Generic alphabet base class.</p>
<p>This class is used as a base class for other types of alphabets.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>letters - list-like object containing the letters of the alphabet.</dt><dd><p>Usually it is a string when letters are single characters.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>size    - size of the alphabet’s letters (e.g. 1 when letters are</dt><dd><p>single characters).</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>letters</strong></dt><dd></dd>
<dt><strong>size</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.Alphabet.Alphabet.contains" title="Bio.Alphabet.Alphabet.contains"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code></a>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.Alphabet.Alphabet.contains">
<code class="sig-name descname">contains</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Alphabet.Alphabet.contains" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p>
<p>Returns a boolean.  This relies on the Alphabet subclassing
hierarchy only, and does not check the letters property.
This isn’t ideal, and doesn’t seem to work as intended
with the AlphabetEncoder classes.</p>
</dd></dl>

<dl class="attribute">
<dt id="Bio.Alphabet.Alphabet.letters">
<code class="sig-name descname">letters</code><em class="property"> = None</em><a class="headerlink" href="#Bio.Alphabet.Alphabet.letters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Bio.Alphabet.Alphabet.size">
<code class="sig-name descname">size</code><em class="property"> = None</em><a class="headerlink" href="#Bio.Alphabet.Alphabet.size" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.AlphabetEncoder">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">AlphabetEncoder</code><span class="sig-paren">(</span><em class="sig-param">alphabet</em>, <em class="sig-param">new_letters</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Alphabet.AlphabetEncoder" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class to construct a new, extended alphabet from an existing one.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.Alphabet.AlphabetEncoder.contains" title="Bio.Alphabet.AlphabetEncoder.contains"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code></a>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.Alphabet.AlphabetEncoder.contains">
<code class="sig-name descname">contains</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Alphabet.AlphabetEncoder.contains" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p>
<p>This is isn’t implemented for the base AlphabetEncoder,
which will always return 0 (False).</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.DNAAlphabet">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">DNAAlphabet</code><a class="headerlink" href="#Bio.Alphabet.DNAAlphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.NucleotideAlphabet" title="Bio.Alphabet.NucleotideAlphabet"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.NucleotideAlphabet</span></code></a></p>
<p>Generic single letter DNA alphabet.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>letters</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.Gapped">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">Gapped</code><span class="sig-paren">(</span><em class="sig-param">alphabet</em>, <em class="sig-param">gap_char='-'</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Alphabet.Gapped" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.AlphabetEncoder" title="Bio.Alphabet.AlphabetEncoder"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.AlphabetEncoder</span></code></a></p>
<p>Alphabets which contain a gap character.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.Alphabet.Gapped.contains" title="Bio.Alphabet.Gapped.contains"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code></a>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.Alphabet.Gapped.contains">
<code class="sig-name descname">contains</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Alphabet.Gapped.contains" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p>
<p>Returns a boolean.  This relies on the Alphabet subclassing
hierarchy, and attempts to check the gap character.  This fails
if the other alphabet does not have a gap character!</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.HasStopCodon">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">HasStopCodon</code><span class="sig-paren">(</span><em class="sig-param">alphabet</em>, <em class="sig-param">stop_symbol='*'</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Alphabet.HasStopCodon" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.AlphabetEncoder" title="Bio.Alphabet.AlphabetEncoder"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.AlphabetEncoder</span></code></a></p>
<p>Alphabets which contain a stop symbol.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.Alphabet.HasStopCodon.contains" title="Bio.Alphabet.HasStopCodon.contains"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code></a>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.Alphabet.HasStopCodon.contains">
<code class="sig-name descname">contains</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Alphabet.HasStopCodon.contains" title="Permalink to this definition">¶</a></dt>
<dd><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p>
<p>Returns a boolean.  This relies on the Alphabet subclassing
hierarchy, and attempts to check the stop symbol.  This fails
if the other alphabet does not have a stop symbol!</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.NucleotideAlphabet">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">NucleotideAlphabet</code><a class="headerlink" href="#Bio.Alphabet.NucleotideAlphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.SingleLetterAlphabet" title="Bio.Alphabet.SingleLetterAlphabet"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.SingleLetterAlphabet</span></code></a></p>
<p>Generic single letter nucleotide alphabet.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>letters</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.ProteinAlphabet">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">ProteinAlphabet</code><a class="headerlink" href="#Bio.Alphabet.ProteinAlphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.SingleLetterAlphabet" title="Bio.Alphabet.SingleLetterAlphabet"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.SingleLetterAlphabet</span></code></a></p>
<p>Generic single letter protein alphabet.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>letters</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.RNAAlphabet">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">RNAAlphabet</code><a class="headerlink" href="#Bio.Alphabet.RNAAlphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.NucleotideAlphabet" title="Bio.Alphabet.NucleotideAlphabet"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.NucleotideAlphabet</span></code></a></p>
<p>Generic single letter RNA alphabet.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>letters</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.SecondaryStructure">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">SecondaryStructure</code><a class="headerlink" href="#Bio.Alphabet.SecondaryStructure" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.SingleLetterAlphabet" title="Bio.Alphabet.SingleLetterAlphabet"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.SingleLetterAlphabet</span></code></a></p>
<p>Alphabet used to describe secondary structure.</p>
<p>Letters are ‘H’ (helix), ‘S’ (strand), ‘T’ (turn) and ‘C’ (coil).</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="Bio.Alphabet.SecondaryStructure.letters">
<code class="sig-name descname">letters</code><em class="property"> = 'HSTC'</em><a class="headerlink" href="#Bio.Alphabet.SecondaryStructure.letters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.SingleLetterAlphabet">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">SingleLetterAlphabet</code><a class="headerlink" href="#Bio.Alphabet.SingleLetterAlphabet" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.Alphabet" title="Bio.Alphabet.Alphabet"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.Alphabet</span></code></a></p>
<p>Generic alphabet with letters of size one.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>letters</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="Bio.Alphabet.SingleLetterAlphabet.letters">
<code class="sig-name descname">letters</code><em class="property"> = None</em><a class="headerlink" href="#Bio.Alphabet.SingleLetterAlphabet.letters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Bio.Alphabet.SingleLetterAlphabet.size">
<code class="sig-name descname">size</code><em class="property"> = 1</em><a class="headerlink" href="#Bio.Alphabet.SingleLetterAlphabet.size" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.Alphabet.ThreeLetterProtein">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Alphabet.</code><code class="sig-name descname">ThreeLetterProtein</code><a class="headerlink" href="#Bio.Alphabet.ThreeLetterProtein" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Alphabet.Alphabet" title="Bio.Alphabet.Alphabet"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Alphabet.Alphabet</span></code></a></p>
<p>Three letter protein alphabet.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">contains</span></code>(self, other)</p></td>
<td><p>Test if the other alphabet is contained in this one (OBSOLETE?).</p></td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="Bio.Alphabet.ThreeLetterProtein.letters">
<code class="sig-name descname">letters</code><em class="property"> = ['Ala', 'Asx', 'Cys', 'Asp', 'Glu', 'Phe', 'Gly', 'His', 'Ile', 'Lys', 'Leu', 'Met', 'Asn', 'Pro', 'Gln', 'Arg', 'Ser', 'Thr', 'Sec', 'Val', 'Trp', 'Xaa', 'Tyr', 'Glx']</em><a class="headerlink" href="#Bio.Alphabet.ThreeLetterProtein.letters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Bio.Alphabet.ThreeLetterProtein.size">
<code class="sig-name descname">size</code><em class="property"> = 3</em><a class="headerlink" href="#Bio.Alphabet.ThreeLetterProtein.size" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.Alphabet.IUPAC.html" class="btn btn-neutral float-right" title="Bio.Alphabet.IUPAC module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.AlignIO.StockholmIO.html" class="btn btn-neutral float-left" title="Bio.AlignIO.StockholmIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2017, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>