<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.codonalign.codonseq module &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/biopython.css?v=d1305275" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=49cfa8a4"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.motifs package" href="Bio.motifs.html" />
    <link rel="prev" title="Bio.codonalign.codonalignment module" href="Bio.codonalign.codonalignment.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.codonalign.html">Bio.codonalign package</a></li>
      <li class="breadcrumb-item active">Bio.codonalign.codonseq module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/codonalign/codonseq.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.codonalign.codonalignment.html" class="btn btn-neutral float-left" title="Bio.codonalign.codonalignment module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.motifs.html" class="btn btn-neutral float-right" title="Bio.motifs package" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.codonalign.codonseq">
<span id="bio-codonalign-codonseq-module"></span><h1>Bio.codonalign.codonseq module<a class="headerlink" href="#module-Bio.codonalign.codonseq" title="Link to this heading">ÔÉÅ</a></h1>
<p>Code for dealing with coding sequence.</p>
<p>CodonSeq class is inherited from Seq class. This is the core class to
deal with sequences in CodonAlignment in biopython.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.codonalign.codonseq.</span></span><span class="sig-name descname"><span class="pre">CodonSeq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rf_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Seq.html#Bio.Seq.Seq" title="Bio.Seq.Seq"><code class="xref py py-class docutils literal notranslate"><span class="pre">Seq</span></code></a></p>
<p>CodonSeq is designed to be within the SeqRecords of a CodonAlignment class.</p>
<p>CodonSeq is useful as it allows the user to specify
reading frame when translate CodonSeq</p>
<p>CodonSeq also accepts codon style slice by calling
get_codon() method.</p>
<p><strong>Important:</strong> Ungapped CodonSeq can be any length if you
specify the rf_table. Gapped CodonSeq should be a
multiple of three.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">codonseq</span> <span class="o">=</span> <span class="n">CodonSeq</span><span class="p">(</span><span class="s2">&quot;AAATTTGGGCCAAATTT&quot;</span><span class="p">,</span> <span class="n">rf_table</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">14</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">codonseq</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>
<span class="go">KFGAKF</span>
</pre></div>
</div>
<p>test get_full_rf_table method</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">CodonSeq</span><span class="p">(</span><span class="s1">&#39;AAATTTCCCGG-TGGGTTTAA&#39;</span><span class="p">,</span> <span class="n">rf_table</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">17</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">full_rf_table</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">get_full_rf_table</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">full_rf_table</span><span class="p">)</span>
<span class="go">[0, 3, 6, 9, 12, 15, 18]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">rf_table</span><span class="o">=</span><span class="n">full_rf_table</span><span class="p">,</span> <span class="n">ungap_seq</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="go">KFPPWV*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">CodonSeq</span><span class="p">(</span><span class="s1">&#39;AAATTTCCCGGGAA-TTTTAA&#39;</span><span class="p">,</span> <span class="n">rf_table</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span> <span class="mi">17</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">get_full_rf_table</span><span class="p">())</span>
<span class="go">[0, 3, 6, 9, 12.0, 15, 18]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">CodonSeq</span><span class="p">(</span><span class="s1">&#39;AAA------------TAA&#39;</span><span class="p">,</span> <span class="n">rf_table</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">get_full_rf_table</span><span class="p">())</span>
<span class="go">[0, 3.0, 6.0, 9.0, 12.0, 15]</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rf_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.__init__" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.get_codon">
<span class="sig-name descname"><span class="pre">get_codon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.get_codon" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Get the index codon from the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.get_codon_num">
<span class="sig-name descname"><span class="pre">get_codon_num</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.get_codon_num" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Return the number of codons in the CodonSeq.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'*'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rf_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ungap_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.translate" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Translate the CodonSeq based on the reading frame in rf_table.</p>
<p>It is possible for the user to specify
a rf_table at this point. If you want to include
gaps in the translated sequence, this is the only
way. ungap_seq should be set to true for this
purpose.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.toSeq">
<span class="sig-name descname"><span class="pre">toSeq</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.toSeq" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Convert DNA to seq object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.get_full_rf_table">
<span class="sig-name descname"><span class="pre">get_full_rf_table</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.get_full_rf_table" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Return full rf_table of the CodonSeq records.</p>
<p>A full rf_table is different from a normal rf_table in that
it translate gaps in CodonSeq. It is helpful to construct
alignment containing frameshift.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.full_translate">
<span class="sig-name descname"><span class="pre">full_translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'*'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.full_translate" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Apply full translation with gaps considered.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.ungap">
<span class="sig-name descname"><span class="pre">ungap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.ungap" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Return a copy of the sequence without the gap character(s).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.from_seq">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rf_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.from_seq" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Get codon sequence from sequence data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.CodonSeq.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.codonalign.codonseq.CodonSeq.__abstractmethods__" title="Link to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.codonalign.codonseq.cal_dn_ds">
<span class="sig-prename descclassname"><span class="pre">Bio.codonalign.codonseq.</span></span><span class="sig-name descname"><span class="pre">cal_dn_ds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codon_seq2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'NG86'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codon_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cfreq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonseq.cal_dn_ds" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Calculate dN and dS of the given two sequences.</p>
<dl class="simple">
<dt>Available methods:</dt><dd><ul class="simple">
<li><p>NG86  - <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/3444411">Nei and Gojobori (1986)</a> (PMID 3444411).</p></li>
<li><p>LWL85 - <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/3916709">Li et al. (1985)</a> (PMID 3916709).</p></li>
<li><p>ML    - <a class="reference external" href="http://mbe.oxfordjournals.org/content/11/5/725">Goldman and Yang (1994)</a> (PMID 7968486).</p></li>
<li><p>YN00  - <a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a026236">Yang and Nielsen (2000)</a> (PMID 10666704).</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>codon_seq1 - CodonSeq or or SeqRecord that contains a CodonSeq</p></li>
<li><p>codon_seq2 - CodonSeq or or SeqRecord that contains a CodonSeq</p></li>
<li><p>w  - transition/transversion ratio</p></li>
<li><p>cfreq - Current codon frequency vector can only be specified
when you are using ML method. Possible ways of
getting cfreq are: F1x4, F3x4 and F61.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.codonalign.codonalignment.html" class="btn btn-neutral float-left" title="Bio.codonalign.codonalignment module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.motifs.html" class="btn btn-neutral float-right" title="Bio.motifs package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>