<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.TogoWS package &mdash; Biopython 1.80.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.UniGene package" href="Bio.UniGene.html" />
    <link rel="prev" title="Bio.SwissProt.KeyWList module" href="Bio.SwissProt.KeyWList.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
            <a href="index.html">
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.TogoWS package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-Bio.TogoWS">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
      <li>Bio.TogoWS package</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/TogoWS/__init__.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.SwissProt.KeyWList.html" class="btn btn-neutral float-left" title="Bio.SwissProt.KeyWList module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.UniGene.html" class="btn btn-neutral float-right" title="Bio.UniGene package" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bio-togows-package">
<h1>Bio.TogoWS package<a class="headerlink" href="#bio-togows-package" title="Permalink to this headline"></a></h1>
<section id="module-Bio.TogoWS">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.TogoWS" title="Permalink to this headline"></a></h2>
<p>Provides code to access the TogoWS integrated websevices of DBCLS, Japan.</p>
<p>This module aims to make the TogoWS (from DBCLS, Japan) easier to use. See:
<a class="reference external" href="http://togows.dbcls.jp/">http://togows.dbcls.jp/</a></p>
<p>The TogoWS REST service provides simple access to a range of databases, acting
as a proxy to shield you from all the different provider APIs. This works using
simple URLs (which this module will construct for you). For more details, see
<a class="reference external" href="http://togows.dbcls.jp/site/en/rest.html">http://togows.dbcls.jp/site/en/rest.html</a></p>
<p>The functionality is somewhat similar to Biopython’s Bio.Entrez module which
provides access to the NCBI’s Entrez Utilities (E-Utils) which also covers a
wide range of databases.</p>
<p>Currently TogoWS does not provide any usage guidelines (unlike the NCBI whose
requirements are reasonably clear). To avoid risking overloading the service,
Biopython will only allow three calls per second.</p>
<p>The TogoWS SOAP service offers a more complex API for calling web services
(essentially calling remote functions) provided by DDBJ, KEGG and PDBj. For
example, this allows you to run a remote BLAST search at the DDBJ. This is
not yet covered by this module, however there are lots of Python examples
on the TogoWS website using the SOAPpy python library. See:
<a class="reference external" href="http://togows.dbcls.jp/site/en/soap.html">http://togows.dbcls.jp/site/en/soap.html</a>
<a class="reference external" href="http://soapy.sourceforge.net/">http://soapy.sourceforge.net/</a></p>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.TogoWS.entry">
<span class="sig-prename descclassname"><span class="pre">Bio.TogoWS.</span></span><span class="sig-name descname"><span class="pre">entry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.TogoWS.entry" title="Permalink to this definition"></a></dt>
<dd><p>Call TogoWS ‘entry’ to fetch a record.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>db - database (string), see list below.</p></li>
<li><p>id - identier (string) or a list of identifiers (either as a list of
strings or a single string with comma separators).</p></li>
<li><p>format - return data file format (string), options depend on the database
e.g. “xml”, “json”, “gff”, “fasta”, “ttl” (RDF Turtle)</p></li>
<li><p>field - specific field from within the database record (string)
e.g. “au” or “authors” for pubmed.</p></li>
</ul>
</dd>
</dl>
<p>At the time of writing, this includes the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">KEGG</span><span class="p">:</span> <span class="n">compound</span><span class="p">,</span> <span class="n">drug</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">glycan</span><span class="p">,</span> <span class="n">orthology</span><span class="p">,</span> <span class="n">reaction</span><span class="p">,</span>
      <span class="n">module</span><span class="p">,</span> <span class="n">pathway</span>
<span class="n">DDBj</span><span class="p">:</span> <span class="n">ddbj</span><span class="p">,</span> <span class="n">dad</span><span class="p">,</span> <span class="n">pdb</span>
<span class="n">NCBI</span><span class="p">:</span> <span class="n">nuccore</span><span class="p">,</span> <span class="n">nucest</span><span class="p">,</span> <span class="n">nucgss</span><span class="p">,</span> <span class="n">nucleotide</span><span class="p">,</span> <span class="n">protein</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">onim</span><span class="p">,</span>
      <span class="n">homologue</span><span class="p">,</span> <span class="n">snp</span><span class="p">,</span> <span class="n">mesh</span><span class="p">,</span> <span class="n">pubmed</span>
<span class="n">EBI</span><span class="p">:</span>  <span class="n">embl</span><span class="p">,</span> <span class="n">uniprot</span><span class="p">,</span> <span class="n">uniparc</span><span class="p">,</span> <span class="n">uniref100</span><span class="p">,</span> <span class="n">uniref90</span><span class="p">,</span> <span class="n">uniref50</span>
</pre></div>
</div>
<p>For the current list, please see <a class="reference external" href="http://togows.dbcls.jp/entry/">http://togows.dbcls.jp/entry/</a></p>
<p>This function is essentially equivalent to the NCBI Entrez service
EFetch, available in Biopython as Bio.Entrez.efetch(…), but that
does not offer field extraction.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.TogoWS.search_count">
<span class="sig-prename descclassname"><span class="pre">Bio.TogoWS.</span></span><span class="sig-name descname"><span class="pre">search_count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.TogoWS.search_count" title="Permalink to this definition"></a></dt>
<dd><p>Call TogoWS search count to see how many matches a search gives.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>db - database (string), see <a class="reference external" href="http://togows.dbcls.jp/search">http://togows.dbcls.jp/search</a></p></li>
<li><p>query - search term (string)</p></li>
</ul>
</dd>
</dl>
<p>You could then use the count to download a large set of search results in
batches using the offset and limit options to Bio.TogoWS.search(). In
general however the Bio.TogoWS.search_iter() function is simpler to use.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.TogoWS.search_iter">
<span class="sig-prename descclassname"><span class="pre">Bio.TogoWS.</span></span><span class="sig-name descname"><span class="pre">search_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.TogoWS.search_iter" title="Permalink to this definition"></a></dt>
<dd><p>Call TogoWS search iterating over the results (generator function).</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>db - database (string), see <a class="reference external" href="http://togows.dbcls.jp/search">http://togows.dbcls.jp/search</a></p></li>
<li><p>query - search term (string)</p></li>
<li><p>limit - optional upper bound on number of search results</p></li>
<li><p>batch - number of search results to pull back each time talk to
TogoWS (currently limited to 100).</p></li>
</ul>
</dd>
</dl>
<p>You would use this function within a for loop, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">TogoWS</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">TogoWS</span><span class="o">.</span><span class="n">search_iter</span><span class="p">(</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="s2">&quot;diabetes+human&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;PubMed ID: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span><span class="nb">id</span><span class="p">)</span> <span class="c1"># maybe fetch data with entry?</span>
<span class="go">PubMed ID: ...</span>
</pre></div>
</div>
<p>Internally this first calls the Bio.TogoWS.search_count() and then
uses Bio.TogoWS.search() to get the results in batches.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.TogoWS.search">
<span class="sig-prename descclassname"><span class="pre">Bio.TogoWS.</span></span><span class="sig-name descname"><span class="pre">search</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.TogoWS.search" title="Permalink to this definition"></a></dt>
<dd><p>Call TogoWS search.</p>
<p>This is a low level wrapper for the TogoWS search function, which
can return results in a several formats. In general, the search_iter
function is more suitable for end users.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>db - database (string), see <a class="reference external" href="http://togows.dbcls.jp/search/">http://togows.dbcls.jp/search/</a></p></li>
<li><p>query - search term (string)</p></li>
<li><p>offset, limit - optional integers specifying which result to start from
(1 based) and the number of results to return.</p></li>
<li><p>format - return data file format (string), e.g. “json”, “ttl” (RDF)
By default plain text is returned, one result per line.</p></li>
</ul>
</dd>
</dl>
<p>At the time of writing, TogoWS applies a default count limit of 100
search results, and this is an upper bound. To access more results,
use the offset argument or the search_iter(…) function.</p>
<p>TogoWS supports a long list of databases, including many from the NCBI
(e.g. “ncbi-pubmed” or “pubmed”, “ncbi-genbank” or “genbank”, and
“ncbi-taxonomy”), EBI (e.g. “ebi-ebml” or “embl”, “ebi-uniprot” or
“uniprot, “ebi-go”), and KEGG (e.g. “kegg-compound” or “compound”).
For the current list, see <a class="reference external" href="http://togows.dbcls.jp/search/">http://togows.dbcls.jp/search/</a></p>
<p>The NCBI provide the Entrez Search service (ESearch) which is similar,
available in Biopython as the Bio.Entrez.esearch() function.</p>
<p>See also the function Bio.TogoWS.search_count() which returns the number
of matches found, and the Bio.TogoWS.search_iter() function which allows
you to iterate over the search results (taking care of batching for you).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.TogoWS.convert">
<span class="sig-prename descclassname"><span class="pre">Bio.TogoWS.</span></span><span class="sig-name descname"><span class="pre">convert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">in_format</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_format</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.TogoWS.convert" title="Permalink to this definition"></a></dt>
<dd><p>Call TogoWS for file format conversion.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>data - string or handle containing input record(s)</p></li>
<li><p>in_format - string describing the input file format (e.g. “genbank”)</p></li>
<li><p>out_format - string describing the requested output format (e.g. “fasta”)</p></li>
</ul>
</dd>
</dl>
<p>For a list of supported conversions (e.g. “genbank” to “fasta”), see
<a class="reference external" href="http://togows.dbcls.jp/convert/">http://togows.dbcls.jp/convert/</a></p>
<p>Note that Biopython has built in support for conversion of sequence and
alignnent file formats (functions Bio.SeqIO.convert and Bio.AlignIO.convert)</p>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.SwissProt.KeyWList.html" class="btn btn-neutral float-left" title="Bio.SwissProt.KeyWList module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.UniGene.html" class="btn btn-neutral float-right" title="Bio.UniGene package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2020, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.80.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.78/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>