

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Bio.UniProt.GOA module &mdash; Biopython 1.79.dev0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Wise package" href="Bio.Wise.html" />
    <link rel="prev" title="Bio.UniProt package" href="Bio.UniProt.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.UniProt.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html#module-Bio.UniProt">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          



















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.UniProt.html">Bio.UniProt package</a> &raquo;</li>
        
      <li>Bio.UniProt.GOA module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/UniProt/GOA.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Wise.html" class="btn btn-neutral float-right" title="Bio.Wise package" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
      
      
        <a href="Bio.UniProt.html" class="btn btn-neutral float-left" title="Bio.UniProt package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.UniProt.GOA">
<span id="bio-uniprot-goa-module"></span><h1>Bio.UniProt.GOA module<a class="headerlink" href="#module-Bio.UniProt.GOA" title="Permalink to this headline">¶</a></h1>
<p>Parsers for the GAF, GPA and GPI formats from UniProt-GOA.</p>
<p>Uniprot-GOA README + GAF format description:
<a class="reference external" href="ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/README">ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/README</a></p>
<p>GAF formats:
<a class="reference external" href="http://geneontology.org/docs/go-annotation-file-gaf-format-2.1/">http://geneontology.org/docs/go-annotation-file-gaf-format-2.1/</a>
<a class="reference external" href="http://geneontology.org/docs/go-annotation-file-gaf-format-2.0/">http://geneontology.org/docs/go-annotation-file-gaf-format-2.0/</a></p>
<p>gp_association (GPA format) README:
<a class="reference external" href="http://geneontology.org/docs/gene-product-association-data-gpad-format/">http://geneontology.org/docs/gene-product-association-data-gpad-format/</a></p>
<p>gp_information (GPI format) README:
<a class="reference external" href="http://geneontology.org/docs/gene-product-information-gpi-format/">http://geneontology.org/docs/gene-product-information-gpi-format/</a></p>
<dl class="py function">
<dt id="Bio.UniProt.GOA.gpi_iterator">
<code class="sig-prename descclassname">Bio.UniProt.GOA.</code><code class="sig-name descname">gpi_iterator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniProt.GOA.gpi_iterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Read GPI format files.</p>
<p>This function should be called to read a
gp_information.goa_uniprot file. At the moment, there is
only one format, but this may change, so
this function is a placeholder a future wrapper.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.UniProt.GOA.gpa_iterator">
<code class="sig-prename descclassname">Bio.UniProt.GOA.</code><code class="sig-name descname">gpa_iterator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniProt.GOA.gpa_iterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Read GPA format files.</p>
<p>This function should be called to read a
gene_association.goa_uniprot file. Reads the first record and
returns a gpa 1.1 or a gpa 1.0 iterator as needed</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.UniProt.GOA.gafbyproteiniterator">
<code class="sig-prename descclassname">Bio.UniProt.GOA.</code><code class="sig-name descname">gafbyproteiniterator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniProt.GOA.gafbyproteiniterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over records in a gene association file.</p>
<p>Returns a list of all consecutive records with the same DB_Object_ID
This function should be called to read a
gene_association.goa_uniprot file. Reads the first record and
returns a gaf 2.0 or a gaf 1.0 iterator as needed
2016-04-09: added GAF 2.1 iterator &amp; fixed bug in iterator assignment
In the meantime GAF 2.1 uses the GAF 2.0 iterator</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.UniProt.GOA.gafiterator">
<code class="sig-prename descclassname">Bio.UniProt.GOA.</code><code class="sig-name descname">gafiterator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniProt.GOA.gafiterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over a GAF 1.0 or 2.0 file.</p>
<p>This function should be called to read a
gene_association.goa_uniprot file. Reads the first record and
returns a gaf 2.0 or a gaf 1.0 iterator as needed</p>
<p>Example: open, read, interat and filter results.</p>
<p>Original data file has been trimed to ~600 rows.</p>
<p>Original source <a class="reference external" href="ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/YEAST/goa_yeast.gaf.gz">ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/YEAST/goa_yeast.gaf.gz</a></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.UniProt.GOA</span> <span class="kn">import</span> <span class="n">gafiterator</span><span class="p">,</span> <span class="n">record_has</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Evidence</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Evidence&#39;</span><span class="p">:</span> <span class="nb">set</span><span class="p">([</span><span class="s1">&#39;ND&#39;</span><span class="p">])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Synonym</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Synonym&#39;</span><span class="p">:</span> <span class="nb">set</span><span class="p">([</span><span class="s1">&#39;YA19A_YEAST&#39;</span><span class="p">,</span> <span class="s1">&#39;YAL019W-A&#39;</span><span class="p">])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Taxon_ID</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Taxon_ID&#39;</span><span class="p">:</span> <span class="nb">set</span><span class="p">([</span><span class="s1">&#39;taxon:559292&#39;</span><span class="p">])}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;UniProt/goa_yeast.gaf&#39;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">gafiterator</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>        <span class="k">if</span> <span class="n">record_has</span><span class="p">(</span><span class="n">rec</span><span class="p">,</span> <span class="n">Taxon_ID</span><span class="p">)</span> <span class="ow">and</span> <span class="n">record_has</span><span class="p">(</span><span class="n">rec</span><span class="p">,</span> <span class="n">Evidence</span><span class="p">)</span> <span class="ow">and</span> <span class="n">record_has</span><span class="p">(</span><span class="n">rec</span><span class="p">,</span> <span class="n">Synonym</span><span class="p">):</span>
<span class="gp">... </span>            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;DB_Object_Name&#39;</span><span class="p">,</span> <span class="s1">&#39;Evidence&#39;</span><span class="p">,</span> <span class="s1">&#39;Synonym&#39;</span><span class="p">,</span> <span class="s1">&#39;Taxon_ID&#39;</span><span class="p">):</span>
<span class="gp">... </span>                <span class="nb">print</span><span class="p">(</span><span class="n">rec</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">Putative uncharacterized protein YAL019W-A</span>
<span class="go">ND</span>
<span class="go">[&#39;YA19A_YEAST&#39;, &#39;YAL019W-A&#39;]</span>
<span class="go">[&#39;taxon:559292&#39;]</span>
<span class="go">Putative uncharacterized protein YAL019W-A</span>
<span class="go">ND</span>
<span class="go">[&#39;YA19A_YEAST&#39;, &#39;YAL019W-A&#39;]</span>
<span class="go">[&#39;taxon:559292&#39;]</span>
<span class="go">Putative uncharacterized protein YAL019W-A</span>
<span class="go">ND</span>
<span class="go">[&#39;YA19A_YEAST&#39;, &#39;YAL019W-A&#39;]</span>
<span class="go">[&#39;taxon:559292&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt id="Bio.UniProt.GOA.writerec">
<code class="sig-prename descclassname">Bio.UniProt.GOA.</code><code class="sig-name descname">writerec</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">outrec</span></em>, <em class="sig-param"><span class="n">handle</span></em>, <em class="sig-param"><span class="n">fields</span><span class="o">=</span><span class="default_value">['DB', 'DB_Object_ID', 'DB_Object_Symbol', 'Qualifier', 'GO_ID', 'DB:Reference', 'Evidence', 'With', 'Aspect', 'DB_Object_Name', 'Synonym', 'DB_Object_Type', 'Taxon_ID', 'Date', 'Assigned_By', 'Annotation_Extension', 'Gene_Product_Form_ID']</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniProt.GOA.writerec" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a single UniProt-GOA record to an output stream.</p>
<p>Caller should know the  format version. Default: gaf-2.0
If header has a value, then it is assumed this is the first record,
a header is written.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.UniProt.GOA.writebyproteinrec">
<code class="sig-prename descclassname">Bio.UniProt.GOA.</code><code class="sig-name descname">writebyproteinrec</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">outprotrec</span></em>, <em class="sig-param"><span class="n">handle</span></em>, <em class="sig-param"><span class="n">fields</span><span class="o">=</span><span class="default_value">['DB', 'DB_Object_ID', 'DB_Object_Symbol', 'Qualifier', 'GO_ID', 'DB:Reference', 'Evidence', 'With', 'Aspect', 'DB_Object_Name', 'Synonym', 'DB_Object_Type', 'Taxon_ID', 'Date', 'Assigned_By', 'Annotation_Extension', 'Gene_Product_Form_ID']</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniProt.GOA.writebyproteinrec" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a list of GAF records to an output stream.</p>
<p>Caller should know the  format version. Default: gaf-2.0
If header has a value, then it is assumed this is the first record,
a header is written. Typically the list is the one read by fafbyproteinrec, which
contains all consecutive lines with the same DB_Object_ID</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.UniProt.GOA.record_has">
<code class="sig-prename descclassname">Bio.UniProt.GOA.</code><code class="sig-name descname">record_has</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">inrec</span></em>, <em class="sig-param"><span class="n">fieldvals</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniProt.GOA.record_has" title="Permalink to this definition">¶</a></dt>
<dd><p>Accept a record, and a dictionary of field values.</p>
<p>The format is {‘field_name’: set([val1, val2])}.
If any field in the record has  a matching value, the function returns
True. Otherwise, returns False.</p>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="Bio.Wise.html" class="btn btn-neutral float-right" title="Bio.Wise package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="Bio.UniProt.html" class="btn btn-neutral float-left" title="Bio.UniProt package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 1999-2020, The Biopython Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.79.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.77/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>