

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.SeqUtils.IsoelectricPoint module &mdash; Biopython 1.77.dev0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SeqUtils.MeltingTemp module" href="Bio.SeqUtils.MeltingTemp.html" />
    <link rel="prev" title="Bio.SeqUtils.CodonUsageIndices module" href="Bio.SeqUtils.CodonUsageIndices.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.77.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SeqUtils.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html#module-Bio.SeqUtils">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.SeqUtils.html">Bio.SeqUtils package</a> &raquo;</li>
        
      <li>Bio.SeqUtils.IsoelectricPoint module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SeqUtils/IsoelectricPoint.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SeqUtils.MeltingTemp.html" class="btn btn-neutral float-right" title="Bio.SeqUtils.MeltingTemp module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SeqUtils.CodonUsageIndices.html" class="btn btn-neutral float-left" title="Bio.SeqUtils.CodonUsageIndices module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.SeqUtils.IsoelectricPoint">
<span id="bio-sequtils-isoelectricpoint-module"></span><h1>Bio.SeqUtils.IsoelectricPoint module<a class="headerlink" href="#module-Bio.SeqUtils.IsoelectricPoint" title="Permalink to this headline">¶</a></h1>
<p>Calculate isoelectric points of polypeptides using methods of Bjellqvist.</p>
<p>pK values and the methos are taken from:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">*</span> <span class="n">Bjellqvist</span><span class="p">,</span> <span class="n">B</span><span class="o">.</span><span class="p">,</span><span class="n">Hughes</span><span class="p">,</span> <span class="n">G</span><span class="o">.</span><span class="n">J</span><span class="o">.</span><span class="p">,</span> <span class="n">Pasquali</span><span class="p">,</span> <span class="n">Ch</span><span class="o">.</span><span class="p">,</span> <span class="n">Paquet</span><span class="p">,</span> <span class="n">N</span><span class="o">.</span><span class="p">,</span> <span class="n">Ravier</span><span class="p">,</span> <span class="n">F</span><span class="o">.</span><span class="p">,</span>
<span class="n">Sanchez</span><span class="p">,</span> <span class="n">J</span><span class="o">.-</span><span class="n">Ch</span><span class="o">.</span><span class="p">,</span> <span class="n">Frutiger</span><span class="p">,</span> <span class="n">S</span><span class="o">.</span> <span class="o">&amp;</span> <span class="n">Hochstrasser</span><span class="p">,</span> <span class="n">D</span><span class="o">.</span><span class="n">F</span><span class="o">.</span>
<span class="n">The</span> <span class="n">focusing</span> <span class="n">positions</span> <span class="n">of</span> <span class="n">polypeptides</span> <span class="ow">in</span> <span class="n">immobilized</span> <span class="n">pH</span> <span class="n">gradients</span> <span class="n">can</span> <span class="n">be</span>
<span class="n">predicted</span> <span class="kn">from</span> <span class="nn">their</span> <span class="n">amino</span> <span class="n">acid</span> <span class="n">sequences</span><span class="o">.</span> <span class="n">Electrophoresis</span> <span class="mi">1993</span><span class="p">,</span> <span class="mi">14</span><span class="p">,</span>
<span class="mi">1023</span><span class="o">-</span><span class="mf">1031.</span>

<span class="o">*</span> <span class="n">Bjellqvist</span><span class="p">,</span> <span class="n">B</span><span class="o">.</span><span class="p">,</span> <span class="n">Basse</span><span class="p">,</span> <span class="n">B</span><span class="o">.</span><span class="p">,</span> <span class="n">Olsen</span><span class="p">,</span> <span class="n">E</span><span class="o">.</span> <span class="ow">and</span> <span class="n">Celis</span><span class="p">,</span> <span class="n">J</span><span class="o">.</span><span class="n">E</span><span class="o">.</span>
<span class="n">Reference</span> <span class="n">points</span> <span class="k">for</span> <span class="n">comparisons</span> <span class="n">of</span> <span class="n">two</span><span class="o">-</span><span class="n">dimensional</span> <span class="n">maps</span> <span class="n">of</span> <span class="n">proteins</span> <span class="kn">from</span>
<span class="nn">different</span> <span class="n">human</span> <span class="n">cell</span> <span class="n">types</span> <span class="n">defined</span> <span class="ow">in</span> <span class="n">a</span> <span class="n">pH</span> <span class="n">scale</span> <span class="n">where</span> <span class="n">isoelectric</span> <span class="n">points</span>
<span class="n">correlate</span> <span class="k">with</span> <span class="n">polypeptide</span> <span class="n">compositions</span><span class="o">.</span> <span class="n">Electrophoresis</span> <span class="mi">1994</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">529</span><span class="o">-</span><span class="mf">539.</span>
</pre></div>
</div>
<p>I designed the algorithm according to a note by David L. Tabb, available at:
<a class="reference external" href="http://fields.scripps.edu/DTASelect/20010710-pI-Algorithm.pdf">http://fields.scripps.edu/DTASelect/20010710-pI-Algorithm.pdf</a></p>
<dl class="py class">
<dt id="Bio.SeqUtils.IsoelectricPoint.IsoelectricPoint">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SeqUtils.IsoelectricPoint.</code><code class="sig-name descname">IsoelectricPoint</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">protein_sequence</span></em>, <em class="sig-param"><span class="n">aa_content</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.IsoelectricPoint.IsoelectricPoint" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class for calculating the IEP or charge at given pH of a protein.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>:protein_sequence: A ``Bio.Seq`` or string object containing a protein</strong></dt><dd><p>sequence.</p>
</dd>
<dt><strong>:aa_content: A dictionary with amino acid letters as keys and it’s</strong></dt><dd><p>occurences as integers, e.g. <code class="docutils literal notranslate"><span class="pre">{&quot;A&quot;:</span> <span class="pre">3,</span> <span class="pre">&quot;C&quot;:</span> <span class="pre">0,</span> <span class="pre">...}</span></code>.
Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the dic will be calculated
from the given sequence.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>The methods of this class can either be accessed from the class itself
or from a <code class="docutils literal notranslate"><span class="pre">ProtParam.ProteinAnalysis</span></code> object (with partially different
names):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils.IsoelectricPoint</span> <span class="kn">import</span> <span class="n">IsoelectricPoint</span> <span class="k">as</span> <span class="n">IP</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein</span> <span class="o">=</span> <span class="n">IP</span><span class="p">(</span><span class="s2">&quot;INGAR&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;IEP of peptide </span><span class="si">{</span><span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="s2"> is </span><span class="si">{</span><span class="n">protein</span><span class="o">.</span><span class="n">pi</span><span class="p">()</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">IEP of peptide INGAR is 9.75</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;It&#39;s charge at pH 7 is </span><span class="si">{</span><span class="n">protein</span><span class="o">.</span><span class="n">charge_at_pH</span><span class="p">(</span><span class="mf">7.0</span><span class="p">)</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">It&#39;s charge at pH 7 is 0.76</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils.ProtParam</span> <span class="kn">import</span> <span class="n">ProteinAnalysis</span> <span class="k">as</span> <span class="n">PA</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein</span> <span class="o">=</span> <span class="n">PA</span><span class="p">(</span><span class="s2">&quot;PETER&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;IEP of </span><span class="si">{</span><span class="n">protein</span><span class="o">.</span><span class="n">sequence</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">protein</span><span class="o">.</span><span class="n">isoelectric_point</span><span class="p">()</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">IEP of PETER: 4.53</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Charge at pH 4.53: </span><span class="si">{</span><span class="n">protein</span><span class="o">.</span><span class="n">charge_at_pH</span><span class="p">(</span><span class="mf">4.53</span><span class="p">)</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">Charge at pH 4.53: 0.00</span>
</pre></div>
</div>
<p class="rubric">Methods</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 88%" />
<col style="width: 12%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>:charge_at_pH(pH):  Calculates the charge of the protein for a given pH</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>:pi():              Calculates the isoelectric point</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="Bio.SeqUtils.IsoelectricPoint.IsoelectricPoint.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">protein_sequence</span></em>, <em class="sig-param"><span class="n">aa_content</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.IsoelectricPoint.IsoelectricPoint.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqUtils.IsoelectricPoint.IsoelectricPoint.charge_at_pH">
<code class="sig-name descname">charge_at_pH</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">pH</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.IsoelectricPoint.IsoelectricPoint.charge_at_pH" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the charge of a protein at given pH.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SeqUtils.IsoelectricPoint.IsoelectricPoint.pi">
<code class="sig-name descname">pi</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">pH</span><span class="o">=</span><span class="default_value">7.775</span></em>, <em class="sig-param"><span class="n">min_</span><span class="o">=</span><span class="default_value">4.05</span></em>, <em class="sig-param"><span class="n">max_</span><span class="o">=</span><span class="default_value">12</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.IsoelectricPoint.IsoelectricPoint.pi" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate and return the isoelectric point as float.</p>
<p>This is a recursive function that uses bisection method.
Wiki on bisection: <a class="reference external" href="https://en.wikipedia.org/wiki/Bisection_method">https://en.wikipedia.org/wiki/Bisection_method</a></p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>pH: the pH at which the current charge of the protein is computed.
This pH lies at the centre of the interval (mean of <cite>min_</cite> and <cite>max_</cite>).</p></li>
<li><p>min_: the minimum of the interval. Initial value defaults to 4.05,
which is below the theoretical minimum, when the protein is composed
exclusively of aspartate.</p></li>
<li><p>max_: the maximum of the the interval. Initial value defaults to 12,
which is above the theoretical maximum, when the protein is composed
exclusively of arginine.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.SeqUtils.MeltingTemp.html" class="btn btn-neutral float-right" title="Bio.SeqUtils.MeltingTemp module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SeqUtils.CodonUsageIndices.html" class="btn btn-neutral float-left" title="Bio.SeqUtils.CodonUsageIndices module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2019, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>