<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.Graphics.BasicChromosome module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.Graphics.ColorSpiral module" href="Bio.Graphics.ColorSpiral.html" />
    <link rel="prev" title="Bio.Graphics.GenomeDiagram package" href="Bio.Graphics.GenomeDiagram.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.Graphics.html">Bio.Graphics package</a></li>
      <li class="breadcrumb-item active">Bio.Graphics.BasicChromosome module</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/api/Bio.Graphics.BasicChromosome.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Graphics.GenomeDiagram.html" class="btn btn-neutral float-left" title="Bio.Graphics.GenomeDiagram package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Graphics.ColorSpiral.html" class="btn btn-neutral float-right" title="Bio.Graphics.ColorSpiral module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.Graphics.BasicChromosome">
<span id="bio-graphics-basicchromosome-module"></span><h1>Bio.Graphics.BasicChromosome module<a class="headerlink" href="#module-Bio.Graphics.BasicChromosome" title="Permalink to this headline">ÔÉÅ</a></h1>
<p>Draw representations of organism chromosomes with added information.</p>
<p>These classes are meant to model the drawing of pictures of chromosomes.
This can be useful for lots of things, including displaying markers on
a chromosome (ie. for genetic mapping) and showing syteny between two
chromosomes.</p>
<p>The structure of these classes is intended to be a Composite, so that
it will be easy to plug in and switch different parts without
breaking the general drawing capabilities of the system. The
relationship between classes is that everything derives from
_ChromosomeComponent, which specifies the overall interface. The parts
then are related so that an Organism contains Chromosomes, and these
Chromosomes contain ChromosomeSegments. This representation differs
from the canonical composite structure in that we don‚Äôt really have
‚Äòleaf‚Äô nodes here ‚Äì all components can potentially hold sub-components.</p>
<p>Most of the time the ChromosomeSegment class is what you‚Äôll want to
customize for specific drawing tasks.</p>
<p>For providing drawing capabilities, these classes use reportlab:</p>
<p><a class="reference external" href="http://www.reportlab.com">http://www.reportlab.com</a></p>
<p>This provides nice output in PDF, SVG and postscript.  If you have
reportlab‚Äôs renderPM module installed you can also use PNG etc.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.Organism">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Graphics.BasicChromosome.</span></span><span class="sig-name descname"><span class="pre">Organism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.Organism" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Graphics.BasicChromosome._ChromosomeComponent</span></code></p>
<p>Top level class for drawing chromosomes.</p>
<p>This class holds information about an organism and all of its
chromosomes, and provides the top level object which could be used
for drawing a chromosome representation of an organism.</p>
<p>Chromosomes should be added and removed from the Organism via the
add and remove functions.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.Organism.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.Organism.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.Organism.draw">
<span class="sig-name descname"><span class="pre">draw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.Organism.draw" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Draw out the information for the Organism.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>output_file ‚Äì The name of a file specifying where the
document should be saved, or a handle to be written to.
The output format is set when creating the Organism object.
Alternatively, output_file=None will return the drawing using
the low-level ReportLab objects (for further processing, such
as adding additional graphics, before writing).</p></li>
<li><p>title ‚Äì The output title of the produced document.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.Chromosome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Graphics.BasicChromosome.</span></span><span class="sig-name descname"><span class="pre">Chromosome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.Chromosome" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Graphics.BasicChromosome._ChromosomeComponent</span></code></p>
<p>Class for drawing a chromosome of an organism.</p>
<p>This organizes the drawing of a single organisms chromosome. This
class can be instantiated directly, but the draw method makes the
most sense to be called in the context of an organism.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.Chromosome.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.Chromosome.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize a Chromosome for drawing.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>chromosome_name - The label for the chromosome.</p></li>
</ul>
</dd>
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>start_x_position, end_x_position - The x positions on the page
where the chromosome should be drawn. This allows multiple
chromosomes to be drawn on a single page.</p></li>
<li><p>start_y_position, end_y_position - The y positions on the page
where the chromosome should be contained.</p></li>
</ul>
</dd>
<dt>Configuration Attributes:</dt><dd><ul class="simple">
<li><p>title_size - The size of the chromosome title.</p></li>
<li><p>scale_num - A number of scale the drawing by. This is useful if
you want to draw multiple chromosomes of different sizes at the
same scale. If this is not set, then the chromosome drawing will
be scaled by the number of segments in the chromosome (so each
chromosome will be the exact same final size).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.Chromosome.subcomponent_size">
<span class="sig-name descname"><span class="pre">subcomponent_size</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.Chromosome.subcomponent_size" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the scaled size of all subcomponents of this component.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.Chromosome.draw">
<span class="sig-name descname"><span class="pre">draw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cur_drawing</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.Chromosome.draw" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Draw a chromosome on the specified template.</p>
<p>Ideally, the x_position and y_*_position attributes should be
set prior to drawing ‚Äì otherwise we‚Äôre going to have some problems.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.ChromosomeSegment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Graphics.BasicChromosome.</span></span><span class="sig-name descname"><span class="pre">ChromosomeSegment</span></span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.ChromosomeSegment" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Graphics.BasicChromosome._ChromosomeComponent</span></code></p>
<p>Draw a segment of a chromosome.</p>
<p>This class provides the important configurable functionality of drawing
a Chromosome. Each segment has some customization available here, or can
be subclassed to define additional functionality. Most of the interesting
drawing stuff is likely to happen at the ChromosomeSegment level.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.ChromosomeSegment.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.ChromosomeSegment.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize a ChromosomeSegment.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>start_x_position, end_x_position - Defines the x range we have
to draw things in.</p></li>
<li><p>start_y_position, end_y_position - Defines the y range we have
to draw things in.</p></li>
</ul>
</dd>
<dt>Configuration Attributes:</dt><dd><ul class="simple">
<li><p>scale - A scaling value for the component. By default this is
set at 1 (ie ‚Äì has the same scale as everything else). Higher
values give more size to the component, smaller values give less.</p></li>
<li><p>fill_color - A color to fill in the segment with. Colors are
available in reportlab.lib.colors</p></li>
<li><p>label - A label to place on the chromosome segment. This should
be a text string specifying what is to be included in the label.</p></li>
<li><p>label_size - The size of the label.</p></li>
<li><p>chr_percent - The percentage of area that the chromosome
segment takes up.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.ChromosomeSegment.draw">
<span class="sig-name descname"><span class="pre">draw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cur_drawing</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.ChromosomeSegment.draw" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Draw a chromosome segment.</p>
<p>Before drawing, the range we are drawing in needs to be set.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.AnnotatedChromosomeSegment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Graphics.BasicChromosome.</span></span><span class="sig-name descname"><span class="pre">AnnotatedChromosomeSegment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bp_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_feature_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">colors.blue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_qualifiers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('gene',</span> <span class="pre">'label',</span> <span class="pre">'name',</span> <span class="pre">'locus_tag',</span> <span class="pre">'product')</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.AnnotatedChromosomeSegment" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Graphics.BasicChromosome.ChromosomeSegment" title="Bio.Graphics.BasicChromosome.ChromosomeSegment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Graphics.BasicChromosome.ChromosomeSegment</span></code></a></p>
<p>Annotated chromosome segment.</p>
<p>This is like the ChromosomeSegment, but accepts a list of features.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.AnnotatedChromosomeSegment.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bp_length</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_feature_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">colors.blue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_qualifiers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('gene',</span> <span class="pre">'label',</span> <span class="pre">'name',</span> <span class="pre">'locus_tag',</span> <span class="pre">'product')</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.AnnotatedChromosomeSegment.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize.</p>
<p>The features can either be SeqFeature objects, or tuples of values:
start (int), end (int), strand (+1, -1, O or None), label (string),
ReportLab color (string or object), and optional ReportLab fill color.</p>
<p>Note we require 0 &lt;= start &lt;= end &lt;= bp_length, and within the vertical
space allocated to this segment lines will be places according to the
start/end coordinates (starting from the top).</p>
<p>Positive stand features are drawn on the right, negative on the left,
otherwise all the way across.</p>
<p>We recommend using consistent units for all the segment‚Äôs scale values
(e.g. their length in base pairs).</p>
<p>When providing features as SeqFeature objects, the default color
is used, unless the feature‚Äôs qualifiers include an Artemis colour
string (functionality also in GenomeDiagram). The caption also follows
the GenomeDiagram approach and takes the first qualifier from the list
or tuple specified in name_qualifiers.</p>
<p>Note additional attribute label_sep_percent controls the percentage of
area that the chromosome segment takes up, by default half of the
chr_percent attribute (half of 25%, thus 12.5%)</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.TelomereSegment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Graphics.BasicChromosome.</span></span><span class="sig-name descname"><span class="pre">TelomereSegment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inverted</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.TelomereSegment" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Graphics.BasicChromosome.ChromosomeSegment" title="Bio.Graphics.BasicChromosome.ChromosomeSegment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Graphics.BasicChromosome.ChromosomeSegment</span></code></a></p>
<p>A segment that is located at the end of a linear chromosome.</p>
<p>This is just like a regular segment, but it draws the end of a chromosome
which is represented by a half circle. This just overrides the
_draw_segment class of ChromosomeSegment to provide that specialized
drawing.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.TelomereSegment.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inverted</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.TelomereSegment.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize a segment at the end of a chromosome.</p>
<p>See ChromosomeSegment for all of the attributes that can be
customized in a TelomereSegments.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>inverted ‚Äì Whether or not the telomere should be inverted
(ie. drawn on the bottom of a chromosome)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.SpacerSegment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Graphics.BasicChromosome.</span></span><span class="sig-name descname"><span class="pre">SpacerSegment</span></span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.SpacerSegment" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Graphics.BasicChromosome.ChromosomeSegment" title="Bio.Graphics.BasicChromosome.ChromosomeSegment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Graphics.BasicChromosome.ChromosomeSegment</span></code></a></p>
<p>A segment that is located at the end of a linear chromosome.</p>
<p>Doesn‚Äôt draw anything, just empty space which can be helpful
for layout purposes (e.g. making room for feature labels).</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Graphics.BasicChromosome.SpacerSegment.draw">
<span class="sig-name descname"><span class="pre">draw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cur_diagram</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Graphics.BasicChromosome.SpacerSegment.draw" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Draw nothing to the current diagram (dummy method).</p>
<p>The segment spacer has no actual image in the diagram,
so this method therefore does nothing, but is defined
to match the expected API of the other segment objects.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Graphics.GenomeDiagram.html" class="btn btn-neutral float-left" title="Bio.Graphics.GenomeDiagram package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Graphics.ColorSpiral.html" class="btn btn-neutral float-right" title="Bio.Graphics.ColorSpiral module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>