<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.PDB.Polypeptide module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.PDB.Residue module" href="Bio.PDB.Residue.html" />
    <link rel="prev" title="Bio.PDB.PSEA module" href="Bio.PDB.PSEA.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PDB.html">Bio.PDB package</a></li>
      <li class="breadcrumb-item active">Bio.PDB.Polypeptide module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/PDB/Polypeptide.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.PDB.PSEA.html" class="btn btn-neutral float-left" title="Bio.PDB.PSEA module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PDB.Residue.html" class="btn btn-neutral float-right" title="Bio.PDB.Residue module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.PDB.Polypeptide">
<span id="bio-pdb-polypeptide-module"></span><h1>Bio.PDB.Polypeptide module<a class="headerlink" href="#module-Bio.PDB.Polypeptide" title="Permalink to this headline">ÔÉÅ</a></h1>
<p>Polypeptide-related classes (construction and representation).</p>
<p>Simple example with multiple chains,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.PDB.PDBParser</span> <span class="kn">import</span> <span class="n">PDBParser</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.PDB.Polypeptide</span> <span class="kn">import</span> <span class="n">PPBuilder</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">structure</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s1">&#39;2BEG&#39;</span><span class="p">,</span> <span class="s1">&#39;PDB/2BEG.pdb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ppb</span><span class="o">=</span><span class="n">PPBuilder</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">pp</span> <span class="ow">in</span> <span class="n">ppb</span><span class="o">.</span><span class="n">build_peptides</span><span class="p">(</span><span class="n">structure</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">pp</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">())</span>
<span class="go">LVFFAEDVGSNKGAIIGLMVGGVVIA</span>
<span class="go">LVFFAEDVGSNKGAIIGLMVGGVVIA</span>
<span class="go">LVFFAEDVGSNKGAIIGLMVGGVVIA</span>
<span class="go">LVFFAEDVGSNKGAIIGLMVGGVVIA</span>
<span class="go">LVFFAEDVGSNKGAIIGLMVGGVVIA</span>
</pre></div>
</div>
<p>Example with non-standard amino acids using HETATM lines in the PDB file,
in this case selenomethionine (MSE):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.PDB.PDBParser</span> <span class="kn">import</span> <span class="n">PDBParser</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.PDB.Polypeptide</span> <span class="kn">import</span> <span class="n">PPBuilder</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">structure</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s1">&#39;1A8O&#39;</span><span class="p">,</span> <span class="s1">&#39;PDB/1A8O.pdb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ppb</span><span class="o">=</span><span class="n">PPBuilder</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">pp</span> <span class="ow">in</span> <span class="n">ppb</span><span class="o">.</span><span class="n">build_peptides</span><span class="p">(</span><span class="n">structure</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">pp</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">())</span>
<span class="go">DIRQGPKEPFRDYVDRFYKTLRAEQASQEVKNW</span>
<span class="go">TETLLVQNANPDCKTILKALGPGATLEE</span>
<span class="go">TACQG</span>
</pre></div>
</div>
<p>If you want to, you can include non-standard amino acids in the peptides:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">pp</span> <span class="ow">in</span> <span class="n">ppb</span><span class="o">.</span><span class="n">build_peptides</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="n">aa_only</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">pp</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">())</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">pp</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span> <span class="n">pp</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()))</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">pp</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">()[</span><span class="o">-</span><span class="mi">7</span><span class="p">],</span> <span class="n">pp</span><span class="p">[</span><span class="o">-</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()))</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">pp</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">()[</span><span class="o">-</span><span class="mi">6</span><span class="p">],</span> <span class="n">pp</span><span class="p">[</span><span class="o">-</span><span class="mi">6</span><span class="p">]</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()))</span>
<span class="go">MDIRQGPKEPFRDYVDRFYKTLRAEQASQEVKNWMTETLLVQNANPDCKTILKALGPGATLEEMMTACQG</span>
<span class="go">M MSE</span>
<span class="go">M MSE</span>
<span class="go">M MSE</span>
</pre></div>
</div>
<p>In this case the selenomethionines (the first and also seventh and sixth from
last residues) have been shown as M (methionine) by the get_sequence method.</p>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.index_to_one">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">index_to_one</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.index_to_one" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Index to corresponding one letter amino acid name.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">index_to_one</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="go">&#39;A&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index_to_one</span><span class="p">(</span><span class="mi">19</span><span class="p">)</span>
<span class="go">&#39;Y&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.one_to_index">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">one_to_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.one_to_index" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>One letter code to index.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">one_to_index</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">one_to_index</span><span class="p">(</span><span class="s1">&#39;Y&#39;</span><span class="p">)</span>
<span class="go">19</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.index_to_three">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">index_to_three</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.index_to_three" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Index to corresponding three letter amino acid name.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">index_to_three</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="go">&#39;ALA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index_to_three</span><span class="p">(</span><span class="mi">19</span><span class="p">)</span>
<span class="go">&#39;TYR&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.three_to_index">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">three_to_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.three_to_index" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Three letter code to index.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">three_to_index</span><span class="p">(</span><span class="s1">&#39;ALA&#39;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">three_to_index</span><span class="p">(</span><span class="s1">&#39;TYR&#39;</span><span class="p">)</span>
<span class="go">19</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.is_aa">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">is_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">residue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">standard</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.is_aa" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return True if residue object/string is an amino acid.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>residue</strong> (<em>L{Residue}</em><em> or </em><em>string</em>) ‚Äì a L{Residue} object OR a three letter amino acid code</p></li>
<li><p><strong>standard</strong> (<em>boolean</em>) ‚Äì flag to check for the 20 AA (default false)</p></li>
</ul>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">is_aa</span><span class="p">(</span><span class="s1">&#39;ALA&#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Known three letter codes for modified amino acids are supported,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">is_aa</span><span class="p">(</span><span class="s1">&#39;FME&#39;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_aa</span><span class="p">(</span><span class="s1">&#39;FME&#39;</span><span class="p">,</span> <span class="n">standard</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.is_nucleic">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">is_nucleic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">residue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">standard</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.is_nucleic" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return True if residue object/string is a nucleic acid.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>residue</strong> (<em>L{Residue}</em><em> or </em><em>string</em>) ‚Äì a L{Residue} object OR a three letter code</p></li>
<li><p><strong>standard</strong> (<em>boolean</em>) ‚Äì flag to check for the 8 (DNA + RNA) canonical bases.
Default is False.</p></li>
</ul>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">is_nucleic</span><span class="p">(</span><span class="s1">&#39;DA &#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">is_nucleic</span><span class="p">(</span><span class="s1">&#39;A  &#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Known three letter codes for modified nucleotides are supported,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">is_nucleic</span><span class="p">(</span><span class="s1">&#39;A2L&#39;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">is_nucleic</span><span class="p">(</span><span class="s1">&#39;A2L&#39;</span><span class="p">,</span> <span class="n">standard</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.Polypeptide">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">Polypeptide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">iterable</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">/</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.Polypeptide" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>A polypeptide is simply a list of L{Residue} objects.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.Polypeptide.get_ca_list">
<span class="sig-name descname"><span class="pre">get_ca_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.Polypeptide.get_ca_list" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Get list of C-alpha atoms in the polypeptide.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the list of C-alpha atoms</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>[L{Atom}, L{Atom}, ‚Ä¶]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.Polypeptide.get_phi_psi_list">
<span class="sig-name descname"><span class="pre">get_phi_psi_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.Polypeptide.get_phi_psi_list" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the list of phi/psi dihedral angles.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.Polypeptide.get_tau_list">
<span class="sig-name descname"><span class="pre">get_tau_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.Polypeptide.get_tau_list" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>List of tau torsions angles for all 4 consecutive Calpha atoms.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.Polypeptide.get_theta_list">
<span class="sig-name descname"><span class="pre">get_theta_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.Polypeptide.get_theta_list" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>List of theta angles for all 3 consecutive Calpha atoms.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.Polypeptide.get_sequence">
<span class="sig-name descname"><span class="pre">get_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.Polypeptide.get_sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the AA sequence as a Seq object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>polypeptide sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>L{Seq}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.Polypeptide.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.Polypeptide.__repr__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return string representation of the polypeptide.</p>
<p>Return &lt;Polypeptide start=START end=END&gt;, where START
and END are sequence identifiers of the outer residues.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.CaPPBuilder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">CaPPBuilder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4.3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.CaPPBuilder" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.PDB.Polypeptide._PPBuilder</span></code></p>
<p>Use CA‚ÄìCA distance to find polypeptides.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.CaPPBuilder.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4.3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.CaPPBuilder.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.PPBuilder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.Polypeptide.</span></span><span class="sig-name descname"><span class="pre">PPBuilder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.8</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.PPBuilder" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.PDB.Polypeptide._PPBuilder</span></code></p>
<p>Use C‚ÄìN distance to find polypeptides.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.Polypeptide.PPBuilder.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.8</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.Polypeptide.PPBuilder.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.PDB.PSEA.html" class="btn btn-neutral float-left" title="Bio.PDB.PSEA module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PDB.Residue.html" class="btn btn-neutral float-right" title="Bio.PDB.Residue module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>