

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.AlignIO.PhylipIO module &mdash; Biopython 1.79.dev0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.AlignIO.StockholmIO module" href="Bio.AlignIO.StockholmIO.html" />
    <link rel="prev" title="Bio.AlignIO.NexusIO module" href="Bio.AlignIO.NexusIO.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.AlignIO.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html#module-Bio.AlignIO">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.AlignIO.html">Bio.AlignIO package</a> &raquo;</li>
        
      <li>Bio.AlignIO.PhylipIO module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/AlignIO/PhylipIO.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.AlignIO.StockholmIO.html" class="btn btn-neutral float-right" title="Bio.AlignIO.StockholmIO module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.AlignIO.NexusIO.html" class="btn btn-neutral float-left" title="Bio.AlignIO.NexusIO module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.AlignIO.PhylipIO">
<span id="bio-alignio-phylipio-module"></span><h1>Bio.AlignIO.PhylipIO module<a class="headerlink" href="#module-Bio.AlignIO.PhylipIO" title="Permalink to this headline">¶</a></h1>
<p>AlignIO support for “phylip” format from Joe Felsenstein’s PHYLIP tools.</p>
<p>You are expected to use this module via the Bio.AlignIO functions (or the
Bio.SeqIO functions if you want to work directly with the gapped sequences).</p>
<p>Support for “relaxed phylip” format is also provided. Relaxed phylip differs
from standard phylip format in the following ways:</p>
<blockquote>
<div><ul class="simple">
<li><p>No whitespace is allowed in the sequence ID.</p></li>
<li><p>No truncation is performed. Instead, sequence IDs are padded to the longest
ID length, rather than 10 characters. A space separates the sequence
identifier from the sequence.</p></li>
</ul>
</div></blockquote>
<p>Relaxed phylip is supported by RAxML and PHYML.</p>
<div class="section" id="note">
<h2>Note<a class="headerlink" href="#note" title="Permalink to this headline">¶</a></h2>
<p>In TREE_PUZZLE (Schmidt et al. 2003) and PHYML (Guindon and Gascuel 2003)
a dot/period (“.”) in a sequence is interpreted as meaning the same
character as in the first sequence.  The PHYLIP documentation from 3.3 to 3.69
<a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/sequence.html">http://evolution.genetics.washington.edu/phylip/doc/sequence.html</a> says:</p>
<p>“a period was also previously allowed but it is no longer allowed,
because it sometimes is used in different senses in other programs”</p>
<p>Biopython 1.58 or later treats dots/periods in the sequence as invalid, both
for reading and writing. Older versions did nothing special with a dot/period.</p>
<dl class="py class">
<dt id="Bio.AlignIO.PhylipIO.PhylipWriter">
<em class="property">class </em><code class="sig-prename descclassname">Bio.AlignIO.PhylipIO.</code><code class="sig-name descname">PhylipWriter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.PhylipWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.AlignIO.Interfaces.html#Bio.AlignIO.Interfaces.SequentialAlignmentWriter" title="Bio.AlignIO.Interfaces.SequentialAlignmentWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.AlignIO.Interfaces.SequentialAlignmentWriter</span></code></a></p>
<p>Phylip alignment writer.</p>
<dl class="py method">
<dt id="Bio.AlignIO.PhylipIO.PhylipWriter.write_alignment">
<code class="sig-name descname">write_alignment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">alignment</span></em>, <em class="sig-param"><span class="n">id_width</span><span class="o">=</span><span class="default_value">10</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.PhylipWriter.write_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Use this to write (another) single alignment to an open file.</p>
<p>This code will write interlaced alignments (when the sequences are
longer than 50 characters).</p>
<p>Note that record identifiers are strictly truncated to id_width,
defaulting to the value required to comply with the PHYLIP standard.</p>
<p>For more information on the file format, please see:
<a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/sequence.html">http://evolution.genetics.washington.edu/phylip/doc/sequence.html</a>
<a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/main.html#inputfiles">http://evolution.genetics.washington.edu/phylip/doc/main.html#inputfiles</a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.AlignIO.PhylipIO.PhylipIterator">
<em class="property">class </em><code class="sig-prename descclassname">Bio.AlignIO.PhylipIO.</code><code class="sig-name descname">PhylipIterator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em>, <em class="sig-param"><span class="n">seq_count</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.PhylipIterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.AlignIO.Interfaces.html#Bio.AlignIO.Interfaces.AlignmentIterator" title="Bio.AlignIO.Interfaces.AlignmentIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.AlignIO.Interfaces.AlignmentIterator</span></code></a></p>
<p>Reads a Phylip alignment file returning a MultipleSeqAlignment iterator.</p>
<p>Record identifiers are limited to at most 10 characters.</p>
<p>It only copes with interlaced phylip files!  Sequential files won’t work
where the sequences are split over multiple lines.</p>
<p>For more information on the file format, please see:
<a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/sequence.html">http://evolution.genetics.washington.edu/phylip/doc/sequence.html</a>
<a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/main.html#inputfiles">http://evolution.genetics.washington.edu/phylip/doc/main.html#inputfiles</a></p>
<dl class="py attribute">
<dt id="Bio.AlignIO.PhylipIO.PhylipIterator.id_width">
<code class="sig-name descname">id_width</code><em class="property"> = 10</em><a class="headerlink" href="#Bio.AlignIO.PhylipIO.PhylipIterator.id_width" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="Bio.AlignIO.PhylipIO.PhylipIterator.__next__">
<code class="sig-name descname">__next__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.PhylipIterator.__next__" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the next alignment from the handle.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.AlignIO.PhylipIO.RelaxedPhylipWriter">
<em class="property">class </em><code class="sig-prename descclassname">Bio.AlignIO.PhylipIO.</code><code class="sig-name descname">RelaxedPhylipWriter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.RelaxedPhylipWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.AlignIO.PhylipIO.PhylipWriter" title="Bio.AlignIO.PhylipIO.PhylipWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.AlignIO.PhylipIO.PhylipWriter</span></code></a></p>
<p>Relaxed Phylip format writer.</p>
<dl class="py method">
<dt id="Bio.AlignIO.PhylipIO.RelaxedPhylipWriter.write_alignment">
<code class="sig-name descname">write_alignment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">alignment</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.RelaxedPhylipWriter.write_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a relaxed phylip alignment.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.AlignIO.PhylipIO.RelaxedPhylipIterator">
<em class="property">class </em><code class="sig-prename descclassname">Bio.AlignIO.PhylipIO.</code><code class="sig-name descname">RelaxedPhylipIterator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em>, <em class="sig-param"><span class="n">seq_count</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.RelaxedPhylipIterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.AlignIO.PhylipIO.PhylipIterator" title="Bio.AlignIO.PhylipIO.PhylipIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.AlignIO.PhylipIO.PhylipIterator</span></code></a></p>
<p>Relaxed Phylip format Iterator.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.AlignIO.PhylipIO.SequentialPhylipWriter">
<em class="property">class </em><code class="sig-prename descclassname">Bio.AlignIO.PhylipIO.</code><code class="sig-name descname">SequentialPhylipWriter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.SequentialPhylipWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.AlignIO.Interfaces.html#Bio.AlignIO.Interfaces.SequentialAlignmentWriter" title="Bio.AlignIO.Interfaces.SequentialAlignmentWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.AlignIO.Interfaces.SequentialAlignmentWriter</span></code></a></p>
<p>Sequential Phylip format Writer.</p>
<dl class="py method">
<dt id="Bio.AlignIO.PhylipIO.SequentialPhylipWriter.write_alignment">
<code class="sig-name descname">write_alignment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">alignment</span></em>, <em class="sig-param"><span class="n">id_width</span><span class="o">=</span><span class="default_value">10</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.SequentialPhylipWriter.write_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a Phylip alignment to the handle.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.AlignIO.PhylipIO.SequentialPhylipIterator">
<em class="property">class </em><code class="sig-prename descclassname">Bio.AlignIO.PhylipIO.</code><code class="sig-name descname">SequentialPhylipIterator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">handle</span></em>, <em class="sig-param"><span class="n">seq_count</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.SequentialPhylipIterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.AlignIO.PhylipIO.PhylipIterator" title="Bio.AlignIO.PhylipIO.PhylipIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.AlignIO.PhylipIO.PhylipIterator</span></code></a></p>
<p>Sequential Phylip format Iterator.</p>
<p>The sequential format carries the same restrictions as the normal
interleaved one, with the difference being that the sequences are listed
sequentially, each sequence written in its entirety before the start of
the next. According to the PHYLIP documentation for input file
formatting, newlines and spaces may optionally be entered at any point
in the sequences.</p>
<dl class="py method">
<dt id="Bio.AlignIO.PhylipIO.SequentialPhylipIterator.__next__">
<code class="sig-name descname">__next__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.SequentialPhylipIterator.__next__" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse the next alignment from the handle.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.AlignIO.PhylipIO.sanitize_name">
<code class="sig-prename descclassname">Bio.AlignIO.PhylipIO.</code><code class="sig-name descname">sanitize_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">width</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.AlignIO.PhylipIO.sanitize_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Sanitise sequence identifier for output.</p>
<p>Removes the banned characters “[]()” and replaces the characters “:;”
with “|”. The name is truncated to “width” characters if specified.</p>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.AlignIO.StockholmIO.html" class="btn btn-neutral float-right" title="Bio.AlignIO.StockholmIO module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.AlignIO.NexusIO.html" class="btn btn-neutral float-left" title="Bio.AlignIO.NexusIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.79.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.77/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>