<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.SeqIO.SeqXmlIO module &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.SeqIO.SffIO module" href="Bio.SeqIO.SffIO.html" />
    <link rel="prev" title="Bio.SeqIO.QualityIO module" href="Bio.SeqIO.QualityIO.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
      <li class="breadcrumb-item active">Bio.SeqIO.SeqXmlIO module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SeqIO/SeqXmlIO.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.SeqIO.QualityIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.QualityIO module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqIO.SffIO.html" class="btn btn-neutral float-right" title="Bio.SeqIO.SffIO module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.SeqIO.SeqXmlIO">
<span id="bio-seqio-seqxmlio-module"></span><h1>Bio.SeqIO.SeqXmlIO module<a class="headerlink" href="#module-Bio.SeqIO.SeqXmlIO" title="Permalink to this headline"></a></h1>
<p>Bio.SeqIO support for the “seqxml” file format, SeqXML.</p>
<p>This module is for reading and writing SeqXML format files as
SeqRecord objects, and is expected to be used via the Bio.SeqIO API.</p>
<p>SeqXML is a lightweight XML format which is supposed be an alternative for
FASTA files. For more Information see <a class="reference external" href="http://www.seqXML.org">http://www.seqXML.org</a> and Schmitt et al
(2011), <a class="reference external" href="https://doi.org/10.1093/bib/bbr025">https://doi.org/10.1093/bib/bbr025</a></p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.SeqXmlIO.</span></span><span class="sig-name descname"><span class="pre">ContentHandler</span></span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">xml.sax.handler.ContentHandler</span></code></p>
<p>Handles XML events generated by the parser (PRIVATE).</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Create a handler to handle XML events.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startDocument">
<span class="sig-name descname"><span class="pre">startDocument</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startDocument" title="Permalink to this definition"></a></dt>
<dd><p>Set XML handlers when an XML declaration is found.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startSeqXMLElement">
<span class="sig-name descname"><span class="pre">startSeqXMLElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startSeqXMLElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle start of a seqXML element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.endSeqXMLElement">
<span class="sig-name descname"><span class="pre">endSeqXMLElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.endSeqXMLElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle end of the seqXML element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startEntryElement">
<span class="sig-name descname"><span class="pre">startEntryElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startEntryElement" title="Permalink to this definition"></a></dt>
<dd><p>Set new entry with id and the optional entry source (PRIVATE).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.endEntryElement">
<span class="sig-name descname"><span class="pre">endEntryElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.endEntryElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle end of an entry element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startEntryFieldElement">
<span class="sig-name descname"><span class="pre">startEntryFieldElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startEntryFieldElement" title="Permalink to this definition"></a></dt>
<dd><p>Receive a field of an entry element and forward it.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startSpeciesElement">
<span class="sig-name descname"><span class="pre">startSpeciesElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startSpeciesElement" title="Permalink to this definition"></a></dt>
<dd><p>Parse the species information.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.endSpeciesElement">
<span class="sig-name descname"><span class="pre">endSpeciesElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.endSpeciesElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle end of a species element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startDescriptionElement">
<span class="sig-name descname"><span class="pre">startDescriptionElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startDescriptionElement" title="Permalink to this definition"></a></dt>
<dd><p>Parse the description.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.endDescriptionElement">
<span class="sig-name descname"><span class="pre">endDescriptionElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.endDescriptionElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle the end of a description element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startSequenceElement">
<span class="sig-name descname"><span class="pre">startSequenceElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startSequenceElement" title="Permalink to this definition"></a></dt>
<dd><p>Parse DNA, RNA, or protein sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.endSequenceElement">
<span class="sig-name descname"><span class="pre">endSequenceElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.endSequenceElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle the end of a sequence element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startDBRefElement">
<span class="sig-name descname"><span class="pre">startDBRefElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startDBRefElement" title="Permalink to this definition"></a></dt>
<dd><p>Parse a database cross reference.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.endDBRefElement">
<span class="sig-name descname"><span class="pre">endDBRefElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.endDBRefElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle the end of a DBRef element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.startPropertyElement">
<span class="sig-name descname"><span class="pre">startPropertyElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.startPropertyElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle the start of a property element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.endPropertyElement">
<span class="sig-name descname"><span class="pre">endPropertyElement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.endPropertyElement" title="Permalink to this definition"></a></dt>
<dd><p>Handle the end of a property element.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.ContentHandler.characters">
<span class="sig-name descname"><span class="pre">characters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.ContentHandler.characters" title="Permalink to this definition"></a></dt>
<dd><p>Handle character data.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.SeqXmlIO.</span></span><span class="sig-name descname"><span class="pre">SeqXmlIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream_or_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceIterator" title="Bio.SeqIO.Interfaces.SequenceIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceIterator</span></code></a></p>
<p>Parser for seqXML files.</p>
<p>Parses seqXML files and creates SeqRecords.
Assumes valid seqXML please validate beforehand.
It is assumed that all information for one record can be found within a
record element or above. Two types of methods are called when the start
tag of an element is reached. To receive only the attributes of an
element before its end tag is reached implement _attr_TAGNAME.
To get an element and its children as a DOM tree implement _elem_TAGNAME.
Everything that is part of the DOM tree will not trigger any further
method calls.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlIterator.BLOCK">
<span class="sig-name descname"><span class="pre">BLOCK</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1024</span></em><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlIterator.BLOCK" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream_or_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">namespace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlIterator.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Create the object and initialize the XML parser.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlIterator.parse">
<span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlIterator.parse" title="Permalink to this definition"></a></dt>
<dd><p>Start parsing the file, and return a SeqRecord generator.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlIterator.iterate">
<span class="sig-name descname"><span class="pre">iterate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlIterator.iterate" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over the records in the XML file.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlIterator.__parameters__">
<span class="sig-name descname"><span class="pre">__parameters__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlIterator.__parameters__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.SeqXmlIO.</span></span><span class="sig-name descname"><span class="pre">SeqXmlWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbiTaxId</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlWriter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceWriter" title="Bio.SeqIO.Interfaces.SequenceWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceWriter</span></code></a></p>
<p>Writes SeqRecords into seqXML file.</p>
<p>SeqXML requires the SeqRecord annotations to specify the molecule_type;
the molecule type is required to contain the term “DNA”, “RNA”, or
“protein”.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlWriter.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncbiTaxId</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlWriter.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Create Object and start the xml generator.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>target - Output stream opened in binary mode, or a path to a file.</p></li>
<li><p>source - The source program/database of the file, for example
UniProt.</p></li>
<li><p>source_version - The version or release number of the source
program or database from which the data originated.</p></li>
<li><p>species - The scientific name of the species of origin of all
entries in the file.</p></li>
<li><p>ncbiTaxId - The NCBI taxonomy identifier of the species of origin.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlWriter.write_header">
<span class="sig-name descname"><span class="pre">write_header</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlWriter.write_header" title="Permalink to this definition"></a></dt>
<dd><p>Write root node with document metadata.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlWriter.write_record">
<span class="sig-name descname"><span class="pre">write_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlWriter.write_record" title="Permalink to this definition"></a></dt>
<dd><p>Write one record.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.SeqXmlIO.SeqXmlWriter.write_footer">
<span class="sig-name descname"><span class="pre">write_footer</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.SeqXmlIO.SeqXmlWriter.write_footer" title="Permalink to this definition"></a></dt>
<dd><p>Close the root node and finish the XML document.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.SeqIO.QualityIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.QualityIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqIO.SffIO.html" class="btn btn-neutral float-right" title="Bio.SeqIO.SffIO module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>