<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.PDB.PICIO module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.PDB.PSEA module" href="Bio.PDB.PSEA.html" />
    <link rel="prev" title="Bio.PDB.PDBParser module" href="Bio.PDB.PDBParser.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PDB.html">Bio.PDB package</a></li>
      <li class="breadcrumb-item active">Bio.PDB.PICIO module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/PDB/PICIO.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.PDB.PDBParser.html" class="btn btn-neutral float-left" title="Bio.PDB.PDBParser module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PDB.PSEA.html" class="btn btn-neutral float-right" title="Bio.PDB.PSEA module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.PDB.PICIO">
<span id="bio-pdb-picio-module"></span><h1>Bio.PDB.PICIO module<a class="headerlink" href="#module-Bio.PDB.PICIO" title="Permalink to this headline"></a></h1>
<p>PICIO: read and write Protein Internal Coordinate (.pic) data files.</p>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.PICIO.read_PIC">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.PICIO.</span></span><span class="sig-name descname"><span class="pre">read_PIC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">TextIO</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quick</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defaults</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="Bio.PDB.Structure.html#Bio.PDB.Structure.Structure" title="Bio.PDB.Structure.Structure"><span class="pre">Bio.PDB.Structure.Structure</span></a></span></span><a class="headerlink" href="#Bio.PDB.PICIO.read_PIC" title="Permalink to this definition"></a></dt>
<dd><p>Load Protein Internal Coordinate (.pic) data from file.</p>
<dl class="simple">
<dt>PIC file format:</dt><dd><ul class="simple">
<li><p>comment lines start with #</p></li>
<li><dl class="simple">
<dt>(optional) PDB HEADER record</dt><dd><ul>
<li><p>idcode and deposition date recommended but optional</p></li>
<li><p>deposition date in PDB format or as changed by Biopython</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>(optional) PDB TITLE record</p></li>
<li><dl class="simple">
<dt>repeat:</dt><dd><ul>
<li><p>Biopython Residue Full ID - sets residue IDs of returned structure</p></li>
<li><p>(optional) PDB N, CA, C ATOM records for chain start</p></li>
<li><p>(optional) PIC Hedra records for residue</p></li>
<li><p>(optional) PIC Dihedra records for residue</p></li>
<li><p>(optional) BFAC records listing AtomKeys and b-factors</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<p>An improvement would define relative positions for HOH (water) entries.</p>
<p>Defaults will be supplied for any value if defaults=True.  Default values
are supplied in ic_data.py, but structures degrade quickly with any
deviation from true coordinates.  Experiment with
<a class="reference internal" href="Bio.PDB.internal_coords.html#Bio.PDB.internal_coords.IC_Residue.pic_flags" title="Bio.PDB.internal_coords.IC_Residue.pic_flags"><code class="xref py py-data docutils literal notranslate"><span class="pre">Bio.PDB.internal_coords.IC_Residue.pic_flags</span></code></a> options to
<a class="reference internal" href="#Bio.PDB.PICIO.write_PIC" title="Bio.PDB.PICIO.write_PIC"><code class="xref py py-func docutils literal notranslate"><span class="pre">write_PIC()</span></code></a> to verify this.</p>
<p>N.B. dihedron (i-1)C-N-CA-CB is ignored in assembly if O exists.</p>
<p>C-beta is by default placed using O-C-CA-CB, but O is missing
in some PDB file residues, which means the sidechain cannot be
placed.  The alternate CB path (i-1)C-N-CA-CB is provided to
circumvent this, but if this is needed then it must be adjusted in
conjunction with PHI ((i-1)C-N-CA-C) as they overlap (see <a class="reference internal" href="Bio.PDB.internal_coords.html#Bio.PDB.internal_coords.IC_Residue.bond_set" title="Bio.PDB.internal_coords.IC_Residue.bond_set"><code class="xref py py-meth docutils literal notranslate"><span class="pre">bond_set()</span></code></a>
and <a class="reference internal" href="Bio.PDB.internal_coords.html#Bio.PDB.internal_coords.IC_Residue.bond_rotate" title="Bio.PDB.internal_coords.IC_Residue.bond_rotate"><code class="xref py py-meth docutils literal notranslate"><span class="pre">bond_rotate()</span></code></a> to handle this automatically).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>Bio.File</em>) – <a class="reference internal" href="Bio.File.html#Bio.File.as_handle" title="Bio.File.as_handle"><code class="xref py py-func docutils literal notranslate"><span class="pre">as_handle()</span></code></a> file name or handle</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – complain when lines not as expected</p></li>
<li><p><strong>quick</strong> (<em>bool</em>) – don’t check residues for all dihedra (no default values)</p></li>
<li><p><strong>defaults</strong> (<em>bool</em>) – create di/hedra as needed from reference database.
Amide proton created if ‘H’ is in IC_Residue.accept_atoms</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Biopython Structure object, Residues with .internal_coord
attributes but no coordinates except for chain start N, CA, C atoms if
supplied, <strong>OR</strong> None on parse fail (silent unless verbose=True)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.PICIO.read_PIC_seq">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.PICIO.</span></span><span class="sig-name descname"><span class="pre">read_PIC_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqRec</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="Bio.SeqRecord.html#Bio.SeqRecord.SeqRecord" title="Bio.SeqRecord.SeqRecord"><span class="pre">Bio.SeqRecord.SeqRecord</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdbid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="Bio.PDB.Structure.html#Bio.PDB.Structure.Structure" title="Bio.PDB.Structure.Structure"><span class="pre">Bio.PDB.Structure.Structure</span></a></span></span><a class="headerlink" href="#Bio.PDB.PICIO.read_PIC_seq" title="Permalink to this definition"></a></dt>
<dd><p>Read <a class="reference internal" href="Bio.SeqRecord.html#Bio.SeqRecord.SeqRecord" title="Bio.SeqRecord.SeqRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">SeqRecord</span></code></a> into Structure with default internal coords.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.PICIO.enumerate_atoms">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.PICIO.</span></span><span class="sig-name descname"><span class="pre">enumerate_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PICIO.enumerate_atoms" title="Permalink to this definition"></a></dt>
<dd><p>Ensure all atoms in entity have serial_number set.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.PICIO.pdb_date">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.PICIO.</span></span><span class="sig-name descname"><span class="pre">pdb_date</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datestr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#Bio.PDB.PICIO.pdb_date" title="Permalink to this definition"></a></dt>
<dd><p>Convert yyyy-mm-dd date to dd-month-yy.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.PDB.PICIO.write_PIC">
<span class="sig-prename descclassname"><span class="pre">Bio.PDB.PICIO.</span></span><span class="sig-name descname"><span class="pre">write_PIC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">entity</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdbid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chainid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">picFlags</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">IC_Residue.picFlagsDefault</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hCut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pCut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PICIO.write_PIC" title="Permalink to this definition"></a></dt>
<dd><p>Write Protein Internal Coordinates (PIC) to file.</p>
<p>See <a class="reference internal" href="#Bio.PDB.PICIO.read_PIC" title="Bio.PDB.PICIO.read_PIC"><code class="xref py py-func docutils literal notranslate"><span class="pre">read_PIC()</span></code></a> for file format.
See <code class="xref py py-data docutils literal notranslate"><span class="pre">IC_Residue.pic_accuracy</span></code> to vary numeric accuracy.
Recurses to lower entity levels (M, C, R).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>entity</strong> (<a class="reference internal" href="Bio.PDB.Entity.html#Bio.PDB.Entity.Entity" title="Bio.PDB.Entity.Entity"><em>Entity</em></a>) – Biopython PDB Entity object: S, M, C or R</p></li>
<li><p><strong>file</strong> (<em>Bio.File</em>) – <a class="reference internal" href="Bio.File.html#Bio.File.as_handle" title="Bio.File.as_handle"><code class="xref py py-func docutils literal notranslate"><span class="pre">as_handle()</span></code></a> file name or handle</p></li>
<li><p><strong>pdbid</strong> (<em>str</em>) – PDB idcode, read from entity if not supplied</p></li>
<li><p><strong>chainid</strong> (<em>char</em>) – PDB Chain ID, set from C level entity.id if needed</p></li>
<li><p><strong>picFlags</strong> (<em>int</em>) – <p>boolean flags controlling output, defined in
<a class="reference internal" href="Bio.PDB.internal_coords.html#Bio.PDB.internal_coords.IC_Residue.pic_flags" title="Bio.PDB.internal_coords.IC_Residue.pic_flags"><code class="xref py py-data docutils literal notranslate"><span class="pre">Bio.PDB.internal_coords.IC_Residue.pic_flags</span></code></a></p>
<ul>
<li><p>”psi”,</p></li>
<li><p>”omg”,</p></li>
<li><p>”phi”,</p></li>
<li><p>”tau”,  # tau hedron (N-Ca-C)</p></li>
<li><p>”chi1”,</p></li>
<li><p>”chi2”,</p></li>
<li><p>”chi3”,</p></li>
<li><p>”chi4”,</p></li>
<li><p>”chi5”,</p></li>
<li><p>”pomg”,  # proline omega</p></li>
<li><p>”chi”,   # chi1 through chi5</p></li>
<li><p>”classic_b”,  # psi | phi | tau | pomg</p></li>
<li><p>”classic”,    # classic_b | chi</p></li>
<li><p>”hedra”,      # all hedra including bond lengths</p></li>
<li><p>”primary”,    # all primary dihedra</p></li>
<li><p>”secondary”,  # all secondary dihedra (fixed angle from primary dihedra)</p></li>
<li><p>”all”,        # hedra | primary | secondary</p></li>
<li><p>”initAtoms”,  # XYZ coordinates of initial Tau (N-Ca-C)</p></li>
<li><p>”bFactors”</p></li>
</ul>
<p>default is everything:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">picFlagsDefault</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pic_flags</span><span class="o">.</span><span class="n">all</span> <span class="o">|</span> <span class="n">pic_flags</span><span class="o">.</span><span class="n">initAtoms</span> <span class="o">|</span> <span class="n">pic_flags</span><span class="o">.</span><span class="n">bFactors</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Usage in your code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># just primary dihedra and all hedra</span>
<span class="n">picFlags</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">IC_Residue</span><span class="o">.</span><span class="n">pic_flags</span><span class="o">.</span><span class="n">primary</span> <span class="o">|</span> <span class="n">IC_Residue</span><span class="o">.</span><span class="n">pic_flags</span><span class="o">.</span><span class="n">hedra</span>
<span class="p">)</span>

<span class="c1"># no B-factors:</span>
<span class="n">picFlags</span> <span class="o">=</span> <span class="n">IC_Residue</span><span class="o">.</span><span class="n">picFlagsDefault</span>
<span class="n">picFlags</span> <span class="o">&amp;=</span> <span class="o">~</span><span class="n">IC_Residue</span><span class="o">.</span><span class="n">pic_flags</span><span class="o">.</span><span class="n">bFactors</span>
</pre></div>
</div>
<p><a class="reference internal" href="#Bio.PDB.PICIO.read_PIC" title="Bio.PDB.PICIO.read_PIC"><code class="xref py py-func docutils literal notranslate"><span class="pre">read_PIC()</span></code></a> with <cite>(defaults=True)</cite> will use default values for
anything left out</p>
</p></li>
<li><p><strong>hCut</strong> (<em>float</em>) – default None
only write hedra with ref db angle std dev greater than this value</p></li>
<li><p><strong>pCut</strong> (<em>float</em>) – default None
only write primary dihedra with ref db angle std dev greater than this
value</p></li>
</ul>
</dd>
</dl>
<p><strong>Default values</strong>:</p>
<p>Data averaged from Sep 2019 Dunbrack cullpdb_pc20_res2.2_R1.0.</p>
<p>Please see</p>
<p><a class="reference external" href="https://dunbrack.fccc.edu/pisces/">PISCES: A Protein Sequence Culling Server</a></p>
<p>‘G. Wang and R. L. Dunbrack, Jr. PISCES: a protein sequence culling
server. Bioinformatics, 19:1589-1591, 2003.’</p>
<p>‘primary’ and ‘secondary’ dihedra are defined in ic_data.py.  Specifically,
secondary dihedra can be determined as a fixed rotation from another known
angle, for example N-Ca-C-O can be estimated from N-Ca-C-N (psi).</p>
<p>Standard deviations are listed in
&lt;biopython distribution&gt;/Bio/PDB/ic_data.py for default values, and can be
used to limit which hedra and dihedra are defaulted vs. output exact
measurements from structure (see hCut and pCut above).  Default values for
primary dihedra (psi, phi, omega, chi1, etc.) are chosen as the most common
integer value, not an average.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><a class="reference internal" href="Bio.PDB.PDBExceptions.html#Bio.PDB.PDBExceptions.PDBException" title="Bio.PDB.PDBExceptions.PDBException"><strong>PDBException</strong></a> – if entity level is A (Atom)</p></li>
<li><p><strong>Exception</strong> – if entity does not have .level attribute</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.PDB.PDBParser.html" class="btn btn-neutral float-left" title="Bio.PDB.PDBParser module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PDB.PSEA.html" class="btn btn-neutral float-right" title="Bio.PDB.PSEA module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>