<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.bgzf module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.kNN module" href="Bio.kNN.html" />
    <link rel="prev" title="Bio.SeqRecord module" href="Bio.SeqRecord.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.File.html">Bio.File module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.LogisticRegression.html">Bio.LogisticRegression module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.MarkovModel.html">Bio.MarkovModel module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.MaxEntropy.html">Bio.MaxEntropy module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NaiveBayes.html">Bio.NaiveBayes module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Seq.html">Bio.Seq module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqFeature.html">Bio.SeqFeature module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqRecord.html">Bio.SeqRecord module</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">Bio.bgzf module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.kNN.html">Bio.kNN module</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.pairwise2.html">Bio.pairwise2 module</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
      <li class="breadcrumb-item active">Bio.bgzf module</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/api/Bio.bgzf.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.SeqRecord.html" class="btn btn-neutral float-left" title="Bio.SeqRecord module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.kNN.html" class="btn btn-neutral float-right" title="Bio.kNN module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.bgzf">
<span id="bio-bgzf-module"></span><h1>Bio.bgzf module<a class="headerlink" href="#module-Bio.bgzf" title="Permalink to this headline"></a></h1>
<p>Read and write BGZF compressed files (the GZIP variant used in BAM).</p>
<p>The SAM/BAM file format (Sequence Alignment/Map) comes in a plain text
format (SAM), and a compressed binary format (BAM). The latter uses a
modified form of gzip compression called BGZF (Blocked GNU Zip Format),
which can be applied to any file format to provide compression with
efficient random access. BGZF is described together with the SAM/BAM
file format at <a class="reference external" href="https://samtools.sourceforge.net/SAM1.pdf">https://samtools.sourceforge.net/SAM1.pdf</a></p>
<p>Please read the text below about ‘virtual offsets’ before using BGZF
files for random access.</p>
<section id="aim-of-this-module">
<h2>Aim of this module<a class="headerlink" href="#aim-of-this-module" title="Permalink to this headline"></a></h2>
<p>The Python gzip library can be used to read BGZF files, since for
decompression they are just (specialised) gzip files. What this
module aims to facilitate is random access to BGZF files (using the
‘virtual offset’ idea), and writing BGZF files (which means using
suitably sized gzip blocks and writing the extra ‘BC’ field in the
gzip headers). As in the gzip library, the zlib library is used
internally.</p>
<p>In addition to being required for random access to and writing of
BAM files, the BGZF format can also be used on other sequential
data (in the sense of one record after another), such as most of
the sequence data formats supported in Bio.SeqIO (like FASTA,
FASTQ, GenBank, etc) or large MAF alignments.</p>
<p>The Bio.SeqIO indexing functions use this module to support BGZF files.</p>
</section>
<section id="technical-introduction-to-bgzf">
<h2>Technical Introduction to BGZF<a class="headerlink" href="#technical-introduction-to-bgzf" title="Permalink to this headline"></a></h2>
<p>The gzip file format allows multiple compressed blocks, each of which
could be a stand alone gzip file. As an interesting bonus, this means
you can use Unix <code class="docutils literal notranslate"><span class="pre">cat</span></code> to combine two or more gzip files into one by
concatenating them. Also, each block can have one of several compression
levels (including uncompressed, which actually takes up a little bit
more space due to the gzip header).</p>
<p>What the BAM designers realised was that while random access to data
stored in traditional gzip files was slow, breaking the file into
gzip blocks would allow fast random access to each block. To access
a particular piece of the decompressed data, you just need to know
which block it starts in (the offset of the gzip block start), and
how far into the (decompressed) contents of the block you need to
read.</p>
<p>One problem with this is finding the gzip block sizes efficiently.
You can do it with a standard gzip file, but it requires every block
to be decompressed – and that would be rather slow. Additionally
typical gzip files may use very large blocks.</p>
<p>All that differs in BGZF is that compressed size of each gzip block
is limited to 2^16 bytes, and an extra ‘BC’ field in the gzip header
records this size. Traditional decompression tools can ignore this,
and unzip the file just like any other gzip file.</p>
<p>The point of this is you can look at the first BGZF block, find out
how big it is from this ‘BC’ header, and thus seek immediately to
the second block, and so on.</p>
<p>The BAM indexing scheme records read positions using a 64 bit
‘virtual offset’, comprising <code class="docutils literal notranslate"><span class="pre">coffset</span> <span class="pre">&lt;&lt;</span> <span class="pre">16</span> <span class="pre">|</span> <span class="pre">uoffset</span></code>, where <code class="docutils literal notranslate"><span class="pre">coffset</span></code>
is the file offset of the BGZF block containing the start of the read
(unsigned integer using up to 64-16 = 48 bits), and <code class="docutils literal notranslate"><span class="pre">uoffset</span></code> is the
offset within the (decompressed) block (unsigned 16 bit integer).</p>
<p>This limits you to BAM files where the last block starts by 2^48
bytes, or 256 petabytes, and the decompressed size of each block
is at most 2^16 bytes, or 64kb. Note that this matches the BGZF
‘BC’ field size which limits the compressed size of each block to
2^16 bytes, allowing for BAM files to use BGZF with no gzip
compression (useful for intermediate files in memory to reduce
CPU load).</p>
</section>
<section id="warning-about-namespaces">
<h2>Warning about namespaces<a class="headerlink" href="#warning-about-namespaces" title="Permalink to this headline"></a></h2>
<p>It is considered a bad idea to use “from XXX import <code class="docutils literal notranslate"><span class="pre">*</span></code>” in Python, because
it pollutes the namespace. This is a real issue with Bio.bgzf (and the
standard Python library gzip) because they contain a function called open
i.e. Suppose you do this:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.bgzf</span> <span class="kn">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">open</span><span class="o">.</span><span class="vm">__module__</span><span class="p">)</span>
<span class="go">Bio.bgzf</span>
</pre></div>
</div>
<p>Or,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">gzip</span> <span class="kn">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">open</span><span class="o">.</span><span class="vm">__module__</span><span class="p">)</span>
<span class="go">gzip</span>
</pre></div>
</div>
<p>Notice that the open function has been replaced. You can “fix” this if you
need to by importing the built-in open function:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">builtins</span> <span class="kn">import</span> <span class="nb">open</span>
</pre></div>
</div>
<p>However, what we recommend instead is to use the explicit namespace, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">bgzf</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bgzf</span><span class="o">.</span><span class="n">open</span><span class="o">.</span><span class="vm">__module__</span><span class="p">)</span>
<span class="go">Bio.bgzf</span>
</pre></div>
</div>
</section>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline"></a></h2>
<p>This is an ordinary GenBank file compressed using BGZF, so it can
be decompressed using gzip,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gzip</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;GenBank/NC_000932.gb.bgz&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">0</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">line</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">80</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">line</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">143</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">70000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">70143</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>We can also access the file using the BGZF reader - but pay
attention to the file offsets which will be explained below:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">BgzfReader</span><span class="p">(</span><span class="s2">&quot;GenBank/NC_000932.gb.bgz&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">0</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">())</span>
<span class="go">LOCUS       NC_000932             154478 bp    DNA     circular PLN 15-APR-2009</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">80</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">())</span>
<span class="go">DEFINITION  Arabidopsis thaliana chloroplast, complete genome.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">143</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">70000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">987828735</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">())</span>
<span class="go">f=&quot;GeneID:844718&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">())</span>
<span class="go">     CDS             complement(join(84337..84771,85454..85843))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">offset</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">55074</span><span class="p">,</span> <span class="mi">126</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">())</span>
<span class="go">    68521 tatgtcattc gaaattgtat aaagacaact cctatttaat agagctattt gtgcaagtat</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Notice the handle’s offset looks different as a BGZF file. This
brings us to the key point about BGZF, which is the block structure:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;GenBank/NC_000932.gb.bgz&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">BgzfBlocks</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Raw start </span><span class="si">%i</span><span class="s2">, raw length </span><span class="si">%i</span><span class="s2">; data start </span><span class="si">%i</span><span class="s2">, data length </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">values</span><span class="p">)</span>
<span class="go">Raw start 0, raw length 15073; data start 0, data length 65536</span>
<span class="go">Raw start 15073, raw length 17857; data start 65536, data length 65536</span>
<span class="go">Raw start 32930, raw length 22144; data start 131072, data length 65536</span>
<span class="go">Raw start 55074, raw length 22230; data start 196608, data length 65536</span>
<span class="go">Raw start 77304, raw length 14939; data start 262144, data length 43478</span>
<span class="go">Raw start 92243, raw length 28; data start 305622, data length 0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>In this example the first three blocks are ‘full’ and hold 65536 bytes
of uncompressed data. The fourth block isn’t full and holds 43478 bytes.
Finally there is a special empty fifth block which takes 28 bytes on
disk and serves as an ‘end of file’ (EOF) marker. If this is missing,
it is possible your BGZF file is incomplete.</p>
<p>By reading ahead 70,000 bytes we moved into the second BGZF block,
and at that point the BGZF virtual offsets start to look different
to a simple offset into the decompressed data as exposed by the gzip
library.</p>
<p>As an example, consider seeking to the decompressed position 196734.
Since 196734 = 65536 + 65536 + 65536 + 126 = 65536*3 + 126, this
is equivalent to jumping the first three blocks (which in this
specific example are all size 65536 after decompression - which
does not always hold) and starting at byte 126 of the fourth block
(after decompression). For BGZF, we need to know the fourth block’s
offset of 55074 and the offset within the block of 126 to get the
BGZF virtual offset.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="mi">55074</span> <span class="o">&lt;&lt;</span> <span class="mi">16</span> <span class="o">|</span> <span class="mi">126</span><span class="p">)</span>
<span class="go">3609329790</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bgzf</span><span class="o">.</span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">55074</span><span class="p">,</span> <span class="mi">126</span><span class="p">))</span>
<span class="go">3609329790</span>
</pre></div>
</div>
<p>Thus for this BGZF file, decompressed position 196734 corresponds
to the virtual offset 3609329790. However, another BGZF file with
different contents would have compressed more or less efficiently,
so the compressed blocks would be different sizes. What this means
is the mapping between the uncompressed offset and the compressed
virtual offset depends on the BGZF file you are using.</p>
<p>If you are accessing a BGZF file via this module, just use the
handle.tell() method to note the virtual offset of a position you
may later want to return to using handle.seek().</p>
<p>The catch with BGZF virtual offsets is while they can be compared
(which offset comes first in the file), you cannot safely subtract
them to get the size of the data between them, nor add/subtract
a relative offset.</p>
<p>Of course you can parse this file with Bio.SeqIO using BgzfReader,
although there isn’t any benefit over using gzip.open(…), unless
you want to index BGZF compressed sequence files:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">BgzfReader</span><span class="p">(</span><span class="s2">&quot;GenBank/NC_000932.gb.bgz&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="go">NC_000932.1</span>
</pre></div>
</div>
</section>
<section id="text-mode">
<h2>Text Mode<a class="headerlink" href="#text-mode" title="Permalink to this headline"></a></h2>
<p>Like the standard library gzip.open(…), the BGZF code defaults to opening
files in binary mode.</p>
<p>You can request the file be opened in text mode, but beware that this is hard
coded to the simple “latin1” (aka “iso-8859-1”) encoding (which includes all
the ASCII characters), which works well with most Western European languages.
However, it is not fully compatible with the more widely used UTF-8 encoding.</p>
<p>In variable width encodings like UTF-8, some single characters in the unicode
text output are represented by multiple bytes in the raw binary form. This is
problematic with BGZF, as we cannot always decode each block in isolation - a
single unicode character could be split over two blocks. This can even happen
with fixed width unicode encodings, as the BGZF block size is not fixed.</p>
<p>Therefore, this module is currently restricted to only support single byte
unicode encodings, such as ASCII, “latin1” (which is a superset of ASCII), or
potentially other character maps (not implemented).</p>
<p>Furthermore, unlike the default text mode on Python 3, we do not attempt to
implement universal new line mode. This transforms the various operating system
new line conventions like Windows (CR LF or “rn”), Unix (just LF, “n”), or
old Macs (just CR, “r”), into just LF (”n”). Here we have the same problem -
is “r” at the end of a block an incomplete Windows style new line?</p>
<p>Instead, you will get the CR (”r”) and LF (”n”) characters as is.</p>
<p>If your data is in UTF-8 or any other incompatible encoding, you must use
binary mode, and decode the appropriate fragments yourself.</p>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.bgzf.open">
<span class="sig-prename descclassname"><span class="pre">Bio.bgzf.</span></span><span class="sig-name descname"><span class="pre">open</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rb'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.open" title="Permalink to this definition"></a></dt>
<dd><p>Open a BGZF file for reading, writing or appending.</p>
<p>If text mode is requested, in order to avoid multi-byte characters, this is
hard coded to use the “latin1” encoding, and “r” and “n” are passed as is
(without implementing universal new line mode).</p>
<p>If your data is in UTF-8 or any other incompatible encoding, you must use
binary mode, and decode the appropriate fragments yourself.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.bgzf.make_virtual_offset">
<span class="sig-prename descclassname"><span class="pre">Bio.bgzf.</span></span><span class="sig-name descname"><span class="pre">make_virtual_offset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">block_start_offset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within_block_offset</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.make_virtual_offset" title="Permalink to this definition"></a></dt>
<dd><p>Compute a BGZF virtual offset from block start and within block offsets.</p>
<p>The BAM indexing scheme records read positions using a 64 bit
‘virtual offset’, comprising in C terms:</p>
<p>block_start_offset &lt;&lt; 16 | within_block_offset</p>
<p>Here block_start_offset is the file offset of the BGZF block
start (unsigned integer using up to 64-16 = 48 bits), and
within_block_offset within the (decompressed) block (unsigned
16 bit integer).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="o">**</span><span class="mi">16</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">65535</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="o">**</span><span class="mi">16</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Require 0 &lt;= within_block_offset &lt; 2**16, got 65536</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="mi">65536</span> <span class="o">==</span> <span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">65537</span> <span class="o">==</span> <span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">131071</span> <span class="o">==</span> <span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="o">**</span><span class="mi">16</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="mi">6553600000</span> <span class="o">==</span> <span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">100000</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">6553600001</span> <span class="o">==</span> <span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">100000</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">6553600010</span> <span class="o">==</span> <span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">100000</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">2</span><span class="o">**</span><span class="mi">48</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Require 0 &lt;= block_start_offset &lt; 2**48, got 281474976710656</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.bgzf.split_virtual_offset">
<span class="sig-prename descclassname"><span class="pre">Bio.bgzf.</span></span><span class="sig-name descname"><span class="pre">split_virtual_offset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">virtual_offset</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.split_virtual_offset" title="Permalink to this definition"></a></dt>
<dd><p>Divides a 64-bit BGZF virtual offset into block start &amp; within block offsets.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="mi">100000</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="n">split_virtual_offset</span><span class="p">(</span><span class="mi">6553600000</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="mi">100000</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span> <span class="o">==</span> <span class="n">split_virtual_offset</span><span class="p">(</span><span class="mi">6553600010</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfBlocks">
<span class="sig-prename descclassname"><span class="pre">Bio.bgzf.</span></span><span class="sig-name descname"><span class="pre">BgzfBlocks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfBlocks" title="Permalink to this definition"></a></dt>
<dd><p>Low level debugging function to inspect BGZF blocks.</p>
<p>Expects a BGZF compressed file opened in binary read mode using
the builtin open function. Do not use a handle from this bgzf
module or the gzip module’s open function which will decompress
the file.</p>
<p>Returns the block start offset (see virtual offsets), the block
length (add these for the start of the next block), and the
decompressed length of the blocks contents (limited to 65536 in
BGZF), as an iterator - one tuple per BGZF block.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">builtins</span> <span class="kn">import</span> <span class="nb">open</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;SamBam/ex1.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">BgzfBlocks</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Raw start </span><span class="si">%i</span><span class="s2">, raw length </span><span class="si">%i</span><span class="s2">; data start </span><span class="si">%i</span><span class="s2">, data length </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">values</span><span class="p">)</span>
<span class="go">Raw start 0, raw length 18239; data start 0, data length 65536</span>
<span class="go">Raw start 18239, raw length 18223; data start 65536, data length 65536</span>
<span class="go">Raw start 36462, raw length 18017; data start 131072, data length 65536</span>
<span class="go">Raw start 54479, raw length 17342; data start 196608, data length 65536</span>
<span class="go">Raw start 71821, raw length 17715; data start 262144, data length 65536</span>
<span class="go">Raw start 89536, raw length 17728; data start 327680, data length 65536</span>
<span class="go">Raw start 107264, raw length 17292; data start 393216, data length 63398</span>
<span class="go">Raw start 124556, raw length 28; data start 456614, data length 0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Indirectly we can tell this file came from an old version of
samtools because all the blocks (except the final one and the
dummy empty EOF marker block) are 65536 bytes.  Later versions
avoid splitting a read between two blocks, and give the header
its own block (useful to speed up replacing the header). You
can see this in ex1_refresh.bam created using samtools 0.1.18:</p>
<p>samtools view -b ex1.bam &gt; ex1_refresh.bam</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;SamBam/ex1_refresh.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">BgzfBlocks</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Raw start </span><span class="si">%i</span><span class="s2">, raw length </span><span class="si">%i</span><span class="s2">; data start </span><span class="si">%i</span><span class="s2">, data length </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">values</span><span class="p">)</span>
<span class="go">Raw start 0, raw length 53; data start 0, data length 38</span>
<span class="go">Raw start 53, raw length 18195; data start 38, data length 65434</span>
<span class="go">Raw start 18248, raw length 18190; data start 65472, data length 65409</span>
<span class="go">Raw start 36438, raw length 18004; data start 130881, data length 65483</span>
<span class="go">Raw start 54442, raw length 17353; data start 196364, data length 65519</span>
<span class="go">Raw start 71795, raw length 17708; data start 261883, data length 65411</span>
<span class="go">Raw start 89503, raw length 17709; data start 327294, data length 65466</span>
<span class="go">Raw start 107212, raw length 17390; data start 392760, data length 63854</span>
<span class="go">Raw start 124602, raw length 28; data start 456614, data length 0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>The above example has no embedded SAM header (thus the first block
is very small at just 38 bytes of decompressed data), while the next
example does (a larger block of 103 bytes). Notice that the rest of
the blocks show the same sizes (they contain the same read data):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;SamBam/ex1_header.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">BgzfBlocks</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Raw start </span><span class="si">%i</span><span class="s2">, raw length </span><span class="si">%i</span><span class="s2">; data start </span><span class="si">%i</span><span class="s2">, data length </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">values</span><span class="p">)</span>
<span class="go">Raw start 0, raw length 104; data start 0, data length 103</span>
<span class="go">Raw start 104, raw length 18195; data start 103, data length 65434</span>
<span class="go">Raw start 18299, raw length 18190; data start 65537, data length 65409</span>
<span class="go">Raw start 36489, raw length 18004; data start 130946, data length 65483</span>
<span class="go">Raw start 54493, raw length 17353; data start 196429, data length 65519</span>
<span class="go">Raw start 71846, raw length 17708; data start 261948, data length 65411</span>
<span class="go">Raw start 89554, raw length 17709; data start 327359, data length 65466</span>
<span class="go">Raw start 107263, raw length 17390; data start 392825, data length 63854</span>
<span class="go">Raw start 124653, raw length 28; data start 456679, data length 0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.bgzf.</span></span><span class="sig-name descname"><span class="pre">BgzfReader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fileobj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cache</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>BGZF reader, acts like a read only handle but seek/tell differ.</p>
<p>Let’s use the BgzfBlocks function to have a peek at the BGZF blocks
in an example BAM file,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">builtins</span> <span class="kn">import</span> <span class="nb">open</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;SamBam/ex1.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">values</span> <span class="ow">in</span> <span class="n">BgzfBlocks</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Raw start </span><span class="si">%i</span><span class="s2">, raw length </span><span class="si">%i</span><span class="s2">; data start </span><span class="si">%i</span><span class="s2">, data length </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">values</span><span class="p">)</span>
<span class="go">Raw start 0, raw length 18239; data start 0, data length 65536</span>
<span class="go">Raw start 18239, raw length 18223; data start 65536, data length 65536</span>
<span class="go">Raw start 36462, raw length 18017; data start 131072, data length 65536</span>
<span class="go">Raw start 54479, raw length 17342; data start 196608, data length 65536</span>
<span class="go">Raw start 71821, raw length 17715; data start 262144, data length 65536</span>
<span class="go">Raw start 89536, raw length 17728; data start 327680, data length 65536</span>
<span class="go">Raw start 107264, raw length 17292; data start 393216, data length 63398</span>
<span class="go">Raw start 124556, raw length 28; data start 456614, data length 0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Now let’s see how to use this block information to jump to
specific parts of the decompressed BAM file:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">BgzfReader</span><span class="p">(</span><span class="s2">&quot;SamBam/ex1.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">0</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">magic</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">4</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
</pre></div>
</div>
<p>So far nothing so strange, we got the magic marker used at the
start of a decompressed BAM file, and the handle position makes
sense. Now however, let’s jump to the end of this block and 4
bytes into the next block by reading 65536 bytes,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="mi">65536</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="go">65536</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="mi">1195311108</span> <span class="o">==</span> <span class="n">handle</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
</pre></div>
</div>
<p>Expecting 4 + 65536 = 65540 were you? Well this is a BGZF 64-bit
virtual offset, which means:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">split_virtual_offset</span><span class="p">(</span><span class="mi">1195311108</span><span class="p">)</span>
<span class="go">(18239, 4)</span>
</pre></div>
</div>
<p>You should spot 18239 as the start of the second BGZF block, while
the 4 is the offset into this block. See also make_virtual_offset,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">18239</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="go">1195311108</span>
</pre></div>
</div>
<p>Let’s jump back to almost the start of the file,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Note that you can use the max_cache argument to limit the number of
BGZF blocks cached in memory. The default is 100, and since each
block can be up to 64kb, the default cache could take up to 6MB of
RAM. The cache is not important for reading through the file in one
pass, but is important for improving performance of random access.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fileobj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cache</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class for reading a BGZF file.</p>
<p>You would typically use the top level <code class="docutils literal notranslate"><span class="pre">bgzf.open(...)</span></code> function
which will call this class internally. Direct use is discouraged.</p>
<p>Either the <code class="docutils literal notranslate"><span class="pre">filename</span></code> (string) or <code class="docutils literal notranslate"><span class="pre">fileobj</span></code> (input file object in
binary mode) arguments must be supplied, but not both.</p>
<p>Argument <code class="docutils literal notranslate"><span class="pre">mode</span></code> controls if the data will be returned as strings in
text mode (“rt”, “tr”, or default “r”), or bytes binary mode (“rb”
or “br”). The argument name matches the built-in <code class="docutils literal notranslate"><span class="pre">open(...)</span></code> and
standard library <code class="docutils literal notranslate"><span class="pre">gzip.open(...)</span></code> function.</p>
<p>If text mode is requested, in order to avoid multi-byte characters,
this is hard coded to use the “latin1” encoding, and “r” and “n”
are passed as is (without implementing universal new line mode). There
is no <code class="docutils literal notranslate"><span class="pre">encoding</span></code> argument.</p>
<p>If your data is in UTF-8 or any other incompatible encoding, you must
use binary mode, and decode the appropriate fragments yourself.</p>
<p>Argument <code class="docutils literal notranslate"><span class="pre">max_cache</span></code> controls the maximum number of BGZF blocks to
cache in memory. Each can be up to 64kb thus the default of 100 blocks
could take up to 6MB of RAM. This is important for efficient random
access, a small value is fine for reading the file in one pass.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.tell">
<span class="sig-name descname"><span class="pre">tell</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.tell" title="Permalink to this definition"></a></dt>
<dd><p>Return a 64-bit unsigned BGZF virtual offset.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.seek">
<span class="sig-name descname"><span class="pre">seek</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">virtual_offset</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.seek" title="Permalink to this definition"></a></dt>
<dd><p>Seek to a 64-bit unsigned BGZF virtual offset.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.read">
<span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.read" title="Permalink to this definition"></a></dt>
<dd><p>Read method for the BGZF module.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.readline">
<span class="sig-name descname"><span class="pre">readline</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.readline" title="Permalink to this definition"></a></dt>
<dd><p>Read a single line for the BGZF file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.__next__">
<span class="sig-name descname"><span class="pre">__next__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.__next__" title="Permalink to this definition"></a></dt>
<dd><p>Return the next line.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.__iter__">
<span class="sig-name descname"><span class="pre">__iter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.__iter__" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over the lines in the BGZF file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.close" title="Permalink to this definition"></a></dt>
<dd><p>Close BGZF file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.seekable">
<span class="sig-name descname"><span class="pre">seekable</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.seekable" title="Permalink to this definition"></a></dt>
<dd><p>Return True indicating the BGZF supports random access.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.isatty">
<span class="sig-name descname"><span class="pre">isatty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.isatty" title="Permalink to this definition"></a></dt>
<dd><p>Return True if connected to a TTY device.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.fileno">
<span class="sig-name descname"><span class="pre">fileno</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.fileno" title="Permalink to this definition"></a></dt>
<dd><p>Return integer file descriptor.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.__enter__">
<span class="sig-name descname"><span class="pre">__enter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.__enter__" title="Permalink to this definition"></a></dt>
<dd><p>Open a file operable with WITH statement.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfReader.__exit__">
<span class="sig-name descname"><span class="pre">__exit__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">traceback</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfReader.__exit__" title="Permalink to this definition"></a></dt>
<dd><p>Close a file with WITH statement.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.bgzf.</span></span><span class="sig-name descname"><span class="pre">BgzfWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'w'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fileobj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compresslevel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Define a BGZFWriter object.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'w'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fileobj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compresslevel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initilize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.write" title="Permalink to this definition"></a></dt>
<dd><p>Write method for the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.flush" title="Permalink to this definition"></a></dt>
<dd><p>Flush data explicitally.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.close" title="Permalink to this definition"></a></dt>
<dd><p>Flush data, write 28 bytes BGZF EOF marker, and close BGZF file.</p>
<p>samtools will look for a magic EOF marker, just a 28 byte empty BGZF
block, and if it is missing warns the BAM file may be truncated. In
addition to samtools writing this block, so too does bgzip - so this
implementation does too.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.tell">
<span class="sig-name descname"><span class="pre">tell</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.tell" title="Permalink to this definition"></a></dt>
<dd><p>Return a BGZF 64-bit virtual offset.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.seekable">
<span class="sig-name descname"><span class="pre">seekable</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.seekable" title="Permalink to this definition"></a></dt>
<dd><p>Return True indicating the BGZF supports random access.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.isatty">
<span class="sig-name descname"><span class="pre">isatty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.isatty" title="Permalink to this definition"></a></dt>
<dd><p>Return True if connected to a TTY device.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.fileno">
<span class="sig-name descname"><span class="pre">fileno</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.fileno" title="Permalink to this definition"></a></dt>
<dd><p>Return integer file descriptor.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.__enter__">
<span class="sig-name descname"><span class="pre">__enter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.__enter__" title="Permalink to this definition"></a></dt>
<dd><p>Open a file operable with WITH statement.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.bgzf.BgzfWriter.__exit__">
<span class="sig-name descname"><span class="pre">__exit__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">traceback</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.bgzf.BgzfWriter.__exit__" title="Permalink to this definition"></a></dt>
<dd><p>Close a file with WITH statement.</p>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.SeqRecord.html" class="btn btn-neutral float-left" title="Bio.SeqRecord module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.kNN.html" class="btn btn-neutral float-right" title="Bio.kNN module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>