<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.SearchIO.HmmerIO package &mdash; Biopython 1.80.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SearchIO.HmmerIO.hmmer2_text module" href="Bio.SearchIO.HmmerIO.hmmer2_text.html" />
    <link rel="prev" title="Bio.SearchIO.HHsuiteIO.hhsuite2_text module" href="Bio.SearchIO.HHsuiteIO.hhsuite2_text.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
            <a href="index.html">
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SearchIO.html#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html#module-Bio.SearchIO">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
          <li><a href="Bio.SearchIO.html">Bio.SearchIO package</a> &raquo;</li>
      <li>Bio.SearchIO.HmmerIO package</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SearchIO/HmmerIO/__init__.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.SearchIO.HHsuiteIO.hhsuite2_text.html" class="btn btn-neutral float-left" title="Bio.SearchIO.HHsuiteIO.hhsuite2_text module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SearchIO.HmmerIO.hmmer2_text.html" class="btn btn-neutral float-right" title="Bio.SearchIO.HmmerIO.hmmer2_text module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bio-searchio-hmmerio-package">
<h1>Bio.SearchIO.HmmerIO package<a class="headerlink" href="#bio-searchio-hmmerio-package" title="Permalink to this headline"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.SearchIO.HmmerIO.hmmer2_text.html">Bio.SearchIO.HmmerIO.hmmer2_text module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SearchIO.HmmerIO.hmmer3_domtab.html">Bio.SearchIO.HmmerIO.hmmer3_domtab module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SearchIO.HmmerIO.hmmer3_tab.html">Bio.SearchIO.HmmerIO.hmmer3_tab module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SearchIO.HmmerIO.hmmer3_text.html">Bio.SearchIO.HmmerIO.hmmer3_text module</a></li>
</ul>
</div>
</section>
<section id="module-Bio.SearchIO.HmmerIO">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.SearchIO.HmmerIO" title="Permalink to this headline"></a></h2>
<p>Bio.SearchIO support for HMMER output formats.</p>
<p>This module adds support for parsing HMMER outputs. HMMER is a
suite of programs implementing the profile hidden Markov models to find
similarity across protein sequences.</p>
<p>Bio.SearchIO.HmmerIO was tested on the following HMMER versions and flavors:</p>
<blockquote>
<div><ul class="simple">
<li><p>HMMER3 flavors: hmmscan, hmmsearch, phmmer</p></li>
<li><p>HMMER2 flavors: hmmpfam, hmmsearch</p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>More information on HMMER are available through these links:</dt><dd><ul class="simple">
<li><p>Web page: <a class="reference external" href="http://hmmer.janelia.org/">http://hmmer.janelia.org/</a></p></li>
<li><p>User guide: <a class="reference external" href="ftp://selab.janelia.org/pub/software/hmmer3/3.0/Userguide.pdf">ftp://selab.janelia.org/pub/software/hmmer3/3.0/Userguide.pdf</a></p></li>
</ul>
</dd>
</dl>
<section id="supported-formats">
<h3>Supported formats<a class="headerlink" href="#supported-formats" title="Permalink to this headline"></a></h3>
<p>Bio.SearchIO.HmmerIO supports the following HMMER output formats:</p>
<blockquote>
<div><ul class="simple">
<li><p>Plain text, v3.0   - ‘hmmer3-text’    - parsing, indexing</p></li>
<li><p>Table, v3.0        - ‘hmmer3-tab’     - parsing, indexing, writing</p></li>
<li><p>Domain table, v3.0 - ‘hmmer3-domtab’* - parsing, indexing, writing</p></li>
<li><p>Plain text, v2.x   - ‘hmmer2-text’    - parsing, indexing</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>For the domain table output, due to the way HMMER outputs the sequence
coordinates, you have to specify what HMMER flavor produced the output as the
file format. So instead of using ‘hmmer3-domtab’, you have to use either
‘hmmscan3-domtab’, ‘hmmsearch3-domtab’, or ‘phmmer3-domtab’ as the file format
name.</p></li>
</ul>
<p>Note that for all output formats, HMMER uses its own convention of input and
output coordinates. It does not use the term ‘hit’ or ‘query’, instead it
uses ‘hmm’ or ‘ali’. For example, ‘hmmfrom’ is the start coordinate of the HMM
sequence while ‘alifrom’ is the start coordinate of the protein sequence.</p>
<p>HmmerIO is aware of this different naming scheme and will adjust them
accordingly to fit SearchIO’s object model. If HmmerIO sees that the output file
to parse was written by hmmsearch or phmmer, all ‘hmm’ coordinates will be the
hit coordinates and ‘ali’ coordinates will be the query coordinates. Conversely,
if the HMMER flavor is hmmscan, ‘hmm’ will be query and ‘ali’ will be hit.</p>
<p>This is why the ‘hmmer3-domtab’ format has to be specified with the source HMMER
flavor. The parsers need to know which is the hit and which is the query.
‘hmmer3-text’ has its source program information present in the file, while
‘hmmer3-tab’ does not output any coordinates. That’s why both of these formats
do not need direct flavor specification like ‘hmmer3-domtab’.</p>
<p>Also note that when using the domain table format writers, it will use HMMER’s
naming convention (‘hmm’ and ‘ali’) so the files you write will be similar to
files written by a real HMMER program.</p>
</section>
<section id="hmmer2-text-and-hmmer3-text">
<h3>hmmer2-text and hmmer3-text<a class="headerlink" href="#hmmer2-text-and-hmmer3-text" title="Permalink to this headline"></a></h3>
<p>The parser for HMMER 3.0 plain text output can parse output files with alignment
blocks (default) or without (with the ‘–noali’ flag). If the alignment blocks
are present, you can also parse files with variable alignment width (using the
‘–notextw’ or ‘–textw’ flag).</p>
<p>The following SearchIO objects attributes are provided. Rows marked with ‘*’
denotes attributes not available in the hmmer2-text format:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 22%" />
<col style="width: 33%" />
<col style="width: 45%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Object</p></th>
<th class="head"><p>Attribute</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td rowspan="7"><p>QueryResult</p></td>
<td><p>accession</p></td>
<td><p>accession (if present)</p></td>
</tr>
<tr class="row-odd"><td><p>description</p></td>
<td><p>query sequence description</p></td>
</tr>
<tr class="row-even"><td><p>id</p></td>
<td><p>query sequence ID</p></td>
</tr>
<tr class="row-odd"><td><p>program</p></td>
<td><p>HMMER flavor</p></td>
</tr>
<tr class="row-even"><td><p>seq_len*</p></td>
<td><p>full length of query sequence</p></td>
</tr>
<tr class="row-odd"><td><p>target</p></td>
<td><p>target search database</p></td>
</tr>
<tr class="row-even"><td><p>version</p></td>
<td><p>BLAST version</p></td>
</tr>
<tr class="row-odd"><td rowspan="8"><p>Hit</p></td>
<td><p>bias*</p></td>
<td><p>hit-level bias</p></td>
</tr>
<tr class="row-even"><td><p>bitscore</p></td>
<td><p>hit-level score</p></td>
</tr>
<tr class="row-odd"><td><p>description</p></td>
<td><p>hit sequence description</p></td>
</tr>
<tr class="row-even"><td><p>domain_exp_num*</p></td>
<td><p>expected number of domains in
the hit (exp column)</p></td>
</tr>
<tr class="row-odd"><td><p>domain_obs_num</p></td>
<td><p>observed number of domains in
the hit (N column)</p></td>
</tr>
<tr class="row-even"><td><p>evalue</p></td>
<td><p>hit-level e-value</p></td>
</tr>
<tr class="row-odd"><td><p>id</p></td>
<td><p>hit sequence ID</p></td>
</tr>
<tr class="row-even"><td><p>is_included*</p></td>
<td><p>boolean, whether the hit is in
the inclusion threshold or not</p></td>
</tr>
<tr class="row-odd"><td rowspan="12"><p>HSP</p></td>
<td><p>acc_avg*</p></td>
<td><p>expected accuracy per alignment
residue (acc column)</p></td>
</tr>
<tr class="row-even"><td><p>bias*</p></td>
<td><p>hsp-level bias</p></td>
</tr>
<tr class="row-odd"><td><p>bitscore</p></td>
<td><p>hsp-level score</p></td>
</tr>
<tr class="row-even"><td><p>domain_index</p></td>
<td><p>the domain index set by HMMER</p></td>
</tr>
<tr class="row-odd"><td><p>env_end*</p></td>
<td><p>end coordinate of the envelope</p></td>
</tr>
<tr class="row-even"><td><p>env_endtype*</p></td>
<td><p>envelope end types (e.g. ‘[]’,
‘..’, ‘[.’, etc.)</p></td>
</tr>
<tr class="row-odd"><td><p>env_start*</p></td>
<td><p>start coordinate of the envelope</p></td>
</tr>
<tr class="row-even"><td><p>evalue</p></td>
<td><p>hsp-level independent e-value</p></td>
</tr>
<tr class="row-odd"><td><p>evalue_cond*</p></td>
<td><p>hsp-level conditional e-value</p></td>
</tr>
<tr class="row-even"><td><p>hit_endtype</p></td>
<td><p>hit sequence end types</p></td>
</tr>
<tr class="row-odd"><td><p>is_included*</p></td>
<td><p>boolean, whether the hit of the
hsp is in the inclusion
threshold</p></td>
</tr>
<tr class="row-even"><td><p>query_endtype</p></td>
<td><p>query sequence end types</p></td>
</tr>
<tr class="row-odd"><td rowspan="10"><p>HSPFragment
(also via HSP)</p></td>
<td><p>aln_annotation</p></td>
<td><p>alignment similarity string and
other annotations (e.g. PP, CS)</p></td>
</tr>
<tr class="row-even"><td><p>aln_span</p></td>
<td><p>length of alignment fragment</p></td>
</tr>
<tr class="row-odd"><td><p>hit</p></td>
<td><p>hit sequence</p></td>
</tr>
<tr class="row-even"><td><p>hit_end</p></td>
<td><p>hit sequence end coordinate, may
be ‘hmmto’ or ‘alito’ depending
on the HMMER flavor</p></td>
</tr>
<tr class="row-odd"><td><p>hit_start</p></td>
<td><p>hit sequence start coordinate,
may be ‘hmmfrom’ or ‘alifrom’
depending on the HMMER flavor</p></td>
</tr>
<tr class="row-even"><td><p>hit_strand</p></td>
<td><p>hit sequence strand</p></td>
</tr>
<tr class="row-odd"><td><p>query</p></td>
<td><p>query sequence</p></td>
</tr>
<tr class="row-even"><td><p>query_end</p></td>
<td><p>query sequence end coordinate,
may be ‘hmmto’ or ‘alito’
depending on the HMMER flavor</p></td>
</tr>
<tr class="row-odd"><td><p>query_start</p></td>
<td><p>query sequence start coordinate,
may be ‘hmmfrom’ or ‘alifrom’
depending on the HMMER flavor</p></td>
</tr>
<tr class="row-even"><td><p>query_strand</p></td>
<td><p>query sequence strand</p></td>
</tr>
</tbody>
</table>
</section>
<section id="hmmer3-tab">
<h3>hmmer3-tab<a class="headerlink" href="#hmmer3-tab" title="Permalink to this headline"></a></h3>
<p>The following SearchIO objects attributes are provided:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 22%" />
<col style="width: 33%" />
<col style="width: 45%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Object</p></th>
<th class="head"><p>Attribute</p></th>
<th class="head"><p>Column / Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td rowspan="3"><p>QueryResult</p></td>
<td><p>accession</p></td>
<td><p>query accession (if present)</p></td>
</tr>
<tr class="row-odd"><td><p>description</p></td>
<td><p>query sequence description</p></td>
</tr>
<tr class="row-even"><td><p>id</p></td>
<td><p>query name</p></td>
</tr>
<tr class="row-odd"><td rowspan="14"><p>Hit</p></td>
<td><p>accession</p></td>
<td><p>hit accession</p></td>
</tr>
<tr class="row-even"><td><p>bias</p></td>
<td><p>hit-level bias</p></td>
</tr>
<tr class="row-odd"><td><p>bitscore</p></td>
<td><p>hit-level score</p></td>
</tr>
<tr class="row-even"><td><p>description</p></td>
<td><p>hit sequence description</p></td>
</tr>
<tr class="row-odd"><td><p>cluster_num</p></td>
<td><p>clu column</p></td>
</tr>
<tr class="row-even"><td><p>domain_exp_num</p></td>
<td><p>exp column</p></td>
</tr>
<tr class="row-odd"><td><p>domain_included_num</p></td>
<td><p>inc column</p></td>
</tr>
<tr class="row-even"><td><p>domain_obs_num</p></td>
<td><p>dom column</p></td>
</tr>
<tr class="row-odd"><td><p>domain_reported_num</p></td>
<td><p>rep column</p></td>
</tr>
<tr class="row-even"><td><p>env_num</p></td>
<td><p>env column</p></td>
</tr>
<tr class="row-odd"><td><p>evalue</p></td>
<td><p>hit-level evalue</p></td>
</tr>
<tr class="row-even"><td><p>id</p></td>
<td><p>target name</p></td>
</tr>
<tr class="row-odd"><td><p>overlap_num</p></td>
<td><p>ov column</p></td>
</tr>
<tr class="row-even"><td><p>region_num</p></td>
<td><p>reg column</p></td>
</tr>
<tr class="row-odd"><td rowspan="3"><p>HSP</p></td>
<td><p>bias</p></td>
<td><p>bias of the best domain</p></td>
</tr>
<tr class="row-even"><td><p>bitscore</p></td>
<td><p>bitscore of the best domain</p></td>
</tr>
<tr class="row-odd"><td><p>evalue</p></td>
<td><p>evalue of the best domain</p></td>
</tr>
</tbody>
</table>
</section>
<section id="hmmer3-domtab">
<h3>hmmer3-domtab<a class="headerlink" href="#hmmer3-domtab" title="Permalink to this headline"></a></h3>
<p>To parse domain table files, you must use the HMMER flavor that produced the
file. So instead of using ‘hmmer3-domtab’, use either ‘hmmsearch3-domtab’,
‘hmmscan3-domtab’, or ‘phmmer3-domtab’.</p>
<p>The following SearchIO objects attributes are provided:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 22%" />
<col style="width: 33%" />
<col style="width: 45%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Object</p></th>
<th class="head"><p>Attribute</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td rowspan="4"><p>QueryResult</p></td>
<td><p>accession</p></td>
<td><p>accession</p></td>
</tr>
<tr class="row-odd"><td><p>description</p></td>
<td><p>query sequence description</p></td>
</tr>
<tr class="row-even"><td><p>id</p></td>
<td><p>query sequence ID</p></td>
</tr>
<tr class="row-odd"><td><p>seq_len</p></td>
<td><p>full length of query sequence</p></td>
</tr>
<tr class="row-even"><td rowspan="7"><p>Hit</p></td>
<td><p>accession</p></td>
<td><p>accession</p></td>
</tr>
<tr class="row-odd"><td><p>bias</p></td>
<td><p>hit-level bias</p></td>
</tr>
<tr class="row-even"><td><p>bitscore</p></td>
<td><p>hit-level score</p></td>
</tr>
<tr class="row-odd"><td><p>description</p></td>
<td><p>hit sequence description</p></td>
</tr>
<tr class="row-even"><td><p>evalue</p></td>
<td><p>hit-level e-value</p></td>
</tr>
<tr class="row-odd"><td><p>id</p></td>
<td><p>hit sequence ID</p></td>
</tr>
<tr class="row-even"><td><p>seq_len</p></td>
<td><p>length of hit sequence or HMM</p></td>
</tr>
<tr class="row-odd"><td rowspan="8"><p>HSP</p></td>
<td><p>acc_avg</p></td>
<td><p>expected accuracy per alignment
residue (acc column)</p></td>
</tr>
<tr class="row-even"><td><p>bias</p></td>
<td><p>hsp-level bias</p></td>
</tr>
<tr class="row-odd"><td><p>bitscore</p></td>
<td><p>hsp-level score</p></td>
</tr>
<tr class="row-even"><td><p>domain_index</p></td>
<td><p>the domain index set by HMMER</p></td>
</tr>
<tr class="row-odd"><td><p>env_end</p></td>
<td><p>end coordinate of the envelope</p></td>
</tr>
<tr class="row-even"><td><p>env_start</p></td>
<td><p>start coordinate of the envelope</p></td>
</tr>
<tr class="row-odd"><td><p>evalue</p></td>
<td><p>hsp-level independent e-value</p></td>
</tr>
<tr class="row-even"><td><p>evalue_cond</p></td>
<td><p>hsp-level conditional e-value</p></td>
</tr>
<tr class="row-odd"><td rowspan="6"><p>HSPFragment
(also via HSP)</p></td>
<td><p>hit_end</p></td>
<td><p>hit sequence end coordinate, may
be ‘hmmto’ or ‘alito’ depending
on the HMMER flavor</p></td>
</tr>
<tr class="row-even"><td><p>hit_start</p></td>
<td><p>hit sequence start coordinate,
may be ‘hmmfrom’ or ‘alifrom’
depending on the HMMER flavor</p></td>
</tr>
<tr class="row-odd"><td><p>hit_strand</p></td>
<td><p>hit sequence strand</p></td>
</tr>
<tr class="row-even"><td><p>query_end</p></td>
<td><p>query sequence end coordinate,
may be ‘hmmto’ or ‘alito’
depending on the HMMER flavor</p></td>
</tr>
<tr class="row-odd"><td><p>query_start</p></td>
<td><p>query sequence start coordinate,
may be ‘hmmfrom’ or ‘alifrom’
depending on the HMMER flavor</p></td>
</tr>
<tr class="row-even"><td><p>query_strand</p></td>
<td><p>query sequence strand</p></td>
</tr>
</tbody>
</table>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.SearchIO.HHsuiteIO.hhsuite2_text.html" class="btn btn-neutral float-left" title="Bio.SearchIO.HHsuiteIO.hhsuite2_text module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SearchIO.HmmerIO.hmmer2_text.html" class="btn btn-neutral float-right" title="Bio.SearchIO.HmmerIO.hmmer2_text module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2020, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.80.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.78/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>