

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Phylo.BaseTree module &mdash; Biopython 1.78.dev0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Phylo.CDAO module" href="Bio.Phylo.CDAO.html" />
    <link rel="prev" title="Bio.Phylo.PAML.yn00 module" href="Bio.Phylo.PAML.yn00.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.78.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html#subpackages">Subpackages</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Phylo.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html#module-Bio.Phylo">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.Phylo.html">Bio.Phylo package</a> &raquo;</li>
        
      <li>Bio.Phylo.BaseTree module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Phylo/BaseTree.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Phylo.CDAO.html" class="btn btn-neutral float-right" title="Bio.Phylo.CDAO module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Phylo.PAML.yn00.html" class="btn btn-neutral float-left" title="Bio.Phylo.PAML.yn00 module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.Phylo.BaseTree">
<span id="bio-phylo-basetree-module"></span><h1>Bio.Phylo.BaseTree module<a class="headerlink" href="#module-Bio.Phylo.BaseTree" title="Permalink to this headline">¶</a></h1>
<p>Base classes for Bio.Phylo objects.</p>
<p>All object representations for phylogenetic trees should derive from these base
classes in order to use the common methods defined on them.</p>
<dl class="py class">
<dt id="Bio.Phylo.BaseTree.TreeElement">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.BaseTree.</code><code class="sig-name descname">TreeElement</code><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeElement" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for all Bio.Phylo classes.</p>
<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeElement.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeElement.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Show this object’s constructor with its primitive arguments.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeElement.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeElement.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Show this object’s constructor with its primitive arguments.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.BaseTree.TreeMixin">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.BaseTree.</code><code class="sig-name descname">TreeMixin</code><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Methods for Tree- and Clade-based classes.</p>
<p>This lets <code class="docutils literal notranslate"><span class="pre">Tree</span></code> and <code class="docutils literal notranslate"><span class="pre">Clade</span></code> support the same traversal and searching
operations without requiring Clade to inherit from Tree, so Clade isn’t
required to have all of Tree’s attributes – just <code class="docutils literal notranslate"><span class="pre">root</span></code> (a Clade
instance) and <code class="docutils literal notranslate"><span class="pre">is_terminal</span></code>.</p>
<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.find_any">
<code class="sig-name descname">find_any</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.find_any" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the first element found by find_elements(), or None.</p>
<p>This is also useful for checking whether any matching element exists in
the tree, and can be used in a conditional expression.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.find_elements">
<code class="sig-name descname">find_elements</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">terminal</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">'preorder'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.find_elements" title="Permalink to this definition">¶</a></dt>
<dd><p>Find all tree elements matching the given attributes.</p>
<p>The arbitrary keyword arguments indicate the attribute name of the
sub-element and the value to match: string, integer or boolean. Strings
are evaluated as regular expression matches; integers are compared
directly for equality, and booleans evaluate the attribute’s truth value
(True or False) before comparing. To handle nonzero floats, search with
a boolean argument, then filter the result manually.</p>
<p>If no keyword arguments are given, then just the class type is used for
matching.</p>
<p>The result is an iterable through all matching objects, by depth-first
search. (Not necessarily the same order as the elements appear in the
source file!)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>target<span class="classifier">TreeElement instance, type, dict, or callable</span></dt><dd><p>Specifies the characteristics to search for. (The default,
TreeElement, matches any standard Bio.Phylo type.)</p>
</dd>
<dt>terminal<span class="classifier">bool</span></dt><dd><p>A boolean value to select for or against terminal nodes (a.k.a.
leaf nodes). True searches for only terminal nodes, False
excludes terminal nodes, and the default, None, searches both
terminal and non-terminal nodes, as well as any tree elements
lacking the <code class="docutils literal notranslate"><span class="pre">is_terminal</span></code> method.</p>
</dd>
<dt>order<span class="classifier">{‘preorder’, ‘postorder’, ‘level’}</span></dt><dd><p>Tree traversal order: ‘preorder’ (default) is depth-first
search, ‘postorder’ is DFS with child nodes preceding parents,
and ‘level’ is breadth-first search.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Phylo</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">phx</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">PhyloXMLIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;PhyloXML/phyloxml_examples.xml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">matches</span> <span class="o">=</span> <span class="n">phx</span><span class="o">.</span><span class="n">phylogenies</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">find_elements</span><span class="p">(</span><span class="n">code</span><span class="o">=</span><span class="s1">&#39;OCTVU&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span>
<span class="go">Taxonomy(code=&#39;OCTVU&#39;, scientific_name=&#39;Octopus vulgaris&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.find_clades">
<code class="sig-name descname">find_clades</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">terminal</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">'preorder'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.find_clades" title="Permalink to this definition">¶</a></dt>
<dd><p>Find each clade containing a matching element.</p>
<p>That is, find each element as with find_elements(), but return the
corresponding clade object. (This is usually what you want.)</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>an iterable through all matching objects, searching
depth-first (preorder) by default.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.get_path">
<code class="sig-name descname">get_path</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.get_path" title="Permalink to this definition">¶</a></dt>
<dd><p>List the clades directly between this root and the given target.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of all clade objects along this path, ending with the
given target, but excluding the root clade.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.get_nonterminals">
<code class="sig-name descname">get_nonterminals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">'preorder'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.get_nonterminals" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of all of this tree’s nonterminal (internal) nodes.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.get_terminals">
<code class="sig-name descname">get_terminals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">order</span><span class="o">=</span><span class="default_value">'preorder'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.get_terminals" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of all of this tree’s terminal (leaf) nodes.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.trace">
<code class="sig-name descname">trace</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">finish</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.trace" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all clade object between two targets in this tree.</p>
<p>Excluding <code class="docutils literal notranslate"><span class="pre">start</span></code>, including <code class="docutils literal notranslate"><span class="pre">finish</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.common_ancestor">
<code class="sig-name descname">common_ancestor</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">targets</span></em>, <em class="sig-param"><span class="o">*</span><span class="n">more_targets</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.common_ancestor" title="Permalink to this definition">¶</a></dt>
<dd><p>Most recent common ancestor (clade) of all the given targets.</p>
<dl class="simple">
<dt>Edge cases:</dt><dd><ul class="simple">
<li><p>If no target is given, returns self.root</p></li>
<li><p>If 1 target is given, returns the target</p></li>
<li><p>If any target is not found in this tree, raises a ValueError</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.count_terminals">
<code class="sig-name descname">count_terminals</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.count_terminals" title="Permalink to this definition">¶</a></dt>
<dd><p>Count the number of terminal (leaf) nodes within this tree.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.depths">
<code class="sig-name descname">depths</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">unit_branch_lengths</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.depths" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a mapping of tree clades to depths (by branch length).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>unit_branch_lengths<span class="classifier">bool</span></dt><dd><p>If True, count only the number of branches (levels in the tree).
By default the distance is the cumulative branch length leading
to the clade.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict of {clade: depth}, where keys are all of the Clade
instances in the tree, and values are the distance from the root to
each clade (including terminals).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.distance">
<code class="sig-name descname">distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target1</span></em>, <em class="sig-param"><span class="n">target2</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the sum of the branch lengths between two targets.</p>
<p>If only one target is specified, the other is the root of this tree.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.is_bifurcating">
<code class="sig-name descname">is_bifurcating</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.is_bifurcating" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if tree downstream of node is strictly bifurcating.</p>
<p>I.e., all nodes have either 2 or 0 children (internal or external,
respectively). The root may have 3 descendents and still be considered
part of a bifurcating tree, because it has no ancestor.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.is_monophyletic">
<code class="sig-name descname">is_monophyletic</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">terminals</span></em>, <em class="sig-param"><span class="o">*</span><span class="n">more_terminals</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.is_monophyletic" title="Permalink to this definition">¶</a></dt>
<dd><p>MRCA of terminals if they comprise a complete subclade, or False.</p>
<p>I.e., there exists a clade such that its terminals are the same set as
the given targets.</p>
<p>The given targets must be terminals of the tree.</p>
<p>To match both <code class="docutils literal notranslate"><span class="pre">Bio.Nexus.Trees</span></code> and the other multi-target methods in
Bio.Phylo, arguments to this method can be specified either of two ways:
(i) as a single list of targets, or (ii) separately specified targets,
e.g. is_monophyletic(t1, t2, t3) – but not both.</p>
<p>For convenience, this method returns the common ancestor (MCRA) of the
targets if they are monophyletic (instead of the value True), and False
otherwise.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>common ancestor if terminals are monophyletic, otherwise False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.is_parent_of">
<code class="sig-name descname">is_parent_of</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.is_parent_of" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if target is a descendent of this tree.</p>
<p>Not required to be a direct descendent.</p>
<p>To check only direct descendents of a clade, simply use list membership
testing: <code class="docutils literal notranslate"><span class="pre">if</span> <span class="pre">subclade</span> <span class="pre">in</span> <span class="pre">clade:</span> <span class="pre">...</span></code></p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.is_preterminal">
<code class="sig-name descname">is_preterminal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.is_preterminal" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if all direct descendents are terminal.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.total_branch_length">
<code class="sig-name descname">total_branch_length</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.total_branch_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the sum of all the branch lengths in this tree.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.collapse">
<code class="sig-name descname">collapse</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.collapse" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete target from the tree, relinking its children to its parent.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the parent clade.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.collapse_all">
<code class="sig-name descname">collapse_all</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.collapse_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Collapse all the descendents of this tree, leaving only terminals.</p>
<p>Total branch lengths are preserved, i.e. the distance to each terminal
stays the same.</p>
<p>For example, this will safely collapse nodes with poor bootstrap
support:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Phylo</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;PhyloXML/apaf.xml&#39;</span><span class="p">,</span> <span class="s1">&#39;phyloxml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total branch length </span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree</span><span class="o">.</span><span class="n">total_branch_length</span><span class="p">())</span>
<span class="go">Total branch length 20.44</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">collapse_all</span><span class="p">(</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">confidence</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">c</span><span class="o">.</span><span class="n">confidence</span> <span class="o">&lt;</span> <span class="mi">70</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total branch length </span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree</span><span class="o">.</span><span class="n">total_branch_length</span><span class="p">())</span>
<span class="go">Total branch length 21.37</span>
</pre></div>
</div>
<p>This implementation avoids strange side-effects by using level-order
traversal and testing all clade properties (versus the target
specification) up front. In particular, if a clade meets the target
specification in the original tree, it will be collapsed.  For example,
if the condition is:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Phylo</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;PhyloXML/apaf.xml&#39;</span><span class="p">,</span> <span class="s1">&#39;phyloxml&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total branch length </span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree</span><span class="o">.</span><span class="n">total_branch_length</span><span class="p">())</span>
<span class="go">Total branch length 20.44</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tree</span><span class="o">.</span><span class="n">collapse_all</span><span class="p">(</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">branch_length</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total branch length </span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">tree</span><span class="o">.</span><span class="n">total_branch_length</span><span class="p">())</span>
<span class="go">Total branch length 21.13</span>
</pre></div>
</div>
<p>Collapsing a clade’s parent node adds the parent’s branch length to the
child, so during the execution of collapse_all, a clade’s branch_length
may increase. In this implementation, clades are collapsed according to
their properties in the original tree, not the properties when tree
traversal reaches the clade. (It’s easier to debug.) If you want the
other behavior (incremental testing), modifying the source code of this
function is straightforward.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.ladderize">
<code class="sig-name descname">ladderize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">reverse</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.ladderize" title="Permalink to this definition">¶</a></dt>
<dd><p>Sort clades in-place according to the number of terminal nodes.</p>
<p>Deepest clades are last by default. Use <code class="docutils literal notranslate"><span class="pre">reverse=True</span></code> to sort clades
deepest-to-shallowest.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.prune">
<code class="sig-name descname">prune</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.prune" title="Permalink to this definition">¶</a></dt>
<dd><p>Prunes a terminal clade from the tree.</p>
<p>If taxon is from a bifurcation, the connecting node will be collapsed
and its branch length added to remaining terminal node. This might be no
longer be a meaningful value.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>parent clade of the pruned target</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.TreeMixin.split">
<code class="sig-name descname">split</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">n</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">1.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.TreeMixin.split" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate n (default 2) new descendants.</p>
<p>In a species tree, this is a speciation event.</p>
<p>New clades have the given branch_length and the same name as this
clade’s root plus an integer suffix (counting from 0). For example,
splitting a clade named “A” produces sub-clades named “A0” and “A1”.
If the clade has no name, the prefix “n” is used for child nodes, e.g.
“n0” and “n1”.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.BaseTree.Tree">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.BaseTree.</code><code class="sig-name descname">Tree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">root</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rooted</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">id</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Phylo.BaseTree.TreeElement" title="Bio.Phylo.BaseTree.TreeElement"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.BaseTree.TreeElement</span></code></a>, <a class="reference internal" href="#Bio.Phylo.BaseTree.TreeMixin" title="Bio.Phylo.BaseTree.TreeMixin"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.BaseTree.TreeMixin</span></code></a></p>
<p>A phylogenetic tree, containing global info for the phylogeny.</p>
<p>The structure and node-specific data is accessible through the ‘root’
clade attached to the Tree instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>root<span class="classifier">Clade</span></dt><dd><p>The starting node of the tree. If the tree is rooted, this will
usually be the root node.</p>
</dd>
<dt>rooted<span class="classifier">bool</span></dt><dd><p>Whether or not the tree is rooted. By default, a tree is assumed to
be rooted.</p>
</dd>
<dt>id<span class="classifier">str</span></dt><dd><p>The identifier of the tree, if there is one.</p>
</dd>
<dt>name<span class="classifier">str</span></dt><dd><p>The name of the tree, in essence a label.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">root</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rooted</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">id</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize parameter for phylogenetic tree.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.from_clade">
<em class="property">classmethod </em><code class="sig-name descname">from_clade</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">clade</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.from_clade" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new Tree object given a clade.</p>
<p>Keyword arguments are the usual <code class="docutils literal notranslate"><span class="pre">Tree</span></code> constructor parameters.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.randomized">
<em class="property">classmethod </em><code class="sig-name descname">randomized</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxa</span></em>, <em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">branch_stdev</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.randomized" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a randomized bifurcating tree given a list of taxa.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>taxa</strong> – Either an integer specifying the number of taxa to create
(automatically named taxon#), or an iterable of taxon names, as
strings.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tree of the same type as this class.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.clade">
<em class="property">property </em><code class="sig-name descname">clade</code><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.clade" title="Permalink to this definition">¶</a></dt>
<dd><p>Return first clade in this tree (not itself).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.as_phyloxml">
<code class="sig-name descname">as_phyloxml</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.as_phyloxml" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert this tree to a PhyloXML-compatible Phylogeny.</p>
<p>This lets you use the additional annotation types PhyloXML defines, and
save this information when you write this tree as ‘phyloxml’.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.root_with_outgroup">
<code class="sig-name descname">root_with_outgroup</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">outgroup_targets</span></em>, <em class="sig-param"><span class="o">*</span><span class="n">more_targets</span></em>, <em class="sig-param"><span class="n">outgroup_branch_length</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.root_with_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p>Reroot this tree with the outgroup clade containing outgroup_targets.</p>
<p>Operates in-place.</p>
<dl class="simple">
<dt>Edge cases:</dt><dd><ul class="simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">outgroup</span> <span class="pre">==</span> <span class="pre">self.root</span></code>, no change</p></li>
<li><p>If outgroup is terminal, create new bifurcating root node with a
0-length branch to the outgroup</p></li>
<li><p>If outgroup is internal, use the given outgroup node as the new
trifurcating root, keeping branches the same</p></li>
<li><p>If the original root was bifurcating, drop it from the tree,
preserving total branch lengths</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>outgroup_branch_length</strong> – <p>length of the branch leading to the
outgroup after rerooting. If not specified (None), then:</p>
<ul class="simple">
<li><p>If the outgroup is an internal node (not a single terminal taxon),
then use that node as the new root.</p></li>
<li><p>Otherwise, create a new root node as the parent of the outgroup.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.root_at_midpoint">
<code class="sig-name descname">root_at_midpoint</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.root_at_midpoint" title="Permalink to this definition">¶</a></dt>
<dd><p>Root the tree at the midpoint of the two most distant taxa.</p>
<p>This operates in-place, leaving a bifurcating root. The topology of the
tree is otherwise retained, though no guarantees are made about the
stability of clade/node/taxon ordering.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.is_terminal">
<code class="sig-name descname">is_terminal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.is_terminal" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the root of this tree is terminal.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.__format__">
<code class="sig-name descname">__format__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">format_spec</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.__format__" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the tree as a string in the specified file format.</p>
<p>This method supports the <code class="docutils literal notranslate"><span class="pre">format</span></code> built-in function added in Python
2.6/3.0.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>format_spec</strong> – a lower-case string supported by <code class="docutils literal notranslate"><span class="pre">Bio.Phylo.write</span></code>
as an output file format.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.format">
<code class="sig-name descname">format</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">format</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.format" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the tree as a string in the specified file format.</p>
<p>This duplicates the __format__ magic method for pre-2.6 Pythons.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Tree.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Tree.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a string representation of the entire tree.</p>
<p>Serialize each sub-clade recursively using <code class="docutils literal notranslate"><span class="pre">repr</span></code> to create a summary
of the object structure.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.BaseTree.Clade">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.BaseTree.</code><code class="sig-name descname">Clade</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">clades</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">confidence</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">color</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">width</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Phylo.BaseTree.TreeElement" title="Bio.Phylo.BaseTree.TreeElement"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.BaseTree.TreeElement</span></code></a>, <a class="reference internal" href="#Bio.Phylo.BaseTree.TreeMixin" title="Bio.Phylo.BaseTree.TreeMixin"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.BaseTree.TreeMixin</span></code></a></p>
<p>A recursively defined sub-tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>branch_length<span class="classifier">str</span></dt><dd><p>The length of the branch leading to the root node of this clade.</p>
</dd>
<dt>name<span class="classifier">str</span></dt><dd><p>The clade’s name (a label).</p>
</dd>
<dt>clades<span class="classifier">list</span></dt><dd><p>Sub-trees rooted directly under this tree’s root.</p>
</dd>
<dt>confidence<span class="classifier">number</span></dt><dd><p>Support.</p>
</dd>
<dt>color<span class="classifier">BranchColor</span></dt><dd><p>The display color of the branch and descendents.</p>
</dd>
<dt>width<span class="classifier">number</span></dt><dd><p>The display width of the branch and descendents.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">branch_length</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">clades</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">confidence</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">color</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">width</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Define parameters for the Clade tree.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.root">
<em class="property">property </em><code class="sig-name descname">root</code><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.root" title="Permalink to this definition">¶</a></dt>
<dd><p>Allow TreeMixin methods to traverse clades properly.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.is_terminal">
<code class="sig-name descname">is_terminal</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.is_terminal" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if this is a terminal (leaf) node.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Get clades by index (integer or slice).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.__iter__">
<code class="sig-name descname">__iter__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.__iter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate through this tree’s direct descendent clades (sub-trees).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.__len__">
<code class="sig-name descname">__len__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.__len__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the number of clades directy under the root.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.__bool__">
<code class="sig-name descname">__bool__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.__bool__" title="Permalink to this definition">¶</a></dt>
<dd><p>Boolean value of an instance of this class (True).</p>
<p>NB: If this method is not defined, but <code class="docutils literal notranslate"><span class="pre">__len__</span></code>  is, then the object
is considered true if the result of <code class="docutils literal notranslate"><span class="pre">__len__()</span></code> is nonzero. We want
Clade instances to always be considered True.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return name of the class instance.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.Clade.color">
<em class="property">property </em><code class="sig-name descname">color</code><a class="headerlink" href="#Bio.Phylo.BaseTree.Clade.color" title="Permalink to this definition">¶</a></dt>
<dd><p>Branch color.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.BaseTree.BranchColor">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.BaseTree.</code><code class="sig-name descname">BranchColor</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">red</span></em>, <em class="sig-param"><span class="n">green</span></em>, <em class="sig-param"><span class="n">blue</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Indicates the color of a clade when rendered graphically.</p>
<p>The color should be interpreted by client code (e.g. visualization
programs) as applying to the whole clade, unless overwritten by the
color(s) of sub-clades.</p>
<p>Color values must be integers from 0 to 255.</p>
<dl class="py attribute">
<dt id="Bio.Phylo.BaseTree.BranchColor.color_names">
<code class="sig-name descname">color_names</code><em class="property"> = {'aqua': (0, 255, 255), 'b': (0, 0, 255), 'black': (0, 0, 0), 'blue': (0, 0, 255), 'brown': (165, 42, 42), 'c': (0, 255, 255), 'cyan': (0, 255, 255), 'fuchsia': (255, 0, 255), 'g': (0, 128, 0), 'gold': (255, 215, 0), 'gray': (128, 128, 128), 'green': (0, 128, 0), 'grey': (128, 128, 128), 'k': (0, 0, 0), 'lime': (0, 255, 0), 'm': (255, 0, 255), 'magenta': (255, 0, 255), 'maroon': (128, 0, 0), 'navy': (0, 0, 128), 'olive': (128, 128, 0), 'orange': (255, 165, 0), 'pink': (255, 192, 203), 'purple': (128, 0, 128), 'r': (255, 0, 0), 'red': (255, 0, 0), 'salmon': (250, 128, 114), 'silver': (192, 192, 192), 'tan': (210, 180, 140), 'teal': (0, 128, 128), 'w': (255, 255, 255), 'white': (255, 255, 255), 'y': (255, 255, 0), 'yellow': (255, 255, 0)}</em><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor.color_names" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.BranchColor.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">red</span></em>, <em class="sig-param"><span class="n">green</span></em>, <em class="sig-param"><span class="n">blue</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize BranchColor for a tree.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.BranchColor.from_hex">
<em class="property">classmethod </em><code class="sig-name descname">from_hex</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">hexstr</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor.from_hex" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct a BranchColor object from a hexadecimal string.</p>
<p>The string format is the same style used in HTML and CSS, such as
‘#FF8000’ for an RGB value of (255, 128, 0).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.BranchColor.from_name">
<em class="property">classmethod </em><code class="sig-name descname">from_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">colorname</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor.from_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct a BranchColor object by the color’s name.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.BranchColor.to_hex">
<code class="sig-name descname">to_hex</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor.to_hex" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a 24-bit hexadecimal RGB representation of this color.</p>
<p>The returned string is suitable for use in HTML/CSS, as a color
parameter in matplotlib, and perhaps other situations.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bc</span> <span class="o">=</span> <span class="n">BranchColor</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bc</span><span class="o">.</span><span class="n">to_hex</span><span class="p">()</span>
<span class="go">&#39;#0cc864&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.BranchColor.to_rgb">
<code class="sig-name descname">to_rgb</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor.to_rgb" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a tuple of RGB values (0 to 255) representing this color.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bc</span> <span class="o">=</span> <span class="n">BranchColor</span><span class="p">(</span><span class="mi">255</span><span class="p">,</span> <span class="mi">165</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bc</span><span class="o">.</span><span class="n">to_rgb</span><span class="p">()</span>
<span class="go">(255, 165, 0)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.BranchColor.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Preserve the standard RGB order when representing this object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.BaseTree.BranchColor.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.BaseTree.BranchColor.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the color’s RGB values.</p>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.Phylo.CDAO.html" class="btn btn-neutral float-right" title="Bio.Phylo.CDAO module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Phylo.PAML.yn00.html" class="btn btn-neutral float-left" title="Bio.Phylo.PAML.yn00 module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>