<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.UniGene package &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.UniProt package" href="Bio.UniProt.html" />
    <link rel="prev" title="Bio.TogoWS package" href="Bio.TogoWS.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
      <li class="breadcrumb-item active">Bio.UniGene package</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/UniGene/__init__.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.TogoWS.html" class="btn btn-neutral float-left" title="Bio.TogoWS package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.UniProt.html" class="btn btn-neutral float-right" title="Bio.UniProt package" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bio-unigene-package">
<h1>Bio.UniGene package<a class="headerlink" href="#bio-unigene-package" title="Permalink to this headline"></a></h1>
<section id="module-Bio.UniGene">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.UniGene" title="Permalink to this headline"></a></h2>
<p>Parse Unigene flat file format files such as the Hs.data file.</p>
<p>Here is an overview of the flat file format that this parser deals with:</p>
<blockquote>
<div><p>Line types/qualifiers:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ID</span>           <span class="n">UniGene</span> <span class="n">cluster</span> <span class="n">ID</span>
<span class="n">TITLE</span>        <span class="n">Title</span> <span class="k">for</span> <span class="n">the</span> <span class="n">cluster</span>
<span class="n">GENE</span>         <span class="n">Gene</span> <span class="n">symbol</span>
<span class="n">CYTOBAND</span>     <span class="n">Cytological</span> <span class="n">band</span>
<span class="n">EXPRESS</span>      <span class="n">Tissues</span> <span class="n">of</span> <span class="n">origin</span> <span class="k">for</span> <span class="n">ESTs</span> <span class="ow">in</span> <span class="n">cluster</span>
<span class="n">RESTR_EXPR</span>   <span class="n">Single</span> <span class="n">tissue</span> <span class="ow">or</span> <span class="n">development</span> <span class="n">stage</span> <span class="n">contributes</span>
             <span class="n">more</span> <span class="n">than</span> <span class="n">half</span> <span class="n">the</span> <span class="n">total</span> <span class="n">EST</span> <span class="n">frequency</span> <span class="k">for</span> <span class="n">this</span> <span class="n">gene</span><span class="o">.</span>
<span class="n">GNM_TERMINUS</span> <span class="n">genomic</span> <span class="n">confirmation</span> <span class="n">of</span> <span class="n">presence</span> <span class="n">of</span> <span class="n">a</span> <span class="mi">3</span><span class="s1">&#39; terminus;</span>
             <span class="n">T</span> <span class="k">if</span> <span class="n">a</span> <span class="n">non</span><span class="o">-</span><span class="n">templated</span> <span class="n">polyA</span> <span class="n">tail</span> <span class="ow">is</span> <span class="n">found</span> <span class="n">among</span>
             <span class="n">a</span> <span class="n">cluster</span><span class="s1">&#39;s sequences; else</span>
             <span class="n">I</span> <span class="k">if</span> <span class="n">templated</span> <span class="n">As</span> <span class="n">are</span> <span class="n">found</span> <span class="ow">in</span> <span class="n">genomic</span> <span class="n">sequence</span> <span class="ow">or</span>
             <span class="n">S</span> <span class="k">if</span> <span class="n">a</span> <span class="n">canonical</span> <span class="n">polyA</span> <span class="n">signal</span> <span class="ow">is</span> <span class="n">found</span> <span class="n">on</span>
               <span class="n">the</span> <span class="n">genomic</span> <span class="n">sequence</span>
<span class="n">GENE_ID</span>      <span class="n">Entrez</span> <span class="n">gene</span> <span class="n">identifier</span> <span class="n">associated</span> <span class="k">with</span> <span class="n">at</span> <span class="n">least</span> <span class="n">one</span>
             <span class="n">sequence</span> <span class="ow">in</span> <span class="n">this</span> <span class="n">cluster</span><span class="p">;</span>
             <span class="n">to</span> <span class="n">be</span> <span class="n">used</span> <span class="n">instead</span> <span class="n">of</span> <span class="n">LocusLink</span><span class="o">.</span>
<span class="n">LOCUSLINK</span>    <span class="n">LocusLink</span> <span class="n">identifier</span> <span class="n">associated</span> <span class="k">with</span> <span class="n">at</span> <span class="n">least</span> <span class="n">one</span>
             <span class="n">sequence</span> <span class="ow">in</span> <span class="n">this</span> <span class="n">cluster</span><span class="p">;</span>
             <span class="n">deprecated</span> <span class="ow">in</span> <span class="n">favor</span> <span class="n">of</span> <span class="n">GENE_ID</span>
<span class="n">HOMOL</span>        <span class="n">Homology</span><span class="p">;</span>
<span class="n">CHROMOSOME</span>   <span class="n">Chromosome</span><span class="o">.</span>  <span class="n">For</span> <span class="n">plants</span><span class="p">,</span> <span class="n">CHROMOSOME</span> <span class="n">refers</span> <span class="n">to</span> <span class="n">mapping</span>
             <span class="n">on</span> <span class="n">the</span> <span class="n">arabidopsis</span> <span class="n">genome</span><span class="o">.</span>
<span class="n">STS</span>          <span class="n">STS</span>
     <span class="n">ACC</span><span class="o">=</span>         <span class="n">GenBank</span><span class="o">/</span><span class="n">EMBL</span><span class="o">/</span><span class="n">DDBJ</span> <span class="n">accession</span> <span class="n">number</span> <span class="n">of</span> <span class="n">STS</span>
                  <span class="p">[</span><span class="n">optional</span> <span class="n">field</span><span class="p">]</span>
     <span class="n">UNISTS</span><span class="o">=</span>      <span class="n">identifier</span> <span class="ow">in</span> <span class="n">NCBI</span><span class="s1">&#39;s UNISTS database</span>
<span class="n">TXMAP</span>        <span class="n">Transcript</span> <span class="nb">map</span> <span class="n">interval</span>
     <span class="n">MARKER</span><span class="o">=</span>      <span class="n">Marker</span> <span class="n">found</span> <span class="n">on</span> <span class="n">at</span> <span class="n">least</span> <span class="n">one</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">this</span>
                  <span class="n">cluster</span>
     <span class="n">RHPANEL</span><span class="o">=</span>     <span class="n">Radiation</span> <span class="n">Hybrid</span> <span class="n">panel</span> <span class="n">used</span> <span class="n">to</span> <span class="n">place</span> <span class="n">marker</span>
<span class="n">PROTSIM</span>      <span class="n">Protein</span> <span class="n">Similarity</span> <span class="n">data</span> <span class="k">for</span> <span class="n">the</span> <span class="n">sequence</span> <span class="k">with</span>
             <span class="n">highest</span><span class="o">-</span><span class="n">scoring</span> <span class="n">protein</span> <span class="n">similarity</span> <span class="ow">in</span> <span class="n">this</span> <span class="n">cluster</span>
     <span class="n">ORG</span><span class="o">=</span>         <span class="n">Organism</span>
     <span class="n">PROTGI</span><span class="o">=</span>      <span class="n">Sequence</span> <span class="n">GI</span> <span class="n">of</span> <span class="n">protein</span>
     <span class="n">PROTID</span><span class="o">=</span>      <span class="n">Sequence</span> <span class="n">ID</span> <span class="n">of</span> <span class="n">protein</span>
     <span class="n">PCT</span><span class="o">=</span>         <span class="n">Percent</span> <span class="n">alignment</span>
     <span class="n">ALN</span><span class="o">=</span>         <span class="n">length</span> <span class="n">of</span> <span class="n">aligned</span> <span class="n">region</span> <span class="p">(</span><span class="n">aa</span><span class="p">)</span>
<span class="n">SCOUNT</span>       <span class="n">Number</span> <span class="n">of</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">cluster</span>
<span class="n">SEQUENCE</span>     <span class="n">Sequence</span>
     <span class="n">ACC</span><span class="o">=</span>         <span class="n">GenBank</span><span class="o">/</span><span class="n">EMBL</span><span class="o">/</span><span class="n">DDBJ</span> <span class="n">accession</span> <span class="n">number</span> <span class="n">of</span> <span class="n">sequence</span>
     <span class="n">NID</span><span class="o">=</span>         <span class="n">Unique</span> <span class="n">nucleotide</span> <span class="n">sequence</span> <span class="n">identifier</span> <span class="p">(</span><span class="n">gi</span><span class="p">)</span>
     <span class="n">PID</span><span class="o">=</span>         <span class="n">Unique</span> <span class="n">protein</span> <span class="n">sequence</span> <span class="n">identifier</span> <span class="p">(</span><span class="n">used</span> <span class="k">for</span>
                  <span class="n">non</span><span class="o">-</span><span class="n">ESTs</span><span class="p">)</span>
     <span class="n">CLONE</span><span class="o">=</span>       <span class="n">Clone</span> <span class="n">identifier</span> <span class="p">(</span><span class="n">used</span> <span class="k">for</span> <span class="n">ESTs</span> <span class="n">only</span><span class="p">)</span>
     <span class="n">END</span><span class="o">=</span>         <span class="n">End</span> <span class="p">(</span><span class="mi">5</span><span class="s1">&#39;/3&#39;</span><span class="p">)</span> <span class="n">of</span> <span class="n">clone</span> <span class="n">insert</span> <span class="n">read</span> <span class="p">(</span><span class="n">used</span> <span class="k">for</span>
                  <span class="n">ESTs</span> <span class="n">only</span><span class="p">)</span>
     <span class="n">LID</span><span class="o">=</span>         <span class="n">Library</span> <span class="n">ID</span><span class="p">;</span> <span class="n">see</span> <span class="n">Hs</span><span class="o">.</span><span class="n">lib</span><span class="o">.</span><span class="n">info</span> <span class="k">for</span> <span class="n">library</span> <span class="n">name</span>
                  <span class="ow">and</span> <span class="n">tissue</span>
     <span class="n">MGC</span><span class="o">=</span>         <span class="mi">5</span><span class="s1">&#39; CDS-completeness indicator; if present, the</span>
                  <span class="n">clone</span> <span class="n">associated</span> <span class="k">with</span> <span class="n">this</span> <span class="n">sequence</span> <span class="ow">is</span> <span class="n">believed</span>
                  <span class="n">CDS</span><span class="o">-</span><span class="n">complete</span><span class="o">.</span> <span class="n">A</span> <span class="n">value</span> <span class="n">greater</span> <span class="n">than</span> <span class="mi">511</span> <span class="ow">is</span> <span class="n">the</span> <span class="n">gi</span>
                  <span class="n">of</span> <span class="n">the</span> <span class="n">CDS</span><span class="o">-</span><span class="n">complete</span> <span class="n">mRNA</span> <span class="n">matched</span> <span class="n">by</span> <span class="n">the</span> <span class="n">EST</span><span class="p">,</span>
                  <span class="n">otherwise</span> <span class="n">the</span> <span class="n">value</span> <span class="ow">is</span> <span class="n">an</span> <span class="n">indicator</span> <span class="n">of</span> <span class="n">the</span>
                  <span class="n">reliability</span> <span class="n">of</span> <span class="n">the</span> <span class="n">test</span> <span class="n">indicating</span> <span class="n">CDS</span>
                  <span class="n">completeness</span><span class="p">;</span> <span class="n">higher</span> <span class="n">values</span> <span class="n">indicate</span> <span class="n">more</span>
                  <span class="n">reliable</span> <span class="n">CDS</span><span class="o">-</span><span class="n">completeness</span> <span class="n">predictions</span><span class="o">.</span>
    <span class="n">SEQTYPE</span><span class="o">=</span>      <span class="n">Description</span> <span class="n">of</span> <span class="n">the</span> <span class="n">nucleotide</span> <span class="n">sequence</span><span class="o">.</span>
                  <span class="n">Possible</span> <span class="n">values</span> <span class="n">are</span> <span class="n">mRNA</span><span class="p">,</span> <span class="n">EST</span> <span class="ow">and</span> <span class="n">HTC</span><span class="o">.</span>
    <span class="n">TRACE</span><span class="o">=</span>        <span class="n">The</span> <span class="n">Trace</span> <span class="n">ID</span> <span class="n">of</span> <span class="n">the</span> <span class="n">EST</span> <span class="n">sequence</span><span class="p">,</span> <span class="k">as</span> <span class="n">provided</span> <span class="n">by</span>
                  <span class="n">NCBI</span> <span class="n">Trace</span> <span class="n">Archive</span>
</pre></div>
</div>
</div></blockquote>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.UniGene.SequenceLine">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.UniGene.</span></span><span class="sig-name descname"><span class="pre">SequenceLine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.SequenceLine" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Store the information for one SEQUENCE line from a Unigene file.</p>
<p>Initialize with the text part of the SEQUENCE line, or nothing.</p>
<dl class="simple">
<dt>Attributes and descriptions (access as LOWER CASE):</dt><dd><ul class="simple">
<li><p>ACC=         GenBank/EMBL/DDBJ accession number of sequence</p></li>
<li><p>NID=         Unique nucleotide sequence identifier (gi)</p></li>
<li><p>PID=         Unique protein sequence identifier (used for non-ESTs)</p></li>
<li><p>CLONE=       Clone identifier (used for ESTs only)</p></li>
<li><p>END=         End (5’/3’) of clone insert read (used for ESTs only)</p></li>
<li><p>LID=         Library ID; see Hs.lib.info for library name and tissue</p></li>
<li><p>MGC=         5’ CDS-completeness indicator; if present,
the clone associated with this sequence
is believed CDS-complete. A value greater than 511
is the gi of the CDS-complete mRNA matched by the EST,
otherwise the value is an indicator of the reliability
of the test indicating CDS completeness;
higher values indicate more reliable CDS-completeness
predictions.</p></li>
<li><p>SEQTYPE=     Description of the nucleotide sequence. Possible values
are mRNA, EST and HTC.</p></li>
<li><p>TRACE=       The Trace ID of the EST sequence, as provided by NCBI
Trace Archive</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.UniGene.SequenceLine.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.SequenceLine.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.UniGene.SequenceLine.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.SequenceLine.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Return UniGene SequenceLine object as a string.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.UniGene.ProtsimLine">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.UniGene.</span></span><span class="sig-name descname"><span class="pre">ProtsimLine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.ProtsimLine" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Store the information for one PROTSIM line from a Unigene file.</p>
<p>Initialize with the text part of the PROTSIM line, or nothing.</p>
<p>Attributes and descriptions (access as LOWER CASE)
ORG=         Organism
PROTGI=      Sequence GI of protein
PROTID=      Sequence ID of protein
PCT=         Percent alignment
ALN=         length of aligned region (aa)</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.UniGene.ProtsimLine.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.ProtsimLine.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.UniGene.ProtsimLine.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.ProtsimLine.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Return UniGene ProtsimLine object as a string.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.UniGene.STSLine">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.UniGene.</span></span><span class="sig-name descname"><span class="pre">STSLine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.STSLine" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Store the information for one STS line from a Unigene file.</p>
<p>Initialize with the text part of the STS line, or nothing.</p>
<p>Attributes and descriptions (access as LOWER CASE)</p>
<p>ACC=         GenBank/EMBL/DDBJ accession number of STS [optional field]
UNISTS=      identifier in NCBI’s UNISTS database</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.UniGene.STSLine.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.STSLine.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.UniGene.STSLine.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.STSLine.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Return UniGene STSLine object as a string.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.UniGene.Record">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.UniGene.</span></span><span class="sig-name descname"><span class="pre">Record</span></span><a class="headerlink" href="#Bio.UniGene.Record" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Store a Unigene record.</p>
<p>Here is what is stored:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="bp">self</span><span class="o">.</span><span class="n">ID</span>           <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># ID line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">species</span>      <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># Hs, Bt, etc.</span>
<span class="bp">self</span><span class="o">.</span><span class="n">title</span>        <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># TITLE line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">symbol</span>       <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># GENE line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">cytoband</span>     <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># CYTOBAND line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">express</span>      <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># EXPRESS line, parsed on &#39;;&#39;</span>
                        <span class="c1"># Will be an array of strings</span>
<span class="bp">self</span><span class="o">.</span><span class="n">restr_expr</span>   <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># RESTR_EXPR line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">gnm_terminus</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># GNM_TERMINUS line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">gene_id</span>      <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># GENE_ID line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">locuslink</span>    <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># LOCUSLINK line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">homol</span>        <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># HOMOL line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">chromosome</span>   <span class="o">=</span> <span class="s1">&#39;&#39;</span>  <span class="c1"># CHROMOSOME line</span>
<span class="bp">self</span><span class="o">.</span><span class="n">protsim</span>      <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># PROTSIM entries, array of Protsims</span>
                        <span class="c1"># Type ProtsimLine</span>
<span class="bp">self</span><span class="o">.</span><span class="n">sequence</span>     <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># SEQUENCE entries, array of Sequence entries</span>
                        <span class="c1"># Type SequenceLine</span>
<span class="bp">self</span><span class="o">.</span><span class="n">sts</span>          <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># STS entries, array of STS entries</span>
                        <span class="c1"># Type STSLine</span>
<span class="bp">self</span><span class="o">.</span><span class="n">txmap</span>        <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># TXMAP entries, array of TXMap entries</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.UniGene.Record.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.Record.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.UniGene.Record.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.Record.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Represent the UniGene Record object as a string for debugging.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.UniGene.parse">
<span class="sig-prename descclassname"><span class="pre">Bio.UniGene.</span></span><span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.parse" title="Permalink to this definition"></a></dt>
<dd><p>Read and load a UniGene records, for files containing multiple records.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.UniGene.read">
<span class="sig-prename descclassname"><span class="pre">Bio.UniGene.</span></span><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.UniGene.read" title="Permalink to this definition"></a></dt>
<dd><p>Read and load a UniGene record, one record per file.</p>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.TogoWS.html" class="btn btn-neutral float-left" title="Bio.TogoWS package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.UniProt.html" class="btn btn-neutral float-right" title="Bio.UniProt package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>