<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.SeqIO.InsdcIO module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SeqIO.Interfaces module" href="Bio.SeqIO.Interfaces.html" />
    <link rel="prev" title="Bio.SeqIO.IgIO module" href="Bio.SeqIO.IgIO.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="index.html">
            
              <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SeqIO.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html#module-Bio.SeqIO">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
      <li class="breadcrumb-item active">Bio.SeqIO.InsdcIO module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SeqIO/InsdcIO.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.SeqIO.IgIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.IgIO module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqIO.Interfaces.html" class="btn btn-neutral float-right" title="Bio.SeqIO.Interfaces module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.SeqIO.InsdcIO">
<span id="bio-seqio-insdcio-module"></span><h1>Bio.SeqIO.InsdcIO module<a class="headerlink" href="#module-Bio.SeqIO.InsdcIO" title="Permalink to this headline"></a></h1>
<p>Bio.SeqIO support for the “genbank” and “embl” file formats.</p>
<p>You are expected to use this module via the Bio.SeqIO functions.
Note that internally this module calls Bio.GenBank to do the actual
parsing of GenBank, EMBL and IMGT files.</p>
<p>See Also:
International Nucleotide Sequence Database Collaboration
<a class="reference external" href="http://www.insdc.org/">http://www.insdc.org/</a></p>
<p>GenBank
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/Genbank/">http://www.ncbi.nlm.nih.gov/Genbank/</a></p>
<p>EMBL Nucleotide Sequence Database
<a class="reference external" href="http://www.ebi.ac.uk/embl/">http://www.ebi.ac.uk/embl/</a></p>
<p>DDBJ (DNA Data Bank of Japan)
<a class="reference external" href="http://www.ddbj.nig.ac.jp/">http://www.ddbj.nig.ac.jp/</a></p>
<p>IMGT (use a variant of EMBL format with longer feature indents)
<a class="reference external" href="http://imgt.cines.fr/download/LIGM-DB/userman_doc.html">http://imgt.cines.fr/download/LIGM-DB/userman_doc.html</a>
<a class="reference external" href="http://imgt.cines.fr/download/LIGM-DB/ftable_doc.html">http://imgt.cines.fr/download/LIGM-DB/ftable_doc.html</a>
<a class="reference external" href="http://www.ebi.ac.uk/imgt/hla/docs/manual.html">http://www.ebi.ac.uk/imgt/hla/docs/manual.html</a></p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.InsdcIO.</span></span><span class="sig-name descname"><span class="pre">GenBankIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceIterator" title="Bio.SeqIO.Interfaces.SequenceIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceIterator</span></code></a></p>
<p>Parser for GenBank files.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankIterator.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Break up a Genbank file into SeqRecord objects.</p>
<p>Argument source is a file-like object opened in text mode or a path to a file.
Every section from the LOCUS line to the terminating // becomes
a single SeqRecord with associated annotation and features.</p>
<p>Note that for genomes or chromosomes, there is typically only
one record.</p>
<p>This gets called internally by Bio.SeqIO for the GenBank file format:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;GenBank/cor6_6.gb&quot;</span><span class="p">,</span> <span class="s2">&quot;gb&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">X55053.1</span>
<span class="go">X62281.1</span>
<span class="go">M81224.1</span>
<span class="go">AJ237582.1</span>
<span class="go">L31939.1</span>
<span class="go">AF297471.1</span>
</pre></div>
</div>
<p>Equivalently,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;GenBank/cor6_6.gb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">GenBankIterator</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">X55053.1</span>
<span class="go">X62281.1</span>
<span class="go">M81224.1</span>
<span class="go">AJ237582.1</span>
<span class="go">L31939.1</span>
<span class="go">AF297471.1</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankIterator.parse">
<span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankIterator.parse" title="Permalink to this definition"></a></dt>
<dd><p>Start parsing the file, and return a SeqRecord generator.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankIterator.__parameters__">
<span class="sig-name descname"><span class="pre">__parameters__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankIterator.__parameters__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.InsdcIO.</span></span><span class="sig-name descname"><span class="pre">EmblIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceIterator" title="Bio.SeqIO.Interfaces.SequenceIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceIterator</span></code></a></p>
<p>Parser for EMBL files.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblIterator.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Break up an EMBL file into SeqRecord objects.</p>
<p>Argument source is a file-like object opened in text mode or a path to a file.
Every section from the LOCUS line to the terminating // becomes
a single SeqRecord with associated annotation and features.</p>
<p>Note that for genomes or chromosomes, there is typically only
one record.</p>
<p>This gets called internally by Bio.SeqIO for the EMBL file format:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;EMBL/epo_prt_selection.embl&quot;</span><span class="p">,</span> <span class="s2">&quot;embl&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">A00022.1</span>
<span class="go">A00028.1</span>
<span class="go">A00031.1</span>
<span class="go">A00034.1</span>
<span class="go">A00060.1</span>
<span class="go">A00071.1</span>
<span class="go">A00072.1</span>
<span class="go">A00078.1</span>
<span class="go">CQ797900.1</span>
</pre></div>
</div>
<p>Equivalently,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;EMBL/epo_prt_selection.embl&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">EmblIterator</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">A00022.1</span>
<span class="go">A00028.1</span>
<span class="go">A00031.1</span>
<span class="go">A00034.1</span>
<span class="go">A00060.1</span>
<span class="go">A00071.1</span>
<span class="go">A00072.1</span>
<span class="go">A00078.1</span>
<span class="go">CQ797900.1</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblIterator.parse">
<span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblIterator.parse" title="Permalink to this definition"></a></dt>
<dd><p>Start parsing the file, and return a SeqRecord generator.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblIterator.__parameters__">
<span class="sig-name descname"><span class="pre">__parameters__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblIterator.__parameters__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.InsdcIO.</span></span><span class="sig-name descname"><span class="pre">ImgtIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceIterator" title="Bio.SeqIO.Interfaces.SequenceIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceIterator</span></code></a></p>
<p>Parser for IMGT files.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtIterator.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Break up an IMGT file into SeqRecord objects.</p>
<p>Argument source is a file-like object opened in text mode or a path to a file.
Every section from the LOCUS line to the terminating // becomes
a single SeqRecord with associated annotation and features.</p>
<p>Note that for genomes or chromosomes, there is typically only
one record.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtIterator.parse">
<span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtIterator.parse" title="Permalink to this definition"></a></dt>
<dd><p>Start parsing the file, and return a SeqRecord generator.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtIterator.__parameters__">
<span class="sig-name descname"><span class="pre">__parameters__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtIterator.__parameters__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.InsdcIO.</span></span><span class="sig-name descname"><span class="pre">GenBankCdsFeatureIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceIterator" title="Bio.SeqIO.Interfaces.SequenceIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceIterator</span></code></a></p>
<p>Parser for GenBank files, creating a SeqRecord for each CDS feature.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Break up a Genbank file into SeqRecord objects for each CDS feature.</p>
<p>Argument source is a file-like object opened in text mode or a path to a file.</p>
<p>Every section from the LOCUS line to the terminating // can contain
many CDS features.  These are returned as with the stated amino acid
translation sequence (if given).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator.parse">
<span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator.parse" title="Permalink to this definition"></a></dt>
<dd><p>Start parsing the file, and return a SeqRecord generator.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator.__parameters__">
<span class="sig-name descname"><span class="pre">__parameters__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankCdsFeatureIterator.__parameters__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.InsdcIO.</span></span><span class="sig-name descname"><span class="pre">EmblCdsFeatureIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqIO.Interfaces.html#Bio.SeqIO.Interfaces.SequenceIterator" title="Bio.SeqIO.Interfaces.SequenceIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceIterator</span></code></a></p>
<p>Parser for EMBL files, creating a SeqRecord for each CDS feature.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Break up a EMBL file into SeqRecord objects for each CDS feature.</p>
<p>Argument source is a file-like object opened in text mode or a path to a file.</p>
<p>Every section from the LOCUS line to the terminating // can contain
many CDS features.  These are returned as with the stated amino acid
translation sequence (if given).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator.parse">
<span class="sig-name descname"><span class="pre">parse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator.parse" title="Permalink to this definition"></a></dt>
<dd><p>Start parsing the file, and return a SeqRecord generator.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator.__parameters__">
<span class="sig-name descname"><span class="pre">__parameters__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblCdsFeatureIterator.__parameters__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.InsdcIO.</span></span><span class="sig-name descname"><span class="pre">GenBankWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">IO</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">os.PathLike</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'w'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.InsdcIO._InsdcWriter</span></code></p>
<p>GenBank writer.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter.HEADER_WIDTH">
<span class="sig-name descname"><span class="pre">HEADER_WIDTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">12</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter.HEADER_WIDTH" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter.QUALIFIER_INDENT">
<span class="sig-name descname"><span class="pre">QUALIFIER_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">21</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter.QUALIFIER_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter.STRUCTURED_COMMENT_START">
<span class="sig-name descname"><span class="pre">STRUCTURED_COMMENT_START</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'-START##'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter.STRUCTURED_COMMENT_START" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter.STRUCTURED_COMMENT_END">
<span class="sig-name descname"><span class="pre">STRUCTURED_COMMENT_END</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'-END##'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter.STRUCTURED_COMMENT_END" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter.STRUCTURED_COMMENT_DELIM">
<span class="sig-name descname"><span class="pre">STRUCTURED_COMMENT_DELIM</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'</span> <span class="pre">::</span> <span class="pre">'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter.STRUCTURED_COMMENT_DELIM" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter.LETTERS_PER_LINE">
<span class="sig-name descname"><span class="pre">LETTERS_PER_LINE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">60</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter.LETTERS_PER_LINE" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter.SEQUENCE_INDENT">
<span class="sig-name descname"><span class="pre">SEQUENCE_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">9</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter.SEQUENCE_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.GenBankWriter.write_record">
<span class="sig-name descname"><span class="pre">write_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.GenBankWriter.write_record" title="Permalink to this definition"></a></dt>
<dd><p>Write a single record to the output file.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.InsdcIO.</span></span><span class="sig-name descname"><span class="pre">EmblWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">IO</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">os.PathLike</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'w'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.InsdcIO._InsdcWriter</span></code></p>
<p>EMBL writer.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.HEADER_WIDTH">
<span class="sig-name descname"><span class="pre">HEADER_WIDTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">5</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.HEADER_WIDTH" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.QUALIFIER_INDENT">
<span class="sig-name descname"><span class="pre">QUALIFIER_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">21</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.QUALIFIER_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.QUALIFIER_INDENT_STR">
<span class="sig-name descname"><span class="pre">QUALIFIER_INDENT_STR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'FT</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.QUALIFIER_INDENT_STR" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.QUALIFIER_INDENT_TMP">
<span class="sig-name descname"><span class="pre">QUALIFIER_INDENT_TMP</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'FT</span>&#160;&#160; <span class="pre">%s</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.QUALIFIER_INDENT_TMP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.FEATURE_HEADER">
<span class="sig-name descname"><span class="pre">FEATURE_HEADER</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'FH</span>&#160;&#160; <span class="pre">Key</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">Location/Qualifiers\nFH\n'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.FEATURE_HEADER" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.LETTERS_PER_BLOCK">
<span class="sig-name descname"><span class="pre">LETTERS_PER_BLOCK</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">10</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.LETTERS_PER_BLOCK" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.BLOCKS_PER_LINE">
<span class="sig-name descname"><span class="pre">BLOCKS_PER_LINE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">6</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.BLOCKS_PER_LINE" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.LETTERS_PER_LINE">
<span class="sig-name descname"><span class="pre">LETTERS_PER_LINE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">60</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.LETTERS_PER_LINE" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.POSITION_PADDING">
<span class="sig-name descname"><span class="pre">POSITION_PADDING</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">10</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.POSITION_PADDING" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.EmblWriter.write_record">
<span class="sig-name descname"><span class="pre">write_record</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">record</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.EmblWriter.write_record" title="Permalink to this definition"></a></dt>
<dd><p>Write a single record to the output file.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqIO.InsdcIO.</span></span><span class="sig-name descname"><span class="pre">ImgtWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">IO</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">os.PathLike</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">bytes</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'w'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtWriter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.SeqIO.InsdcIO.EmblWriter" title="Bio.SeqIO.InsdcIO.EmblWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.InsdcIO.EmblWriter</span></code></a></p>
<p>IMGT writer (EMBL format variant).</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtWriter.HEADER_WIDTH">
<span class="sig-name descname"><span class="pre">HEADER_WIDTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">5</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtWriter.HEADER_WIDTH" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtWriter.QUALIFIER_INDENT">
<span class="sig-name descname"><span class="pre">QUALIFIER_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">25</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtWriter.QUALIFIER_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtWriter.QUALIFIER_INDENT_STR">
<span class="sig-name descname"><span class="pre">QUALIFIER_INDENT_STR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'FT</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtWriter.QUALIFIER_INDENT_STR" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtWriter.QUALIFIER_INDENT_TMP">
<span class="sig-name descname"><span class="pre">QUALIFIER_INDENT_TMP</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'FT</span>&#160;&#160; <span class="pre">%s</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtWriter.QUALIFIER_INDENT_TMP" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.SeqIO.InsdcIO.ImgtWriter.FEATURE_HEADER">
<span class="sig-name descname"><span class="pre">FEATURE_HEADER</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'FH</span>&#160;&#160; <span class="pre">Key</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">Location/Qualifiers\nFH\n'</span></em><a class="headerlink" href="#Bio.SeqIO.InsdcIO.ImgtWriter.FEATURE_HEADER" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.SeqIO.IgIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.IgIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqIO.Interfaces.html" class="btn btn-neutral float-right" title="Bio.SeqIO.Interfaces module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2023, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.81/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>