<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.GenBank.Record module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.GenBank.Scanner module" href="Bio.GenBank.Scanner.html" />
    <link rel="prev" title="Bio.GenBank package" href="Bio.GenBank.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.GenBank.html">Bio.GenBank package</a></li>
      <li class="breadcrumb-item active">Bio.GenBank.Record module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/GenBank/Record.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.GenBank.html" class="btn btn-neutral float-left" title="Bio.GenBank package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.GenBank.Scanner.html" class="btn btn-neutral float-right" title="Bio.GenBank.Scanner module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.GenBank.Record">
<span id="bio-genbank-record-module"></span><h1>Bio.GenBank.Record module<a class="headerlink" href="#module-Bio.GenBank.Record" title="Permalink to this headline"></a></h1>
<p>Hold GenBank data in a straightforward format.</p>
<dl class="simple">
<dt>Classes:</dt><dd><ul class="simple">
<li><p>Record - All of the information in a GenBank record.</p></li>
<li><p>Reference - hold reference data for a record.</p></li>
<li><p>Feature - Hold the information in a Feature Table.</p></li>
<li><p>Qualifier - Qualifiers on a Feature.</p></li>
</ul>
</dd>
</dl>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.GenBank.Record.</span></span><span class="sig-name descname"><span class="pre">Record</span></span><a class="headerlink" href="#Bio.GenBank.Record.Record" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Hold GenBank information in a format similar to the original record.</p>
<p>The Record class is meant to make data easy to get to when you are
just interested in looking at GenBank data.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>locus - The name specified after the LOCUS keyword in the GenBank
record. This may be the accession number, or a clone id or something else.</p></li>
<li><p>size - The size of the record.</p></li>
<li><p>residue_type - The type of residues making up the sequence in this
record. Normally something like RNA, DNA or PROTEIN, but may be as
esoteric as ‘ss-RNA circular’.</p></li>
<li><p>data_file_division - The division this record is stored under in
GenBank (ie. PLN -&gt; plants; PRI -&gt; humans, primates; BCT -&gt; bacteria…)</p></li>
<li><p>date - The date of submission of the record, in a form like ‘28-JUL-1998’</p></li>
<li><p>accession - list of all accession numbers for the sequence.</p></li>
<li><p>nid - Nucleotide identifier number.</p></li>
<li><p>pid - Proteint identifier number</p></li>
<li><p>version - The accession number + version (ie. AB01234.2)</p></li>
<li><p>db_source - Information about the database the record came from</p></li>
<li><p>gi - The NCBI gi identifier for the record.</p></li>
<li><p>keywords - A list of keywords related to the record.</p></li>
<li><p>segment - If the record is one of a series, this is info about which
segment this record is (something like ‘1 of 6’).</p></li>
<li><p>source - The source of material where the sequence came from.</p></li>
<li><p>organism - The genus and species of the organism (ie. ‘Homo sapiens’)</p></li>
<li><p>taxonomy - A listing of the taxonomic classification of the organism,
starting general and getting more specific.</p></li>
<li><p>references - A list of Reference objects.</p></li>
<li><p>comment - Text with any kind of comment about the record.</p></li>
<li><p>features - A listing of Features making up the feature table.</p></li>
<li><p>base_counts - A string with the counts of bases for the sequence.</p></li>
<li><p>origin - A string specifying info about the origin of the sequence.</p></li>
<li><p>sequence - A string with the sequence itself.</p></li>
<li><p>contig - A string of location information for a CONTIG in a RefSeq file</p></li>
<li><p>project - The genome sequencing project numbers
(will be replaced by the dblink cross-references in 2009).</p></li>
<li><p>dblinks - The genome sequencing project number(s) and other links.
(will replace the project information in 2009).</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.GB_LINE_LENGTH">
<span class="sig-name descname"><span class="pre">GB_LINE_LENGTH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">79</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.GB_LINE_LENGTH" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.GB_BASE_INDENT">
<span class="sig-name descname"><span class="pre">GB_BASE_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">12</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.GB_BASE_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.GB_FEATURE_INDENT">
<span class="sig-name descname"><span class="pre">GB_FEATURE_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">21</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.GB_FEATURE_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.GB_INTERNAL_INDENT">
<span class="sig-name descname"><span class="pre">GB_INTERNAL_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">2</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.GB_INTERNAL_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.GB_OTHER_INTERNAL_INDENT">
<span class="sig-name descname"><span class="pre">GB_OTHER_INTERNAL_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">3</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.GB_OTHER_INTERNAL_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.GB_FEATURE_INTERNAL_INDENT">
<span class="sig-name descname"><span class="pre">GB_FEATURE_INTERNAL_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">5</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.GB_FEATURE_INTERNAL_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.GB_SEQUENCE_INDENT">
<span class="sig-name descname"><span class="pre">GB_SEQUENCE_INDENT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">9</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.GB_SEQUENCE_INDENT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.BASE_FORMAT">
<span class="sig-name descname"><span class="pre">BASE_FORMAT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'%-12s'</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.BASE_FORMAT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.INTERNAL_FORMAT">
<span class="sig-name descname"><span class="pre">INTERNAL_FORMAT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'</span>&#160; <span class="pre">%-10s'</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.INTERNAL_FORMAT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.OTHER_INTERNAL_FORMAT">
<span class="sig-name descname"><span class="pre">OTHER_INTERNAL_FORMAT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'</span>&#160;&#160; <span class="pre">%-9s'</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.OTHER_INTERNAL_FORMAT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.BASE_FEATURE_FORMAT">
<span class="sig-name descname"><span class="pre">BASE_FEATURE_FORMAT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'%-21s'</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.BASE_FEATURE_FORMAT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.INTERNAL_FEATURE_FORMAT">
<span class="sig-name descname"><span class="pre">INTERNAL_FEATURE_FORMAT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'</span>&#160;&#160;&#160;&#160; <span class="pre">%-16s'</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.INTERNAL_FEATURE_FORMAT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.SEQUENCE_FORMAT">
<span class="sig-name descname"><span class="pre">SEQUENCE_FORMAT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'%9s'</span></em><a class="headerlink" href="#Bio.GenBank.Record.Record.SEQUENCE_FORMAT" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Record.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Record.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Record.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Provide a GenBank formatted output option for a Record.</p>
<p>The objective of this is to provide an easy way to read in a GenBank
record, modify it somehow, and then output it in ‘GenBank format.’
We are striving to make this work so that a parsed Record that is
output using this function will look exactly like the original
record.</p>
<p>Much of the output is based on format description info at:</p>
<p><a class="reference external" href="ftp://ncbi.nlm.nih.gov/genbank/gbrel.txt">ftp://ncbi.nlm.nih.gov/genbank/gbrel.txt</a></p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Reference">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.GenBank.Record.</span></span><span class="sig-name descname"><span class="pre">Reference</span></span><a class="headerlink" href="#Bio.GenBank.Record.Reference" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Hold information from a GenBank reference.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>number - The number of the reference in the listing of references.</p></li>
<li><p>bases - The bases in the sequence the reference refers to.</p></li>
<li><p>authors - String with all of the authors.</p></li>
<li><p>consrtm - Consortium the authors belong to.</p></li>
<li><p>title - The title of the reference.</p></li>
<li><p>journal - Information about the journal where the reference appeared.</p></li>
<li><p>medline_id - The medline id for the reference.</p></li>
<li><p>pubmed_id - The pubmed_id for the reference.</p></li>
<li><p>remark - Free-form remarks about the reference.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Reference.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Reference.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Reference.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Reference.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Convert the reference to a GenBank format string.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Feature">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.GenBank.Record.</span></span><span class="sig-name descname"><span class="pre">Feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Feature" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Hold information about a Feature in the Feature Table of GenBank record.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>key - The key name of the feature (ie. source)</p></li>
<li><p>location - The string specifying the location of the feature.</p></li>
<li><p>qualifiers - A list of Qualifier objects in the feature.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Feature.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Feature.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Feature.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Feature.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Representation of the object for debugging or logging.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Feature.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Feature.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Return feature as a GenBank format string.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Qualifier">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.GenBank.Record.</span></span><span class="sig-name descname"><span class="pre">Qualifier</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Qualifier" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Hold information about a qualifier in a GenBank feature.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>key - The key name of the qualifier (ie. /organism=)</p></li>
<li><p>value - The value of the qualifier (“Dictyostelium discoideum”).</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Qualifier.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Qualifier.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Qualifier.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Qualifier.__repr__" title="Permalink to this definition"></a></dt>
<dd><p>Representation of the object for debugging or logging.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.GenBank.Record.Qualifier.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.GenBank.Record.Qualifier.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Return feature qualifier as a GenBank format string.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.GenBank.html" class="btn btn-neutral float-left" title="Bio.GenBank package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.GenBank.Scanner.html" class="btn btn-neutral float-right" title="Bio.GenBank.Scanner module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>