<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.Align.interfaces module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.Align.maf module" href="Bio.Align.maf.html" />
    <link rel="prev" title="Bio.Align.hhr module" href="Bio.Align.hhr.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.Align.html">Bio.Align package</a></li>
      <li class="breadcrumb-item active">Bio.Align.interfaces module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Align/interfaces.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Align.hhr.html" class="btn btn-neutral float-left" title="Bio.Align.hhr module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Align.maf.html" class="btn btn-neutral float-right" title="Bio.Align.maf module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.Align.interfaces">
<span id="bio-align-interfaces-module"></span><h1>Bio.Align.interfaces module<a class="headerlink" href="#module-Bio.Align.interfaces" title="Permalink to this headline"></a></h1>
<p>Bio.Align support module (not for general use).</p>
<p>Unless you are writing a new parser or writer for Bio.Align, you should not
use this module.  It provides base classes to try and simplify things.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.interfaces.</span></span><span class="sig-name descname"><span class="pre">AlignmentIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Align.html#Bio.Align.AlignmentsAbstractBaseClass" title="Bio.Align.AlignmentsAbstractBaseClass"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Align.AlignmentsAbstractBaseClass</span></code></a></p>
<p>Base class for building Alignment iterators.</p>
<p>You should write a parse method that returns an Alignment generator.  You
may wish to redefine the __init__ method as well.</p>
<p>Subclasses may define the following class attributes:
- mode   - ‘t’ or ‘b’ for text or binary files, respectively
- fmt    - a human-readable name for the file format.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.mode">
<span class="sig-name descname"><span class="pre">mode</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'t'</span></em><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.mode" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.fmt">
<span class="sig-name descname"><span class="pre">fmt</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.fmt" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Create an AlignmentIterator object.</p>
<p>Arguments:
- source - input file stream, or path to input file</p>
<p>This method MAY be overridden by any subclass.</p>
<p>Note when subclassing:
- there should be a single non-optional argument, the source.
- you can add additional optional arguments.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.__next__">
<span class="sig-name descname"><span class="pre">__next__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.__next__" title="Permalink to this definition"></a></dt>
<dd><p>Return the next alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.__len__" title="Permalink to this definition"></a></dt>
<dd><p>Return the number of alignments.</p>
<p>The number of alignments is cached. If not yet calculated, the iterator
is rewound to the beginning, and the number of alignments is calculated
by iterating over the alignments. The iterator is then returned to its
original position in the file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.__enter__">
<span class="sig-name descname"><span class="pre">__enter__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.__enter__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.__exit__">
<span class="sig-name descname"><span class="pre">__exit__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exc_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exc_value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exc_traceback</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.__exit__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.__getitem__" title="Permalink to this definition"></a></dt>
<dd><p>Return the alignments as an Alignments object (which inherits from list).</p>
<p>Only an index of the form [:] (i.e., a full slice) is supported.
The file stream is returned to its zero position, and the file header
is read and stored in an Alignments object. Next, we iterate over the
alignments and store them in the Alignments object.  The iterator
is then returned to its original position in the file, and the
Alignments object is returned. The Alignments object contains the exact
same information as the Alignment iterator self, but stores the
alignments in a list instead of as an iterator, allowing indexing.</p>
<p>Typical usage is</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Blat/dna_rna.psl&quot;</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;psLayout version&#39;: &#39;3&#39;}</span>
</pre></div>
</div>
<p>As <cite>alignments</cite> is an iterator and not a list, we cannot retrieve an
alignment by its index:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>  
<span class="gt">Traceback (most recent call last):</span>
<span class="w">  </span><span class="c">...</span>
<span class="gr">KeyError</span>: <span class="n">&#39;only [:] (a full slice) can be used as the index&#39;</span>
</pre></div>
</div>
<p>So we use the iterator to create a list-like Alignments object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[:]</span>
</pre></div>
</div>
<p>While <cite>alignments</cite> is a list-like object, it has the same <cite>metadata</cite>
attribute representing the information stored in the file header:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;psLayout version&#39;: &#39;3&#39;}</span>
</pre></div>
</div>
<p>Now we can index individual alignments:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">id</span>
<span class="go">(&#39;chr3&#39;, &#39;NR_111921.1&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.rewind">
<span class="sig-name descname"><span class="pre">rewind</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.rewind" title="Permalink to this definition"></a></dt>
<dd><p>Rewind the iterator to let it loop over the alignments from the beginning.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({'_read_next_alignment'})</span></em><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentIterator.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'fmt':</span> <span class="pre">typing.Optional[str]}</span></em><a class="headerlink" href="#Bio.Align.interfaces.AlignmentIterator.__annotations__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.interfaces.</span></span><span class="sig-name descname"><span class="pre">AlignmentWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">abc.ABC</span></code></p>
<p>Base class for alignment writers. This class should be subclassed.</p>
<p>It is intended for alignment file formats with an (optional)
header, one or more alignments, and an (optional) footer.</p>
<p>The user may call the write_file() method to write a complete
file containing the alignments.</p>
<p>Alternatively, users may call the write_header(), followed
by multiple calls to format_alignment() and/or write_alignments(),
followed finally by write_footer().</p>
<p>Subclasses may define the following class attributes:
- mode   - ‘w’ or ‘wb’ for text or binary files, respectively
- fmt    - a human-readable name for the file format.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.mode">
<span class="sig-name descname"><span class="pre">mode</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'w'</span></em><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.mode" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.fmt">
<span class="sig-name descname"><span class="pre">fmt</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.fmt" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Create the writer object.</p>
<p>Arguments:
- target - output file stream, or path to output file</p>
<p>This method MAY be overridden by any subclass.</p>
<p>Note when subclassing:
- there should be a single non-optional argument, the target.
- you can add additional optional arguments.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.write_header">
<span class="sig-name descname"><span class="pre">write_header</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.write_header" title="Permalink to this definition"></a></dt>
<dd><p>Write the file header to the output file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.write_footer">
<span class="sig-name descname"><span class="pre">write_footer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.write_footer" title="Permalink to this definition"></a></dt>
<dd><p>Write the file footer to the output file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.format_alignment">
<span class="sig-name descname"><span class="pre">format_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.format_alignment" title="Permalink to this definition"></a></dt>
<dd><p>Format a single alignment as a string.</p>
<p>alignment - an Alignment object</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.write_single_alignment">
<span class="sig-name descname"><span class="pre">write_single_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.write_single_alignment" title="Permalink to this definition"></a></dt>
<dd><p>Write a single alignment to the output file, and return 1.</p>
<p>alignments - A list or iterator returning Alignment objects
stream     - Output file stream.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.write_multiple_alignments">
<span class="sig-name descname"><span class="pre">write_multiple_alignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.write_multiple_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Write alignments to the output file, and return the number of alignments.</p>
<p>alignments - A list or iterator returning Alignment objects
stream     - Output file stream.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.write_alignments">
<span class="sig-name descname"><span class="pre">write_alignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.write_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Write alignments to the output file, and return the number of alignments.</p>
<p>alignments - A list or iterator returning Alignment objects
stream     - Output file stream.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.write_file">
<span class="sig-name descname"><span class="pre">write_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.write_file" title="Permalink to this definition"></a></dt>
<dd><p>Write the alignments to the file strenm, and return the number of alignments.</p>
<p>alignments - A list or iterator returning Alignment objects
stream     - Output file stream.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.write" title="Permalink to this definition"></a></dt>
<dd><p>Write a file with the alignments, and return the number of alignments.</p>
<p>alignments - A list or iterator returning Alignment objects</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.interfaces.AlignmentWriter.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'fmt':</span> <span class="pre">typing.Optional[str]}</span></em><a class="headerlink" href="#Bio.Align.interfaces.AlignmentWriter.__annotations__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Align.hhr.html" class="btn btn-neutral float-left" title="Bio.Align.hhr module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Align.maf.html" class="btn btn-neutral float-right" title="Bio.Align.maf module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>