<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.SCOP package &mdash; Biopython 1.81.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SCOP.Cla module" href="Bio.SCOP.Cla.html" />
    <link rel="prev" title="Bio.Restriction.Restriction_Dictionary module" href="Bio.Restriction.Restriction_Dictionary.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
            <a href="index.html">
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.SCOP package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-Bio.SCOP">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
      <li class="breadcrumb-item active">Bio.SCOP package</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SCOP/__init__.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Restriction.Restriction_Dictionary.html" class="btn btn-neutral float-left" title="Bio.Restriction.Restriction_Dictionary module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SCOP.Cla.html" class="btn btn-neutral float-right" title="Bio.SCOP.Cla module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bio-scop-package">
<h1>Bio.SCOP package<a class="headerlink" href="#bio-scop-package" title="Permalink to this headline"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.SCOP.Cla.html">Bio.SCOP.Cla module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SCOP.Des.html">Bio.SCOP.Des module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SCOP.Dom.html">Bio.SCOP.Dom module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SCOP.Hie.html">Bio.SCOP.Hie module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SCOP.Raf.html">Bio.SCOP.Raf module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SCOP.Residues.html">Bio.SCOP.Residues module</a></li>
</ul>
</div>
</section>
<section id="module-Bio.SCOP">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.SCOP" title="Permalink to this headline"></a></h2>
<p>SCOP: Structural Classification of Proteins.</p>
<p>The SCOP database aims to provide a manually constructed classification of
all know protein structures into a hierarchy, the main levels of which
are family, superfamily and fold.</p>
<ul class="simple">
<li><p>“SCOP”:<a class="reference external" href="http://scop.mrc-lmb.cam.ac.uk/legacy/">http://scop.mrc-lmb.cam.ac.uk/legacy/</a></p></li>
<li><p>“Introduction”:<a class="reference external" href="http://scop.mrc-lmb.cam.ac.uk/legacy/intro.html">http://scop.mrc-lmb.cam.ac.uk/legacy/intro.html</a></p></li>
<li><p>“SCOP parsable files”:<a class="reference external" href="http://scop.mrc-lmb.cam.ac.uk/legacy/parse/">http://scop.mrc-lmb.cam.ac.uk/legacy/parse/</a></p></li>
</ul>
<p>The Scop object in this module represents the entire SCOP classification. It
can be built from the three SCOP parsable files, modified is so desired, and
converted back to the same file formats. A single SCOP domain (represented
by the Domain class) can be obtained from Scop using the domain’s SCOP
identifier (sid).</p>
<ul class="simple">
<li><dl class="simple">
<dt>nodeCodeDict – A mapping between known 2 letter node codes and a longer</dt><dd><p>description. The known node types are ‘cl’ (class), ‘cf’
(fold), ‘sf’ (superfamily), ‘fa’ (family), ‘dm’ (domain),
‘sp’ (species), ‘px’ (domain). Additional node types may
be added in the future.</p>
</dd>
</dl>
</li>
</ul>
<p>This module also provides code to access SCOP over the WWW.</p>
<dl class="simple">
<dt>Functions:</dt><dd><ul class="simple">
<li><p>search        – Access the main CGI script.</p></li>
<li><p>_open         – Internally used function.</p></li>
</ul>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="Bio.SCOP.cmp_sccs">
<span class="sig-prename descclassname"><span class="pre">Bio.SCOP.</span></span><span class="sig-name descname"><span class="pre">cmp_sccs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sccs1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sccs2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.cmp_sccs" title="Permalink to this definition"></a></dt>
<dd><p>Order SCOP concise classification strings (sccs).</p>
<p>a.4.5.1 &lt; a.4.5.11 &lt; b.1.1.1</p>
<p>A sccs (e.g. a.4.5.11) compactly represents a domain’s classification.
The letter represents the class, and the numbers are the fold,
superfamily, and family, respectively.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SCOP.parse_domain">
<span class="sig-prename descclassname"><span class="pre">Bio.SCOP.</span></span><span class="sig-name descname"><span class="pre">parse_domain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">term</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.parse_domain" title="Permalink to this definition"></a></dt>
<dd><p>Convert an ASTRAL header string into a Scop domain.</p>
<p>An ASTRAL (<a class="reference external" href="http://astral.stanford.edu/">http://astral.stanford.edu/</a>) header contains a concise
description of a SCOP domain. A very similar format is used when a
Domain object is converted into a string.  The Domain returned by this
method contains most of the SCOP information, but it will not be located
within the SCOP hierarchy (i.e. The parent node will be None). The
description is composed of the SCOP protein and species descriptions.</p>
<p>A typical ASTRAL header looks like –
&gt;d1tpt_1 a.46.2.1 (1-70) Thymidine phosphorylase {Escherichia coli}</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SCOP.Scop">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SCOP.</span></span><span class="sig-name descname"><span class="pre">Scop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cla_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">des_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hie_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The entire SCOP hierarchy.</p>
<p>root – The root node of the hierarchy</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cla_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">des_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hie_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Build the SCOP hierarchy from the SCOP parsable files, or a sql backend.</p>
<p>If no file handles are given, then a Scop object with a single
empty root node is returned.</p>
<p>If a directory and version are given (with dir_path=.., version=…) or
file handles for each file, the whole scop tree will be built in memory.</p>
<p>If a MySQLdb database handle is given, the tree will be built as needed,
minimising construction times.  To build the SQL database to the methods
write_xxx_sql to create the tables.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.getRoot">
<span class="sig-name descname"><span class="pre">getRoot</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.getRoot" title="Permalink to this definition"></a></dt>
<dd><p>Get root node.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.getDomainBySid">
<span class="sig-name descname"><span class="pre">getDomainBySid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sid</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.getDomainBySid" title="Permalink to this definition"></a></dt>
<dd><p>Return a domain from its sid.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.getNodeBySunid">
<span class="sig-name descname"><span class="pre">getNodeBySunid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sunid</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.getNodeBySunid" title="Permalink to this definition"></a></dt>
<dd><p>Return a node from its sunid.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.getDomains">
<span class="sig-name descname"><span class="pre">getDomains</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.getDomains" title="Permalink to this definition"></a></dt>
<dd><p>Return an ordered tuple of all SCOP Domains.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.write_hie">
<span class="sig-name descname"><span class="pre">write_hie</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.write_hie" title="Permalink to this definition"></a></dt>
<dd><p>Build an HIE SCOP parsable file from this object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.write_des">
<span class="sig-name descname"><span class="pre">write_des</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.write_des" title="Permalink to this definition"></a></dt>
<dd><p>Build a DES SCOP parsable file from this object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.write_cla">
<span class="sig-name descname"><span class="pre">write_cla</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.write_cla" title="Permalink to this definition"></a></dt>
<dd><p>Build a CLA SCOP parsable file from this object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.getDomainFromSQL">
<span class="sig-name descname"><span class="pre">getDomainFromSQL</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sunid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.getDomainFromSQL" title="Permalink to this definition"></a></dt>
<dd><p>Load a node from the SQL backend using sunid or sid.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.getAscendentFromSQL">
<span class="sig-name descname"><span class="pre">getAscendentFromSQL</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.getAscendentFromSQL" title="Permalink to this definition"></a></dt>
<dd><p>Get ascendents using SQL backend.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.getDescendentsFromSQL">
<span class="sig-name descname"><span class="pre">getDescendentsFromSQL</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.getDescendentsFromSQL" title="Permalink to this definition"></a></dt>
<dd><p>Get descendents of a node using the database backend.</p>
<p>This avoids repeated iteration of SQL calls and is therefore much
quicker than repeatedly calling node.getChildren().</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.write_hie_sql">
<span class="sig-name descname"><span class="pre">write_hie_sql</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.write_hie_sql" title="Permalink to this definition"></a></dt>
<dd><p>Write HIE data to SQL database.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.write_cla_sql">
<span class="sig-name descname"><span class="pre">write_cla_sql</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.write_cla_sql" title="Permalink to this definition"></a></dt>
<dd><p>Write CLA data to SQL database.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Scop.write_des_sql">
<span class="sig-name descname"><span class="pre">write_des_sql</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Scop.write_des_sql" title="Permalink to this definition"></a></dt>
<dd><p>Write DES data to SQL database.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SCOP.Node">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SCOP.</span></span><span class="sig-name descname"><span class="pre">Node</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A node in the Scop hierarchy.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>sunid  – SCOP unique identifiers. e.g. ‘14986’</p></li>
<li><p>parent – The parent node</p></li>
<li><p>children – A list of child nodes</p></li>
<li><p>sccs     – SCOP concise classification string. e.g. ‘a.1.1.2’</p></li>
<li><p>type     – A 2 letter node type code. e.g. ‘px’ for domains</p></li>
<li><p>description – Description text.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Node.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize a Node in the scop hierarchy.</p>
<p>If a Scop instance is provided to the constructor, this will be used
to lookup related references using the SQL methods.  If no instance
is provided, it is assumed the whole tree exists and is connected.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Node.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Represent the node as a string.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Node.toHieRecord">
<span class="sig-name descname"><span class="pre">toHieRecord</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node.toHieRecord" title="Permalink to this definition"></a></dt>
<dd><p>Return an Hie.Record.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Node.toDesRecord">
<span class="sig-name descname"><span class="pre">toDesRecord</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node.toDesRecord" title="Permalink to this definition"></a></dt>
<dd><p>Return a Des.Record.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Node.getChildren">
<span class="sig-name descname"><span class="pre">getChildren</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node.getChildren" title="Permalink to this definition"></a></dt>
<dd><p>Return a list of children of this Node.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Node.getParent">
<span class="sig-name descname"><span class="pre">getParent</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node.getParent" title="Permalink to this definition"></a></dt>
<dd><p>Return the parent of this Node.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Node.getDescendents">
<span class="sig-name descname"><span class="pre">getDescendents</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node_type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node.getDescendents" title="Permalink to this definition"></a></dt>
<dd><p>Return a list of all descendant nodes of the given type.</p>
<p>Node type can be a two letter code or longer description,
e.g. ‘fa’ or ‘family’.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Node.getAscendent">
<span class="sig-name descname"><span class="pre">getAscendent</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node_type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Node.getAscendent" title="Permalink to this definition"></a></dt>
<dd><p>Return the ancenstor node of the given type, or None.</p>
<p>Node type can be a two letter code or longer description,
e.g. ‘fa’ or ‘family’.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SCOP.Domain">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SCOP.</span></span><span class="sig-name descname"><span class="pre">Domain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Domain" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.SCOP.Node" title="Bio.SCOP.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SCOP.Node</span></code></a></p>
<p>A SCOP domain. A leaf node in the Scop hierarchy.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>sid - The SCOP domain identifier. e.g. <code class="docutils literal notranslate"><span class="pre">&quot;d5hbib_&quot;</span></code></p></li>
<li><p>residues - A Residue object. It defines the collection of PDB
atoms that make up this domain.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Domain.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Domain.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize a SCOP Domain object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Domain.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Domain.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Represent the SCOP Domain as a string.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Domain.toDesRecord">
<span class="sig-name descname"><span class="pre">toDesRecord</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Domain.toDesRecord" title="Permalink to this definition"></a></dt>
<dd><p>Return a Des.Record.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Domain.toClaRecord">
<span class="sig-name descname"><span class="pre">toClaRecord</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Domain.toClaRecord" title="Permalink to this definition"></a></dt>
<dd><p>Return a Cla.Record.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.SCOP.Astral">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SCOP.</span></span><span class="sig-name descname"><span class="pre">Astral</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">astral_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Representation of the ASTRAL database.</p>
<p>Abstraction of the ASTRAL database, which has sequences for all the SCOP domains,
as well as clusterings by percent id or evalue.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dir_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">astral_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the astral database.</p>
<dl class="simple">
<dt>You must provide either a directory of SCOP files:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>dir_path - string, the path to location of the scopseq-x.xx directory</dt><dd><p>(not the directory itself), and</p>
</dd>
</dl>
</li>
<li><p>version   -a version number.</p></li>
</ul>
</dd>
<dt>or, a FASTA file:</dt><dd><ul class="simple">
<li><p>astral_file - string, a path to a fasta file (which will be loaded in memory)</p></li>
</ul>
</dd>
<dt>or, a MYSQL database:</dt><dd><ul class="simple">
<li><p>db_handle - a database handle for a MYSQL database containing a table
‘astral’ with the astral data in it.  This can be created
using writeToSQL.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.domainsClusteredByEv">
<span class="sig-name descname"><span class="pre">domainsClusteredByEv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.domainsClusteredByEv" title="Permalink to this definition"></a></dt>
<dd><p>Get domains clustered by evalue.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.domainsClusteredById">
<span class="sig-name descname"><span class="pre">domainsClusteredById</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.domainsClusteredById" title="Permalink to this definition"></a></dt>
<dd><p>Get domains clustered by percentage identity.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.getAstralDomainsFromFile">
<span class="sig-name descname"><span class="pre">getAstralDomainsFromFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_handle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.getAstralDomainsFromFile" title="Permalink to this definition"></a></dt>
<dd><p>Get the scop domains from a file containing a list of sids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.getAstralDomainsFromSQL">
<span class="sig-name descname"><span class="pre">getAstralDomainsFromSQL</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">column</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.getAstralDomainsFromSQL" title="Permalink to this definition"></a></dt>
<dd><p>Load ASTRAL domains from the MySQL database.</p>
<p>Load a set of astral domains from a column in the astral table of a MYSQL
database (which can be created with writeToSQL(…).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.getSeqBySid">
<span class="sig-name descname"><span class="pre">getSeqBySid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">domain</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.getSeqBySid" title="Permalink to this definition"></a></dt>
<dd><p>Get the seq record of a given domain from its sid.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.getSeq">
<span class="sig-name descname"><span class="pre">getSeq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">domain</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.getSeq" title="Permalink to this definition"></a></dt>
<dd><p>Return seq associated with domain.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.hashedDomainsById">
<span class="sig-name descname"><span class="pre">hashedDomainsById</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.hashedDomainsById" title="Permalink to this definition"></a></dt>
<dd><p>Get domains clustered by sequence identity in a dict.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.hashedDomainsByEv">
<span class="sig-name descname"><span class="pre">hashedDomainsByEv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.hashedDomainsByEv" title="Permalink to this definition"></a></dt>
<dd><p>Get domains clustered by evalue in a dict.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.isDomainInId">
<span class="sig-name descname"><span class="pre">isDomainInId</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.isDomainInId" title="Permalink to this definition"></a></dt>
<dd><p>Return true if the domain is in the astral clusters for percent ID.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.isDomainInEv">
<span class="sig-name descname"><span class="pre">isDomainInEv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dom</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.isDomainInEv" title="Permalink to this definition"></a></dt>
<dd><p>Return true if the domain is in the ASTRAL clusters for evalues.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SCOP.Astral.writeToSQL">
<span class="sig-name descname"><span class="pre">writeToSQL</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_handle</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.Astral.writeToSQL" title="Permalink to this definition"></a></dt>
<dd><p>Write the ASTRAL database to a MYSQL database.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.SCOP.search">
<span class="sig-prename descclassname"><span class="pre">Bio.SCOP.</span></span><span class="sig-name descname"><span class="pre">search</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cgi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'http://scop.mrc-lmb.cam.ac.uk/legacy/search.cgi'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">keywds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SCOP.search" title="Permalink to this definition"></a></dt>
<dd><p>Access SCOP search and return a handle to the results.</p>
<p>Access search.cgi and return a handle to the results.  See the
online help file for an explanation of the parameters:
<a class="reference external" href="http://scop.mrc-lmb.cam.ac.uk/legacy/help.html">http://scop.mrc-lmb.cam.ac.uk/legacy/help.html</a></p>
<p>Raises an IOError if there’s a network error.</p>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Restriction.Restriction_Dictionary.html" class="btn btn-neutral float-left" title="Bio.Restriction.Restriction_Dictionary module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SCOP.Cla.html" class="btn btn-neutral float-right" title="Bio.SCOP.Cla module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2020, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.81.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.79/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>