<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.SeqUtils.ProtParam module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.SeqUtils.ProtParamData module" href="Bio.SeqUtils.ProtParamData.html" />
    <link rel="prev" title="Bio.SeqUtils.MeltingTemp module" href="Bio.SeqUtils.MeltingTemp.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
      <li class="breadcrumb-item active">Bio.SeqUtils.ProtParam module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SeqUtils/ProtParam.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.SeqUtils.MeltingTemp.html" class="btn btn-neutral float-left" title="Bio.SeqUtils.MeltingTemp module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqUtils.ProtParamData.html" class="btn btn-neutral float-right" title="Bio.SeqUtils.ProtParamData module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.SeqUtils.ProtParam">
<span id="bio-sequtils-protparam-module"></span><h1>Bio.SeqUtils.ProtParam module<a class="headerlink" href="#module-Bio.SeqUtils.ProtParam" title="Permalink to this headline"></a></h1>
<p>Simple protein analysis.</p>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline"></a></h2>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils.ProtParam</span> <span class="kn">import</span> <span class="n">ProteinAnalysis</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">X</span> <span class="o">=</span> <span class="n">ProteinAnalysis</span><span class="p">(</span><span class="s2">&quot;MAEGEITTFTALTEKFNLPPGNYKKPKLLYCSNGGHFLRILPDGTVDGT&quot;</span>
<span class="gp">... </span>                    <span class="s2">&quot;RDRSDQHIQLQLSAESVGEVYIKSTETGQYLAMDTSGLLYGSQTPSEEC&quot;</span>
<span class="gp">... </span>                    <span class="s2">&quot;LFLERLEENHYNTYTSKKHAEKNWFVGLKKNGSCKRGPRTHYGQKAILF&quot;</span>
<span class="gp">... </span>                    <span class="s2">&quot;LPLPV&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">count_amino_acids</span><span class="p">()[</span><span class="s1">&#39;A&#39;</span><span class="p">])</span>
<span class="go">6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">count_amino_acids</span><span class="p">()[</span><span class="s1">&#39;E&#39;</span><span class="p">])</span>
<span class="go">12</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">X</span><span class="o">.</span><span class="n">get_amino_acids_percent</span><span class="p">()[</span><span class="s1">&#39;A&#39;</span><span class="p">])</span>
<span class="go">0.04</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">X</span><span class="o">.</span><span class="n">get_amino_acids_percent</span><span class="p">()[</span><span class="s1">&#39;L&#39;</span><span class="p">])</span>
<span class="go">0.12</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">X</span><span class="o">.</span><span class="n">molecular_weight</span><span class="p">())</span>
<span class="go">17103.16</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">X</span><span class="o">.</span><span class="n">aromaticity</span><span class="p">())</span>
<span class="go">0.10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">X</span><span class="o">.</span><span class="n">instability_index</span><span class="p">())</span>
<span class="go">41.98</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">X</span><span class="o">.</span><span class="n">isoelectric_point</span><span class="p">())</span>
<span class="go">7.72</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sec_struc</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">secondary_structure_fraction</span><span class="p">()</span>  <span class="c1"># [helix, turn, sheet]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">sec_struc</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>  <span class="c1"># helix</span>
<span class="go">0.33</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">sec_struc</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>  <span class="c1"># turn</span>
<span class="go">0.29</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%0.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">sec_struc</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>  <span class="c1"># sheet</span>
<span class="go">0.37</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">epsilon_prot</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">molar_extinction_coefficient</span><span class="p">()</span>  <span class="c1"># [reduced, oxidized]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">epsilon_prot</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>  <span class="c1"># with reduced cysteines</span>
<span class="go">17420</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">epsilon_prot</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>  <span class="c1"># with disulfid bridges</span>
<span class="go">17545</span>
</pre></div>
</div>
<dl class="simple">
<dt>Other public methods are:</dt><dd><ul class="simple">
<li><p>gravy</p></li>
<li><p>protein_scale</p></li>
<li><p>flexibility</p></li>
<li><p>charge_at_pH</p></li>
</ul>
</dd>
</dl>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.SeqUtils.ProtParam.</span></span><span class="sig-name descname"><span class="pre">ProteinAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">monoisotopic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class containing methods for protein analysis.</p>
<p>The constructor takes two arguments.
The first is the protein sequence as a string or a Seq object.</p>
<p>The second argument is optional. If set to True, the weight of the amino
acids will be calculated using their monoisotopic mass (the weight of the
most abundant isotopes for each element), instead of the average molecular
mass (the averaged weight of all stable isotopes for each element).
If set to false (the default value) or left out, the IUPAC average
molecular mass will be used for the calculation.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">monoisotopic</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.count_amino_acids">
<span class="sig-name descname"><span class="pre">count_amino_acids</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.count_amino_acids" title="Permalink to this definition"></a></dt>
<dd><p>Count standard amino acids, return a dict.</p>
<p>Counts the number times each amino acid is in the protein
sequence. Returns a dictionary {AminoAcid:Number}.</p>
<p>The return value is cached in self.amino_acids_content.
It is not recalculated upon subsequent calls.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.get_amino_acids_percent">
<span class="sig-name descname"><span class="pre">get_amino_acids_percent</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.get_amino_acids_percent" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the amino acid content in percentages.</p>
<p>The same as count_amino_acids only returns the Number in percentage of
entire sequence. Returns a dictionary of {AminoAcid:percentage}.</p>
<p>The return value is cached in self.amino_acids_percent.</p>
<p>input is the dictionary self.amino_acids_content.
output is a dictionary with amino acids as keys.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.molecular_weight">
<span class="sig-name descname"><span class="pre">molecular_weight</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.molecular_weight" title="Permalink to this definition"></a></dt>
<dd><p>Calculate MW from Protein sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.aromaticity">
<span class="sig-name descname"><span class="pre">aromaticity</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.aromaticity" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the aromaticity according to Lobry, 1994.</p>
<p>Calculates the aromaticity value of a protein according to Lobry, 1994.
It is simply the relative frequency of Phe+Trp+Tyr.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.instability_index">
<span class="sig-name descname"><span class="pre">instability_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.instability_index" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the instability index according to Guruprasad et al 1990.</p>
<p>Implementation of the method of Guruprasad et al. 1990 to test a
protein for stability. Any value above 40 means the protein is unstable
(has a short half life).</p>
<p>See: Guruprasad K., Reddy B.V.B., Pandit M.W.
Protein Engineering 4:155-161(1990).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.flexibility">
<span class="sig-name descname"><span class="pre">flexibility</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.flexibility" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the flexibility according to Vihinen, 1994.</p>
<p>No argument to change window size because parameters are specific for
a window=9. The parameters used are optimized for determining the
flexibility.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.gravy">
<span class="sig-name descname"><span class="pre">gravy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'KyteDoolitle'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.gravy" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the GRAVY (Grand Average of Hydropathy) according to Kyte and Doolitle, 1982.</p>
<p>Utilizes the given Hydrophobicity scale, by default uses the original
proposed by Kyte and Doolittle (KyteDoolitle). Other options are:
Aboderin, AbrahamLeo, Argos, BlackMould, BullBreese, Casari, Cid,
Cowan3.4, Cowan7.5, Eisenberg, Engelman, Fasman, Fauchere, GoldSack,
Guy, Jones, Juretic, Kidera, Miyazawa, Parker,Ponnuswamy, Rose,
Roseman, Sweet, Tanford, Wilson and Zimmerman.</p>
<p>New scales can be added in ProtParamData.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.protein_scale">
<span class="sig-name descname"><span class="pre">protein_scale</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">param_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.protein_scale" title="Permalink to this definition"></a></dt>
<dd><p>Compute a profile by any amino acid scale.</p>
<p>An amino acid scale is defined by a numerical value assigned to each
type of amino acid. The most frequently used scales are the
hydrophobicity or hydrophilicity scales and the secondary structure
conformational parameters scales, but many other scales exist which
are based on different chemical and physical properties of the
amino acids.  You can set several parameters that control the
computation of a scale profile, such as the window size and the window
edge relative weight value.</p>
<p>WindowSize: The window size is the length of the interval to use for
the profile computation. For a window size n, we use the i-(n-1)/2
neighboring residues on each side to compute the score for residue i.
The score for residue i is the sum of the scaled values for these
amino acids, optionally weighted according to their position in the
window.</p>
<p>Edge: The central amino acid of the window always has a weight of 1.
By default, the amino acids at the remaining window positions have the
same weight, but you can make the residue at the center of the window
have a larger weight than the others by setting the edge value for the
residues at the beginning and end of the interval to a value between
0 and 1. For instance, for Edge=0.4 and a window size of 5 the weights
will be: 0.4, 0.7, 1.0, 0.7, 0.4.</p>
<p>The method returns a list of values which can be plotted to view the
change along a protein sequence.  Many scales exist. Just add your
favorites to the ProtParamData modules.</p>
<p>Similar to expasy’s ProtScale:
<a class="reference external" href="http://www.expasy.org/cgi-bin/protscale.pl">http://www.expasy.org/cgi-bin/protscale.pl</a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.isoelectric_point">
<span class="sig-name descname"><span class="pre">isoelectric_point</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.isoelectric_point" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the isoelectric point.</p>
<p>Uses the module IsoelectricPoint to calculate the pI of a protein.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.charge_at_pH">
<span class="sig-name descname"><span class="pre">charge_at_pH</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pH</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.charge_at_pH" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the charge of a protein at given pH.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.secondary_structure_fraction">
<span class="sig-name descname"><span class="pre">secondary_structure_fraction</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.secondary_structure_fraction" title="Permalink to this definition"></a></dt>
<dd><p>Calculate fraction of helix, turn and sheet.</p>
<p>Returns a list of the fraction of amino acids which tend
to be in Helix, Turn or Sheet, according to Haimov and Srebnik, 2016;
Hutchinson and Thornton, 1994; and Kim and Berg, 1993, respectively.</p>
<p>Amino acids in helix: E, M, A, L, K.
Amino acids in turn: N, P, G, S, D.
Amino acids in sheet: V, I, Y, F, W, L, T.</p>
<p>Note that, prior to v1.82, this method wrongly returned
(Sheet, Turn, Helix) while claiming to return (Helix, Turn, Sheet).</p>
<p>Returns a tuple of three floats (Helix, Turn, Sheet).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.SeqUtils.ProtParam.ProteinAnalysis.molar_extinction_coefficient">
<span class="sig-name descname"><span class="pre">molar_extinction_coefficient</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqUtils.ProtParam.ProteinAnalysis.molar_extinction_coefficient" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the molar extinction coefficient.</p>
<p>Calculates the molar extinction coefficient assuming cysteines
(reduced) and cystines residues (Cys-Cys-bond)</p>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.SeqUtils.MeltingTemp.html" class="btn btn-neutral float-left" title="Bio.SeqUtils.MeltingTemp module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.SeqUtils.ProtParamData.html" class="btn btn-neutral float-right" title="Bio.SeqUtils.ProtParamData module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>