<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.motifs.matrix module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.motifs.meme module" href="Bio.motifs.meme.html" />
    <link rel="prev" title="Bio.motifs.mast module" href="Bio.motifs.mast.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Welcome to Biopython’s API documentation!</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Welcome to Biopython’s API documentation!</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.motifs.html">Bio.motifs package</a></li>
      <li class="breadcrumb-item active">Bio.motifs.matrix module</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/api/Bio.motifs.matrix.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.motifs.mast.html" class="btn btn-neutral float-left" title="Bio.motifs.mast module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.motifs.meme.html" class="btn btn-neutral float-right" title="Bio.motifs.meme module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.motifs.matrix">
<span id="bio-motifs-matrix-module"></span><h1>Bio.motifs.matrix module<a class="headerlink" href="#module-Bio.motifs.matrix" title="Permalink to this headline"></a></h1>
<p>Support for various forms of sequence motif matrices.</p>
<p>Implementation of frequency (count) matrices, position-weight matrices,
and position-specific scoring matrices.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.motifs.matrix.</span></span><span class="sig-name descname"><span class="pre">GenericPositionMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
<p>Base class for the support of position matrix operations.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.__str__" title="Permalink to this definition"></a></dt>
<dd><p>Return a string containing nucleotides and counts of the alphabet in the Matrix.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.__getitem__" title="Permalink to this definition"></a></dt>
<dd><p>Return the position matrix of index key.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.consensus">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">consensus</span></span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.consensus" title="Permalink to this definition"></a></dt>
<dd><p>Return the consensus sequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.anticonsensus">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">anticonsensus</span></span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.anticonsensus" title="Permalink to this definition"></a></dt>
<dd><p>Return the anticonsensus sequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.degenerate_consensus">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">degenerate_consensus</span></span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.degenerate_consensus" title="Permalink to this definition"></a></dt>
<dd><p>Return the degenerate consensus sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.calculate_consensus">
<span class="sig-name descname"><span class="pre">calculate_consensus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">substitution_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plurality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">setcase</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.calculate_consensus" title="Permalink to this definition"></a></dt>
<dd><p>Return the consensus sequence (as a string) for the given parameters.</p>
<p>This function largely follows the conventions of the EMBOSS <cite>cons</cite> tool.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>substitution_matrix - the scoring matrix used when comparing
sequences. By default, it is None, in which case we simply count the
frequency of each letter.
Instead of the default value, you can use the substitution matrices
available in Bio.Align.substitution_matrices. Common choices are
BLOSUM62 (also known as EBLOSUM62) for protein, and NUC.4.4 (also
known as EDNAFULL) for nucleotides. NOTE: This has not yet been
implemented.</p></li>
<li><p>plurality           - threshold value for the number of positive
matches, divided by the total count in a column, required to reach
consensus. If substitution_matrix is None, then this argument must
be None, and is ignored; a ValueError is raised otherwise. If
substitution_matrix is not None, then the default value of the
plurality is 0.5.</p></li>
<li><p>identity            - number of identities, divided by the total
count in a column, required to define a consensus value. If the
number of identities is less than identity * total count in a column,
then the undefined character (‘N’ for nucleotides and ‘X’ for amino
acid sequences) is used in the consensus sequence. If identity is
1.0, then only columns of identical letters contribute to the
consensus. Default value is zero.</p></li>
<li><p>setcase             - threshold for the positive matches, divided by
the total count in a column, above which the consensus is is
upper-case and below which the consensus is in lower-case. By
default, this is equal to 0.5.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.gc_content">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gc_content</span></span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.gc_content" title="Permalink to this definition"></a></dt>
<dd><p>Compute the fraction GC content.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.GenericPositionMatrix.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.GenericPositionMatrix.reverse_complement" title="Permalink to this definition"></a></dt>
<dd><p>Compute reverse complement.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.motifs.matrix.FrequencyPositionMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.motifs.matrix.</span></span><span class="sig-name descname"><span class="pre">FrequencyPositionMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.FrequencyPositionMatrix" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.motifs.matrix.GenericPositionMatrix" title="Bio.motifs.matrix.GenericPositionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.motifs.matrix.GenericPositionMatrix</span></code></a></p>
<p>Class for the support of frequency calculations on the Position Matrix.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.FrequencyPositionMatrix.normalize">
<span class="sig-name descname"><span class="pre">normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pseudocounts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.FrequencyPositionMatrix.normalize" title="Permalink to this definition"></a></dt>
<dd><p>Create and return a position-weight matrix by normalizing the counts matrix.</p>
<p>If pseudocounts is None (default), no pseudocounts are added
to the counts.</p>
<p>If pseudocounts is a number, it is added to the counts before
calculating the position-weight matrix.</p>
<p>Alternatively, the pseudocounts can be a dictionary with a key
for each letter in the alphabet associated with the motif.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionWeightMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.motifs.matrix.</span></span><span class="sig-name descname"><span class="pre">PositionWeightMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionWeightMatrix" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.motifs.matrix.GenericPositionMatrix" title="Bio.motifs.matrix.GenericPositionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.motifs.matrix.GenericPositionMatrix</span></code></a></p>
<p>Class for the support of weight calculations on the Position Matrix.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionWeightMatrix.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionWeightMatrix.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionWeightMatrix.log_odds">
<span class="sig-name descname"><span class="pre">log_odds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionWeightMatrix.log_odds" title="Permalink to this definition"></a></dt>
<dd><p>Return the Position-Specific Scoring Matrix.</p>
<p>The Position-Specific Scoring Matrix (PSSM) contains the log-odds
scores computed from the probability matrix and the background
probabilities. If the background is None, a uniform background
distribution is assumed.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.motifs.matrix.</span></span><span class="sig-name descname"><span class="pre">PositionSpecificScoringMatrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.motifs.matrix.GenericPositionMatrix" title="Bio.motifs.matrix.GenericPositionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.motifs.matrix.GenericPositionMatrix</span></code></a></p>
<p>Class for the support of Position Specific Scoring Matrix calculations.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.calculate">
<span class="sig-name descname"><span class="pre">calculate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.calculate" title="Permalink to this definition"></a></dt>
<dd><p>Return the PWM score for a given sequence for all positions.</p>
<dl class="simple">
<dt>Notes:</dt><dd><ul class="simple">
<li><p>the sequence can only be a DNA sequence</p></li>
<li><p>the search is performed only on one strand</p></li>
<li><p>if the sequence and the motif have the same length, a single
number is returned</p></li>
<li><p>otherwise, the result is a one-dimensional numpy array</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.search">
<span class="sig-name descname"><span class="pre">search</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">both</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunksize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span> <span class="pre">**</span> <span class="pre">6</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.search" title="Permalink to this definition"></a></dt>
<dd><p>Find hits with PWM score above given threshold.</p>
<p>A generator function, returning found hits in the given sequence
with the pwm score higher than the threshold.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.max">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max</span></span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.max" title="Permalink to this definition"></a></dt>
<dd><p>Maximal possible score for this motif.</p>
<p>returns the score computed for the consensus sequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.min">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">min</span></span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.min" title="Permalink to this definition"></a></dt>
<dd><p>Minimal possible score for this motif.</p>
<p>returns the score computed for the anticonsensus sequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.gc_content">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gc_content</span></span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.gc_content" title="Permalink to this definition"></a></dt>
<dd><p>Compute the GC-ratio.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.mean">
<span class="sig-name descname"><span class="pre">mean</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.mean" title="Permalink to this definition"></a></dt>
<dd><p>Return expected value of the score of a motif.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.std">
<span class="sig-name descname"><span class="pre">std</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.std" title="Permalink to this definition"></a></dt>
<dd><p>Return standard deviation of the score of a motif.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.dist_pearson">
<span class="sig-name descname"><span class="pre">dist_pearson</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.dist_pearson" title="Permalink to this definition"></a></dt>
<dd><p>Return the similarity score based on pearson correlation for the given motif against self.</p>
<p>We use the Pearson’s correlation of the respective probabilities.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.dist_pearson_at">
<span class="sig-name descname"><span class="pre">dist_pearson_at</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.dist_pearson_at" title="Permalink to this definition"></a></dt>
<dd><p>Return the similarity score based on pearson correlation at the given offset.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.matrix.PositionSpecificScoringMatrix.distribution">
<span class="sig-name descname"><span class="pre">distribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span> <span class="pre">**</span> <span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.matrix.PositionSpecificScoringMatrix.distribution" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the distribution of the scores at the given precision.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.motifs.mast.html" class="btn btn-neutral float-left" title="Bio.motifs.mast module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.motifs.meme.html" class="btn btn-neutral float-right" title="Bio.motifs.meme module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.81/">Previous</a></dd>
        
          <dd><a href="../../latest/">Latest</a></dd>
        
          <dd><a href="../../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>