

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Phylo.TreeConstruction module &mdash; Biopython 1.78.dev0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.PopGen package" href="Bio.PopGen.html" />
    <link rel="prev" title="Bio.Phylo.PhyloXMLIO module" href="Bio.Phylo.PhyloXMLIO.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.78.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html#subpackages">Subpackages</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Phylo.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html#module-Bio.Phylo">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.Phylo.html">Bio.Phylo package</a> &raquo;</li>
        
      <li>Bio.Phylo.TreeConstruction module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Phylo/TreeConstruction.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.PopGen.html" class="btn btn-neutral float-right" title="Bio.PopGen package" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Phylo.PhyloXMLIO.html" class="btn btn-neutral float-left" title="Bio.Phylo.PhyloXMLIO module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.Phylo.TreeConstruction">
<span id="bio-phylo-treeconstruction-module"></span><h1>Bio.Phylo.TreeConstruction module<a class="headerlink" href="#module-Bio.Phylo.TreeConstruction" title="Permalink to this headline">¶</a></h1>
<p>Classes and methods for tree construction.</p>
<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.DistanceMatrix">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">DistanceMatrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">names</span></em>, <em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.TreeConstruction._Matrix</span></code></p>
<p>Distance matrix class that can be used for distance based tree algorithms.</p>
<p>All diagonal elements will be zero no matter what the users provide.</p>
<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceMatrix.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">names</span></em>, <em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceMatrix.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceMatrix.__setitem__">
<code class="sig-name descname">__setitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">item</span></em>, <em class="sig-param"><span class="n">value</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceMatrix.__setitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Set Matrix’s items to values.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceMatrix.format_phylip">
<code class="sig-name descname">format_phylip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">handle</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceMatrix.format_phylip" title="Permalink to this definition">¶</a></dt>
<dd><p>Write data in Phylip format to a given file-like object or handle.</p>
<p>The output stream is the input distance matrix format used with Phylip
programs (e.g. ‘neighbor’). See:
<a class="reference external" href="http://evolution.genetics.washington.edu/phylip/doc/neighbor.html">http://evolution.genetics.washington.edu/phylip/doc/neighbor.html</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>handle<span class="classifier">file or file-like object</span></dt><dd><p>A writeable text mode file handle or other object supporting
the ‘write’ method, such as StringIO or sys.stdout.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.DistanceCalculator">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">DistanceCalculator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">'identity'</span></em>, <em class="sig-param"><span class="n">skip_letters</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceCalculator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to calculate the distance matrix from a DNA or Protein.</p>
<p>Multiple Sequence Alignment(MSA) and the given name of the
substitution model.</p>
<p>Currently only scoring matrices are used.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>model<span class="classifier">str</span></dt><dd><p>Name of the model matrix to be used to calculate distance.
The attribute <code class="docutils literal notranslate"><span class="pre">dna_models</span></code> contains the available model
names for DNA sequences and <code class="docutils literal notranslate"><span class="pre">protein_models</span></code> for protein
sequences.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Loading a small PHYLIP alignment from which to compute distances:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Phylo.TreeConstruction</span> <span class="kn">import</span> <span class="n">DistanceCalculator</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;TreeConstruction/msa.phy&#39;</span><span class="p">),</span> <span class="s1">&#39;phylip&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Alignment</span> <span class="k">with</span> <span class="mi">5</span> <span class="n">rows</span> <span class="ow">and</span> <span class="mi">13</span> <span class="n">columns</span>
<span class="n">AACGTGGCCACAT</span> <span class="n">Alpha</span>
<span class="n">AAGGTCGCCACAC</span> <span class="n">Beta</span>
<span class="n">CAGTTCGCCACAA</span> <span class="n">Gamma</span>
<span class="n">GAGATTTCCGCCT</span> <span class="n">Delta</span>
<span class="n">GAGATCTCCGCCC</span> <span class="n">Epsilon</span>
</pre></div>
</div>
<p>DNA calculator with ‘identity’ model:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">calculator</span> <span class="o">=</span> <span class="n">DistanceCalculator</span><span class="p">(</span><span class="s1">&#39;identity&#39;</span><span class="p">)</span>
<span class="n">dm</span> <span class="o">=</span> <span class="n">calculator</span><span class="o">.</span><span class="n">get_distance</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Alpha</span>   <span class="mi">0</span>
<span class="n">Beta</span>    <span class="mf">0.23076923076923073</span>     <span class="mi">0</span>
<span class="n">Gamma</span>   <span class="mf">0.3846153846153846</span>      <span class="mf">0.23076923076923073</span>     <span class="mi">0</span>
<span class="n">Delta</span>   <span class="mf">0.5384615384615384</span>      <span class="mf">0.5384615384615384</span>      <span class="mf">0.5384615384615384</span>      <span class="mi">0</span>
<span class="n">Epsilon</span> <span class="mf">0.6153846153846154</span>      <span class="mf">0.3846153846153846</span>      <span class="mf">0.46153846153846156</span>     <span class="mf">0.15384615384615385</span>     <span class="mi">0</span>
    <span class="n">Alpha</span>       <span class="n">Beta</span>    <span class="n">Gamma</span>   <span class="n">Delta</span>   <span class="n">Epsilon</span>
</pre></div>
</div>
<p>Protein calculator with ‘blosum62’ model:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">calculator</span> <span class="o">=</span> <span class="n">DistanceCalculator</span><span class="p">(</span><span class="s1">&#39;blosum62&#39;</span><span class="p">)</span>
<span class="n">dm</span> <span class="o">=</span> <span class="n">calculator</span><span class="o">.</span><span class="n">get_distance</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Alpha</span>   <span class="mi">0</span>
<span class="n">Beta</span>    <span class="mf">0.36904761904761907</span>     <span class="mi">0</span>
<span class="n">Gamma</span>   <span class="mf">0.49397590361445787</span>     <span class="mf">0.25</span>    <span class="mi">0</span>
<span class="n">Delta</span>   <span class="mf">0.5853658536585367</span>      <span class="mf">0.5476190476190477</span>      <span class="mf">0.5662650602409638</span>      <span class="mi">0</span>
<span class="n">Epsilon</span> <span class="mf">0.7</span>     <span class="mf">0.3555555555555555</span>      <span class="mf">0.48888888888888893</span>     <span class="mf">0.2222222222222222</span>      <span class="mi">0</span>
    <span class="n">Alpha</span>       <span class="n">Beta</span>    <span class="n">Gamma</span>   <span class="n">Delta</span>   <span class="n">Epsilon</span>
</pre></div>
</div>
<dl class="py attribute">
<dt id="Bio.Phylo.TreeConstruction.DistanceCalculator.dna_models">
<code class="sig-name descname">dna_models</code><em class="property"> = ['benner22', 'benner6', 'benner74', 'dayhoff', 'feng', 'genetic', 'gonnet1992', 'hoxd70', 'johnson', 'jones', 'levin', 'mclachlan', 'mdm78', 'blastn', 'rao', 'risler', 'schneider', 'str', 'trans']</em><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceCalculator.dna_models" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="Bio.Phylo.TreeConstruction.DistanceCalculator.protein_models">
<code class="sig-name descname">protein_models</code><em class="property"> = ['blosum45', 'blosum50', 'blosum62', 'blosum80', 'blosum90', 'pam250', 'pam30', 'pam70']</em><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceCalculator.protein_models" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="Bio.Phylo.TreeConstruction.DistanceCalculator.models">
<code class="sig-name descname">models</code><em class="property"> = ['identity', 'benner22', 'benner6', 'benner74', 'dayhoff', 'feng', 'genetic', 'gonnet1992', 'hoxd70', 'johnson', 'jones', 'levin', 'mclachlan', 'mdm78', 'blastn', 'rao', 'risler', 'schneider', 'str', 'trans', 'blosum45', 'blosum50', 'blosum62', 'blosum80', 'blosum90', 'pam250', 'pam30', 'pam70']</em><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceCalculator.models" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceCalculator.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">'identity'</span></em>, <em class="sig-param"><span class="n">skip_letters</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceCalculator.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize with a distance model.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceCalculator.get_distance">
<code class="sig-name descname">get_distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">msa</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceCalculator.get_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a DistanceMatrix for MSA object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>msa<span class="classifier">MultipleSeqAlignment</span></dt><dd><p>DNA or Protein multiple sequence alignment.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.TreeConstructor">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">TreeConstructor</code><a class="headerlink" href="#Bio.Phylo.TreeConstruction.TreeConstructor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for all tree constructor.</p>
<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.TreeConstructor.build_tree">
<code class="sig-name descname">build_tree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">msa</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.TreeConstructor.build_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Caller to built the tree from a MultipleSeqAlignment object.</p>
<p>This should be implemented in subclass.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.DistanceTreeConstructor">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">DistanceTreeConstructor</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">distance_calculator</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">'nj'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceTreeConstructor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Phylo.TreeConstruction.TreeConstructor" title="Bio.Phylo.TreeConstruction.TreeConstructor"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.TreeConstruction.TreeConstructor</span></code></a></p>
<p>Distance based tree constructor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>method<span class="classifier">str</span></dt><dd><p>Distance tree construction method, ‘nj’(default) or ‘upgma’.</p>
</dd>
<dt>distance_calculator<span class="classifier">DistanceCalculator</span></dt><dd><p>The distance matrix calculator for multiple sequence alignment.
It must be provided if <code class="docutils literal notranslate"><span class="pre">build_tree</span></code> will be called.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Loading a small PHYLIP alignment from which to compute distances, and then
build a upgma Tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Phylo.TreeConstruction</span> <span class="kn">import</span> <span class="n">DistanceTreeConstructor</span>
<span class="kn">from</span> <span class="nn">Bio.Phylo.TreeConstruction</span> <span class="kn">import</span> <span class="n">DistanceCalculator</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;TreeConstruction/msa.phy&#39;</span><span class="p">),</span> <span class="s1">&#39;phylip&#39;</span><span class="p">)</span>
<span class="n">constructor</span> <span class="o">=</span> <span class="n">DistanceTreeConstructor</span><span class="p">()</span>
<span class="n">calculator</span> <span class="o">=</span> <span class="n">DistanceCalculator</span><span class="p">(</span><span class="s1">&#39;identity&#39;</span><span class="p">)</span>
<span class="n">dm</span> <span class="o">=</span> <span class="n">calculator</span><span class="o">.</span><span class="n">get_distance</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="n">upgmatree</span> <span class="o">=</span> <span class="n">constructor</span><span class="o">.</span><span class="n">upgma</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">upgmatree</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span><span class="p">(</span><span class="n">rooted</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner4&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.18749999999999994</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner1&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.07692307692307693</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Epsilon&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.07692307692307693</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Delta&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.11057692307692304</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner3&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.038461538461538464</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner2&#39;</span><span class="p">)</span>
                <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.11538461538461536</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Gamma&#39;</span><span class="p">)</span>
                <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.11538461538461536</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Beta&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.15384615384615383</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Alpha&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Build a NJ Tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">njtree</span> <span class="o">=</span> <span class="n">constructor</span><span class="o">.</span><span class="n">nj</span><span class="p">(</span><span class="n">dm</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">njtree</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span><span class="p">(</span><span class="n">rooted</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner3&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.18269230769230765</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Alpha&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.04807692307692307</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Beta&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.04807692307692307</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner2&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.27884615384615385</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner1&#39;</span><span class="p">)</span>
                <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.051282051282051266</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Epsilon&#39;</span><span class="p">)</span>
                <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.10256410256410259</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Delta&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.14423076923076922</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Gamma&#39;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py attribute">
<dt id="Bio.Phylo.TreeConstruction.DistanceTreeConstructor.methods">
<code class="sig-name descname">methods</code><em class="property"> = ['nj', 'upgma']</em><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceTreeConstructor.methods" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceTreeConstructor.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">distance_calculator</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">method</span><span class="o">=</span><span class="default_value">'nj'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceTreeConstructor.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceTreeConstructor.build_tree">
<code class="sig-name descname">build_tree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">msa</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceTreeConstructor.build_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct and return a Tree, Neighbor Joining or UPGMA.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceTreeConstructor.upgma">
<code class="sig-name descname">upgma</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">distance_matrix</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceTreeConstructor.upgma" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct and return an UPGMA tree.</p>
<p>Constructs and returns an Unweighted Pair Group Method
with Arithmetic mean (UPGMA) tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>distance_matrix<span class="classifier">DistanceMatrix</span></dt><dd><p>The distance matrix for tree construction.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.DistanceTreeConstructor.nj">
<code class="sig-name descname">nj</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">distance_matrix</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.DistanceTreeConstructor.nj" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct and return a Neighbor Joining tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>distance_matrix<span class="classifier">DistanceMatrix</span></dt><dd><p>The distance matrix for tree construction.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.Scorer">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">Scorer</code><a class="headerlink" href="#Bio.Phylo.TreeConstruction.Scorer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for all tree scoring methods.</p>
<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.Scorer.get_score">
<code class="sig-name descname">get_score</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">tree</span></em>, <em class="sig-param"><span class="n">alignment</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.Scorer.get_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Caller to get the score of a tree for the given alignment.</p>
<p>This should be implemented in subclass.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.TreeSearcher">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">TreeSearcher</code><a class="headerlink" href="#Bio.Phylo.TreeConstruction.TreeSearcher" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for all tree searching methods.</p>
<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.TreeSearcher.search">
<code class="sig-name descname">search</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">starting_tree</span></em>, <em class="sig-param"><span class="n">alignment</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.TreeSearcher.search" title="Permalink to this definition">¶</a></dt>
<dd><p>Caller to search the best tree with a starting tree.</p>
<p>This should be implemented in subclass.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.NNITreeSearcher">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">NNITreeSearcher</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scorer</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.NNITreeSearcher" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Phylo.TreeConstruction.TreeSearcher" title="Bio.Phylo.TreeConstruction.TreeSearcher"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.TreeConstruction.TreeSearcher</span></code></a></p>
<p>Tree searching with Nearest Neighbor Interchanges (NNI) algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>scorer<span class="classifier">ParsimonyScorer</span></dt><dd><p>parsimony scorer to calculate the parsimony score of
different trees during NNI algorithm.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.NNITreeSearcher.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">scorer</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.NNITreeSearcher.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.NNITreeSearcher.search">
<code class="sig-name descname">search</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">starting_tree</span></em>, <em class="sig-param"><span class="n">alignment</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.NNITreeSearcher.search" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the TreeSearcher.search method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>starting_tree<span class="classifier">Tree</span></dt><dd><p>starting tree of NNI method.</p>
</dd>
<dt>alignment<span class="classifier">MultipleSeqAlignment</span></dt><dd><p>multiple sequence alignment used to calculate parsimony
score of different NNI trees.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.ParsimonyScorer">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">ParsimonyScorer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.ParsimonyScorer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Phylo.TreeConstruction.Scorer" title="Bio.Phylo.TreeConstruction.Scorer"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.TreeConstruction.Scorer</span></code></a></p>
<p>Parsimony scorer with a scoring matrix.</p>
<p>This is a combination of Fitch algorithm and Sankoff algorithm.
See ParsimonyTreeConstructor for usage.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>matrix<span class="classifier">_Matrix</span></dt><dd><p>scoring matrix used in parsimony score calculation.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.ParsimonyScorer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.ParsimonyScorer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.ParsimonyScorer.get_score">
<code class="sig-name descname">get_score</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">tree</span></em>, <em class="sig-param"><span class="n">alignment</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.ParsimonyScorer.get_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate parsimony score using the Fitch algorithm.</p>
<p>Calculate and return the parsimony score given a tree and the
MSA using either the Fitch algorithm (without a penalty matrix)
or the Sankoff algorithm (with a matrix).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Phylo.TreeConstruction.ParsimonyTreeConstructor">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Phylo.TreeConstruction.</code><code class="sig-name descname">ParsimonyTreeConstructor</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">searcher</span></em>, <em class="sig-param"><span class="n">starting_tree</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.ParsimonyTreeConstructor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Phylo.TreeConstruction.TreeConstructor" title="Bio.Phylo.TreeConstruction.TreeConstructor"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Phylo.TreeConstruction.TreeConstructor</span></code></a></p>
<p>Parsimony tree constructor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>searcher<span class="classifier">TreeSearcher</span></dt><dd><p>tree searcher to search the best parsimony tree.</p>
</dd>
<dt>starting_tree<span class="classifier">Tree</span></dt><dd><p>starting tree provided to the searcher.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>We will load an alignment, and then load various trees which have already been computed from it:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span><span class="p">,</span> <span class="n">Phylo</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;TreeConstruction/msa.phy&#39;</span><span class="p">),</span> <span class="s1">&#39;phylip&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Alignment</span> <span class="k">with</span> <span class="mi">5</span> <span class="n">rows</span> <span class="ow">and</span> <span class="mi">13</span> <span class="n">columns</span>
<span class="n">AACGTGGCCACAT</span> <span class="n">Alpha</span>
<span class="n">AAGGTCGCCACAC</span> <span class="n">Beta</span>
<span class="n">CAGTTCGCCACAA</span> <span class="n">Gamma</span>
<span class="n">GAGATTTCCGCCT</span> <span class="n">Delta</span>
<span class="n">GAGATCTCCGCCC</span> <span class="n">Epsilon</span>
</pre></div>
</div>
<p>Load a starting tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">starting_tree</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;TreeConstruction/nj.tre&#39;</span><span class="p">,</span> <span class="s1">&#39;newick&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">starting_tree</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span><span class="p">(</span><span class="n">rooted</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
    <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner3&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.01421</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner2&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.23927</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner1&#39;</span><span class="p">)</span>
                <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.08531</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Epsilon&#39;</span><span class="p">)</span>
                <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.13691</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Delta&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.2923</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Alpha&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.07477</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Beta&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.17523</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Gamma&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Build the Parsimony tree from the starting tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">scorer</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">TreeConstruction</span><span class="o">.</span><span class="n">ParsimonyScorer</span><span class="p">()</span>
<span class="n">searcher</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">TreeConstruction</span><span class="o">.</span><span class="n">NNITreeSearcher</span><span class="p">(</span><span class="n">scorer</span><span class="p">)</span>
<span class="n">constructor</span> <span class="o">=</span> <span class="n">Phylo</span><span class="o">.</span><span class="n">TreeConstruction</span><span class="o">.</span><span class="n">ParsimonyTreeConstructor</span><span class="p">(</span><span class="n">searcher</span><span class="p">,</span> <span class="n">starting_tree</span><span class="p">)</span>
<span class="n">pars_tree</span> <span class="o">=</span> <span class="n">constructor</span><span class="o">.</span><span class="n">build_tree</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pars_tree</span><span class="p">)</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tree</span><span class="p">(</span><span class="n">rooted</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
    <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.0</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.19732999999999998</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner1&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.13691</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Delta&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.08531</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Epsilon&#39;</span><span class="p">)</span>
        <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.04194000000000003</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner2&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.01421</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Inner3&#39;</span><span class="p">)</span>
                <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.17523</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Gamma&#39;</span><span class="p">)</span>
                <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.07477</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Beta&#39;</span><span class="p">)</span>
            <span class="n">Clade</span><span class="p">(</span><span class="n">branch_length</span><span class="o">=</span><span class="mf">0.2923</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Alpha&#39;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.ParsimonyTreeConstructor.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">searcher</span></em>, <em class="sig-param"><span class="n">starting_tree</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.ParsimonyTreeConstructor.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Phylo.TreeConstruction.ParsimonyTreeConstructor.build_tree">
<code class="sig-name descname">build_tree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">alignment</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Phylo.TreeConstruction.ParsimonyTreeConstructor.build_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Build the tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt>alignment<span class="classifier">MultipleSeqAlignment</span></dt><dd><p>multiple sequence alignment to calculate parsimony tree.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.PopGen.html" class="btn btn-neutral float-right" title="Bio.PopGen package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Phylo.PhyloXMLIO.html" class="btn btn-neutral float-left" title="Bio.Phylo.PhyloXMLIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>