<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.PDB.PDBList module &mdash; Biopython 1.81.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.PDB.PDBParser module" href="Bio.PDB.PDBParser.html" />
    <link rel="prev" title="Bio.PDB.PDBIO module" href="Bio.PDB.PDBIO.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
            <a href="index.html">
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html#subpackages">Subpackages</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.PDB.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html#module-Bio.PDB">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.PDB.html">Bio.PDB package</a></li>
      <li class="breadcrumb-item active">Bio.PDB.PDBList module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/PDB/PDBList.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.PDB.PDBIO.html" class="btn btn-neutral float-left" title="Bio.PDB.PDBIO module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PDB.PDBParser.html" class="btn btn-neutral float-right" title="Bio.PDB.PDBParser module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.PDB.PDBList">
<span id="bio-pdb-pdblist-module"></span><h1>Bio.PDB.PDBList module<a class="headerlink" href="#module-Bio.PDB.PDBList" title="Permalink to this headline"></a></h1>
<p>Access the PDB over the internet (e.g. to download structures).</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.PDB.PDBList.</span></span><span class="sig-name descname"><span class="pre">PDBList</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">server</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ftp://ftp.wwpdb.org'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsolete_pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Quick access to the structure lists on the PDB or its mirrors.</p>
<p>This class provides quick access to the structure lists on the
PDB server or its mirrors. The structure lists contain
four-letter PDB codes, indicating that structures are
new, have been modified or are obsolete. The lists are released
on a weekly basis.</p>
<p>It also provides a function to retrieve PDB files from the server.
To use it properly, prepare a directory /pdb or the like,
where PDB files are stored.</p>
<p>All available file formats (PDB, PDBx/mmCif, PDBML, mmtf) are supported.
Please note that large structures (containing &gt;62 chains
and/or 99999 ATOM lines) are no longer stored as a single PDB file
and by default (when PDB format selected) are not downloaded.</p>
<p>Large structures can be downloaded in other formats, including PDBx/mmCif
or as a .tar file (a collection of PDB-like formatted files for a given
structure).</p>
<p>If you want to use this module from inside a proxy, add
the proxy variable to your environment, e.g. in Unix:
export HTTP_PROXY=’<a class="reference external" href="http://realproxy.charite.de:888">http://realproxy.charite.de:888</a>’
(This can also be added to ~/.bashrc)</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.PDB_REF">
<span class="sig-name descname"><span class="pre">PDB_REF</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'\n</span>&#160;&#160;&#160; <span class="pre">The</span> <span class="pre">Protein</span> <span class="pre">Data</span> <span class="pre">Bank:</span> <span class="pre">a</span> <span class="pre">computer-based</span> <span class="pre">archival</span> <span class="pre">file</span> <span class="pre">for</span> <span class="pre">macromolecular</span> <span class="pre">structures.\n</span>&#160;&#160;&#160; <span class="pre">F.C.Bernstein,</span> <span class="pre">T.F.Koetzle,</span> <span class="pre">G.J.B.Williams,</span> <span class="pre">E.F.Meyer</span> <span class="pre">Jr,</span> <span class="pre">M.D.Brice,</span> <span class="pre">J.R.Rodgers,</span> <span class="pre">O.Kennard,</span> <span class="pre">T.Shimanouchi,</span> <span class="pre">M.Tasumi\n</span>&#160;&#160;&#160; <span class="pre">J.</span> <span class="pre">Mol.</span> <span class="pre">Biol.</span> <span class="pre">112</span> <span class="pre">pp.</span> <span class="pre">535-542</span> <span class="pre">(1977)\n</span>&#160;&#160;&#160; <span class="pre">http://www.pdb.org/.\n</span>&#160;&#160;&#160; <span class="pre">'</span></em><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.PDB_REF" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">server</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ftp://ftp.wwpdb.org'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsolete_pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class with the default server or a custom one.</p>
<p>Argument pdb is the local path to use, defaulting to the current
directory at the moment of initialisation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.get_status_list">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_status_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.get_status_list" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve a list of pdb codes in the weekly pdb status file from given URL.</p>
<p>Used by get_recent_changes. Typical contents of the list files parsed
by this method is now very simply - one PDB name per line.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.get_recent_changes">
<span class="sig-name descname"><span class="pre">get_recent_changes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.get_recent_changes" title="Permalink to this definition"></a></dt>
<dd><p>Return three lists of the newest weekly files (added,mod,obsolete).</p>
<p>Reads the directories with changed entries from the PDB server and
returns a tuple of three URL’s to the files of new, modified and
obsolete entries from the most recent list. The directory with the
largest numerical name is used.
Returns None if something goes wrong.</p>
<p>Contents of the data/status dir (20031013 would be used);:</p>
<blockquote>
<div><p>drwxrwxr-x   2 1002     sysadmin     512 Oct  6 18:28 20031006
drwxrwxr-x   2 1002     sysadmin     512 Oct 14 02:14 20031013
-rw-r–r–   1 1002     sysadmin    1327 Mar 12  2001 README</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.get_all_entries">
<span class="sig-name descname"><span class="pre">get_all_entries</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.get_all_entries" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve the big file containing all the PDB entries and some annotation.</p>
<p>Returns a list of PDB codes in the index file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.get_all_obsolete">
<span class="sig-name descname"><span class="pre">get_all_obsolete</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.get_all_obsolete" title="Permalink to this definition"></a></dt>
<dd><p>Return a list of all obsolete entries ever in the PDB.</p>
<p>Returns a list of all obsolete pdb codes that have ever been
in the PDB.</p>
<p>Gets and parses the file from the PDB server in the format
(the first pdb_code column is the one used). The file looks
like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">LIST</span> <span class="n">OF</span> <span class="n">OBSOLETE</span> <span class="n">COORDINATE</span> <span class="n">ENTRIES</span> <span class="n">AND</span> <span class="n">SUCCESSORS</span>
<span class="n">OBSLTE</span>    <span class="mi">31</span><span class="o">-</span><span class="n">JUL</span><span class="o">-</span><span class="mi">94</span> <span class="mi">116</span><span class="n">L</span>     <span class="mi">216</span><span class="n">L</span>
<span class="o">...</span>
<span class="n">OBSLTE</span>    <span class="mi">29</span><span class="o">-</span><span class="n">JAN</span><span class="o">-</span><span class="mi">96</span> <span class="mi">1</span><span class="n">HFT</span>     <span class="mi">2</span><span class="n">HFT</span>
<span class="n">OBSLTE</span>    <span class="mi">21</span><span class="o">-</span><span class="n">SEP</span><span class="o">-</span><span class="mi">06</span> <span class="mi">1</span><span class="n">HFV</span>     <span class="mi">2</span><span class="n">J5X</span>
<span class="n">OBSLTE</span>    <span class="mi">21</span><span class="o">-</span><span class="n">NOV</span><span class="o">-</span><span class="mi">03</span> <span class="mi">1</span><span class="n">HG6</span>
<span class="n">OBSLTE</span>    <span class="mi">18</span><span class="o">-</span><span class="n">JUL</span><span class="o">-</span><span class="mi">84</span> <span class="mi">1</span><span class="n">HHB</span>     <span class="mi">2</span><span class="n">HHB</span> <span class="mi">3</span><span class="n">HHB</span>
<span class="n">OBSLTE</span>    <span class="mi">08</span><span class="o">-</span><span class="n">NOV</span><span class="o">-</span><span class="mi">96</span> <span class="mi">1</span><span class="n">HID</span>     <span class="mi">2</span><span class="n">HID</span>
<span class="n">OBSLTE</span>    <span class="mi">01</span><span class="o">-</span><span class="n">APR</span><span class="o">-</span><span class="mi">97</span> <span class="mi">1</span><span class="n">HIU</span>     <span class="mi">2</span><span class="n">HIU</span>
<span class="n">OBSLTE</span>    <span class="mi">14</span><span class="o">-</span><span class="n">JAN</span><span class="o">-</span><span class="mi">04</span> <span class="mi">1</span><span class="n">HKE</span>     <span class="mi">1</span><span class="n">UUZ</span>
<span class="o">...</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.retrieve_pdb_file">
<span class="sig-name descname"><span class="pre">retrieve_pdb_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pdb_code</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsolete</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.retrieve_pdb_file" title="Permalink to this definition"></a></dt>
<dd><p>Fetch PDB structure file from PDB server, and store it locally.</p>
<p>The PDB structure’s file name is returned as a single string.
If obsolete <code class="docutils literal notranslate"><span class="pre">==</span></code> True, the file will be saved in a special file tree.</p>
<p>NOTE. The default download format has changed from PDB to PDBx/mmCif</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pdb_code</strong> (<em>string</em>) – 4-symbols structure Id from PDB (e.g. 3J92).</p></li>
<li><p><strong>file_format</strong> (<em>string</em>) – <p>File format. Available options:</p>
<ul>
<li><p>”mmCif” (default, PDBx/mmCif file),</p></li>
<li><p>”pdb” (format PDB),</p></li>
<li><p>”xml” (PDBML/XML format),</p></li>
<li><p>”mmtf” (highly compressed),</p></li>
<li><p>”bundle” (PDB formatted archive for large structure)</p></li>
</ul>
</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – if set to True, existing structure files will be overwritten. Default: False</p></li>
<li><p><strong>obsolete</strong> (<em>bool</em>) – Has a meaning only for obsolete structures. If True, download the obsolete structure
to ‘obsolete’ folder, otherwise download won’t be performed.
This option doesn’t work for mmtf format as obsoleted structures aren’t stored in mmtf.
Also doesn’t have meaning when parameter pdir is specified.
Note: make sure that you are about to download the really obsolete structure.
Trying to download non-obsolete structure into obsolete folder will not work
and you face the “structure doesn’t exists” error.
Default: False</p></li>
<li><p><strong>pdir</strong> (<em>string</em>) – put the file in this directory (default: create a PDB-style directory tree)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>filename</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.update_pdb">
<span class="sig-name descname"><span class="pre">update_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_assemblies</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.update_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Update your local copy of the PDB files.</p>
<p>I guess this is the ‘most wanted’ function from this module.
It gets the weekly lists of new and modified pdb entries and
automatically downloads the according PDB files.
You can call this module as a weekly cron job.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.download_pdb_files">
<span class="sig-name descname"><span class="pre">download_pdb_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pdb_codes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsolete</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.download_pdb_files" title="Permalink to this definition"></a></dt>
<dd><p>Fetch set of PDB structure files from the PDB server and stores them locally.</p>
<p>The PDB structure’s file name is returned as a single string.
If obsolete <code class="docutils literal notranslate"><span class="pre">==</span></code> True, the files will be saved in a special file tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pdb_codes</strong> (<em>list of strings</em>) – a list of 4-symbols structure Ids from PDB</p></li>
<li><p><strong>file_format</strong> – <p>File format. Available options:</p>
<ul>
<li><p>”mmCif” (default, PDBx/mmCif file),</p></li>
<li><p>”pdb” (format PDB),</p></li>
<li><p>”xml” (PMDML/XML format),</p></li>
<li><p>”mmtf” (highly compressed),</p></li>
<li><p>”bundle” (PDB formatted archive for large structure)</p></li>
</ul>
</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – if set to True, existing structure files will be overwritten. Default: False</p></li>
<li><p><strong>obsolete</strong> (<em>bool</em>) – Has a meaning only for obsolete structures.
If True, download the obsolete structure
to ‘obsolete’ folder, otherwise download won’t be performed.
This option doesn’t work for mmtf format as obsoleted structures are not available as mmtf.
(default: False)</p></li>
<li><p><strong>pdir</strong> (<em>string</em>) – put the file in this directory (default: create a PDB-style directory tree)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>filenames</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.get_all_assemblies">
<span class="sig-name descname"><span class="pre">get_all_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mmCif'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.get_all_assemblies" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve the list of PDB entries with an associated bio assembly.</p>
<p>The requested list will be cached to avoid multiple calls to the FTP
server.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file_format</strong> (<em>str</em><em>, </em><em>optional</em>) – format in which to download the entries. Available
options are “mmCif” or “pdb”. Defaults to mmCif.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.retrieve_assembly_file">
<span class="sig-name descname"><span class="pre">retrieve_assembly_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pdb_code</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly_num</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.retrieve_assembly_file" title="Permalink to this definition"></a></dt>
<dd><p>Fetch one or more assembly structures associated with a PDB entry.</p>
<p>Unless noted below, parameters are described in <code class="docutils literal notranslate"><span class="pre">retrieve_pdb_file</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>assembly_num</strong> (<em>int</em>) – assembly number to download.</p>
</dd>
</dl>
<p>:rtype : str
:return: file name of the downloaded assembly file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.download_all_assemblies">
<span class="sig-name descname"><span class="pre">download_all_assemblies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">listfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.download_all_assemblies" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve all biological assemblies not in the local PDB copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>listfile</strong> (<em>str</em><em>, </em><em>optional</em>) – file name to which all assembly codes will be written</p></li>
<li><p><strong>file_format</strong> (<em>str</em><em>, </em><em>optional</em>) – format in which to download the entries. Available
options are “mmCif” or “pdb”. Defaults to mmCif.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.download_entire_pdb">
<span class="sig-name descname"><span class="pre">download_entire_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">listfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.download_entire_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve all PDB entries not present in the local PDB copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>listfile</strong> – filename to which all PDB codes will be written (optional)</p></li>
<li><p><strong>file_format</strong> – <p>File format. Available options:</p>
<ul>
<li><p>”mmCif” (default, PDBx/mmCif file),</p></li>
<li><p>”pdb” (format PDB),</p></li>
<li><p>”xml” (PMDML/XML format),</p></li>
<li><p>”mmtf” (highly compressed),</p></li>
<li><p>”bundle” (PDB formatted archive for large structure)</p></li>
</ul>
</p></li>
</ul>
</dd>
</dl>
<p>NOTE. The default download format has changed from PDB to PDBx/mmCif</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.download_obsolete_entries">
<span class="sig-name descname"><span class="pre">download_obsolete_entries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">listfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.download_obsolete_entries" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve all obsolete PDB entries not present in local obsolete PDB copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>listfile</strong> – filename to which all PDB codes will be written (optional)</p></li>
<li><p><strong>file_format</strong> – file format. Available options:
“mmCif” (default, PDBx/mmCif file),
“pdb” (format PDB),
“xml” (PMDML/XML format),</p></li>
</ul>
</dd>
</dl>
<p>NOTE. The default download format has changed from PDB to PDBx/mmCif</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.PDB.PDBList.PDBList.get_seqres_file">
<span class="sig-name descname"><span class="pre">get_seqres_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">savefile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdb_seqres.txt'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.PDB.PDBList.PDBList.get_seqres_file" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve and save a (big) file containing all the sequences of PDB entries.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.PDB.PDBIO.html" class="btn btn-neutral float-left" title="Bio.PDB.PDBIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.PDB.PDBParser.html" class="btn btn-neutral float-right" title="Bio.PDB.PDBParser module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2020, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.81.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.79/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>