<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.Align.bigbed module &mdash; Biopython 1.82.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Align.bigmaf module" href="Bio.Align.bigmaf.html" />
    <link rel="prev" title="Bio.Align.bed module" href="Bio.Align.bed.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="index.html">
            
              <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html#subpackages">Subpackages</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Align.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html#module-Bio.Align">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.Align.html">Bio.Align package</a></li>
      <li class="breadcrumb-item active">Bio.Align.bigbed module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Align/bigbed.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Align.bed.html" class="btn btn-neutral float-left" title="Bio.Align.bed module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Align.bigmaf.html" class="btn btn-neutral float-right" title="Bio.Align.bigmaf module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.Align.bigbed">
<span id="bio-align-bigbed-module"></span><h1>Bio.Align.bigbed module<a class="headerlink" href="#module-Bio.Align.bigbed" title="Permalink to this headline">ÔÉÅ</a></h1>
<p>Bio.Align support for alignment files in the bigBed format.</p>
<p>The bigBed format stores a series of pairwise alignments in a single indexed
binary file. Typically they are used for transcript to genome alignments. As
in the BED format, the alignment positions and alignment scores are stored,
but the aligned sequences are not.</p>
<p>See <a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/bigBed.html">http://genome.ucsc.edu/goldenPath/help/bigBed.html</a> for more information.</p>
<p>You are expected to use this module via the Bio.Align functions.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.bigbed.Field">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.bigbed.</span></span><span class="sig-name descname"><span class="pre">Field</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">as_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.Field" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.Field.__getnewargs__">
<span class="sig-name descname"><span class="pre">__getnewargs__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.Field.__getnewargs__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return self as a plain tuple.  Used by copy and pickle.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.Field.__new__">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">__new__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">_cls</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.Field.__new__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Create new instance of Field(as_type, name, comment)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.Field.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.Field.__repr__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return a nicely formatted representation string</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.Field.__slots__">
<span class="sig-name descname"><span class="pre">__slots__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">()</span></em><a class="headerlink" href="#Bio.Align.bigbed.Field.__slots__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.Field.as_type">
<span class="sig-name descname"><span class="pre">as_type</span></span><a class="headerlink" href="#Bio.Align.bigbed.Field.as_type" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.Field.comment">
<span class="sig-name descname"><span class="pre">comment</span></span><a class="headerlink" href="#Bio.Align.bigbed.Field.comment" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.Field.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#Bio.Align.bigbed.Field.name" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.bigbed.</span></span><span class="sig-name descname"><span class="pre">AutoSQLTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fields</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>AutoSQL table describing the columns of an (possibly extended) BED format.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable.default">
<span class="sig-name descname"><span class="pre">default</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#Bio.Align.bigbed.AutoSQLTable" title="Bio.Align.bigbed.AutoSQLTable"><span class="pre">Bio.Align.bigbed.AutoSQLTable</span></a></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">[Field(as_type='string',</span> <span class="pre">name='chrom',</span> <span class="pre">comment='Reference</span> <span class="pre">sequence</span> <span class="pre">chromosome</span> <span class="pre">or</span> <span class="pre">scaffold'),</span> <span class="pre">Field(as_type='uint',</span> <span class="pre">name='chromStart',</span> <span class="pre">comment='Start</span> <span class="pre">position</span> <span class="pre">in</span> <span class="pre">chromosome'),</span> <span class="pre">Field(as_type='uint',</span> <span class="pre">name='chromEnd',</span> <span class="pre">comment='End</span> <span class="pre">position</span> <span class="pre">in</span> <span class="pre">chromosome'),</span> <span class="pre">Field(as_type='string',</span> <span class="pre">name='name',</span> <span class="pre">comment='Name</span> <span class="pre">of</span> <span class="pre">item.'),</span> <span class="pre">Field(as_type='uint',</span> <span class="pre">name='score',</span> <span class="pre">comment='Score</span> <span class="pre">(0-1000)'),</span> <span class="pre">Field(as_type='char[1]',</span> <span class="pre">name='strand',</span> <span class="pre">comment='+</span> <span class="pre">or</span> <span class="pre">-</span> <span class="pre">for</span> <span class="pre">strand'),</span> <span class="pre">Field(as_type='uint',</span> <span class="pre">name='thickStart',</span> <span class="pre">comment='Start</span> <span class="pre">of</span> <span class="pre">where</span> <span class="pre">display</span> <span class="pre">should</span> <span class="pre">be</span> <span class="pre">thick</span> <span class="pre">(start</span> <span class="pre">codon)'),</span> <span class="pre">Field(as_type='uint',</span> <span class="pre">name='thickEnd',</span> <span class="pre">comment='End</span> <span class="pre">of</span> <span class="pre">where</span> <span class="pre">display</span> <span class="pre">should</span> <span class="pre">be</span> <span class="pre">thick</span> <span class="pre">(stop</span> <span class="pre">codon)'),</span> <span class="pre">Field(as_type='uint',</span> <span class="pre">name='reserved',</span> <span class="pre">comment='Used</span> <span class="pre">as</span> <span class="pre">itemRgb</span> <span class="pre">as</span> <span class="pre">of</span> <span class="pre">2004-11-22'),</span> <span class="pre">Field(as_type='int',</span> <span class="pre">name='blockCount',</span> <span class="pre">comment='Number</span> <span class="pre">of</span> <span class="pre">blocks'),</span> <span class="pre">Field(as_type='int[blockCount]',</span> <span class="pre">name='blockSizes',</span> <span class="pre">comment='Comma</span> <span class="pre">separated</span> <span class="pre">list</span> <span class="pre">of</span> <span class="pre">block</span> <span class="pre">sizes'),</span> <span class="pre">Field(as_type='int[blockCount]',</span> <span class="pre">name='chromStarts',</span> <span class="pre">comment='Start</span> <span class="pre">positions</span> <span class="pre">relative</span> <span class="pre">to</span> <span class="pre">chromStart')]</span></em><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable.default" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fields</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Create an AutoSQL table describing the columns of an (extended) BED format.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable.from_bytes">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_bytes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable.from_bytes" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return an AutoSQLTable initialized using the bytes object data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable.from_string">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable.from_string" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return an AutoSQLTable initialized using the string object data.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable.__str__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return str(self).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable.__bytes__">
<span class="sig-name descname"><span class="pre">__bytes__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable.__bytes__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">i</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable.__getitem__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>x.__getitem__(y) &lt;==&gt; x[y]</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AutoSQLTable.__annotations__">
<span class="sig-name descname"><span class="pre">__annotations__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'default':</span> <span class="pre">'AutoSQLTable'}</span></em><a class="headerlink" href="#Bio.Align.bigbed.AutoSQLTable.__annotations__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.bigbed.</span></span><span class="sig-name descname"><span class="pre">AlignmentWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bedN</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">declaration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extraIndex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AlignmentWriter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Align.interfaces.html#Bio.Align.interfaces.AlignmentWriter" title="Bio.Align.interfaces.AlignmentWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Align.interfaces.AlignmentWriter</span></code></a></p>
<p>Alignment file writer for the bigBed file format.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentWriter.fmt">
<span class="sig-name descname"><span class="pre">fmt</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'bigBed'</span></em><a class="headerlink" href="#Bio.Align.bigbed.AlignmentWriter.fmt" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentWriter.mode">
<span class="sig-name descname"><span class="pre">mode</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'wb'</span></em><a class="headerlink" href="#Bio.Align.bigbed.AlignmentWriter.mode" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentWriter.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bedN</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">declaration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">targets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extraIndex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AlignmentWriter.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Create an AlignmentWriter object.</p>
<dl>
<dt>Arguments:</dt><dd><ul>
<li><p>target      - output stream or file name.</p></li>
<li><dl class="simple">
<dt>bedN        - number of columns in the BED file.</dt><dd><p>This must be between 3 and 12; default value is 12.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>declaration - an AutoSQLTable object declaring the fields in the BED file.</dt><dd><p>Required only if the BED file contains extra (custom) fields.
Default value is None.</p>
</dd>
</dl>
</li>
<li><dl>
<dt>targets     - A list of SeqRecord objects with the chromosomes in the</dt><dd><p>order as they appear in the alignments. The sequence
contents in each SeqRecord may be undefined, but the
sequence length must be defined, as in this example:</p>
<p>SeqRecord(Seq(None, length=248956422), id=‚Äùchr1‚Äù)</p>
<p>If targets is None (the default value), the alignments
must have an attribute .targets providing the list of
SeqRecord objects.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>compress    - If True (default), compress data using zlib.</dt><dd><p>If False, do not compress data.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>extraIndex  - List of strings with the names of extra columns to be</dt><dd><p>indexed.
Default value is an empty list.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentWriter.write_file">
<span class="sig-name descname"><span class="pre">write_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stream</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AlignmentWriter.write_file" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write the alignments to the file strenm, and return the number of alignments.</p>
<p>alignments - A list or iterator returning Alignment objects
stream     - Output file stream.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentWriter.write_alignments">
<span class="sig-name descname"><span class="pre">write_alignments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignments</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reductions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_indices</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AlignmentWriter.write_alignments" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write alignments to the output file, and return the number of alignments.</p>
<p>alignments - A list or iterator returning Alignment objects
stream     - Output file stream.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentWriter.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.Align.bigbed.AlignmentWriter.__abstractmethods__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.bigbed.</span></span><span class="sig-name descname"><span class="pre">AlignmentIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AlignmentIterator" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Align.interfaces.html#Bio.Align.interfaces.AlignmentIterator" title="Bio.Align.interfaces.AlignmentIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Align.interfaces.AlignmentIterator</span></code></a></p>
<p>Alignment iterator for bigBed files.</p>
<p>The pairwise alignments stored in the bigBed file are loaded and returned
incrementally.  Additional alignment information is stored as attributes
of each alignment.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentIterator.fmt">
<span class="sig-name descname"><span class="pre">fmt</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'bigBed'</span></em><a class="headerlink" href="#Bio.Align.bigbed.AlignmentIterator.fmt" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentIterator.mode">
<span class="sig-name descname"><span class="pre">mode</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'b'</span></em><a class="headerlink" href="#Bio.Align.bigbed.AlignmentIterator.mode" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentIterator.__len__">
<span class="sig-name descname"><span class="pre">__len__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AlignmentIterator.__len__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return the number of alignments.</p>
<p>The number of alignments is cached. If not yet calculated, the iterator
is rewound to the beginning, and the number of alignments is calculated
by iterating over the alignments. The iterator is then returned to its
original position in the file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentIterator.search">
<span class="sig-name descname"><span class="pre">search</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.bigbed.AlignmentIterator.search" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Iterate over alignments overlapping the specified chromosome region..</p>
<p>This method searches the index to find alignments to the specified
chromosome that fully or partially overlap the chromosome region
between start and end.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>chromosome - chromosome name. If None (default value), include all
alignments.</p></li>
<li><p>start      - starting position on the chromosome. If None (default
value), use 0 as the starting position.</p></li>
<li><p>end        - end position on the chromosome. If None (default value),
use the length of the chromosome as the end position.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.bigbed.AlignmentIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.Align.bigbed.AlignmentIterator.__abstractmethods__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Align.bed.html" class="btn btn-neutral float-left" title="Bio.Align.bed module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Align.bigmaf.html" class="btn btn-neutral float-right" title="Bio.Align.bigmaf module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2023, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.82.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.80/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>