

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.SeqIO.Interfaces module &mdash; Biopython 1.75.dev0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SeqIO.NibIO module" href="Bio.SeqIO.NibIO.html" />
    <link rel="prev" title="Bio.SeqIO.InsdcIO module" href="Bio.SeqIO.InsdcIO.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.75.dev0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SeqIO.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html#module-Bio.SeqIO">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.SeqIO.html">Bio.SeqIO package</a> &raquo;</li>
        
      <li>Bio.SeqIO.Interfaces module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SeqIO.NibIO.html" class="btn btn-neutral float-right" title="Bio.SeqIO.NibIO module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SeqIO.InsdcIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.InsdcIO module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.SeqIO.Interfaces">
<span id="bio-seqio-interfaces-module"></span><h1>Bio.SeqIO.Interfaces module<a class="headerlink" href="#module-Bio.SeqIO.Interfaces" title="Permalink to this headline">¶</a></h1>
<p>Bio.SeqIO support module (not for general use).</p>
<p>Unless you are writing a new parser or writer for Bio.SeqIO, you should not
use this module.  It provides base classes to try and simplify things.</p>
<dl class="class">
<dt id="Bio.SeqIO.Interfaces.SequenceIterator">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SeqIO.Interfaces.</code><code class="sig-name descname">SequenceIterator</code><span class="sig-paren">(</span><em class="sig-param">handle</em>, <em class="sig-param">alphabet=Alphabet()</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequenceIterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for building SeqRecord iterators.</p>
<p>You should write a __next__ method to return SeqRecord  objects.  You may
wish to redefine the __init__ method as well.</p>
</dd></dl>

<dl class="class">
<dt id="Bio.SeqIO.Interfaces.SequenceWriter">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SeqIO.Interfaces.</code><code class="sig-name descname">SequenceWriter</code><span class="sig-paren">(</span><em class="sig-param">handle</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequenceWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for building SeqRecord writers.</p>
<p>Interlaced file formats (e.g. Clustal) should subclass directly.</p>
<p>Sequential file formats (e.g. Fasta, GenBank) should subclass the
SequentialSequenceWriter class instead.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.SeqIO.Interfaces.SequenceWriter.clean" title="Bio.SeqIO.Interfaces.SequenceWriter.clean"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clean</span></code></a>(self, text)</p></td>
<td><p>Use this to avoid getting newlines in the output.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.SeqIO.Interfaces.SequenceWriter.write_file" title="Bio.SeqIO.Interfaces.SequenceWriter.write_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_file</span></code></a>(self, records)</p></td>
<td><p>Use this to write an entire file containing the given records.</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.SeqIO.Interfaces.SequenceWriter.clean">
<code class="sig-name descname">clean</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">text</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequenceWriter.clean" title="Permalink to this definition">¶</a></dt>
<dd><p>Use this to avoid getting newlines in the output.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.SeqIO.Interfaces.SequenceWriter.write_file">
<code class="sig-name descname">write_file</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">records</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequenceWriter.write_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Use this to write an entire file containing the given records.</p>
<p>records - A list or iterator returning SeqRecord objects</p>
<p>Should return the number of records (as an integer).</p>
<p>This method can only be called once.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Bio.SeqIO.Interfaces.SequentialSequenceWriter">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SeqIO.Interfaces.</code><code class="sig-name descname">SequentialSequenceWriter</code><span class="sig-paren">(</span><em class="sig-param">handle</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.SeqIO.Interfaces.SequenceWriter" title="Bio.SeqIO.Interfaces.SequenceWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqIO.Interfaces.SequenceWriter</span></code></a></p>
<p>Base class for sequence writers. This class should be subclassed.</p>
<p>It is intended for sequential file formats with an (optional)
header, repeated records, and an (optional) footer.</p>
<p>In this case (as with interlaced file formats), the user may
simply call the write_file() method and be done.</p>
<p>However, they may also call the write_header(), followed
by multiple calls to write_record() and/or write_records()
followed finally by write_footer().</p>
<p>Users must call write_header() and write_footer() even when
the file format concerned doesn’t have a header or footer.
This is to try and make life as easy as possible when
switching the output format.</p>
<p>Note that write_header() cannot require any assumptions about
the number of records.</p>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">clean</span></code>(self, text)</p></td>
<td><p>Use this to avoid getting newlines in the output.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_file" title="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_file</span></code></a>(self, records)</p></td>
<td><p>Use this to write an entire file containing the given records.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_footer" title="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_footer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_footer</span></code></a>(self)</p></td>
<td><p>Write the file footer.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_header" title="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_header"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_header</span></code></a>(self)</p></td>
<td><p>Write the file header.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_record" title="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_record"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_record</span></code></a>(self, record)</p></td>
<td><p>Write a single record to the output file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_records" title="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_records"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_records</span></code></a>(self, records)</p></td>
<td><p>Write multiple record to the output file.</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_file">
<code class="sig-name descname">write_file</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">records</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Use this to write an entire file containing the given records.</p>
<p>records - A list or iterator returning SeqRecord objects</p>
<p>This method can only be called once.  Returns the number of records
written.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_footer">
<code class="sig-name descname">write_footer</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_footer" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the file footer.</p>
<p>If your file format defines a footer, you should implement this method
in order to write the footer after all the records.</p>
<p>The default implementation checks the private attribute ._footer_written
to ensure the footer is only written once.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_header">
<code class="sig-name descname">write_header</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_header" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the file header.</p>
<p>If your file format defines a header, you should implement this method
in order to write the header before any of the records.</p>
<p>The default implementation checks the private attribute ._header_written
to ensure the header is only written once.</p>
</dd></dl>

<dl class="method">
<dt id="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_record">
<code class="sig-name descname">write_record</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">record</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_record" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a single record to the output file.</p>
<p>record - a SeqRecord object</p>
<p>Once you have called write_header() you can call write_record()
and/or write_records() as many times as needed.  Then call
write_footer() and close().</p>
</dd></dl>

<dl class="method">
<dt id="Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_records">
<code class="sig-name descname">write_records</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">records</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SeqIO.Interfaces.SequentialSequenceWriter.write_records" title="Permalink to this definition">¶</a></dt>
<dd><p>Write multiple record to the output file.</p>
<p>records - A list or iterator returning SeqRecord objects</p>
<p>Once you have called write_header() you can call write_record()
and/or write_records() as many times as needed.  Then call
write_footer() and close().</p>
<p>Returns the number of records written.</p>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.SeqIO.NibIO.html" class="btn btn-neutral float-right" title="Bio.SeqIO.NibIO module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SeqIO.InsdcIO.html" class="btn btn-neutral float-left" title="Bio.SeqIO.InsdcIO module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2017, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>