<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.Blast.Applications module &mdash; Biopython 1.80.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Blast.NCBIWWW module" href="Bio.Blast.NCBIWWW.html" />
    <link rel="prev" title="Bio.Blast package" href="Bio.Blast.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
            <a href="index.html">
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Blast.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html#module-Bio.Blast">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
          <li><a href="Bio.Blast.html">Bio.Blast package</a> &raquo;</li>
      <li>Bio.Blast.Applications module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Blast/Applications.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Blast.html" class="btn btn-neutral float-left" title="Bio.Blast package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Blast.NCBIWWW.html" class="btn btn-neutral float-right" title="Bio.Blast.NCBIWWW module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.Blast.Applications">
<span id="bio-blast-applications-module"></span><h1>Bio.Blast.Applications module<a class="headerlink" href="#module-Bio.Blast.Applications" title="Permalink to this headline"></a></h1>
<p>Definitions for interacting with BLAST related applications (OBSOLETE).</p>
<p>Wrappers for the new NCBI BLAST+ tools (written in C++):</p>
<blockquote>
<div><ul class="simple">
<li><p>NcbiblastpCommandline - Protein-Protein BLAST</p></li>
<li><p>NcbiblastnCommandline - Nucleotide-Nucleotide BLAST</p></li>
<li><p>NcbiblastxCommandline - Translated Query-Protein Subject BLAST</p></li>
<li><p>NcbitblastnCommandline - Protein Query-Translated Subject BLAST</p></li>
<li><p>NcbitblastxCommandline - Translated Query-Protein Subject BLAST</p></li>
<li><p>NcbipsiblastCommandline - Position-Specific Initiated BLAST</p></li>
<li><p>NcbirpsblastCommandline - Reverse Position Specific BLAST</p></li>
<li><p>NcbirpstblastnCommandline - Translated Reverse Position Specific BLAST</p></li>
<li><p>NcbideltablastCommandline - Protein-Protein domain enhanced lookup time accelerated blast</p></li>
<li><p>NcbiblastformatterCommandline - Convert ASN.1 to other BLAST output formats</p></li>
<li><p>NcbimakeblastdbCommandline - Application to create BLAST databases</p></li>
</ul>
</div></blockquote>
<p>For further details, see:</p>
<p>Camacho et al. BLAST+: architecture and applications
BMC Bioinformatics 2009, 10:421
<a class="reference external" href="https://doi.org/10.1186/1471-2105-10-421">https://doi.org/10.1186/1471-2105-10-421</a></p>
<p>We have decided to remove this module in future, and instead recommend
building your command and invoking it via the subprocess module directly.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbiblastpCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blastp'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._NcbiblastMain2SeqCommandline</span></code></p>
<p>Create a commandline for the NCBI BLAST+ program blastp (for proteins).</p>
<p>With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI
replaced the old blastall tool with separate tools for each of the searches.
This wrapper therefore replaces BlastallCommandline with option -p blastp.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbiblastpCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbiblastpCommandline</span><span class="p">(</span><span class="n">query</span><span class="o">=</span><span class="s2">&quot;rosemary.pro&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nr&quot;</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">evalue</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">remote</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ungapped</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbiblastpCommandline(cmd=&#39;blastp&#39;, query=&#39;rosemary.pro&#39;, db=&#39;nr&#39;, evalue=0.001, remote=True, ungapped=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">blastp -query rosemary.pro -db nr -evalue 0.001 -remote -ungapped</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blastp'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.best_hit_overhang">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_overhang</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.best_hit_overhang" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm overhang value (float, recommended value: 0.1)</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.best_hit_score_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_score_edge</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.best_hit_score_edge" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm score edge value (float).</p>
<p>Float between 0.0 and 0.5 inclusive. Recommended value: 0.1</p>
<p>Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.comp_based_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">comp_based_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.comp_based_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use composition-based statistics (string, default 2, i.e. True).</p>
<p>0, F or f: no composition-based statistics</p>
<p>2, T or t, D or d : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties</p>
<p>Note that tblastn also supports values of 1 and 3.</p>
<p>This controls the addition of the -comp_based_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.culling_limit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">culling_limit</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.culling_limit" title="Permalink to this definition"></a></dt>
<dd><p>Hit culling limit (integer).</p>
<p>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit.</p>
<p>Incompatible with: best_hit_overhang, best_hit_score_edge.</p>
<p>This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.db_hard_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_hard_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.db_hard_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for hard masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as hard masking. Incompatible with: db_soft_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_hard_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.db_soft_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_soft_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.db_soft_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for soft masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as soft masking. Incompatible with: db_hard_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_soft_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.gapextend" title="Permalink to this definition"></a></dt>
<dd><p>Cost to extend a gap (integer).</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.gapopen" title="Permalink to this definition"></a></dt>
<dd><p>Cost to open a gap (integer).</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.matrix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matrix</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.matrix" title="Permalink to this definition"></a></dt>
<dd><p>Scoring matrix name (default BLOSUM62).</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.seg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seg</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.seg" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with SEG (string).</p>
<p>Format: “yes”, “window locut hicut”, or “no” to disable
Default is “12 2.2 2.5”</p>
<p>This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.subject" title="Permalink to this definition"></a></dt>
<dd><p>Subject sequence(s) to search.</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask</p>
<p>See also subject_loc.</p>
<p>This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.subject_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.subject_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the subject sequence (Format: start-stop).</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask, remote.</p>
<p>See also subject.</p>
<p>This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.task">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">task</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.task" title="Permalink to this definition"></a></dt>
<dd><p>Task to execute (string, blastp (default), blastp-fast or blastp-short).</p>
<p>This controls the addition of the -task parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimum score for words to be added to the BLAST lookup table (float).</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.ungapped">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ungapped</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.ungapped" title="Permalink to this definition"></a></dt>
<dd><p>Perform ungapped alignment only?</p>
<p>This property controls the addition of the -ungapped switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.use_sw_tback">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">use_sw_tback</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.use_sw_tback" title="Permalink to this definition"></a></dt>
<dd><p>Compute locally optimal Smith-Waterman alignments?</p>
<p>This property controls the addition of the -use_sw_tback switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastpCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastpCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbiblastnCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blastn'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._NcbiblastMain2SeqCommandline</span></code></p>
<p>Wrapper for the NCBI BLAST+ program blastn (for nucleotides).</p>
<p>With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI
replaced the old blastall tool with separate tools for each of the searches.
This wrapper therefore replaces BlastallCommandline with option -p blastn.</p>
<p>For example, to run a search against the “nt” nucleotide database using the
FASTA nucleotide file “m_code.fasta” as the query, with an expectation value
cut off of 0.001, saving the output to a file in XML format:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbiblastnCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbiblastnCommandline</span><span class="p">(</span><span class="n">query</span><span class="o">=</span><span class="s2">&quot;m_cold.fasta&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nt&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;plus&quot;</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">evalue</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="s2">&quot;m_cold.xml&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbiblastnCommandline(cmd=&#39;blastn&#39;, out=&#39;m_cold.xml&#39;, outfmt=5, query=&#39;m_cold.fasta&#39;, db=&#39;nt&#39;, evalue=0.001, strand=&#39;plus&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">blastn -out m_cold.xml -outfmt 5 -query m_cold.fasta -db nt -evalue 0.001 -strand plus</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blastn'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.best_hit_overhang">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_overhang</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.best_hit_overhang" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm overhang value (float, recommended value: 0.1)</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.best_hit_score_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_score_edge</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.best_hit_score_edge" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm score edge value (float).</p>
<p>Float between 0.0 and 0.5 inclusive. Recommended value: 0.1</p>
<p>Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.culling_limit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">culling_limit</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.culling_limit" title="Permalink to this definition"></a></dt>
<dd><p>Hit culling limit (integer).</p>
<p>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit.</p>
<p>Incompatible with: best_hit_overhang, best_hit_score_edge.</p>
<p>This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.db_hard_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_hard_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.db_hard_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for hard masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as hard masking. Incompatible with: db_soft_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_hard_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.db_soft_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_soft_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.db_soft_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for soft masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as soft masking. Incompatible with: db_hard_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_soft_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.dust">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dust</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.dust" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with DUST (string).</p>
<p>Format: ‘yes’, ‘level window linker’, or ‘no’ to disable.</p>
<p>Default = ‘20 64 1’.</p>
<p>This controls the addition of the -dust parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.filtering_db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filtering_db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.filtering_db" title="Permalink to this definition"></a></dt>
<dd><p>BLAST database containing filtering elements (i.e. repeats).</p>
<p>This controls the addition of the -filtering_db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.gapextend" title="Permalink to this definition"></a></dt>
<dd><p>Cost to extend a gap (integer).</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.gapopen" title="Permalink to this definition"></a></dt>
<dd><p>Cost to open a gap (integer).</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.index_name">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">index_name</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.index_name" title="Permalink to this definition"></a></dt>
<dd><p>MegaBLAST database index name.</p>
<p>This controls the addition of the -index_name parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.min_raw_gapped_score">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">min_raw_gapped_score</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.min_raw_gapped_score" title="Permalink to this definition"></a></dt>
<dd><p>Minimum raw gapped score to keep an alignment in the preliminary gapped and traceback stages (integer).</p>
<p>This controls the addition of the -min_raw_gapped_score parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.no_greedy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">no_greedy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.no_greedy" title="Permalink to this definition"></a></dt>
<dd><p>Use non-greedy dynamic programming extension</p>
<p>This property controls the addition of the -no_greedy switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.off_diagonal_range">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">off_diagonal_range</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.off_diagonal_range" title="Permalink to this definition"></a></dt>
<dd><p>Number of off-diagonals to search for the 2nd hit (integer).</p>
<p>Expects a positive integer, or 0 (default) to turn off.Added in BLAST 2.2.23+</p>
<p>This controls the addition of the -off_diagonal_range parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.penalty">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">penalty</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.penalty" title="Permalink to this definition"></a></dt>
<dd><p>Penalty for a nucleotide mismatch (integer, at most zero).</p>
<p>This controls the addition of the -penalty parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.perc_identity">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">perc_identity</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.perc_identity" title="Permalink to this definition"></a></dt>
<dd><p>Percent identity (real, 0 to 100 inclusive).</p>
<p>This controls the addition of the -perc_identity parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.reward">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reward</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.reward" title="Permalink to this definition"></a></dt>
<dd><p>Reward for a nucleotide match (integer, at least zero).</p>
<p>This controls the addition of the -reward parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.strand">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.strand" title="Permalink to this definition"></a></dt>
<dd><p>Query strand(s) to search against database/subject.</p>
<p>Values allowed are “both” (default), “minus”, “plus”.</p>
<p>This controls the addition of the -strand parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.subject" title="Permalink to this definition"></a></dt>
<dd><p>Subject sequence(s) to search.</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask</p>
<p>See also subject_loc.</p>
<p>This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.subject_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.subject_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the subject sequence (Format: start-stop).</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask, remote.</p>
<p>See also subject.</p>
<p>This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.task">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">task</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.task" title="Permalink to this definition"></a></dt>
<dd><p>Task to execute (string, default ‘megablast’)</p>
<p>Allowed values ‘blastn’, ‘blastn-short’, ‘dc-megablast’, ‘megablast’ (the default), or ‘vecscreen’.</p>
<p>This controls the addition of the -task parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.template_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">template_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.template_length" title="Permalink to this definition"></a></dt>
<dd><p>Discontiguous MegaBLAST template length (integer).</p>
<p>Allowed values: 16, 18, 21.</p>
<p>Requires: template_type.</p>
<p>This controls the addition of the -template_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.template_type">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">template_type</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.template_type" title="Permalink to this definition"></a></dt>
<dd><p>Discontiguous MegaBLAST template type (string).</p>
<p>Allowed values: ‘coding’, ‘coding_and_optimal’ or ‘optimal’.
Requires: template_length.</p>
<p>This controls the addition of the -template_type parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.ungapped">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ungapped</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.ungapped" title="Permalink to this definition"></a></dt>
<dd><p>Perform ungapped alignment only?</p>
<p>This property controls the addition of the -ungapped switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.use_index">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">use_index</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.use_index" title="Permalink to this definition"></a></dt>
<dd><p>Use MegaBLAST database index (Boolean, Default = False)</p>
<p>This controls the addition of the -use_index parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.window_masker_db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_masker_db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.window_masker_db" title="Permalink to this definition"></a></dt>
<dd><p>Enable WindowMasker filtering using this repeats database (string).</p>
<p>This controls the addition of the -window_masker_db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.window_masker_taxid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_masker_taxid</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.window_masker_taxid" title="Permalink to this definition"></a></dt>
<dd><p>Enable WindowMasker filtering using a Taxonomic ID (integer).</p>
<p>This controls the addition of the -window_masker_taxid parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastnCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastnCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbiblastxCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blastx'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._NcbiblastMain2SeqCommandline</span></code></p>
<p>Wrapper for the NCBI BLAST+ program blastx (nucleotide query, protein database).</p>
<p>With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI
replaced the old blastall tool with separate tools for each of the searches.
This wrapper therefore replaces BlastallCommandline with option -p blastx.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbiblastxCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbiblastxCommandline</span><span class="p">(</span><span class="n">query</span><span class="o">=</span><span class="s2">&quot;m_cold.fasta&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nr&quot;</span><span class="p">,</span> <span class="n">evalue</span><span class="o">=</span><span class="mf">0.001</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbiblastxCommandline(cmd=&#39;blastx&#39;, query=&#39;m_cold.fasta&#39;, db=&#39;nr&#39;, evalue=0.001)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">blastx -query m_cold.fasta -db nr -evalue 0.001</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blastx'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.best_hit_overhang">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_overhang</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.best_hit_overhang" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm overhang value (float, recommended value: 0.1)</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.best_hit_score_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_score_edge</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.best_hit_score_edge" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm score edge value (float).</p>
<p>Float between 0.0 and 0.5 inclusive. Recommended value: 0.1</p>
<p>Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.comp_based_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">comp_based_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.comp_based_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use composition-based statistics for blastp, blastx, or tblastn.</p>
<p>D or d: default (equivalent to 2 )</p>
<p>0 or F or f: no composition-based statistics</p>
<p>1: Composition-based statistics as in NAR 29:2994-3005, 2001</p>
<p>2 or T or t : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties</p>
<p>3: Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, unconditionally.</p>
<p>For programs other than tblastn, must either be absent or be D, F or 0</p>
<p>Default = 2.</p>
<p>This controls the addition of the -comp_based_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.culling_limit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">culling_limit</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.culling_limit" title="Permalink to this definition"></a></dt>
<dd><p>Hit culling limit (integer).</p>
<p>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit.</p>
<p>Incompatible with: best_hit_overhang, best_hit_score_edge.</p>
<p>This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.db_hard_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_hard_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.db_hard_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for hard masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as hard masking. Incompatible with: db_soft_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_hard_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.db_soft_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_soft_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.db_soft_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for soft masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as soft masking. Incompatible with: db_hard_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_soft_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.frame_shift_penalty">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">frame_shift_penalty</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.frame_shift_penalty" title="Permalink to this definition"></a></dt>
<dd><p>Frame shift penalty (integer, at least 1, default ignored) (OBSOLETE).</p>
<p>This was removed in BLAST 2.2.27+</p>
<p>This controls the addition of the -frame_shift_penalty parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.gapextend" title="Permalink to this definition"></a></dt>
<dd><p>Cost to extend a gap (integer).</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.gapopen" title="Permalink to this definition"></a></dt>
<dd><p>Cost to open a gap (integer).</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.matrix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matrix</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.matrix" title="Permalink to this definition"></a></dt>
<dd><p>Scoring matrix name (default BLOSUM62).</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.max_intron_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_intron_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.max_intron_length" title="Permalink to this definition"></a></dt>
<dd><p>Maximum intron length (integer).</p>
<p>Length of the largest intron allowed in a translated nucleotide sequence when linking multiple distinct alignments (a negative value disables linking). Default zero.</p>
<p>This controls the addition of the -max_intron_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.query_gencode">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_gencode</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.query_gencode" title="Permalink to this definition"></a></dt>
<dd><p>Genetic code to use to translate query (integer, default 1).</p>
<p>This controls the addition of the -query_gencode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.seg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seg</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.seg" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with SEG (string).</p>
<p>Format: “yes”, “window locut hicut”, or “no” to disable.Default is “12 2.2 2.5”</p>
<p>This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.strand">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.strand" title="Permalink to this definition"></a></dt>
<dd><p>Query strand(s) to search against database/subject.</p>
<p>Values allowed are “both” (default), “minus”, “plus”.</p>
<p>This controls the addition of the -strand parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.subject" title="Permalink to this definition"></a></dt>
<dd><p>Subject sequence(s) to search.</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask</p>
<p>See also subject_loc.</p>
<p>This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.subject_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.subject_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the subject sequence (Format: start-stop).</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask, remote.</p>
<p>See also subject.</p>
<p>This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.task">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">task</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.task" title="Permalink to this definition"></a></dt>
<dd><p>Task to execute (string, blastx (default) or blastx-fast).</p>
<p>This controls the addition of the -task parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimum score for words to be added to the BLAST lookup table (float).</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.ungapped">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ungapped</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.ungapped" title="Permalink to this definition"></a></dt>
<dd><p>Perform ungapped alignment only?</p>
<p>This property controls the addition of the -ungapped switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.use_sw_tback">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">use_sw_tback</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.use_sw_tback" title="Permalink to this definition"></a></dt>
<dd><p>Compute locally optimal Smith-Waterman alignments?</p>
<p>This property controls the addition of the -use_sw_tback switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastxCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastxCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbitblastnCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tblastn'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._NcbiblastMain2SeqCommandline</span></code></p>
<p>Wrapper for the NCBI BLAST+ program tblastn.</p>
<p>With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI
replaced the old blastall tool with separate tools for each of the searches.
This wrapper therefore replaces BlastallCommandline with option -p tblastn.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbitblastnCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbitblastnCommandline</span><span class="p">(</span><span class="n">help</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbitblastnCommandline(cmd=&#39;tblastn&#39;, help=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">tblastn -help</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tblastn'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.best_hit_overhang">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_overhang</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.best_hit_overhang" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm overhang value (float, recommended value: 0.1)</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.best_hit_score_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_score_edge</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.best_hit_score_edge" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm score edge value (float).</p>
<p>Float between 0.0 and 0.5 inclusive. Recommended value: 0.1</p>
<p>Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.comp_based_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">comp_based_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.comp_based_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use composition-based statistics (string, default 2, i.e. True).</p>
<p>0, F or f: no composition-based statistics</p>
<p>1: Composition-based statistics as in NAR 29:2994-3005, 2001</p>
<p>2, T or t, D or d : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties</p>
<p>3: Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, unconditionally</p>
<p>Note that only tblastn supports values of 1 and 3.</p>
<p>This controls the addition of the -comp_based_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.culling_limit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">culling_limit</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.culling_limit" title="Permalink to this definition"></a></dt>
<dd><p>Hit culling limit (integer).</p>
<p>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit.</p>
<p>Incompatible with: best_hit_overhang, best_hit_score_edge.</p>
<p>This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.db_gencode">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_gencode</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.db_gencode" title="Permalink to this definition"></a></dt>
<dd><p>Genetic code to use to translate query (integer, default 1).</p>
<p>This controls the addition of the -db_gencode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.db_hard_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_hard_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.db_hard_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for hard masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as hard masking. Incompatible with: db_soft_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_hard_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.db_soft_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_soft_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.db_soft_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for soft masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as soft masking. Incompatible with: db_hard_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_soft_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.frame_shift_penalty">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">frame_shift_penalty</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.frame_shift_penalty" title="Permalink to this definition"></a></dt>
<dd><p>Frame shift penalty (integer, at least 1, default ignored) (OBSOLETE).</p>
<p>This was removed in BLAST 2.2.27+</p>
<p>This controls the addition of the -frame_shift_penalty parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.gapextend" title="Permalink to this definition"></a></dt>
<dd><p>Cost to extend a gap (integer).</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.gapopen" title="Permalink to this definition"></a></dt>
<dd><p>Cost to open a gap (integer).</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.in_pssm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">in_pssm</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.in_pssm" title="Permalink to this definition"></a></dt>
<dd><p>PSI-BLAST checkpoint file.</p>
<p>Incompatible with: remote, query</p>
<p>This controls the addition of the -in_pssm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.matrix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matrix</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.matrix" title="Permalink to this definition"></a></dt>
<dd><p>Scoring matrix name (default BLOSUM62).</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.max_intron_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_intron_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.max_intron_length" title="Permalink to this definition"></a></dt>
<dd><p>Maximum intron length (integer).</p>
<p>Length of the largest intron allowed in a translated nucleotide sequence when linking multiple distinct alignments (a negative value disables linking). Default zero.</p>
<p>This controls the addition of the -max_intron_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.seg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seg</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.seg" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with SEG (string).</p>
<p>Format: “yes”, “window locut hicut”, or “no” to disable.</p>
<p>Default is “12 2.2 2.5”</p>
<p>This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.subject" title="Permalink to this definition"></a></dt>
<dd><p>Subject sequence(s) to search.</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask</p>
<p>See also subject_loc.</p>
<p>This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.subject_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.subject_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the subject sequence (Format: start-stop).</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask, remote.</p>
<p>See also subject.</p>
<p>This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.task">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">task</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.task" title="Permalink to this definition"></a></dt>
<dd><p>Task to execute (string, tblastn (default) or tblastn-fast).</p>
<p>This controls the addition of the -task parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimum score for words to be added to the BLAST lookup table (float).</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.ungapped">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ungapped</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.ungapped" title="Permalink to this definition"></a></dt>
<dd><p>Perform ungapped alignment only?</p>
<p>This property controls the addition of the -ungapped switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.use_sw_tback">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">use_sw_tback</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.use_sw_tback" title="Permalink to this definition"></a></dt>
<dd><p>Compute locally optimal Smith-Waterman alignments?</p>
<p>This property controls the addition of the -use_sw_tback switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastnCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastnCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbitblastxCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tblastx'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._NcbiblastMain2SeqCommandline</span></code></p>
<p>Wrapper for the NCBI BLAST+ program tblastx.</p>
<p>With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI
replaced the old blastall tool with separate tools for each of the searches.
This wrapper therefore replaces BlastallCommandline with option -p tblastx.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbitblastxCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbitblastxCommandline</span><span class="p">(</span><span class="n">help</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbitblastxCommandline(cmd=&#39;tblastx&#39;, help=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">tblastx -help</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tblastx'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.best_hit_overhang">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_overhang</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.best_hit_overhang" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm overhang value (float, recommended value: 0.1)</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.best_hit_score_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_score_edge</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.best_hit_score_edge" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm score edge value (float).</p>
<p>Float between 0.0 and 0.5 inclusive. Recommended value: 0.1</p>
<p>Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.culling_limit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">culling_limit</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.culling_limit" title="Permalink to this definition"></a></dt>
<dd><p>Hit culling limit (integer).</p>
<p>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit.</p>
<p>Incompatible with: best_hit_overhang, best_hit_score_edge.</p>
<p>This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.db_gencode">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_gencode</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.db_gencode" title="Permalink to this definition"></a></dt>
<dd><p>Genetic code to use to translate query (integer, default 1).</p>
<p>This controls the addition of the -db_gencode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.db_hard_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_hard_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.db_hard_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for hard masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as hard masking. Incompatible with: db_soft_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_hard_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.db_soft_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db_soft_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.db_soft_mask" title="Permalink to this definition"></a></dt>
<dd><p>Filtering algorithm for soft masking (integer).</p>
<p>Filtering algorithm ID to apply to BLAST database as soft masking. Incompatible with: db_hard_mask, subject, subject_loc</p>
<p>This controls the addition of the -db_soft_mask parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.gapextend" title="Permalink to this definition"></a></dt>
<dd><p>Cost to extend a gap (integer).</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.gapopen" title="Permalink to this definition"></a></dt>
<dd><p>Cost to open a gap (integer).</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.matrix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matrix</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.matrix" title="Permalink to this definition"></a></dt>
<dd><p>Scoring matrix name (default BLOSUM62).</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.max_intron_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_intron_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.max_intron_length" title="Permalink to this definition"></a></dt>
<dd><p>Maximum intron length (integer).</p>
<p>Length of the largest intron allowed in a translated nucleotide sequence when linking multiple distinct alignments (a negative value disables linking). Default zero.</p>
<p>This controls the addition of the -max_intron_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.query_gencode">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_gencode</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.query_gencode" title="Permalink to this definition"></a></dt>
<dd><p>Genetic code to use to translate query (integer, default 1).</p>
<p>This controls the addition of the -query_gencode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.seg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seg</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.seg" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with SEG (string).</p>
<p>Format: “yes”, “window locut hicut”, or “no” to disable.</p>
<p>Default is “12 2.2 2.5”</p>
<p>This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.strand">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.strand" title="Permalink to this definition"></a></dt>
<dd><p>Query strand(s) to search against database/subject.</p>
<p>Values allowed are “both” (default), “minus”, “plus”.</p>
<p>This controls the addition of the -strand parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.subject" title="Permalink to this definition"></a></dt>
<dd><p>Subject sequence(s) to search.</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask</p>
<p>See also subject_loc.</p>
<p>This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.subject_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.subject_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the subject sequence (Format: start-stop).</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask, remote.</p>
<p>See also subject.</p>
<p>This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimum score for words to be added to the BLAST lookup table (float).</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbitblastxCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbitblastxCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbipsiblastCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'psiblast'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._Ncbiblast2SeqCommandline</span></code></p>
<p>Wrapper for the NCBI BLAST+ program psiblast.</p>
<p>With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI
replaced the old blastpgp tool with a similar tool psiblast. This wrapper
therefore replaces BlastpgpCommandline, the wrapper for blastpgp.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbipsiblastCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbipsiblastCommandline</span><span class="p">(</span><span class="n">help</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbipsiblastCommandline(cmd=&#39;psiblast&#39;, help=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">psiblast -help</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'psiblast'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.best_hit_overhang">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_overhang</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.best_hit_overhang" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm overhang value (float, recommended value: 0.1)</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.best_hit_score_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_score_edge</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.best_hit_score_edge" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm score edge value (float).</p>
<p>Float between 0.0 and 0.5 inclusive. Recommended value: 0.1</p>
<p>Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.comp_based_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">comp_based_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.comp_based_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use composition-based statistics (string, default 2, i.e. True).</p>
<p>0, F or f: no composition-based statistics</p>
<p>2, T or t, D or d : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties</p>
<p>Note that tblastn also supports values of 1 and 3.</p>
<p>This controls the addition of the -comp_based_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.culling_limit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">culling_limit</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.culling_limit" title="Permalink to this definition"></a></dt>
<dd><p>Hit culling limit (integer).</p>
<p>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit.</p>
<p>Incompatible with: best_hit_overhang, best_hit_score_edge.</p>
<p>This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.gap_trigger">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gap_trigger</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.gap_trigger" title="Permalink to this definition"></a></dt>
<dd><p>Number of bits to trigger gapping (float, default 22).</p>
<p>This controls the addition of the -gap_trigger parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.gapextend" title="Permalink to this definition"></a></dt>
<dd><p>Cost to extend a gap (integer).</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.gapopen" title="Permalink to this definition"></a></dt>
<dd><p>Cost to open a gap (integer).</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.ignore_msa_master">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ignore_msa_master</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.ignore_msa_master" title="Permalink to this definition"></a></dt>
<dd><p>Ignore the master sequence when creating PSSM.</p>
<p>Requires: in_msa
Incompatible with: msa_master_idx, in_pssm, query, query_loc, phi_pattern</p>
<p>This property controls the addition of the -ignore_msa_master switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.in_msa">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">in_msa</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.in_msa" title="Permalink to this definition"></a></dt>
<dd><p>File name of multiple sequence alignment to restart PSI-BLAST.</p>
<p>Incompatible with: in_pssm, query</p>
<p>This controls the addition of the -in_msa parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.in_pssm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">in_pssm</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.in_pssm" title="Permalink to this definition"></a></dt>
<dd><p>PSI-BLAST checkpoint file.</p>
<p>Incompatible with: in_msa, query, phi_pattern</p>
<p>This controls the addition of the -in_pssm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.inclusion_ethresh">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">inclusion_ethresh</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.inclusion_ethresh" title="Permalink to this definition"></a></dt>
<dd><p>E-value inclusion threshold for pairwise alignments (float, default 0.002).</p>
<p>This controls the addition of the -inclusion_ethresh parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.matrix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matrix</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.matrix" title="Permalink to this definition"></a></dt>
<dd><p>Scoring matrix name (default BLOSUM62).</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.msa_master_idx">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">msa_master_idx</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.msa_master_idx" title="Permalink to this definition"></a></dt>
<dd><p>Index of sequence to use as master in MSA.</p>
<p>Index (1-based) of sequence to use as the master in the multiple sequence alignment. If not specified, the first sequence is used.</p>
<p>This controls the addition of the -msa_master_idx parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.num_iterations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_iterations</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.num_iterations" title="Permalink to this definition"></a></dt>
<dd><p>Number of iterations to perform (integer, at least one).</p>
<p>Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_iterations parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.out_ascii_pssm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out_ascii_pssm</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.out_ascii_pssm" title="Permalink to this definition"></a></dt>
<dd><p>File name to store ASCII version of PSSM.</p>
<p>This controls the addition of the -out_ascii_pssm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.out_pssm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out_pssm</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.out_pssm" title="Permalink to this definition"></a></dt>
<dd><p>File name to store checkpoint file.</p>
<p>This controls the addition of the -out_pssm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.phi_pattern">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">phi_pattern</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.phi_pattern" title="Permalink to this definition"></a></dt>
<dd><p>File name containing pattern to search.</p>
<p>Incompatible with: in_pssm</p>
<p>This controls the addition of the -phi_pattern parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.pseudocount">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pseudocount</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.pseudocount" title="Permalink to this definition"></a></dt>
<dd><p>Pseudo-count value used when constructing PSSM.</p>
<p>Integer. Default is zero.</p>
<p>This controls the addition of the -pseudocount parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.save_each_pssm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">save_each_pssm</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.save_each_pssm" title="Permalink to this definition"></a></dt>
<dd><p>Save PSSM after each iteration</p>
<p>File name is given in -save_pssm or -save_ascii_pssm options.</p>
<p>This property controls the addition of the -save_each_pssm switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.save_pssm_after_last_round">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">save_pssm_after_last_round</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.save_pssm_after_last_round" title="Permalink to this definition"></a></dt>
<dd><p>Save PSSM after the last database search.</p>
<p>This property controls the addition of the -save_pssm_after_last_round switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.seg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seg</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.seg" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with SEG (string).</p>
<p>Format: “yes”, “window locut hicut”, or “no” to disable. Default is “12 2.2 2.5”</p>
<p>This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.subject" title="Permalink to this definition"></a></dt>
<dd><p>Subject sequence(s) to search.</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask</p>
<p>See also subject_loc.</p>
<p>This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.subject_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.subject_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the subject sequence (Format: start-stop).</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask, remote.</p>
<p>See also subject.</p>
<p>This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimum score for words to be added to the BLAST lookup table (float).</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.use_sw_tback">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">use_sw_tback</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.use_sw_tback" title="Permalink to this definition"></a></dt>
<dd><p>Compute locally optimal Smith-Waterman alignments?</p>
<p>This property controls the addition of the -use_sw_tback switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbipsiblastCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbipsiblastCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbirpsblastCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rpsblast'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._NcbiblastCommandline</span></code></p>
<p>Wrapper for the NCBI BLAST+ program rpsblast.</p>
<p>With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI
replaced the old rpsblast tool with a similar tool of the same name. This
wrapper replaces RpsBlastCommandline, the wrapper for the old rpsblast.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbirpsblastCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbirpsblastCommandline</span><span class="p">(</span><span class="n">help</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbirpsblastCommandline(cmd=&#39;rpsblast&#39;, help=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">rpsblast -help</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rpsblast'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.best_hit_overhang">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_overhang</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.best_hit_overhang" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm overhang value (recommended value: 0.1).</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.best_hit_score_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_score_edge</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.best_hit_score_edge" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm score edge value (recommended value: 0.1).</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.comp_based_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">comp_based_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.comp_based_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use composition-based statistics.</p>
<p>D or d: default (equivalent to 0)</p>
<p>0 or F or f: Simplified Composition-based statistics as in Bioinformatics 15:1000-1011, 1999</p>
<p>1 or T or t: Composition-based statistics as in NAR 29:2994-3005, 2001</p>
<p>Default = 0.</p>
<p>This controls the addition of the -comp_based_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.culling_limit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">culling_limit</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.culling_limit" title="Permalink to this definition"></a></dt>
<dd><p>Hit culling limit (integer).</p>
<p>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit. Incompatible with: best_hit_overhang, best_hit_score_edge.</p>
<p>This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.seg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seg</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.seg" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with SEG (string).</p>
<p>Format: “yes”, “window locut hicut”, or “no” to disable.Default is “12 2.2 2.5”</p>
<p>This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.use_sw_tback">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">use_sw_tback</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.use_sw_tback" title="Permalink to this definition"></a></dt>
<dd><p>Compute locally optimal Smith-Waterman alignments?</p>
<p>This property controls the addition of the -use_sw_tback switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpsblastCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpsblastCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbirpstblastnCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rpstblastn'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._NcbiblastCommandline</span></code></p>
<p>Wrapper for the NCBI BLAST+ program rpstblastn.</p>
<p>With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI
replaced the old rpsblast tool with a similar tool of the same name, and a
separate tool rpstblastn for Translated Reverse Position Specific BLAST.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbirpstblastnCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbirpstblastnCommandline</span><span class="p">(</span><span class="n">help</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbirpstblastnCommandline(cmd=&#39;rpstblastn&#39;, help=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">rpstblastn -help</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rpstblastn'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.comp_based_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">comp_based_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.comp_based_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use composition-based statistics.</p>
<p>D or d: default (equivalent to 0)</p>
<p>0 or F or f: Simplified Composition-based statistics as in Bioinformatics 15:1000-1011, 1999</p>
<p>1 or T or t: Composition-based statistics as in NAR 29:2994-3005, 2001</p>
<p>Default = 0.</p>
<p>This controls the addition of the -comp_based_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.query_gencode">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_gencode</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.query_gencode" title="Permalink to this definition"></a></dt>
<dd><p>Genetic code to use to translate query (integer, default 1).</p>
<p>This controls the addition of the -query_gencode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.seg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seg</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.seg" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with SEG (string).</p>
<p>Format: “yes”, “window locut hicut”, or “no” to disable. Default is “12 2.2 2.5”</p>
<p>This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.strand">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.strand" title="Permalink to this definition"></a></dt>
<dd><p>Query strand(s) to search against database/subject.</p>
<p>Values allowed are “both” (default), “minus”, “plus”.</p>
<p>This controls the addition of the -strand parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.ungapped">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ungapped</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.ungapped" title="Permalink to this definition"></a></dt>
<dd><p>Perform ungapped alignment only?</p>
<p>This property controls the addition of the -ungapped switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.use_sw_tback">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">use_sw_tback</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.use_sw_tback" title="Permalink to this definition"></a></dt>
<dd><p>Compute locally optimal Smith-Waterman alignments?</p>
<p>This property controls the addition of the -use_sw_tback switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbirpstblastnCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbirpstblastnCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbiblastformatterCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blast_formatter'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._NcbibaseblastCommandline</span></code></p>
<p>Wrapper for the NCBI BLAST+ program blast_formatter.</p>
<p>With the release of BLAST 2.2.24+ (i.e. the BLAST suite rewritten in C++
instead of C), the NCBI added the ASN.1 output format option to all the
search tools, and extended the blast_formatter to support this as input.</p>
<p>The blast_formatter command allows you to convert the ASN.1 output into
the other output formats (XML, tabular, plain text, HTML).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbiblastformatterCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbiblastformatterCommandline</span><span class="p">(</span><span class="n">archive</span><span class="o">=</span><span class="s2">&quot;example.asn&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="s2">&quot;example.xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbiblastformatterCommandline(cmd=&#39;blast_formatter&#39;, out=&#39;example.xml&#39;, outfmt=5, archive=&#39;example.asn&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">blast_formatter -out example.xml -outfmt 5 -archive example.asn</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<p>Note that this wrapper is for the version of blast_formatter from BLAST
2.2.24+ (or later) which is when the NCBI first announced the inclusion
this tool. There was actually an early version in BLAST 2.2.23+ (and
possibly in older releases) but this did not have the -archive option
(instead -rid is a mandatory argument), and is not supported by this
wrapper.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'blast_formatter'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.archive">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">archive</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.archive" title="Permalink to this definition"></a></dt>
<dd><p>Archive file of results, not compatible with rid arg.</p>
<p>This controls the addition of the -archive parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep.</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.rid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rid</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.rid" title="Permalink to this definition"></a></dt>
<dd><p>BLAST Request ID (RID), not compatible with archive arg.</p>
<p>This controls the addition of the -rid parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbiblastformatterCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbiblastformatterCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbideltablastCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'deltablast'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Blast.Applications._Ncbiblast2SeqCommandline</span></code></p>
<p>Create a commandline for the NCBI BLAST+ program deltablast (for proteins).</p>
<p>This is a wrapper for the deltablast command line command included in
the NCBI BLAST+ software (not present in the original BLAST).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbideltablastCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbideltablastCommandline</span><span class="p">(</span><span class="n">query</span><span class="o">=</span><span class="s2">&quot;rosemary.pro&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nr&quot;</span><span class="p">,</span>
<span class="gp">... </span>                              <span class="n">evalue</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">remote</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbideltablastCommandline(cmd=&#39;deltablast&#39;, query=&#39;rosemary.pro&#39;, db=&#39;nr&#39;, evalue=0.001, remote=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">deltablast -query rosemary.pro -db nr -evalue 0.001 -remote</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'deltablast'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.best_hit_overhang">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_overhang</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.best_hit_overhang" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm overhang value (float, recommended value: 0.1)</p>
<p>Float between 0.0 and 0.5 inclusive. Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.best_hit_score_edge">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best_hit_score_edge</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.best_hit_score_edge" title="Permalink to this definition"></a></dt>
<dd><p>Best Hit algorithm score edge value (float).</p>
<p>Float between 0.0 and 0.5 inclusive. Recommended value: 0.1</p>
<p>Incompatible with: culling_limit.</p>
<p>This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.comp_based_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">comp_based_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.comp_based_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use composition-based statistics (string, default 2, i.e. True).</p>
<p>0, F or f: no composition-based statistics.</p>
<p>2, T or t, D or d : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties</p>
<p>Note that tblastn also supports values of 1 and 3.</p>
<p>This controls the addition of the -comp_based_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.culling_limit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">culling_limit</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.culling_limit" title="Permalink to this definition"></a></dt>
<dd><p>Hit culling limit (integer).</p>
<p>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit.</p>
<p>Incompatible with: best_hit_overhang, best_hit_score_edge.</p>
<p>This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">db</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.db" title="Permalink to this definition"></a></dt>
<dd><p>The database to BLAST against.</p>
<p>This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.dbsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbsize</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.dbsize" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the database (integer).</p>
<p>This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.domain_inclusion_ethresh">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">domain_inclusion_ethresh</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.domain_inclusion_ethresh" title="Permalink to this definition"></a></dt>
<dd><p>E-value inclusion threshold for alignments with conserved domains.</p>
<p>(float, Default is 0.05)</p>
<p>This controls the addition of the -domain_inclusion_ethresh parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.entrez_query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">entrez_query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.entrez_query" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search with the given Entrez query (requires remote).</p>
<p>This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.evalue">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">evalue</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.evalue" title="Permalink to this definition"></a></dt>
<dd><p>Expectation value cutoff.</p>
<p>This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.export_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">export_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.export_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>File name to record the search strategy used.</p>
<p>Incompatible with: import_search_strategy</p>
<p>This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.gap_trigger">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gap_trigger</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.gap_trigger" title="Permalink to this definition"></a></dt>
<dd><p>Number of bits to trigger gapping. Default = 22.</p>
<p>This controls the addition of the -gap_trigger parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.gapextend" title="Permalink to this definition"></a></dt>
<dd><p>Cost to extend a gap (integer).</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.gapopen" title="Permalink to this definition"></a></dt>
<dd><p>Cost to open a gap (integer).</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of GI’s.</p>
<p>Incompatible with: negative_gilist, seqidlist, negative_seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION;  ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.html">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">html</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.html" title="Permalink to this definition"></a></dt>
<dd><p>Produce HTML output? See also the outfmt option.</p>
<p>This property controls the addition of the -html switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.import_search_strategy">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">import_search_strategy</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.import_search_strategy" title="Permalink to this definition"></a></dt>
<dd><p>Search strategy to use.</p>
<p>Incompatible with: export_search_strategy</p>
<p>This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.inclusion_ethresh">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">inclusion_ethresh</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.inclusion_ethresh" title="Permalink to this definition"></a></dt>
<dd><p>Pairwise alignment e-value inclusion threshold (float, default 0.002).</p>
<p>This controls the addition of the -inclusion_ethresh parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.lcase_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lcase_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.lcase_masking" title="Permalink to this definition"></a></dt>
<dd><p>Use lower case filtering in query and subject sequence(s)?</p>
<p>This property controls the addition of the -lcase_masking switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.line_length">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">line_length</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.line_length" title="Permalink to this definition"></a></dt>
<dd><p>Line length for formatting alignments (integer, at least 1, default 60).</p>
<p>Not applicable for outfmt &gt; 4. Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.matrix">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matrix</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.matrix" title="Permalink to this definition"></a></dt>
<dd><p>Scoring matrix name (default BLOSUM62).</p>
<p>This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.max_hsps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.max_hsps" title="Permalink to this definition"></a></dt>
<dd><p>Set max number of HSPs saved per subject sequence</p>
<p>Ddefault 0 means no limit.</p>
<p>This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.max_hsps_per_subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_hsps_per_subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.max_hsps_per_subject" title="Permalink to this definition"></a></dt>
<dd><p>Override max number of HSPs per subject saved for ungapped searches (integer).</p>
<p>This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.max_target_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_target_seqs</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.max_target_seqs" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of aligned sequences to keep (integer, at least one).</p>
<p>This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.negative_gilist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_gilist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.negative_gilist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except the listed GIs.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.negative_seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">negative_seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.negative_seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to everything except listed SeqID’s.</p>
<p>Incompatible with: gilist, seqidlist, remote, subject, subject_loc</p>
<p>This controls the addition of the -negative_seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.num_alignments">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_alignments</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.num_alignments" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show num_alignments for.</p>
<p>Integer argument (at least zero). Default is 200. See also num_alignments.</p>
<p>This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.num_descriptions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_descriptions</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.num_descriptions" title="Permalink to this definition"></a></dt>
<dd><p>Number of database sequences to show one-line descriptions for.</p>
<p>Integer argument (at least zero). Default is 500. See also num_alignments.</p>
<p>This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.num_iterations">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_iterations</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.num_iterations" title="Permalink to this definition"></a></dt>
<dd><p>Number of iterations to perform. (integer &gt;=1, Default is 1).</p>
<p>Incompatible with: remote</p>
<p>This controls the addition of the -num_iterations parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.num_threads">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_threads</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.num_threads" title="Permalink to this definition"></a></dt>
<dd><p>Number of threads to use in the BLAST search.</p>
<p>Integer, at least one. Default is one. Incompatible with: remote</p>
<p>This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.out_ascii_pssm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out_ascii_pssm</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.out_ascii_pssm" title="Permalink to this definition"></a></dt>
<dd><p>File name to store ASCII version of PSSM.</p>
<p>This controls the addition of the -out_ascii_pssm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.out_pssm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out_pssm</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.out_pssm" title="Permalink to this definition"></a></dt>
<dd><p>File name to store checkpoint file.</p>
<p>This controls the addition of the -out_pssm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.outfmt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfmt</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.outfmt" title="Permalink to this definition"></a></dt>
<dd><p>Alignment view.  Typically an integer 0-14 but for some formats can be named columns like ‘6 qseqid sseqid’.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).</p>
<p>This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.parse_deflines">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_deflines</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.parse_deflines" title="Permalink to this definition"></a></dt>
<dd><p>Should the query and subject defline(s) be parsed?</p>
<p>This property controls the addition of the -parse_deflines switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.pseudocount">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pseudocount</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.pseudocount" title="Permalink to this definition"></a></dt>
<dd><p>Pseudo-count value used when constructing PSSM (integer, default 0).</p>
<p>This controls the addition of the -pseudocount parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.qcov_hsp_perc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">qcov_hsp_perc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.qcov_hsp_perc" title="Permalink to this definition"></a></dt>
<dd><p>Percent query coverage per hsp (float, 0 to 100).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.query">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.query" title="Permalink to this definition"></a></dt>
<dd><p>The sequence to search with.</p>
<p>This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.query_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">query_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.query_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the query sequence (Format: start-stop).</p>
<p>This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.remote">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remote</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.remote" title="Permalink to this definition"></a></dt>
<dd><p>Execute search remotely?</p>
<p>Incompatible with: gilist, negative_gilist, subject_loc, num_threads, …</p>
<p>This property controls the addition of the -remote switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.rpsdb">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rpsdb</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.rpsdb" title="Permalink to this definition"></a></dt>
<dd><p>BLAST domain database name (dtring, Default = ‘cdd_delta’).</p>
<p>This controls the addition of the -rpsdb parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.save_each_pssm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">save_each_pssm</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.save_each_pssm" title="Permalink to this definition"></a></dt>
<dd><p>Save PSSM after each iteration.</p>
<p>File name is given in -save_pssm or -save_ascii_pssm options.</p>
<p>This property controls the addition of the -save_each_pssm switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.save_pssm_after_last_round">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">save_pssm_after_last_round</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.save_pssm_after_last_round" title="Permalink to this definition"></a></dt>
<dd><p>Save PSSM after the last database search.</p>
<p>This property controls the addition of the -save_pssm_after_last_round switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.searchsp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">searchsp</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.searchsp" title="Permalink to this definition"></a></dt>
<dd><p>Effective length of the search space (integer).</p>
<p>This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.seg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seg</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.seg" title="Permalink to this definition"></a></dt>
<dd><p>Filter query sequence with SEG (string).</p>
<p>Format: “yes”, “window locut hicut”, or “no” to disable. Default is “12 2.2 2.5”</p>
<p>This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.seqidlist">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqidlist</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.seqidlist" title="Permalink to this definition"></a></dt>
<dd><p>Restrict search of database to list of SeqID’s.</p>
<p>Incompatible with: gilist, negative_gilist, remote, subject, subject_loc</p>
<p>This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.show_domain_hits">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_domain_hits</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.show_domain_hits" title="Permalink to this definition"></a></dt>
<dd><p>Show domain hits?</p>
<p>Incompatible with: remote, subject</p>
<p>This property controls the addition of the -show_domain_hits switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.show_gis">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">show_gis</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.show_gis" title="Permalink to this definition"></a></dt>
<dd><p>Show NCBI GIs in deflines?</p>
<p>This property controls the addition of the -show_gis switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.soft_masking">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">soft_masking</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.soft_masking" title="Permalink to this definition"></a></dt>
<dd><p>Apply filtering locations as soft masks (Boolean, Default = true).</p>
<p>This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.subject">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.subject" title="Permalink to this definition"></a></dt>
<dd><p>Subject sequence(s) to search.</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask</p>
<p>See also subject_loc.</p>
<p>This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.subject_loc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subject_loc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.subject_loc" title="Permalink to this definition"></a></dt>
<dd><p>Location on the subject sequence (Format: start-stop).</p>
<p>Incompatible with: db, gilist, seqidlist, negative_gilist, negative_seqidlist, db_soft_mask, db_hard_mask, remote.</p>
<p>See also subject.</p>
<p>This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.sum_statistics">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_statistics</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.sum_statistics" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics.</p>
<p>This property controls the addition of the -sum_statistics switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.sum_stats">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sum_stats</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.sum_stats" title="Permalink to this definition"></a></dt>
<dd><p>Use sum statistics (boolean).</p>
<p>Added in BLAST+ 2.2.30.</p>
<p>This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimum score for words to be added to the BLAST lookup table (float).</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.use_sw_tback">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">use_sw_tback</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.use_sw_tback" title="Permalink to this definition"></a></dt>
<dd><p>Compute locally optimal Smith-Waterman alignments?</p>
<p>This property controls the addition of the -use_sw_tback switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.window_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.window_size" title="Permalink to this definition"></a></dt>
<dd><p>Multiple hits window size, use 0 to specify 1-hit algorithm (integer).</p>
<p>This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.word_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">word_size</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.word_size" title="Permalink to this definition"></a></dt>
<dd><p>Word size for wordfinder algorithm.</p>
<p>Integer. Minimum 2.</p>
<p>This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.xdrop_gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.xdrop_gap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions (float).</p>
<p>This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.xdrop_gap_final">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_gap_final</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.xdrop_gap_final" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment (float).</p>
<p>This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbideltablastCommandline.xdrop_ungap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">xdrop_ungap</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbideltablastCommandline.xdrop_ungap" title="Permalink to this definition"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions (float).</p>
<p>This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Blast.Applications.</span></span><span class="sig-name descname"><span class="pre">NcbimakeblastdbCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'makeblastdb'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Application.html#Bio.Application.AbstractCommandline" title="Bio.Application.AbstractCommandline"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Application.AbstractCommandline</span></code></a></p>
<p>Wrapper for the NCBI BLAST+ program makeblastdb.</p>
<p>This is a wrapper for the NCBI BLAST+ makeblastdb application
to create BLAST databases. By default, this creates a blast database
with the same name as the input file. The default output location
is the same directory as the input.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Blast.Applications</span> <span class="kn">import</span> <span class="n">NcbimakeblastdbCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">NcbimakeblastdbCommandline</span><span class="p">(</span><span class="n">dbtype</span><span class="o">=</span><span class="s2">&quot;prot&quot;</span><span class="p">,</span>
<span class="gp">... </span>                                   <span class="n">input_file</span><span class="o">=</span><span class="s2">&quot;NC_005816.faa&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">NcbimakeblastdbCommandline(cmd=&#39;makeblastdb&#39;, dbtype=&#39;prot&#39;, input_file=&#39;NC_005816.faa&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">makeblastdb -dbtype prot -in NC_005816.faa</span>
</pre></div>
</div>
<p>You would typically run the command line with cline() or via the Python
subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'makeblastdb'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.blastdb_version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">blastdb_version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.blastdb_version" title="Permalink to this definition"></a></dt>
<dd><p>Version of BLAST database to be created. Tip: use BLAST database version 4 on 32 bit CPU. Default = 5</p>
<p>This controls the addition of the -blastdb_version parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.dbtype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dbtype</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.dbtype" title="Permalink to this definition"></a></dt>
<dd><p>Molecule type of target db (‘nucl’ or ‘prot’).</p>
<p>This controls the addition of the -dbtype parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.gi_mask">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gi_mask</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.gi_mask" title="Permalink to this definition"></a></dt>
<dd><p>Create GI indexed masking data.</p>
<p>This property controls the addition of the -gi_mask switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.gi_mask_name">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gi_mask_name</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.gi_mask_name" title="Permalink to this definition"></a></dt>
<dd><p>Comma-separated list of masking data output files.</p>
<p>This controls the addition of the -gi_mask_name parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.h">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">h</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.h" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE and DESCRIPTION; ignore other arguments.</p>
<p>This property controls the addition of the -h switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.hash_index">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">hash_index</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.hash_index" title="Permalink to this definition"></a></dt>
<dd><p>Create index of sequence hash values.</p>
<p>This property controls the addition of the -hash_index switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.help" title="Permalink to this definition"></a></dt>
<dd><p>Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.input_file">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_file</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.input_file" title="Permalink to this definition"></a></dt>
<dd><p>Input file/database name.</p>
<p>This controls the addition of the -in parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.input_type">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">input_type</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.input_type" title="Permalink to this definition"></a></dt>
<dd><p>Type of the data specified in input_file.</p>
<p>Default = ‘fasta’. Added in BLAST 2.2.26.</p>
<p>This controls the addition of the -input_type parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.logfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">logfile</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.logfile" title="Permalink to this definition"></a></dt>
<dd><p>File to which the program log should be redirected.</p>
<p>This controls the addition of the -logfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.mask_data">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mask_data</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.mask_data" title="Permalink to this definition"></a></dt>
<dd><p>Comma-separated list of input files containing masking data as produced by NCBI masking applications (e.g. dustmasker, segmasker, windowmasker).</p>
<p>This controls the addition of the -mask_data parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.mask_desc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mask_desc</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.mask_desc" title="Permalink to this definition"></a></dt>
<dd><p>Comma-separated list of free form strings to describe the masking algorithm details.</p>
<p>This controls the addition of the -mask_desc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.mask_id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mask_id</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.mask_id" title="Permalink to this definition"></a></dt>
<dd><p>Comma-separated list of strings to uniquely identify the masking algorithm.</p>
<p>This controls the addition of the -mask_id parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.max_file_sz">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_file_sz</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.max_file_sz" title="Permalink to this definition"></a></dt>
<dd><p>Maximum file size for BLAST database files. Default = ‘1GB’.</p>
<p>This controls the addition of the -max_file_sz parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.out">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">out</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.out" title="Permalink to this definition"></a></dt>
<dd><p>Output file for alignment.</p>
<p>This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.parse_seqids">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_seqids</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.parse_seqids" title="Permalink to this definition"></a></dt>
<dd><p>Option to parse seqid for FASTA input if set.</p>
<p>For all other input types, seqids are parsed automatically</p>
<p>This property controls the addition of the -parse_seqids switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.taxid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">taxid</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.taxid" title="Permalink to this definition"></a></dt>
<dd><p>Taxonomy ID to assign to all sequences.</p>
<p>This controls the addition of the -taxid parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.taxid_map">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">taxid_map</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.taxid_map" title="Permalink to this definition"></a></dt>
<dd><p>Text file mapping sequence IDs to taxonomy IDs.</p>
<p>Format:&lt;SequenceId&gt; &lt;TaxonomyId&gt;&lt;newline&gt;</p>
<p>This controls the addition of the -taxid_map parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.title">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">title</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.title" title="Permalink to this definition"></a></dt>
<dd><p>Title for BLAST database.</p>
<p>This controls the addition of the -title parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Blast.Applications.NcbimakeblastdbCommandline.version">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">version</span></span><a class="headerlink" href="#Bio.Blast.Applications.NcbimakeblastdbCommandline.version" title="Permalink to this definition"></a></dt>
<dd><p>Print version number;  ignore other arguments.</p>
<p>This property controls the addition of the -version switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Blast.html" class="btn btn-neutral float-left" title="Bio.Blast package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Blast.NCBIWWW.html" class="btn btn-neutral float-right" title="Bio.Blast.NCBIWWW module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2020, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.80.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.78/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>