

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Data.CodonTable module &mdash; Biopython 1.79.dev0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Data.IUPACData module" href="Bio.Data.IUPACData.html" />
    <link rel="prev" title="Bio.Data package" href="Bio.Data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Data.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html#module-Bio.Data">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.Data.html">Bio.Data package</a> &raquo;</li>
        
      <li>Bio.Data.CodonTable module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Data/CodonTable.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Data.IUPACData.html" class="btn btn-neutral float-right" title="Bio.Data.IUPACData module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Data.html" class="btn btn-neutral float-left" title="Bio.Data package" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.Data.CodonTable">
<span id="bio-data-codontable-module"></span><h1>Bio.Data.CodonTable module<a class="headerlink" href="#module-Bio.Data.CodonTable" title="Permalink to this headline">¶</a></h1>
<p>Codon tables based on those from the NCBI.</p>
<p>These tables are based on parsing the NCBI file
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/entrez/misc/data/gc.prt">ftp://ftp.ncbi.nih.gov/entrez/misc/data/gc.prt</a>
using Scripts/update_ncbi_codon_table.py</p>
<p>Last updated at Version 4.4 (May 2019)</p>
<dl class="py exception">
<dt id="Bio.Data.CodonTable.TranslationError">
<em class="property"><span class="pre">exception</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">TranslationError</span></code><a class="headerlink" href="#Bio.Data.CodonTable.TranslationError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Container for translation specific exceptions.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.Data.CodonTable.CodonTable">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">CodonTable</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotide_alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forward_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">back_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_codons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.CodonTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A codon-table, or genetic code.</p>
<dl class="py method">
<dt id="Bio.Data.CodonTable.CodonTable.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nucleotide_alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forward_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">back_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_codons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.CodonTable.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py attribute">
<dt id="Bio.Data.CodonTable.CodonTable.forward_table">
<code class="sig-name descname"><span class="pre">forward_table</span></code><em class="property"> <span class="pre">=</span> <span class="pre">{}</span></em><a class="headerlink" href="#Bio.Data.CodonTable.CodonTable.forward_table" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="Bio.Data.CodonTable.CodonTable.back_table">
<code class="sig-name descname"><span class="pre">back_table</span></code><em class="property"> <span class="pre">=</span> <span class="pre">{}</span></em><a class="headerlink" href="#Bio.Data.CodonTable.CodonTable.back_table" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="Bio.Data.CodonTable.CodonTable.start_codons">
<code class="sig-name descname"><span class="pre">start_codons</span></code><em class="property"> <span class="pre">=</span> <span class="pre">[]</span></em><a class="headerlink" href="#Bio.Data.CodonTable.CodonTable.start_codons" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="Bio.Data.CodonTable.CodonTable.stop_codons">
<code class="sig-name descname"><span class="pre">stop_codons</span></code><em class="property"> <span class="pre">=</span> <span class="pre">[]</span></em><a class="headerlink" href="#Bio.Data.CodonTable.CodonTable.stop_codons" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="Bio.Data.CodonTable.CodonTable.__str__">
<code class="sig-name descname"><span class="pre">__str__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.CodonTable.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a simple text representation of the codon table.</p>
<p>e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Bio.Data.CodonTable</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">Data</span><span class="o">.</span><span class="n">CodonTable</span><span class="o">.</span><span class="n">standard_dna_table</span><span class="p">)</span>
<span class="go">Table 1 Standard, SGC0</span>

<span class="go">  |  T      |  C      |  A      |  G      |</span>
<span class="go">--+---------+---------+---------+---------+--</span>
<span class="go">T | TTT F   | TCT S   | TAT Y   | TGT C   | T</span>
<span class="go">T | TTC F   | TCC S   | TAC Y   | TGC C   | C</span>
<span class="gp">...</span>
<span class="go">G | GTA V   | GCA A   | GAA E   | GGA G   | A</span>
<span class="go">G | GTG V   | GCG A   | GAG E   | GGG G   | G</span>
<span class="go">--+---------+---------+---------+---------+--</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">Data</span><span class="o">.</span><span class="n">CodonTable</span><span class="o">.</span><span class="n">generic_by_id</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">Table 1 Standard, SGC0</span>

<span class="go">  |  U      |  C      |  A      |  G      |</span>
<span class="go">--+---------+---------+---------+---------+--</span>
<span class="go">U | UUU F   | UCU S   | UAU Y   | UGU C   | U</span>
<span class="go">U | UUC F   | UCC S   | UAC Y   | UGC C   | C</span>
<span class="gp">...</span>
<span class="go">G | GUA V   | GCA A   | GAA E   | GGA G   | A</span>
<span class="go">G | GUG V   | GCG A   | GAG E   | GGG G   | G</span>
<span class="go">--+---------+---------+---------+---------+--</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.Data.CodonTable.make_back_table">
<code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">make_back_table</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_stop_codon</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.make_back_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Back a back-table (naive single codon mapping).</p>
<p>ONLY RETURNS A SINGLE CODON, chosen from the possible alternatives
based on their sort order.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.Data.CodonTable.NCBICodonTable">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">NCBICodonTable</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_codons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codons</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Data.CodonTable.CodonTable" title="Bio.Data.CodonTable.CodonTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Data.CodonTable.CodonTable</span></code></a></p>
<p>Codon table for generic nucleotide sequences.</p>
<dl class="py attribute">
<dt id="Bio.Data.CodonTable.NCBICodonTable.nucleotide_alphabet">
<code class="sig-name descname"><span class="pre">nucleotide_alphabet</span></code><em class="property"> <span class="pre">=</span> <span class="pre">None</span></em><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTable.nucleotide_alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt id="Bio.Data.CodonTable.NCBICodonTable.protein_alphabet">
<code class="sig-name descname"><span class="pre">protein_alphabet</span></code><em class="property"> <span class="pre">=</span> <span class="pre">'ACDEFGHIKLMNPQRSTVWY'</span></em><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTable.protein_alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="Bio.Data.CodonTable.NCBICodonTable.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_codons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codons</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTable.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Data.CodonTable.NCBICodonTable.__repr__">
<code class="sig-name descname"><span class="pre">__repr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTable.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the NCBI codon table class as a string for debugging.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Data.CodonTable.NCBICodonTableDNA">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">NCBICodonTableDNA</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_codons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codons</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTableDNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Data.CodonTable.NCBICodonTable" title="Bio.Data.CodonTable.NCBICodonTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Data.CodonTable.NCBICodonTable</span></code></a></p>
<p>Codon table for unambiguous DNA sequences.</p>
<dl class="py attribute">
<dt id="Bio.Data.CodonTable.NCBICodonTableDNA.nucleotide_alphabet">
<code class="sig-name descname"><span class="pre">nucleotide_alphabet</span></code><em class="property"> <span class="pre">=</span> <span class="pre">'GATC'</span></em><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTableDNA.nucleotide_alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Data.CodonTable.NCBICodonTableRNA">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">NCBICodonTableRNA</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_codons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codons</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTableRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Data.CodonTable.NCBICodonTable" title="Bio.Data.CodonTable.NCBICodonTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Data.CodonTable.NCBICodonTable</span></code></a></p>
<p>Codon table for unambiguous RNA sequences.</p>
<dl class="py attribute">
<dt id="Bio.Data.CodonTable.NCBICodonTableRNA.nucleotide_alphabet">
<code class="sig-name descname"><span class="pre">nucleotide_alphabet</span></code><em class="property"> <span class="pre">=</span> <span class="pre">'GAUC'</span></em><a class="headerlink" href="#Bio.Data.CodonTable.NCBICodonTableRNA.nucleotide_alphabet" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Data.CodonTable.AmbiguousCodonTable">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">AmbiguousCodonTable</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_nucleotide_alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_nucleotide_values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_protein_alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_protein_values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.AmbiguousCodonTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Data.CodonTable.CodonTable" title="Bio.Data.CodonTable.CodonTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Data.CodonTable.CodonTable</span></code></a></p>
<p>Base codon table for ambiguous sequences.</p>
<dl class="py method">
<dt id="Bio.Data.CodonTable.AmbiguousCodonTable.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_nucleotide_alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_nucleotide_values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_protein_alphabet</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_protein_values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.AmbiguousCodonTable.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Data.CodonTable.AmbiguousCodonTable.__getattr__">
<code class="sig-name descname"><span class="pre">__getattr__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.AmbiguousCodonTable.__getattr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Forward attribute lookups to the original table.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.Data.CodonTable.list_possible_proteins">
<code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">list_possible_proteins</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forward_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_nucleotide_values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.list_possible_proteins" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all possible encoded amino acids for ambiguous codon.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.Data.CodonTable.list_ambiguous_codons">
<code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">list_ambiguous_codons</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_nucleotide_values</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.list_ambiguous_codons" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend a codon list to include all possible ambigous codons.</p>
<p>e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="s1">&#39;TAG&#39;</span><span class="p">,</span> <span class="s1">&#39;TAA&#39;</span><span class="p">]</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="s1">&#39;TAG&#39;</span><span class="p">,</span> <span class="s1">&#39;TAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TAR&#39;</span><span class="p">]</span>
<span class="p">[</span><span class="s1">&#39;UAG&#39;</span><span class="p">,</span> <span class="s1">&#39;UGA&#39;</span><span class="p">]</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="s1">&#39;UAG&#39;</span><span class="p">,</span> <span class="s1">&#39;UGA&#39;</span><span class="p">,</span> <span class="s1">&#39;URA&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Note that [‘TAG’, ‘TGA’] -&gt; [‘TAG’, ‘TGA’], this does not add ‘TRR’
(which could also mean ‘TAA’ or ‘TGG’).
Thus only two more codons are added in the following:</p>
<p>e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="s1">&#39;TGA&#39;</span><span class="p">,</span> <span class="s1">&#39;TAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TAG&#39;</span><span class="p">]</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="s1">&#39;TGA&#39;</span><span class="p">,</span> <span class="s1">&#39;TAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TAG&#39;</span><span class="p">,</span> <span class="s1">&#39;TRA&#39;</span><span class="p">,</span> <span class="s1">&#39;TAR&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Returns a new (longer) list of codon strings.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.Data.CodonTable.AmbiguousForwardTable">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">AmbiguousForwardTable</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">forward_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_nucleotide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_protein</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.AmbiguousForwardTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Forward table for translation of ambiguous nucleotide sequences.</p>
<dl class="py method">
<dt id="Bio.Data.CodonTable.AmbiguousForwardTable.__init__">
<code class="sig-name descname"><span class="pre">__init__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forward_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_nucleotide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous_protein</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.AmbiguousForwardTable.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Data.CodonTable.AmbiguousForwardTable.__contains__">
<code class="sig-name descname"><span class="pre">__contains__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codon</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.AmbiguousForwardTable.__contains__" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if codon works as key for ambiguous forward_table.</p>
<p>Only returns ‘True’ if forward_table[codon] returns a value.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Data.CodonTable.AmbiguousForwardTable.get">
<code class="sig-name descname"><span class="pre">get</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">failobj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.AmbiguousForwardTable.get" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement get for dictionary-like behaviour.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Data.CodonTable.AmbiguousForwardTable.__getitem__">
<code class="sig-name descname"><span class="pre">__getitem__</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">codon</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.AmbiguousForwardTable.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement dictionary-like behaviour for AmbiguousForwardTable.</p>
<p>forward_table[codon] will either return an amino acid letter,
or throws a KeyError (if codon does not encode an amino acid)
or a TranslationError (if codon does encode for an amino acid,
but either is also a stop codon or does encode several amino acids,
for which no unique letter is available in the given alphabet.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.Data.CodonTable.register_ncbi_table">
<code class="sig-prename descclassname"><span class="pre">Bio.Data.CodonTable.</span></code><code class="sig-name descname"><span class="pre">register_ncbi_table</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_codons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codons</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Data.CodonTable.register_ncbi_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn codon table data into objects (PRIVATE).</p>
<p>The data is stored in the dictionaries.</p>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.Data.IUPACData.html" class="btn btn-neutral float-right" title="Bio.Data.IUPACData module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Data.html" class="btn btn-neutral float-left" title="Bio.Data package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.79.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.77/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>