<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.Align.stockholm module &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.Align.tabular module" href="Bio.Align.tabular.html" />
    <link rel="prev" title="Bio.Align.sam module" href="Bio.Align.sam.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Welcome to Biopython’s API documentation!</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Welcome to Biopython’s API documentation!</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.Align.html">Bio.Align package</a></li>
      <li class="breadcrumb-item active">Bio.Align.stockholm module</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/api/Bio.Align.stockholm.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Align.sam.html" class="btn btn-neutral float-left" title="Bio.Align.sam module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Align.tabular.html" class="btn btn-neutral float-right" title="Bio.Align.tabular module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.Align.stockholm">
<span id="bio-align-stockholm-module"></span><h1>Bio.Align.stockholm module<a class="headerlink" href="#module-Bio.Align.stockholm" title="Permalink to this headline"></a></h1>
<p>Bio.Align support for alignment files in the Stockholm file format.</p>
<p>You are expected to use this module via the Bio.Align functions.</p>
<p>For example, consider this alignment from PFAM for the HAT helix motif:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># STOCKHOLM 1.0</span>
<span class="c1">#=GF ID   HAT</span>
<span class="c1">#=GF AC   PF02184.18</span>
<span class="c1">#=GF DE   HAT (Half-A-TPR) repeat</span>
<span class="c1">#=GF AU   SMART;</span>
<span class="c1">#=GF SE   Alignment kindly provided by SMART</span>
<span class="c1">#=GF GA   21.00 21.00;</span>
<span class="c1">#=GF TC   21.00 21.00;</span>
<span class="c1">#=GF NC   20.90 20.90;</span>
<span class="c1">#=GF BM   hmmbuild HMM.ann SEED.ann</span>
<span class="c1">#=GF SM   hmmsearch -Z 57096847 -E 1000 --cpu 4 HMM pfamseq</span>
<span class="c1">#=GF TP   Repeat</span>
<span class="c1">#=GF CL   CL0020</span>
<span class="c1">#=GF RN   [1]</span>
<span class="c1">#=GF RM   9478129</span>
<span class="c1">#=GF RT   The HAT helix, a repetitive motif implicated in RNA processing.</span>
<span class="c1">#=GF RA   Preker PJ, Keller W;</span>
<span class="c1">#=GF RL   Trends Biochem Sci 1998;23:15-16.</span>
<span class="c1">#=GF DR   INTERPRO; IPR003107;</span>
<span class="c1">#=GF DR   SMART; HAT;</span>
<span class="c1">#=GF DR   SO; 0001068; polypeptide_repeat;</span>
<span class="c1">#=GF CC   The HAT (Half A TPR) repeat is found in several RNA processing</span>
<span class="c1">#=GF CC   proteins [1].</span>
<span class="c1">#=GF SQ   3</span>
<span class="c1">#=GS CRN_DROME/191-222     AC P17886.2</span>
<span class="c1">#=GS CLF1_SCHPO/185-216    AC P87312.1</span>
<span class="c1">#=GS CLF1_SCHPO/185-216    DR PDB; 3JB9 R; 185-216;</span>
<span class="c1">#=GS O16376_CAEEL/201-233  AC O16376.2</span>
<span class="n">CRN_DROME</span><span class="o">/</span><span class="mi">191</span><span class="o">-</span><span class="mi">222</span>                <span class="n">KEIDRAREIYERFVYVH</span><span class="o">.</span><span class="n">PDVKNWIKFARFEES</span>
<span class="n">CLF1_SCHPO</span><span class="o">/</span><span class="mi">185</span><span class="o">-</span><span class="mi">216</span>               <span class="n">HENERARGIYERFVVVH</span><span class="o">.</span><span class="n">PEVTNWLRWARFEEE</span>
<span class="c1">#=GR CLF1_SCHPO/185-216    SS    --HHHHHHHHHHHHHHS.--HHHHHHHHHHHHH</span>
<span class="n">O16376_CAEEL</span><span class="o">/</span><span class="mi">201</span><span class="o">-</span><span class="mi">233</span>             <span class="n">KEIDRARSVYQRFLHVHGINVQNWIKYAKFEER</span>
<span class="c1">#=GC SS_cons                     --HHHHHHHHHHHHHHS.--HHHHHHHHHHHHH</span>
<span class="c1">#=GC seq_cons                    KEIDRARuIYERFVaVH.P-VpNWIKaARFEEc</span>
<span class="o">//</span>
</pre></div>
</div>
<p>Parsing this file using Bio.Align stores the alignment, its annotations, as
well as the sequences and their annotations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">stockholm</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">stockholm</span><span class="o">.</span><span class="n">AlignmentIterator</span><span class="p">(</span><span class="s2">&quot;Stockholm/example.sth&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(3, 33)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;KEIDRAREIYERFVYVH-PDVKNWIKFARFEES&#39;</span>
</pre></div>
</div>
<p>Alignment meta-data are stored in alignment.annotations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span>
<span class="go">&#39;PF02184.18&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;references&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;title&quot;</span><span class="p">]</span>
<span class="go">&#39;The HAT helix, a repetitive motif implicated in RNA processing.&#39;</span>
</pre></div>
</div>
<p>Annotations of alignment columns are stored in alignment.column_annotations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">column_annotations</span><span class="p">[</span><span class="s2">&quot;consensus secondary structure&quot;</span><span class="p">]</span>
<span class="go">&#39;--HHHHHHHHHHHHHHS.--HHHHHHHHHHHHH&#39;</span>
</pre></div>
</div>
<p>Sequences and their annotations are stored in alignment.sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;CRN_DROME/191-222&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;KEIDRAREIYERFVYVHPDVKNWIKFARFEES&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s2">&quot;secondary structure&quot;</span><span class="p">]</span>
<span class="go">&#39;--HHHHHHHHHHHHHHS--HHHHHHHHHHHHH&#39;</span>
</pre></div>
</div>
<p>Slicing specific columns of an alignment will slice any per-column-annotations:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">column_annotations</span><span class="p">[</span><span class="s2">&quot;consensus secondary structure&quot;</span><span class="p">]</span>
<span class="go">&#39;--HHHHHHHHHHHHHHS.--HHHHHHHHHHHHH&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">part_alignment</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[:,</span><span class="mi">10</span><span class="p">:</span><span class="mi">20</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">part_alignment</span><span class="o">.</span><span class="n">column_annotations</span><span class="p">[</span><span class="s2">&quot;consensus secondary structure&quot;</span><span class="p">]</span>
<span class="go">&#39;HHHHHHS.--&#39;</span>
</pre></div>
</div>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.stockholm.</span></span><span class="sig-name descname"><span class="pre">AlignmentIterator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Align.interfaces.html#Bio.Align.interfaces.AlignmentIterator" title="Bio.Align.interfaces.AlignmentIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Align.interfaces.AlignmentIterator</span></code></a></p>
<p>Alignment iterator for alignment files in the Stockholm format.</p>
<p>The file may contain multiple concatenated alignments, which are loaded
and returned incrementally.</p>
<p>Alignment meta-data (lines starting with #=GF) are stored in the dictionary
alignment.annotations. Column annotations (lines starting with #=GC) are
stored in the dictionary alignment.column_annotations. Sequence names are
stored in record.id. Sequence record meta-data (lines starting with #=GS)
are stored in the dictionary record.annotations. Sequence letter
annotations (lines starting with #=GR) are stored in the dictionary
record.letter_annotations.</p>
<p>Wrap-around alignments are not supported - each sequence must be on
a single line.</p>
<p>For more information on the file format, please see:
<a class="reference external" href="http://sonnhammer.sbc.su.se/Stockholm.html">http://sonnhammer.sbc.su.se/Stockholm.html</a>
<a class="reference external" href="https://en.wikipedia.org/wiki/Stockholm_format">https://en.wikipedia.org/wiki/Stockholm_format</a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.fmt">
<span class="sig-name descname"><span class="pre">fmt</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'Stockholm'</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.fmt" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.gf_mapping">
<span class="sig-name descname"><span class="pre">gf_mapping</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'**':</span> <span class="pre">'**',</span> <span class="pre">'AC':</span> <span class="pre">'accession',</span> <span class="pre">'AU':</span> <span class="pre">'author',</span> <span class="pre">'BM':</span> <span class="pre">'build</span> <span class="pre">method',</span> <span class="pre">'CB':</span> <span class="pre">'calibration</span> <span class="pre">method',</span> <span class="pre">'CC':</span> <span class="pre">'comment',</span> <span class="pre">'CL':</span> <span class="pre">'clan',</span> <span class="pre">'DE':</span> <span class="pre">'definition',</span> <span class="pre">'GA':</span> <span class="pre">'gathering</span> <span class="pre">method',</span> <span class="pre">'ID':</span> <span class="pre">'identifier',</span> <span class="pre">'NC':</span> <span class="pre">'noise</span> <span class="pre">cutoff',</span> <span class="pre">'PI':</span> <span class="pre">'previous</span> <span class="pre">identifier',</span> <span class="pre">'SE':</span> <span class="pre">'source</span> <span class="pre">of</span> <span class="pre">seed',</span> <span class="pre">'SM':</span> <span class="pre">'search</span> <span class="pre">method',</span> <span class="pre">'SS':</span> <span class="pre">'source</span> <span class="pre">of</span> <span class="pre">structure',</span> <span class="pre">'TC':</span> <span class="pre">'trusted</span> <span class="pre">cutoff',</span> <span class="pre">'TP':</span> <span class="pre">'type',</span> <span class="pre">'WK':</span> <span class="pre">'wikipedia'}</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.gf_mapping" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.gr_mapping">
<span class="sig-name descname"><span class="pre">gr_mapping</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'AS':</span> <span class="pre">'active</span> <span class="pre">site',</span> <span class="pre">'CSA':</span> <span class="pre">'Catalytic</span> <span class="pre">Site</span> <span class="pre">Atlas',</span> <span class="pre">'IN':</span> <span class="pre">'intron',</span> <span class="pre">'LI':</span> <span class="pre">'ligand</span> <span class="pre">binding',</span> <span class="pre">'PP':</span> <span class="pre">'posterior</span> <span class="pre">probability',</span> <span class="pre">'SA':</span> <span class="pre">'surface</span> <span class="pre">accessibility',</span> <span class="pre">'SS':</span> <span class="pre">'secondary</span> <span class="pre">structure',</span> <span class="pre">'TM':</span> <span class="pre">'transmembrane',</span> <span class="pre">'pAS':</span> <span class="pre">'active</span> <span class="pre">site</span> <span class="pre">-</span> <span class="pre">Pfam</span> <span class="pre">predicted',</span> <span class="pre">'sAS':</span> <span class="pre">'active</span> <span class="pre">site</span> <span class="pre">-</span> <span class="pre">from</span> <span class="pre">SwissProt'}</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.gr_mapping" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.gc_mapping">
<span class="sig-name descname"><span class="pre">gc_mapping</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'2L3J_B_SS':</span> <span class="pre">'2L3J</span> <span class="pre">B</span> <span class="pre">SS',</span> <span class="pre">'AS_cons':</span> <span class="pre">'consensus</span> <span class="pre">active</span> <span class="pre">site',</span> <span class="pre">'CORE':</span> <span class="pre">'CORE',</span> <span class="pre">'CSA_cons':</span> <span class="pre">'consensus</span> <span class="pre">Catalytic</span> <span class="pre">Site</span> <span class="pre">Atlas',</span> <span class="pre">'IN_cons':</span> <span class="pre">'consensus</span> <span class="pre">intron',</span> <span class="pre">'LI_cons':</span> <span class="pre">'consensus</span> <span class="pre">ligand</span> <span class="pre">binding',</span> <span class="pre">'MM':</span> <span class="pre">'model</span> <span class="pre">mask',</span> <span class="pre">'PK':</span> <span class="pre">'PK',</span> <span class="pre">'PK_SS':</span> <span class="pre">'PK</span> <span class="pre">SS',</span> <span class="pre">'PP_cons':</span> <span class="pre">'consensus</span> <span class="pre">posterior</span> <span class="pre">probability',</span> <span class="pre">'RF':</span> <span class="pre">'reference</span> <span class="pre">coordinate</span> <span class="pre">annotation',</span> <span class="pre">'RNA_elements':</span> <span class="pre">'RNA</span> <span class="pre">elements',</span> <span class="pre">'RNA_ligand_AdoCbl':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">AdoCbl',</span> <span class="pre">'RNA_ligand_AqCbl':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">AqCbl',</span> <span class="pre">'RNA_ligand_FMN':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">FMN',</span> <span class="pre">'RNA_ligand_Guanidinium':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">Guanidinium',</span> <span class="pre">'RNA_ligand_SAM':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">SAM',</span> <span class="pre">'RNA_ligand_THF_1':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">THF</span> <span class="pre">1',</span> <span class="pre">'RNA_ligand_THF_2':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">THF</span> <span class="pre">2',</span> <span class="pre">'RNA_ligand_TPP':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">TPP',</span> <span class="pre">'RNA_ligand_preQ1':</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">preQ1',</span> <span class="pre">'RNA_motif_k_turn':</span> <span class="pre">'RNA</span> <span class="pre">motif</span> <span class="pre">k</span> <span class="pre">turn',</span> <span class="pre">'RNA_structural_element':</span> <span class="pre">'RNA</span> <span class="pre">structural</span> <span class="pre">element',</span> <span class="pre">'RNA_structural_elements':</span> <span class="pre">'RNA</span> <span class="pre">structural</span> <span class="pre">elements',</span> <span class="pre">'Repeat_unit':</span> <span class="pre">'Repeat</span> <span class="pre">unit',</span> <span class="pre">'SA_cons':</span> <span class="pre">'consensus</span> <span class="pre">surface</span> <span class="pre">accessibility',</span> <span class="pre">'SS_cons':</span> <span class="pre">'consensus</span> <span class="pre">secondary</span> <span class="pre">structure',</span> <span class="pre">'TM_cons':</span> <span class="pre">'consensus</span> <span class="pre">transmembrane',</span> <span class="pre">'cons':</span> <span class="pre">'cons',</span> <span class="pre">'pAS_cons':</span> <span class="pre">'consensus</span> <span class="pre">active</span> <span class="pre">site</span> <span class="pre">-</span> <span class="pre">Pfam</span> <span class="pre">predicted',</span> <span class="pre">'sAS_cons':</span> <span class="pre">'consensus</span> <span class="pre">active</span> <span class="pre">site</span> <span class="pre">-</span> <span class="pre">from</span> <span class="pre">SwissProt',</span> <span class="pre">'scorecons':</span> <span class="pre">'consensus</span> <span class="pre">score',</span> <span class="pre">'scorecons_70':</span> <span class="pre">'consensus</span> <span class="pre">score</span> <span class="pre">70',</span> <span class="pre">'scorecons_80':</span> <span class="pre">'consensus</span> <span class="pre">score</span> <span class="pre">80',</span> <span class="pre">'scorecons_90':</span> <span class="pre">'consensus</span> <span class="pre">score</span> <span class="pre">90',</span> <span class="pre">'seq_cons':</span> <span class="pre">'consensus</span> <span class="pre">sequence'}</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.gc_mapping" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.gs_mapping">
<span class="sig-name descname"><span class="pre">gs_mapping</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'AC':</span> <span class="pre">'accession',</span> <span class="pre">'LO':</span> <span class="pre">'look',</span> <span class="pre">'OC':</span> <span class="pre">'organism</span> <span class="pre">classification',</span> <span class="pre">'OS':</span> <span class="pre">'organism'}</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.gs_mapping" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.key">
<span class="sig-name descname"><span class="pre">key</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'IN'</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.key" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.keyword">
<span class="sig-name descname"><span class="pre">keyword</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'cons'</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.keyword" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentIterator.value">
<span class="sig-name descname"><span class="pre">value</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'intron'</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentIterator.value" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentWriter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Align.stockholm.</span></span><span class="sig-name descname"><span class="pre">AlignmentWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.stockholm.AlignmentWriter" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Align.interfaces.html#Bio.Align.interfaces.AlignmentWriter" title="Bio.Align.interfaces.AlignmentWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Align.interfaces.AlignmentWriter</span></code></a></p>
<p>Alignment file writer for the Stockholm file format.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentWriter.gf_mapping">
<span class="sig-name descname"><span class="pre">gf_mapping</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'**':</span> <span class="pre">'**',</span> <span class="pre">'accession':</span> <span class="pre">'AC',</span> <span class="pre">'author':</span> <span class="pre">'AU',</span> <span class="pre">'build</span> <span class="pre">method':</span> <span class="pre">'BM',</span> <span class="pre">'calibration</span> <span class="pre">method':</span> <span class="pre">'CB',</span> <span class="pre">'clan':</span> <span class="pre">'CL',</span> <span class="pre">'comment':</span> <span class="pre">'CC',</span> <span class="pre">'definition':</span> <span class="pre">'DE',</span> <span class="pre">'gathering</span> <span class="pre">method':</span> <span class="pre">'GA',</span> <span class="pre">'identifier':</span> <span class="pre">'ID',</span> <span class="pre">'noise</span> <span class="pre">cutoff':</span> <span class="pre">'NC',</span> <span class="pre">'previous</span> <span class="pre">identifier':</span> <span class="pre">'PI',</span> <span class="pre">'search</span> <span class="pre">method':</span> <span class="pre">'SM',</span> <span class="pre">'source</span> <span class="pre">of</span> <span class="pre">seed':</span> <span class="pre">'SE',</span> <span class="pre">'source</span> <span class="pre">of</span> <span class="pre">structure':</span> <span class="pre">'SS',</span> <span class="pre">'trusted</span> <span class="pre">cutoff':</span> <span class="pre">'TC',</span> <span class="pre">'type':</span> <span class="pre">'TP',</span> <span class="pre">'wikipedia':</span> <span class="pre">'WK'}</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentWriter.gf_mapping" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentWriter.gs_mapping">
<span class="sig-name descname"><span class="pre">gs_mapping</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'accession':</span> <span class="pre">'AC',</span> <span class="pre">'look':</span> <span class="pre">'LO',</span> <span class="pre">'organism':</span> <span class="pre">'OS',</span> <span class="pre">'organism</span> <span class="pre">classification':</span> <span class="pre">'OC'}</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentWriter.gs_mapping" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentWriter.gr_mapping">
<span class="sig-name descname"><span class="pre">gr_mapping</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'Catalytic</span> <span class="pre">Site</span> <span class="pre">Atlas':</span> <span class="pre">'CSA',</span> <span class="pre">'active</span> <span class="pre">site':</span> <span class="pre">'AS',</span> <span class="pre">'active</span> <span class="pre">site</span> <span class="pre">-</span> <span class="pre">Pfam</span> <span class="pre">predicted':</span> <span class="pre">'pAS',</span> <span class="pre">'active</span> <span class="pre">site</span> <span class="pre">-</span> <span class="pre">from</span> <span class="pre">SwissProt':</span> <span class="pre">'sAS',</span> <span class="pre">'intron':</span> <span class="pre">'IN',</span> <span class="pre">'ligand</span> <span class="pre">binding':</span> <span class="pre">'LI',</span> <span class="pre">'posterior</span> <span class="pre">probability':</span> <span class="pre">'PP',</span> <span class="pre">'secondary</span> <span class="pre">structure':</span> <span class="pre">'SS',</span> <span class="pre">'surface</span> <span class="pre">accessibility':</span> <span class="pre">'SA',</span> <span class="pre">'transmembrane':</span> <span class="pre">'TM'}</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentWriter.gr_mapping" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentWriter.gc_mapping">
<span class="sig-name descname"><span class="pre">gc_mapping</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'2L3J</span> <span class="pre">B</span> <span class="pre">SS':</span> <span class="pre">'2L3J_B_SS',</span> <span class="pre">'CORE':</span> <span class="pre">'CORE',</span> <span class="pre">'PK':</span> <span class="pre">'PK',</span> <span class="pre">'PK</span> <span class="pre">SS':</span> <span class="pre">'PK_SS',</span> <span class="pre">'RNA</span> <span class="pre">elements':</span> <span class="pre">'RNA_elements',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">AdoCbl':</span> <span class="pre">'RNA_ligand_AdoCbl',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">AqCbl':</span> <span class="pre">'RNA_ligand_AqCbl',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">FMN':</span> <span class="pre">'RNA_ligand_FMN',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">Guanidinium':</span> <span class="pre">'RNA_ligand_Guanidinium',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">SAM':</span> <span class="pre">'RNA_ligand_SAM',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">THF</span> <span class="pre">1':</span> <span class="pre">'RNA_ligand_THF_1',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">THF</span> <span class="pre">2':</span> <span class="pre">'RNA_ligand_THF_2',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">TPP':</span> <span class="pre">'RNA_ligand_TPP',</span> <span class="pre">'RNA</span> <span class="pre">ligand</span> <span class="pre">preQ1':</span> <span class="pre">'RNA_ligand_preQ1',</span> <span class="pre">'RNA</span> <span class="pre">motif</span> <span class="pre">k</span> <span class="pre">turn':</span> <span class="pre">'RNA_motif_k_turn',</span> <span class="pre">'RNA</span> <span class="pre">structural</span> <span class="pre">element':</span> <span class="pre">'RNA_structural_element',</span> <span class="pre">'RNA</span> <span class="pre">structural</span> <span class="pre">elements':</span> <span class="pre">'RNA_structural_elements',</span> <span class="pre">'Repeat</span> <span class="pre">unit':</span> <span class="pre">'Repeat_unit',</span> <span class="pre">'cons':</span> <span class="pre">'cons',</span> <span class="pre">'consensus</span> <span class="pre">Catalytic</span> <span class="pre">Site</span> <span class="pre">Atlas':</span> <span class="pre">'CSA_cons',</span> <span class="pre">'consensus</span> <span class="pre">active</span> <span class="pre">site':</span> <span class="pre">'AS_cons',</span> <span class="pre">'consensus</span> <span class="pre">active</span> <span class="pre">site</span> <span class="pre">-</span> <span class="pre">Pfam</span> <span class="pre">predicted':</span> <span class="pre">'pAS_cons',</span> <span class="pre">'consensus</span> <span class="pre">active</span> <span class="pre">site</span> <span class="pre">-</span> <span class="pre">from</span> <span class="pre">SwissProt':</span> <span class="pre">'sAS_cons',</span> <span class="pre">'consensus</span> <span class="pre">intron':</span> <span class="pre">'IN_cons',</span> <span class="pre">'consensus</span> <span class="pre">ligand</span> <span class="pre">binding':</span> <span class="pre">'LI_cons',</span> <span class="pre">'consensus</span> <span class="pre">posterior</span> <span class="pre">probability':</span> <span class="pre">'PP_cons',</span> <span class="pre">'consensus</span> <span class="pre">score':</span> <span class="pre">'scorecons',</span> <span class="pre">'consensus</span> <span class="pre">score</span> <span class="pre">70':</span> <span class="pre">'scorecons_70',</span> <span class="pre">'consensus</span> <span class="pre">score</span> <span class="pre">80':</span> <span class="pre">'scorecons_80',</span> <span class="pre">'consensus</span> <span class="pre">score</span> <span class="pre">90':</span> <span class="pre">'scorecons_90',</span> <span class="pre">'consensus</span> <span class="pre">secondary</span> <span class="pre">structure':</span> <span class="pre">'SS_cons',</span> <span class="pre">'consensus</span> <span class="pre">sequence':</span> <span class="pre">'seq_cons',</span> <span class="pre">'consensus</span> <span class="pre">surface</span> <span class="pre">accessibility':</span> <span class="pre">'SA_cons',</span> <span class="pre">'consensus</span> <span class="pre">transmembrane':</span> <span class="pre">'TM_cons',</span> <span class="pre">'model</span> <span class="pre">mask':</span> <span class="pre">'MM',</span> <span class="pre">'reference</span> <span class="pre">coordinate</span> <span class="pre">annotation':</span> <span class="pre">'RF'}</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentWriter.gc_mapping" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentWriter.fmt">
<span class="sig-name descname"><span class="pre">fmt</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'Stockholm'</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentWriter.fmt" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentWriter.format_alignment">
<span class="sig-name descname"><span class="pre">format_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Align.stockholm.AlignmentWriter.format_alignment" title="Permalink to this definition"></a></dt>
<dd><p>Return a string with a single alignment in the Stockholm format.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Align.stockholm.AlignmentWriter.__abstractmethods__">
<span class="sig-name descname"><span class="pre">__abstractmethods__</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">frozenset({})</span></em><a class="headerlink" href="#Bio.Align.stockholm.AlignmentWriter.__abstractmethods__" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Align.sam.html" class="btn btn-neutral float-left" title="Bio.Align.sam module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Align.tabular.html" class="btn btn-neutral float-right" title="Bio.Align.tabular module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>