<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.Emboss.Applications module &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.Emboss.Primer3 module" href="Bio.Emboss.Primer3.html" />
    <link rel="prev" title="Bio.Emboss package" href="Bio.Emboss.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Tutorial/index.html">Biopython Tutorial &amp; Cookbook</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">API documentation</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">API documentation</a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.Emboss.html">Bio.Emboss package</a></li>
      <li class="breadcrumb-item active">Bio.Emboss.Applications module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Emboss/Applications.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.Emboss.html" class="btn btn-neutral float-left" title="Bio.Emboss package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Emboss.Primer3.html" class="btn btn-neutral float-right" title="Bio.Emboss.Primer3 module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.Emboss.Applications">
<span id="bio-emboss-applications-module"></span><h1>Bio.Emboss.Applications module<a class="headerlink" href="#module-Bio.Emboss.Applications" title="Permalink to this headline">ÔÉÅ</a></h1>
<p>Code to interact with and run various EMBOSS programs (OBSOLETE).</p>
<p>These classes follow the AbstractCommandline interfaces for running
programs.</p>
<p>We have decided to remove this module in future, and instead recommend
building your command and invoking it via the subprocess module directly.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">Primer3Commandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'eprimer3'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the Primer3 interface from EMBOSS.</p>
<p>The precise set of supported arguments depends on your version of EMBOSS.
This version accepts arguments current at EMBOSS 6.1.0:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">Primer3Commandline</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;mysequence.fas&quot;</span><span class="p">,</span> <span class="n">auto</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">hybridprobe</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">explainflag</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">osizeopt</span><span class="o">=</span><span class="mi">20</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">psizeopt</span><span class="o">=</span><span class="mi">200</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;myresults.out&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">bogusparameter</span> <span class="o">=</span> <span class="mi">1967</span>  <span class="c1"># Invalid parameter</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">Option name bogusparameter was not found.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">eprimer3 -auto -outfile=myresults.out -sequence=mysequence.fas -hybridprobe=True -psizeopt=200 -osizeopt=20 -explainflag=True</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'eprimer3'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.dnaconc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dnaconc</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.dnaconc" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Nanomolar concentration of annealing oligos in the PCR.</p>
<p>This controls the addition of the -dnaconc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.excludedregion">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">excludedregion</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.excludedregion" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Regions to exclude from primer picking.</p>
<p>This controls the addition of the -excludedregion parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.explainflag">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">explainflag</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.explainflag" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Produce output tags with eprimer3 statistics</p>
<p>This controls the addition of the -explainflag parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.forwardinput">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">forwardinput</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.forwardinput" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence of a forward primer to check.</p>
<p>This controls the addition of the -forwardinput parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.gcclamp">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gcclamp</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.gcclamp" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>The required number of Gs and Cs at the 3‚Äô of each primer.</p>
<p>This controls the addition of the -gcclamp parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.hybridprobe">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">hybridprobe</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.hybridprobe" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Find an internal oligo to use as a hyb probe.</p>
<p>This controls the addition of the -hybridprobe parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.includedregion">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">includedregion</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.includedregion" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Subregion of the sequence in which to pick primers.</p>
<p>This controls the addition of the -includedregion parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.maxdifftm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxdifftm</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.maxdifftm" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum difference in melting temperatures between forward and reverse primers.</p>
<p>This controls the addition of the -maxdifftm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.maxgc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxgc</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.maxgc" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum GC% for a primer.</p>
<p>This controls the addition of the -maxgc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.maxmispriming">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxmispriming</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.maxmispriming" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum allowed similarity of primers to sequences in library specified by -mispriminglibrary</p>
<p>This controls the addition of the -maxmispriming parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.maxpolyx">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxpolyx</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.maxpolyx" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum allowable mononucleotide repeat length in a primer.</p>
<p>This controls the addition of the -maxpolyx parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.maxsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxsize</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.maxsize" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum length of a primer oligo.</p>
<p>This controls the addition of the -maxsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.maxtm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxtm</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.maxtm" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum melting temperature for a primer oligo.</p>
<p>This controls the addition of the -maxtm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.mingc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mingc</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.mingc" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum GC% for a primer.</p>
<p>This controls the addition of the -mingc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.minsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">minsize</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.minsize" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum length of a primer oligo.</p>
<p>This controls the addition of the -minsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.mintm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mintm</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.mintm" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum melting temperature for a primer oligo.</p>
<p>This controls the addition of the -mintm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.mishyblibraryfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mishyblibraryfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.mishyblibraryfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Library file of seqs to avoid internal oligo hybridisation.</p>
<p>This controls the addition of the -mishyblibraryfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.mispriminglibraryfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mispriminglibraryfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.mispriminglibraryfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>File containing library of sequences to avoid amplifying</p>
<p>This controls the addition of the -mispriminglibraryfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.numreturn">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">numreturn</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.numreturn" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum number of primer pairs to return.</p>
<p>This controls the addition of the -numreturn parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.oanyself">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">oanyself</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.oanyself" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum allowable alignment score for self-complementarity.</p>
<p>This controls the addition of the -oanyself parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.odnaconc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">odnaconc</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.odnaconc" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Nanomolar concentration of internal oligo in the hybridisation.</p>
<p>This controls the addition of the -odnaconc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.oendself">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">oendself</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.oendself" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Max 3‚Äô-anchored self-complementarity global alignment score.</p>
<p>This controls the addition of the -oendself parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.oexcludedregion">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">oexcludedregion</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.oexcludedregion" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Do not pick internal oligos in this region.</p>
<p>This controls the addition of the -oexcludedregion parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.ogcmax">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ogcmax</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.ogcmax" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum GC% for internal oligo.</p>
<p>This controls the addition of the -ogcmax parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.ogcmin">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ogcmin</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.ogcmin" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum GC% for internal oligo.</p>
<p>This controls the addition of the -ogcmin parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.ogcopt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ogcopt</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.ogcopt" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Optimum GC% for internal oligo.</p>
<p>This controls the addition of the -ogcopt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.ogcpercent">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ogcpercent</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.ogcpercent" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Optimum GC% for a primer.</p>
<p>This controls the addition of the -ogcpercent parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.oligoinput">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">oligoinput</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.oligoinput" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence of the internal oligo.</p>
<p>This controls the addition of the -oligoinput parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.omaxsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">omaxsize</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.omaxsize" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum length of internal oligo.</p>
<p>This controls the addition of the -omaxsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.ominsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ominsize</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.ominsize" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum length of internal oligo.</p>
<p>This controls the addition of the -ominsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.omishybmax">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">omishybmax</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.omishybmax" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum alignment score for hybridisation of internal oligo to library specified by -mishyblibraryfile.</p>
<p>This controls the addition of the -omishybmax parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.opolyxmax">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">opolyxmax</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.opolyxmax" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum length of mononucleotide repeat in internal oligo.</p>
<p>This controls the addition of the -opolyxmax parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.opttm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">opttm</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.opttm" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Optimum melting temperature for a primer oligo.</p>
<p>Option added in EMBOSS 6.6.0, replacing -otm</p>
<p>This controls the addition of the -opttm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.osaltconc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">osaltconc</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.osaltconc" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Millimolar concentration of salt in the hybridisation.</p>
<p>This controls the addition of the -osaltconc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.osize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">osize</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.osize" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Optimum length of a primer oligo.</p>
<p>This controls the addition of the -osize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.osizeopt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">osizeopt</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.osizeopt" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Optimum length of internal oligo.</p>
<p>This controls the addition of the -osizeopt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.otm">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">otm</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.otm" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Melting temperature for primer oligo (OBSOLETE).</p>
<p>Option replaced in EMBOSS 6.6.0 by -opttm</p>
<p>This controls the addition of the -otm parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.otmmax">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">otmmax</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.otmmax" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum melting temperature of internal oligo.</p>
<p>This controls the addition of the -otmmax parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.otmmin">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">otmmin</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.otmmin" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum melting temperature of internal oligo.</p>
<p>This controls the addition of the -otmmin parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.otmopt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">otmopt</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.otmopt" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Optimum melting temperature of internal oligo.</p>
<p>This controls the addition of the -otmopt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.prange">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prange</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.prange" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Acceptable range of length for the PCR product.</p>
<p>This controls the addition of the -prange parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.psizeopt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">psizeopt</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.psizeopt" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Optimum size for the PCR product.</p>
<p>This controls the addition of the -psizeopt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.ptmmax">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ptmmax</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.ptmmax" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum allowed melting temperature for the amplicon.</p>
<p>This controls the addition of the -ptmmax parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.ptmmin">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ptmmin</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.ptmmin" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum allowed melting temperature for the amplicon.</p>
<p>This controls the addition of the -ptmmin parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.ptmopt">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ptmopt</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.ptmopt" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Optimum melting temperature for the PCR product.</p>
<p>This controls the addition of the -ptmopt parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.reverseinput">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reverseinput</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.reverseinput" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence of a reverse primer to check.</p>
<p>This controls the addition of the -reverseinput parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.saltconc">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">saltconc</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.saltconc" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Millimolar salt concentration in the PCR.</p>
<p>This controls the addition of the -saltconc parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence to choose primers from.</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.target">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">target</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.target" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence to target for flanking primers.</p>
<p>This controls the addition of the -target parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.task">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">task</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.task" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Tell eprimer3 what task to perform.</p>
<p>This controls the addition of the -task parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Primer3Commandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Primer3Commandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">PrimerSearchCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'primersearch'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the primersearch program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'primersearch'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.infile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">infile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.infile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>File containing the primer pairs to search for.</p>
<p>This controls the addition of the -infile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.mismatchpercent">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatchpercent</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.mismatchpercent" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Allowed percentage mismatch (any integer value, default 0).</p>
<p>This controls the addition of the -mismatchpercent parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.seqall">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqall</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.seqall" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence to look for the primer pairs in.</p>
<p>This controls the addition of the -seqall parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.snucleotide">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">snucleotide</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.snucleotide" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are nucleotide (boolean)</p>
<p>This controls the addition of the -snucleotide parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.sprotein">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sprotein</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.sprotein" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are protein (boolean)</p>
<p>This controls the addition of the -sprotein parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PrimerSearchCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PrimerSearchCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FDNADistCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fdnadist'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fdnadist program from EMBOSS.</p>
<p>fdnadist is an EMBOSS wrapper for the PHYLIP program dnadist for
calculating distance matrices from DNA sequence files.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fdnadist'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.basefreq">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">basefreq</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.basefreq" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>specify basefreqs</p>
<p>This controls the addition of the -basefreq parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.categories">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">categories</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.categories" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>File of substitution rate categories</p>
<p>This controls the addition of the -categories parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.freqsfrom">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">freqsfrom</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.freqsfrom" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>use emprical base freqs</p>
<p>This controls the addition of the -freqsfrom parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.gamma">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gamma</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.gamma" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>gamma [g, i,n]</p>
<p>This controls the addition of the -gamma parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.gammacoefficient">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gammacoefficient</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.gammacoefficient" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>value for gamma (&gt; 0.001)</p>
<p>This controls the addition of the -gammacoefficient parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.invarfrac">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">invarfrac</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.invarfrac" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>proportoin of invariant sites</p>
<p>This controls the addition of the -invarfrac parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.lower">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lower</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.lower" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>lower triangle matrix (y/N)</p>
<p>This controls the addition of the -lower parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.method">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">method</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.method" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>sub. model [f,k,j,l,s]</p>
<p>This controls the addition of the -method parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.ncategories">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ncategories</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.ncategories" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>number of rate catergories (1-9)</p>
<p>This controls the addition of the -ncategories parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.rate">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rate</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.rate" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>rate for each category</p>
<p>This controls the addition of the -rate parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>seq file to use (phylip)</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.ttratio">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ttratio</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.ttratio" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>ts/tv ratio</p>
<p>This controls the addition of the -ttratio parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNADistCommandline.weights">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">weights</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNADistCommandline.weights" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>weights file</p>
<p>This controls the addition of the -weights parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FTreeDistCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ftreedist'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the ftreedist program from EMBOSS.</p>
<p>ftreedist is an EMBOSS wrapper for the PHYLIP program treedist used for
calculating distance measures between phylogentic trees.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ftreedist'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.dtype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dtype</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.dtype" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>distance type ([S]ymetric, [b]ranch score)</p>
<p>This controls the addition of the -dtype parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.intreefile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">intreefile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.intreefile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>tree file to score (phylip)</p>
<p>This controls the addition of the -intreefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.noroot">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">noroot</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.noroot" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>treat trees as rooted [N/y]</p>
<p>This controls the addition of the -noroot parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.outgrno">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outgrno</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.outgrno" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>which taxon to root the trees with (starts from 0)</p>
<p>This controls the addition of the -outgrno parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.pairing">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pairing</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.pairing" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>tree pairing method ([A]djacent pairs, all [p]ossible pairs)</p>
<p>This controls the addition of the -pairing parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.style">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">style</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.style" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>output style - [V]erbose, [f]ill, [s]parse</p>
<p>This controls the addition of the -style parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FTreeDistCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FTreeDistCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FNeighborCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fneighbor'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fneighbor program from EMBOSS.</p>
<p>fneighbor is an EMBOSS wrapper for the PHYLIP program neighbor used for
calculating neighbor-joining or UPGMA trees from distance matrices.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fneighbor'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.datafile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">datafile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.datafile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>dist file to use (phylip)</p>
<p>This controls the addition of the -datafile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.jumble">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">jumble</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.jumble" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>randommise input order (Y/n)</p>
<p>This controls the addition of the -jumble parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.matrixtype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">matrixtype</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.matrixtype" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>is matrix square (S), upper (U) or lower (L)</p>
<p>This controls the addition of the -matrixtype parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.outgrno">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outgrno</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.outgrno" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>taxon to use as OG</p>
<p>This controls the addition of the -outgrno parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.outtreefile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outtreefile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.outtreefile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>filename for output tree</p>
<p>This controls the addition of the -outtreefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.progress">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">progress</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.progress" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>print progress (Y/n)</p>
<p>This controls the addition of the -progress parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.seed">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seed</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.seed" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>provide a random seed</p>
<p>This controls the addition of the -seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.treeprint">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">treeprint</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.treeprint" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>print tree (Y/n)</p>
<p>This controls the addition of the -treeprint parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.treetype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">treetype</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.treetype" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>nj or UPGMA tree (n/u)</p>
<p>This controls the addition of the -treetype parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.trout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">trout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.trout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>write tree (Y/n)</p>
<p>This controls the addition of the -trout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FNeighborCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FNeighborCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FSeqBootCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fseqboot'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fseqboot program from EMBOSS.</p>
<p>fseqboot is an EMBOSS wrapper for the PHYLIP program seqboot used to
pseudo-sample alignment files.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fseqboot'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.blocksize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">blocksize</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.blocksize" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>print progress (Y/n)</p>
<p>This controls the addition of the -blocksize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.catergories">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">catergories</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.catergories" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>file of input categories</p>
<p>This controls the addition of the -categories parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.dotdiff">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dotdiff</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.dotdiff" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Use dot-differencing? [Y/n]</p>
<p>This controls the addition of the -dotdiff parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.fracsample">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fracsample</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.fracsample" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>fraction to resample</p>
<p>This controls the addition of the -fracsample parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.jusweights">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">jusweights</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.jusweights" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>what to write out [D]atasets of just [w]eights</p>
<p>This controls the addition of the -justweights parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.regular">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regular</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.regular" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>absolute number to resample</p>
<p>This controls the addition of the -regular parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.reps">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reps</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.reps" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>how many replicates, defaults to 100)</p>
<p>This controls the addition of the -reps parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.rewriteformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rewriteformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.rewriteformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>output format ([P]hyilp, [n]exus, [x]ml</p>
<p>This controls the addition of the -rewriteformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.seed">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seed</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.seed" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>specify random seed</p>
<p>This controls the addition of the -seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.seqtype">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqtype</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.seqtype" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>output format ([D]na, [p]rotein, [r]na</p>
<p>This controls the addition of the -seqtype parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>seq file to sample (phylip)</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.test">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">test</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.test" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>specify operation, default is bootstrap</p>
<p>This controls the addition of the -test parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FSeqBootCommandline.weights">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">weights</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FSeqBootCommandline.weights" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>weights file</p>
<p>This controls the addition of the -weights parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FDNAParsCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fdnapars'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fdnapars program from EMBOSS.</p>
<p>fdnapars is an EMBOSS version of the PHYLIP program dnapars, for
estimating trees from DNA sequences using parsiomny. Calling this command
without providing a value for the option ‚Äú-intreefile‚Äù will invoke
‚Äúinteractive mode‚Äù (and as a result fail if called with subprocess) if
‚Äú-auto‚Äù is not set to true.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fdnapars'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.dotdiff">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dotdiff</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.dotdiff" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Use dot-differencing? [Y/n]</p>
<p>This controls the addition of the -dotdiff parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.intreefile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">intreefile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.intreefile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Phylip tree file</p>
<p>This controls the addition of the -intreefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.maxtrees">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxtrees</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.maxtrees" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>max trees to save during run</p>
<p>This controls the addition of the -maxtrees parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.njumble">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">njumble</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.njumble" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>number of times to randomise input order (default is 0)</p>
<p>This controls the addition of the -njumble parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.outgrno">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outgrno</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.outgrno" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Specify outgroup</p>
<p>This controls the addition of the -outgrno parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.outtreefile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outtreefile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.outtreefile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>filename for output tree</p>
<p>This controls the addition of the -outtreefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.rearrange">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rearrange</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.rearrange" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Rearrange on just 1 best tree (Y/n)</p>
<p>This controls the addition of the -rearrange parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.seed">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seed</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.seed" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>provide random seed</p>
<p>This controls the addition of the -seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>seq file to use (phylip)</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.thorough">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thorough</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.thorough" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>more thorough search (Y/n)</p>
<p>This controls the addition of the -thorough parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.thresh">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thresh</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.thresh" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Use threshold parsimony (y/N)</p>
<p>This controls the addition of the -thresh parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.threshold" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Threshold value</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.transversion">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">transversion</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.transversion" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Use tranversion parsimony (y/N)</p>
<p>This controls the addition of the -transversion parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.trout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">trout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.trout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write trees to file (Y/n)</p>
<p>This controls the addition of the -trout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FDNAParsCommandline.weights">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">weights</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FDNAParsCommandline.weights" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>weights file</p>
<p>This controls the addition of the -weights parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FProtParsCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fprotpars'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fdnapars program from EMBOSS.</p>
<p>fprotpars is an EMBOSS version of the PHYLIP program protpars, for
estimating trees from protein  sequences using parsiomny. Calling this
command without providing a value for the option ‚Äú-intreefile‚Äù will invoke
‚Äúinteractive mode‚Äù (and as a result fail if called with subprocess) if
‚Äú-auto‚Äù is not set to true.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fprotpars'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.dotdiff">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">dotdiff</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.dotdiff" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Use dot-differencing? [Y/n]</p>
<p>This controls the addition of the -dotdiff parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.intreefile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">intreefile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.intreefile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Phylip tree file to score</p>
<p>This controls the addition of the -intreefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.njumble">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">njumble</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.njumble" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>number of times to randomise input order (default is 0)</p>
<p>This controls the addition of the -njumble parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.outgrno">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outgrno</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.outgrno" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Specify outgroup</p>
<p>This controls the addition of the -outgrno parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.outtreefile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outtreefile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.outtreefile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>phylip tree output file</p>
<p>This controls the addition of the -outtreefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.seed">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seed</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.seed" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>provide random seed</p>
<p>This controls the addition of the -seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>seq file to use (phylip)</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.thresh">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">thresh</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.thresh" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Use threshold parsimony (y/N)</p>
<p>This controls the addition of the -thresh parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.threshold" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Threshold value</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.trout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">trout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.trout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write trees to file (Y/n)</p>
<p>This controls the addition of the -trout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.weights">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">weights</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.weights" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>weights file</p>
<p>This controls the addition of the -weights parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtParsCommandline.whichcode">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">whichcode</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtParsCommandline.whichcode" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>which genetic code, [U,M,V,F,Y]]</p>
<p>This controls the addition of the -whichcode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FProtDistCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fprotdist'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fprotdist program from EMBOSS.</p>
<p>fprotdist is an EMBOSS wrapper for the PHYLIP program protdist used to
estimate trees from protein sequences using parsimony</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fprotdist'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.aacateg">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aacateg</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.aacateg" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Choose the category to use [G,C,H]</p>
<p>This controls the addition of the -aacateg parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.basefreq">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">basefreq</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.basefreq" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>DNA base frequencies (space separated list)</p>
<p>This controls the addition of the -basefreq parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.catergories">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">catergories</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.catergories" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>file of rates</p>
<p>This controls the addition of the -catergories parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.ease">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ease</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.ease" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Pob change catergory (float between -0 and 1)</p>
<p>This controls the addition of the -ease parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.gamma">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gamma</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.gamma" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>gamma [g, i,c]</p>
<p>This controls the addition of the -gamma parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.gammacoefficient">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gammacoefficient</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.gammacoefficient" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>value for gamma (&gt; 0.001)</p>
<p>This controls the addition of the -gammacoefficient parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.invarcoefficient">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">invarcoefficient</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.invarcoefficient" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>float for variation of substitution rate among sites</p>
<p>This controls the addition of the -invarcoefficient parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.method">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">method</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.method" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>sub. model [j,h,d,k,s,c]</p>
<p>This controls the addition of the -method parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.ncategories">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ncategories</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.ncategories" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>number of rate catergories (1-9)</p>
<p>This controls the addition of the -ncategories parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.rate">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rate</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.rate" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>rate for each category</p>
<p>This controls the addition of the -rate parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>seq file to use (phylip)</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.ttratio">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ttratio</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.ttratio" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Transition/transversion ratio (0-1)</p>
<p>This controls the addition of the -ttratio parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.weights">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">weights</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.weights" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>weights file</p>
<p>This controls the addition of the -weights parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FProtDistCommandline.whichcode">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">whichcode</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FProtDistCommandline.whichcode" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>genetic code [c,m,v,f,y]</p>
<p>This controls the addition of the -whichcode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FConsenseCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fconsense'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fconsense program from EMBOSS.</p>
<p>fconsense is an EMBOSS wrapper for the PHYLIP program consense used to
calculate consensus trees.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fconsense'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.intreefile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">intreefile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.intreefile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>file with phylip trees to make consensus from</p>
<p>This controls the addition of the -intreefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.method">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">method</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.method" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>consensus method [s, mr, MRE, ml]</p>
<p>This controls the addition of the -method parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.mlfrac">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mlfrac</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.mlfrac" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>cut-off freq for branch to appear in consensus (0.5-1.0)</p>
<p>This controls the addition of the -mlfrac parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.outgrno">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outgrno</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.outgrno" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>OTU to use as outgroup (starts from 0)</p>
<p>This controls the addition of the -outgrno parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.outtreefile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outtreefile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.outtreefile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Phylip tree output file (optional)</p>
<p>This controls the addition of the -outtreefile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.root">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">root</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.root" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>treat trees as rooted (YES, no)</p>
<p>This controls the addition of the -root parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.trout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">trout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.trout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>treat trees as rooted (YES, no)</p>
<p>This controls the addition of the -trout parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FConsenseCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FConsenseCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">WaterCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'water'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the water program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'water'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.aformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.aformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display output in a different specified output format</p>
<p>This controls the addition of the -aformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.asequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">asequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.asequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>First sequence to align</p>
<p>This controls the addition of the -asequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.brief">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">brief</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.brief" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display brief identity and similarity</p>
<p>This property controls the addition of the -brief switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.bsequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bsequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.bsequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Second sequence to align</p>
<p>This controls the addition of the -bsequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.datafile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">datafile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.datafile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Matrix file</p>
<p>This controls the addition of the -datafile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.gapextend" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap extension penalty</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.gapopen" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap open penalty</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.nobrief">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nobrief</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.nobrief" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display extended identity and similarity</p>
<p>This property controls the addition of the -nobrief switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.similarity">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">similarity</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.similarity" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display percent identity and similarity</p>
<p>This controls the addition of the -similarity parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.snucleotide">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">snucleotide</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.snucleotide" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are nucleotide (boolean)</p>
<p>This controls the addition of the -snucleotide parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.sprotein">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sprotein</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.sprotein" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are protein (boolean)</p>
<p>This controls the addition of the -sprotein parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.WaterCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.WaterCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">NeedleCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'needle'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the needle program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'needle'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.aformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.aformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display output in a different specified output format</p>
<p>This controls the addition of the -aformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.asequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">asequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.asequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>First sequence to align</p>
<p>This controls the addition of the -asequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.brief">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">brief</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.brief" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display brief identity and similarity</p>
<p>This property controls the addition of the -brief switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.bsequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bsequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.bsequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Second sequence to align</p>
<p>This controls the addition of the -bsequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.datafile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">datafile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.datafile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Matrix file</p>
<p>This controls the addition of the -datafile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.endextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">endextend</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.endextend" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>The score added to the end gap penalty for each base or residue in the end gap.</p>
<p>This controls the addition of the -endextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.endopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">endopen</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.endopen" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>The score taken away when an end gap is created.</p>
<p>This controls the addition of the -endopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.endweight">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">endweight</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.endweight" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Apply And gap penalties</p>
<p>This controls the addition of the -endweight parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.gapextend" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap extension penalty</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.gapopen" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap open penalty</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.nobrief">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nobrief</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.nobrief" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display extended identity and similarity</p>
<p>This property controls the addition of the -nobrief switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.similarity">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">similarity</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.similarity" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display percent identity and similarity</p>
<p>This controls the addition of the -similarity parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.snucleotide">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">snucleotide</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.snucleotide" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are nucleotide (boolean)</p>
<p>This controls the addition of the -snucleotide parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.sprotein">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sprotein</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.sprotein" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are protein (boolean)</p>
<p>This controls the addition of the -sprotein parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">NeedleallCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'needleall'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the needleall program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'needleall'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.aformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.aformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display output in a different specified output format</p>
<p>This controls the addition of the -aformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.asequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">asequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.asequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>First sequence to align</p>
<p>This controls the addition of the -asequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.brief">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">brief</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.brief" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display brief identity and similarity</p>
<p>This property controls the addition of the -brief switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.bsequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bsequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.bsequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Second sequence to align</p>
<p>This controls the addition of the -bsequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.datafile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">datafile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.datafile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Matrix file</p>
<p>This controls the addition of the -datafile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.endextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">endextend</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.endextend" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>The score added to the end gap penalty for each base or residue in the end gap.</p>
<p>This controls the addition of the -endextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.endopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">endopen</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.endopen" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>The score taken away when an end gap is created.</p>
<p>This controls the addition of the -endopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.endweight">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">endweight</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.endweight" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Apply And gap penalties</p>
<p>This controls the addition of the -endweight parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.errorfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">errorfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.errorfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Error file to be written to.</p>
<p>This controls the addition of the -errorfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.gapextend" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap extension penalty</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.gapopen" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap open penalty</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.minscore">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">minscore</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.minscore" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Exclude alignments with scores below this threshold score.</p>
<p>This controls the addition of the -minscore parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.nobrief">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nobrief</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.nobrief" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display extended identity and similarity</p>
<p>This property controls the addition of the -nobrief switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.similarity">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">similarity</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.similarity" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display percent identity and similarity</p>
<p>This controls the addition of the -similarity parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.snucleotide">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">snucleotide</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.snucleotide" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are nucleotide (boolean)</p>
<p>This controls the addition of the -snucleotide parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.sprotein">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sprotein</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.sprotein" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are protein (boolean)</p>
<p>This controls the addition of the -sprotein parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.NeedleallCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.NeedleallCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">StretcherCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'stretcher'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the stretcher program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'stretcher'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.aformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.aformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display output in a different specified output format</p>
<p>This controls the addition of the -aformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.asequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">asequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.asequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>First sequence to align</p>
<p>This controls the addition of the -asequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.bsequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bsequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.bsequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Second sequence to align</p>
<p>This controls the addition of the -bsequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.datafile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">datafile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.datafile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Matrix file</p>
<p>This controls the addition of the -datafile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.gapextend">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapextend</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.gapextend" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap extension penalty</p>
<p>This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.gapopen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gapopen</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.gapopen" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap open penalty</p>
<p>This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.snucleotide">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">snucleotide</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.snucleotide" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are nucleotide (boolean)</p>
<p>This controls the addition of the -snucleotide parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.sprotein">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sprotein</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.sprotein" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequences are protein (boolean)</p>
<p>This controls the addition of the -sprotein parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.StretcherCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.StretcherCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FuzznucCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fuzznuc'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fuzznuc program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fuzznuc'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.complement">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">complement</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.complement" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Search complementary strand</p>
<p>This controls the addition of the -complement parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.pattern">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pattern</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.pattern" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Search pattern, using standard IUPAC one-letter codes</p>
<p>This controls the addition of the -pattern parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.pmismatch">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pmismatch</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.pmismatch" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Number of mismatches</p>
<p>This controls the addition of the -pmismatch parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.rformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.rformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Specify the report format to output in.</p>
<p>This controls the addition of the -rformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence database USA</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzznucCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzznucCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">FuzzproCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fuzzpro'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the fuzzpro program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'fuzzpro'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.pattern">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pattern</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.pattern" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Search pattern, using standard IUPAC one-letter codes</p>
<p>This controls the addition of the -pattern parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.pmismatch">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pmismatch</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.pmismatch" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Number of mismatches</p>
<p>This controls the addition of the -pmismatch parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.rformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.rformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Specify the report format to output in.</p>
<p>This controls the addition of the -rformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence database USA</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.FuzzproCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.FuzzproCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">Est2GenomeCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'est2genome'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the est2genome program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'est2genome'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.align">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">align</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.align" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Show the alignment.</p>
<p>This controls the addition of the -align parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.best">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">best</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.best" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>You can print out all comparisons instead of just the best</p>
<p>This controls the addition of the -best parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.est">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">est</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.est" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>EST sequence(s)</p>
<p>This controls the addition of the -est parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.gappenalty">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gappenalty</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.gappenalty" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Cost for deleting a single base in either sequence, excluding introns</p>
<p>This controls the addition of the -gappenalty parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.genome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genome</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.genome" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Genomic sequence</p>
<p>This controls the addition of the -genome parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.intronpenalty">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">intronpenalty</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.intronpenalty" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Cost for an intron, independent of length.</p>
<p>This controls the addition of the -intronpenalty parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.match">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">match</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.match" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Score for matching two bases</p>
<p>This controls the addition of the -match parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.minscore">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">minscore</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.minscore" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Exclude alignments with scores below this threshold score.</p>
<p>This controls the addition of the -minscore parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.mismatch">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatch</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.mismatch" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Cost for mismatching two bases</p>
<p>This controls the addition of the -mismatch parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.mode">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mode</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.mode" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>This determines the comparison mode. ‚Äòboth‚Äô, ‚Äòforward‚Äô, or ‚Äòreverse‚Äô</p>
<p>This controls the addition of the -mode parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.reverse">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reverse</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.reverse" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Reverse the orientation of the EST sequence</p>
<p>This controls the addition of the -reverse parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.seed">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seed</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.seed" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Random number seed</p>
<p>This controls the addition of the -seed parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.shuffle">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shuffle</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.shuffle" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Shuffle</p>
<p>This controls the addition of the -shuffle parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.space">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">space</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.space" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>for linear-space recursion.</p>
<p>This controls the addition of the -space parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.splice">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">splice</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.splice" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Use donor and acceptor splice sites.</p>
<p>This controls the addition of the -splice parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.splicepenalty">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">splicepenalty</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.splicepenalty" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Cost for an intron, independent of length and starting/ending on donor-acceptor sites</p>
<p>This controls the addition of the -splicepenalty parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.Est2GenomeCommandline.width">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">width</span></span><a class="headerlink" href="#Bio.Emboss.Applications.Est2GenomeCommandline.width" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Alignment width</p>
<p>This controls the addition of the -width parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">ETandemCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'etandem'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the etandem program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'etandem'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.maxrepeat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxrepeat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.maxrepeat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum repeat size</p>
<p>This controls the addition of the -maxrepeat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.minrepeat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">minrepeat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.minrepeat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum repeat size</p>
<p>This controls the addition of the -minrepeat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.mismatch">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatch</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.mismatch" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Allow N as a mismatch</p>
<p>This controls the addition of the -mismatch parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.rformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.rformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output report format</p>
<p>This controls the addition of the -rformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.threshold" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Threshold score</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.uniform">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">uniform</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.uniform" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Allow uniform consensus</p>
<p>This controls the addition of the -uniform parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.ETandemCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.ETandemCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">EInvertedCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'einverted'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the einverted program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'einverted'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.gap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gap</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.gap" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Gap penalty</p>
<p>This controls the addition of the -gap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.match">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">match</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.match" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Match score</p>
<p>This controls the addition of the -match parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.maxrepeat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxrepeat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.maxrepeat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum separation between the start and end of repeat</p>
<p>This controls the addition of the -maxrepeat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.mismatch">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mismatch</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.mismatch" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Mismatch score</p>
<p>This controls the addition of the -mismatch parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.threshold">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">threshold</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.threshold" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum score threshold</p>
<p>This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.EInvertedCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.EInvertedCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">PalindromeCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'palindrome'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the palindrome program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'palindrome'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.gaplimit">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gaplimit</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.gaplimit" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum gap between repeats</p>
<p>This controls the addition of the -gaplimit parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.maxpallen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maxpallen</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.maxpallen" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Maximum palindrome length</p>
<p>This controls the addition of the -maxpallen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.minpallen">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">minpallen</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.minpallen" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Minimum palindrome length</p>
<p>This controls the addition of the -minpallen parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.nummismatches">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">nummismatches</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.nummismatches" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Number of mismatches allowed</p>
<p>This controls the addition of the -nummismatches parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.overlap">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">overlap</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.overlap" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report overlapping matches</p>
<p>This controls the addition of the -overlap parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Sequence</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.PalindromeCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.PalindromeCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">TranalignCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tranalign'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the tranalign program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tranalign'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.asequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">asequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.asequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Nucleotide sequences to be aligned.</p>
<p>This controls the addition of the -asequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.bsequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bsequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.bsequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Protein sequence alignment</p>
<p>This controls the addition of the -bsequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.outseq">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outseq</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.outseq" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output sequence file.</p>
<p>This controls the addition of the -outseq parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.table">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">table</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.table" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Code to use</p>
<p>This controls the addition of the -table parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.TranalignCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.TranalignCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">DiffseqCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'diffseq'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the diffseq program from EMBOSS.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'diffseq'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.aoutfeat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aoutfeat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.aoutfeat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>File for output of first sequence‚Äôs features</p>
<p>This controls the addition of the -aoutfeat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.asequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">asequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.asequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>First sequence to compare</p>
<p>This controls the addition of the -asequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.boutfeat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">boutfeat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.boutfeat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>File for output of second sequence‚Äôs features</p>
<p>This controls the addition of the -boutfeat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.bsequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bsequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.bsequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Second sequence to compare</p>
<p>This controls the addition of the -bsequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.rformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.rformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output report file format</p>
<p>This controls the addition of the -rformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.DiffseqCommandline.wordsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">wordsize</span></span><a class="headerlink" href="#Bio.Emboss.Applications.DiffseqCommandline.wordsize" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Word size to use for comparisons (10 default)</p>
<p>This controls the addition of the -wordsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">IepCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'iep'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline for EMBOSS iep: calculated isoelectric point and charge.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Emboss.Applications</span> <span class="kn">import</span> <span class="n">IepCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">iep_cline</span> <span class="o">=</span> <span class="n">IepCommandline</span><span class="p">(</span><span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;proteins.faa&quot;</span><span class="p">,</span>
<span class="gp">... </span>                           <span class="n">outfile</span><span class="o">=</span><span class="s2">&quot;proteins.txt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">iep_cline</span><span class="p">)</span>
<span class="go">iep -outfile=proteins.txt -sequence=proteins.faa</span>
</pre></div>
</div>
<p>You would typically run the command line with iep_cline() or via the
Python subprocess module, as described in the Biopython tutorial.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'iep'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.amino">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">amino</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.amino" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Number of N-termini</p>
<blockquote>
<div><p>Integer 0 (default) or more.</p>
</div></blockquote>
<p>This controls the addition of the -amino parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.carboxyl">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">carboxyl</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.carboxyl" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Number of C-termini</p>
<blockquote>
<div><p>Integer 0 (default) or more.</p>
</div></blockquote>
<p>This controls the addition of the -carboxyl parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.disulphides">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">disulphides</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.disulphides" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Number of disulphide bridges</p>
<blockquote>
<div><p>Integer 0 (default) or more.</p>
</div></blockquote>
<p>This controls the addition of the -disulphides parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.lysinemodified">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lysinemodified</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.lysinemodified" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Number of modified lysines</p>
<blockquote>
<div><p>Integer 0 (default) or more.</p>
</div></blockquote>
<p>This controls the addition of the -lysinemodified parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.notermini">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">notermini</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.notermini" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Exclude (True) or include (False) charge at N and C terminus.</p>
<p>This controls the addition of the -notermini parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Protein sequence(s) filename</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.IepCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.IepCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">SeqretCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seqret'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossMinimalCommandLine</span></code></p>
<p>Commandline object for the seqret program from EMBOSS.</p>
<p>This tool allows you to interconvert between different sequence file
formats (e.g. GenBank to FASTA). Combining Biopython‚Äôs Bio.SeqIO module
with seqret using a suitable intermediate file format can allow you to
read/write to an even wider range of file formats.</p>
<p>This wrapper currently only supports the core functionality, things like
feature tables (in EMBOSS 6.1.0 onwards) are not yet included.</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seqret'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.osformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">osformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.osformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output sequence(s) format (e.g. fasta, genbank)</p>
<p>This controls the addition of the -osformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.outseq">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outseq</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.outseq" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output sequence file.</p>
<p>This controls the addition of the -outseq parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Input sequence(s) filename</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.sformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.sformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Input sequence(s) format (e.g. fasta, genbank)</p>
<p>This controls the addition of the -sformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqretCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqretCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.Emboss.Applications.</span></span><span class="sig-name descname"><span class="pre">SeqmatchallCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seqmatchall'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Emboss.Applications._EmbossCommandLine</span></code></p>
<p>Commandline object for the seqmatchall program from EMBOSS.</p>
<p>e.g.
&gt;&gt;&gt; cline = SeqmatchallCommandline(sequence=‚Äùopuntia.fasta‚Äù, outfile=‚Äùopuntia.txt‚Äù)
&gt;&gt;&gt; cline.auto = True
&gt;&gt;&gt; cline.wordsize = 18
&gt;&gt;&gt; cline.aformat = ‚Äúpair‚Äù
&gt;&gt;&gt; print(cline)
seqmatchall -auto -outfile=opuntia.txt -sequence=opuntia.fasta -wordsize=18 -aformat=pair</p>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seqmatchall'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.aformat">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aformat</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.aformat" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Display output in a different specified output format</p>
<p>This controls the addition of the -aformat parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.auto">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">auto</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.auto" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn off prompts.</p>
<p>Automatic mode disables prompting, so we recommend you set this argument all the time when calling an EMBOSS tool from Biopython.</p>
<p>This property controls the addition of the -auto switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.debug">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">debug</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.debug" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write debug output to program.dbg.</p>
<p>This property controls the addition of the -debug switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.die">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">die</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.die" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report dying program messages.</p>
<p>This property controls the addition of the -die switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.error">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">error</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.error" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report errors.</p>
<p>This property controls the addition of the -error switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.filter">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">filter</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.filter" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Read standard input, write standard output.</p>
<p>This property controls the addition of the -filter switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.help">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">help</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.help" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report command line options.</p>
<p>More information on associated and general qualifiers can be found with -help -verbose</p>
<p>This property controls the addition of the -help switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.options">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">options</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.options" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Prompt for standard and additional values.</p>
<p>If you are calling an EMBOSS tool from within Biopython, we DO NOT recommend using this option.</p>
<p>This property controls the addition of the -options switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.outfile">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">outfile</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.outfile" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Output filename</p>
<p>This controls the addition of the -outfile parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.sequence">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.sequence" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Readable set of sequences</p>
<p>This controls the addition of the -sequence parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.stdout">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">stdout</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.stdout" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write standard output.</p>
<p>This property controls the addition of the -stdout switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.verbose">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">verbose</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.verbose" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report some/full command line options</p>
<p>This property controls the addition of the -verbose switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.warning">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">warning</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.warning" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Report warnings.</p>
<p>This property controls the addition of the -warning switch, treat this property as a boolean.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="Bio.Emboss.Applications.SeqmatchallCommandline.wordsize">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">wordsize</span></span><a class="headerlink" href="#Bio.Emboss.Applications.SeqmatchallCommandline.wordsize" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Word size (Integer 2 or more, default 4)</p>
<p>This controls the addition of the -wordsize parameter and its associated value.  Set this property to the argument value required.</p>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.Emboss.html" class="btn btn-neutral float-left" title="Bio.Emboss package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.Emboss.Primer3.html" class="btn btn-neutral float-right" title="Bio.Emboss.Primer3 module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.82/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>