<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.motifs.transfac module &mdash; Biopython 1.82.dev0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.motifs.xms module" href="Bio.motifs.xms.html" />
    <link rel="prev" title="Bio.motifs.thresholds module" href="Bio.motifs.thresholds.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="index.html">
            
              <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html#subpackages">Subpackages</a></li>
<li class="toctree-l4 current"><a class="reference internal" href="Bio.motifs.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.motifs.html#module-Bio.motifs">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="Bio.html">Bio package</a></li>
          <li class="breadcrumb-item"><a href="Bio.motifs.html">Bio.motifs package</a></li>
      <li class="breadcrumb-item active">Bio.motifs.transfac module</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/motifs/transfac.py" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="Bio.motifs.thresholds.html" class="btn btn-neutral float-left" title="Bio.motifs.thresholds module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.motifs.xms.html" class="btn btn-neutral float-right" title="Bio.motifs.xms module" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-Bio.motifs.transfac">
<span id="bio-motifs-transfac-module"></span><h1>Bio.motifs.transfac module<a class="headerlink" href="#module-Bio.motifs.transfac" title="Permalink to this headline">ÔÉÅ</a></h1>
<p>Parsing TRANSFAC files.</p>
<dl class="py class">
<dt class="sig sig-object py" id="Bio.motifs.transfac.Motif">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.motifs.transfac.</span></span><span class="sig-name descname"><span class="pre">Motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ACGT'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alignment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">instances</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.transfac.Motif" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.motifs.html#Bio.motifs.Motif" title="Bio.motifs.Motif"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.motifs.Motif</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
<p>Store the information for one TRANSFAC motif.</p>
<p>This class inherits from the Bio.motifs.Motif base class, as well
as from a Python dictionary. All motif information found by the parser
is stored as attributes of the base class when possible; see the
Bio.motifs.Motif base class for a description of these attributes. All
other information associated with the motif is stored as (key, value)
pairs in the dictionary, where the key is the two-letter fields as found
in the TRANSFAC file. References are an exception: These are stored in
the .references attribute.</p>
<p>These fields are commonly found in TRANSFAC files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AC</span><span class="p">:</span>    <span class="n">Accession</span> <span class="n">number</span>
<span class="n">AS</span><span class="p">:</span>    <span class="n">Accession</span> <span class="n">numbers</span><span class="p">,</span> <span class="n">secondary</span>
<span class="n">BA</span><span class="p">:</span>    <span class="n">Statistical</span> <span class="n">basis</span>
<span class="n">BF</span><span class="p">:</span>    <span class="n">Binding</span> <span class="n">factors</span>
<span class="n">BS</span><span class="p">:</span>    <span class="n">Factor</span> <span class="n">binding</span> <span class="n">sites</span> <span class="n">underlying</span> <span class="n">the</span> <span class="n">matrix</span>
       <span class="p">[</span><span class="n">sequence</span><span class="p">;</span> <span class="n">SITE</span> <span class="n">accession</span> <span class="n">number</span><span class="p">;</span> <span class="n">start</span> <span class="n">position</span> <span class="k">for</span> <span class="n">matrix</span>
       <span class="n">sequence</span><span class="p">;</span> <span class="n">length</span> <span class="n">of</span> <span class="n">sequence</span> <span class="n">used</span><span class="p">;</span> <span class="n">number</span> <span class="n">of</span> <span class="n">gaps</span> <span class="n">inserted</span><span class="p">;</span>
       <span class="n">strand</span> <span class="n">orientation</span><span class="o">.</span><span class="p">]</span>
<span class="n">CC</span><span class="p">:</span>    <span class="n">Comments</span>
<span class="n">CO</span><span class="p">:</span>    <span class="n">Copyright</span> <span class="n">notice</span>
<span class="n">DE</span><span class="p">:</span>    <span class="n">Short</span> <span class="n">factor</span> <span class="n">description</span>
<span class="n">DR</span><span class="p">:</span>    <span class="n">External</span> <span class="n">databases</span>
       <span class="p">[</span><span class="n">database</span> <span class="n">name</span><span class="p">:</span> <span class="n">database</span> <span class="n">accession</span> <span class="n">number</span><span class="p">]</span>
<span class="n">DT</span><span class="p">:</span>    <span class="n">Date</span> <span class="n">created</span><span class="o">/</span><span class="n">updated</span>
<span class="n">HC</span><span class="p">:</span>    <span class="n">Subfamilies</span>
<span class="n">HP</span><span class="p">:</span>    <span class="n">Superfamilies</span>
<span class="n">ID</span><span class="p">:</span>    <span class="n">Identifier</span>
<span class="n">NA</span><span class="p">:</span>    <span class="n">Name</span> <span class="n">of</span> <span class="n">the</span> <span class="n">binding</span> <span class="n">factor</span>
<span class="n">OC</span><span class="p">:</span>    <span class="n">Taxonomic</span> <span class="n">classification</span>
<span class="n">OS</span><span class="p">:</span>    <span class="n">Species</span><span class="o">/</span><span class="n">Taxon</span>
<span class="n">OV</span><span class="p">:</span>    <span class="n">Older</span> <span class="n">version</span>
<span class="n">PV</span><span class="p">:</span>    <span class="n">Preferred</span> <span class="n">version</span>
<span class="n">TY</span><span class="p">:</span>    <span class="n">Type</span>
<span class="n">XX</span><span class="p">:</span>    <span class="n">Empty</span> <span class="n">line</span><span class="p">;</span> <span class="n">these</span> <span class="n">are</span> <span class="ow">not</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">Record</span><span class="o">.</span>
</pre></div>
</div>
<p>References are stored in an .references attribute, which is a list of
dictionaries with the following keys:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">RN</span><span class="p">:</span>    <span class="n">Reference</span> <span class="n">number</span>
<span class="n">RA</span><span class="p">:</span>    <span class="n">Reference</span> <span class="n">authors</span>
<span class="n">RL</span><span class="p">:</span>    <span class="n">Reference</span> <span class="n">data</span>
<span class="n">RT</span><span class="p">:</span>    <span class="n">Reference</span> <span class="n">title</span>
<span class="n">RX</span><span class="p">:</span>    <span class="n">PubMed</span> <span class="n">ID</span>
</pre></div>
</div>
<p>For more information, see the TRANSFAC documentation.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.motifs.transfac.Motif.multiple_value_keys">
<span class="sig-name descname"><span class="pre">multiple_value_keys</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'BF',</span> <span class="pre">'BS',</span> <span class="pre">'DR',</span> <span class="pre">'DT',</span> <span class="pre">'HC',</span> <span class="pre">'HP',</span> <span class="pre">'OV'}</span></em><a class="headerlink" href="#Bio.motifs.transfac.Motif.multiple_value_keys" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.motifs.transfac.Motif.reference_keys">
<span class="sig-name descname"><span class="pre">reference_keys</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'RA',</span> <span class="pre">'RL',</span> <span class="pre">'RT',</span> <span class="pre">'RX'}</span></em><a class="headerlink" href="#Bio.motifs.transfac.Motif.reference_keys" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.transfac.Motif.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.transfac.Motif.__getitem__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Return a new Motif object for the positions included in key.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">motifs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">motif</span> <span class="o">=</span> <span class="n">motifs</span><span class="o">.</span><span class="n">create</span><span class="p">([</span><span class="s2">&quot;AACGCCA&quot;</span><span class="p">,</span> <span class="s2">&quot;ACCGCCC&quot;</span><span class="p">,</span> <span class="s2">&quot;AACTCCG&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span>
<span class="go">AACGCCA</span>
<span class="go">ACCGCCC</span>
<span class="go">AACTCCG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">motif</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="go">AACGCC</span>
<span class="go">ACCGCC</span>
<span class="go">AACTCC</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.motifs.transfac.Record">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">Bio.motifs.transfac.</span></span><span class="sig-name descname"><span class="pre">Record</span></span><a class="headerlink" href="#Bio.motifs.transfac.Record" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></p>
<p>Store the information in a TRANSFAC matrix table.</p>
<p>The record inherits from a list containing the individual motifs.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>version - The version number, corresponding to the ‚ÄòVV‚Äô field
in the TRANSFAC file;</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.transfac.Record.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.transfac.Record.__init__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.motifs.transfac.Record.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.transfac.Record.__str__" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Turn the TRANSFAC matrix into a string.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.motifs.transfac.read">
<span class="sig-prename descclassname"><span class="pre">Bio.motifs.transfac.</span></span><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.transfac.read" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Parse a transfac format handle into a Record object.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="Bio.motifs.transfac.write">
<span class="sig-prename descclassname"><span class="pre">Bio.motifs.transfac.</span></span><span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motifs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.motifs.transfac.write" title="Permalink to this definition">ÔÉÅ</a></dt>
<dd><p>Write the representation of a motif in TRANSFAC format.</p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Bio.motifs.thresholds.html" class="btn btn-neutral float-left" title="Bio.motifs.thresholds module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Bio.motifs.xms.html" class="btn btn-neutral float-right" title="Bio.motifs.xms module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2023, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.82.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.80/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>