<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Introduction &mdash; Biopython 1.83.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Quick Start – What can you do with Biopython?" href="chapter_quick_start.html" />
    <link rel="prev" title="Biopython Tutorial &amp; Cookbook" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Biopython Tutorial &amp; Cookbook</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Introduction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#what-is-biopython">What is Biopython?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#what-can-i-find-in-the-biopython-package">What can I find in the Biopython package</a></li>
<li class="toctree-l3"><a class="reference internal" href="#installing-biopython">Installing Biopython</a></li>
<li class="toctree-l3"><a class="reference internal" href="#frequently-asked-questions-faq">Frequently Asked Questions (FAQ)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chapter_quick_start.html">Quick Start – What can you do with Biopython?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_objects.html">Sequence objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_annot.html">Sequence annotation objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seqio.html">Sequence Input/Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_align.html">Sequence alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise.html">Pairwise sequence alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_msa.html">Multiple Sequence Alignment objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise2.html">Pairwise alignments using pairwise2</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html">BLAST (new)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html#blast-old">BLAST (old)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_searchio.html">BLAST and other sequence search tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_entrez.html">Accessing NCBI’s Entrez databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_uniprot.html">Swiss-Prot and ExPASy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pdb.html">Going 3D: The PDB module</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_popgen.html">Bio.PopGen: Population genetics</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phylo.html">Phylogenetics with Bio.Phylo</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_motifs.html">Sequence motif analysis using Bio.motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cluster.html">Cluster analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_graphics.html">Graphics including GenomeDiagram</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_kegg.html">KEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phenotype.html">Bio.phenotype: analyze phenotypic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cookbook.html">Cookbook – Cool things to do with it</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_testing.html">The Biopython testing framework</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_contributing.html">Where to go from here – contributing to Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_appendix.html">Appendix: Useful stuff about Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_bibliography.html">Bibliography</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html">API documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html#indices-and-tables">Indices and tables</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Biopython Tutorial &amp; Cookbook</a></li>
      <li class="breadcrumb-item active">Introduction</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/Tutorial/chapter_introduction.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="index.html" class="btn btn-neutral float-left" title="Biopython Tutorial &amp; Cookbook" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_quick_start.html" class="btn btn-neutral float-right" title="Quick Start – What can you do with Biopython?" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="introduction">
<span id="chapter-introduction"></span><h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline"></a></h1>
<section id="what-is-biopython">
<h2>What is Biopython?<a class="headerlink" href="#what-is-biopython" title="Permalink to this headline"></a></h2>
<p>Biopython is a collection of freely available Python
(<a class="reference external" href="https://www.python.org">https://www.python.org</a>) modules for computational molecular biology.
Python is an object oriented, interpreted, flexible language that is
widely used for scientific computing. Python is easy to learn, has a
very clear syntax and can easily be extended with modules written in C,
C++ or FORTRAN. Since its inception in 2000 <a class="reference internal" href="chapter_bibliography.html#chapman2000" id="id1"><span>[Chapman2000]</span></a>, Biopython
has been continuously developed and maintained by a large group of
volunteers worldwide.</p>
<p>The Biopython web site (<a class="reference external" href="http://www.biopython.org">http://www.biopython.org</a>) provides an online
resource for modules, scripts, and web links for developers of
Python-based software for bioinformatics use and research. Biopython
includes parsers for various bioinformatics file formats (BLAST,
Clustalw, FASTA, Genbank,…), access to online services (NCBI,
Expasy,…), a standard sequence class, sequence alignment and motif
analysis tools, clustering algorithms, a module for structural biology,
and a module for phylogenetics analysis.</p>
</section>
<section id="what-can-i-find-in-the-biopython-package">
<h2>What can I find in the Biopython package<a class="headerlink" href="#what-can-i-find-in-the-biopython-package" title="Permalink to this headline"></a></h2>
<p>The main Biopython releases have lots of functionality, including:</p>
<ul class="simple">
<li><p>The ability to parse bioinformatics files into Python utilizable data
structures, including support for the following formats:</p>
<ul>
<li><p>Blast output – both from standalone and WWW Blast</p></li>
<li><p>Clustalw</p></li>
<li><p>FASTA</p></li>
<li><p>GenBank</p></li>
<li><p>PubMed and Medline</p></li>
<li><p>ExPASy files, like Enzyme and Prosite</p></li>
<li><p>SCOP, including ‘dom’ and ‘lin’ files</p></li>
<li><p>UniGene</p></li>
<li><p>SwissProt</p></li>
</ul>
</li>
<li><p>Files in the supported formats can be iterated over record by record
or indexed and accessed via a Dictionary interface.</p></li>
<li><p>Code to deal with popular on-line bioinformatics destinations such
as:</p>
<ul>
<li><p>NCBI – Blast, Entrez and PubMed services</p></li>
<li><p>ExPASy – Swiss-Prot and Prosite entries, as well as Prosite
searches</p></li>
</ul>
</li>
<li><p>Interfaces to common bioinformatics programs such as:</p>
<ul>
<li><p>Standalone Blast from NCBI</p></li>
<li><p>Clustalw alignment program</p></li>
<li><p>EMBOSS command line tools</p></li>
</ul>
</li>
<li><p>A standard sequence class that deals with sequences, ids on
sequences, and sequence features.</p></li>
<li><p>Tools for performing common operations on sequences, such as
translation, transcription and weight calculations.</p></li>
<li><p>Code to perform classification of data using k Nearest Neighbors,
Naive Bayes or Support Vector Machines.</p></li>
<li><p>Code for dealing with alignments, including a standard way to create
and deal with substitution matrices.</p></li>
<li><p>Code making it easy to split up parallelizable tasks into separate
processes.</p></li>
<li><p>GUI-based programs to do basic sequence manipulations, translations,
BLASTing, etc.</p></li>
<li><p>Extensive documentation and help with using the modules, including
this file, on-line wiki documentation, the web site, and the mailing
list.</p></li>
<li><p>Integration with BioSQL, a sequence database schema also supported by
the BioPerl and BioJava projects.</p></li>
</ul>
<p>We hope this gives you plenty of reasons to download and start using
Biopython!</p>
</section>
<section id="installing-biopython">
<h2>Installing Biopython<a class="headerlink" href="#installing-biopython" title="Permalink to this headline"></a></h2>
<p>All of the installation information for Biopython was separated from
this document to make it easier to keep updated.</p>
<p>The short version is use <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">biopython</span></code>, see the <a class="reference external" href="https://github.com/biopython/biopython/blob/master/README.rst">main
README</a>
file for other options.</p>
</section>
<section id="frequently-asked-questions-faq">
<h2>Frequently Asked Questions (FAQ)<a class="headerlink" href="#frequently-asked-questions-faq" title="Permalink to this headline"></a></h2>
<ol class="arabic">
<li><div class="line-block">
<div class="line"><em>How do I cite Biopython in a scientific publication?</em></div>
<div class="line">Please cite our application note <a class="reference internal" href="chapter_bibliography.html#cock2009" id="id2"><span>[Cock2009]</span></a> as the
main Biopython reference. In addition, please cite any publications
from the following list if appropriate, in particular as a
reference for specific modules within Biopython (more information
can be found on our website):</div>
</div>
<ul class="simple">
<li><p>For the official project announcement:
Chapman and Chang, 2000 <a class="reference internal" href="chapter_bibliography.html#chapman2000" id="id3"><span>[Chapman2000]</span></a>;</p></li>
<li><p>For <code class="docutils literal notranslate"><span class="pre">Bio.PDB</span></code>:
Hamelryck and Manderick, 2003 <a class="reference internal" href="chapter_bibliography.html#hamelryck2003a" id="id4"><span>[Hamelryck2003A]</span></a>;</p></li>
<li><p>For <code class="docutils literal notranslate"><span class="pre">Bio.Cluster</span></code>:
de Hoon <em>et al.</em>, 2004 <a class="reference internal" href="chapter_bibliography.html#dehoon2004" id="id5"><span>[deHoon2004]</span></a>;</p></li>
<li><p>For <code class="docutils literal notranslate"><span class="pre">Bio.Graphics.GenomeDiagram</span></code>:
Pritchard <em>et al.</em>, 2006 <a class="reference internal" href="chapter_bibliography.html#pritchard2006" id="id6"><span>[Pritchard2006]</span></a>;</p></li>
<li><p>For <code class="docutils literal notranslate"><span class="pre">Bio.Phylo</span></code> and <code class="docutils literal notranslate"><span class="pre">Bio.Phylo.PAML</span></code>:
Talevich <em>et al.</em> 2012 <a class="reference internal" href="chapter_bibliography.html#talevich2012" id="id7"><span>[Talevich2012]</span></a>;</p></li>
<li><p>For the FASTQ file format as supported in Biopython, BioPerl,
BioRuby, BioJava, and EMBOSS:
Cock <em>et al.</em>, 2010 <a class="reference internal" href="chapter_bibliography.html#cock2010" id="id8"><span>[Cock2010]</span></a>.</p></li>
</ul>
</li>
<li><div class="line-block">
<div class="line"><em>How should I capitalize “Biopython”? Is “BioPython” OK?</em></div>
<div class="line">The correct capitalization is “Biopython”, not “BioPython” (even
though that would have matched BioPerl, BioJava and BioRuby).</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>How is the Biopython software licensed?</em></div>
<div class="line">Biopython is distributed under the <em>Biopython License Agreement</em>.
However, since the release of Biopython 1.69, some files are
explicitly dual licensed under your choice of the <em>Biopython
License Agreement</em> or the <em>BSD 3-Clause License</em>. This is with the
intention of later offering all of Biopython under this dual
licensing approach.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>What is the Biopython logo and how is it licensed?</em></div>
<div class="line">As of July 2017 and the Biopython 1.70 release, the Biopython logo
is a yellow and blue snake forming a double helix above the word
“biopython” in lower case. It was designed by Patrick Kunzmann and
this logo is dual licensed under your choice of the <em>Biopython
License Agreement</em> or the <em>BSD 3-Clause License</em>.</div>
<div class="line"><a class="reference internal" href="../_images/biopython_logo_m.png"><img alt="new-logo" src="../_images/biopython_logo_m.png" style="width: 6cm;" /></a></div>
<div class="line">Prior to this, the Biopython logo was two yellow snakes forming a
double helix around the word “BIOPYTHON”, designed by Henrik
Vestergaard and Thomas Hamelryck in 2003 as part of an open
competition.</div>
<div class="line"><a class="reference internal" href="../_images/biopython_logo_old.jpg"><img alt="old-logo" src="../_images/biopython_logo_old.jpg" style="width: 7cm;" /></a></div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Do you have a change-log listing what’s new in each release?</em></div>
<div class="line">See the file <code class="docutils literal notranslate"><span class="pre">NEWS.rst</span></code> included with the source code (originally
called just <code class="docutils literal notranslate"><span class="pre">NEWS</span></code>), or read the <a class="reference external" href="https://github.com/biopython/biopython/blob/master/NEWS.rst">latest NEWS file on
GitHub</a>.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>What is going wrong with my print commands?</em></div>
<div class="line">As of Biopython 1.77, we only support Python 3, so this tutorial
uses the Python 3 style print <em>function</em>.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>How do I find out what version of Biopython I have installed?</em></div>
<div class="line">Use this:</div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Bio</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Bio</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
</pre></div>
</div>
<p>If the “<code class="docutils literal notranslate"><span class="pre">import</span> <span class="pre">Bio</span></code>” line fails, Biopython is not installed. Note
that those are double underscores before and after version. If the
second line fails, your version is <em>very</em> out of date.</p>
<p>If the version string ends with a plus like “<code class="docutils literal notranslate"><span class="pre">1.66+</span></code>”, you don’t
have an official release, but an old snapshot of the in development
code <em>after</em> that version was released. This naming was used until
June 2016 in the run-up to Biopython 1.68.</p>
<p>If the version string ends with “<code class="docutils literal notranslate"><span class="pre">.dev&lt;number&gt;</span></code>” like
“<code class="docutils literal notranslate"><span class="pre">1.68.dev0</span></code>”, again you don’t have an official release, but
instead a snapshot of the in development code <em>before</em> that version
was released.</p>
</li>
<li><div class="line-block">
<div class="line"><em>Where is the latest version of this document?</em></div>
<div class="line">If you download a Biopython source code archive, it will include
the relevant version in both HTML and PDF formats. The latest
published version of this document (updated at each release) is
online:</div>
</div>
<ul class="simple">
<li><p><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">http://biopython.org/DIST/docs/tutorial/Tutorial.html</a></p></li>
<li><p><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.pdf">http://biopython.org/DIST/docs/tutorial/Tutorial.pdf</a></p></li>
</ul>
</li>
<li><div class="line-block">
<div class="line"><em>What is wrong with my sequence comparisons?</em></div>
<div class="line">There was a major change in Biopython 1.65 making the <code class="docutils literal notranslate"><span class="pre">Seq</span></code> and
<code class="docutils literal notranslate"><span class="pre">MutableSeq</span></code> classes (and subclasses) use simple string-based
comparison which you can do explicitly with
<code class="docutils literal notranslate"><span class="pre">str(seq1)</span> <span class="pre">==</span> <span class="pre">str(seq2)</span></code>.</div>
</div>
<p>Older versions of Biopython would use instance-based comparison for
<code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects which you can do explicitly with
<code class="docutils literal notranslate"><span class="pre">id(seq1)</span> <span class="pre">==</span> <span class="pre">id(seq2)</span></code>.</p>
<p>If you still need to support old versions of Biopython, use these
explicit forms to avoid problems. See
Section <a class="reference internal" href="chapter_seq_objects.html#sec-seq-comparison"><span class="std std-ref">Comparing Seq objects</span></a>.</p>
</li>
<li><div class="line-block">
<div class="line"><em>What file formats do</em> <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> <em>and</em> <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO</span></code> <em>read
and write?</em></div>
<div class="line">Check the built in docstrings (<code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">Bio</span> <span class="pre">import</span> <span class="pre">SeqIO</span></code>, then
<code class="docutils literal notranslate"><span class="pre">help(SeqIO)</span></code>), or see <a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a> and
<a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a> on the wiki for the latest
listing.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why won’t the</em> <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> <em>and</em> <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO</span></code> <em>functions</em>
<code class="docutils literal notranslate"><span class="pre">parse</span></code><em>,</em> <code class="docutils literal notranslate"><span class="pre">read</span></code> <em>and</em> <code class="docutils literal notranslate"><span class="pre">write</span></code> <em>take filenames? They
insist on handles!</em></div>
<div class="line">You need Biopython 1.54 or later, or just use handles explicitly
(see Section <a class="reference internal" href="chapter_appendix.html#sec-appendix-handles"><span class="std std-ref">What the heck is a handle?</span></a>).
It is especially important to remember to close output handles
explicitly after writing your data.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why won’t the</em> <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.write()</span></code> <em>and</em> <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.write()</span></code>
<em>functions accept a single record or alignment? They insist on a
list or iterator!</em></div>
<div class="line">You need Biopython 1.54 or later, or just wrap the item with
<code class="docutils literal notranslate"><span class="pre">[...]</span></code> to create a list of one element.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why doesn’t</em> <code class="docutils literal notranslate"><span class="pre">str(...)</span></code> <em>give me the full sequence of a</em> <code class="docutils literal notranslate"><span class="pre">Seq</span></code>
<em>object?</em></div>
<div class="line">You need Biopython 1.45 or later.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why doesn’t</em> <code class="docutils literal notranslate"><span class="pre">Bio.Blast</span></code> <em>work with the latest plain text NCBI
blast output?</em></div>
<div class="line">The NCBI keep tweaking the plain text output from the BLAST tools,
and keeping our parser up to date is/was an ongoing struggle. If
you aren’t using the latest version of Biopython, you could try
upgrading. However, we (and the NCBI) recommend you use the XML
output instead, which is designed to be read by a computer program.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why has my script using</em> <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch()</span></code> <em>stopped
working?</em></div>
<div class="line">This could be due to NCBI changes in February 2012 introducing
EFetch 2.0. First, they changed the default return modes - you
probably want to add <code class="docutils literal notranslate"><span class="pre">retmode=&quot;text&quot;</span></code> to your call. Second, they
are now stricter about how to provide a list of IDs – Biopython
1.59 onwards turns a list into a comma separated string
automatically.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why doesn’t</em> <code class="docutils literal notranslate"><span class="pre">Bio.Blast.NCBIWWW.qblast()</span></code> <em>give the same results
as the NCBI BLAST website?</em></div>
<div class="line">You need to specify the same options – the NCBI often adjust the
default settings on the website, and they do not match the QBLAST
defaults anymore. Check things like the gap penalties and
expectation threshold.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why can’t I add</em> <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> <em>objects together?</em></div>
<div class="line">You need Biopython 1.53 or later.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why doesn’t</em> <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index_db()</span></code> <em>work? The module imports
fine but there is no ``index_db`` function!</em></div>
<div class="line">You need Biopython 1.57 or later (and a Python with SQLite3
support).</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Where is the</em> <code class="docutils literal notranslate"><span class="pre">MultipleSeqAlignment</span></code> <em>object? The</em> <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code>
<em>module imports fine but this class isn’t there!</em></div>
<div class="line">You need Biopython 1.54 or later. Alternatively, the older
<code class="docutils literal notranslate"><span class="pre">Bio.Align.Generic.Alignment</span></code> class supports some of its
functionality, but using this is now discouraged.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why can’t I run command line tools directly from the application
wrappers?</em></div>
<div class="line">You need Biopython 1.55 or later, but these were deprecated in
Biopython 1.78. Consider using the Python <code class="docutils literal notranslate"><span class="pre">subprocess</span></code> module
directly.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>I looked in a directory for code, but I couldn’t find the code
that does something. Where’s it hidden?</em></div>
<div class="line">One thing to know is that we put code in <code class="docutils literal notranslate"><span class="pre">__init__.py</span></code> files. If
you are not used to looking for code in this file this can be
confusing. The reason we do this is to make the imports easier for
users. For instance, instead of having to do a “repetitive” import
like <code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">Bio.GenBank</span> <span class="pre">import</span> <span class="pre">GenBank</span></code>, you can just use
<code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">Bio</span> <span class="pre">import</span> <span class="pre">GenBank</span></code>.</div>
</div>
</li>
<li><div class="line-block">
<div class="line"><em>Why doesn’t</em> <code class="docutils literal notranslate"><span class="pre">Bio.Fasta</span></code> <em>work?</em></div>
<div class="line">We deprecated the <code class="docutils literal notranslate"><span class="pre">Bio.Fasta</span></code> module in Biopython 1.51 (August
2009) and removed it in Biopython 1.55 (August 2010). There is a
brief example showing how to convert old code to use <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code>
instead in the
<a class="reference external" href="https://github.com/biopython/biopython/blob/master/DEPRECATED.rst">DEPRECATED.rst</a>
file.</div>
</div>
</li>
</ol>
<p>For more general questions, the Python FAQ pages
<a class="reference external" href="https://docs.python.org/3/faq/index.html">https://docs.python.org/3/faq/index.html</a> may be useful.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Biopython Tutorial &amp; Cookbook" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_quick_start.html" class="btn btn-neutral float-right" title="Quick Start – What can you do with Biopython?" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.83.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../1.81/">Previous</a></dd>
        
          <dd><a href="../latest/">Latest</a></dd>
        
          <dd><a href="../dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>