<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bio.PopGen: Population genetics &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/biopython.css?v=d1305275" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=49cfa8a4"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Phylogenetics with Bio.Phylo" href="chapter_phylo.html" />
    <link rel="prev" title="Going 3D: The PDB module" href="chapter_pdb.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Biopython Tutorial &amp; Cookbook</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="chapter_introduction.html">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_quick_start.html">Quick Start – What can you do with Biopython?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_objects.html">Sequence objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_annot.html">Sequence annotation objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seqio.html">Sequence Input/Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_align.html">Sequence alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise.html">Pairwise sequence alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_msa.html">Multiple Sequence Alignment objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise2.html">Pairwise alignments using pairwise2</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html">BLAST (new)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html#blast-old">BLAST (old)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_searchio.html">BLAST and other sequence search tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_entrez.html">Accessing NCBI’s Entrez databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_uniprot.html">Swiss-Prot and ExPASy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pdb.html">Going 3D: The PDB module</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Bio.PopGen: Population genetics</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genepop">GenePop</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phylo.html">Phylogenetics with Bio.Phylo</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_motifs.html">Sequence motif analysis using Bio.motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cluster.html">Cluster analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_graphics.html">Graphics including GenomeDiagram</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_kegg.html">KEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phenotype.html">Bio.phenotype: analyze phenotypic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cookbook.html">Cookbook – Cool things to do with it</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_testing.html">The Biopython testing framework</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_contributing.html">Where to go from here – contributing to Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_appendix.html">Appendix: Useful stuff about Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_bibliography.html">Bibliography</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Biopython Tutorial &amp; Cookbook</a></li>
      <li class="breadcrumb-item active">Bio.PopGen: Population genetics</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/Tutorial/chapter_popgen.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="chapter_pdb.html" class="btn btn-neutral float-left" title="Going 3D: The PDB module" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_phylo.html" class="btn btn-neutral float-right" title="Phylogenetics with Bio.Phylo" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bio-popgen-population-genetics">
<span id="chapter-popgen"></span><h1>Bio.PopGen: Population genetics<a class="headerlink" href="#bio-popgen-population-genetics" title="Link to this heading"></a></h1>
<p><code class="docutils literal notranslate"><span class="pre">Bio.PopGen</span></code> is a Biopython module supporting population genetics,
available in Biopython 1.44 onwards. The objective for the module is to
support widely used data formats, applications and databases.</p>
<section id="genepop">
<h2>GenePop<a class="headerlink" href="#genepop" title="Link to this heading"></a></h2>
<p>GenePop (<a class="reference external" href="http://genepop.curtin.edu.au/">http://genepop.curtin.edu.au/</a>) is a popular population genetics
software package supporting Hardy-Weinberg tests, linkage
disequilibrium, population differentiation, basic statistics,
<span class="math notranslate nohighlight">\(F_{st}\)</span> and migration estimates, among others. GenePop does not
supply sequence based statistics as it doesn’t handle sequence data. The
GenePop file format is supported by a wide range of other population
genetic software applications, thus making it a relevant format in the
population genetics field.</p>
<p><code class="docutils literal notranslate"><span class="pre">Bio.PopGen</span></code> provides a parser and generator of GenePop file format.
Utilities to manipulate the content of a record are also provided. Here
is an example on how to read a GenePop file (you can find example
GenePop data files in the Test/PopGen directory of Biopython):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.PopGen</span> <span class="kn">import</span> <span class="n">GenePop</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;example.gen&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
    <span class="n">rec</span> <span class="o">=</span> <span class="n">GenePop</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>
</div>
<p>This will read a file called example.gen and parse it. If you do print
rec, the record will be output again, in GenePop format.</p>
<p>The most important information in rec will be the loci names and
population information (but there is more – use help(GenePop.Record) to
check the API documentation). Loci names can be found on rec.loci_list.
Population information can be found on rec.populations. Populations is a
list with one element per population. Each element is itself a list of
individuals, each individual is a pair composed by individual name and a
list of alleles (2 per marker), here is an example for rec.populations:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">[</span>
    <span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;Ind1&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="mi">201</span><span class="p">)]),</span>
        <span class="p">(</span><span class="s2">&quot;Ind2&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">2</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)]),</span>
    <span class="p">],</span>
    <span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;Other1&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="mi">200</span><span class="p">)]),</span>
    <span class="p">],</span>
<span class="p">]</span>
</pre></div>
</div>
<p>So we have two populations, the first with two individuals, the second
with only one. The first individual of the first population is called
Ind1, allelic information for each of the 3 loci follows. Please note
that for any locus, information might be missing (see as an example,
Ind2 above).</p>
<p>A few utility functions to manipulate GenePop records are made
available, here is an example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.PopGen</span> <span class="kn">import</span> <span class="n">GenePop</span>

<span class="c1"># Imagine that you have loaded rec, as per the code snippet above...</span>

<span class="n">rec</span><span class="o">.</span><span class="n">remove_population</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
<span class="c1"># Removes a population from a record, pos is the population position in</span>
<span class="c1"># rec.populations, remember that it starts on position 0.</span>
<span class="c1"># rec is altered.</span>

<span class="n">rec</span><span class="o">.</span><span class="n">remove_locus_by_position</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
<span class="c1"># Removes a locus by its position, pos is the locus position in</span>
<span class="c1">#  rec.loci_list, remember that it starts on position 0.</span>
<span class="c1">#  rec is altered.</span>

<span class="n">rec</span><span class="o">.</span><span class="n">remove_locus_by_name</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
<span class="c1"># Removes a locus by its name, name is the locus name as in</span>
<span class="c1"># rec.loci_list. If the name doesn&#39;t exist the function fails</span>
<span class="c1"># silently.</span>
<span class="c1"># rec is altered.</span>

<span class="n">rec_loci</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">split_in_loci</span><span class="p">()</span>
<span class="c1"># Splits a record in loci, that is, for each loci, it creates a new</span>
<span class="c1"># record, with a single loci and all populations.</span>
<span class="c1"># The result is returned in a dictionary, being each key the locus name.</span>
<span class="c1"># The value is the GenePop record.</span>
<span class="c1"># rec is not altered.</span>

<span class="n">rec_pops</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">split_in_pops</span><span class="p">(</span><span class="n">pop_names</span><span class="p">)</span>
<span class="c1"># Splits a record in populations, that is, for each population, it creates</span>
<span class="c1"># a new record, with a single population and all loci.</span>
<span class="c1"># The result is returned in a dictionary, being each key</span>
<span class="c1"># the population name. As population names are not available in GenePop,</span>
<span class="c1"># they are passed in array (pop_names).</span>
<span class="c1"># The value of each dictionary entry is the GenePop record.</span>
<span class="c1"># rec is not altered.</span>
</pre></div>
</div>
<p>GenePop does not support population names, a limitation which can be
cumbersome at times. Functionality to enable population names is
currently being planned for Biopython. These extensions won’t break
compatibility in any way with the standard format. In the medium term,
we would also like to support the GenePop web service.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="chapter_pdb.html" class="btn btn-neutral float-left" title="Going 3D: The PDB module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_phylo.html" class="btn btn-neutral float-right" title="Phylogenetics with Bio.Phylo" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>