<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Sequence alignments &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/biopython.css?v=d1305275" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=49cfa8a4"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Pairwise sequence alignment" href="chapter_pairwise.html" />
    <link rel="prev" title="Sequence Input/Output" href="chapter_seqio.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Biopython Tutorial &amp; Cookbook</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="chapter_introduction.html">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_quick_start.html">Quick Start – What can you do with Biopython?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_objects.html">Sequence objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_annot.html">Sequence annotation objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seqio.html">Sequence Input/Output</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Sequence alignments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#alignment-objects">Alignment objects</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#creating-an-alignment-object-from-sequences-and-coordinates">Creating an Alignment object from sequences and coordinates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#creating-an-alignment-object-from-aligned-sequences">Creating an Alignment object from aligned sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#common-alignment-attributes">Common alignment attributes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#slicing-and-indexing-an-alignment">Slicing and indexing an alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#getting-information-about-the-alignment">Getting information about the alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#alignment-shape">Alignment shape</a></li>
<li class="toctree-l4"><a class="reference internal" href="#comparing-alignments">Comparing alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#finding-the-indices-of-aligned-sequences">Finding the indices of aligned sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#counting-identities-mismatches-and-gaps">Counting identities, mismatches, and gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="#letter-frequencies">Letter frequencies</a></li>
<li class="toctree-l4"><a class="reference internal" href="#substitutions">Substitutions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#alignments-as-arrays">Alignments as arrays</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#operations-on-an-alignment">Operations on an alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#sorting-an-alignment">Sorting an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reverse-complementing-the-alignment">Reverse-complementing the alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#adding-alignments">Adding alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mapping-a-pairwise-sequence-alignment">Mapping a pairwise sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mapping-a-multiple-sequence-alignment">Mapping a multiple sequence alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#the-alignments-class">The Alignments class</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reading-and-writing-alignments">Reading and writing alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#reading-alignments">Reading alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#writing-alignments">Writing alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#printing-alignments">Printing alignments</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#alignment-file-formats">Alignment file formats</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#aligned-fasta">Aligned FASTA</a></li>
<li class="toctree-l4"><a class="reference internal" href="#clustalw">ClustalW</a></li>
<li class="toctree-l4"><a class="reference internal" href="#stockholm">Stockholm</a></li>
<li class="toctree-l4"><a class="reference internal" href="#phylip-output-files">PHYLIP output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#emboss">EMBOSS</a></li>
<li class="toctree-l4"><a class="reference internal" href="#gcg-multiple-sequence-format-msf">GCG Multiple Sequence Format (MSF)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#exonerate">Exonerate</a></li>
<li class="toctree-l4"><a class="reference internal" href="#nexus">NEXUS</a></li>
<li class="toctree-l4"><a class="reference internal" href="#tabular-output-from-blast-or-fasta">Tabular output from BLAST or FASTA</a></li>
<li class="toctree-l4"><a class="reference internal" href="#hh-suite-output-files">HH-suite output files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#a2m">A2M</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mauve-extended-multi-fasta-xmfa-format">Mauve eXtended Multi-FastA (xmfa) format</a></li>
<li class="toctree-l4"><a class="reference internal" href="#sequence-alignment-map-sam">Sequence Alignment/Map (SAM)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#browser-extensible-data-bed">Browser Extensible Data (BED)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bigbed">bigBed</a></li>
<li class="toctree-l4"><a class="reference internal" href="#pattern-space-layout-psl">Pattern Space Layout (PSL)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bigpsl">bigPsl</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multiple-alignment-format-maf">Multiple Alignment Format (MAF)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bigmaf">bigMaf</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ucsc-chain-file-format">UCSC chain file format</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise.html">Pairwise sequence alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_msa.html">Multiple Sequence Alignment objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise2.html">Pairwise alignments using pairwise2</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html">BLAST (new)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html#blast-old">BLAST (old)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_searchio.html">BLAST and other sequence search tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_entrez.html">Accessing NCBI’s Entrez databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_uniprot.html">Swiss-Prot and ExPASy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pdb.html">Going 3D: The PDB module</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_popgen.html">Bio.PopGen: Population genetics</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phylo.html">Phylogenetics with Bio.Phylo</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_motifs.html">Sequence motif analysis using Bio.motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cluster.html">Cluster analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_graphics.html">Graphics including GenomeDiagram</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_kegg.html">KEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phenotype.html">Bio.phenotype: analyze phenotypic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cookbook.html">Cookbook – Cool things to do with it</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_testing.html">The Biopython testing framework</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_contributing.html">Where to go from here – contributing to Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_appendix.html">Appendix: Useful stuff about Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_bibliography.html">Bibliography</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Biopython Tutorial &amp; Cookbook</a></li>
      <li class="breadcrumb-item active">Sequence alignments</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/Tutorial/chapter_align.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="chapter_seqio.html" class="btn btn-neutral float-left" title="Sequence Input/Output" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_pairwise.html" class="btn btn-neutral float-right" title="Pairwise sequence alignment" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="sequence-alignments">
<span id="chapter-align"></span><h1>Sequence alignments<a class="headerlink" href="#sequence-alignments" title="Link to this heading"></a></h1>
<p>Sequence alignments are a collection of two or more sequences that have
been aligned to each other – usually with the insertion of gaps, and the
addition of leading or trailing gaps – such that all the sequence
strings are the same length.</p>
<p>Alignments may extend over the full length of each sequence, or may be
limited to a subsection of each sequence. In Biopython, all sequence
alignments are represented by an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object, described in
section <a class="reference internal" href="#sec-alignmentobject"><span class="std std-ref">Alignment objects</span></a>. <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects can be
obtained by parsing the output of alignment software such as Clustal or
BLAT (described in section <a class="reference internal" href="#sec-alignmentparsers"><span class="std std-ref">Reading and writing alignments</span></a>. or by using
Biopython’s pairwise sequence aligner, which can align two sequences to
each other (described in
Chapter <a class="reference internal" href="chapter_pairwise.html#chapter-pairwise"><span class="std std-ref">Pairwise sequence alignment</span></a>).</p>
<p>See Chapter <a class="reference internal" href="chapter_msa.html#chapter-msa"><span class="std std-ref">Multiple Sequence Alignment objects</span></a> for a description of the
older <code class="docutils literal notranslate"><span class="pre">MultipleSeqAlignment</span></code> class and the parsers in <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO</span></code>
that parse the output of sequence alignment software, generating
<code class="docutils literal notranslate"><span class="pre">MultipleSeqAlignment</span></code> objects.</p>
<section id="alignment-objects">
<span id="sec-alignmentobject"></span><h2>Alignment objects<a class="headerlink" href="#alignment-objects" title="Link to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> class is defined in <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code>. Usually you would
get an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object by parsing the output of alignment programs
(section <a class="reference internal" href="#sec-alignmentparsers"><span class="std std-ref">Reading and writing alignments</span></a>) or by running Biopython’s
pairwise aligner (Chapter <a class="reference internal" href="chapter_pairwise.html#chapter-pairwise"><span class="std std-ref">Pairwise sequence alignment</span></a>).
For the benefit of this section, however, we will create an
<code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from scratch.</p>
<section id="creating-an-alignment-object-from-sequences-and-coordinates">
<span id="subsec-align-sequences-coordinates"></span><h3>Creating an Alignment object from sequences and coordinates<a class="headerlink" href="#creating-an-alignment-object-from-sequences-and-coordinates" title="Link to this heading"></a></h3>
<p>Suppose you have three sequences:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqA</span> <span class="o">=</span> <span class="s2">&quot;CCGGTTTTT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqB</span> <span class="o">=</span> <span class="s2">&quot;AGTTTAA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqC</span> <span class="o">=</span> <span class="s2">&quot;AGGTTT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">seqA</span><span class="p">,</span> <span class="n">seqB</span><span class="p">,</span> <span class="n">seqC</span><span class="p">]</span>
</pre></div>
</div>
<p>To create an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object, we also need the coordinates that
define how the sequences are aligned to each other. We use a NumPy array
for that:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">]])</span>
</pre></div>
</div>
<p>These coordinates define the alignment for the following sequence
segments:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">SeqA[1:3]</span></code>, <code class="docutils literal notranslate"><span class="pre">SeqB[0:2]</span></code>, and <code class="docutils literal notranslate"><span class="pre">SeqC[0:2]</span></code> are aligned to each
other;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqA[3:4]</span></code> and <code class="docutils literal notranslate"><span class="pre">SeqC[2:3]</span></code> are aligned to each other, with a gap
of one nucleotide in <code class="docutils literal notranslate"><span class="pre">seqB</span></code>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqA[4:7]</span></code>, <code class="docutils literal notranslate"><span class="pre">SeqB[2:5]</span></code>, and <code class="docutils literal notranslate"><span class="pre">SeqC[3:6]</span></code> are aligned to each
other;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqA[7:9]</span></code> is not aligned to <code class="docutils literal notranslate"><span class="pre">seqB</span></code> or <code class="docutils literal notranslate"><span class="pre">seqC</span></code>.</p></li>
</ul>
<p>Note that the alignment does not include the first nucleotide of
<code class="docutils literal notranslate"><span class="pre">seqA</span></code> and last two nucleotides of <code class="docutils literal notranslate"><span class="pre">seqB</span></code>.</p>
<p>Now we can create the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">Alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">coordinates</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (3 rows x 8 columns) at ...&gt;</span>
</pre></div>
</div>
<p>The alignment object has an attribute <code class="docutils literal notranslate"><span class="pre">sequences</span></code> pointing to the
sequences included in this alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;CCGGTTTTT&#39;, &#39;AGTTTAA&#39;, &#39;AGGTTT&#39;]</span>
</pre></div>
</div>
<p>and an attribute <code class="docutils literal notranslate"><span class="pre">coordinates</span></code> with the alignment coordinates:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[1, 3, 4, 7, 9],</span>
<span class="go">       [0, 2, 2, 5, 5],</span>
<span class="go">       [0, 2, 3, 6, 6]])</span>
</pre></div>
</div>
<p>Print the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object to show the alignment explicitly:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">                  1 CGGTTTTT 9</span>
<span class="go">                  0 AG-TTT-- 5</span>
<span class="go">                  0 AGGTTT-- 6</span>
</pre></div>
</div>
<p>with the starting and end coordinate for each sequence are shown to the
left and right, respectively, of the alignment.</p>
</section>
<section id="creating-an-alignment-object-from-aligned-sequences">
<span id="subsec-align-infer-coordinates"></span><h3>Creating an Alignment object from aligned sequences<a class="headerlink" href="#creating-an-alignment-object-from-aligned-sequences" title="Link to this heading"></a></h3>
<p>If you start out with the aligned sequences, with dashes representing
gaps, then you can calculate the coordinates using the
<code class="docutils literal notranslate"><span class="pre">infer_coordinates</span></code> class method. This method is primarily employed in
Biopython’s alignment parsers (see
Section <a class="reference internal" href="#sec-alignmentparsers"><span class="std std-ref">Reading and writing alignments</span></a>), but it may be useful for other
purposes. For example, you can construct the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from
aligned sequences as follows:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligned_sequences</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CGGTTTTT&quot;</span><span class="p">,</span> <span class="s2">&quot;AG-TTT--&quot;</span><span class="p">,</span> <span class="s2">&quot;AGGTTT--&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">aligned_sequence</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="gp">... </span>             <span class="k">for</span> <span class="n">aligned_sequence</span> <span class="ow">in</span> <span class="n">aligned_sequences</span><span class="p">]</span>  <span class="c1"># fmt: skip</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span>
<span class="go">[&#39;CGGTTTTT&#39;, &#39;AGTTT&#39;, &#39;AGGTTT&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span> <span class="o">=</span> <span class="n">Alignment</span><span class="o">.</span><span class="n">infer_coordinates</span><span class="p">(</span><span class="n">aligned_sequences</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span>
<span class="go">array([[0, 2, 3, 6, 8],</span>
<span class="go">       [0, 2, 2, 5, 5],</span>
<span class="go">       [0, 2, 3, 6, 6]])</span>
</pre></div>
</div>
<p>The initial <code class="docutils literal notranslate"><span class="pre">G</span></code> nucleotide of <code class="docutils literal notranslate"><span class="pre">seqA</span></code> and the final <code class="docutils literal notranslate"><span class="pre">CC</span></code>
nucleotides of <code class="docutils literal notranslate"><span class="pre">seqB</span></code> were not included in the alignment and is
therefore missing here. But this is easy to fix:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;C&quot;</span> <span class="o">+</span> <span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;AA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span>
<span class="go">[&#39;CCGGTTTTT&#39;, &#39;AGTTTAA&#39;, &#39;AGGTTT&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span>
<span class="go">array([[1, 3, 4, 7, 9],</span>
<span class="go">       [0, 2, 2, 5, 5],</span>
<span class="go">       [0, 2, 3, 6, 6]])</span>
</pre></div>
</div>
<p>Now we can create the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">coordinates</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">                  1 CGGTTTTT 9</span>
<span class="go">                  0 AG-TTT-- 5</span>
<span class="go">                  0 AGGTTT-- 6</span>
</pre></div>
</div>
<p>which identical to the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object created above in
section <a class="reference internal" href="#subsec-align-sequences-coordinates"><span class="std std-ref">Creating an Alignment object from sequences and coordinates</span></a>.</p>
<p>By default, the <code class="docutils literal notranslate"><span class="pre">coordinates</span></code> argument to the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code>
initializer is <code class="docutils literal notranslate"><span class="pre">None</span></code>, which assumes that there are no gaps in the
alignment. All sequences in an ungapped alignment must have the same
length. If the <code class="docutils literal notranslate"><span class="pre">coordinates</span></code> argument is <code class="docutils literal notranslate"><span class="pre">None</span></code>, then the
initializer will fill in the <code class="docutils literal notranslate"><span class="pre">coordinates</span></code> attribute of the
<code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object for you:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ungapped_alignment</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">([</span><span class="s2">&quot;ACGTACGT&quot;</span><span class="p">,</span> <span class="s2">&quot;AAGTACGT&quot;</span><span class="p">,</span> <span class="s2">&quot;ACGTACCT&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ungapped_alignment</span>  
<span class="go">&lt;Alignment object (3 rows x 8 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ungapped_alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[0, 8],</span>
<span class="go">       [0, 8],</span>
<span class="go">       [0, 8]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ungapped_alignment</span><span class="p">)</span>
<span class="go">                  0 ACGTACGT 8</span>
<span class="go">                  0 AAGTACGT 8</span>
<span class="go">                  0 ACGTACCT 8</span>
</pre></div>
</div>
</section>
<section id="common-alignment-attributes">
<span id="subsec-align-common-attributes"></span><h3>Common alignment attributes<a class="headerlink" href="#common-alignment-attributes" title="Link to this heading"></a></h3>
<p>The following attributes are commonly found on <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">sequences</span></code>: This is a list of the sequences aligned to each other.
Depending on how the alignment was created, the sequences can have
the following types:</p>
<ul class="simple">
<li><p>plain Python string;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq</span></code>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">MutableSeq</span></code>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bytes</span></code>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bytearray</span></code>;</p></li>
<li><p>NumPy array with data type <code class="docutils literal notranslate"><span class="pre">numpy.int32</span></code>;</p></li>
<li><p>any other object with a contiguous buffer of format <code class="docutils literal notranslate"><span class="pre">&quot;c&quot;</span></code>,
<code class="docutils literal notranslate"><span class="pre">&quot;B&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;i&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;I&quot;</span></code>;</p></li>
<li><p>lists or tuples of objects defined in the <code class="docutils literal notranslate"><span class="pre">alphabet</span></code> attribute
of the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object that created the alignment (see
section <a class="reference internal" href="chapter_pairwise.html#sec-generalized-pairwise"><span class="std std-ref">Generalized pairwise alignments</span></a>).</p></li>
</ul>
<p>For pairwise alignments (meaning an alignment of two sequences), the
properties <code class="docutils literal notranslate"><span class="pre">target</span></code> and <code class="docutils literal notranslate"><span class="pre">query</span></code> are aliases for <code class="docutils literal notranslate"><span class="pre">sequences[0]</span></code>
and <code class="docutils literal notranslate"><span class="pre">sequences[1]</span></code>, respectively.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">coordinates</span></code>: A NumPy array of integers storing the sequence
indices defining how the sequences are aligned to each other;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">score</span></code>: The alignment score, as found by the parser in the
alignment file, or as calculated by the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> (see
section <a class="reference internal" href="chapter_pairwise.html#sec-pairwise-basic"><span class="std std-ref">Basic usage</span></a>);</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">annotations</span></code>: A dictionary storing most other annotations
associated with the alignment;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">column_annotations</span></code>: A dictionary storing annotations that extend
along the alignment and have the same length as the alignment, such
as a consensus sequence (see
section <a class="reference internal" href="#subsec-align-clustal"><span class="std std-ref">ClustalW</span></a> for an example).</p></li>
</ul>
<p>An <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object created by the parser in <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> may have
additional attributes, depending on the alignment file format from which
the alignment was read.</p>
</section>
</section>
<section id="slicing-and-indexing-an-alignment">
<h2>Slicing and indexing an alignment<a class="headerlink" href="#slicing-and-indexing-an-alignment" title="Link to this heading"></a></h2>
<p>Slices of the form <code class="docutils literal notranslate"><span class="pre">alignment[k,</span> <span class="pre">i:j]</span></code>, where <code class="docutils literal notranslate"><span class="pre">k</span></code> is an integer and
<code class="docutils literal notranslate"><span class="pre">i</span></code> and <code class="docutils literal notranslate"><span class="pre">j</span></code> are integers or are absent, return a string showing the
aligned sequence (including gaps) for the target (if <code class="docutils literal notranslate"><span class="pre">k=0</span></code>) or the
query (if <code class="docutils literal notranslate"><span class="pre">k=1</span></code>) that includes only the columns <code class="docutils literal notranslate"><span class="pre">i</span></code> through <code class="docutils literal notranslate"><span class="pre">j</span></code> in
the printed alignment.</p>
<p>To illustrate this, in the following example the printed alignment has 8
columns:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">                  1 CGGTTTTT 9</span>
<span class="go">                  0 AG-TTT-- 5</span>
<span class="go">                  0 AGGTTT-- 6</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">length</span>
<span class="go">8</span>
</pre></div>
</div>
<p>To get the aligned sequence strings individually, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;CGGTTTTT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;AG-TTT--&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;AGGTTT--&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span>
<span class="go">&#39;CGGTTTTT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
<span class="go">&#39;AG-TTT--&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;GGTTTT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;G-TTT-&#39;</span>
</pre></div>
</div>
<p>Columns to be included can also be selected using an iterable over
integers:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)]</span>
<span class="go">&#39;GGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">)]</span>
<span class="go">&#39;A-T&#39;</span>
</pre></div>
</div>
<p>To get the letter at position <code class="docutils literal notranslate"><span class="pre">[i,</span> <span class="pre">j]</span></code> of the printed alignment, use
<code class="docutils literal notranslate"><span class="pre">alignment[i,</span> <span class="pre">j]</span></code>; this will return <code class="docutils literal notranslate"><span class="pre">&quot;-&quot;</span></code> if a gap is found at that
position:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;G&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">]</span>
<span class="go">&#39;-&#39;</span>
</pre></div>
</div>
<p>To get specific columns in the alignment, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;CAA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;GGG&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">]</span>
<span class="go">&#39;G-G&#39;</span>
</pre></div>
</div>
<p>Slices of the form <code class="docutils literal notranslate"><span class="pre">alignment[i:j:k]</span></code> return a new <code class="docutils literal notranslate"><span class="pre">Alignment</span></code>
object including only sequences <code class="docutils literal notranslate"><span class="pre">[i:j:k]</span></code> of the alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>  
<span class="go">&lt;Alignment object (2 rows x 6 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
<span class="go">target            0 AG-TTT 5</span>
<span class="go">                  0 ||-||| 6</span>
<span class="go">query             0 AGGTTT 6</span>
</pre></div>
</div>
<p>Slices of the form <code class="docutils literal notranslate"><span class="pre">alignment[:,</span> <span class="pre">i:j]</span></code>, where <code class="docutils literal notranslate"><span class="pre">i</span></code> and <code class="docutils literal notranslate"><span class="pre">j</span></code> are
integers or are absent, return a new <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object that includes
only the columns <code class="docutils literal notranslate"><span class="pre">i</span></code> through <code class="docutils literal notranslate"><span class="pre">j</span></code> in the printed alignment.</p>
<p>Extracting the first 4 columns for the example alignment above gives:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">]</span>  
<span class="go">&lt;Alignment object (3 rows x 4 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">])</span>
<span class="go">                  1 CGGT 5</span>
<span class="go">                  0 AG-T 3</span>
<span class="go">                  0 AGGT 4</span>
</pre></div>
</div>
<p>Similarly, extracting the last 6 columns gives:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">6</span><span class="p">:]</span>  
<span class="go">&lt;Alignment object (3 rows x 6 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">6</span><span class="p">:])</span>
<span class="go">                  3 GTTTTT 9</span>
<span class="go">                  2 -TTT-- 5</span>
<span class="go">                  2 GTTT-- 6</span>
</pre></div>
</div>
<p>The column index can also be an iterable of integers:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">)])</span>
<span class="go">                  0 GTC 3</span>
<span class="go">                  0 GTA 3</span>
<span class="go">                  0 GTA 3</span>
</pre></div>
</div>
<p>Calling <code class="docutils literal notranslate"><span class="pre">alignment[:,</span> <span class="pre">:]</span></code> returns a copy of the alignment.</p>
</section>
<section id="getting-information-about-the-alignment">
<h2>Getting information about the alignment<a class="headerlink" href="#getting-information-about-the-alignment" title="Link to this heading"></a></h2>
<section id="alignment-shape">
<h3>Alignment shape<a class="headerlink" href="#alignment-shape" title="Link to this heading"></a></h3>
<p>The number of aligned sequences is returned by <code class="docutils literal notranslate"><span class="pre">len(alignment)</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
<p>The alignment length is defined as the number of columns in the
alignment as printed. This is equal to the sum of the number of matches,
number of mismatches, and the total length of gaps in each sequence:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">length</span>
<span class="go">8</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">shape</span></code> property returns a tuple consisting of the length of the
alignment and the number of columns in the alignment as printed:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(3, 8)</span>
</pre></div>
</div>
</section>
<section id="comparing-alignments">
<h3>Comparing alignments<a class="headerlink" href="#comparing-alignments" title="Link to this heading"></a></h3>
<p>Two alignments are equal to each other (meaning that
<code class="docutils literal notranslate"><span class="pre">alignment1</span> <span class="pre">==</span> <span class="pre">alignment2</span></code> evaluates to <code class="docutils literal notranslate"><span class="pre">True</span></code>) if each of the
sequences in <code class="docutils literal notranslate"><span class="pre">alignment1.sequences</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment2.sequences</span></code> are
equal to each other, and <code class="docutils literal notranslate"><span class="pre">alignment1.coordinates</span></code> and
<code class="docutils literal notranslate"><span class="pre">alignment2.coordinates</span></code> contain the same coordinates. If either of
these conditions is not fulfilled, then <code class="docutils literal notranslate"><span class="pre">alignment1</span> <span class="pre">==</span> <span class="pre">alignment2</span></code>
evaluates to <code class="docutils literal notranslate"><span class="pre">False</span></code>. Inequality of two alignments (e.g.,
<code class="docutils literal notranslate"><span class="pre">alignment1</span> <span class="pre">&lt;</span> <span class="pre">alignment2</span></code>) is established by first comparing
<code class="docutils literal notranslate"><span class="pre">alignment1.sequences</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment2.sequences</span></code>, and if they are
equal, by comparing <code class="docutils literal notranslate"><span class="pre">alignment1.coordinates</span></code> to
<code class="docutils literal notranslate"><span class="pre">alignment2.coordinates</span></code>.</p>
</section>
<section id="finding-the-indices-of-aligned-sequences">
<h3>Finding the indices of aligned sequences<a class="headerlink" href="#finding-the-indices-of-aligned-sequences" title="Link to this heading"></a></h3>
<p>For pairwise alignments, the <code class="docutils literal notranslate"><span class="pre">aligned</span></code> property of an alignment
returns the start and end indices of subsequences in the target and
query sequence that were aligned to each other. If the alignment between
target (t) and query (q) consists of <span class="math notranslate nohighlight">\(N\)</span> chunks, you get two
tuples of length <span class="math notranslate nohighlight">\(N\)</span>:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>(((t_start1, t_end1), (t_start2, t_end2), ..., (t_startN, t_endN)),
 ((q_start1, q_end1), (q_start2, q_end2), ..., (q_startN, q_endN)))
</pre></div>
</div>
<p>For example,</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pairwise_alignment</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[:</span><span class="mi">2</span><span class="p">,</span> <span class="p">:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_alignment</span><span class="p">)</span>
<span class="go">target            1 CGGTTTTT 9</span>
<span class="go">                  0 .|-|||-- 8</span>
<span class="go">query             0 AG-TTT-- 5</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">pairwise_alignment</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[1, 3],</span>
<span class="go">        [4, 7]],</span>

<span class="go">       [[0, 2],</span>
<span class="go">        [2, 5]]])</span>
</pre></div>
</div>
<p>Note that different alignments may have the same subsequences aligned to
each other. In particular, this may occur if alignments differ from each
other in terms of their gap placement only:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pairwise_alignment1</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">([</span><span class="s2">&quot;AAACAAA&quot;</span><span class="p">,</span> <span class="s2">&quot;AAAGAAA&quot;</span><span class="p">],</span>
<span class="gp">... </span>                                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">]]))</span>  <span class="c1"># fmt: skip</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pairwise_alignment2</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">([</span><span class="s2">&quot;AAACAAA&quot;</span><span class="p">,</span> <span class="s2">&quot;AAAGAAA&quot;</span><span class="p">],</span>
<span class="gp">... </span>                                <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">]]))</span>  <span class="c1"># fmt: skip</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_alignment1</span><span class="p">)</span>
<span class="go">target            0 AAAC-AAA 7</span>
<span class="go">                  0 |||--||| 8</span>
<span class="go">query             0 AAA-GAAA 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_alignment2</span><span class="p">)</span>
<span class="go">target            0 AAA-CAAA 7</span>
<span class="go">                  0 |||--||| 8</span>
<span class="go">query             0 AAAG-AAA 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">pairwise_alignment1</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 3],</span>
<span class="go">        [4, 7]],</span>

<span class="go">       [[0, 3],</span>
<span class="go">        [4, 7]]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pairwise_alignment2</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 3],</span>
<span class="go">        [4, 7]],</span>

<span class="go">       [[0, 3],</span>
<span class="go">        [4, 7]]])</span>
</pre></div>
</div>
<p>The property <code class="docutils literal notranslate"><span class="pre">indices</span></code> returns a 2D NumPy array with the sequence
index of each letter in the alignment, with gaps indicated by -1:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">                  1 CGGTTTTT 9</span>
<span class="go">                  0 AG-TTT-- 5</span>
<span class="go">                  0 AGGTTT-- 6</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">indices</span>
<span class="go">array([[ 1,  2,  3,  4,  5,  6,  7,  8],</span>
<span class="go">       [ 0,  1, -1,  2,  3,  4, -1, -1],</span>
<span class="go">       [ 0,  1,  2,  3,  4,  5, -1, -1]])</span>
</pre></div>
</div>
<p>The property <code class="docutils literal notranslate"><span class="pre">inverse_indices</span></code> returns a list of 1D NumPy arrays, one
for each of the aligned sequences, with the column index in the
alignment for each letter in the sequence. Letters not included in the
alignment are indicated by -1:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;CCGGTTTTT&#39;, &#39;AGTTTAA&#39;, &#39;AGGTTT&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">inverse_indices</span>  
<span class="go">[array([-1,  0,  1,  2,  3,  4,  5,  6,  7]),</span>
<span class="go"> array([ 0,  1,  3,  4,  5, -1, -1]),</span>
<span class="go"> array([0, 1, 2, 3, 4, 5])]</span>
</pre></div>
</div>
</section>
<section id="counting-identities-mismatches-and-gaps">
<h3>Counting identities, mismatches, and gaps<a class="headerlink" href="#counting-identities-mismatches-and-gaps" title="Link to this heading"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">counts</span></code> method calculates the number of identities, mismatches,
and gaps of a pairwise alignment. For an alignment of more than two
sequences, the number of identities, mismatches, and gaps are calculated
and summed for all pairs of sequences in the alignment. The three
numbers are returned as an <code class="docutils literal notranslate"><span class="pre">AlignmentCounts</span></code> object, which is a
<code class="docutils literal notranslate"><span class="pre">namedtuple</span></code> with fields <code class="docutils literal notranslate"><span class="pre">gaps</span></code>, <code class="docutils literal notranslate"><span class="pre">identities</span></code>, and <code class="docutils literal notranslate"><span class="pre">mismatches</span></code>.
This method currently takes no arguments, but in the future will likely
be modified to accept optional arguments allowing its behavior to be
customized.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_alignment</span><span class="p">)</span>
<span class="go">target            1 CGGTTTTT 9</span>
<span class="go">                  0 .|-|||-- 8</span>
<span class="go">query             0 AG-TTT-- 5</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">pairwise_alignment</span><span class="o">.</span><span class="n">counts</span><span class="p">()</span>
<span class="go">AlignmentCounts(gaps=3, identities=4, mismatches=1)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">                  1 CGGTTTTT 9</span>
<span class="go">                  0 AG-TTT-- 5</span>
<span class="go">                  0 AGGTTT-- 6</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">counts</span><span class="p">()</span>
<span class="go">AlignmentCounts(gaps=8, identities=14, mismatches=2)</span>
</pre></div>
</div>
</section>
<section id="letter-frequencies">
<h3>Letter frequencies<a class="headerlink" href="#letter-frequencies" title="Link to this heading"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">frequencies</span></code> method calculates how often each letter appears in
each column of the alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">frequencies</span>  
<span class="go">{&#39;C&#39;: array([1., 0., 0., 0., 0., 0., 0., 0.]),</span>
<span class="go"> &#39;G&#39;: array([0., 3., 2., 0., 0., 0., 0., 0.]),</span>
<span class="go"> &#39;T&#39;: array([0., 0., 0., 3., 3., 3., 1., 1.]),</span>
<span class="go"> &#39;A&#39;: array([2., 0., 0., 0., 0., 0., 0., 0.]),</span>
<span class="go"> &#39;-&#39;: array([0., 0., 1., 0., 0., 0., 2., 2.])}</span>
</pre></div>
</div>
</section>
<section id="substitutions">
<h3>Substitutions<a class="headerlink" href="#substitutions" title="Link to this heading"></a></h3>
<p>Use the <code class="docutils literal notranslate"><span class="pre">substitutions</span></code> method to find the number of substitutions
between each pair of nucleotides:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">substitutions</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="go">    A   C   G   T</span>
<span class="go">A 1.0 0.0 0.0 0.0</span>
<span class="go">C 2.0 0.0 0.0 0.0</span>
<span class="go">G 0.0 0.0 4.0 0.0</span>
<span class="go">T 0.0 0.0 0.0 9.0</span>
</pre></div>
</div>
<p>Note that the matrix is not symmetric: The counts for a row letter R and
a column letter C is the number of times letter R in a sequence is
replaced by letter C in a sequence appearing below it. For example, the
number of <code class="docutils literal notranslate"><span class="pre">C</span></code>’s that are aligned to an <code class="docutils literal notranslate"><span class="pre">A</span></code> in a later sequence is</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">]</span>
<span class="go">2.0</span>
</pre></div>
</div>
<p>while the number of A’s that are aligned to a C in a later sequence is</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>
<span class="go">0.0</span>
</pre></div>
</div>
<p>To get a symmetric matrix, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">+=</span> <span class="n">m</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">/=</span> <span class="mf">2.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="go">    A   C   G   T</span>
<span class="go">A 1.0 1.0 0.0 0.0</span>
<span class="go">C 1.0 0.0 0.0 0.0</span>
<span class="go">G 0.0 0.0 4.0 0.0</span>
<span class="go">T 0.0 0.0 0.0 9.0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>
<span class="go">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">]</span>
<span class="go">1.0</span>
</pre></div>
</div>
<p>The total number of substitutions between <code class="docutils literal notranslate"><span class="pre">A</span></code>’s and <code class="docutils literal notranslate"><span class="pre">T</span></code>’s in the
alignment is 1.0 + 1.0 = 2.</p>
</section>
<section id="alignments-as-arrays">
<h3>Alignments as arrays<a class="headerlink" href="#alignments-as-arrays" title="Link to this heading"></a></h3>
<p>Using NumPy, you can turn the <code class="docutils literal notranslate"><span class="pre">alignment</span></code> object into an array of
letters. In particular, this may be useful for fast calculations on the
alignment content.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align_array</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(3, 8)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align_array</span>  
<span class="go">array([[b&#39;C&#39;, b&#39;G&#39;, b&#39;G&#39;, b&#39;T&#39;, b&#39;T&#39;, b&#39;T&#39;, b&#39;T&#39;, b&#39;T&#39;],</span>
<span class="go">       [b&#39;A&#39;, b&#39;G&#39;, b&#39;-&#39;, b&#39;T&#39;, b&#39;T&#39;, b&#39;T&#39;, b&#39;-&#39;, b&#39;-&#39;],</span>
<span class="go">       [b&#39;A&#39;, b&#39;G&#39;, b&#39;G&#39;, b&#39;T&#39;, b&#39;T&#39;, b&#39;T&#39;, b&#39;-&#39;, b&#39;-&#39;]], dtype=&#39;|S1&#39;)</span>
</pre></div>
</div>
<p>By default, this will give you an array of <code class="docutils literal notranslate"><span class="pre">bytes</span></code> characters (with
data type <code class="docutils literal notranslate"><span class="pre">dtype='|S1'</span></code>). You can create an array of Unicode (Python
string) characters by using <code class="docutils literal notranslate"><span class="pre">dtype='U'</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">align_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;U&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">align_array</span>  
<span class="go">array([[&#39;C&#39;, &#39;G&#39;, &#39;G&#39;, &#39;T&#39;, &#39;T&#39;, &#39;T&#39;, &#39;T&#39;, &#39;T&#39;],</span>
<span class="go">       [&#39;A&#39;, &#39;G&#39;, &#39;-&#39;, &#39;T&#39;, &#39;T&#39;, &#39;T&#39;, &#39;-&#39;, &#39;-&#39;],</span>
<span class="go">       [&#39;A&#39;, &#39;G&#39;, &#39;G&#39;, &#39;T&#39;, &#39;T&#39;, &#39;T&#39;, &#39;-&#39;, &#39;-&#39;]], dtype=&#39;&lt;U1&#39;)</span>
</pre></div>
</div>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">alignment</span></code> object and the NumPy array <code class="docutils literal notranslate"><span class="pre">align_array</span></code>
are separate objects in memory - editing one will not update the other!</p>
</section>
</section>
<section id="operations-on-an-alignment">
<h2>Operations on an alignment<a class="headerlink" href="#operations-on-an-alignment" title="Link to this heading"></a></h2>
<section id="sorting-an-alignment">
<h3>Sorting an alignment<a class="headerlink" href="#sorting-an-alignment" title="Link to this heading"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">sort</span></code> method sorts the alignment sequences. By default, sorting
is done based on the <code class="docutils literal notranslate"><span class="pre">id</span></code> attribute of each sequence if available, or
the sequence contents otherwise.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">                  1 CGGTTTTT 9</span>
<span class="go">                  0 AG-TTT-- 5</span>
<span class="go">                  0 AGGTTT-- 6</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">                  0 AGGTTT-- 6</span>
<span class="go">                  0 AG-TTT-- 5</span>
<span class="go">                  1 CGGTTTTT 9</span>
</pre></div>
</div>
<p>Alternatively, you can supply a <code class="docutils literal notranslate"><span class="pre">key</span></code> function to determine the sort
order. For example, you can sort the sequences by increasing GC content:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">gc_fraction</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">gc_fraction</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>  <span class="c1"># CHEEMPIE</span>
<span class="go">                  0 AG-TTT-- 5</span>
<span class="go">                  0 AGGTTT-- 6</span>
<span class="go">                  1 CGGTTTTT 9</span>
</pre></div>
</div>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">key</span></code> function is applied to the full sequence
(including the initial <code class="docutils literal notranslate"><span class="pre">A</span></code> and final <code class="docutils literal notranslate"><span class="pre">GG</span></code> nucleotides of <code class="docutils literal notranslate"><span class="pre">seqB</span></code>),
not just to the aligned part.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">reverse</span></code> argument lets you reverse the sort order to obtain the
sequences in decreasing GC content:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">gc_fraction</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">                  1 CGGTTTTT 9</span>
<span class="go">                  0 AGGTTT-- 6</span>
<span class="go">                  0 AG-TTT-- 5</span>
</pre></div>
</div>
</section>
<section id="reverse-complementing-the-alignment">
<h3>Reverse-complementing the alignment<a class="headerlink" href="#reverse-complementing-the-alignment" title="Link to this heading"></a></h3>
<p>Reverse-complementing an alignment will take the reverse complement of
each sequence, and recalculate the coordinates:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;CCGGTTTTT&#39;, &#39;AGGTTT&#39;, &#39;AGTTTAA&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rc_alignment</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rc_alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">)</span>
<span class="go">[&#39;AAAAACCGG&#39;, &#39;AAACCT&#39;, &#39;TTAAACT&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rc_alignment</span><span class="p">)</span>
<span class="go">                  0 AAAAACCG 8</span>
<span class="go">                  0 --AAACCT 6</span>
<span class="go">                  2 --AAA-CT 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;CCGGTTTTT&#39;, &#39;AGGTTT&#39;, &#39;AGTTTAA&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">])</span>
<span class="go">                  1 CGGT 5</span>
<span class="go">                  0 AGGT 4</span>
<span class="go">                  0 AG-T 3</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">rc_alignment</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rc_alignment</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[&#39;AAAAACCGG&#39;, &#39;AAACCT&#39;, &#39;TTAAACT&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">rc_alignment</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">4</span><span class="p">])</span>
<span class="go">                  4 ACCG 8</span>
<span class="go">                  2 ACCT 6</span>
<span class="go">                  4 A-CT 7</span>
</pre></div>
</div>
<p>Reverse-complementing an alignment preserves its column annotations (in
reverse order), but discards all other annotations.</p>
</section>
<section id="adding-alignments">
<h3>Adding alignments<a class="headerlink" href="#adding-alignments" title="Link to this heading"></a></h3>
<p>Alignments can be added together to form an extended alignment if they
have the same number of rows. As an example, let’s first create two
alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAAC&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAC&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAAG&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GTT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GT&quot;</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">left</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span>
<span class="gp">... </span>    <span class="p">[</span><span class="n">a1</span><span class="p">,</span> <span class="n">b1</span><span class="p">,</span> <span class="n">c1</span><span class="p">],</span> <span class="n">coordinates</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">]])</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">left</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;demo&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;start&quot;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">left</span><span class="o">.</span><span class="n">column_annotations</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;stats&quot;</span><span class="p">:</span> <span class="s2">&quot;CCCXC&quot;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">right</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span>
<span class="gp">... </span>    <span class="p">[</span><span class="n">a2</span><span class="p">,</span> <span class="n">b2</span><span class="p">,</span> <span class="n">c2</span><span class="p">],</span> <span class="n">coordinates</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]])</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">right</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;tool&quot;</span><span class="p">:</span> <span class="s2">&quot;demo&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;end&quot;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">right</span><span class="o">.</span><span class="n">column_annotations</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;stats&quot;</span><span class="p">:</span> <span class="s2">&quot;CXC&quot;</span><span class="p">}</span>
</pre></div>
</div>
<p>Now, let’s look at these two alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">left</span><span class="p">)</span>
<span class="go">Alpha             0 AAAAC 5</span>
<span class="go">Beta              0 AAA-C 4</span>
<span class="go">Gamma             0 AAAAG 5</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">right</span><span class="p">)</span>
<span class="go">Alpha             0 GTT 3</span>
<span class="go">Beta              0 -TT 2</span>
<span class="go">Gamma             0 G-T 2</span>
</pre></div>
</div>
<p>Adding the two alignments will combine the two alignments row-wise:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">combined</span> <span class="o">=</span> <span class="n">left</span> <span class="o">+</span> <span class="n">right</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">combined</span><span class="p">)</span>
<span class="go">Alpha             0 AAAACGTT 8</span>
<span class="go">Beta              0 AAA-C-TT 6</span>
<span class="go">Gamma             0 AAAAGG-T 7</span>
</pre></div>
</div>
<p>For this to work, both alignments must have the same number of sequences
(here they both have 3 rows):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">left</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">right</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">combined</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
<p>The sequences are <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> objects, which can be added together.
Refer to Chapter <a class="reference internal" href="chapter_seq_annot.html#chapter-seq-annot"><span class="std std-ref">Sequence annotation objects</span></a> for
details of how the annotation is handled. This example is a special case
in that both original alignments shared the same names, meaning when the
rows are added they also get the same name.</p>
<p>Any common annotations are preserved, but differing annotation is lost.
This is the same behavior used in the <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> annotations and is
designed to prevent accidental propagation of inappropriate values:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">combined</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;tool&#39;: &#39;demo&#39;}</span>
</pre></div>
</div>
<p>Similarly any common per-column-annotations are combined:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">combined</span><span class="o">.</span><span class="n">column_annotations</span>
<span class="go">{&#39;stats&#39;: &#39;CCCXCCXC&#39;}</span>
</pre></div>
</div>
</section>
<section id="mapping-a-pairwise-sequence-alignment">
<h3>Mapping a pairwise sequence alignment<a class="headerlink" href="#mapping-a-pairwise-sequence-alignment" title="Link to this heading"></a></h3>
<p>Suppose you have a pairwise alignment of a transcript to a chromosome:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chromosome</span> <span class="o">=</span> <span class="s2">&quot;AAAAAAAACCCCCCCAAAAAAAAAAAGGGGGGAAAAAAAA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">transcript</span> <span class="o">=</span> <span class="s2">&quot;CCCCCCCGGGGGG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences1</span> <span class="o">=</span> <span class="p">[</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">transcript</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">26</span><span class="p">,</span> <span class="mi">32</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">13</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">sequences1</span><span class="p">,</span> <span class="n">coordinates1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment1</span><span class="p">)</span>
<span class="go">target            8 CCCCCCCAAAAAAAAAAAGGGGGG 32</span>
<span class="go">                  0 |||||||-----------|||||| 24</span>
<span class="go">query             0 CCCCCCC-----------GGGGGG 13</span>
</pre></div>
</div>
<p>and a pairwise alignment between the transcript and a sequence (e.g.,
obtained by RNA-seq):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rnaseq</span> <span class="o">=</span> <span class="s2">&quot;CCCCGGGG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences2</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">,</span> <span class="n">rnaseq</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">11</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">8</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">sequences2</span><span class="p">,</span> <span class="n">coordinates2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="go">target            3 CCCCGGGG 11</span>
<span class="go">                  0 ||||||||  8</span>
<span class="go">query             0 CCCCGGGG  8</span>
</pre></div>
</div>
<p>Use the <code class="docutils literal notranslate"><span class="pre">map</span></code> method on <code class="docutils literal notranslate"><span class="pre">alignment1</span></code>, with <code class="docutils literal notranslate"><span class="pre">alignment2</span></code> as
argument, to find the alignment of the RNA-sequence to the genome:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment3</span> <span class="o">=</span> <span class="n">alignment1</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment3</span><span class="p">)</span>
<span class="go">target           11 CCCCAAAAAAAAAAAGGGG 30</span>
<span class="go">                  0 ||||-----------|||| 19</span>
<span class="go">query             0 CCCC-----------GGGG  8</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment3</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[11, 15, 26, 30],</span>
<span class="go">       [ 0,  4,  4,  8]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment3</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="go">&#39;8\t0\t0\t0\t0\t0\t1\t11\t+\tquery\t8\t0\t8\ttarget\t40\t11\t30\t2\t4,4,\t0,4,\t11,26,\n&#39;</span>
</pre></div>
</div>
<p>To be able to print the sequences, in this example we constructed
<code class="docutils literal notranslate"><span class="pre">alignment1</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment2</span></code> using sequences with a defined
sequence contents. However, mapping the alignment does not depend on the
sequence contents; only the coordinates of <code class="docutils literal notranslate"><span class="pre">alignment1</span></code> and
<code class="docutils literal notranslate"><span class="pre">alignment2</span></code> are used to construct the coordinates for <code class="docutils literal notranslate"><span class="pre">alignment3</span></code>.</p>
<p>The map method can also be used to lift over an alignment between
different genome assemblies. In this case, self is a DNA alignment
between two genome assemblies, and the argument is an alignment of a
transcript against one of the genome assemblies:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chain</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;Blat/panTro5ToPanTro6.over.chain&quot;</span><span class="p">,</span> <span class="s2">&quot;chain&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;chr1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">228573443</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;chr1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">224244399</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">set_printoptions</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>  <span class="c1"># print 5 array elements per row</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">coordinates</span><span class="p">)</span>  
<span class="go">[[122250000 122250400 122250400 ... 122909818 122909819 122909835]</span>
<span class="go"> [111776384 111776784 111776785 ... 112019962 112019962 112019978]]</span>
</pre></div>
</div>
<p>showing that the range 122250000:122909835 of chr1 on chimpanzee genome
assembly panTro5 aligns to range 111776384:112019978 of chr1 of
chimpanzee genome assembly panTro6. See section
<a class="reference internal" href="#subsec-align-chain"><span class="std std-ref">UCSC chain file format</span></a> for more information about the chain
file format.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">transcript</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;Blat/est.panTro5.psl&quot;</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;chr1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">228573443</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;DC525629&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">407</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">coordinates</span><span class="p">)</span>
<span class="go">[[122835789 122835847 122840993 122841145 122907212 122907314]</span>
<span class="go"> [       32        90        90       242       242       344]]</span>
</pre></div>
</div>
<p>This shows that nucleotide range 32:344 of expressed sequence tag
DC525629 aligns to range 122835789:122907314 of chr1 of chimpanzee
genome assembly panTro5. Note that the target sequence
<code class="docutils literal notranslate"><span class="pre">chain.sequences[0].seq</span></code> and the target sequence
<code class="docutils literal notranslate"><span class="pre">transcript.sequences[0]</span></code> have the same length:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>We swap the target and query of the chain such that the query of
<code class="docutils literal notranslate"><span class="pre">chain</span></code> corresponds to the target of <code class="docutils literal notranslate"><span class="pre">transcript</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chain</span> <span class="o">=</span> <span class="n">chain</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;chr1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">224244399</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;chr1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">228573443</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">coordinates</span><span class="p">)</span>  
<span class="go">[[111776384 111776784 111776785 ... 112019962 112019962 112019978]</span>
<span class="go"> [122250000 122250400 122250400 ... 122909818 122909819 122909835]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">set_printoptions</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>  <span class="c1"># reset the print options</span>
</pre></div>
</div>
<p>Now we can get the coordinates of DC525629 against chimpanzee genome
assembly panTro6 by calling <code class="docutils literal notranslate"><span class="pre">chain.map</span></code>, with <code class="docutils literal notranslate"><span class="pre">transcript</span></code> as the
argument:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lifted_transcript</span> <span class="o">=</span> <span class="n">chain</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lifted_transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;chr1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">lifted_transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">224244399</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lifted_transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;DC525629&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">lifted_transcript</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="go">407</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lifted_transcript</span><span class="o">.</span><span class="n">coordinates</span><span class="p">)</span>
<span class="go">[[111982717 111982775 111987921 111988073 112009200 112009302]</span>
<span class="go"> [       32        90        90       242       242       344]]</span>
</pre></div>
</div>
<p>This shows that nucleotide range 32:344 of expressed sequence tag
DC525629 aligns to range 111982717:112009302 of chr1 of chimpanzee
genome assembly panTro6. Note that the genome span of DC525629 on
chimpanzee genome assembly panTro5 is 122907314 - 122835789 = 71525 bp,
while on panTro6 the genome span is 112009302 - 111982717 = 26585 bp.</p>
</section>
<section id="mapping-a-multiple-sequence-alignment">
<h3>Mapping a multiple sequence alignment<a class="headerlink" href="#mapping-a-multiple-sequence-alignment" title="Link to this heading"></a></h3>
<p>Consider a multiple alignment of genomic sequences of chimpanzee, human,
macaque, marmoset, mouse, and rat:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;Blat/panTro5.maf&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome_alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">genome_alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="gp">...</span>
<span class="go">panTro5.chr1 228573443</span>
<span class="go">hg19.chr1 249250621</span>
<span class="go">rheMac8.chr1 225584828</span>
<span class="go">calJac3.chr18 47448759</span>
<span class="go">mm10.chr3 160039680</span>
<span class="go">rn6.chr2 266435125</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome_alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[133922962, 133922962, 133922970, 133922970, 133922972, 133922972,</span>
<span class="go">        133922995, 133922998, 133923010],</span>
<span class="go">       [155784573, 155784573, 155784581, 155784581, 155784583, 155784583,</span>
<span class="go">        155784606, 155784609, 155784621],</span>
<span class="go">       [130383910, 130383910, 130383918, 130383918, 130383920, 130383920,</span>
<span class="go">        130383943, 130383946, 130383958],</span>
<span class="go">       [  9790455,   9790455,   9790463,   9790463,   9790465,   9790465,</span>
<span class="go">          9790488,   9790491,   9790503],</span>
<span class="go">       [ 88858039,  88858036,  88858028,  88858026,  88858024,  88858020,</span>
<span class="go">         88857997,  88857997,  88857985],</span>
<span class="go">       [188162970, 188162967, 188162959, 188162959, 188162957, 188162953,</span>
<span class="go">        188162930, 188162930, 188162918]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">genome_alignment</span><span class="p">)</span>
<span class="go">panTro5.c 133922962 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc</span>
<span class="go">hg19.chr1 155784573 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc</span>
<span class="go">rheMac8.c 130383910 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc</span>
<span class="go">calJac3.c   9790455 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAGCTTAAAggct</span>
<span class="go">mm10.chr3  88858039 TATAATAATTGTATATGTCACAGAAAAAAATGAATTTTCAAT---GACTTAATAGCC</span>
<span class="go">rn6.chr2  188162970 TACAATAATTG--TATGTCATAGAAAAAAATGAATTTTCAAT---AACTTAATAGCC</span>

<span class="go">panTro5.c 133923010</span>
<span class="go">hg19.chr1 155784621</span>
<span class="go">rheMac8.c 130383958</span>
<span class="go">calJac3.c   9790503</span>
<span class="go">mm10.chr3  88857985</span>
<span class="go">rn6.chr2  188162918</span>
</pre></div>
</div>
<p>Suppose we want to replace the older versions of the genome assemblies
(<code class="docutils literal notranslate"><span class="pre">panTro5</span></code>, <code class="docutils literal notranslate"><span class="pre">hg19</span></code>, <code class="docutils literal notranslate"><span class="pre">rheMac8</span></code>, <code class="docutils literal notranslate"><span class="pre">calJac3</span></code>, <code class="docutils literal notranslate"><span class="pre">mm10</span></code>, and <code class="docutils literal notranslate"><span class="pre">rn6</span></code>)
by their current versions (<code class="docutils literal notranslate"><span class="pre">panTro6</span></code>, <code class="docutils literal notranslate"><span class="pre">hg38</span></code>, <code class="docutils literal notranslate"><span class="pre">rheMac10</span></code>,
<code class="docutils literal notranslate"><span class="pre">calJac4</span></code>, <code class="docutils literal notranslate"><span class="pre">mm39</span></code>, and <code class="docutils literal notranslate"><span class="pre">rn7</span></code>). To do so, we need the pairwise
alignment between the old and the new assembly version for each species.
These are provided by UCSC as chain files, typically used for UCSC’s
<code class="docutils literal notranslate"><span class="pre">liftOver</span></code> tool. The <code class="docutils literal notranslate"><span class="pre">.chain</span></code> files in the <code class="docutils literal notranslate"><span class="pre">Tests/Align</span></code>
subdirectory in the Biopython source distribution were extracted from
UCSC’s <code class="docutils literal notranslate"><span class="pre">.chain</span></code> files to only include the relevant genomic region. For
example, to lift over <code class="docutils literal notranslate"><span class="pre">panTro5</span></code> to <code class="docutils literal notranslate"><span class="pre">panTro6</span></code>, we use the file
<code class="docutils literal notranslate"><span class="pre">panTro5ToPanTro6.chain</span></code> with the following contents:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>chain 1198066 chr1 228573443 + 133919957 133932620 chr1 224244399 + 130607995 130620657 1
4990    0   2
1362    3   0
6308
</pre></div>
</div>
<p>To lift over the genome assembly for each species, we read in the
corresponding <code class="docutils literal notranslate"><span class="pre">.chain</span></code> file:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">paths</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>    <span class="s2">&quot;Blat/panTro5ToPanTro6.chain&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;Blat/hg19ToHg38.chain&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;Blat/rheMac8ToRheMac10.chain&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;Blat/calJac3ToCalJac4.chain&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;Blat/mm10ToMm39.chain&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;Blat/rn6ToRn7.chain&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">liftover_alignments</span> <span class="o">=</span> <span class="p">[</span><span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;chain&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">liftover_alignment</span> <span class="ow">in</span> <span class="n">liftover_alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">liftover_alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">liftover_alignment</span><span class="o">.</span><span class="n">coordinates</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:])</span>
<span class="gp">...</span>
<span class="go">chr1 [133919957 133924947 133924947 133926309 133926312 133932620]</span>
<span class="go">chr1 [155184381 156354347 156354348 157128497 157128497 157137496]</span>
<span class="go">chr1 [130382477 130383872 130383872 130384222 130384222 130388520]</span>
<span class="go">chr18 [9786631 9787941 9788508 9788508 9795062 9795065 9795737]</span>
<span class="go">chr3 [66807541 74196805 74196831 94707528 94707528 94708176 94708178 94708718]</span>
<span class="go">chr2 [188111581 188158351 188158351 188171225 188171225 188228261 188228261</span>
<span class="go"> 188236997]</span>
</pre></div>
</div>
<p>Note that the order of species is the same in <code class="docutils literal notranslate"><span class="pre">liftover_alignments</span></code>
and <code class="docutils literal notranslate"><span class="pre">genome_alignment.sequences</span></code>. Now we can lift over the multiple
sequence alignment to the new genome assembly versions:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">genome_alignment</span> <span class="o">=</span> <span class="n">genome_alignment</span><span class="o">.</span><span class="n">mapall</span><span class="p">(</span><span class="n">liftover_alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">genome_alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
<span class="gp">...</span>
<span class="go">chr1 224244399</span>
<span class="go">chr1 248956422</span>
<span class="go">chr1 223616942</span>
<span class="go">chr18 47031477</span>
<span class="go">chr3 159745316</span>
<span class="go">chr2 249053267</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genome_alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[130611000, 130611000, 130611008, 130611008, 130611010, 130611010,</span>
<span class="go">        130611033, 130611036, 130611048],</span>
<span class="go">       [155814782, 155814782, 155814790, 155814790, 155814792, 155814792,</span>
<span class="go">        155814815, 155814818, 155814830],</span>
<span class="go">       [ 95186253,  95186253,  95186245,  95186245,  95186243,  95186243,</span>
<span class="go">         95186220,  95186217,  95186205],</span>
<span class="go">       [  9758318,   9758318,   9758326,   9758326,   9758328,   9758328,</span>
<span class="go">          9758351,   9758354,   9758366],</span>
<span class="go">       [ 88765346,  88765343,  88765335,  88765333,  88765331,  88765327,</span>
<span class="go">         88765304,  88765304,  88765292],</span>
<span class="go">       [174256702, 174256699, 174256691, 174256691, 174256689, 174256685,</span>
<span class="go">        174256662, 174256662, 174256650]])</span>
</pre></div>
</div>
<p>As the <code class="docutils literal notranslate"><span class="pre">.chain</span></code> files do not include the sequence contents, we cannot
print the sequence alignment directly. Instead, we read in the genomic
sequence separately (as a <code class="docutils literal notranslate"><span class="pre">.2bit</span></code> file, as it allows lazy loading; see
section <a class="reference internal" href="chapter_seqio.html#sec-seqio-directionaries"><span class="std std-ref">Sequence files as Dictionaries</span></a>) for
each species:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">names</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;panTro6&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">,</span> <span class="s2">&quot;rheMac10&quot;</span><span class="p">,</span> <span class="s2">&quot;calJac4&quot;</span><span class="p">,</span> <span class="s2">&quot;mm39&quot;</span><span class="p">,</span> <span class="s2">&quot;rn7&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">names</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">filename</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">.2bit&quot;</span>
<span class="gp">... </span>    <span class="n">genome</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;twobit&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">chromosome</span> <span class="o">=</span> <span class="n">genome_alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="gp">... </span>    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome_alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="n">chromosome</span><span class="p">])</span>
<span class="gp">... </span>    <span class="n">genome_alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span>
<span class="gp">... </span>    <span class="n">genome_alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">chromosome</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">genome_alignment</span><span class="p">)</span>
<span class="go">panTro6.c 130611000 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc</span>
<span class="go">hg38.chr1 155814782 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc</span>
<span class="go">rheMac10.  95186253 ---ACTAGTTA--CA----GTAACAGAAAATAAAATTTAAATAGAAACTTAAAggcc</span>
<span class="go">calJac4.c   9758318 ---ACTAGTTA--CA----GTAACAGAaaataaaatttaaatagaagcttaaaggct</span>
<span class="go">mm39.chr3  88765346 TATAATAATTGTATATGTCACAGAAAAAAATGAATTTTCAAT---GACTTAATAGCC</span>
<span class="go">rn7.chr2  174256702 TACAATAATTG--TATGTCATAGAAAAAAATGAATTTTCAAT---AACTTAATAGCC</span>

<span class="go">panTro6.c 130611048</span>
<span class="go">hg38.chr1 155814830</span>
<span class="go">rheMac10.  95186205</span>
<span class="go">calJac4.c   9758366</span>
<span class="go">mm39.chr3  88765292</span>
<span class="go">rn7.chr2  174256650</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">mapall</span></code> method can also be used to create a multiple alignment of
codon sequences from a multiple sequence alignment of the corresponding
amino acid sequences (see Section <a class="reference internal" href="chapter_pairwise.html#sec-msa-codons"><span class="std std-ref">Generating a multiple sequence alignment of codon sequences</span></a>
for details).</p>
</section>
</section>
<section id="the-alignments-class">
<span id="sec-alignments"></span><h2>The Alignments class<a class="headerlink" href="#the-alignments-class" title="Link to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> (plural) class inherits from
<code class="docutils literal notranslate"><span class="pre">AlignmentsAbstractBaseClass</span></code> and from <code class="docutils literal notranslate"><span class="pre">list</span></code>, and can be used as a
list to store <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects. The behavior of <code class="docutils literal notranslate"><span class="pre">Alignments</span></code>
objects is different from that of <code class="docutils literal notranslate"><span class="pre">list</span></code> objects in two important
ways:</p>
<ul>
<li><p>An <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> object is its own iterator, consistent with iterators
returned by <code class="docutils literal notranslate"><span class="pre">Bio.Align.parse</span></code> (see section <a class="reference internal" href="#subsec-align-reading"><span class="std std-ref">Reading alignments</span></a>) or
iterators returned by the pairwise aligner (see Section
<a class="reference internal" href="chapter_pairwise.html#chapter-pairwise"><span class="std std-ref">Pairwise sequence alignment</span></a>). Calling <code class="docutils literal notranslate"><span class="pre">iter</span></code> on the iterator will
always return the <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> object itself. In contrast, calling
<code class="docutils literal notranslate"><span class="pre">iter</span></code> on a list object creates a new iterator each time, allowing you to
have multiple independent iterators for a given list.</p>
<p>In this example, <code class="docutils literal notranslate"><span class="pre">alignment_iterator1</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment_iterator2</span></code> are
obtained from a list and act independently of each other:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">alignment1</span><span class="p">,</span> <span class="n">alignment2</span><span class="p">,</span> <span class="n">alignment3</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_iterator1</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_iterator2</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator1</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 24 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator2</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 24 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator1</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 8 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator1</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 19 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator2</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 8 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator2</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 19 columns) at ...&gt;</span>
</pre></div>
</div>
<p>In contrast, <code class="docutils literal notranslate"><span class="pre">alignment_iterator1</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment_iterator2</span></code> obtained
by calling <code class="docutils literal notranslate"><span class="pre">iter</span></code> on an <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> object are identical to each other:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">Alignments</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Alignments</span><span class="p">([</span><span class="n">alignment1</span><span class="p">,</span> <span class="n">alignment2</span><span class="p">,</span> <span class="n">alignment3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_iterator1</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_iterator2</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_iterator1</span> <span class="ow">is</span> <span class="n">alignment_iterator2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator1</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 24 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator2</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 8 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator1</span><span class="p">)</span>  
<span class="go">&lt;Alignment object (2 rows x 19 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">alignment_iterator2</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
<span class="gr">StopIteration</span>
</pre></div>
</div>
<p>Calling <code class="docutils literal notranslate"><span class="pre">iter</span></code> on an <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> object resets the iterator to its
first item, so you can loop over it again. You can also iterate over the
alignments multiple times using a <code class="docutils literal notranslate"><span class="pre">for</span></code>-loop, which implicitly calls
<code class="docutils literal notranslate"><span class="pre">iter</span></code> on the iterator:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">item</span><span class="p">))</span>  
<span class="gp">...</span>
<span class="go">&lt;Alignment object (2 rows x 24 columns) at ...&gt;</span>
<span class="go">&lt;Alignment object (2 rows x 8 columns) at ...&gt;</span>
<span class="go">&lt;Alignment object (2 rows x 19 columns) at ...&gt;</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">item</span><span class="p">))</span>  
<span class="gp">...</span>
<span class="go">&lt;Alignment object (2 rows x 24 columns) at ...&gt;</span>
<span class="go">&lt;Alignment object (2 rows x 8 columns) at ...&gt;</span>
<span class="go">&lt;Alignment object (2 rows x 19 columns) at ...&gt;</span>
</pre></div>
</div>
<p>This behavior is consistent with regular Python lists, and with iterators
returned by <code class="docutils literal notranslate"><span class="pre">Bio.Align.parse</span></code> (see section <a class="reference internal" href="#subsec-align-reading"><span class="std std-ref">Reading alignments</span></a>) or
by the pairwise aligner (see Section <a class="reference internal" href="chapter_pairwise.html#chapter-pairwise"><span class="std std-ref">Pairwise sequence alignment</span></a>).</p>
</li>
<li><p>Metadata can be stored as attributes on an <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> object,
whereas a plain <code class="docutils literal notranslate"><span class="pre">list</span></code> does not accept attributes:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_list</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="mi">100</span>  
<span class="gt">Traceback (most recent call last):</span>
<span class="w"> </span><span class="c">...</span>
<span class="gr">AttributeError</span>: <span class="n">&#39;list&#39; object has no attribute &#39;score&#39;...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="mi">100</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">score</span>
<span class="go">100</span>
</pre></div>
</div>
</li>
</ul>
</section>
<section id="reading-and-writing-alignments">
<span id="sec-alignmentparsers"></span><h2>Reading and writing alignments<a class="headerlink" href="#reading-and-writing-alignments" title="Link to this heading"></a></h2>
<p>Output from sequence alignment software such as Clustal can be parsed
into <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects by the <code class="docutils literal notranslate"><span class="pre">Bio.Align.read</span></code> and
<code class="docutils literal notranslate"><span class="pre">Bio.Align.parse</span></code> functions. Their usage is analogous to the <code class="docutils literal notranslate"><span class="pre">read</span></code>
and <code class="docutils literal notranslate"><span class="pre">parse</span></code> functions in <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> (see
Section <a class="reference internal" href="chapter_seqio.html#sec-bio-seqio-input"><span class="std std-ref">Parsing or Reading Sequences</span></a>): The <code class="docutils literal notranslate"><span class="pre">read</span></code>
function is used to read an output file containing a single alignment
and returns an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object, while the <code class="docutils literal notranslate"><span class="pre">parse</span></code> function
returns an iterator to iterate over alignments stored in an output file
containing one or more alignments. Section <a class="reference internal" href="#sec-alignformats"><span class="std std-ref">Alignment file formats</span></a>
describes the alignment formats that can be parsed in <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code>.
<code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> also provides a <code class="docutils literal notranslate"><span class="pre">write</span></code> function that can write
alignments in most of these formats.</p>
<section id="reading-alignments">
<span id="subsec-align-reading"></span><h3>Reading alignments<a class="headerlink" href="#reading-alignments" title="Link to this heading"></a></h3>
<p>Use <code class="docutils literal notranslate"><span class="pre">Bio.Align.parse</span></code> to parse a file of sequence alignments. For
example, the file <code class="docutils literal notranslate"><span class="pre">ucsc_mm9_chr10.maf</span></code> contains 48 multiple sequence
alignments in the MAF (Multiple Alignment Format) format (see section
<a class="reference internal" href="#subsec-align-maf"><span class="std std-ref">Multiple Alignment Format (MAF)</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;MAF/ucsc_mm9_chr10.maf&quot;</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span>  
<span class="go">&lt;Bio.Align.maf.AlignmentIterator object at 0x...&gt;</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">&quot;maf&quot;</span></code> is the file format. The alignments object returned by
<code class="docutils literal notranslate"><span class="pre">Bio.Align.parse</span></code> may contain attributes that store metadata found in
the file, such as the version number of the software that was used to
create the alignments. The specific attributes stored for each file
format are described in Section <a class="reference internal" href="#sec-alignformats"><span class="std std-ref">Alignment file formats</span></a>. For MAF
files, we can obtain the file format version and the scoring scheme that
was used:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;MAF Version&#39;: &#39;1&#39;, &#39;Scoring&#39;: &#39;autoMZ.v1&#39;}</span>
</pre></div>
</div>
<p>As alignment files can be very large, <code class="docutils literal notranslate"><span class="pre">Align.parse</span></code> returns an
iterator over the alignments, so you won’t have to store all alignments
in memory at the same time. You can iterate over these alignments and
print out, for example, the number of aligned sequences in each
alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">sequences</span><span class="p">))</span>  
<span class="gp">...</span>
<span class="go">2</span>
<span class="go">4</span>
<span class="go">5</span>
<span class="go">6</span>
<span class="go">...</span>
<span class="go">15</span>
<span class="go">14</span>
<span class="go">7</span>
<span class="go">6</span>
</pre></div>
</div>
<p>You can also call <code class="docutils literal notranslate"><span class="pre">len</span></code> on the alignments to obtain the number of
alignments.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">48</span>
</pre></div>
</div>
<p>Depending on the file format, the number of alignments may be explicitly
stored in the file (for example in the case of bigBed, bigPsl, and
bigMaf files), or otherwise the number of alignments is counted by
looping over them once (and returning the iterator to its original
position). If the file is large, it may therefore take a considerable
amount of time for <code class="docutils literal notranslate"><span class="pre">len</span></code> to return. However, as the number of
alignments is cached, subsequent calls to <code class="docutils literal notranslate"><span class="pre">len</span></code> will return quickly.</p>
<p>If the number of alignments is not excessively large and will fit in
memory, you can convert the alignments iterator to a list of alignments.
To do so, you could call <code class="docutils literal notranslate"><span class="pre">list</span></code> on the <code class="docutils literal notranslate"><span class="pre">alignments</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">)</span>
<span class="go">48</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_list</span><span class="p">[</span><span class="mi">27</span><span class="p">]</span>  
<span class="go">&lt;Alignment object (3 rows x 91 columns) at 0x...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment_list</span><span class="p">[</span><span class="mi">27</span><span class="p">])</span>
<span class="go">mm9.chr10   3019377 CCCCAGCATTCTGGCAGACACAGTG-AAAAGAGACAGATGGTCACTAATAAAATCTGT-A</span>
<span class="go">felCat3.s     46845 CCCAAGTGTTCTGATAGCTAATGTGAAAAAGAAGCATGTGCCCACCAGTAAGCTTTGTGG</span>
<span class="go">canFam2.c  47545247 CCCAAGTGTTCTGATTGCCTCTGTGAAAAAGAAACATGGGCCCGCTAATAagatttgcaa</span>

<span class="go">mm9.chr10   3019435 TAAATTAG-ATCTCAGAGGATGGATGGACCA  3019465</span>
<span class="go">felCat3.s     46785 TGAACTAGAATCTCAGAGGATG---GGACTC    46757</span>
<span class="go">canFam2.c  47545187 tgacctagaatctcagaggatg---ggactc 47545159</span>
</pre></div>
</div>
<p>But this will lose the metadata information:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment_list</span><span class="o">.</span><span class="n">metadata</span>  
<span class="gt">Traceback (most recent call last):</span>
<span class="w">  </span><span class="c">...</span>
<span class="gr">AttributeError</span>: <span class="n">&#39;list&#39; object has no attribute &#39;metadata&#39;</span>
</pre></div>
</div>
<p>Instead, you can ask for a full slice of the alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">&lt;class &#39;Bio.Align.maf.AlignmentIterator&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">&lt;class &#39;Bio.Align.Alignments&#39;&gt;</span>
</pre></div>
</div>
<p>This returns a <code class="docutils literal notranslate"><span class="pre">Bio.Align.Alignments</span></code> object, which can be used as a
list, while keeping the metadata information:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">48</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">11</span><span class="p">])</span>
<span class="go">mm9.chr10   3014742 AAGTTCCCTCCATAATTCCTTCCTCCCACCCCCACA 3014778</span>
<span class="go">calJac1.C      6283 AAATGTA-----TGATCTCCCCATCCTGCCCTG---    6311</span>
<span class="go">otoGar1.s    175262 AGATTTC-----TGATGCCCTCACCCCCTCCGTGCA  175231</span>
<span class="go">loxAfr1.s      9317 AGGCTTA-----TG----CCACCCCCCACCCCCACA    9290</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;MAF Version&#39;: &#39;1&#39;, &#39;Scoring&#39;: &#39;autoMZ.v1&#39;}</span>
</pre></div>
</div>
</section>
<section id="writing-alignments">
<span id="subsec-align-writing"></span><h3>Writing alignments<a class="headerlink" href="#writing-alignments" title="Link to this heading"></a></h3>
<p>To write alignments to a file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;myfile.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">alignments</span></code> is either a single alignment or a list of
alignments, <code class="docutils literal notranslate"><span class="pre">target</span></code> is a file name or an open file-like object, and
<code class="docutils literal notranslate"><span class="pre">&quot;clustal&quot;</span></code> is the file format to be used. As some file formats allow
or require metadata to be stored with the alignments, you may want to
use the <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> (plural) class instead of a plain list of
alignments (see Section <a class="reference internal" href="#sec-alignments"><span class="std std-ref">The Alignments class</span></a>), allowing you to
store a metadata dictionary as an attribute on the <code class="docutils literal notranslate"><span class="pre">alignments</span></code>
object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">Alignments</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Program&quot;</span><span class="p">:</span> <span class="s2">&quot;Biopython&quot;</span><span class="p">,</span> <span class="s2">&quot;Version&quot;</span><span class="p">:</span> <span class="s2">&quot;1.81&quot;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;myfile.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="printing-alignments">
<span id="subsec-align-printing"></span><h3>Printing alignments<a class="headerlink" href="#printing-alignments" title="Link to this heading"></a></h3>
<p>For text (non-binary) formats, you can call Python’s built-in <code class="docutils literal notranslate"><span class="pre">format</span></code>
function on an alignment to get a string showing the alignment in the
requested format, or use <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects in formatted (f-)
strings. If called without an argument, the <code class="docutils literal notranslate"><span class="pre">format</span></code> function returns
the string representation of the alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">&#39;                  1 CGGTTTTT 9\n                  0 AGGTTT-- 6\n                  0 AG-TTT-- 5\n&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">&#39;                  1 CGGTTTTT 9\n                  0 AGGTTT-- 6\n                  0 AG-TTT-- 5\n&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">))</span>
<span class="go">                  1 CGGTTTTT 9</span>
<span class="go">                  0 AGGTTT-- 6</span>
<span class="go">                  0 AG-TTT-- 5</span>
</pre></div>
</div>
<p>By specifying one of the formats shown in
Section <a class="reference internal" href="#sec-alignformats"><span class="std std-ref">Alignment file formats</span></a>, <code class="docutils literal notranslate"><span class="pre">format</span></code> will create a string
showing the alignment in the requested format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
<span class="go">&#39;sequence_0                          CGGTTTTT\nsequence_1                          AGGTTT--\nsequence_2                          AG-TTT--\n\n\n&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">))</span>
<span class="go">sequence_0                          CGGTTTTT</span>
<span class="go">sequence_1                          AGGTTT--</span>
<span class="go">sequence_2                          AG-TTT--</span>



<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;*** this is the alignment in Clustal format: ***</span><span class="se">\n</span><span class="si">{</span><span class="n">alignment</span><span class="si">:</span><span class="s2">clustal</span><span class="si">}</span><span class="se">\n</span><span class="s2">***&quot;</span><span class="p">)</span>
<span class="go">*** this is the alignment in Clustal format: ***</span>
<span class="go">sequence_0                          CGGTTTTT</span>
<span class="go">sequence_1                          AGGTTT--</span>
<span class="go">sequence_2                          AG-TTT--</span>



<span class="go">***</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">)</span>
<span class="go">&#39;a\ns sequence_0 1 8 + 9 CGGTTTTT\ns sequence_1 0 6 + 6 AGGTTT--\ns sequence_2 0 5 + 7 AG-TTT--\n\n&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">))</span>
<span class="go">a</span>
<span class="go">s sequence_0 1 8 + 9 CGGTTTTT</span>
<span class="go">s sequence_1 0 6 + 6 AGGTTT--</span>
<span class="go">s sequence_2 0 5 + 7 AG-TTT--</span>

</pre></div>
</div>
<p>As optional keyword arguments cannot be used with Python’s built-in
<code class="docutils literal notranslate"><span class="pre">format</span></code> function or with formatted strings, the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> class
has a <code class="docutils literal notranslate"><span class="pre">format</span></code> method with optional arguments to customize the
alignment format, as described in the subsections below. For example, we
can print the alignment in BED format (see
section <a class="reference internal" href="#subsec-align-bed"><span class="std std-ref">Browser Extensible Data (BED)</span></a>) with a specific number of
columns:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_alignment</span><span class="p">)</span>
<span class="go">target            1 CGGTTTTT 9</span>
<span class="go">                  0 .|-|||-- 8</span>
<span class="go">query             0 AG-TTT-- 5</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">pairwise_alignment</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">target  1   7   query   0   +   1   7   0   2   2,3,    0,3,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">target  1   7   query   0   +   1   7   0   2   2,3,    0,3,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">,</span> <span class="n">bedN</span><span class="o">=</span><span class="mi">3</span><span class="p">))</span>  
<span class="go">target  1   7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">,</span> <span class="n">bedN</span><span class="o">=</span><span class="mi">6</span><span class="p">))</span>  
<span class="go">target  1   7   query   0   +</span>
</pre></div>
</div>
</section>
</section>
<section id="alignment-file-formats">
<span id="sec-alignformats"></span><h2>Alignment file formats<a class="headerlink" href="#alignment-file-formats" title="Link to this heading"></a></h2>
<p>The table below shows the alignment formats that can be parsed in
Bio.Align. The format argument <code class="docutils literal notranslate"><span class="pre">fmt</span></code> used in <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> functions
to specify the file format is case-insensitive. Most of these file
formats can also be written by <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code>, as shown in the table.</p>
<div class="center docutils container">
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>File format
<code class="docutils literal notranslate"><span class="pre">fmt</span></code></p></td>
<td><p>Description</p></td>
<td><p>text /
binary</p></td>
<td><p>Supported
by
<code class="docutils literal notranslate"><span class="pre">write</span></code></p></td>
<td><p>Subsection</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">a2m</span></code></p></td>
<td><p>A2M</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_a2m">1.7.11</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">bed</span></code></p></td>
<td><p>Browser
Extensible
Data (BED)</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_bed">1.7.14</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">bigbed</span></code></p></td>
<td><p>bigBed</p></td>
<td><p>binary</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_bigbed">1.7.15</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">bigmaf</span></code></p></td>
<td><p>bigMaf</p></td>
<td><p>binary</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_bigmaf">1.7.19</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">bigpsl</span></code></p></td>
<td><p>bigPsl</p></td>
<td><p>binary</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_bigpsl">1.7.17</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">chain</span></code></p></td>
<td><p>UCSC chain
file</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_chain">1.7.20</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">clustal</span></code></p></td>
<td><p>ClustalW</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_clustal">1.7.2</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">emboss</span></code></p></td>
<td><p>EMBOSS</p></td>
<td><p>text</p></td>
<td><p>no</p></td>
<td><p><a class="reference external" href="#subsec:align_emboss">1.7.5</a></p></td>
</tr>
<tr class="row-even"><td><p>``
exonerate``</p></td>
<td><p>Exonerate</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_exonerate">1
.7.7</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">fasta</span></code></p></td>
<td><p>Aligned
FASTA</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_fasta">1.7.1</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">hhr</span></code></p></td>
<td><p>HH-suite
output
files</p></td>
<td><p>text</p></td>
<td><p>no</p></td>
<td><p><a class="reference external" href="#subsec:align_hhr">1.7.10</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">maf</span></code></p></td>
<td><p>Multiple
Alignment
Format
(MAF)</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_maf">1.7.18</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">mauve</span></code></p></td>
<td><p>Mauve
eXtended
Multi-FastA
(xmfa)
format</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_mauve">1.7.12</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">msf</span></code></p></td>
<td><p>GCG
Multiple
Sequence
Format
(MSF)</p></td>
<td><p>text</p></td>
<td><p>no</p></td>
<td><p><a class="reference external" href="#subsec:align_msf">1.7.6</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nexus</span></code></p></td>
<td><p>NEXUS</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_nexus">1.7.8</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">phylip</span></code></p></td>
<td><p>PHYLIP
output
files</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_phylip">1.7.4</a></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">psl</span></code></p></td>
<td><p>Pattern
Space
Layout
(PSL)</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_psl">1.7.16</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">sam</span></code></p></td>
<td><p>Sequence
Al
ignment/Map
(SAM)</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_sam">1.7.13</a></p></td>
</tr>
<tr class="row-even"><td><p>``
stockholm``</p></td>
<td><p>Stockholm</p></td>
<td><p>text</p></td>
<td><p>yes</p></td>
<td><p><a class="reference external" href="#subsec:align_stockholm">1
.7.3</a></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">tabular</span></code></p></td>
<td><p>Tabular
output from
BLAST or
FASTA</p></td>
<td><p>text</p></td>
<td><p>no</p></td>
<td><p><a class="reference external" href="#subsec:align_tabular">1.7.9</a></p></td>
</tr>
</tbody>
</table>
</div>
<section id="aligned-fasta">
<span id="subsec-align-fasta"></span><h3>Aligned FASTA<a class="headerlink" href="#aligned-fasta" title="Link to this heading"></a></h3>
<p>Files in the aligned FASTA format store exactly one (pairwise or
multiple) sequence alignment, in which gaps in the alignment are
represented by dashes (<code class="docutils literal notranslate"><span class="pre">-</span></code>). Use <code class="docutils literal notranslate"><span class="pre">fmt=&quot;fasta&quot;</span></code> to read or write
files in the aligned FASTA format. Note that this is different from
output generated by William Pearson’s FASTA alignment program (parsing
such output is described in section <a class="reference internal" href="#subsec-align-tabular"><span class="std std-ref">Tabular output from BLAST or FASTA</span></a>
instead).</p>
<p>The file <code class="docutils literal notranslate"><span class="pre">probcons.fa</span></code> in Biopython’s test suite stores one multiple
alignment in the aligned FASTA format. The contents of this file is as
follows:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;plas_horvu
D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-VD-VSKISQEEYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVTV
&gt;plas_chlre
--VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-VN-ADAISRDDYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKIIV
&gt;plas_anava
--VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKSLSHKQLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKITV
&gt;plas_proho
VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-ES-APALSNTKLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTITV
&gt;azup_achcy
VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-AE-A-------FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVEV
</pre></div>
</div>
<p>To read this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;probcons.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (5 rows x 101 columns) at ...&gt;</span>
</pre></div>
</div>
<p>We can print the alignment to see its default representation:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">plas_horv         0 D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-VD-VSKISQE</span>
<span class="go">plas_chlr         0 --VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-VN-ADAISRD</span>
<span class="go">plas_anav         0 --VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKSLSHK</span>
<span class="go">plas_proh         0 VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-ES-APALSNT</span>
<span class="go">azup_achc         0 VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-AE-A------</span>

<span class="go">plas_horv        57 EYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVTV 95</span>
<span class="go">plas_chlr        56 DYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKIIV 94</span>
<span class="go">plas_anav        58 QLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKITV 99</span>
<span class="go">plas_proh        56 KLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTITV 94</span>
<span class="go">azup_achc        51 -FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVEV 88</span>
</pre></div>
</div>
<p>or we can print it in the aligned FASTA format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>
<span class="go">&gt;plas_horvu</span>
<span class="go">D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-VD-VSKISQEEYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVTV</span>
<span class="go">&gt;plas_chlre</span>
<span class="go">--VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-VN-ADAISRDDYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKIIV</span>
<span class="go">&gt;plas_anava</span>
<span class="go">--VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKSLSHKQLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKITV</span>
<span class="go">&gt;plas_proho</span>
<span class="go">VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-ES-APALSNTKLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTITV</span>
<span class="go">&gt;azup_achcy</span>
<span class="go">VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-AE-A-------FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVEV</span>
</pre></div>
</div>
<p>or any other available format, for example Clustal (see
section <a class="reference internal" href="#subsec-align-clustal"><span class="std std-ref">ClustalW</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">))</span>
<span class="go">plas_horvu                          D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-</span>
<span class="go">plas_chlre                          --VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-</span>
<span class="go">plas_anava                          --VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKS</span>
<span class="go">plas_proho                          VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-</span>
<span class="go">azup_achcy                          VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-</span>

<span class="go">plas_horvu                          VD-VSKISQEEYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVT</span>
<span class="go">plas_chlre                          VN-ADAISRDDYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKII</span>
<span class="go">plas_anava                          ADLAKSLSHKQLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKIT</span>
<span class="go">plas_proho                          ES-APALSNTKLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTIT</span>
<span class="go">azup_achcy                          AE-A-------FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVE</span>

<span class="go">plas_horvu                          V</span>
<span class="go">plas_chlre                          V</span>
<span class="go">plas_anava                          V</span>
<span class="go">plas_proho                          V</span>
<span class="go">azup_achcy                          V</span>


</pre></div>
</div>
<p>The sequences associated with the alignment are <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> objects:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[SeqRecord(seq=Seq(&#39;DVLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSGVDVSKI...VTV&#39;), id=&#39;plas_horvu&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[]), SeqRecord(seq=Seq(&#39;VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSGVNADAIS...IIV&#39;), id=&#39;plas_chlre&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[]), SeqRecord(seq=Seq(&#39;VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKS...ITV&#39;), id=&#39;plas_anava&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[]), SeqRecord(seq=Seq(&#39;VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDKVPAGESAPALS...ITV&#39;), id=&#39;plas_proho&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[]), SeqRecord(seq=Seq(&#39;VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDKGHNVETIKGMIPDGAEAFKS...VEV&#39;), id=&#39;azup_achcy&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[])]</span>
</pre></div>
</div>
<p>Note that these sequences do not contain gaps (”<code class="docutils literal notranslate"><span class="pre">-</span></code>” characters), as
the alignment information is stored in the <code class="docutils literal notranslate"><span class="pre">coordinates</span></code> attribute
instead:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[ 0,  1,  1, 33, 34, 42, 44, 48, 48, 50, 50, 51, 58, 73, 73, 95],</span>
<span class="go">       [ 0,  0,  0, 32, 33, 41, 43, 47, 47, 49, 49, 50, 57, 72, 72, 94],</span>
<span class="go">       [ 0,  0,  0, 32, 33, 41, 43, 47, 48, 50, 51, 52, 59, 74, 77, 99],</span>
<span class="go">       [ 0,  1,  2, 34, 35, 43, 43, 47, 47, 49, 49, 50, 57, 72, 72, 94],</span>
<span class="go">       [ 0,  1,  2, 34, 34, 42, 44, 48, 48, 50, 50, 51, 51, 66, 66, 88]])</span>
</pre></div>
</div>
<p>Use <code class="docutils literal notranslate"><span class="pre">Align.write</span></code> to write this alignment to a file (here, we’ll use a
<code class="docutils literal notranslate"><span class="pre">StringIO</span></code> object instead of a file):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;FASTA&quot;</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">stream</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>
<span class="go">&gt;plas_horvu</span>
<span class="go">D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-VD-VSKISQEEYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVTV</span>
<span class="go">&gt;plas_chlre</span>
<span class="go">--VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-VN-ADAISRDDYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKIIV</span>
<span class="go">&gt;plas_anava</span>
<span class="go">--VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKSLSHKQLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKITV</span>
<span class="go">&gt;plas_proho</span>
<span class="go">VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-ES-APALSNTKLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTITV</span>
<span class="go">&gt;azup_achcy</span>
<span class="go">VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-AE-A-------FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVEV</span>
</pre></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">Align.write</span></code> returns the number of alignments written (1,
in this case).</p>
</section>
<section id="clustalw">
<span id="subsec-align-clustal"></span><h3>ClustalW<a class="headerlink" href="#clustalw" title="Link to this heading"></a></h3>
<p>Clustal is a set of multiple sequence alignment programs that are
available both as standalone programs as as web servers. The file
<code class="docutils literal notranslate"><span class="pre">opuntia.aln</span></code> (available online or in the <code class="docutils literal notranslate"><span class="pre">Doc/examples</span></code>
subdirectory of the Biopython source code) is an output file generated
by Clustal. Its first few lines are</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>CLUSTAL 2.1 multiple sequence alignment


gi|6273285|gb|AF191659.1|AF191      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA
gi|6273284|gb|AF191658.1|AF191      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA
gi|6273287|gb|AF191661.1|AF191      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA
gi|6273286|gb|AF191660.1|AF191      TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA
gi|6273290|gb|AF191664.1|AF191      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA
gi|6273289|gb|AF191663.1|AF191      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA
gi|6273291|gb|AF191665.1|AF191      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA
                                    ******* **** *************************************

...
</pre></div>
</div>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;opuntia.aln&quot;</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> attribute on <code class="docutils literal notranslate"><span class="pre">alignments</span></code> stores the information
shown in the file header:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;Program&#39;: &#39;CLUSTAL&#39;, &#39;Version&#39;: &#39;2.1&#39;}</span>
</pre></div>
</div>
<p>You can call <code class="docutils literal notranslate"><span class="pre">next</span></code> on the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> to pull out the first (and
only) alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>  
<span class="go">gi|627328         0 TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>
<span class="go">gi|627328         0 TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>
<span class="go">gi|627328         0 TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>
<span class="go">gi|627328         0 TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>
<span class="go">gi|627329         0 TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>
<span class="go">gi|627328         0 TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>
<span class="go">gi|627329         0 TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>

<span class="go">gi|627328        60 CTAAATGATATACGATTCCACTATGTAAGGTCTTTGAATCATATCATAAAAGACAATGTA</span>
<span class="go">gi|627328        60 CTAAATGATATACGATTCCACTATGTAAGGTCTTTGAATCATATCATAAAAGACAATGTA</span>
<span class="go">gi|627328        60 CTAAATGATATACGATTCCACTATGTAAGGTCTTTGAATCATATCATAAAAGACAATGTA</span>
<span class="go">gi|627328        60 CTAAATGATATACGATTCCACTA...</span>
</pre></div>
</div>
<p>If you are not interested in the metadata, then it is more convenient to
use the <code class="docutils literal notranslate"><span class="pre">Align.read</span></code> function, as anyway each Clustal file contains
only one alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;opuntia.aln&quot;</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The consensus line below each alignment block in the Clustal output file
contains an asterisk if the sequence is conserved at each position. This
information is stored in the <code class="docutils literal notranslate"><span class="pre">column_annotations</span></code> attribute of the
<code class="docutils literal notranslate"><span class="pre">alignment</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">column_annotations</span>  
<span class="go">{&#39;clustal_consensus&#39;: &#39;******* **** **********************************...</span>
</pre></div>
</div>
<p>Printing the <code class="docutils literal notranslate"><span class="pre">alignment</span></code> in <code class="docutils literal notranslate"><span class="pre">clustal</span></code> format will show the sequence
alignment, but does not include the metadata:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">))</span>  
<span class="go">gi|6273285|gb|AF191659.1|AF191      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA</span>
<span class="go">gi|6273284|gb|AF191658.1|AF191      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA</span>
<span class="go">gi|6273287|gb|AF191661.1|AF191      TATACATT...</span>
</pre></div>
</div>
<p>Writing the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> in <code class="docutils literal notranslate"><span class="pre">clustal</span></code> format will include both the
metadata and the sequence alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">stream</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>  
<span class="go">CLUSTAL 2.1 multiple sequence alignment</span>


<span class="go">gi|6273285|gb|AF191659.1|AF191      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA</span>
<span class="go">gi|6273284|gb|AF191658.1|AF191      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA</span>
<span class="go">gi|6273287|gb|AF191661.1|AF191      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAA</span>
<span class="go">gi|6273286|gb|AF191660.1|AF191      TATACATAAAAGAAG...</span>
</pre></div>
</div>
<p>Use an <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> (plural) object (see
Section <a class="reference internal" href="#sec-alignments"><span class="std std-ref">The Alignments class</span></a>) if you are creating alignments by
hand, and would like to include metadata information in the output.</p>
</section>
<section id="stockholm">
<span id="subsec-align-stockholm"></span><h3>Stockholm<a class="headerlink" href="#stockholm" title="Link to this heading"></a></h3>
<p>This is an example of a protein sequence alignment in the Stockholm file
format used by PFAM:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># STOCKHOLM 1.0
#=GF ID   7kD_DNA_binding
#=GF AC   PF02294.20
#=GF DE   7kD DNA-binding domain
#=GF AU   Mian N;0000-0003-4284-4749
#=GF AU   Bateman A;0000-0002-6982-4660
#=GF SE   Pfam-B_8148 (release 5.2)
#=GF GA   25.00 25.00;
#=GF TC   26.60 46.20;
#=GF NC   23.20 19.20;
#=GF BM   hmmbuild HMM.ann SEED.ann
#=GF SM   hmmsearch -Z 57096847 -E 1000 --cpu 4 HMM pfamseq
#=GF TP   Domain
#=GF CL   CL0049
#=GF RN   [1]
#=GF RM   3130377
#=GF RT   Microsequence analysis of DNA-binding proteins 7a, 7b, and 7e
#=GF RT   from the archaebacterium Sulfolobus acidocaldarius.
#=GF RA   Choli T, Wittmann-Liebold B, Reinhardt R;
#=GF RL   J Biol Chem 1988;263:7087-7093.
#=GF DR   INTERPRO; IPR003212;
#=GF DR   SCOP; 1sso; fa;
#=GF DR   SO; 0000417; polypeptide_domain;
#=GF CC   This family contains members of the hyper-thermophilic
#=GF CC   archaebacterium  7kD DNA-binding/endoribonuclease P2 family.
#=GF CC   There are five 7kD DNA-binding proteins, 7a-7e, found as
#=GF CC   monomers in the cell. Protein 7e shows the  tightest DNA-binding
#=GF CC   ability.
#=GF SQ   3
#=GS DN7_METS5/4-61   AC A4YEA2.1
#=GS DN7A_SACS2/3-61  AC P61991.2
#=GS DN7A_SACS2/3-61  DR PDB; 1SSO A; 2-60;
#=GS DN7A_SACS2/3-61  DR PDB; 1JIC A; 2-60;
#=GS DN7A_SACS2/3-61  DR PDB; 2CVR A; 2-60;
#=GS DN7A_SACS2/3-61  DR PDB; 1B4O A; 2-60;
#=GS DN7E_SULAC/3-60  AC P13125.2
DN7_METS5/4-61              KIKFKYKGQDLEVDISKVKKVWKVGKMVSFTYDD.NGKTGRGAVSEKDAPKELLNMIGK
DN7A_SACS2/3-61             TVKFKYKGEEKQVDISKIKKVWRVGKMISFTYDEGGGKTGRGAVSEKDAPKELLQMLEK
#=GR DN7A_SACS2/3-61  SS    EEEEESSSSEEEEETTTEEEEEESSSSEEEEEE-SSSSEEEEEEETTTS-CHHHHHHTT
DN7E_SULAC/3-60             KVRFKYKGEEKEVDTSKIKKVWRVGKMVSFTYDD.NGKTGRGAVSEKDAPKELMDMLAR
#=GC SS_cons                EEEEESSSSEEEEETTTEEEEEESSSSEEEEEE-SSSSEEEEEEETTTS-CHHHHHHTT
#=GC seq_cons               KVKFKYKGEEKEVDISKIKKVWRVGKMVSFTYDD.NGKTGRGAVSEKDAPKELLsMLuK
//
</pre></div>
</div>
<p>This is the seed alignment for the 7kD_DNA_binding (PF02294.20) PFAM
entry, downloaded from the InterPro website
(<a class="reference external" href="https://www.ebi.ac.uk/interpro/">https://www.ebi.ac.uk/interpro/</a>). This version of the PFAM entry is
also available in the Biopython source distribution as the file
<code class="docutils literal notranslate"><span class="pre">pfam2.seed.txt</span></code> in the subdirectory <code class="docutils literal notranslate"><span class="pre">Tests/Stockholm/</span></code>. We can load
this file as follows:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;pfam2.seed.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (3 rows x 59 columns) at ...&gt;</span>
</pre></div>
</div>
<p>We can print out a summary of the alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">DN7_METS5         0 KIKFKYKGQDLEVDISKVKKVWKVGKMVSFTYDD-NGKTGRGAVSEKDAPKELLNMIGK</span>
<span class="go">DN7A_SACS         0 TVKFKYKGEEKQVDISKIKKVWRVGKMISFTYDEGGGKTGRGAVSEKDAPKELLQMLEK</span>
<span class="go">DN7E_SULA         0 KVRFKYKGEEKEVDTSKIKKVWRVGKMVSFTYDD-NGKTGRGAVSEKDAPKELMDMLAR</span>

<span class="go">DN7_METS5        58</span>
<span class="go">DN7A_SACS        59</span>
<span class="go">DN7E_SULA        58</span>
</pre></div>
</div>
<p>You could also call Python’s built-in <code class="docutils literal notranslate"><span class="pre">format</span></code> function on the
alignment object to show it in a particular file format (see
section <a class="reference internal" href="#subsec-align-printing"><span class="std std-ref">Printing alignments</span></a> for details), for example in
the Stockholm format to regenerate the file:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">))</span>
<span class="go"># STOCKHOLM 1.0</span>
<span class="go">#=GF ID   7kD_DNA_binding</span>
<span class="go">#=GF AC   PF02294.20</span>
<span class="go">#=GF DE   7kD DNA-binding domain</span>
<span class="go">#=GF AU   Mian N;0000-0003-4284-4749</span>
<span class="go">#=GF AU   Bateman A;0000-0002-6982-4660</span>
<span class="go">#=GF SE   Pfam-B_8148 (release 5.2)</span>
<span class="go">#=GF GA   25.00 25.00;</span>
<span class="go">#=GF TC   26.60 46.20;</span>
<span class="go">#=GF NC   23.20 19.20;</span>
<span class="go">#=GF BM   hmmbuild HMM.ann SEED.ann</span>
<span class="go">#=GF SM   hmmsearch -Z 57096847 -E 1000 --cpu 4 HMM pfamseq</span>
<span class="go">#=GF TP   Domain</span>
<span class="go">#=GF CL   CL0049</span>
<span class="go">#=GF RN   [1]</span>
<span class="go">#=GF RM   3130377</span>
<span class="go">#=GF RT   Microsequence analysis of DNA-binding proteins 7a, 7b, and 7e from</span>
<span class="go">#=GF RT   the archaebacterium Sulfolobus acidocaldarius.</span>
<span class="go">#=GF RA   Choli T, Wittmann-Liebold B, Reinhardt R;</span>
<span class="go">#=GF RL   J Biol Chem 1988;263:7087-7093.</span>
<span class="go">#=GF DR   INTERPRO; IPR003212;</span>
<span class="go">#=GF DR   SCOP; 1sso; fa;</span>
<span class="go">#=GF DR   SO; 0000417; polypeptide_domain;</span>
<span class="go">#=GF CC   This family contains members of the hyper-thermophilic</span>
<span class="go">#=GF CC   archaebacterium  7kD DNA-binding/endoribonuclease P2 family. There</span>
<span class="go">#=GF CC   are five 7kD DNA-binding proteins, 7a-7e, found as monomers in the</span>
<span class="go">#=GF CC   cell. Protein 7e shows the  tightest DNA-binding ability.</span>
<span class="go">#=GF SQ   3</span>
<span class="go">#=GS DN7_METS5/4-61   AC A4YEA2.1</span>
<span class="go">#=GS DN7A_SACS2/3-61  AC P61991.2</span>
<span class="go">#=GS DN7A_SACS2/3-61  DR PDB; 1SSO A; 2-60;</span>
<span class="go">#=GS DN7A_SACS2/3-61  DR PDB; 1JIC A; 2-60;</span>
<span class="go">#=GS DN7A_SACS2/3-61  DR PDB; 2CVR A; 2-60;</span>
<span class="go">#=GS DN7A_SACS2/3-61  DR PDB; 1B4O A; 2-60;</span>
<span class="go">#=GS DN7E_SULAC/3-60  AC P13125.2</span>
<span class="go">DN7_METS5/4-61                  KIKFKYKGQDLEVDISKVKKVWKVGKMVSFTYDD.NGKTGRGAVSEKDAPKELLNMIGK</span>
<span class="go">DN7A_SACS2/3-61                 TVKFKYKGEEKQVDISKIKKVWRVGKMISFTYDEGGGKTGRGAVSEKDAPKELLQMLEK</span>
<span class="go">#=GR DN7A_SACS2/3-61  SS        EEEEESSSSEEEEETTTEEEEEESSSSEEEEEE-SSSSEEEEEEETTTS-CHHHHHHTT</span>
<span class="go">DN7E_SULAC/3-60                 KVRFKYKGEEKEVDTSKIKKVWRVGKMVSFTYDD.NGKTGRGAVSEKDAPKELMDMLAR</span>
<span class="go">#=GC SS_cons                    EEEEESSSSEEEEETTTEEEEEESSSSEEEEEE-SSSSEEEEEEETTTS-CHHHHHHTT</span>
<span class="go">#=GC seq_cons                   KVKFKYKGEEKEVDISKIKKVWRVGKMVSFTYDD.NGKTGRGAVSEKDAPKELLsMLuK</span>
<span class="go">//</span>
</pre></div>
</div>
<p>or alternatively as aligned FASTA (see section
<a class="reference internal" href="#subsec-align-fasta"><span class="std std-ref">Aligned FASTA</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>
<span class="go">&gt;DN7_METS5/4-61</span>
<span class="go">KIKFKYKGQDLEVDISKVKKVWKVGKMVSFTYDD-NGKTGRGAVSEKDAPKELLNMIGK</span>
<span class="go">&gt;DN7A_SACS2/3-61</span>
<span class="go">TVKFKYKGEEKQVDISKIKKVWRVGKMISFTYDEGGGKTGRGAVSEKDAPKELLQMLEK</span>
<span class="go">&gt;DN7E_SULAC/3-60</span>
<span class="go">KVRFKYKGEEKEVDTSKIKKVWRVGKMVSFTYDD-NGKTGRGAVSEKDAPKELMDMLAR</span>
</pre></div>
</div>
<p>or in the PHYLIP format (see section <a class="reference internal" href="#subsec-align-phylip"><span class="std std-ref">PHYLIP output files</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">))</span>
<span class="go">3 59</span>
<span class="go">DN7_METS5/KIKFKYKGQDLEVDISKVKKVWKVGKMVSFTYDD-NGKTGRGAVSEKDAPKELLNMIGK</span>
<span class="go">DN7A_SACS2TVKFKYKGEEKQVDISKIKKVWRVGKMISFTYDEGGGKTGRGAVSEKDAPKELLQMLEK</span>
<span class="go">DN7E_SULACKVRFKYKGEEKEVDTSKIKKVWRVGKMVSFTYDD-NGKTGRGAVSEKDAPKELMDMLAR</span>
</pre></div>
</div>
<p>General information of the alignment is stored under the <code class="docutils literal notranslate"><span class="pre">annotations</span></code>
attribute of the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object, for example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;identifier&quot;</span><span class="p">]</span>
<span class="go">&#39;7kD_DNA_binding&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;clan&quot;</span><span class="p">]</span>
<span class="go">&#39;CL0049&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;database references&quot;</span><span class="p">]</span>
<span class="go">[{&#39;reference&#39;: &#39;INTERPRO; IPR003212;&#39;}, {&#39;reference&#39;: &#39;SCOP; 1sso; fa;&#39;}, {&#39;reference&#39;: &#39;SO; 0000417; polypeptide_domain;&#39;}]</span>
</pre></div>
</div>
<p>The individual sequences in this alignment are stored under
<code class="docutils literal notranslate"><span class="pre">alignment.sequences</span></code> as <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code>s, including any annotations
associated with each sequence record:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;accession&quot;</span><span class="p">],</span> <span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">))</span>
<span class="gp">...</span>
<span class="go">DN7_METS5/4-61 A4YEA2.1 []</span>
<span class="go">DN7A_SACS2/3-61 P61991.2 [&#39;PDB; 1SSO A; 2-60;&#39;, &#39;PDB; 1JIC A; 2-60;&#39;, &#39;PDB; 2CVR A; 2-60;&#39;, &#39;PDB; 1B4O A; 2-60;&#39;]</span>
<span class="go">DN7E_SULAC/3-60 P13125.2 []</span>
</pre></div>
</div>
<p>The secondary structure of the second sequence (<code class="docutils literal notranslate"><span class="pre">DN7A_SACS2/3-61</span></code>) is
stored in the <code class="docutils literal notranslate"><span class="pre">letter_annotations</span></code> attribute of the <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">letter_annotations</span>
<span class="go">{}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">letter_annotations</span>
<span class="go">{&#39;secondary structure&#39;: &#39;EEEEESSSSEEEEETTTEEEEEESSSSEEEEEE-SSSSEEEEEEETTTS-CHHHHHHTT&#39;}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">letter_annotations</span>
<span class="go">{}</span>
</pre></div>
</div>
<p>The consensus sequence and secondary structure are associated with the
sequence alignment as a whole, and are therefore stored in the
<code class="docutils literal notranslate"><span class="pre">column_annotations</span></code> attribute of the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">column_annotations</span>  
<span class="go">{&#39;consensus secondary structure&#39;: &#39;EEEEESSSSEEEEETTTEEEEEESSSSEEEEEE-SSSSEEEEEEETTTS-CHHHHHHTT&#39;,</span>
<span class="go"> &#39;consensus sequence&#39;: &#39;KVKFKYKGEEKEVDISKIKKVWRVGKMVSFTYDD.NGKTGRGAVSEKDAPKELLsMLuK&#39;}</span>
</pre></div>
</div>
</section>
<section id="phylip-output-files">
<span id="subsec-align-phylip"></span><h3>PHYLIP output files<a class="headerlink" href="#phylip-output-files" title="Link to this heading"></a></h3>
<p>The PHYLIP format for sequence alignments is derived from the PHYLogeny
Interference Package from Joe Felsenstein. Files in the PHYLIP format
start with two numbers for the number of rows and columns in the printed
alignment. The sequence alignment itself can be in sequential format or
in interleaved format. An example of the former is the
<code class="docutils literal notranslate"><span class="pre">sequential.phy</span></code> file (provided in <code class="docutils literal notranslate"><span class="pre">Tests/Phylip/</span></code> in the Biopython
source distribution):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span> 3 384
CYS1_DICDI   -----MKVIL LFVLAVFTVF VSS------- --------RG IPPEEQ---- --------SQ
             FLEFQDKFNK KY-SHEEYLE RFEIFKSNLG KIEELNLIAI NHKADTKFGV NKFADLSSDE
             FKNYYLNNKE AIFTDDLPVA DYLDDEFINS IPTAFDWRTR G-AVTPVKNQ GQCGSCWSFS
             TTGNVEGQHF ISQNKLVSLS EQNLVDCDHE CMEYEGEEAC DEGCNGGLQP NAYNYIIKNG
             GIQTESSYPY TAETGTQCNF NSANIGAKIS NFTMIP-KNE TVMAGYIVST GPLAIAADAV
             E-WQFYIGGV F-DIPCN--P NSLDHGILIV GYSAKNTIFR KNMPYWIVKN SWGADWGEQG
             YIYLRRGKNT CGVSNFVSTS II--
ALEU_HORVU   MAHARVLLLA LAVLATAAVA VASSSSFADS NPIRPVTDRA ASTLESAVLG ALGRTRHALR
             FARFAVRYGK SYESAAEVRR RFRIFSESLE EVRSTN---- RKGLPYRLGI NRFSDMSWEE
             FQATRL-GAA QTCSATLAGN HLMRDA--AA LPETKDWRED G-IVSPVKNQ AHCGSCWTFS
             TTGALEAAYT QATGKNISLS EQQLVDCAGG FNNF------ --GCNGGLPS QAFEYIKYNG
             GIDTEESYPY KGVNGV-CHY KAENAAVQVL DSVNITLNAE DELKNAVGLV RPVSVAFQVI
             DGFRQYKSGV YTSDHCGTTP DDVNHAVLAV GYGVENGV-- ---PYWLIKN SWGADWGDNG
             YFKMEMGKNM CAIATCASYP VVAA
CATH_HUMAN   ------MWAT LPLLCAGAWL LGV------- -PVCGAAELS VNSLEK---- --------FH
             FKSWMSKHRK TY-STEEYHH RLQTFASNWR KINAHN---- NGNHTFKMAL NQFSDMSFAE
             IKHKYLWSEP QNCSAT--KS NYLRGT--GP YPPSVDWRKK GNFVSPVKNQ GACGSCWTFS
             TTGALESAIA IATGKMLSLA EQQLVDCAQD FNNY------ --GCQGGLPS QAFEYILYNK
             GIMGEDTYPY QGKDGY-CKF QPGKAIGFVK DVANITIYDE EAMVEAVALY NPVSFAFEVT
             QDFMMYRTGI YSSTSCHKTP DKVNHAVLAV GYGEKNGI-- ---PYWIVKN SWGPQWGMNG
             YFLIERGKNM CGLAACASYP IPLV
</pre></div>
</div>
<p>In the sequential format, the complete alignment for one sequence is
shown before proceeding to the next sequence. In the interleaved format,
the alignments for different sequences are next to each other, for
example in the file <code class="docutils literal notranslate"><span class="pre">interlaced.phy</span></code> (provided in <code class="docutils literal notranslate"><span class="pre">Tests/Phylip/</span></code> in
the Biopython source distribution):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span> 3 384
CYS1_DICDI   -----MKVIL LFVLAVFTVF VSS------- --------RG IPPEEQ---- --------SQ
ALEU_HORVU   MAHARVLLLA LAVLATAAVA VASSSSFADS NPIRPVTDRA ASTLESAVLG ALGRTRHALR
CATH_HUMAN   ------MWAT LPLLCAGAWL LGV------- -PVCGAAELS VNSLEK---- --------FH

             FLEFQDKFNK KY-SHEEYLE RFEIFKSNLG KIEELNLIAI NHKADTKFGV NKFADLSSDE
             FARFAVRYGK SYESAAEVRR RFRIFSESLE EVRSTN---- RKGLPYRLGI NRFSDMSWEE
             FKSWMSKHRK TY-STEEYHH RLQTFASNWR KINAHN---- NGNHTFKMAL NQFSDMSFAE

             FKNYYLNNKE AIFTDDLPVA DYLDDEFINS IPTAFDWRTR G-AVTPVKNQ GQCGSCWSFS
             FQATRL-GAA QTCSATLAGN HLMRDA--AA LPETKDWRED G-IVSPVKNQ AHCGSCWTFS
             IKHKYLWSEP QNCSAT--KS NYLRGT--GP YPPSVDWRKK GNFVSPVKNQ GACGSCWTFS

             TTGNVEGQHF ISQNKLVSLS EQNLVDCDHE CMEYEGEEAC DEGCNGGLQP NAYNYIIKNG
             TTGALEAAYT QATGKNISLS EQQLVDCAGG FNNF------ --GCNGGLPS QAFEYIKYNG
             TTGALESAIA IATGKMLSLA EQQLVDCAQD FNNY------ --GCQGGLPS QAFEYILYNK

             GIQTESSYPY TAETGTQCNF NSANIGAKIS NFTMIP-KNE TVMAGYIVST GPLAIAADAV
             GIDTEESYPY KGVNGV-CHY KAENAAVQVL DSVNITLNAE DELKNAVGLV RPVSVAFQVI
             GIMGEDTYPY QGKDGY-CKF QPGKAIGFVK DVANITIYDE EAMVEAVALY NPVSFAFEVT

             E-WQFYIGGV F-DIPCN--P NSLDHGILIV GYSAKNTIFR KNMPYWIVKN SWGADWGEQG
             DGFRQYKSGV YTSDHCGTTP DDVNHAVLAV GYGVENGV-- ---PYWLIKN SWGADWGDNG
             QDFMMYRTGI YSSTSCHKTP DKVNHAVLAV GYGEKNGI-- ---PYWIVKN SWGPQWGMNG

             YIYLRRGKNT CGVSNFVSTS II--
             YFKMEMGKNM CAIATCASYP VVAA
             YFLIERGKNM CGLAACASYP IPLV
</pre></div>
</div>
<p>The parser in <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> detects from the file contents if it is in
the sequential or in the interleaved format, and then parses it
appropriately.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;sequential.phy&quot;</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (3 rows x 384 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;interlaced.phy&quot;</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span>  
<span class="go">&lt;Alignment object (3 rows x 384 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">==</span> <span class="n">alignment2</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Here, two alignments are considered to be equal if they have the same
sequence contents and the same alignment coordinates.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(3, 384)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">CYS1_DICD         0 -----MKVILLFVLAVFTVFVSS---------------RGIPPEEQ------------SQ</span>
<span class="go">ALEU_HORV         0 MAHARVLLLALAVLATAAVAVASSSSFADSNPIRPVTDRAASTLESAVLGALGRTRHALR</span>
<span class="go">CATH_HUMA         0 ------MWATLPLLCAGAWLLGV--------PVCGAAELSVNSLEK------------FH</span>

<span class="go">CYS1_DICD        28 FLEFQDKFNKKY-SHEEYLERFEIFKSNLGKIEELNLIAINHKADTKFGVNKFADLSSDE</span>
<span class="go">ALEU_HORV        60 FARFAVRYGKSYESAAEVRRRFRIFSESLEEVRSTN----RKGLPYRLGINRFSDMSWEE</span>
<span class="go">CATH_HUMA        34 FKSWMSKHRKTY-STEEYHHRLQTFASNWRKINAHN----NGNHTFKMALNQFSDMSFAE</span>

<span class="go">CYS1_DICD        87 FKNYYLNNKEAIFTDDLPVADYLDDEFINSIPTAFDWRTRG-AVTPVKNQGQCGSCWSFS</span>
<span class="go">ALEU_HORV       116 FQATRL-GAAQTCSATLAGNHLMRDA--AALPETKDWREDG-IVSPVKNQAHCGSCWTFS</span>
<span class="go">CATH_HUMA        89 IKHKYLWSEPQNCSAT--KSNYLRGT--GPYPPSVDWRKKGNFVSPVKNQGACGSCWTFS</span>

<span class="go">CYS1_DICD       146 TTGNVEGQHFISQNKLVSLSEQNLVDCDHECMEYEGEEACDEGCNGGLQPNAYNYIIKNG</span>
<span class="go">ALEU_HORV       172 TTGALEAAYTQATGKNISLSEQQLVDCAGGFNNF--------GCNGGLPSQAFEYIKYNG</span>
<span class="go">CATH_HUMA       145 TTGALESAIAIATGKMLSLAEQQLVDCAQDFNNY--------GCQGGLPSQAFEYILYNK</span>

<span class="go">CYS1_DICD       206 GIQTESSYPYTAETGTQCNFNSANIGAKISNFTMIP-KNETVMAGYIVSTGPLAIAADAV</span>
<span class="go">ALEU_HORV       224 GIDTEESYPYKGVNGV-CHYKAENAAVQVLDSVNITLNAEDELKNAVGLVRPVSVAFQVI</span>
<span class="go">CATH_HUMA       197 GIMGEDTYPYQGKDGY-CKFQPGKAIGFVKDVANITIYDEEAMVEAVALYNPVSFAFEVT</span>

<span class="go">CYS1_DICD       265 E-WQFYIGGVF-DIPCN--PNSLDHGILIVGYSAKNTIFRKNMPYWIVKNSWGADWGEQG</span>
<span class="go">ALEU_HORV       283 DGFRQYKSGVYTSDHCGTTPDDVNHAVLAVGYGVENGV-----PYWLIKNSWGADWGDNG</span>
<span class="go">CATH_HUMA       256 QDFMMYRTGIYSSTSCHKTPDKVNHAVLAVGYGEKNGI-----PYWIVKNSWGPQWGMNG</span>

<span class="go">CYS1_DICD       321 YIYLRRGKNTCGVSNFVSTSII-- 343</span>
<span class="go">ALEU_HORV       338 YFKMEMGKNMCAIATCASYPVVAA 362</span>
<span class="go">CATH_HUMA       311 YFLIERGKNMCGLAACASYPIPLV 335</span>
</pre></div>
</div>
<p>When outputting the alignment in PHYLIP format, <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> writes
each of the aligned sequences on one line:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">))</span>
<span class="go">3 384</span>
<span class="go">CYS1_DICDI-----MKVILLFVLAVFTVFVSS---------------RGIPPEEQ------------SQFLEFQDKFNKKY-SHEEYLERFEIFKSNLGKIEELNLIAINHKADTKFGVNKFADLSSDEFKNYYLNNKEAIFTDDLPVADYLDDEFINSIPTAFDWRTRG-AVTPVKNQGQCGSCWSFSTTGNVEGQHFISQNKLVSLSEQNLVDCDHECMEYEGEEACDEGCNGGLQPNAYNYIIKNGGIQTESSYPYTAETGTQCNFNSANIGAKISNFTMIP-KNETVMAGYIVSTGPLAIAADAVE-WQFYIGGVF-DIPCN--PNSLDHGILIVGYSAKNTIFRKNMPYWIVKNSWGADWGEQGYIYLRRGKNTCGVSNFVSTSII--</span>
<span class="go">ALEU_HORVUMAHARVLLLALAVLATAAVAVASSSSFADSNPIRPVTDRAASTLESAVLGALGRTRHALRFARFAVRYGKSYESAAEVRRRFRIFSESLEEVRSTN----RKGLPYRLGINRFSDMSWEEFQATRL-GAAQTCSATLAGNHLMRDA--AALPETKDWREDG-IVSPVKNQAHCGSCWTFSTTGALEAAYTQATGKNISLSEQQLVDCAGGFNNF--------GCNGGLPSQAFEYIKYNGGIDTEESYPYKGVNGV-CHYKAENAAVQVLDSVNITLNAEDELKNAVGLVRPVSVAFQVIDGFRQYKSGVYTSDHCGTTPDDVNHAVLAVGYGVENGV-----PYWLIKNSWGADWGDNGYFKMEMGKNMCAIATCASYPVVAA</span>
<span class="go">CATH_HUMAN------MWATLPLLCAGAWLLGV--------PVCGAAELSVNSLEK------------FHFKSWMSKHRKTY-STEEYHHRLQTFASNWRKINAHN----NGNHTFKMALNQFSDMSFAEIKHKYLWSEPQNCSAT--KSNYLRGT--GPYPPSVDWRKKGNFVSPVKNQGACGSCWTFSTTGALESAIAIATGKMLSLAEQQLVDCAQDFNNY--------GCQGGLPSQAFEYILYNKGIMGEDTYPYQGKDGY-CKFQPGKAIGFVKDVANITIYDEEAMVEAVALYNPVSFAFEVTQDFMMYRTGIYSSTSCHKTPDKVNHAVLAVGYGEKNGI-----PYWIVKNSWGPQWGMNGYFLIERGKNMCGLAACASYPIPLV</span>
</pre></div>
</div>
<p>We can write the alignment in PHYLIP format, parse the result, and
confirm it is the same as the original alignment object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment3</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">==</span> <span class="n">alignment3</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">]</span>
<span class="go">[&#39;CYS1_DICDI&#39;, &#39;ALEU_HORVU&#39;, &#39;CATH_HUMAN&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment3</span><span class="o">.</span><span class="n">sequences</span><span class="p">]</span>
<span class="go">[&#39;CYS1_DICDI&#39;, &#39;ALEU_HORVU&#39;, &#39;CATH_HUMAN&#39;]</span>
</pre></div>
</div>
</section>
<section id="emboss">
<span id="subsec-align-emboss"></span><h3>EMBOSS<a class="headerlink" href="#emboss" title="Link to this heading"></a></h3>
<p>EMBOSS (European Molecular Biology Open Software Suite) is a set of
open-source software tools for molecular biology and
bioinformatics <a class="reference internal" href="chapter_bibliography.html#rice2000" id="id1"><span>[Rice2000]</span></a>. It includes software such
as <code class="docutils literal notranslate"><span class="pre">needle</span></code> and <code class="docutils literal notranslate"><span class="pre">water</span></code> for pairwise sequence alignment. This is an
example of output generated by the <code class="docutils literal notranslate"><span class="pre">water</span></code> program for Smith-Waterman
local pairwise sequence alignment (available as <code class="docutils literal notranslate"><span class="pre">water.txt</span></code> in the
<code class="docutils literal notranslate"><span class="pre">Tests/Emboss</span></code> directory of the Biopython distribution):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>########################################
# Program:  water
# Rundate:  Wed Jan 16 17:23:19 2002
# Report_file: stdout
########################################
#=======================================
#
# Aligned_sequences: 2
# 1: IXI_234
# 2: IXI_235
# Matrix: EBLOSUM62
# Gap_penalty: 10.0
# Extend_penalty: 0.5
#
# Length: 131
# Identity:     112/131 (85.5%)
# Similarity:   112/131 (85.5%)
# Gaps:          19/131 (14.5%)
# Score: 591.5
#
#
#=======================================

IXI_234            1 TSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGRPCCSAAPRRPQAT     50
                     |||||||||||||||         ||||||||||||||||||||||||||
IXI_235            1 TSPASIRPPAGPSSR---------RPSPPGPRRPTGRPCCSAAPRRPQAT     41

IXI_234           51 GGWKTCSGTCTTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSRSAG    100
                     ||||||||||||||||||||||||          ||||||||||||||||
IXI_235           42 GGWKTCSGTCTTSTSTRHRGRSGW----------RASRKSMRAACSRSAG     81

IXI_234          101 SRPNRFAPTLMSSCITSTTGPPAWAGDRSHE    131
                     |||||||||||||||||||||||||||||||
IXI_235           82 SRPNRFAPTLMSSCITSTTGPPAWAGDRSHE    112


#---------------------------------------
#---------------------------------------
</pre></div>
</div>
<p>As this output file contains only one alignment, we can use
<code class="docutils literal notranslate"><span class="pre">Align.read</span></code> to extract it directly. Here, instead we will use
<code class="docutils literal notranslate"><span class="pre">Align.parse</span></code> so we can see the metadata of this <code class="docutils literal notranslate"><span class="pre">water</span></code> run:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;water.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;emboss&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">metadata</span></code> attribute of <code class="docutils literal notranslate"><span class="pre">alignments</span></code> stores the information
shown in the header of the file, including the program used to generate
the output, the date and time the program was run, the output file name,
and the specific alignment file format that was used (assumed to be
<code class="docutils literal notranslate"><span class="pre">srspair</span></code> by default):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;Align_format&#39;: &#39;srspair&#39;, &#39;Program&#39;: &#39;water&#39;, &#39;Rundate&#39;: &#39;Wed Jan 16 17:23:19 2002&#39;, &#39;Report_file&#39;: &#39;stdout&#39;}</span>
</pre></div>
</div>
<p>To pull out the alignment, we use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (2 rows x 131 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(2, 131)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">IXI_234           0 TSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGTC</span>
<span class="go">                  0 |||||||||||||||---------||||||||||||||||||||||||||||||||||||</span>
<span class="go">IXI_235           0 TSPASIRPPAGPSSR---------RPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGTC</span>

<span class="go">IXI_234          60 TTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTTG</span>
<span class="go">                 60 ||||||||||||||----------||||||||||||||||||||||||||||||||||||</span>
<span class="go">IXI_235          51 TTSTSTRHRGRSGW----------RASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTTG</span>

<span class="go">IXI_234         120 PPAWAGDRSHE 131</span>
<span class="go">                120 ||||||||||| 131</span>
<span class="go">IXI_235         101 PPAWAGDRSHE 112</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[  0,  15,  24,  74,  84, 131],</span>
<span class="go">       [  0,  15,  15,  65,  65, 112]])</span>
</pre></div>
</div>
<p>We can use indices to extract specific parts of the alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;TSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGTCTTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTTGPPAWAGDRSHE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;TSPASIRPPAGPSSR---------RPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGTCTTSTSTRHRGRSGW----------RASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTTGPPAWAGDRSHE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">:</span><span class="mi">30</span><span class="p">]</span>
<span class="go">&#39;GPSSR---------RPSPPG&#39;</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">annotations</span></code> attribute of the <code class="docutils literal notranslate"><span class="pre">alignment</span></code> stores the
information associated with this alignment specifically:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;Matrix&#39;: &#39;EBLOSUM62&#39;, &#39;Gap_penalty&#39;: 10.0, &#39;Extend_penalty&#39;: 0.5, &#39;Identity&#39;: 112, &#39;Similarity&#39;: 112, &#39;Gaps&#39;: 19, &#39;Score&#39;: 591.5}</span>
</pre></div>
</div>
<p>The number of gaps, identities, and mismatches can also be obtained by
calling the <code class="docutils literal notranslate"><span class="pre">counts</span></code> method on the <code class="docutils literal notranslate"><span class="pre">alignment</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">counts</span><span class="p">()</span>
<span class="go">AlignmentCounts(gaps=19, identities=112, mismatches=0)</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">AlignmentCounts</span></code> is a <code class="docutils literal notranslate"><span class="pre">namedtuple</span></code> in the <code class="docutils literal notranslate"><span class="pre">collections</span></code>
module in Python’s standard library.</p>
<p>The consensus line shown between the two sequences is stored in the
<code class="docutils literal notranslate"><span class="pre">column_annotations</span></code> attribute:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">column_annotations</span>
<span class="go">{&#39;emboss_consensus&#39;: &#39;|||||||||||||||         ||||||||||||||||||||||||||||||||||||||||||||||||||          |||||||||||||||||||||||||||||||||||||||||||||||&#39;}</span>
</pre></div>
</div>
<p>Use the <code class="docutils literal notranslate"><span class="pre">format</span></code> function (or the <code class="docutils literal notranslate"><span class="pre">format</span></code> method) to print the
alignment in other formats, for example in the PHYLIP format (see
section <a class="reference internal" href="#subsec-align-phylip"><span class="std std-ref">PHYLIP output files</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">))</span>
<span class="go">2 131</span>
<span class="go">IXI_234   TSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGTCTTSTSTRHRGRSGWSARTTTAACLRASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTTGPPAWAGDRSHE</span>
<span class="go">IXI_235   TSPASIRPPAGPSSR---------RPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGTCTTSTSTRHRGRSGW----------RASRKSMRAACSRSAGSRPNRFAPTLMSSCITSTTGPPAWAGDRSHE</span>
</pre></div>
</div>
<p>We can use <code class="docutils literal notranslate"><span class="pre">alignment.sequences</span></code> to get the individual sequences.
However, as this is a pairwise alignment, we can also use
<code class="docutils literal notranslate"><span class="pre">alignment.target</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment.query</span></code> to get the target and query
sequences:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span>
<span class="go">SeqRecord(seq=Seq(&#39;TSPASIRPPAGPSSRPAMVSSRRTRPSPPGPRRPTGRPCCSAAPRRPQATGGWK...SHE&#39;), id=&#39;IXI_234&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&lt;unknown description&gt;&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span>
<span class="go">SeqRecord(seq=Seq(&#39;TSPASIRPPAGPSSRRPSPPGPRRPTGRPCCSAAPRRPQATGGWKTCSGTCTTS...SHE&#39;), id=&#39;IXI_235&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&lt;unknown description&gt;&#39;, dbxrefs=[])</span>
</pre></div>
</div>
<p>Currently, Biopython does not support writing sequence alignments in the
output formats defined by EMBOSS.</p>
</section>
<section id="gcg-multiple-sequence-format-msf">
<span id="subsec-align-msf"></span><h3>GCG Multiple Sequence Format (MSF)<a class="headerlink" href="#gcg-multiple-sequence-format-msf" title="Link to this heading"></a></h3>
<p>The Multiple Sequence Format (MSF) was created to store multiple
sequence alignments generated by the GCG (Genetics Computer Group) set
of programs. The file <code class="docutils literal notranslate"><span class="pre">W_prot.msf</span></code> in the <code class="docutils literal notranslate"><span class="pre">Tests/msf</span></code> directory of
the Biopython distribution is an example of a sequence alignment file in
the MSF format This file shows an alignment of 11 protein sequences:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>!!AA_MULTIPLE_ALIGNMENT

   MSF: 99  Type: P  Oct 18, 2017  11:35  Check: 0 ..

 Name: W*01:01:01:01    Len:    99  Check: 7236  Weight:  1.00
 Name: W*01:01:01:02    Len:    99  Check: 7236  Weight:  1.00
 Name: W*01:01:01:03    Len:    99  Check: 7236  Weight:  1.00
 Name: W*01:01:01:04    Len:    99  Check: 7236  Weight:  1.00
 Name: W*01:01:01:05    Len:    99  Check: 7236  Weight:  1.00
 Name: W*01:01:01:06    Len:    99  Check: 7236  Weight:  1.00
 Name: W*02:01          Len:    93  Check: 9483  Weight:  1.00
 Name: W*03:01:01:01    Len:    93  Check: 9974  Weight:  1.00
 Name: W*03:01:01:02    Len:    93  Check: 9974  Weight:  1.00
 Name: W*04:01          Len:    93  Check: 9169  Weight:  1.00
 Name: W*05:01          Len:    99  Check: 7331  Weight:  1.00
//

  W*01:01:01:01  GLTPFNGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ
  W*01:01:01:02  GLTPFNGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ
  W*01:01:01:03  GLTPFNGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ
  W*01:01:01:04  GLTPFNGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ
  W*01:01:01:05  GLTPFNGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ
  W*01:01:01:06  GLTPFNGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ
        W*02:01  GLTPSNGYTA ATWTRTAASS VGMNIPYDGA SYLVRNQELR SWTAADKAAQ
  W*03:01:01:01  GLTPSSGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ
  W*03:01:01:02  GLTPSSGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ
        W*04:01  GLTPSNGYTA ATWTRTAASS VGMNIPYDGA SYLVRNQELR SWTAADKAAQ
        W*05:01  GLTPSSGYTA ATWTRTAVSS VGMNIPYHGA SYLVRNQELR SWTAADKAAQ

  W*01:01:01:01  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK DSHDPPPHL
  W*01:01:01:02  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK DSHDPPPHL
  W*01:01:01:03  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK DSHDPPPHL
  W*01:01:01:04  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK DSHDPPPHL
  W*01:01:01:05  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK DSHDPPPHL
  W*01:01:01:06  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK DSHDPPPHL
        W*02:01  MPWRRNMQSC SKPTCREGGR SGSAKSLRMG RRRCTAQNPK RLT
  W*03:01:01:01  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK RLT
  W*03:01:01:02  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK RLT
        W*04:01  MPWRRNMQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK RLT
        W*05:01  MPWRRNRQSC SKPTCREGGR SGSAKSLRMG RRGCSAQNPK DSHDPPPHL
</pre></div>
</div>
<p>To parse this file with Biopython, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;W_prot.msf&quot;</span><span class="p">,</span> <span class="s2">&quot;msf&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The parser skips all lines up to and including the line starting with
“<code class="docutils literal notranslate"><span class="pre">MSF:</span></code>”. The following lines (until the “<code class="docutils literal notranslate"><span class="pre">//</span></code>” demarcation) are
read by the parser to verify the length of each sequence. The alignment
section (after the “<code class="docutils literal notranslate"><span class="pre">//</span></code>” demarcation) is read by the parser and
stored as an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (11 rows x 99 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">W*01:01:0         0 GLTPFNGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>
<span class="go">W*01:01:0         0 GLTPFNGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>
<span class="go">W*01:01:0         0 GLTPFNGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>
<span class="go">W*01:01:0         0 GLTPFNGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>
<span class="go">W*01:01:0         0 GLTPFNGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>
<span class="go">W*01:01:0         0 GLTPFNGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>
<span class="go">W*02:01           0 GLTPSNGYTAATWTRTAASSVGMNIPYDGASYLVRNQELRSWTAADKAAQMPWRRNMQSC</span>
<span class="go">W*03:01:0         0 GLTPSSGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>
<span class="go">W*03:01:0         0 GLTPSSGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>
<span class="go">W*04:01           0 GLTPSNGYTAATWTRTAASSVGMNIPYDGASYLVRNQELRSWTAADKAAQMPWRRNMQSC</span>
<span class="go">W*05:01           0 GLTPSSGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWRRNRQSC</span>

<span class="go">W*01:01:0        60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKDSHDPPPHL 99</span>
<span class="go">W*01:01:0        60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKDSHDPPPHL 99</span>
<span class="go">W*01:01:0        60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKDSHDPPPHL 99</span>
<span class="go">W*01:01:0        60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKDSHDPPPHL 99</span>
<span class="go">W*01:01:0        60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKDSHDPPPHL 99</span>
<span class="go">W*01:01:0        60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKDSHDPPPHL 99</span>
<span class="go">W*02:01          60 SKPTCREGGRSGSAKSLRMGRRRCTAQNPKRLT------ 93</span>
<span class="go">W*03:01:0        60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKRLT------ 93</span>
<span class="go">W*03:01:0        60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKRLT------ 93</span>
<span class="go">W*04:01          60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKRLT------ 93</span>
<span class="go">W*05:01          60 SKPTCREGGRSGSAKSLRMGRRGCSAQNPKDSHDPPPHL 99</span>
</pre></div>
</div>
<p>The sequences and their names are stored in the <code class="docutils literal notranslate"><span class="pre">alignment.sequences</span></code>
attribute:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">)</span>
<span class="go">11</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;W*01:01:01:01&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;GLTPFNGYTAATWTRTAVSSVGMNIPYHGASYLVRNQELRSWTAADKAAQMPWR...PHL&#39;)</span>
</pre></div>
</div>
<p>The alignment coordinates are stored in the <code class="docutils literal notranslate"><span class="pre">alignment.coordinates</span></code>
attribute:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[ 0, 93, 99],</span>
<span class="go">       [ 0, 93, 99],</span>
<span class="go">       [ 0, 93, 99],</span>
<span class="go">       [ 0, 93, 99],</span>
<span class="go">       [ 0, 93, 99],</span>
<span class="go">       [ 0, 93, 99],</span>
<span class="go">       [ 0, 93, 93],</span>
<span class="go">       [ 0, 93, 93],</span>
<span class="go">       [ 0, 93, 93],</span>
<span class="go">       [ 0, 93, 93],</span>
<span class="go">       [ 0, 93, 99]])</span>
</pre></div>
</div>
<p>Currently, Biopython does not support writing sequence alignments in the
MSF format.</p>
</section>
<section id="exonerate">
<span id="subsec-align-exonerate"></span><h3>Exonerate<a class="headerlink" href="#exonerate" title="Link to this heading"></a></h3>
<p>Exonerate is a generic program for pairwise sequence
alignments <a class="reference internal" href="chapter_bibliography.html#slater2005" id="id2"><span>[Slater2005]</span></a>. The sequence alignments found
by Exonerate can be output in a human-readable form, in the “cigar”
(Compact Idiosyncratic Gapped Alignment Report) format, or in the
“vulgar” (Verbose Useful Labelled Gapped Alignment Report) format. The
user can request to include one or more of these formats in the output.
The parser in <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> can only parse alignments in the cigar or
vulgar formats, and will not parse output that includes alignments in
human-readable format.</p>
<p>The file <code class="docutils literal notranslate"><span class="pre">exn_22_m_cdna2genome_vulgar.exn</span></code> in the Biopython test suite
is an example of an Exonerate output file showing the alignments in
vulgar format:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>Command line: [exonerate -m cdna2genome ../scer_cad1.fa /media/Waterloo/Downloads/genomes/scer_s288c/scer_s288c.fa --bestn 3 --showalignment no --showcigar no --showvulgar yes]
Hostname: [blackbriar]
vulgar: gi|296143771|ref|NM_001180731.1| 0 1230 + gi|330443520|ref|NC_001136.10| 1319275 1318045 - 6146 M 1 1 C 3 3 M 1226 1226
vulgar: gi|296143771|ref|NM_001180731.1| 1230 0 - gi|330443520|ref|NC_001136.10| 1318045 1319275 + 6146 M 129 129 C 3 3 M 1098 1098
vulgar: gi|296143771|ref|NM_001180731.1| 0 516 + gi|330443688|ref|NC_001145.3| 85010 667216 + 518 M 11 11 G 1 0 M 15 15 G 2 0 M 4 4 G 1 0 M 1 1 G 1 0 M 8 8 G 4 0 M 17 17 5 0 2 I 0 168904 3 0 2 M 4 4 G 0 1 M 8 8 G 2 0 M 3 3 G 1 0 M 33 33 G 0 2 M 7 7 G 0 1 M 102 102 5 0 2 I 0 96820 3 0 2 M 14 14 G 0 2 M 10 10 G 2 0 M 5 5 G 0 2 M 10 10 G 2 0 M 4 4 G 0 1 M 20 20 G 1 0 M 15 15 G 0 1 M 5 5 G 3 0 M 4 4 5 0 2 I 0 122114 3 0 2 M 20 20 G 0 5 M 6 6 5 0 2 I 0 193835 3 0 2 M 12 12 G 0 2 M 5 5 G 1 0 M 7 7 G 0 2 M 1 1 G 0 1 M 12 12 C 75 75 M 6 6 G 1 0 M 4 4 G 0 1 M 2 2 G 0 1 M 3 3 G 0 1 M 41 41
-- completed exonerate analysis
</pre></div>
</div>
<p>This file includes three alignments. To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;exn_22_m_cdna2genome_vulgar.exn&quot;</span><span class="p">,</span> <span class="s2">&quot;exonerate&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The dictionary <code class="docutils literal notranslate"><span class="pre">alignments.metadata</span></code> stores general information about
these alignments, shown at the top of the output file:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>  
<span class="go">{&#39;Program&#39;: &#39;exonerate&#39;,</span>
<span class="go"> &#39;Command line&#39;: &#39;exonerate -m cdna2genome ../scer_cad1.fa /media/Waterloo/Downloads/genomes/scer_s288c/scer_s288c.fa --bestn 3 --showalignment no --showcigar no --showvulgar yes&#39;,</span>
<span class="go"> &#39;Hostname&#39;: &#39;blackbriar&#39;}</span>
</pre></div>
</div>
<p>Now we can iterate over the alignments. The first alignment, with
alignment score 6146.0, has no gaps:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span>
<span class="go">6146.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[1319275, 1319274, 1319271, 1318045],</span>
<span class="go">       [      0,       1,       4,    1230]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>  
<span class="go">gi|330443   1319275 ????????????????????????????????????????????????????????????</span>
<span class="go">                  0 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="go">gi|296143         0 ????????????????????????????????????????????????????????????</span>
<span class="go">...</span>
<span class="go">gi|330443   1318075 ?????????????????????????????? 1318045</span>
<span class="go">               1200 ||||||||||||||||||||||||||||||    1230</span>
<span class="go">gi|296143      1200 ??????????????????????????????    1230</span>
</pre></div>
</div>
<p>Note that the target (the first sequence) in the printed alignment is on
the reverse strand while the query (the second sequence) is on the
forward strand, with the target coordinate decreasing and the query
coordinate increasing. Printing this alignment in <code class="docutils literal notranslate"><span class="pre">exonerate</span></code> format
using Python’s built-in <code class="docutils literal notranslate"><span class="pre">format</span></code> function writes a vulgar line:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;exonerate&quot;</span><span class="p">))</span>
<span class="go">vulgar: gi|296143771|ref|NM_001180731.1| 0 1230 + gi|330443520|ref|NC_001136.10| 1319275 1318045 - 6146 M 1 1 C 3 3 M 1226 1226</span>
</pre></div>
</div>
<p>Using the <code class="docutils literal notranslate"><span class="pre">format</span></code> method allows us to request either a vulgar line
(default) or a cigar line:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;exonerate&quot;</span><span class="p">,</span> <span class="s2">&quot;vulgar&quot;</span><span class="p">))</span>
<span class="go">vulgar: gi|296143771|ref|NM_001180731.1| 0 1230 + gi|330443520|ref|NC_001136.10| 1319275 1318045 - 6146 M 1 1 C 3 3 M 1226 1226</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;exonerate&quot;</span><span class="p">,</span> <span class="s2">&quot;cigar&quot;</span><span class="p">))</span>
<span class="go">cigar: gi|296143771|ref|NM_001180731.1| 0 1230 + gi|330443520|ref|NC_001136.10| 1319275 1318045 - 6146 M 1 M 3 M 1226</span>
</pre></div>
</div>
<p>The vulgar line contains information about the alignment (in the section
<code class="docutils literal notranslate"><span class="pre">M</span> <span class="pre">1</span> <span class="pre">1</span> <span class="pre">C</span> <span class="pre">3</span> <span class="pre">3</span> <span class="pre">M</span> <span class="pre">1226</span></code>) that is missing from the cigar line
<code class="docutils literal notranslate"><span class="pre">M</span> <span class="pre">1</span> <span class="pre">M</span> <span class="pre">3</span> <span class="pre">M</span> <span class="pre">1226</span></code>. The vulgar line specifies that the alignment starts
with a single aligned nucleotides, followed by three aligned nucleotides
that form a codon (<code class="docutils literal notranslate"><span class="pre">C</span></code>), followed by 1226 aligned nucleotides. In the
cigar line, we see a single aligned nucleotide, followed by three
aligned nucleotides, followed by 1226 aligned nucleotides; it does not
specify that the three aligned nucleotides form a codon. This
information from the vulgar line is stored in the <code class="docutils literal notranslate"><span class="pre">operations</span></code>
attribute:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">operations</span>
<span class="go">bytearray(b&#39;MCM&#39;)</span>
</pre></div>
</div>
<p>See the Exonerate documentation for the definition of other operation
codes.</p>
<p>Similarly, the <code class="docutils literal notranslate"><span class="pre">&quot;vulgar&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;cigar&quot;</span></code> argument can be used when
calling <code class="docutils literal notranslate"><span class="pre">Bio.Align.write</span></code> to write a file with vulgar or cigar
alignment lines.</p>
<p>We can print the alignment in BED and PSL format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">gi|330443520|ref|NC_001136.10|  1318045 1319275 gi|296143771|ref|NM_001180731.1| 6146   -   1318045 1319275 0   3   1226,3,1,   0,1226,1229,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">))</span>  
<span class="go">1230    0   0   0   0   0   0   0   -   gi|296143771|ref|NM_001180731.1|    1230    0   1230    gi|330443520|ref|NC_001136.10|  1319275 1318045 1319275 3   1226,3,1,   0,1226,1229,    1318045,1319271,1319274,</span>
</pre></div>
</div>
<p>The SAM format parser defines its own (optional) <code class="docutils literal notranslate"><span class="pre">operations</span></code>
attribute (section <a class="reference internal" href="#subsec-align-sam"><span class="std std-ref">Sequence Alignment/Map (SAM)</span></a>), which is not quite
consistent with the <code class="docutils literal notranslate"><span class="pre">operations</span></code> attribute defined in the Exonerate
format parser. As the <code class="docutils literal notranslate"><span class="pre">operations</span></code> attribute is optional, we delete it
before printing the alignment in SAM format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">alignment</span><span class="o">.</span><span class="n">operations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">))</span>  
<span class="go">gi|296143771|ref|NM_001180731.1|    16  gi|330443520|ref|NC_001136.10|  1318046 255 1226M3M1M   *   0   0   *   *   AS:i:6146</span>
</pre></div>
</div>
<p>The third alignment contains four long gaps:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>  <span class="c1"># second alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>  <span class="c1"># third alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>  
<span class="go">gi|330443     85010 ???????????-???????????????--????-?-????????----????????????</span>
<span class="go">                  0 |||||||||||-|||||||||||||||--||||-|-||||||||----||||||||||||</span>
<span class="go">gi|296143         0 ????????????????????????????????????????????????????????????</span>

<span class="go">gi|330443     85061 ????????????????????????????????????????????????????????????</span>
<span class="go">                 60 |||||-------------------------------------------------------</span>
<span class="go">gi|296143        60 ?????-------------------------------------------------------</span>
<span class="go">...</span>
<span class="go">gi|330443    666990 ????????????????????????????????????????????????????????????</span>
<span class="go">             582000 --------------------------------------------------||||||||||</span>
<span class="go">gi|296143       346 --------------------------------------------------??????????</span>

<span class="go">gi|330443    667050 ?????????-??????????????????????????????????????????????????</span>
<span class="go">             582060 ||--|||||-|||||||--|-|||||||||||||||||||||||||||||||||||||||</span>
<span class="go">gi|296143       356 ??--?????????????--?-???????????????????????????????????????</span>

<span class="go">gi|330443    667109 ??????????????????????????????????????????????????????-?????</span>
<span class="go">             582120 ||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||-</span>
<span class="go">gi|296143       411 ???????????????????????????????????????????????????????????-</span>

<span class="go">gi|330443    667168 ???????????????????????????????????????????????? 667216</span>
<span class="go">             582180 ||-|||-||||||||||||||||||||||||||||||||||||||||| 582228</span>
<span class="go">gi|296143       470 ??-???-?????????????????????????????????????????    516</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;exonerate&quot;</span><span class="p">))</span>  
<span class="go">vulgar: gi|296143771|ref|NM_001180731.1| 0 516 + gi|330443688|ref|NC_001145.3|</span>
<span class="go">85010 667216 + 518 M 11 11 G 1 0 M 15 15 G 2 0 M 4 4 G 1 0 M 1 1 G 1 0 M 8 8</span>
<span class="go"> G 4 0 M 17 17 5 0 2 I 0 168904 3 0 2 M 4 4 G 0 1 M 8 8 G 2 0 M 3 3 G 1 0</span>
<span class="go"> M 33 33 G 0 2 M 7 7 G 0 1 M 102 102 5 0 2 I 0 96820 3 0 2 M 14 14 G 0 2 M 10 10</span>
<span class="go"> G 2 0 M 5 5 G 0 2 M 10 10 G 2 0 M 4 4 G 0 1 M 20 20 G 1 0 M 15 15 G 0 1 M 5 5</span>
<span class="go"> G 3 0 M 4 4 5 0 2 I 0 122114 3 0 2 M 20 20 G 0 5 M 6 6 5 0 2 I 0 193835 3 0 2</span>
<span class="go"> M 12 12 G 0 2 M 5 5 G 1 0 M 7 7 G 0 2 M 1 1 G 0 1 M 12 12 C 75 75 M 6 6 G 1 0</span>
<span class="go"> M 4 4 G 0 1 M 2 2 G 0 1 M 3 3 G 0 1 M 41 41</span>
</pre></div>
</div>
</section>
<section id="nexus">
<span id="subsec-align-nexus"></span><h3>NEXUS<a class="headerlink" href="#nexus" title="Link to this heading"></a></h3>
<p>The NEXUS file format <a class="reference internal" href="chapter_bibliography.html#maddison1997" id="id3"><span>[Maddison1997]</span></a> is used by
several programs to store phylogenetic information. This is an example
of a file in the NEXUS format (available as <code class="docutils literal notranslate"><span class="pre">codonposset.nex</span></code> in the
<code class="docutils literal notranslate"><span class="pre">Tests/Nexus</span></code> subdirectory in the Biopython distribution):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>#NEXUS
[MacClade 4.05 registered to Computational Biologist, University]


BEGIN DATA;
       DIMENSIONS  NTAX=2 NCHAR=22;
       FORMAT DATATYPE=DNA  MISSING=? GAP=- ;
MATRIX
[                           10        20 ]
[                           .         .  ]

Aegotheles         AAAAAGGCATTGTGGTGGGAAT   [22]
Aerodramus         ?????????TTGTGGTGGGAAT   [13]
;
END;


BEGIN CODONS;
       CODONPOSSET * CodonPositions =
               N: 1-10,
               1: 11-22\3,
               2: 12-22\3,
               3: 13-22\3;
       CODESET  * UNTITLED = Universal: all ;
END;
</pre></div>
</div>
<p>In general, files in the NEXUS format can be much more complex.
<code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> relies heavily on NEXUS parser in <code class="docutils literal notranslate"><span class="pre">Bio.Nexus</span></code> (see
Chapter <a class="reference internal" href="chapter_phylo.html#chapter-phylo"><span class="std std-ref">Phylogenetics with Bio.Phylo</span></a>) to extract <code class="docutils literal notranslate"><span class="pre">Alignment</span></code>
objects from NEXUS files.</p>
<p>To read the alignment in this NEXUS file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;codonposset.nex&quot;</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">Aegothele         0 AAAAAGGCATTGTGGTGGGAAT 22</span>
<span class="go">                  0 .........||||||||||||| 22</span>
<span class="go">Aerodramu         0 ?????????TTGTGGTGGGAAT 22</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(2, 22)</span>
</pre></div>
</div>
<p>The sequences are stored under the <code class="docutils literal notranslate"><span class="pre">sequences</span></code> attribute:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;Aegotheles&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;AAAAAGGCATTGTGGTGGGAAT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;molecule_type&#39;: &#39;DNA&#39;}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;Aerodramus&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;?????????TTGTGGTGGGAAT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;molecule_type&#39;: &#39;DNA&#39;}</span>
</pre></div>
</div>
<p>To print this alignment in the NEXUS format, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;nexus&quot;</span><span class="p">))</span>
<span class="go">#NEXUS</span>
<span class="go">begin data;</span>
<span class="go">dimensions ntax=2 nchar=22;</span>
<span class="go">format datatype=dna missing=? gap=-;</span>
<span class="go">matrix</span>
<span class="go">Aegotheles AAAAAGGCATTGTGGTGGGAAT</span>
<span class="go">Aerodramus ?????????TTGTGGTGGGAAT</span>
<span class="go">;</span>
<span class="go">end;</span>
</pre></div>
</div>
<p>Similarly, you can use
<code class="docutils literal notranslate"><span class="pre">Align.write(alignment,</span> <span class="pre">&quot;myfilename.nex&quot;,</span> <span class="pre">&quot;nexus&quot;)</span></code> to write the
alignment in the NEXUS format to the file <code class="docutils literal notranslate"><span class="pre">myfilename.nex</span></code>.</p>
</section>
<section id="tabular-output-from-blast-or-fasta">
<span id="subsec-align-tabular"></span><h3>Tabular output from BLAST or FASTA<a class="headerlink" href="#tabular-output-from-blast-or-fasta" title="Link to this heading"></a></h3>
<p>Alignment output in tabular output is generated by the FASTA
aligner <a class="reference internal" href="chapter_bibliography.html#pearson1988" id="id4"><span>[Pearson1988]</span></a> run with the <code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">8CB</span></code> or
<code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">8CC</span></code> argument, or by BLAST <a class="reference internal" href="chapter_bibliography.html#altschul1990" id="id5"><span>[Altschul1990]</span></a> run
with the <code class="docutils literal notranslate"><span class="pre">-outfmt</span> <span class="pre">7</span></code> argument.</p>
<p>The file <code class="docutils literal notranslate"><span class="pre">nucleotide_m8CC.txt</span></code> in the <code class="docutils literal notranslate"><span class="pre">Tests/Fasta</span></code> subdirectory of
the Biopython source distribution is an example of an output file
generated by FASTA with the <code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">8CC</span></code> argument:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># fasta36 -m 8CC seq/mgstm1.nt seq/gst.nlib
# FASTA 36.3.8h May, 2020
# Query: pGT875  - 657 nt
# Database: seq/gst.nlib
# Fields: query id, subject id, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score, aln_code
# 12 hits found
pGT875  pGT875  100.00  657 0   0   1   657 38  694 4.6e-191    655.6   657M
pGT875  RABGLTR 79.10   646 135 0   1   646 34  679 1.6e-116    408.0   646M
pGT875  BTGST   59.56   413 167 21  176 594 228 655 1.9e-07 45.7    149M1D7M1I17M3D60M5I6M1I13M2I13M4I30M2I6M2D112M
pGT875  RABGSTB 66.93   127 42  8   159 289 157 287 3.2e-07 45.0    15M2I17M2D11M1I58M1I11M1D7M1D8M
pGT875  OCDHPR  91.30   23  2   1   266 289 2303    2325    0.012   29.7    17M1D6M
...
# FASTA processed 1 queries
</pre></div>
</div>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;nucleotide_m8CC.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;tabular&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Information shown in the file header is stored in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
attribute of <code class="docutils literal notranslate"><span class="pre">alignments</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>  
<span class="go">{&#39;Command line&#39;: &#39;fasta36 -m 8CC seq/mgstm1.nt seq/gst.nlib&#39;,</span>
<span class="go"> &#39;Program&#39;: &#39;FASTA&#39;,</span>
<span class="go"> &#39;Version&#39;: &#39;36.3.8h May, 2020&#39;,</span>
<span class="go"> &#39;Database&#39;: &#39;seq/gst.nlib&#39;}</span>
</pre></div>
</div>
<p>Extract a specific alignment by iterating over the <code class="docutils literal notranslate"><span class="pre">alignments</span></code>. As an
example, let’s go to the fourth alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">RABGSTB         156 ??????????????????????????????????--????????????????????????</span>
<span class="go">                  0 |||||||||||||||--|||||||||||||||||--|||||||||||-||||||||||||</span>
<span class="go">pGT875          158 ???????????????--??????????????????????????????-????????????</span>

<span class="go">RABGSTB         214 ??????????????????????????????????????????????????????????-?</span>
<span class="go">                 60 ||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||-|</span>
<span class="go">pGT875          215 ??????????????????????????????????????????????-?????????????</span>

<span class="go">RABGSTB         273 ??????-???????? 287</span>
<span class="go">                120 ||||||-|||||||| 135</span>
<span class="go">pGT875          274 ??????????????? 289</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>  
<span class="go">array([[156, 171, 173, 190, 190, 201, 202, 260, 261, 272, 272, 279, 279, 287],</span>
<span class="go">       [158, 173, 173, 190, 192, 203, 203, 261, 261, 272, 273, 280, 281, 289]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[156, 171],</span>
<span class="go">        [173, 190],</span>
<span class="go">        [190, 201],</span>
<span class="go">        [202, 260],</span>
<span class="go">        [261, 272],</span>
<span class="go">        [272, 279],</span>
<span class="go">        [279, 287]],</span>

<span class="go">       [[158, 173],</span>
<span class="go">        [173, 190],</span>
<span class="go">        [192, 203],</span>
<span class="go">        [203, 261],</span>
<span class="go">        [261, 272],</span>
<span class="go">        [273, 280],</span>
<span class="go">        [281, 289]]])</span>
</pre></div>
</div>
<p>The sequence information of the target and query sequences is stored in
the <code class="docutils literal notranslate"><span class="pre">target</span></code> and <code class="docutils literal notranslate"><span class="pre">query</span></code> attributes (as well as under
<code class="docutils literal notranslate"><span class="pre">alignment.sequences</span></code>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span>
<span class="go">SeqRecord(seq=Seq(None, length=287), id=&#39;RABGSTB&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&lt;unknown description&gt;&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span>
<span class="go">SeqRecord(seq=Seq(None, length=657), id=&#39;pGT875&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&lt;unknown description&gt;&#39;, dbxrefs=[])</span>
</pre></div>
</div>
<p>Information of the alignment is stored under the <code class="docutils literal notranslate"><span class="pre">annotations</span></code>
attribute of the <code class="docutils literal notranslate"><span class="pre">alignment</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span>  
<span class="go">{&#39;% identity&#39;: 66.93,</span>
<span class="go"> &#39;mismatches&#39;: 42,</span>
<span class="go"> &#39;gap opens&#39;: 8,</span>
<span class="go"> &#39;evalue&#39;: 3.2e-07,</span>
<span class="go"> &#39;bit score&#39;: 45.0}</span>
</pre></div>
</div>
<p>BLAST in particular offers many options to customize tabular output by
including or excluding specific columns; see the BLAST documentation for
details. This information is stored in the dictionaries
<code class="docutils literal notranslate"><span class="pre">alignment.annotations</span></code>, <code class="docutils literal notranslate"><span class="pre">alignment.target.annotations</span></code>, or
<code class="docutils literal notranslate"><span class="pre">alignment.query.annotations</span></code>, as appropriate.</p>
</section>
<section id="hh-suite-output-files">
<span id="subsec-align-hhr"></span><h3>HH-suite output files<a class="headerlink" href="#hh-suite-output-files" title="Link to this heading"></a></h3>
<p>Alignment files in the <code class="docutils literal notranslate"><span class="pre">hhr</span></code> format are generated by <code class="docutils literal notranslate"><span class="pre">hhsearch</span></code> or
<code class="docutils literal notranslate"><span class="pre">hhblits</span></code> in HH-suite <a class="reference internal" href="chapter_bibliography.html#steinegger2019" id="id6"><span>[Steinegger2019]</span></a>. As an
example, see the file <code class="docutils literal notranslate"><span class="pre">2uvo_hhblits.hhr</span></code> in Biopython’s test suite:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>Query         2UVO:A|PDBID|CHAIN|SEQUENCE
Match_columns 171
No_of_seqs    1560 out of 4005
Neff          8.3
Searched_HMMs 34
Date          Fri Feb 15 16:34:13 2019
Command       hhblits -i 2uvoAh.fasta -d /pdb70

 No Hit                             Prob E-value P-value  Score    SS Cols Query HMM  Template HMM
  1 2uvo_A Agglutinin isolectin 1; 100.0 3.7E-34 4.8E-38  210.3   0.0  171    1-171     1-171 (171)
  2 2wga   ; lectin (agglutinin);   99.9 1.1E-30 1.4E-34  190.4   0.0  162    2-169     2-163 (164)
  3 1ulk_A Lectin-C; chitin-bindin  99.8 5.2E-24 6.6E-28  148.2   0.0  120    1-124     2-121 (126)
...
 31 4z8i_A BBTPGRP3, peptidoglycan  79.6    0.12 1.5E-05   36.1   0.0   37    1-37      9-54  (236)
 32 1wga   ; lectin (agglutinin);   40.4     2.6 0.00029   25.9   0.0  106   54-163    11-116 (164)

No 1
&gt;2uvo_A Agglutinin isolectin 1; carbohydrate-binding protein, hevein domain, chitin-binding, GERM agglutinin, chitin-binding protein; HET: NDG NAG GOL; 1.40A {Triticum aestivum} PDB: 1wgc_A* 2cwg_A* 2x3t_A* 4aml_A* 7wga_A 9wga_A 2wgc_A 1wgt_A 1k7t_A* 1k7v_A* 1k7u_A 2x52_A* 1t0w_A*
Probab=99.95  E-value=3.7e-34  Score=210.31  Aligned_cols=171  Identities=100%  Similarity=2.050  Sum_probs=166.9

Q 2UVO:A|PDBID|C    1 ERCGEQGSNMECPNNLCCSQYGYCGMGGDYCGKGCQNGACWTSKRCGSQAGGATCTNNQCCSQYGYCGFGAEYCGAGCQG   80 (171)
Q Consensus         1 ~~cg~~~~~~~c~~~~CCs~~g~CG~~~~~c~~~c~~~~c~~~~~Cg~~~~~~~c~~~~CCs~~g~CG~~~~~c~~~c~~   80 (171)
                      ||||++.++..||++.|||+|+|||.+.+||+++||.+.|++..+|+++++.++|....|||.++||+.+.+||+.+||.
T Consensus         1 ~~cg~~~~~~~c~~~~CCS~~g~Cg~~~~~Cg~gC~~~~c~~~~~cg~~~~~~~c~~~~CCs~~g~Cg~~~~~c~~~c~~   80 (171)
T 2uvo_A            1 ERCGEQGSNMECPNNLCCSQYGYCGMGGDYCGKGCQNGACWTSKRCGSQAGGATCTNNQCCSQYGYCGFGAEYCGAGCQG   80 (171)
T ss_dssp             CBCBGGGTTBBCGGGCEECTTSBEEBSHHHHSTTCCBSSCSSCCBCBGGGTTBCCSTTCEECTTSBEEBSHHHHSTTCCB
T ss_pred             CCCCCCCCCcCCCCCCeeCCCCeECCCcccccCCccccccccccccCcccCCcccCCccccCCCceeCCCccccCCCccc
Confidence            79999999999999999999999999999999999999999999999999999999999999999999999999999999


Q 2UVO:A|PDBID|C   81 GPCRADIKCGSQAGGKLCPNNLCCSQWGFCGLGSEFCGGGCQSGACSTDKPCGKDAGGRVCTNNYCCSKWGSCGIGPGYC  160 (171)
Q Consensus        81 ~~~~~~~~Cg~~~~~~~c~~~~CCS~~G~CG~~~~~C~~~Cq~~~c~~~~~Cg~~~~~~~c~~~~CCS~~G~CG~~~~~C  160 (171)
                      +++++|+.|+...+++.||++.|||.|||||...+||+.+||+++|++|.+|++.+++++|..+.|||+++-||+...||
T Consensus        81 ~~~~~~~~cg~~~~~~~c~~~~CCs~~g~CG~~~~~C~~gCq~~~c~~~~~cg~~~~~~~c~~~~ccs~~g~Cg~~~~~C  160 (171)
T 2uvo_A           81 GPCRADIKCGSQAGGKLCPNNLCCSQWGFCGLGSEFCGGGCQSGACSTDKPCGKDAGGRVCTNNYCCSKWGSCGIGPGYC  160 (171)
T ss_dssp             SSCSSCCBCBGGGTTBCCGGGCEECTTSBEEBSHHHHSTTCCBSSCSSCCCCBTTTTTBCCSTTCEECTTSCEEBSHHHH
T ss_pred             ccccccccccccccCCCCCCCcccCCCCccCCCcccccCCCcCCccccccccccccccccCCCCCCcCCCCEecCchhhc
Confidence            99999999999988999999999999999999999999999999999999999999999999999999999999999999


Q 2UVO:A|PDBID|C  161 GAGCQSGGCDG  171 (171)
Q Consensus       161 ~~gCq~~~c~~  171 (171)
                      +++||++.|||
T Consensus       161 ~~~cq~~~~~~  171 (171)
T 2uvo_A          161 GAGCQSGGCDG  171 (171)
T ss_dssp             STTCCBSSCC-
T ss_pred             ccccccCCCCC
Confidence            99999999986


No 2
...


No 32
&gt;1wga   ; lectin (agglutinin); NMR {}
Probab=40.43  E-value=2.6  Score=25.90  Aligned_cols=106  Identities=20%  Similarity=0.652  Sum_probs=54.7

Q 2UVO:A|PDBID|C   54 TCTNNQCCSQYGYCGFGAEYCGAGCQGGPCRADIKCGSQAGGKLCPNNLCCSQWGFCGLGSEFCGGGCQSGACSTDKPCG  133 (171)
Q Consensus        54 ~c~~~~CCs~~g~CG~~~~~c~~~c~~~~~~~~~~Cg~~~~~~~c~~~~CCS~~G~CG~~~~~C~~~Cq~~~c~~~~~Cg  133 (171)
                      .|....||.....|......|...|....|.....|...  ...|....||.....|......|...|....+.....|.
T Consensus        11 ~c~~~~cc~~~~~c~~~~~~c~~~c~~~~c~~~~~c~~~--~~~c~~~~cc~~~~~c~~~~~~c~~~c~~~~c~~~~~c~   88 (164)
T 1wga             11 XCXXXXCCXXXXXCXXXXXXCXXXCXXXXCXXXXXCXXX--XXXCXXXXCCXXXXXCXXXXXXCXXXCXXXXCXXXXXCX   88 (164)
T ss_pred             ccccccccccccccccccccccccccccccccccccccc--ccccccccccccccccccccccccccccccccccccccc
Confidence            344556666666666666566555543333223333321  234666677777777777766666655544332223333


Q 2UVO:A|PDBID|C  134 KDAGGRVCTNNYCCSKWGSCGIGPGYCGAG  163 (171)
Q Consensus       134 ~~~~~~~c~~~~CCS~~G~CG~~~~~C~~g  163 (171)
                      ..  ...|....||.....|......|...
T Consensus        89 ~~--~~~c~~~~cc~~~~~c~~~~~~c~~~  116 (164)
T 1wga             89 XX--XXXCXXXXCCXXXXXCXXXXXXCXXX  116 (164)
T ss_pred             cc--cccccccccccccccccccccccccc
Confidence            22  23344455555555555555544433


Done!
</pre></div>
</div>
<p>The file contains three sections:</p>
<ul class="simple">
<li><p>A header with general information about the alignments;</p></li>
<li><p>A summary with one line for each of the alignments obtained;</p></li>
<li><p>The alignments shown consecutively in detail.</p></li>
</ul>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;2uvo_hhblits.hhr&quot;</span><span class="p">,</span> <span class="s2">&quot;hhr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Most of the header information is stored in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> attribute
of <code class="docutils literal notranslate"><span class="pre">alignments</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>  
<span class="go">{&#39;Match_columns&#39;: 171,</span>
<span class="go"> &#39;No_of_seqs&#39;: (1560, 4005),</span>
<span class="go"> &#39;Neff&#39;: 8.3,</span>
<span class="go"> &#39;Searched_HMMs&#39;: 34,</span>
<span class="go"> &#39;Rundate&#39;: &#39;Fri Feb 15 16:34:13 2019&#39;,</span>
<span class="go"> &#39;Command line&#39;: &#39;hhblits -i 2uvoAh.fasta -d /pdb70&#39;}</span>
</pre></div>
</div>
<p>except the query name, which is stored as an attribute:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">query_name</span>
<span class="go">&#39;2UVO:A|PDBID|CHAIN|SEQUENCE&#39;</span>
</pre></div>
</div>
<p>as it will reappear in each of the alignments.</p>
<p>Iterate over the alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>  
<span class="gp">...</span>
<span class="go">2uvo_A</span>
<span class="go">2wga</span>
<span class="go">1ulk_A</span>
<span class="go">...</span>
<span class="go">4z8i_A</span>
<span class="go">1wga</span>
</pre></div>
</div>
<p>Let’s look at the first alignment in more detail:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (2 rows x 171 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">2uvo_A            0 ERCGEQGSNMECPNNLCCSQYGYCGMGGDYCGKGCQNGACWTSKRCGSQAGGATCTNNQC</span>
<span class="go">                  0 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="go">2UVO:A|PD         0 ERCGEQGSNMECPNNLCCSQYGYCGMGGDYCGKGCQNGACWTSKRCGSQAGGATCTNNQC</span>

<span class="go">2uvo_A           60 CSQYGYCGFGAEYCGAGCQGGPCRADIKCGSQAGGKLCPNNLCCSQWGFCGLGSEFCGGG</span>
<span class="go">                 60 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="go">2UVO:A|PD        60 CSQYGYCGFGAEYCGAGCQGGPCRADIKCGSQAGGKLCPNNLCCSQWGFCGLGSEFCGGG</span>

<span class="go">2uvo_A          120 CQSGACSTDKPCGKDAGGRVCTNNYCCSKWGSCGIGPGYCGAGCQSGGCDG 171</span>
<span class="go">                120 ||||||||||||||||||||||||||||||||||||||||||||||||||| 171</span>
<span class="go">2UVO:A|PD       120 CQSGACSTDKPCGKDAGGRVCTNNYCCSKWGSCGIGPGYCGAGCQSGGCDG 171</span>
</pre></div>
</div>
<p>The target and query sequences are stored in <code class="docutils literal notranslate"><span class="pre">alignment.sequences</span></code>. As
these are pairwise alignments, we can also access them through
<code class="docutils literal notranslate"><span class="pre">alignment.target</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment.query</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span> <span class="ow">is</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span> <span class="ow">is</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The ID of the query is set from the <code class="docutils literal notranslate"><span class="pre">alignments.query_name</span></code> (note that
the query ID printed in the alignment in the <code class="docutils literal notranslate"><span class="pre">hhr</span></code> file is
abbreviated):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;2UVO:A|PDBID|CHAIN|SEQUENCE&#39;</span>
</pre></div>
</div>
<p>The ID of the target is taken from the sequence alignment block (the
line starting with <code class="docutils literal notranslate"><span class="pre">T</span> <span class="pre">2uvo_A</span></code>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;2uvo_A&#39;</span>
</pre></div>
</div>
<p>The sequence contents of the target and query are filled in from the
information available in this alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;ERCGEQGSNMECPNNLCCSQYGYCGMGGDYCGKGCQNGACWTSKRCGSQAGGAT...CDG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;ERCGEQGSNMECPNNLCCSQYGYCGMGGDYCGKGCQNGACWTSKRCGSQAGGAT...CDG&#39;)</span>
</pre></div>
</div>
<p>The sequence contents will be incomplete (a partially defined sequence;
see Section <a class="reference internal" href="chapter_seq_objects.html#sec-partial-seq"><span class="std std-ref">Sequences with partially defined sequence contents</span></a>) if the alignment
does not extend over the full sequence.</p>
<p>The second line of this alignment block, starting with “<code class="docutils literal notranslate"><span class="pre">&gt;</span></code>”, shows
the name and description of the Hidden Markov Model from which the
target sequence was taken. These are stored under the keys
<code class="docutils literal notranslate"><span class="pre">&quot;hmm_name&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;hmm_description&quot;</span></code> in the
<code class="docutils literal notranslate"><span class="pre">alignment.target.annotations</span></code> dictionary:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">annotations</span>  
<span class="go">{&#39;hmm_name&#39;: &#39;2uvo_A&#39;,</span>
<span class="go"> &#39;hmm_description&#39;: &#39;Agglutinin isolectin 1; carbohydrate-binding protein, hevein domain, chitin-binding, GERM agglutinin, chitin-binding protein; HET: NDG NAG GOL; 1.40A {Triticum aestivum} PDB: 1wgc_A* 2cwg_A* 2x3t_A* 4aml_A* 7wga_A 9wga_A 2wgc_A 1wgt_A 1k7t_A* 1k7v_A* 1k7u_A 2x52_A* 1t0w_A*&#39;}</span>
</pre></div>
</div>
<p>The dictionary <code class="docutils literal notranslate"><span class="pre">alignment.target.letter_annotations</span></code> stores the target
alignent consensus sequence, the secondary structure as predicted by
PSIPRED, and the target secondary structure as determined by DSSP:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">letter_annotations</span>  
<span class="go">{&#39;Consensus&#39;: &#39;~~cg~~~~~~~c~~~~CCS~~g~Cg~~~~~Cg~gC~~~~c~~~~~cg~~~~~~~c~~~~CCs~~g~Cg~~~~~c~~~c~~~~~~~~~~cg~~~~~~~c~~~~CCs~~g~CG~~~~~C~~gCq~~~c~~~~~cg~~~~~~~c~~~~ccs~~g~Cg~~~~~C~~~cq~~~~~~&#39;,</span>
<span class="go"> &#39;ss_pred&#39;: &#39;CCCCCCCCCcCCCCCCeeCCCCeECCCcccccCCccccccccccccCcccCCcccCCccccCCCceeCCCccccCCCcccccccccccccccccCCCCCCCcccCCCCccCCCcccccCCCcCCccccccccccccccccCCCCCCcCCCCEecCchhhcccccccCCCCC&#39;,</span>
<span class="go"> &#39;ss_dssp&#39;: &#39;CBCBGGGTTBBCGGGCEECTTSBEEBSHHHHSTTCCBSSCSSCCBCBGGGTTBCCSTTCEECTTSBEEBSHHHHSTTCCBSSCSSCCBCBGGGTTBCCGGGCEECTTSBEEBSHHHHSTTCCBSSCSSCCCCBTTTTTBCCSTTCEECTTSCEEBSHHHHSTTCCBSSCC &#39;}</span>
</pre></div>
</div>
<p>In this example, for the query sequence only the consensus sequence is
available:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">letter_annotations</span>
<span class="go">{&#39;Consensus&#39;: &#39;~~cg~~~~~~~c~~~~CCs~~g~CG~~~~~c~~~c~~~~c~~~~~Cg~~~~~~~c~~~~CCs~~g~CG~~~~~c~~~c~~~~~~~~~~Cg~~~~~~~c~~~~CCS~~G~CG~~~~~C~~~Cq~~~c~~~~~Cg~~~~~~~c~~~~CCS~~G~CG~~~~~C~~gCq~~~c~~&#39;}</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">alignment.annotations</span></code> dictionary stores information about the
alignment shown on the third line of the alignment block:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span>  
<span class="go">{&#39;Probab&#39;: 99.95,</span>
<span class="go"> &#39;E-value&#39;: 3.7e-34,</span>
<span class="go"> &#39;Score&#39;: 210.31,</span>
<span class="go"> &#39;Identities&#39;: 100.0,</span>
<span class="go"> &#39;Similarity&#39;: 2.05,</span>
<span class="go"> &#39;Sum_probs&#39;: 166.9}</span>
</pre></div>
</div>
<p>Confidence values for the pairwise alignment are stored under the
<code class="docutils literal notranslate"><span class="pre">&quot;Confidence&quot;</span></code> key in the <code class="docutils literal notranslate"><span class="pre">alignment.column_annotations</span></code> dictionary.
This dictionary also stores the score for each column, shown between the
query and the target section of each alignment block:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">column_annotations</span>  
<span class="go">{&#39;column score&#39;: &#39;||||++.++..||++.|||+|+|||.+.+||+++||.+.|++..+|+++++.++|....|||.++||+.+.+||+.+||.+++++|+.|+...+++.||++.|||.|||||...+||+.+||+++|++|.+|++.+++++|..+.|||+++-||+...||+++||++.|||&#39;,</span>
<span class="go"> &#39;Confidence&#39;: &#39;799999999999999999999999999999999999999999999999999999999999999999999999999999999999999999998899999999999999999999999999999999999999999999999999999999999999999999999999986&#39;}</span>
</pre></div>
</div>
</section>
<section id="a2m">
<span id="subsec-align-a2m"></span><h3>A2M<a class="headerlink" href="#a2m" title="Link to this heading"></a></h3>
<p>A2M files are alignment files created by <code class="docutils literal notranslate"><span class="pre">align2model</span></code> or <code class="docutils literal notranslate"><span class="pre">hmmscore</span></code>
in the SAM Sequence Alignment and Modeling Software
System <a class="reference internal" href="chapter_bibliography.html#krogh1994" id="id7"><span>[Krogh1994]</span></a>, <a class="reference internal" href="chapter_bibliography.html#hughey1996" id="id8"><span>[Hughey1996]</span></a>. An A2M file contains
one multiple alignment. The A2M file format is similar to aligned FASTA
(see section <a class="reference internal" href="#subsec-align-fasta"><span class="std std-ref">Aligned FASTA</span></a>). However, to distinguish
insertions from deletions, A2M uses both dashes and periods to represent
gaps, and both upper and lower case characters in the aligned sequences.
Matches are represented by upper case letters and deletions by dashes in
alignment columns containing matches or deletions only. Insertions are
represented by lower case letters, with gaps aligned to the insertion
shown as periods. Header lines start with “<code class="docutils literal notranslate"><span class="pre">&gt;</span></code>” followed by the name
of the sequence, and optionally a description.</p>
<p>The file <code class="docutils literal notranslate"><span class="pre">probcons.a2m</span></code> in Biopython’s test suite is an example of an
A2M file (see section <a class="reference internal" href="#subsec-align-fasta"><span class="std std-ref">Aligned FASTA</span></a> for the same
alignment in aligned FASTA format):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;plas_horvu
D.VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG.VD.VSKISQEEYLTAPGETFSVTLTV...PGTYGFYCEPHAGAGMVGKVT
V
&gt;plas_chlre
-.VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG.VN.ADAISRDDYLNAPGETYSVKLTA...AGEYGYYCEPHQGAGMVGKII
V
&gt;plas_anava
-.VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKsADlAKSLSHKQLLMSPGQSTSTTFPAdapAGEYTFYCEPHRGAGMVGKIT
V
&gt;plas_proho
VqIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG.ES.APALSNTKLRIAPGSFYSVTLGT...PGTYSFYCTPHRGAGMVGTIT
V
&gt;azup_achcy
VhMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG.AE.A-------FKSKINENYKVTFTA...PGVYGVKCTPHYGMGMVGVVE
V
</pre></div>
</div>
<p>To parse this alignment, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;probcons.a2m&quot;</span><span class="p">,</span> <span class="s2">&quot;a2m&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (5 rows x 101 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">plas_horv         0 D-VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG-VD-VSKISQE</span>
<span class="go">plas_chlr         0 --VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG-VN-ADAISRD</span>
<span class="go">plas_anav         0 --VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKSADLAKSLSHK</span>
<span class="go">plas_proh         0 VQIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG-ES-APALSNT</span>
<span class="go">azup_achc         0 VHMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG-AE-A------</span>

<span class="go">plas_horv        57 EYLTAPGETFSVTLTV---PGTYGFYCEPHAGAGMVGKVTV 95</span>
<span class="go">plas_chlr        56 DYLNAPGETYSVKLTA---AGEYGYYCEPHQGAGMVGKIIV 94</span>
<span class="go">plas_anav        58 QLLMSPGQSTSTTFPADAPAGEYTFYCEPHRGAGMVGKITV 99</span>
<span class="go">plas_proh        56 KLRIAPGSFYSVTLGT---PGTYSFYCTPHRGAGMVGTITV 94</span>
<span class="go">azup_achc        51 -FKSKINENYKVTFTA---PGVYGVKCTPHYGMGMVGVVEV 88</span>
</pre></div>
</div>
<p>The parser analyzes the pattern of dashes, periods, and lower and upper
case letters in the A2M file to determine if a column is an
match/mismatch/deletion (”<code class="docutils literal notranslate"><span class="pre">D</span></code>”) or an insertion (”<code class="docutils literal notranslate"><span class="pre">I</span></code>”). This
information is stored under the <code class="docutils literal notranslate"><span class="pre">match</span></code> key of the
<code class="docutils literal notranslate"><span class="pre">alignment.column_annotations</span></code> dictionary:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">column_annotations</span>
<span class="go">{&#39;state&#39;: &#39;DIDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDIDDIDDDDDDDDDDDDDDDDDDDDDDDIIIDDDDDDDDDDDDDDDDDDDDDD&#39;}</span>
</pre></div>
</div>
<p>As the state information is stored in the <code class="docutils literal notranslate"><span class="pre">alignment</span></code>, we can print
the alignment in the A2M format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;a2m&quot;</span><span class="p">))</span>
<span class="go">&gt;plas_horvu</span>
<span class="go">D.VLLGANGGVLVFEPNDFSVKAGETITFKNNAGYPHNVVFDEDAVPSG.VD.VSKISQEEYLTAPGETFSVTLTV...PGTYGFYCEPHAGAGMVGKVTV</span>
<span class="go">&gt;plas_chlre</span>
<span class="go">-.VKLGADSGALEFVPKTLTIKSGETVNFVNNAGFPHNIVFDEDAIPSG.VN.ADAISRDDYLNAPGETYSVKLTA...AGEYGYYCEPHQGAGMVGKIIV</span>
<span class="go">&gt;plas_anava</span>
<span class="go">-.VKLGSDKGLLVFEPAKLTIKPGDTVEFLNNKVPPHNVVFDAALNPAKsADlAKSLSHKQLLMSPGQSTSTTFPAdapAGEYTFYCEPHRGAGMVGKITV</span>
<span class="go">&gt;plas_proho</span>
<span class="go">VqIKMGTDKYAPLYEPKALSISAGDTVEFVMNKVGPHNVIFDK--VPAG.ES.APALSNTKLRIAPGSFYSVTLGT...PGTYSFYCTPHRGAGMVGTITV</span>
<span class="go">&gt;azup_achcy</span>
<span class="go">VhMLNKGKDGAMVFEPASLKVAPGDTVTFIPTDK-GHNVETIKGMIPDG.AE.A-------FKSKINENYKVTFTA...PGVYGVKCTPHYGMGMVGVVEV</span>
</pre></div>
</div>
<p>Similarly, the alignment can be written in the A2M format to an output
file using <code class="docutils literal notranslate"><span class="pre">Align.write</span></code> (see
section <a class="reference internal" href="#subsec-align-writing"><span class="std std-ref">Writing alignments</span></a>).</p>
</section>
<section id="mauve-extended-multi-fasta-xmfa-format">
<span id="subsec-align-mauve"></span><h3>Mauve eXtended Multi-FastA (xmfa) format<a class="headerlink" href="#mauve-extended-multi-fasta-xmfa-format" title="Link to this heading"></a></h3>
<p>Mauve <a class="reference internal" href="chapter_bibliography.html#darling2004" id="id9"><span>[Darling2004]</span></a> is a software package for
constructing multiple genome alignments. These alignments are stored in
the eXtended Multi-FastA (xmfa) format. Depending on how exactly
<code class="docutils literal notranslate"><span class="pre">progressiveMauve</span></code> (the aligner program in Mauve) was called, the xmfa
format is slightly different.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">progressiveMauve</span></code> is called with a single sequence input file, as
in</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>progressiveMauve combined.fasta  --output=combined.xmfa ...
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">combined.fasta</span></code> contains the genome sequences:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&gt;equCab1
GAAAAGGAAAGTACGGCCCGGCCACTCCGGGTGTGTGCTAGGAGGGCTTA
&gt;mm9
GAAGAGGAAAAGTAGATCCCTGGCGTCCGGAGCTGGGACGT
&gt;canFam2
CAAGCCCTGCGCGCTCAGCCGGAGTGTCCCGGGCCCTGCTTTCCTTTTC
</pre></div>
</div>
<p>then the output file <code class="docutils literal notranslate"><span class="pre">combined.xmfa</span></code> is as follows:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>#FormatVersion Mauve1
#Sequence1File  combined.fa
#Sequence1Entry 1
#Sequence1Format    FastA
#Sequence2File  combined.fa
#Sequence2Entry 2
#Sequence2Format    FastA
#Sequence3File  combined.fa
#Sequence3Entry 3
#Sequence3Format    FastA
#BackboneFile   combined.xmfa.bbcols
&gt; 1:2-49 - combined.fa
AAGCCCTCCTAGCACACACCCGGAGTGG-CCGGGCCGTACTTTCCTTTT
&gt; 2:0-0 + combined.fa
-------------------------------------------------
&gt; 3:2-48 + combined.fa
AAGCCCTGC--GCGCTCAGCCGGAGTGTCCCGGGCCCTGCTTTCCTTTT
=
&gt; 1:1-1 + combined.fa
G
=
&gt; 1:50-50 + combined.fa
A
=
&gt; 2:1-41 + combined.fa
GAAGAGGAAAAGTAGATCCCTGGCGTCCGGAGCTGGGACGT
=
&gt; 3:1-1 + combined.fa
C
=
&gt; 3:49-49 + combined.fa
C
=
</pre></div>
</div>
<p>with numbers (1, 2, 3) referring to the input genome sequences for horse
(<code class="docutils literal notranslate"><span class="pre">equCab1</span></code>), mouse (<code class="docutils literal notranslate"><span class="pre">mm9</span></code>), and dog (<code class="docutils literal notranslate"><span class="pre">canFam2</span></code>), respectively.
This xmfa file consists of six alignment blocks, separated by <code class="docutils literal notranslate"><span class="pre">=</span></code>
characters. Use <code class="docutils literal notranslate"><span class="pre">Align.parse</span></code> to extract these alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;combined.xmfa&quot;</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The file header data are stored in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> attribute:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>  
<span class="go">{&#39;FormatVersion&#39;: &#39;Mauve1&#39;,</span>
<span class="go"> &#39;BackboneFile&#39;: &#39;combined.xmfa.bbcols&#39;,</span>
<span class="go"> &#39;File&#39;: &#39;combined.fa&#39;}</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">identifiers</span></code> attribute stores the sequence identifiers for the
three sequences, which in this case is the three numbers:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">identifiers</span>
<span class="go">[&#39;0&#39;, &#39;1&#39;, &#39;2&#39;]</span>
</pre></div>
</div>
<p>These identifiers are used in the individual alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">([</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">])</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;******&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">[&#39;0&#39;, &#39;1&#39;, &#39;2&#39;]</span>
<span class="go">0                49 AAGCCCTCCTAGCACACACCCGGAGTGG-CCGGGCCGTACTTTCCTTTT  1</span>
<span class="go">1                 0 -------------------------------------------------  0</span>
<span class="go">2                 1 AAGCCCTGC--GCGCTCAGCCGGAGTGTCCCGGGCCCTGCTTTCCTTTT 48</span>

<span class="go">******</span>
<span class="go">[&#39;0&#39;]</span>
<span class="go">0                 0 G 1</span>

<span class="go">******</span>
<span class="go">[&#39;0&#39;]</span>
<span class="go">0                49 A 50</span>

<span class="go">******</span>
<span class="go">[&#39;1&#39;]</span>
<span class="go">1                 0 GAAGAGGAAAAGTAGATCCCTGGCGTCCGGAGCTGGGACGT 41</span>

<span class="go">******</span>
<span class="go">[&#39;2&#39;]</span>
<span class="go">2                 0 C 1</span>

<span class="go">******</span>
<span class="go">[&#39;2&#39;]</span>
<span class="go">2                48 C 49</span>

<span class="go">******</span>
</pre></div>
</div>
<p>Note that only the first block is a real alignment; the other blocks
contain only a single sequence. By including these blocks, the xmfa file
contains the full sequence that was provided in the <code class="docutils literal notranslate"><span class="pre">combined.fa</span></code>
input file.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">progressiveMauve</span></code> is called with a separate input file for each
genome, as in</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>progressiveMauve equCab1.fa canFam2.fa mm9.fa --output=separate.xmfa ...
</pre></div>
</div>
<p>where each Fasta file contains the genome sequence for one species only,
then the output file <code class="docutils literal notranslate"><span class="pre">separate.xmfa</span></code> is as follows:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>#FormatVersion Mauve1
#Sequence1File  equCab1.fa
#Sequence1Format    FastA
#Sequence2File  canFam2.fa
#Sequence2Format    FastA
#Sequence3File  mm9.fa
#Sequence3Format    FastA
#BackboneFile   separate.xmfa.bbcols
&gt; 1:1-50 - equCab1.fa
TAAGCCCTCCTAGCACACACCCGGAGTGGCC-GGGCCGTAC-TTTCCTTTTC
&gt; 2:1-49 + canFam2.fa
CAAGCCCTGC--GCGCTCAGCCGGAGTGTCCCGGGCCCTGC-TTTCCTTTTC
&gt; 3:1-19 - mm9.fa
---------------------------------GGATCTACTTTTCCTCTTC
=
&gt; 3:20-41 + mm9.fa
CTGGCGTCCGGAGCTGGGACGT
=
</pre></div>
</div>
<p>The identifiers <code class="docutils literal notranslate"><span class="pre">equCab1</span></code> for horse, <code class="docutils literal notranslate"><span class="pre">mm9</span></code> for mouse, and
<code class="docutils literal notranslate"><span class="pre">canFam2</span></code> for dog are now shown explicitly in the output file. This
xmfa file consists of two alignment blocks, separated by <code class="docutils literal notranslate"><span class="pre">=</span></code>
characters. Use <code class="docutils literal notranslate"><span class="pre">Align.parse</span></code> to extract these alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;separate.xmfa&quot;</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The file header data now does not include the input file name:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>  
<span class="go">{&#39;FormatVersion&#39;: &#39;Mauve1&#39;,</span>
<span class="go"> &#39;BackboneFile&#39;: &#39;separate.xmfa.bbcols&#39;}</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">identifiers</span></code> attribute stores the sequence identifiers for the
three sequences:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">identifiers</span>
<span class="go">[&#39;equCab1.fa&#39;, &#39;canFam2.fa&#39;, &#39;mm9.fa&#39;]</span>
</pre></div>
</div>
<p>These identifiers are used in the individual alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">([</span><span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">])</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;******&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">[&#39;equCab1.fa&#39;, &#39;canFam2.fa&#39;, &#39;mm9.fa&#39;]</span>
<span class="go">equCab1.f        50 TAAGCCCTCCTAGCACACACCCGGAGTGGCC-GGGCCGTAC-TTTCCTTTTC  0</span>
<span class="go">canFam2.f         0 CAAGCCCTGC--GCGCTCAGCCGGAGTGTCCCGGGCCCTGC-TTTCCTTTTC 49</span>
<span class="go">mm9.fa           19 ---------------------------------GGATCTACTTTTCCTCTTC  0</span>

<span class="go">******</span>
<span class="go">[&#39;mm9.fa&#39;]</span>
<span class="go">mm9.fa           19 CTGGCGTCCGGAGCTGGGACGT 41</span>

<span class="go">******</span>
</pre></div>
</div>
<p>To output the alignments in Mauve format, use <code class="docutils literal notranslate"><span class="pre">Align.write</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;separate.xmfa&quot;</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">stream</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>  
<span class="go">#FormatVersion Mauve1</span>
<span class="go">#Sequence1File  equCab1.fa</span>
<span class="go">#Sequence1Format    FastA</span>
<span class="go">#Sequence2File  canFam2.fa</span>
<span class="go">#Sequence2Format    FastA</span>
<span class="go">#Sequence3File  mm9.fa</span>
<span class="go">#Sequence3Format    FastA</span>
<span class="go">#BackboneFile   separate.xmfa.bbcols</span>
<span class="go">&gt; 1:1-50 - equCab1.fa</span>
<span class="go">TAAGCCCTCCTAGCACACACCCGGAGTGGCC-GGGCCGTAC-TTTCCTTTTC</span>
<span class="go">&gt; 2:1-49 + canFam2.fa</span>
<span class="go">CAAGCCCTGC--GCGCTCAGCCGGAGTGTCCCGGGCCCTGC-TTTCCTTTTC</span>
<span class="go">&gt; 3:1-19 - mm9.fa</span>
<span class="go">---------------------------------GGATCTACTTTTCCTCTTC</span>
<span class="go">=</span>
<span class="go">&gt; 3:20-41 + mm9.fa</span>
<span class="go">CTGGCGTCCGGAGCTGGGACGT</span>
<span class="go">=</span>
</pre></div>
</div>
<p>Here, the writer makes use of the information stored in
<code class="docutils literal notranslate"><span class="pre">alignments.metadata</span></code> and <code class="docutils literal notranslate"><span class="pre">alignments.identifiers</span></code> to create this
format. If your <code class="docutils literal notranslate"><span class="pre">alignments</span></code> object does not have these attributes,
you can provide them as keyword arguments to <code class="docutils literal notranslate"><span class="pre">Align.write</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;separate.xmfa&quot;</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">metadata</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">identifiers</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">identifiers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>  <span class="c1"># this drops the attributes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>  
<span class="gt">Traceback (most recent call last):</span>
<span class="w"> </span><span class="c">...</span>
<span class="gr">AttributeError</span>: <span class="n">&#39;list&#39; object has no attribute &#39;metadata&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">identifiers</span>  
<span class="gt">Traceback (most recent call last):</span>
<span class="w"> </span><span class="c">...</span>
<span class="gr">AttributeError</span>: <span class="n">&#39;list&#39; object has no attribute &#39;identifiers&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span> <span class="n">identifiers</span><span class="o">=</span><span class="n">identifiers</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">stream</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>  
<span class="go">#FormatVersion Mauve1</span>
<span class="go">#Sequence1File  equCab1.fa</span>
<span class="go">#Sequence1Format    FastA</span>
<span class="go">#Sequence2File  canFam2.fa</span>
<span class="go">#Sequence2Format    FastA</span>
<span class="go">#Sequence3File  mm9.fa</span>
<span class="go">#Sequence3Format    FastA</span>
<span class="go">#BackboneFile   separate.xmfa.bbcols</span>
<span class="go">&gt; 1:1-50 - equCab1.fa</span>
<span class="go">TAAGCCCTCCTAGCACACACCCGGAGTGGCC-GGGCCGTAC-TTTCCTTTTC</span>
<span class="go">&gt; 2:1-49 + canFam2.fa</span>
<span class="go">CAAGCCCTGC--GCGCTCAGCCGGAGTGTCCCGGGCCCTGC-TTTCCTTTTC</span>
<span class="go">&gt; 3:1-19 - mm9.fa</span>
<span class="go">---------------------------------GGATCTACTTTTCCTCTTC</span>
<span class="go">=</span>
<span class="go">&gt; 3:20-41 + mm9.fa</span>
<span class="go">CTGGCGTCCGGAGCTGGGACGT</span>
<span class="go">=</span>
</pre></div>
</div>
<p>Python does not allow you to add these attributes to the <code class="docutils literal notranslate"><span class="pre">alignments</span></code>
object directly, as in this example it was converted to a plain list.
However, you can construct an <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> object and add attributes
to it (see Section <a class="reference internal" href="#sec-alignments"><span class="std std-ref">The Alignments class</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">Alignments</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">identifiers</span> <span class="o">=</span> <span class="n">identifiers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span> <span class="n">identifiers</span><span class="o">=</span><span class="n">identifiers</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">stream</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>  
<span class="go">#FormatVersion Mauve1</span>
<span class="go">#Sequence1File  equCab1.fa</span>
<span class="go">#Sequence1Format    FastA</span>
<span class="go">#Sequence2File  canFam2.fa</span>
<span class="go">#Sequence2Format    FastA</span>
<span class="go">#Sequence3File  mm9.fa</span>
<span class="go">#Sequence3Format    FastA</span>
<span class="go">#BackboneFile   separate.xmfa.bbcols</span>
<span class="go">&gt; 1:1-50 - equCab1.fa</span>
<span class="go">TAAGCCCTCCTAGCACACACCCGGAGTGGCC-GGGCCGTAC-TTTCCTTTTC</span>
<span class="go">&gt; 2:1-49 + canFam2.fa</span>
<span class="go">CAAGCCCTGC--GCGCTCAGCCGGAGTGTCCCGGGCCCTGC-TTTCCTTTTC</span>
<span class="go">&gt; 3:1-19 - mm9.fa</span>
<span class="go">---------------------------------GGATCTACTTTTCCTCTTC</span>
<span class="go">=</span>
<span class="go">&gt; 3:20-41 + mm9.fa</span>
<span class="go">CTGGCGTCCGGAGCTGGGACGT</span>
<span class="go">=</span>
</pre></div>
</div>
<p>When printing a single alignment in <code class="docutils literal notranslate"><span class="pre">Mauve</span></code> format, use keyword
arguments to provide the metadata and identifiers:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;mauve&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span> <span class="n">identifiers</span><span class="o">=</span><span class="n">identifiers</span><span class="p">))</span>
<span class="go">&gt; 1:1-50 - equCab1.fa</span>
<span class="go">TAAGCCCTCCTAGCACACACCCGGAGTGGCC-GGGCCGTAC-TTTCCTTTTC</span>
<span class="go">&gt; 2:1-49 + canFam2.fa</span>
<span class="go">CAAGCCCTGC--GCGCTCAGCCGGAGTGTCCCGGGCCCTGC-TTTCCTTTTC</span>
<span class="go">&gt; 3:1-19 - mm9.fa</span>
<span class="go">---------------------------------GGATCTACTTTTCCTCTTC</span>
<span class="go">=</span>
</pre></div>
</div>
</section>
<section id="sequence-alignment-map-sam">
<span id="subsec-align-sam"></span><h3>Sequence Alignment/Map (SAM)<a class="headerlink" href="#sequence-alignment-map-sam" title="Link to this heading"></a></h3>
<p>Files in the Sequence Alignment/Map (SAM) format
<a class="reference internal" href="chapter_bibliography.html#li2009" id="id10"><span>[Li2009]</span></a> store pairwise sequence alignments, usually
of next-generation sequencing data against a reference genome. The file
<code class="docutils literal notranslate"><span class="pre">ex1.sam</span></code> in Biopython’s test suite is an example of a minimal file in
the SAM format. Its first few lines are as follows:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>EAS56_57:6:190:289:82   69      chr1    100     0       *       =       100     0       CTCAAGGTTGTTGCAAGGGGGTCTATGTGAACAAA     &lt;&lt;&lt;7&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;8;;&lt;7;4&lt;;&lt;;;;;;94&lt;;     MF:i:192
EAS56_57:6:190:289:82   137     chr1    100     73      35M     =       100     0       AGGGGTGCAGAGCCGAGTCACGGGGTTGCCAGCAC     &lt;&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;;&lt;&lt;;&lt;&lt;&lt;&lt;;8&lt;6;9;;2;     MF:i:64 Aq:i:0  NM:i:0  UQ:i:0  H0:i:1  H1:i:0
EAS51_64:3:190:727:308  99      chr1    103     99      35M     =       263     195     GGTGCAGAGCCGAGTCACGGGGTTGCCAGCACAGG     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;::&lt;&lt;&lt;844     MF:i:18 Aq:i:73 NM:i:0  UQ:i:0  H0:i:1  H1:i:0
...
</pre></div>
</div>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;ex1.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">flag</span></code> of the first line is 69. According to the SAM/BAM file
format specification, lines for which the flag contains the bitwise flag
4 are unmapped. As 69 has the bit corresponding to this position set to
True, this sequence is unmapped and was not aligned to the genome (in
spite of the first line showing <code class="docutils literal notranslate"><span class="pre">chr1</span></code>). The target of this alignment
(or the first item in <code class="docutils literal notranslate"><span class="pre">alignment.sequences</span></code>) is therefore <code class="docutils literal notranslate"><span class="pre">None</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">flag</span>
<span class="go">69</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">bin</span><span class="p">(</span><span class="mi">69</span><span class="p">)</span>
<span class="go">&#39;0b1000101&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">bin</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="go">&#39;0b100&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">if</span> <span class="n">alignment</span><span class="o">.</span><span class="n">flag</span> <span class="o">&amp;</span> <span class="mi">4</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;unmapped&quot;</span><span class="p">)</span>
<span class="gp">... </span><span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;mapped&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">unmapped</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span>
<span class="go">[None, SeqRecord(seq=Seq(&#39;CTCAAGGTTGTTGCAAGGGGGTCTATGTGAACAAA&#39;), id=&#39;EAS56_57:6:190:289:82&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[])]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span> <span class="ow">is</span> <span class="kc">None</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The second line represents an alignment to chromosome 1:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">if</span> <span class="n">alignment</span><span class="o">.</span><span class="n">flag</span> <span class="o">&amp;</span> <span class="mi">4</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;unmapped&quot;</span><span class="p">)</span>
<span class="gp">... </span><span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;mapped&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">mapped</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span>
<span class="go">SeqRecord(seq=None, id=&#39;chr1&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[])</span>
</pre></div>
</div>
<p>As this SAM file does not store the genome sequence information for each
alignment, we cannot print the alignment. However, we can print the
alignment information in SAM format or any other format (such as BED,
see section <a class="reference internal" href="#subsec-align-bed"><span class="std std-ref">Browser Extensible Data (BED)</span></a>) that does not require the
target sequence information:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">)</span>
<span class="go">&#39;EAS56_57:6:190:289:82\t137\tchr1\t100\t73\t35M\t=\t100\t0\tAGGGGTGCAGAGCCGAGTCACGGGGTTGCCAGCAC\t&lt;&lt;&lt;&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;;&lt;&lt;;&lt;&lt;&lt;&lt;;8&lt;6;9;;2;\tMF:i:64\tAq:i:0\tNM:i:0\tUQ:i:0\tH0:i:1\tH1:i:0\n&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">)</span>
<span class="go">&#39;chr1\t99\t134\tEAS56_57:6:190:289:82\t0\t+\t99\t134\t0\t1\t35,\t0,\n&#39;</span>
</pre></div>
</div>
<p>However, we cannot print the alignment in PSL format (see
section <a class="reference internal" href="#subsec-align-psl"><span class="std std-ref">Pattern Space Layout (PSL)</span></a>) as that would require knowing
the size of the target sequence chr1:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>  
<span class="gt">Traceback (most recent call last):</span>
<span class="w"> </span><span class="c">...</span>
<span class="gr">TypeError</span>: <span class="n">...</span>
</pre></div>
</div>
<p>If you know the size of the target sequences, you can set them by hand:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">1575</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;chr1&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">target</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>  
<span class="go">&#39;35\t0\t0\t0\t0\t0\t0\t0\t+\tEAS56_57:6:190:289:82\t35\t0\t35\tchr1\t1575\t99\t134\t1\t35,\t0,\t99,\n&#39;</span>
</pre></div>
</div>
<p>The file <code class="docutils literal notranslate"><span class="pre">ex1_header.sam</span></code> in Biopython’s test suite contains the same
alignments, but now also includes a header. Its first few lines are as
follows:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>@HD\tVN:1.3\tSO:coordinate
@SQ\tSN:chr1\tLN:1575
@SQ\tSN:chr2\tLN:1584
EAS56_57:6:190:289:82   69      chr1    100     0       *       =       100     0       CTCAAGGTTGTTGCAAGGGGGTCTATGTGAACAAA     &lt;&lt;&lt;7&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;8;;&lt;7;4&lt;;&lt;;;;;;94&lt;;     MF:i:192
...
</pre></div>
</div>
<p>The header stores general information about the alignments, including
the size of the target chromosomes. The target information is stored in
the <code class="docutils literal notranslate"><span class="pre">targets</span></code> attribute of the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;ex1_header.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">targets</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">targets</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">SeqRecord(seq=Seq(None, length=1575), id=&#39;chr1&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">targets</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">SeqRecord(seq=Seq(None, length=1584), id=&#39;chr2&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[])</span>
</pre></div>
</div>
<p>Other information provided in the header is stored in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code>
attribute:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;HD&#39;: {&#39;VN&#39;: &#39;1.3&#39;, &#39;SO&#39;: &#39;coordinate&#39;}}</span>
</pre></div>
</div>
<p>With the target information, we can now also print the alignment in PSL
format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>  <span class="c1"># the unmapped sequence; skip it</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="go">&#39;35\t0\t0\t0\t0\t0\t0\t0\t+\tEAS56_57:6:190:289:82\t35\t0\t35\tchr1\t1575\t99\t134\t1\t35,\t0,\t99,\n&#39;</span>
</pre></div>
</div>
<p>We can now also print the alignment in human-readable form, but note
that the target sequence contents is not available from this file:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">chr1             99 ??????????????????????????????????? 134</span>
<span class="go">                  0 ...................................  35</span>
<span class="go">EAS56_57:         0 AGGGGTGCAGAGCCGAGTCACGGGGTTGCCAGCAC  35</span>
</pre></div>
</div>
<p>Alignments in the file <code class="docutils literal notranslate"><span class="pre">sam1.sam</span></code> in the Biopython test suite contain
an additional <code class="docutils literal notranslate"><span class="pre">MD</span></code> tag that shows how the query sequence differs from
the target sequence:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>@SQ     SN:1    LN:239940
@PG     ID:bwa  PN:bwa  VN:0.6.2-r126
HWI-1KL120:88:D0LRBACXX:1:1101:1780:2146        77      *       0       0       *       *       0       0       GATGGGAAACCCATGGCCGAGTGGGAAGAAACCAGCTGAGGTCACATCACCAGAGGAGGGAGAGTGTGGCCCCTGACTCAGTCCATCAGCTTGTGGAGCTG   @=?DDDDBFFFF7A;E?GGEGE8BB?FF?F&gt;G@F=GIIDEIBCFF&lt;FEFEC@EEEE2?8B8/=@((-;?@2&lt;B9@##########################
...
HWI-1KL120:88:D0LRBACXX:1:1101:2852:2134        137     1       136186  25      101M    =       136186  0       TCACGGTGGCCTGTTGAGGCAGGGGCTCACGCTGACCTCTCTCGGCGTGGGAGGGGCCGGTGTGAGGCAAGGGCTCACGCTGACCTCTCTCGGCGTGGGAG   @C@FFFDFHGHHHJJJIJJJJIJJJGEDHHGGHGBGIIGIIAB@GEE=BDBBCCDD@D@B7@;@DDD?&lt;A?DD728:&gt;8()009&gt;:&gt;&gt;C@&gt;5??B######   XT:A:U  NM:i:5  SM:i:25 AM:i:0  X0:i:1  X1:i:0  XM:i:5  XO:i:0  XG:i:0  MD:Z:25G14G2C34A12A9
</pre></div>
</div>
<p>The parser reconstructs the local genome sequence from the <code class="docutils literal notranslate"><span class="pre">MD</span></code> tag,
allowing us to see the target sequence explicitly when printing the
alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;sam1.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="ow">not</span> <span class="n">alignment</span><span class="o">.</span><span class="n">flag</span> <span class="o">&amp;</span> <span class="mi">4</span><span class="p">:</span>  <span class="c1"># Skip the unmapped lines</span>
<span class="gp">... </span>        <span class="k">break</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (2 rows x 101 columns) at ...&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">1            136185 TCACGGTGGCCTGTTGAGGCAGGGGGTCACGCTGACCTCTGTCCGCGTGGGAGGGGCCGG</span>
<span class="go">                  0 |||||||||||||||||||||||||.||||||||||||||.||.||||||||||||||||</span>
<span class="go">HWI-1KL12         0 TCACGGTGGCCTGTTGAGGCAGGGGCTCACGCTGACCTCTCTCGGCGTGGGAGGGGCCGG</span>

<span class="go">1            136245 TGTGAGGCAAGGGCTCACACTGACCTCTCTCAGCGTGGGAG 136286</span>
<span class="go">                 60 ||||||||||||||||||.||||||||||||.|||||||||    101</span>
<span class="go">HWI-1KL12        60 TGTGAGGCAAGGGCTCACGCTGACCTCTCTCGGCGTGGGAG    101</span>
</pre></div>
</div>
<p>SAM files may include additional information to distinguish simple
sequence insertions and deletions from skipped regions of the genome
(e.g. introns), hard and soft clipping, and padded sequence regions. As
this information cannot be stored in the <code class="docutils literal notranslate"><span class="pre">coordinates</span></code> attribute of an
<code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object, and is stored in a dedicated <code class="docutils literal notranslate"><span class="pre">operations</span></code>
attribute instead. Let’s use the third alignment in this SAM file as an
example:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;dna_rna.sam&quot;</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;SAM&quot;</span><span class="p">))</span>  
<span class="go">NR_111921.1 0   chr3    48663768    0   46M1827N82M3376N76M12H  *   0   0   CACGAGAGGAGCGGAGGCGAGGGGTGAACGCGGAGCACTCCAATCGCTCCCAACTAGAGGTCCACCCAGGACCCAGAGACCTGGATTTGAGGCTGCTGGGCGGCAGATGGAGCGATCAGAAGACCAGGAGACGGGAGCTGGAGTGCAGTGGCTGTTCACAAGCGTGAAAGCAAAGATTAAAAAATTTGTTTTTATATTAAAAAA    *   AS:i:1000   NM:i:0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[48663767, 48663813, 48665640, 48665722, 48669098, 48669174],</span>
<span class="go">       [       0,       46,       46,      128,      128,      204]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">operations</span>
<span class="go">bytearray(b&#39;MNMNM&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;hard_clip_right&quot;</span><span class="p">]</span>
<span class="go">12</span>
</pre></div>
</div>
<p>In this alignment, the cigar string <code class="docutils literal notranslate"><span class="pre">63M1062N75M468N43M</span></code> defines 46
aligned nucleotides, an intron of 1827 nucleotides, 82 aligned
nucleotides, an intron of 3376 nucleotides, 76 aligned nucleotides, and
12 hard-clipped nucleotides. These operations are shown in the
<code class="docutils literal notranslate"><span class="pre">operations</span></code> attribute, except for hard-clipping, which is stored in
<code class="docutils literal notranslate"><span class="pre">alignment.query.annotations[&quot;hard_clip_right&quot;]</span></code> (or
<code class="docutils literal notranslate"><span class="pre">alignment.query.annotations[&quot;hard_clip_left&quot;]</span></code>, if applicable)
instead.</p>
<p>To write a SAM file with alignments created from scratch, use an
<code class="docutils literal notranslate"><span class="pre">Alignments</span></code> (plural) object (see Section <a class="reference internal" href="#sec-alignments"><span class="std std-ref">The Alignments class</span></a>)
to store the alignments as well as the metadata and targets:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">Alignments</span><span class="p">()</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;chr1&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">15000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">seq2</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;chr2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">targets</span> <span class="o">=</span> <span class="p">[</span><span class="n">target1</span><span class="p">,</span> <span class="n">target2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;HD&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;VN&quot;</span><span class="p">:</span> <span class="s2">&quot;1.3&quot;</span><span class="p">,</span> <span class="s2">&quot;SO&quot;</span><span class="p">:</span> <span class="s2">&quot;coordinate&quot;</span><span class="p">}}</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">seqA</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">queryA</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">seqA</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;readA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">target1</span><span class="p">,</span> <span class="n">queryA</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">4300</span><span class="p">,</span> <span class="mi">4320</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">Alignment</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">coordinates</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">seqB</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">queryB</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">seqB</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;readB&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">target1</span><span class="p">,</span> <span class="n">queryB</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">5900</span><span class="p">,</span> <span class="mi">5925</span><span class="p">],</span> <span class="p">[</span><span class="mi">25</span><span class="p">,</span> <span class="mi">0</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">Alignment</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">coordinates</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">seqC</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">queryC</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">seqC</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;readC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">target2</span><span class="p">,</span> <span class="n">queryC</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">12300</span><span class="p">,</span> <span class="mi">12318</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">18</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">Alignment</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">coordinates</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">stream</span><span class="o">.</span><span class="n">getvalue</span><span class="p">())</span>  
<span class="go">@HD VN:1.3  SO:coordinate</span>
<span class="go">@SQ SN:chr1 LN:10000</span>
<span class="go">@SQ SN:chr2 LN:15000</span>
<span class="go">readA   0   chr1    4301    255 20M *   0   0   *   *</span>
<span class="go">readB   16  chr1    5901    255 25M *   0   0   *   *</span>
<span class="go">readC   0   chr2    12301   255 18M22S  *   0   0   *       *</span>
</pre></div>
</div>
</section>
<section id="browser-extensible-data-bed">
<span id="subsec-align-bed"></span><h3>Browser Extensible Data (BED)<a class="headerlink" href="#browser-extensible-data-bed" title="Link to this heading"></a></h3>
<p>BED (Browser Extensible Data) files are typically used to store the
alignments of gene transcripts to the genome. See the <a class="reference external" href="http://genome.cse.ucsc.edu/FAQ/FAQformat.html#format1">description from
UCSC</a> for a
full explanation of the BED format.</p>
<p>BED files have three required fields and nine optional fields. The file
<code class="docutils literal notranslate"><span class="pre">bed12.bed</span></code> in subdirectory <code class="docutils literal notranslate"><span class="pre">Tests/Blat</span></code> is an example of a BED file
with 12 fields:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>chr22   1000    5000    mRNA1   960 +   1200    4900    255,0,0 2   567,488,    0,3512,
chr22   2000    6000    mRNA2   900 -   2300    5960    0,255,0 2   433,399,    0,3601,
</pre></div>
</div>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;bed12.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">[[1000 1567 4512 5000]</span>
<span class="go"> [   0  567  567 1055]]</span>
<span class="go">[[2000 2433 5601 6000]</span>
<span class="go"> [ 832  399  399    0]]</span>
</pre></div>
</div>
<p>Note that the first sequence (”<code class="docutils literal notranslate"><span class="pre">mRNA1</span></code>”) was mapped to the forward
strand, while the second sequence (”<code class="docutils literal notranslate"><span class="pre">mRNA2</span></code>”) was mapped to the
reverse strand.</p>
<p>As a BED file does not store the length of each chromosome, the length
of the target sequence is set to its maximum:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span>
<span class="go">SeqRecord(seq=Seq(None, length=9223372036854775807), id=&#39;chr22&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[])</span>
</pre></div>
</div>
<p>The length of the query sequence can be inferred from its alignment
information:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span>
<span class="go">SeqRecord(seq=Seq(None, length=832), id=&#39;mRNA2&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&#39;, dbxrefs=[])</span>
</pre></div>
</div>
<p>The alignment score (field 5) and information stored in fields 7-9
(referred to as <code class="docutils literal notranslate"><span class="pre">thickStart</span></code>, <code class="docutils literal notranslate"><span class="pre">thickEnd</span></code>, and <code class="docutils literal notranslate"><span class="pre">itemRgb</span></code> in the BED
format specification) are stored as attributes on the <code class="docutils literal notranslate"><span class="pre">alignment</span></code>
object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span>
<span class="go">900.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">thickStart</span>
<span class="go">2300</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">thickEnd</span>
<span class="go">5960</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">itemRgb</span>
<span class="go">&#39;0,255,0&#39;</span>
</pre></div>
</div>
<p>To print an alignment in the BED format, you can use Python’s built-in
<code class="docutils literal notranslate"><span class="pre">format</span></code> function:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">chr22   2000    6000    mRNA2   900 -   2300    5960    0,255,0 2   433,399,    0,3601,</span>
</pre></div>
</div>
<p>or you can use the <code class="docutils literal notranslate"><span class="pre">format</span></code> method of the <code class="docutils literal notranslate"><span class="pre">alignment</span></code> object. This
allows you to specify the number of fields to be written as the <code class="docutils literal notranslate"><span class="pre">bedN</span></code>
keyword argument:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">chr22   2000    6000    mRNA2   900 -   2300    5960    0,255,0 2   433,399,    0,3601,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>  
<span class="go">chr22   2000    6000</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>  
<span class="go">chr22   2000    6000    mRNA2   900 -</span>
</pre></div>
</div>
<p>The same keyword argument can be used with <code class="docutils literal notranslate"><span class="pre">Align.write</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="s2">&quot;mybed3file.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">,</span> <span class="n">bedN</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="s2">&quot;mybed6file.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">,</span> <span class="n">bedN</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="s2">&quot;mybed12file.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
</section>
<section id="bigbed">
<span id="subsec-align-bigbed"></span><h3>bigBed<a class="headerlink" href="#bigbed" title="Link to this heading"></a></h3>
<p>The bigBed file format is an indexed binary version of a BED
file <a class="reference internal" href="#subsec-align-bed"><span class="std std-ref">Browser Extensible Data (BED)</span></a>. To create a bigBed file, you can
either use the <code class="docutils literal notranslate"><span class="pre">bedToBigBed</span></code> program from UCSC
(`) &lt;<a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bigBed.html">https://genome.ucsc.edu/goldenPath/help/bigBed.html</a>&gt;`__. or you can
use Biopython for it by calling the <code class="docutils literal notranslate"><span class="pre">Bio.Align.write</span></code> function with
<code class="docutils literal notranslate"><span class="pre">fmt=&quot;bigbed&quot;</span></code>. While the two methods should result in identical
bigBed files, using <code class="docutils literal notranslate"><span class="pre">bedToBigBed</span></code> is much faster and may be more
reliable, as it is the gold standard. As bigBed files come with a
built-in index, it allows you to quickly search a specific genomic
region.</p>
<p>As an example, let’s parse the bigBed file <code class="docutils literal notranslate"><span class="pre">dna_rna.bb</span></code>, available in
the <code class="docutils literal notranslate"><span class="pre">Tests/Blat</span></code> subdirectory in the Biopython distribution:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;dna_rna.bb&quot;</span><span class="p">,</span> <span class="s2">&quot;bigbed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">declaration</span><span class="p">)</span>  
<span class="go">table bed</span>
<span class="go">&quot;Browser Extensible Data&quot;</span>
<span class="go">(</span>
<span class="go">   string          chrom;          &quot;Reference sequence chromosome or scaffold&quot;</span>
<span class="go">   uint            chromStart;     &quot;Start position in chromosome&quot;</span>
<span class="go">   uint            chromEnd;       &quot;End position in chromosome&quot;</span>
<span class="go">   string          name;           &quot;Name of item.&quot;</span>
<span class="go">   uint            score;          &quot;Score (0-1000)&quot;</span>
<span class="go">   char[1]         strand;         &quot;+ or - for strand&quot;</span>
<span class="go">   uint            thickStart;     &quot;Start of where display should be thick (start codon)&quot;</span>
<span class="go">   uint            thickEnd;       &quot;End of where display should be thick (stop codon)&quot;</span>
<span class="go">   uint            reserved;       &quot;Used as itemRgb as of 2004-11-22&quot;</span>
<span class="go">   int             blockCount;     &quot;Number of blocks&quot;</span>
<span class="go">   int[blockCount] blockSizes;     &quot;Comma separated list of block sizes&quot;</span>
<span class="go">   int[blockCount] chromStarts;    &quot;Start positions relative to chromStart&quot;</span>
<span class="go">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">declaration</span></code> contains the specification of the columns, in
AutoSql format, that was used to create the bigBed file. Target
sequences (typically, the chromosomes against which the sequences were
aligned) are stored in the <code class="docutils literal notranslate"><span class="pre">targets</span></code> attribute. In the bigBed format,
only the identifier and the size of each target is stored. In this
example, there is only a single chromosome:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">targets</span>
<span class="go">[SeqRecord(seq=Seq(None, length=198295559), id=&#39;chr3&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&lt;unknown description&gt;&#39;, dbxrefs=[])]</span>
</pre></div>
</div>
<p>Let’s look at the individual alignments. The alignment information is
stored in the same way as for a BED file (see section
<a class="reference internal" href="#subsec-align-bed"><span class="std std-ref">Browser Extensible Data (BED)</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;chr3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;NR_046654.1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[42530895, 42530958, 42532020, 42532095, 42532563, 42532606],</span>
<span class="go">       [     181,      118,      118,       43,       43,        0]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">thickStart</span>
<span class="go">42530895</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">thickEnd</span>
<span class="go">42532606</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>  
<span class="go">chr3       42530895 ????????????????????????????????????????????????????????????</span>
<span class="go">                  0 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
<span class="go">NR_046654       181 ????????????????????????????????????????????????????????????</span>

<span class="go">chr3       42530955 ????????????????????????????????????????????????????????????</span>
<span class="go">                 60 |||---------------------------------------------------------</span>
<span class="go">NR_046654       121 ???---------------------------------------------------------</span>
<span class="go">...</span>
<span class="go">chr3       42532515 ????????????????????????????????????????????????????????????</span>
<span class="go">               1620 ------------------------------------------------||||||||||||</span>
<span class="go">NR_046654        43 ------------------------------------------------????????????</span>

<span class="go">chr3       42532575 ??????????????????????????????? 42532606</span>
<span class="go">               1680 |||||||||||||||||||||||||||||||     1711</span>
<span class="go">NR_046654        31 ???????????????????????????????        0</span>
</pre></div>
</div>
<p>The default bigBed format does not store the sequence contents of the
target and query. If these are available elsewhere (for example, a Fasta
file), you can set <code class="docutils literal notranslate"><span class="pre">alignment.target.seq</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment.query.seq</span></code>
to show the sequence contents when printing the alignment, or to write
the alignment in formats that require the sequence contents (such as
Clustal, see section <a class="reference internal" href="#subsec-align-clustal"><span class="std std-ref">ClustalW</span></a>). The test script
<code class="docutils literal notranslate"><span class="pre">test_Align_bigbed.py</span></code> in the <code class="docutils literal notranslate"><span class="pre">Tests</span></code> subdirectory in the Biopython
distribution gives some examples on how to do that.</p>
<p>Now let’s see how to search for a sequence region. These are the
sequences stored in the bigBed file, printed in BED format (see section
<a class="reference internal" href="#subsec-align-bed"><span class="std std-ref">Browser Extensible Data (BED)</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="gp">...</span>
<span class="go">chr3    42530895    42532606    NR_046654.1 1000    -   42530895    42532606    0   3   63,75,43,   0,1125,1668,</span>

<span class="go">chr3    42530895    42532606    NR_046654.1_modified    978 -   42530895    42532606    0   5   27,36,17,56,43, 0,27,1125,1144,1668,</span>

<span class="go">chr3    48663767    48669174    NR_111921.1 1000    +   48663767    48669174    0   3   46,82,76,   0,1873,5331,</span>

<span class="go">chr3    48663767    48669174    NR_111921.1_modified    972 +   48663767    48669174    0   5   28,17,76,6,76,  0,29,1873,1949,5331,</span>
</pre></div>
</div>
<p>Use the <code class="docutils literal notranslate"><span class="pre">search</span></code> method on the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> object to find regions
on chr3 between positions 48000000 and 49000000. This method returns an
iterator:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selected_alignments</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;chr3&quot;</span><span class="p">,</span> <span class="mi">48000000</span><span class="p">,</span> <span class="mi">49000000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">selected_alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">NR_111921.1</span>
<span class="go">NR_111921.1_modified</span>
</pre></div>
</div>
<p>The chromosome name may be <code class="docutils literal notranslate"><span class="pre">None</span></code> to include all chromosomes, and the
start and end positions may be <code class="docutils literal notranslate"><span class="pre">None</span></code> to start searching from position
0 or to continue searching until the end of the chromosome,
respectively.</p>
<p>Writing alignments in the bigBed format is as easy as calling
<code class="docutils literal notranslate"><span class="pre">Bio.Align.write</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="s2">&quot;output.bb&quot;</span><span class="p">,</span> <span class="s2">&quot;bigbed&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can specify the number of BED fields to be included in the bigBed
file. For example, to write a BED6 file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignments</span><span class="p">,</span> <span class="s2">&quot;output.bb&quot;</span><span class="p">,</span> <span class="s2">&quot;bigbed&quot;</span><span class="p">,</span> <span class="n">bedN</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
</pre></div>
</div>
<p>Same as for <code class="docutils literal notranslate"><span class="pre">bedToBigBed</span></code>, you can include additional columns in the
bigBed output. Suppose the file <code class="docutils literal notranslate"><span class="pre">bedExample2.as</span></code> (available in the
<code class="docutils literal notranslate"><span class="pre">Tests/Blat</span></code> subdirectory of the Biopython distribution) stores the
declaration of the included BED fields in AutoSql format. We can read
this declaration as follows:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">bigbed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;bedExample2.as&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stream</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">autosql_data</span> <span class="o">=</span> <span class="n">stream</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">declaration</span> <span class="o">=</span> <span class="n">bigbed</span><span class="o">.</span><span class="n">AutoSQLTable</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="n">autosql_data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">declaration</span><span class="p">)</span>
<span class="go">&lt;class &#39;Bio.Align.bigbed.AutoSQLTable&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">declaration</span><span class="p">)</span>
<span class="go">table hg18KGchr7</span>
<span class="go">&quot;UCSC Genes for chr7 with color plus GeneSymbol and SwissProtID&quot;</span>
<span class="go">(</span>
<span class="go">   string  chrom;         &quot;Reference sequence chromosome or scaffold&quot;</span>
<span class="go">   uint    chromStart;    &quot;Start position of feature on chromosome&quot;</span>
<span class="go">   uint    chromEnd;      &quot;End position of feature on chromosome&quot;</span>
<span class="go">   string  name;          &quot;Name of gene&quot;</span>
<span class="go">   uint    score;         &quot;Score&quot;</span>
<span class="go">   char[1] strand;        &quot;+ or - for strand&quot;</span>
<span class="go">   uint    thickStart;    &quot;Coding region start&quot;</span>
<span class="go">   uint    thickEnd;      &quot;Coding region end&quot;</span>
<span class="go">   uint    reserved;      &quot;Green on + strand, Red on - strand&quot;</span>
<span class="go">   string  geneSymbol;    &quot;Gene Symbol&quot;</span>
<span class="go">   string  spID;          &quot;SWISS-PROT protein Accession number&quot;</span>
<span class="go">)</span>
</pre></div>
</div>
<p>Now we can write a bigBed file with the 9 BED fields plus the additional
fields <code class="docutils literal notranslate"><span class="pre">geneSymbol</span></code> and <code class="docutils literal notranslate"><span class="pre">spID</span></code> by calling</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Align</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">alignments</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;output.bb&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s2">&quot;bigbed&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">bedN</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">declaration</span><span class="o">=</span><span class="n">declaration</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">extraIndex</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;geneSymbol&quot;</span><span class="p">],</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
<p>Here, we also requested to include additional indices on the <code class="docutils literal notranslate"><span class="pre">name</span></code>
and <code class="docutils literal notranslate"><span class="pre">geneSymbol</span></code> in the bigBed file. <code class="docutils literal notranslate"><span class="pre">Align.write</span></code> expects to find
the keys <code class="docutils literal notranslate"><span class="pre">geneSymbol</span></code> and <code class="docutils literal notranslate"><span class="pre">spID</span></code> in the <code class="docutils literal notranslate"><span class="pre">alignment.annotations</span></code>
dictionary. Please refer to the test script <code class="docutils literal notranslate"><span class="pre">test_Align_bigbed.py</span></code> in
the <code class="docutils literal notranslate"><span class="pre">Tests</span></code> subdirectory in the Biopython distribution for more
examples of writing alignment files in the bigBed format.</p>
</section>
<section id="pattern-space-layout-psl">
<span id="subsec-align-psl"></span><h3>Pattern Space Layout (PSL)<a class="headerlink" href="#pattern-space-layout-psl" title="Link to this heading"></a></h3>
<p>PSL (Pattern Space Layout) files are are generated by the BLAST-Like
Alignment Tool BLAT <a class="reference internal" href="chapter_bibliography.html#kent2002" id="id11"><span>[Kent2002]</span></a>. Like BED files (see
section <a class="reference internal" href="#subsec-align-bed"><span class="std std-ref">Browser Extensible Data (BED)</span></a>), PSL files are typically used to
store alignments of transcripts to genomes. This is an example of a
short BLAT file (available as <code class="docutils literal notranslate"><span class="pre">dna_rna.psl</span></code> in the <code class="docutils literal notranslate"><span class="pre">Tests/Blat</span></code>
subdirectory of the Biopython distribution), with the standard PSL
header consisting of 5 lines:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>psLayout version 3

match   mis-    rep.    N&#39;s Q gap   Q gap   T gap   T gap   strand  Q           Q       Q       Q   T           T       T       T   block   blockSizes  qStarts  tStarts
        match   match       count   bases   count   bases           name        size    start   end name        size    start   end count
---------------------------------------------------------------------------------------------------------------------------------------------------------------
165 0   39  0   0   0   2   5203    +   NR_111921.1 216 0   204 chr3    198295559   48663767    48669174    3   46,82,76,   0,46,128,   48663767,48665640,48669098,
175 0   6   0   0   0   2   1530    -   NR_046654.1 181 0   181 chr3    198295559   42530895    42532606    3   63,75,43,   0,63,138,   42530895,42532020,42532563,
162 2   39  0   1   2   3   5204    +   NR_111921.1_modified    220 3   208 chr3    198295559   48663767    48669174    5   28,17,76,6,76,  3,31,48,126,132,    48663767,48663796,48665640,48665716,48669098,
172 1   6   0   1   3   3   1532    -   NR_046654.1_modified    190 3   185 chr3    198295559   42530895    42532606    5   27,36,17,56,43, 5,35,71,88,144, 42530895,42530922,42532020,42532039,42532563,
</pre></div>
</div>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;dna_rna.psl&quot;</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>
<span class="go">{&#39;psLayout version&#39;: &#39;3&#39;}</span>
</pre></div>
</div>
<p>Iterate over the alignments to get one <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object for each
line:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">chr3 NR_046654.1</span>
<span class="go">chr3 NR_046654.1_modified</span>
<span class="go">chr3 NR_111921.1</span>
<span class="go">chr3 NR_111921.1_modified</span>
</pre></div>
</div>
<p>Let’s look at the last alignment in more detail. The first four columns
in the PSL file show the number of matches, the number of mismatches,
the number of nucleotides aligned to repeat regions, and the number of
nucleotides aligned to N (unknown) characters. These values are stored
as attributes to the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">matches</span>
<span class="go">162</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">misMatches</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">repMatches</span>
<span class="go">39</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">nCount</span>
<span class="go">0</span>
</pre></div>
</div>
<p>As the sequence data of the target and query are not stored explicitly
in the PSL file, the sequence content of <code class="docutils literal notranslate"><span class="pre">alignment.target</span></code> and
<code class="docutils literal notranslate"><span class="pre">alignment.query</span></code> is undefined. However, their sequence lengths are
known:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span>  
<span class="go">SeqRecord(seq=Seq(None, length=198295559), id=&#39;chr3&#39;, ...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span>  
<span class="go">SeqRecord(seq=Seq(None, length=220), id=&#39;NR_111921.1_modified&#39;, ...)</span>
</pre></div>
</div>
<p>We can print the alignment in BED or PSL format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">chr3    48663767    48669174    NR_111921.1_modified    0   +   48663767    48669174    0   5   28,17,76,6,76,  0,29,1873,1949,5331,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">))</span>  
<span class="go">162 2   39  0   1   2   3   5204    +   NR_111921.1_modified    220 3   208 chr3    198295559   48663767    48669174    5   28,17,76,6,76,  3,31,48,126,132,    48663767,48663796,48665640,48665716,48669098,</span>
</pre></div>
</div>
<p>Here, the number of matches, mismatches, repeat region matches, and
matches to unknown nucleotides were taken from the corresponding
attributes of the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object. If these attributes are not
available, for example if the alignment did not come from a PSL file,
then these numbers are calculated using the sequence contents, if
available. Repeat regions in the target sequence are indicated by
masking the sequence as lower-case or upper-case characters, as defined
by the following values for the <code class="docutils literal notranslate"><span class="pre">mask</span></code> keyword argument:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">False</span></code> (default): Do not count matches to masked sequences
separately;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;lower&quot;</span></code>: Count and report matches to lower-case characters as
matches to repeat regions;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;upper&quot;</span></code>: Count and report matches to upper-case characters as
matches to repeat regions;</p></li>
</ul>
<p>The character used for unknown nucleotides is defined by the
<code class="docutils literal notranslate"><span class="pre">wildcard</span></code> argument. For consistency with BLAT, the wildcard character
is <code class="docutils literal notranslate"><span class="pre">&quot;N&quot;</span></code> by default. Use <code class="docutils literal notranslate"><span class="pre">wildcard=None</span></code> if you don’t want to count
matches to any unknown nucleotides separately.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;GGTGGGGG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;AAAAAAAggggGGNGAAAAA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coordinates</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">Alignment</span><span class="p">([</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">],</span> <span class="n">coordinates</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 AAAAAAAggggGGNGAAAAA 20</span>
<span class="go">                  0 -------....||.|----- 20</span>
<span class="go">query             0 -------GGTGGGGG-----  8</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">line</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>  
<span class="go">6   1   0   1   0   0   0   0   +   query   8   0   8   target   20   7   15   1   8,   0,   7,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">line</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;psl&quot;</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="s2">&quot;lower&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>  
<span class="go">3   1   3   1   0   0   0   0   +   query   8   0   8   target   20   7   15   1   8,   0,   7,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">line</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;psl&quot;</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="s2">&quot;lower&quot;</span><span class="p">,</span> <span class="n">wildcard</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>  
<span class="go">3   2   3   0   0   0   0   0   +   query   8   0   8   target   20   7   15   1   8,   0,   7,</span>
</pre></div>
</div>
<p>The same arguments can be used when writing alignments to an output file
in PSL format using <code class="docutils literal notranslate"><span class="pre">Bio.Align.write</span></code>. This function has an additional
keyword <code class="docutils literal notranslate"><span class="pre">header</span></code> (<code class="docutils literal notranslate"><span class="pre">True</span></code> by default) specifying if the PSL header
should be written.</p>
<p>In addition to the <code class="docutils literal notranslate"><span class="pre">format</span></code> method, you can use Python’s built-in
<code class="docutils literal notranslate"><span class="pre">format</span></code> function:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">))</span>  
<span class="go">6   1   0   1   0   0   0   0   +   query   8   0   8   target   20   7   15   1   8,   0,   7,</span>
</pre></div>
</div>
<p>allowing <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects to be used in formatted (f-) strings in
Python:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">line</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;The alignment in PSL format is &#39;</span><span class="si">{</span><span class="n">alignment</span><span class="si">:</span><span class="s2">psl</span><span class="si">}</span><span class="s2">&#39;.&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>  
<span class="go">The alignment in PSL format is &#39;6   1   0   1   0   0   0   0   +   query   8   0   8   target   20   7   15   1   8,   0,   7,</span>
<span class="go">&#39;</span>
</pre></div>
</div>
<p>Note that optional keyword arguments cannot be used with the <code class="docutils literal notranslate"><span class="pre">format</span></code>
function or with formatted strings.</p>
</section>
<section id="bigpsl">
<span id="subsec-align-bigpsl"></span><h3>bigPsl<a class="headerlink" href="#bigpsl" title="Link to this heading"></a></h3>
<p>A bigPsl file is a bigBed file with a BED12+13 format consisting of the
12 predefined BED fields and 13 custom fields defined in the AutoSql
file <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bigPsl.html">bigPsl.as</a>
provided by UCSC, creating an indexed binary version of a PSL file (see
section <a class="reference internal" href="#subsec-align-psl"><span class="std std-ref">Pattern Space Layout (PSL)</span></a>). To create a bigPsl file, you
can either use the <code class="docutils literal notranslate"><span class="pre">pslToBigPsl</span></code> and <code class="docutils literal notranslate"><span class="pre">bedToBigBed</span></code> programs from
UCSC. or you can use Biopython by calling the <code class="docutils literal notranslate"><span class="pre">Bio.Align.write</span></code>
function with <code class="docutils literal notranslate"><span class="pre">fmt=&quot;bigpsl&quot;</span></code>. While the two methods should result in
identical bigPsl files, the UCSC tools are much faster and may be more
reliable, as it is the gold standard. As bigPsl files are bigBed files,
they come with a built-in index, allowing you to quickly search a
specific genomic region.</p>
<p>As an example, let’s parse the bigBed file <code class="docutils literal notranslate"><span class="pre">dna_rna.psl.bb</span></code>, available
in the <code class="docutils literal notranslate"><span class="pre">Tests/Blat</span></code> subdirectory in the Biopython distribution. This
file is the bigPsl equivalent of the bigBed file <code class="docutils literal notranslate"><span class="pre">dna_rna.bb</span></code> (see
section <a class="reference internal" href="#subsec-align-bigbed"><span class="std std-ref">bigBed</span></a>) and of the PSL file
<code class="docutils literal notranslate"><span class="pre">dna_rna.psl</span></code> (see section <a class="reference internal" href="#subsec-align-psl"><span class="std std-ref">Pattern Space Layout (PSL)</span></a>).</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;dna_rna.psl.bb&quot;</span><span class="p">,</span> <span class="s2">&quot;bigpsl&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">declaration</span><span class="p">)</span>  
<span class="go">table bigPsl</span>
<span class="go">&quot;bigPsl pairwise alignment&quot;</span>
<span class="go">(</span>
<span class="go">   string          chrom;           &quot;Reference sequence chromosome or scaffold&quot;</span>
<span class="go">   uint            chromStart;      &quot;Start position in chromosome&quot;</span>
<span class="go">   uint            chromEnd;        &quot;End position in chromosome&quot;</span>
<span class="go">   string          name;            &quot;Name or ID of item, ideally both human readable and unique&quot;</span>
<span class="go">   uint            score;           &quot;Score (0-1000)&quot;</span>
<span class="go">   char[1]         strand;          &quot;+ or - indicates whether the query aligns to the + or - strand on the reference&quot;</span>
<span class="go">   uint            thickStart;      &quot;Start of where display should be thick (start codon)&quot;</span>
<span class="go">   uint            thickEnd;        &quot;End of where display should be thick (stop codon)&quot;</span>
<span class="go">   uint            reserved;        &quot;RGB value (use R,G,B string in input file)&quot;</span>
<span class="go">   int             blockCount;      &quot;Number of blocks&quot;</span>
<span class="go">   int[blockCount] blockSizes;      &quot;Comma separated list of block sizes&quot;</span>
<span class="go">   int[blockCount] chromStarts;     &quot;Start positions relative to chromStart&quot;</span>
<span class="go">   uint            oChromStart;     &quot;Start position in other chromosome&quot;</span>
<span class="go">   uint            oChromEnd;       &quot;End position in other chromosome&quot;</span>
<span class="go">   char[1]         oStrand;         &quot;+ or -, - means that psl was reversed into BED-compatible coordinates&quot;</span>
<span class="go">   uint            oChromSize;      &quot;Size of other chromosome.&quot;</span>
<span class="go">   int[blockCount] oChromStarts;    &quot;Start positions relative to oChromStart or from oChromStart+oChromSize depending on strand&quot;</span>
<span class="go">   lstring         oSequence;       &quot;Sequence on other chrom (or edit list, or empty)&quot;</span>
<span class="go">   string          oCDS;            &quot;CDS in NCBI format&quot;</span>
<span class="go">   uint            chromSize;       &quot;Size of target chromosome&quot;</span>
<span class="go">   uint            match;           &quot;Number of bases matched.&quot;</span>
<span class="go">   uint            misMatch;        &quot;Number of bases that don&#39;t match&quot;</span>
<span class="go">   uint            repMatch;        &quot;Number of bases that match but are part of repeats&quot;</span>
<span class="go">   uint            nCount;          &quot;Number of &#39;N&#39; bases&quot;</span>
<span class="go">   uint            seqType;         &quot;0=empty, 1=nucleotide, 2=amino_acid&quot;</span>
<span class="go">)</span>
</pre></div>
</div>
<p>The declaration contains the specification of the columns as defined by
the <code class="docutils literal notranslate"><span class="pre">bigPsl.as</span></code> AutoSql file from UCSC. Target sequences (typically,
the chromosomes against which the sequences were aligned) are stored in
the <code class="docutils literal notranslate"><span class="pre">targets</span></code> attribute. In the bigBed format, only the identifier and
the size of each target is stored. In this example, there is only a
single chromosome:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">targets</span>
<span class="go">[SeqRecord(seq=Seq(None, length=198295559), id=&#39;chr3&#39;, name=&#39;&lt;unknown name&gt;&#39;, description=&#39;&lt;unknown description&gt;&#39;, dbxrefs=[])]</span>
</pre></div>
</div>
<p>Iterating over the alignments gives one Alignment object for each line:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">chr3 NR_046654.1</span>
<span class="go">chr3 NR_046654.1_modified</span>
<span class="go">chr3 NR_111921.1</span>
<span class="go">chr3 NR_111921.1_modified</span>
</pre></div>
</div>
<p>Let’s look at the individual alignments. The alignment information is
stored in the same way as for the corresponding PSL file (see
section <a class="reference internal" href="#subsec-align-psl"><span class="std std-ref">Pattern Space Layout (PSL)</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[48663767, 48663795, 48663796, 48663813, 48665640, 48665716,</span>
<span class="go">        48665716, 48665722, 48669098, 48669174],</span>
<span class="go">       [       3,       31,       31,       48,       48,      124,</span>
<span class="go">             126,      132,      132,      208]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">thickStart</span>
<span class="go">48663767</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">thickEnd</span>
<span class="go">48669174</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">matches</span>
<span class="go">162</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">misMatches</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">repMatches</span>
<span class="go">39</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">nCount</span>
<span class="go">0</span>
</pre></div>
</div>
<p>We can print the alignment in BED or PSL format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">chr3    48663767    48669174    NR_111921.1_modified    1000    +   48663767    48669174    0   5   28,17,76,6,76,  0,29,1873,1949,5331,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">))</span>  
<span class="go">162 2   39  0   1   2   3   5204    +   NR_111921.1_modified    220 3   208 chr3    198295559   48663767    48669174    5   28,17,76,6,76,  3,31,48,126,132,    48663767,48663796,48665640,48665716,48669098,</span>
</pre></div>
</div>
<p>As a bigPsl file is a special case of a bigBed file, you can use the
<code class="docutils literal notranslate"><span class="pre">search</span></code> method on the alignments object to find alignments to
specific genomic regions. For example, we can look for regions on chr3
between positions 48000000 and 49000000:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selected_alignments</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;chr3&quot;</span><span class="p">,</span> <span class="mi">48000000</span><span class="p">,</span> <span class="mi">49000000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">selected_alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">NR_111921.1</span>
<span class="go">NR_111921.1_modified</span>
</pre></div>
</div>
<p>The chromosome name may be <code class="docutils literal notranslate"><span class="pre">None</span></code> to include all chromosomes, and the
start and end positions may be <code class="docutils literal notranslate"><span class="pre">None</span></code> to start searching from position
0 or to continue searching until the end of the chromosome,
respectively.</p>
<p>To write a bigPsl file with Biopython, use
<code class="docutils literal notranslate"><span class="pre">Bio.Align.write(alignments,</span> <span class="pre">&quot;myfilename.bb&quot;,</span> <span class="pre">fmt=&quot;bigpsl&quot;)</span></code>, where
<code class="docutils literal notranslate"><span class="pre">myfilename.bb</span></code> is the name of the output bigPsl file. Alternatively,
you can use a (binary) stream for output. Additional options are</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">compress</span></code>: If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default), compress data using zlib; if
<code class="docutils literal notranslate"><span class="pre">False</span></code>, do not compress data.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">extraIndex</span></code>: List of strings with the names of extra columns to be
indexed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cds</span></code>: If <code class="docutils literal notranslate"><span class="pre">True</span></code>, look for a query feature of type CDS and write
it in NCBI style in the PSL file (default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fa</span></code>: If <code class="docutils literal notranslate"><span class="pre">True</span></code>, include the query sequence in the PSL file
(default: <code class="docutils literal notranslate"><span class="pre">False</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">mask</span></code>: Specify if repeat regions in the target sequence are masked
and should be reported in the <code class="docutils literal notranslate"><span class="pre">repMatches</span></code> field instead of in the
<code class="docutils literal notranslate"><span class="pre">matches</span></code> field. Acceptable values are</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">None</span></code>: no masking (default);</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;lower&quot;</span></code>: masking by lower-case characters;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;upper&quot;</span></code>: masking by upper-case characters.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">wildcard</span></code>: Report alignments to the wildcard character
(representing unknown nucleotides) in the target or query sequence in
the <code class="docutils literal notranslate"><span class="pre">nCount</span></code> field instead of in the <code class="docutils literal notranslate"><span class="pre">matches</span></code>, <code class="docutils literal notranslate"><span class="pre">misMatches</span></code>,
or <code class="docutils literal notranslate"><span class="pre">repMatches</span></code> fields. Default value is <code class="docutils literal notranslate"><span class="pre">&quot;N&quot;</span></code>.</p></li>
</ul>
<p>See section <a class="reference internal" href="#subsec-align-psl"><span class="std std-ref">Pattern Space Layout (PSL)</span></a> for an explanation on how the
number of matches, mismatches, repeat region matches, and matches to
unknown nucleotides are obtained.</p>
</section>
<section id="multiple-alignment-format-maf">
<span id="subsec-align-maf"></span><h3>Multiple Alignment Format (MAF)<a class="headerlink" href="#multiple-alignment-format-maf" title="Link to this heading"></a></h3>
<p>MAF (Multiple Alignment Format) files store a series of multiple
sequence alignments in a human-readable format. MAF files are typically
used to store alignment s of genomes to each other. The file
<code class="docutils literal notranslate"><span class="pre">ucsc_test.maf</span></code> in the <code class="docutils literal notranslate"><span class="pre">Tests/MAF</span></code> subdirectory of the Biopython
distribution is an example of a simple MAF file:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>track name=euArc visibility=pack mafDot=off frames=&quot;multiz28wayFrames&quot; speciesOrder=&quot;hg16 panTro1 baboon mm4 rn3&quot; description=&quot;A sample alignment&quot;
##maf version=1 scoring=tba.v8
# tba.v8 (((human chimp) baboon) (mouse rat))
# multiz.v7
# maf_project.v5 _tba_right.maf3 mouse _tba_C
# single_cov2.v4 single_cov2 /dev/stdin

a score=23262.0
s hg16.chr7    27578828 38 + 158545518 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG
s panTro1.chr6 28741140 38 + 161576975 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG
s baboon         116834 38 +   4622798 AAA-GGGAATGTTAACCAAATGA---GTTGTCTCTTATGGTG
s mm4.chr6     53215344 38 + 151104725 -AATGGGAATGTTAAGCAAACGA---ATTGTCTCTCAGTGTG
s rn3.chr4     81344243 40 + 187371129 -AA-GGGGATGCTAAGCCAATGAGTTGTTGTCTCTCAATGTG

a score=5062.0
s hg16.chr7    27699739 6 + 158545518 TAAAGA
s panTro1.chr6 28862317 6 + 161576975 TAAAGA
s baboon         241163 6 +   4622798 TAAAGA
s mm4.chr6     53303881 6 + 151104725 TAAAGA
s rn3.chr4     81444246 6 + 187371129 taagga

a score=6636.0
s hg16.chr7    27707221 13 + 158545518 gcagctgaaaaca
s panTro1.chr6 28869787 13 + 161576975 gcagctgaaaaca
s baboon         249182 13 +   4622798 gcagctgaaaaca
s mm4.chr6     53310102 13 + 151104725 ACAGCTGAAAATA
</pre></div>
</div>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;ucsc_test.maf&quot;</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Information shown in the file header (the track line and subsequent
lines starting with “<code class="docutils literal notranslate"><span class="pre">#</span></code>”)) is stored in the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> attribute of
the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">metadata</span>  
<span class="go">{&#39;name&#39;: &#39;euArc&#39;,</span>
<span class="go"> &#39;visibility&#39;: &#39;pack&#39;,</span>
<span class="go"> &#39;mafDot&#39;: &#39;off&#39;,</span>
<span class="go"> &#39;frames&#39;: &#39;multiz28wayFrames&#39;,</span>
<span class="go"> &#39;speciesOrder&#39;: [&#39;hg16&#39;, &#39;panTro1&#39;, &#39;baboon&#39;, &#39;mm4&#39;, &#39;rn3&#39;],</span>
<span class="go"> &#39;description&#39;: &#39;A sample alignment&#39;,</span>
<span class="go"> &#39;MAF Version&#39;: &#39;1&#39;,</span>
<span class="go"> &#39;Scoring&#39;: &#39;tba.v8&#39;,</span>
<span class="go"> &#39;Comments&#39;: [&#39;tba.v8 (((human chimp) baboon) (mouse rat))&#39;,</span>
<span class="go">              &#39;multiz.v7&#39;,</span>
<span class="go">              &#39;maf_project.v5 _tba_right.maf3 mouse _tba_C&#39;,</span>
<span class="go">              &#39;single_cov2.v4 single_cov2 /dev/stdin&#39;]}</span>
</pre></div>
</div>
<p>By iterating over the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> we obtain one <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object
for each alignment block in the MAF file:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span>
<span class="go">23262.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">{</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">}</span>  
<span class="go">{&#39;hg16.chr7&#39;: 158545518,</span>
<span class="go"> &#39;panTro1.chr6&#39;: 161576975,</span>
<span class="go"> &#39;baboon&#39;: 4622798,</span>
<span class="go"> &#39;mm4.chr6&#39;: 151104725,</span>
<span class="go"> &#39;rn3.chr4&#39;: 187371129}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>  
<span class="go">array([[27578828, 27578829, 27578831, 27578831, 27578850, 27578850, 27578866],</span>
<span class="go">       [28741140, 28741141, 28741143, 28741143, 28741162, 28741162, 28741178],</span>
<span class="go">       [  116834,   116835,   116837,   116837,   116856,   116856, 116872],</span>
<span class="go">       [53215344, 53215344, 53215346, 53215347, 53215366, 53215366, 53215382],</span>
<span class="go">       [81344243, 81344243, 81344245, 81344245, 81344264, 81344267, 81344283]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">hg16.chr7  27578828 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG 27578866</span>
<span class="go">panTro1.c  28741140 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG 28741178</span>
<span class="go">baboon       116834 AAA-GGGAATGTTAACCAAATGA---GTTGTCTCTTATGGTG   116872</span>
<span class="go">mm4.chr6   53215344 -AATGGGAATGTTAAGCAAACGA---ATTGTCTCTCAGTGTG 53215382</span>
<span class="go">rn3.chr4   81344243 -AA-GGGGATGCTAAGCCAATGAGTTGTTGTCTCTCAATGTG 81344283</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">))</span>
<span class="go">5 42</span>
<span class="go">hg16.chr7 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG</span>
<span class="go">panTro1.chAAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG</span>
<span class="go">baboon    AAA-GGGAATGTTAACCAAATGA---GTTGTCTCTTATGGTG</span>
<span class="go">mm4.chr6  -AATGGGAATGTTAAGCAAACGA---ATTGTCTCTCAGTGTG</span>
<span class="go">rn3.chr4  -AA-GGGGATGCTAAGCCAATGAGTTGTTGTCTCTCAATGTG</span>
</pre></div>
</div>
<p>In addition to the “<code class="docutils literal notranslate"><span class="pre">a</span></code>” (alignment block) and “<code class="docutils literal notranslate"><span class="pre">s</span></code>” (sequence)
lines, MAF files may contain “<code class="docutils literal notranslate"><span class="pre">i</span></code>” lines with information about the
genome sequence before and after this block, “<code class="docutils literal notranslate"><span class="pre">e</span></code>” lines with
information about empty parts of the alignment, and “<code class="docutils literal notranslate"><span class="pre">q</span></code>” lines
showing the quality of each aligned base. This is an example of an
alignment block including such lines:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>a score=19159.000000
s mm9.chr10                         3014644 45 + 129993255 CCTGTACC---CTTTGGTGAGAATTTTTGTTTCAGTGTTAAAAGTTTG
s hg18.chr6                        15870786 46 - 170899992 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT
i hg18.chr6                        I 9085 C 0
s panTro2.chr6                     16389355 46 - 173908612 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT
q panTro2.chr6                                             99999999999999999999999-9999999999999999999-9999
i panTro2.chr6                     I 9106 C 0
s calJac1.Contig6394                   6182 46 +    133105 CCTATACCTTTCTTTCATGAGAA-TTTTGTTTGAATCCTAAAC-TTTT
i calJac1.Contig6394               N 0 C 0
s loxAfr1.scaffold_75566               1167 34 -     10574 ------------TTTGGTTAGAA-TTATGCTTTAATTCAAAAC-TTCC
q loxAfr1.scaffold_75566                                   ------------99999699899-9999999999999869998-9997
i loxAfr1.scaffold_75566           N 0 C 0
e tupBel1.scaffold_114895.1-498454   167376 4145 -    498454 I
e echTel1.scaffold_288249             87661 7564 +    100002 I
e otoGar1.scaffold_334.1-359464      181217 2931 -    359464 I
e ponAbe2.chr6                     16161448 8044 - 174210431 I
</pre></div>
</div>
<p>This is the 10th alignment block in the file <code class="docutils literal notranslate"><span class="pre">ucsc_mm9_chr10.maf</span></code>
(available in the <code class="docutils literal notranslate"><span class="pre">Tests/MAF</span></code> subdirectory of the Biopython
distribution):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;ucsc_mm9_chr10.maf&quot;</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span>
<span class="go">19159.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">mm9.chr10   3014644 CCTGTACC---CTTTGGTGAGAATTTTTGTTTCAGTGTTAAAAGTTTG   3014689</span>
<span class="go">hg18.chr6 155029206 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT 155029160</span>
<span class="go">panTro2.c 157519257 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT 157519211</span>
<span class="go">calJac1.C      6182 CCTATACCTTTCTTTCATGAGAA-TTTTGTTTGAATCCTAAAC-TTTT      6228</span>
<span class="go">loxAfr1.s      9407 ------------TTTGGTTAGAA-TTATGCTTTAATTCAAAAC-TTCC      9373</span>
</pre></div>
</div>
<p>The “<code class="docutils literal notranslate"><span class="pre">i</span></code>” lines show the relationship between the sequence in the
current alignment block to the ones in the preceding and subsequent
alignment block. This information is stored in the <code class="docutils literal notranslate"><span class="pre">annotations</span></code>
attribute of the corresponding sequence:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;leftStatus&#39;: &#39;I&#39;, &#39;leftCount&#39;: 9085, &#39;rightStatus&#39;: &#39;C&#39;, &#39;rightCount&#39;: 0}</span>
</pre></div>
</div>
<p>showing that there are 9085 bases inserted (”<code class="docutils literal notranslate"><span class="pre">I</span></code>”) between this block
and the preceding one, while the block is contiguous (”<code class="docutils literal notranslate"><span class="pre">C</span></code>”) with the
subsequent one. See the <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5">UCSC
documentation</a>
for the full description of these fields and status characters.</p>
<p>The “<code class="docutils literal notranslate"><span class="pre">q</span></code>” lines show the sequence quality, which is stored under the
“<code class="docutils literal notranslate"><span class="pre">quality</span></code>” dictionary key of the<code class="docutils literal notranslate"><span class="pre">annotations</span></code> attribute of the
corresponding sequence:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;quality&quot;</span><span class="p">]</span>
<span class="go">&#39;9999999999999999999999999999999999999999999999&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;quality&quot;</span><span class="p">]</span>
<span class="go">&#39;9999969989999999999999998699989997&#39;</span>
</pre></div>
</div>
<p>The “<code class="docutils literal notranslate"><span class="pre">e</span></code>” lines show information about species with a contiguous
sequence before and after this alignment bloack, but with no aligning
nucleotides in this alignment block. This is stored under the
“<code class="docutils literal notranslate"><span class="pre">empty</span></code>” key of the <code class="docutils literal notranslate"><span class="pre">alignment.annotations</span></code> dictionary:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;empty&quot;</span><span class="p">]</span>  
<span class="go">[(SeqRecord(seq=Seq(None, length=498454), id=&#39;tupBel1.scaffold_114895.1-498454&#39;, name=&#39;&#39;, description=&#39;&#39;, dbxrefs=[]), (331078, 326933), &#39;I&#39;),</span>
<span class="go"> (SeqRecord(seq=Seq(None, length=100002), id=&#39;echTel1.scaffold_288249&#39;, name=&#39;&#39;, description=&#39;&#39;, dbxrefs=[]), (87661, 95225), &#39;I&#39;),</span>
<span class="go"> (SeqRecord(seq=Seq(None, length=359464), id=&#39;otoGar1.scaffold_334.1-359464&#39;, name=&#39;&#39;, description=&#39;&#39;, dbxrefs=[]), (178247, 175316), &#39;I&#39;),</span>
<span class="go"> (SeqRecord(seq=Seq(None, length=174210431), id=&#39;ponAbe2.chr6&#39;, name=&#39;&#39;, description=&#39;&#39;, dbxrefs=[]), (158048983, 158040939), &#39;I&#39;)]</span>
</pre></div>
</div>
<p>This shows for example that there were non-aligning bases inserted
(”<code class="docutils literal notranslate"><span class="pre">I</span></code>”) from position 158040939 to 158048983 on the opposite strand of
the <code class="docutils literal notranslate"><span class="pre">ponAbe2.chr6</span></code> genomic sequence. Again, see the <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5">UCSC
documentation</a>
for the full definition of “<code class="docutils literal notranslate"><span class="pre">e</span></code>” lines.</p>
<p>To print an alignment in MAF format, you can use Python’s built-in
<code class="docutils literal notranslate"><span class="pre">format</span></code> function:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;MAF&quot;</span><span class="p">))</span>
<span class="go">a score=19159.000000</span>
<span class="go">s mm9.chr10                         3014644   45 + 129993255 CCTGTACC---CTTTGGTGAGAATTTTTGTTTCAGTGTTAAAAGTTTG</span>
<span class="go">s hg18.chr6                        15870786   46 - 170899992 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT</span>
<span class="go">i hg18.chr6                        I 9085 C 0</span>
<span class="go">s panTro2.chr6                     16389355   46 - 173908612 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT</span>
<span class="go">q panTro2.chr6                                               99999999999999999999999-9999999999999999999-9999</span>
<span class="go">i panTro2.chr6                     I 9106 C 0</span>
<span class="go">s calJac1.Contig6394                   6182   46 +    133105 CCTATACCTTTCTTTCATGAGAA-TTTTGTTTGAATCCTAAAC-TTTT</span>
<span class="go">i calJac1.Contig6394               N 0 C 0</span>
<span class="go">s loxAfr1.scaffold_75566               1167   34 -     10574 ------------TTTGGTTAGAA-TTATGCTTTAATTCAAAAC-TTCC</span>
<span class="go">q loxAfr1.scaffold_75566                                     ------------99999699899-9999999999999869998-9997</span>
<span class="go">i loxAfr1.scaffold_75566           N 0 C 0</span>
<span class="go">e tupBel1.scaffold_114895.1-498454   167376 4145 -    498454 I</span>
<span class="go">e echTel1.scaffold_288249             87661 7564 +    100002 I</span>
<span class="go">e otoGar1.scaffold_334.1-359464      181217 2931 -    359464 I</span>
<span class="go">e ponAbe2.chr6                     16161448 8044 - 174210431 I</span>

</pre></div>
</div>
<p>To write a complete MAF file, use
<code class="docutils literal notranslate"><span class="pre">Bio.Align.write(alignments,</span> <span class="pre">&quot;myfilename.maf&quot;,</span> <span class="pre">fmt=&quot;maf&quot;)</span></code>, where
<code class="docutils literal notranslate"><span class="pre">myfilename.maf</span></code> is the name of the output MAF file. Alternatively,
you can use a (text) stream for output. File header information will be
taken from the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> attribute of the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> object. If
you are creating the alignments from scratch, you can use the
<code class="docutils literal notranslate"><span class="pre">Alignments</span></code> (plural) class to create a list-like <code class="docutils literal notranslate"><span class="pre">alignments</span></code>
object (see Section <a class="reference internal" href="#sec-alignments"><span class="std std-ref">The Alignments class</span></a>) and give it a
<code class="docutils literal notranslate"><span class="pre">metadata</span></code> attribute.</p>
</section>
<section id="bigmaf">
<span id="subsec-align-bigmaf"></span><h3>bigMaf<a class="headerlink" href="#bigmaf" title="Link to this heading"></a></h3>
<p>A bigMaf file is a bigBed file with a BED3+1 format consisting of the 3
required BED fields plus a custom field that stores a MAF alignment
block as a string, crearing an indexed binary version of a MAF file (see
section <a class="reference internal" href="#subsec-align-bigmaf"><span class="std std-ref">bigMaf</span></a>). The associated AutoSql file
<a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/examples/bigMaf.as">bigMaf.as</a>
is provided by UCSC. To create a bigMaf file, you can either use the
<code class="docutils literal notranslate"><span class="pre">mafToBigMaf</span></code> and <code class="docutils literal notranslate"><span class="pre">bedToBigBed</span></code> programs from UCSC. or you can use
Biopython by calling the Bio.Align.write function with <code class="docutils literal notranslate"><span class="pre">fmt=&quot;bigmaf&quot;</span></code>.
While the two methods should result in identical bigMaf files, the UCSC
tools are much faster and may be more reliable, as it is the gold
standard. As bigMaf files are bigBed files, they come with a built-in
index, allowing you to quickly search a specific region of the reference
genome.</p>
<p>The file <code class="docutils literal notranslate"><span class="pre">ucsc_test.bb</span></code> in the <code class="docutils literal notranslate"><span class="pre">Tests/MAF</span></code> subdirectory of the
Biopython distribution is an example of a bigMaf file. This file is
equivalent to the MAF file <code class="docutils literal notranslate"><span class="pre">ucsc_test.maf</span></code> (see
section <a class="reference internal" href="#subsec-align-maf"><span class="std std-ref">Multiple Alignment Format (MAF)</span></a>). To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;ucsc_test.bb&quot;</span><span class="p">,</span> <span class="s2">&quot;bigmaf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">declaration</span><span class="p">)</span>  
<span class="go">table bedMaf</span>
<span class="go">&quot;Bed3 with MAF block&quot;</span>
<span class="go">(</span>
<span class="go">   string  chrom;         &quot;Reference sequence chromosome or scaffold&quot;</span>
<span class="go">   uint    chromStart;    &quot;Start position in chromosome&quot;</span>
<span class="go">   uint    chromEnd;      &quot;End position in chromosome&quot;</span>
<span class="go">   lstring mafBlock;      &quot;MAF block&quot;</span>
<span class="go">)</span>
</pre></div>
</div>
<p>The declaration contains the specification of the columns as defined by
the bigMaf.as AutoSql file from UCSC.</p>
<p>The bigMaf file does not store the header information found in the MAF
file, but it does define a reference genome. The corresponding
<code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> is stored in the <code class="docutils literal notranslate"><span class="pre">targets</span></code> attribute of the
<code class="docutils literal notranslate"><span class="pre">alignments</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">reference</span>
<span class="go">&#39;hg16&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">targets</span>  
<span class="go">[SeqRecord(seq=Seq(None, length=158545518), id=&#39;hg16.chr7&#39;, ...)]</span>
</pre></div>
</div>
<p>By iterating over the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> we obtain one <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object
for each alignment block in the bigMaf file:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span>
<span class="go">23262.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">{</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">}</span>  
<span class="go">{&#39;hg16.chr7&#39;: 158545518,</span>
<span class="go"> &#39;panTro1.chr6&#39;: 161576975,</span>
<span class="go"> &#39;baboon&#39;: 4622798,</span>
<span class="go"> &#39;mm4.chr6&#39;: 151104725,</span>
<span class="go"> &#39;rn3.chr4&#39;: 187371129}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>  
<span class="go">array([[27578828, 27578829, 27578831, 27578831, 27578850, 27578850, 27578866],</span>
<span class="go">       [28741140, 28741141, 28741143, 28741143, 28741162, 28741162, 28741178],</span>
<span class="go">       [  116834,   116835,   116837,   116837,   116856,   116856, 116872],</span>
<span class="go">       [53215344, 53215344, 53215346, 53215347, 53215366, 53215366, 53215382],</span>
<span class="go">       [81344243, 81344243, 81344245, 81344245, 81344264, 81344267, 81344283]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">hg16.chr7  27578828 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG 27578866</span>
<span class="go">panTro1.c  28741140 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG 28741178</span>
<span class="go">baboon       116834 AAA-GGGAATGTTAACCAAATGA---GTTGTCTCTTATGGTG   116872</span>
<span class="go">mm4.chr6   53215344 -AATGGGAATGTTAAGCAAACGA---ATTGTCTCTCAGTGTG 53215382</span>
<span class="go">rn3.chr4   81344243 -AA-GGGGATGCTAAGCCAATGAGTTGTTGTCTCTCAATGTG 81344283</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">))</span>
<span class="go">5 42</span>
<span class="go">hg16.chr7 AAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG</span>
<span class="go">panTro1.chAAA-GGGAATGTTAACCAAATGA---ATTGTCTCTTACGGTG</span>
<span class="go">baboon    AAA-GGGAATGTTAACCAAATGA---GTTGTCTCTTATGGTG</span>
<span class="go">mm4.chr6  -AATGGGAATGTTAAGCAAACGA---ATTGTCTCTCAGTGTG</span>
<span class="go">rn3.chr4  -AA-GGGGATGCTAAGCCAATGAGTTGTTGTCTCTCAATGTG</span>
</pre></div>
</div>
<p>Information in the “<code class="docutils literal notranslate"><span class="pre">i</span></code>”, “<code class="docutils literal notranslate"><span class="pre">e</span></code>”, and “<code class="docutils literal notranslate"><span class="pre">q</span></code>” lines is stored in the
same way as in the corresponding MAF file (see
section <a class="reference internal" href="#subsec-align-maf"><span class="std std-ref">Multiple Alignment Format (MAF)</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;ucsc_mm9_chr10.bb&quot;</span><span class="p">,</span> <span class="s2">&quot;bigmaf&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span>
<span class="go">19159.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">mm9.chr10   3014644 CCTGTACC---CTTTGGTGAGAATTTTTGTTTCAGTGTTAAAAGTTTG   3014689</span>
<span class="go">hg18.chr6 155029206 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT 155029160</span>
<span class="go">panTro2.c 157519257 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT 157519211</span>
<span class="go">calJac1.C      6182 CCTATACCTTTCTTTCATGAGAA-TTTTGTTTGAATCCTAAAC-TTTT      6228</span>
<span class="go">loxAfr1.s      9407 ------------TTTGGTTAGAA-TTATGCTTTAATTCAAAAC-TTCC      9373</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;MAF&quot;</span><span class="p">))</span>
<span class="go">a score=19159.000000</span>
<span class="go">s mm9.chr10                         3014644   45 + 129993255 CCTGTACC---CTTTGGTGAGAATTTTTGTTTCAGTGTTAAAAGTTTG</span>
<span class="go">s hg18.chr6                        15870786   46 - 170899992 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT</span>
<span class="go">i hg18.chr6                        I 9085 C 0</span>
<span class="go">s panTro2.chr6                     16389355   46 - 173908612 CCTATACCTTTCTTTTATGAGAA-TTTTGTTTTAATCCTAAAC-TTTT</span>
<span class="go">q panTro2.chr6                                               99999999999999999999999-9999999999999999999-9999</span>
<span class="go">i panTro2.chr6                     I 9106 C 0</span>
<span class="go">s calJac1.Contig6394                   6182   46 +    133105 CCTATACCTTTCTTTCATGAGAA-TTTTGTTTGAATCCTAAAC-TTTT</span>
<span class="go">i calJac1.Contig6394               N 0 C 0</span>
<span class="go">s loxAfr1.scaffold_75566               1167   34 -     10574 ------------TTTGGTTAGAA-TTATGCTTTAATTCAAAAC-TTCC</span>
<span class="go">q loxAfr1.scaffold_75566                                     ------------99999699899-9999999999999869998-9997</span>
<span class="go">i loxAfr1.scaffold_75566           N 0 C 0</span>
<span class="go">e tupBel1.scaffold_114895.1-498454   167376 4145 -    498454 I</span>
<span class="go">e echTel1.scaffold_288249             87661 7564 +    100002 I</span>
<span class="go">e otoGar1.scaffold_334.1-359464      181217 2931 -    359464 I</span>
<span class="go">e ponAbe2.chr6                     16161448 8044 - 174210431 I</span>


<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;leftStatus&#39;: &#39;I&#39;, &#39;leftCount&#39;: 9085, &#39;rightStatus&#39;: &#39;C&#39;, &#39;rightCount&#39;: 0}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;quality&quot;</span><span class="p">]</span>
<span class="go">&#39;9999999999999999999999999999999999999999999999&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;quality&quot;</span><span class="p">]</span>
<span class="go">&#39;9999969989999999999999998699989997&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;empty&quot;</span><span class="p">]</span>  
<span class="go">[(SeqRecord(seq=Seq(None, length=498454), id=&#39;tupBel1.scaffold_114895.1-498454&#39;, name=&#39;&#39;, description=&#39;&#39;, dbxrefs=[]), (331078, 326933), &#39;I&#39;),</span>
<span class="go"> (SeqRecord(seq=Seq(None, length=100002), id=&#39;echTel1.scaffold_288249&#39;, name=&#39;&#39;, description=&#39;&#39;, dbxrefs=[]), (87661, 95225), &#39;I&#39;),</span>
<span class="go"> (SeqRecord(seq=Seq(None, length=359464), id=&#39;otoGar1.scaffold_334.1-359464&#39;, name=&#39;&#39;, description=&#39;&#39;, dbxrefs=[]), (178247, 175316), &#39;I&#39;),</span>
<span class="go"> (SeqRecord(seq=Seq(None, length=174210431), id=&#39;ponAbe2.chr6&#39;, name=&#39;&#39;, description=&#39;&#39;, dbxrefs=[]), (158048983, 158040939), &#39;I&#39;)]</span>
</pre></div>
</div>
<p>To write a complete bigMaf file, use
<code class="docutils literal notranslate"><span class="pre">Bio.Align.write(alignments,</span> <span class="pre">&quot;myfilename.bb&quot;,</span> <span class="pre">fmt=&quot;bigMaf&quot;)</span></code>, where
<code class="docutils literal notranslate"><span class="pre">myfilename.bb</span></code> is the name of the output bigMaf file. Alternatively,
you can use a (binary) stream for output. If you are creating the
alignments from scratch, you can use the <code class="docutils literal notranslate"><span class="pre">Alignments</span></code> (plural) class
to create a list-like <code class="docutils literal notranslate"><span class="pre">alignments</span></code> object (see
Section <a class="reference internal" href="#sec-alignments"><span class="std std-ref">The Alignments class</span></a>) and give it a <code class="docutils literal notranslate"><span class="pre">targets</span></code> attribute.
The latter must be a list of <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> objects for the chromosomes
for the reference species in the order in which they appear in the
alignments. Alternatively, you can use the <code class="docutils literal notranslate"><span class="pre">targets</span></code> keyword argument
when calling <code class="docutils literal notranslate"><span class="pre">Bio.Align.write</span></code>. The <code class="docutils literal notranslate"><span class="pre">id</span></code> of each <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> must
be of the form <code class="docutils literal notranslate"><span class="pre">reference.chromosome</span></code>, where <code class="docutils literal notranslate"><span class="pre">reference</span></code> refers to
the reference species. <code class="docutils literal notranslate"><span class="pre">Bio.Align.write</span></code> has the additional keyword
argument <code class="docutils literal notranslate"><span class="pre">compress</span></code> (<code class="docutils literal notranslate"><span class="pre">True</span></code> by default) specifying whether the data
should be compressed using zlib.</p>
<p>As a bigMaf file is a special case of a bigBed file, you can use the
<code class="docutils literal notranslate"><span class="pre">search</span></code> method on the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> object to find alignments to
specific regions of the reference species. For example, we can look for
regions on chr10 between positions 3018000 and 3019000 on chromosome 10:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selected_alignments</span> <span class="o">=</span> <span class="n">alignments</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;mm9.chr10&quot;</span><span class="p">,</span> <span class="mi">3018000</span><span class="p">,</span> <span class="mi">3019000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">selected_alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">3017743 3018161</span>
<span class="go">3018161 3018230</span>
<span class="go">3018230 3018359</span>
<span class="go">3018359 3018482</span>
<span class="go">3018482 3018644</span>
<span class="go">3018644 3018822</span>
<span class="go">3018822 3018932</span>
<span class="go">3018932 3019271</span>
</pre></div>
</div>
<p>The chromosome name may be <code class="docutils literal notranslate"><span class="pre">None</span></code> to include all chromosomes, and the
start and end positions may be <code class="docutils literal notranslate"><span class="pre">None</span></code> to start searching from position
0 or to continue searching until the end of the chromosome,
respectively. Note that we can search on genomic position for the
reference species only.</p>
</section>
<section id="ucsc-chain-file-format">
<span id="subsec-align-chain"></span><h3>UCSC chain file format<a class="headerlink" href="#ucsc-chain-file-format" title="Link to this heading"></a></h3>
<p>Chain files describe a pairwise alignment between two nucleotide
sequences, allowing gaps in both sequences. Only the length of each
aligned subsequences and the gap lengths are stored in a chain file; the
sequences themselves are not stored. Chain files are typically used to
store alignments between two genome assembly versions, allowing
alignments to one genome assembly version to be lifted over to the other
genome assembly. This is an example of a chain file (available as
<code class="docutils literal notranslate"><span class="pre">psl_34_001.chain</span></code> in the <code class="docutils literal notranslate"><span class="pre">Tests/Blat</span></code> subdirectory of the Biopython
distribution):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>chain 16 chr4 191154276 + 61646095 61646111 hg18_dna 33 + 11 27 1
16
chain 33 chr1 249250621 + 10271783 10271816 hg18_dna 33 + 0 33 2
33
chain 17 chr2 243199373 + 53575980 53575997 hg18_dna 33 - 8 25 3
17
chain 35 chr9 141213431 + 85737865 85737906 hg19_dna 50 + 9 50 4
41
chain 41 chr8 146364022 + 95160479 95160520 hg19_dna 50 + 8 49 5
41
chain 30 chr22 51304566 + 42144400 42144436 hg19_dna 50 + 11 47 6
36
chain 41 chr2 243199373 + 183925984 183926028 hg19_dna 50 + 1 49 7
6       0       4
38
chain 31 chr19 59128983 + 35483340 35483510 hg19_dna 50 + 10 46 8
25      134     0
11
chain 39 chr18 78077248 + 23891310 23891349 hg19_dna 50 + 10 49 9
39
...
</pre></div>
</div>
<p>This file was generated by running UCSC’s <code class="docutils literal notranslate"><span class="pre">pslToChain</span></code> program on the
PSL file <code class="docutils literal notranslate"><span class="pre">psl_34_001.psl</span></code>. According to the chain file format
specification, there should be a blank line after each chain block, but
some tools (including <code class="docutils literal notranslate"><span class="pre">pslToChain</span></code>) apparently do not follow this
rule.</p>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;psl_34_001.chain&quot;</span><span class="p">,</span> <span class="s2">&quot;chain&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Iterate over alignments to get one <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object for each chain:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">alignment</span><span class="o">.</span><span class="n">query</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>  
<span class="gp">...</span>
<span class="go">chr4 hg18_dna</span>
<span class="go">chr1 hg18_dna</span>
<span class="go">chr2 hg18_dna</span>
<span class="go">chr9 hg19_dna</span>
<span class="go">chr8 hg19_dna</span>
<span class="go">chr22 hg19_dna</span>
<span class="go">chr2 hg19_dna</span>
<span class="go">...</span>
<span class="go">chr1 hg19_dna</span>
</pre></div>
</div>
<p>Iterate from the start until we reach the seventh alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">7</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>Check the alignment score and chain ID (the first and last number,
respectively, in the header line of each chain block) to confirm that we
got the seventh alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span>
<span class="go">41.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
<span class="go">&#39;7&#39;</span>
</pre></div>
</div>
<p>We can print the alignment in the chain file format. The alignment
coordinates are consistent with the information in the chain block, with
an aligned section of 6 nucleotides, a gap of 4 nucleotides, and an
aligned section of 38 nucleotides:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;chain&quot;</span><span class="p">))</span>  
<span class="go">chain 41 chr2 243199373 + 183925984 183926028 hg19_dna 50 + 1 49 7</span>
<span class="go">6   0   4</span>
<span class="go">38</span>


<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[183925984, 183925990, 183925990, 183926028],</span>
<span class="go">       [        1,         7,        11,        49]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">chr2      183925984 ??????----?????????????????????????????????????? 183926028</span>
<span class="go">                  0 ||||||----||||||||||||||||||||||||||||||||||||||        48</span>
<span class="go">hg19_dna          1 ????????????????????????????????????????????????        49</span>
</pre></div>
</div>
<p>We can also print the alignment in a few other alignment fite formats:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;BED&quot;</span><span class="p">))</span>  
<span class="go">chr2    183925984   183926028   hg19_dna    41  +   183925984   183926028   0   2   6,38,   0,6,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;PSL&quot;</span><span class="p">))</span>  
<span class="go">44  0   0   0   1   4   0   0   +   hg19_dna    50  1   49  chr2    243199373   183925984   183926028   2   6,38,   1,11,   183925984,183925990,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;exonerate&quot;</span><span class="p">))</span>
<span class="go">vulgar: hg19_dna 1 49 + chr2 183925984 183926028 + 41 M 6 6 G 4 0 M 38 38</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;exonerate&quot;</span><span class="p">,</span> <span class="s2">&quot;cigar&quot;</span><span class="p">))</span>
<span class="go">cigar: hg19_dna 1 49 + chr2 183925984 183926028 + 41 M 6 I 4 M 38</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s2">&quot;sam&quot;</span><span class="p">))</span>  
<span class="go">hg19_dna    0   chr2    183925985   255 1S6M4I38M1S *   0   0   *   *   AS:i:41 id:A:7</span>
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="chapter_seqio.html" class="btn btn-neutral float-left" title="Sequence Input/Output" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_pairwise.html" class="btn btn-neutral float-right" title="Pairwise sequence alignment" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>