<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>KEGG &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Bio.phenotype: analyze phenotypic data" href="chapter_phenotype.html" />
    <link rel="prev" title="Graphics including GenomeDiagram" href="chapter_graphics.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Biopython Tutorial &amp; Cookbook</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="chapter_introduction.html">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_quick_start.html">Quick Start – What can you do with Biopython?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_objects.html">Sequence objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_annot.html">Sequence annotation objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seqio.html">Sequence Input/Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_align.html">Sequence alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise.html">Pairwise sequence alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_msa.html">Multiple Sequence Alignment objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise2.html">Pairwise alignments using pairwise2</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html">BLAST (new)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html#blast-old">BLAST (old)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_searchio.html">BLAST and other sequence search tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_entrez.html">Accessing NCBI’s Entrez databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_uniprot.html">Swiss-Prot and ExPASy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pdb.html">Going 3D: The PDB module</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_popgen.html">Bio.PopGen: Population genetics</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phylo.html">Phylogenetics with Bio.Phylo</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_motifs.html">Sequence motif analysis using Bio.motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cluster.html">Cluster analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_graphics.html">Graphics including GenomeDiagram</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">KEGG</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#parsing-kegg-records">Parsing KEGG records</a></li>
<li class="toctree-l3"><a class="reference internal" href="#querying-the-kegg-api">Querying the KEGG API</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phenotype.html">Bio.phenotype: analyze phenotypic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cookbook.html">Cookbook – Cool things to do with it</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_testing.html">The Biopython testing framework</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_contributing.html">Where to go from here – contributing to Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_appendix.html">Appendix: Useful stuff about Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_bibliography.html">Bibliography</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Biopython Tutorial &amp; Cookbook</a></li>
      <li class="breadcrumb-item active">KEGG</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/Tutorial/chapter_kegg.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="chapter_graphics.html" class="btn btn-neutral float-left" title="Graphics including GenomeDiagram" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_phenotype.html" class="btn btn-neutral float-right" title="Bio.phenotype: analyze phenotypic data" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="kegg">
<span id="chapter-kegg"></span><h1>KEGG<a class="headerlink" href="#kegg" title="Permalink to this headline"></a></h1>
<p>KEGG (<a class="reference external" href="https://www.kegg.jp/">https://www.kegg.jp/</a>) is a database resource for understanding
high-level functions and utilities of the biological system, such as the
cell, the organism and the ecosystem, from molecular-level information,
especially large-scale molecular datasets generated by genome sequencing
and other high-throughput experimental technologies.</p>
<p>Please note that the KEGG parser implementation in Biopython is
incomplete. While the KEGG website indicates many flat file formats,
only parsers and writers for compound, enzyme, and map are currently
implemented. However, a generic parser is implemented to handle the
other formats.</p>
<section id="parsing-kegg-records">
<h2>Parsing KEGG records<a class="headerlink" href="#parsing-kegg-records" title="Permalink to this headline"></a></h2>
<p>Parsing a KEGG record is as simple as using any other file format parser
in Biopython. (Before running the following codes, please open
<a class="reference external" href="http://rest.kegg.jp/get/ec:5.4.2.2">http://rest.kegg.jp/get/ec:5.4.2.2</a> with your web browser and save it as
<code class="docutils literal notranslate"><span class="pre">ec_5.4.2.2.txt</span></code>.)</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.KEGG</span> <span class="kn">import</span> <span class="n">Enzyme</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Enzyme</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;ec_5.4.2.2.txt&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">records</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">classname</span>
<span class="go">[&#39;Isomerases;&#39;, &#39;Intramolecular transferases;&#39;, &#39;Phosphotransferases (phosphomutases)&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">entry</span>
<span class="go">&#39;5.4.2.2&#39;</span>
</pre></div>
</div>
<p>Alternatively, if the input KEGG file has exactly one entry, you can use
<code class="docutils literal notranslate"><span class="pre">read</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.KEGG</span> <span class="kn">import</span> <span class="n">Enzyme</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Enzyme</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;ec_5.4.2.2.txt&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">classname</span>
<span class="go">[&#39;Isomerases;&#39;, &#39;Intramolecular transferases;&#39;, &#39;Phosphotransferases (phosphomutases)&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">entry</span>
<span class="go">&#39;5.4.2.2&#39;</span>
</pre></div>
</div>
<p>The following section will shows how to download the above enzyme using
the KEGG api as well as how to use the generic parser with data that
does not have a custom parser implemented.</p>
</section>
<section id="querying-the-kegg-api">
<h2>Querying the KEGG API<a class="headerlink" href="#querying-the-kegg-api" title="Permalink to this headline"></a></h2>
<p>Biopython has full support for the querying of the KEGG api. Querying
all KEGG endpoints are supported; all methods documented by KEGG
(<a class="reference external" href="https://www.kegg.jp/kegg/rest/keggapi.html">https://www.kegg.jp/kegg/rest/keggapi.html</a>) are supported. The
interface has some validation of queries which follow rules defined on
the KEGG site. However, invalid queries which return a 400 or 404 must
be handled by the user.</p>
<p>First, here is how to extend the above example by downloading the
relevant enzyme and passing it through the Enzyme parser.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.KEGG</span> <span class="kn">import</span> <span class="n">REST</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.KEGG</span> <span class="kn">import</span> <span class="n">Enzyme</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">request</span> <span class="o">=</span> <span class="n">REST</span><span class="o">.</span><span class="n">kegg_get</span><span class="p">(</span><span class="s2">&quot;ec:5.4.2.2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;ec_5.4.2.2.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">request</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Enzyme</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;ec_5.4.2.2.txt&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">records</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">classname</span>
<span class="go">[&#39;Isomerases;&#39;, &#39;Intramolecular transferases;&#39;, &#39;Phosphotransferases (phosphomutases)&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">entry</span>
<span class="go">&#39;5.4.2.2&#39;</span>
</pre></div>
</div>
<p>Now, here’s a more realistic example which shows a combination of
querying the KEGG API. This will demonstrate how to extract a unique set
of all human pathway gene symbols which relate to DNA repair. The steps
that need to be taken to do so are as follows. First, we need to get a
list of all human pathways. Secondly, we need to filter those for ones
which relate to “repair”. Lastly, we need to get a list of all the gene
symbols in all repair pathways.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.KEGG</span> <span class="kn">import</span> <span class="n">REST</span>

<span class="n">human_pathways</span> <span class="o">=</span> <span class="n">REST</span><span class="o">.</span><span class="n">kegg_list</span><span class="p">(</span><span class="s2">&quot;pathway&quot;</span><span class="p">,</span> <span class="s2">&quot;hsa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

<span class="c1"># Filter all human pathways for repair pathways</span>
<span class="n">repair_pathways</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">human_pathways</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">):</span>
    <span class="n">entry</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;repair&quot;</span> <span class="ow">in</span> <span class="n">description</span><span class="p">:</span>
        <span class="n">repair_pathways</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">entry</span><span class="p">)</span>

<span class="c1"># Get the genes for pathways and add them to a list</span>
<span class="n">repair_genes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">repair_pathways</span><span class="p">:</span>
    <span class="n">pathway_file</span> <span class="o">=</span> <span class="n">REST</span><span class="o">.</span><span class="n">kegg_get</span><span class="p">(</span><span class="n">pathway</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>  <span class="c1"># query and read each pathway</span>

    <span class="c1"># iterate through each KEGG pathway file, keeping track of which section</span>
    <span class="c1"># of the file we&#39;re in, only read the gene in each pathway</span>
    <span class="n">current_section</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">pathway_file</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">):</span>
        <span class="n">section</span> <span class="o">=</span> <span class="n">line</span><span class="p">[:</span><span class="mi">12</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>  <span class="c1"># section names are within 12 columns</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">section</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">current_section</span> <span class="o">=</span> <span class="n">section</span>

        <span class="k">if</span> <span class="n">current_section</span> <span class="o">==</span> <span class="s2">&quot;GENE&quot;</span><span class="p">:</span>
            <span class="n">gene_identifiers</span><span class="p">,</span> <span class="n">gene_description</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">12</span><span class="p">:]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;; &quot;</span><span class="p">)</span>
            <span class="n">gene_id</span><span class="p">,</span> <span class="n">gene_symbol</span> <span class="o">=</span> <span class="n">gene_identifiers</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_symbol</span> <span class="ow">in</span> <span class="n">repair_genes</span><span class="p">:</span>
                <span class="n">repair_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;There are </span><span class="si">%d</span><span class="s2"> repair pathways and </span><span class="si">%d</span><span class="s2"> repair genes. The genes are:&quot;</span>
    <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">repair_pathways</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">repair_genes</span><span class="p">))</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">repair_genes</span><span class="p">))</span>
</pre></div>
</div>
<p>The KEGG API wrapper is compatible with all endpoints. Usage is
essentially replacing all slashes in the url with commas and using that
list as arguments to the corresponding method in the KEGG module. Here
are a few examples from the api documentation
(<a class="reference external" href="https://www.kegg.jp/kegg/docs/keggapi.html">https://www.kegg.jp/kegg/docs/keggapi.html</a>).</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>/list/hsa:10458+ece:Z5100            -&gt; REST.kegg_list([&quot;hsa:10458&quot;, &quot;ece:Z5100&quot;])
/find/compound/300-310/mol_weight    -&gt; REST.kegg_find(&quot;compound&quot;, &quot;300-310&quot;, &quot;mol_weight&quot;)
/get/hsa:10458+ece:Z5100/aaseq      -&gt; REST.kegg_get([&quot;hsa:10458&quot;, &quot;ece:Z5100&quot;], &quot;aaseq&quot;)
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="chapter_graphics.html" class="btn btn-neutral float-left" title="Graphics including GenomeDiagram" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_phenotype.html" class="btn btn-neutral float-right" title="Bio.phenotype: analyze phenotypic data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>