<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bibliography &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="API documentation" href="../api/index.html" />
    <link rel="prev" title="Appendix: Useful stuff about Python" href="chapter_appendix.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Biopython Tutorial &amp; Cookbook</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="chapter_introduction.html">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_quick_start.html">Quick Start – What can you do with Biopython?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_objects.html">Sequence objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_annot.html">Sequence annotation objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seqio.html">Sequence Input/Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_align.html">Sequence alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise.html">Pairwise sequence alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_msa.html">Multiple Sequence Alignment objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise2.html">Pairwise alignments using pairwise2</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html">BLAST (new)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html#blast-old">BLAST (old)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_searchio.html">BLAST and other sequence search tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_entrez.html">Accessing NCBI’s Entrez databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_uniprot.html">Swiss-Prot and ExPASy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pdb.html">Going 3D: The PDB module</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_popgen.html">Bio.PopGen: Population genetics</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phylo.html">Phylogenetics with Bio.Phylo</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_motifs.html">Sequence motif analysis using Bio.motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cluster.html">Cluster analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_graphics.html">Graphics including GenomeDiagram</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_kegg.html">KEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phenotype.html">Bio.phenotype: analyze phenotypic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cookbook.html">Cookbook – Cool things to do with it</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_testing.html">The Biopython testing framework</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_contributing.html">Where to go from here – contributing to Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_appendix.html">Appendix: Useful stuff about Python</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Bibliography</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Biopython Tutorial &amp; Cookbook</a></li>
      <li class="breadcrumb-item active">Bibliography</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/Tutorial/chapter_bibliography.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="chapter_appendix.html" class="btn btn-neutral float-left" title="Appendix: Useful stuff about Python" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../api/index.html" class="btn btn-neutral float-right" title="API documentation" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bibliography">
<span id="chapter-bibliography"></span><h1>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline"></a></h1>
<dl class="citation">
<dt class="label" id="altschul1990"><span class="brackets">Altschul1990</span></dt>
<dd><p>Stephen F. Altschul, Warren Gish, Webb Miller, Eugene W. Myers, David J. Lipman: Basic Local Alignment Search Tool. <em>Journal of Molecular Biology</em> <strong>215</strong> (3): 403–410 (1990). <a class="reference external" href="https://doi.org/10.1016/S0022-2836(05)80360-2">https://doi.org/10.1016/S0022-2836(05)80360-2</a></p>
</dd>
<dt class="label" id="bailey1994"><span class="brackets">Bailey1994</span></dt>
<dd><p>Timothy L. Bailey and Charles Elkan: Fitting a mixture model by expectation maximization to discover motifs in biopolymers, <em>Proceedings of the Second International Conference on Intelligent Systems for Molecular Biology</em> 28–36. AAAI Press, Menlo Park, California (1994).</p>
</dd>
<dt class="label" id="cavener1987"><span class="brackets">Cavener1987</span></dt>
<dd><p>Douglas R. Cavener: Comparison of the consensus sequence flanking translational start sites in Drosophila and vertebrates. <em>Nucleic Acids Research</em> <strong>15</strong> (4): 1353–1361 (1987). <a class="reference external" href="https://doi.org/10.1093/nar/15.4.1353">https://doi.org/10.1093/nar/15.4.1353</a></p>
</dd>
<dt class="label" id="chapman2000"><span class="brackets">Chapman2000</span></dt>
<dd><p>Brad Chapman and Jeff Chang: Biopython: Python tools for computational biology. <em>ACM SIGBIO Newsletter</em> <strong>20</strong> (2): 15–19 (August 2000).</p>
</dd>
<dt class="label" id="cock2009"><span class="brackets">Cock2009</span></dt>
<dd><p>Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman, Cymon J. Cox, Andrew Dalke, Iddo Friedberg, Thomas Hamelryck, Frank Kauff, Bartek Wilczyński, Michiel J. L. de Hoon: Biopython: freely available Python tools for computational molecular biology and bioinformatics. <em>Bioinformatics</em> <strong>25</strong> (11), 1422–1423 (2009). <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp163">https://doi.org/10.1093/bioinformatics/btp163</a></p>
</dd>
<dt class="label" id="cock2010"><span class="brackets">Cock2010</span></dt>
<dd><p>Peter J. A. Cock, Christopher J. Fields, Naohisa Goto, Michael L. Heuer, Peter M. Rice: The Sanger FASTQ file format for sequences with quality scores, and the Solexa/Illumina FASTQ variants. <em>Nucleic Acids Research</em> <strong>38</strong> (6): 1767–1771 (2010). <a class="reference external" href="https://doi.org/10.1093/nar/gkp1137">https://doi.org/10.1093/nar/gkp1137</a></p>
</dd>
<dt class="label" id="cornish1985"><span class="brackets">Cornish1985</span></dt>
<dd><p>Athel Cornish-Bowden: Nomenclature for incompletely specified bases in nucleic acid sequences: Recommendations 1984. <em>Nucleic Acids Research</em> <strong>13</strong> (9): 3021–3030 (1985). <a class="reference external" href="https://doi.org/10.1093/nar/13.9.3021">https://doi.org/10.1093/nar/13.9.3021</a></p>
</dd>
<dt class="label" id="darling2004"><span class="brackets">Darling2004</span></dt>
<dd><p>Aaron E. Darling, Bob Mau, Frederick R. Blattner, Nicole T. Perna: Mauve: Multiple alignment of conserved genomic sequence with rearrangements. <em>Genome Research</em> <strong>14</strong> (7): 1394–1403 (2004). <a class="reference external" href="https://doi.org/10.1101/gr.2289704">https://doi.org/10.1101/gr.2289704</a></p>
</dd>
<dt class="label" id="dayhoff1978"><span class="brackets">Dayhoff1978</span></dt>
<dd><p>M.O. Dayhoff, R.M. Schwartz, and B.C. Orcutt: A Model of Evolutionary Change in Proteins. <em>Atlas of Protein Sequence and Structure</em>, Volume 5, Supplement 3, 1978: 345–352. The National Biomedical Research Foundation, 1979.</p>
</dd>
<dt class="label" id="dehoon2004"><span class="brackets">DeHoon2004</span></dt>
<dd><p>Michiel J. L. de Hoon, Seiya Imoto, John Nolan, Satoru Miyano: Open source clustering software. <em>Bioinformatics</em> <strong>20</strong> (9): 1453–1454 (2004). <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bth078">https://doi.org/10.1093/bioinformatics/bth078</a></p>
</dd>
<dt class="label" id="durbin1998"><span class="brackets">Durbin1998</span></dt>
<dd><p>Richard Durbin, Sean R. Eddy, Anders Krogh, Graeme Mitchison: Biological sequence analysis: Probabilistic models of proteins and nucleic acids. Cambridge University Press, Cambridge, UK (1998).</p>
</dd>
<dt class="label" id="eisen1998"><span class="brackets">Eisen1998</span></dt>
<dd><p>Michiel B. Eisen, Paul T. Spellman, Patrick O. Brown, David Botstein: Cluster analysis and display of genome-wide expression patterns. <em>Proceedings of the National Academy of Sciences USA</em> <strong>95</strong> (25): 14863–14868 (1998). <a class="reference external" href="https://doi.org/10.1073/pnas.95.25.14863">https://doi.org/10.1073/pnas.95.25.14863</a></p>
</dd>
<dt class="label" id="goldman1994"><span class="brackets">Goldman1994</span></dt>
<dd><p>Nick Goldman and Ziheng Yang: A codon-based model of nucleotide substitution for protein-coding DNA sequences. <em>Molecular Biology and Evolution</em> <strong>11</strong> (5) 725–736 (1994). <a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a040153">https://doi.org/10.1093/oxfordjournals.molbev.a040153</a></p>
</dd>
<dt class="label" id="golub1971"><span class="brackets">Golub1971</span></dt>
<dd><p>Gene H. Golub, Christian Reinsch: Singular value decomposition and least squares solutions. In <em>Handbook for Automatic Computation</em>, <strong>2</strong>, (Linear Algebra) (J. H. Wilkinson and C. Reinsch, eds), 134–151. New York: Springer-Verlag (1971).</p>
</dd>
<dt class="label" id="golub1989"><span class="brackets">Golub1989</span></dt>
<dd><p>Gene H. Golub, Charles F. Van Loan: <em>Matrix computations</em>, 2nd edition (1989).</p>
</dd>
<dt class="label" id="hamelryck2003a"><span class="brackets">Hamelryck2003A</span></dt>
<dd><p>Thomas Hamelryck and Bernard Manderick: PDB parser and structure class implemented in Python. <em>Bioinformatics</em> <strong>19</strong> (17): 2308–2310 (2003) <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btg299">https://doi.org/10.1093/bioinformatics/btg299</a></p>
</dd>
<dt class="label" id="hamelryck2003b"><span class="brackets">Hamelryck2003B</span></dt>
<dd><p>Thomas Hamelryck: Efficient identification of side-chain patterns using a multidimensional index tree. <em>Proteins</em> <strong>51</strong> (1): 96–108 (2003).  <a class="reference external" href="https://doi.org/10.1002/prot.10338">https://doi.org/10.1002/prot.10338</a></p>
</dd>
<dt class="label" id="hamelryck2005"><span class="brackets">Hamelryck2005</span></dt>
<dd><p>Thomas Hamelryck: An amino acid has two sides; A new 2D measure provides a different view of solvent exposure. <em>Proteins</em> <strong>59</strong> (1): 29–48 (2005).  <a class="reference external" href="https://doi.org/10.1002/prot.20379">https://doi.org/10.1002/prot.20379</a></p>
</dd>
<dt class="label" id="henikoff1992"><span class="brackets">Henikoff1992</span></dt>
<dd><p>Steven Henikoff, Jorja G. Henikoff: Amino acid substitution matrices from protein blocks. <em>Proceedings of the National Academy of Sciences USA</em> <strong>89</strong> (2): 10915–10919 (1992). <a class="reference external" href="https://doi.org/10.1073/pnas.89.22.10915">https://doi.org/10.1073/pnas.89.22.10915</a></p>
</dd>
<dt class="label" id="hihara2001"><span class="brackets">Hihara2001</span></dt>
<dd><p>Yukako Hihara, Ayako Kamei, Minoru Kanehisa, Aaron Kaplan and Masahiko Ikeuchi: DNA microarray analysis of cyanobacterial gene expression during acclimation to high light. <em>Plant Cell</em> <strong>13</strong> (4): 793–806 (2001). <a class="reference external" href="https://doi.org/10.1105/tpc.13.4.793">https://doi.org/10.1105/tpc.13.4.793</a></p>
</dd>
<dt class="label" id="hughey1996"><span class="brackets">Hughey1996</span></dt>
<dd><p>Richard Hughey, Anders Krogh: Hidden Markov models for sequence analysis: extension and analysis of the basic method. <em>Computer Applications in the Biosciences: CABIOS</em> <strong>12</strong> (2): 95–107 (1996).  <a class="reference external" href="https://doi.org/10.1093/bioinformatics/12.2.95">https://doi.org/10.1093/bioinformatics/12.2.95</a></p>
</dd>
<dt class="label" id="jupe2012"><span class="brackets">Jupe2012</span></dt>
<dd><p>Florian Jupe, Leighton Pritchard, Graham J. Etherington, Katrin MacKenzie, Peter JA Cock, Frank Wright, Sanjeev Kumar Sharma, Dan Bolser, Glenn J Bryan, Jonathan DG Jones, Ingo Hein: Identification and localisation of the NB-LRR gene family within the potato genome. <em>BMC Genomics</em> <strong>13</strong>: 75 (2012).  <a class="reference external" href="https://doi.org/10.1186/1471-2164-13-75">https://doi.org/10.1186/1471-2164-13-75</a></p>
</dd>
<dt class="label" id="kachitvichyanukul1988"><span class="brackets">Kachitvichyanukul1988</span></dt>
<dd><p>Voratas Kachitvichyanukul, Bruce W. Schmeiser: Binomial Random Variate Generation. <em>Communications of the ACM</em> <strong>31</strong> (2): 216–222 (1988). <a class="reference external" href="https://doi.org/10.1145/42372.42381">https://doi.org/10.1145/42372.42381</a></p>
</dd>
<dt class="label" id="kent2002"><span class="brackets">Kent2002</span></dt>
<dd><p>W. James Kent: BLAT – The BLAST-Like Alignment Tool. <em>Genome Research</em> <strong>12</strong>: 656–664 (2002). <a class="reference external" href="https://doi.org/10.1101/gr.229202">https://doi.org/10.1101/gr.229202</a></p>
</dd>
<dt class="label" id="kohonen1997"><span class="brackets">Kohonen1997</span></dt>
<dd><p>Teuvo Kohonen: Self-organizing maps, 2nd Edition. Berlin; New York: Springer-Verlag (1997).</p>
</dd>
<dt class="label" id="krogh1994"><span class="brackets">Krogh1994</span></dt>
<dd><p>Anders Krogh, Michael Brown, I. Saira Mian, Kimmen Sjölander, David Haussler: Hidden Markov Models in computational biology: Applications to protein modeling. <em>Journal of Molecular Biology</em> <strong>235</strong> (5): 1501–1531 (1994). <a class="reference external" href="https://doi.org/10.1006/jmbi.1994.1104">https://doi.org/10.1006/jmbi.1994.1104</a></p>
</dd>
<dt class="label" id="lecuyer1988"><span class="brackets">Lecuyer1988</span></dt>
<dd><p>Pierre L’Ecuyer: Efficient and Portable Combined Random Number Generators. <em>Communications of the ACM</em> <strong>31</strong> (6): 742–749,774 (1988). <a class="reference external" href="https://doi.org/10.1145/62959.62969">https://doi.org/10.1145/62959.62969</a></p>
</dd>
<dt class="label" id="li1985"><span class="brackets">Li1985</span></dt>
<dd><p>Wen-Hsiung Li, Chung-I Wu, Chi-Cheng Luo: A new method for estimating synonymous and nonsynonymous rates of nucleotide substitution considering the relative likelihood of nucleotide and codon changes. <em>Molecular Biology and Evolution</em> <strong>2</strong> (2): 150–174 (1985). <a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a040343">https://doi.org/10.1093/oxfordjournals.molbev.a040343</a></p>
</dd>
<dt class="label" id="li2009"><span class="brackets">Li2009</span></dt>
<dd><p>Heng Li, Bob Handsaker, Alec Wysoker, Tim Fennell, Jue Ruan, Nils Homer, Gabor Marth, Goncalo Abecasis, Richard Durbin: The Sequence Alignment/Map format and SAMtools. <em>Bioinformatics</em> <strong>25</strong> (16): 2078–2079 (2009). <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp352">https://doi.org/10.1093/bioinformatics/btp352</a></p>
</dd>
<dt class="label" id="maddison1997"><span class="brackets">Maddison1997</span></dt>
<dd><p>David R. Maddison, David L. Swofford, Wayne P. Maddison: Nexus: An Extensible File Format for Systematic Information. <em>Systematic Biology</em> <strong>46</strong> (4): 590–621 (1997).  <a class="reference external" href="https://doi.org/10.1093/sysbio/46.4.590">https://doi.org/10.1093/sysbio/46.4.590</a></p>
</dd>
<dt class="label" id="majumdar2005"><span class="brackets">Majumdar2005</span></dt>
<dd><p>Indraneel Majumdar, S. Sri Krishna, Nick V. Grishin: PALSSE: A program to delineate linear secondary structural elements from protein structures. <em>BMC Bioinformatics</em> <strong>6</strong>: 202 (2005). <a class="reference external" href="https://doi.org/10.1186/1471-2105-6-202">https://doi.org/10.1186/1471-2105-6-202</a>.</p>
</dd>
<dt class="label" id="matys2003"><span class="brackets">Matys2003</span></dt>
<dd><p>V. Matys, E. Fricke, R. Geffers, E. Gößling, M. Haubrock, R. Hehl, K. Hornischer, D. Karas, A.E. Kel, O.V. Kel-Margoulis, D.U. Kloos, S. Land, B. Lewicki-Potapov, H. Michael, R. Münch, I. Reuter, S. Rotert, H. Saxel, M. Scheer, S. Thiele, E. Wingender E: TRANSFAC: transcriptional regulation, from patterns to profiles. <em>Nucleic Acids Research</em> <strong>31</strong> (1): 374–378 (2003).  <a class="reference external" href="https://doi.org/10.1093/nar/gkg108">https://doi.org/10.1093/nar/gkg108</a></p>
</dd>
<dt class="label" id="nei1986"><span class="brackets">Nei1986</span></dt>
<dd><p>Masatoshi Nei and Takashi Gojobori: Simple methods for estimating the numbers of synonymous and nonsynonymous nucleotide substitutions. <em>Molecular Biology and Evolution</em> <strong>3</strong> (5): 418–426 (1986). <a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a040410">https://doi.org/10.1093/oxfordjournals.molbev.a040410</a></p>
</dd>
<dt class="label" id="pearson1988"><span class="brackets">Pearson1988</span></dt>
<dd><p>William R. Pearson, David J. Lipman: Improved tools for biological sequence comparison. <em>Proceedings of the National Academy of Sciences USA</em> <strong>85</strong> (8): 2444–2448 (1988). <a class="reference external" href="https://doi.org/10.1073/pnas.85.8.2444">https://doi.org/10.1073/pnas.85.8.2444</a></p>
</dd>
<dt class="label" id="pritchard2006"><span class="brackets">Pritchard2006</span></dt>
<dd><p>Leighton Pritchard, Jennifer A. White, Paul R.J. Birch, Ian K. Toth: GenomeDiagram: a python package for the visualization of large-scale genomic data. <em>Bioinformatics</em> <strong>22</strong> (5): 616–617 (2006). <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btk021">https://doi.org/10.1093/bioinformatics/btk021</a></p>
</dd>
<dt class="label" id="proux2002"><span class="brackets">Proux2002</span></dt>
<dd><p>Caroline Proux, Douwe van Sinderen, Juan Suarez, Pilar Garcia, Victor Ladero, Gerald F. Fitzgerald, Frank Desiere, Harald Brüssow: The dilemma of phage taxonomy illustrated by comparative genomics of Sfi21-Like Siphoviridae in lactic acid bacteria. <em>Journal of Bacteriology</em> <strong>184</strong> (21): 6026–6036 (2002). <a class="reference external" href="https://doi.org/10.1128/JB.184.21.6026-6036.2002">https://doi.org/10.1128/JB.184.21.6026-6036.2002</a></p>
</dd>
<dt class="label" id="rice2000"><span class="brackets">Rice2000</span></dt>
<dd><p>Peter Rice, Ian Longden, Alan Bleasby: EMBOSS: The European Molecular Biology Open Software Suite. <em>Trends in Genetics</em> <strong>16</strong> (6): 276–277 (2000). <a class="reference external" href="https://doi.org/10.1016/S0168-9525(00)02024-2">https://doi.org/10.1016/S0168-9525(00)02024-2</a></p>
</dd>
<dt class="label" id="saldanha2004"><span class="brackets">Saldanha2004</span></dt>
<dd><p>Alok Saldanha: Java Treeview—extensible visualization of microarray data. <em>Bioinformatics</em> <strong>20</strong> (17): 3246–3248 (2004). <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bth349">https://doi.org/10.1093/bioinformatics/bth349</a></p>
</dd>
<dt class="label" id="schneider1986"><span class="brackets">Schneider1986</span></dt>
<dd><p>Thomas D. Schneider, Gary D. Stormo, Larry Gold: Information content of binding sites on nucleotide sequences. <em>Journal of Molecular Biology</em> <strong>188</strong> (3): 415–431 (1986). <a class="reference external" href="https://doi.org/10.1016/0022-2836(86)90165-8">https://doi.org/10.1016/0022-2836(86)90165-8</a></p>
</dd>
<dt class="label" id="schneider2005"><span class="brackets">Schneider2005</span></dt>
<dd><p>Adrian Schneider, Gina M. Cannarozzi, and Gaston H. Gonnet: Empirical codon substitution matrix. <em>BMC Bioinformatics</em> <strong>6</strong>: 134 (2005). <a class="reference external" href="https://doi.org/10.1186/1471-2105-6-134">https://doi.org/10.1186/1471-2105-6-134</a></p>
</dd>
<dt class="label" id="sibson1973"><span class="brackets">Sibson1973</span></dt>
<dd><p>Robin Sibson: SLINK: An optimally efficient algorithm for the single-link cluster method. <em>The Computer Journal</em> <strong>16</strong> (1): 30–34 (1973). <a class="reference external" href="https://doi.org/10.1093/comjnl/16.1.30">https://doi.org/10.1093/comjnl/16.1.30</a></p>
</dd>
<dt class="label" id="slater2005"><span class="brackets">Slater2005</span></dt>
<dd><p>Guy St C. Slater, Ewan Birney: Automated generation of heuristics for biological sequence comparison. <em>BMC Bioinformatics</em> <strong>6</strong>: 31 (2005). <a class="reference external" href="https://doi.org/10.1186/1471-2105-6-31">https://doi.org/10.1186/1471-2105-6-31</a></p>
</dd>
<dt class="label" id="snedecor1989"><span class="brackets">Snedecor1989</span></dt>
<dd><p>George W. Snedecor, William G. Cochran: <em>Statistical methods</em>. Ames, Iowa: Iowa State University Press (1989).</p>
</dd>
<dt class="label" id="steinegger2019"><span class="brackets">Steinegger2019</span></dt>
<dd><p>Martin Steinegger, Markus Meier, Milot Mirdita, Harald Vöhringer, Stephan J. Haunsberger, Johannes Söding: HH-suite3 for fast remote homology detection and deep protein annotation. <em>BMC Bioinformatics</em> <strong>20</strong>: 473 (2019). <a class="reference external" href="https://doi.org/10.1186/s12859-019-3019-7">https://doi.org/10.1186/s12859-019-3019-7</a></p>
</dd>
<dt class="label" id="talevich2012"><span class="brackets">Talevich2012</span></dt>
<dd><p>Eric Talevich, Brandon M. Invergo, Peter J.A. Cock, Brad A. Chapman: Bio.Phylo: A unified toolkit for processing, analyzing and visualizing phylogenetic trees in Biopython. <em>BMC Bioinformatics</em> <strong>13</strong>: 209 (2012). <a class="reference external" href="https://doi.org/10.1186/1471-2105-13-209">https://doi.org/10.1186/1471-2105-13-209</a></p>
</dd>
<dt class="label" id="tamayo1999"><span class="brackets">Tamayo1999</span></dt>
<dd><p>Pablo Tamayo, Donna Slonim, Jill Mesirov, Qing Zhu, Sutisak Kitareewan, Ethan Dmitrovsky, Eric S. Lander, Todd R. Golub: Interpreting patterns of gene expression with self-organizing maps: Methods and application to hematopoietic differentiation. <em>Proceedings of the National Academy of Sciences USA</em> <strong>96</strong> (6): 2907–2912 (1999). <a class="reference external" href="https://doi.org/10.1073/pnas.96.6.2907">https://doi.org/10.1073/pnas.96.6.2907</a></p>
</dd>
<dt class="label" id="toth2006"><span class="brackets">Toth2006</span></dt>
<dd><p>Ian K. Toth, Leighton Pritchard, Paul R. J. Birch: Comparative genomics reveals what makes an enterobacterial plant pathogen. <em>Annual Review of Phytopathology</em> <strong>44</strong>: 305–336 (2006). <a class="reference external" href="https://doi.org/10.1146/annurev.phyto.44.070505.143444">https://doi.org/10.1146/annurev.phyto.44.070505.143444</a></p>
</dd>
<dt class="label" id="vanderauwera2009"><span class="brackets">Vanderauwera2009</span></dt>
<dd><p>Géraldine A. van der Auwera, Jaroslaw E. Król, Haruo Suzuki, Brian Foster, Rob van Houdt, Celeste J. Brown, Max Mergeay, Eva M. Top: Plasmids captured in C. metallidurans CH34: defining the PromA family of broad-host-range plasmids. <em>Antonie van Leeuwenhoek</em> <strong>96</strong> (2): 193–204 (2009). <a class="reference external" href="https://doi.org/10.1007/s10482-009-9316-9">https://doi.org/10.1007/s10482-009-9316-9</a></p>
</dd>
<dt class="label" id="waterman1987"><span class="brackets">Waterman1987</span></dt>
<dd><p>Michael S. Waterman, Mark Eggert: A new algorithm for best subsequence alignments with application to tRNA-rRNA comparisons. <em>Journal of Molecular Biology</em> <strong>197</strong> (4): 723–728 (1987). <a class="reference external" href="https://doi.org/10.1016/0022-2836(87)90478-5">https://doi.org/10.1016/0022-2836(87)90478-5</a></p>
</dd>
<dt class="label" id="yang2000"><span class="brackets">Yang2000</span></dt>
<dd><p>Ziheng Yang and Rasmus Nielsen: Estimating synonymous and nonsynonymous substitution rates under realistic evolutionary models. <em>Molecular Biology and Evolution</em> <strong>17</strong> (1): 32–43 (2000). <a class="reference external" href="https://doi.org/10.1093/oxfordjournals.molbev.a026236">https://doi.org/10.1093/oxfordjournals.molbev.a026236</a></p>
</dd>
<dt class="label" id="yeung2001"><span class="brackets">Yeung2001</span></dt>
<dd><p>Ka Yee Yeung, Walter L. Ruzzo: Principal Component Analysis for clustering gene expression data. <em>Bioinformatics</em> <strong>17</strong> (9): 763–774 (2001). <a class="reference external" href="https://doi.org/10.1093/bioinformatics/17.9.763">https://doi.org/10.1093/bioinformatics/17.9.763</a></p>
</dd>
</dl>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="chapter_appendix.html" class="btn btn-neutral float-left" title="Appendix: Useful stuff about Python" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../api/index.html" class="btn btn-neutral float-right" title="API documentation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>