<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Accessing NCBI’s Entrez databases &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/biopython.css?v=d1305275" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=49cfa8a4"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Swiss-Prot and ExPASy" href="chapter_uniprot.html" />
    <link rel="prev" title="BLAST and other sequence search tools" href="chapter_searchio.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Biopython Tutorial &amp; Cookbook</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="chapter_introduction.html">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_quick_start.html">Quick Start – What can you do with Biopython?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_objects.html">Sequence objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_annot.html">Sequence annotation objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seqio.html">Sequence Input/Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_align.html">Sequence alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise.html">Pairwise sequence alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_msa.html">Multiple Sequence Alignment objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise2.html">Pairwise alignments using pairwise2</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html">BLAST (new)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html#blast-old">BLAST (old)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_searchio.html">BLAST and other sequence search tools</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Accessing NCBI’s Entrez databases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#entrez-guidelines">Entrez Guidelines</a></li>
<li class="toctree-l3"><a class="reference internal" href="#einfo-obtaining-information-about-the-entrez-databases">EInfo: Obtaining information about the Entrez databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="#esearch-searching-the-entrez-databases">ESearch: Searching the Entrez databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="#epost-uploading-a-list-of-identifiers">EPost: Uploading a list of identifiers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#esummary-retrieving-summaries-from-primary-ids">ESummary: Retrieving summaries from primary IDs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#efetch-downloading-full-records-from-entrez">EFetch: Downloading full records from Entrez</a></li>
<li class="toctree-l3"><a class="reference internal" href="#elink-searching-for-related-items-in-ncbi-entrez">ELink: Searching for related items in NCBI Entrez</a></li>
<li class="toctree-l3"><a class="reference internal" href="#egquery-global-query-counts-for-search-terms">EGQuery: Global Query - counts for search terms</a></li>
<li class="toctree-l3"><a class="reference internal" href="#espell-obtaining-spelling-suggestions">ESpell: Obtaining spelling suggestions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parsing-huge-entrez-xml-files">Parsing huge Entrez XML files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#html-escape-characters">HTML escape characters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#handling-errors">Handling errors</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#the-file-is-not-an-xml-file">The file is not an XML file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-file-ends-prematurely-or-is-otherwise-corrupted">The file ends prematurely or is otherwise corrupted</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-file-contains-items-that-are-missing-from-the-associated-dtd">The file contains items that are missing from the associated DTD</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-file-contains-an-error-message">The file contains an error message</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#specialized-parsers">Specialized parsers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#parsing-medline-records">Parsing Medline records</a></li>
<li class="toctree-l4"><a class="reference internal" href="#parsing-geo-records">Parsing GEO records</a></li>
<li class="toctree-l4"><a class="reference internal" href="#parsing-unigene-records">Parsing UniGene records</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#using-a-proxy">Using a proxy</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pubmed-and-medline">PubMed and Medline</a></li>
<li class="toctree-l4"><a class="reference internal" href="#searching-downloading-and-parsing-entrez-nucleotide-records">Searching, downloading, and parsing Entrez Nucleotide records</a></li>
<li class="toctree-l4"><a class="reference internal" href="#searching-downloading-and-parsing-genbank-records">Searching, downloading, and parsing GenBank records</a></li>
<li class="toctree-l4"><a class="reference internal" href="#finding-the-lineage-of-an-organism">Finding the lineage of an organism</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#using-the-history-and-webenv">Using the history and WebEnv</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#searching-for-and-downloading-sequences-using-the-history">Searching for and downloading sequences using the history</a></li>
<li class="toctree-l4"><a class="reference internal" href="#searching-for-and-downloading-abstracts-using-the-history">Searching for and downloading abstracts using the history</a></li>
<li class="toctree-l4"><a class="reference internal" href="#searching-for-citations">Searching for citations</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chapter_uniprot.html">Swiss-Prot and ExPASy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pdb.html">Going 3D: The PDB module</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_popgen.html">Bio.PopGen: Population genetics</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phylo.html">Phylogenetics with Bio.Phylo</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_motifs.html">Sequence motif analysis using Bio.motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cluster.html">Cluster analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_graphics.html">Graphics including GenomeDiagram</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_kegg.html">KEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phenotype.html">Bio.phenotype: analyze phenotypic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cookbook.html">Cookbook – Cool things to do with it</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_testing.html">The Biopython testing framework</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_contributing.html">Where to go from here – contributing to Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_appendix.html">Appendix: Useful stuff about Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_bibliography.html">Bibliography</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Biopython Tutorial &amp; Cookbook</a></li>
      <li class="breadcrumb-item active">Accessing NCBI’s Entrez databases</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/Tutorial/chapter_entrez.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="chapter_searchio.html" class="btn btn-neutral float-left" title="BLAST and other sequence search tools" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_uniprot.html" class="btn btn-neutral float-right" title="Swiss-Prot and ExPASy" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="accessing-ncbis-entrez-databases">
<span id="chapter-entrez"></span><h1>Accessing NCBI’s Entrez databases<a class="headerlink" href="#accessing-ncbis-entrez-databases" title="Link to this heading"></a></h1>
<p>Entrez (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html">https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html</a>) is a data
retrieval system that provides users access to NCBI’s databases such as
PubMed, GenBank, GEO, and many others. You can access Entrez from a web
browser to manually enter queries, or you can use Biopython’s
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> module for programmatic access to Entrez. The latter
allows you for example to search PubMed or download GenBank records from
within a Python script.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> module makes use of the Entrez Programming Utilities
(also known as EUtils), consisting of eight tools that are described in
detail on NCBI’s page at <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25501/">https://www.ncbi.nlm.nih.gov/books/NBK25501/</a>.
Each of these tools corresponds to one Python function in the
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> module, as described in the sections below. This module
makes sure that the correct URL is used for the queries, and that NCBI’s
guidelines for responsible data access are being followed.</p>
<p>The output returned by the Entrez Programming Utilities is typically in
XML format. To parse such output, you have several options:</p>
<ol class="arabic simple">
<li><p>Use <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code>’s parser to parse the XML output into a Python
object;</p></li>
<li><p>Use one of the XML parsers available in Python’s standard library;</p></li>
<li><p>Read the XML output as raw text, and parse it by string searching and
manipulation.</p></li>
</ol>
<p>See the Python documentation for a description of the XML parsers in
Python’s standard library. Here, we discuss the parser in Biopython’s
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> module. This parser can be used to parse data as provided
through <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code>’s programmatic access functions to Entrez, but
can also be used to parse XML data from NCBI Entrez that are stored in a
file. In the latter case, the XML file should be opened in binary mode
(e.g. <code class="docutils literal notranslate"><span class="pre">open(&quot;myfile.xml&quot;,</span> <span class="pre">&quot;rb&quot;)</span></code>) for the XML parser in <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code>
to work correctly. Alternatively, you can pass the file name or path to
the XML file, and let <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> take care of opening and closing
the file.</p>
<p>NCBI uses DTD (Document Type Definition) files to describe the structure
of the information contained in XML files. Most of the DTD files used by
NCBI are included in the Biopython distribution. The <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code>
parser makes use of the DTD files when parsing an XML file returned by
NCBI Entrez.</p>
<p>Occasionally, you may find that the DTD file associated with a specific
XML file is missing in the Biopython distribution. In particular, this
may happen when NCBI updates its DTD files. If this happens,
<code class="docutils literal notranslate"><span class="pre">Entrez.read</span></code> will show a warning message with the name and URL of the
missing DTD file. The parser will proceed to access the missing DTD file
through the internet, allowing the parsing of the XML file to continue.
However, the parser is much faster if the DTD file is available locally.
For this purpose, please download the DTD file from the URL in the
warning message and place it in the directory
<code class="docutils literal notranslate"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></code>, containing the other DTD files. If
you don’t have write access to this directory, you can also place the
DTD file in <code class="docutils literal notranslate"><span class="pre">~/.biopython/Bio/Entrez/DTDs</span></code>, where <code class="docutils literal notranslate"><span class="pre">~</span></code> represents
your home directory. Since this directory is read before the directory
<code class="docutils literal notranslate"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></code>, you can also put newer versions of
DTD files there if the ones in <code class="docutils literal notranslate"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></code>
become outdated. Alternatively, if you installed Biopython from source,
you can add the DTD file to the source code’s <code class="docutils literal notranslate"><span class="pre">Bio/Entrez/DTDs</span></code>
directory, and reinstall Biopython. This will install the new DTD file
in the correct location together with the other DTD files.</p>
<p>The Entrez Programming Utilities can also generate output in other
formats, such as the Fasta or GenBank file formats for sequence
databases, or the MedLine format for the literature database, discussed
in Section <a class="reference internal" href="#sec-entrez-specialized-parsers"><span class="std std-ref">Specialized parsers</span></a>.</p>
<p>The functions in <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> for programmatic access to Entrez return
data either in binary format or in text format, depending on the type of
data requested. In most cases, these functions return data in text
format by decoding the data obtained from NCBI Entrez to Python strings
under the assumption that the encoding is UTF-8. However, XML data are
returned in binary format. The reason for this is that the encoding is
specified in the XML document itself, which means that we won’t know the
correct encoding to use until we start parsing the file.
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code>’s parser therefore accepts data in binary format,
extracts the encoding from the XML, and uses it to decode all text in
the XML document to Python strings, ensuring that all text (in
particular in languages other than English) are interpreted correctly.
This is also the reason why you should open an XML file a binary mode
when you want to use <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code>’s parser to parse the file.</p>
<section id="entrez-guidelines">
<span id="sec-entrez-guidelines"></span><h2>Entrez Guidelines<a class="headerlink" href="#entrez-guidelines" title="Link to this heading"></a></h2>
<p>Before using Biopython to access the NCBI’s online resources (via
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> or some of the other modules), please read the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25497/">NCBI’s
Entrez User
Requirements</a>. If the
NCBI finds you are abusing their systems, they can and will ban your
access!</p>
<p>To paraphrase:</p>
<ul>
<li><p>For any series of more than 100 requests, do this at weekends or
outside USA peak times. This is up to you to obey.</p></li>
<li><p>Use the <a class="reference external" href="https://eutils.ncbi.nlm.nih.gov">https://eutils.ncbi.nlm.nih.gov</a> address, not the standard
NCBI Web address. Biopython uses this web address.</p></li>
<li><p>If you are using a API key, you can make at most 10 queries per
second, otherwise at most 3 queries per second. This is automatically
enforced by Biopython. Include <code class="docutils literal notranslate"><span class="pre">api_key=&quot;MyAPIkey&quot;</span></code> in the argument
list or set it as a module level variable:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">api_key</span> <span class="o">=</span> <span class="s2">&quot;MyAPIkey&quot;</span>
</pre></div>
</div>
</li>
<li><p>Use the optional email parameter so the NCBI can contact you if there
is a problem. You can either explicitly set this as a parameter with
each call to Entrez (e.g. include <code class="docutils literal notranslate"><span class="pre">email=&quot;A.N.Other&#64;example.com&quot;</span></code>
in the argument list), or you can set a global email address:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> will then use this email address with each call to
Entrez. The <code class="docutils literal notranslate"><span class="pre">example.com</span></code> address is a reserved domain name
specifically for documentation (RFC 2606). Please DO NOT use a random
email – it’s better not to give an email at all. The email parameter
has been mandatory since June 1, 2010. In case of excessive usage,
NCBI will attempt to contact a user at the e-mail address provided
prior to blocking access to the E-utilities.</p>
</li>
<li><p>If you are using Biopython within some larger software suite, use the
tool parameter to specify this. You can either explicitly set the
tool name as a parameter with each call to Entrez (e.g. include
<code class="docutils literal notranslate"><span class="pre">tool=&quot;MyLocalScript&quot;</span></code> in the argument list), or you can set a
global tool name:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">tool</span> <span class="o">=</span> <span class="s2">&quot;MyLocalScript&quot;</span>
</pre></div>
</div>
<p>The tool parameter will default to Biopython.</p>
</li>
<li><p>For large queries, the NCBI also recommend using their session
history feature (the WebEnv session cookie string, see
Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a>). This is only slightly more
complicated.</p></li>
</ul>
<p>In conclusion, be sensible with your usage levels. If you plan to
download lots of data, consider other options. For example, if you want
easy access to all the human genes, consider fetching each chromosome by
FTP as a GenBank file, and importing these into your own BioSQL database
(see Section <a class="reference internal" href="chapter_cookbook.html#sec-biosql"><span class="std std-ref">BioSQL – storing sequences in a relational database</span></a>).</p>
</section>
<section id="einfo-obtaining-information-about-the-entrez-databases">
<span id="sec-entrez-einfo"></span><h2>EInfo: Obtaining information about the Entrez databases<a class="headerlink" href="#einfo-obtaining-information-about-the-entrez-databases" title="Link to this heading"></a></h2>
<p>EInfo provides field index term counts, last update, and available links
for each of NCBI’s databases. In addition, you can use EInfo to obtain a
list of all database names accessible through the Entrez utilities:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">einfo</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">stream</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>The variable <code class="docutils literal notranslate"><span class="pre">result</span></code> now contains a list of databases in XML format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="go">&lt;?xml version=&quot;1.0&quot;?&gt;</span>
<span class="go">&lt;!DOCTYPE eInfoResult PUBLIC &quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot;</span>
<span class="go"> &quot;https://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;&gt;</span>
<span class="go">&lt;eInfoResult&gt;</span>
<span class="go">&lt;DbList&gt;</span>
<span class="go">        &lt;DbName&gt;pubmed&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;protein&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nucleotide&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nuccore&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nucgss&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nucest&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;structure&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;genome&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;books&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;cancerchromosomes&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;cdd&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;gap&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;domains&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;gene&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;genomeprj&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;gensat&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;geo&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;gds&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;homologene&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;journals&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;mesh&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;ncbisearch&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nlmcatalog&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;omia&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;omim&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;pmc&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;popset&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;probe&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;proteinclusters&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;pcassay&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;pccompound&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;pcsubstance&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;snp&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;taxonomy&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;toolkit&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;unigene&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;unists&lt;/DbName&gt;</span>
<span class="go">&lt;/DbList&gt;</span>
<span class="go">&lt;/eInfoResult&gt;</span>
</pre></div>
</div>
<p>Since this is a fairly simple XML file, we could extract the information
it contains simply by string searching. Using <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code>’s parser
instead, we can directly parse this XML file into a Python object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">einfo</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
</pre></div>
</div>
<p>Now <code class="docutils literal notranslate"><span class="pre">record</span></code> is a dictionary with exactly one key:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">dict_keys([&#39;DbList&#39;])</span>
</pre></div>
</div>
<p>The values stored in this key is the list of database names shown in the
XML above:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;DbList&quot;</span><span class="p">]</span>
<span class="go">[&#39;pubmed&#39;, &#39;protein&#39;, &#39;nucleotide&#39;, &#39;nuccore&#39;, &#39;nucgss&#39;, &#39;nucest&#39;,</span>
<span class="go"> &#39;structure&#39;, &#39;genome&#39;, &#39;books&#39;, &#39;cancerchromosomes&#39;, &#39;cdd&#39;, &#39;gap&#39;,</span>
<span class="go"> &#39;domains&#39;, &#39;gene&#39;, &#39;genomeprj&#39;, &#39;gensat&#39;, &#39;geo&#39;, &#39;gds&#39;, &#39;homologene&#39;,</span>
<span class="go"> &#39;journals&#39;, &#39;mesh&#39;, &#39;ncbisearch&#39;, &#39;nlmcatalog&#39;, &#39;omia&#39;, &#39;omim&#39;, &#39;pmc&#39;,</span>
<span class="go"> &#39;popset&#39;, &#39;probe&#39;, &#39;proteinclusters&#39;, &#39;pcassay&#39;, &#39;pccompound&#39;,</span>
<span class="go"> &#39;pcsubstance&#39;, &#39;snp&#39;, &#39;taxonomy&#39;, &#39;toolkit&#39;, &#39;unigene&#39;, &#39;unists&#39;]</span>
</pre></div>
</div>
<p>For each of these databases, we can use EInfo again to obtain more
information:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">einfo</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;DbInfo&quot;</span><span class="p">][</span><span class="s2">&quot;Description&quot;</span><span class="p">]</span>
<span class="go">&#39;PubMed bibliographic record&#39;</span>
</pre></div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;DbInfo&quot;</span><span class="p">][</span><span class="s2">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">&#39;17989604&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;DbInfo&quot;</span><span class="p">][</span><span class="s2">&quot;LastUpdate&quot;</span><span class="p">]</span>
<span class="go">&#39;2008/05/24 06:45&#39;</span>
</pre></div>
</div>
<p>Try <code class="docutils literal notranslate"><span class="pre">record[&quot;DbInfo&quot;].keys()</span></code> for other information stored in this
record. One of the most useful is a list of possible search fields for
use with ESearch:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;DbInfo&quot;</span><span class="p">][</span><span class="s2">&quot;FieldList&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%(Name)s</span><span class="s2">, </span><span class="si">%(FullName)s</span><span class="s2">, </span><span class="si">%(Description)s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">field</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">ALL, All Fields, All terms from all searchable fields</span>
<span class="go">UID, UID, Unique number assigned to publication</span>
<span class="go">FILT, Filter, Limits the records</span>
<span class="go">TITL, Title, Words in title of publication</span>
<span class="go">WORD, Text Word, Free text associated with publication</span>
<span class="go">MESH, MeSH Terms, Medical Subject Headings assigned to publication</span>
<span class="go">MAJR, MeSH Major Topic, MeSH terms of major importance to publication</span>
<span class="go">AUTH, Author, Author(s) of publication</span>
<span class="go">JOUR, Journal, Journal abbreviation of publication</span>
<span class="go">AFFL, Affiliation, Author&#39;s institutional affiliation and address</span>
<span class="go">...</span>
</pre></div>
</div>
<p>That’s a long list, but indirectly this tells you that for the PubMed
database, you can do things like <code class="docutils literal notranslate"><span class="pre">Jones[AUTH]</span></code> to search the author
field, or <code class="docutils literal notranslate"><span class="pre">Sanger[AFFL]</span></code> to restrict to authors at the Sanger Centre.
This can be very handy - especially if you are not so familiar with a
particular database.</p>
</section>
<section id="esearch-searching-the-entrez-databases">
<span id="sec-entrez-esearch"></span><h2>ESearch: Searching the Entrez databases<a class="headerlink" href="#esearch-searching-the-entrez-databases" title="Link to this heading"></a></h2>
<p>To search any of these databases, we use <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.esearch()</span></code>. For
example, let’s search in PubMed for publications that include Biopython
in their title:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;biopython[title]&quot;</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="s2">&quot;40&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;19304878&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">])</span>
<span class="go">[&#39;22909249&#39;, &#39;19304878&#39;]</span>
</pre></div>
</div>
<p>In this output, you see PubMed IDs (including 19304878 which is the PMID
for the Biopython application note), which can be retrieved by EFetch
(see section <a class="reference internal" href="#sec-efetch"><span class="std std-ref">EFetch: Downloading full records from Entrez</span></a>).</p>
<p>You can also use ESearch to search GenBank. Here we’ll do a quick search
for the <em>matK</em> gene in <em>Cypripedioideae</em> orchids (see
Section <a class="reference internal" href="#sec-entrez-einfo"><span class="std std-ref">EInfo: Obtaining information about the Entrez databases</span></a> about EInfo for one way to find out
which fields you can search in each Entrez database):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;Cypripedioideae[Orgn] AND matK[Gene]&quot;</span><span class="p">,</span> <span class="n">idtype</span><span class="o">=</span><span class="s2">&quot;acc&quot;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">&#39;348&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;JQ660909.1&#39;, &#39;JQ660908.1&#39;, &#39;JQ660907.1&#39;, &#39;JQ660906.1&#39;, ..., &#39;JQ660890.1&#39;]</span>
</pre></div>
</div>
<p>Each of the IDs (JQ660909.1, JQ660908.1, JQ660907.1, …) is a GenBank
identifier (Accession number). See section <a class="reference internal" href="#sec-efetch"><span class="std std-ref">EFetch: Downloading full records from Entrez</span></a> for
information on how to actually download these GenBank records.</p>
<p>Note that instead of a species name like <code class="docutils literal notranslate"><span class="pre">Cypripedioideae[Orgn]</span></code>, you
can restrict the search using an NCBI taxon identifier, here this would
be <code class="docutils literal notranslate"><span class="pre">txid158330[Orgn]</span></code>. This isn’t currently documented on the ESearch
help page - the NCBI explained this in reply to an email query. You can
often deduce the search term formatting by playing with the Entrez web
interface. For example, including <code class="docutils literal notranslate"><span class="pre">complete[prop]</span></code> in a genome search
restricts to just completed genomes.</p>
<p>As a final example, let’s get a list of computational journal titles:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nlmcatalog&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;computational[Journal]&quot;</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="s2">&quot;20&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> computational journals found&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">]))</span>
<span class="go">117 computational Journals found</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The first 20 are</span><span class="se">\n</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]))</span>
<span class="go">[&#39;101660833&#39;, &#39;101664671&#39;, &#39;101661657&#39;, &#39;101659814&#39;, &#39;101657941&#39;,</span>
<span class="go"> &#39;101653734&#39;, &#39;101669877&#39;, &#39;101649614&#39;, &#39;101647835&#39;, &#39;101639023&#39;,</span>
<span class="go"> &#39;101627224&#39;, &#39;101647801&#39;, &#39;101589678&#39;, &#39;101585369&#39;, &#39;101645372&#39;,</span>
<span class="go"> &#39;101586429&#39;, &#39;101582229&#39;, &#39;101574747&#39;, &#39;101564639&#39;, &#39;101671907&#39;]</span>
</pre></div>
</div>
<p>Again, we could use EFetch to obtain more information for each of these
journal IDs.</p>
<p>ESearch has many useful options — see the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESearch">ESearch help
page</a>
for more information.</p>
</section>
<section id="epost-uploading-a-list-of-identifiers">
<h2>EPost: Uploading a list of identifiers<a class="headerlink" href="#epost-uploading-a-list-of-identifiers" title="Link to this heading"></a></h2>
<p>EPost uploads a list of UIs for use in subsequent search strategies; see
the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EPost">EPost help
page</a>
for more information. It is available from Biopython through the
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez.epost()</span></code> function.</p>
<p>To give an example of when this is useful, suppose you have a long list
of IDs you want to download using EFetch (maybe sequences, maybe
citations – anything). When you make a request with EFetch your list of
IDs, the database etc, are all turned into a long URL sent to the
server. If your list of IDs is long, this URL gets long, and long URLs
can break (e.g. some proxies don’t cope well).</p>
<p>Instead, you can break this up into two steps, first uploading the list
of IDs using EPost (this uses an “HTML post” internally, rather than an
“HTML get”, getting round the long URL problem). With the history
support, you can then refer to this long list of IDs, and download the
associated data with EFetch.</p>
<p>Let’s look at a simple example to see how EPost works – uploading some
PubMed identifiers:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">id_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;19304878&quot;</span><span class="p">,</span> <span class="s2">&quot;18606172&quot;</span><span class="p">,</span> <span class="s2">&quot;16403221&quot;</span><span class="p">,</span> <span class="s2">&quot;16377612&quot;</span><span class="p">,</span> <span class="s2">&quot;14871861&quot;</span><span class="p">,</span> <span class="s2">&quot;14630660&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Entrez</span><span class="o">.</span><span class="n">epost</span><span class="p">(</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">id_list</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="go">&lt;?xml version=&quot;1.0&quot;?&gt;</span>
<span class="go">&lt;!DOCTYPE ePostResult PUBLIC &quot;-//NLM//DTD ePostResult, 11 May 2002//EN&quot;</span>
<span class="go"> &quot;https://www.ncbi.nlm.nih.gov/entrez/query/DTD/ePost_020511.dtd&quot;&gt;</span>
<span class="go">&lt;ePostResult&gt;</span>
<span class="go">    &lt;QueryKey&gt;1&lt;/QueryKey&gt;</span>
<span class="go">    &lt;WebEnv&gt;NCID_01_206841095_130.14.22.101_9001_1242061629&lt;/WebEnv&gt;</span>
<span class="go">&lt;/ePostResult&gt;</span>
</pre></div>
</div>
<p>The returned XML includes two important strings, <code class="docutils literal notranslate"><span class="pre">QueryKey</span></code> and
<code class="docutils literal notranslate"><span class="pre">WebEnv</span></code> which together define your history session. You would extract
these values for use with another Entrez call such as EFetch:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">id_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;19304878&quot;</span><span class="p">,</span> <span class="s2">&quot;18606172&quot;</span><span class="p">,</span> <span class="s2">&quot;16403221&quot;</span><span class="p">,</span> <span class="s2">&quot;16377612&quot;</span><span class="p">,</span> <span class="s2">&quot;14871861&quot;</span><span class="p">,</span> <span class="s2">&quot;14630660&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">search_results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">Entrez</span><span class="o">.</span><span class="n">epost</span><span class="p">(</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">id_list</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">webenv</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;WebEnv&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query_key</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;QueryKey&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a> shows how to use the history
feature.</p>
</section>
<section id="esummary-retrieving-summaries-from-primary-ids">
<h2>ESummary: Retrieving summaries from primary IDs<a class="headerlink" href="#esummary-retrieving-summaries-from-primary-ids" title="Link to this heading"></a></h2>
<p>ESummary retrieves document summaries from a list of primary IDs (see
the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESummary">ESummary help
page</a>
for more information). In Biopython, ESummary is available as
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez.esummary()</span></code>. Using the search result above, we can for
example find out more about the journal with ID 30367:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esummary</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nlmcatalog&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;101660833&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">info</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;TitleMainList&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Journal info</span><span class="se">\n</span><span class="s2">id: </span><span class="si">{}</span><span class="se">\n</span><span class="s2">Title: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Id&quot;</span><span class="p">],</span> <span class="n">info</span><span class="p">[</span><span class="s2">&quot;Title&quot;</span><span class="p">]))</span>
<span class="go">Journal info</span>
<span class="go">id: 101660833</span>
<span class="go">Title: IEEE transactions on computational imaging.</span>
</pre></div>
</div>
</section>
<section id="efetch-downloading-full-records-from-entrez">
<span id="sec-efetch"></span><h2>EFetch: Downloading full records from Entrez<a class="headerlink" href="#efetch-downloading-full-records-from-entrez" title="Link to this heading"></a></h2>
<p>EFetch is what you use when you want to retrieve a full record from
Entrez. This covers several possible databases, as described on the main
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK3837/">EFetch Help page</a>.</p>
<p>For most of their databases, the NCBI support several different file
formats. Requesting a specific file format from Entrez using
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch()</span></code> requires specifying the <code class="docutils literal notranslate"><span class="pre">rettype</span></code> and/or
<code class="docutils literal notranslate"><span class="pre">retmode</span></code> optional arguments. The different combinations are described
for each database type on the pages linked to on <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch">NCBI efetch
webpage</a>.</p>
<p>One common usage is downloading sequences in the FASTA or
GenBank/GenPept plain text formats (which can then be parsed with
<code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code>, see
Sections <a class="reference internal" href="chapter_seqio.html#sec-seqio-genbank-online"><span class="std std-ref">Parsing GenBank records from the net</span></a>
and <a class="reference internal" href="#sec-efetch"><span class="std std-ref">EFetch: Downloading full records from Entrez</span></a>). From the <em>Cypripedioideae</em> example above, we
can download GenBank record EU490707 using <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;EU490707&quot;</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">stream</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="go">LOCUS       EU490707                1302 bp    DNA     linear   PLN 26-JUL-2016</span>
<span class="go">DEFINITION  Selenipedium aequinoctiale maturase K (matK) gene, partial cds;</span>
<span class="go">            chloroplast.</span>
<span class="go">ACCESSION   EU490707</span>
<span class="go">VERSION     EU490707.1</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      chloroplast Selenipedium aequinoctiale</span>
<span class="go">  ORGANISM  Selenipedium aequinoctiale</span>
<span class="go">            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;</span>
<span class="go">            Spermatophyta; Magnoliopsida; Liliopsida; Asparagales; Orchidaceae;</span>
<span class="go">            Cypripedioideae; Selenipedium.</span>
<span class="go">REFERENCE   1  (bases 1 to 1302)</span>
<span class="go">  AUTHORS   Neubig,K.M., Whitten,W.M., Carlsward,B.S., Blanco,M.A., Endara,L.,</span>
<span class="go">            Williams,N.H. and Moore,M.</span>
<span class="go">  TITLE     Phylogenetic utility of ycf1 in orchids: a plastid gene more</span>
<span class="go">            variable than matK</span>
<span class="go">  JOURNAL   Plant Syst. Evol. 277 (1-2), 75-84 (2009)</span>
<span class="go">REFERENCE   2  (bases 1 to 1302)</span>
<span class="go">  AUTHORS   Neubig,K.M., Whitten,W.M., Carlsward,B.S., Blanco,M.A.,</span>
<span class="go">            Endara,C.L., Williams,N.H. and Moore,M.J.</span>
<span class="go">  TITLE     Direct Submission</span>
<span class="go">  JOURNAL   Submitted (14-FEB-2008) Department of Botany, University of</span>
<span class="go">            Florida, 220 Bartram Hall, Gainesville, FL 32611-8526, USA</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">     source          1..1302</span>
<span class="go">                     /organism=&quot;Selenipedium aequinoctiale&quot;</span>
<span class="go">                     /organelle=&quot;plastid:chloroplast&quot;</span>
<span class="go">                     /mol_type=&quot;genomic DNA&quot;</span>
<span class="go">                     /specimen_voucher=&quot;FLAS:Blanco 2475&quot;</span>
<span class="go">                     /db_xref=&quot;taxon:256374&quot;</span>
<span class="go">     gene            &lt;1..&gt;1302</span>
<span class="go">                     /gene=&quot;matK&quot;</span>
<span class="go">     CDS             &lt;1..&gt;1302</span>
<span class="go">                     /gene=&quot;matK&quot;</span>
<span class="go">                     /codon_start=1</span>
<span class="go">                     /transl_table=11</span>
<span class="go">                     /product=&quot;maturase K&quot;</span>
<span class="go">                     /protein_id=&quot;ACC99456.1&quot;</span>
<span class="go">                     /translation=&quot;IFYEPVEIFGYDNKSSLVLVKRLITRMYQQNFLISSVNDSNQKG</span>
<span class="go">                     FWGHKHFFSSHFSSQMVSEGFGVILEIPFSSQLVSSLEEKKIPKYQNLRSIHSIFPFL</span>
<span class="go">                     EDKFLHLNYVSDLLIPHPIHLEILVQILQCRIKDVPSLHLLRLLFHEYHNLNSLITSK</span>
<span class="go">                     KFIYAFSKRKKRFLWLLYNSYVYECEYLFQFLRKQSSYLRSTSSGVFLERTHLYVKIE</span>
<span class="go">                     HLLVVCCNSFQRILCFLKDPFMHYVRYQGKAILASKGTLILMKKWKFHLVNFWQSYFH</span>
<span class="go">                     FWSQPYRIHIKQLSNYSFSFLGYFSSVLENHLVVRNQMLENSFIINLLTKKFDTIAPV</span>
<span class="go">                     ISLIGSLSKAQFCTVLGHPISKPIWTDFSDSDILDRFCRICRNLCRYHSGSSKKQVLY</span>
<span class="go">                     RIKYILRLSCARTLARKHKSTVRTFMRRLGSGLLEEFFMEEE&quot;</span>
<span class="go">ORIGIN</span>
<span class="go">        1 attttttacg aacctgtgga aatttttggt tatgacaata aatctagttt agtacttgtg</span>
<span class="go">       61 aaacgtttaa ttactcgaat gtatcaacag aattttttga tttcttcggt taatgattct</span>
<span class="go">      121 aaccaaaaag gattttgggg gcacaagcat tttttttctt ctcatttttc ttctcaaatg</span>
<span class="go">      181 gtatcagaag gttttggagt cattctggaa attccattct cgtcgcaatt agtatcttct</span>
<span class="go">      241 cttgaagaaa aaaaaatacc aaaatatcag aatttacgat ctattcattc aatatttccc</span>
<span class="go">      301 tttttagaag acaaattttt acatttgaat tatgtgtcag atctactaat accccatccc</span>
<span class="go">      361 atccatctgg aaatcttggt tcaaatcctt caatgccgga tcaaggatgt tccttctttg</span>
<span class="go">      421 catttattgc gattgctttt ccacgaatat cataatttga atagtctcat tacttcaaag</span>
<span class="go">      481 aaattcattt acgccttttc aaaaagaaag aaaagattcc tttggttact atataattct</span>
<span class="go">      541 tatgtatatg aatgcgaata tctattccag tttcttcgta aacagtcttc ttatttacga</span>
<span class="go">      601 tcaacatctt ctggagtctt tcttgagcga acacatttat atgtaaaaat agaacatctt</span>
<span class="go">      661 ctagtagtgt gttgtaattc ttttcagagg atcctatgct ttctcaagga tcctttcatg</span>
<span class="go">      721 cattatgttc gatatcaagg aaaagcaatt ctggcttcaa agggaactct tattctgatg</span>
<span class="go">      781 aagaaatgga aatttcatct tgtgaatttt tggcaatctt attttcactt ttggtctcaa</span>
<span class="go">      841 ccgtatagga ttcatataaa gcaattatcc aactattcct tctcttttct ggggtatttt</span>
<span class="go">      901 tcaagtgtac tagaaaatca tttggtagta agaaatcaaa tgctagagaa ttcatttata</span>
<span class="go">      961 ataaatcttc tgactaagaa attcgatacc atagccccag ttatttctct tattggatca</span>
<span class="go">     1021 ttgtcgaaag ctcaattttg tactgtattg ggtcatccta ttagtaaacc gatctggacc</span>
<span class="go">     1081 gatttctcgg attctgatat tcttgatcga ttttgccgga tatgtagaaa tctttgtcgt</span>
<span class="go">     1141 tatcacagcg gatcctcaaa aaaacaggtt ttgtatcgta taaaatatat acttcgactt</span>
<span class="go">     1201 tcgtgtgcta gaactttggc acggaaacat aaaagtacag tacgcacttt tatgcgaaga</span>
<span class="go">     1261 ttaggttcgg gattattaga agaattcttt atggaagaag aa</span>
<span class="go">//</span>

</pre></div>
</div>
<p>Please be aware that as of October 2016 GI identifiers are discontinued
in favor of accession numbers. You can still fetch sequences based on
their GI, but new sequences are no longer given this identifier. You
should instead refer to them by the “Accession number” as done in the
example.</p>
<p>The arguments <code class="docutils literal notranslate"><span class="pre">rettype=&quot;gb&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">retmode=&quot;text&quot;</span></code> let us download
this record in the GenBank format.</p>
<p>Note that until Easter 2009, the Entrez EFetch API let you use “genbank”
as the return type, however the NCBI now insist on using the official
return types of “gb” or “gbwithparts” (or “gp” for proteins) as
described on online. Also note that until Feb 2012, the Entrez EFetch
API would default to returning plain text files, but now defaults to
XML.</p>
<p>Alternatively, you could for example use <code class="docutils literal notranslate"><span class="pre">rettype=&quot;fasta&quot;</span></code> to get the
Fasta-format; see the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch">EFetch Sequences Help
page</a>
for other options. Remember – the available formats depend on which
database you are downloading from - see the main <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch">EFetch Help
page</a>.</p>
<p>If you fetch the record in one of the formats accepted by <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code>
(see Chapter <a class="reference internal" href="chapter_seqio.html#chapter-seqio"><span class="std std-ref">Sequence Input/Output</span></a>), you could directly
parse it into a <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;EU490707&quot;</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="go">EU490707.1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="go">EU490707</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="go">Selenipedium aequinoctiale maturase K (matK) gene, partial cds; chloroplast</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">))</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;ATTTTTTACGAACCTGTGGAAATTTTTGGTTATGACAATAAATCTAGTTTAGTA...GAA&#39;)</span>
</pre></div>
</div>
<p>Note that a more typical use would be to save the sequence data to a
local file, and <em>then</em> parse it with <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code>. This can save you
having to re-download the same file repeatedly while working on your
script, and places less load on the NCBI’s servers. For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>

<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;EU490707.gbk&quot;</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="c1"># Downloading...</span>
    <span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;EU490707&quot;</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>
    <span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">streame</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
    <span class="n">output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saved&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Parsing...&quot;</span><span class="p">)</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</pre></div>
</div>
<p>To get the output in XML format, which you can parse using the
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez.read()</span></code> function, use <code class="docutils literal notranslate"><span class="pre">retmode=&quot;xml&quot;</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;EU490707&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;GBSeq_definition&quot;</span><span class="p">]</span>
<span class="go">&#39;Selenipedium aequinoctiale maturase K (matK) gene, partial cds; chloroplast&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;GBSeq_source&quot;</span><span class="p">]</span>
<span class="go">&#39;chloroplast Selenipedium aequinoctiale&#39;</span>
</pre></div>
</div>
<p>So, that dealt with sequences. For examples of parsing file formats
specific to the other databases (e.g. the <code class="docutils literal notranslate"><span class="pre">MEDLINE</span></code> format used in
PubMed), see Section <a class="reference internal" href="#sec-entrez-specialized-parsers"><span class="std std-ref">Specialized parsers</span></a>.</p>
<p>If you want to perform a search with <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.esearch()</span></code>, and then
download the records with <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch()</span></code>, you should use the
WebEnv history feature – see Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a>.</p>
</section>
<section id="elink-searching-for-related-items-in-ncbi-entrez">
<span id="sec-elink"></span><h2>ELink: Searching for related items in NCBI Entrez<a class="headerlink" href="#elink-searching-for-related-items-in-ncbi-entrez" title="Link to this heading"></a></h2>
<p>ELink, available from Biopython as <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.elink()</span></code>, can be used
to find related items in the NCBI Entrez databases. For example, you can
us this to find nucleotide entries for an entry in the gene database,
and other cool stuff.</p>
<p>Let’s use ELink to find articles related to the Biopython application
note published in <em>Bioinformatics</em> in 2009. The PubMed ID of this
article is 19304878:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pmid</span> <span class="o">=</span> <span class="s2">&quot;19304878&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">Entrez</span><span class="o">.</span><span class="n">elink</span><span class="p">(</span><span class="n">dbfrom</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">pmid</span><span class="p">))</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">record</span></code> variable consists of a Python list, one for each database
in which we searched. Since we specified only one PubMed ID to search
for, <code class="docutils literal notranslate"><span class="pre">record</span></code> contains only one item. This item is a dictionary
containing information about our search term, as well as all the related
items that were found:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;DbFrom&quot;</span><span class="p">]</span>
<span class="go">&#39;pubmed&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;19304878&#39;]</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">&quot;LinkSetDb&quot;</span></code> key contains the search results, stored as a list
consisting of one item for each target database. In our search results,
we only find hits in the PubMed database (although sub-divided into
categories):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;LinkSetDb&quot;</span><span class="p">])</span>
<span class="go">8</span>
</pre></div>
</div>
<p>The exact numbers should increase over time:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">linksetdb</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;LinkSetDb&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">linksetdb</span><span class="p">[</span><span class="s2">&quot;DbTo&quot;</span><span class="p">],</span> <span class="n">linksetdb</span><span class="p">[</span><span class="s2">&quot;LinkName&quot;</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">linksetdb</span><span class="p">[</span><span class="s2">&quot;Link&quot;</span><span class="p">]))</span>
<span class="gp">...</span>
<span class="go">pubmed pubmed_pubmed 284</span>
<span class="go">pubmed pubmed_pubmed_alsoviewed 7</span>
<span class="go">pubmed pubmed_pubmed_citedin 926</span>
<span class="go">pubmed pubmed_pubmed_combined 6</span>
<span class="go">pubmed pubmed_pubmed_five 6</span>
<span class="go">pubmed pubmed_pubmed_refs 17</span>
<span class="go">pubmed pubmed_pubmed_reviews 12</span>
<span class="go">pubmed pubmed_pubmed_reviews_five 6</span>
</pre></div>
</div>
<p>The actual search results are stored as under the <code class="docutils literal notranslate"><span class="pre">&quot;Link&quot;</span></code> key.</p>
<p>Let’s now at the first search result:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Link&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="go">{&#39;Id&#39;: &#39;19304878&#39;}</span>
</pre></div>
</div>
<p>This is the article we searched for, which doesn’t help us much, so
let’s look at the second search result:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Link&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="go">{&#39;Id&#39;: &#39;14630660&#39;}</span>
</pre></div>
</div>
<p>This paper, with PubMed ID 14630660, is about the Biopython PDB parser.</p>
<p>We can use a loop to print out all PubMed IDs:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Link&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">link</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">19304878</span>
<span class="go">14630660</span>
<span class="go">18689808</span>
<span class="go">17121776</span>
<span class="go">16377612</span>
<span class="go">12368254</span>
<span class="go">......</span>
</pre></div>
</div>
<p>Now that was nice, but personally I am often more interested to find out
if a paper has been cited. Well, ELink can do that too – at least for
journals in Pubmed Central (see
Section <a class="reference internal" href="#sec-elink-citations"><span class="std std-ref">Searching for citations</span></a>).</p>
<p>For help on ELink, see the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ELink">ELink help
page</a>.
There is an entire sub-page just for the <a class="reference external" href="https://eutils.ncbi.nlm.nih.gov/corehtml/query/static/entrezlinks.html">link
names</a>,
describing how different databases can be cross referenced.</p>
</section>
<section id="egquery-global-query-counts-for-search-terms">
<h2>EGQuery: Global Query - counts for search terms<a class="headerlink" href="#egquery-global-query-counts-for-search-terms" title="Link to this heading"></a></h2>
<p>EGQuery provides counts for a search term in each of the Entrez
databases (i.e. a global query). This is particularly useful to find out
how many items your search terms would find in each database without
actually performing lots of separate searches with ESearch (see the
example in <a class="reference internal" href="#sec-entrez-example-genbank"><span class="std std-ref">Searching, downloading, and parsing Entrez Nucleotide records</span></a> below).</p>
<p>In this example, we use <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.egquery()</span></code> to obtain the counts
for “Biopython”:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">egquery</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;biopython&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;eGQueryResult&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;DbName&quot;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">pubmed 6</span>
<span class="go">pmc 62</span>
<span class="go">journals 0</span>
<span class="go">...</span>
</pre></div>
</div>
<p>See the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EGQuery">EGQuery help
page</a>
for more information.</p>
</section>
<section id="espell-obtaining-spelling-suggestions">
<h2>ESpell: Obtaining spelling suggestions<a class="headerlink" href="#espell-obtaining-spelling-suggestions" title="Link to this heading"></a></h2>
<p>ESpell retrieves spelling suggestions. In this example, we use
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez.espell()</span></code> to obtain the correct spelling of Biopython:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">espell</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;biopythooon&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;Query&quot;</span><span class="p">]</span>
<span class="go">&#39;biopythooon&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;CorrectedQuery&quot;</span><span class="p">]</span>
<span class="go">&#39;biopython&#39;</span>
</pre></div>
</div>
<p>See the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESpell">ESpell help
page</a>
for more information. The main use of this is for GUI tools to provide
automatic suggestions for search terms.</p>
</section>
<section id="parsing-huge-entrez-xml-files">
<h2>Parsing huge Entrez XML files<a class="headerlink" href="#parsing-huge-entrez-xml-files" title="Link to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Entrez.read</span></code> function reads the entire XML file returned by
Entrez into a single Python object, which is kept in memory. To parse
Entrez XML files too large to fit in memory, you can use the function
<code class="docutils literal notranslate"><span class="pre">Entrez.parse</span></code>. This is a generator function that reads records in the
XML file one by one. This function is only useful if the XML file
reflects a Python list object (in other words, if <code class="docutils literal notranslate"><span class="pre">Entrez.read</span></code> on a
computer with infinite memory resources would return a Python list).</p>
<p>For example, you can download the entire Entrez Gene database for a
given organism as a file from NCBI’s ftp site. These files can be very
large. As an example, on September 4, 2009, the file
<code class="docutils literal notranslate"><span class="pre">Homo_sapiens.ags.gz</span></code>, containing the Entrez Gene database for human,
had a size of 116576 kB. This file, which is in the <code class="docutils literal notranslate"><span class="pre">ASN</span></code> format, can
be converted into an XML file using NCBI’s <code class="docutils literal notranslate"><span class="pre">gene2xml</span></code> program (see
NCBI’s ftp site for more information):</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>gene2xml<span class="w"> </span>-b<span class="w"> </span>T<span class="w"> </span>-i<span class="w"> </span>Homo_sapiens.ags<span class="w"> </span>-o<span class="w"> </span>Homo_sapiens.xml
</pre></div>
</div>
<p>The resulting XML file has a size of 6.1 GB. Attempting <code class="docutils literal notranslate"><span class="pre">Entrez.read</span></code>
on this file will result in a <code class="docutils literal notranslate"><span class="pre">MemoryError</span></code> on many computers.</p>
<p>The XML file <code class="docutils literal notranslate"><span class="pre">Homo_sapiens.xml</span></code> consists of a list of Entrez gene
records, each corresponding to one Entrez gene in human.
<code class="docutils literal notranslate"><span class="pre">Entrez.parse</span></code> retrieves these gene records one by one. You can then
print out or store the relevant information in each record by iterating
over the records. For example, this script iterates over the Entrez gene
records and prints out the gene numbers and names for all current genes:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;Homo_sapiens.xml&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternatively, you can use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Homo_sapiens.xml&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>and let <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> take care of opening and closing the file. This
is safer, as the file will then automatically be closed after parsing
it, or if an error occurs.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">status</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;Entrezgene_track-info&quot;</span><span class="p">][</span><span class="s2">&quot;Gene-track&quot;</span><span class="p">][</span><span class="s2">&quot;Gene-track_status&quot;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">status</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;value&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;discontinued&quot;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">continue</span>
<span class="gp">... </span>    <span class="n">geneid</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;Entrezgene_track-info&quot;</span><span class="p">][</span><span class="s2">&quot;Gene-track&quot;</span><span class="p">][</span><span class="s2">&quot;Gene-track_geneid&quot;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="n">genename</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;Entrezgene_gene&quot;</span><span class="p">][</span><span class="s2">&quot;Gene-ref&quot;</span><span class="p">][</span><span class="s2">&quot;Gene-ref_locus&quot;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">geneid</span><span class="p">,</span> <span class="n">genename</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">1 A1BG</span>
<span class="go">2 A2M</span>
<span class="go">3 A2MP</span>
<span class="go">8 AA</span>
<span class="go">9 NAT1</span>
<span class="go">10 NAT2</span>
<span class="go">11 AACP</span>
<span class="go">12 SERPINA3</span>
<span class="go">13 AADAC</span>
<span class="go">14 AAMP</span>
<span class="go">15 AANAT</span>
<span class="go">16 AARS</span>
<span class="go">17 AAVS1</span>
<span class="go">...</span>
</pre></div>
</div>
</section>
<section id="html-escape-characters">
<h2>HTML escape characters<a class="headerlink" href="#html-escape-characters" title="Link to this heading"></a></h2>
<p>Pubmed records may contain HTML tags to indicate e.g. subscripts,
superscripts, or italic text, as well as mathematical symbols via
MathML. By default, the <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> parser treats all text as plain
text without markup; for example, the fragment “<span class="math notranslate nohighlight">\(P &lt; 0.05\)</span>” in the
abstract of a Pubmed record, which is encoded as</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&lt;i&gt;P&lt;/i&gt; &amp;lt; 0.05
</pre></div>
</div>
<p>in the XML returned by Entrez, is converted to the Python string</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&#39;&lt;i&gt;P&lt;/i&gt; &lt; 0.05&#39;
</pre></div>
</div>
<p>by the <code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> parser. While this is more human-readable, it is
not valid HTML due to the less-than sign, and makes further processing
of the text e.g. by an HTML parser impractical. To ensure that all
strings returned by the parser are valid HTML, call <code class="docutils literal notranslate"><span class="pre">Entrez.read</span></code> or
<code class="docutils literal notranslate"><span class="pre">Entrez.parse</span></code> with the <code class="docutils literal notranslate"><span class="pre">escape</span></code> argument set to <code class="docutils literal notranslate"><span class="pre">True</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">,</span> <span class="n">escape</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>The parser will then replace all characters disallowed in HTML by their
HTML-escaped equivalent; in the example above, the parser will generate</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&#39;&lt;i&gt;P&lt;/i&gt; &amp;lt; 0.05&#39;
</pre></div>
</div>
<p>which is a valid HTML fragment. By default, <code class="docutils literal notranslate"><span class="pre">escape</span></code> is <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
</section>
<section id="handling-errors">
<h2>Handling errors<a class="headerlink" href="#handling-errors" title="Link to this heading"></a></h2>
<section id="the-file-is-not-an-xml-file">
<h3>The file is not an XML file<a class="headerlink" href="#the-file-is-not-an-xml-file" title="Link to this heading"></a></h3>
<p>For example, this error occurs if you try to parse a Fasta file as if it
were an XML file:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;NC_005816.fna&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>  <span class="c1"># a Fasta file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">  </span><span class="c">...</span>
<span class="gr">Bio.Entrez.Parser.NotXMLError</span>: <span class="n">Failed to parse the XML data (syntax error: line 1, column 0). Please make sure that the input data are in XML format.</span>
</pre></div>
</div>
<p>Here, the parser didn’t find the <code class="docutils literal notranslate"><span class="pre">&lt;?xml</span> <span class="pre">...</span></code> tag with which an XML
file is supposed to start, and therefore decides (correctly) that the
file is not an XML file.</p>
</section>
<section id="the-file-ends-prematurely-or-is-otherwise-corrupted">
<h3>The file ends prematurely or is otherwise corrupted<a class="headerlink" href="#the-file-ends-prematurely-or-is-otherwise-corrupted" title="Link to this heading"></a></h3>
<p>When your file is in the XML format but is corrupted (for example, by
ending prematurely), the parser will raise a CorruptedXMLError.</p>
<p>Here is an example of an XML file that ends prematurely:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&lt;?xml version=&quot;1.0&quot;?&gt;
&lt;!DOCTYPE eInfoResult PUBLIC &quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot; &quot;https://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;&gt;
&lt;eInfoResult&gt;
&lt;DbList&gt;
        &lt;DbName&gt;pubmed&lt;/DbName&gt;
        &lt;DbName&gt;protein&lt;/DbName&gt;
        &lt;DbName&gt;nucleotide&lt;/DbName&gt;
        &lt;DbName&gt;nuccore&lt;/DbName&gt;
        &lt;DbName&gt;nucgss&lt;/DbName&gt;
        &lt;DbName&gt;nucest&lt;/DbName&gt;
        &lt;DbName&gt;structure&lt;/DbName&gt;
        &lt;DbName&gt;genome&lt;/DbName&gt;
        &lt;DbName&gt;books&lt;/DbName&gt;
        &lt;DbName&gt;cancerchromosomes&lt;/DbName&gt;
        &lt;DbName&gt;cdd&lt;/DbName&gt;
</pre></div>
</div>
<p>which will generate the following traceback:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">  </span><span class="c">...</span>
<span class="gr">Bio.Entrez.Parser.CorruptedXMLError</span>: <span class="n">Failed to parse the XML data (no element found: line 16, column 0). Please make sure that the input data are not corrupted.</span>
</pre></div>
</div>
<p>Note that the error message tells you at what point in the XML file the
error was detected.</p>
</section>
<section id="the-file-contains-items-that-are-missing-from-the-associated-dtd">
<h3>The file contains items that are missing from the associated DTD<a class="headerlink" href="#the-file-contains-items-that-are-missing-from-the-associated-dtd" title="Link to this heading"></a></h3>
<p>This is an example of an XML file containing tags that do not have a
description in the corresponding DTD file:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>&lt;?xml version=&quot;1.0&quot;?&gt;
&lt;!DOCTYPE eInfoResult PUBLIC &quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot; &quot;https://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;&gt;
&lt;eInfoResult&gt;
        &lt;DbInfo&gt;
        &lt;DbName&gt;pubmed&lt;/DbName&gt;
        &lt;MenuName&gt;PubMed&lt;/MenuName&gt;
        &lt;Description&gt;PubMed bibliographic record&lt;/Description&gt;
        &lt;Count&gt;20161961&lt;/Count&gt;
        &lt;LastUpdate&gt;2010/09/10 04:52&lt;/LastUpdate&gt;
        &lt;FieldList&gt;
                &lt;Field&gt;
...
                &lt;/Field&gt;
        &lt;/FieldList&gt;
        &lt;DocsumList&gt;
                &lt;Docsum&gt;
                        &lt;DsName&gt;PubDate&lt;/DsName&gt;
                        &lt;DsType&gt;4&lt;/DsType&gt;
                        &lt;DsTypeName&gt;string&lt;/DsTypeName&gt;
                &lt;/Docsum&gt;
                &lt;Docsum&gt;
                        &lt;DsName&gt;EPubDate&lt;/DsName&gt;
...
        &lt;/DbInfo&gt;
&lt;/eInfoResult&gt;
</pre></div>
</div>
<p>In this file, for some reason the tag <code class="docutils literal notranslate"><span class="pre">&lt;DocsumList&gt;</span></code> (and several
others) are not listed in the DTD file <code class="docutils literal notranslate"><span class="pre">eInfo_020511.dtd</span></code>, which is
specified on the second line as the DTD for this XML file. By default,
the parser will stop and raise a ValidationError if it cannot find some
tag in the DTD:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;einfo3.xml&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">  </span><span class="c">...</span>
<span class="gr">Bio.Entrez.Parser.ValidationError</span>: <span class="n">Failed to find tag &#39;DocsumList&#39; in the DTD. To skip all tags that are not represented in the DTD, please call Bio.Entrez.read or Bio.Entrez.parse with validate=False.</span>
</pre></div>
</div>
<p>Optionally, you can instruct the parser to skip such tags instead of
raising a ValidationError. This is done by calling <code class="docutils literal notranslate"><span class="pre">Entrez.read</span></code> or
<code class="docutils literal notranslate"><span class="pre">Entrez.parse</span></code> with the argument <code class="docutils literal notranslate"><span class="pre">validate</span></code> equal to False:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;einfo3.xml&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Of course, the information contained in the XML tags that are not in the
DTD are not present in the record returned by <code class="docutils literal notranslate"><span class="pre">Entrez.read</span></code>.</p>
</section>
<section id="the-file-contains-an-error-message">
<h3>The file contains an error message<a class="headerlink" href="#the-file-contains-an-error-message" title="Link to this heading"></a></h3>
<p>This may occur, for example, when you attempt to access a PubMed record
for a nonexistent PubMed ID. By default, this will raise a
<code class="docutils literal notranslate"><span class="pre">RuntimeError</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esummary</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;99999999&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">RuntimeError</span>: <span class="n">UID=99999999: cannot get document summary</span>
</pre></div>
</div>
<p>If you are accessing multiple PubMed records, the <code class="docutils literal notranslate"><span class="pre">RuntimeError</span></code> would
prevent you from receiving results for any of the PubMed records if one
of the PubMed IDs is incorrect. To circumvent this, you can set the
<code class="docutils literal notranslate"><span class="pre">ignore_errors</span></code> argument to <code class="docutils literal notranslate"><span class="pre">True</span></code>. This will return the requested
results for the valid PubMed IDs, and an <code class="docutils literal notranslate"><span class="pre">ErrorElement</span></code> for the
incorrect ID:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esummary</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;19304878,99999999,31278684&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">,</span> <span class="n">ignore_errors</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tag</span>
<span class="go">&#39;DocSum&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Title&quot;</span><span class="p">]</span>
<span class="go">&#39;Biopython: freely available Python tools for computational molecular biology and bioinformatics.&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">tag</span>
<span class="go">&#39;ERROR&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">ErrorElement(&#39;UID=99999999: cannot get document summary&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">tag</span>
<span class="go">&#39;DocSum&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s2">&quot;Title&quot;</span><span class="p">]</span>
<span class="go">&#39;Sharing Programming Resources Between Bio* Projects.&#39;</span>
</pre></div>
</div>
</section>
</section>
<section id="specialized-parsers">
<span id="sec-entrez-specialized-parsers"></span><h2>Specialized parsers<a class="headerlink" href="#specialized-parsers" title="Link to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.read()</span></code> function can parse most (if not all) XML
output returned by Entrez. Entrez typically allows you to retrieve
records in other formats, which may have some advantages compared to the
XML format in terms of readability (or download size).</p>
<p>To request a specific file format from Entrez using
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch()</span></code> requires specifying the <code class="docutils literal notranslate"><span class="pre">rettype</span></code> and/or
<code class="docutils literal notranslate"><span class="pre">retmode</span></code> optional arguments. The different combinations are described
for each database type on the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch">NCBI efetch
webpage</a>.</p>
<p>One obvious case is you may prefer to download sequences in the FASTA or
GenBank/GenPept plain text formats (which can then be parsed with
<code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code>, see
Sections <a class="reference internal" href="chapter_seqio.html#sec-seqio-genbank-online"><span class="std std-ref">Parsing GenBank records from the net</span></a>
and <a class="reference internal" href="#sec-efetch"><span class="std std-ref">EFetch: Downloading full records from Entrez</span></a>). For the literature databases, Biopython
contains a parser for the <code class="docutils literal notranslate"><span class="pre">MEDLINE</span></code> format used in PubMed.</p>
<section id="parsing-medline-records">
<span id="sec-entrez-and-medline"></span><h3>Parsing Medline records<a class="headerlink" href="#parsing-medline-records" title="Link to this heading"></a></h3>
<p>You can find the Medline parser in <code class="docutils literal notranslate"><span class="pre">Bio.Medline</span></code>. Suppose we want to
parse the file <code class="docutils literal notranslate"><span class="pre">pubmed_result1.txt</span></code>, containing one Medline record.
You can find this file in Biopython’s <code class="docutils literal notranslate"><span class="pre">Tests\Medline</span></code> directory. The
file looks like this:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>PMID- 12230038
OWN - NLM
STAT- MEDLINE
DA  - 20020916
DCOM- 20030606
LR  - 20041117
PUBM- Print
IS  - 1467-5463 (Print)
VI  - 3
IP  - 3
DP  - 2002 Sep
TI  - The Bio* toolkits--a brief overview.
PG  - 296-302
AB  - Bioinformatics research is often difficult to do with commercial software. The
      Open Source BioPerl, BioPython and Biojava projects provide toolkits with
...
</pre></div>
</div>
<p>We first open the file and then parse it:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Medline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;pubmed_result1.txt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stream</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">record</span> <span class="o">=</span> <span class="n">Medline</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">record</span></code> now contains the Medline record as a Python dictionary:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;PMID&quot;</span><span class="p">]</span>
<span class="go">&#39;12230038&#39;</span>
</pre></div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;AB&quot;</span><span class="p">]</span>
<span class="go">&#39;Bioinformatics research is often difficult to do with commercial software.</span>
<span class="go">The Open Source BioPerl, BioPython and Biojava projects provide toolkits with</span>
<span class="go">multiple functionality that make it easier to create customized pipelines or</span>
<span class="go">analysis. This review briefly compares the quirks of the underlying languages</span>
<span class="go">and the functionality, documentation, utility and relative advantages of the</span>
<span class="go">Bio counterparts, particularly from the point of view of the beginning</span>
<span class="go">biologist programmer.&#39;</span>
</pre></div>
</div>
<p>The key names used in a Medline record can be rather obscure; use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</pre></div>
</div>
<p>for a brief summary.</p>
<p>To parse a file containing multiple Medline records, you can use the
<code class="docutils literal notranslate"><span class="pre">parse</span></code> function instead:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Medline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;pubmed_result2.txt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stream</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">Medline</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">stream</span><span class="p">):</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;TI&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">A high level interface to SCOP and ASTRAL implemented in python.</span>
<span class="go">GenomeDiagram: a python package for the visualization of large-scale genomic data.</span>
<span class="go">Open source clustering software.</span>
<span class="go">PDB file parser and structure class implemented in Python.</span>
</pre></div>
</div>
<p>Instead of parsing Medline records stored in files, you can also parse
Medline records downloaded by <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch</span></code>. For example, let’s
look at all Medline records in PubMed related to Biopython:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;biopython&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;19304878&#39;, &#39;18606172&#39;, &#39;16403221&#39;, &#39;16377612&#39;, &#39;14871861&#39;, &#39;14630660&#39;, &#39;12230038&#39;]</span>
</pre></div>
</div>
<p>We now use <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch</span></code> to download these Medline records:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">idlist</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">idlist</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;medline&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, we specify <code class="docutils literal notranslate"><span class="pre">rettype=&quot;medline&quot;,</span> <span class="pre">retmode=&quot;text&quot;</span></code> to obtain the
Medline records in plain-text Medline format. Now we use <code class="docutils literal notranslate"><span class="pre">Bio.Medline</span></code>
to parse these records:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Medline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Medline</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;AU&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">[&#39;Cock PJ&#39;, &#39;Antao T&#39;, &#39;Chang JT&#39;, &#39;Chapman BA&#39;, &#39;Cox CJ&#39;, &#39;Dalke A&#39;, ..., &#39;de Hoon MJ&#39;]</span>
<span class="go">[&#39;Munteanu CR&#39;, &#39;Gonzalez-Diaz H&#39;, &#39;Magalhaes AL&#39;]</span>
<span class="go">[&#39;Casbon JA&#39;, &#39;Crooks GE&#39;, &#39;Saqi MA&#39;]</span>
<span class="go">[&#39;Pritchard L&#39;, &#39;White JA&#39;, &#39;Birch PR&#39;, &#39;Toth IK&#39;]</span>
<span class="go">[&#39;de Hoon MJ&#39;, &#39;Imoto S&#39;, &#39;Nolan J&#39;, &#39;Miyano S&#39;]</span>
<span class="go">[&#39;Hamelryck T&#39;, &#39;Manderick B&#39;]</span>
<span class="go">[&#39;Mangalam H&#39;]</span>
</pre></div>
</div>
<p>For comparison, here we show an example using the XML format:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">idlist</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;medline&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">[</span><span class="s2">&quot;PubmedArticle&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;MedlineCitation&quot;</span><span class="p">][</span><span class="s2">&quot;Article&quot;</span><span class="p">][</span><span class="s2">&quot;ArticleTitle&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">Biopython: freely available Python tools for computational molecular biology and</span>
<span class="go"> bioinformatics.</span>
<span class="go">Enzymes/non-enzymes classification model complexity based on composition, sequence,</span>
<span class="go"> 3D and topological indices.</span>
<span class="go">A high level interface to SCOP and ASTRAL implemented in python.</span>
<span class="go">GenomeDiagram: a python package for the visualization of large-scale genomic data.</span>
<span class="go">Open source clustering software.</span>
<span class="go">PDB file parser and structure class implemented in Python.</span>
<span class="go">The Bio* toolkits--a brief overview.</span>
</pre></div>
</div>
<p>Note that in both of these examples, for simplicity we have naively
combined ESearch and EFetch. In this situation, the NCBI would expect
you to use their history feature, as illustrated in
Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a>.</p>
</section>
<section id="parsing-geo-records">
<h3>Parsing GEO records<a class="headerlink" href="#parsing-geo-records" title="Link to this heading"></a></h3>
<p>GEO (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus</a>) is
a data repository of high-throughput gene expression and hybridization
array data. The <code class="docutils literal notranslate"><span class="pre">Bio.Geo</span></code> module can be used to parse GEO-formatted
data.</p>
<p>The following code fragment shows how to parse the example GEO file
<code class="docutils literal notranslate"><span class="pre">GSE16.txt</span></code> into a record and print the record:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Geo</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;GSE16.txt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Geo</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>You can search the “gds” database (GEO datasets) with ESearch:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;gds&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;GSE16&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">&#39;27&#39;</span>
</pre></div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;200000016&#39;, &#39;100000028&#39;, ...]</span>
</pre></div>
</div>
<p>From the Entrez website, UID “200000016” is GDS16 while the other hit
“100000028” is for the associated platform, GPL28. Unfortunately, at the
time of writing the NCBI don’t seem to support downloading GEO files
using Entrez (not as XML, nor in the <em>Simple Omnibus Format in Text</em>
(SOFT) format).</p>
<p>However, it is actually pretty straight forward to download the GEO
files by FTP from <a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/geo/">ftp://ftp.ncbi.nih.gov/pub/geo/</a> instead. In this case
you might want
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SOFT/by_series/GSE16/GSE16_family.soft.gz">ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SOFT/by_series/GSE16/GSE16_family.soft.gz</a>
(a compressed file, see the Python module gzip).</p>
</section>
<section id="parsing-unigene-records">
<h3>Parsing UniGene records<a class="headerlink" href="#parsing-unigene-records" title="Link to this heading"></a></h3>
<p>UniGene is an NCBI database of the transcriptome, with each UniGene
record showing the set of transcripts that are associated with a
particular gene in a specific organism. A typical UniGene record looks
like this:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>ID          Hs.2
TITLE       N-acetyltransferase 2 (arylamine N-acetyltransferase)
GENE        NAT2
CYTOBAND    8p22
GENE_ID     10
LOCUSLINK   10
HOMOL       YES
EXPRESS      bone| connective tissue| intestine| liver| liver tumor| normal| soft tissue/muscle tissue tumor| adult
RESTR_EXPR   adult
CHROMOSOME  8
STS         ACC=PMC310725P3 UNISTS=272646
STS         ACC=WIAF-2120 UNISTS=44576
STS         ACC=G59899 UNISTS=137181
...
STS         ACC=GDB:187676 UNISTS=155563
PROTSIM     ORG=10090; PROTGI=6754794; PROTID=NP_035004.1; PCT=76.55; ALN=288
PROTSIM     ORG=9796; PROTGI=149742490; PROTID=XP_001487907.1; PCT=79.66; ALN=288
PROTSIM     ORG=9986; PROTGI=126722851; PROTID=NP_001075655.1; PCT=76.90; ALN=288
...
PROTSIM     ORG=9598; PROTGI=114619004; PROTID=XP_519631.2; PCT=98.28; ALN=288

SCOUNT      38
SEQUENCE    ACC=BC067218.1; NID=g45501306; PID=g45501307; SEQTYPE=mRNA
SEQUENCE    ACC=NM_000015.2; NID=g116295259; PID=g116295260; SEQTYPE=mRNA
SEQUENCE    ACC=D90042.1; NID=g219415; PID=g219416; SEQTYPE=mRNA
SEQUENCE    ACC=D90040.1; NID=g219411; PID=g219412; SEQTYPE=mRNA
SEQUENCE    ACC=BC015878.1; NID=g16198419; PID=g16198420; SEQTYPE=mRNA
SEQUENCE    ACC=CR407631.1; NID=g47115198; PID=g47115199; SEQTYPE=mRNA
SEQUENCE    ACC=BG569293.1; NID=g13576946; CLONE=IMAGE:4722596; END=5&#39;; LID=6989; SEQTYPE=EST; TRACE=44157214
...
SEQUENCE    ACC=AU099534.1; NID=g13550663; CLONE=HSI08034; END=5&#39;; LID=8800; SEQTYPE=EST
//
</pre></div>
</div>
<p>This particular record shows the set of transcripts (shown in the
<code class="docutils literal notranslate"><span class="pre">SEQUENCE</span></code> lines) that originate from the human gene NAT2, encoding en
N-acetyltransferase. The <code class="docutils literal notranslate"><span class="pre">PROTSIM</span></code> lines show proteins with
significant similarity to NAT2, whereas the <code class="docutils literal notranslate"><span class="pre">STS</span></code> lines show the
corresponding sequence-tagged sites in the genome.</p>
<p>To parse UniGene files, use the <code class="docutils literal notranslate"><span class="pre">Bio.UniGene</span></code> module:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">UniGene</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;myunigenefile.data&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">UniGene</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">record</span></code> returned by <code class="docutils literal notranslate"><span class="pre">UniGene.read</span></code> is a Python object with
attributes corresponding to the fields in the UniGene record. For
example,</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">ID</span>
<span class="go">&quot;Hs.2&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">title</span>
<span class="go">&quot;N-acetyltransferase 2 (arylamine N-acetyltransferase)&quot;</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">EXPRESS</span></code> and <code class="docutils literal notranslate"><span class="pre">RESTR_EXPR</span></code> lines are stored as Python lists of
strings:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="p">[</span>
    <span class="s2">&quot;bone&quot;</span><span class="p">,</span>
    <span class="s2">&quot;connective tissue&quot;</span><span class="p">,</span>
    <span class="s2">&quot;intestine&quot;</span><span class="p">,</span>
    <span class="s2">&quot;liver&quot;</span><span class="p">,</span>
    <span class="s2">&quot;liver tumor&quot;</span><span class="p">,</span>
    <span class="s2">&quot;normal&quot;</span><span class="p">,</span>
    <span class="s2">&quot;soft tissue/muscle tissue tumor&quot;</span><span class="p">,</span>
    <span class="s2">&quot;adult&quot;</span><span class="p">,</span>
<span class="p">]</span>
</pre></div>
</div>
<p>Specialized objects are returned for the <code class="docutils literal notranslate"><span class="pre">STS</span></code>, <code class="docutils literal notranslate"><span class="pre">PROTSIM</span></code>, and
<code class="docutils literal notranslate"><span class="pre">SEQUENCE</span></code> lines, storing the keys shown in each line as attributes:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">sts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">acc</span>
<span class="go">&#39;PMC310725P3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">sts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">unists</span>
<span class="go">&#39;272646&#39;</span>
</pre></div>
</div>
<p>and similarly for the <code class="docutils literal notranslate"><span class="pre">PROTSIM</span></code> and <code class="docutils literal notranslate"><span class="pre">SEQUENCE</span></code> lines.</p>
<p>To parse a file containing more than one UniGene record, use the
<code class="docutils literal notranslate"><span class="pre">parse</span></code> function in <code class="docutils literal notranslate"><span class="pre">Bio.UniGene</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">UniGene</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;unigenerecords.data&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">UniGene</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">ID</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
</section>
</section>
<section id="using-a-proxy">
<h2>Using a proxy<a class="headerlink" href="#using-a-proxy" title="Link to this heading"></a></h2>
<p>Normally you won’t have to worry about using a proxy, but if this is an
issue on your network here is how to deal with it. Internally,
<code class="docutils literal notranslate"><span class="pre">Bio.Entrez</span></code> uses the standard Python library <code class="docutils literal notranslate"><span class="pre">urllib</span></code> for accessing
the NCBI servers. This will check an environment variable called
<code class="docutils literal notranslate"><span class="pre">http_proxy</span></code> to configure any simple proxy automatically.
Unfortunately this module does not support the use of proxies which
require authentication.</p>
<p>You may choose to set the <code class="docutils literal notranslate"><span class="pre">http_proxy</span></code> environment variable once (how
you do this will depend on your operating system). Alternatively you can
set this within Python at the start of your script, for example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>

<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;http_proxy&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;http://proxyhost.example.com:8080&quot;</span>
</pre></div>
</div>
<p>See the <a class="reference external" href="https://docs.python.org/2/library/urllib.html">urllib
documentation</a> for
more details.</p>
</section>
<section id="examples">
<span id="sec-entrez-examples"></span><h2>Examples<a class="headerlink" href="#examples" title="Link to this heading"></a></h2>
<section id="pubmed-and-medline">
<span id="sec-pub-med"></span><h3>PubMed and Medline<a class="headerlink" href="#pubmed-and-medline" title="Link to this heading"></a></h3>
<p>If you are in the medical field or interested in human issues (and many
times even if you are not!), PubMed
(<a class="reference external" href="https://www.ncbi.nlm.nih.gov/PubMed/">https://www.ncbi.nlm.nih.gov/PubMed/</a>) is an excellent source of all
kinds of goodies. So like other things, we’d like to be able to grab
information from it and use it in Python scripts.</p>
<p>In this example, we will query PubMed for all articles having to do with
orchids (see section <a class="reference internal" href="chapter_quick_start.html#sec-orchids"><span class="std std-ref">A usage example</span></a> for our
motivation). We first check how many of such articles there are:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">egquery</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;orchid&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;eGQueryResult&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;DbName&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;pubmed&quot;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">463</span>
</pre></div>
</div>
<p>Now we use the <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch</span></code> function to download the PubMed IDs
of these 463 articles:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;orchid&quot;</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="mi">463</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">idlist</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>This returns a Python list containing all of the PubMed IDs of articles
related to orchids:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">idlist</span><span class="p">)</span>
<span class="go">[&#39;18680603&#39;, &#39;18665331&#39;, &#39;18661158&#39;, &#39;18627489&#39;, &#39;18627452&#39;, &#39;18612381&#39;,</span>
<span class="go">&#39;18594007&#39;, &#39;18591784&#39;, &#39;18589523&#39;, &#39;18579475&#39;, &#39;18575811&#39;, &#39;18575690&#39;,</span>
<span class="go">...</span>
</pre></div>
</div>
<p>Now that we’ve got them, we obviously want to get the corresponding
Medline records and extract the information from them. Here, we’ll
download the Medline records in the Medline flat-file format, and use
the <code class="docutils literal notranslate"><span class="pre">Bio.Medline</span></code> module to parse them:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Medline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">idlist</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;medline&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Medline</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
</pre></div>
</div>
<p>NOTE - We’ve just done a separate search and fetch here, the NCBI much
prefer you to take advantage of their history support in this situation.
See Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a>.</p>
<p>Keep in mind that <code class="docutils literal notranslate"><span class="pre">records</span></code> is an iterator, so you can iterate through
the records only once. If you want to save the records, you can convert
them to a list:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">records</span><span class="p">)</span>
</pre></div>
</div>
<p>Let’s now iterate over the records to print out some information about
each record:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;title:&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;TI&quot;</span><span class="p">,</span> <span class="s2">&quot;?&quot;</span><span class="p">))</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;authors:&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;AU&quot;</span><span class="p">,</span> <span class="s2">&quot;?&quot;</span><span class="p">))</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;source:&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;SO&quot;</span><span class="p">,</span> <span class="s2">&quot;?&quot;</span><span class="p">))</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>The output for this looks like:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>title: Sex pheromone mimicry in the early spider orchid (ophrys sphegodes):
patterns of hydrocarbons as the key mechanism for pollination by sexual
deception [In Process Citation]
authors: [&#39;Schiestl FP&#39;, &#39;Ayasse M&#39;, &#39;Paulus HF&#39;, &#39;Lofstedt C&#39;, &#39;Hansson BS&#39;,
&#39;Ibarra F&#39;, &#39;Francke W&#39;]
source: J Comp Physiol [A] 2000 Jun;186(6):567-74
</pre></div>
</div>
<p>Especially interesting to note is the list of authors, which is returned
as a standard Python list. This makes it easy to manipulate and search
using standard Python tools. For instance, we could loop through a whole
bunch of entries searching for a particular author with code like the
following:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">search_author</span> <span class="o">=</span> <span class="s2">&quot;Waits T&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="ow">not</span> <span class="s2">&quot;AU&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">continue</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">search_author</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;AU&quot;</span><span class="p">]:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Author </span><span class="si">%s</span><span class="s2"> found: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">search_author</span><span class="p">,</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;SO&quot;</span><span class="p">]))</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>Hopefully this section gave you an idea of the power and flexibility of
the Entrez and Medline interfaces and how they can be used together.</p>
</section>
<section id="searching-downloading-and-parsing-entrez-nucleotide-records">
<span id="sec-entrez-example-genbank"></span><h3>Searching, downloading, and parsing Entrez Nucleotide records<a class="headerlink" href="#searching-downloading-and-parsing-entrez-nucleotide-records" title="Link to this heading"></a></h3>
<p>Here we’ll show a simple example of performing a remote Entrez query. In
section <a class="reference internal" href="chapter_quick_start.html#sec-orchids"><span class="std std-ref">A usage example</span></a> of the parsing examples, we
talked about using NCBI’s Entrez website to search the NCBI nucleotide
databases for info on Cypripedioideae, our friends the lady slipper
orchids. Now, we’ll look at how to automate that process using a Python
script. In this example, we’ll just show how to connect, get the
results, and parse them, with the Entrez module doing all of the work.</p>
<p>First, we use EGQuery to find out the number of results we will get
before actually downloading them. EGQuery will tell us how many search
results were found in each of the databases, but for this example we are
only interested in nucleotides:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">egquery</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Cypripedioideae&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;eGQueryResult&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;DbName&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;nuccore&quot;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">4457</span>
</pre></div>
</div>
<p>So, we expect to find 4457 Entrez Nucleotide records (this increased
from 814 records in 2008; it is likely to continue to increase in the
future). If you find some ridiculously high number of hits, you may want
to reconsider if you really want to download all of them, which is our
next step. Let’s use the <code class="docutils literal notranslate"><span class="pre">retmax</span></code> argument to restrict the maximum
number of records retrieved to the number available in 2008:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;Cypripedioideae&quot;</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="mi">814</span><span class="p">,</span> <span class="n">idtype</span><span class="o">=</span><span class="s2">&quot;acc&quot;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Here, <code class="docutils literal notranslate"><span class="pre">record</span></code> is a Python dictionary containing the search results
and some auxiliary information. Just for information, let’s look at what
is stored in this dictionary:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="go">[&#39;Count&#39;, &#39;RetMax&#39;, &#39;IdList&#39;, &#39;TranslationSet&#39;, &#39;RetStart&#39;, &#39;QueryTranslation&#39;]</span>
</pre></div>
</div>
<p>First, let’s check how many results were found:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">])</span>
<span class="go">&#39;4457&#39;</span>
</pre></div>
</div>
<p>You might have expected this to be 814, the maximum number of records we
asked to retrieve. However, <code class="docutils literal notranslate"><span class="pre">Count</span></code> represents the total number of
records available for that search, not how many were retrieved. The
retrieved records are stored in <code class="docutils literal notranslate"><span class="pre">record['IdList']</span></code>, which should
contain the total number we asked for:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">])</span>
<span class="go">814</span>
</pre></div>
</div>
<p>Let’s look at the first five results:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">][:</span><span class="mi">5</span><span class="p">]</span>
<span class="go">[&#39;KX265015.1&#39;, &#39;KX265014.1&#39;, &#39;KX265013.1&#39;, &#39;KX265012.1&#39;, &#39;KX265011.1&#39;]</span>
</pre></div>
</div>
<p>We can download these records using <code class="docutils literal notranslate"><span class="pre">efetch</span></code>. While you could download
these records one by one, to reduce the load on NCBI’s servers, it is
better to fetch a bunch of records at the same time, shown below.
However, in this situation you should ideally be using the history
feature described later in Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a>.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">idlist</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">][:</span><span class="mi">5</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">idlist</span><span class="p">)</span>
<span class="go">KX265015.1, KX265014.1, KX265013.1, KX265012.1, KX265011.1]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">idlist</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span>
<span class="go">5</span>
</pre></div>
</div>
<p>Each of these records corresponds to one GenBank record.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="go">[&#39;GBSeq_moltype&#39;, &#39;GBSeq_source&#39;, &#39;GBSeq_sequence&#39;,</span>
<span class="go"> &#39;GBSeq_primary-accession&#39;, &#39;GBSeq_definition&#39;, &#39;GBSeq_accession-version&#39;,</span>
<span class="go"> &#39;GBSeq_topology&#39;, &#39;GBSeq_length&#39;, &#39;GBSeq_feature-table&#39;,</span>
<span class="go"> &#39;GBSeq_create-date&#39;, &#39;GBSeq_other-seqids&#39;, &#39;GBSeq_division&#39;,</span>
<span class="go"> &#39;GBSeq_taxonomy&#39;, &#39;GBSeq_references&#39;, &#39;GBSeq_update-date&#39;,</span>
<span class="go"> &#39;GBSeq_organism&#39;, &#39;GBSeq_locus&#39;, &#39;GBSeq_strandedness&#39;]</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;GBSeq_primary-accession&quot;</span><span class="p">])</span>
<span class="go">DQ110336</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;GBSeq_other-seqids&quot;</span><span class="p">])</span>
<span class="go">[&#39;gb|DQ110336.1|&#39;, &#39;gi|187237168&#39;]</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;GBSeq_definition&quot;</span><span class="p">])</span>
<span class="go">Cypripedium calceolus voucher Davis 03-03 A maturase (matR) gene, partial cds;</span>
<span class="go">mitochondrial</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;GBSeq_organism&quot;</span><span class="p">])</span>
<span class="go">Cypripedium calceolus</span>
</pre></div>
</div>
<p>You could use this to quickly set up searches – but for heavy usage, see
Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a>.</p>
</section>
<section id="searching-downloading-and-parsing-genbank-records">
<span id="sec-entrez-search-fetch-genbank"></span><h3>Searching, downloading, and parsing GenBank records<a class="headerlink" href="#searching-downloading-and-parsing-genbank-records" title="Link to this heading"></a></h3>
<p>The GenBank record format is a very popular method of holding
information about sequences, sequence features, and other associated
sequence information. The format is a good way to get information from
the NCBI databases at <a class="reference external" href="https://www.ncbi.nlm.nih.gov/">https://www.ncbi.nlm.nih.gov/</a>.</p>
<p>In this example we’ll show how to query the NCBI databases,to retrieve
the records from the query, and then parse them using <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> -
something touched on in
Section <a class="reference internal" href="chapter_seqio.html#sec-seqio-genbank-online"><span class="std std-ref">Parsing GenBank records from the net</span></a>. For
simplicity, this example <em>does not</em> take advantage of the WebEnv history
feature – see Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a> for this.</p>
<p>First, we want to make a query and find out the ids of the records to
retrieve. Here we’ll do a quick search for one of our favorite
organisms, <em>Opuntia</em> (prickly-pear cacti). We can do quick search and
get back the GIs (GenBank identifiers) for all of the corresponding
records. First we check how many records there are:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">egquery</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s2">&quot;Opuntia AND rpl16&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;eGQueryResult&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;DbName&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;nuccore&quot;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">9</span>
</pre></div>
</div>
<p>Now we download the list of GenBank identifiers:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nuccore&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;Opuntia AND rpl16&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gi_list</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gi_list</span>
<span class="go">[&#39;57240072&#39;, &#39;57240071&#39;, &#39;6273287&#39;, &#39;6273291&#39;, &#39;6273290&#39;, &#39;6273289&#39;, &#39;6273286&#39;,</span>
<span class="go">&#39;6273285&#39;, &#39;6273284&#39;]</span>
</pre></div>
</div>
<p>Now we use these GIs to download the GenBank records - note that with
older versions of Biopython you had to supply a comma separated list of
GI numbers to Entrez, as of Biopython 1.59 you can pass a list and this
is converted for you:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gi_str</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gi_list</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nuccore&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">gi_str</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>If you want to look at the raw GenBank files, you can read from this
stream and print out the result:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">text</span> <span class="o">=</span> <span class="n">stream</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">text</span><span class="p">)</span>
<span class="go">LOCUS       AY851612                 892 bp    DNA     linear   PLN 10-APR-2007</span>
<span class="go">DEFINITION  Opuntia subulata rpl16 gene, intron; chloroplast.</span>
<span class="go">ACCESSION   AY851612</span>
<span class="go">VERSION     AY851612.1  GI:57240072</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      chloroplast Austrocylindropuntia subulata</span>
<span class="go">  ORGANISM  Austrocylindropuntia subulata</span>
<span class="go">            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;</span>
<span class="go">            Spermatophyta; Magnoliophyta; eudicotyledons; core eudicotyledons;</span>
<span class="go">            Caryophyllales; Cactaceae; Opuntioideae; Austrocylindropuntia.</span>
<span class="go">REFERENCE   1  (bases 1 to 892)</span>
<span class="go">  AUTHORS   Butterworth,C.A. and Wallace,R.S.</span>
<span class="go">...</span>
</pre></div>
</div>
<p>In this case, we are just getting the raw records. To get the records in
a more Python-friendly form, we can use <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> to parse the
GenBank data into <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> objects, including <code class="docutils literal notranslate"><span class="pre">SeqFeature</span></code>
objects (see Chapter <a class="reference internal" href="chapter_seqio.html#chapter-seqio"><span class="std std-ref">Sequence Input/Output</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;nuccore&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">gi_str</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">stream</span><span class="p">,</span> <span class="s2">&quot;gb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can now step through the records and look at the information we are
interested in:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">, length </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)</span><span class="si">}</span><span class="s2">, with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">)</span><span class="si">}</span><span class="s2"> features&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">AY851612, length 892, with 3 features</span>
<span class="go">AY851611, length 881, with 3 features</span>
<span class="go">AF191661, length 895, with 3 features</span>
<span class="go">AF191665, length 902, with 3 features</span>
<span class="go">AF191664, length 899, with 3 features</span>
<span class="go">AF191663, length 899, with 3 features</span>
<span class="go">AF191660, length 893, with 3 features</span>
<span class="go">AF191659, length 894, with 3 features</span>
<span class="go">AF191658, length 896, with 3 features</span>
</pre></div>
</div>
<p>Using these automated query retrieval functionality is a big plus over
doing things by hand. Although the module should obey the NCBI’s max
three queries per second rule, the NCBI have other recommendations like
avoiding peak hours. See Section <a class="reference internal" href="#sec-entrez-guidelines"><span class="std std-ref">Entrez Guidelines</span></a>. In
particular, please note that for simplicity, this example does not use
the WebEnv history feature. You should use this for any non-trivial
search and download work, see Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a>.</p>
<p>Finally, if plan to repeat your analysis, rather than downloading the
files from the NCBI and parsing them immediately (as shown in this
example), you should just download the records <em>once</em> and save them to
your hard disk, and then parse the local file.</p>
</section>
<section id="finding-the-lineage-of-an-organism">
<h3>Finding the lineage of an organism<a class="headerlink" href="#finding-the-lineage-of-an-organism" title="Link to this heading"></a></h3>
<p>Staying with a plant example, let’s now find the lineage of the
Cypripedioideae orchid family. First, we search the Taxonomy database
for Cypripedioideae, which yields exactly one NCBI taxonomy identifier:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;Taxonomy&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;Cypripedioideae&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;158330&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;158330&#39;</span>
</pre></div>
</div>
<p>Now, we use <code class="docutils literal notranslate"><span class="pre">efetch</span></code> to download this entry in the Taxonomy database,
and then parse it:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s2">&quot;Taxonomy&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;158330&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
</pre></div>
</div>
<p>Again, this record stores lots of information:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[&#39;Lineage&#39;, &#39;Division&#39;, &#39;ParentTaxId&#39;, &#39;PubDate&#39;, &#39;LineageEx&#39;,</span>
<span class="go"> &#39;CreateDate&#39;, &#39;TaxId&#39;, &#39;Rank&#39;, &#39;GeneticCode&#39;, &#39;ScientificName&#39;,</span>
<span class="go"> &#39;MitoGeneticCode&#39;, &#39;UpdateDate&#39;]</span>
</pre></div>
</div>
<p>We can get the lineage directly from this record:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Lineage&quot;</span><span class="p">]</span>
<span class="go">&#39;cellular organisms; Eukaryota; Viridiplantae; Streptophyta; Streptophytina;</span>
<span class="go"> Embryophyta; Tracheophyta; Euphyllophyta; Spermatophyta; Magnoliopsida;</span>
<span class="go"> Liliopsida; Asparagales; Orchidaceae&#39;</span>
</pre></div>
</div>
<p>The record data contains much more than just the information shown here
- for example look under <code class="docutils literal notranslate"><span class="pre">&quot;LineageEx&quot;</span></code> instead of <code class="docutils literal notranslate"><span class="pre">&quot;Lineage&quot;</span></code> and
you’ll get the NCBI taxon identifiers of the lineage entries too.</p>
</section>
</section>
<section id="using-the-history-and-webenv">
<span id="sec-entrez-webenv"></span><h2>Using the history and WebEnv<a class="headerlink" href="#using-the-history-and-webenv" title="Link to this heading"></a></h2>
<p>Often you will want to make a series of linked queries. Most typically,
running a search, perhaps refining the search, and then retrieving
detailed search results. You <em>can</em> do this by making a series of
separate calls to Entrez. However, the NCBI prefer you to take advantage
of their history support - for example combining ESearch and EFetch.</p>
<p>Another typical use of the history support would be to combine EPost and
EFetch. You use EPost to upload a list of identifiers, which starts a
new history session. You then download the records with EFetch by
referring to the session (instead of the identifiers).</p>
<section id="searching-for-and-downloading-sequences-using-the-history">
<h3>Searching for and downloading sequences using the history<a class="headerlink" href="#searching-for-and-downloading-sequences-using-the-history" title="Link to this heading"></a></h3>
<p>Suppose we want to search and download all the <em>Opuntia</em> rpl16
nucleotide sequences, and store them in a FASTA file. As shown in
Section <a class="reference internal" href="#sec-entrez-search-fetch-genbank"><span class="std std-ref">Searching, downloading, and parsing GenBank records</span></a>, we can naively
combine <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.esearch()</span></code> to get a list of Accession numbers, and
then call <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch()</span></code> to download them all.</p>
<p>However, the approved approach is to run the search with the history
feature. Then, we can fetch the results by reference to the search
results - which the NCBI can anticipate and cache.</p>
<p>To do this, call <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.esearch()</span></code> as normal, but with the
additional argument of <code class="docutils literal notranslate"><span class="pre">usehistory=&quot;y&quot;</span></code>,</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;history.user@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;Opuntia[orgn] and rpl16&quot;</span><span class="p">,</span> <span class="n">usehistory</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="n">idtype</span><span class="o">=</span><span class="s2">&quot;acc&quot;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">search_results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">stream</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>As before (see Section <a class="reference internal" href="#sec-entrez-example-genbank"><span class="std std-ref">Searching, downloading, and parsing Entrez Nucleotide records</span></a>), the
XML output includes the first <code class="docutils literal notranslate"><span class="pre">retmax</span></code> search results, with <code class="docutils literal notranslate"><span class="pre">retmax</span></code>
defaulting to 20:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">acc_list</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;IdList&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">acc_list</span><span class="p">)</span>
<span class="go">20</span>
</pre></div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">count</span>
<span class="go">28</span>
</pre></div>
</div>
<p>You also get given two additional pieces of information, the <code class="docutils literal notranslate"><span class="pre">WebEnv</span></code>
session cookie, and the <code class="docutils literal notranslate"><span class="pre">QueryKey</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">webenv</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;WebEnv&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query_key</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;QueryKey&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>Having stored these values in variables <code class="docutils literal notranslate"><span class="pre">session_cookie</span></code> and
<code class="docutils literal notranslate"><span class="pre">query_key</span></code> we can use them as parameters to <code class="docutils literal notranslate"><span class="pre">Bio.Entrez.efetch()</span></code>
instead of giving the GI numbers as identifiers.</p>
<p>While for small searches you might be OK downloading everything at once,
it is better to download in batches. You use the <code class="docutils literal notranslate"><span class="pre">retstart</span></code> and
<code class="docutils literal notranslate"><span class="pre">retmax</span></code> parameters to specify which range of search results you want
returned (starting entry using zero-based counting, and maximum number
of results to return). Note that if Biopython encounters a transient
failure like a HTTP 500 response when communicating with NCBI, it will
automatically try again a couple of times. For example,</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># This assumes you have already run a search as shown above,</span>
<span class="c1"># and set the variables count, webenv, query_key</span>

<span class="n">batch_size</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;orchid_rpl16.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">count</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">):</span>
    <span class="n">end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">start</span> <span class="o">+</span> <span class="n">batch_size</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Going to download record </span><span class="si">%i</span><span class="s2"> to </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
    <span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span>
        <span class="n">db</span><span class="o">=</span><span class="s2">&quot;nucleotide&quot;</span><span class="p">,</span>
        <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">,</span>
        <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">,</span>
        <span class="n">retstart</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
        <span class="n">retmax</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
        <span class="n">webenv</span><span class="o">=</span><span class="n">webenv</span><span class="p">,</span>
        <span class="n">query_key</span><span class="o">=</span><span class="n">query_key</span><span class="p">,</span>
        <span class="n">idtype</span><span class="o">=</span><span class="s2">&quot;acc&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">stream</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>For illustrative purposes, this example downloaded the FASTA records in
batches of three. Unless you are downloading genomes or chromosomes, you
would normally pick a larger batch size.</p>
</section>
<section id="searching-for-and-downloading-abstracts-using-the-history">
<h3>Searching for and downloading abstracts using the history<a class="headerlink" href="#searching-for-and-downloading-abstracts-using-the-history" title="Link to this heading"></a></h3>
<p>Here is another history example, searching for papers published in the
last year about the <em>Opuntia</em>, and then downloading them into a file in
MedLine format:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>

<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;history.user@example.com&quot;</span>
<span class="n">search_results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span>
        <span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s2">&quot;Opuntia[ORGN]&quot;</span><span class="p">,</span> <span class="n">reldate</span><span class="o">=</span><span class="mi">365</span><span class="p">,</span> <span class="n">datetype</span><span class="o">=</span><span class="s2">&quot;pdat&quot;</span><span class="p">,</span> <span class="n">usehistory</span><span class="o">=</span><span class="s2">&quot;y&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;Count&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Found </span><span class="si">%i</span><span class="s2"> results&quot;</span> <span class="o">%</span> <span class="n">count</span><span class="p">)</span>

<span class="n">batch_size</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;recent_orchid_papers.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">count</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">):</span>
    <span class="n">end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">start</span> <span class="o">+</span> <span class="n">batch_size</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Going to download record </span><span class="si">%i</span><span class="s2"> to </span><span class="si">%i</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
    <span class="n">stream</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span>
        <span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span>
        <span class="n">rettype</span><span class="o">=</span><span class="s2">&quot;medline&quot;</span><span class="p">,</span>
        <span class="n">retmode</span><span class="o">=</span><span class="s2">&quot;text&quot;</span><span class="p">,</span>
        <span class="n">retstart</span><span class="o">=</span><span class="n">start</span><span class="p">,</span>
        <span class="n">retmax</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
        <span class="n">webenv</span><span class="o">=</span><span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;WebEnv&quot;</span><span class="p">],</span>
        <span class="n">query_key</span><span class="o">=</span><span class="n">search_results</span><span class="p">[</span><span class="s2">&quot;QueryKey&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">stream</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">stream</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>At the time of writing, this gave 28 matches - but because this is a
date dependent search, this will of course vary. As described in
Section <a class="reference internal" href="#sec-entrez-and-medline"><span class="std std-ref">Parsing Medline records</span></a> above, you can then use
<code class="docutils literal notranslate"><span class="pre">Bio.Medline</span></code> to parse the saved records.</p>
</section>
<section id="searching-for-citations">
<span id="sec-elink-citations"></span><h3>Searching for citations<a class="headerlink" href="#searching-for-citations" title="Link to this heading"></a></h3>
<p>Back in Section <a class="reference internal" href="#sec-elink"><span class="std std-ref">ELink: Searching for related items in NCBI Entrez</span></a> we mentioned ELink can be used to
search for citations of a given paper. Unfortunately this only covers
journals indexed for PubMed Central (doing it for all the journals in
PubMed would mean a lot more work for the NIH). Let’s try this for the
Biopython PDB parser paper, PubMed ID 14630660:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s2">&quot;A.N.Other@example.com&quot;</span>  <span class="c1"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pmid</span> <span class="o">=</span> <span class="s2">&quot;14630660&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">Entrez</span><span class="o">.</span><span class="n">elink</span><span class="p">(</span><span class="n">dbfrom</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="s2">&quot;pmc&quot;</span><span class="p">,</span> <span class="n">LinkName</span><span class="o">=</span><span class="s2">&quot;pubmed_pmc_refs&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">pmid</span><span class="p">)</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pmc_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">link</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">results</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Link&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pmc_ids</span>
<span class="go">[&#39;2744707&#39;, &#39;2705363&#39;, &#39;2682512&#39;, ..., &#39;1190160&#39;]</span>
</pre></div>
</div>
<p>Great - eleven articles. But why hasn’t the Biopython application note
been found (PubMed ID 19304878)? Well, as you might have guessed from
the variable names, there are not actually PubMed IDs, but PubMed
Central IDs. Our application note is the third citing paper in that
list, PMCID 2682512.</p>
<p>So, what if (like me) you’d rather get back a list of PubMed IDs? Well
we can call ELink again to translate them. This becomes a two step
process, so by now you should expect to use the history feature to
accomplish it (Section <a class="reference internal" href="#sec-entrez-webenv"><span class="std std-ref">Using the history and WebEnv</span></a>).</p>
<p>But first, taking the more straightforward approach of making a second
(separate) call to ELink:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">results2</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">Entrez</span><span class="o">.</span><span class="n">elink</span><span class="p">(</span><span class="n">dbfrom</span><span class="o">=</span><span class="s2">&quot;pmc&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="s2">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">LinkName</span><span class="o">=</span><span class="s2">&quot;pmc_pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pmc_ids</span><span class="p">))</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pubmed_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">link</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">results2</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Link&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pubmed_ids</span>
<span class="go">[&#39;19698094&#39;, &#39;19450287&#39;, &#39;19304878&#39;, ..., &#39;15985178&#39;]</span>
</pre></div>
</div>
<p>This time you can immediately spot the Biopython application note as the
third hit (PubMed ID 19304878).</p>
<p>Now, let’s do that all again but with the history … <em>TODO</em>.</p>
<p>And finally, don’t forget to include your <em>own</em> email address in the
Entrez calls.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="chapter_searchio.html" class="btn btn-neutral float-left" title="BLAST and other sequence search tools" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_uniprot.html" class="btn btn-neutral float-right" title="Swiss-Prot and ExPASy" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>