<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Pairwise sequence alignment &mdash; Biopython 1.84.dev0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/biopython.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Multiple Sequence Alignment objects" href="chapter_msa.html" />
    <link rel="prev" title="Sequence alignments" href="chapter_align.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >

          
          
          <a href="../index.html">
            
              <img src="../_static/biopython_logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of contents</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Biopython Tutorial &amp; Cookbook</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="chapter_introduction.html">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_quick_start.html">Quick Start – What can you do with Biopython?</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_objects.html">Sequence objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seq_annot.html">Sequence annotation objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_seqio.html">Sequence Input/Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_align.html">Sequence alignments</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Pairwise sequence alignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#basic-usage">Basic usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#the-pairwise-aligner-object">The pairwise aligner object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#substitution-scores">Substitution scores</a></li>
<li class="toctree-l3"><a class="reference internal" href="#affine-gap-scores">Affine gap scores</a></li>
<li class="toctree-l3"><a class="reference internal" href="#general-gap-scores">General gap scores</a></li>
<li class="toctree-l3"><a class="reference internal" href="#using-a-pre-defined-substitution-matrix-and-gap-scores">Using a pre-defined substitution matrix and gap scores</a></li>
<li class="toctree-l3"><a class="reference internal" href="#iterating-over-alignments">Iterating over alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#aligning-to-the-reverse-strand">Aligning to the reverse strand</a></li>
<li class="toctree-l3"><a class="reference internal" href="#substitution-matrices">Substitution matrices</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#array-objects">Array objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="#calculating-a-substitution-matrix-from-a-pairwise-sequence-alignment">Calculating a substitution matrix from a pairwise sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#calculating-a-substitution-matrix-from-a-multiple-sequence-alignment">Calculating a substitution matrix from a multiple sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reading-array-objects-from-file">Reading <code class="docutils literal notranslate"><span class="pre">Array</span></code> objects from file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#loading-predefined-substitution-matrices">Loading predefined substitution matrices</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a></li>
<li class="toctree-l3"><a class="reference internal" href="#generalized-pairwise-alignments">Generalized pairwise alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#generalized-pairwise-alignments-using-a-substitution-matrix-and-alphabet">Generalized pairwise alignments using a substitution matrix and alphabet</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generalized-pairwise-alignments-using-match-mismatch-scores-and-an-alphabet">Generalized pairwise alignments using match/mismatch scores and an alphabet</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generalized-pairwise-alignments-using-match-mismatch-scores-and-integer-sequences">Generalized pairwise alignments using match/mismatch scores and integer sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generalized-pairwise-alignments-using-a-substitution-matrix-and-integer-sequences">Generalized pairwise alignments using a substitution matrix and integer sequences</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#codon-alignments">Codon alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#aligning-a-nucleotide-sequence-to-an-amino-acid-sequence">Aligning a nucleotide sequence to an amino acid sequence</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generating-a-multiple-sequence-alignment-of-codon-sequences">Generating a multiple sequence alignment of codon sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="#analyzing-a-codon-alignment">Analyzing a codon alignment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chapter_msa.html">Multiple Sequence Alignment objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pairwise2.html">Pairwise alignments using pairwise2</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html">BLAST (new)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_blast.html#blast-old">BLAST (old)</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_searchio.html">BLAST and other sequence search tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_entrez.html">Accessing NCBI’s Entrez databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_uniprot.html">Swiss-Prot and ExPASy</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_pdb.html">Going 3D: The PDB module</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_popgen.html">Bio.PopGen: Population genetics</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phylo.html">Phylogenetics with Bio.Phylo</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_motifs.html">Sequence motif analysis using Bio.motifs</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cluster.html">Cluster analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_graphics.html">Graphics including GenomeDiagram</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_kegg.html">KEGG</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_phenotype.html">Bio.phenotype: analyze phenotypic data</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_cookbook.html">Cookbook – Cool things to do with it</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_testing.html">The Biopython testing framework</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_contributing.html">Where to go from here – contributing to Biopython</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_appendix.html">Appendix: Useful stuff about Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="chapter_bibliography.html">Bibliography</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api/index.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #10100F" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Biopython</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Biopython Tutorial &amp; Cookbook</a></li>
      <li class="breadcrumb-item active">Pairwise sequence alignment</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/biopython/biopython/blob/master/Doc/Tutorial/chapter_pairwise.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul><div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Sequential page navigation">
        <a href="chapter_align.html" class="btn btn-neutral float-left" title="Sequence alignments" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_msa.html" class="btn btn-neutral float-right" title="Multiple Sequence Alignment objects" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
  </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="pairwise-sequence-alignment">
<span id="chapter-pairwise"></span><h1>Pairwise sequence alignment<a class="headerlink" href="#pairwise-sequence-alignment" title="Permalink to this headline"></a></h1>
<p>Pairwise sequence alignment is the process of aligning two sequences to
each other by optimizing the similarity score between them. The
<code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> module contains the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> class for global
and local alignments using the Needleman-Wunsch, Smith-Waterman, Gotoh
(three-state), and Waterman-Smith-Beyer global and local pairwise
alignment algorithms, with numerous options to change the alignment
parameters. We refer to Durbin <em>et al.</em> <a class="reference internal" href="chapter_bibliography.html#durbin1998" id="id1"><span>[Durbin1998]</span></a>
for in-depth information on sequence alignment algorithms.</p>
<section id="basic-usage">
<span id="sec-pairwise-basic"></span><h2>Basic usage<a class="headerlink" href="#basic-usage" title="Permalink to this headline"></a></h2>
<p>To generate pairwise alignments, first create a <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code>
object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object <code class="docutils literal notranslate"><span class="pre">aligner</span></code> (see
Section <a class="reference internal" href="#sec-pairwise-aligner"><span class="std std-ref">The pairwise aligner object</span></a>) stores the alignment parameters
to be used for the pairwise alignments. These attributes can be set in
the constructor of the object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">(</span><span class="n">match_score</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
</pre></div>
</div>
<p>or after the object is made:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">match_score</span> <span class="o">=</span> <span class="mf">1.0</span>
</pre></div>
</div>
<p>Use the <code class="docutils literal notranslate"><span class="pre">aligner.score</span></code> method to calculate the alignment score
between two sequences:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;GAACT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;GAT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">3.0</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">aligner.align</span></code> method returns <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects, each
representing one alignment between the two sequences:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span>  
<span class="go">&lt;Alignment object (2 rows x 5 columns) at ...&gt;</span>
</pre></div>
</div>
<p>Iterate over the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects and print them to see the
alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">target            0 GAACT 5</span>
<span class="go">                  0 ||--| 5</span>
<span class="go">query             0 GA--T 3</span>

<span class="go">target            0 GAACT 5</span>
<span class="go">                  0 |-|-| 5</span>
<span class="go">query             0 G-A-T 3</span>
</pre></div>
</div>
<p>Each alignment stores the alignment score:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span>
<span class="go">3.0</span>
</pre></div>
</div>
<p>as well as pointers to the sequences that were aligned:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">target</span>
<span class="go">&#39;GAACT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">query</span>
<span class="go">&#39;GAT&#39;</span>
</pre></div>
</div>
<p>Internally, the alignment is stored in terms of the sequence
coordinates:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[0, 2, 4, 5],</span>
<span class="go">       [0, 2, 2, 3]])</span>
</pre></div>
</div>
<p>Here, the two rows refer to the target and query sequence. These
coordinates show that the alignment consists of the following three
blocks:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">target[0:2]</span></code> aligned to <code class="docutils literal notranslate"><span class="pre">query[0:2]</span></code>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">target[2:4]</span></code> aligned to a gap, since <code class="docutils literal notranslate"><span class="pre">query[2:2]</span></code> is an empty
string;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">target[4:5]</span></code> aligned to <code class="docutils literal notranslate"><span class="pre">query[2:3]</span></code>.</p></li>
</ul>
<p>The number of aligned sequences is always 2 for a pairwise alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<p>The alignment length is defined as the number of columns in the
alignment as printed. This is equal to the sum of the number of matches,
number of mismatches, and the total length of gaps in the target and
query:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">length</span>
<span class="go">5</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">aligned</span></code> property, which returns the start and end indices of
aligned subsequences, returns two tuples of length 2 for the first
alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 2],</span>
<span class="go">        [4, 5]],</span>

<span class="go">       [[0, 2],</span>
<span class="go">        [2, 3]]])</span>
</pre></div>
</div>
<p>while for the alternative alignment, two tuples of length 3 are
returned:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 GAACT 5</span>
<span class="go">                  0 |-|-| 5</span>
<span class="go">query             0 G-A-T 3</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 1],</span>
<span class="go">        [2, 3],</span>
<span class="go">        [4, 5]],</span>

<span class="go">       [[0, 1],</span>
<span class="go">        [1, 2],</span>
<span class="go">        [2, 3]]])</span>
</pre></div>
</div>
<p>Note that different alignments may have the same subsequences aligned to
each other. In particular, this may occur if alignments differ from each
other in terms of their gap placement only:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;global&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mismatch_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;AAACAAA&quot;</span><span class="p">,</span> <span class="s2">&quot;AAAGAAA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 AAAC-AAA 7</span>
<span class="go">                  0 |||--||| 8</span>
<span class="go">query             0 AAA-GAAA 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 3],</span>
<span class="go">        [4, 7]],</span>

<span class="go">       [[0, 3],</span>
<span class="go">        [4, 7]]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">target            0 AAA-CAAA 7</span>
<span class="go">                  0 |||--||| 8</span>
<span class="go">query             0 AAAG-AAA 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">aligned</span>
<span class="go">array([[[0, 3],</span>
<span class="go">        [4, 7]],</span>

<span class="go">       [[0, 3],</span>
<span class="go">        [4, 7]]])</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">map</span></code> method can be applied on a pairwise alignment <code class="docutils literal notranslate"><span class="pre">alignment1</span></code>
to find the pairwise alignment of the query of <code class="docutils literal notranslate"><span class="pre">alignment2</span></code> to the
target of <code class="docutils literal notranslate"><span class="pre">alignment1</span></code>, where the target of <code class="docutils literal notranslate"><span class="pre">alignment2</span></code> and the
query of <code class="docutils literal notranslate"><span class="pre">alignment1</span></code> are identical. A typical example is where
<code class="docutils literal notranslate"><span class="pre">alignment1</span></code> is the pairwise alignment between a chromosome and a
transcript, <code class="docutils literal notranslate"><span class="pre">alignment2</span></code> is the pairwise alignment between the
transcript and a sequence (e.g., an RNA-seq read), and we want to find
the alignment of the sequence to the chromosome:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;local&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">open_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">extend_gap_score</span> <span class="o">=</span> <span class="mi">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">chromosome</span> <span class="o">=</span> <span class="s2">&quot;AAAAAAAACCCCCCCAAAAAAAAAAAGGGGGGAAAAAAAA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">transcript</span> <span class="o">=</span> <span class="s2">&quot;CCCCCCCGGGGGG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments1</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">transcript</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments1</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span> <span class="o">=</span> <span class="n">alignments1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment1</span><span class="p">)</span>
<span class="go">target            8 CCCCCCCAAAAAAAAAAAGGGGGG 32</span>
<span class="go">                  0 |||||||-----------|||||| 24</span>
<span class="go">query             0 CCCCCCC-----------GGGGGG 13</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;CCCCGGGG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments2</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">transcript</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments2</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span> <span class="o">=</span> <span class="n">alignments2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="go">target            3 CCCCGGGG 11</span>
<span class="go">                  0 ||||||||  8</span>
<span class="go">query             0 CCCCGGGG  8</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">mapped_alignment</span> <span class="o">=</span> <span class="n">alignment1</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mapped_alignment</span><span class="p">)</span>
<span class="go">target           11 CCCCAAAAAAAAAAAGGGG 30</span>
<span class="go">                  0 ||||-----------|||| 19</span>
<span class="go">query             0 CCCC-----------GGGG  8</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">mapped_alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="go">&#39;8\t0\t0\t0\t0\t0\t1\t11\t+\tquery\t8\t0\t8\ttarget\t40\t11\t30\t2\t4,4,\t0,4,\t11,26,\n&#39;</span>
</pre></div>
</div>
<p>Mapping the alignment does not depend on the sequence contents. If we
delete the sequence contents, the same alignment is found in PSL format
(though we obviously lose the ability to print the sequence alignment):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment1</span><span class="o">.</span><span class="n">target</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span><span class="o">.</span><span class="n">query</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment1</span><span class="o">.</span><span class="n">query</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">target</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span><span class="o">.</span><span class="n">query</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">query</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mapped_alignment</span> <span class="o">=</span> <span class="n">alignment1</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">format</span><span class="p">(</span><span class="n">mapped_alignment</span><span class="p">,</span> <span class="s2">&quot;psl&quot;</span><span class="p">)</span>
<span class="go">&#39;8\t0\t0\t0\t0\t0\t1\t11\t+\tquery\t8\t0\t8\ttarget\t40\t11\t30\t2\t4,4,\t0,4,\t11,26,\n&#39;</span>
</pre></div>
</div>
<p>By default, a global pairwise alignment is performed, which finds the
optimal alignment over the whole length of <code class="docutils literal notranslate"><span class="pre">target</span></code> and <code class="docutils literal notranslate"><span class="pre">query</span></code>.
Instead, a local alignment will find the subsequence of <code class="docutils literal notranslate"><span class="pre">target</span></code> and
<code class="docutils literal notranslate"><span class="pre">query</span></code> with the highest alignment score. Local alignments can be
generated by setting <code class="docutils literal notranslate"><span class="pre">aligner.mode</span></code> to <code class="docutils literal notranslate"><span class="pre">&quot;local&quot;</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;local&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;AGAACTC&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;GAACT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">5.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">target            1 GAACT 6</span>
<span class="go">                  0 ||||| 5</span>
<span class="go">query             0 GAACT 5</span>
</pre></div>
</div>
<p>Note that there is some ambiguity in the definition of the best local
alignments if segments with a score 0 can be added to the alignment. We
follow the suggestion by Waterman &amp; Eggert
<a class="reference internal" href="chapter_bibliography.html#waterman1987" id="id2"><span>[Waterman1987]</span></a> and disallow such extensions.</p>
</section>
<section id="the-pairwise-aligner-object">
<span id="sec-pairwise-aligner"></span><h2>The pairwise aligner object<a class="headerlink" href="#the-pairwise-aligner-object" title="Permalink to this headline"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object stores all alignment parameters to be
used for the pairwise alignments. To see an overview of the values for
all parameters, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">(</span><span class="n">match_score</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;local&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">aligner</span><span class="p">)</span>
<span class="go">Pairwise sequence aligner with parameters</span>
<span class="go">  wildcard: None</span>
<span class="go">  match_score: 1.000000</span>
<span class="go">  mismatch_score: 0.000000</span>
<span class="go">  target_internal_open_gap_score: 0.000000</span>
<span class="go">  target_internal_extend_gap_score: 0.000000</span>
<span class="go">  target_left_open_gap_score: 0.000000</span>
<span class="go">  target_left_extend_gap_score: 0.000000</span>
<span class="go">  target_right_open_gap_score: 0.000000</span>
<span class="go">  target_right_extend_gap_score: 0.000000</span>
<span class="go">  query_internal_open_gap_score: 0.000000</span>
<span class="go">  query_internal_extend_gap_score: 0.000000</span>
<span class="go">  query_left_open_gap_score: 0.000000</span>
<span class="go">  query_left_extend_gap_score: 0.000000</span>
<span class="go">  query_right_open_gap_score: 0.000000</span>
<span class="go">  query_right_extend_gap_score: 0.000000</span>
<span class="go">  mode: local</span>
</pre></div>
</div>
<p>See Sections <a class="reference internal" href="#sec-pairwise-substitution-scores"><span class="std std-ref">Substitution scores</span></a>,
<a class="reference internal" href="#sec-pairwise-affine-gapscores"><span class="std std-ref">Affine gap scores</span></a>, and
<a class="reference internal" href="#sec-pairwise-general-gapscores"><span class="std std-ref">General gap scores</span></a> below for the definition of
these parameters. The attribute <code class="docutils literal notranslate"><span class="pre">mode</span></code> (described above in
Section <a class="reference internal" href="#sec-pairwise-basic"><span class="std std-ref">Basic usage</span></a>) can be set equal to <code class="docutils literal notranslate"><span class="pre">&quot;global&quot;</span></code>
or <code class="docutils literal notranslate"><span class="pre">&quot;local&quot;</span></code> to specify global or local pairwise alignment,
respectively.</p>
<p>Depending on the gap scoring parameters (see
Sections <a class="reference internal" href="#sec-pairwise-affine-gapscores"><span class="std std-ref">Affine gap scores</span></a> and
<a class="reference internal" href="#sec-pairwise-general-gapscores"><span class="std std-ref">General gap scores</span></a>) and mode, a
<code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object automatically chooses the appropriate
algorithm to use for pairwise sequence alignment. To verify the selected
algorithm, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">algorithm</span>
<span class="go">&#39;Smith-Waterman&#39;</span>
</pre></div>
</div>
<p>This attribute is read-only.</p>
<p>A <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object also stores the precision <span class="math notranslate nohighlight">\(\epsilon\)</span>
to be used during alignment. The value of <span class="math notranslate nohighlight">\(\epsilon\)</span> is stored in
the attribute <code class="docutils literal notranslate"><span class="pre">aligner.epsilon</span></code>, and by default is equal to
<span class="math notranslate nohighlight">\(10^{-6}\)</span>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">epsilon</span>
<span class="go">1e-06</span>
</pre></div>
</div>
<p>Two scores will be considered equal to each other for the purpose of the
alignment if the absolute difference between them is less than
<span class="math notranslate nohighlight">\(\epsilon\)</span>.</p>
</section>
<section id="substitution-scores">
<span id="sec-pairwise-substitution-scores"></span><h2>Substitution scores<a class="headerlink" href="#substitution-scores" title="Permalink to this headline"></a></h2>
<p>Substitution scores define the value to be added to the total score when
two letters (nucleotides or amino acids) are aligned to each other. The
substitution scores to be used by the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> can be
specified in two ways:</p>
<ul>
<li><p>By specifying a match score for identical letters, and a mismatch
scores for mismatched letters. Nucleotide sequence alignments are
typically based on match and mismatch scores. For example, by default
BLAST <a class="reference internal" href="chapter_bibliography.html#altschul1990" id="id3"><span>[Altschul1990]</span></a> uses a match score of
<span class="math notranslate nohighlight">\(+1\)</span> and a mismatch score of <span class="math notranslate nohighlight">\(-2\)</span> for nucleotide
alignments by <code class="docutils literal notranslate"><span class="pre">megablast</span></code>, with a gap penalty of 2.5 (see section
<a class="reference internal" href="#sec-pairwise-affine-gapscores"><span class="std std-ref">Affine gap scores</span></a> for more information on gap
scores). Match and mismatch scores can be specified by setting the
<code class="docutils literal notranslate"><span class="pre">match</span></code> and <code class="docutils literal notranslate"><span class="pre">mismatch</span></code> attributes of the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code>
object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">match_score</span>
<span class="go">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mismatch_score</span>
<span class="go">0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="s2">&quot;ACAT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="go">3.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">match_score</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mismatch_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">2.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">2.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="s2">&quot;ACAT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="go">1.0</span>
</pre></div>
</div>
<p>When using match and mismatch scores, you can specify a wildcard
character (<code class="docutils literal notranslate"><span class="pre">None</span></code> by default) for unknown letters. These will get a
zero score in alignments, irrespective of the value of the match or
mismatch score:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">wildcard</span> <span class="o">=</span> <span class="s2">&quot;?&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="s2">&quot;AC?T&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="go">3.0</span>
</pre></div>
</div>
</li>
<li><p>Alternatively, you can use the <code class="docutils literal notranslate"><span class="pre">substitution_matrix</span></code> attribute of
the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object to specify a substitution matrix. This
allows you to apply different scores for different pairs of matched
and mismatched letters. This is typically used for amino acid
sequence alignments. For example, by default BLAST
<a class="reference internal" href="chapter_bibliography.html#altschul1990" id="id4"><span>[Altschul1990]</span></a> uses the BLOSUM62 substitution
matrix for protein alignments by <code class="docutils literal notranslate"><span class="pre">blastp</span></code>. This substitution matrix
is available from Biopython:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">substitution_matrices</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>  
<span class="go">[&#39;BENNER22&#39;, &#39;BENNER6&#39;, &#39;BENNER74&#39;, &#39;BLASTN&#39;, &#39;BLASTP&#39;, &#39;BLOSUM45&#39;, &#39;BLOSUM50&#39;, &#39;BLOSUM62&#39;, ..., &#39;TRANS&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">matrix</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;BLOSUM62&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>  
<span class="go">#  Matrix made by matblas from blosum62.iij</span>
<span class="go">...</span>
<span class="go">     A    R    N    D    C    Q ...</span>
<span class="go">A  4.0 -1.0 -2.0 -2.0  0.0 -1.0 ...</span>
<span class="go">R -1.0  5.0  0.0 -2.0 -3.0  1.0 ...</span>
<span class="go">N -2.0  0.0  6.0  1.0 -3.0  0.0 ...</span>
<span class="go">D -2.0 -2.0  1.0  6.0 -3.0  0.0 ...</span>
<span class="go">C  0.0 -3.0 -3.0 -3.0  9.0 -3.0 ...</span>
<span class="go">Q -1.0  1.0  0.0  0.0 -3.0  5.0 ...</span>
<span class="go">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="n">matrix</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="s2">&quot;ACDQ&quot;</span><span class="p">,</span> <span class="s2">&quot;ACDQ&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">24.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="s2">&quot;ACDQ&quot;</span><span class="p">,</span> <span class="s2">&quot;ACNQ&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">19.0</span>
</pre></div>
</div>
<p>When using a substitution matrix, <code class="docutils literal notranslate"><span class="pre">X</span></code> is <em>not</em> interpreted as an
unknown character. Instead, the score provided by the substitution
matrix will be used:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">matrix</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;X&quot;</span><span class="p">]</span>
<span class="go">-1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="s2">&quot;ACDQ&quot;</span><span class="p">,</span> <span class="s2">&quot;ACXQ&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span>
<span class="go">17.0</span>
</pre></div>
</div>
</li>
</ul>
<p>By default, <code class="docutils literal notranslate"><span class="pre">aligner.substitution_matrix</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code>. The attributes
<code class="docutils literal notranslate"><span class="pre">aligner.match_score</span></code> and <code class="docutils literal notranslate"><span class="pre">aligner.mismatch_score</span></code> are ignored if
<code class="docutils literal notranslate"><span class="pre">aligner.substitution_matrix</span></code> is not <code class="docutils literal notranslate"><span class="pre">None</span></code>. Setting
<code class="docutils literal notranslate"><span class="pre">aligner.match_score</span></code> or <code class="docutils literal notranslate"><span class="pre">aligner.mismatch_score</span></code> to valid values
will reset <code class="docutils literal notranslate"><span class="pre">aligner.substitution_matrix</span></code> to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</section>
<section id="affine-gap-scores">
<span id="sec-pairwise-affine-gapscores"></span><h2>Affine gap scores<a class="headerlink" href="#affine-gap-scores" title="Permalink to this headline"></a></h2>
<p>Affine gap scores are defined by a score to open a gap, and a score to
extend an existing gap:</p>
<p><span class="math notranslate nohighlight">\(\textrm{gap score} = \textrm{open gap score} + (n-1) \times \textrm{extend gap score}\)</span>,</p>
<p>where <span class="math notranslate nohighlight">\(n\)</span> is the length of the gap. Biopython’s pairwise sequence
aligner allows fine-grained control over the gap scoring scheme by
specifying the following twelve attributes of a <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code>
object:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 49%" />
<col style="width: 51%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p><strong>Opening scores</strong></p></th>
<th class="head"><p><strong>Extending scores</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">query_left_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_left_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">query_internal_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_internal_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">query_right_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_right_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">target_left_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_left_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">target_internal_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_internal_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">target_right_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_right_extend_gap_score</span></code></p></td>
</tr>
</tbody>
</table>
<p>These attributes allow for different gap scores for internal gaps and on
either end of the sequence, as shown in this example:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 18%" />
<col style="width: 63%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p><strong>target</strong></p></th>
<th class="head"><p><strong>query</strong></p></th>
<th class="head"><p><strong>score</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>A</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>query left open gap score</p></td>
</tr>
<tr class="row-odd"><td><p>C</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>query left extend gap score</p></td>
</tr>
<tr class="row-even"><td><p>C</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>query left extend gap score</p></td>
</tr>
<tr class="row-odd"><td><p>G</p></td>
<td><p>G</p></td>
<td><p>match score</p></td>
</tr>
<tr class="row-even"><td><p>G</p></td>
<td><p>T</p></td>
<td><p>mismatch score</p></td>
</tr>
<tr class="row-odd"><td><p>G</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>query internal open gap score</p></td>
</tr>
<tr class="row-even"><td><p>A</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>query internal extend gap score</p></td>
</tr>
<tr class="row-odd"><td><p>A</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>query internal extend gap score</p></td>
</tr>
<tr class="row-even"><td><p>T</p></td>
<td><p>T</p></td>
<td><p>match score</p></td>
</tr>
<tr class="row-odd"><td><p>A</p></td>
<td><p>A</p></td>
<td><p>match score</p></td>
</tr>
<tr class="row-even"><td><p>G</p></td>
<td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>query internal open gap score</p></td>
</tr>
<tr class="row-odd"><td><p>C</p></td>
<td><p>C</p></td>
<td><p>match score</p></td>
</tr>
<tr class="row-even"><td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>C</p></td>
<td><p>target internal open gap score</p></td>
</tr>
<tr class="row-odd"><td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>C</p></td>
<td><p>target internal extend gap score</p></td>
</tr>
<tr class="row-even"><td><p>C</p></td>
<td><p>C</p></td>
<td><p>match score</p></td>
</tr>
<tr class="row-odd"><td><p>T</p></td>
<td><p>G</p></td>
<td><p>mismatch score</p></td>
</tr>
<tr class="row-even"><td><p>C</p></td>
<td><p>C</p></td>
<td><p>match score</p></td>
</tr>
<tr class="row-odd"><td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>C</p></td>
<td><p>target internal open gap score</p></td>
</tr>
<tr class="row-even"><td><p>A</p></td>
<td><p>A</p></td>
<td><p>match score</p></td>
</tr>
<tr class="row-odd"><td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>T</p></td>
<td><p>target right open gap score</p></td>
</tr>
<tr class="row-even"><td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>A</p></td>
<td><p>target right extend gap score</p></td>
</tr>
<tr class="row-odd"><td><ul class="simple">
<li></li>
</ul>
</td>
<td><p>A</p></td>
<td><p>target right extend gap score</p></td>
</tr>
</tbody>
</table>
<p>For convenience, <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> objects have additional attributes
that refer to a number of these values collectively, as shown
(hierarchically) in Table <a class="reference internal" href="#table-align-meta-attributes"><span class="std std-ref">Meta-attributes of the pairwise aligner objects.</span></a>.</p>
<table class="docutils align-default" id="table-align-meta-attributes">
<caption><span class="caption-number">Table 1 </span><span class="caption-text">Meta-attributes of the pairwise aligner objects.</span><a class="headerlink" href="#table-align-meta-attributes" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 46%" />
<col style="width: 54%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Meta-attribute</p></th>
<th class="head"><p>Attributes it maps to</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_open_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">internal_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_internal_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_internal_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">internal_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_internal_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_internal_open_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">internal_extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_internal_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_internal_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">end_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_end_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_end_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">end_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_end_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_end_open_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">end_extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_end_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_end_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">left_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_left_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_left_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">right_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_right_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_right_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">left_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_left_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_left_open_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">left_extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_left_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_left_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">right_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_right_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_right_open_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">right_extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_right_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_right_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">target_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_internal_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_left_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_right_open_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">target_extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_internal_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_left_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_right_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">target_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">query_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_internal_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_left_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_right_open_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">query_extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_internal_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_left_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_right_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">query_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">target_internal_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_internal_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_internal_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">target_end_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_end_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_end_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">target_end_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_left_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_right_open_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">target_end_extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_left_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_right_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">target_left_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_left_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_left_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">target_right_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">target_right_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">target_right_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">query_end_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_end_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_end_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">query_end_open_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_left_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_right_open_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">query_end_extend_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_left_extend_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_right_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">query_internal_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_internal_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_internal_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">query_left_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_left_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_left_extend_gap_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">query_right_gap_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">query_right_open_gap_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">query_right_extend_gap_score</span></code></p></td>
</tr>
</tbody>
</table>
</section>
<section id="general-gap-scores">
<span id="sec-pairwise-general-gapscores"></span><h2>General gap scores<a class="headerlink" href="#general-gap-scores" title="Permalink to this headline"></a></h2>
<p>For even more fine-grained control over the gap scores, you can specify
a gap scoring function. For example, the gap scoring function below
disallows a gap after two nucleotides in the query sequence:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">my_gap_score_function</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">length</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">start</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="o">-</span><span class="mi">1000</span>
<span class="gp">... </span>    <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">length</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">query_gap_score</span> <span class="o">=</span> <span class="n">my_gap_score_function</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;AACTT&quot;</span><span class="p">,</span> <span class="s2">&quot;AATT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">target            0 AACTT 5</span>
<span class="go">                  0 -|.|| 5</span>
<span class="go">query             0 -AATT 4</span>

<span class="go">target            0 AACTT 5</span>
<span class="go">                  0 |-.|| 5</span>
<span class="go">query             0 A-ATT 4</span>

<span class="go">target            0 AACTT 5</span>
<span class="go">                  0 ||.-| 5</span>
<span class="go">query             0 AAT-T 4</span>

<span class="go">target            0 AACTT 5</span>
<span class="go">                  0 ||.|- 5</span>
<span class="go">query             0 AATT- 4</span>
</pre></div>
</div>
</section>
<section id="using-a-pre-defined-substitution-matrix-and-gap-scores">
<span id="sec-pairwise-predefined-scoring"></span><h2>Using a pre-defined substitution matrix and gap scores<a class="headerlink" href="#using-a-pre-defined-substitution-matrix-and-gap-scores" title="Permalink to this headline"></a></h2>
<p>By default, a <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object is initialized with a match
score of +1.0, a mismatch score of 0.0, and all gap scores equal to 0.0,
While this has the benefit of being a simple scoring scheme, in general
it does not give the best performance. Instead, you can use the argument
<code class="docutils literal notranslate"><span class="pre">scoring</span></code> to select a predefined scoring scheme when initializing a
<code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object. Currently, the provided scoring schemes are
<code class="docutils literal notranslate"><span class="pre">blastn</span></code> and <code class="docutils literal notranslate"><span class="pre">megablast</span></code>, which are suitable for nucleotide
alignments, and <code class="docutils literal notranslate"><span class="pre">blastp</span></code>, which is suitable for protein alignments.
Selecting these scoring schemes will initialize the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code>
object to the default scoring parameters used by BLASTN, MegaBLAST, and
BLASTP, respectively.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">(</span><span class="n">scoring</span><span class="o">=</span><span class="s2">&quot;blastn&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">aligner</span><span class="p">)</span>  
<span class="go">Pairwise sequence aligner with parameters</span>
<span class="go">  substitution_matrix: &lt;Array object at ...&gt;</span>
<span class="go">  target_internal_open_gap_score: -7.000000</span>
<span class="go">  target_internal_extend_gap_score: -2.000000</span>
<span class="go">  target_left_open_gap_score: -7.000000</span>
<span class="go">  target_left_extend_gap_score: -2.000000</span>
<span class="go">  target_right_open_gap_score: -7.000000</span>
<span class="go">  target_right_extend_gap_score: -2.000000</span>
<span class="go">  query_internal_open_gap_score: -7.000000</span>
<span class="go">  query_internal_extend_gap_score: -2.000000</span>
<span class="go">  query_left_open_gap_score: -7.000000</span>
<span class="go">  query_left_extend_gap_score: -2.000000</span>
<span class="go">  query_right_open_gap_score: -7.000000</span>
<span class="go">  query_right_extend_gap_score: -2.000000</span>
<span class="go">  mode: global</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span><span class="p">[:,</span> <span class="p">:])</span>
<span class="go">     A    T    G    C    S    W    R    Y    K    M    B    V    H    D    N</span>
<span class="go">A  2.0 -3.0 -3.0 -3.0 -3.0 -1.0 -1.0 -3.0 -3.0 -1.0 -3.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">T -3.0  2.0 -3.0 -3.0 -3.0 -1.0 -3.0 -1.0 -1.0 -3.0 -1.0 -3.0 -1.0 -1.0 -2.0</span>
<span class="go">G -3.0 -3.0  2.0 -3.0 -1.0 -3.0 -1.0 -3.0 -1.0 -3.0 -1.0 -1.0 -3.0 -1.0 -2.0</span>
<span class="go">C -3.0 -3.0 -3.0  2.0 -1.0 -3.0 -3.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -3.0 -2.0</span>
<span class="go">S -3.0 -3.0 -1.0 -1.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">W -1.0 -1.0 -3.0 -3.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">R -1.0 -3.0 -1.0 -3.0 -1.0 -1.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">Y -3.0 -1.0 -3.0 -1.0 -1.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">K -3.0 -1.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">M -1.0 -3.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">B -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">V -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">H -1.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">D -1.0 -1.0 -1.0 -3.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0</span>
<span class="go">N -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0</span>
</pre></div>
</div>
</section>
<section id="iterating-over-alignments">
<h2>Iterating over alignments<a class="headerlink" href="#iterating-over-alignments" title="Permalink to this headline"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">alignments</span></code> returned by <code class="docutils literal notranslate"><span class="pre">aligner.align</span></code> are a kind of immutable
iterable objects (similar to <code class="docutils literal notranslate"><span class="pre">range</span></code>). While they appear similar to a
<code class="docutils literal notranslate"><span class="pre">tuple</span></code> or <code class="docutils literal notranslate"><span class="pre">list</span></code> of <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects, they are different in
the sense that each <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object is created dynamically when it
is needed. This approach was chosen because the number of alignments can
be extremely large, in particular for poor alignments (see
Section <a class="reference internal" href="#sec-pairwise-examples"><span class="std std-ref">Examples</span></a> for an example).</p>
<p>You can perform the following operations on <code class="docutils literal notranslate"><span class="pre">alignments</span></code>:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">len(alignments)</span></code> returns the number of alignments stored. This
function returns quickly, even if the number of alignments is huge.
If the number of alignments is extremely large (typically, larger
than 9,223,372,036,854,775,807, which is the largest integer that can
be stored as a <code class="docutils literal notranslate"><span class="pre">long</span> <span class="pre">int</span></code> on 64 bit machines), <code class="docutils literal notranslate"><span class="pre">len(alignments)</span></code>
will raise an <code class="docutils literal notranslate"><span class="pre">OverflowError</span></code>. A large number of alignments
suggests that the alignment quality is low.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span> <span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
</li>
<li><p>You can extract a specific alignment by index:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span> <span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="go">target            0 AAA 3</span>
<span class="go">                  0 -|| 3</span>
<span class="go">query             0 -AA 2</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 AAA 3</span>
<span class="go">                  0 ||- 3</span>
<span class="go">query             0 AA- 2</span>
</pre></div>
</div>
</li>
<li><p>You can iterate over alignments, for example as in</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>By calling <code class="docutils literal notranslate"><span class="pre">alignments.rewind</span></code>, you can rewind the <code class="docutils literal notranslate"><span class="pre">alignments</span></code>
iterator to the first alignment and iterate over the alignments from
the beginning:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;AAA&quot;</span><span class="p">,</span> <span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">target            0 AAA 3</span>
<span class="go">                  0 ||- 3</span>
<span class="go">query             0 AA- 2</span>

<span class="go">target            0 AAA 3</span>
<span class="go">                  0 |-| 3</span>
<span class="go">query             0 A-A 2</span>

<span class="go">target            0 AAA 3</span>
<span class="go">                  0 -|| 3</span>
<span class="go">query             0 -AA 2</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span><span class="o">.</span><span class="n">rewind</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">target            0 AAA 3</span>
<span class="go">                  0 ||- 3</span>
<span class="go">query             0 AA- 2</span>

<span class="go">target            0 AAA 3</span>
<span class="go">                  0 |-| 3</span>
<span class="go">query             0 A-A 2</span>

<span class="go">target            0 AAA 3</span>
<span class="go">                  0 -|| 3</span>
<span class="go">query             0 -AA 2</span>
</pre></div>
</div>
<p>You can also convert the <code class="docutils literal notranslate"><span class="pre">alignments</span></code> iterator into a <code class="docutils literal notranslate"><span class="pre">list</span></code> or
<code class="docutils literal notranslate"><span class="pre">tuple</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
</pre></div>
</div>
<p>It is wise to check the number of alignments by calling
<code class="docutils literal notranslate"><span class="pre">len(alignments)</span></code> before attempting to call <code class="docutils literal notranslate"><span class="pre">list(alignments)</span></code> to
save all alignments as a list.</p>
</li>
<li><p>The alignment score (which has the same value for each alignment in
<code class="docutils literal notranslate"><span class="pre">alignments</span></code>) is stored as an attribute. This allows you to check
the alignment score before proceeding to extract individual
alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">2.0</span>
</pre></div>
</div>
</li>
</ul>
</section>
<section id="aligning-to-the-reverse-strand">
<h2>Aligning to the reverse strand<a class="headerlink" href="#aligning-to-the-reverse-strand" title="Permalink to this headline"></a></h2>
<p>By default, the pairwise aligner aligns the forward strand of the query
to the forward strand of the target. To calculate the alignment score
for <code class="docutils literal notranslate"><span class="pre">query</span></code> to the reverse strand of <code class="docutils literal notranslate"><span class="pre">target</span></code>, use <code class="docutils literal notranslate"><span class="pre">strand=&quot;-&quot;</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">reverse_complement</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">target</span> <span class="o">=</span> <span class="s2">&quot;AAAACCC&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;AACC&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mismatch_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">internal_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>  <span class="c1"># strand is &quot;+&quot; by default</span>
<span class="go">4.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">query</span><span class="p">),</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="go">4.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="go">0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">query</span><span class="p">))</span>
<span class="go">0.0</span>
</pre></div>
</div>
<p>The alignments against the reverse strand can be obtained by specifying
<code class="docutils literal notranslate"><span class="pre">strand=&quot;-&quot;</span></code> when calling <code class="docutils literal notranslate"><span class="pre">aligner.align</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 AAAACCC 7</span>
<span class="go">                  0 --||||- 7</span>
<span class="go">query             0 --AACC- 4</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">target   2   6   query   4   +   2   6   0   1   4,   0,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">query</span><span class="p">),</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 AAAACCC 7</span>
<span class="go">                  0 --||||- 7</span>
<span class="go">query             4 --AACC- 0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">))</span>  
<span class="go">target   2   6   query   4   -   2   6   0   1   4,   0,</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 AAAACCC----  7</span>
<span class="go">                  0 ----------- 11</span>
<span class="go">query             4 -------GGTT  0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">target            0 ----AAAACCC  7</span>
<span class="go">                  0 ----------- 11</span>
<span class="go">query             4 GGTT-------  0</span>
</pre></div>
</div>
<p>Note that the score for aligning <code class="docutils literal notranslate"><span class="pre">query</span></code> to the reverse strand of
<code class="docutils literal notranslate"><span class="pre">target</span></code> may be different from the score for aligning the reverse
complement of <code class="docutils literal notranslate"><span class="pre">query</span></code> to the forward strand of <code class="docutils literal notranslate"><span class="pre">target</span></code> if the left
and right gap scores are different:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">left_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">right_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="go">2.8</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">query</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 AAAACCC 7</span>
<span class="go">                  0 --||||- 7</span>
<span class="go">query             0 --AACC- 4</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">query</span><span class="p">),</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="go">3.1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">target</span><span class="p">,</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">query</span><span class="p">),</span> <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 AAAACCC 7</span>
<span class="go">                  0 --||||- 7</span>
<span class="go">query             4 --AACC- 0</span>
</pre></div>
</div>
</section>
<section id="substitution-matrices">
<span id="sec-substitution-matrices"></span><h2>Substitution matrices<a class="headerlink" href="#substitution-matrices" title="Permalink to this headline"></a></h2>
<p>Substitution matrices <a class="reference internal" href="chapter_bibliography.html#durbin1998" id="id5"><span>[Durbin1998]</span></a> provide the scoring
terms for classifying how likely two different residues are to
substitute for each other. This is essential in doing sequence
comparisons. Biopython provides a ton of common substitution matrices,
including the famous PAM and BLOSUM series of matrices, and also
provides functionality for creating your own substitution matrices.</p>
<section id="array-objects">
<h3>Array objects<a class="headerlink" href="#array-objects" title="Permalink to this headline"></a></h3>
<p>You can think of substitutions matrices as 2D arrays in which the
indices are letters (nucleotides or amino acids) rather than integers.
The <code class="docutils literal notranslate"><span class="pre">Array</span></code> class in <code class="docutils literal notranslate"><span class="pre">Bio.Align.substitution_matrices</span></code> is a subclass
of numpy arrays that supports indexing both by integers and by specific
strings. An <code class="docutils literal notranslate"><span class="pre">Array</span></code> instance can either be a one-dimensional array or
a square two-dimensional arrays. A one-dimensional <code class="docutils literal notranslate"><span class="pre">Array</span></code> object can
for example be used to store the nucleotide frequency of a DNA sequence,
while a two-dimensional <code class="docutils literal notranslate"><span class="pre">Array</span></code> object can be used to represent a
scoring matrix for sequence alignments.</p>
<p>To create a one-dimensional <code class="docutils literal notranslate"><span class="pre">Array</span></code>, only the alphabet of allowed
letters needs to be specified:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.substitution_matrices</span> <span class="kn">import</span> <span class="n">Array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">Array</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">A 0.0</span>
<span class="go">C 0.0</span>
<span class="go">G 0.0</span>
<span class="go">T 0.0</span>
</pre></div>
</div>
<p>The allowed letters are stored in the <code class="docutils literal notranslate"><span class="pre">alphabet</span></code> property:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">&#39;ACGT&#39;</span>
</pre></div>
</div>
<p>This property is read-only; modifying the underlying <code class="docutils literal notranslate"><span class="pre">_alphabet</span></code>
attribute may lead to unexpected results. Elements can be accessed both
by letter and by integer index:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="mi">7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">A  0.0</span>
<span class="go">C -3.0</span>
<span class="go">G  7.0</span>
<span class="go">T  0.0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">-3.0</span>
</pre></div>
</div>
<p>Using a letter that is not in the alphabet, or an index that is out of
bounds, will cause a <code class="docutils literal notranslate"><span class="pre">IndexError</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;U&quot;</span><span class="p">]</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">IndexError</span>: <span class="n">&#39;U&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">6</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">IndexError</span>: <span class="n">&#39;X&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">IndexError</span>: <span class="n">index 7 is out of bounds for axis 0 with size 4</span>
</pre></div>
</div>
<p>A two-dimensional <code class="docutils literal notranslate"><span class="pre">Array</span></code> can be created by specifying <code class="docutils literal notranslate"><span class="pre">dims=2</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.substitution_matrices</span> <span class="kn">import</span> <span class="n">Array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">Array</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="n">dims</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">    A   C   G   T</span>
<span class="go">A 0.0 0.0 0.0 0.0</span>
<span class="go">C 0.0 0.0 0.0 0.0</span>
<span class="go">G 0.0 0.0 0.0 0.0</span>
<span class="go">T 0.0 0.0 0.0 0.0</span>
</pre></div>
</div>
<p>Again, both letters and integers can be used for indexing, and
specifying a letter that is not in the alphabet will cause an
<code class="docutils literal notranslate"><span class="pre">IndexError</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">12.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mf">5.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">    A    C   G    T</span>
<span class="go">A 0.0 12.0 0.0  0.0</span>
<span class="go">C 0.0  0.0 0.0  0.0</span>
<span class="go">G 0.0  5.0 0.0  0.0</span>
<span class="go">T 0.0  0.0 0.0 -2.0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">IndexError</span>: <span class="n">&#39;X&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">5</span><span class="p">]</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">IndexError</span>: <span class="n">index 5 is out of bounds for axis 1 with size 4</span>
</pre></div>
</div>
<p>Selecting a row or column from the two-dimensional array will return a
one-dimensional <code class="docutils literal notranslate"><span class="pre">Array</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">Array</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="n">dims</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">12.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mf">5.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span>
</pre></div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s2">&quot;G&quot;</span><span class="p">]</span>
<span class="go">Array([0., 5., 0., 0.],</span>
<span class="go">      alphabet=&#39;ACGT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[:,</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span>
<span class="go">Array([12.,  0.,  5.,  0.],</span>
<span class="go">      alphabet=&#39;ACGT&#39;)</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Array</span></code> objects can thus be used as an array and as a dictionary. They
can be converted to plain numpy arrays or plain dictionary objects:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">Array</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span>
</pre></div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">Array([0., 5., 0., 0.],</span>
<span class="go">      alphabet=&#39;ACGT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>  <span class="c1"># create a plain numpy array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">array([0., 5., 0., 0.])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>  <span class="c1"># create a plain dictionary</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span>
<span class="go">{&#39;A&#39;: 0.0, &#39;C&#39;: 5.0, &#39;G&#39;: 0.0, &#39;T&#39;: 0.0}</span>
</pre></div>
</div>
<p>While the alphabet of an <code class="docutils literal notranslate"><span class="pre">Array</span></code> is usually a string, you may also use
a tuple of (immutable) objects. This is used for example for a codon
substitution matrix (as in the
<code class="docutils literal notranslate"><span class="pre">substitution_matrices.load(&quot;SCHNEIDER&quot;)</span></code> example shown later), where
the keys are not individual nucleotides or amino acids but instead
three-nucleotide codons.</p>
<p>While the <code class="docutils literal notranslate"><span class="pre">alphabet</span></code> property of an <code class="docutils literal notranslate"><span class="pre">Array</span></code> is immutable, you can
create a new <code class="docutils literal notranslate"><span class="pre">Array</span></code> object by selecting the letters you are
interested in from the alphabet. For example,</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">Array</span><span class="p">(</span><span class="s2">&quot;ABCD&quot;</span><span class="p">,</span> <span class="n">dims</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">16</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">     A    B    C    D</span>
<span class="go">A  0.0  1.0  2.0  3.0</span>
<span class="go">B  4.0  5.0  6.0  7.0</span>
<span class="go">C  8.0  9.0 10.0 11.0</span>
<span class="go">D 12.0 13.0 14.0 15.0</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;CAD&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">b</span><span class="p">)</span>
<span class="go">     C    A    D</span>
<span class="go">C 10.0  8.0 11.0</span>
<span class="go">A  2.0  0.0  3.0</span>
<span class="go">D 14.0 12.0 15.0</span>
</pre></div>
</div>
<p>Note that this also allows you to reorder the alphabet.</p>
<p>Data for letters that are not found in the alphabet are set to zero:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;DEC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
<span class="go">     D   E    C</span>
<span class="go">D 15.0 0.0 14.0</span>
<span class="go">E  0.0 0.0  0.0</span>
<span class="go">C 11.0 0.0 10.0</span>
</pre></div>
</div>
<p>As the <code class="docutils literal notranslate"><span class="pre">Array</span></code> class is a subclass of numpy array, it can be used as
such. A <code class="docutils literal notranslate"><span class="pre">ValueError</span></code> is triggered if the <code class="docutils literal notranslate"><span class="pre">Array</span></code> objects appearing
in a mathematical operation have different alphabets, for example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.substitution_matrices</span> <span class="kn">import</span> <span class="n">Array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">Array</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="n">Array</span><span class="p">(</span><span class="s2">&quot;ACGU&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">+</span> <span class="n">r</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">alphabets are inconsistent</span>
</pre></div>
</div>
</section>
<section id="calculating-a-substitution-matrix-from-a-pairwise-sequence-alignment">
<h3>Calculating a substitution matrix from a pairwise sequence alignment<a class="headerlink" href="#calculating-a-substitution-matrix-from-a-pairwise-sequence-alignment" title="Permalink to this headline"></a></h3>
<p>As <code class="docutils literal notranslate"><span class="pre">Array</span></code> is a subclass of a numpy array, you can apply mathematical
operations on an <code class="docutils literal notranslate"><span class="pre">Array</span></code> object in much the same way. Here, we
illustrate this by calculating a scoring matrix from the alignment of
the 16S ribosomal RNA gene sequences of <em>Escherichia coli</em> and <em>Bacillus
subtilis</em>. First, we create a <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> object (see
Chapter <a class="reference internal" href="#chapter-pairwise"><span class="std std-ref">Pairwise sequence alignment</span></a>) and initialize it with the default
scores used by <code class="docutils literal notranslate"><span class="pre">blastn</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">(</span><span class="n">scoring</span><span class="o">=</span><span class="s2">&quot;blastn&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;local&quot;</span>
</pre></div>
</div>
<p>Next, we read in the 16S ribosomal RNA gene sequence of <em>Escherichia
coli</em> and <em>Bacillus subtilis</em> (provided in <code class="docutils literal notranslate"><span class="pre">Tests/Align/ecoli.fa</span></code> and
<code class="docutils literal notranslate"><span class="pre">Tests/Align/bsubtilis.fa</span></code>), and align them to each other:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence1</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;ecoli.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequence2</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;bsubtilis.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">sequence1</span><span class="p">,</span> <span class="n">sequence2</span><span class="p">)</span>
</pre></div>
</div>
<p>The number of alignments generated is very large:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1990656</span>
</pre></div>
</div>
<p>However, as they only differ trivially from each other, we arbitrarily
choose the first alignment, and count the number of each substitution:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">substitutions</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">substitutions</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">substitutions</span><span class="p">)</span>
<span class="go">      A     C     G     T</span>
<span class="go">A 307.0  19.0  34.0  19.0</span>
<span class="go">C  15.0 280.0  25.0  29.0</span>
<span class="go">G  34.0  24.0 401.0  20.0</span>
<span class="go">T  24.0  36.0  20.0 228.0</span>
</pre></div>
</div>
<p>We normalize against the total number to find the probability of each
substitution, and create a symmetric matrix of observed frequencies:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">observed_frequencies</span> <span class="o">=</span> <span class="n">substitutions</span> <span class="o">/</span> <span class="n">substitutions</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">observed_frequencies</span> <span class="o">=</span> <span class="p">(</span><span class="n">observed_frequencies</span> <span class="o">+</span> <span class="n">observed_frequencies</span><span class="o">.</span><span class="n">transpose</span><span class="p">())</span> <span class="o">/</span> <span class="mf">2.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">observed_frequencies</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">%.4f</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="go">       A      C      G      T</span>
<span class="go">A 0.2026 0.0112 0.0224 0.0142</span>
<span class="go">C 0.0112 0.1848 0.0162 0.0215</span>
<span class="go">G 0.0224 0.0162 0.2647 0.0132</span>
<span class="go">T 0.0142 0.0215 0.0132 0.1505</span>
</pre></div>
</div>
<p>The background probability is the probability of finding an A, C, G, or
T nucleotide in each sequence separately. This can be calculated as the
sum of each row or column:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">background</span> <span class="o">=</span> <span class="n">observed_frequencies</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">background</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">%.4f</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="go">A 0.2505</span>
<span class="go">C 0.2337</span>
<span class="go">G 0.3165</span>
<span class="go">T 0.1993</span>
</pre></div>
</div>
<p>The number of substitutions expected at random is simply the product of
the background distribution with itself:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">expected_frequencies</span> <span class="o">=</span> <span class="n">background</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">background</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">expected_frequencies</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">%.4f</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="go">       A      C      G      T</span>
<span class="go">A 0.0627 0.0585 0.0793 0.0499</span>
<span class="go">C 0.0585 0.0546 0.0740 0.0466</span>
<span class="go">G 0.0793 0.0740 0.1002 0.0631</span>
<span class="go">T 0.0499 0.0466 0.0631 0.0397</span>
</pre></div>
</div>
<p>The scoring matrix can then be calculated as the logarithm of the
odds-ratio of the observed and the expected probabilities:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">oddsratios</span> <span class="o">=</span> <span class="n">observed_frequencies</span> <span class="o">/</span> <span class="n">expected_frequencies</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scoring_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">oddsratios</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scoring_matrix</span><span class="p">)</span>
<span class="go">     A    C    G    T</span>
<span class="go">A  1.7 -2.4 -1.8 -1.8</span>
<span class="go">C -2.4  1.8 -2.2 -1.1</span>
<span class="go">G -1.8 -2.2  1.4 -2.3</span>
<span class="go">T -1.8 -1.1 -2.3  1.9</span>
</pre></div>
</div>
<p>The matrix can be used to set the substitution matrix for the pairwise
aligner (see Chapter <a class="reference internal" href="#chapter-pairwise"><span class="std std-ref">Pairwise sequence alignment</span></a>):</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="n">scoring_matrix</span>
</pre></div>
</div>
</section>
<section id="calculating-a-substitution-matrix-from-a-multiple-sequence-alignment">
<span id="subsec-subs-mat-ex"></span><h3>Calculating a substitution matrix from a multiple sequence alignment<a class="headerlink" href="#calculating-a-substitution-matrix-from-a-multiple-sequence-alignment" title="Permalink to this headline"></a></h3>
<p>In this example, we’ll first read a protein sequence alignment from the
Clustalw file <a class="reference external" href="examples/protein.aln">protein.aln</a> (also available
online
<a class="reference external" href="https://raw.githubusercontent.com/biopython/biopython/master/Tests/Clustalw/protein.aln">here</a>)</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;protein.aln&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Section <a class="reference internal" href="chapter_align.html#subsec-align-clustal"><span class="std std-ref">ClustalW</span></a> contains more
information on doing this.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">substitutions</span></code> property of the alignment stores the number of
times different residues substitute for each other:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">substitutions</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">substitutions</span>
</pre></div>
</div>
<p>To make the example more readable, we’ll select only amino acids with
polar charged side chains:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">substitutions</span> <span class="o">=</span> <span class="n">substitutions</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;DEHKR&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">substitutions</span><span class="p">)</span>
<span class="go">       D      E      H      K      R</span>
<span class="go">D 2360.0  270.0   15.0    1.0   48.0</span>
<span class="go">E  241.0 3305.0   15.0   45.0    2.0</span>
<span class="go">H    0.0   18.0 1235.0    8.0    0.0</span>
<span class="go">K    0.0    9.0   24.0 3218.0  130.0</span>
<span class="go">R    2.0    2.0   17.0  103.0 2079.0</span>
</pre></div>
</div>
<p>Rows and columns for other amino acids were removed from the matrix.</p>
<p>Next, we normalize the matrix and make it symmetric.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">observed_frequencies</span> <span class="o">=</span> <span class="n">substitutions</span> <span class="o">/</span> <span class="n">substitutions</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">observed_frequencies</span> <span class="o">=</span> <span class="p">(</span><span class="n">observed_frequencies</span> <span class="o">+</span> <span class="n">observed_frequencies</span><span class="o">.</span><span class="n">transpose</span><span class="p">())</span> <span class="o">/</span> <span class="mf">2.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">observed_frequencies</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">%.4f</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="go">       D      E      H      K      R</span>
<span class="go">D 0.1795 0.0194 0.0006 0.0000 0.0019</span>
<span class="go">E 0.0194 0.2514 0.0013 0.0021 0.0002</span>
<span class="go">H 0.0006 0.0013 0.0939 0.0012 0.0006</span>
<span class="go">K 0.0000 0.0021 0.0012 0.2448 0.0089</span>
<span class="go">R 0.0019 0.0002 0.0006 0.0089 0.1581</span>
</pre></div>
</div>
<p>Summing over rows or columns gives the relative frequency of occurrence
of each residue:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">background</span> <span class="o">=</span> <span class="n">observed_frequencies</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">background</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">%.4f</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="go">D 0.2015</span>
<span class="go">E 0.2743</span>
<span class="go">H 0.0976</span>
<span class="go">K 0.2569</span>
<span class="go">R 0.1697</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">background</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="go">1.0</span>
</pre></div>
</div>
<p>The expected frequency of residue pairs is then</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">expected_frequencies</span> <span class="o">=</span> <span class="n">background</span><span class="p">[:,</span> <span class="kc">None</span><span class="p">]</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">background</span><span class="p">[</span><span class="kc">None</span><span class="p">,</span> <span class="p">:])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">expected_frequencies</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">%.4f</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="go">       D      E      H      K      R</span>
<span class="go">D 0.0406 0.0553 0.0197 0.0518 0.0342</span>
<span class="go">E 0.0553 0.0752 0.0268 0.0705 0.0465</span>
<span class="go">H 0.0197 0.0268 0.0095 0.0251 0.0166</span>
<span class="go">K 0.0518 0.0705 0.0251 0.0660 0.0436</span>
<span class="go">R 0.0342 0.0465 0.0166 0.0436 0.0288</span>
</pre></div>
</div>
<p>Here, <code class="docutils literal notranslate"><span class="pre">background[:,</span> <span class="pre">None]</span></code> creates a 2D array consisting of a single
column with the values of <code class="docutils literal notranslate"><span class="pre">expected_frequencies</span></code>, and
<code class="docutils literal notranslate"><span class="pre">rxpected_frequencies[None,</span> <span class="pre">:]</span></code> a 2D array with these values as a
single row. Taking their dot product (inner product) creates a matrix of
expected frequencies where each entry consists of two
<code class="docutils literal notranslate"><span class="pre">expected_frequencies</span></code> values multiplied with each other. For example,
<code class="docutils literal notranslate"><span class="pre">expected_frequencies['D',</span> <span class="pre">'E']</span></code> is equal to
<code class="docutils literal notranslate"><span class="pre">residue_frequencies['D']</span> <span class="pre">*</span> <span class="pre">residue_frequencies['E']</span></code>.</p>
<p>We can now calculate the log-odds matrix by dividing the observed
frequencies by the expected frequencies and taking the logarithm:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scoring_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">observed_frequencies</span> <span class="o">/</span> <span class="n">expected_frequencies</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">scoring_matrix</span><span class="p">)</span>
<span class="go">      D    E    H     K    R</span>
<span class="go">D   2.1 -1.5 -5.1 -10.4 -4.2</span>
<span class="go">E  -1.5  1.7 -4.4  -5.1 -8.3</span>
<span class="go">H  -5.1 -4.4  3.3  -4.4 -4.7</span>
<span class="go">K -10.4 -5.1 -4.4   1.9 -2.3</span>
<span class="go">R  -4.2 -8.3 -4.7  -2.3  2.5</span>
</pre></div>
</div>
<p>This matrix can be used as the substitution matrix when performing
alignments. For example,</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="n">scoring_matrix</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">3.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;DEHEK&quot;</span><span class="p">,</span> <span class="s2">&quot;DHHKK&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 DEHEK 5</span>
<span class="go">                  0 |.|.| 5</span>
<span class="go">query             0 DHHKK 5</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">alignments</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">-2.18</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="n">scoring_matrix</span><span class="p">[</span><span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="o">+</span> <span class="n">scoring_matrix</span><span class="p">[</span><span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;H&quot;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="o">+</span> <span class="n">scoring_matrix</span><span class="p">[</span><span class="s2">&quot;H&quot;</span><span class="p">,</span> <span class="s2">&quot;H&quot;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="o">+</span> <span class="n">scoring_matrix</span><span class="p">[</span><span class="s2">&quot;E&quot;</span><span class="p">,</span> <span class="s2">&quot;K&quot;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="o">+</span> <span class="n">scoring_matrix</span><span class="p">[</span><span class="s2">&quot;K&quot;</span><span class="p">,</span> <span class="s2">&quot;K&quot;</span><span class="p">]</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">score</span><span class="p">)</span>
<span class="go">-2.18</span>
</pre></div>
</div>
<p>(see Chapter <a class="reference internal" href="#chapter-pairwise"><span class="std std-ref">Pairwise sequence alignment</span></a> for details).</p>
</section>
<section id="reading-array-objects-from-file">
<h3>Reading <code class="docutils literal notranslate"><span class="pre">Array</span></code> objects from file<a class="headerlink" href="#reading-array-objects-from-file" title="Permalink to this headline"></a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Bio.Align.substitution_matrices</span></code> includes a parser to read one- and
two-dimensional <code class="docutils literal notranslate"><span class="pre">Array</span></code> objects from file. One-dimensional arrays are
represented by a simple two-column format, with the first column
containing the key and the second column the corresponding value. For
example, the file <code class="docutils literal notranslate"><span class="pre">hg38.chrom.sizes</span></code> (obtained from UCSC), available
in the <code class="docutils literal notranslate"><span class="pre">Tests/Align</span></code> subdirectory of the Biopython distribution,
contains the size in nucleotides of each chromosome in human genome
assembly hg38:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>chr1    248956422
chr2    242193529
chr3    198295559
chr4    190214555
...
chrUn_KI270385v1    990
chrUn_KI270423v1    981
chrUn_KI270392v1    971
chrUn_KI270394v1    970
</pre></div>
</div>
<p>To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">substitution_matrices</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;hg38.chrom.sizes&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">table</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>  
<span class="go">chr1 248956422.0</span>
<span class="go">chr2 242193529.0</span>
<span class="go">chr3 198295559.0</span>
<span class="go">chr4 190214555.0</span>
<span class="go">...</span>
<span class="go">chrUn_KI270423v1       981.0</span>
<span class="go">chrUn_KI270392v1       971.0</span>
<span class="go">chrUn_KI270394v1       970.0</span>
</pre></div>
</div>
<p>Use <code class="docutils literal notranslate"><span class="pre">dtype=int</span></code> to read the values as integers:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;hg38.chrom.sizes&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">table</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">table</span><span class="p">)</span>  
<span class="go">chr1 248956422</span>
<span class="go">chr2 242193529</span>
<span class="go">chr3 198295559</span>
<span class="go">chr4 190214555</span>
<span class="go">...</span>
<span class="go">chrUn_KI270423v1       981</span>
<span class="go">chrUn_KI270392v1       971</span>
<span class="go">chrUn_KI270394v1       970</span>
</pre></div>
</div>
<p>For two-dimensional arrays, we follow the file format of substitution
matrices provided by NCBI. For example, the BLOSUM62 matrix, which is
the default substitution matrix for NCBI’s protein-protein BLAST
<a class="reference internal" href="chapter_bibliography.html#altschul1990" id="id6"><span>[Altschul1990]</span></a> program <code class="docutils literal notranslate"><span class="pre">blastp</span></code>, is stored as
follows:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>#  Matrix made by matblas from blosum62.iij
#  * column uses minimum score
#  BLOSUM Clustered Scoring Matrix in 1/2 Bit Units
#  Blocks Database = /data/blocks_5.0/blocks.dat
#  Cluster Percentage: &gt;= 62
#  Entropy =   0.6979, Expected =  -0.5209
   A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  Z  X  *
A  4 -1 -2 -2  0 -1 -1  0 -2 -1 -1 -1 -1 -2 -1  1  0 -3 -2  0 -2 -1  0 -4
R -1  5  0 -2 -3  1  0 -2  0 -3 -2  2 -1 -3 -2 -1 -1 -3 -2 -3 -1  0 -1 -4
N -2  0  6  1 -3  0  0  0  1 -3 -3  0 -2 -3 -2  1  0 -4 -2 -3  3  0 -1 -4
D -2 -2  1  6 -3  0  2 -1 -1 -3 -4 -1 -3 -3 -1  0 -1 -4 -3 -3  4  1 -1 -4
C  0 -3 -3 -3  9 -3 -4 -3 -3 -1 -1 -3 -1 -2 -3 -1 -1 -2 -2 -1 -3 -3 -2 -4
Q -1  1  0  0 -3  5  2 -2  0 -3 -2  1  0 -3 -1  0 -1 -2 -1 -2  0  3 -1 -4
E -1  0  0  2 -4  2  5 -2  0 -3 -3  1 -2 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4
G  0 -2  0 -1 -3 -2 -2  6 -2 -4 -4 -2 -3 -3 -2  0 -2 -2 -3 -3 -1 -2 -1 -4
H -2  0  1 -1 -3  0  0 -2  8 -3 -3 -1 -2 -1 -2 -1 -2 -2  2 -3  0  0 -1 -4
...
</pre></div>
</div>
<p>This file is included in the Biopython distribution under
<code class="docutils literal notranslate"><span class="pre">Bio/Align/substitution_matrices/data</span></code>. To parse this file, use</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">substitution_matrices</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;BLOSUM62&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">matrix</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">matrix</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)</span>
<span class="go">ARNDCQEGHILKMFPSTWYVBZX*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">matrix</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">])</span>
<span class="go">-2.0</span>
</pre></div>
</div>
<p>The header lines starting with <code class="docutils literal notranslate"><span class="pre">#</span></code> are stored in the attribute
<code class="docutils literal notranslate"><span class="pre">header</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">matrix</span><span class="o">.</span><span class="n">header</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;Matrix made by matblas from blosum62.iij&#39;</span>
</pre></div>
</div>
<p>We can now use this matrix as the substitution matrix on an aligner
object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="n">matrix</span>
</pre></div>
</div>
<p>To save an Array object, create a string first:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">text</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">text</span><span class="p">)</span>  
<span class="go">#  Matrix made by matblas from blosum62.iij</span>
<span class="go">#  * column uses minimum score</span>
<span class="go">#  BLOSUM Clustered Scoring Matrix in 1/2 Bit Units</span>
<span class="go">#  Blocks Database = /data/blocks_5.0/blocks.dat</span>
<span class="go">#  Cluster Percentage: &gt;= 62</span>
<span class="go">#  Entropy =   0.6979, Expected =  -0.5209</span>
<span class="go">     A    R    N    D    C    Q    E    G    H    I    L    K    M    F    P    S ...</span>
<span class="go">A  4.0 -1.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0 ...</span>
<span class="go">R -1.0  5.0  0.0 -2.0 -3.0  1.0  0.0 -2.0  0.0 -3.0 -2.0  2.0 -1.0 -3.0 -2.0 -1.0 ...</span>
<span class="go">N -2.0  0.0  6.0  1.0 -3.0  0.0  0.0  0.0  1.0 -3.0 -3.0  0.0 -2.0 -3.0 -2.0  1.0 ...</span>
<span class="go">D -2.0 -2.0  1.0  6.0 -3.0  0.0  2.0 -1.0 -1.0 -3.0 -4.0 -1.0 -3.0 -3.0 -1.0  0.0 ...</span>
<span class="go">C  0.0 -3.0 -3.0 -3.0  9.0 -3.0 -4.0 -3.0 -3.0 -1.0 -1.0 -3.0 -1.0 -2.0 -3.0 -1.0 ...</span>
<span class="go">...</span>
</pre></div>
</div>
<p>and write the <code class="docutils literal notranslate"><span class="pre">text</span></code> to a file.</p>
</section>
<section id="loading-predefined-substitution-matrices">
<h3>Loading predefined substitution matrices<a class="headerlink" href="#loading-predefined-substitution-matrices" title="Permalink to this headline"></a></h3>
<p>Biopython contains a large set of substitution matrices defined in the
literature, including BLOSUM (Blocks Substitution Matrix)
<a class="reference internal" href="chapter_bibliography.html#henikoff1992" id="id7"><span>[Henikoff1992]</span></a> and PAM (Point Accepted Mutation)
matrices <a class="reference internal" href="chapter_bibliography.html#dayhoff1978" id="id8"><span>[Dayhoff1978]</span></a>. These matrices are available
as flat files in the <code class="docutils literal notranslate"><span class="pre">Bio/Align/substitution_matrices/data</span></code> directory,
and can be loaded into Python using the <code class="docutils literal notranslate"><span class="pre">load</span></code> function in the
<code class="docutils literal notranslate"><span class="pre">substitution_matrices</span></code> submodule. For example, the BLOSUM62 matrix
can be loaded by running</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">substitution_matrices</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;BLOSUM62&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>This substitution matrix has an alphabet consisting of the 20 amino
acids used in the genetic code, the three ambiguous amino acids B
(asparagine or aspartic acid), Z (glutamine or glutamic acid), and X
(representing any amino acid), and the stop codon represented by an
asterisk:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="o">.</span><span class="n">alphabet</span>
<span class="go">&#39;ARNDCQEGHILKMFPSTWYVBZX*&#39;</span>
</pre></div>
</div>
<p>To get a full list of available substitution matrices, use <code class="docutils literal notranslate"><span class="pre">load</span></code>
without an argument:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>  
<span class="go">[&#39;BENNER22&#39;, &#39;BENNER6&#39;, &#39;BENNER74&#39;, &#39;BLASTN&#39;, &#39;BLASTP&#39;, &#39;BLOSUM45&#39;, &#39;BLOSUM50&#39;, ..., &#39;TRANS&#39;]</span>
</pre></div>
</div>
<p>Note that the substitution matrix provided by Schneider <em>et al.</em>
<a class="reference internal" href="chapter_bibliography.html#schneider2005" id="id9"><span>[Schneider2005]</span></a> uses an alphabet consisting of
three-nucleotide codons:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;SCHNEIDER&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="o">.</span><span class="n">alphabet</span>  
<span class="go">(&#39;AAA&#39;, &#39;AAC&#39;, &#39;AAG&#39;, &#39;AAT&#39;, &#39;ACA&#39;, &#39;ACC&#39;, &#39;ACG&#39;, &#39;ACT&#39;, ..., &#39;TTG&#39;, &#39;TTT&#39;)</span>
</pre></div>
</div>
</section>
</section>
<section id="examples">
<span id="sec-pairwise-examples"></span><h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline"></a></h2>
<p>Suppose you want to do a global pairwise alignment between the same two
hemoglobin sequences from above (<code class="docutils literal notranslate"><span class="pre">HBA_HUMAN</span></code>, <code class="docutils literal notranslate"><span class="pre">HBB_HUMAN</span></code>) stored in
<code class="docutils literal notranslate"><span class="pre">alpha.faa</span></code> and <code class="docutils literal notranslate"><span class="pre">beta.faa</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;alpha.faa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;beta.faa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">seq1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">seq2</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="go">72.0</span>
</pre></div>
</div>
<p>showing an alignment score of 72.0. To see the individual alignments, do</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">seq1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">seq2</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, the total number of optimal alignments is huge (more
than <span class="math notranslate nohighlight">\(4 \times 10^{37}\)</span>), and calling <code class="docutils literal notranslate"><span class="pre">len(alignments)</span></code> will
raise an <code class="docutils literal notranslate"><span class="pre">OverflowError</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">OverflowError</span>: <span class="n">number of optimal alignments is larger than 9223372036854775807</span>
</pre></div>
</div>
<p>Let’s have a look at the first alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<p>The alignment object stores the alignment score, as well as the
alignment itself:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">72.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            0 MV-LS-PAD--KTN--VK-AA-WGKV-----GAHAGEYGAEALE-RMFLSF----P-TTK</span>
<span class="go">                  0 ||-|--|----|----|--|--||||-----|---||--|--|--|--|------|-|--</span>
<span class="go">query             0 MVHL-TP--EEK--SAV-TA-LWGKVNVDEVG---GE--A--L-GR--L--LVVYPWT--</span>

<span class="go">target           41 TY--FPHF----DLSHGS---AQVK-G------HGKKV--A--DA-LTNAVAHV-DDMPN</span>
<span class="go">                 60 ----|--|----|||------|-|--|------|||||--|--|--|--|--|--|---|</span>
<span class="go">query            39 --QRF--FESFGDLS---TPDA-V-MGNPKVKAHGKKVLGAFSD-GL--A--H-LD---N</span>

<span class="go">target           79 ALS----A-LSD-LHAH--KLR-VDPV-NFK-LLSHC---LLVT--LAAHLPA----EFT</span>
<span class="go">                120 -|-----|-||--||----||--|||--||--||------|-|---||-|-------|||</span>
<span class="go">query            81 -L-KGTFATLS-ELH--CDKL-HVDP-ENF-RLL---GNVL-V-CVLA-H---HFGKEFT</span>

<span class="go">target          119 PA-VH-ASLDKFLAS---VSTV------LTS--KYR- 142</span>
<span class="go">                180 |--|--|------|----|--|------|----||-- 217</span>
<span class="go">query           124 P-PV-QA------A-YQKV--VAGVANAL--AHKY-H 147</span>
</pre></div>
</div>
<p>Better alignments are usually obtained by penalizing gaps: higher costs
for opening a gap and lower costs for extending an existing gap. For
amino acid sequences match scores are usually encoded in matrices like
<code class="docutils literal notranslate"><span class="pre">PAM</span></code> or <code class="docutils literal notranslate"><span class="pre">BLOSUM</span></code>. Thus, a more meaningful alignment for our example
can be obtained by using the BLOSUM62 matrix, together with a gap open
penalty of 10 and a gap extension penalty of 0.5:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">substitution_matrices</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;alpha.faa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;beta.faa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">open_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">extend_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;BLOSUM62&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">seq1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">seq2</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="go">292.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">seq1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">seq2</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">292.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">target            0 MV-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DLS-----HGS</span>
<span class="go">                  0 ||-|.|..|..|.|.||||--...|.|.|||.|.....|.|...|..|-|||-----.|.</span>
<span class="go">query             0 MVHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGN</span>

<span class="go">target           53 AQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAH</span>
<span class="go">                 60 ..||.|||||..|.....||.|........||.||..||.|||.||.||...|...||.|</span>
<span class="go">query            58 PKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHH</span>

<span class="go">target          113 LPAEFTPAVHASLDKFLASVSTVLTSKYR 142</span>
<span class="go">                120 ...||||.|.|...|..|.|...|..||. 149</span>
<span class="go">query           118 FGKEFTPPVQAAYQKVVAGVANALAHKYH 147</span>
</pre></div>
</div>
<p>This alignment has the same score that we obtained earlier with EMBOSS
needle using the same sequences and the same parameters.</p>
<p>To perform a local alignment, set <code class="docutils literal notranslate"><span class="pre">aligner.mode</span></code> to <code class="docutils literal notranslate"><span class="pre">'local'</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s2">&quot;local&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">open_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">extend_gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="s2">&quot;LSPADKTNVKAA&quot;</span><span class="p">,</span> <span class="s2">&quot;PEEKSAV&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">target            2 PADKTNV 9</span>
<span class="go">                  0 |..|..| 7</span>
<span class="go">query             0 PEEKSAV 7</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">16.0</span>
</pre></div>
</div>
</section>
<section id="generalized-pairwise-alignments">
<span id="sec-generalized-pairwise"></span><h2>Generalized pairwise alignments<a class="headerlink" href="#generalized-pairwise-alignments" title="Permalink to this headline"></a></h2>
<p>In most cases, <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> is used to perform alignments of
sequences (strings or <code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects) consisting of single-letter
nucleotides or amino acids. More generally, <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> can also
be applied to lists or tuples of arbitrary objects. This section will
describe some examples of such generalized pairwise alignments.</p>
<section id="generalized-pairwise-alignments-using-a-substitution-matrix-and-alphabet">
<h3>Generalized pairwise alignments using a substitution matrix and alphabet<a class="headerlink" href="#generalized-pairwise-alignments-using-a-substitution-matrix-and-alphabet" title="Permalink to this headline"></a></h3>
<p>Schneider <em>et al.</em> <a class="reference internal" href="chapter_bibliography.html#schneider2005" id="id10"><span>[Schneider2005]</span></a> created a
substitution matrix for aligning three-nucleotide codons (see
<a class="reference external" href="#codonmatrix">below</a> in section <a class="reference internal" href="#sec-substitution-matrices"><span class="std std-ref">Substitution matrices</span></a>
for more information). This substitution matrix is associated with an
alphabet consisting of all three-letter codons:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">substitution_matrices</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">substitution_matrices</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;SCHNEIDER&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="o">.</span><span class="n">alphabet</span>  
<span class="go">(&#39;AAA&#39;, &#39;AAC&#39;, &#39;AAG&#39;, &#39;AAT&#39;, &#39;ACA&#39;, &#39;ACC&#39;, &#39;ACG&#39;, &#39;ACT&#39;, ..., &#39;TTG&#39;, &#39;TTT&#39;)</span>
</pre></div>
</div>
<p>We can use this matrix to align codon sequences to each other:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="n">m</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mf">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;AAT&quot;</span><span class="p">,</span> <span class="s2">&quot;CTG&quot;</span><span class="p">,</span> <span class="s2">&quot;TTT&quot;</span><span class="p">,</span> <span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;AAT&quot;</span><span class="p">,</span> <span class="s2">&quot;TTA&quot;</span><span class="p">,</span> <span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">AAT CTG TTT TTT</span>
<span class="go">||| ... ||| ---</span>
<span class="go">AAT TTA TTT ---</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">AAT CTG TTT TTT</span>
<span class="go">||| ... --- |||</span>
<span class="go">AAT TTA --- TTT</span>
</pre></div>
</div>
<p>Note that aligning <code class="docutils literal notranslate"><span class="pre">TTT</span></code> to <code class="docutils literal notranslate"><span class="pre">TTA</span></code>, as in this example:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="go">AAT CTG TTT TTT</span>
<span class="go">||| --- ... |||</span>
<span class="go">AAT --- TTA TTT</span>
</pre></div>
</div>
<p>would get a much lower score:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;CTG&quot;</span><span class="p">,</span> <span class="s2">&quot;TTA&quot;</span><span class="p">])</span>
<span class="go">7.6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;TTT&quot;</span><span class="p">,</span> <span class="s2">&quot;TTA&quot;</span><span class="p">])</span>
<span class="go">-0.3</span>
</pre></div>
</div>
<p>presumably because <code class="docutils literal notranslate"><span class="pre">CTG</span></code> and <code class="docutils literal notranslate"><span class="pre">TTA</span></code> both code for leucine, while
<code class="docutils literal notranslate"><span class="pre">TTT</span></code> codes for phenylalanine. The three-letter codon substitution
matrix also reveals a preference among codons representing the same
amino acid. For example, <code class="docutils literal notranslate"><span class="pre">TTA</span></code> has a preference for <code class="docutils literal notranslate"><span class="pre">CTG</span></code> preferred
compared to <code class="docutils literal notranslate"><span class="pre">CTC</span></code>, though all three code for leucine:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;AAT&quot;</span><span class="p">,</span> <span class="s2">&quot;CTG&quot;</span><span class="p">,</span> <span class="s2">&quot;CTC&quot;</span><span class="p">,</span> <span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;AAT&quot;</span><span class="p">,</span> <span class="s2">&quot;TTA&quot;</span><span class="p">,</span> <span class="s2">&quot;TTT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">AAT CTG CTC TTT</span>
<span class="go">||| ... --- |||</span>
<span class="go">AAT TTA --- TTT</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">[</span><span class="s2">&quot;CTC&quot;</span><span class="p">,</span> <span class="s2">&quot;TTA&quot;</span><span class="p">])</span>
<span class="go">6.5</span>
</pre></div>
</div>
</section>
<section id="generalized-pairwise-alignments-using-match-mismatch-scores-and-an-alphabet">
<h3>Generalized pairwise alignments using match/mismatch scores and an alphabet<a class="headerlink" href="#generalized-pairwise-alignments-using-match-mismatch-scores-and-an-alphabet" title="Permalink to this headline"></a></h3>
<p>Using the three-letter amino acid symbols, the sequences above translate
to</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Asn&quot;</span><span class="p">,</span> <span class="s2">&quot;Leu&quot;</span><span class="p">,</span> <span class="s2">&quot;Leu&quot;</span><span class="p">,</span> <span class="s2">&quot;Phe&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Asn&quot;</span><span class="p">,</span> <span class="s2">&quot;Leu&quot;</span><span class="p">,</span> <span class="s2">&quot;Phe&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can align these sequences directly to each other by using a
three-letter amino acid alphabet:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">alphabet</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Ala&#39;</span><span class="p">,</span> <span class="s1">&#39;Arg&#39;</span><span class="p">,</span> <span class="s1">&#39;Asn&#39;</span><span class="p">,</span> <span class="s1">&#39;Asp&#39;</span><span class="p">,</span> <span class="s1">&#39;Cys&#39;</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="s1">&#39;Gln&#39;</span><span class="p">,</span> <span class="s1">&#39;Glu&#39;</span><span class="p">,</span> <span class="s1">&#39;Gly&#39;</span><span class="p">,</span> <span class="s1">&#39;His&#39;</span><span class="p">,</span> <span class="s1">&#39;Ile&#39;</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="s1">&#39;Leu&#39;</span><span class="p">,</span> <span class="s1">&#39;Lys&#39;</span><span class="p">,</span> <span class="s1">&#39;Met&#39;</span><span class="p">,</span> <span class="s1">&#39;Phe&#39;</span><span class="p">,</span> <span class="s1">&#39;Pro&#39;</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="s1">&#39;Ser&#39;</span><span class="p">,</span> <span class="s1">&#39;Thr&#39;</span><span class="p">,</span> <span class="s1">&#39;Trp&#39;</span><span class="p">,</span> <span class="s1">&#39;Tyr&#39;</span><span class="p">,</span> <span class="s1">&#39;Val&#39;</span><span class="p">]</span>  <span class="c1"># fmt: skip</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>We use +6/-1 match and mismatch scores as an approximation of the
BLOSUM62 matrix, and align these sequences to each other:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">match</span> <span class="o">=</span> <span class="o">+</span><span class="mi">6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mismatch</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">))</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">Asn Leu Leu Phe</span>
<span class="go">||| ||| --- |||</span>
<span class="go">Asn Leu --- Phe</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">Asn Leu Leu Phe</span>
<span class="go">||| --- ||| |||</span>
<span class="go">Asn --- Leu Phe</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">18.0</span>
</pre></div>
</div>
</section>
<section id="generalized-pairwise-alignments-using-match-mismatch-scores-and-integer-sequences">
<h3>Generalized pairwise alignments using match/mismatch scores and integer sequences<a class="headerlink" href="#generalized-pairwise-alignments-using-match-mismatch-scores-and-integer-sequences" title="Permalink to this headline"></a></h3>
<p>Internally, the first step when performing an alignment is to replace
the two sequences by integer arrays consisting of the indices of each
letter in each sequence in the alphabet associated with the aligner.
This step can be bypassed by passing integer arrays directly:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">13</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">13</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">match</span> <span class="o">=</span> <span class="o">+</span><span class="mi">6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">mismatch</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">))</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">2 10 10 13</span>
<span class="go">| || -- ||</span>
<span class="go">2 10 -- 13</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">2 10 10 13</span>
<span class="go">| -- || ||</span>
<span class="go">2 -- 10 13</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">18.0</span>
</pre></div>
</div>
<p>Note that the indices should consist of 32-bit integers, as specified in
this example by <code class="docutils literal notranslate"><span class="pre">numpy.int32</span></code>.</p>
<p>Unknown letters can again be included by defining a wildcard character,
and using the corresponding Unicode code point number as the index:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">wildcard</span> <span class="o">=</span> <span class="s2">&quot;?&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">ord</span><span class="p">(</span><span class="n">aligner</span><span class="o">.</span><span class="n">wildcard</span><span class="p">)</span>
<span class="go">63</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">63</span><span class="p">,</span> <span class="mi">13</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">))</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">2 10 10 13</span>
<span class="go">| .. -- ||</span>
<span class="go">2 63 -- 13</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">2 10 10 13</span>
<span class="go">| -- .. ||</span>
<span class="go">2 -- 63 13</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">9.0</span>
</pre></div>
</div>
</section>
<section id="generalized-pairwise-alignments-using-a-substitution-matrix-and-integer-sequences">
<h3>Generalized pairwise alignments using a substitution matrix and integer sequences<a class="headerlink" href="#generalized-pairwise-alignments-using-a-substitution-matrix-and-integer-sequences" title="Permalink to this headline"></a></h3>
<p>Integer sequences can also be aligned using a substitution matrix, in
this case a numpy square array without an alphabet associated with it.
In this case, all index values must be non-negative, and smaller than
the size of the substitution matrix:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">PairwiseAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">:]</span> <span class="o">=</span> <span class="n">m</span><span class="p">[</span><span class="mi">1</span><span class="p">:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="mi">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="go">[[ 1. -2. -2. -2. -2.]</span>
<span class="go"> [-2.  1.  0.  0.  0.]</span>
<span class="go"> [-2.  0.  3.  0.  0.]</span>
<span class="go"> [-2.  0.  0.  1.  0.]</span>
<span class="go"> [-2.  0.  0.  0.  1.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="n">m</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span><span class="o">.</span><span class="n">gap_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">s1</span><span class="p">,</span> <span class="n">s2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">))</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">0 - 2 3 4</span>
<span class="go">| - | . -</span>
<span class="go">0 3 2 1 -</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">0 - 2 3 4</span>
<span class="go">| - | - .</span>
<span class="go">0 3 2 - 1</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="o">.</span><span class="n">score</span><span class="p">)</span>
<span class="go">2.0</span>
</pre></div>
</div>
</section>
</section>
<section id="codon-alignments">
<span id="sec-codon-alignments"></span><h2>Codon alignments<a class="headerlink" href="#codon-alignments" title="Permalink to this headline"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">CodonAligner</span></code> class in the <code class="docutils literal notranslate"><span class="pre">Bio.Align</span></code> module implements a
specialized aligner for aligning a nucleotide sequence to the amino acid
sequence it encodes. Such alignments are non-trivial if frameshifts
occur during translation.</p>
<section id="aligning-a-nucleotide-sequence-to-an-amino-acid-sequence">
<h3>Aligning a nucleotide sequence to an amino acid sequence<a class="headerlink" href="#aligning-a-nucleotide-sequence-to-an-amino-acid-sequence" title="Permalink to this headline"></a></h3>
<p>To align a nucleotide sequence to an amino acid sequence, first create a
<code class="docutils literal notranslate"><span class="pre">CodonAligner</span></code> object:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">CodonAligner</span><span class="p">()</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">CodonAligner</span></code> object <code class="docutils literal notranslate"><span class="pre">aligner</span></code> stores the alignment parameters
to be used for the alignments:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">aligner</span><span class="p">)</span>
<span class="go">Codon aligner with parameters</span>
<span class="go">  wildcard: &#39;X&#39;</span>
<span class="go">  match_score: 1.0</span>
<span class="go">  mismatch_score: 0.0</span>
<span class="go">  frameshift_minus_two_score: -3.0</span>
<span class="go">  frameshift_minus_one_score: -3.0</span>
<span class="go">  frameshift_plus_one_score: -3.0</span>
<span class="go">  frameshift_plus_two_score: -3.0</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">wildcard</span></code>, <code class="docutils literal notranslate"><span class="pre">match_score</span></code>, and <code class="docutils literal notranslate"><span class="pre">mismatch_score</span></code> parameters are
defined in the same was as for the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> class described
above (see Section <a class="reference internal" href="#sec-pairwise-aligner"><span class="std std-ref">The pairwise aligner object</span></a>). The values
specified by the <code class="docutils literal notranslate"><span class="pre">frameshift_minus_two_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">frameshift_minus_one_score</span></code>, <code class="docutils literal notranslate"><span class="pre">frameshift_plus_one_score</span></code>, and
<code class="docutils literal notranslate"><span class="pre">frameshift_plus_two_score</span></code> parameters are added to the alignment
score whenever a -2, -1, +1, or +2 frame shift, respectively, occurs in
the alignment. By default, the frame shift scores are set to -3.0.
Similar to the <code class="docutils literal notranslate"><span class="pre">PairwiseAligner</span></code> class
(Table <a class="reference internal" href="#table-align-meta-attributes"><span class="std std-ref">Meta-attributes of the pairwise aligner objects.</span></a>), the <code class="docutils literal notranslate"><span class="pre">CodonAligner</span></code>
class defines additional attributes that refer to a number of these
values collectively, as shown in
Table <a class="reference internal" href="#table-codonalign-meta-attributes"><span class="std std-ref">Meta-attributes of CodonAligner objects.</span></a>.</p>
<table class="docutils align-default" id="table-codonalign-meta-attributes">
<caption><span class="caption-number">Table 2 </span><span class="caption-text">Meta-attributes of CodonAligner objects.</span><a class="headerlink" href="#table-codonalign-meta-attributes" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 46%" />
<col style="width: 54%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Meta-attribute</p></th>
<th class="head"><p>Attributes it maps to</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">frameshift_minus_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">frameshift_minus_two_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">frameshift_minus_one_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">frameshift_plus_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">frameshift_plus_two_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">frameshift_plus_one_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">frameshift_two_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">frameshift_minus_two_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">frameshift_plus_two_score</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">frameshift_one_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">frameshift_minus_one_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">frameshift_plus_one_score</span></code></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">frameshift_score</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">frameshift_minus_two_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">frameshift_minus_one_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">frameshift_plus_one_score</span></code>,
<code class="docutils literal notranslate"><span class="pre">frameshift_plus_two_score</span></code></p></td>
</tr>
</tbody>
</table>
<p>Now let’s consider two nucleotide sequences and the amino acid sequences
they encode:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nuc1</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TCAGGGACTGCGAGAACCAAGCTACTGCTGCTGCTGGCTGCGCTCTGCGCCGCAGGTGGGGCGCTGGAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">nuc1</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;rna1&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nuc2</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TCAGGGACTTCGAGAACCAAGCGCTCCTGCTGCTGGCTGCGCTCGGCGCCGCAGGTGGAGCACTGGAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">nuc2</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;rna2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aa1</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;SGTARTKLLLLLAALCAAGGALE&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aa2</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;SGTSRTKRLLLLAALGAAGGALE&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pro1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">aa1</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;pro1&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pro2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">aa2</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;pro2&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>While the two protein sequences both consist of 23 amino acids, the
first nucleotide sequence consists of <span class="math notranslate nohighlight">\(3 \times 23 = 69\)</span>
nucleotides while the second nucleotide sequence tonsists of only 68
nucleotides:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">pro1</span><span class="p">)</span>
<span class="go">23</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">pro2</span><span class="p">)</span>
<span class="go">23</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">rna1</span><span class="p">)</span>
<span class="go">69</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">rna2</span><span class="p">)</span>
<span class="go">68</span>
</pre></div>
</div>
<p>This is due to a -1 frame shift event during translation of the second
nucleotide sequence. Use <code class="docutils literal notranslate"><span class="pre">CodonAligner.align</span></code> to align <code class="docutils literal notranslate"><span class="pre">rna1</span></code> to
<code class="docutils literal notranslate"><span class="pre">pro1</span></code>, and <code class="docutils literal notranslate"><span class="pre">rna2</span></code> to <code class="docutils literal notranslate"><span class="pre">pro2</span></code>, returning an iterator of
<code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments1</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">pro1</span><span class="p">,</span> <span class="n">rna1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments1</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment1</span><span class="p">)</span>
<span class="go">pro1              0 S  G  T  A  R  T  K  L  L  L  L  L  A  A  L  C  A  A  G  G</span>
<span class="go">rna1              0 TCAGGGACTGCGAGAACCAAGCTACTGCTGCTGCTGGCTGCGCTCTGCGCCGCAGGTGGG</span>

<span class="go">pro1             20 A  L  E   23</span>
<span class="go">rna1             60 GCGCTGGAG 69</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[ 0, 23],</span>
<span class="go">       [ 0, 69]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;SGTARTKLLLLLAALCAAGGALE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;TCAGGGACTGCGAGAACCAAGCTACTGCTGCTGCTGGCTGCGCTCTGCGCCGCAGGTGGGGCGCTGGAG&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments2</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">pro2</span><span class="p">,</span> <span class="n">rna2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments2</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment2</span><span class="p">)</span>
<span class="go">pro2              0 S  G  T  S  R  T  K  R   8</span>
<span class="go">rna2              0 TCAGGGACTTCGAGAACCAAGCGC 24</span>

<span class="go">pro2              8 L  L  L  L  A  A  L  G  A  A  G  G  A  L  E   23</span>
<span class="go">rna2             23 CTCCTGCTGCTGGCTGCGCTCGGCGCCGCAGGTGGAGCACTGGAG 68</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;SGTSRTKRLLLLAALGAAGGALE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="go">&#39;TCAGGGACTTCGAGAACCAAGCGCCTCCTGCTGCTGGCTGCGCTCGGCGCCGCAGGTGGAGCACTGGAG&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span><span class="o">.</span><span class="n">coordinates</span>
<span class="go">array([[ 0,  8,  8, 23],</span>
<span class="go">       [ 0, 24, 23, 68]])</span>
</pre></div>
</div>
<p>While <code class="docutils literal notranslate"><span class="pre">alignment1</span></code> is a continuous alignment of the 69 nucleotides to
the 23 amino acids, in <code class="docutils literal notranslate"><span class="pre">alignment2</span></code> we find a -1 frame shift after 24
nucleotides. As <code class="docutils literal notranslate"><span class="pre">alignment2[1]</span></code> contains the nucleotide sequence after
applying the -1 frame shift, it is one nucleotide longer than <code class="docutils literal notranslate"><span class="pre">nuc2</span></code>
and can be translated directly, resulting in the amino acid sequence
<code class="docutils literal notranslate"><span class="pre">aa2</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">translate</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">nuc2</span><span class="p">)</span>
<span class="go">68</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignment2</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">69</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">alignment2</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="go">&#39;SGTSRTKRLLLLAALGAAGGALE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">==</span> <span class="n">aa2</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The alignment score is stored as an attribute on the <code class="docutils literal notranslate"><span class="pre">alignments1</span></code> and
<code class="docutils literal notranslate"><span class="pre">alignments2</span></code> iterators, and on the individual alignments
<code class="docutils literal notranslate"><span class="pre">alignment1</span></code> and <code class="docutils literal notranslate"><span class="pre">alignment2</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments1</span><span class="o">.</span><span class="n">score</span>
<span class="go">23.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment1</span><span class="o">.</span><span class="n">score</span>
<span class="go">23.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignments2</span><span class="o">.</span><span class="n">score</span>
<span class="go">20.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment2</span><span class="o">.</span><span class="n">score</span>
<span class="go">20.0</span>
</pre></div>
</div>
<p>where the score of the <code class="docutils literal notranslate"><span class="pre">rna1</span></code>-<code class="docutils literal notranslate"><span class="pre">pro1</span></code> alignment is equal to the
number of aligned amino acids, and the score of the <code class="docutils literal notranslate"><span class="pre">rna2</span></code>-<code class="docutils literal notranslate"><span class="pre">pro2</span></code>
alignment is 3 less due to the penalty for the frame shift. To calculate
the alignment score without calculating the alignment itself, the
<code class="docutils literal notranslate"><span class="pre">score</span></code> method can be used:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">pro1</span><span class="p">,</span> <span class="n">rna1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="go">23.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">pro2</span><span class="p">,</span> <span class="n">rna2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="go">20.0</span>
</pre></div>
</div>
</section>
<section id="generating-a-multiple-sequence-alignment-of-codon-sequences">
<span id="sec-msa-codons"></span><h3>Generating a multiple sequence alignment of codon sequences<a class="headerlink" href="#generating-a-multiple-sequence-alignment-of-codon-sequences" title="Permalink to this headline"></a></h3>
<p>Suppose we have a third related amino acid sequence and its associated
nucleotide sequence:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aa3</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MGTALLLLLAALCAAGGALE&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pro3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">aa3</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;pro3&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nuc3</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGGAACCGCGCTGCTTTTGCTACTGGCCGCGCTCTGCGCCGCAGGTGGGGCCCTGGAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rna3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">nuc3</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;rna3&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nuc3</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span> <span class="o">==</span> <span class="n">aa3</span>
<span class="go">True</span>
</pre></div>
</div>
<p>As above, we use the <code class="docutils literal notranslate"><span class="pre">CodonAligner</span></code> to align the nucleotide sequence
to the amino acid sequence:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">alignments3</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">pro3</span><span class="p">,</span> <span class="n">rna3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">alignments3</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment3</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment3</span><span class="p">)</span>
<span class="go">pro3              0 M  G  T  A  L  L  L  L  L  A  A  L  C  A  A  G  G  A  L  E</span>
<span class="go">rna3              0 ATGGGAACCGCGCTGCTTTTGCTACTGGCCGCGCTCTGCGCCGCAGGTGGGGCCCTGGAG</span>

<span class="go">pro3             20</span>
<span class="go">rna3             60</span>
</pre></div>
</div>
<p>The three amino acid sequences can be aligned to each other, for example
using ClustalW. Here, we create the alignment by hand:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">Alignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sequences</span> <span class="o">=</span> <span class="p">[</span><span class="n">pro1</span><span class="p">,</span> <span class="n">pro2</span><span class="p">,</span> <span class="n">pro3</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_alignment</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">sequences</span><span class="p">,</span> <span class="n">coordinates</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">23</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">23</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">20</span><span class="p">]])</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">protein_alignment</span><span class="p">)</span>
<span class="go">pro1              0 SGTARTKLLLLLAALCAAGGALE 23</span>
<span class="go">pro2              0 SGTSRTKRLLLLAALGAAGGALE 23</span>
<span class="go">pro3              0 MGTA---LLLLLAALCAAGGALE 20</span>
</pre></div>
</div>
<p>Now we can use the <code class="docutils literal notranslate"><span class="pre">mapall</span></code> method on the protein alignment, with the
nucleotide-to-protein pairwise alignments as the argument, to obtain the
corresponding codon alignment:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">codon_alignment</span> <span class="o">=</span> <span class="n">protein_alignment</span><span class="o">.</span><span class="n">mapall</span><span class="p">([</span><span class="n">alignment1</span><span class="p">,</span> <span class="n">alignment2</span><span class="p">,</span> <span class="n">alignment3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">codon_alignment</span><span class="p">)</span>
<span class="go">rna1              0 TCAGGGACTGCGAGAACCAAGCTA 24</span>
<span class="go">rna2              0 TCAGGGACTTCGAGAACCAAGCGC 24</span>
<span class="go">rna3              0 ATGGGAACCGCG---------CTG 15</span>

<span class="go">rna1             24 CTGCTGCTGCTGGCTGCGCTCTGCGCCGCAGGTGGGGCGCTGGAG 69</span>
<span class="go">rna2             23 CTCCTGCTGCTGGCTGCGCTCGGCGCCGCAGGTGGAGCACTGGAG 68</span>
<span class="go">rna3             15 CTTTTGCTACTGGCCGCGCTCTGCGCCGCAGGTGGGGCCCTGGAG 60</span>
</pre></div>
</div>
</section>
<section id="analyzing-a-codon-alignment">
<h3>Analyzing a codon alignment<a class="headerlink" href="#analyzing-a-codon-alignment" title="Permalink to this headline"></a></h3>
<section id="calculating-the-number-of-nonsynonymous-and-synonymous-substitutions-per-site">
<h4>Calculating the number of nonsynonymous and synonymous substitutions per site<a class="headerlink" href="#calculating-the-number-of-nonsynonymous-and-synonymous-substitutions-per-site" title="Permalink to this headline"></a></h4>
<p>The most important application of a codon alignment is to estimate the
number of nonsynonymous substitutions per site (dN) and synonymous
substitutions per site (dS). These can be calculated by the
<code class="docutils literal notranslate"><span class="pre">calculate_dn_ds</span></code> function in <code class="docutils literal notranslate"><span class="pre">Bio.Align.analysis</span></code>. This function
takes a pairwise codon alignment and input, as well as the optional
arguments <code class="docutils literal notranslate"><span class="pre">method</span></code> specifying the calculation method, <code class="docutils literal notranslate"><span class="pre">codon_table</span></code>
(defaulting to the Standard Code), the ratio <code class="docutils literal notranslate"><span class="pre">k</span></code> of the transition and
transversion rates, and <code class="docutils literal notranslate"><span class="pre">cfreq</span></code> to specify the equilibrium codon
frequency. Biopython currently supports three counting based methods
(<code class="docutils literal notranslate"><span class="pre">NG86</span></code>, <code class="docutils literal notranslate"><span class="pre">LWL85</span></code>, <code class="docutils literal notranslate"><span class="pre">YN00</span></code>) as well as the maximum likelihood method
(<code class="docutils literal notranslate"><span class="pre">ML</span></code>) to estimate dN and dS:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">NG86</span></code>: Nei and Gojobori (1986) <a class="reference internal" href="chapter_bibliography.html#nei1986" id="id11"><span>[Nei1986]</span></a>
(default). With this method, you can also specify the ratio of the
transition and transversion rates via the argument <code class="docutils literal notranslate"><span class="pre">k</span></code>, defaulting
to <code class="docutils literal notranslate"><span class="pre">1.0</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">LWL85</span></code>: Li <em>et al.</em> (1985) <a class="reference internal" href="chapter_bibliography.html#li1985" id="id12"><span>[Li1985]</span></a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">YN00</span></code>: Yang and Nielsen (2000) <a class="reference internal" href="chapter_bibliography.html#yang2000" id="id13"><span>[Yang2000]</span></a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ML</span></code>: Goldman and Yang (1994) <a class="reference internal" href="chapter_bibliography.html#goldman1994" id="id14"><span>[Goldman1994]</span></a>. With
this method, you can also specify the equilibrium codon frequency via
the <code class="docutils literal notranslate"><span class="pre">cfreq</span></code> argument, with the following options:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">F1x4</span></code>: count the nucleotide frequency in the provided codon
sequences, and use it to calculate the background codon frequency;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">F3x4</span></code>: (default) count the nucleotide frequency separately for
the first, second, and third position in the provided codons, and
use it to calculate the background codon frequency;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">F61</span></code>: count the frequency of codons from the provided codon
sequences, with a pseudocount of 0.1.</p></li>
</ul>
</li>
</ul>
<p>The <code class="docutils literal notranslate"><span class="pre">calculate_dN_dS</span></code> method can be applied to a pairwise codon
alignment. In general, the different calculation methods will result in
slightly different estimates for dN and dS:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">analysis</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pairwise_codon_alignment</span> <span class="o">=</span> <span class="n">codon_alignment</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pairwise_codon_alignment</span><span class="p">)</span>
<span class="go">rna1              0 TCAGGGACTGCGAGAACCAAGCTA 24</span>
<span class="go">                  0 |||||||||.||||||||||||..</span>
<span class="go">rna2              0 TCAGGGACTTCGAGAACCAAGCGC 24</span>

<span class="go">rna1             24 CTGCTGCTGCTGGCTGCGCTCTGCGCCGCAGGTGGGGCGCTGGAG 69</span>
<span class="go">                 24 ||.||||||||||||||||||.|||||||||||||.||.|||||| 69</span>
<span class="go">rna2             23 CTCCTGCTGCTGGCTGCGCTCGGCGCCGCAGGTGGAGCACTGGAG 68</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">analysis</span><span class="o">.</span><span class="n">calculate_dn_ds</span><span class="p">(</span><span class="n">pairwise_codon_alignment</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;NG86&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span><span class="p">)</span>  
<span class="go">0.067715... 0.201197...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">analysis</span><span class="o">.</span><span class="n">calculate_dn_ds</span><span class="p">(</span><span class="n">pairwise_codon_alignment</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;LWL85&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span><span class="p">)</span>  
<span class="go">0.068728... 0.207551...</span>
</pre></div>
</div>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">analysis</span><span class="o">.</span><span class="n">calculate_dn_ds</span><span class="p">(</span><span class="n">pairwise_codon_alignment</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;YN00&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span><span class="p">)</span>  
<span class="go">0.081468... 0.127706...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">analysis</span><span class="o">.</span><span class="n">calculate_dn_ds</span><span class="p">(</span><span class="n">pairwise_codon_alignment</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;ML&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span><span class="p">)</span>  
<span class="go">0.069475... 0.205754...</span>
</pre></div>
</div>
<p>For a multiple alignment of codon sequences, you can calculate a matrix
of dN and dS values:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dN</span><span class="p">,</span> <span class="n">dS</span> <span class="o">=</span> <span class="n">analysis</span><span class="o">.</span><span class="n">calculate_dn_ds_matrix</span><span class="p">(</span><span class="n">codon_alignment</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;NG86&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dN</span><span class="p">)</span>
<span class="go">rna1    0.000000</span>
<span class="go">rna2    0.067715    0.000000</span>
<span class="go">rna3    0.060204    0.145469    0.000000</span>
<span class="go">    rna1    rna2    rna3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dS</span><span class="p">)</span>
<span class="go">rna1    0.000000</span>
<span class="go">rna2    0.201198    0.000000</span>
<span class="go">rna3    0.664268    0.798957    0.000000</span>
<span class="go">    rna1    rna2    rna3</span>
</pre></div>
</div>
<p>The objects <code class="docutils literal notranslate"><span class="pre">dN</span></code> and <code class="docutils literal notranslate"><span class="pre">dS</span></code> returned by <code class="docutils literal notranslate"><span class="pre">calculate_dn_ds_matrix</span></code> are
instances of the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> class in
<code class="docutils literal notranslate"><span class="pre">Bio.Phylo.TreeConstruction</span></code>. This function only takes <code class="docutils literal notranslate"><span class="pre">codon_table</span></code>
as an optional argument.</p>
<p>From these two sequences, you can create a dN tree and a dS tree using
<code class="docutils literal notranslate"><span class="pre">Bio.Phylo.TreeConstruction</span></code>:</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Phylo.TreeConstruction</span> <span class="kn">import</span> <span class="n">DistanceTreeConstructor</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dn_constructor</span> <span class="o">=</span> <span class="n">DistanceTreeConstructor</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds_constructor</span> <span class="o">=</span> <span class="n">DistanceTreeConstructor</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dn_tree</span> <span class="o">=</span> <span class="n">dn_constructor</span><span class="o">.</span><span class="n">upgma</span><span class="p">(</span><span class="n">dN</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds_tree</span> <span class="o">=</span> <span class="n">ds_constructor</span><span class="o">.</span><span class="n">upgma</span><span class="p">(</span><span class="n">dS</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">dn_tree</span><span class="p">))</span>
<span class="go">&lt;class &#39;Bio.Phylo.BaseTree.Tree&#39;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">dn_tree</span><span class="p">)</span>  
<span class="go">Tree(rooted=True)</span>
<span class="go">    Clade(branch_length=0, name=&#39;Inner2&#39;)</span>
<span class="go">        Clade(branch_length=0.053296..., name=&#39;rna2&#39;)</span>
<span class="go">        Clade(branch_length=0.023194..., name=&#39;Inner1&#39;)</span>
<span class="go">            Clade(branch_length=0.0301021..., name=&#39;rna3&#39;)</span>
<span class="go">            Clade(branch_length=0.0301021..., name=&#39;rna1&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ds_tree</span><span class="p">)</span>  
<span class="go">Tree(rooted=True)</span>
<span class="go">    Clade(branch_length=0, name=&#39;Inner2&#39;)</span>
<span class="go">        Clade(branch_length=0.365806..., name=&#39;rna3&#39;)</span>
<span class="go">        Clade(branch_length=0.265207..., name=&#39;Inner1&#39;)</span>
<span class="go">            Clade(branch_length=0.100598..., name=&#39;rna2&#39;)</span>
<span class="go">            Clade(branch_length=0.100598..., name=&#39;rna1&#39;)</span>
</pre></div>
</div>
</section>
<section id="performing-the-mcdonald-kreitman-test">
<h4>Performing the McDonald-Kreitman test<a class="headerlink" href="#performing-the-mcdonald-kreitman-test" title="Permalink to this headline"></a></h4>
<p>The McDonald-Kreitman test assesses the amount of adaptive evolution by
comparing the within species synonymous substitutions and nonsynonymous
substitutions to the between species synonymous substitutions and
nonsynonymous substitutions to see if they are from the same
evolutionary process. The test requires gene sequences sampled from
different individuals of the same species. In the following example, we
will use Adh gene from fruit fly. The data includes 11 individuals from
<em>Drosophila melanogaster</em>, 4 individuals from <em>Drosophila simulans</em>, and
12 individuals from <em>Drosophila yakuba</em>. The protein alignment data and
the nucleotide sequences are available in the <code class="docutils literal notranslate"><span class="pre">Tests/codonalign</span></code>
directory as the files <code class="docutils literal notranslate"><span class="pre">adh.aln</span></code> and <code class="docutils literal notranslate"><span class="pre">drosophila.fasta</span></code>,
respectively, in the Biopython distribution. The function <code class="docutils literal notranslate"><span class="pre">mktest</span></code> in
<code class="docutils literal notranslate"><span class="pre">Bio.Align.analysis</span></code> implements the Mcdonald-Kreitman test.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Align</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">CodonAligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align.analysis</span> <span class="kn">import</span> <span class="n">mktest</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">CodonAligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nucleotide_records</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;drosophila.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">nucleotide_record</span> <span class="ow">in</span> <span class="n">nucleotide_records</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">nucleotide_record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>  
<span class="gp">...</span>
<span class="go">gi|9097|emb|X57361.1| Drosophila simulans (individual c) ...</span>
<span class="go">gi|9099|emb|X57362.1| Drosophila simulans (individual d) ...</span>
<span class="go">gi|9101|emb|X57363.1| Drosophila simulans (individual e) ...</span>
<span class="go">gi|9103|emb|X57364.1| Drosophila simulans (individual f) ...</span>
<span class="go">gi|9217|emb|X57365.1| Drosophila yakuba (individual a) ...</span>
<span class="go">gi|9219|emb|X57366.1| Drosophila yakuba (individual b) ...</span>
<span class="go">gi|9221|emb|X57367.1| Drosophila yakuba (individual c) ...</span>
<span class="go">gi|9223|emb|X57368.1| Drosophila yakuba (individual d) ...</span>
<span class="go">gi|9225|emb|X57369.1| Drosophila yakuba (individual e) ...</span>
<span class="go">gi|9227|emb|X57370.1| Drosophila yakuba (individual f) ...</span>
<span class="go">gi|9229|emb|X57371.1| Drosophila yakuba (individual g) ...</span>
<span class="go">gi|9231|emb|X57372.1| Drosophila yakuba (individual h) ...</span>
<span class="go">gi|9233|emb|X57373.1| Drosophila yakuba (individual i) ...</span>
<span class="go">gi|9235|emb|X57374.1| Drosophila yakuba (individual j) ...</span>
<span class="go">gi|9237|emb|X57375.1| Drosophila yakuba (individual k) ...</span>
<span class="go">gi|9239|emb|X57376.1| Drosophila yakuba (individual l) ...</span>
<span class="go">gi|156879|gb|M17837.1|DROADHCK D.melanogaster (strain Ja-F) ...</span>
<span class="go">gi|156863|gb|M19547.1|DROADHCC D.melanogaster (strain Af-S) ...</span>
<span class="go">gi|156877|gb|M17836.1|DROADHCJ D.melanogaster (strain Af-F) ...</span>
<span class="go">gi|156875|gb|M17835.1|DROADHCI D.melanogaster (strain Wa-F) ...</span>
<span class="go">gi|156873|gb|M17834.1|DROADHCH D.melanogaster (strain Fr-F) ...</span>
<span class="go">gi|156871|gb|M17833.1|DROADHCG D.melanogaster (strain Fl-F) ...</span>
<span class="go">gi|156869|gb|M17832.1|DROADHCF D.melanogaster (strain Ja-S) ...</span>
<span class="go">gi|156867|gb|M17831.1|DROADHCE D.melanogaster (strain Fl-2S) ...</span>
<span class="go">gi|156865|gb|M17830.1|DROADHCD D.melanogaster (strain Fr-S) ...</span>
<span class="go">gi|156861|gb|M17828.1|DROADHCB D.melanogaster (strain Fl-1S) ...</span>
<span class="go">gi|156859|gb|M17827.1|DROADHCA D.melanogaster (strain Wa-S) ...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">protein_alignment</span> <span class="o">=</span> <span class="n">Align</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;adh.aln&quot;</span><span class="p">,</span> <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">protein_alignment</span><span class="p">)</span>
<span class="go">27</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">protein_alignment</span><span class="p">)</span>  
<span class="go">gi|9217|e         0 MAFTLTNKNVVFVAGLGGIGLDTSKELVKRDLKNLVILDRIENPAAIAELKAINPKVTVT</span>
<span class="go">gi|9219|e         0 MAFTLTNKNVVFVAGLGGIGLDTSKELVKRDLKNLVILDRIENPAAIAELKAINPKVTVT</span>
<span class="go">gi|9221|e         0 MAFTLTNKNVVFVAGLGGIGLDTSKELVKRDLKNLVILDRIENPAAIAELKAINPKVTVT</span>
<span class="go">...</span>
<span class="go">gi|156859         0 MSFTLTNKNVIFVAGLGGIGLDTSKELLKRDLKNLVILDRIENPAAIAELKAINPKVTVT</span>

<span class="go">...</span>

<span class="go">gi|9217|e       240 GTLEAIQWSKHWDSGI 256</span>
<span class="go">gi|9219|e       240 GTLEAIQWSKHWDSGI 256</span>
<span class="go">gi|9221|e       240 GTLEAIQWSKHWDSGI 256</span>
<span class="go">...</span>
<span class="go">gi|156859       240 GTLEAIQWTKHWDSGI 256</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">codon_alignments</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">protein_record</span> <span class="ow">in</span> <span class="n">protein_alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">nucleotide_record</span> <span class="o">=</span> <span class="n">nucleotide_records</span><span class="p">[</span><span class="n">protein_record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
<span class="gp">... </span>    <span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">protein_record</span><span class="p">,</span> <span class="n">nucleotide_record</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="gp">... </span>    <span class="n">codon_alignment</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">codon_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon_alignment</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">codon_alignment</span><span class="p">)</span>  
<span class="go">gi|156859         0 M  S  F  T  L  T  N  K  N  V  I  F  V  A  G  L  G  G  I  G</span>
<span class="go">gi|156859         0 ATGTCGTTTACTTTGACCAACAAGAACGTGATTTTCGTTGCCGGTCTGGGAGGCATTGGT</span>

<span class="go">gi|156859        20 L  D  T  S  K  E  L  L  K  R  D  L  K  N  L  V  I  L  D  R</span>
<span class="go">gi|156859        60 CTGGACACCAGCAAGGAGCTGCTCAAGCGCGATCTGAAGAACCTGGTGATCCTCGACCGC</span>

<span class="go">...</span>

<span class="go">gi|156859       240 G  T  L  E  A  I  Q  W  T  K  H  W  D  S  G  I   256</span>
<span class="go">gi|156859       720 GGCACCCTGGAGGCCATCCAGTGGACCAAGCACTGGGACTCCGGCATC 768</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">nucleotide_records</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>  <span class="c1"># Close indexed FASTA file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">protein_alignment</span><span class="o">.</span><span class="n">mapall</span><span class="p">(</span><span class="n">codon_alignments</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>  
<span class="go">gi|9217|e         0 ATGGCGTTTACCTTGACCAACAAGAACGTGGTTTTCGTGGCCGGTCTGGGAGGCATTGGT</span>
<span class="go">gi|9219|e         0 ATGGCGTTTACCTTGACCAACAAGAACGTGGTTTTCGTGGCCGGTCTGGGAGGCATTGGT</span>
<span class="go">gi|9221|e         0 ATGGCGTTTACCTTGACCAACAAGAACGTGGTTTTCGTGGCCGGTCTGGGAGGCATTGGT</span>
<span class="go">...</span>
<span class="go">gi|156859         0 ATGTCGTTTACTTTGACCAACAAGAACGTGATTTTCGTTGCCGGTCTGGGAGGCATTGGT</span>

<span class="go">...</span>

<span class="go">gi|9217|e       720 GGCACCCTGGAGGCCATCCAGTGGTCCAAGCACTGGGACTCCGGCATC 768</span>
<span class="go">gi|9219|e       720 GGCACCCTGGAGGCCATCCAGTGGTCCAAGCACTGGGACTCCGGCATC 768</span>
<span class="go">gi|9221|e       720 GGTACCCTGGAGGCCATCCAGTGGTCCAAGCACTGGGACTCCGGCATC 768</span>
<span class="go">...</span>
<span class="go">gi|156859       720 GGCACCCTGGAGGCCATCCAGTGGACCAAGCACTGGGACTCCGGCATC 768</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">unique_species</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Drosophila simulans&quot;</span><span class="p">,</span> <span class="s2">&quot;Drosophila yakuba&quot;</span><span class="p">,</span> <span class="s2">&quot;D.melanogaster&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">species</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">alignment</span><span class="o">.</span><span class="n">sequences</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">description</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">description</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">unique_species</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">if</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">description</span><span class="p">:</span>
<span class="gp">... </span>            <span class="k">break</span>
<span class="gp">... </span>    <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to find species for </span><span class="si">{</span><span class="n">description</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">species</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">species</span><span class="p">)</span>
<span class="go">[&#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila yakuba&#39;, &#39;Drosophila simulans&#39;, &#39;Drosophila simulans&#39;, &#39;Drosophila simulans&#39;, &#39;Drosophila simulans&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;, &#39;D.melanogaster&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pvalue</span> <span class="o">=</span> <span class="n">mktest</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>  
<span class="go">0.00206457...</span>
</pre></div>
</div>
<p>In addition to the multiple codon alignment, the function <code class="docutils literal notranslate"><span class="pre">mktest</span></code>
takes as input the species to which each sequence in the alignment
belongs to. The codon table can be provided as an optional argument
<code class="docutils literal notranslate"><span class="pre">codon_table</span></code>.</p>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="chapter_align.html" class="btn btn-neutral float-left" title="Sequence alignments" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="chapter_msa.html" class="btn btn-neutral float-right" title="Multiple Sequence Alignment objects" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 1999-2024, The Biopython Contributors. See the Biopython license terms.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.84.dev0
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="/docs/1.83/">Previous</a></dd>
        
          <dd><a href="/docs/latest/">Latest</a></dd>
        
          <dd><a href="/docs/dev/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>
<script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>