

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Bio.Application package &mdash; Biopython 1.79 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Blast package" href="Bio.Blast.html" />
    <link rel="prev" title="Bio.Alphabet package" href="Bio.Alphabet.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.Application package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#module-Bio.Application">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          



















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
      <li>Bio.Application package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Application/__init__.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Blast.html" class="btn btn-neutral float-right" title="Bio.Blast package" accesskey="n">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
      
      
        <a href="Bio.Alphabet.html" class="btn btn-neutral float-left" title="Bio.Alphabet package" accesskey="p"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="bio-application-package">
<h1>Bio.Application package<a class="headerlink" href="#bio-application-package" title="Permalink to this headline">¶</a></h1>
<section id="module-Bio.Application">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.Application" title="Permalink to this headline">¶</a></h2>
<p>General mechanisms to access applications in Biopython (OBSOLETE).</p>
<p>This module is not intended for direct use. It provides the basic objects which
are subclassed by our command line wrappers, such as:</p>
<blockquote>
<div><ul class="simple">
<li><p>Bio.Align.Applications</p></li>
<li><p>Bio.Blast.Applications</p></li>
<li><p>Bio.Emboss.Applications</p></li>
<li><p>Bio.Sequencing.Applications</p></li>
</ul>
</div></blockquote>
<p>These modules provide wrapper classes for command line tools to help you
construct command line strings by setting the values of each parameter.
The finished command line strings are then normally invoked via the built-in
Python module subprocess.</p>
<p>Due to the on going maintainance burden or keeping command line application
wrappers up to date, we have decided to deprecate and eventually remove them.
We instead now recommend building your command line and invoking it directly
with the subprocess module.</p>
<dl class="py exception">
<dt class="sig sig-object py" id="Bio.Application.ApplicationError">
<em class="property"><span class="pre">exception</span> </em><span class="sig-prename descclassname"><span class="pre">Bio.Application.</span></span><span class="sig-name descname"><span class="pre">ApplicationError</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">returncode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stdout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stderr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.ApplicationError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">subprocess.CalledProcessError</span></code></p>
<p>Raised when an application returns a non-zero exit status (OBSOLETE).</p>
<p>The exit status will be stored in the returncode attribute, similarly
the command line string used in the cmd attribute, and (if captured)
stdout and stderr as strings.</p>
<p>This exception is a subclass of subprocess.CalledProcessError.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">err</span> <span class="o">=</span> <span class="n">ApplicationError</span><span class="p">(</span><span class="o">-</span><span class="mi">11</span><span class="p">,</span> <span class="s2">&quot;helloworld&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;Some error text&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">err</span><span class="o">.</span><span class="n">returncode</span><span class="p">,</span> <span class="n">err</span><span class="o">.</span><span class="n">cmd</span><span class="p">,</span> <span class="n">err</span><span class="o">.</span><span class="n">stdout</span><span class="p">,</span> <span class="n">err</span><span class="o">.</span><span class="n">stderr</span>
<span class="go">(-11, &#39;helloworld&#39;, &#39;&#39;, &#39;Some error text&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">err</span><span class="p">)</span>
<span class="go">Non-zero return code -11 from &#39;helloworld&#39;, message &#39;Some error text&#39;</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.ApplicationError.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">returncode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stdout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stderr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.ApplicationError.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.ApplicationError.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.ApplicationError.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Format the error as a string.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.ApplicationError.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.ApplicationError.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent the error as a string.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="Bio.Application.AbstractCommandline">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">Bio.Application.</span></span><span class="sig-name descname"><span class="pre">AbstractCommandline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmd</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.AbstractCommandline" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Generic interface for constructing command line strings (OBSOLETE).</p>
<p>This class shouldn’t be called directly; it should be subclassed to
provide an implementation for a specific application.</p>
<p>For a usage example we’ll show one of the EMBOSS wrappers.  You can set
options when creating the wrapper object using keyword arguments - or
later using their corresponding properties:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Emboss.Applications</span> <span class="kn">import</span> <span class="n">WaterCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">WaterCommandline</span><span class="p">(</span><span class="n">gapopen</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">gapextend</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">WaterCommandline(cmd=&#39;water&#39;, gapopen=10, gapextend=0.5)</span>
</pre></div>
</div>
<p>You can instead manipulate the parameters via their properties, e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">gapopen</span>
<span class="go">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">gapopen</span> <span class="o">=</span> <span class="mi">20</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">WaterCommandline(cmd=&#39;water&#39;, gapopen=20, gapextend=0.5)</span>
</pre></div>
</div>
<p>You can clear a parameter you have already added by ‘deleting’ the
corresponding property:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">cline</span><span class="o">.</span><span class="n">gapopen</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">gapopen</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">WaterCommandline(cmd=&#39;water&#39;, gapextend=0.5)</span>
</pre></div>
</div>
<p>Once you have set the parameters you need, you can turn the object into
a string (e.g. to log the command):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">You must either set outfile (output filename), or enable filter or stdout (output to stdout).</span>
</pre></div>
</div>
<p>In this case the wrapper knows certain arguments are required to construct
a valid command line for the tool.  For a complete example,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Emboss.Applications</span> <span class="kn">import</span> <span class="n">WaterCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">water_cmd</span> <span class="o">=</span> <span class="n">WaterCommandline</span><span class="p">(</span><span class="n">gapopen</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">gapextend</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">water_cmd</span><span class="o">.</span><span class="n">asequence</span> <span class="o">=</span> <span class="s2">&quot;asis:ACCCGGGCGCGGT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">water_cmd</span><span class="o">.</span><span class="n">bsequence</span> <span class="o">=</span> <span class="s2">&quot;asis:ACCCGAGCGCGGT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">water_cmd</span><span class="o">.</span><span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;temp_water.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">water_cmd</span><span class="p">)</span>
<span class="go">water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">water_cmd</span>
<span class="go">WaterCommandline(cmd=&#39;water&#39;, outfile=&#39;temp_water.txt&#39;, asequence=&#39;asis:ACCCGGGCGCGGT&#39;, bsequence=&#39;asis:ACCCGAGCGCGGT&#39;, gapopen=10, gapextend=0.5)</span>
</pre></div>
</div>
<p>You would typically run the command line via a standard Python operating
system call using the subprocess module for full control. For the simple
case where you just want to run the command and get the output:</p>
<p>stdout, stderr = water_cmd()</p>
<p>Note that by default we assume the underlying tool is installed on the
system $PATH environment variable. This is normal under Linux/Unix, but
may need to be done manually under Windows. Alternatively, you can specify
the full path to the binary as the first argument (cmd):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Emboss.Applications</span> <span class="kn">import</span> <span class="n">WaterCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">water_cmd</span> <span class="o">=</span> <span class="n">WaterCommandline</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;C:\Program Files\EMBOSS\water.exe&quot;</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="n">gapopen</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">gapextend</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="n">asequence</span><span class="o">=</span><span class="s2">&quot;asis:ACCCGGGCGCGGT&quot;</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="n">bsequence</span><span class="o">=</span><span class="s2">&quot;asis:ACCCGAGCGCGGT&quot;</span><span class="p">,</span>
<span class="gp">... </span>                             <span class="n">outfile</span><span class="o">=</span><span class="s2">&quot;temp_water.txt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">water_cmd</span><span class="p">)</span>
<span class="go">&quot;C:\Program Files\EMBOSS\water.exe&quot; -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5</span>
</pre></div>
</div>
<p>Notice that since the path name includes a space it has automatically
been quoted.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="Bio.Application.AbstractCommandline.parameters">
<span class="sig-name descname"><span class="pre">parameters</span></span><em class="property"> <span class="pre">=</span> <span class="pre">None</span></em><a class="headerlink" href="#Bio.Application.AbstractCommandline.parameters" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.AbstractCommandline.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmd</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.AbstractCommandline.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance of a command line wrapper object.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.AbstractCommandline.__str__">
<span class="sig-name descname"><span class="pre">__str__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.AbstractCommandline.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Make the commandline string with the currently set options.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Emboss.Applications</span> <span class="kn">import</span> <span class="n">WaterCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">WaterCommandline</span><span class="p">(</span><span class="n">gapopen</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">gapextend</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">asequence</span> <span class="o">=</span> <span class="s2">&quot;asis:ACCCGGGCGCGGT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">bsequence</span> <span class="o">=</span> <span class="s2">&quot;asis:ACCCGAGCGCGGT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;temp_water.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">&#39;water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.AbstractCommandline.__repr__">
<span class="sig-name descname"><span class="pre">__repr__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.AbstractCommandline.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the command line object for debugging.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Emboss.Applications</span> <span class="kn">import</span> <span class="n">WaterCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">WaterCommandline</span><span class="p">(</span><span class="n">gapopen</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">gapextend</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">asequence</span> <span class="o">=</span> <span class="s2">&quot;asis:ACCCGGGCGCGGT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">bsequence</span> <span class="o">=</span> <span class="s2">&quot;asis:ACCCGAGCGCGGT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;temp_water.txt&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span>
<span class="go">WaterCommandline(cmd=&#39;water&#39;, outfile=&#39;temp_water.txt&#39;, asequence=&#39;asis:ACCCGGGCGCGGT&#39;, bsequence=&#39;asis:ACCCGAGCGCGGT&#39;, gapopen=10, gapextend=0.5)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.AbstractCommandline.set_parameter">
<span class="sig-name descname"><span class="pre">set_parameter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.AbstractCommandline.set_parameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a commandline option for a program (OBSOLETE).</p>
<p>Every parameter is available via a property and as a named
keyword when creating the instance. Using either of these is
preferred to this legacy set_parameter method which is now
OBSOLETE, and likely to be DEPRECATED and later REMOVED in
future releases.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.AbstractCommandline.__setattr__">
<span class="sig-name descname"><span class="pre">__setattr__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.AbstractCommandline.__setattr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Set attribute name to value (PRIVATE).</p>
<p>This code implements a workaround for a user interface issue.
Without this __setattr__ attribute-based assignment of parameters
will silently accept invalid parameters, leading to known instances
of the user assuming that parameters for the application are set,
when they are not.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Emboss.Applications</span> <span class="kn">import</span> <span class="n">WaterCommandline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span> <span class="o">=</span> <span class="n">WaterCommandline</span><span class="p">(</span><span class="n">gapopen</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">gapextend</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">asequence</span> <span class="o">=</span> <span class="s2">&quot;a.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">bsequence</span> <span class="o">=</span> <span class="s2">&quot;b.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cline</span><span class="o">.</span><span class="n">csequence</span> <span class="o">=</span> <span class="s2">&quot;c.fasta&quot;</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Option name csequence was not found.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cline</span><span class="p">)</span>
<span class="go">water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5</span>
</pre></div>
</div>
<p>This workaround uses a whitelist of object attributes, and sets the
object attribute list as normal, for these.  Other attributes are
assumed to be parameters, and passed to the self.set_parameter method
for validation and assignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="Bio.Application.AbstractCommandline.__call__">
<span class="sig-name descname"><span class="pre">__call__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">self</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stdin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stdout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stderr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cwd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">env</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Application.AbstractCommandline.__call__" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute command, wait for it to finish, return (stdout, stderr).</p>
<p>Runs the command line tool and waits for it to finish. If it returns
a non-zero error level, an exception is raised. Otherwise two strings
are returned containing stdout and stderr.</p>
<p>The optional stdin argument should be a string of data which will be
passed to the tool as standard input.</p>
<p>The optional stdout and stderr argument may be filenames (string),
but otherwise are treated as a booleans, and control if the output
should be captured as strings (True, default), or ignored by sending
it to /dev/null to avoid wasting memory (False). If sent to a file
or ignored, then empty string(s) are returned.</p>
<p>The optional cwd argument is a string giving the working directory
to run the command from. See Python’s subprocess module documentation
for more details.</p>
<p>The optional env argument is a dictionary setting the environment
variables to be used in the new process. By default the current
process’ environment variables are used. See Python’s subprocess
module documentation for more details.</p>
<p>Default example usage:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.Emboss.Applications</span> <span class="kn">import</span> <span class="n">WaterCommandline</span>
<span class="n">water_cmd</span> <span class="o">=</span> <span class="n">WaterCommandline</span><span class="p">(</span><span class="n">gapopen</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">gapextend</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
                             <span class="n">stdout</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">auto</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                             <span class="n">asequence</span><span class="o">=</span><span class="s2">&quot;a.fasta&quot;</span><span class="p">,</span> <span class="n">bsequence</span><span class="o">=</span><span class="s2">&quot;b.fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;About to run: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">water_cmd</span><span class="p">)</span>
<span class="n">std_output</span><span class="p">,</span> <span class="n">err_output</span> <span class="o">=</span> <span class="n">water_cmd</span><span class="p">()</span>
</pre></div>
</div>
<p>This functionality is similar to subprocess.check_output(). In general
if you require more control over running the command, use subprocess
directly.</p>
<p>When the program called returns a non-zero error level, a custom
ApplicationError exception is raised. This includes any stdout and
stderr strings captured as attributes of the exception object, since
they may be useful for diagnosing what went wrong.</p>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="Bio.Blast.html" class="btn btn-neutral float-right" title="Bio.Blast package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="Bio.Alphabet.html" class="btn btn-neutral float-left" title="Bio.Alphabet package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 1999-2020, The Biopython Contributors.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  


  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.79
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.78/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>
      <dl>
        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>