

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Nexus.Nexus module &mdash; Biopython 1.75 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.Nexus.Nodes module" href="Bio.Nexus.Nodes.html" />
    <link rel="prev" title="Bio.Nexus package" href="Bio.Nexus.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.Nexus.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.Nexus.html#module-Bio.Nexus">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.Nexus.html">Bio.Nexus package</a> &raquo;</li>
        
      <li>Bio.Nexus.Nexus module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Nexus/Nexus.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.Nexus.Nodes.html" class="btn btn-neutral float-right" title="Bio.Nexus.Nodes module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Nexus.html" class="btn btn-neutral float-left" title="Bio.Nexus package" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.Nexus.Nexus">
<span id="bio-nexus-nexus-module"></span><h1>Bio.Nexus.Nexus module<a class="headerlink" href="#module-Bio.Nexus.Nexus" title="Permalink to this headline">Â¶</a></h1>
<p>Nexus class. Parse the contents of a NEXUS file.</p>
<p>Based upon âNEXUS: An extensible file format for systematic informationâ
Maddison, Swofford, Maddison. 1997. Syst. Biol. 46(4):590-621</p>
<dl class="py exception">
<dt id="Bio.Nexus.Nexus.NexusError">
<em class="property">exception </em><code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">NexusError</code><a class="headerlink" href="#Bio.Nexus.Nexus.NexusError" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Provision for the management of Nexus exceptions.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.Nexus.Nexus.CharBuffer">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">CharBuffer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">string</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Helps reading NEXUS-words and characters from a buffer (semi-PRIVATE).</p>
<p>This class is not intended for public use (any more).</p>
<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">string</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.__init__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.peek">
<code class="sig-name descname">peek</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.peek" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the first character from the buffer.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.peek_nonwhitespace">
<code class="sig-name descname">peek_nonwhitespace</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.peek_nonwhitespace" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the first character from the buffer, do not include spaces.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.__next__">
<code class="sig-name descname">__next__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.__next__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Iterate over NEXUS characters in the file.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.next_nonwhitespace">
<code class="sig-name descname">next_nonwhitespace</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.next_nonwhitespace" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Check for next non whitespace character in NEXUS file.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.skip_whitespace">
<code class="sig-name descname">skip_whitespace</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.skip_whitespace" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Skip whitespace characters in NEXUS file.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.next_until">
<code class="sig-name descname">next_until</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">target</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.next_until" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Iterate over the NEXUS file until a target character is reached.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.peek_word">
<code class="sig-name descname">peek_word</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">word</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.peek_word" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a word stored in the buffer.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.next_word">
<code class="sig-name descname">next_word</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.next_word" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the next NEXUS word from a string.</p>
<p>This deals with single and double quotes, whitespace and punctuation.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.CharBuffer.rest">
<code class="sig-name descname">rest</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.CharBuffer.rest" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the rest of the string without parsing.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Nexus.Nexus.StepMatrix">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">StepMatrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">symbols</span></em>, <em class="sig-param"><span class="n">gap</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.StepMatrix" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Calculate a stepmatrix for weighted parsimony.</p>
<p>See :
COMBINATORIAL WEIGHTS IN PHYLOGENETIC ANALYSIS - A STATISTICAL PARSIMONY PROCEDURE
Wheeler (1990), Cladistics 6:269-275.</p>
<dl class="py method">
<dt id="Bio.Nexus.Nexus.StepMatrix.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">symbols</span></em>, <em class="sig-param"><span class="n">gap</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.StepMatrix.__init__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.StepMatrix.set">
<code class="sig-name descname">set</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">x</span></em>, <em class="sig-param"><span class="n">y</span></em>, <em class="sig-param"><span class="n">value</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.StepMatrix.set" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Set a given value in the matrixâs position.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.StepMatrix.add">
<code class="sig-name descname">add</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">x</span></em>, <em class="sig-param"><span class="n">y</span></em>, <em class="sig-param"><span class="n">value</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.StepMatrix.add" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Add the given value to existing, in matrixâs position.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.StepMatrix.sum">
<code class="sig-name descname">sum</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.StepMatrix.sum" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Calculate the associations, makes matrix of associations.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.StepMatrix.transformation">
<code class="sig-name descname">transformation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.StepMatrix.transformation" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Calculate the transformation matrix.</p>
<p>Normalizes the columns of the matrix of associations.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.StepMatrix.weighting">
<code class="sig-name descname">weighting</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.StepMatrix.weighting" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Calculate the Phylogenetic weight matrix.</p>
<p>Constructed from the logarithmic transformation of the
transformation matrix.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.StepMatrix.smprint">
<code class="sig-name descname">smprint</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">'your_name_here'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.StepMatrix.smprint" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Print a stepmatrix.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.Nexus.Nexus.safename">
<code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">safename</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">mrbayes</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.safename" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a taxon identifier according to NEXUS standard.</p>
<p>Wrap quotes around names with punctuation or whitespace, and double
single quotes.</p>
<p>mrbayes=True: write names without quotes, whitespace or punctuation
for the mrbayes software package.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.Nexus.Nexus.quotestrip">
<code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">quotestrip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">word</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.quotestrip" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Remove quotes and/or double quotes around identifiers.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.Nexus.Nexus.get_start_end">
<code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">get_start_end</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em>, <em class="sig-param"><span class="n">skiplist</span><span class="o">=</span><span class="default_value">'-', '?'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.get_start_end" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return position of first and last character which is not in skiplist.</p>
<p>Skiplist defaults to [â-â,â?â].</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.Nexus.Nexus.combine">
<code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">combine</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">matrices</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.combine" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Combine matrices in [(name,nexus-instance),â¦] and return new nexus instance.</p>
<p>combined_matrix=combine([(name1,nexus_instance1),(name2,nexus_instance2),â¦]
Character sets, character partitions and taxon sets are prefixed, readjusted
and present in the combined matrix.</p>
</dd></dl>

<dl class="py class">
<dt id="Bio.Nexus.Nexus.Commandline">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">Commandline</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">line</span></em>, <em class="sig-param"><span class="n">title</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Commandline" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represent a commandline as command and options.</p>
<dl class="py method">
<dt id="Bio.Nexus.Nexus.Commandline.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">line</span></em>, <em class="sig-param"><span class="n">title</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Commandline.__init__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Nexus.Nexus.Block">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">Block</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">title</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Block" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represent a NEXUS block with block name and list of commandlines.</p>
<dl class="py method">
<dt id="Bio.Nexus.Nexus.Block.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">title</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Block.__init__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Nexus.Nexus.Nexus">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Nexus.Nexus.</code><code class="sig-name descname">Nexus</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">input</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Create the Nexus class, main class for the management of Nexus files.</p>
<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">input</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.__init__" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.get_original_taxon_order">
<code class="sig-name descname">get_original_taxon_order</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.get_original_taxon_order" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Included for backwards compatibility (DEPRECATED).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.set_original_taxon_order">
<code class="sig-name descname">set_original_taxon_order</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">value</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.set_original_taxon_order" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Included for backwards compatibility (DEPRECATED).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.original_taxon_order">
<em class="property">property </em><code class="sig-name descname">original_taxon_order</code><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.original_taxon_order" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Included for backwards compatibility (DEPRECATED).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.read">
<code class="sig-name descname">read</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">input</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.read" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Read and parse NEXUS input (a filename, file-handle, or string).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.write_nexus_data_partitions">
<code class="sig-name descname">write_nexus_data_partitions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">filename</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">blocksize</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">interleave</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">exclude</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">delete</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">charpartition</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">comment</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">mrbayes</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.write_nexus_data_partitions" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Write a nexus file for each partition in charpartition.</p>
<p>Only non-excluded characters and non-deleted taxa are included,
just the data block is written.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.write_nexus_data">
<code class="sig-name descname">write_nexus_data</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">filename</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">exclude</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">delete</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">blocksize</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">interleave</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">interleave_by_partition</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">comment</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">omit_NEXUS</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">append_sets</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">mrbayes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">codons_block</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.write_nexus_data" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Write a nexus file with data and sets block to a file or handle.</p>
<p>Character sets and partitions are appended by default, and are
adjusted according to excluded characters (i.e. character sets
still point to the same sites (not necessarily same positions),
without including the deleted characters.</p>
<ul class="simple">
<li><p>filename - Either a filename as a string (which will be opened,
written to and closed), or a handle object (which will
be written to but NOT closed).</p></li>
<li><p>interleave_by_partition - Optional name of partition (string)</p></li>
<li><p>omit_NEXUS - Boolean.  If true, the â#NEXUSâ line normally at the
start of the file is omitted.</p></li>
</ul>
<p>Returns the filename/handle used to write the data.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.append_sets">
<code class="sig-name descname">append_sets</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">exclude</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">delete</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">mrbayes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">include_codons</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">codons_only</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.append_sets" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a sets block.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.export_fasta">
<code class="sig-name descname">export_fasta</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">filename</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">width</span><span class="o">=</span><span class="default_value">70</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.export_fasta" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Write matrix into a fasta file.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.export_phylip">
<code class="sig-name descname">export_phylip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">filename</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.export_phylip" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Write matrix into a PHYLIP file.</p>
<p>Note that this writes a relaxed PHYLIP format file, where the names
are not truncated, nor checked for invalid characters.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.constant">
<code class="sig-name descname">constant</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">delete</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">exclude</span><span class="o">=</span><span class="default_value">()</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.constant" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a list with all constant characters.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.cstatus">
<code class="sig-name descname">cstatus</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">site</span></em>, <em class="sig-param"><span class="n">delete</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">narrow</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.cstatus" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Summarize character.</p>
<p>narrow=True:  paup-mode (a c ? â&gt; ac; ? ? ? â&gt; ?)
narrow=false:           (a c ? â&gt; a c g t -; ? ? ? â&gt; a c g t -)</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.weighted_stepmatrix">
<code class="sig-name descname">weighted_stepmatrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">'your_name_here'</span></em>, <em class="sig-param"><span class="n">exclude</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">delete</span><span class="o">=</span><span class="default_value">()</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.weighted_stepmatrix" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Calculate a stepmatrix for weighted parsimony.</p>
<p>See Wheeler (1990), Cladistics 6:269-275 and
Felsenstein (1981), Biol. J. Linn. Soc. 16:183-196</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.crop_matrix">
<code class="sig-name descname">crop_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">delete</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">exclude</span><span class="o">=</span><span class="default_value">()</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.crop_matrix" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a matrix without deleted taxa and excluded characters.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.bootstrap">
<code class="sig-name descname">bootstrap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">delete</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">exclude</span><span class="o">=</span><span class="default_value">()</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.bootstrap" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a bootstrapped matrix.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.add_sequence">
<code class="sig-name descname">add_sequence</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">sequence</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.add_sequence" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Add a sequence (string) to the matrix.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.insert_gap">
<code class="sig-name descname">insert_gap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">pos</span></em>, <em class="sig-param"><span class="n">n</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">leftgreedy</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.insert_gap" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Add a gap into the matrix and adjust charsets and partitions.</p>
<p>pos=0: first position
pos=nchar: last position</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.invert">
<code class="sig-name descname">invert</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">charlist</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.invert" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return all character indices that are not in charlist.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.gaponly">
<code class="sig-name descname">gaponly</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">include_missing</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.gaponly" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return gap-only sites.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Nexus.Nexus.Nexus.terminal_gap_to_missing">
<code class="sig-name descname">terminal_gap_to_missing</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">missing</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">skip_n</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Nexus.Nexus.Nexus.terminal_gap_to_missing" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Replace all terminal gaps with missing character.</p>
<p>Mixtures like ???ââ??ââ- are properly resolved.</p>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.Nexus.Nodes.html" class="btn btn-neutral float-right" title="Bio.Nexus.Nodes module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Nexus.html" class="btn btn-neutral float-left" title="Bio.Nexus package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.75
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.74/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>

      <dl>

        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>