

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.SubsMat package &mdash; Biopython 1.75 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SubsMat.FreqTable module" href="Bio.SubsMat.FreqTable.html" />
    <link rel="prev" title="Bio.Statistics.lowess module" href="Bio.Statistics.lowess.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.SubsMat package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-Bio.SubsMat">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
      <li>Bio.SubsMat package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SubsMat/__init__.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SubsMat.FreqTable.html" class="btn btn-neutral float-right" title="Bio.SubsMat.FreqTable module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Statistics.lowess.html" class="btn btn-neutral float-left" title="Bio.Statistics.lowess module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="bio-subsmat-package">
<h1>Bio.SubsMat package<a class="headerlink" href="#bio-subsmat-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.SubsMat.FreqTable.html">Bio.SubsMat.FreqTable module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SubsMat.MatrixInfo.html">Bio.SubsMat.MatrixInfo module</a></li>
</ul>
</div>
</div>
<div class="section" id="module-Bio.SubsMat">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.SubsMat" title="Permalink to this headline">¶</a></h2>
<p>Substitution matrices, log odds matrices, and operations on them.</p>
<div class="section" id="general">
<h3>General:<a class="headerlink" href="#general" title="Permalink to this headline">¶</a></h3>
<p>This module provides a class and a few routines for generating
substitution matrices, similar ot BLOSUM or PAM matrices, but based on
user-provided data.
The class used for these matrices is SeqMat</p>
<p>Matrices are implemented as a dictionary. Each index contains a 2-tuple,
which are the two residue/nucleotide types replaced. The value differs
according to the matrix’s purpose: e.g in a log-odds frequency matrix, the
value would be log(Pij/(Pi*Pj)) where:
Pij: frequency of substitution of letter (residue/nucleotide) i by j
Pi, Pj: expected frequencies of i and j, respectively.</p>
</div>
<div class="section" id="usage">
<h3>Usage:<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h3>
<p>The following section is laid out in the order by which most people wish
to generate a log-odds matrix. Of course, interim matrices can be
generated and investigated. Most people just want a log-odds matrix,
that’s all.</p>
</div>
<div class="section" id="generating-an-accepted-replacement-matrix">
<h3>Generating an Accepted Replacement Matrix:<a class="headerlink" href="#generating-an-accepted-replacement-matrix" title="Permalink to this headline">¶</a></h3>
<p>Initially, you should generate an accepted replacement matrix (ARM)
from your data. The values in ARM are the _counted_ number of
replacements according to your data. The data could be a set of pairs
or multiple alignments. So for instance if Alanine was replaced by
Cysteine 10 times, and Cysteine by Alanine 12 times, the corresponding
ARM entries would be:
[‘A’,’C’]: 10,
[‘C’,’A’] 12
As order doesn’t matter, user can already provide only one entry:
[‘A’,’C’]: 22
A SeqMat instance may be initialized with either a full (first
method of counting: 10, 12) or half (the latter method, 22) matrix. A
Full protein alphabet matrix would be of the size 20x20 = 400. A Half
matrix of that alphabet would be 20x20/2 + 20/2 = 210. That is because
same-letter entries don’t change. (The matrix diagonal). Given an
alphabet size of N:
Full matrix size:N*N
Half matrix size: N(N+1)/2</p>
<p>If you provide a full matrix, the constructor will create a half-matrix
automatically.
If you provide a half-matrix, make sure of a (low, high) sorted order in
the keys: there should only be
a (‘A’,’C’) not a (‘C’,’A’).</p>
<p>Internal functions:</p>
</div>
<div class="section" id="generating-the-observed-frequency-matrix-ofm">
<h3>Generating the observed frequency matrix (OFM):<a class="headerlink" href="#generating-the-observed-frequency-matrix-ofm" title="Permalink to this headline">¶</a></h3>
<p>Use: OFM = _build_obs_freq_mat(ARM)
The OFM is generated from the ARM, only instead of replacement counts, it
contains replacement frequencies.</p>
</div>
<div class="section" id="generating-an-expected-frequency-matrix-efm">
<h3>Generating an expected frequency matrix (EFM):<a class="headerlink" href="#generating-an-expected-frequency-matrix-efm" title="Permalink to this headline">¶</a></h3>
<p>Use: EFM = _build_exp_freq_mat(OFM,exp_freq_table)
exp_freq_table: should be a freqTableC instantiation. See freqTable.py for
detailed information. Briefly, the expected frequency table has the
frequencies of appearance for each member of the alphabet</p>
</div>
<div class="section" id="generating-a-substitution-frequency-matrix-sfm">
<h3>Generating a substitution frequency matrix (SFM):<a class="headerlink" href="#generating-a-substitution-frequency-matrix-sfm" title="Permalink to this headline">¶</a></h3>
<p>Use: SFM = _build_subs_mat(OFM,EFM)
Accepts an OFM, EFM. Provides the division product of the corresponding
values.</p>
</div>
<div class="section" id="generating-a-log-odds-matrix-lom">
<h3>Generating a log-odds matrix (LOM):<a class="headerlink" href="#generating-a-log-odds-matrix-lom" title="Permalink to this headline">¶</a></h3>
<p>Use: LOM=_build_log_odds_mat(SFM[,logbase=10,factor=10.0,roundit=1])
Accepts an SFM. logbase: base of the logarithm used to generate the
log-odds values. factor: factor used to multiply the log-odds values.
roundit: default - true. Whether to round the values.
Each entry is generated by log(LOM[key])*factor
And rounded if required.</p>
</div>
<div class="section" id="external">
<h3>External:<a class="headerlink" href="#external" title="Permalink to this headline">¶</a></h3>
<p>In most cases, users will want to generate a log-odds matrix only, without
explicitly calling the OFM –&gt; EFM –&gt; SFM stages. The function
build_log_odds_matrix does that. User provides an ARM and an expected
frequency table. The function returns the log-odds matrix.</p>
</div>
<div class="section" id="methods-for-subtraction-addition-and-multiplication-of-matrices">
<h3>Methods for subtraction, addition and multiplication of matrices:<a class="headerlink" href="#methods-for-subtraction-addition-and-multiplication-of-matrices" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Generation of an expected frequency table from an observed frequency
matrix.</p></li>
<li><p>Calculation of linear correlation coefficient between two matrices.</p></li>
<li><p>Calculation of relative entropy is now done using the
_make_relative_entropy method and is stored in the member
self.relative_entropy</p></li>
<li><p>Calculation of entropy is now done using the _make_entropy method and
is stored in the member self.entropy.</p></li>
<li><p>Jensen-Shannon distance between the distributions from which the
matrices are derived. This is a distance function based on the
distribution’s entropies.</p></li>
</ul>
<dl class="py class">
<dt id="Bio.SubsMat.SeqMat">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">SeqMat</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mat_name</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">build_later</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></p>
<p>A Generic sequence matrix class.</p>
<p>The key is a 2-tuple containing the letter indices of the matrix. Those
should be sorted in the tuple (low, high). Because each matrix is dealt
with as a half-matrix.</p>
<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mat_name</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">build_later</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize.</p>
<p>User may supply:</p>
<ul class="simple">
<li><p>data: matrix itself</p></li>
<li><p>mat_name: its name. See below.</p></li>
<li><p>alphabet: an instance of Bio.Alphabet, or a subclass. If not
supplied, constructor builds its own from that matrix.</p></li>
<li><p>build_later: skip the matrix size assertion. User will build the
matrix after creating the instance. Constructor builds a half matrix
filled with zeroes.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.make_entropy">
<code class="sig-name descname">make_entropy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.make_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate and set the entropy attribute.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.sum">
<code class="sig-name descname">sum</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Return sum of the results.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.format">
<code class="sig-name descname">format</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">fmt</span><span class="o">=</span><span class="default_value">'%4d'</span></em>, <em class="sig-param"><span class="n">letterfmt</span><span class="o">=</span><span class="default_value">'%4s'</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">non_sym</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">full</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.format" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a string with the bottom-half (default) or a full matrix.</p>
<p>User may pass own alphabet, which should contain all letters in the
alphabet of the matrix, but may be in a different order. This
order will be the order of the letters on the axes.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.print_full_mat">
<code class="sig-name descname">print_full_mat</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">f</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">format</span><span class="o">=</span><span class="default_value">'%4d'</span></em>, <em class="sig-param"><span class="n">topformat</span><span class="o">=</span><span class="default_value">'%4s'</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">factor</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">non_sym</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.print_full_mat" title="Permalink to this definition">¶</a></dt>
<dd><p>Print the full matrix to the file handle f or stdout.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.print_mat">
<code class="sig-name descname">print_mat</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">f</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">format</span><span class="o">=</span><span class="default_value">'%4d'</span></em>, <em class="sig-param"><span class="n">bottomformat</span><span class="o">=</span><span class="default_value">'%4s'</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">factor</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.print_mat" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a nice half-matrix.</p>
<p>f=sys.stdout to see on the screen.</p>
<p>User may pass own alphabet, which should contain all letters in the
alphabet of the matrix, but may be in a different order. This
order will be the order of the letters on the axes.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a nice half-matrix.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.__sub__">
<code class="sig-name descname">__sub__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.__sub__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return integer subtraction product of the two matrices.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.__mul__">
<code class="sig-name descname">__mul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.__mul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Element-wise matrix multiplication.</p>
<p>Returns a new matrix created by multiplying each element by other (if
other is scalar), or by performing element-wise multiplication of the
two matrices (if other is a matrix of the same size).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.__rmul__">
<code class="sig-name descname">__rmul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.__rmul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Element-wise matrix multiplication.</p>
<p>Returns a new matrix created by multiplying each element by other (if
other is scalar), or by performing element-wise multiplication of the
two matrices (if other is a matrix of the same size).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.SubsMat.SeqMat.__add__">
<code class="sig-name descname">__add__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SeqMat.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix addition.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SubsMat.SubstitutionMatrix">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">SubstitutionMatrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mat_name</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">build_later</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SubstitutionMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.SubsMat.SeqMat" title="Bio.SubsMat.SeqMat"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SubsMat.SeqMat</span></code></a></p>
<p>Substitution matrix.</p>
<dl class="py method">
<dt id="Bio.SubsMat.SubstitutionMatrix.calculate_relative_entropy">
<code class="sig-name descname">calculate_relative_entropy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">obs_freq_mat</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.SubstitutionMatrix.calculate_relative_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate and return relative entropy w.r.t. observed frequency matrix.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SubsMat.LogOddsMatrix">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">LogOddsMatrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mat_name</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">build_later</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.LogOddsMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.SubsMat.SeqMat" title="Bio.SubsMat.SeqMat"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SubsMat.SeqMat</span></code></a></p>
<p>Log odds matrix.</p>
<dl class="py method">
<dt id="Bio.SubsMat.LogOddsMatrix.calculate_relative_entropy">
<code class="sig-name descname">calculate_relative_entropy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">obs_freq_mat</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.LogOddsMatrix.calculate_relative_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate and return relative entropy w.r.t. observed frequency matrix.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.SubsMat.make_log_odds_matrix">
<code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">make_log_odds_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">acc_rep_mat</span></em>, <em class="sig-param"><span class="n">exp_freq_table</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">logbase</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">factor</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">round_digit</span><span class="o">=</span><span class="default_value">9</span></em>, <em class="sig-param"><span class="n">keep_nd</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.make_log_odds_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Make log-odds matrix.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.SubsMat.observed_frequency_to_substitution_matrix">
<code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">observed_frequency_to_substitution_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">obs_freq_mat</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.observed_frequency_to_substitution_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert observed frequency table into substitution matrix.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.SubsMat.read_text_matrix">
<code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">read_text_matrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_file</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.read_text_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a matrix from a text file.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.SubsMat.two_mat_relative_entropy">
<code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">two_mat_relative_entropy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mat_1</span></em>, <em class="sig-param"><span class="n">mat_2</span></em>, <em class="sig-param"><span class="n">logbase</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">diag</span><span class="o">=</span><span class="default_value">3</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.two_mat_relative_entropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return relative entropy of two matrices.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.SubsMat.two_mat_correlation">
<code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">two_mat_correlation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mat_1</span></em>, <em class="sig-param"><span class="n">mat_2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.two_mat_correlation" title="Permalink to this definition">¶</a></dt>
<dd><p>Return linear correlation coefficient between two matrices.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.SubsMat.two_mat_DJS">
<code class="sig-prename descclassname">Bio.SubsMat.</code><code class="sig-name descname">two_mat_DJS</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mat_1</span></em>, <em class="sig-param"><span class="n">mat_2</span></em>, <em class="sig-param"><span class="n">pi_1</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">pi_2</span><span class="o">=</span><span class="default_value">0.5</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SubsMat.two_mat_DJS" title="Permalink to this definition">¶</a></dt>
<dd><p>Return Jensen-Shannon Distance between two observed frequence matrices.</p>
</dd></dl>

</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.SubsMat.FreqTable.html" class="btn btn-neutral float-right" title="Bio.SubsMat.FreqTable module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.Statistics.lowess.html" class="btn btn-neutral float-left" title="Bio.Statistics.lowess module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.75
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.74/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>

      <dl>

        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>