

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.codonalign.codonalignment module &mdash; Biopython 1.75 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.codonalign.codonalphabet module" href="Bio.codonalign.codonalphabet.html" />
    <link rel="prev" title="Bio.codonalign.chisq module" href="Bio.codonalign.chisq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.codonalign.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.codonalign.html#module-Bio.codonalign">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.codonalign.html">Bio.codonalign package</a> &raquo;</li>
        
      <li>Bio.codonalign.codonalignment module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/codonalign/codonalignment.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.codonalign.codonalphabet.html" class="btn btn-neutral float-right" title="Bio.codonalign.codonalphabet module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.codonalign.chisq.html" class="btn btn-neutral float-left" title="Bio.codonalign.chisq module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.codonalign.codonalignment">
<span id="bio-codonalign-codonalignment-module"></span><h1>Bio.codonalign.codonalignment module<a class="headerlink" href="#module-Bio.codonalign.codonalignment" title="Permalink to this headline">¶</a></h1>
<p>Code for dealing with Codon Alignment.</p>
<p>CodonAlignment class is inherited from MultipleSeqAlignment class. This is
the core class to deal with codon alignment in biopython.</p>
<dl class="py class">
<dt id="Bio.codonalign.codonalignment.CodonAlignment">
<em class="property">class </em><code class="sig-prename descclassname">Bio.codonalign.codonalignment.</code><code class="sig-name descname">CodonAlignment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">records</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">CodonAlphabet(Standard)</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.Align.html#Bio.Align.MultipleSeqAlignment" title="Bio.Align.MultipleSeqAlignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Align.MultipleSeqAlignment</span></code></a></p>
<p>Codon Alignment class that inherits from MultipleSeqAlignment.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">generic_dna</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span><span class="p">,</span> <span class="n">Gapped</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">CodonSeq</span><span class="p">(</span><span class="s2">&quot;AAAACGTCG&quot;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">default_codon_alphabet</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">CodonSeq</span><span class="p">(</span><span class="s2">&quot;AAA---TCG&quot;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">default_codon_alphabet</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">CodonSeq</span><span class="p">(</span><span class="s2">&quot;AAAAGGTGG&quot;</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">default_codon_alphabet</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">CodonAlignment</span><span class="p">([</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">]))</span>
<span class="go">CodonAlphabet(Standard) CodonAlignment with 3 rows and 9 columns (3 codons)</span>
<span class="go">AAAACGTCG Alpha</span>
<span class="go">AAA---TCG Beta</span>
<span class="go">AAAAGGTGG Gamma</span>
</pre></div>
</div>
<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">records</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">CodonAlphabet(Standard)</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a multi-line string summary of the alignment.</p>
<p>This output is indicated to be readable, but large alignment
is shown truncated. A maximum of 20 rows (sequences) and
60 columns (20 codons) are shown, with the record identifiers.
This should fit nicely on a single screen. e.g.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">index</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a CodonAlignment object for single indexing.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.__add__">
<code class="sig-name descname">__add__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Combine two codonalignments with the same number of rows by adding them.</p>
<p>The method also allows to combine a CodonAlignment object with a
MultipleSeqAlignment object. The following rules apply:</p>
<blockquote>
<div><ul class="simple">
<li><p>CodonAlignment + CodonAlignment -&gt; CodonAlignment</p></li>
<li><p>CodonAlignment + MultipleSeqAlignment -&gt; MultipleSeqAlignment</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.get_aln_length">
<code class="sig-name descname">get_aln_length</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.get_aln_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Get aligment length.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.toMultipleSeqAlignment">
<code class="sig-name descname">toMultipleSeqAlignment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.toMultipleSeqAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert the CodonAlignment to a MultipleSeqAlignment.</p>
<p>Return a MultipleSeqAlignment containing all the
SeqRecord in the CodonAlignment using Seq to store
sequences</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.get_dn_ds_matrix">
<code class="sig-name descname">get_dn_ds_matrix</code><span class="sig-paren">(</span><em class="sig-param">self, method='NG86', codon_table=NCBICodonTable(id=1, names=['Standard', 'SGC0'], ...)</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.get_dn_ds_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Available methods include NG86, LWL85, YN00 and ML.</p>
<dl class="simple">
<dt>Argument:</dt><dd><ul class="simple">
<li><p>method       - Available methods include NG86, LWL85, YN00 and ML.</p></li>
<li><p>codon_table  - Codon table to use for forward translation.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.get_dn_ds_tree">
<code class="sig-name descname">get_dn_ds_tree</code><span class="sig-paren">(</span><em class="sig-param">self, dn_ds_method='NG86', tree_method='UPGMA', codon_table=NCBICodonTable(id=1, names=['Standard', 'SGC0'], ...)</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.get_dn_ds_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Cnstruct dn tree and ds tree.</p>
<dl class="simple">
<dt>Argument:</dt><dd><ul class="simple">
<li><p>dn_ds_method - Available methods include NG86, LWL85, YN00 and ML.</p></li>
<li><p>tree_method  - Available methods include UPGMA and NJ.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.codonalign.codonalignment.CodonAlignment.from_msa">
<em class="property">classmethod </em><code class="sig-name descname">from_msa</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">align</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">CodonAlphabet(Standard)</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.CodonAlignment.from_msa" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a MultipleSeqAlignment to CodonAlignment.</p>
<p>Function to convert a MultipleSeqAlignment to CodonAlignment.
It is the user’s responsibility to ensure all the requirement
needed by CodonAlignment is met.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.codonalign.codonalignment.mktest">
<code class="sig-prename descclassname">Bio.codonalign.codonalignment.</code><code class="sig-name descname">mktest</code><span class="sig-paren">(</span><em class="sig-param">codon_alns, codon_table=NCBICodonTable(id=1, names=['Standard', 'SGC0'], ...), alpha=0.05</em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.codonalign.codonalignment.mktest" title="Permalink to this definition">¶</a></dt>
<dd><p>McDonald-Kreitman test for neutrality.</p>
<p>Implement the McDonald-Kreitman test for neutrality (PMID: 1904993)
This method counts changes rather than sites
(<a class="reference external" href="http://mkt.uab.es/mkt/help_mkt.asp">http://mkt.uab.es/mkt/help_mkt.asp</a>).</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>codon_alns  - list of CodonAlignment to compare (each
CodonAlignment object corresponds to gene sampled from a species)</p></li>
</ul>
</dd>
</dl>
<p>Return the p-value of test result.</p>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.codonalign.codonalphabet.html" class="btn btn-neutral float-right" title="Bio.codonalign.codonalphabet module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.codonalign.chisq.html" class="btn btn-neutral float-left" title="Bio.codonalign.chisq module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.75
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.74/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>

      <dl>

        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>