

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.Seq module &mdash; Biopython 1.78 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SeqFeature module" href="Bio.SeqFeature.html" />
    <link rel="prev" title="Bio.NaiveBayes module" href="Bio.NaiveBayes.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.File.html">Bio.File module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.LogisticRegression.html">Bio.LogisticRegression module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.MarkovModel.html">Bio.MarkovModel module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.MaxEntropy.html">Bio.MaxEntropy module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NaiveBayes.html">Bio.NaiveBayes module</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.Seq module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqFeature.html">Bio.SeqFeature module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqRecord.html">Bio.SeqRecord module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.bgzf.html">Bio.bgzf module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.kNN.html">Bio.kNN module</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.pairwise2.html">Bio.pairwise2 module</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
      <li>Bio.Seq module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/Seq.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SeqFeature.html" class="btn btn-neutral float-right" title="Bio.SeqFeature module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.NaiveBayes.html" class="btn btn-neutral float-left" title="Bio.NaiveBayes module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-Bio.Seq">
<span id="bio-seq-module"></span><h1>Bio.Seq module<a class="headerlink" href="#module-Bio.Seq" title="Permalink to this headline">¶</a></h1>
<p>Provide objects to represent biological sequences.</p>
<dl class="simple">
<dt>See also the <a class="reference external" href="http://biopython.org/wiki/Seq">Seq</a> wiki and the chapter in our tutorial:</dt><dd><ul class="simple">
<li><p><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">HTML Tutorial</a></p></li>
<li><p><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.pdf">PDF Tutorial</a></p></li>
</ul>
</dd>
</dl>
<dl class="py class">
<dt id="Bio.Seq.Seq">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">Seq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Read-only sequence object (essentially a string with biological methods).</p>
<p>Like normal python strings, our basic sequence object is immutable.
This prevents you from doing my_seq[5] = “A” for example, but does allow
Seq objects to be used as dictionary keys.</p>
<p>The Seq object provides a number of string like methods (such as count,
find, split and strip).</p>
<p>The Seq object also provides some biological methods, such as complement,
reverse_complement, transcribe, back_transcribe and translate (which are
not applicable to protein sequences).</p>
<dl class="py method">
<dt id="Bio.Seq.Seq.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a Seq object.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>data - Sequence, required (string)</p></li>
</ul>
</dd>
</dl>
<p>You will typically use Bio.SeqIO to read in sequences from files as
SeqRecord objects, whose sequence will be exposed as a Seq object via
the seq property.</p>
<p>However, will often want to create your own Seq objects directly:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Seq(&#39;MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="go">MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return (truncated) representation of the sequence for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the full sequence as a python string, use str(my_seq).</p>
<p>Note that Biopython 1.44 and earlier would give a truncated
version of repr(my_seq) for str(my_seq).  If you are writing code
which need to be backwards compatible with really old Biopython,
you should continue to use my_seq.tostring() as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">try</span><span class="p">:</span>
    <span class="c1"># The old way, removed in Biopython 1.73</span>
    <span class="n">as_string</span> <span class="o">=</span> <span class="n">seq_obj</span><span class="o">.</span><span class="n">tostring</span><span class="p">()</span>
<span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
    <span class="c1"># The new way, needs Biopython 1.45 or later.</span>
    <span class="c1"># Don&#39;t use this on Biopython 1.44 or older as truncates</span>
    <span class="n">as_string</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq_obj</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__hash__">
<code class="sig-name descname">__hash__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__hash__" title="Permalink to this definition">¶</a></dt>
<dd><p>Hash of the sequence as a string for comparison.</p>
<p>See Seq object comparison documentation (method <code class="docutils literal notranslate"><span class="pre">__eq__</span></code> in
particular) as this has changed in Biopython 1.65. Older versions
would hash on object identity.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Compare the sequence to another sequence or a string (README).</p>
<p>Historically comparing Seq objects has done Python object comparison.
After considerable discussion (keeping in mind constraints of the
Python language, hashes and dictionary support), Biopython now uses
simple string comparison (with a warning about the change).</p>
<p>If you still need to support releases prior to Biopython 1.65, please
just do explicit comparisons:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">id</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">id</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>The new behaviour is to use string-like equality:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">==</span> <span class="n">seq2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">==</span> <span class="s2">&quot;ACGT&quot;</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__lt__">
<code class="sig-name descname">__lt__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__lt__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the less-than operand.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__le__">
<code class="sig-name descname">__le__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__le__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the less-than or equal operand.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__gt__">
<code class="sig-name descname">__gt__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__gt__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the greater-than operand.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__ge__">
<code class="sig-name descname">__ge__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__ge__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the greater-than or equal operand.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__len__">
<code class="sig-name descname">__len__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__len__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the sequence, use len(my_seq).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a subsequence of single letter, use my_seq[index].</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
<span class="go">&#39;A&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__add__">
<code class="sig-name descname">__add__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Add another sequence or string to this sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MELKI&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;LV&quot;</span>
<span class="go">Seq(&#39;MELKILV&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__radd__">
<code class="sig-name descname">__radd__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__radd__" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a sequence on the left.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;LV&quot;</span> <span class="o">+</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MELKI&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;LVMELKI&#39;)</span>
</pre></div>
</div>
<p>Adding two Seq (like) objects is handled via the __add__ method.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__mul__">
<code class="sig-name descname">__mul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__mul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply Seq by integer.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s1">&#39;ATG&#39;</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span>
<span class="go">Seq(&#39;ATGATG&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__rmul__">
<code class="sig-name descname">__rmul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__rmul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply integer by Seq.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">2</span> <span class="o">*</span> <span class="n">Seq</span><span class="p">(</span><span class="s1">&#39;ATG&#39;</span><span class="p">)</span>
<span class="go">Seq(&#39;ATGATG&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__imul__">
<code class="sig-name descname">__imul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__imul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply Seq in-place.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s1">&#39;ATG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">*=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">Seq(&#39;ATGATG&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.tomutable">
<code class="sig-name descname">tomutable</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.tomutable" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the full sequence as a MutableSeq object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MKQHKAMIVALIVICITAVVAAL&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Seq(&#39;MKQHKAMIVALIVICITAVVAAL&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">tomutable</span><span class="p">()</span>
<span class="go">MutableSeq(&#39;MKQHKAMIVALIVICITAVVAAL&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.count">
<code class="sig-name descname">count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.count" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a non-overlapping count, like that of a python string.</p>
<p>This behaves like the python string method of the same name,
which does a non-overlapping count!</p>
<p>For an overlapping search use the newer count_overlap() method.</p>
<p>Returns an integer, the number of occurrences of substring
argument sub in the (sub)sequence given by [start:end].
Optional arguments start and end are interpreted as in slice
notation.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>sub - a string or another Seq object to look for</p></li>
<li><p>start - optional integer, slice start</p></li>
<li><p>end - optional integer, slice end</p></li>
</ul>
</dd>
</dl>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAATGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">))</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">)))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="go">1</span>
</pre></div>
</div>
<p>HOWEVER, please note because python strings and Seq objects (and
MutableSeq objects) do a non-overlapping search, this may not give
the answer you expect:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;AAAA&quot;</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">))</span>
<span class="go">2</span>
</pre></div>
</div>
<p>An overlapping search, as implemented in .count_overlap(),
would give the answer as three!</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.count_overlap">
<code class="sig-name descname">count_overlap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.count_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an overlapping count.</p>
<p>For a non-overlapping search use the count() method.</p>
<p>Returns an integer, the number of occurrences of substring
argument sub in the (sub)sequence given by [start:end].
Optional arguments start and end are interpreted as in slice
notation.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>sub - a string or another Seq object to look for</p></li>
<li><p>start - optional integer, slice start</p></li>
<li><p>end - optional integer, slice end</p></li>
</ul>
</dd>
</dl>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">))</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATATATATA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;ATA&quot;</span><span class="p">))</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATATATATA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;ATA&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="go">1</span>
</pre></div>
</div>
<p>Where substrings do not overlap, should behave the same as
the count() method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAAATGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">))</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">)))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">))</span> <span class="o">==</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>HOWEVER, do not use this method for such cases because the
count() method is much for efficient.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.__contains__">
<code class="sig-name descname">__contains__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">char</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.__contains__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the ‘in’ keyword, like a python string.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATATGAAATTTGAAAA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;AAA&quot;</span> <span class="ow">in</span> <span class="n">my_dna</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA&quot;</span><span class="p">)</span> <span class="ow">in</span> <span class="n">my_dna</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.find">
<code class="sig-name descname">find</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.find" title="Permalink to this definition">¶</a></dt>
<dd><p>Find method, like that of a python string.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Returns an integer, the index of the first occurrence of substring
argument sub in the (sub)sequence given by [start:end].</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>sub - a string or another Seq object to look for</p></li>
<li><p>start - optional integer, slice start</p></li>
<li><p>end - optional integer, slice end</p></li>
</ul>
</dd>
</dl>
<p>Returns -1 if the subsequence is NOT found.</p>
<p>e.g. Locating the first typical start codon, AUG, in an RNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;AUG&quot;</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.rfind">
<code class="sig-name descname">rfind</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.rfind" title="Permalink to this definition">¶</a></dt>
<dd><p>Find from right method, like that of a python string.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Returns an integer, the index of the last (right most) occurrence of
substring argument sub in the (sub)sequence given by [start:end].</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>sub - a string or another Seq object to look for</p></li>
<li><p>start - optional integer, slice start</p></li>
<li><p>end - optional integer, slice end</p></li>
</ul>
</dd>
</dl>
<p>Returns -1 if the subsequence is NOT found.</p>
<p>e.g. Locating the last typical start codon, AUG, in an RNA sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">rfind</span><span class="p">(</span><span class="s2">&quot;AUG&quot;</span><span class="p">)</span>
<span class="go">15</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.index">
<code class="sig-name descname">index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Like find() but raise ValueError when the substring is not found.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">)</span>
<span class="go">-1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
           <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">substring not found...</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.rindex">
<code class="sig-name descname">rindex</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.rindex" title="Permalink to this definition">¶</a></dt>
<dd><p>Like rfind() but raise ValueError when the substring is not found.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.startswith">
<code class="sig-name descname">startswith</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">prefix</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.startswith" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if the Seq starts with the given prefix, False otherwise.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Return True if the sequence starts with the specified prefix
(a string or another Seq object), False otherwise.
With optional start, test sequence beginning at that position.
With optional end, stop comparing sequence at that position.
prefix can also be a tuple of strings to try.  e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GUC&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;AUG&quot;</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;AUG&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">startswith</span><span class="p">((</span><span class="s2">&quot;UCC&quot;</span><span class="p">,</span> <span class="s2">&quot;UCA&quot;</span><span class="p">,</span> <span class="s2">&quot;UCG&quot;</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.endswith">
<code class="sig-name descname">endswith</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">suffix</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.endswith" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if the Seq ends with the given suffix, False otherwise.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Return True if the sequence ends with the specified suffix
(a string or another Seq object), False otherwise.
With optional start, test sequence beginning at that position.
With optional end, stop comparing sequence at that position.
suffix can also be a tuple of strings to try.  e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;UUG&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;AUG&quot;</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;AUG&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">18</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span><span class="o">.</span><span class="n">endswith</span><span class="p">((</span><span class="s2">&quot;UCC&quot;</span><span class="p">,</span> <span class="s2">&quot;UCA&quot;</span><span class="p">,</span> <span class="s2">&quot;UUG&quot;</span><span class="p">))</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.split">
<code class="sig-name descname">split</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">maxsplit</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.split" title="Permalink to this definition">¶</a></dt>
<dd><p>Split method, like that of a python string.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Return a list of the ‘words’ in the string (as Seq objects),
using sep as the delimiter string.  If maxsplit is given, at
most maxsplit splits are done.  If maxsplit is omitted, all
splits are made.</p>
<p>Following the python string method, sep will by default be any
white space (tabs, spaces, newlines) but this is unlikely to
apply to biological sequences.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_aa</span> <span class="o">=</span> <span class="n">my_rna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_aa</span>
<span class="go">Seq(&#39;VMAIVMGR*KGAR*L&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">pep</span> <span class="ow">in</span> <span class="n">my_aa</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">pep</span>
<span class="go">Seq(&#39;VMAIVMGR&#39;)</span>
<span class="go">Seq(&#39;KGAR&#39;)</span>
<span class="go">Seq(&#39;L&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">pep</span> <span class="ow">in</span> <span class="n">my_aa</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">pep</span>
<span class="go">Seq(&#39;VMAIVMGR&#39;)</span>
<span class="go">Seq(&#39;KGAR*L&#39;)</span>
</pre></div>
</div>
<p>See also the rsplit method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">pep</span> <span class="ow">in</span> <span class="n">my_aa</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">pep</span>
<span class="go">Seq(&#39;VMAIVMGR*KGAR&#39;)</span>
<span class="go">Seq(&#39;L&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.rsplit">
<code class="sig-name descname">rsplit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sep</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">maxsplit</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.rsplit" title="Permalink to this definition">¶</a></dt>
<dd><p>Do a right split method, like that of a python string.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Return a list of the ‘words’ in the string (as Seq objects),
using sep as the delimiter string.  If maxsplit is given, at
most maxsplit splits are done COUNTING FROM THE RIGHT.
If maxsplit is omitted, all splits are made.</p>
<p>Following the python string method, sep will by default be any
white space (tabs, spaces, newlines) but this is unlikely to
apply to biological sequences.</p>
<p>e.g. print(my_seq.rsplit(“*”,1))</p>
<p>See also the split method.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.strip">
<code class="sig-name descname">strip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">chars</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.strip" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new Seq object with leading and trailing ends stripped.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Optional argument chars defines which characters to remove.  If
omitted or None (default) then as for the python string method,
this defaults to removing any white space.</p>
<p>e.g. print(my_seq.strip(“-“))</p>
<p>See also the lstrip and rstrip methods.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.lstrip">
<code class="sig-name descname">lstrip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">chars</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.lstrip" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new Seq object with leading (left) end stripped.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Optional argument chars defines which characters to remove.  If
omitted or None (default) then as for the python string method,
this defaults to removing any white space.</p>
<p>e.g. print(my_seq.lstrip(“-“))</p>
<p>See also the strip and rstrip methods.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.rstrip">
<code class="sig-name descname">rstrip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">chars</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.rstrip" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new Seq object with trailing (right) end stripped.</p>
<p>This behaves like the python string method of the same name.</p>
<p>Optional argument chars defines which characters to remove.  If
omitted or None (default) then as for the python string method,
this defaults to removing any white space.</p>
<p>e.g. Removing a nucleotide sequence’s polyadenylation (poly-A tail):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGGTACGCTTATGTCACGTAGAAAAAA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Seq(&#39;CGGTACGCTTATGTCACGTAGAAAAAA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;CGGTACGCTTATGTCACGTAG&#39;)</span>
</pre></div>
</div>
<p>See also the strip and lstrip methods.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.upper">
<code class="sig-name descname">upper</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.upper" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an upper case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;VHLTPeeK*&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Seq(&#39;VHLTPeeK*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="go">Seq(&#39;vhltpeek*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="go">Seq(&#39;VHLTPEEK*&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.lower">
<code class="sig-name descname">lower</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.lower" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a lower case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGGTACGCTTATGTCACGTAGAAAAAA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Seq(&#39;CGGTACGCTTATGTCACGTAGAAAAAA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="go">Seq(&#39;cggtacgcttatgtcacgtagaaaaaa&#39;)</span>
</pre></div>
</div>
<p>See also the upper method.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.encode">
<code class="sig-name descname">encode</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">encoding</span><span class="o">=</span><span class="default_value">'utf-8'</span></em>, <em class="sig-param"><span class="n">errors</span><span class="o">=</span><span class="default_value">'strict'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an encoded version of the sequence as a bytes object.</p>
<p>The Seq object aims to match the interface of a Python string.</p>
<p>This is essentially to save you doing str(my_seq).encode() when
you need a bytes string, for example for computing a hash:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s2">&quot;ascii&quot;</span><span class="p">)</span>
<span class="go">b&#39;ACGT&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.complement">
<code class="sig-name descname">complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the complement sequence by creating a new Seq object.</p>
<p>This method is intended for use with DNA sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CCCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span>
<span class="go">Seq(&#39;CCCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;GGGGGCTATC&#39;)</span>
</pre></div>
</div>
<p>You can of course used mixed case sequences,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CCCCCgatA-GD&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span>
<span class="go">Seq(&#39;CCCCCgatA-GD&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;GGGGGctaT-CH&#39;)</span>
</pre></div>
</div>
<p>Note in the above example, ambiguous character D denotes
G, A or T so its complement is H (for C, T or A).</p>
<p>Note that if the sequence contains neither T nor U, we
assume it is DNA and map any A character to T:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;GCT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;GCTA&#39;)</span>
</pre></div>
</div>
<p>If you actually have RNA, this currently works but we
may deprecate this later. We recommend using the new
complement_rna method instead:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAU&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;GCUA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAU&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;GCUA&#39;)</span>
</pre></div>
</div>
<p>If the sequence contains both T and U, an exception is
raised:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Mixed RNA/DNA found</span>
</pre></div>
</div>
<p>Trying to complement a protein sequence gives a meaningless
sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGR&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">Seq(&#39;KTIBKCY&#39;)</span>
</pre></div>
</div>
<p>Here “M” was interpreted as the IUPAC ambiguity code for
“A” or “C”, with complement “K” for “T” or “G”. Likewise
“A” has complement “T”. The letter “I” has no defined
meaning under the IUPAC convention, and is unchanged.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the reverse complement sequence by creating a new Seq object.</p>
<p>This method is intended for use with DNA sequences:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CCCCCGATAGNR&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span>
<span class="go">Seq(&#39;CCCCCGATAGNR&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;YNCTATCGGGGG&#39;)</span>
</pre></div>
</div>
<p>Note in the above example, since R = G or A, its complement
is Y (which denotes C or T).</p>
<p>You can of course used mixed case sequences,</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CCCCCgatA-G&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span>
<span class="go">Seq(&#39;CCCCCgatA-G&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;C-TatcGGGGG&#39;)</span>
</pre></div>
</div>
<p>As discussed for the complement method, if the sequence
contains neither T nor U, is is assumed to be DNA and
will map any letter A to T.</p>
<p>If you are dealing with RNA you should use the new
reverse_complement_rna method instead</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>  <span class="c1"># defaults to DNA</span>
<span class="go">Seq(&#39;TCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;UCG&#39;)</span>
</pre></div>
</div>
<p>If the sequence contains both T and U, an exception is raised:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Mixed RNA/DNA found</span>
</pre></div>
</div>
<p>Trying to reverse complement a protein sequence will give
a meaningless sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGR&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="go">Seq(&#39;YCKBITK&#39;)</span>
</pre></div>
</div>
<p>Here “M” was interpretted as the IUPAC ambiguity code for
“A” or “C”, with complement “K” for “T” or “G” - and so on.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.complement_rna">
<code class="sig-name descname">complement_rna</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.complement_rna" title="Permalink to this definition">¶</a></dt>
<dd><p>Complement of an RNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>  <span class="c1"># defaults to DNA</span>
<span class="go">Seq(&#39;GCT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;GCU&#39;)</span>
</pre></div>
</div>
<p>If the sequence contains both T and U, an exception is raised:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;CGAUT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Mixed RNA/DNA found</span>
</pre></div>
</div>
<p>If the sequence contains T, an exception is raised:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">complement_rna</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">DNA found, RNA expected</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.reverse_complement_rna">
<code class="sig-name descname">reverse_complement_rna</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.reverse_complement_rna" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse complement of an RNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACG&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">reverse_complement_rna</span><span class="p">()</span>
<span class="go">Seq(&#39;CGU&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.transcribe">
<code class="sig-name descname">transcribe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.transcribe" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the RNA sequence from a DNA sequence by creating a new Seq object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span>
<span class="go">Seq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
</pre></div>
</div>
<p>Trying to transcribe an RNA sequence should have no effect.
If you have a nucleotide sequence which might be DNA or RNA
(or even a mixture), calling the transcribe method will ensure
any T becomes U.</p>
<p>Trying to transcribe a protein sequence will replace any
T for Threonine with U for Selenocysteine, which has no
biologically plausible rational. Older versions of Biopython
would throw an exception.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGRT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;MAIVMGRU&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.back_transcribe">
<code class="sig-name descname">back_transcribe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.back_transcribe" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the DNA sequence from an RNA sequence by creating a new Seq object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">messenger_rna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">messenger_rna</span>
<span class="go">Seq(&#39;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">messenger_rna</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&#39;)</span>
</pre></div>
</div>
<p>Trying to back-transcribe DNA has no effect, If you have a nucleotide
sequence which might be DNA or RNA (or even a mixture), calling the
back-transcribe method will ensure any T becomes U.</p>
<p>Trying to back-transcribe a protein sequence will replace any U for
Selenocysteine with T for Threonine, which is biologically meaningless.
Older versions of Biopython would raise an exception here:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MAIVMGRU&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">()</span>
<span class="go">Seq(&#39;MAIVMGRT&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.translate">
<code class="sig-name descname">translate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">'Standard'</span></em>, <em class="sig-param"><span class="n">stop_symbol</span><span class="o">=</span><span class="default_value">'*'</span></em>, <em class="sig-param"><span class="n">to_stop</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">cds</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">gap</span><span class="o">=</span><span class="default_value">'-'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn a nucleotide sequence into a protein sequence by creating a new Seq object.</p>
<p>This method will translate DNA or RNA sequences. It should not
be used on protein sequences as any result will be biologically
meaningless.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>table - Which codon table to use?  This can be either a name
(string), an NCBI identifier (integer), or a CodonTable
object (useful for non-standard genetic codes).  This
defaults to the “Standard” table.</p></li>
<li><p>stop_symbol - Single character string, what to use for
terminators.  This defaults to the asterisk, “*”.</p></li>
<li><p>to_stop - Boolean, defaults to False meaning do a full
translation continuing on past any stop codons (translated as the
specified stop_symbol).  If True, translation is terminated at
the first in frame stop codon (and the stop_symbol is not
appended to the returned protein sequence).</p></li>
<li><p>cds - Boolean, indicates this is a complete CDS.  If True,
this checks the sequence starts with a valid alternative start
codon (which will be translated as methionine, M), that the
sequence length is a multiple of three, and that there is a
single in frame stop codon at the end (this will be excluded
from the protein sequence, regardless of the to_stop option).
If these tests fail, an exception is raised.</p></li>
<li><p>gap - Single character string to denote symbol used for gaps.
Defaults to the minus sign.</p></li>
</ul>
</dd>
</dl>
<p>e.g. Using the standard table:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;VAIVMGR*KGAR*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">stop_symbol</span><span class="o">=</span><span class="s2">&quot;@&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;VAIVMGR@KGAR@&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Seq(&#39;VAIVMGR&#39;)</span>
</pre></div>
</div>
<p>Now using NCBI table 2, where TGA is not a stop codon:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">Seq(&#39;VAIVMGRWKGAR*&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Seq(&#39;VAIVMGRWKGAR&#39;)</span>
</pre></div>
</div>
<p>In fact, GTG is an alternative start codon under NCBI table 2, meaning
this sequence could be a complete CDS:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Seq(&#39;MAIVMGRWKGAR&#39;)</span>
</pre></div>
</div>
<p>It isn’t a valid CDS under NCBI table 1, due to both the start codon
and also the in frame stop codons:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">Bio.Data.CodonTable.TranslationError</span>: <span class="n">First codon &#39;GTG&#39; is not a start codon</span>
</pre></div>
</div>
<p>If the sequence has no in-frame stop codon, then the to_stop argument
has no effect:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna2</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TTGGCCATTGTAATGGGCCGC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna2</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="go">Seq(&#39;LAIVMGR&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna2</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Seq(&#39;LAIVMGR&#39;)</span>
</pre></div>
</div>
<p>NOTE - Ambiguous codons like “TAN” or “NNN” could be an amino acid
or a stop codon.  These are translated as “X”.  Any invalid codon
(e.g. “TA?” or “T-A”) will throw a TranslationError.</p>
<p>NOTE - This does NOT behave like the python string’s translate
method.  For that use str(my_seq).translate(…) instead</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.ungap">
<code class="sig-name descname">ungap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">gap</span><span class="o">=</span><span class="default_value">'-'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.ungap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy of the sequence without the gap character(s).</p>
<p>The gap character now defaults to the minus sign, and can only
be specified via the method argument. This is no longer possible
via the sequence’s alphabet (as was possible up to Biopython 1.77):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;-ATA--TGAAAT-TTGAAAA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span>
<span class="go">Seq(&#39;-ATA--TGAAAT-TTGAAAA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span><span class="o">.</span><span class="n">ungap</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;ATATGAAATTTGAAAA&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.Seq.join">
<code class="sig-name descname">join</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.Seq.join" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a merge of the sequences in other, spaced by the sequence from self.</p>
<p>Accepts either a Seq or string (and iterates over the letters), or an
iterable containing Seq or string objects. These arguments will be
concatenated with the calling sequence as the spacer:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">concatenated</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s1">&#39;NNNNN&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA&quot;</span><span class="p">),</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TTT&quot;</span><span class="p">),</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;PPP&quot;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">concatenated</span>
<span class="go">Seq(&#39;AAANNNNNTTTNNNNNPPP&#39;)</span>
</pre></div>
</div>
<p>Joining the letters of a single sequence:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s1">&#39;NNNNN&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">))</span>
<span class="go">Seq(&#39;ANNNNNCNNNNNGNNNNNT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s1">&#39;NNNNN&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;ANNNNNCNNNNNGNNNNNT&#39;)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Seq.UnknownSeq">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">UnknownSeq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">length</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">character</span><span class="o">=</span><span class="default_value">'?'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Bio.Seq.Seq" title="Bio.Seq.Seq"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.Seq.Seq</span></code></a></p>
<p>Read-only sequence object of known length but unknown contents.</p>
<p>If you have an unknown sequence, you can represent this with a normal
Seq object, for example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="o">*</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">Seq(&#39;NNNNN&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="go">NNNNN</span>
</pre></div>
</div>
<p>However, this is rather wasteful of memory (especially for large
sequences), which is where this class is most useful:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">unk_five</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unk_five</span>
<span class="go">UnknownSeq(5, character=&#39;?&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">unk_five</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">unk_five</span><span class="p">)</span>
<span class="go">?????</span>
</pre></div>
</div>
<p>You can add unknown sequence together. Provided the characters are the
same, you get another memory saving UnknownSeq:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">unk_four</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unk_four</span>
<span class="go">UnknownSeq(4, character=&#39;?&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unk_four</span> <span class="o">+</span> <span class="n">unk_five</span>
<span class="go">UnknownSeq(9, character=&#39;?&#39;)</span>
</pre></div>
</div>
<p>If the characters are different, addition gives an ordinary Seq object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">unk_nnnn</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unk_nnnn</span>
<span class="go">UnknownSeq(4, character=&#39;N&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unk_nnnn</span> <span class="o">+</span> <span class="n">unk_four</span>
<span class="go">Seq(&#39;NNNN????&#39;)</span>
</pre></div>
</div>
<p>Combining with a real Seq gives a new Seq object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">known_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unk_four</span> <span class="o">+</span> <span class="n">known_seq</span>
<span class="go">Seq(&#39;????ACGT&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">known_seq</span> <span class="o">+</span> <span class="n">unk_four</span>
<span class="go">Seq(&#39;ACGT????&#39;)</span>
</pre></div>
</div>
<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">length</span></em>, <em class="sig-param"><span class="n">alphabet</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">character</span><span class="o">=</span><span class="default_value">'?'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new UnknownSeq object.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>length - Integer, required.</p></li>
<li><p>alphabet - no longer used, must be None.</p></li>
<li><p>character - single letter string, default “?”. Typically “N”
for nucleotides, “X” for proteins, and “?” otherwise.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__len__">
<code class="sig-name descname">__len__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__len__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the stated length of the unknown sequence.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the unknown sequence as full string of the given length.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return (truncated) representation of the sequence for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__add__">
<code class="sig-name descname">__add__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Add another sequence or string to this sequence.</p>
<p>Adding two UnknownSeq objects returns another UnknownSeq object
provided the character is the same.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s1">&#39;X&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s1">&#39;X&#39;</span><span class="p">)</span>
<span class="go">UnknownSeq(15, character=&#39;X&#39;)</span>
</pre></div>
</div>
<p>If the characters differ, an UnknownSeq object cannot be used, so a
Seq object is returned:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s1">&#39;X&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;XXXXXXXXXXxxxxx&#39;)</span>
</pre></div>
</div>
<p>If adding a string to an UnknownSeq, a new Seq is returned:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s1">&#39;X&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;LV&quot;</span>
<span class="go">Seq(&#39;XXXXXLV&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__radd__">
<code class="sig-name descname">__radd__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__radd__" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a sequence on the left.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__mul__">
<code class="sig-name descname">__mul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__mul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply UnknownSeq by integer.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span>
<span class="go">UnknownSeq(6, character=&#39;?&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span>
<span class="go">UnknownSeq(6, character=&#39;N&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__rmul__">
<code class="sig-name descname">__rmul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__rmul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply integer by UnknownSeq.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">2</span> <span class="o">*</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">UnknownSeq(6, character=&#39;?&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">2</span> <span class="o">*</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="go">UnknownSeq(6, character=&#39;N&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__imul__">
<code class="sig-name descname">__imul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__imul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply UnknownSeq in-place.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">*=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">UnknownSeq(6, character=&#39;N&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a subsequence from the UnknownSeq object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">unk</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">unk</span><span class="p">[:])</span>
<span class="go">NNNNNNNN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">unk</span><span class="p">[</span><span class="mi">5</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">unk</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="go">NNNNNN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">unk</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">])</span>
<span class="go">NNN</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.count">
<code class="sig-name descname">count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.count" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a non-overlapping count, like that of a python string.</p>
<p>This behaves like the python string (and Seq object) method of the
same name, which does a non-overlapping count!</p>
<p>For an overlapping search use the newer count_overlap() method.</p>
<p>Returns an integer, the number of occurrences of substring
argument sub in the (sub)sequence given by [start:end].
Optional arguments start and end are interpreted as in slice
notation.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>sub - a string or another Seq object to look for</p></li>
<li><p>start - optional integer, slice start</p></li>
<li><p>end - optional integer, slice end</p></li>
</ul>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;NNNN&quot;</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;NNNN&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="go">0</span>
</pre></div>
</div>
<p>HOWEVER, please note because that python strings and Seq objects (and
MutableSeq objects) do a non-overlapping search, this may not give
the answer you expect:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;NN&quot;</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;NNN&quot;</span><span class="p">)</span>
<span class="go">1</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.count_overlap">
<code class="sig-name descname">count_overlap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.count_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an overlapping count.</p>
<p>For a non-overlapping search use the count() method.</p>
<p>Returns an integer, the number of occurrences of substring
argument sub in the (sub)sequence given by [start:end].
Optional arguments start and end are interpreted as in slice
notation.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>sub - a string or another Seq object to look for</p></li>
<li><p>start - optional integer, slice start</p></li>
<li><p>end - optional integer, slice end</p></li>
</ul>
</dd>
</dl>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;NN&quot;</span><span class="p">)</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;NNN&quot;</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<p>Where substrings do not overlap, should behave the same as
the count() method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.complement">
<code class="sig-name descname">complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the complement of an unknown nucleotide equals itself.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_nuc</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_nuc</span>
<span class="go">UnknownSeq(8, character=&#39;?&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_nuc</span><span class="p">)</span>
<span class="go">????????</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_nuc</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="go">UnknownSeq(8, character=&#39;?&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_nuc</span><span class="o">.</span><span class="n">complement</span><span class="p">())</span>
<span class="go">????????</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.complement_rna">
<code class="sig-name descname">complement_rna</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.complement_rna" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the complement assuming it is RNA.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the reverse complement of an unknown sequence.</p>
<p>The reverse complement of an unknown nucleotide equals itself:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">example</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">example</span><span class="p">)</span>
<span class="go">NNNNNN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">example</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">())</span>
<span class="go">NNNNNN</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.reverse_complement_rna">
<code class="sig-name descname">reverse_complement_rna</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.reverse_complement_rna" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the reverse complement assuming it is RNA.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.transcribe">
<code class="sig-name descname">transcribe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.transcribe" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an unknown RNA sequence from an unknown DNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span>
<span class="go">UnknownSeq(10, character=&#39;N&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_dna</span><span class="p">)</span>
<span class="go">NNNNNNNNNN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">my_dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span>
<span class="go">UnknownSeq(10, character=&#39;N&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_rna</span><span class="p">)</span>
<span class="go">NNNNNNNNNN</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.back_transcribe">
<code class="sig-name descname">back_transcribe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.back_transcribe" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an unknown DNA sequence from an unknown RNA sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_rna</span>
<span class="go">UnknownSeq(20, character=&#39;N&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_rna</span><span class="p">)</span>
<span class="go">NNNNNNNNNNNNNNNNNNNN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">my_rna</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span>
<span class="go">UnknownSeq(20, character=&#39;N&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_dna</span><span class="p">)</span>
<span class="go">NNNNNNNNNNNNNNNNNNNN</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.upper">
<code class="sig-name descname">upper</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.upper" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an upper case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;n&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">UnknownSeq(20, character=&#39;n&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="go">nnnnnnnnnnnnnnnnnnnn</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="go">UnknownSeq(20, character=&#39;N&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>
<span class="go">NNNNNNNNNNNNNNNNNNNN</span>
</pre></div>
</div>
<p>See also the lower method.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.lower">
<code class="sig-name descname">lower</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.lower" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a lower case copy of the sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;X&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">UnknownSeq(20, character=&#39;X&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="go">XXXXXXXXXXXXXXXXXXXX</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
<span class="go">UnknownSeq(20, character=&#39;x&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span>
<span class="go">xxxxxxxxxxxxxxxxxxxx</span>
</pre></div>
</div>
<p>See also the upper method.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.translate">
<code class="sig-name descname">translate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">'Standard'</span></em>, <em class="sig-param"><span class="n">stop_symbol</span><span class="o">=</span><span class="default_value">'*'</span></em>, <em class="sig-param"><span class="n">to_stop</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">cds</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">gap</span><span class="o">=</span><span class="default_value">'-'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate an unknown nucleotide sequence into an unknown protein.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;N&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="go">NNNNNNNNN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span>
<span class="go">UnknownSeq(3, character=&#39;X&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_protein</span><span class="p">)</span>
<span class="go">XXX</span>
</pre></div>
</div>
<p>In comparison, using a normal Seq object:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;NNNNNNNNN&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="go">NNNNNNNNN</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_protein</span>
<span class="go">Seq(&#39;XXX&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_protein</span><span class="p">)</span>
<span class="go">XXX</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.ungap">
<code class="sig-name descname">ungap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">gap</span><span class="o">=</span><span class="default_value">'-'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.ungap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy of the sequence without the gap character(s).</p>
<p>The gap character now defaults to the minus sign, and can only
be specified via the method argument. This is no longer possible
via the sequence’s alphabet (as was possible up to Biopython 1.77):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">UnknownSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s1">&#39;N&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span>
<span class="go">UnknownSeq(20, character=&#39;N&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span><span class="o">.</span><span class="n">ungap</span><span class="p">()</span>  <span class="c1"># using default</span>
<span class="go">UnknownSeq(20, character=&#39;N&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_dna</span><span class="o">.</span><span class="n">ungap</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="go">UnknownSeq(20, character=&#39;N&#39;)</span>
</pre></div>
</div>
<p>If the UnknownSeq is using the gap character, then an empty Seq is
returned:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_gap</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="n">character</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_gap</span>
<span class="go">UnknownSeq(20, character=&#39;-&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_gap</span><span class="o">.</span><span class="n">ungap</span><span class="p">()</span>  <span class="c1"># using default</span>
<span class="go">Seq(&#39;&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_gap</span><span class="o">.</span><span class="n">ungap</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.UnknownSeq.join">
<code class="sig-name descname">join</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.UnknownSeq.join" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a merge of the sequences in other, spaced by the sequence from self.</p>
<p>Accepts either a Seq or string (and iterates over the letters), or an
iterable containing Seq or string objects. These arguments will be
concatenated with the calling sequence as the spacer:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">concatenated</span> <span class="o">=</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA&quot;</span><span class="p">),</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TTT&quot;</span><span class="p">),</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;PPP&quot;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">concatenated</span>
<span class="go">Seq(&#39;AAA?????TTT?????PPP&#39;)</span>
</pre></div>
</div>
<p>If all the inputs are also UnknownSeq using the same character, then it
returns a new UnknownSeq:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span> <span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">)])</span>
<span class="go">UnknownSeq(19, character=&#39;?&#39;)</span>
</pre></div>
</div>
<p>Examples taking a single sequence and joining the letters:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="go">Seq(&#39;A???C???G???T&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">UnknownSeq</span><span class="p">(</span><span class="mi">4</span><span class="p">))</span>
<span class="go">UnknownSeq(13, character=&#39;?&#39;)</span>
</pre></div>
</div>
<p>Will only return an UnknownSeq object if all of the objects to be joined are
also UnknownSeqs with the same character as the spacer, similar to how the
addition of an UnknownSeq and another UnknownSeq would work.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.Seq.MutableSeq">
<em class="property">class </em><code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">MutableSeq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An editable sequence object.</p>
<p>Unlike normal python strings and our basic sequence object (the Seq class)
which are immutable, the MutableSeq lets you edit the sequence in place.
However, this means you cannot use a MutableSeq object as a dictionary key.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ACTCGTCGTCG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ACTCGTCGTCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
<span class="go">&#39;T&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ACTCGACGTCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
<span class="go">&#39;A&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="p">[</span><span class="mi">5</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;NNN&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ACTCGNNNTCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="go">11</span>
</pre></div>
</div>
<p>Note that the MutableSeq object does not support as many string-like
or biological methods as the Seq object.</p>
<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return (truncated) representation of the sequence for debugging.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the full sequence as a python string.</p>
<p>Note that Biopython 1.44 and earlier would give a truncated
version of repr(my_seq) for str(my_seq).  If you are writing code
which needs to be backwards compatible with old Biopython, you
should continue to use my_seq.tostring() rather than str(my_seq).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Compare the sequence to another sequence or a string.</p>
<p>Historically comparing DNA to RNA, or Nucleotide to Protein would
raise an exception. This was later downgraded to a warning, but since
Biopython 1.78 the alphabet is ignored for comparisons.</p>
<p>If you need to support older Biopython versions, please just do
explicit comparisons:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq2</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">id</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">id</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">str</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Biopython now does:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">==</span> <span class="n">seq2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">==</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq1</span> <span class="o">==</span> <span class="s2">&quot;ACGT&quot;</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__lt__">
<code class="sig-name descname">__lt__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__lt__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the less-than operand.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__le__">
<code class="sig-name descname">__le__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__le__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the less-than or equal operand.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__gt__">
<code class="sig-name descname">__gt__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__gt__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the greater-than operand.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__ge__">
<code class="sig-name descname">__ge__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__ge__" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement the greater-than or equal operand.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__len__">
<code class="sig-name descname">__len__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__len__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the sequence, use len(my_seq).</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a subsequence of single letter, use my_seq[index].</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
<span class="go">&#39;A&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__setitem__">
<code class="sig-name descname">__setitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">index</span></em>, <em class="sig-param"><span class="n">value</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__setitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Set a subsequence of single letter via value parameter.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;T&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;TCTCGACGTCG&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__delitem__">
<code class="sig-name descname">__delitem__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">index</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__delitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Delete a subsequence of single letter.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">my_seq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;CTCGACGTCG&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__add__">
<code class="sig-name descname">__add__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__add__" title="Permalink to this definition">¶</a></dt>
<dd><p>Add another sequence or string to this sequence.</p>
<p>Returns a new MutableSeq object.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__radd__">
<code class="sig-name descname">__radd__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__radd__" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a sequence on the left.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;LV&quot;</span> <span class="o">+</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;MELKI&quot;</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;LVMELKI&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__mul__">
<code class="sig-name descname">__mul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__mul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply MutableSeq by integer.</p>
<p>Note this is not in-place and returns a new object,
matching native Python list multiplication.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ATG&#39;</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span>
<span class="go">MutableSeq(&#39;ATGATG&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__rmul__">
<code class="sig-name descname">__rmul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__rmul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply integer by MutableSeq.</p>
<p>Note this is not in-place and returns a new object,
matching native Python list multiplication.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="mi">2</span> <span class="o">*</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ATG&#39;</span><span class="p">)</span>
<span class="go">MutableSeq(&#39;ATGATG&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.__imul__">
<code class="sig-name descname">__imul__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.__imul__" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiply MutableSeq in-place.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ATG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">*=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span>
<span class="go">MutableSeq(&#39;ATGATG&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.append">
<code class="sig-name descname">append</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">c</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a subsequence to the mutable sequence object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ACTCGACGTCGA&#39;)</span>
</pre></div>
</div>
<p>No return value.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.insert">
<code class="sig-name descname">insert</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">i</span></em>, <em class="sig-param"><span class="n">c</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.insert" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a subsequence to the mutable sequence object at a given index.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;AACTCGACGTCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="s1">&#39;G&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;AACTCGACGGTCG&#39;)</span>
</pre></div>
</div>
<p>No return value.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.pop">
<code class="sig-name descname">pop</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">i</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.pop" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove a subsequence of a single letter at given index.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="go">&#39;G&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ACTCGACGTC&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
<span class="go">&#39;C&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ACTCGACGT&#39;)</span>
</pre></div>
</div>
<p>Returns the last character of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.remove">
<code class="sig-name descname">remove</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">item</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.remove" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove a subsequence of a single letter from mutable sequence.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ATCGACGTCG&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;TCGACGTCG&#39;)</span>
</pre></div>
</div>
<p>No return value.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.count">
<code class="sig-name descname">count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.count" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a non-overlapping count, like that of a python string.</p>
<p>This behaves like the python string method of the same name,
which does a non-overlapping count!</p>
<p>For an overlapping search use the newer count_overlap() method.</p>
<p>Returns an integer, the number of occurrences of substring
argument sub in the (sub)sequence given by [start:end].
Optional arguments start and end are interpreted as in slice
notation.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>sub - a string or another Seq object to look for</p></li>
<li><p>start - optional integer, slice start</p></li>
<li><p>end - optional integer, slice end</p></li>
</ul>
</dd>
</dl>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;AAAATGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">))</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">)))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="go">1</span>
</pre></div>
</div>
<p>HOWEVER, please note because that python strings, Seq objects and
MutableSeq objects do a non-overlapping search, this may not give
the answer you expect:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="s2">&quot;AAAA&quot;</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;AAAA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">))</span>
<span class="go">2</span>
</pre></div>
</div>
<p>An overlapping search would give the answer as three!</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.count_overlap">
<code class="sig-name descname">count_overlap</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">sub</span></em>, <em class="sig-param"><span class="n">start</span><span class="o">=</span><span class="default_value">0</span></em>, <em class="sig-param"><span class="n">end</span><span class="o">=</span><span class="default_value">9223372036854775807</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.count_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an overlapping count.</p>
<p>For a non-overlapping search use the count() method.</p>
<p>Returns an integer, the number of occurrences of substring
argument sub in the (sub)sequence given by [start:end].
Optional arguments start and end are interpreted as in slice
notation.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>sub - a string or another Seq object to look for</p></li>
<li><p>start - optional integer, slice start</p></li>
<li><p>end - optional integer, slice end</p></li>
</ul>
</dd>
</dl>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;AAAA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">))</span>
<span class="go">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ATATATATA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;ATA&quot;</span><span class="p">))</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ATATATATA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;ATA&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="go">1</span>
</pre></div>
</div>
<p>Where substrings do not overlap, should behave the same as
the count() method:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;AAAATGA&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">))</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">my_mseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">my_mseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;ATG&quot;</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">)))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">))</span> <span class="o">==</span> <span class="n">my_mseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span><span class="o">.</span><span class="n">count_overlap</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="n">my_mseq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AT&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>HOWEVER, do not use this method for such cases because the
count() method is much for efficient.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.index">
<code class="sig-name descname">index</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">item</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Return first occurrence position of a single entry (i.e. letter).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;ACTCGACGTCG&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">)</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;T&quot;</span><span class="p">))</span>
<span class="go">2</span>
</pre></div>
</div>
<p>Note unlike a Biopython Seq object, or Python string, multi-letter
subsequences are not supported.  Instead this acts like an array or
a list of the entries. There is therefore no <code class="docutils literal notranslate"><span class="pre">.rindex()</span></code> method.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.reverse">
<code class="sig-name descname">reverse</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.reverse" title="Permalink to this definition">¶</a></dt>
<dd><p>Modify the mutable sequence to reverse itself.</p>
<p>No return value.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.complement">
<code class="sig-name descname">complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Modify the mutable sequence to take on its complement.</p>
<p>No return value.</p>
<p>If the sequence contains neither T nor U, DNA is assumed
and any A will be mapped to T.</p>
<p>If the sequence contains both T and U, an exception is raised.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.reverse_complement">
<code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Modify the mutable sequence to take on its reverse complement.</p>
<p>No return value.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.extend">
<code class="sig-name descname">extend</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a sequence to the original mutable sequence object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;ACTCGACGTCG&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ACTCGACGTCGA&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="s1">&#39;TTT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_seq</span>
<span class="go">MutableSeq(&#39;ACTCGACGTCGATTT&#39;)</span>
</pre></div>
</div>
<p>No return value.</p>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.toseq">
<code class="sig-name descname">toseq</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.toseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the full sequence as a new immutable Seq object.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">MutableSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s2">&quot;MKQHKAMIVALIVICITAVVAAL&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span>
<span class="go">MutableSeq(&#39;MKQHKAMIVALIVICITAVVAAL&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_mseq</span><span class="o">.</span><span class="n">toseq</span><span class="p">()</span>
<span class="go">Seq(&#39;MKQHKAMIVALIVICITAVVAAL&#39;)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt id="Bio.Seq.MutableSeq.join">
<code class="sig-name descname">join</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.MutableSeq.join" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a merge of the sequences in other, spaced by the sequence from self.</p>
<p>Accepts all Seq objects and Strings as objects to be concatenated with the spacer</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">concatenated</span> <span class="o">=</span> <span class="n">MutableSeq</span><span class="p">(</span><span class="s1">&#39;NNNNN&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;AAA&quot;</span><span class="p">),</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;TTT&quot;</span><span class="p">),</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;PPP&quot;</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">concatenated</span>
<span class="go">Seq(&#39;AAANNNNNTTTNNNNNPPP&#39;)</span>
</pre></div>
</div>
<p>Throws error if other is not an iterable and if objects inside of the iterable
are not Seq or String objects</p>
</dd></dl>

<dl class="py attribute">
<dt id="Bio.Seq.MutableSeq.__hash__">
<code class="sig-name descname">__hash__</code><em class="property"> = None</em><a class="headerlink" href="#Bio.Seq.MutableSeq.__hash__" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt id="Bio.Seq.transcribe">
<code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">transcribe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dna</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.transcribe" title="Permalink to this definition">¶</a></dt>
<dd><p>Transcribe a DNA sequence into RNA.</p>
<p>If given a string, returns a new string object.</p>
<p>Given a Seq or MutableSeq, returns a new Seq object.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">transcribe</span><span class="p">(</span><span class="s2">&quot;ACTGN&quot;</span><span class="p">)</span>
<span class="go">&#39;ACUGN&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt id="Bio.Seq.back_transcribe">
<code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">back_transcribe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rna</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.back_transcribe" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the RNA sequence back-transcribed into DNA.</p>
<p>If given a string, returns a new string object.</p>
<p>Given a Seq or MutableSeq, returns a new Seq object.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">back_transcribe</span><span class="p">(</span><span class="s2">&quot;ACUGN&quot;</span><span class="p">)</span>
<span class="go">&#39;ACTGN&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt id="Bio.Seq.translate">
<code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">translate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em>, <em class="sig-param"><span class="n">table</span><span class="o">=</span><span class="default_value">'Standard'</span></em>, <em class="sig-param"><span class="n">stop_symbol</span><span class="o">=</span><span class="default_value">'*'</span></em>, <em class="sig-param"><span class="n">to_stop</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">cds</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">gap</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate a nucleotide sequence into amino acids.</p>
<p>If given a string, returns a new string object. Given a Seq or
MutableSeq, returns a Seq object.</p>
<dl class="simple">
<dt>Arguments:</dt><dd><ul class="simple">
<li><p>table - Which codon table to use?  This can be either a name
(string), an NCBI identifier (integer), or a CodonTable object
(useful for non-standard genetic codes).  Defaults to the “Standard”
table.</p></li>
<li><p>stop_symbol - Single character string, what to use for any
terminators, defaults to the asterisk, “*”.</p></li>
<li><p>to_stop - Boolean, defaults to False meaning do a full
translation continuing on past any stop codons
(translated as the specified stop_symbol).  If
True, translation is terminated at the first in
frame stop codon (and the stop_symbol is not
appended to the returned protein sequence).</p></li>
<li><p>cds - Boolean, indicates this is a complete CDS.  If True, this
checks the sequence starts with a valid alternative start
codon (which will be translated as methionine, M), that the
sequence length is a multiple of three, and that there is a
single in frame stop codon at the end (this will be excluded
from the protein sequence, regardless of the to_stop option).
If these tests fail, an exception is raised.</p></li>
<li><p>gap - Single character string to denote symbol used for gaps.
Defaults to None.</p></li>
</ul>
</dd>
</dl>
<p>A simple string example using the default (standard) genetic code:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna</span> <span class="o">=</span> <span class="s2">&quot;GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">)</span>
<span class="go">&#39;VAIVMGR*KGAR*&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">,</span> <span class="n">stop_symbol</span><span class="o">=</span><span class="s2">&quot;@&quot;</span><span class="p">)</span>
<span class="go">&#39;VAIVMGR@KGAR@&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">,</span> <span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">&#39;VAIVMGR&#39;</span>
</pre></div>
</div>
<p>Now using NCBI table 2, where TGA is not a stop codon:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;VAIVMGRWKGAR*&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">&#39;VAIVMGRWKGAR&#39;</span>
</pre></div>
</div>
<p>In fact this example uses an alternative start codon valid under NCBI
table 2, GTG, which means this example is a complete valid CDS which
when translated should really start with methionine (not valine):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">&#39;MAIVMGRWKGAR&#39;</span>
</pre></div>
</div>
<p>Note that if the sequence has no in-frame stop codon, then the to_stop
argument has no effect:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna2</span> <span class="o">=</span> <span class="s2">&quot;GTGGCCATTGTAATGGGCCGC&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna2</span><span class="p">)</span>
<span class="go">&#39;VAIVMGR&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna2</span><span class="p">,</span> <span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">&#39;VAIVMGR&#39;</span>
</pre></div>
</div>
<p>NOTE - Ambiguous codons like “TAN” or “NNN” could be an amino acid
or a stop codon.  These are translated as “X”.  Any invalid codon
(e.g. “TA?” or “T-A”) will throw a TranslationError.</p>
<p>It will however translate either DNA or RNA.</p>
<p>NOTE - Since version 1.71 Biopython contains codon tables with ‘ambiguous
stop codons’. These are stop codons with unambiguous sequence but which
have a context dependent coding as STOP or as amino acid. With these tables
‘to_stop’ must be False (otherwise a ValueError is raised). The dual
coding codons will always be translated as amino acid, except for
‘cds=True’, where the last codon will be translated as STOP.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">coding_dna3</span> <span class="o">=</span> <span class="s2">&quot;ATGGCACGGAAGTGA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna3</span><span class="p">)</span>
<span class="go">&#39;MARK*&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna3</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="mi">27</span><span class="p">)</span>  <span class="c1"># Table 27: TGA -&gt; STOP or W</span>
<span class="go">&#39;MARKW&#39;</span>
</pre></div>
</div>
<p>It will however raise a BiopythonWarning (not shown).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna3</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="mi">27</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">&#39;MARK&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">translate</span><span class="p">(</span><span class="n">coding_dna3</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="mi">27</span><span class="p">,</span> <span class="n">to_stop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">You cannot use &#39;to_stop=True&#39; with this table ...</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt id="Bio.Seq.reverse_complement">
<code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the reverse complement sequence of a nucleotide string.</p>
<p>If given a string, returns a new string object.
Given a Seq or a MutableSeq, returns a new Seq object.</p>
<p>Supports unambiguous and ambiguous nucleotide sequences.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reverse_complement</span><span class="p">(</span><span class="s2">&quot;ACTG-NH&quot;</span><span class="p">)</span>
<span class="go">&#39;DN-CAGT&#39;</span>
</pre></div>
</div>
<p>If neither T nor U is present, DNA is assumed and A is mapped to T:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reverse_complement</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">&#39;T&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt id="Bio.Seq.complement">
<code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">complement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the complement sequence of a DNA string.</p>
<p>If given a string, returns a new string object.</p>
<p>Given a Seq or a MutableSeq, returns a new Seq object.</p>
<p>Supports unambiguous and ambiguous nucleotide sequences.</p>
<p>e.g.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">complement</span><span class="p">(</span><span class="s2">&quot;ACTG-NH&quot;</span><span class="p">)</span>
<span class="go">&#39;TGAC-ND&#39;</span>
</pre></div>
</div>
<p>If neither T nor U is present, DNA is assumed and A is mapped to T:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">complement</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="go">&#39;T&#39;</span>
</pre></div>
</div>
<p>However, this may not be supported in future. Please use the
complement_rna function if you have RNA.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.Seq.complement_rna">
<code class="sig-prename descclassname">Bio.Seq.</code><code class="sig-name descname">complement_rna</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.Seq.complement_rna" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the complement sequence of an RNA string.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">complement</span><span class="p">(</span><span class="s2">&quot;ACG&quot;</span><span class="p">)</span>  <span class="c1"># assumed DNA</span>
<span class="go">&#39;TGC&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">complement_rna</span><span class="p">(</span><span class="s2">&quot;ACG&quot;</span><span class="p">)</span>
<span class="go">&#39;UGC&#39;</span>
</pre></div>
</div>
<p>If the sequence contains a T, and error is raised.</p>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.SeqFeature.html" class="btn btn-neutral float-right" title="Bio.SeqFeature module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.NaiveBayes.html" class="btn btn-neutral float-left" title="Bio.NaiveBayes module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.78
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.77/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>

      <dl>

        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>