

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.SwissProt package &mdash; Biopython 1.78 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SwissProt.KeyWList module" href="Bio.SwissProt.KeyWList.html" />
    <link rel="prev" title="Bio.SubsMat.MatrixInfo module" href="Bio.SubsMat.MatrixInfo.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.78
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Bio.SwissProt package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-Bio.SwissProt">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
      <li>Bio.SwissProt package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SwissProt/__init__.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SwissProt.KeyWList.html" class="btn btn-neutral float-right" title="Bio.SwissProt.KeyWList module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SubsMat.MatrixInfo.html" class="btn btn-neutral float-left" title="Bio.SubsMat.MatrixInfo module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="bio-swissprot-package">
<h1>Bio.SwissProt package<a class="headerlink" href="#bio-swissprot-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.SwissProt.KeyWList.html">Bio.SwissProt.KeyWList module</a></li>
</ul>
</div>
</div>
<div class="section" id="module-Bio.SwissProt">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.SwissProt" title="Permalink to this headline">¶</a></h2>
<p>Code to work with the sprotXX.dat file from SwissProt.</p>
<p><a class="reference external" href="https://web.expasy.org/docs/userman.html">https://web.expasy.org/docs/userman.html</a></p>
<dl class="simple">
<dt>Classes:</dt><dd><ul class="simple">
<li><p>Record             Holds SwissProt data.</p></li>
<li><p>Reference          Holds reference data from a SwissProt record.</p></li>
</ul>
</dd>
<dt>Functions:</dt><dd><ul class="simple">
<li><p>read               Read one SwissProt record</p></li>
<li><p>parse              Read multiple SwissProt records</p></li>
</ul>
</dd>
</dl>
<dl class="py exception">
<dt id="Bio.SwissProt.SwissProtParserError">
<em class="property">exception </em><code class="sig-prename descclassname">Bio.SwissProt.</code><code class="sig-name descname">SwissProtParserError</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="n">line</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SwissProt.SwissProtParserError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code></p>
<p>An error occurred while parsing a SwissProt file.</p>
<dl class="py method">
<dt id="Bio.SwissProt.SwissProtParserError.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em>, <em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="n">line</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SwissProt.SwissProtParserError.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a SwissProtParserError object with the offending line.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SwissProt.Record">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SwissProt.</code><code class="sig-name descname">Record</code><a class="headerlink" href="#Bio.SwissProt.Record" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Holds information from a SwissProt record.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>entry_name        Name of this entry, e.g. RL1_ECOLI.</p></li>
<li><p>data_class        Either ‘STANDARD’ or ‘PRELIMINARY’.</p></li>
<li><p>molecule_type     Type of molecule, ‘PRT’,</p></li>
<li><p>sequence_length   Number of residues.</p></li>
<li><p>accessions        List of the accession numbers, e.g. [‘P00321’]</p></li>
<li><p>created           A tuple of (date, release).</p></li>
<li><p>sequence_update   A tuple of (date, release).</p></li>
<li><p>annotation_update A tuple of (date, release).</p></li>
<li><p>description       Free-format description.</p></li>
<li><p>gene_name         Gene name.  See userman.txt for description.</p></li>
<li><p>organism          The source of the sequence.</p></li>
<li><p>organelle         The origin of the sequence.</p></li>
<li><p>organism_classification  The taxonomy classification.  List of strings.
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/">http://www.ncbi.nlm.nih.gov/Taxonomy/</a>)</p></li>
<li><p>taxonomy_id       A list of NCBI taxonomy id’s.</p></li>
<li><p>host_organism     A list of names of the hosts of a virus, if any.</p></li>
<li><p>host_taxonomy_id  A list of NCBI taxonomy id’s of the hosts, if any.</p></li>
<li><p>references        List of Reference objects.</p></li>
<li><p>comments          List of strings.</p></li>
<li><p>cross_references  List of tuples (db, id1[, id2][, id3]).  See the docs.</p></li>
<li><p>keywords          List of the keywords.</p></li>
<li><p>features          List of tuples (key name, from, to, description).
from and to can be either integers for the residue
numbers, ‘&lt;’, ‘&gt;’, or ‘?’</p></li>
<li><p>protein_existence Numerical value describing the evidence for the existence of the protein.</p></li>
<li><p>seqinfo           tuple of (length, molecular weight, CRC32 value)</p></li>
<li><p>sequence          The sequence.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SwissProt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">example_filename</span> <span class="o">=</span> <span class="s2">&quot;SwissProt/sp008&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">example_filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">records</span> <span class="o">=</span> <span class="n">SwissProt</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">entry_name</span><span class="p">)</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">accessions</span><span class="p">))</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">keywords</span><span class="p">)</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">organism</span><span class="p">))</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">sequence</span><span class="p">[:</span><span class="mi">20</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;...&quot;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">1A02_HUMAN</span>
<span class="go">P01892,P06338,P30514,P30444,P30445,P30446,Q29680,Q29899,Q95352,Q29837,Q95380</span>
<span class="go">[&#39;MHC I&#39;, &#39;Transmembrane&#39;, &#39;Glycoprotein&#39;, &#39;Signal&#39;, &#39;Polymorphism&#39;, &#39;3D-structure&#39;]</span>
<span class="go">&#39;Homo sapiens (Human).&#39;</span>
<span class="go">MAVMAPRTLVLLLSGALALT...</span>
</pre></div>
</div>
<dl class="py method">
<dt id="Bio.SwissProt.Record.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SwissProt.Record.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SwissProt.Reference">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SwissProt.</code><code class="sig-name descname">Reference</code><a class="headerlink" href="#Bio.SwissProt.Reference" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Holds information from one reference in a SwissProt entry.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><ul class="simple">
<li><p>number      Number of reference in an entry.</p></li>
<li><p>evidence    Evidence code.  List of strings.</p></li>
<li><p>positions   Describes extent of work.  List of strings.</p></li>
<li><p>comments    Comments.  List of (token, text).</p></li>
<li><p>references  References.  List of (dbname, identifier).</p></li>
<li><p>authors     The authors of the work.</p></li>
<li><p>title       Title of the work.</p></li>
<li><p>location    A citation for the work.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="Bio.SwissProt.Reference.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SwissProt.Reference.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the class.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="Bio.SwissProt.FeatureTable">
<em class="property">class </em><code class="sig-prename descclassname">Bio.SwissProt.</code><code class="sig-name descname">FeatureTable</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">location</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">type</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">location_operator</span><span class="o">=</span><span class="default_value">''</span></em>, <em class="sig-param"><span class="n">strand</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">id</span><span class="o">=</span><span class="default_value">'&lt;unknown id&gt;'</span></em>, <em class="sig-param"><span class="n">qualifiers</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sub_features</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ref_db</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SwissProt.FeatureTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="Bio.SeqFeature.html#Bio.SeqFeature.SeqFeature" title="Bio.SeqFeature.SeqFeature"><code class="xref py py-class docutils literal notranslate"><span class="pre">Bio.SeqFeature.SeqFeature</span></code></a></p>
<p>Stores feature annotations for specific regions of the sequence.</p>
<p>This is a subclass of SeqFeature, defined in Bio.SeqFeature, where the
attributes are used as follows:</p>
<blockquote>
<div><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">location</span></code>: location of the feature on the canonical or isoform
sequence; the location is stored as an instance of FeatureLocation,
defined in Bio.SeqFeature, with the ref attribute set to the isoform
ID referring to the canonical or isoform sequence on which the feature
is defined</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">id</span></code>: unique and stable identifier (FTId), only provided for features
belonging to the types CARBOHYD, CHAIN, PEPTIDE, PROPEP, VARIANT, or
VAR_SEQ</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">type</span></code>: indicates the type of feature, as defined by the UniProt
Knowledgebase documentation:</p>
<blockquote>
<div><ul class="simple">
<li><p>ACT_SITE: amino acid(s) involved in the activity of an enzyme</p></li>
<li><p>BINDING:  binding site for any chemical group</p></li>
<li><p>CARBOHYD: glycosylation site; an FTId identifier to the GlyConnect
database is provided if annotated there</p></li>
<li><p>CA_BIND:  calcium-binding region</p></li>
<li><p>CHAIN:    polypeptide chain in the mature protein</p></li>
<li><p>COILED:   coiled-coil region</p></li>
<li><p>COMPBIAS: compositionally biased region</p></li>
<li><p>CONFLICT: different sources report differing sequences</p></li>
<li><p>CROSSLNK: posttransationally formed amino acid bond</p></li>
<li><p>DISULFID: disulfide bond</p></li>
<li><p>DNA_BIND: DNA-binding region</p></li>
<li><p>DOMAIN:   domain, defined as a specific combination of secondary
structures organized into a characteristic three-dimensional
structure or fold</p></li>
<li><p>INIT_MET: initiator methionine</p></li>
<li><p>INTRAMEM: region located in a membrane without crossing it</p></li>
<li><p>HELIX:    alpha-, 3(10)-, or pi-helix secondary structure</p></li>
<li><p>LIPID:    covalent binding of a lipid moiety</p></li>
<li><p>METAL:    binding site for a metal ion</p></li>
<li><p>MOD_RES:  posttranslational modification (PTM) of a residue,
annotated by the controlled vocabulary defined by the ptmlist.txt
document on the UniProt website</p></li>
<li><p>MOTIF:    short sequence motif of biological interest</p></li>
<li><p>MUTAGEN:  site experimentally altered by mutagenesis</p></li>
<li><p>NON_CONS: non-consecutive residues</p></li>
<li><p>NON_STD:  non-standard amino acid</p></li>
<li><p>NON_TER:  the residue at an extremity of the sequence is not the
terminal residue</p></li>
<li><p>NP_BIND:  nucleotide phosphate-binding region</p></li>
<li><p>PEPTIDE:  released active mature polypeptide</p></li>
<li><p>PROPEP:   any processed propeptide</p></li>
<li><p>REGION:   region of interest in the sequence</p></li>
<li><p>REPEAT:   internal sequence repetition</p></li>
<li><p>SIGNAL:   signal sequence (prepeptide)</p></li>
<li><p>SITE:     amino-acid site of interest not represented by another
feature key</p></li>
<li><p>STRAND:   beta-strand secondary structure; either a hydrogen-bonded
extended beta strand or a residue in an isolated beta-bridge</p></li>
<li><p>TOPO_DOM: topological domain</p></li>
<li><p>TRANSIT:  transit peptide (mitochondrion, chloroplast, thylakoid,
cyanelle, peroxisome, etc.)</p></li>
<li><p>TRANSMEM: transmembrane region</p></li>
<li><p>TURN:     H-bonded turn (3-, 4-, or 5-turn)</p></li>
<li><p>UNSURE:   uncertainties in the sequence</p></li>
<li><p>VARIANT:  sequence variant; an FTId is provided for protein sequence
variants of Hominidae (great apes and humans)</p></li>
<li><p>VAR_SEQ:  sequence variant produced by alternative splicing,
alternative promoter usage, alternative initiation, or ribosomal
frameshifting</p></li>
<li><p>ZN_FING:  zinc finger region</p></li>
</ul>
</div></blockquote>
</li>
<li><p>qualifiers   A dictionary of additional information, which may include
the feature evidence and free-text notes. While SwissProt includes the
feature identifier code (FTId) as a qualifier, it is stored as the
attribute ID of the FeatureTable object.</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt id="Bio.SwissProt.parse">
<code class="sig-prename descclassname">Bio.SwissProt.</code><code class="sig-name descname">parse</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">source</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SwissProt.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>Read multiple SwissProt records from file.</p>
<p>Argument source is a file-like object or a path to a file.</p>
<p>Returns a generator object which yields Bio.SwissProt.Record() objects.</p>
</dd></dl>

<dl class="py function">
<dt id="Bio.SwissProt.read">
<code class="sig-prename descclassname">Bio.SwissProt.</code><code class="sig-name descname">read</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">source</span></em><span class="sig-paren">)</span><a class="headerlink" href="#Bio.SwissProt.read" title="Permalink to this definition">¶</a></dt>
<dd><p>Read one SwissProt record from file.</p>
<p>Argument source is a file-like object or a path to a file.</p>
<p>Returns a Record() object.</p>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.SwissProt.KeyWList.html" class="btn btn-neutral float-right" title="Bio.SwissProt.KeyWList module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SubsMat.MatrixInfo.html" class="btn btn-neutral float-left" title="Bio.SubsMat.MatrixInfo module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>