

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bio.SearchIO.ExonerateIO package &mdash; Biopython 1.76 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/biopython.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bio.SearchIO.ExonerateIO.exonerate_cigar module" href="Bio.SearchIO.ExonerateIO.exonerate_cigar.html" />
    <link rel="prev" title="Bio.SearchIO.BlastIO.blast_xml module" href="Bio.SearchIO.BlastIO.blast_xml.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #10100F" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/biopython_logo_white.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">API Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="Bio.html">Bio package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="Bio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="Bio.Affy.html">Bio.Affy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Align.html">Bio.Align package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.AlignIO.html">Bio.AlignIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Alphabet.html">Bio.Alphabet package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Application.html">Bio.Application package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Blast.html">Bio.Blast package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.CAPS.html">Bio.CAPS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Cluster.html">Bio.Cluster package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Compass.html">Bio.Compass package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Crystal.html">Bio.Crystal package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Data.html">Bio.Data package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Emboss.html">Bio.Emboss package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Entrez.html">Bio.Entrez package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.ExPASy.html">Bio.ExPASy package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.FSSP.html">Bio.FSSP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.GenBank.html">Bio.GenBank package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Geo.html">Bio.Geo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Graphics.html">Bio.Graphics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.HMM.html">Bio.HMM package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KDTree.html">Bio.KDTree package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.KEGG.html">Bio.KEGG package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Medline.html">Bio.Medline package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.NMR.html">Bio.NMR package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Nexus.html">Bio.Nexus package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PDB.html">Bio.PDB package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Pathway.html">Bio.Pathway package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Phylo.html">Bio.Phylo package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.PopGen.html">Bio.PopGen package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Restriction.html">Bio.Restriction package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SCOP.html">Bio.SCOP package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SVDSuperimposer.html">Bio.SVDSuperimposer package</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="Bio.SearchIO.html">Bio.SearchIO package</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="Bio.SearchIO.html#subpackages">Subpackages</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="Bio.SearchIO.html#module-Bio.SearchIO">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqIO.html">Bio.SeqIO package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SeqUtils.html">Bio.SeqUtils package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Sequencing.html">Bio.Sequencing package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Statistics.html">Bio.Statistics package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SubsMat.html">Bio.SubsMat package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.SwissProt.html">Bio.SwissProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.TogoWS.html">Bio.TogoWS package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniGene.html">Bio.UniGene package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.UniProt.html">Bio.UniProt package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.Wise.html">Bio.Wise package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.codonalign.html">Bio.codonalign package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.motifs.html">Bio.motifs package</a></li>
<li class="toctree-l3"><a class="reference internal" href="Bio.phenotype.html">Bio.phenotype package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="Bio.html#module-Bio">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BioSQL.html">BioSQL package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Biopython</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="Bio.html">Bio package</a> &raquo;</li>
        
          <li><a href="Bio.SearchIO.html">Bio.SearchIO package</a> &raquo;</li>
        
      <li>Bio.SearchIO.ExonerateIO package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/biopython/biopython/blob/master/Bio/SearchIO/ExonerateIO/__init__.py" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="breadcrumb navigation">
      
        <a href="Bio.SearchIO.ExonerateIO.exonerate_cigar.html" class="btn btn-neutral float-right" title="Bio.SearchIO.ExonerateIO.exonerate_cigar module" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SearchIO.BlastIO.blast_xml.html" class="btn btn-neutral float-left" title="Bio.SearchIO.BlastIO.blast_xml module" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
  </div>
  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="bio-searchio-exonerateio-package">
<h1>Bio.SearchIO.ExonerateIO package<a class="headerlink" href="#bio-searchio-exonerateio-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="Bio.SearchIO.ExonerateIO.exonerate_cigar.html">Bio.SearchIO.ExonerateIO.exonerate_cigar module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SearchIO.ExonerateIO.exonerate_text.html">Bio.SearchIO.ExonerateIO.exonerate_text module</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bio.SearchIO.ExonerateIO.exonerate_vulgar.html">Bio.SearchIO.ExonerateIO.exonerate_vulgar module</a></li>
</ul>
</div>
</div>
<div class="section" id="module-Bio.SearchIO.ExonerateIO">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Bio.SearchIO.ExonerateIO" title="Permalink to this headline">¶</a></h2>
<p>Bio.SearchIO support for Exonerate output formats.</p>
<p>This module adds support for handling Exonerate outputs. Exonerate is a generic
tool for pairwise sequence comparison that allows you to align sequences using
several different models.</p>
<p>Bio.SearchIO.ExonerateIO was tested on the following Exonerate versions and
models:</p>
<blockquote>
<div><ul class="simple">
<li><p>version: 2.2</p></li>
<li><p>models:
- affine:local                - cdna2genome
- coding2coding               - est2genome
- genome2genome               - ner
- protein2dna                 - protein2genome
- ungapped                    - ungapped:translated</p></li>
</ul>
</div></blockquote>
<p>Although model testing were not exhaustive, ExonerateIO should be able to cope
with all Exonerate models. Please file a bug report if you stumble upon an
unparseable file.</p>
<p>More information on Exonerate is available on its home page at
www.ebi.ac.uk/~guy/exonerate/</p>
<div class="section" id="supported-formats">
<h3>Supported Formats<a class="headerlink" href="#supported-formats" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul class="simple">
<li><p>Plain text alignment - ‘exonerate-text’   - parsing, indexing</p></li>
<li><p>Vulgar line          - ‘exonerate-vulgar’ - parsing, indexing</p></li>
<li><p>Cigar line           - ‘exonerate-cigar’  - parsing, indexing</p></li>
</ul>
</div></blockquote>
<p>On Exonerate, these output formats are not exclusive to one another. For
example, you may have both plain text and vulgar output in the same file.
ExonerateIO can only handle one of these at a time, however. If you have a file
containing both plain text and vulgar lines, for example, you have to pick
either ‘exonerate-text’ or ‘exonerate-vulgar’ to parse it.</p>
<p>Due to the cigar format specification, many features of the alignments such as
introns or frameshifts may be collapsed into a single feature (in this case,
they are labelled ‘D’ for ‘deletion’). The parser does not attempt to guess
whether the D label it encounters is a real deletion or a collapsed feature.
As such, parsing or indexing using ‘exonerate-cigar’ may yield different results
compared to ‘exonerate-text’ or ‘exonerate-vulgar’.</p>
</div>
<div class="section" id="exonerate-text">
<h3>exonerate-text<a class="headerlink" href="#exonerate-text" title="Permalink to this headline">¶</a></h3>
<p>The plain text output / C4 alignment is the output triggered by the
‘–showalignemnt’ flag. Compared to the two other output formats, this format
contains the most information, having the complete query and hit sequences of
the alignment.</p>
<p>Here are some examples of the C4 output alignment that ExonerateIO can handle
(coordinates not written in scale):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>1. simple ungapped alignments

       1 : ATGGGCAATATCCTTCGGAAAGGTCAGCAAAT :      56
           ||||||||||||||||||||||||||||||||
 1319275 : ATGGGCAATATCCTTCGGAAAGGTCAGCAAAT : 1319220

2. alignments with frameshifts:

     129 : -TGCCGTTACCAT----GACGAAAGTATTAAT : 160
           -CysArgTyrHis----AspGluSerIleAsn
           #||||||||||||####|||||||||||||||
           #CysArgTyrHis####AspGluSerIleAsn
 1234593 : GTGCCGTTACCATCGGTGACGAAAGTATTAAT : 1234630

3. alignments with introns and split codons:

    382 :    {A}                             {CC}AAA                 :    358
          AAA{T}  &gt;&gt;&gt;&gt; Target Intron 3 &gt;&gt;&gt;&gt;  {hr}LysATGAGCGATGAAAATA
          || { }++         55423 bp        ++{  } !  |||  ||||||||||
          AAC{L}gt.........................ag{eu}AspTTGAATGATGAAAATA
  42322 :    {C}                             {TG}GAT                 :  97769

4. alignments with NER blocks

    111 : CAGAAAA--&lt;   31  &gt;--CTGCCCAGAAT--&lt;   10  &gt;--AACGAGCGTTCCG- :    184
          | |||||--&lt; NER 1 &gt;--| ||||| | |--&lt; NER 2 &gt;--|||  | ||||||-
 297911 : CTGAAAA--&lt;   29  &gt;--CCGCCCAAAGT--&lt;   13  &gt;--AACTGGAGTTCCG- : 297993
</pre></div>
</div>
<p>ExonerateIO utilizes the HSPFragment model quite extensively to deal with non-
ungapped alignments. For any single HSPFragment, if ExonerateIO sees an intron,
a NER block, or a frameshift, it will break the fragment into two HSPFragment
objects and adjust each of their start and end coordinate appropriately.</p>
<p>You may notice that Exonerate always uses the three letter amino acid codes to
display protein sequences. If the protein itself is part of the query sequence,
such as in the protein2dna model, ExonerateIO will transform the protein
sequence into using one letter codes. This is because the SeqRecord objects that
store the sequences are designed for single-letter sequences only. If Exonerate
also outputs the underlying nucleotide sequence, it will be saved into an
<code class="docutils literal notranslate"><span class="pre">aln_annotation</span></code> entry as a list of triplets.</p>
<p>If the protein sequence is not part of the actual alignment, such as in the
est2genome or genome2genome models, ExonerateIO will keep the three letter codes
and store them as <code class="docutils literal notranslate"><span class="pre">aln_annotation</span></code> entries. In these cases, the hit and
query sequences may be used directly as SeqRecord objects as they are one-letter
nucleotide codes. The three-letter protein sequences are then stored as entries
in the <code class="docutils literal notranslate"><span class="pre">aln_annotation</span></code> dictionary.</p>
<p>For ‘exonerate-text’, ExonerateIO provides the following object attributes:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 22%" />
<col style="width: 33%" />
<col style="width: 45%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Object</p></th>
<th class="head"><p>Attribute</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td rowspan="4"><p>QueryResult</p></td>
<td><p>description</p></td>
<td><p>query sequence description</p></td>
</tr>
<tr class="row-odd"><td><p>id</p></td>
<td><p>query sequence ID</p></td>
</tr>
<tr class="row-even"><td><p>model</p></td>
<td><p>alignment model</p></td>
</tr>
<tr class="row-odd"><td><p>program</p></td>
<td><p>‘exonerate’</p></td>
</tr>
<tr class="row-even"><td rowspan="2"><p>Hit</p></td>
<td><p>description</p></td>
<td><p>hit sequence description</p></td>
</tr>
<tr class="row-odd"><td><p>id</p></td>
<td><p>hit sequence ID</p></td>
</tr>
<tr class="row-even"><td rowspan="3"><p>HSP</p></td>
<td><p>hit_split_codons</p></td>
<td><p>list of split codon coordinates
in the hit sequence</p></td>
</tr>
<tr class="row-odd"><td><p>score</p></td>
<td><p>alignment score</p></td>
</tr>
<tr class="row-even"><td><p>query_split_codons</p></td>
<td><p>list of split codon coordinates
in the query sequence</p></td>
</tr>
<tr class="row-odd"><td rowspan="11"><p>HSPFragment</p></td>
<td><p>aln_annotation</p></td>
<td><p>alignment similarity string, hit
sequence annotation, and/or
query sequence annotation</p></td>
</tr>
<tr class="row-even"><td><p>hit</p></td>
<td><p>hit sequence</p></td>
</tr>
<tr class="row-odd"><td><p>hit_end</p></td>
<td><p>hit sequence end coordinate</p></td>
</tr>
<tr class="row-even"><td><p>hit_frame</p></td>
<td><p>hit sequence reading frame</p></td>
</tr>
<tr class="row-odd"><td><p>hit_start</p></td>
<td><p>hit sequence start coordinate</p></td>
</tr>
<tr class="row-even"><td><p>hit_strand</p></td>
<td><p>hit sequence strand</p></td>
</tr>
<tr class="row-odd"><td><p>query</p></td>
<td><p>query sequence</p></td>
</tr>
<tr class="row-even"><td><p>query_end</p></td>
<td><p>query sequence end coordinate</p></td>
</tr>
<tr class="row-odd"><td><p>query_frame</p></td>
<td><p>query sequence reading frame</p></td>
</tr>
<tr class="row-even"><td><p>query_start</p></td>
<td><p>query sequence start coordinate</p></td>
</tr>
<tr class="row-odd"><td><p>query_strand</p></td>
<td><p>query sequence strand</p></td>
</tr>
</tbody>
</table>
<p>Note that you can also use the default HSP or HSPFragment properties. For
example, to check the intron coordinates of your result you can use the
<code class="docutils literal notranslate"><span class="pre">query_inter_ranges</span></code> or <code class="docutils literal notranslate"><span class="pre">hit_inter_ranges</span></code> properties:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SearchIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Exonerate/exn_22_m_genome2genome.exn&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">all_qresult</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SearchIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s1">&#39;exonerate-text&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hsp</span> <span class="o">=</span> <span class="n">all_qresult</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>   <span class="c1"># last qresult, last hit, last hsp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hsp</span>
<span class="go">HSP(...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hsp</span><span class="o">.</span><span class="n">query_inter_ranges</span>
<span class="go">[(388, 449), (284, 319), (198, 198), (114, 161)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hsp</span><span class="o">.</span><span class="n">hit_inter_ranges</span>
<span class="go">[(487387, 641682), (386207, 487327), (208677, 386123), (71917, 208639)]</span>
</pre></div>
</div>
<p>Here you can see that for both query and hit introns, the coordinates
in each tuple is always (start, end) where start &lt;= end. But when you compare
each tuple to the next, the coordinates decrease. This is an indication that
both the query and hit sequences lie on the minus strand. Exonerate outputs
minus strand results in a decreasing manner; the start coordinate is always
bigger than the end coordinate. ExonerateIO preserves the fragment ordering as a
whole, but uses its own standard to store an individual fragment’s start and end
coordinates.</p>
<p>You may also notice that the third tuple in <code class="docutils literal notranslate"><span class="pre">query_inter_ranges</span></code> is (198, 198),
two exact same numbers. This means that the query sequence does not have any
gaps at that position. The gap is only present in the hit sequence, where we see
that the third tuple contains (208677, 386123), a gap of about 177k bases.</p>
<p>Another example is to use the <code class="docutils literal notranslate"><span class="pre">hit_frame_all</span></code> and <code class="docutils literal notranslate"><span class="pre">query_frame_all</span></code> to see if
there are any frameshifts in your alignment:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SearchIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Exonerate/exn_22_m_coding2coding_fshifts.exn&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">qresult</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">SearchIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s1">&#39;exonerate-text&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hsp</span> <span class="o">=</span> <span class="n">qresult</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>      <span class="c1"># first hit, first hsp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hsp</span>
<span class="go">HSP(...)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hsp</span><span class="o">.</span><span class="n">query_frame_all</span>
<span class="go">[1, 2, 2, 2]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hsp</span><span class="o">.</span><span class="n">hit_frame_all</span>
<span class="go">[1, 1, 3, 1]</span>
</pre></div>
</div>
<p>Here you can see that the alignment as a whole has three frameshifts. The first
one occurs in the query sequence, after the first fragment (1 -&gt; 2 shift), the
second one occurs in the hit sequence, after the second fragment (1 -&gt; 3 shift),
and the last one also occurs in the hit sequence, before the last fragment (3 -&gt;
1 shift).</p>
<p>There are other default HSP properties that you can use to ease your workflow.
Please refer to the HSP object documentation for more details.</p>
</div>
<div class="section" id="exonerate-vulgar">
<h3>exonerate-vulgar<a class="headerlink" href="#exonerate-vulgar" title="Permalink to this headline">¶</a></h3>
<p>The vulgar format provides a compact way of representing alignments created by
Exonerate. In general, it contains the same information as the plain text output
except for the ‘model’ information and the actual sequences themselves. You can
expect that the coordinates obtained from using ‘exonerate-text’ and
‘exonerate-vulgar’ to be the same. Both formats also creates HSPFragment using
the same triggers: introns, NER blocks, and/or frameshifts.</p>
</div>
<div class="section" id="exonerate-cigar">
<h3>exonerate-cigar<a class="headerlink" href="#exonerate-cigar" title="Permalink to this headline">¶</a></h3>
<p>The cigar format provides an even more compact representation of Exonerate
alignments. However, this comes with a cost of losing information. In the cigar
format, for example, introns are treated as simple deletions. This makes it
impossible for the parser to distinguish between simple deletions or intron
regions. As such, ‘exonerate-cigar’ may produce different sets of coordinates
and fragments compared to ‘exonerate-vulgar’ or ‘exonerate-text’.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Bio.SearchIO.ExonerateIO.exonerate_cigar.html" class="btn btn-neutral float-right" title="Bio.SearchIO.ExonerateIO.exonerate_cigar module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bio.SearchIO.BlastIO.blast_xml.html" class="btn btn-neutral float-left" title="Bio.SearchIO.BlastIO.blast_xml module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 1999-2020, The Biopython Contributors

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      Biopython v: 1.76
      <span class="fa fa-caret-down"></span>
    </span>
    <div class="rst-other-versions">
      <dl>
        <dt>Versions</dt>
        
          <dd><a href="../../1.75/api/">Previous</a></dd>
        
          <dd><a href="../../latest/api/">Latest</a></dd>
        
          <dd><a href="../../dev/api/">Develop</a></dd>
        
      </dl>

      <dl>

        <dt>Biopython Project</dt>
          <dd>
            <a href="https://biopython.org" class="icon icon-home">  Homepage</a>
          </dd>
          <dd>
            <a href="https://github.com/biopython/biopython" class="fa fa-github">  On GitHub</a>
          </dd>
      </dl>
    </div>
  </div>


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>